{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "toc": true
   },
   "source": [
    "<h1>Table of Contents<span class=\"tocSkip\"></span></h1>\n",
    "<div class=\"toc\"><ul class=\"toc-item\"><li><span><a href=\"#Python-Setup\" data-toc-modified-id=\"Python-Setup-1\"><span class=\"toc-item-num\">1&nbsp;&nbsp;</span>Python Setup</a></span><ul class=\"toc-item\"><li><span><a href=\"#Change-the-width-of-the-page\" data-toc-modified-id=\"Change-the-width-of-the-page-1.1\"><span class=\"toc-item-num\">1.1&nbsp;&nbsp;</span>Change the width of the page</a></span></li><li><span><a href=\"#Import-packages\" data-toc-modified-id=\"Import-packages-1.2\"><span class=\"toc-item-num\">1.2&nbsp;&nbsp;</span>Import packages</a></span></li></ul></li><li><span><a href=\"#Colours\" data-toc-modified-id=\"Colours-2\"><span class=\"toc-item-num\">2&nbsp;&nbsp;</span>Colours</a></span><ul class=\"toc-item\"><li><span><a href=\"#Colour-line-graph\" data-toc-modified-id=\"Colour-line-graph-2.1\"><span class=\"toc-item-num\">2.1&nbsp;&nbsp;</span>Colour line graph</a></span></li></ul></li><li><span><a href=\"#Totals-for-studies\" data-toc-modified-id=\"Totals-for-studies-3\"><span class=\"toc-item-num\">3&nbsp;&nbsp;</span>Totals for studies</a></span></li><li><span><a href=\"#Functions-for-calculating-trinucleotide-context-specific-mutation-rates\" data-toc-modified-id=\"Functions-for-calculating-trinucleotide-context-specific-mutation-rates-4\"><span class=\"toc-item-num\">4&nbsp;&nbsp;</span>Functions for calculating trinucleotide-context specific mutation rates</a></span><ul class=\"toc-item\"><li><span><a href=\"#Calculating-mutation-rates-for-individual-variants\" data-toc-modified-id=\"Calculating-mutation-rates-for-individual-variants-4.1\"><span class=\"toc-item-num\">4.1&nbsp;&nbsp;</span>Calculating mutation rates for individual variants</a></span><ul class=\"toc-item\"><li><span><a href=\"#DNMT3A\" data-toc-modified-id=\"DNMT3A-4.1.1\"><span class=\"toc-item-num\">4.1.1&nbsp;&nbsp;</span>DNMT3A</a></span></li><li><span><a href=\"#TET2\" data-toc-modified-id=\"TET2-4.1.2\"><span class=\"toc-item-num\">4.1.2&nbsp;&nbsp;</span>TET2</a></span></li><li><span><a href=\"#ASXL1\" data-toc-modified-id=\"ASXL1-4.1.3\"><span class=\"toc-item-num\">4.1.3&nbsp;&nbsp;</span>ASXL1</a></span></li><li><span><a href=\"#TP53\" data-toc-modified-id=\"TP53-4.1.4\"><span class=\"toc-item-num\">4.1.4&nbsp;&nbsp;</span>TP53</a></span></li></ul></li><li><span><a href=\"#Calculating-mutation-rates-from-a-.csv-file-of-variants\" data-toc-modified-id=\"Calculating-mutation-rates-from-a-.csv-file-of-variants-4.2\"><span class=\"toc-item-num\">4.2&nbsp;&nbsp;</span>Calculating mutation rates from a .csv file of variants</a></span><ul class=\"toc-item\"><li><span><a href=\"#DNMT3A\" data-toc-modified-id=\"DNMT3A-4.2.1\"><span class=\"toc-item-num\">4.2.1&nbsp;&nbsp;</span>DNMT3A</a></span></li><li><span><a href=\"#TET2\" data-toc-modified-id=\"TET2-4.2.2\"><span class=\"toc-item-num\">4.2.2&nbsp;&nbsp;</span>TET2</a></span></li><li><span><a href=\"#ASXL1\" data-toc-modified-id=\"ASXL1-4.2.3\"><span class=\"toc-item-num\">4.2.3&nbsp;&nbsp;</span>ASXL1</a></span></li><li><span><a href=\"#TP53\" data-toc-modified-id=\"TP53-4.2.4\"><span class=\"toc-item-num\">4.2.4&nbsp;&nbsp;</span>TP53</a></span></li></ul></li><li><span><a href=\"#Calculating-mutation-rates-from-a-list-of-variants\" data-toc-modified-id=\"Calculating-mutation-rates-from-a-list-of-variants-4.3\"><span class=\"toc-item-num\">4.3&nbsp;&nbsp;</span>Calculating mutation rates from a list of variants</a></span><ul class=\"toc-item\"><li><span><a href=\"#DNMT3A\" data-toc-modified-id=\"DNMT3A-4.3.1\"><span class=\"toc-item-num\">4.3.1&nbsp;&nbsp;</span>DNMT3A</a></span></li><li><span><a href=\"#TET2\" data-toc-modified-id=\"TET2-4.3.2\"><span class=\"toc-item-num\">4.3.2&nbsp;&nbsp;</span>TET2</a></span></li><li><span><a href=\"#ASXL1\" data-toc-modified-id=\"ASXL1-4.3.3\"><span class=\"toc-item-num\">4.3.3&nbsp;&nbsp;</span>ASXL1</a></span></li><li><span><a href=\"#TP53\" data-toc-modified-id=\"TP53-4.3.4\"><span class=\"toc-item-num\">4.3.4&nbsp;&nbsp;</span>TP53</a></span></li></ul></li></ul></li><li><span><a href=\"#Lists-of-variants-targeted-by-each-study\" data-toc-modified-id=\"Lists-of-variants-targeted-by-each-study-5\"><span class=\"toc-item-num\">5&nbsp;&nbsp;</span>Lists of variants targeted by each study</a></span><ul class=\"toc-item\"><li><span><a href=\"#Jaiswal-2014\" data-toc-modified-id=\"Jaiswal-2014-5.1\"><span class=\"toc-item-num\">5.1&nbsp;&nbsp;</span>Jaiswal 2014</a></span></li><li><span><a href=\"#Genovese-2014\" data-toc-modified-id=\"Genovese-2014-5.2\"><span class=\"toc-item-num\">5.2&nbsp;&nbsp;</span>Genovese 2014</a></span></li><li><span><a href=\"#McKerrel-2015\" data-toc-modified-id=\"McKerrel-2015-5.3\"><span class=\"toc-item-num\">5.3&nbsp;&nbsp;</span>McKerrel 2015</a></span></li><li><span><a href=\"#Zink-2017\" data-toc-modified-id=\"Zink-2017-5.4\"><span class=\"toc-item-num\">5.4&nbsp;&nbsp;</span>Zink 2017</a></span></li><li><span><a href=\"#Coombs-2017\" data-toc-modified-id=\"Coombs-2017-5.5\"><span class=\"toc-item-num\">5.5&nbsp;&nbsp;</span>Coombs 2017</a></span></li><li><span><a href=\"#Young-2016-&amp;-2019\" data-toc-modified-id=\"Young-2016-&amp;-2019-5.6\"><span class=\"toc-item-num\">5.6&nbsp;&nbsp;</span>Young 2016 &amp; 2019</a></span></li><li><span><a href=\"#Desai-2018\" data-toc-modified-id=\"Desai-2018-5.7\"><span class=\"toc-item-num\">5.7&nbsp;&nbsp;</span>Desai 2018</a></span></li><li><span><a href=\"#Acuna-Hidalgo-2017\" data-toc-modified-id=\"Acuna-Hidalgo-2017-5.8\"><span class=\"toc-item-num\">5.8&nbsp;&nbsp;</span>Acuna-Hidalgo 2017</a></span></li></ul></li><li><span><a href=\"#Lists-of-all-possible-variants-in-DNMT3A,-TET2,-ASXL1,-TP53\" data-toc-modified-id=\"Lists-of-all-possible-variants-in-DNMT3A,-TET2,-ASXL1,-TP53-6\"><span class=\"toc-item-num\">6&nbsp;&nbsp;</span>Lists of all possible variants in DNMT3A, TET2, ASXL1, TP53</a></span><ul class=\"toc-item\"><li><span><a href=\"#DNMT3A\" data-toc-modified-id=\"DNMT3A-6.1\"><span class=\"toc-item-num\">6.1&nbsp;&nbsp;</span>DNMT3A</a></span></li><li><span><a href=\"#TET2\" data-toc-modified-id=\"TET2-6.2\"><span class=\"toc-item-num\">6.2&nbsp;&nbsp;</span>TET2</a></span></li><li><span><a href=\"#ASXL1\" data-toc-modified-id=\"ASXL1-6.3\"><span class=\"toc-item-num\">6.3&nbsp;&nbsp;</span>ASXL1</a></span></li><li><span><a href=\"#TP53\" data-toc-modified-id=\"TP53-6.4\"><span class=\"toc-item-num\">6.4&nbsp;&nbsp;</span>TP53</a></span></li></ul></li><li><span><a href=\"#Actual-number-of-observations-of-each-variant\" data-toc-modified-id=\"Actual-number-of-observations-of-each-variant-7\"><span class=\"toc-item-num\">7&nbsp;&nbsp;</span>Actual number of observations of each variant</a></span><ul class=\"toc-item\"><li><ul class=\"toc-item\"><li><span><a href=\"#DNMT3A\" data-toc-modified-id=\"DNMT3A-7.0.1\"><span class=\"toc-item-num\">7.0.1&nbsp;&nbsp;</span>DNMT3A</a></span></li><li><span><a href=\"#TET2\" data-toc-modified-id=\"TET2-7.0.2\"><span class=\"toc-item-num\">7.0.2&nbsp;&nbsp;</span>TET2</a></span></li><li><span><a href=\"#ASXL1\" data-toc-modified-id=\"ASXL1-7.0.3\"><span class=\"toc-item-num\">7.0.3&nbsp;&nbsp;</span>ASXL1</a></span></li><li><span><a href=\"#TP53\" data-toc-modified-id=\"TP53-7.0.4\"><span class=\"toc-item-num\">7.0.4&nbsp;&nbsp;</span>TP53</a></span></li></ul></li></ul></li><li><span><a href=\"#Functions-for-calculating-the-expected-number-of-observations-of-a-variant\" data-toc-modified-id=\"Functions-for-calculating-the-expected-number-of-observations-of-a-variant-8\"><span class=\"toc-item-num\">8&nbsp;&nbsp;</span>Functions for calculating the expected number of observations of a variant</a></span></li><li><span><a href=\"#Maximum-Likelihood-Estimation-for-s\" data-toc-modified-id=\"Maximum-Likelihood-Estimation-for-s-9\"><span class=\"toc-item-num\">9&nbsp;&nbsp;</span>Maximum Likelihood Estimation for s</a></span><ul class=\"toc-item\"><li><span><a href=\"#DNMT3A-variants\" data-toc-modified-id=\"DNMT3A-variants-9.1\"><span class=\"toc-item-num\">9.1&nbsp;&nbsp;</span>DNMT3A variants</a></span></li><li><span><a href=\"#TET2-variants\" data-toc-modified-id=\"TET2-variants-9.2\"><span class=\"toc-item-num\">9.2&nbsp;&nbsp;</span>TET2 variants</a></span></li><li><span><a href=\"#ASXL1-variants\" data-toc-modified-id=\"ASXL1-variants-9.3\"><span class=\"toc-item-num\">9.3&nbsp;&nbsp;</span>ASXL1 variants</a></span></li><li><span><a href=\"#TP53-variants\" data-toc-modified-id=\"TP53-variants-9.4\"><span class=\"toc-item-num\">9.4&nbsp;&nbsp;</span>TP53 variants</a></span></li></ul></li></ul></div>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Python Setup"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Change the width of the page"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 275,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>.container { width:80% !important; }</style>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "from IPython.core.display import display, HTML\n",
    "display(HTML(\"<style>.container { width:80% !important; }</style>\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Import packages"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 276,
   "metadata": {
    "code_folding": []
   },
   "outputs": [],
   "source": [
    "# imported packages\n",
    "import random\n",
    "import numpy as np\n",
    "import matplotlib\n",
    "import matplotlib.pyplot as plt\n",
    "import math\n",
    "import matplotlib.ticker as plticker\n",
    "from matplotlib.ticker import LinearLocator, FormatStrFormatter\n",
    "from matplotlib.patches import Polygon\n",
    "from matplotlib.patches import Patch\n",
    "from matplotlib.lines import Line2D\n",
    "from matplotlib import cm\n",
    "import scipy.special\n",
    "import scipy.integrate as it\n",
    "from scipy import integrate\n",
    "from scipy.interpolate import interp1d\n",
    "from scipy.stats import kde\n",
    "import copy\n",
    "import glob, os\n",
    "import re\n",
    "from sklearn import datasets, linear_model\n",
    "import pandas as pd\n",
    "from decimal import *\n",
    "from operator import itemgetter    \n",
    "from collections import OrderedDict\n",
    "import timeit"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Colours"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 277,
   "metadata": {
    "code_folding": []
   },
   "outputs": [],
   "source": [
    "# Lists of colors for plots\n",
    "c0 = (0.76, 0.76, 0.76)\n",
    "c1 = (1.00, 0.18, 0.33);\n",
    "c2 = (1.00, 0.23, 0.19);\n",
    "c3 = (1.00, 0.58, 0.00);\n",
    "c4 = (1.00, 0.80, 0.00);\n",
    "c5 = (0.30, 0.85, 0.39);\n",
    "c6 = (0.35, 0.78, 0.98);\n",
    "c7 = (0.20, 0.67, 0.86);\n",
    "c8 = (0.00, 0.48, 1.00);\n",
    "c9 = (0.35, 0.34, 0.84);\n",
    "c10 = (0.00, 0.31, 0.57);\n",
    "c11 = (0.12, 0.29, 0.69);\n",
    "c12 = (0.17, 0.17, 0.42);\n",
    "c13 = (1.00, 1.00, 1.00);\n",
    "c14 = (0.77, 0.04, 0.00);"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 278,
   "metadata": {
    "code_folding": [
     0
    ]
   },
   "outputs": [],
   "source": [
    "#define the colors from colorbrewer2\n",
    "orange1 = '#feedde'\n",
    "orange2 = '#fdbe85'\n",
    "orange3 = '#fd8d3c'\n",
    "orange4 = '#e6550d'\n",
    "orange5 = '#a63603'\n",
    "blue1 = '#eff3ff'\n",
    "blue2 = '#bdd7e7'\n",
    "blue3 = '#6baed6'\n",
    "blue4 = '#3182bd'\n",
    "blue5 = '#08519c'\n",
    "green1 = '#edf8e9'\n",
    "green2 = '#bae4b3'\n",
    "green3 = '#74c476'\n",
    "green4 = '#31a354'\n",
    "green5 = '#006d2c'\n",
    "grey1 = '#f7f7f7'\n",
    "grey2 = '#cccccc'\n",
    "grey3 = '#969696'\n",
    "grey4 = '#636363'\n",
    "grey5 = '#252525'\n",
    "purple1 = '#f2f0f7'\n",
    "purple2 = '#cbc9e2'\n",
    "purple3 = '#9e9ac8'\n",
    "purple4 = '#756bb1'\n",
    "purple5 = '#54278f'\n",
    "red1 = '#fee5d9'\n",
    "red2 = '#fcae91'\n",
    "red3 = '#fb6a4a'\n",
    "red4 = '#de2d26'\n",
    "red5 = '#a50f15'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Colour line graph"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 279,
   "metadata": {
    "code_folding": [
     0
    ]
   },
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAlkAAAExCAYAAABPt7ftAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDIuMi4zLCBodHRwOi8vbWF0cGxvdGxpYi5vcmcvIxREBQAAF0xJREFUeJzt3X2QnWd53/Hvj2KcENtgrTAOkW11hpeCiIDhOIEqGEcwQNrBocUW5EWyOmZ2mBiHiSs6oetpAq5CQsDQDs6MRcuwQMybGqfEaUiYJVZCB5geWfbSJVGdpn6R7WLWshkgDl7JV//Y48x22dUenec82j2r72dmZ3Tu+36uc9l//ea+n/M8qSokSZI0XE9Z7QYkSZLWI0OWJElSCwxZkiRJLTBkSZIktcCQJUmS1AJDliRJUgsMWZJGVpJLkxxZ7T4kaSmGLElrQpJfTNJN8r0kDyb5kyQ/s9p9SdKgDFmSVl2Sa4EPA78FPBu4EPg94OdXqZ+nrsb3SlpfDFmSVlWSZwDvBa6uqj+oqu9X1VxV/VFVvSvJmUk+nOSB3t+Hk5y5TK0XJrktyaNJZpJctmDutiRvW/B5d5KvLPhcSa5OchdwV+Z9KMlDSb6TZDrJi1v8XyFpnTFkSVptrwR+BLhlmfkJ4BXAS4GXAD8FXLd4UZIzgD8C/gw4D7gG+P0kLziJXt4E/DTwIuB1wCXA84FnAm8BHj6JWpJOc4YsSattDJitqmPLzP8S8N6qeqiqvg28B9i5xLpXAGcBv11Vj1fVl4FbgV84iV7eV1VHq+oxYA44G/gnQKrqr6rqwZOoJek0Z8iStNoeBjae4D6o5wD3LPh8T29sqXX3VdUTi9b+xEn0ct+T/+iFtI8ANwLfSrIvyTknUUvSac6QJWm1fRX4e+aP6pbyAHDRgs8X9saWWndBkqcsWnt/79/fB56+YO78JWrU//eh6j9W1cuBLcwfG75rmR4l6YcYsiStqqr6DvDvgBuTvCnJ05OckeTnkrwf+DRwXZJnJdnYW/upJUp9nfkg9W96118KvBH4TG/+DuBf9uo/F7jqRH0luTjJT/fu9fo+80HwePP/YkmnC0OWpFVXVTcA1zJ/Q/u3mT+2ewfwh8C/B7rANPAN4Pbe2OIajwOXAT8HzDL/CIhdVfXXvSUfAh4HvgVMAr+/QlvnAB8FHmH+2PFh4AOD/jdKOv2kqlZeJUmSpJPiTpYkSVILDFmSJEktMGRJkiS1wJAlSZLUAkOWJElSC1b9TfMbN26szZs3r3YbkiRJKzp48OBsVT2rn7WrHrI2b95Mt9td7TYkSZJWlOSelVfN87hQkiSpBYYsSZKkFhiyJEmSWmDIkiRJaoEhS5IkqQWGLEmSpBYYsiRJklpgyJIkSWqBIUuSJKkFhixJkqQWGLIkSZJaYMiSJElqwUAhK8klSW5PcizJ5UvMn5Pk/iQfad6iJEnS6Bl0J+teYDdw8zLz1wMHBqwtSZI08p7az6Iku4A9QAHTVbWzN/7EEmtfDjwb+CLQWan2Y3cd5htvuPQkWpYkSVr7VgxZSbYAE8C2qppNsuEEa58CfBDYCbzmBOvGgXGA5/3omSfbsyRJ0prXz07WdmB/Vc0CVNXRE6z9FeC/VdV9SZZdVFX7gH0AnU6nfvKLt/XdsCRJ0qo5Qb5ZrJ+QFeaPCfvxSuBVSX4FOAt4WpLvVdWv992RJEnSOtDPje9TwI4kYwAnOi6sql+qqgurajPz93B9woAlSZJORyuGrKqaAfYCB5LcCdyQ5OIkR4ArgJuSzLTcpyRJ0kjp69eFVTUJTC4a3rTCNR8HPj5QV5IkSSPOJ75LkiS1wJAlSZLUAkOWJElSCwxZkiRJLTBkSZIktcCQJUmS1AJDliRJUgsMWZIkSS0YKGQluSTJ7UmOJbl8wfhFSQ4muSPJTJK3D69VSZKk0dHXE9+XcC+wm/n3Ey70IPBPq+oHSc4C/meSL1TVAw16lCRJGjl9hawku5gPVAVMV9XO3vgTC9dV1eMLPp6Jx5GSJOk0tWLISrIFmAC2VdVskg0rrL8A+GPgucC73MWSJEmno352mrYD+6tqFqCqjp5ocVXdV1VbmQ9ZVyZ59uI1ScaTdJN05+bmBulbkiRpTesnZIX5Y8KT0tvBmgFetcTcvqrqVFXnjDPOONnSkiRJa14/IWsK2JFkDOBEx4VJNiX50d6/zwW2AYeH0agkSdIoWfGerKqaSbIXOJDkOHAoyY3ALcC5wBuTvKeqtgAvBD6YpJjfAftAVX2jxf4lSZLWpL5+XVhVk8DkouFNS6z7ErB1CH1JkiSNNB+xIEmS1AJDliRJUgsMWZIkSS0wZEmSJLXAkCVJktQCQ5YkSVILDFmSJEktMGRJkiS1wJAlSZLUgoFCVpJLktye5FiSyxeMvzTJV5PMJJlO8pbhtSpJkjQ6+nqtzhLuBXYDexaN/x2wq6ruSvIc4GCSP62qRxv0KEmSNHL6CllJdjEfqAqYrqqdvfEnFq6rqv+14N8PJHkIeBawbMi6++5vsXv3hwdoXZIkae1aMWQl2QJMANuqajbJhn4KJ/kp4GnA/15ibhwYBzjnnGefVMOSJEmjIFV14gXJNcD5VTWxxNzHgVurav+i8R8HbgOurKqvnah+p9Opbrd7km1LkiSdekkOVlWnn7X93Pge5o8J+/3yc4A/Bq5bKWBJkiStV/2ErClgR5IxgBMdFyZ5GnAL8Imq+vxwWpQkSRo9K4asqpoB9gIHktwJ3JDk4iRHgCuAm5LM9JbvAC4Bdie5o/f30raalyRJWqv6+nVhVU0Ck4uGNy2x7lPAp4bQlyRJ0kjzie+SJEktMGRJkiS1wJAlSZLUAkOWJElSCwxZkiRJLTBkSZIktcCQJUmS1AJDliRJUgsGCllJLklye5JjSS5fNPfFJI8muXU4LUqSJI2eQXey7gV2AzcvMfe7wM5BG5IkSVoP+gpZSXYlmU5yZ5JPVtXdVTUNPLF4bVVNAd8ddqOSJEmjZMV3FybZAkwA26pqNsmGYTbwt/d/j7f8+leGWVKSJGnV9bOTtR3YX1WzAFV1tOmXJhlP0k3SnXt8rmk5SZKkNWfFnSwgQA3zS6tqH7APoNPp1Gd/+2eGWV6SJKkVn/ud/tf2s5M1BexIMgYw7ONCSZKk9WjFkFVVM8Be4ECSO4Ebklyc5AhwBXBTkpkn1yf5S+DzwGuSHEny+pZ6lyRJWrP6OS6kqiaByUXDm5ZZ+6qmTUmSJI06n/guSZLUAkOWJElSCwxZkiRJLTBkSZIktcCQJUmS1AJDliRJUgsMWZIkSS0wZEmSJLXAkCVJktSCgUJWkkuS3J7kWJLLF81dmeSu3t+Vw2lTkiRptPT1Wp0l3AvsBvYsHOy9PPo3gA5QwMEkX6iqR5o0KUmSNGr6CllJdjEfqAqYrqqdvfEnFi19PfClqjram/8S8Abg08vVPnz/I1w68V8GaF2SJGntWjFkJdkCTADbqmq2t1u1nJ8A7lvw+UhvbHHNcWAc4MyxC06qYUmSpFHQz07WdmB/Vc0CPLlLtYwsMVY/NFC1D9gH0Ol06ra9b+6jDUmSpNWV3+p/bT83voclgtIyjgALt6Y2AQ/0344kSdL60E/ImgJ2JBmDf7i5fTl/CrwuyblJzgVe1xuTJEk6rax4XFhVM0n2AgeSHAcOJbkRuAU4F3hjkvdU1ZaqOprkeuB/9C5/7wrHi5IkSetSX78urKpJYHLR8KZl1n4M+FjDviRJkkaaT3yXJElqgSFLkiSpBYYsSZKkFhiyJEmSWmDIkiRJaoEhS5IkqQWGLEmSpBYYsiRJklow1JCV5KIkU0mmk9yWZMkHlkqSJK13w97J+gDwiaraCrwXeN+Q60uSJI2ERiErya7ertWdST4JvIj5F0oD/Dnw800blCRJGkV9vbtwKUm2ABPAtqqaTbIB+AjwZuA/AP8CODvJWFU9vFydBx94nOvfc8+gbUiSJK1JTXaytgP7q2oWoKqOAnuAVyc5BLwauB84tvjCJONJukm6c3NzDVqQJElam1JVg12Y/CpwXlVdt8z8WcBfV9UJb37vdDrV7XYH6kGSJOlUSnKwqjr9rG2ykzUF7Egy1vvSDUk2Jnmy5ruBjzWoL0mSNLIGvierqmaS7AUOJDkOHAJuBd6XpIC/AK4eTpuSJEmjZeCQBVBVk8DkouH9TWpKkiStBz7xXZIkqQWGLEmSpBYYsiRJklpgyJIkSWqBIUuSJKkFhixJkqQWGLIkSZJaYMiSJElqgSFLkiSpBUMNWUkuTPLnSQ4lmU7yz4ZZX5IkaVQMeyfrOuBzVfUy4K3A7w25viRJ0kho9O7CJLuAPUAB08DfAef0pp8BPLBSjcOzcOl/atKFJEnS2jNwyEqyBZgAtlXVbJINwJnAnyW5Bvgx4LXLXDsOjAOc+Zytg7YgSZK0ZqWqBrtwPkidX1UTC8au7dX8YJJXAv8ZeHFVPbFcnU6nU91ud6AeJEmSTqUkB6uq08/aJvdkhfljwoWuAj4HUFVfBX4E2NjgOyRJkkZSk5A1BexIMgbQOy68F3hN7/MLmQ9Z327apCRJ0qgZ+J6sqppJshc4kOQ4cAj418BHk/wa87tcu2vQ80hJkqQR1ujXhVU1CUwuGt7WpKYkSdJ64BPfJUmSWmDIkiRJaoEhS5IkqQWGLEmSpBYYsiRJklpgyJIkSWqBIUuSJKkFhixJkqQWNHoY6WJJPgT8bO/j04HzquqZw/wOSZKkUTDUkFVVv/bkv5NcA7xsmPUlSZJGRaPjwiS7kkwnuTPJJxdN/wLw6Sb1JUmSRtXAO1lJtgATwLaqmk2yYcHcRcA/Br68Up17vvs447cdGbQNSZKkNanJTtZ2YH9VzQJU1dEFc2/tzR1f6sIk40m6Sbpzc3MNWpAkSVqbmtyTFaCWmXsrcPVyF1bVPmAfQKfTqX2XbmrQhiRJ0qnx0ZNY22QnawrYkWQM4MnjwiQvAM4FvtqgtiRJ0kgbeCerqmaS7AUOJDkOHAJ2M3/D+2eqarldLkmSpHWv0SMcqmoSmFw09ptNakqSJK0HPvFdkiSpBYYsSZKkFhiyJEmSWmDIkiRJaoEhS5IkqQWGLEmSpBYYsiRJklpgyJIkSWqBIUuSJKkFQw9ZSXYk+WaSmSQ3D7u+JEnSKGj0Wp3FkjwPeDewraoeSXLeMOtLkiSNikYhK8kuYA9QwDTwIHBjVT0CUFUPrVTjwceK62eONWlDkiRpzRn4uDDJFmAC2F5VLwHeCTwfeH6S/57ka0nesMy140m6Sbpzc3ODtiBJkrRmpaoGuzC5Bji/qiYWjN0KzAE7gE3AXwIvrqpHl6vT6XSq2+0O1IMkSdKplORgVXX6Wdvkxvcwf0y40BHgv1bVXFX9H+Aw8LwG3yFJkjSSmoSsKWBHkjGAJBuAPwR+tvd5I/PHh3/btElJkqRRM/CN71U1k2QvcCDJceAQ8K+A1yX5JnAceFdVPTycViVJkkZHo18XVtUkMLlo+NrenyRJ0mnLJ75LkiS1wJAlSZLUAkOWJElSCwxZkiRJLTBkSZIktcCQJUmS1AJDliRJUgsMWZIkSS0wZEmSJLVgqCErye4k305yR+/vbcOsL0mSNCoavVZnGZ+tqne0UFeSJGlkNApZSXYBe4ACpoGpk61x/+P/l3973/ubtCFJkrTmDHxcmGQLMAFsr6qXAO/sTb05yXSS/UkuWOba8STdJN25ublBW5AkSVqzUlWDXZhcA5xfVRMLxsaA71XVD5K8HdhRVdtPVKfT6VS32x2oB0mSpFMpycGq6vSztsmN72H+mPAfVNXDVfWD3sePAi9vUF+SJGlkNQlZU8CO3u4VSTYk+fEF85cBf9WkOUmSpFE18I3vVTWTZC9wIMlx4BDwYJLLgGPAUWD3ULqUJEkaMY1+XVhVk8DkouF3N6kpSZK0HvjEd0mSpBYYsiRJklpgyJIkSWqBIUuSJKkFhixJkqQWGLIkSZJaYMiSJElqgSFLkiSpBa2ErCSXJ6kkfb1AUZIkab0ZeshKcjbwq8DXh11bkiRpVDR6rU6SXcAeoIDpqtoJXA+8vze+sr8/DIcvbdKGJEnSmjPwTlaSLcAEsL2qXgK8M8nLgAuq6tYVrh1P0k3SnZubG7QFSZKkNavJTtZ2YH9VzfY+Pwr8AbB7pQurah+wD6DT6RQvuK1BG5IkSadK+l7Z5J6sMH9M+KSzgRcDtyW5G3gF8AVvfpckSaejJiFrCtiRZKz3+R9V1caq2lxVm4GvAZdVVbdpk5IkSaNm4OPCqppJshc4kOQ4cIg+jgolSZJOB41+XVhVk8DkMnOXNqktSZI0ynziuyRJUgsMWZIkSS0wZEmSJLXAkCVJktQCQ5YkSVILDFmSJEktMGRJkiS1wJAlSZLUAkOWJElSC4YaspK8Pck3ktyR5CtJXjTM+pIkSaNi2DtZN1fVT1bVS4H3AzcMub4kSdJIaPTuwiS7gD1AAdNVtXPB9I/1xk/s0cNwy6VN2pAkSVpzBg5ZSbYAE8C2qppNsqE3fjVwLfA0YPsy144D4wBbLzxz0BYkSZLWrFStvNm05IXJNcD5VTWxzPwvAq+vqitPVKfT6VS32x2oB0mSpFMpycGq6vSztsk9WeHEx4GfAd7UoL4kSdLIahKypoAdScYAkmxI8rwF8/8cuKtJc5IkSaNq4HuyqmomyV7gQJLjwCHgO0leC8wBjwAnPCqUJElarxr9urCqJoHJIfUiSZK0bvjEd0mSpBYYsiRJklpgyJIkSWqBIUuSJKkFhixJkqQWGLIkSZJaYMiSJElqgSFLkiSpBUMNWUmuTfLNJNNJppJcNMz6kiRJo2LYO1mHgE5VbQX2A+8fcn1JkqSR0Oi1Okl2AXuAAqaraueC6a8Bv7xikXvvgXe8rUkbkiRJa87AISvJFmAC2FZVs0k2LFpyFfAny1w7DowDbH3mOYO2IEmStGalqga7MLkGOL+qJpaY+2XgHcCrq+oHJ6rT6XSq2+0O1IMkSdKplORgVXX6WdvkuDDMHxMu/vLXMr/DtWLAkiRJWq+a3Pg+BexIMgaQZEOSlwE3AZdV1UPDaFCSJGkUDbyTVVUzSfYCB5IcZ/6XhZuAs4DPJwG4t6ouG0qnkiRJI6TRrwurahKYHFIvkiRJ64ZPfJckSWqBIUuSJKkFhixJkqQWGLIkSZJaYMiSJElqgSFLkiSpBYYsSZKkFhiyJEmSWmDIkiRJasFQQ1aSS5LcnuRYksuHWVuSJGmUDHsn615gN3DzkOtKkiSNlEbvLkyyC9gDFDBdVTt740/0XeTuB+Cq32jShiRJ0pozcMhKsgWYALZV1WySDSdx7TgwDrD1nI2DtiBJkrRmpaoGuzC5Bji/qiaWmPs4cGtV7V+pTqfTqW63O1APkiRJp1KSg1XV6Wdtk3uywvwxoSRJkhZpErKmgB1JxgBO5rhQkiRpvRs4ZFXVDLAXOJDkTuCGJBcnOQJcAdyUZGZIfUqSJI2URr8urKpJYHLR8KYmNSVJktYDn/guSZLUAkOWJElSCwxZkiRJLTBkSZIktcCQJUmS1AJDliRJUgsMWZIkSS0wZEmSJLVgqCEryZlJPpvkb5J8PcnmYdaXJEkaFcPeyboKeKSqngt8CPidIdeXJEkaCY1eq5NkF7AHKGAaOA/4zd70fuAjSVJV1eR7JEmSRs3AISvJFmAC2FZVs0k2AH8B3AdQVceSfAcYA2aXq/PYY48xPT09aBuSJElrUpPjwu3A/qqaBaiqo0CWWPdDu1hJxpN0k3Tn5uYatCBJkrQ2NTkuDD8coI4AFwBHkjwVeAZwdPGFVbUP2AfQ6XRq69atDdqQJElae5rsZE0BO5KMAfSOC78AXNmbvxz4svdjSZKk09HAO1lVNZNkL3AgyXHgEPB24JNJ/ob5Hay3DqdNSZKk0dLo14VVNQlMLhq+oklNSZKk9cAnvkuSJLXAkCVJktQCQ5YkSVILsto//kvyXeDwqjYhSZLUnxdU1dn9LGx04/uQHK6qzmo3IUmStJIk3X7XelwoSZLUAkOWJElSC9ZCyNq32g1IkiT1qe/csuo3vkuSJK1Ha2EnS5Ikad0xZEmSJLXAkCVJktQCQ5YkSVILDFmSJEkt+H8LopM0sc1fuQAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<Figure size 720x360 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "#Colour graph\n",
    "import matplotlib.pyplot as plt\n",
    "plt.close('all')\n",
    "f, (col) = plt.subplots(1, 1, sharey=True, figsize=(10, 5))\n",
    "col.set_xlim(1, 5)\n",
    "col.set_ylim(-0.5,15)\n",
    "col.set_title('Colours')\n",
    "\n",
    "x_major_ticks = [1, 5]\n",
    "x_major_tick_labels = [\"\", \"\"]\n",
    "col.set_xticks(x_major_ticks)\n",
    "col.set_xticklabels(x_major_tick_labels)\n",
    "\n",
    "y_major_ticks = [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14]\n",
    "y_major_tick_labels = [\"c0\", \"c1\", \"c2\", \"c3\", \"c4\", \"c5\", \"c6\", \"c7\", \"c8\", \"c9\", \"c10\", \"c11\", \"c12\", \"c13\", \"c14\"]\n",
    "col.set_yticks(y_major_ticks)\n",
    "col.set_yticklabels(y_major_tick_labels)\n",
    " \n",
    "# c0 points\n",
    "x0 = [1,5]\n",
    "y0 = [0,0]\n",
    "# plotting the c1 points \n",
    "plt.plot(x0, y0, label = \"c0\", color=c0)\n",
    " \n",
    "# c1 points\n",
    "x1 = [1,5]\n",
    "y1 = [1,1]\n",
    "# plotting the c1 points \n",
    "plt.plot(x1, y1, label = \"c1\", color=c1)\n",
    "\n",
    "# c2 points\n",
    "x2 = [1,5]\n",
    "y2 = [2,2]\n",
    "# plotting the c2 points \n",
    "plt.plot(x2, y2, label = \"c2\", color=c2)\n",
    "\n",
    "# c3 points\n",
    "x3 = [1,5]\n",
    "y3 = [3,3]\n",
    "# plotting the c3 points \n",
    "plt.plot(x3, y3, label = \"c3\", color=c3)\n",
    "\n",
    "# c4 points\n",
    "x4 = [1,5]\n",
    "y4 = [4, 4]\n",
    "# plotting the c4 points \n",
    "plt.plot(x4, y4, label = \"c4\", color=c4)\n",
    "\n",
    "# c5 points\n",
    "x5 = [1,5]\n",
    "y5 = [5, 5]\n",
    "# plotting the c5 points \n",
    "plt.plot(x5, y5, label = \"c5\", color=c5)\n",
    "\n",
    "# c6 points\n",
    "x6 = [1,5]\n",
    "y6 = [6, 6]\n",
    "# plotting the c6 points \n",
    "plt.plot(x6, y6, label = \"c6\", color=c6)\n",
    "\n",
    "# c7 points\n",
    "x7 = [1,5]\n",
    "y7 = [7, 7]\n",
    "# plotting the c7 points \n",
    "plt.plot(x7, y7, label = \"c7\", color=c7)\n",
    "\n",
    "# c8 points\n",
    "x8 = [1,5]\n",
    "y8 = [8, 8]\n",
    "# plotting the c8 points \n",
    "plt.plot(x8, y8, label = \"c8\", color=c8)\n",
    "\n",
    "# c9 points\n",
    "x9 = [1,5]\n",
    "y9 = [9, 9]\n",
    "# plotting the c9 points \n",
    "plt.plot(x9, y9, label = \"c9\", color=c9)\n",
    "\n",
    "# c10 points\n",
    "x10 = [1,5]\n",
    "y10 = [10, 10]\n",
    "# plotting the c4 points \n",
    "plt.plot(x10, y10, label = \"c10\", color=c10)\n",
    "\n",
    "# c11 points\n",
    "x11 = [1,5]\n",
    "y11 = [11, 11]\n",
    "# plotting the c11 points \n",
    "plt.plot(x11, y11, label = \"c11\", color=c11)\n",
    "\n",
    "# c12 points\n",
    "x12 = [1,5]\n",
    "y12 = [12, 12]\n",
    "# plotting the c12 points \n",
    "plt.plot(x12, y12, label = \"c12\", color=c12)\n",
    "\n",
    "# c13 points\n",
    "x13 = [1,5]\n",
    "y13 = [13, 13]\n",
    "# plotting the c13 points \n",
    "plt.plot(x13, y13, label = \"c13\", color=c13)\n",
    "\n",
    "# c14 points\n",
    "x14 = [1,5]\n",
    "y14 = [14, 14]\n",
    "# plotting the c14 points \n",
    "plt.plot(x14, y14, label = \"c14\", color=c14)\n",
    "\n",
    "# # show a legend on the plot\n",
    "# plt.legend()\n",
    " \n",
    "# function to show the plot\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 280,
   "metadata": {},
   "outputs": [],
   "source": [
    "missense = c6\n",
    "nonsense = c1\n",
    "synonymous = c4"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Totals for studies"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 281,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Number of individuals in each study\n",
    "Jaiswaltotal = 17182\n",
    "Genovesetotal = 12380\n",
    "McKerreltotal = 4219\n",
    "Youngtotal = 20\n",
    "Young2019totalcontrols = 69 #controls only\n",
    "ZinkWGStotal = 11262\n",
    "Acunatotal = 2006\n",
    "Coombstotal = 1591 #chemotherapy and radiotherapy naive individuals\n",
    "Desaitotal = 181 #controls only"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Functions for calculating trinucleotide-context specific mutation rates"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 282,
   "metadata": {
    "code_folding": [
     0,
     8
    ]
   },
   "outputs": [],
   "source": [
    "# Amino Acid Code:\n",
    "\n",
    "Amino_acid_abbreviations = {'Ala': 'A', 'Arg': 'R', 'Asn': 'N', 'Asp': 'D',\\\n",
    "                           'Cys': 'C', 'Glu': 'E', 'Gln': 'Q', 'Gly': 'G',\\\n",
    "                           'His': 'H', 'Ile': 'I', 'Leu': 'L', 'Lys': 'K',\\\n",
    "                           'Met': 'M', 'Phe': 'F', 'Pro': 'P', 'Ser': 'S',\\\n",
    "                           'Thr': 'T', 'Trp': 'W', 'Tyr': 'Y', 'Val': 'V'}\n",
    "\n",
    "Genetic_Code = {'TTT': 'Phe', 'TTC': 'Phe', 'TTA': 'Leu', 'TTG': 'Leu',\\\n",
    "               'TCT': 'Ser', 'TCC': 'Ser', 'TCA': 'Ser', 'TCG': 'Ser',\\\n",
    "               'TAT': 'Tyr', 'TAC': 'Tyr', 'TAA': '*', 'TAG': '*',\\\n",
    "               'TGT': 'Cys', 'TGC': 'Cys', 'TGA': '*', 'TGG': 'Trp',\\\n",
    "               'CTT': 'Leu', 'CTC': 'Leu', 'CTA': 'Leu', 'CTG': 'Leu',\\\n",
    "               'CCT': 'Pro', 'CCC': 'Pro', 'CCA': 'Pro', 'CCG': 'Pro',\\\n",
    "               'CAT': 'His', 'CAC': 'His', 'CAA': 'Gln', 'CAG': 'Gln',\\\n",
    "               'CGT': 'Arg', 'CGC': 'Arg', 'CGA': 'Arg', 'CGG': 'Arg',\\\n",
    "               'ATT': 'Ile', 'ATC': 'Ile', 'ATA': 'Ile', 'ATG': 'Met',\\\n",
    "               'ACT': 'Thr', 'ACC': 'Thr', 'ACA': 'Thr', 'ACG': 'Thr',\\\n",
    "               'AAT': 'Asn', 'AAC': 'Asn', 'AAA': 'Lys', 'AAG': 'Lys',\\\n",
    "               'AGT': 'Ser', 'AGC': 'Ser', 'AGA': 'Arg', 'AGG': 'Arg',\\\n",
    "               'GTT': 'Val', 'GTC': 'Val', 'GTA': 'Val', 'GTG': 'Val',\\\n",
    "               'GCT': 'Ala', 'GCC': 'Ala', 'GCA': 'Ala', 'GCG': 'Ala',\\\n",
    "               'GAT': 'Asp', 'GAC': 'Asp', 'GAA': 'Glu', 'GAG': 'Glu',\\\n",
    "               'GGT': 'Gly', 'GGC': 'Gly', 'GGA': 'Gly', 'GGG': 'Gly'}\n",
    "\n",
    "Possible_Amino_Acid_Changes = {'TTT': ('Ill', 'Leu', 'Val', 'Tyr', 'Ser', 'Cys', 'Leu', 'Phe', 'Leu'),\\\n",
    "                          'TTC': ('Ile', 'Leu', 'Val', 'Tyr', 'Ser', 'Cys', 'Leu', 'Phe', 'Leu'),\\\n",
    "                          'TTA': ('Ile', 'Leu', 'Val', '*', 'Ser', '*', 'Phe', 'Leu', 'Phe'),\\\n",
    "                          'TTG': ('Met', 'Leu', 'Val', '*', 'Ser', 'Val', 'Leu', 'Phe', 'Phe'),\\\n",
    "                          'TCT': ('Thr', 'Pro', 'Ala', 'Tyr', 'Phe', 'Cys', 'Ser', 'Ser', 'Ser'),\\\n",
    "                          'TCC': ('Thr', 'Pro', 'Ala', 'Tyr', 'Phe', 'Cys', 'Ser', 'Ser', 'Ser'),\\\n",
    "                          'TCA': ('Thr', 'Pro', 'Ala', '*', 'Leu', '*', 'Ser', 'Ser', 'Ser'),\\\n",
    "                          'TCG': ('Thr', 'Pro', 'Ala', '*', 'Leu', 'Trp', 'Ser', 'Ser', 'Ser'),\\\n",
    "                          'TAT': ('Ile', 'His', 'Asp', 'Ser', 'Phe', 'Cys', '*', 'Tyr', '*'),\\\n",
    "                          'TAC': ('Asn', 'His', 'Asp', 'Ser', 'Phe', 'Cys', '*', 'Tyr', '*'),\\\n",
    "                          'TAA': ('Lys', 'Gln', 'Glu', 'Ser', 'Leu', '*', 'Tyr', '*', 'Tyr'),\\\n",
    "                          'TAG': ('Lys', 'Gln', 'Glu', 'Ser', 'Leu', 'Trp', 'Tyr', '*', 'Tyr'),\\\n",
    "                          'TGT': ('Ser', 'Arg', 'Gly', 'Tyr', 'Ser', 'Phe', '*', 'Cys', 'Trp'),\\\n",
    "                          'TGC': ('Ser', 'Arg', 'Gly', 'Tyr', 'Ser', 'Phe', '*', 'Cys', 'Trp'),\\\n",
    "                          'TGA': ('Arg', 'Arg', 'Gly', '*', 'Ser', 'Leu', 'Cys', 'Trp', 'Cys'),\\\n",
    "                          'TGG': ('Arg', 'Arg', 'Gly', '*', 'Ser', 'Leu', '*', 'Cys', 'Cys'),\\\n",
    "                          'CTT': ('Ile', 'Phe', 'Val', 'His', 'Pro', 'Arg', 'Leu', 'Leu', 'Leu'),\\\n",
    "                          'CTC': ('Ile', 'Phe', 'Val', 'His', 'Pro', 'Arg', 'Leu', 'Leu', 'Leu'),\\\n",
    "                          'CTA': ('Ile', 'Leu', 'Val', 'Gln', 'Pro', 'Arg', 'Leu', 'Leu', 'Leu'),\\\n",
    "                          'CTG': ('Met', 'Leu', 'Val', 'Gln', 'Pro', 'Arg', 'Leu', 'Leu', 'Leu'),\\\n",
    "                          'CCT': ('Thr', 'Ser', 'Ala', 'His', 'Leu', 'Arg', 'Pro', 'Pro', 'Pro'),\\\n",
    "                          'CCC': ('Thr', 'Ser', 'Ala', 'His', 'Leu', 'Arg', 'Pro', 'Pro', 'Pro'),\\\n",
    "                          'CCA': ('Thr', 'Ser', 'Ala', 'Gln', 'Leu', 'Arg', 'Pro', 'Pro', 'Pro'),\\\n",
    "                          'CCG': ('Thr', 'Ser', 'Ala', 'Gln', 'Leu', 'Arg', 'Pro', 'Pro', 'Pro'),\\\n",
    "                          'CAT': ('Asn', 'Tyr', 'Asp', 'Pro', 'Leu', 'Arg', 'Gln', 'His', 'Gln'),\\\n",
    "                          'CAC': ('Asn', 'Tyr', 'Asp', 'Pro', 'Leu', 'Arg', 'Gln', 'His', 'Gln'),\\\n",
    "                          'CAA': ('Lys', '*', 'Glu', 'Pro', 'Leu', 'Arg', 'His', 'Gln', 'His'),\\\n",
    "                          'CAG': ('Lys', '*', 'Glu', 'Pro', 'Leu', 'Arg', 'His', 'Gln', 'His'),\\\n",
    "                          'CGT': ('Ser', 'Cys', 'Gly', 'His', 'Pro', 'Leu', 'Arg', 'Arg', 'Arg'),\\\n",
    "                          'CGC': ('Ser', 'Cys', 'Gly', 'His', 'Pro', 'Leu', 'Arg', 'Arg', 'Arg'),\\\n",
    "                          'CGA': ('Arg', '*', 'Gly', 'Gln', 'Pro', 'Leu', 'Arg', 'Arg', 'Arg'),\\\n",
    "                          'CGG': ('Arg', 'Trp', 'Gly', 'Gln', 'Pro', 'Leu', 'Arg', 'Arg', 'Arg'),\\\n",
    "                               'ATT': ('Leu', 'Phe', 'Val', 'Asn', 'Thr', 'Ser', 'Ile', 'Ile', 'Met'),\\\n",
    "                              'ATC': ('Leu', 'Phe', 'Val', 'Asn', 'Thr', 'Ser', 'Ile', 'Ile', 'Met'),\\\n",
    "                              'ATA': ('Leu', 'Leu', 'Val', 'Lys', 'Thr', 'Arg', 'Ile', 'Met', 'Ile'),\\\n",
    "                              'ATG': ('Leu', 'Leu', 'Val', 'Lys', 'Thr', 'Arg', 'Ile', 'Ile', 'Ile'),\\\n",
    "                              'ACT': ('Pro', 'Ser', 'Ala', 'Asn', 'Ile', 'Ser', 'Thr', 'Thr', 'Thr'),\\\n",
    "                              'ACC': ('Pro', 'Ser', 'Ala', 'Asn', 'Ile', 'Ser', 'Thr', 'Thr', 'Thr'),\\\n",
    "                              'ACA': ('Pro', 'Ser', 'Ala', 'Lys', 'Ile', 'Arg', 'Thr', 'Thr', 'Thr'),\\\n",
    "                              'ACG': ('Pro', 'Ser', 'Ala', 'Lys', 'Met', 'Arg', 'Thr', 'Thr', 'Thr'),\\\n",
    "                              'AAT': ('His', 'Tyr', 'Asp', 'Thr', 'Ile', 'Ser', 'Lys', 'Asn', 'Lys'),\\\n",
    "                              'AAC': ('His', 'Tyr', 'Asp', 'Thr', 'Ile', 'Ser', 'Lys', 'Asn', 'Lys'),\\\n",
    "                              'AAA': ('Gln', '*', 'Glu', 'Thr', 'Ile', 'Arg', 'Asn', 'Lys', 'Asn'),\\\n",
    "                              'AAG': ('Gln', '*', 'Glu', 'Thr', 'Met', 'Arg', 'Asn', 'Lys', 'Asn'),\\\n",
    "                              'AGT': ('Arg', 'Cys', 'Gly', 'Asn', 'Thr', 'Ile', 'Arg', 'Ser', 'Arg'),\\\n",
    "                              'AGC': ('Arg', 'Cys', 'Gly', 'Asn', 'Thr', 'Ile', 'Arg', 'Ser', 'Arg'),\\\n",
    "                              'AGA': ('Arg', '*', 'Gly', 'Lys', 'Thr', 'Ile', 'Ser', 'Arg', 'Ser'),\\\n",
    "                              'AGG': ('Arg', 'Trp', 'Gly', 'Lys', 'Thr', 'Met', 'Arg', 'Ser', 'Ser'),\\\n",
    "                              'GTT': ('Ile', 'Leu', 'Phe', 'Asp', 'Ala', 'Gly', 'Val', 'Val', 'Val'),\\\n",
    "                              'GTC': ('Ile', 'Leu', 'Phe', 'Asp', 'Ala', 'Gly', 'Val', 'Val', 'Val'),\\\n",
    "                              'GTA': ('Ile', 'Leu', 'Leu', 'Glu', 'Ala', 'Gly', 'Val', 'Val', 'Val'),\\\n",
    "                              'GTG': ('Met', 'Leu', 'Leu', 'Glu', 'Ala', 'Gly', 'Val', 'Val', 'Val'),\\\n",
    "                              'GCT': ('Thr', 'Pro', 'Ser', 'Asp', 'Val', 'Gly', 'Ala', 'Ala', 'Ala'),\\\n",
    "                              'GCC': ('Thr', 'Pro', 'Ser', 'Asp', 'Val', 'Gly', 'Ala', 'Ala', 'Ala'),\\\n",
    "                              'GCA': ('Thr', 'Pro', 'Ser', 'Glu', 'Val', 'Gly', 'Ala', 'Ala', 'Ala'),\\\n",
    "                              'GCG': ('Thr', 'Pro', 'Ser', 'Glu', 'Val', 'Gly', 'Ala', 'Ala', 'Ala'),\\\n",
    "                              'GAT': ('Asn', 'His', 'Tyr', 'Ala', 'Val', 'Gly', 'Glu', 'Asp', 'Glu'),\\\n",
    "                              'GAC': ('Asn', 'His', 'Tyr', 'Ala', 'Val', 'Gly', 'Glu', 'Asp', 'Glu'),\\\n",
    "                              'GAA': ('Lys', 'Gln', '*', 'Ala', 'Val', 'Gly', 'Asp', 'Glu', 'Asp'),\\\n",
    "                              'GAG': ('Lys', 'Gln', '*', 'Ala', 'Val','Gly', 'Asp', 'Glu', 'Asp'),\\\n",
    "                              'GGT': ('Ser', 'Arg', 'Cys', 'Asp', 'Ala', 'Val', 'Gly','Gly', 'Gly'),\\\n",
    "                              'GGC': ('Ser', 'Arg', 'Cys', 'Asp', 'Ala', 'Val', 'Gly', 'Gly', 'Gly'),\\\n",
    "                              'GGA': ('Arg', 'Arg', '*', 'Glu', 'Ala', 'Val', 'Gly', 'Gly', 'Gly'),\\\n",
    "                              'GGG': ('Arg', 'Arg', 'Trp', 'Glu', 'Ala', 'Val', 'Gly', 'Gly', 'Gly')}\n",
    "\n",
    "Possible_Amino_Acid_Changes_letters = {'TTT': ('I', 'L', 'V', 'Y', 'S', 'C', 'L', 'F', 'L'),\\\n",
    "                          'TTC': ('I', 'L', 'V', 'Y', 'S', 'C', 'L', 'F', 'L'),\\\n",
    "                          'TTA': ('I', 'L', 'V', '*', 'S', '*', 'F', 'L', 'F'),\\\n",
    "                          'TTG': ('M', 'L', 'V', '*', 'S', 'V', 'L', 'F', 'F'),\\\n",
    "                          'TCT': ('T', 'P', 'A', 'Y', 'F', 'C', 'S', 'S', 'S'),\\\n",
    "                          'TCC': ('T', 'P', 'A', 'Y', 'F', 'C', 'S', 'S', 'S'),\\\n",
    "                          'TCA': ('T', 'P', 'A', '*', 'L', '*', 'S', 'S', 'S'),\\\n",
    "                          'TCG': ('T', 'P', 'A', '*', 'L', 'W', 'S', 'S', 'S'),\\\n",
    "                          'TAT': ('I', 'H', 'D', 'S', 'F', 'C', '*', 'Y', '*'),\\\n",
    "                          'TAC': ('N', 'H', 'D', 'S', 'F', 'C', '*', 'Y', '*'),\\\n",
    "                          'TAA': ('K', 'Q', 'E', 'S', 'L', '*', 'Y', '*', 'Y'),\\\n",
    "                          'TAG': ('K', 'Q', 'E', 'S', 'L', 'W', 'Y', '*', 'Y'),\\\n",
    "                          'TGT': ('S', 'R', 'G', 'Y', 'S', 'F', '*', 'C', 'W'),\\\n",
    "                          'TGC': ('S', 'R', 'G', 'Y', 'S', 'F', '*', 'C', 'W'),\\\n",
    "                          'TGA': ('R', 'R', 'G', '*', 'S', 'L', 'C', 'W', 'C'),\\\n",
    "                          'TGG': ('R', 'R', 'G', '*', 'S', 'L', '*', 'C', 'C'),\\\n",
    "                          'CTT': ('I', 'F', 'V', 'H', 'P', 'R', 'L', 'L', 'L'),\\\n",
    "                          'CTC': ('I', 'F', 'V', 'H', 'P', 'R', 'L', 'L', 'L'),\\\n",
    "                          'CTA': ('I', 'L', 'V', 'Q', 'P', 'R', 'L', 'L', 'L'),\\\n",
    "                          'CTG': ('M', 'L', 'V', 'Q', 'P', 'R', 'L', 'L', 'L'),\\\n",
    "                          'CCT': ('T', 'S', 'A', 'H', 'L', 'R', 'P', 'P', 'P'),\\\n",
    "                          'CCC': ('T', 'S', 'A', 'H', 'L', 'R', 'P', 'P', 'P'),\\\n",
    "                          'CCA': ('T', 'S', 'A', 'Q', 'L', 'R', 'P', 'P', 'P'),\\\n",
    "                          'CCG': ('T', 'S', 'A', 'Q', 'L', 'R', 'P', 'P', 'P'),\\\n",
    "                          'CAT': ('N', 'Y', 'D', 'P', 'L', 'R', 'Q', 'H', 'Q'),\\\n",
    "                          'CAC': ('N', 'Y', 'D', 'P', 'L', 'R', 'Q', 'H', 'Q'),\\\n",
    "                          'CAA': ('K', '*', 'E', 'P', 'L', 'R', 'H', 'Q', 'H'),\\\n",
    "                          'CAG': ('K', '*', 'E', 'P', 'L', 'R', 'H', 'Q', 'H'),\\\n",
    "                          'CGT': ('S', 'C', 'G', 'H', 'P', 'L', 'R', 'R', 'R'),\\\n",
    "                          'CGC': ('S', 'C', 'G', 'H', 'P', 'L', 'R', 'R', 'R'),\\\n",
    "                          'CGA': ('R', '*', 'G', 'Q', 'P', 'L', 'R', 'R', 'R'),\\\n",
    "                          'CGG': ('R', 'W', 'G', 'Q', 'P', 'L', 'R', 'R', 'R'),\\\n",
    "                               'ATT': ('L', 'F', 'V', 'N', 'T', 'S', 'I', 'I', 'M'),\\\n",
    "                              'ATC': ('L', 'F', 'V', 'N', 'T', 'S', 'I', 'I', 'M'),\\\n",
    "                              'ATA': ('L', 'L', 'V', 'K', 'T', 'R', 'I', 'M', 'I'),\\\n",
    "                              'ATG': ('L', 'L', 'V', 'K', 'T', 'R', 'I', 'I', 'I'),\\\n",
    "                              'ACT': ('P', 'S', 'A', 'N', 'I', 'S', 'T', 'T', 'T'),\\\n",
    "                              'ACC': ('P', 'S', 'A', 'N', 'I', 'S', 'T', 'T', 'T'),\\\n",
    "                              'ACA': ('P', 'S', 'A', 'K', 'I', 'R', 'T', 'T', 'T'),\\\n",
    "                              'ACG': ('P', 'S', 'A', 'K', 'M', 'R', 'T', 'T', 'T'),\\\n",
    "                              'AAT': ('H', 'Y', 'D', 'T', 'I', 'S', 'K', 'N', 'K'),\\\n",
    "                              'AAC': ('H', 'Y', 'D', 'T', 'I', 'S', 'K', 'N', 'K'),\\\n",
    "                              'AAA': ('Q', '*', 'E', 'T', 'I', 'R', 'N', 'K', 'N'),\\\n",
    "                              'AAG': ('Q', '*', 'E', 'T', 'M', 'R', 'N', 'K', 'N'),\\\n",
    "                              'AGT': ('R', 'C', 'G', 'N', 'T', 'I', 'R', 'S', 'R'),\\\n",
    "                              'AGC': ('R', 'C', 'G', 'N', 'T', 'I', 'R', 'S', 'R'),\\\n",
    "                              'AGA': ('R', '*', 'G', 'K', 'T', 'I', 'S', 'R', 'S'),\\\n",
    "                              'AGG': ('R', 'W', 'G', 'K', 'T', 'M', 'R', 'S', 'S'),\\\n",
    "                              'GTT': ('I', 'L', 'F', 'D', 'A', 'G', 'V', 'V', 'V'),\\\n",
    "                              'GTC': ('I', 'L', 'F', 'D', 'A', 'G', 'V', 'V', 'V'),\\\n",
    "                              'GTA': ('I', 'L', 'L', 'E', 'A', 'G', 'V', 'V', 'V'),\\\n",
    "                              'GTG': ('M', 'L', 'L', 'E', 'A', 'G', 'V', 'V', 'V'),\\\n",
    "                              'GCT': ('T', 'P', 'S', 'D', 'V', 'G', 'A', 'A', 'A'),\\\n",
    "                              'GCC': ('T', 'P', 'S', 'D', 'V', 'G', 'A', 'A', 'A'),\\\n",
    "                              'GCA': ('T', 'P', 'S', 'E', 'V', 'G', 'A', 'A', 'A'),\\\n",
    "                              'GCG': ('T', 'P', 'S', 'E', 'V', 'G', 'A', 'A', 'A'),\\\n",
    "                              'GAT': ('N', 'H', 'Y', 'A', 'V', 'G', 'E', 'D', 'E'),\\\n",
    "                              'GAC': ('N', 'H', 'Y', 'A', 'V', 'G', 'E', 'D', 'E'),\\\n",
    "                              'GAA': ('K', 'Q', '*', 'A', 'V', 'G', 'D', 'E', 'D'),\\\n",
    "                              'GAG': ('K', 'Q', '*', 'A', 'V','G', 'D', 'E', 'D'),\\\n",
    "                              'GGT': ('S', 'R', 'C', 'D', 'A', 'V', 'G','G', 'G'),\\\n",
    "                              'GGC': ('S', 'R', 'C', 'D', 'A', 'V', 'G', 'G', 'G'),\\\n",
    "                              'GGA': ('R', 'R', '*', 'E', 'A', 'V', 'G', 'G', 'G'),\\\n",
    "                              'GGG': ('R', 'R', 'W', 'E', 'A', 'V', 'G', 'G', 'G')}\n",
    "\n",
    "Possible_Base_Changes = {'TTT': ('A', 'C', 'G', 'A', 'C', 'G', 'A', 'C', 'G'),\\\n",
    "                          'TTC': ('A', 'C', 'G', 'A', 'C', 'G', 'A', 'T', 'G'),\\\n",
    "                          'TTA': ('A', 'C', 'G', 'A', 'C', 'G', 'C', 'G', 'T'),\\\n",
    "                          'TTG': ('A', 'C', 'G', 'A', 'C', 'G', 'A', 'C', 'T'),\\\n",
    "                          'TCT': ('A', 'C', 'G', 'A', 'T', 'G', 'A', 'C', 'G'),\\\n",
    "                          'TCC': ('A', 'C', 'G', 'A', 'T', 'G', 'A', 'T', 'G'),\\\n",
    "                          'TCA': ('A', 'C', 'G', 'A', 'T', 'G', 'C', 'T', 'G'),\\\n",
    "                          'TCG': ('A', 'C', 'G', 'A', 'T', 'G', 'A', 'C', 'T'),\\\n",
    "                          'TAT': ('A', 'C', 'G', 'C', 'T', 'G', 'A', 'C', 'G'),\\\n",
    "                          'TAC': ('A', 'C', 'G', 'C', 'T', 'G', 'A', 'T', 'G'),\\\n",
    "                          'TAA': ('A', 'C', 'G', 'C', 'T', 'G', 'C', 'G', 'T'),\\\n",
    "                          'TAG': ('A', 'C', 'G', 'C', 'T', 'G', 'T', 'A', 'C'),\\\n",
    "                          'TGT': ('A', 'C', 'G', 'A', 'C', 'T', 'A', 'C', 'G'),\\\n",
    "                          'TGC': ('A', 'C', 'G', 'A', 'C', 'T', 'A', 'T', 'G'),\\\n",
    "                          'TGA': ('A', 'C', 'G', 'A', 'C', 'T', 'C', 'G', 'T'),\\\n",
    "                          'TGG': ('A', 'C', 'G', 'A', 'C', 'T', 'A', 'C', 'T'),\\\n",
    "                          'CTT': ('A', 'T', 'G', 'A', 'C', 'G', 'A', 'C', 'G'),\\\n",
    "                          'CTC': ('A', 'T', 'G', 'A', 'C', 'G', 'A', 'T', 'G'),\\\n",
    "                          'CTA': ('A', 'T', 'G', 'A', 'C', 'G', 'C', 'G', 'T'),\\\n",
    "                          'CTG': ('A', 'T', 'G', 'A', 'C', 'G', 'A', 'C', 'T'),\\\n",
    "                          'CCT': ('A', 'T', 'G', 'A', 'T', 'G', 'A', 'C', 'G'),\\\n",
    "                          'CCC': ('A', 'T', 'G', 'A', 'T', 'G', 'A', 'T', 'G'),\\\n",
    "                          'CCA': ('A', 'T', 'G', 'A', 'T', 'G', 'C', 'T', 'G'),\\\n",
    "                          'CCG': ('A', 'T', 'G', 'A', 'T', 'G', 'A', 'C', 'T'),\\\n",
    "                          'CAT': ('A', 'T', 'G', 'C', 'T', 'G', 'A', 'C', 'G'),\\\n",
    "                          'CAC': ('A', 'T', 'G', 'C', 'T', 'G', 'A', 'T', 'G'),\\\n",
    "                          'CAA': ('A', 'T', 'G', 'C', 'T', 'G', 'C', 'G', 'T'),\\\n",
    "                          'CAG': ('A', 'T', 'G', 'C', 'T', 'G', 'T', 'A', 'C'),\\\n",
    "                          'CGT': ('A', 'T', 'G', 'A', 'C', 'T', 'A', 'C', 'G'),\\\n",
    "                          'CGC': ('A', 'T', 'G', 'A', 'C', 'T', 'A', 'T', 'G'),\\\n",
    "                          'CGA': ('A', 'T', 'G', 'A', 'C', 'T', 'C', 'G', 'T'),\\\n",
    "                          'CGG': ('A', 'T', 'G', 'A', 'C', 'T', 'A', 'C', 'T'),\\\n",
    "                               'ATT': ('C', 'T', 'G', 'A', 'C', 'G', 'A', 'C', 'G'),\\\n",
    "                              'ATC': ('C', 'T', 'G', 'A', 'C', 'G', 'A', 'T', 'G'),\\\n",
    "                              'ATA': ('C', 'T', 'G', 'A', 'C', 'G', 'C', 'G', 'T'),\\\n",
    "                              'ATG': ('C', 'T', 'G', 'A', 'C', 'G', 'A', 'C', 'T'),\\\n",
    "                              'ACT': ('C', 'T', 'G', 'A', 'T', 'G', 'A', 'C', 'G'),\\\n",
    "                              'ACC': ('C', 'T', 'G', 'A', 'T', 'G', 'A', 'T', 'G'),\\\n",
    "                              'ACA': ('C', 'T', 'G', 'A', 'T', 'G', 'C', 'T', 'G'),\\\n",
    "                              'ACG': ('C', 'T', 'G', 'A', 'T', 'G', 'A', 'C', 'T'),\\\n",
    "                              'AAT': ('C', 'T', 'G', 'C', 'T', 'G', 'A', 'C', 'G'),\\\n",
    "                              'AAC': ('C', 'T', 'G', 'C', 'T', 'G', 'A', 'T', 'G'),\\\n",
    "                              'AAA': ('C', 'T', 'G', 'C', 'T', 'G', 'C', 'G', 'T'),\\\n",
    "                              'AAG': ('C', 'T', 'G', 'C', 'T', 'G', 'T', 'A', 'C'),\\\n",
    "                              'AGT': ('C', 'T', 'G', 'A', 'C', 'T', 'A', 'C', 'G'),\\\n",
    "                              'AGC': ('C', 'T', 'G', 'A', 'C', 'T', 'A', 'T', 'G'),\\\n",
    "                              'AGA': ('C', 'T', 'G', 'A', 'C', 'T', 'C', 'G', 'T'),\\\n",
    "                              'AGG': ('C', 'T', 'G', 'A', 'C', 'T', 'A', 'C', 'T'),\\\n",
    "                              'GTT': ('A', 'C', 'T', 'A', 'C', 'G', 'A', 'C', 'G'),\\\n",
    "                              'GTC': ('A', 'C', 'T', 'A', 'C', 'G', 'A', 'T', 'G'),\\\n",
    "                              'GTA': ('A', 'C', 'T', 'A', 'C', 'G', 'C', 'G', 'T'),\\\n",
    "                              'GTG': ('A', 'C', 'T', 'A', 'C', 'G', 'A', 'C', 'T'),\\\n",
    "                              'GCT': ('A', 'C', 'T', 'A', 'T', 'G', 'A', 'C', 'G'),\\\n",
    "                              'GCC': ('A', 'C', 'T', 'A', 'T', 'G', 'A', 'T', 'G'),\\\n",
    "                              'GCA': ('A', 'C', 'T', 'A', 'T', 'G', 'C', 'T', 'G'),\\\n",
    "                              'GCG': ('A', 'C', 'T', 'A', 'T', 'G', 'A', 'C', 'T'),\\\n",
    "                              'GAT': ('A', 'C', 'T', 'C', 'T', 'G', 'A', 'C', 'G'),\\\n",
    "                              'GAC': ('A', 'C', 'T', 'C', 'T', 'G', 'A', 'T', 'G'),\\\n",
    "                              'GAA': ('A', 'C', 'T', 'C', 'T', 'G', 'C', 'G', 'T'),\\\n",
    "                              'GAG': ('A', 'C', 'T', 'C', 'T', 'G', 'T', 'A', 'C'),\\\n",
    "                              'GGT': ('A', 'C', 'T', 'A', 'C', 'T', 'A','C', 'G'),\\\n",
    "                              'GGC': ('A', 'C', 'T', 'A', 'C', 'T', 'A', 'T', 'G'),\\\n",
    "                              'GGA': ('A', 'C', 'T', 'A', 'C', 'T', 'C', 'G', 'T'),\\\n",
    "                              'GGG': ('A', 'C', 'T', 'A', 'C', 'T', 'A', 'C', 'T')}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 283,
   "metadata": {
    "code_folding": []
   },
   "outputs": [],
   "source": [
    "#Import mutation rates\n",
    "import csv\n",
    "filename=\"Trinucleotide_context_mutation_rates.csv\" #csv file containing trinucleotide-context specific mutation rates (calculated from Lee-Six et al 2018 data)\n",
    "\n",
    "with open(filename, 'r') as csvfile:\n",
    "    read_reader = csv.reader(csvfile)  #csv.reader returns a reader object which will iterate over lines in the csvfile\n",
    "    row_count=0\n",
    "    Mutation_rates = {} #create a dictionary where key = trincucleotide site, value = mutation rate\n",
    "\n",
    "    for row in read_reader:\n",
    "        if row_count>0:\n",
    "            site = row[0] #trinucleotide site\n",
    "            comp_site = row[2] #complementary site\n",
    "            if site != '':\n",
    "                mutation_rate = float(row[7])\n",
    "                Mutation_rates[site] = mutation_rate\n",
    "                Mutation_rates[comp_site] = mutation_rate\n",
    "        row_count = row_count+1"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Calculating mutation rates for individual variants"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### DNMT3A"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 284,
   "metadata": {
    "code_folding": [
     0
    ]
   },
   "outputs": [],
   "source": [
    "def mutation_rates_variants_DNMT3A(variant):\n",
    "    intron_exon_boundary_caution_base_DNMT3A_NP072046 = [1279, 1429, 1474, 1554, 1852, 2083, 2082, 2173, 2322, 2478]\n",
    "\n",
    "    #The mutation rate at a base is calculated based on its trinucleotide context.  To obtain the trinucleotide context, the base either side of the mutated base needs to be ascertained.\n",
    "    #The coding sequence (+ an extra base at each end) of each gene is used when determining the trinucleotide context of the base change.  The problem with using the coding sequence is that\n",
    "    #if a base change falls at the end of an exon, its neighbouring base will actuall be an intronic base, which may be different to the 'neighbouring' base in the coding sequence.\n",
    "    #To account for this, there is an \"intron_exon_check\" function that is called at the intron/exon boundary if the neighbouring base in the intron is different to the neighbouring base in the coding sequence.\n",
    "    \n",
    "    def intron_exon_check_DNMT3A_NP072046(position_in_gene_number):\n",
    "        if position_in_gene_number == 1279:\n",
    "            gene_pos = 1279\n",
    "            first_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos-1]\n",
    "            middle_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "    #             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 1429:\n",
    "            gene_pos = 1429\n",
    "            first_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos-1]\n",
    "            middle_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "    #             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 1474:\n",
    "            gene_pos = 1474\n",
    "            first_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos-1]\n",
    "            middle_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "    #             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 1554:\n",
    "            gene_pos = 1554\n",
    "            first_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos-1]\n",
    "            middle_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "    #             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 1852:\n",
    "            gene_pos = 1852\n",
    "            first_base = 'G'\n",
    "            middle_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos]\n",
    "            third_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos+1]\n",
    "            triplet = first_base+middle_base+third_base\n",
    "    #             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 2083:\n",
    "            gene_pos = 2083\n",
    "            first_base = 'G'\n",
    "            middle_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos]\n",
    "            third_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos+1]\n",
    "            triplet = first_base+middle_base+third_base\n",
    "    #             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 2082:\n",
    "            gene_pos = 2082\n",
    "            first_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos-1]\n",
    "            middle_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "    #             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 2173:\n",
    "            gene_pos = 2173\n",
    "            first_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos-1]\n",
    "            middle_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "    #             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 2322:\n",
    "            gene_pos = 2322\n",
    "            first_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos-1]\n",
    "            middle_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "    #             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 2478:\n",
    "            gene_pos = 2478\n",
    "            first_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos-1]\n",
    "            middle_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "    #             print(triplet)\n",
    "\n",
    "        return triplet\n",
    "\n",
    "    #DNMT3A transcript used:\n",
    "    #Downloaded from http://grch37.ensembl.org/Homo_sapiens/Transcript/Exons?db=core;g=ENSG00000119772;r=2:25455845-25564754;t=ENST00000321117\n",
    "    DNMT3A_exons_NP_072046 = 'ATGCCCGCCATGCCCTCCAGCGGCCCCGGGGACACCAGCAGCTCTGCTGCGGAGCGGGAGGAGGACCGAAAGGACGGAGAGGAGCAGGAGGAGCCGCGTGGCAAGGAGGAGCGCCAAGAGCCCAGCACCACGGCACGGAAGGTGGGGCGGCCTGGGAGGAAGCGCAAGCACCCCCCGGTGGAAAGCGGTGACACGCCAAAGGACCCTGCGGTGATCTCCAAGTCCCCATCCATGGCCCAGGACTCAGGCGCCTCAGAGCTATTACCCAATGGGGACTTGGAGAAGCGGAGTGAGCCCCAGCCAGAGGAGGGGAGCCCTGCTGGGGGGCAGAAGGGCGGGGCCCCAGCAGAGGGAGAGGGTGCAGCTGAGACCCTGCCTGAAGCCTCAAGAGCAGTGGAAAATGGCTGCTGCACCCCCAAGGAGGGCCGAGGAGCCCCTGCAGAAGCGGGCAAAGAACAGAAGGAGACCAACATCGAATCCATGAAAATGGAGGGCTCCCGGGGCCGGCTGCGGGGTGGCTTGGGCTGGGAGTCCAGCCTCCGTCAGCGGCCCATGCCGAGGCTCACCTTCCAGGCGGGGGACCCCTACTACATCAGCAAGCGCAAGCGGGACGAGTGGCTGGCACGCTGGAAAAGGGAGGCTGAGAAGAAAGCCAAGGTCATTGCAGGAATGAATGCTGTGGAAGAAAACCAGGGGCCCGGGGAGTCTCAGAAGGTGGAGGAGGCCAGCCCTCCTGCTGTGCAGCAGCCCACTGACCCCGCATCCCCCACTGTGGCTACCACGCCTGAGCCCGTGGGGTCCGATGCTGGGGACAAGAATGCCACCAAAGCAGGCGATGACGAGCCAGAGTACGAGGACGGCCGGGGCTTTGGCATTGGGGAGCTGGTGTGGGGGAAACTGCGGGGCTTCTCCTGGTGGCCAGGCCGCATTGTGTCTTGGTGGATGACGGGCCGGAGCCGAGCAGCTGAAGGCACCCGCTGGGTCATGTGGTTCGGAGACGGCAAATTCTCAGTGGTGTGTGTTGAGAAGCTGATGCCGCTGAGCTCGTTTTGCAGTGCGTTCCACCAGGCCACGTACAACAAGCAGCCCATGTACCGCAAAGCCATCTACGAGGTCCTGCAGGTGGCCAGCAGCCGCGCGGGGAAGCTGTTCCCGGTGTGCCACGACAGCGATGAGAGTGACACTGCCAAGGCCGTGGAGGTGCAGAACAAGCCCATGATTGAATGGGCCCTGGGGGGCTTCCAGCCTTCTGGCCCTAAGGGCCTGGAGCCACCAGAAGAAGAGAAGAATCCCTACAAAGAAGTGTACACGGACATGTGGGTGGAACCTGAGGCAGCTGCCTACGCACCACCTCCACCAGCCAAAAAGCCCCGGAAGAGCACAGCGGAGAAGCCCAAGGTCAAGGAGATTATTGATGAGCGCACAAGAGAGCGGCTGGTGTACGAGGTGCGGCAGAAGTGCCGGAACATTGAGGACATCTGCATCTCCTGTGGGAGCCTCAATGTTACCCTGGAACACCCCCTCTTCGTTGGAGGAATGTGCCAAAACTGCAAGAACTGCTTTCTGGAGTGTGCGTACCAGTACGACGACGACGGCTACCAGTCCTACTGCACCATCTGCTGTGGGGGCCGTGAGGTGCTCATGTGCGGAAACAACAACTGCTGCAGGTGCTTTTGCGTGGAGTGTGTGGACCTCTTGGTGGGGCCGGGGGCTGCCCAGGCAGCCATTAAGGAAGACCCCTGGAACTGCTACATGTGCGGGCACAAGGGTACCTACGGGCTGCTGCGGCGGCGAGAGGACTGGCCCTCCCGGCTCCAGATGTTCTTCGCTAATAACCACGACCAGGAATTTGACCCTCCAAAGGTTTACCCACCTGTCCCAGCTGAGAAGAGGAAGCCCATCCGGGTGCTGTCTCTCTTTGATGGAATCGCTACAGGGCTCCTGGTGCTGAAGGACTTGGGCATTCAGGTGGACCGCTACATTGCCTCGGAGGTGTGTGAGGACTCCATCACGGTGGGCATGGTGCGGCACCAGGGGAAGATCATGTACGTCGGGGACGTCCGCAGCGTCACACAGAAGCATATCCAGGAGTGGGGCCCATTCGATCTGGTGATTGGGGGCAGTCCCTGCAATGACCTCTCCATCGTCAACCCTGCTCGCAAGGGCCTCTACGAGGGCACTGGCCGGCTCTTCTTTGAGTTCTACCGCCTCCTGCATGATGCGCGGCCCAAGGAGGGAGATGATCGCCCCTTCTTCTGGCTCTTTGAGAATGTGGTGGCCATGGGCGTTAGTGACAAGAGGGACATCTCGCGATTTCTCGAGTCCAACCCTGTGATGATTGATGCCAAAGAAGTGTCAGCTGCACACAGGGCCCGCTACTTCTGGGGTAACCTTCCCGGTATGAACAGGCCGTTGGCATCCACTGTGAATGATAAGCTGGAGCTGCAGGAGTGTCTGGAGCATGGCAGGATAGCCAAGTTCAGCAAAGTGAGGACCATTACTACGAGGTCAAACTCCATAAAGCAGGGCAAAGACCAGCATTTTCCTGTCTTCATGAATGAGAAAGAGGACATCTTATGGTGCACTGAAATGGAAAGGGTATTTGGTTTCCCAGTCCACTATACTGACGTCTCCAACATGAGCCGCTTGGCGAGGCAGAGACTGCTGGGCCGGTCATGGAGCGTGCCAGTCATCCGCCACCTCTTCGCTCCGCTGAAGGAGTATTTTGCGTGTGTGTAA'\n",
    "\n",
    "    #Split the sequence into codons \n",
    "    DNMT3A_codons_NP_072046 = [DNMT3A_exons_NP_072046[i:i+3] for i in range(0, len(DNMT3A_exons_NP_072046), 3)]\n",
    "\n",
    "    #DNMT3A transcript used:\n",
    "    #Downloaded from http://grch37.ensembl.org/Homo_sapiens/Transcript/Exons?db=core;g=ENSG00000119772;r=2:25455845-25564754;t=ENST00000321117\n",
    "    DNMT3A_exons_extra_bases_NP_072046 = 'GATGCCCGCCATGCCCTCCAGCGGCCCCGGGGACACCAGCAGCTCTGCTGCGGAGCGGGAGGAGGACCGAAAGGACGGAGAGGAGCAGGAGGAGCCGCGTGGCAAGGAGGAGCGCCAAGAGCCCAGCACCACGGCACGGAAGGTGGGGCGGCCTGGGAGGAAGCGCAAGCACCCCCCGGTGGAAAGCGGTGACACGCCAAAGGACCCTGCGGTGATCTCCAAGTCCCCATCCATGGCCCAGGACTCAGGCGCCTCAGAGCTATTACCCAATGGGGACTTGGAGAAGCGGAGTGAGCCCCAGCCAGAGGAGGGGAGCCCTGCTGGGGGGCAGAAGGGCGGGGCCCCAGCAGAGGGAGAGGGTGCAGCTGAGACCCTGCCTGAAGCCTCAAGAGCAGTGGAAAATGGCTGCTGCACCCCCAAGGAGGGCCGAGGAGCCCCTGCAGAAGCGGGCAAAGAACAGAAGGAGACCAACATCGAATCCATGAAAATGGAGGGCTCCCGGGGCCGGCTGCGGGGTGGCTTGGGCTGGGAGTCCAGCCTCCGTCAGCGGCCCATGCCGAGGCTCACCTTCCAGGCGGGGGACCCCTACTACATCAGCAAGCGCAAGCGGGACGAGTGGCTGGCACGCTGGAAAAGGGAGGCTGAGAAGAAAGCCAAGGTCATTGCAGGAATGAATGCTGTGGAAGAAAACCAGGGGCCCGGGGAGTCTCAGAAGGTGGAGGAGGCCAGCCCTCCTGCTGTGCAGCAGCCCACTGACCCCGCATCCCCCACTGTGGCTACCACGCCTGAGCCCGTGGGGTCCGATGCTGGGGACAAGAATGCCACCAAAGCAGGCGATGACGAGCCAGAGTACGAGGACGGCCGGGGCTTTGGCATTGGGGAGCTGGTGTGGGGGAAACTGCGGGGCTTCTCCTGGTGGCCAGGCCGCATTGTGTCTTGGTGGATGACGGGCCGGAGCCGAGCAGCTGAAGGCACCCGCTGGGTCATGTGGTTCGGAGACGGCAAATTCTCAGTGGTGTGTGTTGAGAAGCTGATGCCGCTGAGCTCGTTTTGCAGTGCGTTCCACCAGGCCACGTACAACAAGCAGCCCATGTACCGCAAAGCCATCTACGAGGTCCTGCAGGTGGCCAGCAGCCGCGCGGGGAAGCTGTTCCCGGTGTGCCACGACAGCGATGAGAGTGACACTGCCAAGGCCGTGGAGGTGCAGAACAAGCCCATGATTGAATGGGCCCTGGGGGGCTTCCAGCCTTCTGGCCCTAAGGGCCTGGAGCCACCAGAAGAAGAGAAGAATCCCTACAAAGAAGTGTACACGGACATGTGGGTGGAACCTGAGGCAGCTGCCTACGCACCACCTCCACCAGCCAAAAAGCCCCGGAAGAGCACAGCGGAGAAGCCCAAGGTCAAGGAGATTATTGATGAGCGCACAAGAGAGCGGCTGGTGTACGAGGTGCGGCAGAAGTGCCGGAACATTGAGGACATCTGCATCTCCTGTGGGAGCCTCAATGTTACCCTGGAACACCCCCTCTTCGTTGGAGGAATGTGCCAAAACTGCAAGAACTGCTTTCTGGAGTGTGCGTACCAGTACGACGACGACGGCTACCAGTCCTACTGCACCATCTGCTGTGGGGGCCGTGAGGTGCTCATGTGCGGAAACAACAACTGCTGCAGGTGCTTTTGCGTGGAGTGTGTGGACCTCTTGGTGGGGCCGGGGGCTGCCCAGGCAGCCATTAAGGAAGACCCCTGGAACTGCTACATGTGCGGGCACAAGGGTACCTACGGGCTGCTGCGGCGGCGAGAGGACTGGCCCTCCCGGCTCCAGATGTTCTTCGCTAATAACCACGACCAGGAATTTGACCCTCCAAAGGTTTACCCACCTGTCCCAGCTGAGAAGAGGAAGCCCATCCGGGTGCTGTCTCTCTTTGATGGAATCGCTACAGGGCTCCTGGTGCTGAAGGACTTGGGCATTCAGGTGGACCGCTACATTGCCTCGGAGGTGTGTGAGGACTCCATCACGGTGGGCATGGTGCGGCACCAGGGGAAGATCATGTACGTCGGGGACGTCCGCAGCGTCACACAGAAGCATATCCAGGAGTGGGGCCCATTCGATCTGGTGATTGGGGGCAGTCCCTGCAATGACCTCTCCATCGTCAACCCTGCTCGCAAGGGCCTCTACGAGGGCACTGGCCGGCTCTTCTTTGAGTTCTACCGCCTCCTGCATGATGCGCGGCCCAAGGAGGGAGATGATCGCCCCTTCTTCTGGCTCTTTGAGAATGTGGTGGCCATGGGCGTTAGTGACAAGAGGGACATCTCGCGATTTCTCGAGTCCAACCCTGTGATGATTGATGCCAAAGAAGTGTCAGCTGCACACAGGGCCCGCTACTTCTGGGGTAACCTTCCCGGTATGAACAGGCCGTTGGCATCCACTGTGAATGATAAGCTGGAGCTGCAGGAGTGTCTGGAGCATGGCAGGATAGCCAAGTTCAGCAAAGTGAGGACCATTACTACGAGGTCAAACTCCATAAAGCAGGGCAAAGACCAGCATTTTCCTGTCTTCATGAATGAGAAAGAGGACATCTTATGGTGCACTGAAATGGAAAGGGTATTTGGTTTCCCAGTCCACTATACTGACGTCTCCAACATGAGCCGCTTGGCGAGGCAGAGACTGCTGGGCCGGTCATGGAGCGTGCCAGTCATCCGCCACCTCTTCGCTCCGCTGAAGGAGTATTTTGCGTGTGTGTAAG'\n",
    "\n",
    "    #Create a list of the codons they called missense variants at\n",
    "    DNMT3A_variant_list_NP_072046 = []\n",
    "    DNMT3A_missense_codons_NP_072046 =[]\n",
    "    DNMT3A_missense_variant_DICT_NP_072046 = {}\n",
    "\n",
    "    variant_codon = int(((re.findall('\\d+', variant)))[0])\n",
    "\n",
    "    codon0_NP_072046 = variant_codon-1\n",
    "    codon1_NP_072046 = variant_codon\n",
    "    variant_NP_072046 = variant\n",
    "    DNMT3A_missense_codons_NP_072046.append(codon0_NP_072046) #List of codon numbers that they called variants at\n",
    "    DNMT3A_missense_variant_DICT_NP_072046[row_count]=(codon1_NP_072046, variant_NP_072046) #Dictionary:  K = codon number, V = variant at that codon (e.g. 76: p.V76L)\n",
    "\n",
    "    #Looking at what possible amino acid changes are possible at those codons\n",
    "    DNMT3A_missense_codons_NP_072046_DICT = {}\n",
    "    calls_DNMT3A_missense_NP_072046 = {}\n",
    "\n",
    "    for i in DNMT3A_missense_codons_NP_072046: #Iterates over the list of codon numbers\n",
    "        DNMT3A_missense_codons_NP_072046_DICT[i+1]= (DNMT3A_codons_NP_072046[i]) #Creates of dicitonary of key = codon number, value = bases (sequence) at that codon\n",
    "        #e.g. 12: CGT, 882: GCA\n",
    "\n",
    "    for k,v in DNMT3A_missense_codons_NP_072046_DICT.items(): #iterating over the list of codon numbers (and codon sequences)\n",
    "        if v in Possible_Amino_Acid_Changes_letters.keys(): #assign the list of all the possible amino acid changes to each codon they called\n",
    "            changes = Possible_Amino_Acid_Changes_letters[v]\n",
    "    #         print(Possible_Amino_Acid_Changes.values())\n",
    "            calls_DNMT3A_missense_NP_072046[k]=(v, changes)\n",
    "        #e.g. 12: (CGT, 'K', '*', 'T'etc...)\n",
    "\n",
    "    #Changing the original amino acid to a Word (e.g. Phe) rather than a codon\n",
    "    calls_DNMT3A_missense_NP_072046_extra = {}\n",
    "\n",
    "    for k, v in calls_DNMT3A_missense_NP_072046.items():\n",
    "        if v[0] in Genetic_Code:\n",
    "            Amino_acid_change = Genetic_Code.get(v[0])\n",
    "            calls_DNMT3A_missense_NP_072046_extra[k] = (Amino_acid_change, v[1])\n",
    "            #e.g. 12: (Phe, 'K', '*', 'T' etc..)\n",
    "\n",
    "    #Changing the original amino acid to a LETTER (rather than the codon)\n",
    "    calls_DNMT3A_missense_NP_072046_extra2 = {}\n",
    "\n",
    "    for k, v in calls_DNMT3A_missense_NP_072046_extra.items():\n",
    "        if v[0] in Amino_acid_abbreviations:\n",
    "            Amino_acid_abbrev = Amino_acid_abbreviations.get(v[0])\n",
    "            calls_DNMT3A_missense_NP_072046_extra2[k] = (Amino_acid_abbrev, v[1])\n",
    "            #e.g. 12: ((F), ('K', '*', 'T' etc..))\n",
    "\n",
    "    #Create a dictionary containing info on which position in the codon is affected to create the variant\n",
    "    DNMT3A_position_in_list_dict_missense_NP_072046={}\n",
    "    for k, v in DNMT3A_missense_variant_DICT_NP_072046.items(): #Iterates over the dictionary of (key = row_count, value = aa_pos, variant (e.g. 3: (76, p.V76L))\n",
    "        observed_change=(v[1])[-1] #Selects the amino acid change (e.g. the L from p.V76L)\n",
    "        possible_changes=calls_DNMT3A_missense_NP_072046[v[0]][1] #possible changes are the second value in the dictionary of (k = codon number, value = (original amino acid, possible changes))\n",
    "        possible_changes_info=calls_DNMT3A_missense_NP_072046[v[0]]\n",
    "        codon_seq=calls_DNMT3A_missense_NP_072046[v[0]][0]\n",
    "        possible_changes_string = \"\".join(possible_changes)\n",
    "        if v[1][2] == calls_DNMT3A_missense_NP_072046_extra2[v[0]][0]:  #v[1][2] = the original amino acid letter (i.e if this aa is in this transcript)\n",
    "            position_in_list = -1\n",
    "            positions_in_list = []\n",
    "            while True:\n",
    "                position_in_list = possible_changes_string.find(observed_change, position_in_list+1) #counts position in possible changes (if 0-2 = 1st base of codon, if 3-5 = 2nd base, if 6-8 = 3rd base)\n",
    "                if position_in_list == -1: break\n",
    "                positions_in_list.append(position_in_list)\n",
    "                DNMT3A_position_in_list_dict_missense_NP_072046[k]=(positions_in_list, possible_changes_info, codon_seq, v[0], v[1])\n",
    "\n",
    "    aa_positions_variants_called = []\n",
    "    for k, v in DNMT3A_position_in_list_dict_missense_NP_072046.items():\n",
    "        aa_positions_variants_called.append(v[3])\n",
    "    #     print('positions at which missense variants called = ', aa_positions_variants_called)\n",
    "\n",
    "    #Create a dictionary that contains the position in the list of possible AA where the variant is and what it changes to\n",
    "    DNMT3A_bases_missense_NP_072046 = {}\n",
    "    for k, v in DNMT3A_position_in_list_dict_missense_NP_072046.items():\n",
    "        codon_seq = v[2]\n",
    "        base_changes=[]\n",
    "        if codon_seq in Possible_Base_Changes.keys():\n",
    "            list_of_possible_bases = Possible_Base_Changes.get(codon_seq)\n",
    "        for i in v[0]:\n",
    "            base_change_position = i\n",
    "            base_change = list_of_possible_bases[base_change_position]\n",
    "            base_changes.append(base_change)\n",
    "        DNMT3A_bases_missense_NP_072046[k]=(v, base_changes)\n",
    "\n",
    "    #Create dictionary which contains original codon seq and details on which base is changed\n",
    "    DNMT3A_bases_missense2_NP_072046 = {}\n",
    "\n",
    "    for k, v in DNMT3A_bases_missense_NP_072046.items():\n",
    "        base_change_positions = v[0][0]\n",
    "        codon_positions = []\n",
    "        positions_in_gene = []\n",
    "        for i in base_change_positions:\n",
    "            base_change_position = i\n",
    "            if base_change_position <= 2:\n",
    "                codon_pos = 1\n",
    "            if 2 < base_change_position <= 5:\n",
    "                codon_pos = 2\n",
    "            if 5 < base_change_position <= 8:\n",
    "                codon_pos = 3\n",
    "            codon_positions.append(codon_pos)\n",
    "        DNMT3A_bases_missense2_NP_072046[k] = (v, codon_positions)\n",
    "\n",
    "    #Create dictionary which contains original codon seq, details on which base is changed and it's position in the coding region\n",
    "    DNMT3A_bases_missense3_NP_072046 = {}\n",
    "\n",
    "    for k, v in DNMT3A_bases_missense2_NP_072046.items():\n",
    "        codon_positions = v[1]\n",
    "        base_changes = v[0][1]\n",
    "        original_bases = []\n",
    "        positions_in_gene = []\n",
    "        codon_number = v[0][0][3]\n",
    "        variant = v[0][0][4]\n",
    "        for i in codon_positions:\n",
    "            codon_pos = i\n",
    "            position_in_gene = ((int(v[0][0][3])-1)*3)+codon_pos #codon number (v[0][0][3]) minus 1 (e.g. codon 307 -1 = 306, then times by 3 takes you to beginning of codon 307 (then add the position))  \n",
    "            positions_in_gene.append(position_in_gene)\n",
    "            original_base = (v[0][0][2][codon_pos-1])\n",
    "            original_bases.append(original_base)\n",
    "        DNMT3A_bases_missense3_NP_072046[k] = (v[0][0][1], positions_in_gene, original_bases, base_changes, codon_number, variant)\n",
    "\n",
    "    #Create dictionary which contains original codon seq, details on which base is changed and its neighbouring bases\n",
    "    DNMT3A_bases_missense4_NP_072046 = {}\n",
    "\n",
    "    for k, v in DNMT3A_bases_missense3_NP_072046.items():\n",
    "        position_in_gene = v[1]\n",
    "        sites_trio_missense=[]\n",
    "\n",
    "        for i in position_in_gene:\n",
    "    #         print(i)\n",
    "            if i in intron_exon_boundary_caution_base_DNMT3A_NP072046:\n",
    "                adapted_trio = intron_exon_check_DNMT3A_NP072046(i)\n",
    "                sites_trio_missense.append([adapted_trio])\n",
    "                DNMT3A_bases_missense4_NP_072046[k] = (v, sites_trio_missense)\n",
    "\n",
    "            else:\n",
    "                for i in position_in_gene:\n",
    "                    gene_pos = i\n",
    "                    site_trio_missense = [DNMT3A_exons_extra_bases_NP_072046[gene_pos:gene_pos+3] for gene_pos in range (gene_pos-1, gene_pos, 3)]\n",
    "                    sites_trio_missense.append(site_trio_missense)\n",
    "                    DNMT3A_bases_missense4_NP_072046[k] = (v, sites_trio_missense)\n",
    "\n",
    "    #Create a dictionary containing pos in coding region, original base, new base, base affected with its neighbours\n",
    "    #e.g. 938, G, A, TGG = base 938 on coding region, original base = G, changes to base A, base with neighbours either side = TGG\n",
    "    DNMT3A_sites_missense5_NP_072046 = {}\n",
    "\n",
    "    a=0\n",
    "    for k, v in DNMT3A_bases_missense4_NP_072046.items():\n",
    "        codon_n = v[0][1] #base number in codon (out of 3 x 912)\n",
    "        codon_n_912 = v[0][4]\n",
    "        original_base = v[0][2]\n",
    "        new_base = v[0][3]\n",
    "        site_trio = v[1]\n",
    "        info=[]\n",
    "        variant = v[0][5]\n",
    "\n",
    "        change = 0\n",
    "        for i in codon_n:\n",
    "            codon_number = i\n",
    "            orig_base = original_base[change]\n",
    "            n_base = new_base[change]\n",
    "            site_tri = site_trio[change]\n",
    "            info.append([codon_number, orig_base, n_base, site_tri, codon_n_912, variant])\n",
    "            change = change+1\n",
    "        DNMT3A_sites_missense5_NP_072046[a]=info\n",
    "        a=a+1\n",
    "\n",
    "    # Write in a way that can iterate over mutation rates (i.e. put in to form C[C>A]G etc)   \n",
    "    DNMT3A_sites_missense6_NP_072046={}\n",
    "\n",
    "    separating_positions = {} # separate out all the possible changes to just 1 effective item in the list\n",
    "    a=0\n",
    "    for k, v in DNMT3A_sites_missense5_NP_072046.items():\n",
    "        for i in v: #because some contain more than 1 change from that codon:\n",
    "            separating_positions[a]=i\n",
    "            a=a+1\n",
    "\n",
    "    for k, v in separating_positions.items():\n",
    "        base_pos = v[0]\n",
    "        original_base = v[1]\n",
    "        new_base = v[2]\n",
    "        site_trio = v[3]\n",
    "        codon_n_912 = v[4]\n",
    "        variant = v[5]\n",
    "        sequence = (original_base,new_base)\n",
    "        changes = '>'.join(sequence)\n",
    "        left_base = site_trio[0][0]\n",
    "        right_base = site_trio[0][2]\n",
    "        sequence2 = (changes, right_base)\n",
    "        changes2 = ']'.join(sequence2)\n",
    "        sequence3 = (left_base, changes2)\n",
    "        changes3 = '['.join(sequence3)\n",
    "        DNMT3A_sites_missense6_NP_072046[k] = (v[0],changes3, codon_n_912, variant) #site_trio = the affected base and its neighbours\n",
    "\n",
    "\n",
    "    #Assign the mutation rates to each site...\n",
    "    DNMT3A_mutation_rates_missense_NP_072046 = {}\n",
    "    for k, v in DNMT3A_sites_missense6_NP_072046.items():\n",
    "        base_change_info = v[1]\n",
    "        if base_change_info in Mutation_rates.keys():\n",
    "            mutation_rate = Mutation_rates.get(base_change_info)\n",
    "            DNMT3A_mutation_rates_missense_NP_072046[k] = (v, mutation_rate)\n",
    "\n",
    "    DNMT3A_mutation_rate_missense_NP_072046_sum = []\n",
    "    mutation_rates_list = []\n",
    "    for k, v in DNMT3A_mutation_rates_missense_NP_072046.items():\n",
    "        aa_pos = v[0][2]\n",
    "        site = v[0][1]\n",
    "        mutation_rate_value = v[1] \n",
    "        variant = v[0][3]\n",
    "        mutation_rates_list.append((aa_pos, site, mutation_rate_value, variant))\n",
    "        DNMT3A_mutation_rate_missense_NP_072046_sum.append(mutation_rate_value)\n",
    "\n",
    "    DNMT3A_Summed_mutation_rate_missense_NP_072046_added = np.sum(DNMT3A_mutation_rate_missense_NP_072046_sum)\n",
    "\n",
    "    mutation_rates_list_all = mutation_rates_list\n",
    "    return DNMT3A_Summed_mutation_rate_missense_NP_072046_added"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### TET2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 285,
   "metadata": {
    "code_folding": [
     0
    ]
   },
   "outputs": [],
   "source": [
    "def mutation_rates_variants_TET2(variant):\n",
    "    intron_exon_boundary_caution_base_TET2 = [3409, 3804, 4044, 4537]\n",
    "\n",
    "    #The mutation rate at a base is calculated based on its trinucleotide context.  To obtain the trinucleotide context, the base either side of the mutated base needs to be ascertained.\n",
    "    #The coding sequence (+ an extra base at each end) of each gene is used when determining the trinucleotide context of the base change.  The problem with using the coding sequence is that\n",
    "    #if a base change falls at the end of an exon, its neighbouring base will actuall be an intronic base, which may be different to the 'neighbouring' base in the coding sequence.\n",
    "    #To account for this, there is an \"intron_exon_check\" function that is called at the intron/exon boundary if the neighbouring base in the intron is different to the neighbouring base in the coding sequence.\n",
    "\n",
    "    def intron_exon_check_TET2(position_in_gene_number):\n",
    "        if position_in_gene_number == 3409:\n",
    "            gene_pos = 3409\n",
    "            first_base = TET2_exons_extra_bases_NP[gene_pos-1]\n",
    "            middle_base = TET2_exons_extra_bases_NP[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 3804:\n",
    "            gene_pos = 3804\n",
    "            first_base = 'G'\n",
    "            middle_base = TET2_exons_extra_bases_NP[gene_pos]\n",
    "            third_base = TET2_exons_extra_bases_NP[gene_pos+1]\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 4044:\n",
    "            gene_pos = 4044\n",
    "            first_base = TET2_exons_extra_bases_NP[gene_pos-1]\n",
    "            middle_base = TET2_exons_extra_bases_NP[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 4537:\n",
    "            gene_pos = 4537\n",
    "            first_base = TET2_exons_extra_bases_NP[gene_pos-1]\n",
    "            middle_base = TET2_exons_extra_bases_NP[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        return triplet\n",
    "    \n",
    "    #TET2 transcript (NP):\n",
    "    #Downloaded from http://grch37.ensembl.org/Homo_sapiens/Transcript/Exons?db=core;g=ENSG00000168769;r=4:106067032-106163928;t=ENST00000380013\n",
    "    TET2_exons_NP = 'ATGGAACAGGATAGAACCAACCATGTTGAGGGCAACAGACTAAGTCCATTCCTGATACCATCACCTCCCATTTGCCAGACAGAACCTCTGGCTACAAAGCTCCAGAATGGAAGCCCACTGCCTGAGAGAGCTCATCCAGAAGTAAATGGAGACACCAAGTGGCACTCTTTCAAAAGTTATTATGGAATACCCTGTATGAAGGGAAGCCAGAATAGTCGTGTGAGTCCTGACTTTACACAAGAAAGTAGAGGGTATTCCAAGTGTTTGCAAAATGGAGGAATAAAACGCACAGTTAGTGAACCTTCTCTCTCTGGGCTCCTTCAGATCAAGAAATTGAAACAAGACCAAAAGGCTAATGGAGAAAGACGTAACTTCGGGGTAAGCCAAGAAAGAAATCCAGGTGAAAGCAGTCAACCAAATGTCTCCGATTTGAGTGATAAGAAAGAATCTGTGAGTTCTGTAGCCCAAGAAAATGCAGTTAAAGATTTCACCAGTTTTTCAACACATAACTGCAGTGGGCCTGAAAATCCAGAGCTTCAGATTCTGAATGAGCAGGAGGGGAAAAGTGCTAATTACCATGACAAGAACATTGTATTACTTAAAAACAAGGCAGTGCTAATGCCTAATGGTGCTACAGTTTCTGCCTCTTCCGTGGAACACACACATGGTGAACTCCTGGAAAAAACACTGTCTCAATATTATCCAGATTGTGTTTCCATTGCGGTGCAGAAAACCACATCTCACATAAATGCCATTAACAGTCAGGCTACTAATGAGTTGTCCTGTGAGATCACTCACCCATCGCATACCTCAGGGCAGATCAATTCCGCACAGACCTCTAACTCTGAGCTGCCTCCAAAGCCAGCTGCAGTGGTGAGTGAGGCCTGTGATGCTGATGATGCTGATAATGCCAGTAAACTAGCTGCAATGCTAAATACCTGTTCCTTTCAGAAACCAGAACAACTACAACAACAAAAATCAGTTTTTGAGATATGCCCATCTCCTGCAGAAAATAACATCCAGGGAACCACAAAGCTAGCGTCTGGTGAAGAATTCTGTTCAGGTTCCAGCAGCAATTTGCAAGCTCCTGGTGGCAGCTCTGAACGGTATTTAAAACAAAATGAAATGAATGGTGCTTACTTCAAGCAAAGCTCAGTGTTCACTAAGGATTCCTTTTCTGCCACTACCACACCACCACCACCATCACAATTGCTTCTTTCTCCCCCTCCTCCTCTTCCACAGGTTCCTCAGCTTCCTTCAGAAGGAAAAAGCACTCTGAATGGTGGAGTTTTAGAAGAACACCACCACTACCCCAACCAAAGTAACACAACACTTTTAAGGGAAGTGAAAATAGAGGGTAAACCTGAGGCACCACCTTCCCAGAGTCCTAATCCATCTACACATGTATGCAGCCCTTCTCCGATGCTTTCTGAAAGGCCTCAGAATAATTGTGTGAACAGGAATGACATACAGACTGCAGGGACAATGACTGTTCCATTGTGTTCTGAGAAAACAAGACCAATGTCAGAACACCTCAAGCATAACCCACCAATTTTTGGTAGCAGTGGAGAGCTACAGGACAACTGCCAGCAGTTGATGAGAAACAAAGAGCAAGAGATTCTGAAGGGTCGAGACAAGGAGCAAACACGAGATCTTGTGCCCCCAACACAGCACTATCTGAAACCAGGATGGATTGAATTGAAGGCCCCTCGTTTTCACCAAGCGGAATCCCATCTAAAACGTAATGAGGCATCACTGCCATCAATTCTTCAGTATCAACCCAATCTCTCCAATCAAATGACCTCCAAACAATACACTGGAAATTCCAACATGCCTGGGGGGCTCCCAAGGCAAGCTTACACCCAGAAAACAACACAGCTGGAGCACAAGTCACAAATGTACCAAGTTGAAATGAATCAAGGGCAGTCCCAAGGTACAGTGGACCAACATCTCCAGTTCCAAAAACCCTCACACCAGGTGCACTTCTCCAAAACAGACCATTTACCAAAAGCTCATGTGCAGTCACTGTGTGGCACTAGATTTCATTTTCAACAAAGAGCAGATTCCCAAACTGAAAAACTTATGTCCCCAGTGTTGAAACAGCACTTGAATCAACAGGCTTCAGAGACTGAGCCATTTTCAAACTCACACCTTTTGCAACATAAGCCTCATAAACAGGCAGCACAAACACAACCATCCCAGAGTTCACATCTCCCTCAAAACCAGCAACAGCAGCAAAAATTACAAATAAAGAATAAAGAGGAAATACTCCAGACTTTTCCTCACCCCCAAAGCAACAATGATCAGCAAAGAGAAGGATCATTCTTTGGCCAGACTAAAGTGGAAGAATGTTTTCATGGTGAAAATCAGTATTCAAAATCAAGCGAGTTCGAGACTCATAATGTCCAAATGGGACTGGAGGAAGTACAGAATATAAATCGTAGAAATTCCCCTTATAGTCAGACCATGAAATCAAGTGCATGCAAAATACAGGTTTCTTGTTCAAACAATACACACCTAGTTTCAGAGAATAAAGAACAGACTACACATCCTGAACTTTTTGCAGGAAACAAGACCCAAAACTTGCATCACATGCAATATTTTCCAAATAATGTGATCCCAAAGCAAGATCTTCTTCACAGGTGCTTTCAAGAACAGGAGCAGAAGTCACAACAAGCTTCAGTTCTACAGGGATATAAAAATAGAAACCAAGATATGTCTGGTCAACAAGCTGCGCAACTTGCTCAGCAAAGGTACTTGATACATAACCATGCAAATGTTTTTCCTGTGCCTGACCAGGGAGGAAGTCACACTCAGACCCCTCCCCAGAAGGACACTCAAAAGCATGCTGCTCTAAGGTGGCATCTCTTACAGAAGCAAGAACAGCAGCAAACACAGCAACCCCAAACTGAGTCTTGCCATAGTCAGATGCACAGGCCAATTAAGGTGGAACCTGGATGCAAGCCACATGCCTGTATGCACACAGCACCACCAGAAAACAAAACATGGAAAAAGGTAACTAAGCAAGAGAATCCACCTGCAAGCTGTGATAATGTGCAGCAAAAGAGCATCATTGAGACCATGGAGCAGCATCTGAAGCAGTTTCACGCCAAGTCGTTATTTGACCATAAGGCTCTTACTCTCAAATCACAGAAGCAAGTAAAAGTTGAAATGTCAGGGCCAGTCACAGTTTTGACTAGACAAACCACTGCTGCAGAACTTGATAGCCACACCCCAGCTTTAGAGCAGCAAACAACTTCTTCAGAAAAGACACCAACCAAAAGAACAGCTGCTTCTGTTCTCAATAATTTTATAGAGTCACCTTCCAAATTACTAGATACTCCTATAAAAAATTTATTGGATACACCTGTCAAGACTCAATATGATTTCCCATCTTGCAGATGTGTAGAGCAAATTATTGAAAAAGATGAAGGTCCTTTTTATACCCATCTAGGAGCAGGTCCTAATGTGGCAGCTATTAGAGAAATCATGGAAGAAAGGTTTGGACAGAAGGGTAAAGCTATTAGGATTGAAAGAGTCATCTATACTGGTAAAGAAGGCAAAAGTTCTCAGGGATGTCCTATTGCTAAGTGGGTGGTTCGCAGAAGCAGCAGTGAAGAGAAGCTACTGTGTTTGGTGCGGGAGCGAGCTGGCCACACCTGTGAGGCTGCAGTGATTGTGATTCTCATCCTGGTGTGGGAAGGAATCCCGCTGTCTCTGGCTGACAAACTCTACTCGGAGCTTACCGAGACGCTGAGGAAATACGGCACGCTCACCAATCGCCGGTGTGCCTTGAATGAAGAGAGAACTTGCGCCTGTCAGGGGCTGGATCCAGAAACCTGTGGTGCCTCCTTCTCTTTTGGTTGTTCATGGAGCATGTACTACAATGGATGTAAGTTTGCCAGAAGCAAGATCCCAAGGAAGTTTAAGCTGCTTGGGGATGACCCAAAAGAGGAAGAGAAACTGGAGTCTCATTTGCAAAACCTGTCCACTCTTATGGCACCAACATATAAGAAACTTGCACCTGATGCATATAATAATCAGATTGAATATGAACACAGAGCACCAGAGTGCCGTCTGGGTCTGAAGGAAGGCCGTCCATTCTCAGGGGTCACTGCATGTTTGGACTTCTGTGCTCATGCCCACAGAGACTTGCACAACATGCAGAATGGCAGCACATTGGTATGCACTCTCACTAGAGAAGACAATCGAGAATTTGGAGGAAAACCTGAGGATGAGCAGCTTCACGTTCTGCCTTTATACAAAGTCTCTGACGTGGATGAGTTTGGGAGTGTGGAAGCTCAGGAGGAGAAAAAACGGAGTGGTGCCATTCAGGTACTGAGTTCTTTTCGGCGAAAAGTCAGGATGTTAGCAGAGCCAGTCAAGACTTGCCGACAAAGGAAACTAGAAGCCAAGAAAGCTGCAGCTGAAAAGCTTTCCTCCCTGGAGAACAGCTCAAATAAAAATGAAAAGGAAAAGTCAGCCCCATCACGTACAAAACAAACTGAAAACGCAAGCCAGGCTAAACAGTTGGCAGAACTTTTGCGACTTTCAGGACCAGTCATGCAGCAGTCCCAGCAGCCCCAGCCTCTACAGAAGCAGCCACCACAGCCCCAGCAGCAGCAGAGACCCCAGCAGCAGCAGCCACATCACCCTCAGACAGAGTCTGTCAACTCTTATTCTGCTTCTGGATCCACCAATCCATACATGAGACGGCCCAATCCAGTTAGTCCTTATCCAAACTCTTCACACACTTCAGATATCTATGGAAGCACCAGCCCTATGAACTTCTATTCCACCTCATCTCAAGCTGCAGGTTCATATTTGAATTCTTCTAATCCCATGAACCCTTACCCTGGGCTTTTGAATCAGAATACCCAATATCCATCATATCAATGCAATGGAAACCTATCAGTGGACAACTGCTCCCCATATCTGGGTTCCTATTCTCCCCAGTCTCAGCCGATGGATCTGTATAGGTATCCAAGCCAAGACCCTCTGTCTAAGCTCAGTCTACCACCCATCCATACACTTTACCAGCCAAGGTTTGGAAATAGCCAGAGTTTTACATCTAAATACTTAGGTTATGGAAACCAAAATATGCAGGGAGATGGTTTCAGCAGTTGTACCATTAGACCAAATGTACATCATGTAGGGAAATTGCCTCCTTATCCCACTCATGAGATGGATGGCCACTTCATGGGAGCCACCTCTAGATTACCACCCAATCTGAGCAATCCAAACATGGACTATAAAAATGGTGAACATCATTCACCTTCTCACATAATCCATAACTACAGTGCAGCTCCGGGCATGTTCAACAGCTCTCTTCATGCCCTGCATCTCCAAAACAAGGAGAATGACATGCTTTCCCACACAGCTAATGGGTTATCAAAGATGCTTCCAGCTCTTAACCATGATAGAACTGCTTGTGTCCAAGGAGGCTTACACAAATTAAGTGATGCTAATGGTCAGGAAAAGCAGCCATTGGCACTAGTCCAGGGTGTGGCTTCTGGTGCAGAGGACAACGATGAGGTCTGGTCAGACAGCGAGCAGAGCTTTCTGGATCCTGACATTGGGGGAGTGGCCGTGGCTCCAACTCATGGGTCAATTCTCATTGAGTGTGCAAAGCGTGAGCTGCATGCCACAACCCCTTTAAAGAATCCCAATAGGAATCACCCCACCAGGATCTCCCTCGTCTTTTACCAGCATAAGAGCATGAATGAGCCAAAACATGGCTTGGCTCTTTGGGAAGCCAAAATGGCTGAAAAAGCCCGTGAGAAAGAGGAAGAGTGTGAAAAGTATGGCCCAGACTATGTGCCTCAGAAATCCCATGGCAAAAAAGTGAAACGGGAGCCTGCTGAGCCACATGAAACTTCAGAGCCCACTTACCTGCGTTTCATCAAGTCTCTTGCCGAAAGGACCATGTCCGTGACCACAGACTCCACAGTAACTACATCTCCATATGCCTTCACTCGGGTCACAGGGCCTTACAACAGATATATATGA'\n",
    "\n",
    "    #Split the sequence into codons \n",
    "    TET2_codons_NP = [TET2_exons_NP[i:i+3] for i in range(0, len(TET2_exons_NP), 3)]\n",
    "\n",
    "    #TET2 transcript (transcript variant 2 (TET2-001), NM_017628, NP_001120680):\n",
    "    #Downloaded from http://grch37.ensembl.org/Homo_sapiens/Transcript/Exons?db=core;g=ENSG00000168769;r=4:106067032-106163928;t=ENST00000380013\n",
    "    TET2_exons_extra_bases_NP = 'GATGGAACAGGATAGAACCAACCATGTTGAGGGCAACAGACTAAGTCCATTCCTGATACCATCACCTCCCATTTGCCAGACAGAACCTCTGGCTACAAAGCTCCAGAATGGAAGCCCACTGCCTGAGAGAGCTCATCCAGAAGTAAATGGAGACACCAAGTGGCACTCTTTCAAAAGTTATTATGGAATACCCTGTATGAAGGGAAGCCAGAATAGTCGTGTGAGTCCTGACTTTACACAAGAAAGTAGAGGGTATTCCAAGTGTTTGCAAAATGGAGGAATAAAACGCACAGTTAGTGAACCTTCTCTCTCTGGGCTCCTTCAGATCAAGAAATTGAAACAAGACCAAAAGGCTAATGGAGAAAGACGTAACTTCGGGGTAAGCCAAGAAAGAAATCCAGGTGAAAGCAGTCAACCAAATGTCTCCGATTTGAGTGATAAGAAAGAATCTGTGAGTTCTGTAGCCCAAGAAAATGCAGTTAAAGATTTCACCAGTTTTTCAACACATAACTGCAGTGGGCCTGAAAATCCAGAGCTTCAGATTCTGAATGAGCAGGAGGGGAAAAGTGCTAATTACCATGACAAGAACATTGTATTACTTAAAAACAAGGCAGTGCTAATGCCTAATGGTGCTACAGTTTCTGCCTCTTCCGTGGAACACACACATGGTGAACTCCTGGAAAAAACACTGTCTCAATATTATCCAGATTGTGTTTCCATTGCGGTGCAGAAAACCACATCTCACATAAATGCCATTAACAGTCAGGCTACTAATGAGTTGTCCTGTGAGATCACTCACCCATCGCATACCTCAGGGCAGATCAATTCCGCACAGACCTCTAACTCTGAGCTGCCTCCAAAGCCAGCTGCAGTGGTGAGTGAGGCCTGTGATGCTGATGATGCTGATAATGCCAGTAAACTAGCTGCAATGCTAAATACCTGTTCCTTTCAGAAACCAGAACAACTACAACAACAAAAATCAGTTTTTGAGATATGCCCATCTCCTGCAGAAAATAACATCCAGGGAACCACAAAGCTAGCGTCTGGTGAAGAATTCTGTTCAGGTTCCAGCAGCAATTTGCAAGCTCCTGGTGGCAGCTCTGAACGGTATTTAAAACAAAATGAAATGAATGGTGCTTACTTCAAGCAAAGCTCAGTGTTCACTAAGGATTCCTTTTCTGCCACTACCACACCACCACCACCATCACAATTGCTTCTTTCTCCCCCTCCTCCTCTTCCACAGGTTCCTCAGCTTCCTTCAGAAGGAAAAAGCACTCTGAATGGTGGAGTTTTAGAAGAACACCACCACTACCCCAACCAAAGTAACACAACACTTTTAAGGGAAGTGAAAATAGAGGGTAAACCTGAGGCACCACCTTCCCAGAGTCCTAATCCATCTACACATGTATGCAGCCCTTCTCCGATGCTTTCTGAAAGGCCTCAGAATAATTGTGTGAACAGGAATGACATACAGACTGCAGGGACAATGACTGTTCCATTGTGTTCTGAGAAAACAAGACCAATGTCAGAACACCTCAAGCATAACCCACCAATTTTTGGTAGCAGTGGAGAGCTACAGGACAACTGCCAGCAGTTGATGAGAAACAAAGAGCAAGAGATTCTGAAGGGTCGAGACAAGGAGCAAACACGAGATCTTGTGCCCCCAACACAGCACTATCTGAAACCAGGATGGATTGAATTGAAGGCCCCTCGTTTTCACCAAGCGGAATCCCATCTAAAACGTAATGAGGCATCACTGCCATCAATTCTTCAGTATCAACCCAATCTCTCCAATCAAATGACCTCCAAACAATACACTGGAAATTCCAACATGCCTGGGGGGCTCCCAAGGCAAGCTTACACCCAGAAAACAACACAGCTGGAGCACAAGTCACAAATGTACCAAGTTGAAATGAATCAAGGGCAGTCCCAAGGTACAGTGGACCAACATCTCCAGTTCCAAAAACCCTCACACCAGGTGCACTTCTCCAAAACAGACCATTTACCAAAAGCTCATGTGCAGTCACTGTGTGGCACTAGATTTCATTTTCAACAAAGAGCAGATTCCCAAACTGAAAAACTTATGTCCCCAGTGTTGAAACAGCACTTGAATCAACAGGCTTCAGAGACTGAGCCATTTTCAAACTCACACCTTTTGCAACATAAGCCTCATAAACAGGCAGCACAAACACAACCATCCCAGAGTTCACATCTCCCTCAAAACCAGCAACAGCAGCAAAAATTACAAATAAAGAATAAAGAGGAAATACTCCAGACTTTTCCTCACCCCCAAAGCAACAATGATCAGCAAAGAGAAGGATCATTCTTTGGCCAGACTAAAGTGGAAGAATGTTTTCATGGTGAAAATCAGTATTCAAAATCAAGCGAGTTCGAGACTCATAATGTCCAAATGGGACTGGAGGAAGTACAGAATATAAATCGTAGAAATTCCCCTTATAGTCAGACCATGAAATCAAGTGCATGCAAAATACAGGTTTCTTGTTCAAACAATACACACCTAGTTTCAGAGAATAAAGAACAGACTACACATCCTGAACTTTTTGCAGGAAACAAGACCCAAAACTTGCATCACATGCAATATTTTCCAAATAATGTGATCCCAAAGCAAGATCTTCTTCACAGGTGCTTTCAAGAACAGGAGCAGAAGTCACAACAAGCTTCAGTTCTACAGGGATATAAAAATAGAAACCAAGATATGTCTGGTCAACAAGCTGCGCAACTTGCTCAGCAAAGGTACTTGATACATAACCATGCAAATGTTTTTCCTGTGCCTGACCAGGGAGGAAGTCACACTCAGACCCCTCCCCAGAAGGACACTCAAAAGCATGCTGCTCTAAGGTGGCATCTCTTACAGAAGCAAGAACAGCAGCAAACACAGCAACCCCAAACTGAGTCTTGCCATAGTCAGATGCACAGGCCAATTAAGGTGGAACCTGGATGCAAGCCACATGCCTGTATGCACACAGCACCACCAGAAAACAAAACATGGAAAAAGGTAACTAAGCAAGAGAATCCACCTGCAAGCTGTGATAATGTGCAGCAAAAGAGCATCATTGAGACCATGGAGCAGCATCTGAAGCAGTTTCACGCCAAGTCGTTATTTGACCATAAGGCTCTTACTCTCAAATCACAGAAGCAAGTAAAAGTTGAAATGTCAGGGCCAGTCACAGTTTTGACTAGACAAACCACTGCTGCAGAACTTGATAGCCACACCCCAGCTTTAGAGCAGCAAACAACTTCTTCAGAAAAGACACCAACCAAAAGAACAGCTGCTTCTGTTCTCAATAATTTTATAGAGTCACCTTCCAAATTACTAGATACTCCTATAAAAAATTTATTGGATACACCTGTCAAGACTCAATATGATTTCCCATCTTGCAGATGTGTAGAGCAAATTATTGAAAAAGATGAAGGTCCTTTTTATACCCATCTAGGAGCAGGTCCTAATGTGGCAGCTATTAGAGAAATCATGGAAGAAAGGTTTGGACAGAAGGGTAAAGCTATTAGGATTGAAAGAGTCATCTATACTGGTAAAGAAGGCAAAAGTTCTCAGGGATGTCCTATTGCTAAGTGGGTGGTTCGCAGAAGCAGCAGTGAAGAGAAGCTACTGTGTTTGGTGCGGGAGCGAGCTGGCCACACCTGTGAGGCTGCAGTGATTGTGATTCTCATCCTGGTGTGGGAAGGAATCCCGCTGTCTCTGGCTGACAAACTCTACTCGGAGCTTACCGAGACGCTGAGGAAATACGGCACGCTCACCAATCGCCGGTGTGCCTTGAATGAAGAGAGAACTTGCGCCTGTCAGGGGCTGGATCCAGAAACCTGTGGTGCCTCCTTCTCTTTTGGTTGTTCATGGAGCATGTACTACAATGGATGTAAGTTTGCCAGAAGCAAGATCCCAAGGAAGTTTAAGCTGCTTGGGGATGACCCAAAAGAGGAAGAGAAACTGGAGTCTCATTTGCAAAACCTGTCCACTCTTATGGCACCAACATATAAGAAACTTGCACCTGATGCATATAATAATCAGATTGAATATGAACACAGAGCACCAGAGTGCCGTCTGGGTCTGAAGGAAGGCCGTCCATTCTCAGGGGTCACTGCATGTTTGGACTTCTGTGCTCATGCCCACAGAGACTTGCACAACATGCAGAATGGCAGCACATTGGTATGCACTCTCACTAGAGAAGACAATCGAGAATTTGGAGGAAAACCTGAGGATGAGCAGCTTCACGTTCTGCCTTTATACAAAGTCTCTGACGTGGATGAGTTTGGGAGTGTGGAAGCTCAGGAGGAGAAAAAACGGAGTGGTGCCATTCAGGTACTGAGTTCTTTTCGGCGAAAAGTCAGGATGTTAGCAGAGCCAGTCAAGACTTGCCGACAAAGGAAACTAGAAGCCAAGAAAGCTGCAGCTGAAAAGCTTTCCTCCCTGGAGAACAGCTCAAATAAAAATGAAAAGGAAAAGTCAGCCCCATCACGTACAAAACAAACTGAAAACGCAAGCCAGGCTAAACAGTTGGCAGAACTTTTGCGACTTTCAGGACCAGTCATGCAGCAGTCCCAGCAGCCCCAGCCTCTACAGAAGCAGCCACCACAGCCCCAGCAGCAGCAGAGACCCCAGCAGCAGCAGCCACATCACCCTCAGACAGAGTCTGTCAACTCTTATTCTGCTTCTGGATCCACCAATCCATACATGAGACGGCCCAATCCAGTTAGTCCTTATCCAAACTCTTCACACACTTCAGATATCTATGGAAGCACCAGCCCTATGAACTTCTATTCCACCTCATCTCAAGCTGCAGGTTCATATTTGAATTCTTCTAATCCCATGAACCCTTACCCTGGGCTTTTGAATCAGAATACCCAATATCCATCATATCAATGCAATGGAAACCTATCAGTGGACAACTGCTCCCCATATCTGGGTTCCTATTCTCCCCAGTCTCAGCCGATGGATCTGTATAGGTATCCAAGCCAAGACCCTCTGTCTAAGCTCAGTCTACCACCCATCCATACACTTTACCAGCCAAGGTTTGGAAATAGCCAGAGTTTTACATCTAAATACTTAGGTTATGGAAACCAAAATATGCAGGGAGATGGTTTCAGCAGTTGTACCATTAGACCAAATGTACATCATGTAGGGAAATTGCCTCCTTATCCCACTCATGAGATGGATGGCCACTTCATGGGAGCCACCTCTAGATTACCACCCAATCTGAGCAATCCAAACATGGACTATAAAAATGGTGAACATCATTCACCTTCTCACATAATCCATAACTACAGTGCAGCTCCGGGCATGTTCAACAGCTCTCTTCATGCCCTGCATCTCCAAAACAAGGAGAATGACATGCTTTCCCACACAGCTAATGGGTTATCAAAGATGCTTCCAGCTCTTAACCATGATAGAACTGCTTGTGTCCAAGGAGGCTTACACAAATTAAGTGATGCTAATGGTCAGGAAAAGCAGCCATTGGCACTAGTCCAGGGTGTGGCTTCTGGTGCAGAGGACAACGATGAGGTCTGGTCAGACAGCGAGCAGAGCTTTCTGGATCCTGACATTGGGGGAGTGGCCGTGGCTCCAACTCATGGGTCAATTCTCATTGAGTGTGCAAAGCGTGAGCTGCATGCCACAACCCCTTTAAAGAATCCCAATAGGAATCACCCCACCAGGATCTCCCTCGTCTTTTACCAGCATAAGAGCATGAATGAGCCAAAACATGGCTTGGCTCTTTGGGAAGCCAAAATGGCTGAAAAAGCCCGTGAGAAAGAGGAAGAGTGTGAAAAGTATGGCCCAGACTATGTGCCTCAGAAATCCCATGGCAAAAAAGTGAAACGGGAGCCTGCTGAGCCACATGAAACTTCAGAGCCCACTTACCTGCGTTTCATCAAGTCTCTTGCCGAAAGGACCATGTCCGTGACCACAGACTCCACAGTAACTACATCTCCATATGCCTTCACTCGGGTCACAGGGCCTTACAACAGATATATATGAT'\n",
    "\n",
    "    #Create a list of the codons they called missense variants at\n",
    "    TET2_variant_list_NP = []\n",
    "    TET2_missense_codons_NP =[]\n",
    "    TET2_missense_variant_DICT_NP = {}\n",
    "\n",
    "    variant_codon = int(((re.findall('\\d+', variant)))[0])\n",
    "\n",
    "    codon0_NP = variant_codon-1\n",
    "    codon1_NP = variant_codon\n",
    "    variant_NP = variant\n",
    "    TET2_missense_codons_NP.append(codon0_NP) #List of codon numbers that they called variants at\n",
    "    TET2_missense_variant_DICT_NP[row_count]=(codon1_NP, variant_NP) #Dictionary:  K = codon number, V = variant at that codon (e.g. 76: p.V76L)\n",
    "\n",
    "    #Looking at what possible amino acid changes are possible at those codons\n",
    "    TET2_missense_codons_NP_DICT = {}\n",
    "    calls_TET2_missense_NP = {}\n",
    "\n",
    "    for i in TET2_missense_codons_NP: #Iterates over the list of codon numbers\n",
    "        TET2_missense_codons_NP_DICT[i+1]= (TET2_codons_NP[i]) #Creates of dicitonary of key = codon number, value = bases (sequence) at that codon\n",
    "        #e.g. 12: CGT, 882: GCA\n",
    "\n",
    "    for k,v in TET2_missense_codons_NP_DICT.items(): #iterating over the list of codon numbers (and codon sequences)\n",
    "        if v in Possible_Amino_Acid_Changes_letters.keys(): #assign the list of all the possible amino acid changes to each codon they called\n",
    "            changes = Possible_Amino_Acid_Changes_letters[v]\n",
    "    #         print(Possible_Amino_Acid_Changes.values())\n",
    "            calls_TET2_missense_NP[k]=(v, changes)\n",
    "        #e.g. 12: (CGT, 'K', '*', 'T'etc...)\n",
    "\n",
    "    #Changing the original amino acid to a Word (e.g. Phe) rather than a codon\n",
    "    calls_TET2_missense_NP_extra = {}\n",
    "\n",
    "    for k, v in calls_TET2_missense_NP.items():\n",
    "        if v[0] in Genetic_Code:\n",
    "            Amino_acid_change = Genetic_Code.get(v[0])\n",
    "            calls_TET2_missense_NP_extra[k] = (Amino_acid_change, v[1])\n",
    "            #e.g. 12: (Phe, 'K', '*', 'T' etc..)\n",
    "\n",
    "    #Changing the original amino acid to a LETTER (rather than the codon)\n",
    "    calls_TET2_missense_NP_extra2 = {}\n",
    "\n",
    "    for k, v in calls_TET2_missense_NP_extra.items():\n",
    "        if v[0] in Amino_acid_abbreviations:\n",
    "            Amino_acid_abbrev = Amino_acid_abbreviations.get(v[0])\n",
    "            calls_TET2_missense_NP_extra2[k] = (Amino_acid_abbrev, v[1])\n",
    "            #e.g. 12: ((F), ('K', '*', 'T' etc..))\n",
    "\n",
    "    #Create a dictionary containing info on which position in the codon is affected to create the variant\n",
    "    TET2_position_in_list_dict_missense_NP={}\n",
    "    for k, v in TET2_missense_variant_DICT_NP.items(): #Iterates over the dictionary of (key = row_count, value = aa_pos, variant (e.g. 3: (76, p.V76L))\n",
    "        observed_change=(v[1])[-1] #Selects the amino acid change (e.g. the L from p.V76L)\n",
    "        possible_changes=calls_TET2_missense_NP[v[0]][1] #possible changes are the second value in the dictionary of (k = codon number, value = (original amino acid, possible changes))\n",
    "        possible_changes_info=calls_TET2_missense_NP[v[0]]\n",
    "        codon_seq=calls_TET2_missense_NP[v[0]][0]\n",
    "        possible_changes_string = \"\".join(possible_changes)\n",
    "        if v[1][2] == calls_TET2_missense_NP_extra2[v[0]][0]:  #v[1][2] = the original amino acid letter (i.e if this aa is in this transcript)\n",
    "            position_in_list = -1\n",
    "            positions_in_list = []\n",
    "            while True:\n",
    "                position_in_list = possible_changes_string.find(observed_change, position_in_list+1) #counts position in possible changes (if 0-2 = 1st base of codon, if 3-5 = 2nd base, if 6-8 = 3rd base)\n",
    "                if position_in_list == -1: break\n",
    "                positions_in_list.append(position_in_list)\n",
    "                TET2_position_in_list_dict_missense_NP[k]=(positions_in_list, possible_changes_info, codon_seq, v[0], v[1])\n",
    "\n",
    "    aa_positions_variants_called = []\n",
    "    for k, v in TET2_position_in_list_dict_missense_NP.items():\n",
    "        aa_positions_variants_called.append(v[3])\n",
    "    #     print('positions at which missense variants called = ', aa_positions_variants_called)\n",
    "\n",
    "    #Create a dictionary that contains the position in the list of possible AA where the variant is and what it changes to\n",
    "    TET2_bases_missense_NP = {}\n",
    "    for k, v in TET2_position_in_list_dict_missense_NP.items():\n",
    "        codon_seq = v[2]\n",
    "        base_changes=[]\n",
    "        if codon_seq in Possible_Base_Changes.keys():\n",
    "            list_of_possible_bases = Possible_Base_Changes.get(codon_seq)\n",
    "        for i in v[0]:\n",
    "            base_change_position = i\n",
    "            base_change = list_of_possible_bases[base_change_position]\n",
    "            base_changes.append(base_change)\n",
    "        TET2_bases_missense_NP[k]=(v, base_changes)\n",
    "\n",
    "    #Create dictionary which contains original codon seq and details on which base is changed\n",
    "    TET2_bases_missense2_NP = {}\n",
    "\n",
    "    for k, v in TET2_bases_missense_NP.items():\n",
    "        base_change_positions = v[0][0]\n",
    "        codon_positions = []\n",
    "        positions_in_gene = []\n",
    "        for i in base_change_positions:\n",
    "            base_change_position = i\n",
    "            if base_change_position <= 2:\n",
    "                codon_pos = 1\n",
    "            if 2 < base_change_position <= 5:\n",
    "                codon_pos = 2\n",
    "            if 5 < base_change_position <= 8:\n",
    "                codon_pos = 3\n",
    "            codon_positions.append(codon_pos)\n",
    "        TET2_bases_missense2_NP[k] = (v, codon_positions)\n",
    "\n",
    "    #Create dictionary which contains original codon seq, details on which base is changed and it's position in the coding region\n",
    "    TET2_bases_missense3_NP = {}\n",
    "\n",
    "    for k, v in TET2_bases_missense2_NP.items():\n",
    "        codon_positions = v[1]\n",
    "        base_changes = v[0][1]\n",
    "        original_bases = []\n",
    "        positions_in_gene = []\n",
    "        codon_number = v[0][0][3]\n",
    "        variant = v[0][0][4]\n",
    "        for i in codon_positions:\n",
    "            codon_pos = i\n",
    "            position_in_gene = ((int(v[0][0][3])-1)*3)+codon_pos #codon number (v[0][0][3]) minus 1 (e.g. codon 307 -1 = 306, then times by 3 takes you to beginning of codon 307 (then add the position))  \n",
    "            positions_in_gene.append(position_in_gene)\n",
    "            original_base = (v[0][0][2][codon_pos-1])\n",
    "            original_bases.append(original_base)\n",
    "        TET2_bases_missense3_NP[k] = (v[0][0][1], positions_in_gene, original_bases, base_changes, codon_number, variant)\n",
    "\n",
    "    #Create dictionary which contains original codon seq, details on which base is changed and its neighbouring bases\n",
    "    TET2_bases_missense4_NP = {}\n",
    "\n",
    "    for k, v in TET2_bases_missense3_NP.items():\n",
    "        position_in_gene = v[1]\n",
    "        sites_trio_missense=[]\n",
    "\n",
    "        for i in position_in_gene:\n",
    "    #         print(i)\n",
    "            if i in intron_exon_boundary_caution_base_TET2:\n",
    "    #                 print('positions at intron_exon_boundary =',i)\n",
    "                adapted_trio = intron_exon_check_TET2(i)\n",
    "                sites_trio_missense.append([adapted_trio])\n",
    "                TET2_bases_missense4_NP[k] = (v, sites_trio_missense)\n",
    "\n",
    "            else:\n",
    "                for i in position_in_gene:\n",
    "                    gene_pos = i\n",
    "                    site_trio_missense = [TET2_exons_extra_bases_NP[gene_pos:gene_pos+3] for gene_pos in range (gene_pos-1, gene_pos, 3)]\n",
    "                    sites_trio_missense.append(site_trio_missense)\n",
    "                    TET2_bases_missense4_NP[k] = (v, sites_trio_missense)\n",
    "\n",
    "    #Create a dictionary containing pos in coding region, original base, new base, base affected with its neighbours\n",
    "    #e.g. 938, G, A, TGG = base 938 on coding region, original base = G, changes to base A, base with neighbours either side = TGG\n",
    "    TET2_sites_missense5_NP = {}\n",
    "\n",
    "    a=0\n",
    "    for k, v in TET2_bases_missense4_NP.items():\n",
    "        codon_n = v[0][1] #base number in codon (out of 3 x 912)\n",
    "        codon_n_912 = v[0][4]\n",
    "        original_base = v[0][2]\n",
    "        new_base = v[0][3]\n",
    "        site_trio = v[1]\n",
    "        info=[]\n",
    "        variant = v[0][5]\n",
    "\n",
    "        change = 0\n",
    "        for i in codon_n:\n",
    "            codon_number = i\n",
    "            orig_base = original_base[change]\n",
    "            n_base = new_base[change]\n",
    "            site_tri = site_trio[change]\n",
    "            info.append([codon_number, orig_base, n_base, site_tri, codon_n_912, variant])\n",
    "            change = change+1\n",
    "        TET2_sites_missense5_NP[a]=info\n",
    "        a=a+1\n",
    "\n",
    "    # Write in a way that can iterate over mutation rates (i.e. put in to form C[C>A]G etc)   \n",
    "    TET2_sites_missense6_NP={}\n",
    "\n",
    "    separating_positions = {} # separate out all the possible changes to just 1 effective item in the list\n",
    "    a=0\n",
    "    for k, v in TET2_sites_missense5_NP.items():\n",
    "        for i in v: #because some contain more than 1 change from that codon:\n",
    "            separating_positions[a]=i\n",
    "            a=a+1\n",
    "\n",
    "    for k, v in separating_positions.items():\n",
    "        base_pos = v[0]\n",
    "        original_base = v[1]\n",
    "        new_base = v[2]\n",
    "        site_trio = v[3]\n",
    "        codon_n_912 = v[4]\n",
    "        variant = v[5]\n",
    "        sequence = (original_base,new_base)\n",
    "        changes = '>'.join(sequence)\n",
    "        left_base = site_trio[0][0]\n",
    "        right_base = site_trio[0][2]\n",
    "        sequence2 = (changes, right_base)\n",
    "        changes2 = ']'.join(sequence2)\n",
    "        sequence3 = (left_base, changes2)\n",
    "        changes3 = '['.join(sequence3)\n",
    "        TET2_sites_missense6_NP[k] = (v[0],changes3, codon_n_912, variant) #site_trio = the affected base and its neighbours\n",
    "\n",
    "\n",
    "    #Assign the mutation rates to each site...\n",
    "    TET2_mutation_rates_missense_NP = {}\n",
    "    for k, v in TET2_sites_missense6_NP.items():\n",
    "        base_change_info = v[1]\n",
    "        if base_change_info in Mutation_rates.keys():\n",
    "            mutation_rate = Mutation_rates.get(base_change_info)\n",
    "            TET2_mutation_rates_missense_NP[k] = (v, mutation_rate)\n",
    "\n",
    "    TET2_mutation_rate_missense_NP_sum = []\n",
    "    mutation_rates_list = []\n",
    "    for k, v in TET2_mutation_rates_missense_NP.items():\n",
    "    #     print(v[1])\n",
    "        aa_pos = v[0][2]\n",
    "        site = v[0][1]\n",
    "        mutation_rate_value = v[1] \n",
    "        variant = v[0][3]\n",
    "        mutation_rates_list.append((aa_pos, site, mutation_rate_value, variant))\n",
    "        TET2_mutation_rate_missense_NP_sum.append(mutation_rate_value)\n",
    "\n",
    "    TET2_Summed_mutation_rate_missense_NP_added = np.sum(TET2_mutation_rate_missense_NP_sum)\n",
    "\n",
    "    mutation_rates_list_all = mutation_rates_list\n",
    "    return TET2_Summed_mutation_rate_missense_NP_added"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### ASXL1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 286,
   "metadata": {
    "code_folding": [
     0
    ]
   },
   "outputs": [],
   "source": [
    "def mutation_rates_variants_ASXL1(variant):\n",
    "    intron_exon_boundary_caution_base_ASXL1 = [140, 143, 252, 373, 1086, 1085]\n",
    "   \n",
    "    #The mutation rate at a base is calculated based on its trinucleotide context.  To obtain the trinucleotide context, the base either side of the mutated base needs to be ascertained.\n",
    "    #The coding sequence (+ an extra base at each end) of each gene is used when determining the trinucleotide context of the base change.  The problem with using the coding sequence is that\n",
    "    #if a base change falls at the end of an exon, its neighbouring base will actuall be an intronic base, which may be different to the 'neighbouring' base in the coding sequence.\n",
    "    #To account for this, there is an \"intron_exon_check\" function that is called at the intron/exon boundary if the neighbouring base in the intron is different to the neighbouring base in the coding sequence.\n",
    "\n",
    "    def intron_exon_check(position_in_gene_number):\n",
    "        if position_in_gene_number == 140:\n",
    "            gene_pos = 140\n",
    "            first_base = ASXL1_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = ASXL1_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 143:\n",
    "            gene_pos = 143\n",
    "            first_base = ASXL1_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = ASXL1_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 252:\n",
    "            gene_pos = 252\n",
    "            first_base = ASXL1_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = ASXL1_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 373:\n",
    "            gene_pos = 373\n",
    "            first_base = ASXL1_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = ASXL1_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 1086:\n",
    "            gene_pos = 1086\n",
    "            first_base = 'G'\n",
    "            middle_base = ASXL1_exons_extra_bases[gene_pos]\n",
    "            third_base = ASXL1_exons_extra_bases[gene_pos+1]\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 1085:\n",
    "            gene_pos = 1085\n",
    "            first_base = ASXL1_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = ASXL1_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        return triplet\n",
    "    \n",
    "    #ASXL1 transcript:\n",
    "    #Downloaded from http://grch37.ensembl.org\n",
    "    ASXL1_exons = 'ATGAAGGACAAACAGAAGAAGAAGAAGGAGCGCACGTGGGCCGAGGCCGCGCGCCTGGTATTAGAAAACTACTCGGATGCTCCAATGACACCAAAACAGATTCTGCAGGTCATAGAGGCAGAAGGACTAAAGGAAATGAGAAGTGGGACTTCCCCTCTCGCATGCCTCAATGCTATGCTACATTCCAATTCAAGAGGAGGAGAGGGGTTGTTTTATAAACTGCCTGGCCGAATCAGCCTTTTCACGCTCAAGAAGGATGCCCTGCAGTGGTCTCGCCATCCAGCTACAGTGGAGGGAGAGGAGCCAGAGGACACGGCTGATGTGGAGAGCTGTGGGTCTAATGAAGCCAGCACTGTGAGTGGTGAAAACGATGTATCTCTTGATGAAACATCTTCGAACGCATCCTGTTCTACAGAATCTCAGAGTCGACCTCTTTCCAATCCCAGGGACAGCTACAGAGCTTCCTCACAGGCGAACAAACAAAAGAAAAAGACTGGGGTGATGCTGCCTCGAGTTGTCCTGACTCCTCTGAAGGTAAACGGGGCCCACGTGGAATCTGCATCAGGGTTCTCGGGCTGCCACGCCGATGGCGAGAGCGGCAGCCCGTCCAGCAGCAGCAGCGGCTCTCTGGCCCTGGGCAGCGCTGCTATTCGTGGCCAGGCCGAGGTCACCCAGGACCCTGCCCCGCTCCTGAGAGGCTTCCGGAAGCCAGCCACAGGTCAAATGAAGCGCAACAGAGGGGAAGAAATAGATTTTGAGACACCTGGGTCCATTCTTGTCAACACCAACCTCCGTGCCCTGATCAACTCTCGGACCTTCCATGCCTTACCATCACACTTCCAGCAGCAGCTCCTCTTCCTCCTGCCTGAAGTAGACAGACAGGTGGGGACGGATGGCCTGTTGCGTCTCAGCAGCAGTGCACTAAATAACGAGTTTTTTACCCATGCGGCTCAGAGCTGGCGGGAGCGCCTGGCTGATGGTGAATTTACTCATGAGATGCAAGTCAGGATACGACAGGAAATGGAGAAGGAAAAGAAGGTGGAACAATGGAAAGAAAAGTTCTTTGAAGACTACTATGGACAGAAGCTGGGTTTGACCAAAGAAGAGTCATTGCAGCAGAACGTGGGCCAGGAGGAGGCTGAAATCAAAAGTGGCTTGTGTGTCCCAGGAGAATCAGTGCGTATACAGCGTGGTCCAGCCACCCGACAGCGAGATGGGCATTTTAAGAAACGCTCTCGGCCAGATCTCCGAACCAGAGCCAGAAGGAATCTGTACAAAAAACAGGAGTCAGAACAAGCAGGGGTTGCTAAGGATGCAAAATCTGTGGCCTCAGATGTTCCCCTCTACAAGGATGGGGAGGCTAAGACTGACCCAGCAGGGCTGAGCAGTCCCCATCTGCCAGGCACATCCTCTGCAGCACCCGACCTGGAGGGTCCCGAATTCCCAGTTGAGTCTGTGGCTTCTCGGATCCAGGCTGAGCCAGACAACTTGGCACGTGCCTCTGCATCTCCAGACAGAATTCCTAGCCTGCCTCAGGAAACTGTGGATCAGGAACCCAAGGATCAGAAGAGGAAATCCTTTGAGCAGGCGGCCTCTGCATCCTTTCCCGAAAAGAAGCCCCGGCTTGAAGATCGTCAGTCCTTTCGTAACACAATTGAAAGTGTTCACACCGAAAAGCCACAGCCCACTAAAGAGGAGCCCAAAGTCCCGCCCATCCGGATTCAACTTTCACGTATCAAACCACCCTGGGTGGTTAAAGGTCAGCCCACTTACCAGATATGCCCCCGGATCATCCCCACCACGGAGTCCTCCTGCCGGGGTTGGACTGGCGCCAGGACCCTCGCAGACATTAAAGCCCGTGCTCTGCAGGTCCGAGGGGCGAGAGGTCACCACTGCCATAGAGAGGCGGCCACCACTGCCATCGGAGGGGGGGGTGGCCCGGGTGGAGGTGGCGGCGGGGCCACCGATGAGGGAGGTGGCAGAGGCAGCAGCAGTGGTGATGGTGGTGAGGCCTGTGGCCACCCTGAGCCCAGGGGAGGCCCGAGCACCCCTGGAAAGTGTACGTCAGATCTACAGCGAACACAACTACTGCCGCCTTATCCTCTAAATGGGGAGCATACCCAGGCCGGAACTGCCATGTCCAGAGCTAGGAGAGAGGACCTGCCTTCTCTGAGAAAGGAGGAAAGCTGCCTACTACAGAGGGCTACAGTTGGACTCACAGATGGGCTAGGAGATGCCTCCCAACTCCCCGTTGCTCCCACTGGGGACCAGCCATGCCAGGCCTTGCCCCTACTGTCCTCCCAAACCTCAGTAGCTGAGAGATTAGTGGAGCAGCCTCAGTTGCATCCGGATGTTAGAACTGAATGTGAGTCTGGCACCACTTCCTGGGAAAGTGATGATGAGGAGCAAGGACCCACCGTTCCTGCAGACAATGGTCCCATTCTGTCTCTAGTGGGAGATGATACATTAGAGAAAGGAACTGGCCAAGCTCTTGACAGTCATCCCACTATGAAGGATCCTGTAAATGTGACCCCCAGTTCCACACCTGAATCCTCACCGACTGATTGCCTGCAGAACAGAGCATTTGATGACGAATTAGGGCTTGGTGGCTCATGCCCTCCTATGAGGGAAAGTGATACTAGACAAGAAAACTTGAAAACCAAGGCTCTCGTTTCTAACAGTTCTTTGCATTGGATACCCATCCCATCGAATGATGAGGTAGTGAAACAGCCCAAACCAGAATCCAGAGAACACATACCATCTGTTGAGCCCCAGGTTGGAGAGGAGTGGGAGAAAGCTGCTCCCACCCCTCCTGCATTGCCTGGGGATTTGACAGCTGAGGAGGGTCTAGATCCTCTTGACAGCCTTACTTCACTCTGGACTGTGCCATCTCGAGGAGGCAGTGACAGCAATGGCAGTTACTGTCAACAGGTGGACATTGAAAAGCTGAAAATCAACGGAGACTCTGAAGCACTGAGTCCTCACGGTGAGTCCACGGATACAGCCTCTGACTTTGAAGGTCACCTCACGGAGGACAGCAGTGAGGCTGACACTAGAGAAGCTGCAGTGACAAAGGGATCTTCGGTGGACAAGGATGAGAAACCCAATTGGAACCAATCTGCCCCACTGTCCAAGGTGAATGGTGACATGCGTCTGGTTACAAGGACAGATGGGATGGTTGCTCCTCAGAGCTGGGTGTCTCGAGTATGTGCGGTCCGCCAAAAGATCCCAGATTCCCTACTGCTGGCCAGTACTGAGTACCAGCCAAGAGCCGTGTGCCTGTCCATGCCTGGGTCCTCAGTGGAGGCCACTAACCCACTTGTGATGCAGTTGCTGCAGGGTAGCTTGCCCCTAGAGAAGGTTCTTCCACCAGCCCACGATGACAGCATGTCAGAATCCCCACAAGTACCACTTACAAAAGACCAGAGCCATGGCTCGCTACGCATGGGATCTTTACATGGTCTTGGAAAAAACAGTGGCATGGTTGATGGAAGCAGCCCCAGTTCTTTAAGGGCTTTGAAGGAGCCTCTTCTGCCAGATAGCTGTGAAACAGGCACTGGTCTTGCCAGGATTGAGGCCACCCAGGCTCCTGGAGCACCCCAAAAGAATTGCAAGGCAGTCCCAAGTTTTGACTCCCTCCATCCAGTGACAAATCCCATTACATCCTCTAGGAAACTGGAAGAAATGGATTCCAAAGAGCAGTTCTCTTCCTTTAGTTGTGAAGATCAGAAGGAAGTCCGTGCTATGTCACAGGACAGTAATTCAAATGCTGCTCCAGGAAAGAGCCCAGGAGATCTTACTACCTCGAGAACACCTCGTTTCTCATCTCCAAATGTGATCTCCTTTGGTCCAGAGCAGACAGGTCGGGCCCTGGGTGATCAGAGCAATGTTACAGGCCAAGGGAAGAAGCTTTTTGGCTCTGGGAATGTGGCTGCAACCCTTCAGCGCCCCAGGCCTGCGGACCCGATGCCTCTTCCTGCTGAGATCCCTCCAGTTTTTCCCAGTGGGAAGTTGGGACCAAGCACAAACTCCATGTCTGGTGGGGTACAGACTCCAAGGGAAGACTGGGCTCCAAAGCCACATGCCTTTGTTGGCAGCGTCAAGAATGAGAAGACTTTTGTGGGGGGTCCTCTTAAGGCAAATGCCGAGAACAGGAAAGCTACTGGGCATAGTCCCCTGGAACTGGTGGGTCACTTGGAAGGGATGCCCTTTGTCATGGACTTGCCCTTCTGGAAATTACCCCGAGAGCCAGGGAAGGGGCTCAGTGAGCCTCTGGAGCCTTCTTCTCTCCCCTCCCAACTCAGCATCAAGCAGGCATTTTATGGGAAGCTTTCTAAACTCCAACTGAGTTCCACCAGCTTTAATTATTCCTCTAGCTCTCCCACCTTTCCCAAAGGCCTTGCTGGAAGTGTGGTGCAGCTGAGCCACAAAGCAAACTTTGGTGCGAGCCACAGTGCATCACTTTCCTTGCAAATGTTCACTGACAGCAGCACGGTGGAAAGCATCTCGCTCCAGTGTGCGTGCAGCCTGAAAGCCATGATCATGTGCCAAGGCTGCGGTGCGTTCTGTCACGATGACTGTATTGGACCCTCAAAGCTCTGTGTATTGTGCCTTGTGGTGAGATAA'\n",
    "    \n",
    "    #Split the sequence into codons \n",
    "    ASXL1_codons = [ASXL1_exons[i:i+3] for i in range(0, len(ASXL1_exons), 3)]\n",
    "\n",
    "    #ASXL1 transcript:\n",
    "    #Downloaded from http://grch37.ensembl.org\n",
    "    ASXL1_exons_extra_bases = 'GATGAAGGACAAACAGAAGAAGAAGAAGGAGCGCACGTGGGCCGAGGCCGCGCGCCTGGTATTAGAAAACTACTCGGATGCTCCAATGACACCAAAACAGATTCTGCAGGTCATAGAGGCAGAAGGACTAAAGGAAATGAGAAGTGGGACTTCCCCTCTCGCATGCCTCAATGCTATGCTACATTCCAATTCAAGAGGAGGAGAGGGGTTGTTTTATAAACTGCCTGGCCGAATCAGCCTTTTCACGCTCAAGAAGGATGCCCTGCAGTGGTCTCGCCATCCAGCTACAGTGGAGGGAGAGGAGCCAGAGGACACGGCTGATGTGGAGAGCTGTGGGTCTAATGAAGCCAGCACTGTGAGTGGTGAAAACGATGTATCTCTTGATGAAACATCTTCGAACGCATCCTGTTCTACAGAATCTCAGAGTCGACCTCTTTCCAATCCCAGGGACAGCTACAGAGCTTCCTCACAGGCGAACAAACAAAAGAAAAAGACTGGGGTGATGCTGCCTCGAGTTGTCCTGACTCCTCTGAAGGTAAACGGGGCCCACGTGGAATCTGCATCAGGGTTCTCGGGCTGCCACGCCGATGGCGAGAGCGGCAGCCCGTCCAGCAGCAGCAGCGGCTCTCTGGCCCTGGGCAGCGCTGCTATTCGTGGCCAGGCCGAGGTCACCCAGGACCCTGCCCCGCTCCTGAGAGGCTTCCGGAAGCCAGCCACAGGTCAAATGAAGCGCAACAGAGGGGAAGAAATAGATTTTGAGACACCTGGGTCCATTCTTGTCAACACCAACCTCCGTGCCCTGATCAACTCTCGGACCTTCCATGCCTTACCATCACACTTCCAGCAGCAGCTCCTCTTCCTCCTGCCTGAAGTAGACAGACAGGTGGGGACGGATGGCCTGTTGCGTCTCAGCAGCAGTGCACTAAATAACGAGTTTTTTACCCATGCGGCTCAGAGCTGGCGGGAGCGCCTGGCTGATGGTGAATTTACTCATGAGATGCAAGTCAGGATACGACAGGAAATGGAGAAGGAAAAGAAGGTGGAACAATGGAAAGAAAAGTTCTTTGAAGACTACTATGGACAGAAGCTGGGTTTGACCAAAGAAGAGTCATTGCAGCAGAACGTGGGCCAGGAGGAGGCTGAAATCAAAAGTGGCTTGTGTGTCCCAGGAGAATCAGTGCGTATACAGCGTGGTCCAGCCACCCGACAGCGAGATGGGCATTTTAAGAAACGCTCTCGGCCAGATCTCCGAACCAGAGCCAGAAGGAATCTGTACAAAAAACAGGAGTCAGAACAAGCAGGGGTTGCTAAGGATGCAAAATCTGTGGCCTCAGATGTTCCCCTCTACAAGGATGGGGAGGCTAAGACTGACCCAGCAGGGCTGAGCAGTCCCCATCTGCCAGGCACATCCTCTGCAGCACCCGACCTGGAGGGTCCCGAATTCCCAGTTGAGTCTGTGGCTTCTCGGATCCAGGCTGAGCCAGACAACTTGGCACGTGCCTCTGCATCTCCAGACAGAATTCCTAGCCTGCCTCAGGAAACTGTGGATCAGGAACCCAAGGATCAGAAGAGGAAATCCTTTGAGCAGGCGGCCTCTGCATCCTTTCCCGAAAAGAAGCCCCGGCTTGAAGATCGTCAGTCCTTTCGTAACACAATTGAAAGTGTTCACACCGAAAAGCCACAGCCCACTAAAGAGGAGCCCAAAGTCCCGCCCATCCGGATTCAACTTTCACGTATCAAACCACCCTGGGTGGTTAAAGGTCAGCCCACTTACCAGATATGCCCCCGGATCATCCCCACCACGGAGTCCTCCTGCCGGGGTTGGACTGGCGCCAGGACCCTCGCAGACATTAAAGCCCGTGCTCTGCAGGTCCGAGGGGCGAGAGGTCACCACTGCCATAGAGAGGCGGCCACCACTGCCATCGGAGGGGGGGGTGGCCCGGGTGGAGGTGGCGGCGGGGCCACCGATGAGGGAGGTGGCAGAGGCAGCAGCAGTGGTGATGGTGGTGAGGCCTGTGGCCACCCTGAGCCCAGGGGAGGCCCGAGCACCCCTGGAAAGTGTACGTCAGATCTACAGCGAACACAACTACTGCCGCCTTATCCTCTAAATGGGGAGCATACCCAGGCCGGAACTGCCATGTCCAGAGCTAGGAGAGAGGACCTGCCTTCTCTGAGAAAGGAGGAAAGCTGCCTACTACAGAGGGCTACAGTTGGACTCACAGATGGGCTAGGAGATGCCTCCCAACTCCCCGTTGCTCCCACTGGGGACCAGCCATGCCAGGCCTTGCCCCTACTGTCCTCCCAAACCTCAGTAGCTGAGAGATTAGTGGAGCAGCCTCAGTTGCATCCGGATGTTAGAACTGAATGTGAGTCTGGCACCACTTCCTGGGAAAGTGATGATGAGGAGCAAGGACCCACCGTTCCTGCAGACAATGGTCCCATTCTGTCTCTAGTGGGAGATGATACATTAGAGAAAGGAACTGGCCAAGCTCTTGACAGTCATCCCACTATGAAGGATCCTGTAAATGTGACCCCCAGTTCCACACCTGAATCCTCACCGACTGATTGCCTGCAGAACAGAGCATTTGATGACGAATTAGGGCTTGGTGGCTCATGCCCTCCTATGAGGGAAAGTGATACTAGACAAGAAAACTTGAAAACCAAGGCTCTCGTTTCTAACAGTTCTTTGCATTGGATACCCATCCCATCGAATGATGAGGTAGTGAAACAGCCCAAACCAGAATCCAGAGAACACATACCATCTGTTGAGCCCCAGGTTGGAGAGGAGTGGGAGAAAGCTGCTCCCACCCCTCCTGCATTGCCTGGGGATTTGACAGCTGAGGAGGGTCTAGATCCTCTTGACAGCCTTACTTCACTCTGGACTGTGCCATCTCGAGGAGGCAGTGACAGCAATGGCAGTTACTGTCAACAGGTGGACATTGAAAAGCTGAAAATCAACGGAGACTCTGAAGCACTGAGTCCTCACGGTGAGTCCACGGATACAGCCTCTGACTTTGAAGGTCACCTCACGGAGGACAGCAGTGAGGCTGACACTAGAGAAGCTGCAGTGACAAAGGGATCTTCGGTGGACAAGGATGAGAAACCCAATTGGAACCAATCTGCCCCACTGTCCAAGGTGAATGGTGACATGCGTCTGGTTACAAGGACAGATGGGATGGTTGCTCCTCAGAGCTGGGTGTCTCGAGTATGTGCGGTCCGCCAAAAGATCCCAGATTCCCTACTGCTGGCCAGTACTGAGTACCAGCCAAGAGCCGTGTGCCTGTCCATGCCTGGGTCCTCAGTGGAGGCCACTAACCCACTTGTGATGCAGTTGCTGCAGGGTAGCTTGCCCCTAGAGAAGGTTCTTCCACCAGCCCACGATGACAGCATGTCAGAATCCCCACAAGTACCACTTACAAAAGACCAGAGCCATGGCTCGCTACGCATGGGATCTTTACATGGTCTTGGAAAAAACAGTGGCATGGTTGATGGAAGCAGCCCCAGTTCTTTAAGGGCTTTGAAGGAGCCTCTTCTGCCAGATAGCTGTGAAACAGGCACTGGTCTTGCCAGGATTGAGGCCACCCAGGCTCCTGGAGCACCCCAAAAGAATTGCAAGGCAGTCCCAAGTTTTGACTCCCTCCATCCAGTGACAAATCCCATTACATCCTCTAGGAAACTGGAAGAAATGGATTCCAAAGAGCAGTTCTCTTCCTTTAGTTGTGAAGATCAGAAGGAAGTCCGTGCTATGTCACAGGACAGTAATTCAAATGCTGCTCCAGGAAAGAGCCCAGGAGATCTTACTACCTCGAGAACACCTCGTTTCTCATCTCCAAATGTGATCTCCTTTGGTCCAGAGCAGACAGGTCGGGCCCTGGGTGATCAGAGCAATGTTACAGGCCAAGGGAAGAAGCTTTTTGGCTCTGGGAATGTGGCTGCAACCCTTCAGCGCCCCAGGCCTGCGGACCCGATGCCTCTTCCTGCTGAGATCCCTCCAGTTTTTCCCAGTGGGAAGTTGGGACCAAGCACAAACTCCATGTCTGGTGGGGTACAGACTCCAAGGGAAGACTGGGCTCCAAAGCCACATGCCTTTGTTGGCAGCGTCAAGAATGAGAAGACTTTTGTGGGGGGTCCTCTTAAGGCAAATGCCGAGAACAGGAAAGCTACTGGGCATAGTCCCCTGGAACTGGTGGGTCACTTGGAAGGGATGCCCTTTGTCATGGACTTGCCCTTCTGGAAATTACCCCGAGAGCCAGGGAAGGGGCTCAGTGAGCCTCTGGAGCCTTCTTCTCTCCCCTCCCAACTCAGCATCAAGCAGGCATTTTATGGGAAGCTTTCTAAACTCCAACTGAGTTCCACCAGCTTTAATTATTCCTCTAGCTCTCCCACCTTTCCCAAAGGCCTTGCTGGAAGTGTGGTGCAGCTGAGCCACAAAGCAAACTTTGGTGCGAGCCACAGTGCATCACTTTCCTTGCAAATGTTCACTGACAGCAGCACGGTGGAAAGCATCTCGCTCCAGTGTGCGTGCAGCCTGAAAGCCATGATCATGTGCCAAGGCTGCGGTGCGTTCTGTCACGATGACTGTATTGGACCCTCAAAGCTCTGTGTATTGTGCCTTGTGGTGAGATAAT'\n",
    "  \n",
    "    #Create a list of the codons they called missense variants at\n",
    "    ASXL1_variant_list = []\n",
    "    ASXL1_missense_codons =[]\n",
    "    ASXL1_missense_variant_DICT = {}\n",
    "\n",
    "    variant_codon = int(((re.findall('\\d+', variant)))[0])\n",
    "\n",
    "    codon0 = variant_codon-1\n",
    "    codon1 = variant_codon\n",
    "    variant = variant\n",
    "    ASXL1_missense_codons.append(codon0) #List of codon numbers that they called variants at\n",
    "    ASXL1_missense_variant_DICT[row_count]=(codon1, variant) #Dictionary:  K = codon number, V = variant at that codon (e.g. 76: p.V76L)\n",
    "\n",
    "    #Looking at what possible amino acid changes are possible at those codons\n",
    "    ASXL1_missense_codons_DICT = {}\n",
    "    calls_ASXL1_missense = {}\n",
    "\n",
    "    for i in ASXL1_missense_codons: #Iterates over the list of codon numbers\n",
    "        ASXL1_missense_codons_DICT[i+1]= (ASXL1_codons[i]) #Creates of dicitonary of key = codon number, value = bases (sequence) at that codon\n",
    "        #e.g. 12: CGT, 882: GCA\n",
    "\n",
    "    for k,v in ASXL1_missense_codons_DICT.items(): #iterating over the list of codon numbers (and codon sequences)\n",
    "        if v in Possible_Amino_Acid_Changes_letters.keys(): #assign the list of all the possible amino acid changes to each codon they called\n",
    "            changes = Possible_Amino_Acid_Changes_letters[v]\n",
    "    #         print(Possible_Amino_Acid_Changes.values())\n",
    "            calls_ASXL1_missense[k]=(v, changes)\n",
    "        #e.g. 12: (CGT, 'K', '*', 'T'etc...)\n",
    "\n",
    "    #Changing the original amino acid to a Word (e.g. Phe) rather than a codon\n",
    "    calls_ASXL1_missense_extra = {}\n",
    "\n",
    "    for k, v in calls_ASXL1_missense.items():\n",
    "        if v[0] in Genetic_Code:\n",
    "            Amino_acid_change = Genetic_Code.get(v[0])\n",
    "            calls_ASXL1_missense_extra[k] = (Amino_acid_change, v[1])\n",
    "            #e.g. 12: (Phe, 'K', '*', 'T' etc..)\n",
    "\n",
    "    #Changing the original amino acid to a LETTER (rather than the codon)\n",
    "    calls_ASXL1_missense_extra2 = {}\n",
    "\n",
    "    for k, v in calls_ASXL1_missense_extra.items():\n",
    "        if v[0] in Amino_acid_abbreviations:\n",
    "            Amino_acid_abbrev = Amino_acid_abbreviations.get(v[0])\n",
    "            calls_ASXL1_missense_extra2[k] = (Amino_acid_abbrev, v[1])\n",
    "            #e.g. 12: ((F), ('K', '*', 'T' etc..))\n",
    "\n",
    "    #Create a dictionary containing info on which position in the codon is affected to create the variant\n",
    "    ASXL1_position_in_list_dict_missense={}\n",
    "    for k, v in ASXL1_missense_variant_DICT.items(): #Iterates over the dictionary of (key = row_count, value = aa_pos, variant (e.g. 3: (76, p.V76L))\n",
    "        observed_change=(v[1])[-1] #Selects the amino acid change (e.g. the L from p.V76L)\n",
    "        possible_changes=calls_ASXL1_missense[v[0]][1] #possible changes are the second value in the dictionary of (k = codon number, value = (original amino acid, possible changes))\n",
    "        possible_changes_info=calls_ASXL1_missense[v[0]]\n",
    "        codon_seq=calls_ASXL1_missense[v[0]][0]\n",
    "        possible_changes_string = \"\".join(possible_changes)\n",
    "        if v[1][2] == calls_ASXL1_missense_extra2[v[0]][0]:  #v[1][2] = the original amino acid letter (i.e if this aa is in this transcript)\n",
    "            position_in_list = -1\n",
    "            positions_in_list = []\n",
    "            while True:\n",
    "                position_in_list = possible_changes_string.find(observed_change, position_in_list+1) #counts position in possible changes (if 0-2 = 1st base of codon, if 3-5 = 2nd base, if 6-8 = 3rd base)\n",
    "                if position_in_list == -1: break\n",
    "                positions_in_list.append(position_in_list)\n",
    "                ASXL1_position_in_list_dict_missense[k]=(positions_in_list, possible_changes_info, codon_seq, v[0], v[1])\n",
    "\n",
    "    aa_positions_variants_called = []\n",
    "    for k, v in ASXL1_position_in_list_dict_missense.items():\n",
    "        aa_positions_variants_called.append(v[3])\n",
    "    #     print('positions at which missense variants called = ', aa_positions_variants_called)\n",
    "\n",
    "    #Create a dictionary that contains the position in the list of possible AA where the variant is and what it changes to\n",
    "    ASXL1_bases_missense = {}\n",
    "    for k, v in ASXL1_position_in_list_dict_missense.items():\n",
    "        codon_seq = v[2]\n",
    "        base_changes=[]\n",
    "        if codon_seq in Possible_Base_Changes.keys():\n",
    "            list_of_possible_bases = Possible_Base_Changes.get(codon_seq)\n",
    "        for i in v[0]:\n",
    "            base_change_position = i\n",
    "            base_change = list_of_possible_bases[base_change_position]\n",
    "            base_changes.append(base_change)\n",
    "        ASXL1_bases_missense[k]=(v, base_changes)\n",
    "\n",
    "    #Create dictionary which contains original codon seq and details on which base is changed\n",
    "    ASXL1_bases_missense2 = {}\n",
    "\n",
    "    for k, v in ASXL1_bases_missense.items():\n",
    "        base_change_positions = v[0][0]\n",
    "        codon_positions = []\n",
    "        positions_in_gene = []\n",
    "        for i in base_change_positions:\n",
    "            base_change_position = i\n",
    "            if base_change_position <= 2:\n",
    "                codon_pos = 1\n",
    "            if 2 < base_change_position <= 5:\n",
    "                codon_pos = 2\n",
    "            if 5 < base_change_position <= 8:\n",
    "                codon_pos = 3\n",
    "            codon_positions.append(codon_pos)\n",
    "        ASXL1_bases_missense2[k] = (v, codon_positions)\n",
    "\n",
    "    #Create dictionary which contains original codon seq, details on which base is changed and it's position in the coding region\n",
    "    ASXL1_bases_missense3 = {}\n",
    "\n",
    "    for k, v in ASXL1_bases_missense2.items():\n",
    "        codon_positions = v[1]\n",
    "        base_changes = v[0][1]\n",
    "        original_bases = []\n",
    "        positions_in_gene = []\n",
    "        codon_number = v[0][0][3]\n",
    "        variant = v[0][0][4]\n",
    "        for i in codon_positions:\n",
    "            codon_pos = i\n",
    "            position_in_gene = ((int(v[0][0][3])-1)*3)+codon_pos #codon number (v[0][0][3]) minus 1 (e.g. codon 307 -1 = 306, then times by 3 takes you to beginning of codon 307 (then add the position))  \n",
    "            positions_in_gene.append(position_in_gene)\n",
    "            original_base = (v[0][0][2][codon_pos-1])\n",
    "            original_bases.append(original_base)\n",
    "        ASXL1_bases_missense3[k] = (v[0][0][1], positions_in_gene, original_bases, base_changes, codon_number, variant)\n",
    "\n",
    "    #Create dictionary which contains original codon seq, details on which base is changed and its neighbouring bases\n",
    "    ASXL1_bases_missense4 = {}\n",
    "\n",
    "    for k, v in ASXL1_bases_missense3.items():\n",
    "        position_in_gene = v[1]\n",
    "        sites_trio_missense=[]\n",
    "\n",
    "        for i in position_in_gene:\n",
    "    #         print(i)\n",
    "            if i in intron_exon_boundary_caution_base_ASXL1:\n",
    "    #                 print('positions at intron_exon_boundary =',i)\n",
    "                adapted_trio = intron_exon_check_ASXL1(i)\n",
    "                sites_trio_missense.append([adapted_trio])\n",
    "                ASXL1_bases_missense4[k] = (v, sites_trio_missense)\n",
    "\n",
    "            else:\n",
    "                for i in position_in_gene:\n",
    "                    gene_pos = i\n",
    "                    site_trio_missense = [ASXL1_exons_extra_bases[gene_pos:gene_pos+3] for gene_pos in range (gene_pos-1, gene_pos, 3)]\n",
    "                    sites_trio_missense.append(site_trio_missense)\n",
    "                    ASXL1_bases_missense4[k] = (v, sites_trio_missense)\n",
    "\n",
    "    #Create a dictionary containing pos in coding region, original base, new base, base affected with its neighbours\n",
    "    #e.g. 938, G, A, TGG = base 938 on coding region, original base = G, changes to base A, base with neighbours either side = TGG\n",
    "    ASXL1_sites_missense5 = {}\n",
    "\n",
    "    a=0\n",
    "    for k, v in ASXL1_bases_missense4.items():\n",
    "        codon_n = v[0][1] #base number in codon (out of 3 x 912)\n",
    "        codon_n_912 = v[0][4]\n",
    "        original_base = v[0][2]\n",
    "        new_base = v[0][3]\n",
    "        site_trio = v[1]\n",
    "        info=[]\n",
    "        variant = v[0][5]\n",
    "\n",
    "        change = 0\n",
    "        for i in codon_n:\n",
    "            codon_number = i\n",
    "            orig_base = original_base[change]\n",
    "            n_base = new_base[change]\n",
    "            site_tri = site_trio[change]\n",
    "            info.append([codon_number, orig_base, n_base, site_tri, codon_n_912, variant])\n",
    "            change = change+1\n",
    "        ASXL1_sites_missense5[a]=info\n",
    "        a=a+1\n",
    "\n",
    "    # Write in a way that can iterate over mutation rates (i.e. put in to form C[C>A]G etc)   \n",
    "    ASXL1_sites_missense6={}\n",
    "\n",
    "    separating_positions = {} # separate out all the possible changes to just 1 effective item in the list\n",
    "    a=0\n",
    "    for k, v in ASXL1_sites_missense5.items():\n",
    "        for i in v: #because some contain more than 1 change from that codon:\n",
    "            separating_positions[a]=i\n",
    "            a=a+1\n",
    "\n",
    "    for k, v in separating_positions.items():\n",
    "        base_pos = v[0]\n",
    "        original_base = v[1]\n",
    "        new_base = v[2]\n",
    "        site_trio = v[3]\n",
    "        codon_n_912 = v[4]\n",
    "        variant = v[5]\n",
    "        sequence = (original_base,new_base)\n",
    "        changes = '>'.join(sequence)\n",
    "        left_base = site_trio[0][0]\n",
    "        right_base = site_trio[0][2]\n",
    "        sequence2 = (changes, right_base)\n",
    "        changes2 = ']'.join(sequence2)\n",
    "        sequence3 = (left_base, changes2)\n",
    "        changes3 = '['.join(sequence3)\n",
    "        ASXL1_sites_missense6[k] = (v[0],changes3, codon_n_912, variant) #site_trio = the affected base and its neighbours\n",
    "\n",
    "\n",
    "    #Assign the mutation rates to each site...\n",
    "    ASXL1_mutation_rates_missense = {}\n",
    "    for k, v in ASXL1_sites_missense6.items():\n",
    "        base_change_info = v[1]\n",
    "        if base_change_info in Mutation_rates.keys():\n",
    "            mutation_rate = Mutation_rates.get(base_change_info)\n",
    "            ASXL1_mutation_rates_missense[k] = (v, mutation_rate)\n",
    "\n",
    "    ASXL1_mutation_rate_missense_sum = []\n",
    "    mutation_rates_list = []\n",
    "    for k, v in ASXL1_mutation_rates_missense.items():\n",
    "    #     print(v[1])\n",
    "        aa_pos = v[0][2]\n",
    "        site = v[0][1]\n",
    "        mutation_rate_value = v[1] \n",
    "        variant = v[0][3]\n",
    "        mutation_rates_list.append((aa_pos, site, mutation_rate_value, variant))\n",
    "        ASXL1_mutation_rate_missense_sum.append(mutation_rate_value)\n",
    "\n",
    "    ASXL1_Summed_mutation_rate_missense_added = np.sum(ASXL1_mutation_rate_missense_sum)\n",
    "\n",
    "    mutation_rates_list_all = mutation_rates_list\n",
    "    return ASXL1_Summed_mutation_rate_missense_added"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### TP53"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 287,
   "metadata": {
    "code_folding": [
     0
    ]
   },
   "outputs": [],
   "source": [
    "def mutation_rates_variants_TP53(variant):\n",
    "    intron_exon_boundary_caution_base_TP53 = [74, 75, 96, 375, 672, 782, 919, 993, 1100]\n",
    "\n",
    "    #The mutation rate at a base is calculated based on its trinucleotide context.  To obtain the trinucleotide context, the base either side of the mutated base needs to be ascertained.\n",
    "    #The coding sequence (+ an extra base at each end) of each gene is used when determining the trinucleotide context of the base change.  The problem with using the coding sequence is that\n",
    "    #if a base change falls at the end of an exon, its neighbouring base will actuall be an intronic base, which may be different to the 'neighbouring' base in the coding sequence.\n",
    "    #To account for this, there is an \"intron_exon_check\" function that is called at the intron/exon boundary if the neighbouring base in the intron is different to the neighbouring base in the coding sequence.\n",
    "    \n",
    "    def intron_exon_check(position_in_gene_number):\n",
    "        if position_in_gene_number == 74:\n",
    "            gene_pos = 74\n",
    "            first_base = TP53_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = TP53_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 75:\n",
    "            gene_pos = 75\n",
    "            first_base = 'G'\n",
    "            middle_base = TP53_exons_extra_bases[gene_pos]\n",
    "            third_base = TP53_exons_extra_bases[gene_pos+1]\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 96:\n",
    "            gene_pos = 96\n",
    "            first_base = TP53_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = TP53_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 375:\n",
    "            gene_pos = 375\n",
    "            first_base = TP53_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = TP53_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 672:\n",
    "            gene_pos = 672\n",
    "            first_base = TP53_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = TP53_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 782:\n",
    "            gene_pos = 782\n",
    "            first_base = TP53_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = TP53_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 919:\n",
    "            gene_pos = 919\n",
    "            first_base = TP53_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = TP53_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 993:\n",
    "            gene_pos = 993\n",
    "            first_base = TP53_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = TP53_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 1100:\n",
    "            gene_pos = 1100\n",
    "            first_base = TP53_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = TP53_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        return triplet\n",
    "    \n",
    "    #TP53 transcript:\n",
    "    #Downloaded from http://grch37.ensembl.org\n",
    "    TP53_exons = 'ATGGAGGAGCCGCAGTCAGATCCTAGCGTCGAGCCCCCTCTGAGTCAGGAAACATTTTCAGACCTATGGAAACTACTTCCTGAAAACAACGTTCTGTCCCCCTTGCCGTCCCAAGCAATGGATGATTTGATGCTGTCCCCGGACGATATTGAACAATGGTTCACTGAAGACCCAGGTCCAGATGAAGCTCCCAGAATGCCAGAGGCTGCTCCCCCCGTGGCCCCTGCACCAGCAGCTCCTACACCGGCGGCCCCTGCACCAGCCCCCTCCTGGCCCCTGTCATCTTCTGTCCCTTCCCAGAAAACCTACCAGGGCAGCTACGGTTTCCGTCTGGGCTTCTTGCATTCTGGGACAGCCAAGTCTGTGACTTGCACGTACTCCCCTGCCCTCAACAAGATGTTTTGCCAACTGGCCAAGACCTGCCCTGTGCAGCTGTGGGTTGATTCCACACCCCCGCCCGGCACCCGCGTCCGCGCCATGGCCATCTACAAGCAGTCACAGCACATGACGGAGGTTGTGAGGCGCTGCCCCCACCATGAGCGCTGCTCAGATAGCGATGGTCTGGCCCCTCCTCAGCATCTTATCCGAGTGGAAGGAAATTTGCGTGTGGAGTATTTGGATGACAGAAACACTTTTCGACATAGTGTGGTGGTGCCCTATGAGCCGCCTGAGGTTGGCTCTGACTGTACCACCATCCACTACAACTACATGTGTAACAGTTCCTGCATGGGCGGCATGAACCGGAGGCCCATCCTCACCATCATCACACTGGAAGACTCCAGTGGTAATCTACTGGGACGGAACAGCTTTGAGGTGCGTGTTTGTGCCTGTCCTGGGAGAGACCGGCGCACAGAGGAAGAGAATCTCCGCAAGAAAGGGGAGCCTCACCACGAGCTGCCCCCAGGGAGCACTAAGCGAGCACTGCCCAACAACACCAGCTCCTCTCCCCAGCCAAAGAAGAAACCACTGGATGGAGAATATTTCACCCTTCAGATCCGTGGGCGTGAGCGCTTCGAGATGTTCCGAGAGCTGAATGAGGCCTTGGAACTCAAGGATGCCCAGGCTGGGAAGGAGCCAGGGGGGAGCAGGGCTCACTCCAGCCACCTGAAGTCCAAAAAGGGTCAGTCTACCTCCCGCCATAAAAAACTCATGTTCAAGACAGAAGGGCCTGACTCAGACTGA'\n",
    "    \n",
    "    #Split the sequence into codons \n",
    "    TP53_codons = [TP53_exons[i:i+3] for i in range(0, len(TP53_exons), 3)]\n",
    "\n",
    "    #TP53 transcript:\n",
    "    #Downloaded from http://grch37.ensembl.org\n",
    "    TP53_exons_extra_bases = 'CATGGAGGAGCCGCAGTCAGATCCTAGCGTCGAGCCCCCTCTGAGTCAGGAAACATTTTCAGACCTATGGAAACTACTTCCTGAAAACAACGTTCTGTCCCCCTTGCCGTCCCAAGCAATGGATGATTTGATGCTGTCCCCGGACGATATTGAACAATGGTTCACTGAAGACCCAGGTCCAGATGAAGCTCCCAGAATGCCAGAGGCTGCTCCCCCCGTGGCCCCTGCACCAGCAGCTCCTACACCGGCGGCCCCTGCACCAGCCCCCTCCTGGCCCCTGTCATCTTCTGTCCCTTCCCAGAAAACCTACCAGGGCAGCTACGGTTTCCGTCTGGGCTTCTTGCATTCTGGGACAGCCAAGTCTGTGACTTGCACGTACTCCCCTGCCCTCAACAAGATGTTTTGCCAACTGGCCAAGACCTGCCCTGTGCAGCTGTGGGTTGATTCCACACCCCCGCCCGGCACCCGCGTCCGCGCCATGGCCATCTACAAGCAGTCACAGCACATGACGGAGGTTGTGAGGCGCTGCCCCCACCATGAGCGCTGCTCAGATAGCGATGGTCTGGCCCCTCCTCAGCATCTTATCCGAGTGGAAGGAAATTTGCGTGTGGAGTATTTGGATGACAGAAACACTTTTCGACATAGTGTGGTGGTGCCCTATGAGCCGCCTGAGGTTGGCTCTGACTGTACCACCATCCACTACAACTACATGTGTAACAGTTCCTGCATGGGCGGCATGAACCGGAGGCCCATCCTCACCATCATCACACTGGAAGACTCCAGTGGTAATCTACTGGGACGGAACAGCTTTGAGGTGCGTGTTTGTGCCTGTCCTGGGAGAGACCGGCGCACAGAGGAAGAGAATCTCCGCAAGAAAGGGGAGCCTCACCACGAGCTGCCCCCAGGGAGCACTAAGCGAGCACTGCCCAACAACACCAGCTCCTCTCCCCAGCCAAAGAAGAAACCACTGGATGGAGAATATTTCACCCTTCAGATCCGTGGGCGTGAGCGCTTCGAGATGTTCCGAGAGCTGAATGAGGCCTTGGAACTCAAGGATGCCCAGGCTGGGAAGGAGCCAGGGGGGAGCAGGGCTCACTCCAGCCACCTGAAGTCCAAAAAGGGTCAGTCTACCTCCCGCCATAAAAAACTCATGTTCAAGACAGAAGGGCCTGACTCAGACTGAC'\n",
    "    \n",
    "    #Create a list of the codons they called missense variants at (from COSMIC)\n",
    "    TP53_variant_list = []\n",
    "    TP53_missense_codons =[]\n",
    "    TP53_missense_variant_DICT = {}\n",
    "\n",
    "    variant_codon = int(((re.findall('\\d+', variant)))[0])\n",
    "\n",
    "    codon0 = variant_codon-1\n",
    "    codon1 = variant_codon\n",
    "    variant = variant\n",
    "    TP53_missense_codons.append(codon0) #List of codon numbers that they called variants at\n",
    "    TP53_missense_variant_DICT[row_count]=(codon1, variant) #Dictionary:  K = codon number, V = variant at that codon (e.g. 76: p.V76L)\n",
    "\n",
    "    #Looking at what possible amino acid changes are possible at those codons\n",
    "    TP53_missense_codons_DICT = {}\n",
    "    calls_TP53_missense = {}\n",
    "\n",
    "    for i in TP53_missense_codons: #Iterates over the list of codon numbers\n",
    "        TP53_missense_codons_DICT[i+1]= (TP53_codons[i]) #Creates of dicitonary of key = codon number, value = bases (sequence) at that codon\n",
    "        #e.g. 12: CGT, 882: GCA\n",
    "\n",
    "    for k,v in TP53_missense_codons_DICT.items(): #iterating over the list of codon numbers (and codon sequences)\n",
    "        if v in Possible_Amino_Acid_Changes_letters.keys(): #assign the list of all the possible amino acid changes to each codon they called\n",
    "            changes = Possible_Amino_Acid_Changes_letters[v]\n",
    "    #         print(Possible_Amino_Acid_Changes.values())\n",
    "            calls_TP53_missense[k]=(v, changes)\n",
    "        #e.g. 12: (CGT, 'K', '*', 'T'etc...)\n",
    "\n",
    "    #Changing the original amino acid to a Word (e.g. Phe) rather than a codon\n",
    "    calls_TP53_missense_extra = {}\n",
    "\n",
    "    for k, v in calls_TP53_missense.items():\n",
    "        if v[0] in Genetic_Code:\n",
    "            Amino_acid_change = Genetic_Code.get(v[0])\n",
    "            calls_TP53_missense_extra[k] = (Amino_acid_change, v[1])\n",
    "            #e.g. 12: (Phe, 'K', '*', 'T' etc..)\n",
    "\n",
    "    #Changing the original amino acid to a LETTER (rather than the codon)\n",
    "    calls_TP53_missense_extra2 = {}\n",
    "\n",
    "    for k, v in calls_TP53_missense_extra.items():\n",
    "        if v[0] in Amino_acid_abbreviations:\n",
    "            Amino_acid_abbrev = Amino_acid_abbreviations.get(v[0])\n",
    "            calls_TP53_missense_extra2[k] = (Amino_acid_abbrev, v[1])\n",
    "            #e.g. 12: ((F), ('K', '*', 'T' etc..))\n",
    "\n",
    "    #Create a dictionary containing info on which position in the codon is affected to create the variant\n",
    "    TP53_position_in_list_dict_missense={}\n",
    "    for k, v in TP53_missense_variant_DICT.items(): #Iterates over the dictionary of (key = row_count, value = aa_pos, variant (e.g. 3: (76, p.V76L))\n",
    "        observed_change=(v[1])[-1] #Selects the amino acid change (e.g. the L from p.V76L)\n",
    "        possible_changes=calls_TP53_missense[v[0]][1] #possible changes are the second value in the dictionary of (k = codon number, value = (original amino acid, possible changes))\n",
    "        possible_changes_info=calls_TP53_missense[v[0]]\n",
    "        codon_seq=calls_TP53_missense[v[0]][0]\n",
    "        possible_changes_string = \"\".join(possible_changes)\n",
    "        if v[1][2] == calls_TP53_missense_extra2[v[0]][0]:  #v[1][2] = the original amino acid letter (i.e if this aa is in this transcript)\n",
    "            position_in_list = -1\n",
    "            positions_in_list = []\n",
    "            while True:\n",
    "                position_in_list = possible_changes_string.find(observed_change, position_in_list+1) #counts position in possible changes (if 0-2 = 1st base of codon, if 3-5 = 2nd base, if 6-8 = 3rd base)\n",
    "                if position_in_list == -1: break\n",
    "                positions_in_list.append(position_in_list)\n",
    "                TP53_position_in_list_dict_missense[k]=(positions_in_list, possible_changes_info, codon_seq, v[0], v[1])\n",
    "\n",
    "    aa_positions_variants_called = []\n",
    "    for k, v in TP53_position_in_list_dict_missense.items():\n",
    "        aa_positions_variants_called.append(v[3])\n",
    "    #     print('positions at which missense variants called = ', aa_positions_variants_called)\n",
    "\n",
    "    #Create a dictionary that contains the position in the list of possible AA where the variant is and what it changes to\n",
    "    TP53_bases_missense = {}\n",
    "    for k, v in TP53_position_in_list_dict_missense.items():\n",
    "        codon_seq = v[2]\n",
    "        base_changes=[]\n",
    "        if codon_seq in Possible_Base_Changes.keys():\n",
    "            list_of_possible_bases = Possible_Base_Changes.get(codon_seq)\n",
    "        for i in v[0]:\n",
    "            base_change_position = i\n",
    "            base_change = list_of_possible_bases[base_change_position]\n",
    "            base_changes.append(base_change)\n",
    "        TP53_bases_missense[k]=(v, base_changes)\n",
    "\n",
    "    #Create dictionary which contains original codon seq and details on which base is changed\n",
    "    TP53_bases_missense2 = {}\n",
    "\n",
    "    for k, v in TP53_bases_missense.items():\n",
    "        base_change_positions = v[0][0]\n",
    "        codon_positions = []\n",
    "        positions_in_gene = []\n",
    "        for i in base_change_positions:\n",
    "            base_change_position = i\n",
    "            if base_change_position <= 2:\n",
    "                codon_pos = 1\n",
    "            if 2 < base_change_position <= 5:\n",
    "                codon_pos = 2\n",
    "            if 5 < base_change_position <= 8:\n",
    "                codon_pos = 3\n",
    "            codon_positions.append(codon_pos)\n",
    "        TP53_bases_missense2[k] = (v, codon_positions)\n",
    "\n",
    "    #Create dictionary which contains original codon seq, details on which base is changed and it's position in the coding region\n",
    "    TP53_bases_missense3 = {}\n",
    "\n",
    "    for k, v in TP53_bases_missense2.items():\n",
    "        codon_positions = v[1]\n",
    "        base_changes = v[0][1]\n",
    "        original_bases = []\n",
    "        positions_in_gene = []\n",
    "        codon_number = v[0][0][3]\n",
    "        variant = v[0][0][4]\n",
    "        for i in codon_positions:\n",
    "            codon_pos = i\n",
    "            position_in_gene = ((int(v[0][0][3])-1)*3)+codon_pos #codon number (v[0][0][3]) minus 1 (e.g. codon 307 -1 = 306, then times by 3 takes you to beginning of codon 307 (then add the position))  \n",
    "            positions_in_gene.append(position_in_gene)\n",
    "            original_base = (v[0][0][2][codon_pos-1])\n",
    "            original_bases.append(original_base)\n",
    "        TP53_bases_missense3[k] = (v[0][0][1], positions_in_gene, original_bases, base_changes, codon_number, variant)\n",
    "\n",
    "    #Create dictionary which contains original codon seq, details on which base is changed and its neighbouring bases\n",
    "    TP53_bases_missense4 = {}\n",
    "\n",
    "    for k, v in TP53_bases_missense3.items():\n",
    "        position_in_gene = v[1]\n",
    "        sites_trio_missense=[]\n",
    "\n",
    "        for i in position_in_gene:\n",
    "    #         print(i)\n",
    "            if i in intron_exon_boundary_caution_base_TP53:\n",
    "    #                 print('positions at intron_exon_boundary =',i)\n",
    "                adapted_trio = intron_exon_check(i)\n",
    "                sites_trio_missense.append([adapted_trio])\n",
    "                TP53_bases_missense4[k] = (v, sites_trio_missense)\n",
    "\n",
    "            else:\n",
    "                for i in position_in_gene:\n",
    "                    gene_pos = i\n",
    "                    site_trio_missense = [TP53_exons_extra_bases[gene_pos:gene_pos+3] for gene_pos in range (gene_pos-1, gene_pos, 3)]\n",
    "                    sites_trio_missense.append(site_trio_missense)\n",
    "                    TP53_bases_missense4[k] = (v, sites_trio_missense)\n",
    "\n",
    "    #Create a dictionary containing pos in coding region, original base, new base, base affected with its neighbours\n",
    "    #e.g. 938, G, A, TGG = base 938 on coding region, original base = G, changes to base A, base with neighbours either side = TGG\n",
    "    TP53_sites_missense5 = {}\n",
    "\n",
    "    a=0\n",
    "    for k, v in TP53_bases_missense4.items():\n",
    "        codon_n = v[0][1] #base number in codon (out of 3 x 912)\n",
    "        codon_n_912 = v[0][4]\n",
    "        original_base = v[0][2]\n",
    "        new_base = v[0][3]\n",
    "        site_trio = v[1]\n",
    "        info=[]\n",
    "        variant = v[0][5]\n",
    "\n",
    "        change = 0\n",
    "        for i in codon_n:\n",
    "            codon_number = i\n",
    "            orig_base = original_base[change]\n",
    "            n_base = new_base[change]\n",
    "            site_tri = site_trio[change]\n",
    "            info.append([codon_number, orig_base, n_base, site_tri, codon_n_912, variant])\n",
    "            change = change+1\n",
    "        TP53_sites_missense5[a]=info\n",
    "        a=a+1\n",
    "\n",
    "    # Write in a way that can iterate over mutation rates (i.e. put in to form C[C>A]G etc)   \n",
    "    TP53_sites_missense6={}\n",
    "\n",
    "    separating_positions = {} # separate out all the possible changes to just 1 effective item in the list\n",
    "    a=0\n",
    "    for k, v in TP53_sites_missense5.items():\n",
    "        for i in v: #because some contain more than 1 change from that codon:\n",
    "            separating_positions[a]=i\n",
    "            a=a+1\n",
    "\n",
    "    for k, v in separating_positions.items():\n",
    "        base_pos = v[0]\n",
    "        original_base = v[1]\n",
    "        new_base = v[2]\n",
    "        site_trio = v[3]\n",
    "        codon_n_912 = v[4]\n",
    "        variant = v[5]\n",
    "        sequence = (original_base,new_base)\n",
    "        changes = '>'.join(sequence)\n",
    "        left_base = site_trio[0][0]\n",
    "        right_base = site_trio[0][2]\n",
    "        sequence2 = (changes, right_base)\n",
    "        changes2 = ']'.join(sequence2)\n",
    "        sequence3 = (left_base, changes2)\n",
    "        changes3 = '['.join(sequence3)\n",
    "        TP53_sites_missense6[k] = (v[0],changes3, codon_n_912, variant) #site_trio = the affected base and its neighbours\n",
    "\n",
    "\n",
    "    #Assign the mutation rates to each site...\n",
    "    TP53_mutation_rates_missense = {}\n",
    "    for k, v in TP53_sites_missense6.items():\n",
    "        base_change_info = v[1]\n",
    "        if base_change_info in Mutation_rates.keys():\n",
    "            mutation_rate = Mutation_rates.get(base_change_info)\n",
    "            TP53_mutation_rates_missense[k] = (v, mutation_rate)\n",
    "\n",
    "    TP53_mutation_rate_missense_sum = []\n",
    "    mutation_rates_list = []\n",
    "    for k, v in TP53_mutation_rates_missense.items():\n",
    "    #     print(v[1])\n",
    "        aa_pos = v[0][2]\n",
    "        site = v[0][1]\n",
    "        mutation_rate_value = v[1] \n",
    "        variant = v[0][3]\n",
    "        mutation_rates_list.append((aa_pos, site, mutation_rate_value, variant))\n",
    "        TP53_mutation_rate_missense_sum.append(mutation_rate_value)\n",
    "\n",
    "    TP53_Summed_mutation_rate_missense_added = np.sum(TP53_mutation_rate_missense_sum)\n",
    "\n",
    "    mutation_rates_list_all = mutation_rates_list\n",
    "    return TP53_Summed_mutation_rate_missense_added"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Calculating mutation rates from a .csv file of variants"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### DNMT3A"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 288,
   "metadata": {
    "code_folding": [
     0
    ]
   },
   "outputs": [],
   "source": [
    "def mutation_rates_list_nonsense_variants_DNMT3A_NP072046(csv_file_variants):\n",
    "   \n",
    "    #The mutation rate at a base is calculated based on its trinucleotide context.  To obtain the trinucleotide context, the base either side of the mutated base needs to be ascertained.\n",
    "    #The coding sequence (+ an extra base at each end) of each gene is used when determining the trinucleotide context of the base change.  The problem with using the coding sequence is that\n",
    "    #if a base change falls at the end of an exon, its neighbouring base will actuall be an intronic base, which may be different to the 'neighbouring' base in the coding sequence.\n",
    "    #To account for this, there is an \"intron_exon_check\" function that is called at the intron/exon boundary if the neighbouring base in the intron is different to the neighbouring base in the coding sequence.\n",
    "\n",
    "    intron_exon_boundary_caution_base_DNMT3A_NP072046 = [1279, 1429, 1474, 1554, 1852, 2083, 2082, 2173, 2322, 2478]\n",
    "   \n",
    "    def intron_exon_check_DNMT3A_NP072046(position_in_gene_number):\n",
    "        if position_in_gene_number == 1279:\n",
    "            gene_pos = 1279\n",
    "            first_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos-1]\n",
    "            middle_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 1429:\n",
    "            gene_pos = 1429\n",
    "            first_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos-1]\n",
    "            middle_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 1474:\n",
    "            gene_pos = 1474\n",
    "            first_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos-1]\n",
    "            middle_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 1554:\n",
    "            gene_pos = 1554\n",
    "            first_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos-1]\n",
    "            middle_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 1852:\n",
    "            gene_pos = 1852\n",
    "            first_base = 'G'\n",
    "            middle_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos]\n",
    "            third_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos+1]\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 2083:\n",
    "            gene_pos = 2083\n",
    "            first_base = 'G'\n",
    "            middle_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos]\n",
    "            third_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos+1]\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 2082:\n",
    "            gene_pos = 2082\n",
    "            first_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos-1]\n",
    "            middle_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 2173:\n",
    "            gene_pos = 2173\n",
    "            first_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos-1]\n",
    "            middle_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 2322:\n",
    "            gene_pos = 2322\n",
    "            first_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos-1]\n",
    "            middle_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 2478:\n",
    "            gene_pos = 2478\n",
    "            first_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos-1]\n",
    "            middle_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        return triplet\n",
    "    \n",
    "    #DNMT3A transcript used:\n",
    "    #Downloaded from http://grch37.ensembl.org/Homo_sapiens/Transcript/Exons?db=core;g=ENSG00000119772;r=2:25455845-25564754;t=ENST00000321117\n",
    "    DNMT3A_exons_NP_072046 = 'ATGCCCGCCATGCCCTCCAGCGGCCCCGGGGACACCAGCAGCTCTGCTGCGGAGCGGGAGGAGGACCGAAAGGACGGAGAGGAGCAGGAGGAGCCGCGTGGCAAGGAGGAGCGCCAAGAGCCCAGCACCACGGCACGGAAGGTGGGGCGGCCTGGGAGGAAGCGCAAGCACCCCCCGGTGGAAAGCGGTGACACGCCAAAGGACCCTGCGGTGATCTCCAAGTCCCCATCCATGGCCCAGGACTCAGGCGCCTCAGAGCTATTACCCAATGGGGACTTGGAGAAGCGGAGTGAGCCCCAGCCAGAGGAGGGGAGCCCTGCTGGGGGGCAGAAGGGCGGGGCCCCAGCAGAGGGAGAGGGTGCAGCTGAGACCCTGCCTGAAGCCTCAAGAGCAGTGGAAAATGGCTGCTGCACCCCCAAGGAGGGCCGAGGAGCCCCTGCAGAAGCGGGCAAAGAACAGAAGGAGACCAACATCGAATCCATGAAAATGGAGGGCTCCCGGGGCCGGCTGCGGGGTGGCTTGGGCTGGGAGTCCAGCCTCCGTCAGCGGCCCATGCCGAGGCTCACCTTCCAGGCGGGGGACCCCTACTACATCAGCAAGCGCAAGCGGGACGAGTGGCTGGCACGCTGGAAAAGGGAGGCTGAGAAGAAAGCCAAGGTCATTGCAGGAATGAATGCTGTGGAAGAAAACCAGGGGCCCGGGGAGTCTCAGAAGGTGGAGGAGGCCAGCCCTCCTGCTGTGCAGCAGCCCACTGACCCCGCATCCCCCACTGTGGCTACCACGCCTGAGCCCGTGGGGTCCGATGCTGGGGACAAGAATGCCACCAAAGCAGGCGATGACGAGCCAGAGTACGAGGACGGCCGGGGCTTTGGCATTGGGGAGCTGGTGTGGGGGAAACTGCGGGGCTTCTCCTGGTGGCCAGGCCGCATTGTGTCTTGGTGGATGACGGGCCGGAGCCGAGCAGCTGAAGGCACCCGCTGGGTCATGTGGTTCGGAGACGGCAAATTCTCAGTGGTGTGTGTTGAGAAGCTGATGCCGCTGAGCTCGTTTTGCAGTGCGTTCCACCAGGCCACGTACAACAAGCAGCCCATGTACCGCAAAGCCATCTACGAGGTCCTGCAGGTGGCCAGCAGCCGCGCGGGGAAGCTGTTCCCGGTGTGCCACGACAGCGATGAGAGTGACACTGCCAAGGCCGTGGAGGTGCAGAACAAGCCCATGATTGAATGGGCCCTGGGGGGCTTCCAGCCTTCTGGCCCTAAGGGCCTGGAGCCACCAGAAGAAGAGAAGAATCCCTACAAAGAAGTGTACACGGACATGTGGGTGGAACCTGAGGCAGCTGCCTACGCACCACCTCCACCAGCCAAAAAGCCCCGGAAGAGCACAGCGGAGAAGCCCAAGGTCAAGGAGATTATTGATGAGCGCACAAGAGAGCGGCTGGTGTACGAGGTGCGGCAGAAGTGCCGGAACATTGAGGACATCTGCATCTCCTGTGGGAGCCTCAATGTTACCCTGGAACACCCCCTCTTCGTTGGAGGAATGTGCCAAAACTGCAAGAACTGCTTTCTGGAGTGTGCGTACCAGTACGACGACGACGGCTACCAGTCCTACTGCACCATCTGCTGTGGGGGCCGTGAGGTGCTCATGTGCGGAAACAACAACTGCTGCAGGTGCTTTTGCGTGGAGTGTGTGGACCTCTTGGTGGGGCCGGGGGCTGCCCAGGCAGCCATTAAGGAAGACCCCTGGAACTGCTACATGTGCGGGCACAAGGGTACCTACGGGCTGCTGCGGCGGCGAGAGGACTGGCCCTCCCGGCTCCAGATGTTCTTCGCTAATAACCACGACCAGGAATTTGACCCTCCAAAGGTTTACCCACCTGTCCCAGCTGAGAAGAGGAAGCCCATCCGGGTGCTGTCTCTCTTTGATGGAATCGCTACAGGGCTCCTGGTGCTGAAGGACTTGGGCATTCAGGTGGACCGCTACATTGCCTCGGAGGTGTGTGAGGACTCCATCACGGTGGGCATGGTGCGGCACCAGGGGAAGATCATGTACGTCGGGGACGTCCGCAGCGTCACACAGAAGCATATCCAGGAGTGGGGCCCATTCGATCTGGTGATTGGGGGCAGTCCCTGCAATGACCTCTCCATCGTCAACCCTGCTCGCAAGGGCCTCTACGAGGGCACTGGCCGGCTCTTCTTTGAGTTCTACCGCCTCCTGCATGATGCGCGGCCCAAGGAGGGAGATGATCGCCCCTTCTTCTGGCTCTTTGAGAATGTGGTGGCCATGGGCGTTAGTGACAAGAGGGACATCTCGCGATTTCTCGAGTCCAACCCTGTGATGATTGATGCCAAAGAAGTGTCAGCTGCACACAGGGCCCGCTACTTCTGGGGTAACCTTCCCGGTATGAACAGGCCGTTGGCATCCACTGTGAATGATAAGCTGGAGCTGCAGGAGTGTCTGGAGCATGGCAGGATAGCCAAGTTCAGCAAAGTGAGGACCATTACTACGAGGTCAAACTCCATAAAGCAGGGCAAAGACCAGCATTTTCCTGTCTTCATGAATGAGAAAGAGGACATCTTATGGTGCACTGAAATGGAAAGGGTATTTGGTTTCCCAGTCCACTATACTGACGTCTCCAACATGAGCCGCTTGGCGAGGCAGAGACTGCTGGGCCGGTCATGGAGCGTGCCAGTCATCCGCCACCTCTTCGCTCCGCTGAAGGAGTATTTTGCGTGTGTGTAA'\n",
    "\n",
    "    #Split the sequence into codons \n",
    "    DNMT3A_codons_NP_072046 = [DNMT3A_exons_NP_072046[i:i+3] for i in range(0, len(DNMT3A_exons_NP_072046), 3)]\n",
    "\n",
    "    #DNMT3A transcript used:\n",
    "    #Downloaded from http://grch37.ensembl.org/Homo_sapiens/Transcript/Exons?db=core;g=ENSG00000119772;r=2:25455845-25564754;t=ENST00000321117\n",
    "    DNMT3A_exons_extra_bases_NP_072046 = 'GATGCCCGCCATGCCCTCCAGCGGCCCCGGGGACACCAGCAGCTCTGCTGCGGAGCGGGAGGAGGACCGAAAGGACGGAGAGGAGCAGGAGGAGCCGCGTGGCAAGGAGGAGCGCCAAGAGCCCAGCACCACGGCACGGAAGGTGGGGCGGCCTGGGAGGAAGCGCAAGCACCCCCCGGTGGAAAGCGGTGACACGCCAAAGGACCCTGCGGTGATCTCCAAGTCCCCATCCATGGCCCAGGACTCAGGCGCCTCAGAGCTATTACCCAATGGGGACTTGGAGAAGCGGAGTGAGCCCCAGCCAGAGGAGGGGAGCCCTGCTGGGGGGCAGAAGGGCGGGGCCCCAGCAGAGGGAGAGGGTGCAGCTGAGACCCTGCCTGAAGCCTCAAGAGCAGTGGAAAATGGCTGCTGCACCCCCAAGGAGGGCCGAGGAGCCCCTGCAGAAGCGGGCAAAGAACAGAAGGAGACCAACATCGAATCCATGAAAATGGAGGGCTCCCGGGGCCGGCTGCGGGGTGGCTTGGGCTGGGAGTCCAGCCTCCGTCAGCGGCCCATGCCGAGGCTCACCTTCCAGGCGGGGGACCCCTACTACATCAGCAAGCGCAAGCGGGACGAGTGGCTGGCACGCTGGAAAAGGGAGGCTGAGAAGAAAGCCAAGGTCATTGCAGGAATGAATGCTGTGGAAGAAAACCAGGGGCCCGGGGAGTCTCAGAAGGTGGAGGAGGCCAGCCCTCCTGCTGTGCAGCAGCCCACTGACCCCGCATCCCCCACTGTGGCTACCACGCCTGAGCCCGTGGGGTCCGATGCTGGGGACAAGAATGCCACCAAAGCAGGCGATGACGAGCCAGAGTACGAGGACGGCCGGGGCTTTGGCATTGGGGAGCTGGTGTGGGGGAAACTGCGGGGCTTCTCCTGGTGGCCAGGCCGCATTGTGTCTTGGTGGATGACGGGCCGGAGCCGAGCAGCTGAAGGCACCCGCTGGGTCATGTGGTTCGGAGACGGCAAATTCTCAGTGGTGTGTGTTGAGAAGCTGATGCCGCTGAGCTCGTTTTGCAGTGCGTTCCACCAGGCCACGTACAACAAGCAGCCCATGTACCGCAAAGCCATCTACGAGGTCCTGCAGGTGGCCAGCAGCCGCGCGGGGAAGCTGTTCCCGGTGTGCCACGACAGCGATGAGAGTGACACTGCCAAGGCCGTGGAGGTGCAGAACAAGCCCATGATTGAATGGGCCCTGGGGGGCTTCCAGCCTTCTGGCCCTAAGGGCCTGGAGCCACCAGAAGAAGAGAAGAATCCCTACAAAGAAGTGTACACGGACATGTGGGTGGAACCTGAGGCAGCTGCCTACGCACCACCTCCACCAGCCAAAAAGCCCCGGAAGAGCACAGCGGAGAAGCCCAAGGTCAAGGAGATTATTGATGAGCGCACAAGAGAGCGGCTGGTGTACGAGGTGCGGCAGAAGTGCCGGAACATTGAGGACATCTGCATCTCCTGTGGGAGCCTCAATGTTACCCTGGAACACCCCCTCTTCGTTGGAGGAATGTGCCAAAACTGCAAGAACTGCTTTCTGGAGTGTGCGTACCAGTACGACGACGACGGCTACCAGTCCTACTGCACCATCTGCTGTGGGGGCCGTGAGGTGCTCATGTGCGGAAACAACAACTGCTGCAGGTGCTTTTGCGTGGAGTGTGTGGACCTCTTGGTGGGGCCGGGGGCTGCCCAGGCAGCCATTAAGGAAGACCCCTGGAACTGCTACATGTGCGGGCACAAGGGTACCTACGGGCTGCTGCGGCGGCGAGAGGACTGGCCCTCCCGGCTCCAGATGTTCTTCGCTAATAACCACGACCAGGAATTTGACCCTCCAAAGGTTTACCCACCTGTCCCAGCTGAGAAGAGGAAGCCCATCCGGGTGCTGTCTCTCTTTGATGGAATCGCTACAGGGCTCCTGGTGCTGAAGGACTTGGGCATTCAGGTGGACCGCTACATTGCCTCGGAGGTGTGTGAGGACTCCATCACGGTGGGCATGGTGCGGCACCAGGGGAAGATCATGTACGTCGGGGACGTCCGCAGCGTCACACAGAAGCATATCCAGGAGTGGGGCCCATTCGATCTGGTGATTGGGGGCAGTCCCTGCAATGACCTCTCCATCGTCAACCCTGCTCGCAAGGGCCTCTACGAGGGCACTGGCCGGCTCTTCTTTGAGTTCTACCGCCTCCTGCATGATGCGCGGCCCAAGGAGGGAGATGATCGCCCCTTCTTCTGGCTCTTTGAGAATGTGGTGGCCATGGGCGTTAGTGACAAGAGGGACATCTCGCGATTTCTCGAGTCCAACCCTGTGATGATTGATGCCAAAGAAGTGTCAGCTGCACACAGGGCCCGCTACTTCTGGGGTAACCTTCCCGGTATGAACAGGCCGTTGGCATCCACTGTGAATGATAAGCTGGAGCTGCAGGAGTGTCTGGAGCATGGCAGGATAGCCAAGTTCAGCAAAGTGAGGACCATTACTACGAGGTCAAACTCCATAAAGCAGGGCAAAGACCAGCATTTTCCTGTCTTCATGAATGAGAAAGAGGACATCTTATGGTGCACTGAAATGGAAAGGGTATTTGGTTTCCCAGTCCACTATACTGACGTCTCCAACATGAGCCGCTTGGCGAGGCAGAGACTGCTGGGCCGGTCATGGAGCGTGCCAGTCATCCGCCACCTCTTCGCTCCGCTGAAGGAGTATTTTGCGTGTGTGTAAG'\n",
    "\n",
    "    #Create a list of the codons they called nonsense variants at\n",
    "    import csv\n",
    "    filename = csv_file_variants\n",
    "    with open(filename, 'r') as csvfile:\n",
    "        read_reader = csv.reader(csvfile)  #csv.reader returns a reader object which will iterate over lines in the csvfile\n",
    "        row_count=0\n",
    "\n",
    "        DNMT3A_nonsense_variant_list_NP_072046 = []\n",
    "        DNMT3A_nonsense_codons_NP_072046 =[]\n",
    "        DNMT3A_nonsense_variant_DICT_NP_072046 = {}\n",
    "\n",
    "        for row in read_reader: #each row contains aa number (e.g. 11) and variant (e.g. p.D11A)\n",
    "    #         print(row)\n",
    "            if row_count>0: \n",
    "                codon0_NP_072046 = int(row[0])-1\n",
    "                codon1_NP_072046 = int(row[0])\n",
    "                variant_NP_072046 = row[1]\n",
    "                DNMT3A_nonsense_codons_NP_072046.append(codon0_NP_072046) #List of codon numbers that they called variants at\n",
    "                DNMT3A_nonsense_variant_DICT_NP_072046[row_count]=(codon1_NP_072046, variant_NP_072046) #Dictionary:  K = codon number, V = variant at that codon (e.g. 76: p.V76L)\n",
    "\n",
    "            row_count=row_count+1\n",
    "            \n",
    "    #Looking at what possible amino acid changes are possible at those codons\n",
    "    DNMT3A_nonsense_codons_NP_072046_DICT = {}\n",
    "    calls_DNMT3A_nonsense_NP_072046 = {}\n",
    "\n",
    "    for i in DNMT3A_nonsense_codons_NP_072046: #Iterates over the list of codon numbers\n",
    "        DNMT3A_nonsense_codons_NP_072046_DICT[i+1]= (DNMT3A_codons_NP_072046[i]) #Creates of dicitonary of key = codon number, value = bases (sequence) at that codon\n",
    "        #e.g. 12: CGT, 882: GCA\n",
    "        \n",
    "    for k,v in DNMT3A_nonsense_codons_NP_072046_DICT.items(): #iterating over the list of codon numbers (and codon sequences)\n",
    "        if v in Possible_Amino_Acid_Changes_letters.keys(): #assign the list of all the possible amino acid changes to each codon they called\n",
    "            changes = Possible_Amino_Acid_Changes_letters[v]\n",
    "    #         print(Possible_Amino_Acid_Changes.values())\n",
    "            calls_DNMT3A_nonsense_NP_072046[k]=(v, changes)\n",
    "        #e.g. 12: (CGT, 'K', '*', 'T'etc...)\n",
    "        \n",
    "    #Changing the original amino acid to a Word (e.g. Phe) rather than a codon\n",
    "    calls_DNMT3A_nonsense_NP_072046_extra = {}\n",
    "\n",
    "    for k, v in calls_DNMT3A_nonsense_NP_072046.items():\n",
    "        if v[0] in Genetic_Code:\n",
    "            Amino_acid_change = Genetic_Code.get(v[0])\n",
    "            calls_DNMT3A_nonsense_NP_072046_extra[k] = (Amino_acid_change, v[1])\n",
    "            #e.g. 12: (Phe, 'K', '*', 'T' etc..)\n",
    "\n",
    "    #Changing the original amino acid to a LETTER (rather than the codon)\n",
    "    calls_DNMT3A_nonsense_NP_072046_extra2 = {}\n",
    "\n",
    "    for k, v in calls_DNMT3A_nonsense_NP_072046_extra.items():\n",
    "        if v[0] in Amino_acid_abbreviations:\n",
    "            Amino_acid_abbrev = Amino_acid_abbreviations.get(v[0])\n",
    "            calls_DNMT3A_nonsense_NP_072046_extra2[k] = (Amino_acid_abbrev, v[1])\n",
    "            #e.g. 12: ((F), ('K', '*', 'T' etc..))\n",
    "            \n",
    "    #Create a dictionary containing info on which position in the codon is affected to create the variant\n",
    "    DNMT3A_position_in_list_dict_nonsense_NP_072046={}\n",
    "    for k, v in DNMT3A_nonsense_variant_DICT_NP_072046.items(): #Iterates over the dictionary of (key = row_count, value = aa_pos, variant (e.g. 3: (76, p.V76L))\n",
    "        observed_change=(v[1])[-1] #Selects the amino acid change (e.g. the L from p.V76L)\n",
    "        possible_changes=calls_DNMT3A_nonsense_NP_072046[v[0]][1] #possible changes are the second value in the dictionary of (k = codon number, value = (original amino acid, possible changes))\n",
    "        possible_changes_info=calls_DNMT3A_nonsense_NP_072046[v[0]]\n",
    "        codon_seq=calls_DNMT3A_nonsense_NP_072046[v[0]][0]\n",
    "        possible_changes_string = \"\".join(possible_changes)\n",
    "        if v[1][2] == calls_DNMT3A_nonsense_NP_072046_extra2[v[0]][0]:  #v[1][2] = the original amino acid letter (i.e if this aa is in this transcript)\n",
    "            position_in_list = -1\n",
    "            positions_in_list = []\n",
    "            while True:\n",
    "                position_in_list = possible_changes_string.find(observed_change, position_in_list+1) #counts position in possible changes (if 0-2 = 1st base of codon, if 3-5 = 2nd base, if 6-8 = 3rd base)\n",
    "                if position_in_list == -1: break\n",
    "                positions_in_list.append(position_in_list)\n",
    "                DNMT3A_position_in_list_dict_nonsense_NP_072046[k]=(positions_in_list, possible_changes_info, codon_seq, v[0], v[1])\n",
    "            \n",
    "    aa_positions_variants_called = []\n",
    "    for k, v in DNMT3A_position_in_list_dict_nonsense_NP_072046.items():\n",
    "        aa_positions_variants_called.append(v[3])\n",
    "#     print('positions at which nonsense variants called = ', aa_positions_variants_called)\n",
    "                \n",
    "    #Create a dictionary that contains the position in the list of possible AA where the variant is and what it changes to\n",
    "    DNMT3A_bases_nonsense_NP_072046 = {}\n",
    "    for k, v in DNMT3A_position_in_list_dict_nonsense_NP_072046.items():\n",
    "        codon_seq = v[2]\n",
    "        base_changes=[]\n",
    "        if codon_seq in Possible_Base_Changes.keys():\n",
    "            list_of_possible_bases = Possible_Base_Changes.get(codon_seq)\n",
    "        for i in v[0]:\n",
    "            base_change_position = i\n",
    "            base_change = list_of_possible_bases[base_change_position]\n",
    "            base_changes.append(base_change)\n",
    "        DNMT3A_bases_nonsense_NP_072046[k]=(v, base_changes)\n",
    "        \n",
    "    #Create dictionary which contains original codon seq and details on which base is changed\n",
    "    DNMT3A_bases_nonsense2_NP_072046 = {}\n",
    "\n",
    "    for k, v in DNMT3A_bases_nonsense_NP_072046.items():\n",
    "        base_change_positions = v[0][0]\n",
    "        codon_positions = []\n",
    "        positions_in_gene = []\n",
    "        for i in base_change_positions:\n",
    "            base_change_position = i\n",
    "            if base_change_position <= 2:\n",
    "                codon_pos = 1\n",
    "            if 2 < base_change_position <= 5:\n",
    "                codon_pos = 2\n",
    "            if 5 < base_change_position <= 8:\n",
    "                codon_pos = 3\n",
    "            codon_positions.append(codon_pos)\n",
    "        DNMT3A_bases_nonsense2_NP_072046[k] = (v, codon_positions)\n",
    "        \n",
    "    #Create dictionary which contains original codon seq, details on which base is changed and it's position in the coding region\n",
    "    DNMT3A_bases_nonsense3_NP_072046 = {}\n",
    "\n",
    "    for k, v in DNMT3A_bases_nonsense2_NP_072046.items():\n",
    "        codon_positions = v[1]\n",
    "        base_changes = v[0][1]\n",
    "        original_bases = []\n",
    "        positions_in_gene = []\n",
    "        codon_number = v[0][0][3]\n",
    "        variant = v[0][0][4]\n",
    "        for i in codon_positions:\n",
    "            codon_pos = i\n",
    "            position_in_gene = ((int(v[0][0][3])-1)*3)+codon_pos #codon number (v[0][0][3]) minus 1 (e.g. codon 307 -1 = 306, then times by 3 takes you to beginning of codon 307 (then add the position))  \n",
    "            positions_in_gene.append(position_in_gene)\n",
    "            original_base = (v[0][0][2][codon_pos-1])\n",
    "            original_bases.append(original_base)\n",
    "        DNMT3A_bases_nonsense3_NP_072046[k] = (v[0][0][1], positions_in_gene, original_bases, base_changes, codon_number, variant)\n",
    "\n",
    "    #Create dictionary which contains original codon seq, details on which base is changed and its neighbouring bases\n",
    "    DNMT3A_bases_nonsense4_NP_072046 = {}\n",
    "\n",
    "    for k, v in DNMT3A_bases_nonsense3_NP_072046.items():\n",
    "        position_in_gene = v[1]\n",
    "        sites_trio_nonsense=[]\n",
    "\n",
    "        for i in position_in_gene:\n",
    "    #         print(i)\n",
    "            if i in intron_exon_boundary_caution_base_DNMT3A_NP072046:\n",
    "#                 print('positions at intron_exon_boundary =',i)\n",
    "                adapted_trio = intron_exon_check_DNMT3A_NP072046(i)\n",
    "                sites_trio_nonsense.append([adapted_trio])\n",
    "                DNMT3A_bases_nonsense4_NP_072046[k] = (v, sites_trio_nonsense)\n",
    "\n",
    "            else:\n",
    "                for i in position_in_gene:\n",
    "                    gene_pos = i\n",
    "                    site_trio_nonsense = [DNMT3A_exons_extra_bases_NP_072046[gene_pos:gene_pos+3] for gene_pos in range (gene_pos-1, gene_pos, 3)]\n",
    "                    sites_trio_nonsense.append(site_trio_nonsense)\n",
    "                    DNMT3A_bases_nonsense4_NP_072046[k] = (v, sites_trio_nonsense)\n",
    "                    \n",
    "    #Create a dictionary containing pos in coding region, original base, new base, base affected with its neighbours\n",
    "    #e.g. 938, G, A, TGG = base 938 on coding region, original base = G, changes to base A, base with neighbours either side = TGG\n",
    "    DNMT3A_sites_nonsense5_NP_072046 = {}\n",
    "\n",
    "    a=0\n",
    "    for k, v in DNMT3A_bases_nonsense4_NP_072046.items():\n",
    "        codon_n = v[0][1] #base number in codon (out of 3 x 912)\n",
    "        codon_n_912 = v[0][4]\n",
    "        original_base = v[0][2]\n",
    "        new_base = v[0][3]\n",
    "        site_trio = v[1]\n",
    "        info=[]\n",
    "        variant = v[0][5]\n",
    "\n",
    "        change = 0\n",
    "        for i in codon_n:\n",
    "            codon_number = i\n",
    "            orig_base = original_base[change]\n",
    "            n_base = new_base[change]\n",
    "            site_tri = site_trio[change]\n",
    "            info.append([codon_number, orig_base, n_base, site_tri, codon_n_912, variant])\n",
    "            change = change+1\n",
    "        DNMT3A_sites_nonsense5_NP_072046[a]=info\n",
    "        a=a+1\n",
    "#     print(DNMT3A_sites_nonsense5_NP_072046)\n",
    "        \n",
    "    # Write in a way that can iterate over mutation rates (i.e. put in to form C[C>A]G etc)   \n",
    "    DNMT3A_sites_nonsense6_NP_072046={}\n",
    "\n",
    "    separating_positions = {} # separate out all the possible changes to just 1 effective item in the list\n",
    "    a=0\n",
    "    for k, v in DNMT3A_sites_nonsense5_NP_072046.items():\n",
    "        for i in v: #because some contain more than 1 change from that codon:\n",
    "            separating_positions[a]=i\n",
    "            a=a+1\n",
    "    \n",
    "    for k, v in separating_positions.items():\n",
    "        base_pos = v[0]\n",
    "        original_base = v[1]\n",
    "        new_base = v[2]\n",
    "        site_trio = v[3]\n",
    "        codon_n_912 = v[4]\n",
    "        variant = v[5]\n",
    "        sequence = (str(original_base),str(new_base))\n",
    "        changes = '>'.join(sequence)\n",
    "        left_base = site_trio[0][0]\n",
    "        right_base = site_trio[0][2]\n",
    "        sequence2 = (changes, right_base)\n",
    "        changes2 = ']'.join(sequence2)\n",
    "        sequence3 = (left_base, changes2)\n",
    "        changes3 = '['.join(sequence3)\n",
    "        DNMT3A_sites_nonsense6_NP_072046[k] = (v[0],changes3, codon_n_912, variant) #site_trio = the affected base and its neighbours\n",
    "        \n",
    "        \n",
    "    #Assign the mutation rates to each site...\n",
    "    DNMT3A_mutation_rates_nonsense_NP_072046 = {}\n",
    "    for k, v in DNMT3A_sites_nonsense6_NP_072046.items():\n",
    "        base_change_info = v[1]\n",
    "        if base_change_info in Mutation_rates.keys():\n",
    "            mutation_rate = Mutation_rates.get(base_change_info)\n",
    "            DNMT3A_mutation_rates_nonsense_NP_072046[k] = (v, mutation_rate)\n",
    "\n",
    "    DNMT3A_mutation_rate_nonsense_NP_072046_sum = []\n",
    "    mutation_rates_list = []\n",
    "    for k, v in DNMT3A_mutation_rates_nonsense_NP_072046.items():\n",
    "    #     print(v[1])\n",
    "        aa_pos = v[0][2]\n",
    "        site = v[0][1]\n",
    "        mutation_rate_value = v[1] \n",
    "        variant = v[0][3]\n",
    "        mutation_rates_list.append((aa_pos, site, mutation_rate_value, variant))\n",
    "        DNMT3A_mutation_rate_nonsense_NP_072046_sum.append(mutation_rate_value)\n",
    "\n",
    "    DNMT3A_Summed_mutation_rate_nonsense_NP_072046_added = np.sum(DNMT3A_mutation_rate_nonsense_NP_072046_sum)\n",
    "\n",
    "    print('Total summed mutation rate across DNMT3A nonsense_NP_072046 for this study = ', DNMT3A_Summed_mutation_rate_nonsense_NP_072046_added)\n",
    "    \n",
    "    mutation_rates_list_all = mutation_rates_list\n",
    "    return mutation_rates_list_all #in format: column 1 = 882, column 2 = p.R882*"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 289,
   "metadata": {
    "code_folding": [
     0
    ]
   },
   "outputs": [],
   "source": [
    "def mutation_rates_list_missense_variants_DNMT3A_NP072046(csv_file_variants):\n",
    "    \n",
    "    #The mutation rate at a base is calculated based on its trinucleotide context.  To obtain the trinucleotide context, the base either side of the mutated base needs to be ascertained.\n",
    "    #The coding sequence (+ an extra base at each end) of each gene is used when determining the trinucleotide context of the base change.  The problem with using the coding sequence is that\n",
    "    #if a base change falls at the end of an exon, its neighbouring base will actuall be an intronic base, which may be different to the 'neighbouring' base in the coding sequence.\n",
    "    #To account for this, there is an \"intron_exon_check\" function that is called at the intron/exon boundary if the neighbouring base in the intron is different to the neighbouring base in the coding sequence.\n",
    "    \n",
    "    intron_exon_boundary_caution_base_DNMT3A_NP072046 = [1279, 1429, 1474, 1554, 1852, 2083, 2082, 2173, 2322, 2478]\n",
    "\n",
    "    def intron_exon_check_DNMT3A_NP072046(position_in_gene_number):\n",
    "        if position_in_gene_number == 1279:\n",
    "            gene_pos = 1279\n",
    "            first_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos-1]\n",
    "            middle_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 1429:\n",
    "            gene_pos = 1429\n",
    "            first_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos-1]\n",
    "            middle_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 1474:\n",
    "            gene_pos = 1474\n",
    "            first_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos-1]\n",
    "            middle_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 1554:\n",
    "            gene_pos = 1554\n",
    "            first_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos-1]\n",
    "            middle_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 1852:\n",
    "            gene_pos = 1852\n",
    "            first_base = 'G'\n",
    "            middle_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos]\n",
    "            third_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos+1]\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 2083:\n",
    "            gene_pos = 2083\n",
    "            first_base = 'G'\n",
    "            middle_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos]\n",
    "            third_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos+1]\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 2082:\n",
    "            gene_pos = 2082\n",
    "            first_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos-1]\n",
    "            middle_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 2173:\n",
    "            gene_pos = 2173\n",
    "            first_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos-1]\n",
    "            middle_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 2322:\n",
    "            gene_pos = 2322\n",
    "            first_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos-1]\n",
    "            middle_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 2478:\n",
    "            gene_pos = 2478\n",
    "            first_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos-1]\n",
    "            middle_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        return triplet\n",
    "    \n",
    "    #DNMT3A transcript used:\n",
    "    #Downloaded from http://grch37.ensembl.org/Homo_sapiens/Transcript/Exons?db=core;g=ENSG00000119772;r=2:25455845-25564754;t=ENST00000321117\n",
    "    DNMT3A_exons_NP_072046 = 'ATGCCCGCCATGCCCTCCAGCGGCCCCGGGGACACCAGCAGCTCTGCTGCGGAGCGGGAGGAGGACCGAAAGGACGGAGAGGAGCAGGAGGAGCCGCGTGGCAAGGAGGAGCGCCAAGAGCCCAGCACCACGGCACGGAAGGTGGGGCGGCCTGGGAGGAAGCGCAAGCACCCCCCGGTGGAAAGCGGTGACACGCCAAAGGACCCTGCGGTGATCTCCAAGTCCCCATCCATGGCCCAGGACTCAGGCGCCTCAGAGCTATTACCCAATGGGGACTTGGAGAAGCGGAGTGAGCCCCAGCCAGAGGAGGGGAGCCCTGCTGGGGGGCAGAAGGGCGGGGCCCCAGCAGAGGGAGAGGGTGCAGCTGAGACCCTGCCTGAAGCCTCAAGAGCAGTGGAAAATGGCTGCTGCACCCCCAAGGAGGGCCGAGGAGCCCCTGCAGAAGCGGGCAAAGAACAGAAGGAGACCAACATCGAATCCATGAAAATGGAGGGCTCCCGGGGCCGGCTGCGGGGTGGCTTGGGCTGGGAGTCCAGCCTCCGTCAGCGGCCCATGCCGAGGCTCACCTTCCAGGCGGGGGACCCCTACTACATCAGCAAGCGCAAGCGGGACGAGTGGCTGGCACGCTGGAAAAGGGAGGCTGAGAAGAAAGCCAAGGTCATTGCAGGAATGAATGCTGTGGAAGAAAACCAGGGGCCCGGGGAGTCTCAGAAGGTGGAGGAGGCCAGCCCTCCTGCTGTGCAGCAGCCCACTGACCCCGCATCCCCCACTGTGGCTACCACGCCTGAGCCCGTGGGGTCCGATGCTGGGGACAAGAATGCCACCAAAGCAGGCGATGACGAGCCAGAGTACGAGGACGGCCGGGGCTTTGGCATTGGGGAGCTGGTGTGGGGGAAACTGCGGGGCTTCTCCTGGTGGCCAGGCCGCATTGTGTCTTGGTGGATGACGGGCCGGAGCCGAGCAGCTGAAGGCACCCGCTGGGTCATGTGGTTCGGAGACGGCAAATTCTCAGTGGTGTGTGTTGAGAAGCTGATGCCGCTGAGCTCGTTTTGCAGTGCGTTCCACCAGGCCACGTACAACAAGCAGCCCATGTACCGCAAAGCCATCTACGAGGTCCTGCAGGTGGCCAGCAGCCGCGCGGGGAAGCTGTTCCCGGTGTGCCACGACAGCGATGAGAGTGACACTGCCAAGGCCGTGGAGGTGCAGAACAAGCCCATGATTGAATGGGCCCTGGGGGGCTTCCAGCCTTCTGGCCCTAAGGGCCTGGAGCCACCAGAAGAAGAGAAGAATCCCTACAAAGAAGTGTACACGGACATGTGGGTGGAACCTGAGGCAGCTGCCTACGCACCACCTCCACCAGCCAAAAAGCCCCGGAAGAGCACAGCGGAGAAGCCCAAGGTCAAGGAGATTATTGATGAGCGCACAAGAGAGCGGCTGGTGTACGAGGTGCGGCAGAAGTGCCGGAACATTGAGGACATCTGCATCTCCTGTGGGAGCCTCAATGTTACCCTGGAACACCCCCTCTTCGTTGGAGGAATGTGCCAAAACTGCAAGAACTGCTTTCTGGAGTGTGCGTACCAGTACGACGACGACGGCTACCAGTCCTACTGCACCATCTGCTGTGGGGGCCGTGAGGTGCTCATGTGCGGAAACAACAACTGCTGCAGGTGCTTTTGCGTGGAGTGTGTGGACCTCTTGGTGGGGCCGGGGGCTGCCCAGGCAGCCATTAAGGAAGACCCCTGGAACTGCTACATGTGCGGGCACAAGGGTACCTACGGGCTGCTGCGGCGGCGAGAGGACTGGCCCTCCCGGCTCCAGATGTTCTTCGCTAATAACCACGACCAGGAATTTGACCCTCCAAAGGTTTACCCACCTGTCCCAGCTGAGAAGAGGAAGCCCATCCGGGTGCTGTCTCTCTTTGATGGAATCGCTACAGGGCTCCTGGTGCTGAAGGACTTGGGCATTCAGGTGGACCGCTACATTGCCTCGGAGGTGTGTGAGGACTCCATCACGGTGGGCATGGTGCGGCACCAGGGGAAGATCATGTACGTCGGGGACGTCCGCAGCGTCACACAGAAGCATATCCAGGAGTGGGGCCCATTCGATCTGGTGATTGGGGGCAGTCCCTGCAATGACCTCTCCATCGTCAACCCTGCTCGCAAGGGCCTCTACGAGGGCACTGGCCGGCTCTTCTTTGAGTTCTACCGCCTCCTGCATGATGCGCGGCCCAAGGAGGGAGATGATCGCCCCTTCTTCTGGCTCTTTGAGAATGTGGTGGCCATGGGCGTTAGTGACAAGAGGGACATCTCGCGATTTCTCGAGTCCAACCCTGTGATGATTGATGCCAAAGAAGTGTCAGCTGCACACAGGGCCCGCTACTTCTGGGGTAACCTTCCCGGTATGAACAGGCCGTTGGCATCCACTGTGAATGATAAGCTGGAGCTGCAGGAGTGTCTGGAGCATGGCAGGATAGCCAAGTTCAGCAAAGTGAGGACCATTACTACGAGGTCAAACTCCATAAAGCAGGGCAAAGACCAGCATTTTCCTGTCTTCATGAATGAGAAAGAGGACATCTTATGGTGCACTGAAATGGAAAGGGTATTTGGTTTCCCAGTCCACTATACTGACGTCTCCAACATGAGCCGCTTGGCGAGGCAGAGACTGCTGGGCCGGTCATGGAGCGTGCCAGTCATCCGCCACCTCTTCGCTCCGCTGAAGGAGTATTTTGCGTGTGTGTAA'\n",
    "\n",
    "    #Split the sequence into codons \n",
    "    DNMT3A_codons_NP_072046 = [DNMT3A_exons_NP_072046[i:i+3] for i in range(0, len(DNMT3A_exons_NP_072046), 3)]\n",
    "\n",
    "    #DNMT3A transcript used:\n",
    "    #Downloaded from http://grch37.ensembl.org/Homo_sapiens/Transcript/Exons?db=core;g=ENSG00000119772;r=2:25455845-25564754;t=ENST00000321117\n",
    "    DNMT3A_exons_extra_bases_NP_072046 = 'GATGCCCGCCATGCCCTCCAGCGGCCCCGGGGACACCAGCAGCTCTGCTGCGGAGCGGGAGGAGGACCGAAAGGACGGAGAGGAGCAGGAGGAGCCGCGTGGCAAGGAGGAGCGCCAAGAGCCCAGCACCACGGCACGGAAGGTGGGGCGGCCTGGGAGGAAGCGCAAGCACCCCCCGGTGGAAAGCGGTGACACGCCAAAGGACCCTGCGGTGATCTCCAAGTCCCCATCCATGGCCCAGGACTCAGGCGCCTCAGAGCTATTACCCAATGGGGACTTGGAGAAGCGGAGTGAGCCCCAGCCAGAGGAGGGGAGCCCTGCTGGGGGGCAGAAGGGCGGGGCCCCAGCAGAGGGAGAGGGTGCAGCTGAGACCCTGCCTGAAGCCTCAAGAGCAGTGGAAAATGGCTGCTGCACCCCCAAGGAGGGCCGAGGAGCCCCTGCAGAAGCGGGCAAAGAACAGAAGGAGACCAACATCGAATCCATGAAAATGGAGGGCTCCCGGGGCCGGCTGCGGGGTGGCTTGGGCTGGGAGTCCAGCCTCCGTCAGCGGCCCATGCCGAGGCTCACCTTCCAGGCGGGGGACCCCTACTACATCAGCAAGCGCAAGCGGGACGAGTGGCTGGCACGCTGGAAAAGGGAGGCTGAGAAGAAAGCCAAGGTCATTGCAGGAATGAATGCTGTGGAAGAAAACCAGGGGCCCGGGGAGTCTCAGAAGGTGGAGGAGGCCAGCCCTCCTGCTGTGCAGCAGCCCACTGACCCCGCATCCCCCACTGTGGCTACCACGCCTGAGCCCGTGGGGTCCGATGCTGGGGACAAGAATGCCACCAAAGCAGGCGATGACGAGCCAGAGTACGAGGACGGCCGGGGCTTTGGCATTGGGGAGCTGGTGTGGGGGAAACTGCGGGGCTTCTCCTGGTGGCCAGGCCGCATTGTGTCTTGGTGGATGACGGGCCGGAGCCGAGCAGCTGAAGGCACCCGCTGGGTCATGTGGTTCGGAGACGGCAAATTCTCAGTGGTGTGTGTTGAGAAGCTGATGCCGCTGAGCTCGTTTTGCAGTGCGTTCCACCAGGCCACGTACAACAAGCAGCCCATGTACCGCAAAGCCATCTACGAGGTCCTGCAGGTGGCCAGCAGCCGCGCGGGGAAGCTGTTCCCGGTGTGCCACGACAGCGATGAGAGTGACACTGCCAAGGCCGTGGAGGTGCAGAACAAGCCCATGATTGAATGGGCCCTGGGGGGCTTCCAGCCTTCTGGCCCTAAGGGCCTGGAGCCACCAGAAGAAGAGAAGAATCCCTACAAAGAAGTGTACACGGACATGTGGGTGGAACCTGAGGCAGCTGCCTACGCACCACCTCCACCAGCCAAAAAGCCCCGGAAGAGCACAGCGGAGAAGCCCAAGGTCAAGGAGATTATTGATGAGCGCACAAGAGAGCGGCTGGTGTACGAGGTGCGGCAGAAGTGCCGGAACATTGAGGACATCTGCATCTCCTGTGGGAGCCTCAATGTTACCCTGGAACACCCCCTCTTCGTTGGAGGAATGTGCCAAAACTGCAAGAACTGCTTTCTGGAGTGTGCGTACCAGTACGACGACGACGGCTACCAGTCCTACTGCACCATCTGCTGTGGGGGCCGTGAGGTGCTCATGTGCGGAAACAACAACTGCTGCAGGTGCTTTTGCGTGGAGTGTGTGGACCTCTTGGTGGGGCCGGGGGCTGCCCAGGCAGCCATTAAGGAAGACCCCTGGAACTGCTACATGTGCGGGCACAAGGGTACCTACGGGCTGCTGCGGCGGCGAGAGGACTGGCCCTCCCGGCTCCAGATGTTCTTCGCTAATAACCACGACCAGGAATTTGACCCTCCAAAGGTTTACCCACCTGTCCCAGCTGAGAAGAGGAAGCCCATCCGGGTGCTGTCTCTCTTTGATGGAATCGCTACAGGGCTCCTGGTGCTGAAGGACTTGGGCATTCAGGTGGACCGCTACATTGCCTCGGAGGTGTGTGAGGACTCCATCACGGTGGGCATGGTGCGGCACCAGGGGAAGATCATGTACGTCGGGGACGTCCGCAGCGTCACACAGAAGCATATCCAGGAGTGGGGCCCATTCGATCTGGTGATTGGGGGCAGTCCCTGCAATGACCTCTCCATCGTCAACCCTGCTCGCAAGGGCCTCTACGAGGGCACTGGCCGGCTCTTCTTTGAGTTCTACCGCCTCCTGCATGATGCGCGGCCCAAGGAGGGAGATGATCGCCCCTTCTTCTGGCTCTTTGAGAATGTGGTGGCCATGGGCGTTAGTGACAAGAGGGACATCTCGCGATTTCTCGAGTCCAACCCTGTGATGATTGATGCCAAAGAAGTGTCAGCTGCACACAGGGCCCGCTACTTCTGGGGTAACCTTCCCGGTATGAACAGGCCGTTGGCATCCACTGTGAATGATAAGCTGGAGCTGCAGGAGTGTCTGGAGCATGGCAGGATAGCCAAGTTCAGCAAAGTGAGGACCATTACTACGAGGTCAAACTCCATAAAGCAGGGCAAAGACCAGCATTTTCCTGTCTTCATGAATGAGAAAGAGGACATCTTATGGTGCACTGAAATGGAAAGGGTATTTGGTTTCCCAGTCCACTATACTGACGTCTCCAACATGAGCCGCTTGGCGAGGCAGAGACTGCTGGGCCGGTCATGGAGCGTGCCAGTCATCCGCCACCTCTTCGCTCCGCTGAAGGAGTATTTTGCGTGTGTGTAAG'\n",
    "\n",
    "    #Create a list of the codons they called missense variants at\n",
    "    import csv\n",
    "    filename = csv_file_variants\n",
    "    with open(filename, 'r') as csvfile:\n",
    "        read_reader = csv.reader(csvfile)  #csv.reader returns a reader object which will iterate over lines in the csvfile\n",
    "        row_count=0\n",
    "\n",
    "        DNMT3A_missense_variant_list_NP_072046 = []\n",
    "        DNMT3A_missense_codons_NP_072046 =[]\n",
    "        DNMT3A_missense_variant_DICT_NP_072046 = {}\n",
    "\n",
    "        for row in read_reader: #each row contains aa number (e.g. 11) and variant (e.g. p.D11A)\n",
    "    #         print(row)\n",
    "            if row_count>0: \n",
    "                codon0_NP_072046 = int(row[0])-1\n",
    "                codon1_NP_072046 = int(row[0])\n",
    "                variant_NP_072046 = row[1]\n",
    "                DNMT3A_missense_codons_NP_072046.append(codon0_NP_072046) #List of codon numbers that they called variants at\n",
    "                DNMT3A_missense_variant_DICT_NP_072046[row_count]=(codon1_NP_072046, variant_NP_072046) #Dictionary:  K = codon number, V = variant at that codon (e.g. 76: p.V76L)\n",
    "\n",
    "            row_count=row_count+1\n",
    "            \n",
    "    #Looking at what possible amino acid changes are possible at those codons\n",
    "    DNMT3A_missense_codons_NP_072046_DICT = {}\n",
    "    calls_DNMT3A_missense_NP_072046 = {}\n",
    "\n",
    "    for i in DNMT3A_missense_codons_NP_072046: #Iterates over the list of codon numbers\n",
    "        DNMT3A_missense_codons_NP_072046_DICT[i+1]= (DNMT3A_codons_NP_072046[i]) #Creates of dicitonary of key = codon number, value = bases (sequence) at that codon\n",
    "        #e.g. 12: CGT, 882: GCA\n",
    "        \n",
    "    for k,v in DNMT3A_missense_codons_NP_072046_DICT.items(): #iterating over the list of codon numbers (and codon sequences)\n",
    "        if v in Possible_Amino_Acid_Changes_letters.keys(): #assign the list of all the possible amino acid changes to each codon they called\n",
    "            changes = Possible_Amino_Acid_Changes_letters[v]\n",
    "    #         print(Possible_Amino_Acid_Changes.values())\n",
    "            calls_DNMT3A_missense_NP_072046[k]=(v, changes)\n",
    "        #e.g. 12: (CGT, 'K', '*', 'T'etc...)\n",
    "        \n",
    "    #Changing the original amino acid to a Word (e.g. Phe) rather than a codon\n",
    "    calls_DNMT3A_missense_NP_072046_extra = {}\n",
    "\n",
    "    for k, v in calls_DNMT3A_missense_NP_072046.items():\n",
    "        if v[0] in Genetic_Code:\n",
    "            Amino_acid_change = Genetic_Code.get(v[0])\n",
    "            calls_DNMT3A_missense_NP_072046_extra[k] = (Amino_acid_change, v[1])\n",
    "            #e.g. 12: (Phe, 'K', '*', 'T' etc..)\n",
    "\n",
    "    #Changing the original amino acid to a LETTER (rather than the codon)\n",
    "    calls_DNMT3A_missense_NP_072046_extra2 = {}\n",
    "\n",
    "    for k, v in calls_DNMT3A_missense_NP_072046_extra.items():\n",
    "        if v[0] in Amino_acid_abbreviations:\n",
    "            Amino_acid_abbrev = Amino_acid_abbreviations.get(v[0])\n",
    "            calls_DNMT3A_missense_NP_072046_extra2[k] = (Amino_acid_abbrev, v[1])\n",
    "            #e.g. 12: ((F), ('K', '*', 'T' etc..))\n",
    "            \n",
    "    #Create a dictionary containing info on which position in the codon is affected to create the variant\n",
    "    DNMT3A_position_in_list_dict_missense_NP_072046={}\n",
    "    for k, v in DNMT3A_missense_variant_DICT_NP_072046.items(): #Iterates over the dictionary of (key = row_count, value = aa_pos, variant (e.g. 3: (76, p.V76L))\n",
    "        observed_change=(v[1])[-1] #Selects the amino acid change (e.g. the L from p.V76L)\n",
    "        possible_changes=calls_DNMT3A_missense_NP_072046[v[0]][1] #possible changes are the second value in the dictionary of (k = codon number, value = (original amino acid, possible changes))\n",
    "        possible_changes_info=calls_DNMT3A_missense_NP_072046[v[0]]\n",
    "        codon_seq=calls_DNMT3A_missense_NP_072046[v[0]][0]\n",
    "        possible_changes_string = \"\".join(possible_changes)\n",
    "        if v[1][2] == calls_DNMT3A_missense_NP_072046_extra2[v[0]][0]:  #v[1][2] = the original amino acid letter (i.e if this aa is in this transcript)\n",
    "            position_in_list = -1\n",
    "            positions_in_list = []\n",
    "            while True:\n",
    "                position_in_list = possible_changes_string.find(observed_change, position_in_list+1) #counts position in possible changes (if 0-2 = 1st base of codon, if 3-5 = 2nd base, if 6-8 = 3rd base)\n",
    "                if position_in_list == -1: break\n",
    "                positions_in_list.append(position_in_list)\n",
    "                DNMT3A_position_in_list_dict_missense_NP_072046[k]=(positions_in_list, possible_changes_info, codon_seq, v[0], v[1])\n",
    "            \n",
    "    aa_positions_variants_called = []\n",
    "    for k, v in DNMT3A_position_in_list_dict_missense_NP_072046.items():\n",
    "        aa_positions_variants_called.append(v[3])\n",
    "#     print('positions at which missense variants called = ', aa_positions_variants_called)\n",
    "                \n",
    "    #Create a dictionary that contains the position in the list of possible AA where the variant is and what it changes to\n",
    "    DNMT3A_bases_missense_NP_072046 = {}\n",
    "    for k, v in DNMT3A_position_in_list_dict_missense_NP_072046.items():\n",
    "        codon_seq = v[2]\n",
    "        base_changes=[]\n",
    "        if codon_seq in Possible_Base_Changes.keys():\n",
    "            list_of_possible_bases = Possible_Base_Changes.get(codon_seq)\n",
    "        for i in v[0]:\n",
    "            base_change_position = i\n",
    "            base_change = list_of_possible_bases[base_change_position]\n",
    "            base_changes.append(base_change)\n",
    "        DNMT3A_bases_missense_NP_072046[k]=(v, base_changes)\n",
    "        \n",
    "    #Create dictionary which contains original codon seq and details on which base is changed\n",
    "    DNMT3A_bases_missense2_NP_072046 = {}\n",
    "\n",
    "    for k, v in DNMT3A_bases_missense_NP_072046.items():\n",
    "        base_change_positions = v[0][0]\n",
    "        codon_positions = []\n",
    "        positions_in_gene = []\n",
    "        for i in base_change_positions:\n",
    "            base_change_position = i\n",
    "            if base_change_position <= 2:\n",
    "                codon_pos = 1\n",
    "            if 2 < base_change_position <= 5:\n",
    "                codon_pos = 2\n",
    "            if 5 < base_change_position <= 8:\n",
    "                codon_pos = 3\n",
    "            codon_positions.append(codon_pos)\n",
    "        DNMT3A_bases_missense2_NP_072046[k] = (v, codon_positions)\n",
    "        \n",
    "    #Create dictionary which contains original codon seq, details on which base is changed and it's position in the coding region\n",
    "    DNMT3A_bases_missense3_NP_072046 = {}\n",
    "\n",
    "    for k, v in DNMT3A_bases_missense2_NP_072046.items():\n",
    "        codon_positions = v[1]\n",
    "        base_changes = v[0][1]\n",
    "        original_bases = []\n",
    "        positions_in_gene = []\n",
    "        codon_number = v[0][0][3]\n",
    "        variant = v[0][0][4]\n",
    "        for i in codon_positions:\n",
    "            codon_pos = i\n",
    "            position_in_gene = ((int(v[0][0][3])-1)*3)+codon_pos #codon number (v[0][0][3]) minus 1 (e.g. codon 307 -1 = 306, then times by 3 takes you to beginning of codon 307 (then add the position))  \n",
    "            positions_in_gene.append(position_in_gene)\n",
    "            original_base = (v[0][0][2][codon_pos-1])\n",
    "            original_bases.append(original_base)\n",
    "        DNMT3A_bases_missense3_NP_072046[k] = (v[0][0][1], positions_in_gene, original_bases, base_changes, codon_number, variant)\n",
    "\n",
    "    #Create dictionary which contains original codon seq, details on which base is changed and its neighbouring bases\n",
    "    DNMT3A_bases_missense4_NP_072046 = {}\n",
    "\n",
    "    for k, v in DNMT3A_bases_missense3_NP_072046.items():\n",
    "        position_in_gene = v[1]\n",
    "        sites_trio_missense=[]\n",
    "\n",
    "        for i in position_in_gene:\n",
    "    #         print(i)\n",
    "            if i in intron_exon_boundary_caution_base_DNMT3A_NP072046:\n",
    "#                 print('positions at intron_exon_boundary =',i)\n",
    "                adapted_trio = intron_exon_check_DNMT3A_NP072046(i)\n",
    "                sites_trio_missense.append([adapted_trio])\n",
    "                DNMT3A_bases_missense4_NP_072046[k] = (v, sites_trio_missense)\n",
    "\n",
    "            else:\n",
    "                for i in position_in_gene:\n",
    "                    gene_pos = i\n",
    "                    site_trio_missense = [DNMT3A_exons_extra_bases_NP_072046[gene_pos:gene_pos+3] for gene_pos in range (gene_pos-1, gene_pos, 3)]\n",
    "                    sites_trio_missense.append(site_trio_missense)\n",
    "                    DNMT3A_bases_missense4_NP_072046[k] = (v, sites_trio_missense)\n",
    "                    \n",
    "    #Create a dictionary containing pos in coding region, original base, new base, base affected with its neighbours\n",
    "    #e.g. 938, G, A, TGG = base 938 on coding region, original base = G, changes to base A, base with neighbours either side = TGG\n",
    "    DNMT3A_sites_missense5_NP_072046 = {}\n",
    "\n",
    "    a=0\n",
    "    for k, v in DNMT3A_bases_missense4_NP_072046.items():\n",
    "        codon_n = v[0][1] #base number in codon (out of 3 x 912)\n",
    "        codon_n_912 = v[0][4]\n",
    "        original_base = v[0][2]\n",
    "        new_base = v[0][3]\n",
    "        site_trio = v[1]\n",
    "        info=[]\n",
    "        variant = v[0][5]\n",
    "\n",
    "        change = 0\n",
    "        for i in codon_n:\n",
    "            codon_number = i\n",
    "            orig_base = original_base[change]\n",
    "            n_base = new_base[change]\n",
    "            site_tri = site_trio[change]\n",
    "            info.append([codon_number, orig_base, n_base, site_tri, codon_n_912, variant])\n",
    "            change = change+1\n",
    "        DNMT3A_sites_missense5_NP_072046[a]=info\n",
    "        a=a+1\n",
    "        \n",
    "    # Write in a way that can iterate over mutation rates (i.e. put in to form C[C>A]G etc)   \n",
    "    DNMT3A_sites_missense6_NP_072046={}\n",
    "\n",
    "    separating_positions = {} # separate out all the possible changes to just 1 effective item in the list\n",
    "    a=0\n",
    "    for k, v in DNMT3A_sites_missense5_NP_072046.items():\n",
    "        for i in v: #because some contain more than 1 change from that codon:\n",
    "            separating_positions[a]=i\n",
    "            a=a+1\n",
    "    \n",
    "    for k, v in separating_positions.items():\n",
    "        base_pos = v[0]\n",
    "        original_base = v[1]\n",
    "        new_base = v[2]\n",
    "        site_trio = v[3]\n",
    "        codon_n_912 = v[4]\n",
    "        variant = v[5]\n",
    "        sequence = (str(original_base),str(new_base))\n",
    "        changes = '>'.join(sequence)\n",
    "        left_base = site_trio[0][0]\n",
    "        right_base = site_trio[0][2]\n",
    "        sequence2 = (changes, right_base)\n",
    "        changes2 = ']'.join(sequence2)\n",
    "        sequence3 = (left_base, changes2)\n",
    "        changes3 = '['.join(sequence3)\n",
    "        DNMT3A_sites_missense6_NP_072046[k] = (v[0],changes3, codon_n_912, variant) #site_trio = the affected base and its neighbours\n",
    "        \n",
    "        \n",
    "    #Assign the mutation rates to each site...\n",
    "    DNMT3A_mutation_rates_missense_NP_072046 = {}\n",
    "    for k, v in DNMT3A_sites_missense6_NP_072046.items():\n",
    "        base_change_info = v[1]\n",
    "        if base_change_info in Mutation_rates.keys():\n",
    "            mutation_rate = Mutation_rates.get(base_change_info)\n",
    "            DNMT3A_mutation_rates_missense_NP_072046[k] = (v, mutation_rate)\n",
    "\n",
    "    DNMT3A_mutation_rate_missense_NP_072046_sum = []\n",
    "    mutation_rates_list = []\n",
    "    for k, v in DNMT3A_mutation_rates_missense_NP_072046.items():\n",
    "    #     print(v[1])\n",
    "        aa_pos = v[0][2]\n",
    "        site = v[0][1]\n",
    "        mutation_rate_value = v[1] \n",
    "        variant = v[0][3]\n",
    "        mutation_rates_list.append((aa_pos, site, mutation_rate_value, variant))\n",
    "        DNMT3A_mutation_rate_missense_NP_072046_sum.append(mutation_rate_value)\n",
    "        \n",
    "    DNMT3A_Summed_mutation_rate_missense_NP_072046_added = np.sum(DNMT3A_mutation_rate_missense_NP_072046_sum)\n",
    "\n",
    "    print('Total summed mutation rate across DNMT3A missense_NP_072046 for this study = ', DNMT3A_Summed_mutation_rate_missense_NP_072046_added)\n",
    "    \n",
    "    mutation_rates_list_all = mutation_rates_list\n",
    "    return mutation_rates_list_all #in format: column 1 = 882, column 2 = p.R882H"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### TET2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 290,
   "metadata": {
    "code_folding": [
     0
    ]
   },
   "outputs": [],
   "source": [
    "def mutation_rates_list_nonsense_variants_TET2(csv_file_variants):\n",
    "\n",
    "    #The mutation rate at a base is calculated based on its trinucleotide context.  To obtain the trinucleotide context, the base either side of the mutated base needs to be ascertained.\n",
    "    #The coding sequence (+ an extra base at each end) of each gene is used when determining the trinucleotide context of the base change.  The problem with using the coding sequence is that\n",
    "    #if a base change falls at the end of an exon, its neighbouring base will actuall be an intronic base, which may be different to the 'neighbouring' base in the coding sequence.\n",
    "    #To account for this, there is an \"intron_exon_check\" function that is called at the intron/exon boundary if the neighbouring base in the intron is different to the neighbouring base in the coding sequence.\n",
    "    \n",
    "    intron_exon_boundary_caution_base_TET2_NP01120680 = [3409, 3804, 4044, 4537]\n",
    "   \n",
    "    def intron_exon_check_TET2_NP01120680(position_in_gene_number):\n",
    "        if position_in_gene_number == 3409:\n",
    "            gene_pos = 3409\n",
    "            first_base = TET2_exons_extra_bases_NP_01120680[gene_pos-1]\n",
    "            middle_base = TET2_exons_extra_bases_NP_01120680[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 3804:\n",
    "            gene_pos = 3804\n",
    "            first_base = 'G'\n",
    "            middle_base = TET2_exons_extra_bases_NP_01120680[gene_pos]\n",
    "            third_base = TET2_exons_extra_bases_NP_01120680[gene_pos+1]\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 4044:\n",
    "            gene_pos = 4044\n",
    "            first_base = TET2_exons_extra_bases_NP_01120680[gene_pos-1]\n",
    "            middle_base = TET2_exons_extra_bases_NP_01120680[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 4537:\n",
    "            gene_pos = 4537\n",
    "            first_base = TET2_exons_extra_bases_NP_01120680[gene_pos-1]\n",
    "            middle_base = TET2_exons_extra_bases_NP_01120680[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        return triplet\n",
    "    \n",
    "    #TET2 transcript (NP_01120680):\n",
    "    #Downloaded from http://grch37.ensembl.org/Homo_sapiens/Transcript/Exons?db=core;g=ENSG00000168769;r=4:106067032-106163928;t=ENST00000380013\n",
    "    TET2_exons_NP_01120680 = 'ATGGAACAGGATAGAACCAACCATGTTGAGGGCAACAGACTAAGTCCATTCCTGATACCATCACCTCCCATTTGCCAGACAGAACCTCTGGCTACAAAGCTCCAGAATGGAAGCCCACTGCCTGAGAGAGCTCATCCAGAAGTAAATGGAGACACCAAGTGGCACTCTTTCAAAAGTTATTATGGAATACCCTGTATGAAGGGAAGCCAGAATAGTCGTGTGAGTCCTGACTTTACACAAGAAAGTAGAGGGTATTCCAAGTGTTTGCAAAATGGAGGAATAAAACGCACAGTTAGTGAACCTTCTCTCTCTGGGCTCCTTCAGATCAAGAAATTGAAACAAGACCAAAAGGCTAATGGAGAAAGACGTAACTTCGGGGTAAGCCAAGAAAGAAATCCAGGTGAAAGCAGTCAACCAAATGTCTCCGATTTGAGTGATAAGAAAGAATCTGTGAGTTCTGTAGCCCAAGAAAATGCAGTTAAAGATTTCACCAGTTTTTCAACACATAACTGCAGTGGGCCTGAAAATCCAGAGCTTCAGATTCTGAATGAGCAGGAGGGGAAAAGTGCTAATTACCATGACAAGAACATTGTATTACTTAAAAACAAGGCAGTGCTAATGCCTAATGGTGCTACAGTTTCTGCCTCTTCCGTGGAACACACACATGGTGAACTCCTGGAAAAAACACTGTCTCAATATTATCCAGATTGTGTTTCCATTGCGGTGCAGAAAACCACATCTCACATAAATGCCATTAACAGTCAGGCTACTAATGAGTTGTCCTGTGAGATCACTCACCCATCGCATACCTCAGGGCAGATCAATTCCGCACAGACCTCTAACTCTGAGCTGCCTCCAAAGCCAGCTGCAGTGGTGAGTGAGGCCTGTGATGCTGATGATGCTGATAATGCCAGTAAACTAGCTGCAATGCTAAATACCTGTTCCTTTCAGAAACCAGAACAACTACAACAACAAAAATCAGTTTTTGAGATATGCCCATCTCCTGCAGAAAATAACATCCAGGGAACCACAAAGCTAGCGTCTGGTGAAGAATTCTGTTCAGGTTCCAGCAGCAATTTGCAAGCTCCTGGTGGCAGCTCTGAACGGTATTTAAAACAAAATGAAATGAATGGTGCTTACTTCAAGCAAAGCTCAGTGTTCACTAAGGATTCCTTTTCTGCCACTACCACACCACCACCACCATCACAATTGCTTCTTTCTCCCCCTCCTCCTCTTCCACAGGTTCCTCAGCTTCCTTCAGAAGGAAAAAGCACTCTGAATGGTGGAGTTTTAGAAGAACACCACCACTACCCCAACCAAAGTAACACAACACTTTTAAGGGAAGTGAAAATAGAGGGTAAACCTGAGGCACCACCTTCCCAGAGTCCTAATCCATCTACACATGTATGCAGCCCTTCTCCGATGCTTTCTGAAAGGCCTCAGAATAATTGTGTGAACAGGAATGACATACAGACTGCAGGGACAATGACTGTTCCATTGTGTTCTGAGAAAACAAGACCAATGTCAGAACACCTCAAGCATAACCCACCAATTTTTGGTAGCAGTGGAGAGCTACAGGACAACTGCCAGCAGTTGATGAGAAACAAAGAGCAAGAGATTCTGAAGGGTCGAGACAAGGAGCAAACACGAGATCTTGTGCCCCCAACACAGCACTATCTGAAACCAGGATGGATTGAATTGAAGGCCCCTCGTTTTCACCAAGCGGAATCCCATCTAAAACGTAATGAGGCATCACTGCCATCAATTCTTCAGTATCAACCCAATCTCTCCAATCAAATGACCTCCAAACAATACACTGGAAATTCCAACATGCCTGGGGGGCTCCCAAGGCAAGCTTACACCCAGAAAACAACACAGCTGGAGCACAAGTCACAAATGTACCAAGTTGAAATGAATCAAGGGCAGTCCCAAGGTACAGTGGACCAACATCTCCAGTTCCAAAAACCCTCACACCAGGTGCACTTCTCCAAAACAGACCATTTACCAAAAGCTCATGTGCAGTCACTGTGTGGCACTAGATTTCATTTTCAACAAAGAGCAGATTCCCAAACTGAAAAACTTATGTCCCCAGTGTTGAAACAGCACTTGAATCAACAGGCTTCAGAGACTGAGCCATTTTCAAACTCACACCTTTTGCAACATAAGCCTCATAAACAGGCAGCACAAACACAACCATCCCAGAGTTCACATCTCCCTCAAAACCAGCAACAGCAGCAAAAATTACAAATAAAGAATAAAGAGGAAATACTCCAGACTTTTCCTCACCCCCAAAGCAACAATGATCAGCAAAGAGAAGGATCATTCTTTGGCCAGACTAAAGTGGAAGAATGTTTTCATGGTGAAAATCAGTATTCAAAATCAAGCGAGTTCGAGACTCATAATGTCCAAATGGGACTGGAGGAAGTACAGAATATAAATCGTAGAAATTCCCCTTATAGTCAGACCATGAAATCAAGTGCATGCAAAATACAGGTTTCTTGTTCAAACAATACACACCTAGTTTCAGAGAATAAAGAACAGACTACACATCCTGAACTTTTTGCAGGAAACAAGACCCAAAACTTGCATCACATGCAATATTTTCCAAATAATGTGATCCCAAAGCAAGATCTTCTTCACAGGTGCTTTCAAGAACAGGAGCAGAAGTCACAACAAGCTTCAGTTCTACAGGGATATAAAAATAGAAACCAAGATATGTCTGGTCAACAAGCTGCGCAACTTGCTCAGCAAAGGTACTTGATACATAACCATGCAAATGTTTTTCCTGTGCCTGACCAGGGAGGAAGTCACACTCAGACCCCTCCCCAGAAGGACACTCAAAAGCATGCTGCTCTAAGGTGGCATCTCTTACAGAAGCAAGAACAGCAGCAAACACAGCAACCCCAAACTGAGTCTTGCCATAGTCAGATGCACAGGCCAATTAAGGTGGAACCTGGATGCAAGCCACATGCCTGTATGCACACAGCACCACCAGAAAACAAAACATGGAAAAAGGTAACTAAGCAAGAGAATCCACCTGCAAGCTGTGATAATGTGCAGCAAAAGAGCATCATTGAGACCATGGAGCAGCATCTGAAGCAGTTTCACGCCAAGTCGTTATTTGACCATAAGGCTCTTACTCTCAAATCACAGAAGCAAGTAAAAGTTGAAATGTCAGGGCCAGTCACAGTTTTGACTAGACAAACCACTGCTGCAGAACTTGATAGCCACACCCCAGCTTTAGAGCAGCAAACAACTTCTTCAGAAAAGACACCAACCAAAAGAACAGCTGCTTCTGTTCTCAATAATTTTATAGAGTCACCTTCCAAATTACTAGATACTCCTATAAAAAATTTATTGGATACACCTGTCAAGACTCAATATGATTTCCCATCTTGCAGATGTGTAGAGCAAATTATTGAAAAAGATGAAGGTCCTTTTTATACCCATCTAGGAGCAGGTCCTAATGTGGCAGCTATTAGAGAAATCATGGAAGAAAGGTTTGGACAGAAGGGTAAAGCTATTAGGATTGAAAGAGTCATCTATACTGGTAAAGAAGGCAAAAGTTCTCAGGGATGTCCTATTGCTAAGTGGGTGGTTCGCAGAAGCAGCAGTGAAGAGAAGCTACTGTGTTTGGTGCGGGAGCGAGCTGGCCACACCTGTGAGGCTGCAGTGATTGTGATTCTCATCCTGGTGTGGGAAGGAATCCCGCTGTCTCTGGCTGACAAACTCTACTCGGAGCTTACCGAGACGCTGAGGAAATACGGCACGCTCACCAATCGCCGGTGTGCCTTGAATGAAGAGAGAACTTGCGCCTGTCAGGGGCTGGATCCAGAAACCTGTGGTGCCTCCTTCTCTTTTGGTTGTTCATGGAGCATGTACTACAATGGATGTAAGTTTGCCAGAAGCAAGATCCCAAGGAAGTTTAAGCTGCTTGGGGATGACCCAAAAGAGGAAGAGAAACTGGAGTCTCATTTGCAAAACCTGTCCACTCTTATGGCACCAACATATAAGAAACTTGCACCTGATGCATATAATAATCAGATTGAATATGAACACAGAGCACCAGAGTGCCGTCTGGGTCTGAAGGAAGGCCGTCCATTCTCAGGGGTCACTGCATGTTTGGACTTCTGTGCTCATGCCCACAGAGACTTGCACAACATGCAGAATGGCAGCACATTGGTATGCACTCTCACTAGAGAAGACAATCGAGAATTTGGAGGAAAACCTGAGGATGAGCAGCTTCACGTTCTGCCTTTATACAAAGTCTCTGACGTGGATGAGTTTGGGAGTGTGGAAGCTCAGGAGGAGAAAAAACGGAGTGGTGCCATTCAGGTACTGAGTTCTTTTCGGCGAAAAGTCAGGATGTTAGCAGAGCCAGTCAAGACTTGCCGACAAAGGAAACTAGAAGCCAAGAAAGCTGCAGCTGAAAAGCTTTCCTCCCTGGAGAACAGCTCAAATAAAAATGAAAAGGAAAAGTCAGCCCCATCACGTACAAAACAAACTGAAAACGCAAGCCAGGCTAAACAGTTGGCAGAACTTTTGCGACTTTCAGGACCAGTCATGCAGCAGTCCCAGCAGCCCCAGCCTCTACAGAAGCAGCCACCACAGCCCCAGCAGCAGCAGAGACCCCAGCAGCAGCAGCCACATCACCCTCAGACAGAGTCTGTCAACTCTTATTCTGCTTCTGGATCCACCAATCCATACATGAGACGGCCCAATCCAGTTAGTCCTTATCCAAACTCTTCACACACTTCAGATATCTATGGAAGCACCAGCCCTATGAACTTCTATTCCACCTCATCTCAAGCTGCAGGTTCATATTTGAATTCTTCTAATCCCATGAACCCTTACCCTGGGCTTTTGAATCAGAATACCCAATATCCATCATATCAATGCAATGGAAACCTATCAGTGGACAACTGCTCCCCATATCTGGGTTCCTATTCTCCCCAGTCTCAGCCGATGGATCTGTATAGGTATCCAAGCCAAGACCCTCTGTCTAAGCTCAGTCTACCACCCATCCATACACTTTACCAGCCAAGGTTTGGAAATAGCCAGAGTTTTACATCTAAATACTTAGGTTATGGAAACCAAAATATGCAGGGAGATGGTTTCAGCAGTTGTACCATTAGACCAAATGTACATCATGTAGGGAAATTGCCTCCTTATCCCACTCATGAGATGGATGGCCACTTCATGGGAGCCACCTCTAGATTACCACCCAATCTGAGCAATCCAAACATGGACTATAAAAATGGTGAACATCATTCACCTTCTCACATAATCCATAACTACAGTGCAGCTCCGGGCATGTTCAACAGCTCTCTTCATGCCCTGCATCTCCAAAACAAGGAGAATGACATGCTTTCCCACACAGCTAATGGGTTATCAAAGATGCTTCCAGCTCTTAACCATGATAGAACTGCTTGTGTCCAAGGAGGCTTACACAAATTAAGTGATGCTAATGGTCAGGAAAAGCAGCCATTGGCACTAGTCCAGGGTGTGGCTTCTGGTGCAGAGGACAACGATGAGGTCTGGTCAGACAGCGAGCAGAGCTTTCTGGATCCTGACATTGGGGGAGTGGCCGTGGCTCCAACTCATGGGTCAATTCTCATTGAGTGTGCAAAGCGTGAGCTGCATGCCACAACCCCTTTAAAGAATCCCAATAGGAATCACCCCACCAGGATCTCCCTCGTCTTTTACCAGCATAAGAGCATGAATGAGCCAAAACATGGCTTGGCTCTTTGGGAAGCCAAAATGGCTGAAAAAGCCCGTGAGAAAGAGGAAGAGTGTGAAAAGTATGGCCCAGACTATGTGCCTCAGAAATCCCATGGCAAAAAAGTGAAACGGGAGCCTGCTGAGCCACATGAAACTTCAGAGCCCACTTACCTGCGTTTCATCAAGTCTCTTGCCGAAAGGACCATGTCCGTGACCACAGACTCCACAGTAACTACATCTCCATATGCCTTCACTCGGGTCACAGGGCCTTACAACAGATATATATGA'\n",
    "\n",
    "    #Split the sequence into codons \n",
    "    TET2_codons_NP_01120680 = [TET2_exons_NP_01120680[i:i+3] for i in range(0, len(TET2_exons_NP_01120680), 3)]\n",
    "\n",
    "    #TET2 transcript (transcript variant 2 (TET2-001), NM_017628, NP_001120680):\n",
    "    #Downloaded from http://grch37.ensembl.org/Homo_sapiens/Transcript/Exons?db=core;g=ENSG00000168769;r=4:106067032-106163928;t=ENST00000380013\n",
    "    TET2_exons_extra_bases_NP_01120680 = 'GATGGAACAGGATAGAACCAACCATGTTGAGGGCAACAGACTAAGTCCATTCCTGATACCATCACCTCCCATTTGCCAGACAGAACCTCTGGCTACAAAGCTCCAGAATGGAAGCCCACTGCCTGAGAGAGCTCATCCAGAAGTAAATGGAGACACCAAGTGGCACTCTTTCAAAAGTTATTATGGAATACCCTGTATGAAGGGAAGCCAGAATAGTCGTGTGAGTCCTGACTTTACACAAGAAAGTAGAGGGTATTCCAAGTGTTTGCAAAATGGAGGAATAAAACGCACAGTTAGTGAACCTTCTCTCTCTGGGCTCCTTCAGATCAAGAAATTGAAACAAGACCAAAAGGCTAATGGAGAAAGACGTAACTTCGGGGTAAGCCAAGAAAGAAATCCAGGTGAAAGCAGTCAACCAAATGTCTCCGATTTGAGTGATAAGAAAGAATCTGTGAGTTCTGTAGCCCAAGAAAATGCAGTTAAAGATTTCACCAGTTTTTCAACACATAACTGCAGTGGGCCTGAAAATCCAGAGCTTCAGATTCTGAATGAGCAGGAGGGGAAAAGTGCTAATTACCATGACAAGAACATTGTATTACTTAAAAACAAGGCAGTGCTAATGCCTAATGGTGCTACAGTTTCTGCCTCTTCCGTGGAACACACACATGGTGAACTCCTGGAAAAAACACTGTCTCAATATTATCCAGATTGTGTTTCCATTGCGGTGCAGAAAACCACATCTCACATAAATGCCATTAACAGTCAGGCTACTAATGAGTTGTCCTGTGAGATCACTCACCCATCGCATACCTCAGGGCAGATCAATTCCGCACAGACCTCTAACTCTGAGCTGCCTCCAAAGCCAGCTGCAGTGGTGAGTGAGGCCTGTGATGCTGATGATGCTGATAATGCCAGTAAACTAGCTGCAATGCTAAATACCTGTTCCTTTCAGAAACCAGAACAACTACAACAACAAAAATCAGTTTTTGAGATATGCCCATCTCCTGCAGAAAATAACATCCAGGGAACCACAAAGCTAGCGTCTGGTGAAGAATTCTGTTCAGGTTCCAGCAGCAATTTGCAAGCTCCTGGTGGCAGCTCTGAACGGTATTTAAAACAAAATGAAATGAATGGTGCTTACTTCAAGCAAAGCTCAGTGTTCACTAAGGATTCCTTTTCTGCCACTACCACACCACCACCACCATCACAATTGCTTCTTTCTCCCCCTCCTCCTCTTCCACAGGTTCCTCAGCTTCCTTCAGAAGGAAAAAGCACTCTGAATGGTGGAGTTTTAGAAGAACACCACCACTACCCCAACCAAAGTAACACAACACTTTTAAGGGAAGTGAAAATAGAGGGTAAACCTGAGGCACCACCTTCCCAGAGTCCTAATCCATCTACACATGTATGCAGCCCTTCTCCGATGCTTTCTGAAAGGCCTCAGAATAATTGTGTGAACAGGAATGACATACAGACTGCAGGGACAATGACTGTTCCATTGTGTTCTGAGAAAACAAGACCAATGTCAGAACACCTCAAGCATAACCCACCAATTTTTGGTAGCAGTGGAGAGCTACAGGACAACTGCCAGCAGTTGATGAGAAACAAAGAGCAAGAGATTCTGAAGGGTCGAGACAAGGAGCAAACACGAGATCTTGTGCCCCCAACACAGCACTATCTGAAACCAGGATGGATTGAATTGAAGGCCCCTCGTTTTCACCAAGCGGAATCCCATCTAAAACGTAATGAGGCATCACTGCCATCAATTCTTCAGTATCAACCCAATCTCTCCAATCAAATGACCTCCAAACAATACACTGGAAATTCCAACATGCCTGGGGGGCTCCCAAGGCAAGCTTACACCCAGAAAACAACACAGCTGGAGCACAAGTCACAAATGTACCAAGTTGAAATGAATCAAGGGCAGTCCCAAGGTACAGTGGACCAACATCTCCAGTTCCAAAAACCCTCACACCAGGTGCACTTCTCCAAAACAGACCATTTACCAAAAGCTCATGTGCAGTCACTGTGTGGCACTAGATTTCATTTTCAACAAAGAGCAGATTCCCAAACTGAAAAACTTATGTCCCCAGTGTTGAAACAGCACTTGAATCAACAGGCTTCAGAGACTGAGCCATTTTCAAACTCACACCTTTTGCAACATAAGCCTCATAAACAGGCAGCACAAACACAACCATCCCAGAGTTCACATCTCCCTCAAAACCAGCAACAGCAGCAAAAATTACAAATAAAGAATAAAGAGGAAATACTCCAGACTTTTCCTCACCCCCAAAGCAACAATGATCAGCAAAGAGAAGGATCATTCTTTGGCCAGACTAAAGTGGAAGAATGTTTTCATGGTGAAAATCAGTATTCAAAATCAAGCGAGTTCGAGACTCATAATGTCCAAATGGGACTGGAGGAAGTACAGAATATAAATCGTAGAAATTCCCCTTATAGTCAGACCATGAAATCAAGTGCATGCAAAATACAGGTTTCTTGTTCAAACAATACACACCTAGTTTCAGAGAATAAAGAACAGACTACACATCCTGAACTTTTTGCAGGAAACAAGACCCAAAACTTGCATCACATGCAATATTTTCCAAATAATGTGATCCCAAAGCAAGATCTTCTTCACAGGTGCTTTCAAGAACAGGAGCAGAAGTCACAACAAGCTTCAGTTCTACAGGGATATAAAAATAGAAACCAAGATATGTCTGGTCAACAAGCTGCGCAACTTGCTCAGCAAAGGTACTTGATACATAACCATGCAAATGTTTTTCCTGTGCCTGACCAGGGAGGAAGTCACACTCAGACCCCTCCCCAGAAGGACACTCAAAAGCATGCTGCTCTAAGGTGGCATCTCTTACAGAAGCAAGAACAGCAGCAAACACAGCAACCCCAAACTGAGTCTTGCCATAGTCAGATGCACAGGCCAATTAAGGTGGAACCTGGATGCAAGCCACATGCCTGTATGCACACAGCACCACCAGAAAACAAAACATGGAAAAAGGTAACTAAGCAAGAGAATCCACCTGCAAGCTGTGATAATGTGCAGCAAAAGAGCATCATTGAGACCATGGAGCAGCATCTGAAGCAGTTTCACGCCAAGTCGTTATTTGACCATAAGGCTCTTACTCTCAAATCACAGAAGCAAGTAAAAGTTGAAATGTCAGGGCCAGTCACAGTTTTGACTAGACAAACCACTGCTGCAGAACTTGATAGCCACACCCCAGCTTTAGAGCAGCAAACAACTTCTTCAGAAAAGACACCAACCAAAAGAACAGCTGCTTCTGTTCTCAATAATTTTATAGAGTCACCTTCCAAATTACTAGATACTCCTATAAAAAATTTATTGGATACACCTGTCAAGACTCAATATGATTTCCCATCTTGCAGATGTGTAGAGCAAATTATTGAAAAAGATGAAGGTCCTTTTTATACCCATCTAGGAGCAGGTCCTAATGTGGCAGCTATTAGAGAAATCATGGAAGAAAGGTTTGGACAGAAGGGTAAAGCTATTAGGATTGAAAGAGTCATCTATACTGGTAAAGAAGGCAAAAGTTCTCAGGGATGTCCTATTGCTAAGTGGGTGGTTCGCAGAAGCAGCAGTGAAGAGAAGCTACTGTGTTTGGTGCGGGAGCGAGCTGGCCACACCTGTGAGGCTGCAGTGATTGTGATTCTCATCCTGGTGTGGGAAGGAATCCCGCTGTCTCTGGCTGACAAACTCTACTCGGAGCTTACCGAGACGCTGAGGAAATACGGCACGCTCACCAATCGCCGGTGTGCCTTGAATGAAGAGAGAACTTGCGCCTGTCAGGGGCTGGATCCAGAAACCTGTGGTGCCTCCTTCTCTTTTGGTTGTTCATGGAGCATGTACTACAATGGATGTAAGTTTGCCAGAAGCAAGATCCCAAGGAAGTTTAAGCTGCTTGGGGATGACCCAAAAGAGGAAGAGAAACTGGAGTCTCATTTGCAAAACCTGTCCACTCTTATGGCACCAACATATAAGAAACTTGCACCTGATGCATATAATAATCAGATTGAATATGAACACAGAGCACCAGAGTGCCGTCTGGGTCTGAAGGAAGGCCGTCCATTCTCAGGGGTCACTGCATGTTTGGACTTCTGTGCTCATGCCCACAGAGACTTGCACAACATGCAGAATGGCAGCACATTGGTATGCACTCTCACTAGAGAAGACAATCGAGAATTTGGAGGAAAACCTGAGGATGAGCAGCTTCACGTTCTGCCTTTATACAAAGTCTCTGACGTGGATGAGTTTGGGAGTGTGGAAGCTCAGGAGGAGAAAAAACGGAGTGGTGCCATTCAGGTACTGAGTTCTTTTCGGCGAAAAGTCAGGATGTTAGCAGAGCCAGTCAAGACTTGCCGACAAAGGAAACTAGAAGCCAAGAAAGCTGCAGCTGAAAAGCTTTCCTCCCTGGAGAACAGCTCAAATAAAAATGAAAAGGAAAAGTCAGCCCCATCACGTACAAAACAAACTGAAAACGCAAGCCAGGCTAAACAGTTGGCAGAACTTTTGCGACTTTCAGGACCAGTCATGCAGCAGTCCCAGCAGCCCCAGCCTCTACAGAAGCAGCCACCACAGCCCCAGCAGCAGCAGAGACCCCAGCAGCAGCAGCCACATCACCCTCAGACAGAGTCTGTCAACTCTTATTCTGCTTCTGGATCCACCAATCCATACATGAGACGGCCCAATCCAGTTAGTCCTTATCCAAACTCTTCACACACTTCAGATATCTATGGAAGCACCAGCCCTATGAACTTCTATTCCACCTCATCTCAAGCTGCAGGTTCATATTTGAATTCTTCTAATCCCATGAACCCTTACCCTGGGCTTTTGAATCAGAATACCCAATATCCATCATATCAATGCAATGGAAACCTATCAGTGGACAACTGCTCCCCATATCTGGGTTCCTATTCTCCCCAGTCTCAGCCGATGGATCTGTATAGGTATCCAAGCCAAGACCCTCTGTCTAAGCTCAGTCTACCACCCATCCATACACTTTACCAGCCAAGGTTTGGAAATAGCCAGAGTTTTACATCTAAATACTTAGGTTATGGAAACCAAAATATGCAGGGAGATGGTTTCAGCAGTTGTACCATTAGACCAAATGTACATCATGTAGGGAAATTGCCTCCTTATCCCACTCATGAGATGGATGGCCACTTCATGGGAGCCACCTCTAGATTACCACCCAATCTGAGCAATCCAAACATGGACTATAAAAATGGTGAACATCATTCACCTTCTCACATAATCCATAACTACAGTGCAGCTCCGGGCATGTTCAACAGCTCTCTTCATGCCCTGCATCTCCAAAACAAGGAGAATGACATGCTTTCCCACACAGCTAATGGGTTATCAAAGATGCTTCCAGCTCTTAACCATGATAGAACTGCTTGTGTCCAAGGAGGCTTACACAAATTAAGTGATGCTAATGGTCAGGAAAAGCAGCCATTGGCACTAGTCCAGGGTGTGGCTTCTGGTGCAGAGGACAACGATGAGGTCTGGTCAGACAGCGAGCAGAGCTTTCTGGATCCTGACATTGGGGGAGTGGCCGTGGCTCCAACTCATGGGTCAATTCTCATTGAGTGTGCAAAGCGTGAGCTGCATGCCACAACCCCTTTAAAGAATCCCAATAGGAATCACCCCACCAGGATCTCCCTCGTCTTTTACCAGCATAAGAGCATGAATGAGCCAAAACATGGCTTGGCTCTTTGGGAAGCCAAAATGGCTGAAAAAGCCCGTGAGAAAGAGGAAGAGTGTGAAAAGTATGGCCCAGACTATGTGCCTCAGAAATCCCATGGCAAAAAAGTGAAACGGGAGCCTGCTGAGCCACATGAAACTTCAGAGCCCACTTACCTGCGTTTCATCAAGTCTCTTGCCGAAAGGACCATGTCCGTGACCACAGACTCCACAGTAACTACATCTCCATATGCCTTCACTCGGGTCACAGGGCCTTACAACAGATATATATGAT'\n",
    "\n",
    "    #Create a list of the codons they called nonsense variants at\n",
    "    import csv\n",
    "    filename = csv_file_variants\n",
    "    with open(filename, 'r') as csvfile:\n",
    "        read_reader = csv.reader(csvfile)  #csv.reader returns a reader object which will iterate over lines in the csvfile\n",
    "        row_count=0\n",
    "\n",
    "        TET2_nonsense_variant_list_NP_01120680 = []\n",
    "        TET2_nonsense_codons_NP_01120680 =[]\n",
    "        TET2_nonsense_variant_DICT_NP_01120680 = {}\n",
    "\n",
    "        for row in read_reader: #each row contains aa number (e.g. 11) and variant (e.g. p.D11A)\n",
    "    #         print(row)\n",
    "            if row_count>0: \n",
    "                codon0_NP_01120680 = int(row[0])-1\n",
    "                codon1_NP_01120680 = int(row[0])\n",
    "                variant_NP_01120680 = row[1]\n",
    "                TET2_nonsense_codons_NP_01120680.append(codon0_NP_01120680) #List of codon numbers that they called variants at\n",
    "                TET2_nonsense_variant_DICT_NP_01120680[row_count]=(codon1_NP_01120680, variant_NP_01120680) #Dictionary:  K = codon number, V = variant at that codon (e.g. 76: p.V76L)\n",
    "\n",
    "            row_count=row_count+1\n",
    "            \n",
    "    #Looking at what possible amino acid changes are possible at those codons\n",
    "    TET2_nonsense_codons_NP_01120680_DICT = {}\n",
    "    calls_TET2_nonsense_NP_01120680 = {}\n",
    "\n",
    "    for i in TET2_nonsense_codons_NP_01120680: #Iterates over the list of codon numbers\n",
    "        TET2_nonsense_codons_NP_01120680_DICT[i+1]= (TET2_codons_NP_01120680[i]) #Creates of dicitonary of key = codon number, value = bases (sequence) at that codon\n",
    "        #e.g. 12: CGT, 882: GCA\n",
    "        \n",
    "    for k,v in TET2_nonsense_codons_NP_01120680_DICT.items(): #iterating over the list of codon numbers (and codon sequences)\n",
    "        if v in Possible_Amino_Acid_Changes_letters.keys(): #assign the list of all the possible amino acid changes to each codon they called\n",
    "            changes = Possible_Amino_Acid_Changes_letters[v]\n",
    "    #         print(Possible_Amino_Acid_Changes.values())\n",
    "            calls_TET2_nonsense_NP_01120680[k]=(v, changes)\n",
    "        #e.g. 12: (CGT, 'K', '*', 'T'etc...)\n",
    "        \n",
    "    #Changing the original amino acid to a Word (e.g. Phe) rather than a codon\n",
    "    calls_TET2_nonsense_NP_01120680_extra = {}\n",
    "\n",
    "    for k, v in calls_TET2_nonsense_NP_01120680.items():\n",
    "        if v[0] in Genetic_Code:\n",
    "            Amino_acid_change = Genetic_Code.get(v[0])\n",
    "            calls_TET2_nonsense_NP_01120680_extra[k] = (Amino_acid_change, v[1])\n",
    "            #e.g. 12: (Phe, 'K', '*', 'T' etc..)\n",
    "\n",
    "    #Changing the original amino acid to a LETTER (rather than the codon)\n",
    "    calls_TET2_nonsense_NP_01120680_extra2 = {}\n",
    "\n",
    "    for k, v in calls_TET2_nonsense_NP_01120680_extra.items():\n",
    "        if v[0] in Amino_acid_abbreviations:\n",
    "            Amino_acid_abbrev = Amino_acid_abbreviations.get(v[0])\n",
    "            calls_TET2_nonsense_NP_01120680_extra2[k] = (Amino_acid_abbrev, v[1])\n",
    "            #e.g. 12: ((F), ('K', '*', 'T' etc..))\n",
    "            \n",
    "    #Create a dictionary containing info on which position in the codon is affected to create the variant\n",
    "    TET2_position_in_list_dict_nonsense_NP_01120680={}\n",
    "    for k, v in TET2_nonsense_variant_DICT_NP_01120680.items(): #Iterates over the dictionary of (key = row_count, value = aa_pos, variant (e.g. 3: (76, p.V76L))\n",
    "        observed_change=(v[1])[-1] #Selects the amino acid change (e.g. the L from p.V76L)\n",
    "        possible_changes=calls_TET2_nonsense_NP_01120680[v[0]][1] #possible changes are the second value in the dictionary of (k = codon number, value = (original amino acid, possible changes))\n",
    "        possible_changes_info=calls_TET2_nonsense_NP_01120680[v[0]]\n",
    "        codon_seq=calls_TET2_nonsense_NP_01120680[v[0]][0]\n",
    "        possible_changes_string = \"\".join(possible_changes)\n",
    "        if v[1][2] == calls_TET2_nonsense_NP_01120680_extra2[v[0]][0]:  #v[1][2] = the original amino acid letter (i.e if this aa is in this transcript)\n",
    "            position_in_list = -1\n",
    "            positions_in_list = []\n",
    "            while True:\n",
    "                position_in_list = possible_changes_string.find(observed_change, position_in_list+1) #counts position in possible changes (if 0-2 = 1st base of codon, if 3-5 = 2nd base, if 6-8 = 3rd base)\n",
    "                if position_in_list == -1: break\n",
    "                positions_in_list.append(position_in_list)\n",
    "                TET2_position_in_list_dict_nonsense_NP_01120680[k]=(positions_in_list, possible_changes_info, codon_seq, v[0], v[1])\n",
    "            \n",
    "    aa_positions_variants_called = []\n",
    "    for k, v in TET2_position_in_list_dict_nonsense_NP_01120680.items():\n",
    "        aa_positions_variants_called.append(v[3])\n",
    "#     print('positions at which nonsense variants called = ', aa_positions_variants_called)\n",
    "                \n",
    "    #Create a dictionary that contains the position in the list of possible AA where the variant is and what it changes to\n",
    "    TET2_bases_nonsense_NP_01120680 = {}\n",
    "    for k, v in TET2_position_in_list_dict_nonsense_NP_01120680.items():\n",
    "        codon_seq = v[2]\n",
    "        base_changes=[]\n",
    "        if codon_seq in Possible_Base_Changes.keys():\n",
    "            list_of_possible_bases = Possible_Base_Changes.get(codon_seq)\n",
    "        for i in v[0]:\n",
    "            base_change_position = i\n",
    "            base_change = list_of_possible_bases[base_change_position]\n",
    "            base_changes.append(base_change)\n",
    "        TET2_bases_nonsense_NP_01120680[k]=(v, base_changes)\n",
    "        \n",
    "    #Create dictionary which contains original codon seq and details on which base is changed\n",
    "    TET2_bases_nonsense2_NP_01120680 = {}\n",
    "\n",
    "    for k, v in TET2_bases_nonsense_NP_01120680.items():\n",
    "        base_change_positions = v[0][0]\n",
    "        codon_positions = []\n",
    "        positions_in_gene = []\n",
    "        for i in base_change_positions:\n",
    "            base_change_position = i\n",
    "            if base_change_position <= 2:\n",
    "                codon_pos = 1\n",
    "            if 2 < base_change_position <= 5:\n",
    "                codon_pos = 2\n",
    "            if 5 < base_change_position <= 8:\n",
    "                codon_pos = 3\n",
    "            codon_positions.append(codon_pos)\n",
    "        TET2_bases_nonsense2_NP_01120680[k] = (v, codon_positions)\n",
    "        \n",
    "    #Create dictionary which contains original codon seq, details on which base is changed and it's position in the coding region\n",
    "    TET2_bases_nonsense3_NP_01120680 = {}\n",
    "\n",
    "    for k, v in TET2_bases_nonsense2_NP_01120680.items():\n",
    "        codon_positions = v[1]\n",
    "        base_changes = v[0][1]\n",
    "        original_bases = []\n",
    "        positions_in_gene = []\n",
    "        codon_number = v[0][0][3]\n",
    "        variant = v[0][0][4]\n",
    "        for i in codon_positions:\n",
    "            codon_pos = i\n",
    "            position_in_gene = ((int(v[0][0][3])-1)*3)+codon_pos #codon number (v[0][0][3]) minus 1 (e.g. codon 307 -1 = 306, then times by 3 takes you to beginning of codon 307 (then add the position))  \n",
    "            positions_in_gene.append(position_in_gene)\n",
    "            original_base = (v[0][0][2][codon_pos-1])\n",
    "            original_bases.append(original_base)\n",
    "        TET2_bases_nonsense3_NP_01120680[k] = (v[0][0][1], positions_in_gene, original_bases, base_changes, codon_number, variant)\n",
    "\n",
    "    #Create dictionary which contains original codon seq, details on which base is changed and its neighbouring bases\n",
    "    TET2_bases_nonsense4_NP_01120680 = {}\n",
    "\n",
    "    for k, v in TET2_bases_nonsense3_NP_01120680.items():\n",
    "        position_in_gene = v[1]\n",
    "        sites_trio_nonsense=[]\n",
    "\n",
    "        for i in position_in_gene:\n",
    "    #         print(i)\n",
    "            if i in intron_exon_boundary_caution_base_TET2_NP01120680:\n",
    "#                 print('positions at intron_exon_boundary =',i)\n",
    "                adapted_trio = intron_exon_check_TET2_NP01120680(i)\n",
    "                sites_trio_nonsense.append([adapted_trio])\n",
    "                TET2_bases_nonsense4_NP_01120680[k] = (v, sites_trio_nonsense)\n",
    "\n",
    "            else:\n",
    "                for i in position_in_gene:\n",
    "                    gene_pos = i\n",
    "                    site_trio_nonsense = [TET2_exons_extra_bases_NP_01120680[gene_pos:gene_pos+3] for gene_pos in range (gene_pos-1, gene_pos, 3)]\n",
    "                    sites_trio_nonsense.append(site_trio_nonsense)\n",
    "                    TET2_bases_nonsense4_NP_01120680[k] = (v, sites_trio_nonsense)\n",
    "                    \n",
    "    #Create a dictionary containing pos in coding region, original base, new base, base affected with its neighbours\n",
    "    #e.g. 938, G, A, TGG = base 938 on coding region, original base = G, changes to base A, base with neighbours either side = TGG\n",
    "    TET2_sites_nonsense5_NP_01120680 = {}\n",
    "\n",
    "    a=0\n",
    "    for k, v in TET2_bases_nonsense4_NP_01120680.items():\n",
    "        codon_n = v[0][1] #base number in codon (out of 3 x 912)\n",
    "        codon_n_912 = v[0][4]\n",
    "        original_base = v[0][2]\n",
    "        new_base = v[0][3]\n",
    "        site_trio = v[1]\n",
    "        info=[]\n",
    "        variant = v[0][5]\n",
    "\n",
    "        change = 0\n",
    "        for i in codon_n:\n",
    "            codon_number = i\n",
    "            orig_base = original_base[change]\n",
    "            n_base = new_base[change]\n",
    "            site_tri = site_trio[change]\n",
    "            info.append([codon_number, orig_base, n_base, site_tri, codon_n_912, variant])\n",
    "            change = change+1\n",
    "        TET2_sites_nonsense5_NP_01120680[a]=info\n",
    "        a=a+1\n",
    "#     print(TET2_sites_nonsense5_NP_01120680)\n",
    "        \n",
    "    # Write in a way that can iterate over mutation rates (i.e. put in to form C[C>A]G etc)   \n",
    "    TET2_sites_nonsense6_NP_01120680={}\n",
    "\n",
    "    separating_positions = {} # separate out all the possible changes to just 1 effective item in the list\n",
    "    a=0\n",
    "    for k, v in TET2_sites_nonsense5_NP_01120680.items():\n",
    "        for i in v: #because some contain more than 1 change from that codon:\n",
    "            separating_positions[a]=i\n",
    "            a=a+1\n",
    "    \n",
    "    for k, v in separating_positions.items():\n",
    "        base_pos = v[0]\n",
    "        original_base = v[1]\n",
    "        new_base = v[2]\n",
    "        site_trio = v[3]\n",
    "        codon_n_912 = v[4]\n",
    "        variant = v[5]\n",
    "        sequence = (str(original_base),str(new_base))\n",
    "        changes = '>'.join(sequence)\n",
    "        left_base = site_trio[0][0]\n",
    "        right_base = site_trio[0][2]\n",
    "        sequence2 = (changes, right_base)\n",
    "        changes2 = ']'.join(sequence2)\n",
    "        sequence3 = (left_base, changes2)\n",
    "        changes3 = '['.join(sequence3)\n",
    "        TET2_sites_nonsense6_NP_01120680[k] = (v[0],changes3, codon_n_912, variant) #site_trio = the affected base and its neighbours\n",
    "        \n",
    "        \n",
    "    #Assign the mutation rates to each site...\n",
    "    TET2_mutation_rates_nonsense_NP_01120680 = {}\n",
    "    for k, v in TET2_sites_nonsense6_NP_01120680.items():\n",
    "        base_change_info = v[1]\n",
    "        if base_change_info in Mutation_rates.keys():\n",
    "            mutation_rate = Mutation_rates.get(base_change_info)\n",
    "            TET2_mutation_rates_nonsense_NP_01120680[k] = (v, mutation_rate)\n",
    "\n",
    "    TET2_mutation_rate_nonsense_NP_01120680_sum = []\n",
    "    mutation_rates_list = []\n",
    "    for k, v in TET2_mutation_rates_nonsense_NP_01120680.items():\n",
    "    #     print(v[1])\n",
    "        aa_pos = v[0][2]\n",
    "        site = v[0][1]\n",
    "        mutation_rate_value = v[1] \n",
    "        variant = v[0][3]\n",
    "        mutation_rates_list.append((aa_pos, site, mutation_rate_value, variant))\n",
    "        TET2_mutation_rate_nonsense_NP_01120680_sum.append(mutation_rate_value)\n",
    "        \n",
    "    TET2_Summed_mutation_rate_nonsense_NP_01120680_added = np.sum(TET2_mutation_rate_nonsense_NP_01120680_sum)\n",
    "\n",
    "    print('Total summed mutation rate across TET2 nonsense_NP_01120680 for this study = ', TET2_Summed_mutation_rate_nonsense_NP_01120680_added)\n",
    "    \n",
    "    mutation_rates_list_all = mutation_rates_list\n",
    "    return mutation_rates_list_all #in format: column 1 = 882, column 2 = p.R882* #2002 aa"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 291,
   "metadata": {
    "code_folding": [
     0
    ]
   },
   "outputs": [],
   "source": [
    "def mutation_rates_list_missense_variants_TET2(csv_file_variants):\n",
    "    \n",
    "    #The mutation rate at a base is calculated based on its trinucleotide context.  To obtain the trinucleotide context, the base either side of the mutated base needs to be ascertained.\n",
    "    #The coding sequence (+ an extra base at each end) of each gene is used when determining the trinucleotide context of the base change.  The problem with using the coding sequence is that\n",
    "    #if a base change falls at the end of an exon, its neighbouring base will actuall be an intronic base, which may be different to the 'neighbouring' base in the coding sequence.\n",
    "    #To account for this, there is an \"intron_exon_check\" function that is called at the intron/exon boundary if the neighbouring base in the intron is different to the neighbouring base in the coding sequence.\n",
    "    \n",
    "    intron_exon_boundary_caution_base_TET2_NP01120680 = [3409, 3804, 4044, 4537]\n",
    "   \n",
    "    def intron_exon_check_TET2_NP01120680(position_in_gene_number):\n",
    "        if position_in_gene_number == 3409:\n",
    "            gene_pos = 3409\n",
    "            first_base = TET2_exons_extra_bases_NP_01120680[gene_pos-1]\n",
    "            middle_base = TET2_exons_extra_bases_NP_01120680[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 3804:\n",
    "            gene_pos = 3804\n",
    "            first_base = 'G'\n",
    "            middle_base = TET2_exons_extra_bases_NP_01120680[gene_pos]\n",
    "            third_base = TET2_exons_extra_bases_NP_01120680[gene_pos+1]\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 4044:\n",
    "            gene_pos = 4044\n",
    "            first_base = TET2_exons_extra_bases_NP_01120680[gene_pos-1]\n",
    "            middle_base = TET2_exons_extra_bases_NP_01120680[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 4537:\n",
    "            gene_pos = 4537\n",
    "            first_base = TET2_exons_extra_bases_NP_01120680[gene_pos-1]\n",
    "            middle_base = TET2_exons_extra_bases_NP_01120680[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        return triplet\n",
    "    \n",
    "    #TET2 transcript (NP_01120680):\n",
    "    #Downloaded from http://grch37.ensembl.org/Homo_sapiens/Transcript/Exons?db=core;g=ENSG00000168769;r=4:106067032-106163928;t=ENST00000380013\n",
    "    TET2_exons_NP_01120680 = 'ATGGAACAGGATAGAACCAACCATGTTGAGGGCAACAGACTAAGTCCATTCCTGATACCATCACCTCCCATTTGCCAGACAGAACCTCTGGCTACAAAGCTCCAGAATGGAAGCCCACTGCCTGAGAGAGCTCATCCAGAAGTAAATGGAGACACCAAGTGGCACTCTTTCAAAAGTTATTATGGAATACCCTGTATGAAGGGAAGCCAGAATAGTCGTGTGAGTCCTGACTTTACACAAGAAAGTAGAGGGTATTCCAAGTGTTTGCAAAATGGAGGAATAAAACGCACAGTTAGTGAACCTTCTCTCTCTGGGCTCCTTCAGATCAAGAAATTGAAACAAGACCAAAAGGCTAATGGAGAAAGACGTAACTTCGGGGTAAGCCAAGAAAGAAATCCAGGTGAAAGCAGTCAACCAAATGTCTCCGATTTGAGTGATAAGAAAGAATCTGTGAGTTCTGTAGCCCAAGAAAATGCAGTTAAAGATTTCACCAGTTTTTCAACACATAACTGCAGTGGGCCTGAAAATCCAGAGCTTCAGATTCTGAATGAGCAGGAGGGGAAAAGTGCTAATTACCATGACAAGAACATTGTATTACTTAAAAACAAGGCAGTGCTAATGCCTAATGGTGCTACAGTTTCTGCCTCTTCCGTGGAACACACACATGGTGAACTCCTGGAAAAAACACTGTCTCAATATTATCCAGATTGTGTTTCCATTGCGGTGCAGAAAACCACATCTCACATAAATGCCATTAACAGTCAGGCTACTAATGAGTTGTCCTGTGAGATCACTCACCCATCGCATACCTCAGGGCAGATCAATTCCGCACAGACCTCTAACTCTGAGCTGCCTCCAAAGCCAGCTGCAGTGGTGAGTGAGGCCTGTGATGCTGATGATGCTGATAATGCCAGTAAACTAGCTGCAATGCTAAATACCTGTTCCTTTCAGAAACCAGAACAACTACAACAACAAAAATCAGTTTTTGAGATATGCCCATCTCCTGCAGAAAATAACATCCAGGGAACCACAAAGCTAGCGTCTGGTGAAGAATTCTGTTCAGGTTCCAGCAGCAATTTGCAAGCTCCTGGTGGCAGCTCTGAACGGTATTTAAAACAAAATGAAATGAATGGTGCTTACTTCAAGCAAAGCTCAGTGTTCACTAAGGATTCCTTTTCTGCCACTACCACACCACCACCACCATCACAATTGCTTCTTTCTCCCCCTCCTCCTCTTCCACAGGTTCCTCAGCTTCCTTCAGAAGGAAAAAGCACTCTGAATGGTGGAGTTTTAGAAGAACACCACCACTACCCCAACCAAAGTAACACAACACTTTTAAGGGAAGTGAAAATAGAGGGTAAACCTGAGGCACCACCTTCCCAGAGTCCTAATCCATCTACACATGTATGCAGCCCTTCTCCGATGCTTTCTGAAAGGCCTCAGAATAATTGTGTGAACAGGAATGACATACAGACTGCAGGGACAATGACTGTTCCATTGTGTTCTGAGAAAACAAGACCAATGTCAGAACACCTCAAGCATAACCCACCAATTTTTGGTAGCAGTGGAGAGCTACAGGACAACTGCCAGCAGTTGATGAGAAACAAAGAGCAAGAGATTCTGAAGGGTCGAGACAAGGAGCAAACACGAGATCTTGTGCCCCCAACACAGCACTATCTGAAACCAGGATGGATTGAATTGAAGGCCCCTCGTTTTCACCAAGCGGAATCCCATCTAAAACGTAATGAGGCATCACTGCCATCAATTCTTCAGTATCAACCCAATCTCTCCAATCAAATGACCTCCAAACAATACACTGGAAATTCCAACATGCCTGGGGGGCTCCCAAGGCAAGCTTACACCCAGAAAACAACACAGCTGGAGCACAAGTCACAAATGTACCAAGTTGAAATGAATCAAGGGCAGTCCCAAGGTACAGTGGACCAACATCTCCAGTTCCAAAAACCCTCACACCAGGTGCACTTCTCCAAAACAGACCATTTACCAAAAGCTCATGTGCAGTCACTGTGTGGCACTAGATTTCATTTTCAACAAAGAGCAGATTCCCAAACTGAAAAACTTATGTCCCCAGTGTTGAAACAGCACTTGAATCAACAGGCTTCAGAGACTGAGCCATTTTCAAACTCACACCTTTTGCAACATAAGCCTCATAAACAGGCAGCACAAACACAACCATCCCAGAGTTCACATCTCCCTCAAAACCAGCAACAGCAGCAAAAATTACAAATAAAGAATAAAGAGGAAATACTCCAGACTTTTCCTCACCCCCAAAGCAACAATGATCAGCAAAGAGAAGGATCATTCTTTGGCCAGACTAAAGTGGAAGAATGTTTTCATGGTGAAAATCAGTATTCAAAATCAAGCGAGTTCGAGACTCATAATGTCCAAATGGGACTGGAGGAAGTACAGAATATAAATCGTAGAAATTCCCCTTATAGTCAGACCATGAAATCAAGTGCATGCAAAATACAGGTTTCTTGTTCAAACAATACACACCTAGTTTCAGAGAATAAAGAACAGACTACACATCCTGAACTTTTTGCAGGAAACAAGACCCAAAACTTGCATCACATGCAATATTTTCCAAATAATGTGATCCCAAAGCAAGATCTTCTTCACAGGTGCTTTCAAGAACAGGAGCAGAAGTCACAACAAGCTTCAGTTCTACAGGGATATAAAAATAGAAACCAAGATATGTCTGGTCAACAAGCTGCGCAACTTGCTCAGCAAAGGTACTTGATACATAACCATGCAAATGTTTTTCCTGTGCCTGACCAGGGAGGAAGTCACACTCAGACCCCTCCCCAGAAGGACACTCAAAAGCATGCTGCTCTAAGGTGGCATCTCTTACAGAAGCAAGAACAGCAGCAAACACAGCAACCCCAAACTGAGTCTTGCCATAGTCAGATGCACAGGCCAATTAAGGTGGAACCTGGATGCAAGCCACATGCCTGTATGCACACAGCACCACCAGAAAACAAAACATGGAAAAAGGTAACTAAGCAAGAGAATCCACCTGCAAGCTGTGATAATGTGCAGCAAAAGAGCATCATTGAGACCATGGAGCAGCATCTGAAGCAGTTTCACGCCAAGTCGTTATTTGACCATAAGGCTCTTACTCTCAAATCACAGAAGCAAGTAAAAGTTGAAATGTCAGGGCCAGTCACAGTTTTGACTAGACAAACCACTGCTGCAGAACTTGATAGCCACACCCCAGCTTTAGAGCAGCAAACAACTTCTTCAGAAAAGACACCAACCAAAAGAACAGCTGCTTCTGTTCTCAATAATTTTATAGAGTCACCTTCCAAATTACTAGATACTCCTATAAAAAATTTATTGGATACACCTGTCAAGACTCAATATGATTTCCCATCTTGCAGATGTGTAGAGCAAATTATTGAAAAAGATGAAGGTCCTTTTTATACCCATCTAGGAGCAGGTCCTAATGTGGCAGCTATTAGAGAAATCATGGAAGAAAGGTTTGGACAGAAGGGTAAAGCTATTAGGATTGAAAGAGTCATCTATACTGGTAAAGAAGGCAAAAGTTCTCAGGGATGTCCTATTGCTAAGTGGGTGGTTCGCAGAAGCAGCAGTGAAGAGAAGCTACTGTGTTTGGTGCGGGAGCGAGCTGGCCACACCTGTGAGGCTGCAGTGATTGTGATTCTCATCCTGGTGTGGGAAGGAATCCCGCTGTCTCTGGCTGACAAACTCTACTCGGAGCTTACCGAGACGCTGAGGAAATACGGCACGCTCACCAATCGCCGGTGTGCCTTGAATGAAGAGAGAACTTGCGCCTGTCAGGGGCTGGATCCAGAAACCTGTGGTGCCTCCTTCTCTTTTGGTTGTTCATGGAGCATGTACTACAATGGATGTAAGTTTGCCAGAAGCAAGATCCCAAGGAAGTTTAAGCTGCTTGGGGATGACCCAAAAGAGGAAGAGAAACTGGAGTCTCATTTGCAAAACCTGTCCACTCTTATGGCACCAACATATAAGAAACTTGCACCTGATGCATATAATAATCAGATTGAATATGAACACAGAGCACCAGAGTGCCGTCTGGGTCTGAAGGAAGGCCGTCCATTCTCAGGGGTCACTGCATGTTTGGACTTCTGTGCTCATGCCCACAGAGACTTGCACAACATGCAGAATGGCAGCACATTGGTATGCACTCTCACTAGAGAAGACAATCGAGAATTTGGAGGAAAACCTGAGGATGAGCAGCTTCACGTTCTGCCTTTATACAAAGTCTCTGACGTGGATGAGTTTGGGAGTGTGGAAGCTCAGGAGGAGAAAAAACGGAGTGGTGCCATTCAGGTACTGAGTTCTTTTCGGCGAAAAGTCAGGATGTTAGCAGAGCCAGTCAAGACTTGCCGACAAAGGAAACTAGAAGCCAAGAAAGCTGCAGCTGAAAAGCTTTCCTCCCTGGAGAACAGCTCAAATAAAAATGAAAAGGAAAAGTCAGCCCCATCACGTACAAAACAAACTGAAAACGCAAGCCAGGCTAAACAGTTGGCAGAACTTTTGCGACTTTCAGGACCAGTCATGCAGCAGTCCCAGCAGCCCCAGCCTCTACAGAAGCAGCCACCACAGCCCCAGCAGCAGCAGAGACCCCAGCAGCAGCAGCCACATCACCCTCAGACAGAGTCTGTCAACTCTTATTCTGCTTCTGGATCCACCAATCCATACATGAGACGGCCCAATCCAGTTAGTCCTTATCCAAACTCTTCACACACTTCAGATATCTATGGAAGCACCAGCCCTATGAACTTCTATTCCACCTCATCTCAAGCTGCAGGTTCATATTTGAATTCTTCTAATCCCATGAACCCTTACCCTGGGCTTTTGAATCAGAATACCCAATATCCATCATATCAATGCAATGGAAACCTATCAGTGGACAACTGCTCCCCATATCTGGGTTCCTATTCTCCCCAGTCTCAGCCGATGGATCTGTATAGGTATCCAAGCCAAGACCCTCTGTCTAAGCTCAGTCTACCACCCATCCATACACTTTACCAGCCAAGGTTTGGAAATAGCCAGAGTTTTACATCTAAATACTTAGGTTATGGAAACCAAAATATGCAGGGAGATGGTTTCAGCAGTTGTACCATTAGACCAAATGTACATCATGTAGGGAAATTGCCTCCTTATCCCACTCATGAGATGGATGGCCACTTCATGGGAGCCACCTCTAGATTACCACCCAATCTGAGCAATCCAAACATGGACTATAAAAATGGTGAACATCATTCACCTTCTCACATAATCCATAACTACAGTGCAGCTCCGGGCATGTTCAACAGCTCTCTTCATGCCCTGCATCTCCAAAACAAGGAGAATGACATGCTTTCCCACACAGCTAATGGGTTATCAAAGATGCTTCCAGCTCTTAACCATGATAGAACTGCTTGTGTCCAAGGAGGCTTACACAAATTAAGTGATGCTAATGGTCAGGAAAAGCAGCCATTGGCACTAGTCCAGGGTGTGGCTTCTGGTGCAGAGGACAACGATGAGGTCTGGTCAGACAGCGAGCAGAGCTTTCTGGATCCTGACATTGGGGGAGTGGCCGTGGCTCCAACTCATGGGTCAATTCTCATTGAGTGTGCAAAGCGTGAGCTGCATGCCACAACCCCTTTAAAGAATCCCAATAGGAATCACCCCACCAGGATCTCCCTCGTCTTTTACCAGCATAAGAGCATGAATGAGCCAAAACATGGCTTGGCTCTTTGGGAAGCCAAAATGGCTGAAAAAGCCCGTGAGAAAGAGGAAGAGTGTGAAAAGTATGGCCCAGACTATGTGCCTCAGAAATCCCATGGCAAAAAAGTGAAACGGGAGCCTGCTGAGCCACATGAAACTTCAGAGCCCACTTACCTGCGTTTCATCAAGTCTCTTGCCGAAAGGACCATGTCCGTGACCACAGACTCCACAGTAACTACATCTCCATATGCCTTCACTCGGGTCACAGGGCCTTACAACAGATATATATGA'\n",
    "\n",
    "    #Split the sequence into codons \n",
    "    TET2_codons_NP_01120680 = [TET2_exons_NP_01120680[i:i+3] for i in range(0, len(TET2_exons_NP_01120680), 3)]\n",
    "\n",
    "    #TET2 transcript (transcript variant 2 (TET2-001), NM_017628, NP_001120680):\n",
    "    #Downloaded from http://grch37.ensembl.org/Homo_sapiens/Transcript/Exons?db=core;g=ENSG00000168769;r=4:106067032-106163928;t=ENST00000380013\n",
    "    TET2_exons_extra_bases_NP_01120680 = 'GATGGAACAGGATAGAACCAACCATGTTGAGGGCAACAGACTAAGTCCATTCCTGATACCATCACCTCCCATTTGCCAGACAGAACCTCTGGCTACAAAGCTCCAGAATGGAAGCCCACTGCCTGAGAGAGCTCATCCAGAAGTAAATGGAGACACCAAGTGGCACTCTTTCAAAAGTTATTATGGAATACCCTGTATGAAGGGAAGCCAGAATAGTCGTGTGAGTCCTGACTTTACACAAGAAAGTAGAGGGTATTCCAAGTGTTTGCAAAATGGAGGAATAAAACGCACAGTTAGTGAACCTTCTCTCTCTGGGCTCCTTCAGATCAAGAAATTGAAACAAGACCAAAAGGCTAATGGAGAAAGACGTAACTTCGGGGTAAGCCAAGAAAGAAATCCAGGTGAAAGCAGTCAACCAAATGTCTCCGATTTGAGTGATAAGAAAGAATCTGTGAGTTCTGTAGCCCAAGAAAATGCAGTTAAAGATTTCACCAGTTTTTCAACACATAACTGCAGTGGGCCTGAAAATCCAGAGCTTCAGATTCTGAATGAGCAGGAGGGGAAAAGTGCTAATTACCATGACAAGAACATTGTATTACTTAAAAACAAGGCAGTGCTAATGCCTAATGGTGCTACAGTTTCTGCCTCTTCCGTGGAACACACACATGGTGAACTCCTGGAAAAAACACTGTCTCAATATTATCCAGATTGTGTTTCCATTGCGGTGCAGAAAACCACATCTCACATAAATGCCATTAACAGTCAGGCTACTAATGAGTTGTCCTGTGAGATCACTCACCCATCGCATACCTCAGGGCAGATCAATTCCGCACAGACCTCTAACTCTGAGCTGCCTCCAAAGCCAGCTGCAGTGGTGAGTGAGGCCTGTGATGCTGATGATGCTGATAATGCCAGTAAACTAGCTGCAATGCTAAATACCTGTTCCTTTCAGAAACCAGAACAACTACAACAACAAAAATCAGTTTTTGAGATATGCCCATCTCCTGCAGAAAATAACATCCAGGGAACCACAAAGCTAGCGTCTGGTGAAGAATTCTGTTCAGGTTCCAGCAGCAATTTGCAAGCTCCTGGTGGCAGCTCTGAACGGTATTTAAAACAAAATGAAATGAATGGTGCTTACTTCAAGCAAAGCTCAGTGTTCACTAAGGATTCCTTTTCTGCCACTACCACACCACCACCACCATCACAATTGCTTCTTTCTCCCCCTCCTCCTCTTCCACAGGTTCCTCAGCTTCCTTCAGAAGGAAAAAGCACTCTGAATGGTGGAGTTTTAGAAGAACACCACCACTACCCCAACCAAAGTAACACAACACTTTTAAGGGAAGTGAAAATAGAGGGTAAACCTGAGGCACCACCTTCCCAGAGTCCTAATCCATCTACACATGTATGCAGCCCTTCTCCGATGCTTTCTGAAAGGCCTCAGAATAATTGTGTGAACAGGAATGACATACAGACTGCAGGGACAATGACTGTTCCATTGTGTTCTGAGAAAACAAGACCAATGTCAGAACACCTCAAGCATAACCCACCAATTTTTGGTAGCAGTGGAGAGCTACAGGACAACTGCCAGCAGTTGATGAGAAACAAAGAGCAAGAGATTCTGAAGGGTCGAGACAAGGAGCAAACACGAGATCTTGTGCCCCCAACACAGCACTATCTGAAACCAGGATGGATTGAATTGAAGGCCCCTCGTTTTCACCAAGCGGAATCCCATCTAAAACGTAATGAGGCATCACTGCCATCAATTCTTCAGTATCAACCCAATCTCTCCAATCAAATGACCTCCAAACAATACACTGGAAATTCCAACATGCCTGGGGGGCTCCCAAGGCAAGCTTACACCCAGAAAACAACACAGCTGGAGCACAAGTCACAAATGTACCAAGTTGAAATGAATCAAGGGCAGTCCCAAGGTACAGTGGACCAACATCTCCAGTTCCAAAAACCCTCACACCAGGTGCACTTCTCCAAAACAGACCATTTACCAAAAGCTCATGTGCAGTCACTGTGTGGCACTAGATTTCATTTTCAACAAAGAGCAGATTCCCAAACTGAAAAACTTATGTCCCCAGTGTTGAAACAGCACTTGAATCAACAGGCTTCAGAGACTGAGCCATTTTCAAACTCACACCTTTTGCAACATAAGCCTCATAAACAGGCAGCACAAACACAACCATCCCAGAGTTCACATCTCCCTCAAAACCAGCAACAGCAGCAAAAATTACAAATAAAGAATAAAGAGGAAATACTCCAGACTTTTCCTCACCCCCAAAGCAACAATGATCAGCAAAGAGAAGGATCATTCTTTGGCCAGACTAAAGTGGAAGAATGTTTTCATGGTGAAAATCAGTATTCAAAATCAAGCGAGTTCGAGACTCATAATGTCCAAATGGGACTGGAGGAAGTACAGAATATAAATCGTAGAAATTCCCCTTATAGTCAGACCATGAAATCAAGTGCATGCAAAATACAGGTTTCTTGTTCAAACAATACACACCTAGTTTCAGAGAATAAAGAACAGACTACACATCCTGAACTTTTTGCAGGAAACAAGACCCAAAACTTGCATCACATGCAATATTTTCCAAATAATGTGATCCCAAAGCAAGATCTTCTTCACAGGTGCTTTCAAGAACAGGAGCAGAAGTCACAACAAGCTTCAGTTCTACAGGGATATAAAAATAGAAACCAAGATATGTCTGGTCAACAAGCTGCGCAACTTGCTCAGCAAAGGTACTTGATACATAACCATGCAAATGTTTTTCCTGTGCCTGACCAGGGAGGAAGTCACACTCAGACCCCTCCCCAGAAGGACACTCAAAAGCATGCTGCTCTAAGGTGGCATCTCTTACAGAAGCAAGAACAGCAGCAAACACAGCAACCCCAAACTGAGTCTTGCCATAGTCAGATGCACAGGCCAATTAAGGTGGAACCTGGATGCAAGCCACATGCCTGTATGCACACAGCACCACCAGAAAACAAAACATGGAAAAAGGTAACTAAGCAAGAGAATCCACCTGCAAGCTGTGATAATGTGCAGCAAAAGAGCATCATTGAGACCATGGAGCAGCATCTGAAGCAGTTTCACGCCAAGTCGTTATTTGACCATAAGGCTCTTACTCTCAAATCACAGAAGCAAGTAAAAGTTGAAATGTCAGGGCCAGTCACAGTTTTGACTAGACAAACCACTGCTGCAGAACTTGATAGCCACACCCCAGCTTTAGAGCAGCAAACAACTTCTTCAGAAAAGACACCAACCAAAAGAACAGCTGCTTCTGTTCTCAATAATTTTATAGAGTCACCTTCCAAATTACTAGATACTCCTATAAAAAATTTATTGGATACACCTGTCAAGACTCAATATGATTTCCCATCTTGCAGATGTGTAGAGCAAATTATTGAAAAAGATGAAGGTCCTTTTTATACCCATCTAGGAGCAGGTCCTAATGTGGCAGCTATTAGAGAAATCATGGAAGAAAGGTTTGGACAGAAGGGTAAAGCTATTAGGATTGAAAGAGTCATCTATACTGGTAAAGAAGGCAAAAGTTCTCAGGGATGTCCTATTGCTAAGTGGGTGGTTCGCAGAAGCAGCAGTGAAGAGAAGCTACTGTGTTTGGTGCGGGAGCGAGCTGGCCACACCTGTGAGGCTGCAGTGATTGTGATTCTCATCCTGGTGTGGGAAGGAATCCCGCTGTCTCTGGCTGACAAACTCTACTCGGAGCTTACCGAGACGCTGAGGAAATACGGCACGCTCACCAATCGCCGGTGTGCCTTGAATGAAGAGAGAACTTGCGCCTGTCAGGGGCTGGATCCAGAAACCTGTGGTGCCTCCTTCTCTTTTGGTTGTTCATGGAGCATGTACTACAATGGATGTAAGTTTGCCAGAAGCAAGATCCCAAGGAAGTTTAAGCTGCTTGGGGATGACCCAAAAGAGGAAGAGAAACTGGAGTCTCATTTGCAAAACCTGTCCACTCTTATGGCACCAACATATAAGAAACTTGCACCTGATGCATATAATAATCAGATTGAATATGAACACAGAGCACCAGAGTGCCGTCTGGGTCTGAAGGAAGGCCGTCCATTCTCAGGGGTCACTGCATGTTTGGACTTCTGTGCTCATGCCCACAGAGACTTGCACAACATGCAGAATGGCAGCACATTGGTATGCACTCTCACTAGAGAAGACAATCGAGAATTTGGAGGAAAACCTGAGGATGAGCAGCTTCACGTTCTGCCTTTATACAAAGTCTCTGACGTGGATGAGTTTGGGAGTGTGGAAGCTCAGGAGGAGAAAAAACGGAGTGGTGCCATTCAGGTACTGAGTTCTTTTCGGCGAAAAGTCAGGATGTTAGCAGAGCCAGTCAAGACTTGCCGACAAAGGAAACTAGAAGCCAAGAAAGCTGCAGCTGAAAAGCTTTCCTCCCTGGAGAACAGCTCAAATAAAAATGAAAAGGAAAAGTCAGCCCCATCACGTACAAAACAAACTGAAAACGCAAGCCAGGCTAAACAGTTGGCAGAACTTTTGCGACTTTCAGGACCAGTCATGCAGCAGTCCCAGCAGCCCCAGCCTCTACAGAAGCAGCCACCACAGCCCCAGCAGCAGCAGAGACCCCAGCAGCAGCAGCCACATCACCCTCAGACAGAGTCTGTCAACTCTTATTCTGCTTCTGGATCCACCAATCCATACATGAGACGGCCCAATCCAGTTAGTCCTTATCCAAACTCTTCACACACTTCAGATATCTATGGAAGCACCAGCCCTATGAACTTCTATTCCACCTCATCTCAAGCTGCAGGTTCATATTTGAATTCTTCTAATCCCATGAACCCTTACCCTGGGCTTTTGAATCAGAATACCCAATATCCATCATATCAATGCAATGGAAACCTATCAGTGGACAACTGCTCCCCATATCTGGGTTCCTATTCTCCCCAGTCTCAGCCGATGGATCTGTATAGGTATCCAAGCCAAGACCCTCTGTCTAAGCTCAGTCTACCACCCATCCATACACTTTACCAGCCAAGGTTTGGAAATAGCCAGAGTTTTACATCTAAATACTTAGGTTATGGAAACCAAAATATGCAGGGAGATGGTTTCAGCAGTTGTACCATTAGACCAAATGTACATCATGTAGGGAAATTGCCTCCTTATCCCACTCATGAGATGGATGGCCACTTCATGGGAGCCACCTCTAGATTACCACCCAATCTGAGCAATCCAAACATGGACTATAAAAATGGTGAACATCATTCACCTTCTCACATAATCCATAACTACAGTGCAGCTCCGGGCATGTTCAACAGCTCTCTTCATGCCCTGCATCTCCAAAACAAGGAGAATGACATGCTTTCCCACACAGCTAATGGGTTATCAAAGATGCTTCCAGCTCTTAACCATGATAGAACTGCTTGTGTCCAAGGAGGCTTACACAAATTAAGTGATGCTAATGGTCAGGAAAAGCAGCCATTGGCACTAGTCCAGGGTGTGGCTTCTGGTGCAGAGGACAACGATGAGGTCTGGTCAGACAGCGAGCAGAGCTTTCTGGATCCTGACATTGGGGGAGTGGCCGTGGCTCCAACTCATGGGTCAATTCTCATTGAGTGTGCAAAGCGTGAGCTGCATGCCACAACCCCTTTAAAGAATCCCAATAGGAATCACCCCACCAGGATCTCCCTCGTCTTTTACCAGCATAAGAGCATGAATGAGCCAAAACATGGCTTGGCTCTTTGGGAAGCCAAAATGGCTGAAAAAGCCCGTGAGAAAGAGGAAGAGTGTGAAAAGTATGGCCCAGACTATGTGCCTCAGAAATCCCATGGCAAAAAAGTGAAACGGGAGCCTGCTGAGCCACATGAAACTTCAGAGCCCACTTACCTGCGTTTCATCAAGTCTCTTGCCGAAAGGACCATGTCCGTGACCACAGACTCCACAGTAACTACATCTCCATATGCCTTCACTCGGGTCACAGGGCCTTACAACAGATATATATGAT'\n",
    "\n",
    "\n",
    "    #Create a list of the codons they called missense variants at\n",
    "    import csv\n",
    "    filename = csv_file_variants\n",
    "    with open(filename, 'r') as csvfile:\n",
    "        read_reader = csv.reader(csvfile)  #csv.reader returns a reader object which will iterate over lines in the csvfile\n",
    "        row_count=0\n",
    "\n",
    "        TET2_missense_variant_list_NP_01120680 = []\n",
    "        TET2_missense_codons_NP_01120680 =[]\n",
    "        TET2_missense_variant_DICT_NP_01120680 = {}\n",
    "\n",
    "        for row in read_reader: #each row contains aa number (e.g. 11) and variant (e.g. p.D11A)\n",
    "    #         print(row)\n",
    "            if row_count>0: \n",
    "                codon0_NP_01120680 = int(row[0])-1\n",
    "                codon1_NP_01120680 = int(row[0])\n",
    "                variant_NP_01120680 = row[1]\n",
    "                TET2_missense_codons_NP_01120680.append(codon0_NP_01120680) #List of codon numbers that they called variants at\n",
    "                TET2_missense_variant_DICT_NP_01120680[row_count]=(codon1_NP_01120680, variant_NP_01120680) #Dictionary:  K = codon number, V = variant at that codon (e.g. 76: p.V76L)\n",
    "\n",
    "            row_count=row_count+1\n",
    "            \n",
    "    #Looking at what possible amino acid changes are possible at those codons\n",
    "    TET2_missense_codons_NP_01120680_DICT = {}\n",
    "    calls_TET2_missense_NP_01120680 = {}\n",
    "\n",
    "    for i in TET2_missense_codons_NP_01120680: #Iterates over the list of codon numbers\n",
    "        TET2_missense_codons_NP_01120680_DICT[i+1]= (TET2_codons_NP_01120680[i]) #Creates of dicitonary of key = codon number, value = bases (sequence) at that codon\n",
    "        #e.g. 12: CGT, 882: GCA\n",
    "        \n",
    "    for k,v in TET2_missense_codons_NP_01120680_DICT.items(): #iterating over the list of codon numbers (and codon sequences)\n",
    "        if v in Possible_Amino_Acid_Changes_letters.keys(): #assign the list of all the possible amino acid changes to each codon they called\n",
    "            changes = Possible_Amino_Acid_Changes_letters[v]\n",
    "    #         print(Possible_Amino_Acid_Changes.values())\n",
    "            calls_TET2_missense_NP_01120680[k]=(v, changes)\n",
    "        #e.g. 12: (CGT, 'K', '*', 'T'etc...)\n",
    "        \n",
    "    #Changing the original amino acid to a Word (e.g. Phe) rather than a codon\n",
    "    calls_TET2_missense_NP_01120680_extra = {}\n",
    "\n",
    "    for k, v in calls_TET2_missense_NP_01120680.items():\n",
    "        if v[0] in Genetic_Code:\n",
    "            Amino_acid_change = Genetic_Code.get(v[0])\n",
    "            calls_TET2_missense_NP_01120680_extra[k] = (Amino_acid_change, v[1])\n",
    "            #e.g. 12: (Phe, 'K', '*', 'T' etc..)\n",
    "\n",
    "    #Changing the original amino acid to a LETTER (rather than the codon)\n",
    "    calls_TET2_missense_NP_01120680_extra2 = {}\n",
    "\n",
    "    for k, v in calls_TET2_missense_NP_01120680_extra.items():\n",
    "        if v[0] in Amino_acid_abbreviations:\n",
    "            Amino_acid_abbrev = Amino_acid_abbreviations.get(v[0])\n",
    "            calls_TET2_missense_NP_01120680_extra2[k] = (Amino_acid_abbrev, v[1])\n",
    "            #e.g. 12: ((F), ('K', '*', 'T' etc..))\n",
    "            \n",
    "    #Create a dictionary containing info on which position in the codon is affected to create the variant\n",
    "    TET2_position_in_list_dict_missense_NP_01120680={}\n",
    "    for k, v in TET2_missense_variant_DICT_NP_01120680.items(): #Iterates over the dictionary of (key = row_count, value = aa_pos, variant (e.g. 3: (76, p.V76L))\n",
    "        observed_change=(v[1])[-1] #Selects the amino acid change (e.g. the L from p.V76L)\n",
    "        possible_changes=calls_TET2_missense_NP_01120680[v[0]][1] #possible changes are the second value in the dictionary of (k = codon number, value = (original amino acid, possible changes))\n",
    "        possible_changes_info=calls_TET2_missense_NP_01120680[v[0]]\n",
    "        codon_seq=calls_TET2_missense_NP_01120680[v[0]][0]\n",
    "        possible_changes_string = \"\".join(possible_changes)\n",
    "        if v[1][2] == calls_TET2_missense_NP_01120680_extra2[v[0]][0]:  #v[1][2] = the original amino acid letter (i.e if this aa is in this transcript)\n",
    "            position_in_list = -1\n",
    "            positions_in_list = []\n",
    "            while True:\n",
    "                position_in_list = possible_changes_string.find(observed_change, position_in_list+1) #counts position in possible changes (if 0-2 = 1st base of codon, if 3-5 = 2nd base, if 6-8 = 3rd base)\n",
    "                if position_in_list == -1: break\n",
    "                positions_in_list.append(position_in_list)\n",
    "                TET2_position_in_list_dict_missense_NP_01120680[k]=(positions_in_list, possible_changes_info, codon_seq, v[0], v[1])\n",
    "            \n",
    "    aa_positions_variants_called = []\n",
    "    for k, v in TET2_position_in_list_dict_missense_NP_01120680.items():\n",
    "        aa_positions_variants_called.append(v[3])\n",
    "#     print('positions at which missense variants called = ', aa_positions_variants_called)\n",
    "                \n",
    "    #Create a dictionary that contains the position in the list of possible AA where the variant is and what it changes to\n",
    "    TET2_bases_missense_NP_01120680 = {}\n",
    "    for k, v in TET2_position_in_list_dict_missense_NP_01120680.items():\n",
    "        codon_seq = v[2]\n",
    "        base_changes=[]\n",
    "        if codon_seq in Possible_Base_Changes.keys():\n",
    "            list_of_possible_bases = Possible_Base_Changes.get(codon_seq)\n",
    "        for i in v[0]:\n",
    "            base_change_position = i\n",
    "            base_change = list_of_possible_bases[base_change_position]\n",
    "            base_changes.append(base_change)\n",
    "        TET2_bases_missense_NP_01120680[k]=(v, base_changes)\n",
    "        \n",
    "    #Create dictionary which contains original codon seq and details on which base is changed\n",
    "    TET2_bases_missense2_NP_01120680 = {}\n",
    "\n",
    "    for k, v in TET2_bases_missense_NP_01120680.items():\n",
    "        base_change_positions = v[0][0]\n",
    "        codon_positions = []\n",
    "        positions_in_gene = []\n",
    "        for i in base_change_positions:\n",
    "            base_change_position = i\n",
    "            if base_change_position <= 2:\n",
    "                codon_pos = 1\n",
    "            if 2 < base_change_position <= 5:\n",
    "                codon_pos = 2\n",
    "            if 5 < base_change_position <= 8:\n",
    "                codon_pos = 3\n",
    "            codon_positions.append(codon_pos)\n",
    "        TET2_bases_missense2_NP_01120680[k] = (v, codon_positions)\n",
    "        \n",
    "    #Create dictionary which contains original codon seq, details on which base is changed and it's position in the coding region\n",
    "    TET2_bases_missense3_NP_01120680 = {}\n",
    "\n",
    "    for k, v in TET2_bases_missense2_NP_01120680.items():\n",
    "        codon_positions = v[1]\n",
    "        base_changes = v[0][1]\n",
    "        original_bases = []\n",
    "        positions_in_gene = []\n",
    "        codon_number = v[0][0][3]\n",
    "        variant = v[0][0][4]\n",
    "        for i in codon_positions:\n",
    "            codon_pos = i\n",
    "            position_in_gene = ((int(v[0][0][3])-1)*3)+codon_pos #codon number (v[0][0][3]) minus 1 (e.g. codon 307 -1 = 306, then times by 3 takes you to beginning of codon 307 (then add the position))  \n",
    "            positions_in_gene.append(position_in_gene)\n",
    "            original_base = (v[0][0][2][codon_pos-1])\n",
    "            original_bases.append(original_base)\n",
    "        TET2_bases_missense3_NP_01120680[k] = (v[0][0][1], positions_in_gene, original_bases, base_changes, codon_number, variant)\n",
    "\n",
    "    #Create dictionary which contains original codon seq, details on which base is changed and its neighbouring bases\n",
    "    TET2_bases_missense4_NP_01120680 = {}\n",
    "\n",
    "    for k, v in TET2_bases_missense3_NP_01120680.items():\n",
    "        position_in_gene = v[1]\n",
    "        sites_trio_missense=[]\n",
    "\n",
    "        for i in position_in_gene:\n",
    "    #         print(i)\n",
    "            if i in intron_exon_boundary_caution_base_TET2_NP01120680:\n",
    "#                 print('positions at intron_exon_boundary =',i)\n",
    "                adapted_trio = intron_exon_check_TET2_NP01120680(i)\n",
    "                sites_trio_missense.append([adapted_trio])\n",
    "                TET2_bases_missense4_NP_01120680[k] = (v, sites_trio_missense)\n",
    "\n",
    "            else:\n",
    "                for i in position_in_gene:\n",
    "                    gene_pos = i\n",
    "                    site_trio_missense = [TET2_exons_extra_bases_NP_01120680[gene_pos:gene_pos+3] for gene_pos in range (gene_pos-1, gene_pos, 3)]\n",
    "                    sites_trio_missense.append(site_trio_missense)\n",
    "                    TET2_bases_missense4_NP_01120680[k] = (v, sites_trio_missense)\n",
    "                    \n",
    "    #Create a dictionary containing pos in coding region, original base, new base, base affected with its neighbours\n",
    "    #e.g. 938, G, A, TGG = base 938 on coding region, original base = G, changes to base A, base with neighbours either side = TGG\n",
    "    TET2_sites_missense5_NP_01120680 = {}\n",
    "\n",
    "    a=0\n",
    "    for k, v in TET2_bases_missense4_NP_01120680.items():\n",
    "        codon_n = v[0][1] #base number in codon (out of 3 x 912)\n",
    "        codon_n_912 = v[0][4]\n",
    "        original_base = v[0][2]\n",
    "        new_base = v[0][3]\n",
    "        site_trio = v[1]\n",
    "        info=[]\n",
    "        variant = v[0][5]\n",
    "\n",
    "        change = 0\n",
    "        for i in codon_n:\n",
    "            codon_number = i\n",
    "            orig_base = original_base[change]\n",
    "            n_base = new_base[change]\n",
    "            site_tri = site_trio[change]\n",
    "            info.append([codon_number, orig_base, n_base, site_tri, codon_n_912, variant])\n",
    "            change = change+1\n",
    "        TET2_sites_missense5_NP_01120680[a]=info\n",
    "        a=a+1\n",
    "        \n",
    "    # Write in a way that can iterate over mutation rates (i.e. put in to form C[C>A]G etc)   \n",
    "    TET2_sites_missense6_NP_01120680={}\n",
    "\n",
    "    separating_positions = {} # separate out all the possible changes to just 1 effective item in the list\n",
    "    a=0\n",
    "    for k, v in TET2_sites_missense5_NP_01120680.items():\n",
    "        for i in v: #because some contain more than 1 change from that codon:\n",
    "            separating_positions[a]=i\n",
    "            a=a+1\n",
    "    \n",
    "    for k, v in separating_positions.items():\n",
    "        base_pos = v[0]\n",
    "        original_base = v[1]\n",
    "        new_base = v[2]\n",
    "        site_trio = v[3]\n",
    "        codon_n_912 = v[4]\n",
    "        variant = v[5]\n",
    "        sequence = (str(original_base),str(new_base))\n",
    "        changes = '>'.join(sequence)\n",
    "        left_base = site_trio[0][0]\n",
    "        right_base = site_trio[0][2]\n",
    "        sequence2 = (changes, right_base)\n",
    "        changes2 = ']'.join(sequence2)\n",
    "        sequence3 = (left_base, changes2)\n",
    "        changes3 = '['.join(sequence3)\n",
    "        TET2_sites_missense6_NP_01120680[k] = (v[0],changes3, codon_n_912, variant) #site_trio = the affected base and its neighbours\n",
    "        \n",
    "        \n",
    "    #Assign the mutation rates to each site...\n",
    "    TET2_mutation_rates_missense_NP_01120680 = {}\n",
    "    for k, v in TET2_sites_missense6_NP_01120680.items():\n",
    "        base_change_info = v[1]\n",
    "        if base_change_info in Mutation_rates.keys():\n",
    "            mutation_rate = Mutation_rates.get(base_change_info)\n",
    "            TET2_mutation_rates_missense_NP_01120680[k] = (v, mutation_rate)\n",
    "\n",
    "    TET2_mutation_rate_missense_NP_01120680_sum = []\n",
    "    mutation_rates_list = []\n",
    "    for k, v in TET2_mutation_rates_missense_NP_01120680.items():\n",
    "    #     print(v[1])\n",
    "        aa_pos = v[0][2]\n",
    "        site = v[0][1]\n",
    "        mutation_rate_value = v[1] \n",
    "        variant = v[0][3]\n",
    "        mutation_rates_list.append((aa_pos, site, mutation_rate_value, variant))\n",
    "        TET2_mutation_rate_missense_NP_01120680_sum.append(mutation_rate_value)\n",
    "        \n",
    "    TET2_Summed_mutation_rate_missense_NP_01120680_added = np.sum(TET2_mutation_rate_missense_NP_01120680_sum)\n",
    "\n",
    "    print('Total summed mutation rate across TET2 missense_NP_01120680 for this study = ', TET2_Summed_mutation_rate_missense_NP_01120680_added)\n",
    "    \n",
    "    mutation_rates_list_all = mutation_rates_list\n",
    "    return mutation_rates_list_all #in format: column 1 = 882, column 2 = p.R882H  #2002 aa"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 292,
   "metadata": {
    "code_folding": [
     0
    ]
   },
   "outputs": [],
   "source": [
    "def mutation_rates_list_nonsense_variants_TET2_shorter(csv_file_variants):\n",
    "\n",
    "    #The mutation rate at a base is calculated based on its trinucleotide context.  To obtain the trinucleotide context, the base either side of the mutated base needs to be ascertained.\n",
    "    #The coding sequence (+ an extra base at each end) of each gene is used when determining the trinucleotide context of the base change.  The problem with using the coding sequence is that\n",
    "    #if a base change falls at the end of an exon, its neighbouring base will actuall be an intronic base, which may be different to the 'neighbouring' base in the coding sequence.\n",
    "    #To account for this, there is an \"intron_exon_check\" function that is called at the intron/exon boundary if the neighbouring base in the intron is different to the neighbouring base in the coding sequence.\n",
    "    \n",
    "    #TET2 transcript:\n",
    "    #Downloaded from http://grch37.ensembl.org/Homo_sapiens/Transcript/Summary?db=core;g=ENSG00000168769;r=4:106067032-106200973;t=ENST00000305737\n",
    "    TET2_exons_NP_060098 = 'ATGGAACAGGATAGAACCAACCATGTTGAGGGCAACAGACTAAGTCCATTCCTGATACCATCACCTCCCATTTGCCAGACAGAACCTCTGGCTACAAAGCTCCAGAATGGAAGCCCACTGCCTGAGAGAGCTCATCCAGAAGTAAATGGAGACACCAAGTGGCACTCTTTCAAAAGTTATTATGGAATACCCTGTATGAAGGGAAGCCAGAATAGTCGTGTGAGTCCTGACTTTACACAAGAAAGTAGAGGGTATTCCAAGTGTTTGCAAAATGGAGGAATAAAACGCACAGTTAGTGAACCTTCTCTCTCTGGGCTCCTTCAGATCAAGAAATTGAAACAAGACCAAAAGGCTAATGGAGAAAGACGTAACTTCGGGGTAAGCCAAGAAAGAAATCCAGGTGAAAGCAGTCAACCAAATGTCTCCGATTTGAGTGATAAGAAAGAATCTGTGAGTTCTGTAGCCCAAGAAAATGCAGTTAAAGATTTCACCAGTTTTTCAACACATAACTGCAGTGGGCCTGAAAATCCAGAGCTTCAGATTCTGAATGAGCAGGAGGGGAAAAGTGCTAATTACCATGACAAGAACATTGTATTACTTAAAAACAAGGCAGTGCTAATGCCTAATGGTGCTACAGTTTCTGCCTCTTCCGTGGAACACACACATGGTGAACTCCTGGAAAAAACACTGTCTCAATATTATCCAGATTGTGTTTCCATTGCGGTGCAGAAAACCACATCTCACATAAATGCCATTAACAGTCAGGCTACTAATGAGTTGTCCTGTGAGATCACTCACCCATCGCATACCTCAGGGCAGATCAATTCCGCACAGACCTCTAACTCTGAGCTGCCTCCAAAGCCAGCTGCAGTGGTGAGTGAGGCCTGTGATGCTGATGATGCTGATAATGCCAGTAAACTAGCTGCAATGCTAAATACCTGTTCCTTTCAGAAACCAGAACAACTACAACAACAAAAATCAGTTTTTGAGATATGCCCATCTCCTGCAGAAAATAACATCCAGGGAACCACAAAGCTAGCGTCTGGTGAAGAATTCTGTTCAGGTTCCAGCAGCAATTTGCAAGCTCCTGGTGGCAGCTCTGAACGGTATTTAAAACAAAATGAAATGAATGGTGCTTACTTCAAGCAAAGCTCAGTGTTCACTAAGGATTCCTTTTCTGCCACTACCACACCACCACCACCATCACAATTGCTTCTTTCTCCCCCTCCTCCTCTTCCACAGGTTCCTCAGCTTCCTTCAGAAGGAAAAAGCACTCTGAATGGTGGAGTTTTAGAAGAACACCACCACTACCCCAACCAAAGTAACACAACACTTTTAAGGGAAGTGAAAATAGAGGGTAAACCTGAGGCACCACCTTCCCAGAGTCCTAATCCATCTACACATGTATGCAGCCCTTCTCCGATGCTTTCTGAAAGGCCTCAGAATAATTGTGTGAACAGGAATGACATACAGACTGCAGGGACAATGACTGTTCCATTGTGTTCTGAGAAAACAAGACCAATGTCAGAACACCTCAAGCATAACCCACCAATTTTTGGTAGCAGTGGAGAGCTACAGGACAACTGCCAGCAGTTGATGAGAAACAAAGAGCAAGAGATTCTGAAGGGTCGAGACAAGGAGCAAACACGAGATCTTGTGCCCCCAACACAGCACTATCTGAAACCAGGATGGATTGAATTGAAGGCCCCTCGTTTTCACCAAGCGGAATCCCATCTAAAACGTAATGAGGCATCACTGCCATCAATTCTTCAGTATCAACCCAATCTCTCCAATCAAATGACCTCCAAACAATACACTGGAAATTCCAACATGCCTGGGGGGCTCCCAAGGCAAGCTTACACCCAGAAAACAACACAGCTGGAGCACAAGTCACAAATGTACCAAGTTGAAATGAATCAAGGGCAGTCCCAAGGTACAGTGGACCAACATCTCCAGTTCCAAAAACCCTCACACCAGGTGCACTTCTCCAAAACAGACCATTTACCAAAAGCTCATGTGCAGTCACTGTGTGGCACTAGATTTCATTTTCAACAAAGAGCAGATTCCCAAACTGAAAAACTTATGTCCCCAGTGTTGAAACAGCACTTGAATCAACAGGCTTCAGAGACTGAGCCATTTTCAAACTCACACCTTTTGCAACATAAGCCTCATAAACAGGCAGCACAAACACAACCATCCCAGAGTTCACATCTCCCTCAAAACCAGCAACAGCAGCAAAAATTACAAATAAAGAATAAAGAGGAAATACTCCAGACTTTTCCTCACCCCCAAAGCAACAATGATCAGCAAAGAGAAGGATCATTCTTTGGCCAGACTAAAGTGGAAGAATGTTTTCATGGTGAAAATCAGTATTCAAAATCAAGCGAGTTCGAGACTCATAATGTCCAAATGGGACTGGAGGAAGTACAGAATATAAATCGTAGAAATTCCCCTTATAGTCAGACCATGAAATCAAGTGCATGCAAAATACAGGTTTCTTGTTCAAACAATACACACCTAGTTTCAGAGAATAAAGAACAGACTACACATCCTGAACTTTTTGCAGGAAACAAGACCCAAAACTTGCATCACATGCAATATTTTCCAAATAATGTGATCCCAAAGCAAGATCTTCTTCACAGGTGCTTTCAAGAACAGGAGCAGAAGTCACAACAAGCTTCAGTTCTACAGGGATATAAAAATAGAAACCAAGATATGTCTGGTCAACAAGCTGCGCAACTTGCTCAGCAAAGGTACTTGATACATAACCATGCAAATGTTTTTCCTGTGCCTGACCAGGGAGGAAGTCACACTCAGACCCCTCCCCAGAAGGACACTCAAAAGCATGCTGCTCTAAGGTGGCATCTCTTACAGAAGCAAGAACAGCAGCAAACACAGCAACCCCAAACTGAGTCTTGCCATAGTCAGATGCACAGGCCAATTAAGGTGGAACCTGGATGCAAGCCACATGCCTGTATGCACACAGCACCACCAGAAAACAAAACATGGAAAAAGGTAACTAAGCAAGAGAATCCACCTGCAAGCTGTGATAATGTGCAGCAAAAGAGCATCATTGAGACCATGGAGCAGCATCTGAAGCAGTTTCACGCCAAGTCGTTATTTGACCATAAGGCTCTTACTCTCAAATCACAGAAGCAAGTAAAAGTTGAAATGTCAGGGCCAGTCACAGTTTTGACTAGACAAACCACTGCTGCAGAACTTGATAGCCACACCCCAGCTTTAGAGCAGCAAACAACTTCTTCAGAAAAGACACCAACCAAAAGAACAGCTGCTTCTGTTCTCAATAATTTTATAGAGTCACCTTCCAAATTACTAGATACTCCTATAAAAAATTTATTGGATACACCTGTCAAGACTCAATATGATTTCCCATCTTGCAGATGTGTAGGTAAGTGCCAGAAATGTACTGAGACACATGGCGTTTATCCAGAATTAGCAAATTTATCTTCAGATATGGGATTTTCCTTCTTTTTTTAA'\n",
    "\n",
    "    #Split the sequence into codons \n",
    "    TET2_codons_NP_060098 = [TET2_exons_NP_060098[i:i+3] for i in range(0, len(TET2_exons_NP_060098), 3)]\n",
    "\n",
    "    #TET2 transcript 1 extra base at each end:\n",
    "    #Downloaded from http://grch37.ensembl.org/Homo_sapiens/Transcript/Summary?db=core;g=ENSG00000168769;r=4:106067032-106200973;t=ENST00000305737\n",
    "    TET2_exons_extra_bases_NP_060098 = 'GATGGAACAGGATAGAACCAACCATGTTGAGGGCAACAGACTAAGTCCATTCCTGATACCATCACCTCCCATTTGCCAGACAGAACCTCTGGCTACAAAGCTCCAGAATGGAAGCCCACTGCCTGAGAGAGCTCATCCAGAAGTAAATGGAGACACCAAGTGGCACTCTTTCAAAAGTTATTATGGAATACCCTGTATGAAGGGAAGCCAGAATAGTCGTGTGAGTCCTGACTTTACACAAGAAAGTAGAGGGTATTCCAAGTGTTTGCAAAATGGAGGAATAAAACGCACAGTTAGTGAACCTTCTCTCTCTGGGCTCCTTCAGATCAAGAAATTGAAACAAGACCAAAAGGCTAATGGAGAAAGACGTAACTTCGGGGTAAGCCAAGAAAGAAATCCAGGTGAAAGCAGTCAACCAAATGTCTCCGATTTGAGTGATAAGAAAGAATCTGTGAGTTCTGTAGCCCAAGAAAATGCAGTTAAAGATTTCACCAGTTTTTCAACACATAACTGCAGTGGGCCTGAAAATCCAGAGCTTCAGATTCTGAATGAGCAGGAGGGGAAAAGTGCTAATTACCATGACAAGAACATTGTATTACTTAAAAACAAGGCAGTGCTAATGCCTAATGGTGCTACAGTTTCTGCCTCTTCCGTGGAACACACACATGGTGAACTCCTGGAAAAAACACTGTCTCAATATTATCCAGATTGTGTTTCCATTGCGGTGCAGAAAACCACATCTCACATAAATGCCATTAACAGTCAGGCTACTAATGAGTTGTCCTGTGAGATCACTCACCCATCGCATACCTCAGGGCAGATCAATTCCGCACAGACCTCTAACTCTGAGCTGCCTCCAAAGCCAGCTGCAGTGGTGAGTGAGGCCTGTGATGCTGATGATGCTGATAATGCCAGTAAACTAGCTGCAATGCTAAATACCTGTTCCTTTCAGAAACCAGAACAACTACAACAACAAAAATCAGTTTTTGAGATATGCCCATCTCCTGCAGAAAATAACATCCAGGGAACCACAAAGCTAGCGTCTGGTGAAGAATTCTGTTCAGGTTCCAGCAGCAATTTGCAAGCTCCTGGTGGCAGCTCTGAACGGTATTTAAAACAAAATGAAATGAATGGTGCTTACTTCAAGCAAAGCTCAGTGTTCACTAAGGATTCCTTTTCTGCCACTACCACACCACCACCACCATCACAATTGCTTCTTTCTCCCCCTCCTCCTCTTCCACAGGTTCCTCAGCTTCCTTCAGAAGGAAAAAGCACTCTGAATGGTGGAGTTTTAGAAGAACACCACCACTACCCCAACCAAAGTAACACAACACTTTTAAGGGAAGTGAAAATAGAGGGTAAACCTGAGGCACCACCTTCCCAGAGTCCTAATCCATCTACACATGTATGCAGCCCTTCTCCGATGCTTTCTGAAAGGCCTCAGAATAATTGTGTGAACAGGAATGACATACAGACTGCAGGGACAATGACTGTTCCATTGTGTTCTGAGAAAACAAGACCAATGTCAGAACACCTCAAGCATAACCCACCAATTTTTGGTAGCAGTGGAGAGCTACAGGACAACTGCCAGCAGTTGATGAGAAACAAAGAGCAAGAGATTCTGAAGGGTCGAGACAAGGAGCAAACACGAGATCTTGTGCCCCCAACACAGCACTATCTGAAACCAGGATGGATTGAATTGAAGGCCCCTCGTTTTCACCAAGCGGAATCCCATCTAAAACGTAATGAGGCATCACTGCCATCAATTCTTCAGTATCAACCCAATCTCTCCAATCAAATGACCTCCAAACAATACACTGGAAATTCCAACATGCCTGGGGGGCTCCCAAGGCAAGCTTACACCCAGAAAACAACACAGCTGGAGCACAAGTCACAAATGTACCAAGTTGAAATGAATCAAGGGCAGTCCCAAGGTACAGTGGACCAACATCTCCAGTTCCAAAAACCCTCACACCAGGTGCACTTCTCCAAAACAGACCATTTACCAAAAGCTCATGTGCAGTCACTGTGTGGCACTAGATTTCATTTTCAACAAAGAGCAGATTCCCAAACTGAAAAACTTATGTCCCCAGTGTTGAAACAGCACTTGAATCAACAGGCTTCAGAGACTGAGCCATTTTCAAACTCACACCTTTTGCAACATAAGCCTCATAAACAGGCAGCACAAACACAACCATCCCAGAGTTCACATCTCCCTCAAAACCAGCAACAGCAGCAAAAATTACAAATAAAGAATAAAGAGGAAATACTCCAGACTTTTCCTCACCCCCAAAGCAACAATGATCAGCAAAGAGAAGGATCATTCTTTGGCCAGACTAAAGTGGAAGAATGTTTTCATGGTGAAAATCAGTATTCAAAATCAAGCGAGTTCGAGACTCATAATGTCCAAATGGGACTGGAGGAAGTACAGAATATAAATCGTAGAAATTCCCCTTATAGTCAGACCATGAAATCAAGTGCATGCAAAATACAGGTTTCTTGTTCAAACAATACACACCTAGTTTCAGAGAATAAAGAACAGACTACACATCCTGAACTTTTTGCAGGAAACAAGACCCAAAACTTGCATCACATGCAATATTTTCCAAATAATGTGATCCCAAAGCAAGATCTTCTTCACAGGTGCTTTCAAGAACAGGAGCAGAAGTCACAACAAGCTTCAGTTCTACAGGGATATAAAAATAGAAACCAAGATATGTCTGGTCAACAAGCTGCGCAACTTGCTCAGCAAAGGTACTTGATACATAACCATGCAAATGTTTTTCCTGTGCCTGACCAGGGAGGAAGTCACACTCAGACCCCTCCCCAGAAGGACACTCAAAAGCATGCTGCTCTAAGGTGGCATCTCTTACAGAAGCAAGAACAGCAGCAAACACAGCAACCCCAAACTGAGTCTTGCCATAGTCAGATGCACAGGCCAATTAAGGTGGAACCTGGATGCAAGCCACATGCCTGTATGCACACAGCACCACCAGAAAACAAAACATGGAAAAAGGTAACTAAGCAAGAGAATCCACCTGCAAGCTGTGATAATGTGCAGCAAAAGAGCATCATTGAGACCATGGAGCAGCATCTGAAGCAGTTTCACGCCAAGTCGTTATTTGACCATAAGGCTCTTACTCTCAAATCACAGAAGCAAGTAAAAGTTGAAATGTCAGGGCCAGTCACAGTTTTGACTAGACAAACCACTGCTGCAGAACTTGATAGCCACACCCCAGCTTTAGAGCAGCAAACAACTTCTTCAGAAAAGACACCAACCAAAAGAACAGCTGCTTCTGTTCTCAATAATTTTATAGAGTCACCTTCCAAATTACTAGATACTCCTATAAAAAATTTATTGGATACACCTGTCAAGACTCAATATGATTTCCCATCTTGCAGATGTGTAGGTAAGTGCCAGAAATGTACTGAGACACATGGCGTTTATCCAGAATTAGCAAATTTATCTTCAGATATGGGATTTTCCTTCTTTTTTTAAA'\n",
    "\n",
    "    #Create a list of the codons they called nonsense variants at\n",
    "    import csv\n",
    "    filename = csv_file_variants\n",
    "    with open(filename, 'r') as csvfile:\n",
    "        read_reader = csv.reader(csvfile)  #csv.reader returns a reader object which will iterate over lines in the csvfile\n",
    "        row_count=0\n",
    "\n",
    "        TET2_nonsense_variant_list_NP_060098 = []\n",
    "        TET2_nonsense_codons_NP_060098 =[]\n",
    "        TET2_nonsense_variant_DICT_NP_060098 = {}\n",
    "\n",
    "        for row in read_reader: #each row contains aa number (e.g. 11) and variant (e.g. p.D11A)\n",
    "    #         print(row)\n",
    "            if row_count>0: \n",
    "                codon0_NP_060098 = int(row[0])-1\n",
    "                codon1_NP_060098 = int(row[0])\n",
    "                variant_NP_060098 = row[1]\n",
    "                if codon1_NP_060098 < 1167:\n",
    "                    TET2_nonsense_codons_NP_060098.append(codon0_NP_060098) #List of codon numbers that they called variants at\n",
    "                    TET2_nonsense_variant_DICT_NP_060098[row_count]=(codon1_NP_060098, variant_NP_060098) #Dictionary:  K = codon number, V = variant at that codon (e.g. 76: p.V76L)\n",
    "\n",
    "            row_count=row_count+1\n",
    "            \n",
    "    #Looking at what possible amino acid changes are possible at those codons\n",
    "    TET2_nonsense_codons_NP_060098_DICT = {}\n",
    "    calls_TET2_nonsense_NP_060098 = {}\n",
    "\n",
    "    for i in TET2_nonsense_codons_NP_060098: #Iterates over the list of codon numbers\n",
    "        TET2_nonsense_codons_NP_060098_DICT[i+1]= (TET2_codons_NP_060098[i]) #Creates of dicitonary of key = codon number, value = bases (sequence) at that codon\n",
    "        #e.g. 12: CGT, 882: GCA\n",
    "        \n",
    "    for k,v in TET2_nonsense_codons_NP_060098_DICT.items(): #iterating over the list of codon numbers (and codon sequences)\n",
    "        if v in Possible_Amino_Acid_Changes_letters.keys(): #assign the list of all the possible amino acid changes to each codon they called\n",
    "            changes = Possible_Amino_Acid_Changes_letters[v]\n",
    "    #         print(Possible_Amino_Acid_Changes.values())\n",
    "            calls_TET2_nonsense_NP_060098[k]=(v, changes)\n",
    "        #e.g. 12: (CGT, 'K', '*', 'T'etc...)\n",
    "        \n",
    "    #Changing the original amino acid to a Word (e.g. Phe) rather than a codon\n",
    "    calls_TET2_nonsense_NP_060098_extra = {}\n",
    "\n",
    "    for k, v in calls_TET2_nonsense_NP_060098.items():\n",
    "        if v[0] in Genetic_Code:\n",
    "            Amino_acid_change = Genetic_Code.get(v[0])\n",
    "            calls_TET2_nonsense_NP_060098_extra[k] = (Amino_acid_change, v[1])\n",
    "            #e.g. 12: (Phe, 'K', '*', 'T' etc..)\n",
    "\n",
    "    #Changing the original amino acid to a LETTER (rather than the codon)\n",
    "    calls_TET2_nonsense_NP_060098_extra2 = {}\n",
    "\n",
    "    for k, v in calls_TET2_nonsense_NP_060098_extra.items():\n",
    "        if v[0] in Amino_acid_abbreviations:\n",
    "            Amino_acid_abbrev = Amino_acid_abbreviations.get(v[0])\n",
    "            calls_TET2_nonsense_NP_060098_extra2[k] = (Amino_acid_abbrev, v[1])\n",
    "            #e.g. 12: ((F), ('K', '*', 'T' etc..))\n",
    "            \n",
    "    #Create a dictionary containing info on which position in the codon is affected to create the variant\n",
    "    TET2_position_in_list_dict_nonsense_NP_060098={}\n",
    "    for k, v in TET2_nonsense_variant_DICT_NP_060098.items(): #Iterates over the dictionary of (key = row_count, value = aa_pos, variant (e.g. 3: (76, p.V76L))\n",
    "        observed_change=(v[1])[-1] #Selects the amino acid change (e.g. the L from p.V76L)\n",
    "        possible_changes=calls_TET2_nonsense_NP_060098[v[0]][1] #possible changes are the second value in the dictionary of (k = codon number, value = (original amino acid, possible changes))\n",
    "        possible_changes_info=calls_TET2_nonsense_NP_060098[v[0]]\n",
    "        codon_seq=calls_TET2_nonsense_NP_060098[v[0]][0]\n",
    "        possible_changes_string = \"\".join(possible_changes)\n",
    "        if v[1][2] == calls_TET2_nonsense_NP_060098_extra2[v[0]][0]:  #v[1][2] = the original amino acid letter (i.e if this aa is in this transcript)\n",
    "            position_in_list = -1\n",
    "            positions_in_list = []\n",
    "            while True:\n",
    "                position_in_list = possible_changes_string.find(observed_change, position_in_list+1) #counts position in possible changes (if 0-2 = 1st base of codon, if 3-5 = 2nd base, if 6-8 = 3rd base)\n",
    "                if position_in_list == -1: break\n",
    "                positions_in_list.append(position_in_list)\n",
    "                TET2_position_in_list_dict_nonsense_NP_060098[k]=(positions_in_list, possible_changes_info, codon_seq, v[0], v[1])\n",
    "            \n",
    "    aa_positions_variants_called = []\n",
    "    for k, v in TET2_position_in_list_dict_nonsense_NP_060098.items():\n",
    "        aa_positions_variants_called.append(v[3])\n",
    "#     print('positions at which nonsense variants called = ', aa_positions_variants_called)\n",
    "                \n",
    "    #Create a dictionary that contains the position in the list of possible AA where the variant is and what it changes to\n",
    "    TET2_bases_nonsense_NP_060098 = {}\n",
    "    for k, v in TET2_position_in_list_dict_nonsense_NP_060098.items():\n",
    "        codon_seq = v[2]\n",
    "        base_changes=[]\n",
    "        if codon_seq in Possible_Base_Changes.keys():\n",
    "            list_of_possible_bases = Possible_Base_Changes.get(codon_seq)\n",
    "        for i in v[0]:\n",
    "            base_change_position = i\n",
    "            base_change = list_of_possible_bases[base_change_position]\n",
    "            base_changes.append(base_change)\n",
    "        TET2_bases_nonsense_NP_060098[k]=(v, base_changes)\n",
    "        \n",
    "    #Create dictionary which contains original codon seq and details on which base is changed\n",
    "    TET2_bases_nonsense2_NP_060098 = {}\n",
    "\n",
    "    for k, v in TET2_bases_nonsense_NP_060098.items():\n",
    "        base_change_positions = v[0][0]\n",
    "        codon_positions = []\n",
    "        positions_in_gene = []\n",
    "        for i in base_change_positions:\n",
    "            base_change_position = i\n",
    "            if base_change_position <= 2:\n",
    "                codon_pos = 1\n",
    "            if 2 < base_change_position <= 5:\n",
    "                codon_pos = 2\n",
    "            if 5 < base_change_position <= 8:\n",
    "                codon_pos = 3\n",
    "            codon_positions.append(codon_pos)\n",
    "        TET2_bases_nonsense2_NP_060098[k] = (v, codon_positions)\n",
    "        \n",
    "    #Create dictionary which contains original codon seq, details on which base is changed and it's position in the coding region\n",
    "    TET2_bases_nonsense3_NP_060098 = {}\n",
    "\n",
    "    for k, v in TET2_bases_nonsense2_NP_060098.items():\n",
    "        codon_positions = v[1]\n",
    "        base_changes = v[0][1]\n",
    "        original_bases = []\n",
    "        positions_in_gene = []\n",
    "        codon_number = v[0][0][3]\n",
    "        variant = v[0][0][4]\n",
    "        for i in codon_positions:\n",
    "            codon_pos = i\n",
    "            position_in_gene = ((int(v[0][0][3])-1)*3)+codon_pos #codon number (v[0][0][3]) minus 1 (e.g. codon 307 -1 = 306, then times by 3 takes you to beginning of codon 307 (then add the position))  \n",
    "            positions_in_gene.append(position_in_gene)\n",
    "            original_base = (v[0][0][2][codon_pos-1])\n",
    "            original_bases.append(original_base)\n",
    "        TET2_bases_nonsense3_NP_060098[k] = (v[0][0][1], positions_in_gene, original_bases, base_changes, codon_number, variant)\n",
    "\n",
    "    #Create dictionary which contains original codon seq, details on which base is changed and its neighbouring bases\n",
    "    TET2_bases_nonsense4_NP_060098 = {}\n",
    "\n",
    "    for k, v in TET2_bases_nonsense3_NP_060098.items():\n",
    "        position_in_gene = v[1]\n",
    "        sites_trio_nonsense=[]\n",
    "\n",
    "        for i in position_in_gene:\n",
    "            gene_pos = i\n",
    "            site_trio_nonsense = [TET2_exons_extra_bases_NP_060098[gene_pos:gene_pos+3] for gene_pos in range (gene_pos-1, gene_pos, 3)]\n",
    "            sites_trio_nonsense.append(site_trio_nonsense)\n",
    "            TET2_bases_nonsense4_NP_060098[k] = (v, sites_trio_nonsense)\n",
    "                    \n",
    "    #Create a dictionary containing pos in coding region, original base, new base, base affected with its neighbours\n",
    "    #e.g. 938, G, A, TGG = base 938 on coding region, original base = G, changes to base A, base with neighbours either side = TGG\n",
    "    TET2_sites_nonsense5_NP_060098 = {}\n",
    "\n",
    "    a=0\n",
    "    for k, v in TET2_bases_nonsense4_NP_060098.items():\n",
    "        codon_n = v[0][1] #base number in codon (out of 3 x 912)\n",
    "        codon_n_912 = v[0][4]\n",
    "        original_base = v[0][2]\n",
    "        new_base = v[0][3]\n",
    "        site_trio = v[1]\n",
    "        info=[]\n",
    "        variant = v[0][5]\n",
    "\n",
    "        change = 0\n",
    "        for i in codon_n:\n",
    "            codon_number = i\n",
    "            orig_base = original_base[change]\n",
    "            n_base = new_base[change]\n",
    "            site_tri = site_trio[change]\n",
    "            info.append([codon_number, orig_base, n_base, site_tri, codon_n_912, variant])\n",
    "            change = change+1\n",
    "        TET2_sites_nonsense5_NP_060098[a]=info\n",
    "        a=a+1\n",
    "#     print(TET2_sites_nonsense5_NP_060098)\n",
    "        \n",
    "    # Write in a way that can iterate over mutation rates (i.e. put in to form C[C>A]G etc)   \n",
    "    TET2_sites_nonsense6_NP_060098={}\n",
    "\n",
    "    separating_positions = {} # separate out all the possible changes to just 1 effective item in the list\n",
    "    a=0\n",
    "    for k, v in TET2_sites_nonsense5_NP_060098.items():\n",
    "        for i in v: #because some contain more than 1 change from that codon:\n",
    "            separating_positions[a]=i\n",
    "            a=a+1\n",
    "    \n",
    "    for k, v in separating_positions.items():\n",
    "        base_pos = v[0]\n",
    "        original_base = v[1]\n",
    "        new_base = v[2]\n",
    "        site_trio = v[3]\n",
    "        codon_n_912 = v[4]\n",
    "        variant = v[5]\n",
    "        sequence = (str(original_base),str(new_base))\n",
    "        changes = '>'.join(sequence)\n",
    "        left_base = site_trio[0][0]\n",
    "        right_base = site_trio[0][2]\n",
    "        sequence2 = (changes, right_base)\n",
    "        changes2 = ']'.join(sequence2)\n",
    "        sequence3 = (left_base, changes2)\n",
    "        changes3 = '['.join(sequence3)\n",
    "        TET2_sites_nonsense6_NP_060098[k] = (v[0],changes3, codon_n_912, variant) #site_trio = the affected base and its neighbours\n",
    "        \n",
    "        \n",
    "    #Assign the mutation rates to each site...\n",
    "    TET2_mutation_rates_nonsense_NP_060098 = {}\n",
    "    for k, v in TET2_sites_nonsense6_NP_060098.items():\n",
    "        base_change_info = v[1]\n",
    "        if base_change_info in Mutation_rates.keys():\n",
    "            mutation_rate = Mutation_rates.get(base_change_info)\n",
    "            TET2_mutation_rates_nonsense_NP_060098[k] = (v, mutation_rate)\n",
    "\n",
    "    TET2_mutation_rate_nonsense_NP_060098_sum = []\n",
    "    mutation_rates_list = []\n",
    "    for k, v in TET2_mutation_rates_nonsense_NP_060098.items():\n",
    "    #     print(v[1])\n",
    "        aa_pos = v[0][2]\n",
    "        site = v[0][1]\n",
    "        mutation_rate_value = v[1] \n",
    "        variant = v[0][3]\n",
    "        mutation_rates_list.append((aa_pos, site, mutation_rate_value, variant))\n",
    "        TET2_mutation_rate_nonsense_NP_060098_sum.append(mutation_rate_value)\n",
    "        \n",
    "    TET2_Summed_mutation_rate_nonsense_NP_060098_added = np.sum(TET2_mutation_rate_nonsense_NP_060098_sum)\n",
    "\n",
    "    print('Total summed mutation rate across TET2 nonsense_NP_060098 for this study = ', TET2_Summed_mutation_rate_nonsense_NP_060098_added)\n",
    "    \n",
    "    mutation_rates_list_all = mutation_rates_list\n",
    "    return mutation_rates_list_all #in format: column 1 = 882, column 2 = p.R882*  #1165 aa"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 293,
   "metadata": {
    "code_folding": [
     0
    ]
   },
   "outputs": [],
   "source": [
    "def mutation_rates_list_missense_variants_TET2_shorter(csv_file_variants):\n",
    "    \n",
    "    #The mutation rate at a base is calculated based on its trinucleotide context.  To obtain the trinucleotide context, the base either side of the mutated base needs to be ascertained.\n",
    "    #The coding sequence (+ an extra base at each end) of each gene is used when determining the trinucleotide context of the base change.  The problem with using the coding sequence is that\n",
    "    #if a base change falls at the end of an exon, its neighbouring base will actuall be an intronic base, which may be different to the 'neighbouring' base in the coding sequence.\n",
    "    #To account for this, there is an \"intron_exon_check\" function that is called at the intron/exon boundary if the neighbouring base in the intron is different to the neighbouring base in the coding sequence.\n",
    "    \n",
    "    #TET2 transcript:\n",
    "    #Downloaded from http://grch37.ensembl.org/Homo_sapiens/Transcript/Summary?db=core;g=ENSG00000168769;r=4:106067032-106200973;t=ENST00000305737\n",
    "    TET2_exons_NP_060098 = 'ATGGAACAGGATAGAACCAACCATGTTGAGGGCAACAGACTAAGTCCATTCCTGATACCATCACCTCCCATTTGCCAGACAGAACCTCTGGCTACAAAGCTCCAGAATGGAAGCCCACTGCCTGAGAGAGCTCATCCAGAAGTAAATGGAGACACCAAGTGGCACTCTTTCAAAAGTTATTATGGAATACCCTGTATGAAGGGAAGCCAGAATAGTCGTGTGAGTCCTGACTTTACACAAGAAAGTAGAGGGTATTCCAAGTGTTTGCAAAATGGAGGAATAAAACGCACAGTTAGTGAACCTTCTCTCTCTGGGCTCCTTCAGATCAAGAAATTGAAACAAGACCAAAAGGCTAATGGAGAAAGACGTAACTTCGGGGTAAGCCAAGAAAGAAATCCAGGTGAAAGCAGTCAACCAAATGTCTCCGATTTGAGTGATAAGAAAGAATCTGTGAGTTCTGTAGCCCAAGAAAATGCAGTTAAAGATTTCACCAGTTTTTCAACACATAACTGCAGTGGGCCTGAAAATCCAGAGCTTCAGATTCTGAATGAGCAGGAGGGGAAAAGTGCTAATTACCATGACAAGAACATTGTATTACTTAAAAACAAGGCAGTGCTAATGCCTAATGGTGCTACAGTTTCTGCCTCTTCCGTGGAACACACACATGGTGAACTCCTGGAAAAAACACTGTCTCAATATTATCCAGATTGTGTTTCCATTGCGGTGCAGAAAACCACATCTCACATAAATGCCATTAACAGTCAGGCTACTAATGAGTTGTCCTGTGAGATCACTCACCCATCGCATACCTCAGGGCAGATCAATTCCGCACAGACCTCTAACTCTGAGCTGCCTCCAAAGCCAGCTGCAGTGGTGAGTGAGGCCTGTGATGCTGATGATGCTGATAATGCCAGTAAACTAGCTGCAATGCTAAATACCTGTTCCTTTCAGAAACCAGAACAACTACAACAACAAAAATCAGTTTTTGAGATATGCCCATCTCCTGCAGAAAATAACATCCAGGGAACCACAAAGCTAGCGTCTGGTGAAGAATTCTGTTCAGGTTCCAGCAGCAATTTGCAAGCTCCTGGTGGCAGCTCTGAACGGTATTTAAAACAAAATGAAATGAATGGTGCTTACTTCAAGCAAAGCTCAGTGTTCACTAAGGATTCCTTTTCTGCCACTACCACACCACCACCACCATCACAATTGCTTCTTTCTCCCCCTCCTCCTCTTCCACAGGTTCCTCAGCTTCCTTCAGAAGGAAAAAGCACTCTGAATGGTGGAGTTTTAGAAGAACACCACCACTACCCCAACCAAAGTAACACAACACTTTTAAGGGAAGTGAAAATAGAGGGTAAACCTGAGGCACCACCTTCCCAGAGTCCTAATCCATCTACACATGTATGCAGCCCTTCTCCGATGCTTTCTGAAAGGCCTCAGAATAATTGTGTGAACAGGAATGACATACAGACTGCAGGGACAATGACTGTTCCATTGTGTTCTGAGAAAACAAGACCAATGTCAGAACACCTCAAGCATAACCCACCAATTTTTGGTAGCAGTGGAGAGCTACAGGACAACTGCCAGCAGTTGATGAGAAACAAAGAGCAAGAGATTCTGAAGGGTCGAGACAAGGAGCAAACACGAGATCTTGTGCCCCCAACACAGCACTATCTGAAACCAGGATGGATTGAATTGAAGGCCCCTCGTTTTCACCAAGCGGAATCCCATCTAAAACGTAATGAGGCATCACTGCCATCAATTCTTCAGTATCAACCCAATCTCTCCAATCAAATGACCTCCAAACAATACACTGGAAATTCCAACATGCCTGGGGGGCTCCCAAGGCAAGCTTACACCCAGAAAACAACACAGCTGGAGCACAAGTCACAAATGTACCAAGTTGAAATGAATCAAGGGCAGTCCCAAGGTACAGTGGACCAACATCTCCAGTTCCAAAAACCCTCACACCAGGTGCACTTCTCCAAAACAGACCATTTACCAAAAGCTCATGTGCAGTCACTGTGTGGCACTAGATTTCATTTTCAACAAAGAGCAGATTCCCAAACTGAAAAACTTATGTCCCCAGTGTTGAAACAGCACTTGAATCAACAGGCTTCAGAGACTGAGCCATTTTCAAACTCACACCTTTTGCAACATAAGCCTCATAAACAGGCAGCACAAACACAACCATCCCAGAGTTCACATCTCCCTCAAAACCAGCAACAGCAGCAAAAATTACAAATAAAGAATAAAGAGGAAATACTCCAGACTTTTCCTCACCCCCAAAGCAACAATGATCAGCAAAGAGAAGGATCATTCTTTGGCCAGACTAAAGTGGAAGAATGTTTTCATGGTGAAAATCAGTATTCAAAATCAAGCGAGTTCGAGACTCATAATGTCCAAATGGGACTGGAGGAAGTACAGAATATAAATCGTAGAAATTCCCCTTATAGTCAGACCATGAAATCAAGTGCATGCAAAATACAGGTTTCTTGTTCAAACAATACACACCTAGTTTCAGAGAATAAAGAACAGACTACACATCCTGAACTTTTTGCAGGAAACAAGACCCAAAACTTGCATCACATGCAATATTTTCCAAATAATGTGATCCCAAAGCAAGATCTTCTTCACAGGTGCTTTCAAGAACAGGAGCAGAAGTCACAACAAGCTTCAGTTCTACAGGGATATAAAAATAGAAACCAAGATATGTCTGGTCAACAAGCTGCGCAACTTGCTCAGCAAAGGTACTTGATACATAACCATGCAAATGTTTTTCCTGTGCCTGACCAGGGAGGAAGTCACACTCAGACCCCTCCCCAGAAGGACACTCAAAAGCATGCTGCTCTAAGGTGGCATCTCTTACAGAAGCAAGAACAGCAGCAAACACAGCAACCCCAAACTGAGTCTTGCCATAGTCAGATGCACAGGCCAATTAAGGTGGAACCTGGATGCAAGCCACATGCCTGTATGCACACAGCACCACCAGAAAACAAAACATGGAAAAAGGTAACTAAGCAAGAGAATCCACCTGCAAGCTGTGATAATGTGCAGCAAAAGAGCATCATTGAGACCATGGAGCAGCATCTGAAGCAGTTTCACGCCAAGTCGTTATTTGACCATAAGGCTCTTACTCTCAAATCACAGAAGCAAGTAAAAGTTGAAATGTCAGGGCCAGTCACAGTTTTGACTAGACAAACCACTGCTGCAGAACTTGATAGCCACACCCCAGCTTTAGAGCAGCAAACAACTTCTTCAGAAAAGACACCAACCAAAAGAACAGCTGCTTCTGTTCTCAATAATTTTATAGAGTCACCTTCCAAATTACTAGATACTCCTATAAAAAATTTATTGGATACACCTGTCAAGACTCAATATGATTTCCCATCTTGCAGATGTGTAGGTAAGTGCCAGAAATGTACTGAGACACATGGCGTTTATCCAGAATTAGCAAATTTATCTTCAGATATGGGATTTTCCTTCTTTTTTTAA'\n",
    "\n",
    "    #Split the sequence into codons \n",
    "    TET2_codons_NP_060098 = [TET2_exons_NP_060098[i:i+3] for i in range(0, len(TET2_exons_NP_060098), 3)]\n",
    "\n",
    "    #TET2 transcript 1 extra base at each end:\n",
    "    #Downloaded from http://grch37.ensembl.org/Homo_sapiens/Transcript/Summary?db=core;g=ENSG00000168769;r=4:106067032-106200973;t=ENST00000305737\n",
    "    TET2_exons_extra_bases_NP_060098 = 'GATGGAACAGGATAGAACCAACCATGTTGAGGGCAACAGACTAAGTCCATTCCTGATACCATCACCTCCCATTTGCCAGACAGAACCTCTGGCTACAAAGCTCCAGAATGGAAGCCCACTGCCTGAGAGAGCTCATCCAGAAGTAAATGGAGACACCAAGTGGCACTCTTTCAAAAGTTATTATGGAATACCCTGTATGAAGGGAAGCCAGAATAGTCGTGTGAGTCCTGACTTTACACAAGAAAGTAGAGGGTATTCCAAGTGTTTGCAAAATGGAGGAATAAAACGCACAGTTAGTGAACCTTCTCTCTCTGGGCTCCTTCAGATCAAGAAATTGAAACAAGACCAAAAGGCTAATGGAGAAAGACGTAACTTCGGGGTAAGCCAAGAAAGAAATCCAGGTGAAAGCAGTCAACCAAATGTCTCCGATTTGAGTGATAAGAAAGAATCTGTGAGTTCTGTAGCCCAAGAAAATGCAGTTAAAGATTTCACCAGTTTTTCAACACATAACTGCAGTGGGCCTGAAAATCCAGAGCTTCAGATTCTGAATGAGCAGGAGGGGAAAAGTGCTAATTACCATGACAAGAACATTGTATTACTTAAAAACAAGGCAGTGCTAATGCCTAATGGTGCTACAGTTTCTGCCTCTTCCGTGGAACACACACATGGTGAACTCCTGGAAAAAACACTGTCTCAATATTATCCAGATTGTGTTTCCATTGCGGTGCAGAAAACCACATCTCACATAAATGCCATTAACAGTCAGGCTACTAATGAGTTGTCCTGTGAGATCACTCACCCATCGCATACCTCAGGGCAGATCAATTCCGCACAGACCTCTAACTCTGAGCTGCCTCCAAAGCCAGCTGCAGTGGTGAGTGAGGCCTGTGATGCTGATGATGCTGATAATGCCAGTAAACTAGCTGCAATGCTAAATACCTGTTCCTTTCAGAAACCAGAACAACTACAACAACAAAAATCAGTTTTTGAGATATGCCCATCTCCTGCAGAAAATAACATCCAGGGAACCACAAAGCTAGCGTCTGGTGAAGAATTCTGTTCAGGTTCCAGCAGCAATTTGCAAGCTCCTGGTGGCAGCTCTGAACGGTATTTAAAACAAAATGAAATGAATGGTGCTTACTTCAAGCAAAGCTCAGTGTTCACTAAGGATTCCTTTTCTGCCACTACCACACCACCACCACCATCACAATTGCTTCTTTCTCCCCCTCCTCCTCTTCCACAGGTTCCTCAGCTTCCTTCAGAAGGAAAAAGCACTCTGAATGGTGGAGTTTTAGAAGAACACCACCACTACCCCAACCAAAGTAACACAACACTTTTAAGGGAAGTGAAAATAGAGGGTAAACCTGAGGCACCACCTTCCCAGAGTCCTAATCCATCTACACATGTATGCAGCCCTTCTCCGATGCTTTCTGAAAGGCCTCAGAATAATTGTGTGAACAGGAATGACATACAGACTGCAGGGACAATGACTGTTCCATTGTGTTCTGAGAAAACAAGACCAATGTCAGAACACCTCAAGCATAACCCACCAATTTTTGGTAGCAGTGGAGAGCTACAGGACAACTGCCAGCAGTTGATGAGAAACAAAGAGCAAGAGATTCTGAAGGGTCGAGACAAGGAGCAAACACGAGATCTTGTGCCCCCAACACAGCACTATCTGAAACCAGGATGGATTGAATTGAAGGCCCCTCGTTTTCACCAAGCGGAATCCCATCTAAAACGTAATGAGGCATCACTGCCATCAATTCTTCAGTATCAACCCAATCTCTCCAATCAAATGACCTCCAAACAATACACTGGAAATTCCAACATGCCTGGGGGGCTCCCAAGGCAAGCTTACACCCAGAAAACAACACAGCTGGAGCACAAGTCACAAATGTACCAAGTTGAAATGAATCAAGGGCAGTCCCAAGGTACAGTGGACCAACATCTCCAGTTCCAAAAACCCTCACACCAGGTGCACTTCTCCAAAACAGACCATTTACCAAAAGCTCATGTGCAGTCACTGTGTGGCACTAGATTTCATTTTCAACAAAGAGCAGATTCCCAAACTGAAAAACTTATGTCCCCAGTGTTGAAACAGCACTTGAATCAACAGGCTTCAGAGACTGAGCCATTTTCAAACTCACACCTTTTGCAACATAAGCCTCATAAACAGGCAGCACAAACACAACCATCCCAGAGTTCACATCTCCCTCAAAACCAGCAACAGCAGCAAAAATTACAAATAAAGAATAAAGAGGAAATACTCCAGACTTTTCCTCACCCCCAAAGCAACAATGATCAGCAAAGAGAAGGATCATTCTTTGGCCAGACTAAAGTGGAAGAATGTTTTCATGGTGAAAATCAGTATTCAAAATCAAGCGAGTTCGAGACTCATAATGTCCAAATGGGACTGGAGGAAGTACAGAATATAAATCGTAGAAATTCCCCTTATAGTCAGACCATGAAATCAAGTGCATGCAAAATACAGGTTTCTTGTTCAAACAATACACACCTAGTTTCAGAGAATAAAGAACAGACTACACATCCTGAACTTTTTGCAGGAAACAAGACCCAAAACTTGCATCACATGCAATATTTTCCAAATAATGTGATCCCAAAGCAAGATCTTCTTCACAGGTGCTTTCAAGAACAGGAGCAGAAGTCACAACAAGCTTCAGTTCTACAGGGATATAAAAATAGAAACCAAGATATGTCTGGTCAACAAGCTGCGCAACTTGCTCAGCAAAGGTACTTGATACATAACCATGCAAATGTTTTTCCTGTGCCTGACCAGGGAGGAAGTCACACTCAGACCCCTCCCCAGAAGGACACTCAAAAGCATGCTGCTCTAAGGTGGCATCTCTTACAGAAGCAAGAACAGCAGCAAACACAGCAACCCCAAACTGAGTCTTGCCATAGTCAGATGCACAGGCCAATTAAGGTGGAACCTGGATGCAAGCCACATGCCTGTATGCACACAGCACCACCAGAAAACAAAACATGGAAAAAGGTAACTAAGCAAGAGAATCCACCTGCAAGCTGTGATAATGTGCAGCAAAAGAGCATCATTGAGACCATGGAGCAGCATCTGAAGCAGTTTCACGCCAAGTCGTTATTTGACCATAAGGCTCTTACTCTCAAATCACAGAAGCAAGTAAAAGTTGAAATGTCAGGGCCAGTCACAGTTTTGACTAGACAAACCACTGCTGCAGAACTTGATAGCCACACCCCAGCTTTAGAGCAGCAAACAACTTCTTCAGAAAAGACACCAACCAAAAGAACAGCTGCTTCTGTTCTCAATAATTTTATAGAGTCACCTTCCAAATTACTAGATACTCCTATAAAAAATTTATTGGATACACCTGTCAAGACTCAATATGATTTCCCATCTTGCAGATGTGTAGGTAAGTGCCAGAAATGTACTGAGACACATGGCGTTTATCCAGAATTAGCAAATTTATCTTCAGATATGGGATTTTCCTTCTTTTTTTAAA'\n",
    "\n",
    "    #Create a list of the codons they called missense variants at\n",
    "    import csv\n",
    "    filename = csv_file_variants\n",
    "    with open(filename, 'r') as csvfile:\n",
    "        read_reader = csv.reader(csvfile)  #csv.reader returns a reader object which will iterate over lines in the csvfile\n",
    "        row_count=0\n",
    "\n",
    "        TET2_missense_variant_list_NP_060098 = []\n",
    "        TET2_missense_codons_NP_060098 =[]\n",
    "        TET2_missense_variant_DICT_NP_060098 = {}\n",
    "\n",
    "        for row in read_reader: #each row contains aa number (e.g. 11) and variant (e.g. p.D11A)\n",
    "    #         print(row)\n",
    "            if row_count>0: \n",
    "                codon0_NP_060098 = int(row[0])-1\n",
    "                codon1_NP_060098 = int(row[0])\n",
    "                variant_NP_060098 = row[1]\n",
    "                if codon1_NP_060098 < 1167:\n",
    "                    TET2_missense_codons_NP_060098.append(codon0_NP_060098) #List of codon numbers that they called variants at\n",
    "                    TET2_missense_variant_DICT_NP_060098[row_count]=(codon1_NP_060098, variant_NP_060098) #Dictionary:  K = codon number, V = variant at that codon (e.g. 76: p.V76L)\n",
    "\n",
    "            row_count=row_count+1\n",
    "            \n",
    "    #Looking at what possible amino acid changes are possible at those codons\n",
    "    TET2_missense_codons_NP_060098_DICT = {}\n",
    "    calls_TET2_missense_NP_060098 = {}\n",
    "\n",
    "    for i in TET2_missense_codons_NP_060098: #Iterates over the list of codon numbers\n",
    "        TET2_missense_codons_NP_060098_DICT[i+1]= (TET2_codons_NP_060098[i]) #Creates of dicitonary of key = codon number, value = bases (sequence) at that codon\n",
    "        #e.g. 12: CGT, 882: GCA\n",
    "        \n",
    "    for k,v in TET2_missense_codons_NP_060098_DICT.items(): #iterating over the list of codon numbers (and codon sequences)\n",
    "        if v in Possible_Amino_Acid_Changes_letters.keys(): #assign the list of all the possible amino acid changes to each codon they called\n",
    "            changes = Possible_Amino_Acid_Changes_letters[v]\n",
    "    #         print(Possible_Amino_Acid_Changes.values())\n",
    "            calls_TET2_missense_NP_060098[k]=(v, changes)\n",
    "        #e.g. 12: (CGT, 'K', '*', 'T'etc...)\n",
    "        \n",
    "    #Changing the original amino acid to a Word (e.g. Phe) rather than a codon\n",
    "    calls_TET2_missense_NP_060098_extra = {}\n",
    "\n",
    "    for k, v in calls_TET2_missense_NP_060098.items():\n",
    "        if v[0] in Genetic_Code:\n",
    "            Amino_acid_change = Genetic_Code.get(v[0])\n",
    "            calls_TET2_missense_NP_060098_extra[k] = (Amino_acid_change, v[1])\n",
    "            #e.g. 12: (Phe, 'K', '*', 'T' etc..)\n",
    "\n",
    "    #Changing the original amino acid to a LETTER (rather than the codon)\n",
    "    calls_TET2_missense_NP_060098_extra2 = {}\n",
    "\n",
    "    for k, v in calls_TET2_missense_NP_060098_extra.items():\n",
    "        if v[0] in Amino_acid_abbreviations:\n",
    "            Amino_acid_abbrev = Amino_acid_abbreviations.get(v[0])\n",
    "            calls_TET2_missense_NP_060098_extra2[k] = (Amino_acid_abbrev, v[1])\n",
    "            #e.g. 12: ((F), ('K', '*', 'T' etc..))\n",
    "            \n",
    "    #Create a dictionary containing info on which position in the codon is affected to create the variant\n",
    "    TET2_position_in_list_dict_missense_NP_060098={}\n",
    "    for k, v in TET2_missense_variant_DICT_NP_060098.items(): #Iterates over the dictionary of (key = row_count, value = aa_pos, variant (e.g. 3: (76, p.V76L))\n",
    "        observed_change=(v[1])[-1] #Selects the amino acid change (e.g. the L from p.V76L)\n",
    "        possible_changes=calls_TET2_missense_NP_060098[v[0]][1] #possible changes are the second value in the dictionary of (k = codon number, value = (original amino acid, possible changes))\n",
    "        possible_changes_info=calls_TET2_missense_NP_060098[v[0]]\n",
    "        codon_seq=calls_TET2_missense_NP_060098[v[0]][0]\n",
    "        possible_changes_string = \"\".join(possible_changes)\n",
    "        if v[1][2] == calls_TET2_missense_NP_060098_extra2[v[0]][0]:  #v[1][2] = the original amino acid letter (i.e if this aa is in this transcript)\n",
    "            position_in_list = -1\n",
    "            positions_in_list = []\n",
    "            while True:\n",
    "                position_in_list = possible_changes_string.find(observed_change, position_in_list+1) #counts position in possible changes (if 0-2 = 1st base of codon, if 3-5 = 2nd base, if 6-8 = 3rd base)\n",
    "                if position_in_list == -1: break\n",
    "                positions_in_list.append(position_in_list)\n",
    "                TET2_position_in_list_dict_missense_NP_060098[k]=(positions_in_list, possible_changes_info, codon_seq, v[0], v[1])\n",
    "            \n",
    "    aa_positions_variants_called = []\n",
    "    for k, v in TET2_position_in_list_dict_missense_NP_060098.items():\n",
    "        aa_positions_variants_called.append(v[3])\n",
    "#     print('positions at which missense variants called = ', aa_positions_variants_called)\n",
    "                \n",
    "    #Create a dictionary that contains the position in the list of possible AA where the variant is and what it changes to\n",
    "    TET2_bases_missense_NP_060098 = {}\n",
    "    for k, v in TET2_position_in_list_dict_missense_NP_060098.items():\n",
    "        codon_seq = v[2]\n",
    "        base_changes=[]\n",
    "        if codon_seq in Possible_Base_Changes.keys():\n",
    "            list_of_possible_bases = Possible_Base_Changes.get(codon_seq)\n",
    "        for i in v[0]:\n",
    "            base_change_position = i\n",
    "            base_change = list_of_possible_bases[base_change_position]\n",
    "            base_changes.append(base_change)\n",
    "        TET2_bases_missense_NP_060098[k]=(v, base_changes)\n",
    "        \n",
    "    #Create dictionary which contains original codon seq and details on which base is changed\n",
    "    TET2_bases_missense2_NP_060098 = {}\n",
    "\n",
    "    for k, v in TET2_bases_missense_NP_060098.items():\n",
    "        base_change_positions = v[0][0]\n",
    "        codon_positions = []\n",
    "        positions_in_gene = []\n",
    "        for i in base_change_positions:\n",
    "            base_change_position = i\n",
    "            if base_change_position <= 2:\n",
    "                codon_pos = 1\n",
    "            if 2 < base_change_position <= 5:\n",
    "                codon_pos = 2\n",
    "            if 5 < base_change_position <= 8:\n",
    "                codon_pos = 3\n",
    "            codon_positions.append(codon_pos)\n",
    "        TET2_bases_missense2_NP_060098[k] = (v, codon_positions)\n",
    "        \n",
    "    #Create dictionary which contains original codon seq, details on which base is changed and it's position in the coding region\n",
    "    TET2_bases_missense3_NP_060098 = {}\n",
    "\n",
    "    for k, v in TET2_bases_missense2_NP_060098.items():\n",
    "        codon_positions = v[1]\n",
    "        base_changes = v[0][1]\n",
    "        original_bases = []\n",
    "        positions_in_gene = []\n",
    "        codon_number = v[0][0][3]\n",
    "        variant = v[0][0][4]\n",
    "        for i in codon_positions:\n",
    "            codon_pos = i\n",
    "            position_in_gene = ((int(v[0][0][3])-1)*3)+codon_pos #codon number (v[0][0][3]) minus 1 (e.g. codon 307 -1 = 306, then times by 3 takes you to beginning of codon 307 (then add the position))  \n",
    "            positions_in_gene.append(position_in_gene)\n",
    "            original_base = (v[0][0][2][codon_pos-1])\n",
    "            original_bases.append(original_base)\n",
    "        TET2_bases_missense3_NP_060098[k] = (v[0][0][1], positions_in_gene, original_bases, base_changes, codon_number, variant)\n",
    "\n",
    "    #Create dictionary which contains original codon seq, details on which base is changed and its neighbouring bases\n",
    "    TET2_bases_missense4_NP_060098 = {}\n",
    "\n",
    "    for k, v in TET2_bases_missense3_NP_060098.items():\n",
    "        position_in_gene = v[1]\n",
    "        sites_trio_missense=[]\n",
    "\n",
    "        for i in position_in_gene:\n",
    "            gene_pos = i\n",
    "            site_trio_missense = [TET2_exons_extra_bases_NP_060098[gene_pos:gene_pos+3] for gene_pos in range (gene_pos-1, gene_pos, 3)]\n",
    "            sites_trio_missense.append(site_trio_missense)\n",
    "            TET2_bases_missense4_NP_060098[k] = (v, sites_trio_missense)\n",
    "                    \n",
    "    #Create a dictionary containing pos in coding region, original base, new base, base affected with its neighbours\n",
    "    #e.g. 938, G, A, TGG = base 938 on coding region, original base = G, changes to base A, base with neighbours either side = TGG\n",
    "    TET2_sites_missense5_NP_060098 = {}\n",
    "\n",
    "    a=0\n",
    "    for k, v in TET2_bases_missense4_NP_060098.items():\n",
    "        codon_n = v[0][1] #base number in codon (out of 3 x 912)\n",
    "        codon_n_912 = v[0][4]\n",
    "        original_base = v[0][2]\n",
    "        new_base = v[0][3]\n",
    "        site_trio = v[1]\n",
    "        info=[]\n",
    "        variant = v[0][5]\n",
    "\n",
    "        change = 0\n",
    "        for i in codon_n:\n",
    "            codon_number = i\n",
    "            orig_base = original_base[change]\n",
    "            n_base = new_base[change]\n",
    "            site_tri = site_trio[change]\n",
    "            info.append([codon_number, orig_base, n_base, site_tri, codon_n_912, variant])\n",
    "            change = change+1\n",
    "        TET2_sites_missense5_NP_060098[a]=info\n",
    "        a=a+1\n",
    "        \n",
    "    # Write in a way that can iterate over mutation rates (i.e. put in to form C[C>A]G etc)   \n",
    "    TET2_sites_missense6_NP_060098={}\n",
    "\n",
    "    separating_positions = {} # separate out all the possible changes to just 1 effective item in the list\n",
    "    a=0\n",
    "    for k, v in TET2_sites_missense5_NP_060098.items():\n",
    "        for i in v: #because some contain more than 1 change from that codon:\n",
    "            separating_positions[a]=i\n",
    "            a=a+1\n",
    "    \n",
    "    for k, v in separating_positions.items():\n",
    "        base_pos = v[0]\n",
    "        original_base = v[1]\n",
    "        new_base = v[2]\n",
    "        site_trio = v[3]\n",
    "        codon_n_912 = v[4]\n",
    "        variant = v[5]\n",
    "        sequence = (str(original_base),str(new_base))\n",
    "        changes = '>'.join(sequence)\n",
    "        left_base = site_trio[0][0]\n",
    "        right_base = site_trio[0][2]\n",
    "        sequence2 = (changes, right_base)\n",
    "        changes2 = ']'.join(sequence2)\n",
    "        sequence3 = (left_base, changes2)\n",
    "        changes3 = '['.join(sequence3)\n",
    "        TET2_sites_missense6_NP_060098[k] = (v[0],changes3, codon_n_912, variant) #site_trio = the affected base and its neighbours\n",
    "        \n",
    "        \n",
    "    #Assign the mutation rates to each site...\n",
    "    TET2_mutation_rates_missense_NP_060098 = {}\n",
    "    for k, v in TET2_sites_missense6_NP_060098.items():\n",
    "        base_change_info = v[1]\n",
    "        if base_change_info in Mutation_rates.keys():\n",
    "            mutation_rate = Mutation_rates.get(base_change_info)\n",
    "            TET2_mutation_rates_missense_NP_060098[k] = (v, mutation_rate)\n",
    "\n",
    "    TET2_mutation_rate_missense_NP_060098_sum = []\n",
    "    mutation_rates_list = []\n",
    "    for k, v in TET2_mutation_rates_missense_NP_060098.items():\n",
    "    #     print(v[1])\n",
    "        aa_pos = v[0][2]\n",
    "        site = v[0][1]\n",
    "        mutation_rate_value = v[1] \n",
    "        variant = v[0][3]\n",
    "        mutation_rates_list.append((aa_pos, site, mutation_rate_value, variant))\n",
    "        TET2_mutation_rate_missense_NP_060098_sum.append(mutation_rate_value)\n",
    "        \n",
    "    TET2_Summed_mutation_rate_missense_NP_060098_added = np.sum(TET2_mutation_rate_missense_NP_060098_sum)\n",
    "\n",
    "    print('Total summed mutation rate across TET2 missense_NP_060098 for this study = ', TET2_Summed_mutation_rate_missense_NP_060098_added)\n",
    "    \n",
    "    mutation_rates_list_all = mutation_rates_list\n",
    "    return mutation_rates_list_all #in format: column 1 = 882, column 2 = p.R882H  #1165 aa"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### ASXL1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 294,
   "metadata": {
    "code_folding": [
     0
    ]
   },
   "outputs": [],
   "source": [
    "def mutation_rates_list_nonsense_variants_ASXL1(csv_file_variants):\n",
    "    \n",
    "    #The mutation rate at a base is calculated based on its trinucleotide context.  To obtain the trinucleotide context, the base either side of the mutated base needs to be ascertained.\n",
    "    #The coding sequence (+ an extra base at each end) of each gene is used when determining the trinucleotide context of the base change.  The problem with using the coding sequence is that\n",
    "    #if a base change falls at the end of an exon, its neighbouring base will actuall be an intronic base, which may be different to the 'neighbouring' base in the coding sequence.\n",
    "    #To account for this, there is an \"intron_exon_check\" function that is called at the intron/exon boundary if the neighbouring base in the intron is different to the neighbouring base in the coding sequence.\n",
    "    \n",
    "    intron_exon_boundary_caution_base = [140, 143, 252, 373, 1086, 1085]\n",
    "   \n",
    "    def intron_exon_check(position_in_gene_number):\n",
    "        if position_in_gene_number == 140:\n",
    "            gene_pos = 140\n",
    "            first_base = ASXL1_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = ASXL1_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 143:\n",
    "            gene_pos = 143\n",
    "            first_base = ASXL1_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = ASXL1_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 252:\n",
    "            gene_pos = 252\n",
    "            first_base = ASXL1_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = ASXL1_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 373:\n",
    "            gene_pos = 373\n",
    "            first_base = ASXL1_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = ASXL1_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 1086:\n",
    "            gene_pos = 1086\n",
    "            first_base = 'G'\n",
    "            middle_base = ASXL1_exons_extra_bases[gene_pos]\n",
    "            third_base = ASXL1_exons_extra_bases[gene_pos+1]\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 1085:\n",
    "            gene_pos = 1085\n",
    "            first_base = ASXL1_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = ASXL1_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        return triplet\n",
    "    \n",
    "    #ASXL1 transcript:\n",
    "    #Downloaded from http://grch37.ensembl.org\n",
    "    ASXL1_exons = 'ATGAAGGACAAACAGAAGAAGAAGAAGGAGCGCACGTGGGCCGAGGCCGCGCGCCTGGTATTAGAAAACTACTCGGATGCTCCAATGACACCAAAACAGATTCTGCAGGTCATAGAGGCAGAAGGACTAAAGGAAATGAGAAGTGGGACTTCCCCTCTCGCATGCCTCAATGCTATGCTACATTCCAATTCAAGAGGAGGAGAGGGGTTGTTTTATAAACTGCCTGGCCGAATCAGCCTTTTCACGCTCAAGAAGGATGCCCTGCAGTGGTCTCGCCATCCAGCTACAGTGGAGGGAGAGGAGCCAGAGGACACGGCTGATGTGGAGAGCTGTGGGTCTAATGAAGCCAGCACTGTGAGTGGTGAAAACGATGTATCTCTTGATGAAACATCTTCGAACGCATCCTGTTCTACAGAATCTCAGAGTCGACCTCTTTCCAATCCCAGGGACAGCTACAGAGCTTCCTCACAGGCGAACAAACAAAAGAAAAAGACTGGGGTGATGCTGCCTCGAGTTGTCCTGACTCCTCTGAAGGTAAACGGGGCCCACGTGGAATCTGCATCAGGGTTCTCGGGCTGCCACGCCGATGGCGAGAGCGGCAGCCCGTCCAGCAGCAGCAGCGGCTCTCTGGCCCTGGGCAGCGCTGCTATTCGTGGCCAGGCCGAGGTCACCCAGGACCCTGCCCCGCTCCTGAGAGGCTTCCGGAAGCCAGCCACAGGTCAAATGAAGCGCAACAGAGGGGAAGAAATAGATTTTGAGACACCTGGGTCCATTCTTGTCAACACCAACCTCCGTGCCCTGATCAACTCTCGGACCTTCCATGCCTTACCATCACACTTCCAGCAGCAGCTCCTCTTCCTCCTGCCTGAAGTAGACAGACAGGTGGGGACGGATGGCCTGTTGCGTCTCAGCAGCAGTGCACTAAATAACGAGTTTTTTACCCATGCGGCTCAGAGCTGGCGGGAGCGCCTGGCTGATGGTGAATTTACTCATGAGATGCAAGTCAGGATACGACAGGAAATGGAGAAGGAAAAGAAGGTGGAACAATGGAAAGAAAAGTTCTTTGAAGACTACTATGGACAGAAGCTGGGTTTGACCAAAGAAGAGTCATTGCAGCAGAACGTGGGCCAGGAGGAGGCTGAAATCAAAAGTGGCTTGTGTGTCCCAGGAGAATCAGTGCGTATACAGCGTGGTCCAGCCACCCGACAGCGAGATGGGCATTTTAAGAAACGCTCTCGGCCAGATCTCCGAACCAGAGCCAGAAGGAATCTGTACAAAAAACAGGAGTCAGAACAAGCAGGGGTTGCTAAGGATGCAAAATCTGTGGCCTCAGATGTTCCCCTCTACAAGGATGGGGAGGCTAAGACTGACCCAGCAGGGCTGAGCAGTCCCCATCTGCCAGGCACATCCTCTGCAGCACCCGACCTGGAGGGTCCCGAATTCCCAGTTGAGTCTGTGGCTTCTCGGATCCAGGCTGAGCCAGACAACTTGGCACGTGCCTCTGCATCTCCAGACAGAATTCCTAGCCTGCCTCAGGAAACTGTGGATCAGGAACCCAAGGATCAGAAGAGGAAATCCTTTGAGCAGGCGGCCTCTGCATCCTTTCCCGAAAAGAAGCCCCGGCTTGAAGATCGTCAGTCCTTTCGTAACACAATTGAAAGTGTTCACACCGAAAAGCCACAGCCCACTAAAGAGGAGCCCAAAGTCCCGCCCATCCGGATTCAACTTTCACGTATCAAACCACCCTGGGTGGTTAAAGGTCAGCCCACTTACCAGATATGCCCCCGGATCATCCCCACCACGGAGTCCTCCTGCCGGGGTTGGACTGGCGCCAGGACCCTCGCAGACATTAAAGCCCGTGCTCTGCAGGTCCGAGGGGCGAGAGGTCACCACTGCCATAGAGAGGCGGCCACCACTGCCATCGGAGGGGGGGGTGGCCCGGGTGGAGGTGGCGGCGGGGCCACCGATGAGGGAGGTGGCAGAGGCAGCAGCAGTGGTGATGGTGGTGAGGCCTGTGGCCACCCTGAGCCCAGGGGAGGCCCGAGCACCCCTGGAAAGTGTACGTCAGATCTACAGCGAACACAACTACTGCCGCCTTATCCTCTAAATGGGGAGCATACCCAGGCCGGAACTGCCATGTCCAGAGCTAGGAGAGAGGACCTGCCTTCTCTGAGAAAGGAGGAAAGCTGCCTACTACAGAGGGCTACAGTTGGACTCACAGATGGGCTAGGAGATGCCTCCCAACTCCCCGTTGCTCCCACTGGGGACCAGCCATGCCAGGCCTTGCCCCTACTGTCCTCCCAAACCTCAGTAGCTGAGAGATTAGTGGAGCAGCCTCAGTTGCATCCGGATGTTAGAACTGAATGTGAGTCTGGCACCACTTCCTGGGAAAGTGATGATGAGGAGCAAGGACCCACCGTTCCTGCAGACAATGGTCCCATTCTGTCTCTAGTGGGAGATGATACATTAGAGAAAGGAACTGGCCAAGCTCTTGACAGTCATCCCACTATGAAGGATCCTGTAAATGTGACCCCCAGTTCCACACCTGAATCCTCACCGACTGATTGCCTGCAGAACAGAGCATTTGATGACGAATTAGGGCTTGGTGGCTCATGCCCTCCTATGAGGGAAAGTGATACTAGACAAGAAAACTTGAAAACCAAGGCTCTCGTTTCTAACAGTTCTTTGCATTGGATACCCATCCCATCGAATGATGAGGTAGTGAAACAGCCCAAACCAGAATCCAGAGAACACATACCATCTGTTGAGCCCCAGGTTGGAGAGGAGTGGGAGAAAGCTGCTCCCACCCCTCCTGCATTGCCTGGGGATTTGACAGCTGAGGAGGGTCTAGATCCTCTTGACAGCCTTACTTCACTCTGGACTGTGCCATCTCGAGGAGGCAGTGACAGCAATGGCAGTTACTGTCAACAGGTGGACATTGAAAAGCTGAAAATCAACGGAGACTCTGAAGCACTGAGTCCTCACGGTGAGTCCACGGATACAGCCTCTGACTTTGAAGGTCACCTCACGGAGGACAGCAGTGAGGCTGACACTAGAGAAGCTGCAGTGACAAAGGGATCTTCGGTGGACAAGGATGAGAAACCCAATTGGAACCAATCTGCCCCACTGTCCAAGGTGAATGGTGACATGCGTCTGGTTACAAGGACAGATGGGATGGTTGCTCCTCAGAGCTGGGTGTCTCGAGTATGTGCGGTCCGCCAAAAGATCCCAGATTCCCTACTGCTGGCCAGTACTGAGTACCAGCCAAGAGCCGTGTGCCTGTCCATGCCTGGGTCCTCAGTGGAGGCCACTAACCCACTTGTGATGCAGTTGCTGCAGGGTAGCTTGCCCCTAGAGAAGGTTCTTCCACCAGCCCACGATGACAGCATGTCAGAATCCCCACAAGTACCACTTACAAAAGACCAGAGCCATGGCTCGCTACGCATGGGATCTTTACATGGTCTTGGAAAAAACAGTGGCATGGTTGATGGAAGCAGCCCCAGTTCTTTAAGGGCTTTGAAGGAGCCTCTTCTGCCAGATAGCTGTGAAACAGGCACTGGTCTTGCCAGGATTGAGGCCACCCAGGCTCCTGGAGCACCCCAAAAGAATTGCAAGGCAGTCCCAAGTTTTGACTCCCTCCATCCAGTGACAAATCCCATTACATCCTCTAGGAAACTGGAAGAAATGGATTCCAAAGAGCAGTTCTCTTCCTTTAGTTGTGAAGATCAGAAGGAAGTCCGTGCTATGTCACAGGACAGTAATTCAAATGCTGCTCCAGGAAAGAGCCCAGGAGATCTTACTACCTCGAGAACACCTCGTTTCTCATCTCCAAATGTGATCTCCTTTGGTCCAGAGCAGACAGGTCGGGCCCTGGGTGATCAGAGCAATGTTACAGGCCAAGGGAAGAAGCTTTTTGGCTCTGGGAATGTGGCTGCAACCCTTCAGCGCCCCAGGCCTGCGGACCCGATGCCTCTTCCTGCTGAGATCCCTCCAGTTTTTCCCAGTGGGAAGTTGGGACCAAGCACAAACTCCATGTCTGGTGGGGTACAGACTCCAAGGGAAGACTGGGCTCCAAAGCCACATGCCTTTGTTGGCAGCGTCAAGAATGAGAAGACTTTTGTGGGGGGTCCTCTTAAGGCAAATGCCGAGAACAGGAAAGCTACTGGGCATAGTCCCCTGGAACTGGTGGGTCACTTGGAAGGGATGCCCTTTGTCATGGACTTGCCCTTCTGGAAATTACCCCGAGAGCCAGGGAAGGGGCTCAGTGAGCCTCTGGAGCCTTCTTCTCTCCCCTCCCAACTCAGCATCAAGCAGGCATTTTATGGGAAGCTTTCTAAACTCCAACTGAGTTCCACCAGCTTTAATTATTCCTCTAGCTCTCCCACCTTTCCCAAAGGCCTTGCTGGAAGTGTGGTGCAGCTGAGCCACAAAGCAAACTTTGGTGCGAGCCACAGTGCATCACTTTCCTTGCAAATGTTCACTGACAGCAGCACGGTGGAAAGCATCTCGCTCCAGTGTGCGTGCAGCCTGAAAGCCATGATCATGTGCCAAGGCTGCGGTGCGTTCTGTCACGATGACTGTATTGGACCCTCAAAGCTCTGTGTATTGTGCCTTGTGGTGAGATAA'\n",
    "    \n",
    "    #Split the sequence into codons \n",
    "    ASXL1_codons = [ASXL1_exons[i:i+3] for i in range(0, len(ASXL1_exons), 3)]\n",
    "\n",
    "    #ASXL1 transcript:\n",
    "    #Downloaded from http://grch37.ensembl.org\n",
    "    ASXL1_exons_extra_bases = 'GATGAAGGACAAACAGAAGAAGAAGAAGGAGCGCACGTGGGCCGAGGCCGCGCGCCTGGTATTAGAAAACTACTCGGATGCTCCAATGACACCAAAACAGATTCTGCAGGTCATAGAGGCAGAAGGACTAAAGGAAATGAGAAGTGGGACTTCCCCTCTCGCATGCCTCAATGCTATGCTACATTCCAATTCAAGAGGAGGAGAGGGGTTGTTTTATAAACTGCCTGGCCGAATCAGCCTTTTCACGCTCAAGAAGGATGCCCTGCAGTGGTCTCGCCATCCAGCTACAGTGGAGGGAGAGGAGCCAGAGGACACGGCTGATGTGGAGAGCTGTGGGTCTAATGAAGCCAGCACTGTGAGTGGTGAAAACGATGTATCTCTTGATGAAACATCTTCGAACGCATCCTGTTCTACAGAATCTCAGAGTCGACCTCTTTCCAATCCCAGGGACAGCTACAGAGCTTCCTCACAGGCGAACAAACAAAAGAAAAAGACTGGGGTGATGCTGCCTCGAGTTGTCCTGACTCCTCTGAAGGTAAACGGGGCCCACGTGGAATCTGCATCAGGGTTCTCGGGCTGCCACGCCGATGGCGAGAGCGGCAGCCCGTCCAGCAGCAGCAGCGGCTCTCTGGCCCTGGGCAGCGCTGCTATTCGTGGCCAGGCCGAGGTCACCCAGGACCCTGCCCCGCTCCTGAGAGGCTTCCGGAAGCCAGCCACAGGTCAAATGAAGCGCAACAGAGGGGAAGAAATAGATTTTGAGACACCTGGGTCCATTCTTGTCAACACCAACCTCCGTGCCCTGATCAACTCTCGGACCTTCCATGCCTTACCATCACACTTCCAGCAGCAGCTCCTCTTCCTCCTGCCTGAAGTAGACAGACAGGTGGGGACGGATGGCCTGTTGCGTCTCAGCAGCAGTGCACTAAATAACGAGTTTTTTACCCATGCGGCTCAGAGCTGGCGGGAGCGCCTGGCTGATGGTGAATTTACTCATGAGATGCAAGTCAGGATACGACAGGAAATGGAGAAGGAAAAGAAGGTGGAACAATGGAAAGAAAAGTTCTTTGAAGACTACTATGGACAGAAGCTGGGTTTGACCAAAGAAGAGTCATTGCAGCAGAACGTGGGCCAGGAGGAGGCTGAAATCAAAAGTGGCTTGTGTGTCCCAGGAGAATCAGTGCGTATACAGCGTGGTCCAGCCACCCGACAGCGAGATGGGCATTTTAAGAAACGCTCTCGGCCAGATCTCCGAACCAGAGCCAGAAGGAATCTGTACAAAAAACAGGAGTCAGAACAAGCAGGGGTTGCTAAGGATGCAAAATCTGTGGCCTCAGATGTTCCCCTCTACAAGGATGGGGAGGCTAAGACTGACCCAGCAGGGCTGAGCAGTCCCCATCTGCCAGGCACATCCTCTGCAGCACCCGACCTGGAGGGTCCCGAATTCCCAGTTGAGTCTGTGGCTTCTCGGATCCAGGCTGAGCCAGACAACTTGGCACGTGCCTCTGCATCTCCAGACAGAATTCCTAGCCTGCCTCAGGAAACTGTGGATCAGGAACCCAAGGATCAGAAGAGGAAATCCTTTGAGCAGGCGGCCTCTGCATCCTTTCCCGAAAAGAAGCCCCGGCTTGAAGATCGTCAGTCCTTTCGTAACACAATTGAAAGTGTTCACACCGAAAAGCCACAGCCCACTAAAGAGGAGCCCAAAGTCCCGCCCATCCGGATTCAACTTTCACGTATCAAACCACCCTGGGTGGTTAAAGGTCAGCCCACTTACCAGATATGCCCCCGGATCATCCCCACCACGGAGTCCTCCTGCCGGGGTTGGACTGGCGCCAGGACCCTCGCAGACATTAAAGCCCGTGCTCTGCAGGTCCGAGGGGCGAGAGGTCACCACTGCCATAGAGAGGCGGCCACCACTGCCATCGGAGGGGGGGGTGGCCCGGGTGGAGGTGGCGGCGGGGCCACCGATGAGGGAGGTGGCAGAGGCAGCAGCAGTGGTGATGGTGGTGAGGCCTGTGGCCACCCTGAGCCCAGGGGAGGCCCGAGCACCCCTGGAAAGTGTACGTCAGATCTACAGCGAACACAACTACTGCCGCCTTATCCTCTAAATGGGGAGCATACCCAGGCCGGAACTGCCATGTCCAGAGCTAGGAGAGAGGACCTGCCTTCTCTGAGAAAGGAGGAAAGCTGCCTACTACAGAGGGCTACAGTTGGACTCACAGATGGGCTAGGAGATGCCTCCCAACTCCCCGTTGCTCCCACTGGGGACCAGCCATGCCAGGCCTTGCCCCTACTGTCCTCCCAAACCTCAGTAGCTGAGAGATTAGTGGAGCAGCCTCAGTTGCATCCGGATGTTAGAACTGAATGTGAGTCTGGCACCACTTCCTGGGAAAGTGATGATGAGGAGCAAGGACCCACCGTTCCTGCAGACAATGGTCCCATTCTGTCTCTAGTGGGAGATGATACATTAGAGAAAGGAACTGGCCAAGCTCTTGACAGTCATCCCACTATGAAGGATCCTGTAAATGTGACCCCCAGTTCCACACCTGAATCCTCACCGACTGATTGCCTGCAGAACAGAGCATTTGATGACGAATTAGGGCTTGGTGGCTCATGCCCTCCTATGAGGGAAAGTGATACTAGACAAGAAAACTTGAAAACCAAGGCTCTCGTTTCTAACAGTTCTTTGCATTGGATACCCATCCCATCGAATGATGAGGTAGTGAAACAGCCCAAACCAGAATCCAGAGAACACATACCATCTGTTGAGCCCCAGGTTGGAGAGGAGTGGGAGAAAGCTGCTCCCACCCCTCCTGCATTGCCTGGGGATTTGACAGCTGAGGAGGGTCTAGATCCTCTTGACAGCCTTACTTCACTCTGGACTGTGCCATCTCGAGGAGGCAGTGACAGCAATGGCAGTTACTGTCAACAGGTGGACATTGAAAAGCTGAAAATCAACGGAGACTCTGAAGCACTGAGTCCTCACGGTGAGTCCACGGATACAGCCTCTGACTTTGAAGGTCACCTCACGGAGGACAGCAGTGAGGCTGACACTAGAGAAGCTGCAGTGACAAAGGGATCTTCGGTGGACAAGGATGAGAAACCCAATTGGAACCAATCTGCCCCACTGTCCAAGGTGAATGGTGACATGCGTCTGGTTACAAGGACAGATGGGATGGTTGCTCCTCAGAGCTGGGTGTCTCGAGTATGTGCGGTCCGCCAAAAGATCCCAGATTCCCTACTGCTGGCCAGTACTGAGTACCAGCCAAGAGCCGTGTGCCTGTCCATGCCTGGGTCCTCAGTGGAGGCCACTAACCCACTTGTGATGCAGTTGCTGCAGGGTAGCTTGCCCCTAGAGAAGGTTCTTCCACCAGCCCACGATGACAGCATGTCAGAATCCCCACAAGTACCACTTACAAAAGACCAGAGCCATGGCTCGCTACGCATGGGATCTTTACATGGTCTTGGAAAAAACAGTGGCATGGTTGATGGAAGCAGCCCCAGTTCTTTAAGGGCTTTGAAGGAGCCTCTTCTGCCAGATAGCTGTGAAACAGGCACTGGTCTTGCCAGGATTGAGGCCACCCAGGCTCCTGGAGCACCCCAAAAGAATTGCAAGGCAGTCCCAAGTTTTGACTCCCTCCATCCAGTGACAAATCCCATTACATCCTCTAGGAAACTGGAAGAAATGGATTCCAAAGAGCAGTTCTCTTCCTTTAGTTGTGAAGATCAGAAGGAAGTCCGTGCTATGTCACAGGACAGTAATTCAAATGCTGCTCCAGGAAAGAGCCCAGGAGATCTTACTACCTCGAGAACACCTCGTTTCTCATCTCCAAATGTGATCTCCTTTGGTCCAGAGCAGACAGGTCGGGCCCTGGGTGATCAGAGCAATGTTACAGGCCAAGGGAAGAAGCTTTTTGGCTCTGGGAATGTGGCTGCAACCCTTCAGCGCCCCAGGCCTGCGGACCCGATGCCTCTTCCTGCTGAGATCCCTCCAGTTTTTCCCAGTGGGAAGTTGGGACCAAGCACAAACTCCATGTCTGGTGGGGTACAGACTCCAAGGGAAGACTGGGCTCCAAAGCCACATGCCTTTGTTGGCAGCGTCAAGAATGAGAAGACTTTTGTGGGGGGTCCTCTTAAGGCAAATGCCGAGAACAGGAAAGCTACTGGGCATAGTCCCCTGGAACTGGTGGGTCACTTGGAAGGGATGCCCTTTGTCATGGACTTGCCCTTCTGGAAATTACCCCGAGAGCCAGGGAAGGGGCTCAGTGAGCCTCTGGAGCCTTCTTCTCTCCCCTCCCAACTCAGCATCAAGCAGGCATTTTATGGGAAGCTTTCTAAACTCCAACTGAGTTCCACCAGCTTTAATTATTCCTCTAGCTCTCCCACCTTTCCCAAAGGCCTTGCTGGAAGTGTGGTGCAGCTGAGCCACAAAGCAAACTTTGGTGCGAGCCACAGTGCATCACTTTCCTTGCAAATGTTCACTGACAGCAGCACGGTGGAAAGCATCTCGCTCCAGTGTGCGTGCAGCCTGAAAGCCATGATCATGTGCCAAGGCTGCGGTGCGTTCTGTCACGATGACTGTATTGGACCCTCAAAGCTCTGTGTATTGTGCCTTGTGGTGAGATAAT'\n",
    "    \n",
    "    #Create a list of the codons they called nonsense variants at\n",
    "    import csv\n",
    "    filename = csv_file_variants\n",
    "    with open(filename, 'r') as csvfile:\n",
    "        read_reader = csv.reader(csvfile)  #csv.reader returns a reader object which will iterate over lines in the csvfile\n",
    "        row_count=0\n",
    "\n",
    "        ASXL1_nonsense_variant_list = []\n",
    "        ASXL1_nonsense_codons =[]\n",
    "        ASXL1_nonsense_variant_DICT = {}\n",
    "\n",
    "        for row in read_reader: #each row contains aa number (e.g. 11) and variant (e.g. p.D11A)\n",
    "    #         print(row)\n",
    "            if row_count>0: \n",
    "                codon0 = int(row[0])-1\n",
    "                codon1 = int(row[0])\n",
    "                variant = row[1]\n",
    "                ASXL1_nonsense_codons.append(codon0) #List of codon numbers that they called variants at\n",
    "                ASXL1_nonsense_variant_DICT[row_count]=(codon1, variant) #Dictionary:  K = codon number, V = variant at that codon (e.g. 76: p.V76L)\n",
    "\n",
    "            row_count=row_count+1\n",
    "            \n",
    "    #Looking at what possible amino acid changes are possible at those codons\n",
    "    ASXL1_nonsense_codons_DICT = {}\n",
    "    calls_nonsense = {}\n",
    "\n",
    "    for i in ASXL1_nonsense_codons: #Iterates over the list of codon numbers\n",
    "        ASXL1_nonsense_codons_DICT[i+1]= (ASXL1_codons[i]) #Creates of dicitonary of key = codon number, value = bases (sequence) at that codon\n",
    "        #e.g. 12: CGT, 882: GCA\n",
    "        \n",
    "    for k,v in ASXL1_nonsense_codons_DICT.items(): #iterating over the list of codon numbers (and codon sequences)\n",
    "        if v in Possible_Amino_Acid_Changes_letters.keys(): #assign the list of all the possible amino acid changes to each codon they called\n",
    "            changes = Possible_Amino_Acid_Changes_letters[v]\n",
    "    #         print(Possible_Amino_Acid_Changes.values())\n",
    "            calls_nonsense[k]=(v, changes)\n",
    "        #e.g. 12: (CGT, 'K', '*', 'T'etc...)\n",
    "        \n",
    "    #Changing the original amino acid to a Word (e.g. Phe) rather than a codon\n",
    "    calls_nonsense_extra = {}\n",
    "\n",
    "    for k, v in calls_nonsense.items():\n",
    "        if v[0] in Genetic_Code:\n",
    "            Amino_acid_change = Genetic_Code.get(v[0])\n",
    "            calls_nonsense_extra[k] = (Amino_acid_change, v[1])\n",
    "            #e.g. 12: (Phe, 'K', '*', 'T' etc..)\n",
    "\n",
    "    #Changing the original amino acid to a LETTER (rather than the codon)\n",
    "    calls_nonsense_extra2 = {}\n",
    "\n",
    "    for k, v in calls_nonsense_extra.items():\n",
    "        if v[0] in Amino_acid_abbreviations:\n",
    "            Amino_acid_abbrev = Amino_acid_abbreviations.get(v[0])\n",
    "            calls_nonsense_extra2[k] = (Amino_acid_abbrev, v[1])\n",
    "            #e.g. 12: ((F), ('K', '*', 'T' etc..))\n",
    "            \n",
    "    #Create a dictionary containing info on which position in the codon is affected to create the variant\n",
    "    ASXL1_position_in_list_dict_nonsense={}\n",
    "    for k, v in ASXL1_nonsense_variant_DICT.items(): #Iterates over the dictionary of (key = row_count, value = aa_pos, variant (e.g. 3: (76, p.V76L))\n",
    "        observed_change=(v[1])[-1] #Selects the amino acid change (e.g. the L from p.V76L)\n",
    "        possible_changes=calls_nonsense[v[0]][1] #possible changes are the second value in the dictionary of (k = codon number, value = (original amino acid, possible changes))\n",
    "        possible_changes_info=calls_nonsense[v[0]]\n",
    "        codon_seq=calls_nonsense[v[0]][0]\n",
    "        possible_changes_string = \"\".join(possible_changes)\n",
    "        if v[1][2] == calls_nonsense_extra2[v[0]][0]:  #v[1][2] = the original amino acid letter (i.e if this aa is in this transcript)\n",
    "            position_in_list = -1\n",
    "            positions_in_list = []\n",
    "            while True:\n",
    "                position_in_list = possible_changes_string.find(observed_change, position_in_list+1) #counts position in possible changes (if 0-2 = 1st base of codon, if 3-5 = 2nd base, if 6-8 = 3rd base)\n",
    "                if position_in_list == -1: break\n",
    "                positions_in_list.append(position_in_list)\n",
    "                ASXL1_position_in_list_dict_nonsense[k]=(positions_in_list, possible_changes_info, codon_seq, v[0], v[1])\n",
    "            \n",
    "    aa_positions_variants_called = []\n",
    "    for k, v in ASXL1_position_in_list_dict_nonsense.items():\n",
    "        aa_positions_variants_called.append(v[3])\n",
    "#     print('positions at which nonsense variants called = ', aa_positions_variants_called)\n",
    "                \n",
    "    #Create a dictionary that contains the position in the list of possible AA where the variant is and what it changes to\n",
    "    ASXL1_bases_nonsense = {}\n",
    "    for k, v in ASXL1_position_in_list_dict_nonsense.items():\n",
    "        codon_seq = v[2]\n",
    "        base_changes=[]\n",
    "        if codon_seq in Possible_Base_Changes.keys():\n",
    "            list_of_possible_bases = Possible_Base_Changes.get(codon_seq)\n",
    "        for i in v[0]:\n",
    "            base_change_position = i\n",
    "            base_change = list_of_possible_bases[base_change_position]\n",
    "            base_changes.append(base_change)\n",
    "        ASXL1_bases_nonsense[k]=(v, base_changes)\n",
    "        \n",
    "    #Create dictionary which contains original codon seq and details on which base is changed\n",
    "    ASXL1_bases_nonsense2 = {}\n",
    "\n",
    "    for k, v in ASXL1_bases_nonsense.items():\n",
    "        base_change_positions = v[0][0]\n",
    "        codon_positions = []\n",
    "        positions_in_gene = []\n",
    "        for i in base_change_positions:\n",
    "            base_change_position = i\n",
    "            if base_change_position <= 2:\n",
    "                codon_pos = 1\n",
    "            if 2 < base_change_position <= 5:\n",
    "                codon_pos = 2\n",
    "            if 5 < base_change_position <= 8:\n",
    "                codon_pos = 3\n",
    "            codon_positions.append(codon_pos)\n",
    "        ASXL1_bases_nonsense2[k] = (v, codon_positions)\n",
    "        \n",
    "    #Create dictionary which contains original codon seq, details on which base is changed and it's position in the coding region\n",
    "    ASXL1_bases_nonsense3 = {}\n",
    "\n",
    "    for k, v in ASXL1_bases_nonsense2.items():\n",
    "        codon_positions = v[1]\n",
    "        base_changes = v[0][1]\n",
    "        original_bases = []\n",
    "        positions_in_gene = []\n",
    "        codon_number = v[0][0][3]\n",
    "        variant = v[0][0][4]\n",
    "        for i in codon_positions:\n",
    "            codon_pos = i\n",
    "            position_in_gene = ((int(v[0][0][3])-1)*3)+codon_pos #codon number (v[0][0][3]) minus 1 (e.g. codon 307 -1 = 306, then times by 3 takes you to beginning of codon 307 (then add the position))  \n",
    "            positions_in_gene.append(position_in_gene)\n",
    "            original_base = (v[0][0][2][codon_pos-1])\n",
    "            original_bases.append(original_base)\n",
    "        ASXL1_bases_nonsense3[k] = (v[0][0][1], positions_in_gene, original_bases, base_changes, codon_number, variant)\n",
    "\n",
    "    #Create dictionary which contains original codon seq, details on which base is changed and its neighbouring bases\n",
    "    ASXL1_bases_nonsense4 = {}\n",
    "\n",
    "    for k, v in ASXL1_bases_nonsense3.items():\n",
    "        position_in_gene = v[1]\n",
    "        sites_trio_nonsense=[]\n",
    "\n",
    "        for i in position_in_gene:\n",
    "    #         print(i)\n",
    "            if i in intron_exon_boundary_caution_base:\n",
    "#                 print('positions at intron_exon_boundary =',i)\n",
    "                adapted_trio = intron_exon_check(i)\n",
    "                sites_trio_nonsense.append([adapted_trio])\n",
    "                ASXL1_bases_nonsense4[k] = (v, sites_trio_nonsense)\n",
    "\n",
    "            else:\n",
    "                for i in position_in_gene:\n",
    "                    gene_pos = i\n",
    "                    site_trio_nonsense = [ASXL1_exons_extra_bases[gene_pos:gene_pos+3] for gene_pos in range (gene_pos-1, gene_pos, 3)]\n",
    "                    sites_trio_nonsense.append(site_trio_nonsense)\n",
    "                    ASXL1_bases_nonsense4[k] = (v, sites_trio_nonsense)\n",
    "                    \n",
    "    #Create a dictionary containing pos in coding region, original base, new base, base affected with its neighbours\n",
    "    #e.g. 938, G, A, TGG = base 938 on coding region, original base = G, changes to base A, base with neighbours either side = TGG\n",
    "    ASXL1_sites_nonsense5 = {}\n",
    "\n",
    "    a=0\n",
    "    for k, v in ASXL1_bases_nonsense4.items():\n",
    "        codon_n = v[0][1] #base number in codon (out of 3 x 912)\n",
    "        codon_n_912 = v[0][4]\n",
    "        original_base = v[0][2]\n",
    "        new_base = v[0][3]\n",
    "        site_trio = v[1]\n",
    "        info=[]\n",
    "        variant = v[0][5]\n",
    "\n",
    "        change = 0\n",
    "        for i in codon_n:\n",
    "            codon_number = i\n",
    "            orig_base = original_base[change]\n",
    "            n_base = new_base[change]\n",
    "            site_tri = site_trio[change]\n",
    "            info.append([codon_number, orig_base, n_base, site_tri, codon_n_912, variant])\n",
    "            change = change+1\n",
    "        ASXL1_sites_nonsense5[a]=info\n",
    "        a=a+1\n",
    "#     print(ASXL1_sites_nonsense5)\n",
    "        \n",
    "    # Write in a way that can iterate over mutation rates (i.e. put in to form C[C>A]G etc)   \n",
    "    ASXL1_sites_nonsense6={}\n",
    "\n",
    "    separating_positions = {} # separate out all the possible changes to just 1 effective item in the list\n",
    "    a=0\n",
    "    for k, v in ASXL1_sites_nonsense5.items():\n",
    "        for i in v: #because some contain more than 1 change from that codon:\n",
    "            separating_positions[a]=i\n",
    "            a=a+1\n",
    "    \n",
    "    for k, v in separating_positions.items():\n",
    "        base_pos = v[0]\n",
    "        original_base = v[1]\n",
    "        new_base = v[2]\n",
    "        site_trio = v[3]\n",
    "        codon_n_912 = v[4]\n",
    "        variant = v[5]\n",
    "        sequence = (str(original_base),str(new_base))\n",
    "        changes = '>'.join(sequence)\n",
    "        left_base = site_trio[0][0]\n",
    "        right_base = site_trio[0][2]\n",
    "        sequence2 = (changes, right_base)\n",
    "        changes2 = ']'.join(sequence2)\n",
    "        sequence3 = (left_base, changes2)\n",
    "        changes3 = '['.join(sequence3)\n",
    "        ASXL1_sites_nonsense6[k] = (v[0],changes3, codon_n_912, variant) #site_trio = the affected base and its neighbours\n",
    "        \n",
    "        \n",
    "    #Assign the mutation rates to each site...\n",
    "    ASXL1_mutation_rates_nonsense = {}\n",
    "    for k, v in ASXL1_sites_nonsense6.items():\n",
    "        base_change_info = v[1]\n",
    "        if base_change_info in Mutation_rates.keys():\n",
    "            mutation_rate = Mutation_rates.get(base_change_info)\n",
    "            ASXL1_mutation_rates_nonsense[k] = (v, mutation_rate)\n",
    "\n",
    "    ASXL1_mutation_rate_nonsense_sum = []\n",
    "    mutation_rates_list = []\n",
    "    for k, v in ASXL1_mutation_rates_nonsense.items():\n",
    "    #     print(v[1])\n",
    "        aa_pos = v[0][2]\n",
    "        site = v[0][1]\n",
    "        mutation_rate_value = v[1] \n",
    "        variant = v[0][3]\n",
    "        mutation_rates_list.append((aa_pos, site, mutation_rate_value, variant))\n",
    "        ASXL1_mutation_rate_nonsense_sum.append(mutation_rate_value)\n",
    "        \n",
    "    ASXL1_Summed_mutation_rate_nonsense_added = np.sum(ASXL1_mutation_rate_nonsense_sum)\n",
    "\n",
    "    print('Total summed mutation rate across ASXL1 nonsense for this study = ', ASXL1_Summed_mutation_rate_nonsense_added)\n",
    "    \n",
    "    mutation_rates_list_all = mutation_rates_list\n",
    "    return mutation_rates_list_all #in format: column 1 = 882, column 2 = p.R882*"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 295,
   "metadata": {
    "code_folding": [
     0
    ]
   },
   "outputs": [],
   "source": [
    "def mutation_rates_list_missense_variants_ASXL1(csv_file_variants):\n",
    "    \n",
    "    #The mutation rate at a base is calculated based on its trinucleotide context.  To obtain the trinucleotide context, the base either side of the mutated base needs to be ascertained.\n",
    "    #The coding sequence (+ an extra base at each end) of each gene is used when determining the trinucleotide context of the base change.  The problem with using the coding sequence is that\n",
    "    #if a base change falls at the end of an exon, its neighbouring base will actuall be an intronic base, which may be different to the 'neighbouring' base in the coding sequence.\n",
    "    #To account for this, there is an \"intron_exon_check\" function that is called at the intron/exon boundary if the neighbouring base in the intron is different to the neighbouring base in the coding sequence.\n",
    "    \n",
    "    intron_exon_boundary_caution_base = [140, 143, 252, 373, 1086, 1085]\n",
    "   \n",
    "    def intron_exon_check(position_in_gene_number):\n",
    "        if position_in_gene_number == 140:\n",
    "            gene_pos = 140\n",
    "            first_base = ASXL1_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = ASXL1_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 143:\n",
    "            gene_pos = 143\n",
    "            first_base = ASXL1_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = ASXL1_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 252:\n",
    "            gene_pos = 252\n",
    "            first_base = ASXL1_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = ASXL1_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 373:\n",
    "            gene_pos = 373\n",
    "            first_base = ASXL1_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = ASXL1_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 1086:\n",
    "            gene_pos = 1086\n",
    "            first_base = 'G'\n",
    "            middle_base = ASXL1_exons_extra_bases[gene_pos]\n",
    "            third_base = ASXL1_exons_extra_bases[gene_pos+1]\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 1085:\n",
    "            gene_pos = 1085\n",
    "            first_base = ASXL1_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = ASXL1_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        return triplet\n",
    "    \n",
    "    #ASXL1 transcript:\n",
    "    #Downloaded from http://grch37.ensembl.org\n",
    "    ASXL1_exons = 'ATGAAGGACAAACAGAAGAAGAAGAAGGAGCGCACGTGGGCCGAGGCCGCGCGCCTGGTATTAGAAAACTACTCGGATGCTCCAATGACACCAAAACAGATTCTGCAGGTCATAGAGGCAGAAGGACTAAAGGAAATGAGAAGTGGGACTTCCCCTCTCGCATGCCTCAATGCTATGCTACATTCCAATTCAAGAGGAGGAGAGGGGTTGTTTTATAAACTGCCTGGCCGAATCAGCCTTTTCACGCTCAAGAAGGATGCCCTGCAGTGGTCTCGCCATCCAGCTACAGTGGAGGGAGAGGAGCCAGAGGACACGGCTGATGTGGAGAGCTGTGGGTCTAATGAAGCCAGCACTGTGAGTGGTGAAAACGATGTATCTCTTGATGAAACATCTTCGAACGCATCCTGTTCTACAGAATCTCAGAGTCGACCTCTTTCCAATCCCAGGGACAGCTACAGAGCTTCCTCACAGGCGAACAAACAAAAGAAAAAGACTGGGGTGATGCTGCCTCGAGTTGTCCTGACTCCTCTGAAGGTAAACGGGGCCCACGTGGAATCTGCATCAGGGTTCTCGGGCTGCCACGCCGATGGCGAGAGCGGCAGCCCGTCCAGCAGCAGCAGCGGCTCTCTGGCCCTGGGCAGCGCTGCTATTCGTGGCCAGGCCGAGGTCACCCAGGACCCTGCCCCGCTCCTGAGAGGCTTCCGGAAGCCAGCCACAGGTCAAATGAAGCGCAACAGAGGGGAAGAAATAGATTTTGAGACACCTGGGTCCATTCTTGTCAACACCAACCTCCGTGCCCTGATCAACTCTCGGACCTTCCATGCCTTACCATCACACTTCCAGCAGCAGCTCCTCTTCCTCCTGCCTGAAGTAGACAGACAGGTGGGGACGGATGGCCTGTTGCGTCTCAGCAGCAGTGCACTAAATAACGAGTTTTTTACCCATGCGGCTCAGAGCTGGCGGGAGCGCCTGGCTGATGGTGAATTTACTCATGAGATGCAAGTCAGGATACGACAGGAAATGGAGAAGGAAAAGAAGGTGGAACAATGGAAAGAAAAGTTCTTTGAAGACTACTATGGACAGAAGCTGGGTTTGACCAAAGAAGAGTCATTGCAGCAGAACGTGGGCCAGGAGGAGGCTGAAATCAAAAGTGGCTTGTGTGTCCCAGGAGAATCAGTGCGTATACAGCGTGGTCCAGCCACCCGACAGCGAGATGGGCATTTTAAGAAACGCTCTCGGCCAGATCTCCGAACCAGAGCCAGAAGGAATCTGTACAAAAAACAGGAGTCAGAACAAGCAGGGGTTGCTAAGGATGCAAAATCTGTGGCCTCAGATGTTCCCCTCTACAAGGATGGGGAGGCTAAGACTGACCCAGCAGGGCTGAGCAGTCCCCATCTGCCAGGCACATCCTCTGCAGCACCCGACCTGGAGGGTCCCGAATTCCCAGTTGAGTCTGTGGCTTCTCGGATCCAGGCTGAGCCAGACAACTTGGCACGTGCCTCTGCATCTCCAGACAGAATTCCTAGCCTGCCTCAGGAAACTGTGGATCAGGAACCCAAGGATCAGAAGAGGAAATCCTTTGAGCAGGCGGCCTCTGCATCCTTTCCCGAAAAGAAGCCCCGGCTTGAAGATCGTCAGTCCTTTCGTAACACAATTGAAAGTGTTCACACCGAAAAGCCACAGCCCACTAAAGAGGAGCCCAAAGTCCCGCCCATCCGGATTCAACTTTCACGTATCAAACCACCCTGGGTGGTTAAAGGTCAGCCCACTTACCAGATATGCCCCCGGATCATCCCCACCACGGAGTCCTCCTGCCGGGGTTGGACTGGCGCCAGGACCCTCGCAGACATTAAAGCCCGTGCTCTGCAGGTCCGAGGGGCGAGAGGTCACCACTGCCATAGAGAGGCGGCCACCACTGCCATCGGAGGGGGGGGTGGCCCGGGTGGAGGTGGCGGCGGGGCCACCGATGAGGGAGGTGGCAGAGGCAGCAGCAGTGGTGATGGTGGTGAGGCCTGTGGCCACCCTGAGCCCAGGGGAGGCCCGAGCACCCCTGGAAAGTGTACGTCAGATCTACAGCGAACACAACTACTGCCGCCTTATCCTCTAAATGGGGAGCATACCCAGGCCGGAACTGCCATGTCCAGAGCTAGGAGAGAGGACCTGCCTTCTCTGAGAAAGGAGGAAAGCTGCCTACTACAGAGGGCTACAGTTGGACTCACAGATGGGCTAGGAGATGCCTCCCAACTCCCCGTTGCTCCCACTGGGGACCAGCCATGCCAGGCCTTGCCCCTACTGTCCTCCCAAACCTCAGTAGCTGAGAGATTAGTGGAGCAGCCTCAGTTGCATCCGGATGTTAGAACTGAATGTGAGTCTGGCACCACTTCCTGGGAAAGTGATGATGAGGAGCAAGGACCCACCGTTCCTGCAGACAATGGTCCCATTCTGTCTCTAGTGGGAGATGATACATTAGAGAAAGGAACTGGCCAAGCTCTTGACAGTCATCCCACTATGAAGGATCCTGTAAATGTGACCCCCAGTTCCACACCTGAATCCTCACCGACTGATTGCCTGCAGAACAGAGCATTTGATGACGAATTAGGGCTTGGTGGCTCATGCCCTCCTATGAGGGAAAGTGATACTAGACAAGAAAACTTGAAAACCAAGGCTCTCGTTTCTAACAGTTCTTTGCATTGGATACCCATCCCATCGAATGATGAGGTAGTGAAACAGCCCAAACCAGAATCCAGAGAACACATACCATCTGTTGAGCCCCAGGTTGGAGAGGAGTGGGAGAAAGCTGCTCCCACCCCTCCTGCATTGCCTGGGGATTTGACAGCTGAGGAGGGTCTAGATCCTCTTGACAGCCTTACTTCACTCTGGACTGTGCCATCTCGAGGAGGCAGTGACAGCAATGGCAGTTACTGTCAACAGGTGGACATTGAAAAGCTGAAAATCAACGGAGACTCTGAAGCACTGAGTCCTCACGGTGAGTCCACGGATACAGCCTCTGACTTTGAAGGTCACCTCACGGAGGACAGCAGTGAGGCTGACACTAGAGAAGCTGCAGTGACAAAGGGATCTTCGGTGGACAAGGATGAGAAACCCAATTGGAACCAATCTGCCCCACTGTCCAAGGTGAATGGTGACATGCGTCTGGTTACAAGGACAGATGGGATGGTTGCTCCTCAGAGCTGGGTGTCTCGAGTATGTGCGGTCCGCCAAAAGATCCCAGATTCCCTACTGCTGGCCAGTACTGAGTACCAGCCAAGAGCCGTGTGCCTGTCCATGCCTGGGTCCTCAGTGGAGGCCACTAACCCACTTGTGATGCAGTTGCTGCAGGGTAGCTTGCCCCTAGAGAAGGTTCTTCCACCAGCCCACGATGACAGCATGTCAGAATCCCCACAAGTACCACTTACAAAAGACCAGAGCCATGGCTCGCTACGCATGGGATCTTTACATGGTCTTGGAAAAAACAGTGGCATGGTTGATGGAAGCAGCCCCAGTTCTTTAAGGGCTTTGAAGGAGCCTCTTCTGCCAGATAGCTGTGAAACAGGCACTGGTCTTGCCAGGATTGAGGCCACCCAGGCTCCTGGAGCACCCCAAAAGAATTGCAAGGCAGTCCCAAGTTTTGACTCCCTCCATCCAGTGACAAATCCCATTACATCCTCTAGGAAACTGGAAGAAATGGATTCCAAAGAGCAGTTCTCTTCCTTTAGTTGTGAAGATCAGAAGGAAGTCCGTGCTATGTCACAGGACAGTAATTCAAATGCTGCTCCAGGAAAGAGCCCAGGAGATCTTACTACCTCGAGAACACCTCGTTTCTCATCTCCAAATGTGATCTCCTTTGGTCCAGAGCAGACAGGTCGGGCCCTGGGTGATCAGAGCAATGTTACAGGCCAAGGGAAGAAGCTTTTTGGCTCTGGGAATGTGGCTGCAACCCTTCAGCGCCCCAGGCCTGCGGACCCGATGCCTCTTCCTGCTGAGATCCCTCCAGTTTTTCCCAGTGGGAAGTTGGGACCAAGCACAAACTCCATGTCTGGTGGGGTACAGACTCCAAGGGAAGACTGGGCTCCAAAGCCACATGCCTTTGTTGGCAGCGTCAAGAATGAGAAGACTTTTGTGGGGGGTCCTCTTAAGGCAAATGCCGAGAACAGGAAAGCTACTGGGCATAGTCCCCTGGAACTGGTGGGTCACTTGGAAGGGATGCCCTTTGTCATGGACTTGCCCTTCTGGAAATTACCCCGAGAGCCAGGGAAGGGGCTCAGTGAGCCTCTGGAGCCTTCTTCTCTCCCCTCCCAACTCAGCATCAAGCAGGCATTTTATGGGAAGCTTTCTAAACTCCAACTGAGTTCCACCAGCTTTAATTATTCCTCTAGCTCTCCCACCTTTCCCAAAGGCCTTGCTGGAAGTGTGGTGCAGCTGAGCCACAAAGCAAACTTTGGTGCGAGCCACAGTGCATCACTTTCCTTGCAAATGTTCACTGACAGCAGCACGGTGGAAAGCATCTCGCTCCAGTGTGCGTGCAGCCTGAAAGCCATGATCATGTGCCAAGGCTGCGGTGCGTTCTGTCACGATGACTGTATTGGACCCTCAAAGCTCTGTGTATTGTGCCTTGTGGTGAGATAA'\n",
    "    \n",
    "    #Split the sequence into codons \n",
    "    ASXL1_codons = [ASXL1_exons[i:i+3] for i in range(0, len(ASXL1_exons), 3)]\n",
    "\n",
    "    #ASXL1 transcript:\n",
    "    #Downloaded from http://grch37.ensembl.org\n",
    "    ASXL1_exons_extra_bases = 'GATGAAGGACAAACAGAAGAAGAAGAAGGAGCGCACGTGGGCCGAGGCCGCGCGCCTGGTATTAGAAAACTACTCGGATGCTCCAATGACACCAAAACAGATTCTGCAGGTCATAGAGGCAGAAGGACTAAAGGAAATGAGAAGTGGGACTTCCCCTCTCGCATGCCTCAATGCTATGCTACATTCCAATTCAAGAGGAGGAGAGGGGTTGTTTTATAAACTGCCTGGCCGAATCAGCCTTTTCACGCTCAAGAAGGATGCCCTGCAGTGGTCTCGCCATCCAGCTACAGTGGAGGGAGAGGAGCCAGAGGACACGGCTGATGTGGAGAGCTGTGGGTCTAATGAAGCCAGCACTGTGAGTGGTGAAAACGATGTATCTCTTGATGAAACATCTTCGAACGCATCCTGTTCTACAGAATCTCAGAGTCGACCTCTTTCCAATCCCAGGGACAGCTACAGAGCTTCCTCACAGGCGAACAAACAAAAGAAAAAGACTGGGGTGATGCTGCCTCGAGTTGTCCTGACTCCTCTGAAGGTAAACGGGGCCCACGTGGAATCTGCATCAGGGTTCTCGGGCTGCCACGCCGATGGCGAGAGCGGCAGCCCGTCCAGCAGCAGCAGCGGCTCTCTGGCCCTGGGCAGCGCTGCTATTCGTGGCCAGGCCGAGGTCACCCAGGACCCTGCCCCGCTCCTGAGAGGCTTCCGGAAGCCAGCCACAGGTCAAATGAAGCGCAACAGAGGGGAAGAAATAGATTTTGAGACACCTGGGTCCATTCTTGTCAACACCAACCTCCGTGCCCTGATCAACTCTCGGACCTTCCATGCCTTACCATCACACTTCCAGCAGCAGCTCCTCTTCCTCCTGCCTGAAGTAGACAGACAGGTGGGGACGGATGGCCTGTTGCGTCTCAGCAGCAGTGCACTAAATAACGAGTTTTTTACCCATGCGGCTCAGAGCTGGCGGGAGCGCCTGGCTGATGGTGAATTTACTCATGAGATGCAAGTCAGGATACGACAGGAAATGGAGAAGGAAAAGAAGGTGGAACAATGGAAAGAAAAGTTCTTTGAAGACTACTATGGACAGAAGCTGGGTTTGACCAAAGAAGAGTCATTGCAGCAGAACGTGGGCCAGGAGGAGGCTGAAATCAAAAGTGGCTTGTGTGTCCCAGGAGAATCAGTGCGTATACAGCGTGGTCCAGCCACCCGACAGCGAGATGGGCATTTTAAGAAACGCTCTCGGCCAGATCTCCGAACCAGAGCCAGAAGGAATCTGTACAAAAAACAGGAGTCAGAACAAGCAGGGGTTGCTAAGGATGCAAAATCTGTGGCCTCAGATGTTCCCCTCTACAAGGATGGGGAGGCTAAGACTGACCCAGCAGGGCTGAGCAGTCCCCATCTGCCAGGCACATCCTCTGCAGCACCCGACCTGGAGGGTCCCGAATTCCCAGTTGAGTCTGTGGCTTCTCGGATCCAGGCTGAGCCAGACAACTTGGCACGTGCCTCTGCATCTCCAGACAGAATTCCTAGCCTGCCTCAGGAAACTGTGGATCAGGAACCCAAGGATCAGAAGAGGAAATCCTTTGAGCAGGCGGCCTCTGCATCCTTTCCCGAAAAGAAGCCCCGGCTTGAAGATCGTCAGTCCTTTCGTAACACAATTGAAAGTGTTCACACCGAAAAGCCACAGCCCACTAAAGAGGAGCCCAAAGTCCCGCCCATCCGGATTCAACTTTCACGTATCAAACCACCCTGGGTGGTTAAAGGTCAGCCCACTTACCAGATATGCCCCCGGATCATCCCCACCACGGAGTCCTCCTGCCGGGGTTGGACTGGCGCCAGGACCCTCGCAGACATTAAAGCCCGTGCTCTGCAGGTCCGAGGGGCGAGAGGTCACCACTGCCATAGAGAGGCGGCCACCACTGCCATCGGAGGGGGGGGTGGCCCGGGTGGAGGTGGCGGCGGGGCCACCGATGAGGGAGGTGGCAGAGGCAGCAGCAGTGGTGATGGTGGTGAGGCCTGTGGCCACCCTGAGCCCAGGGGAGGCCCGAGCACCCCTGGAAAGTGTACGTCAGATCTACAGCGAACACAACTACTGCCGCCTTATCCTCTAAATGGGGAGCATACCCAGGCCGGAACTGCCATGTCCAGAGCTAGGAGAGAGGACCTGCCTTCTCTGAGAAAGGAGGAAAGCTGCCTACTACAGAGGGCTACAGTTGGACTCACAGATGGGCTAGGAGATGCCTCCCAACTCCCCGTTGCTCCCACTGGGGACCAGCCATGCCAGGCCTTGCCCCTACTGTCCTCCCAAACCTCAGTAGCTGAGAGATTAGTGGAGCAGCCTCAGTTGCATCCGGATGTTAGAACTGAATGTGAGTCTGGCACCACTTCCTGGGAAAGTGATGATGAGGAGCAAGGACCCACCGTTCCTGCAGACAATGGTCCCATTCTGTCTCTAGTGGGAGATGATACATTAGAGAAAGGAACTGGCCAAGCTCTTGACAGTCATCCCACTATGAAGGATCCTGTAAATGTGACCCCCAGTTCCACACCTGAATCCTCACCGACTGATTGCCTGCAGAACAGAGCATTTGATGACGAATTAGGGCTTGGTGGCTCATGCCCTCCTATGAGGGAAAGTGATACTAGACAAGAAAACTTGAAAACCAAGGCTCTCGTTTCTAACAGTTCTTTGCATTGGATACCCATCCCATCGAATGATGAGGTAGTGAAACAGCCCAAACCAGAATCCAGAGAACACATACCATCTGTTGAGCCCCAGGTTGGAGAGGAGTGGGAGAAAGCTGCTCCCACCCCTCCTGCATTGCCTGGGGATTTGACAGCTGAGGAGGGTCTAGATCCTCTTGACAGCCTTACTTCACTCTGGACTGTGCCATCTCGAGGAGGCAGTGACAGCAATGGCAGTTACTGTCAACAGGTGGACATTGAAAAGCTGAAAATCAACGGAGACTCTGAAGCACTGAGTCCTCACGGTGAGTCCACGGATACAGCCTCTGACTTTGAAGGTCACCTCACGGAGGACAGCAGTGAGGCTGACACTAGAGAAGCTGCAGTGACAAAGGGATCTTCGGTGGACAAGGATGAGAAACCCAATTGGAACCAATCTGCCCCACTGTCCAAGGTGAATGGTGACATGCGTCTGGTTACAAGGACAGATGGGATGGTTGCTCCTCAGAGCTGGGTGTCTCGAGTATGTGCGGTCCGCCAAAAGATCCCAGATTCCCTACTGCTGGCCAGTACTGAGTACCAGCCAAGAGCCGTGTGCCTGTCCATGCCTGGGTCCTCAGTGGAGGCCACTAACCCACTTGTGATGCAGTTGCTGCAGGGTAGCTTGCCCCTAGAGAAGGTTCTTCCACCAGCCCACGATGACAGCATGTCAGAATCCCCACAAGTACCACTTACAAAAGACCAGAGCCATGGCTCGCTACGCATGGGATCTTTACATGGTCTTGGAAAAAACAGTGGCATGGTTGATGGAAGCAGCCCCAGTTCTTTAAGGGCTTTGAAGGAGCCTCTTCTGCCAGATAGCTGTGAAACAGGCACTGGTCTTGCCAGGATTGAGGCCACCCAGGCTCCTGGAGCACCCCAAAAGAATTGCAAGGCAGTCCCAAGTTTTGACTCCCTCCATCCAGTGACAAATCCCATTACATCCTCTAGGAAACTGGAAGAAATGGATTCCAAAGAGCAGTTCTCTTCCTTTAGTTGTGAAGATCAGAAGGAAGTCCGTGCTATGTCACAGGACAGTAATTCAAATGCTGCTCCAGGAAAGAGCCCAGGAGATCTTACTACCTCGAGAACACCTCGTTTCTCATCTCCAAATGTGATCTCCTTTGGTCCAGAGCAGACAGGTCGGGCCCTGGGTGATCAGAGCAATGTTACAGGCCAAGGGAAGAAGCTTTTTGGCTCTGGGAATGTGGCTGCAACCCTTCAGCGCCCCAGGCCTGCGGACCCGATGCCTCTTCCTGCTGAGATCCCTCCAGTTTTTCCCAGTGGGAAGTTGGGACCAAGCACAAACTCCATGTCTGGTGGGGTACAGACTCCAAGGGAAGACTGGGCTCCAAAGCCACATGCCTTTGTTGGCAGCGTCAAGAATGAGAAGACTTTTGTGGGGGGTCCTCTTAAGGCAAATGCCGAGAACAGGAAAGCTACTGGGCATAGTCCCCTGGAACTGGTGGGTCACTTGGAAGGGATGCCCTTTGTCATGGACTTGCCCTTCTGGAAATTACCCCGAGAGCCAGGGAAGGGGCTCAGTGAGCCTCTGGAGCCTTCTTCTCTCCCCTCCCAACTCAGCATCAAGCAGGCATTTTATGGGAAGCTTTCTAAACTCCAACTGAGTTCCACCAGCTTTAATTATTCCTCTAGCTCTCCCACCTTTCCCAAAGGCCTTGCTGGAAGTGTGGTGCAGCTGAGCCACAAAGCAAACTTTGGTGCGAGCCACAGTGCATCACTTTCCTTGCAAATGTTCACTGACAGCAGCACGGTGGAAAGCATCTCGCTCCAGTGTGCGTGCAGCCTGAAAGCCATGATCATGTGCCAAGGCTGCGGTGCGTTCTGTCACGATGACTGTATTGGACCCTCAAAGCTCTGTGTATTGTGCCTTGTGGTGAGATAAT'\n",
    "  \n",
    "    #Create a list of the codons they called missense variants at\n",
    "    import csv\n",
    "    filename = csv_file_variants\n",
    "    with open(filename, 'r') as csvfile:\n",
    "        read_reader = csv.reader(csvfile)  #csv.reader returns a reader object which will iterate over lines in the csvfile\n",
    "        row_count=0\n",
    "\n",
    "        ASXL1_missense_variant_list = []\n",
    "        ASXL1_missense_codons =[]\n",
    "        ASXL1_missense_variant_DICT = {}\n",
    "\n",
    "        for row in read_reader: #each row contains aa number (e.g. 11) and variant (e.g. p.D11A)\n",
    "    #         print(row)\n",
    "            if row_count>0: \n",
    "                codon0 = int(row[0])-1\n",
    "                codon1 = int(row[0])\n",
    "                variant = row[1]\n",
    "                ASXL1_missense_codons.append(codon0) #List of codon numbers that they called variants at\n",
    "                ASXL1_missense_variant_DICT[row_count]=(codon1, variant) #Dictionary:  K = codon number, V = variant at that codon (e.g. 76: p.V76L)\n",
    "\n",
    "            row_count=row_count+1\n",
    "            \n",
    "    #Looking at what possible amino acid changes are possible at those codons\n",
    "    ASXL1_missense_codons_DICT = {}\n",
    "    calls_ASXL1_missense = {}\n",
    "\n",
    "    for i in ASXL1_missense_codons: #Iterates over the list of codon numbers\n",
    "        ASXL1_missense_codons_DICT[i+1]= (ASXL1_codons[i]) #Creates of dicitonary of key = codon number, value = bases (sequence) at that codon\n",
    "        #e.g. 12: CGT, 882: GCA\n",
    "        \n",
    "    for k,v in ASXL1_missense_codons_DICT.items(): #iterating over the list of codon numbers (and codon sequences)\n",
    "        if v in Possible_Amino_Acid_Changes_letters.keys(): #assign the list of all the possible amino acid changes to each codon they called\n",
    "            changes = Possible_Amino_Acid_Changes_letters[v]\n",
    "    #         print(Possible_Amino_Acid_Changes.values())\n",
    "            calls_ASXL1_missense[k]=(v, changes)\n",
    "        #e.g. 12: (CGT, 'K', '*', 'T'etc...)\n",
    "        \n",
    "    #Changing the original amino acid to a Word (e.g. Phe) rather than a codon\n",
    "    calls_ASXL1_missense_extra = {}\n",
    "\n",
    "    for k, v in calls_ASXL1_missense.items():\n",
    "        if v[0] in Genetic_Code:\n",
    "            Amino_acid_change = Genetic_Code.get(v[0])\n",
    "            calls_ASXL1_missense_extra[k] = (Amino_acid_change, v[1])\n",
    "            #e.g. 12: (Phe, 'K', '*', 'T' etc..)\n",
    "\n",
    "    #Changing the original amino acid to a LETTER (rather than the codon)\n",
    "    calls_ASXL1_missense_extra2 = {}\n",
    "\n",
    "    for k, v in calls_ASXL1_missense_extra.items():\n",
    "        if v[0] in Amino_acid_abbreviations:\n",
    "            Amino_acid_abbrev = Amino_acid_abbreviations.get(v[0])\n",
    "            calls_ASXL1_missense_extra2[k] = (Amino_acid_abbrev, v[1])\n",
    "            #e.g. 12: ((F), ('K', '*', 'T' etc..))\n",
    "            \n",
    "    #Create a dictionary containing info on which position in the codon is affected to create the variant\n",
    "    ASXL1_position_in_list_dict_missense={}\n",
    "    for k, v in ASXL1_missense_variant_DICT.items(): #Iterates over the dictionary of (key = row_count, value = aa_pos, variant (e.g. 3: (76, p.V76L))\n",
    "        observed_change=(v[1])[-1] #Selects the amino acid change (e.g. the L from p.V76L)\n",
    "        possible_changes=calls_ASXL1_missense[v[0]][1] #possible changes are the second value in the dictionary of (k = codon number, value = (original amino acid, possible changes))\n",
    "        possible_changes_info=calls_ASXL1_missense[v[0]]\n",
    "        codon_seq=calls_ASXL1_missense[v[0]][0]\n",
    "        possible_changes_string = \"\".join(possible_changes)\n",
    "        if v[1][2] == calls_ASXL1_missense_extra2[v[0]][0]:  #v[1][2] = the original amino acid letter (i.e if this aa is in this transcript)\n",
    "            position_in_list = -1\n",
    "            positions_in_list = []\n",
    "            while True:\n",
    "                position_in_list = possible_changes_string.find(observed_change, position_in_list+1) #counts position in possible changes (if 0-2 = 1st base of codon, if 3-5 = 2nd base, if 6-8 = 3rd base)\n",
    "                if position_in_list == -1: break\n",
    "                positions_in_list.append(position_in_list)\n",
    "                ASXL1_position_in_list_dict_missense[k]=(positions_in_list, possible_changes_info, codon_seq, v[0], v[1])\n",
    "            \n",
    "    aa_positions_variants_called = []\n",
    "    for k, v in ASXL1_position_in_list_dict_missense.items():\n",
    "        aa_positions_variants_called.append(v[3])\n",
    "#     print('positions at which missense variants called = ', aa_positions_variants_called)\n",
    "                \n",
    "    #Create a dictionary that contains the position in the list of possible AA where the variant is and what it changes to\n",
    "    ASXL1_bases_missense = {}\n",
    "    for k, v in ASXL1_position_in_list_dict_missense.items():\n",
    "        codon_seq = v[2]\n",
    "        base_changes=[]\n",
    "        if codon_seq in Possible_Base_Changes.keys():\n",
    "            list_of_possible_bases = Possible_Base_Changes.get(codon_seq)\n",
    "        for i in v[0]:\n",
    "            base_change_position = i\n",
    "            base_change = list_of_possible_bases[base_change_position]\n",
    "            base_changes.append(base_change)\n",
    "        ASXL1_bases_missense[k]=(v, base_changes)\n",
    "        \n",
    "    #Create dictionary which contains original codon seq and details on which base is changed\n",
    "    ASXL1_bases_missense2 = {}\n",
    "\n",
    "    for k, v in ASXL1_bases_missense.items():\n",
    "        base_change_positions = v[0][0]\n",
    "        codon_positions = []\n",
    "        positions_in_gene = []\n",
    "        for i in base_change_positions:\n",
    "            base_change_position = i\n",
    "            if base_change_position <= 2:\n",
    "                codon_pos = 1\n",
    "            if 2 < base_change_position <= 5:\n",
    "                codon_pos = 2\n",
    "            if 5 < base_change_position <= 8:\n",
    "                codon_pos = 3\n",
    "            codon_positions.append(codon_pos)\n",
    "        ASXL1_bases_missense2[k] = (v, codon_positions)\n",
    "        \n",
    "    #Create dictionary which contains original codon seq, details on which base is changed and it's position in the coding region\n",
    "    ASXL1_bases_missense3 = {}\n",
    "\n",
    "    for k, v in ASXL1_bases_missense2.items():\n",
    "        codon_positions = v[1]\n",
    "        base_changes = v[0][1]\n",
    "        original_bases = []\n",
    "        positions_in_gene = []\n",
    "        codon_number = v[0][0][3]\n",
    "        variant = v[0][0][4]\n",
    "        for i in codon_positions:\n",
    "            codon_pos = i\n",
    "            position_in_gene = ((int(v[0][0][3])-1)*3)+codon_pos #codon number (v[0][0][3]) minus 1 (e.g. codon 307 -1 = 306, then times by 3 takes you to beginning of codon 307 (then add the position))  \n",
    "            positions_in_gene.append(position_in_gene)\n",
    "            original_base = (v[0][0][2][codon_pos-1])\n",
    "            original_bases.append(original_base)\n",
    "        ASXL1_bases_missense3[k] = (v[0][0][1], positions_in_gene, original_bases, base_changes, codon_number, variant)\n",
    "\n",
    "    #Create dictionary which contains original codon seq, details on which base is changed and its neighbouring bases\n",
    "    ASXL1_bases_missense4 = {}\n",
    "\n",
    "    for k, v in ASXL1_bases_missense3.items():\n",
    "        position_in_gene = v[1]\n",
    "        sites_trio_missense=[]\n",
    "\n",
    "        for i in position_in_gene:\n",
    "    #         print(i)\n",
    "            if i in intron_exon_boundary_caution_base:\n",
    "#                 print('positions at intron_exon_boundary =',i)\n",
    "                adapted_trio = intron_exon_check(i)\n",
    "                sites_trio_missense.append([adapted_trio])\n",
    "                ASXL1_bases_missense4[k] = (v, sites_trio_missense)\n",
    "\n",
    "            else:\n",
    "                for i in position_in_gene:\n",
    "                    gene_pos = i\n",
    "                    site_trio_missense = [ASXL1_exons_extra_bases[gene_pos:gene_pos+3] for gene_pos in range (gene_pos-1, gene_pos, 3)]\n",
    "                    sites_trio_missense.append(site_trio_missense)\n",
    "                    ASXL1_bases_missense4[k] = (v, sites_trio_missense)\n",
    "                    \n",
    "    #Create a dictionary containing pos in coding region, original base, new base, base affected with its neighbours\n",
    "    #e.g. 938, G, A, TGG = base 938 on coding region, original base = G, changes to base A, base with neighbours either side = TGG\n",
    "    ASXL1_sites_missense5 = {}\n",
    "\n",
    "    a=0\n",
    "    for k, v in ASXL1_bases_missense4.items():\n",
    "        codon_n = v[0][1] #base number in codon (out of 3 x 912)\n",
    "        codon_n_912 = v[0][4]\n",
    "        original_base = v[0][2]\n",
    "        new_base = v[0][3]\n",
    "        site_trio = v[1]\n",
    "        info=[]\n",
    "        variant = v[0][5]\n",
    "\n",
    "        change = 0\n",
    "        for i in codon_n:\n",
    "            codon_number = i\n",
    "            orig_base = original_base[change]\n",
    "            n_base = new_base[change]\n",
    "            site_tri = site_trio[change]\n",
    "            info.append([codon_number, orig_base, n_base, site_tri, codon_n_912, variant])\n",
    "            change = change+1\n",
    "        ASXL1_sites_missense5[a]=info\n",
    "        a=a+1\n",
    "        \n",
    "    # Write in a way that can iterate over mutation rates (i.e. put in to form C[C>A]G etc)   \n",
    "    ASXL1_sites_missense6={}\n",
    "\n",
    "    separating_positions = {} # separate out all the possible changes to just 1 effective item in the list\n",
    "    a=0\n",
    "    for k, v in ASXL1_sites_missense5.items():\n",
    "        for i in v: #because some contain more than 1 change from that codon:\n",
    "            separating_positions[a]=i\n",
    "            a=a+1\n",
    "    \n",
    "    for k, v in separating_positions.items():\n",
    "        base_pos = v[0]\n",
    "        original_base = v[1]\n",
    "        new_base = v[2]\n",
    "        site_trio = v[3]\n",
    "        codon_n_912 = v[4]\n",
    "        variant = v[5]\n",
    "        sequence = (str(original_base),str(new_base))\n",
    "        changes = '>'.join(sequence)\n",
    "        left_base = site_trio[0][0]\n",
    "        right_base = site_trio[0][2]\n",
    "        sequence2 = (changes, right_base)\n",
    "        changes2 = ']'.join(sequence2)\n",
    "        sequence3 = (left_base, changes2)\n",
    "        changes3 = '['.join(sequence3)\n",
    "        ASXL1_sites_missense6[k] = (v[0],changes3, codon_n_912, variant) #site_trio = the affected base and its neighbours\n",
    "        \n",
    "        \n",
    "    #Assign the mutation rates to each site...\n",
    "    ASXL1_mutation_rates_missense = {}\n",
    "    for k, v in ASXL1_sites_missense6.items():\n",
    "        base_change_info = v[1]\n",
    "        if base_change_info in Mutation_rates.keys():\n",
    "            mutation_rate = Mutation_rates.get(base_change_info)\n",
    "            ASXL1_mutation_rates_missense[k] = (v, mutation_rate)\n",
    "\n",
    "    ASXL1_mutation_rate_missense_sum = []\n",
    "    mutation_rates_list = []\n",
    "    for k, v in ASXL1_mutation_rates_missense.items():\n",
    "    #     print(v[1])\n",
    "        aa_pos = v[0][2]\n",
    "        site = v[0][1]\n",
    "        mutation_rate_value = v[1] \n",
    "        variant = v[0][3]\n",
    "        mutation_rates_list.append((aa_pos, site, mutation_rate_value, variant))\n",
    "        ASXL1_mutation_rate_missense_sum.append(mutation_rate_value)\n",
    "        \n",
    "    ASXL1_Summed_mutation_rate_missense_added = np.sum(ASXL1_mutation_rate_missense_sum)\n",
    "\n",
    "    print('Total summed mutation rate across ASXL1 missense for this study = ', ASXL1_Summed_mutation_rate_missense_added)\n",
    "    \n",
    "    mutation_rates_list_all = mutation_rates_list\n",
    "    return mutation_rates_list_all #in format: column 1 = 882, column 2 = p.R882H"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### TP53"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 296,
   "metadata": {
    "code_folding": [
     0
    ]
   },
   "outputs": [],
   "source": [
    "def mutation_rates_list_nonsense_variants_TP53(csv_file_variants):\n",
    "\n",
    "    #The mutation rate at a base is calculated based on its trinucleotide context.  To obtain the trinucleotide context, the base either side of the mutated base needs to be ascertained.\n",
    "    #The coding sequence (+ an extra base at each end) of each gene is used when determining the trinucleotide context of the base change.  The problem with using the coding sequence is that\n",
    "    #if a base change falls at the end of an exon, its neighbouring base will actuall be an intronic base, which may be different to the 'neighbouring' base in the coding sequence.\n",
    "    #To account for this, there is an \"intron_exon_check\" function that is called at the intron/exon boundary if the neighbouring base in the intron is different to the neighbouring base in the coding sequence.\n",
    "    \n",
    "    intron_exon_boundary_caution_base = [74, 75, 96, 375, 672, 782, 919, 993, 1100]\n",
    "   \n",
    "    def intron_exon_check(position_in_gene_number):\n",
    "        if position_in_gene_number == 74:\n",
    "            gene_pos = 74\n",
    "            first_base = TP53_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = TP53_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 75:\n",
    "            gene_pos = 75\n",
    "            first_base = 'G'\n",
    "            middle_base = TP53_exons_extra_bases[gene_pos]\n",
    "            third_base = TP53_exons_extra_bases[gene_pos+1]\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 96:\n",
    "            gene_pos = 96\n",
    "            first_base = TP53_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = TP53_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 375:\n",
    "            gene_pos = 375\n",
    "            first_base = TP53_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = TP53_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 672:\n",
    "            gene_pos = 672\n",
    "            first_base = TP53_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = TP53_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 782:\n",
    "            gene_pos = 782\n",
    "            first_base = TP53_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = TP53_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 919:\n",
    "            gene_pos = 919\n",
    "            first_base = TP53_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = TP53_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 993:\n",
    "            gene_pos = 993\n",
    "            first_base = TP53_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = TP53_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 1100:\n",
    "            gene_pos = 1100\n",
    "            first_base = TP53_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = TP53_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        return triplet\n",
    "    \n",
    "    #TP53 transcript:\n",
    "    #Downloaded from http://grch37.ensembl.org\n",
    "    TP53_exons = 'ATGGAGGAGCCGCAGTCAGATCCTAGCGTCGAGCCCCCTCTGAGTCAGGAAACATTTTCAGACCTATGGAAACTACTTCCTGAAAACAACGTTCTGTCCCCCTTGCCGTCCCAAGCAATGGATGATTTGATGCTGTCCCCGGACGATATTGAACAATGGTTCACTGAAGACCCAGGTCCAGATGAAGCTCCCAGAATGCCAGAGGCTGCTCCCCCCGTGGCCCCTGCACCAGCAGCTCCTACACCGGCGGCCCCTGCACCAGCCCCCTCCTGGCCCCTGTCATCTTCTGTCCCTTCCCAGAAAACCTACCAGGGCAGCTACGGTTTCCGTCTGGGCTTCTTGCATTCTGGGACAGCCAAGTCTGTGACTTGCACGTACTCCCCTGCCCTCAACAAGATGTTTTGCCAACTGGCCAAGACCTGCCCTGTGCAGCTGTGGGTTGATTCCACACCCCCGCCCGGCACCCGCGTCCGCGCCATGGCCATCTACAAGCAGTCACAGCACATGACGGAGGTTGTGAGGCGCTGCCCCCACCATGAGCGCTGCTCAGATAGCGATGGTCTGGCCCCTCCTCAGCATCTTATCCGAGTGGAAGGAAATTTGCGTGTGGAGTATTTGGATGACAGAAACACTTTTCGACATAGTGTGGTGGTGCCCTATGAGCCGCCTGAGGTTGGCTCTGACTGTACCACCATCCACTACAACTACATGTGTAACAGTTCCTGCATGGGCGGCATGAACCGGAGGCCCATCCTCACCATCATCACACTGGAAGACTCCAGTGGTAATCTACTGGGACGGAACAGCTTTGAGGTGCGTGTTTGTGCCTGTCCTGGGAGAGACCGGCGCACAGAGGAAGAGAATCTCCGCAAGAAAGGGGAGCCTCACCACGAGCTGCCCCCAGGGAGCACTAAGCGAGCACTGCCCAACAACACCAGCTCCTCTCCCCAGCCAAAGAAGAAACCACTGGATGGAGAATATTTCACCCTTCAGATCCGTGGGCGTGAGCGCTTCGAGATGTTCCGAGAGCTGAATGAGGCCTTGGAACTCAAGGATGCCCAGGCTGGGAAGGAGCCAGGGGGGAGCAGGGCTCACTCCAGCCACCTGAAGTCCAAAAAGGGTCAGTCTACCTCCCGCCATAAAAAACTCATGTTCAAGACAGAAGGGCCTGACTCAGACTGA'\n",
    "    \n",
    "    #Split the sequence into codons \n",
    "    TP53_codons = [TP53_exons[i:i+3] for i in range(0, len(TP53_exons), 3)]\n",
    "\n",
    "    #TP53 transcript:\n",
    "    #Downloaded from http://grch37.ensembl.org\n",
    "    TP53_exons_extra_bases = 'CATGGAGGAGCCGCAGTCAGATCCTAGCGTCGAGCCCCCTCTGAGTCAGGAAACATTTTCAGACCTATGGAAACTACTTCCTGAAAACAACGTTCTGTCCCCCTTGCCGTCCCAAGCAATGGATGATTTGATGCTGTCCCCGGACGATATTGAACAATGGTTCACTGAAGACCCAGGTCCAGATGAAGCTCCCAGAATGCCAGAGGCTGCTCCCCCCGTGGCCCCTGCACCAGCAGCTCCTACACCGGCGGCCCCTGCACCAGCCCCCTCCTGGCCCCTGTCATCTTCTGTCCCTTCCCAGAAAACCTACCAGGGCAGCTACGGTTTCCGTCTGGGCTTCTTGCATTCTGGGACAGCCAAGTCTGTGACTTGCACGTACTCCCCTGCCCTCAACAAGATGTTTTGCCAACTGGCCAAGACCTGCCCTGTGCAGCTGTGGGTTGATTCCACACCCCCGCCCGGCACCCGCGTCCGCGCCATGGCCATCTACAAGCAGTCACAGCACATGACGGAGGTTGTGAGGCGCTGCCCCCACCATGAGCGCTGCTCAGATAGCGATGGTCTGGCCCCTCCTCAGCATCTTATCCGAGTGGAAGGAAATTTGCGTGTGGAGTATTTGGATGACAGAAACACTTTTCGACATAGTGTGGTGGTGCCCTATGAGCCGCCTGAGGTTGGCTCTGACTGTACCACCATCCACTACAACTACATGTGTAACAGTTCCTGCATGGGCGGCATGAACCGGAGGCCCATCCTCACCATCATCACACTGGAAGACTCCAGTGGTAATCTACTGGGACGGAACAGCTTTGAGGTGCGTGTTTGTGCCTGTCCTGGGAGAGACCGGCGCACAGAGGAAGAGAATCTCCGCAAGAAAGGGGAGCCTCACCACGAGCTGCCCCCAGGGAGCACTAAGCGAGCACTGCCCAACAACACCAGCTCCTCTCCCCAGCCAAAGAAGAAACCACTGGATGGAGAATATTTCACCCTTCAGATCCGTGGGCGTGAGCGCTTCGAGATGTTCCGAGAGCTGAATGAGGCCTTGGAACTCAAGGATGCCCAGGCTGGGAAGGAGCCAGGGGGGAGCAGGGCTCACTCCAGCCACCTGAAGTCCAAAAAGGGTCAGTCTACCTCCCGCCATAAAAAACTCATGTTCAAGACAGAAGGGCCTGACTCAGACTGAC'\n",
    "    \n",
    "    #Create a list of the codons they called nonsense variants at\n",
    "    import csv\n",
    "    filename = csv_file_variants\n",
    "    with open(filename, 'r') as csvfile:\n",
    "        read_reader = csv.reader(csvfile)  #csv.reader returns a reader object which will iterate over lines in the csvfile\n",
    "        row_count=0\n",
    "\n",
    "        TP53_nonsense_variant_list = []\n",
    "        TP53_nonsense_codons =[]\n",
    "        TP53_nonsense_variant_DICT = {}\n",
    "\n",
    "        for row in read_reader: #each row contains aa number (e.g. 11) and variant (e.g. p.D11A)\n",
    "    #         print(row)\n",
    "            if row_count>0: \n",
    "                codon0 = int(row[0])-1\n",
    "                codon1 = int(row[0])\n",
    "                variant = row[1]\n",
    "                TP53_nonsense_codons.append(codon0) #List of codon numbers that they called variants at\n",
    "                TP53_nonsense_variant_DICT[row_count]=(codon1, variant) #Dictionary:  K = codon number, V = variant at that codon (e.g. 76: p.V76L)\n",
    "\n",
    "            row_count=row_count+1\n",
    "            \n",
    "    #Looking at what possible amino acid changes are possible at those codons\n",
    "    TP53_nonsense_codons_DICT = {}\n",
    "    calls_nonsense = {}\n",
    "\n",
    "    for i in TP53_nonsense_codons: #Iterates over the list of codon numbers\n",
    "        TP53_nonsense_codons_DICT[i+1]= (TP53_codons[i]) #Creates of dicitonary of key = codon number, value = bases (sequence) at that codon\n",
    "        #e.g. 12: CGT, 882: GCA\n",
    "        \n",
    "    for k,v in TP53_nonsense_codons_DICT.items(): #iterating over the list of codon numbers (and codon sequences)\n",
    "        if v in Possible_Amino_Acid_Changes_letters.keys(): #assign the list of all the possible amino acid changes to each codon they called\n",
    "            changes = Possible_Amino_Acid_Changes_letters[v]\n",
    "    #         print(Possible_Amino_Acid_Changes.values())\n",
    "            calls_nonsense[k]=(v, changes)\n",
    "        #e.g. 12: (CGT, 'K', '*', 'T'etc...)\n",
    "        \n",
    "    #Changing the original amino acid to a Word (e.g. Phe) rather than a codon\n",
    "    calls_nonsense_extra = {}\n",
    "\n",
    "    for k, v in calls_nonsense.items():\n",
    "        if v[0] in Genetic_Code:\n",
    "            Amino_acid_change = Genetic_Code.get(v[0])\n",
    "            calls_nonsense_extra[k] = (Amino_acid_change, v[1])\n",
    "            #e.g. 12: (Phe, 'K', '*', 'T' etc..)\n",
    "\n",
    "    #Changing the original amino acid to a LETTER (rather than the codon)\n",
    "    calls_nonsense_extra2 = {}\n",
    "\n",
    "    for k, v in calls_nonsense_extra.items():\n",
    "        if v[0] in Amino_acid_abbreviations:\n",
    "            Amino_acid_abbrev = Amino_acid_abbreviations.get(v[0])\n",
    "            calls_nonsense_extra2[k] = (Amino_acid_abbrev, v[1])\n",
    "            #e.g. 12: ((F), ('K', '*', 'T' etc..))\n",
    "            \n",
    "    #Create a dictionary containing info on which position in the codon is affected to create the variant\n",
    "    TP53_position_in_list_dict_nonsense={}\n",
    "    for k, v in TP53_nonsense_variant_DICT.items(): #Iterates over the dictionary of (key = row_count, value = aa_pos, variant (e.g. 3: (76, p.V76L))\n",
    "        observed_change=(v[1])[-1] #Selects the amino acid change (e.g. the L from p.V76L)\n",
    "        possible_changes=calls_nonsense[v[0]][1] #possible changes are the second value in the dictionary of (k = codon number, value = (original amino acid, possible changes))\n",
    "        possible_changes_info=calls_nonsense[v[0]]\n",
    "        codon_seq=calls_nonsense[v[0]][0]\n",
    "        possible_changes_string = \"\".join(possible_changes)\n",
    "        if v[1][2] == calls_nonsense_extra2[v[0]][0]:  #v[1][2] = the original amino acid letter (i.e if this aa is in this transcript)\n",
    "            position_in_list = -1\n",
    "            positions_in_list = []\n",
    "            while True:\n",
    "                position_in_list = possible_changes_string.find(observed_change, position_in_list+1) #counts position in possible changes (if 0-2 = 1st base of codon, if 3-5 = 2nd base, if 6-8 = 3rd base)\n",
    "                if position_in_list == -1: break\n",
    "                positions_in_list.append(position_in_list)\n",
    "                TP53_position_in_list_dict_nonsense[k]=(positions_in_list, possible_changes_info, codon_seq, v[0], v[1])\n",
    "            \n",
    "    aa_positions_variants_called = []\n",
    "    for k, v in TP53_position_in_list_dict_nonsense.items():\n",
    "        aa_positions_variants_called.append(v[3])\n",
    "#     print('positions at which nonsense variants called = ', aa_positions_variants_called)\n",
    "                \n",
    "    #Create a dictionary that contains the position in the list of possible AA where the variant is and what it changes to\n",
    "    TP53_bases_nonsense = {}\n",
    "    for k, v in TP53_position_in_list_dict_nonsense.items():\n",
    "        codon_seq = v[2]\n",
    "        base_changes=[]\n",
    "        if codon_seq in Possible_Base_Changes.keys():\n",
    "            list_of_possible_bases = Possible_Base_Changes.get(codon_seq)\n",
    "        for i in v[0]:\n",
    "            base_change_position = i\n",
    "            base_change = list_of_possible_bases[base_change_position]\n",
    "            base_changes.append(base_change)\n",
    "        TP53_bases_nonsense[k]=(v, base_changes)\n",
    "        \n",
    "    #Create dictionary which contains original codon seq and details on which base is changed\n",
    "    TP53_bases_nonsense2 = {}\n",
    "\n",
    "    for k, v in TP53_bases_nonsense.items():\n",
    "        base_change_positions = v[0][0]\n",
    "        codon_positions = []\n",
    "        positions_in_gene = []\n",
    "        for i in base_change_positions:\n",
    "            base_change_position = i\n",
    "            if base_change_position <= 2:\n",
    "                codon_pos = 1\n",
    "            if 2 < base_change_position <= 5:\n",
    "                codon_pos = 2\n",
    "            if 5 < base_change_position <= 8:\n",
    "                codon_pos = 3\n",
    "            codon_positions.append(codon_pos)\n",
    "        TP53_bases_nonsense2[k] = (v, codon_positions)\n",
    "        \n",
    "    #Create dictionary which contains original codon seq, details on which base is changed and it's position in the coding region\n",
    "    TP53_bases_nonsense3 = {}\n",
    "\n",
    "    for k, v in TP53_bases_nonsense2.items():\n",
    "        codon_positions = v[1]\n",
    "        base_changes = v[0][1]\n",
    "        original_bases = []\n",
    "        positions_in_gene = []\n",
    "        codon_number = v[0][0][3]\n",
    "        variant = v[0][0][4]\n",
    "        for i in codon_positions:\n",
    "            codon_pos = i\n",
    "            position_in_gene = ((int(v[0][0][3])-1)*3)+codon_pos #codon number (v[0][0][3]) minus 1 (e.g. codon 307 -1 = 306, then times by 3 takes you to beginning of codon 307 (then add the position))  \n",
    "            positions_in_gene.append(position_in_gene)\n",
    "            original_base = (v[0][0][2][codon_pos-1])\n",
    "            original_bases.append(original_base)\n",
    "        TP53_bases_nonsense3[k] = (v[0][0][1], positions_in_gene, original_bases, base_changes, codon_number, variant)\n",
    "\n",
    "    #Create dictionary which contains original codon seq, details on which base is changed and its neighbouring bases\n",
    "    TP53_bases_nonsense4 = {}\n",
    "\n",
    "    for k, v in TP53_bases_nonsense3.items():\n",
    "        position_in_gene = v[1]\n",
    "        sites_trio_nonsense=[]\n",
    "\n",
    "        for i in position_in_gene:\n",
    "    #         print(i)\n",
    "            if i in intron_exon_boundary_caution_base:\n",
    "#                 print('positions at intron_exon_boundary =',i)\n",
    "                adapted_trio = intron_exon_check(i)\n",
    "                sites_trio_nonsense.append([adapted_trio])\n",
    "                TP53_bases_nonsense4[k] = (v, sites_trio_nonsense)\n",
    "\n",
    "            else:\n",
    "                for i in position_in_gene:\n",
    "                    gene_pos = i\n",
    "                    site_trio_nonsense = [TP53_exons_extra_bases[gene_pos:gene_pos+3] for gene_pos in range (gene_pos-1, gene_pos, 3)]\n",
    "                    sites_trio_nonsense.append(site_trio_nonsense)\n",
    "                    TP53_bases_nonsense4[k] = (v, sites_trio_nonsense)\n",
    "                    \n",
    "    #Create a dictionary containing pos in coding region, original base, new base, base affected with its neighbours\n",
    "    #e.g. 938, G, A, TGG = base 938 on coding region, original base = G, changes to base A, base with neighbours either side = TGG\n",
    "    TP53_sites_nonsense5 = {}\n",
    "\n",
    "    a=0\n",
    "    for k, v in TP53_bases_nonsense4.items():\n",
    "        codon_n = v[0][1] #base number in codon (out of 3 x 912)\n",
    "        codon_n_912 = v[0][4]\n",
    "        original_base = v[0][2]\n",
    "        new_base = v[0][3]\n",
    "        site_trio = v[1]\n",
    "        info=[]\n",
    "        variant = v[0][5]\n",
    "\n",
    "        change = 0\n",
    "        for i in codon_n:\n",
    "            codon_number = i\n",
    "            orig_base = original_base[change]\n",
    "            n_base = new_base[change]\n",
    "            site_tri = site_trio[change]\n",
    "            info.append([codon_number, orig_base, n_base, site_tri, codon_n_912, variant])\n",
    "            change = change+1\n",
    "        TP53_sites_nonsense5[a]=info\n",
    "        a=a+1\n",
    "#     print(TP53_sites_nonsense5)\n",
    "        \n",
    "    # Write in a way that can iterate over mutation rates (i.e. put in to form C[C>A]G etc)   \n",
    "    TP53_sites_nonsense6={}\n",
    "\n",
    "    separating_positions = {} # separate out all the possible changes to just 1 effective item in the list\n",
    "    a=0\n",
    "    for k, v in TP53_sites_nonsense5.items():\n",
    "        for i in v: #because some contain more than 1 change from that codon:\n",
    "            separating_positions[a]=i\n",
    "            a=a+1\n",
    "    \n",
    "    for k, v in separating_positions.items():\n",
    "        base_pos = v[0]\n",
    "        original_base = v[1]\n",
    "        new_base = v[2]\n",
    "        site_trio = v[3]\n",
    "        codon_n_912 = v[4]\n",
    "        variant = v[5]\n",
    "        sequence = (str(original_base),str(new_base))\n",
    "        changes = '>'.join(sequence)\n",
    "        left_base = site_trio[0][0]\n",
    "        right_base = site_trio[0][2]\n",
    "        sequence2 = (changes, right_base)\n",
    "        changes2 = ']'.join(sequence2)\n",
    "        sequence3 = (left_base, changes2)\n",
    "        changes3 = '['.join(sequence3)\n",
    "        TP53_sites_nonsense6[k] = (v[0],changes3, codon_n_912, variant) #site_trio = the affected base and its neighbours\n",
    "        \n",
    "        \n",
    "    #Assign the mutation rates to each site...\n",
    "    TP53_mutation_rates_nonsense = {}\n",
    "    for k, v in TP53_sites_nonsense6.items():\n",
    "        base_change_info = v[1]\n",
    "        if base_change_info in Mutation_rates.keys():\n",
    "            mutation_rate = Mutation_rates.get(base_change_info)\n",
    "            TP53_mutation_rates_nonsense[k] = (v, mutation_rate)\n",
    "\n",
    "    TP53_mutation_rate_nonsense_sum = []\n",
    "    mutation_rates_list = []\n",
    "    for k, v in TP53_mutation_rates_nonsense.items():\n",
    "    #     print(v[1])\n",
    "        aa_pos = v[0][2]\n",
    "        site = v[0][1]\n",
    "        mutation_rate_value = v[1] \n",
    "        variant = v[0][3]\n",
    "        mutation_rates_list.append((aa_pos, site, mutation_rate_value, variant))\n",
    "        TP53_mutation_rate_nonsense_sum.append(mutation_rate_value)\n",
    "    \n",
    "    TP53_Summed_mutation_rate_nonsense_added = np.sum(TP53_mutation_rate_nonsense_sum)\n",
    "\n",
    "    print('Total summed mutation rate across TP53 nonsense for this study = ', TP53_Summed_mutation_rate_nonsense_added)\n",
    "    \n",
    "    mutation_rates_list_all = mutation_rates_list\n",
    "    return mutation_rates_list_all #in format: column 1 = 882, column 2 = p.R882*"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 297,
   "metadata": {
    "code_folding": [
     0
    ]
   },
   "outputs": [],
   "source": [
    "def mutation_rates_list_missense_variants_TP53(csv_file_variants):\n",
    "    \n",
    "    #The mutation rate at a base is calculated based on its trinucleotide context.  To obtain the trinucleotide context, the base either side of the mutated base needs to be ascertained.\n",
    "    #The coding sequence (+ an extra base at each end) of each gene is used when determining the trinucleotide context of the base change.  The problem with using the coding sequence is that\n",
    "    #if a base change falls at the end of an exon, its neighbouring base will actuall be an intronic base, which may be different to the 'neighbouring' base in the coding sequence.\n",
    "    #To account for this, there is an \"intron_exon_check\" function that is called at the intron/exon boundary if the neighbouring base in the intron is different to the neighbouring base in the coding sequence.\n",
    "    \n",
    "    intron_exon_boundary_caution_base_TP53 = [74, 75, 96, 375, 672, 782, 919, 993, 1100]\n",
    "   \n",
    "    def intron_exon_check(position_in_gene_number):\n",
    "        if position_in_gene_number == 74:\n",
    "            gene_pos = 74\n",
    "            first_base = TP53_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = TP53_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 75:\n",
    "            gene_pos = 75\n",
    "            first_base = 'G'\n",
    "            middle_base = TP53_exons_extra_bases[gene_pos]\n",
    "            third_base = TP53_exons_extra_bases[gene_pos+1]\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 96:\n",
    "            gene_pos = 96\n",
    "            first_base = TP53_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = TP53_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 375:\n",
    "            gene_pos = 375\n",
    "            first_base = TP53_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = TP53_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 672:\n",
    "            gene_pos = 672\n",
    "            first_base = TP53_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = TP53_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 782:\n",
    "            gene_pos = 782\n",
    "            first_base = TP53_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = TP53_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 919:\n",
    "            gene_pos = 919\n",
    "            first_base = TP53_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = TP53_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 993:\n",
    "            gene_pos = 993\n",
    "            first_base = TP53_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = TP53_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 1100:\n",
    "            gene_pos = 1100\n",
    "            first_base = TP53_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = TP53_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        return triplet\n",
    "    \n",
    "    #TP53 transcript:\n",
    "    #Downloaded from http://grch37.ensembl.org\n",
    "    TP53_exons = 'ATGGAGGAGCCGCAGTCAGATCCTAGCGTCGAGCCCCCTCTGAGTCAGGAAACATTTTCAGACCTATGGAAACTACTTCCTGAAAACAACGTTCTGTCCCCCTTGCCGTCCCAAGCAATGGATGATTTGATGCTGTCCCCGGACGATATTGAACAATGGTTCACTGAAGACCCAGGTCCAGATGAAGCTCCCAGAATGCCAGAGGCTGCTCCCCCCGTGGCCCCTGCACCAGCAGCTCCTACACCGGCGGCCCCTGCACCAGCCCCCTCCTGGCCCCTGTCATCTTCTGTCCCTTCCCAGAAAACCTACCAGGGCAGCTACGGTTTCCGTCTGGGCTTCTTGCATTCTGGGACAGCCAAGTCTGTGACTTGCACGTACTCCCCTGCCCTCAACAAGATGTTTTGCCAACTGGCCAAGACCTGCCCTGTGCAGCTGTGGGTTGATTCCACACCCCCGCCCGGCACCCGCGTCCGCGCCATGGCCATCTACAAGCAGTCACAGCACATGACGGAGGTTGTGAGGCGCTGCCCCCACCATGAGCGCTGCTCAGATAGCGATGGTCTGGCCCCTCCTCAGCATCTTATCCGAGTGGAAGGAAATTTGCGTGTGGAGTATTTGGATGACAGAAACACTTTTCGACATAGTGTGGTGGTGCCCTATGAGCCGCCTGAGGTTGGCTCTGACTGTACCACCATCCACTACAACTACATGTGTAACAGTTCCTGCATGGGCGGCATGAACCGGAGGCCCATCCTCACCATCATCACACTGGAAGACTCCAGTGGTAATCTACTGGGACGGAACAGCTTTGAGGTGCGTGTTTGTGCCTGTCCTGGGAGAGACCGGCGCACAGAGGAAGAGAATCTCCGCAAGAAAGGGGAGCCTCACCACGAGCTGCCCCCAGGGAGCACTAAGCGAGCACTGCCCAACAACACCAGCTCCTCTCCCCAGCCAAAGAAGAAACCACTGGATGGAGAATATTTCACCCTTCAGATCCGTGGGCGTGAGCGCTTCGAGATGTTCCGAGAGCTGAATGAGGCCTTGGAACTCAAGGATGCCCAGGCTGGGAAGGAGCCAGGGGGGAGCAGGGCTCACTCCAGCCACCTGAAGTCCAAAAAGGGTCAGTCTACCTCCCGCCATAAAAAACTCATGTTCAAGACAGAAGGGCCTGACTCAGACTGA'\n",
    "    \n",
    "    #Split the sequence into codons \n",
    "    TP53_codons = [TP53_exons[i:i+3] for i in range(0, len(TP53_exons), 3)]\n",
    "\n",
    "    #TP53 transcript:\n",
    "    #Downloaded from http://grch37.ensembl.org\n",
    "    TP53_exons_extra_bases = 'CATGGAGGAGCCGCAGTCAGATCCTAGCGTCGAGCCCCCTCTGAGTCAGGAAACATTTTCAGACCTATGGAAACTACTTCCTGAAAACAACGTTCTGTCCCCCTTGCCGTCCCAAGCAATGGATGATTTGATGCTGTCCCCGGACGATATTGAACAATGGTTCACTGAAGACCCAGGTCCAGATGAAGCTCCCAGAATGCCAGAGGCTGCTCCCCCCGTGGCCCCTGCACCAGCAGCTCCTACACCGGCGGCCCCTGCACCAGCCCCCTCCTGGCCCCTGTCATCTTCTGTCCCTTCCCAGAAAACCTACCAGGGCAGCTACGGTTTCCGTCTGGGCTTCTTGCATTCTGGGACAGCCAAGTCTGTGACTTGCACGTACTCCCCTGCCCTCAACAAGATGTTTTGCCAACTGGCCAAGACCTGCCCTGTGCAGCTGTGGGTTGATTCCACACCCCCGCCCGGCACCCGCGTCCGCGCCATGGCCATCTACAAGCAGTCACAGCACATGACGGAGGTTGTGAGGCGCTGCCCCCACCATGAGCGCTGCTCAGATAGCGATGGTCTGGCCCCTCCTCAGCATCTTATCCGAGTGGAAGGAAATTTGCGTGTGGAGTATTTGGATGACAGAAACACTTTTCGACATAGTGTGGTGGTGCCCTATGAGCCGCCTGAGGTTGGCTCTGACTGTACCACCATCCACTACAACTACATGTGTAACAGTTCCTGCATGGGCGGCATGAACCGGAGGCCCATCCTCACCATCATCACACTGGAAGACTCCAGTGGTAATCTACTGGGACGGAACAGCTTTGAGGTGCGTGTTTGTGCCTGTCCTGGGAGAGACCGGCGCACAGAGGAAGAGAATCTCCGCAAGAAAGGGGAGCCTCACCACGAGCTGCCCCCAGGGAGCACTAAGCGAGCACTGCCCAACAACACCAGCTCCTCTCCCCAGCCAAAGAAGAAACCACTGGATGGAGAATATTTCACCCTTCAGATCCGTGGGCGTGAGCGCTTCGAGATGTTCCGAGAGCTGAATGAGGCCTTGGAACTCAAGGATGCCCAGGCTGGGAAGGAGCCAGGGGGGAGCAGGGCTCACTCCAGCCACCTGAAGTCCAAAAAGGGTCAGTCTACCTCCCGCCATAAAAAACTCATGTTCAAGACAGAAGGGCCTGACTCAGACTGAC'\n",
    "    \n",
    "    #Create a list of the codons they called missense variants at\n",
    "    import csv\n",
    "    filename = csv_file_variants\n",
    "    with open(filename, 'r') as csvfile:\n",
    "        read_reader = csv.reader(csvfile)  #csv.reader returns a reader object which will iterate over lines in the csvfile\n",
    "        row_count=0\n",
    "\n",
    "        TP53_missense_variant_list = []\n",
    "        TP53_missense_codons =[]\n",
    "        TP53_missense_variant_DICT = {}\n",
    "\n",
    "        for row in read_reader: #each row contains aa number (e.g. 11) and variant (e.g. p.D11A)\n",
    "    #         print(row)\n",
    "            if row_count>0: \n",
    "                codon0 = int(row[0])-1\n",
    "                codon1 = int(row[0])\n",
    "                variant = row[1]\n",
    "                TP53_missense_codons.append(codon0) #List of codon numbers that they called variants at\n",
    "                TP53_missense_variant_DICT[row_count]=(codon1, variant) #Dictionary:  K = codon number, V = variant at that codon (e.g. 76: p.V76L)\n",
    "\n",
    "            row_count=row_count+1\n",
    "            \n",
    "    #Looking at what possible amino acid changes are possible at those codons\n",
    "    TP53_missense_codons_DICT = {}\n",
    "    calls_TP53_missense = {}\n",
    "\n",
    "    for i in TP53_missense_codons: #Iterates over the list of codon numbers\n",
    "        TP53_missense_codons_DICT[i+1]= (TP53_codons[i]) #Creates of dicitonary of key = codon number, value = bases (sequence) at that codon\n",
    "        #e.g. 12: CGT, 882: GCA\n",
    "        \n",
    "    for k,v in TP53_missense_codons_DICT.items(): #iterating over the list of codon numbers (and codon sequences)\n",
    "        if v in Possible_Amino_Acid_Changes_letters.keys(): #assign the list of all the possible amino acid changes to each codon they called\n",
    "            changes = Possible_Amino_Acid_Changes_letters[v]\n",
    "    #         print(Possible_Amino_Acid_Changes.values())\n",
    "            calls_TP53_missense[k]=(v, changes)\n",
    "        #e.g. 12: (CGT, 'K', '*', 'T'etc...)\n",
    "        \n",
    "    #Changing the original amino acid to a Word (e.g. Phe) rather than a codon\n",
    "    calls_TP53_missense_extra = {}\n",
    "\n",
    "    for k, v in calls_TP53_missense.items():\n",
    "        if v[0] in Genetic_Code:\n",
    "            Amino_acid_change = Genetic_Code.get(v[0])\n",
    "            calls_TP53_missense_extra[k] = (Amino_acid_change, v[1])\n",
    "            #e.g. 12: (Phe, 'K', '*', 'T' etc..)\n",
    "\n",
    "    #Changing the original amino acid to a LETTER (rather than the codon)\n",
    "    calls_TP53_missense_extra2 = {}\n",
    "\n",
    "    for k, v in calls_TP53_missense_extra.items():\n",
    "        if v[0] in Amino_acid_abbreviations:\n",
    "            Amino_acid_abbrev = Amino_acid_abbreviations.get(v[0])\n",
    "            calls_TP53_missense_extra2[k] = (Amino_acid_abbrev, v[1])\n",
    "            #e.g. 12: ((F), ('K', '*', 'T' etc..))\n",
    "            \n",
    "    #Create a dictionary containing info on which position in the codon is affected to create the variant\n",
    "    TP53_position_in_list_dict_missense={}\n",
    "    for k, v in TP53_missense_variant_DICT.items(): #Iterates over the dictionary of (key = row_count, value = aa_pos, variant (e.g. 3: (76, p.V76L))\n",
    "        observed_change=(v[1])[-1] #Selects the amino acid change (e.g. the L from p.V76L)\n",
    "        possible_changes=calls_TP53_missense[v[0]][1] #possible changes are the second value in the dictionary of (k = codon number, value = (original amino acid, possible changes))\n",
    "        possible_changes_info=calls_TP53_missense[v[0]]\n",
    "        codon_seq=calls_TP53_missense[v[0]][0]\n",
    "        possible_changes_string = \"\".join(possible_changes)\n",
    "        if v[1][2] == calls_TP53_missense_extra2[v[0]][0]:  #v[1][2] = the original amino acid letter (i.e if this aa is in this transcript)\n",
    "            position_in_list = -1\n",
    "            positions_in_list = []\n",
    "            while True:\n",
    "                position_in_list = possible_changes_string.find(observed_change, position_in_list+1) #counts position in possible changes (if 0-2 = 1st base of codon, if 3-5 = 2nd base, if 6-8 = 3rd base)\n",
    "                if position_in_list == -1: break\n",
    "                positions_in_list.append(position_in_list)\n",
    "                TP53_position_in_list_dict_missense[k]=(positions_in_list, possible_changes_info, codon_seq, v[0], v[1])\n",
    "            \n",
    "    aa_positions_variants_called = []\n",
    "    for k, v in TP53_position_in_list_dict_missense.items():\n",
    "        aa_positions_variants_called.append(v[3])\n",
    "#     print('positions at which missense variants called = ', aa_positions_variants_called)\n",
    "                \n",
    "    #Create a dictionary that contains the position in the list of possible AA where the variant is and what it changes to\n",
    "    TP53_bases_missense = {}\n",
    "    for k, v in TP53_position_in_list_dict_missense.items():\n",
    "        codon_seq = v[2]\n",
    "        base_changes=[]\n",
    "        if codon_seq in Possible_Base_Changes.keys():\n",
    "            list_of_possible_bases = Possible_Base_Changes.get(codon_seq)\n",
    "        for i in v[0]:\n",
    "            base_change_position = i\n",
    "            base_change = list_of_possible_bases[base_change_position]\n",
    "            base_changes.append(base_change)\n",
    "        TP53_bases_missense[k]=(v, base_changes)\n",
    "        \n",
    "    #Create dictionary which contains original codon seq and details on which base is changed\n",
    "    TP53_bases_missense2 = {}\n",
    "\n",
    "    for k, v in TP53_bases_missense.items():\n",
    "        base_change_positions = v[0][0]\n",
    "        codon_positions = []\n",
    "        positions_in_gene = []\n",
    "        for i in base_change_positions:\n",
    "            base_change_position = i\n",
    "            if base_change_position <= 2:\n",
    "                codon_pos = 1\n",
    "            if 2 < base_change_position <= 5:\n",
    "                codon_pos = 2\n",
    "            if 5 < base_change_position <= 8:\n",
    "                codon_pos = 3\n",
    "            codon_positions.append(codon_pos)\n",
    "        TP53_bases_missense2[k] = (v, codon_positions)\n",
    "        \n",
    "    #Create dictionary which contains original codon seq, details on which base is changed and it's position in the coding region\n",
    "    TP53_bases_missense3 = {}\n",
    "\n",
    "    for k, v in TP53_bases_missense2.items():\n",
    "        codon_positions = v[1]\n",
    "        base_changes = v[0][1]\n",
    "        original_bases = []\n",
    "        positions_in_gene = []\n",
    "        codon_number = v[0][0][3]\n",
    "        variant = v[0][0][4]\n",
    "        for i in codon_positions:\n",
    "            codon_pos = i\n",
    "            position_in_gene = ((int(v[0][0][3])-1)*3)+codon_pos #codon number (v[0][0][3]) minus 1 (e.g. codon 307 -1 = 306, then times by 3 takes you to beginning of codon 307 (then add the position))  \n",
    "            positions_in_gene.append(position_in_gene)\n",
    "            original_base = (v[0][0][2][codon_pos-1])\n",
    "            original_bases.append(original_base)\n",
    "        TP53_bases_missense3[k] = (v[0][0][1], positions_in_gene, original_bases, base_changes, codon_number, variant)\n",
    "\n",
    "    #Create dictionary which contains original codon seq, details on which base is changed and its neighbouring bases\n",
    "    TP53_bases_missense4 = {}\n",
    "\n",
    "    for k, v in TP53_bases_missense3.items():\n",
    "        position_in_gene = v[1]\n",
    "        sites_trio_missense=[]\n",
    "\n",
    "        for i in position_in_gene:\n",
    "    #         print(i)\n",
    "            if i in intron_exon_boundary_caution_base_TP53:\n",
    "#                 print('positions at intron_exon_boundary =',i)\n",
    "                adapted_trio = intron_exon_check(i)\n",
    "                sites_trio_missense.append([adapted_trio])\n",
    "                TP53_bases_missense4[k] = (v, sites_trio_missense)\n",
    "\n",
    "            else:\n",
    "                for i in position_in_gene:\n",
    "                    gene_pos = i\n",
    "                    site_trio_missense = [TP53_exons_extra_bases[gene_pos:gene_pos+3] for gene_pos in range (gene_pos-1, gene_pos, 3)]\n",
    "                    sites_trio_missense.append(site_trio_missense)\n",
    "                    TP53_bases_missense4[k] = (v, sites_trio_missense)\n",
    "                    \n",
    "    #Create a dictionary containing pos in coding region, original base, new base, base affected with its neighbours\n",
    "    #e.g. 938, G, A, TGG = base 938 on coding region, original base = G, changes to base A, base with neighbours either side = TGG\n",
    "    TP53_sites_missense5 = {}\n",
    "\n",
    "    a=0\n",
    "    for k, v in TP53_bases_missense4.items():\n",
    "        codon_n = v[0][1] #base number in codon (out of 3 x 912)\n",
    "        codon_n_912 = v[0][4]\n",
    "        original_base = v[0][2]\n",
    "        new_base = v[0][3]\n",
    "        site_trio = v[1]\n",
    "        info=[]\n",
    "        variant = v[0][5]\n",
    "\n",
    "        change = 0\n",
    "        for i in codon_n:\n",
    "            codon_number = i\n",
    "            orig_base = original_base[change]\n",
    "            n_base = new_base[change]\n",
    "            site_tri = site_trio[change]\n",
    "            info.append([codon_number, orig_base, n_base, site_tri, codon_n_912, variant])\n",
    "            change = change+1\n",
    "        TP53_sites_missense5[a]=info\n",
    "        a=a+1\n",
    "        \n",
    "    # Write in a way that can iterate over mutation rates (i.e. put in to form C[C>A]G etc)   \n",
    "    TP53_sites_missense6={}\n",
    "\n",
    "    separating_positions = {} # separate out all the possible changes to just 1 effective item in the list\n",
    "    a=0\n",
    "    for k, v in TP53_sites_missense5.items():\n",
    "        for i in v: #because some contain more than 1 change from that codon:\n",
    "            separating_positions[a]=i\n",
    "            a=a+1\n",
    "    \n",
    "    for k, v in separating_positions.items():\n",
    "        base_pos = v[0]\n",
    "        original_base = v[1]\n",
    "        new_base = v[2]\n",
    "        site_trio = v[3]\n",
    "        codon_n_912 = v[4]\n",
    "        variant = v[5]\n",
    "        sequence = (str(original_base),str(new_base))\n",
    "        changes = '>'.join(sequence)\n",
    "        left_base = site_trio[0][0]\n",
    "        right_base = site_trio[0][2]\n",
    "        sequence2 = (changes, right_base)\n",
    "        changes2 = ']'.join(sequence2)\n",
    "        sequence3 = (left_base, changes2)\n",
    "        changes3 = '['.join(sequence3)\n",
    "        TP53_sites_missense6[k] = (v[0],changes3, codon_n_912, variant) #site_trio = the affected base and its neighbours\n",
    "        \n",
    "        \n",
    "    #Assign the mutation rates to each site...\n",
    "    TP53_mutation_rates_missense = {}\n",
    "    for k, v in TP53_sites_missense6.items():\n",
    "        base_change_info = v[1]\n",
    "        if base_change_info in Mutation_rates.keys():\n",
    "            mutation_rate = Mutation_rates.get(base_change_info)\n",
    "            TP53_mutation_rates_missense[k] = (v, mutation_rate)\n",
    "\n",
    "    TP53_mutation_rate_missense_sum = []\n",
    "    mutation_rates_list = []\n",
    "    for k, v in TP53_mutation_rates_missense.items():\n",
    "    #     print(v[1])\n",
    "        aa_pos = v[0][2]\n",
    "        site = v[0][1]\n",
    "        mutation_rate_value = v[1] \n",
    "        variant = v[0][3]\n",
    "        mutation_rates_list.append((aa_pos, site, mutation_rate_value, variant))\n",
    "        TP53_mutation_rate_missense_sum.append(mutation_rate_value)\n",
    "        \n",
    "    TP53_Summed_mutation_rate_missense_added = np.sum(TP53_mutation_rate_missense_sum)\n",
    "\n",
    "    print('Total summed mutation rate across TP53 missense for this study = ', TP53_Summed_mutation_rate_missense_added)\n",
    "    \n",
    "    mutation_rates_list_all = mutation_rates_list\n",
    "    return mutation_rates_list_all #in format: column 1 = 882, column 2 = p.R882H"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Calculating mutation rates from a list of variants"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### DNMT3A"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 298,
   "metadata": {
    "code_folding": [
     0
    ]
   },
   "outputs": [],
   "source": [
    "def mutation_rates_nonsense_range_of_variants_DNMT3A_NP072046(lower_range, upper_range):\n",
    "    \n",
    "    #The mutation rate at a base is calculated based on its trinucleotide context.  To obtain the trinucleotide context, the base either side of the mutated base needs to be ascertained.\n",
    "    #The coding sequence (+ an extra base at each end) of each gene is used when determining the trinucleotide context of the base change.  The problem with using the coding sequence is that\n",
    "    #if a base change falls at the end of an exon, its neighbouring base will actuall be an intronic base, which may be different to the 'neighbouring' base in the coding sequence.\n",
    "    #To account for this, there is an \"intron_exon_check\" function that is called at the intron/exon boundary if the neighbouring base in the intron is different to the neighbouring base in the coding sequence.\n",
    "    \n",
    "    intron_exon_boundary_caution_base_DNMT3A_NP072046 = [1279, 1431, 1476, 1554, 1852, 2083, 2085, 2175, 2322, 2478]\n",
    "\n",
    "    def intron_exon_check_DNMT3A_NP072046(position_in_gene_number):\n",
    "        if position_in_gene_number == 1279:\n",
    "            gene_pos = 1279\n",
    "            first_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos-1]\n",
    "            middle_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 1429:\n",
    "            gene_pos = 1429\n",
    "            first_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos-1]\n",
    "            middle_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 1474:\n",
    "            gene_pos = 1474\n",
    "            first_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos-1]\n",
    "            middle_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 1554:\n",
    "            gene_pos = 1554\n",
    "            first_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos-1]\n",
    "            middle_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 1852:\n",
    "            gene_pos = 1852\n",
    "            first_base = 'G'\n",
    "            middle_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos]\n",
    "            third_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos+1]\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 2083:\n",
    "            gene_pos = 2083\n",
    "            first_base = 'G'\n",
    "            middle_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos]\n",
    "            third_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos+1]\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 2082:\n",
    "            gene_pos = 2082\n",
    "            first_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos-1]\n",
    "            middle_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 2173:\n",
    "            gene_pos = 2173\n",
    "            first_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos-1]\n",
    "            middle_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 2322:\n",
    "            gene_pos = 2322\n",
    "            first_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos-1]\n",
    "            middle_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 2478:\n",
    "            gene_pos = 2478\n",
    "            first_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos-1]\n",
    "            middle_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        return triplet\n",
    "\n",
    "    #DNMT3A transcript used:\n",
    "    #Downloaded from http://grch37.ensembl.org/Homo_sapiens/Transcript/Exons?db=core;g=ENSG00000119772;r=2:25455845-25564754;t=ENST00000321117\n",
    "    DNMT3A_exons_NP_072046 = 'ATGCCCGCCATGCCCTCCAGCGGCCCCGGGGACACCAGCAGCTCTGCTGCGGAGCGGGAGGAGGACCGAAAGGACGGAGAGGAGCAGGAGGAGCCGCGTGGCAAGGAGGAGCGCCAAGAGCCCAGCACCACGGCACGGAAGGTGGGGCGGCCTGGGAGGAAGCGCAAGCACCCCCCGGTGGAAAGCGGTGACACGCCAAAGGACCCTGCGGTGATCTCCAAGTCCCCATCCATGGCCCAGGACTCAGGCGCCTCAGAGCTATTACCCAATGGGGACTTGGAGAAGCGGAGTGAGCCCCAGCCAGAGGAGGGGAGCCCTGCTGGGGGGCAGAAGGGCGGGGCCCCAGCAGAGGGAGAGGGTGCAGCTGAGACCCTGCCTGAAGCCTCAAGAGCAGTGGAAAATGGCTGCTGCACCCCCAAGGAGGGCCGAGGAGCCCCTGCAGAAGCGGGCAAAGAACAGAAGGAGACCAACATCGAATCCATGAAAATGGAGGGCTCCCGGGGCCGGCTGCGGGGTGGCTTGGGCTGGGAGTCCAGCCTCCGTCAGCGGCCCATGCCGAGGCTCACCTTCCAGGCGGGGGACCCCTACTACATCAGCAAGCGCAAGCGGGACGAGTGGCTGGCACGCTGGAAAAGGGAGGCTGAGAAGAAAGCCAAGGTCATTGCAGGAATGAATGCTGTGGAAGAAAACCAGGGGCCCGGGGAGTCTCAGAAGGTGGAGGAGGCCAGCCCTCCTGCTGTGCAGCAGCCCACTGACCCCGCATCCCCCACTGTGGCTACCACGCCTGAGCCCGTGGGGTCCGATGCTGGGGACAAGAATGCCACCAAAGCAGGCGATGACGAGCCAGAGTACGAGGACGGCCGGGGCTTTGGCATTGGGGAGCTGGTGTGGGGGAAACTGCGGGGCTTCTCCTGGTGGCCAGGCCGCATTGTGTCTTGGTGGATGACGGGCCGGAGCCGAGCAGCTGAAGGCACCCGCTGGGTCATGTGGTTCGGAGACGGCAAATTCTCAGTGGTGTGTGTTGAGAAGCTGATGCCGCTGAGCTCGTTTTGCAGTGCGTTCCACCAGGCCACGTACAACAAGCAGCCCATGTACCGCAAAGCCATCTACGAGGTCCTGCAGGTGGCCAGCAGCCGCGCGGGGAAGCTGTTCCCGGTGTGCCACGACAGCGATGAGAGTGACACTGCCAAGGCCGTGGAGGTGCAGAACAAGCCCATGATTGAATGGGCCCTGGGGGGCTTCCAGCCTTCTGGCCCTAAGGGCCTGGAGCCACCAGAAGAAGAGAAGAATCCCTACAAAGAAGTGTACACGGACATGTGGGTGGAACCTGAGGCAGCTGCCTACGCACCACCTCCACCAGCCAAAAAGCCCCGGAAGAGCACAGCGGAGAAGCCCAAGGTCAAGGAGATTATTGATGAGCGCACAAGAGAGCGGCTGGTGTACGAGGTGCGGCAGAAGTGCCGGAACATTGAGGACATCTGCATCTCCTGTGGGAGCCTCAATGTTACCCTGGAACACCCCCTCTTCGTTGGAGGAATGTGCCAAAACTGCAAGAACTGCTTTCTGGAGTGTGCGTACCAGTACGACGACGACGGCTACCAGTCCTACTGCACCATCTGCTGTGGGGGCCGTGAGGTGCTCATGTGCGGAAACAACAACTGCTGCAGGTGCTTTTGCGTGGAGTGTGTGGACCTCTTGGTGGGGCCGGGGGCTGCCCAGGCAGCCATTAAGGAAGACCCCTGGAACTGCTACATGTGCGGGCACAAGGGTACCTACGGGCTGCTGCGGCGGCGAGAGGACTGGCCCTCCCGGCTCCAGATGTTCTTCGCTAATAACCACGACCAGGAATTTGACCCTCCAAAGGTTTACCCACCTGTCCCAGCTGAGAAGAGGAAGCCCATCCGGGTGCTGTCTCTCTTTGATGGAATCGCTACAGGGCTCCTGGTGCTGAAGGACTTGGGCATTCAGGTGGACCGCTACATTGCCTCGGAGGTGTGTGAGGACTCCATCACGGTGGGCATGGTGCGGCACCAGGGGAAGATCATGTACGTCGGGGACGTCCGCAGCGTCACACAGAAGCATATCCAGGAGTGGGGCCCATTCGATCTGGTGATTGGGGGCAGTCCCTGCAATGACCTCTCCATCGTCAACCCTGCTCGCAAGGGCCTCTACGAGGGCACTGGCCGGCTCTTCTTTGAGTTCTACCGCCTCCTGCATGATGCGCGGCCCAAGGAGGGAGATGATCGCCCCTTCTTCTGGCTCTTTGAGAATGTGGTGGCCATGGGCGTTAGTGACAAGAGGGACATCTCGCGATTTCTCGAGTCCAACCCTGTGATGATTGATGCCAAAGAAGTGTCAGCTGCACACAGGGCCCGCTACTTCTGGGGTAACCTTCCCGGTATGAACAGGCCGTTGGCATCCACTGTGAATGATAAGCTGGAGCTGCAGGAGTGTCTGGAGCATGGCAGGATAGCCAAGTTCAGCAAAGTGAGGACCATTACTACGAGGTCAAACTCCATAAAGCAGGGCAAAGACCAGCATTTTCCTGTCTTCATGAATGAGAAAGAGGACATCTTATGGTGCACTGAAATGGAAAGGGTATTTGGTTTCCCAGTCCACTATACTGACGTCTCCAACATGAGCCGCTTGGCGAGGCAGAGACTGCTGGGCCGGTCATGGAGCGTGCCAGTCATCCGCCACCTCTTCGCTCCGCTGAAGGAGTATTTTGCGTGTGTGTAA'\n",
    "\n",
    "    #Split the sequence into codons \n",
    "    DNMT3A_codons_NP_072046 = [DNMT3A_exons_NP_072046[i:i+3] for i in range(0, len(DNMT3A_exons_NP_072046), 3)]\n",
    "\n",
    "    #DNMT3A transcript used:\n",
    "    #Downloaded from http://grch37.ensembl.org/Homo_sapiens/Transcript/Exons?db=core;g=ENSG00000119772;r=2:25455845-25564754;t=ENST00000321117\n",
    "    DNMT3A_exons_extra_bases_NP_072046 = 'GATGCCCGCCATGCCCTCCAGCGGCCCCGGGGACACCAGCAGCTCTGCTGCGGAGCGGGAGGAGGACCGAAAGGACGGAGAGGAGCAGGAGGAGCCGCGTGGCAAGGAGGAGCGCCAAGAGCCCAGCACCACGGCACGGAAGGTGGGGCGGCCTGGGAGGAAGCGCAAGCACCCCCCGGTGGAAAGCGGTGACACGCCAAAGGACCCTGCGGTGATCTCCAAGTCCCCATCCATGGCCCAGGACTCAGGCGCCTCAGAGCTATTACCCAATGGGGACTTGGAGAAGCGGAGTGAGCCCCAGCCAGAGGAGGGGAGCCCTGCTGGGGGGCAGAAGGGCGGGGCCCCAGCAGAGGGAGAGGGTGCAGCTGAGACCCTGCCTGAAGCCTCAAGAGCAGTGGAAAATGGCTGCTGCACCCCCAAGGAGGGCCGAGGAGCCCCTGCAGAAGCGGGCAAAGAACAGAAGGAGACCAACATCGAATCCATGAAAATGGAGGGCTCCCGGGGCCGGCTGCGGGGTGGCTTGGGCTGGGAGTCCAGCCTCCGTCAGCGGCCCATGCCGAGGCTCACCTTCCAGGCGGGGGACCCCTACTACATCAGCAAGCGCAAGCGGGACGAGTGGCTGGCACGCTGGAAAAGGGAGGCTGAGAAGAAAGCCAAGGTCATTGCAGGAATGAATGCTGTGGAAGAAAACCAGGGGCCCGGGGAGTCTCAGAAGGTGGAGGAGGCCAGCCCTCCTGCTGTGCAGCAGCCCACTGACCCCGCATCCCCCACTGTGGCTACCACGCCTGAGCCCGTGGGGTCCGATGCTGGGGACAAGAATGCCACCAAAGCAGGCGATGACGAGCCAGAGTACGAGGACGGCCGGGGCTTTGGCATTGGGGAGCTGGTGTGGGGGAAACTGCGGGGCTTCTCCTGGTGGCCAGGCCGCATTGTGTCTTGGTGGATGACGGGCCGGAGCCGAGCAGCTGAAGGCACCCGCTGGGTCATGTGGTTCGGAGACGGCAAATTCTCAGTGGTGTGTGTTGAGAAGCTGATGCCGCTGAGCTCGTTTTGCAGTGCGTTCCACCAGGCCACGTACAACAAGCAGCCCATGTACCGCAAAGCCATCTACGAGGTCCTGCAGGTGGCCAGCAGCCGCGCGGGGAAGCTGTTCCCGGTGTGCCACGACAGCGATGAGAGTGACACTGCCAAGGCCGTGGAGGTGCAGAACAAGCCCATGATTGAATGGGCCCTGGGGGGCTTCCAGCCTTCTGGCCCTAAGGGCCTGGAGCCACCAGAAGAAGAGAAGAATCCCTACAAAGAAGTGTACACGGACATGTGGGTGGAACCTGAGGCAGCTGCCTACGCACCACCTCCACCAGCCAAAAAGCCCCGGAAGAGCACAGCGGAGAAGCCCAAGGTCAAGGAGATTATTGATGAGCGCACAAGAGAGCGGCTGGTGTACGAGGTGCGGCAGAAGTGCCGGAACATTGAGGACATCTGCATCTCCTGTGGGAGCCTCAATGTTACCCTGGAACACCCCCTCTTCGTTGGAGGAATGTGCCAAAACTGCAAGAACTGCTTTCTGGAGTGTGCGTACCAGTACGACGACGACGGCTACCAGTCCTACTGCACCATCTGCTGTGGGGGCCGTGAGGTGCTCATGTGCGGAAACAACAACTGCTGCAGGTGCTTTTGCGTGGAGTGTGTGGACCTCTTGGTGGGGCCGGGGGCTGCCCAGGCAGCCATTAAGGAAGACCCCTGGAACTGCTACATGTGCGGGCACAAGGGTACCTACGGGCTGCTGCGGCGGCGAGAGGACTGGCCCTCCCGGCTCCAGATGTTCTTCGCTAATAACCACGACCAGGAATTTGACCCTCCAAAGGTTTACCCACCTGTCCCAGCTGAGAAGAGGAAGCCCATCCGGGTGCTGTCTCTCTTTGATGGAATCGCTACAGGGCTCCTGGTGCTGAAGGACTTGGGCATTCAGGTGGACCGCTACATTGCCTCGGAGGTGTGTGAGGACTCCATCACGGTGGGCATGGTGCGGCACCAGGGGAAGATCATGTACGTCGGGGACGTCCGCAGCGTCACACAGAAGCATATCCAGGAGTGGGGCCCATTCGATCTGGTGATTGGGGGCAGTCCCTGCAATGACCTCTCCATCGTCAACCCTGCTCGCAAGGGCCTCTACGAGGGCACTGGCCGGCTCTTCTTTGAGTTCTACCGCCTCCTGCATGATGCGCGGCCCAAGGAGGGAGATGATCGCCCCTTCTTCTGGCTCTTTGAGAATGTGGTGGCCATGGGCGTTAGTGACAAGAGGGACATCTCGCGATTTCTCGAGTCCAACCCTGTGATGATTGATGCCAAAGAAGTGTCAGCTGCACACAGGGCCCGCTACTTCTGGGGTAACCTTCCCGGTATGAACAGGCCGTTGGCATCCACTGTGAATGATAAGCTGGAGCTGCAGGAGTGTCTGGAGCATGGCAGGATAGCCAAGTTCAGCAAAGTGAGGACCATTACTACGAGGTCAAACTCCATAAAGCAGGGCAAAGACCAGCATTTTCCTGTCTTCATGAATGAGAAAGAGGACATCTTATGGTGCACTGAAATGGAAAGGGTATTTGGTTTCCCAGTCCACTATACTGACGTCTCCAACATGAGCCGCTTGGCGAGGCAGAGACTGCTGGGCCGGTCATGGAGCGTGCCAGTCATCCGCCACCTCTTCGCTCCGCTGAAGGAGTATTTTGCGTGTGTGTAAG'\n",
    "\n",
    "    #List of all amino acid positions (across whole coding region) covered (amino acid position minus 1 (as python starts at 0))\n",
    "    DNMT3A_nonsense_codons_DICT_NP_072046={}\n",
    "    for i in range(lower_range, upper_range):\n",
    "        DNMT3A_nonsense_codons_DICT_NP_072046[i+1]= (DNMT3A_codons_NP_072046[i])\n",
    "\n",
    "    calls_DNMT3A_nonsense_NP_072046 = {}\n",
    "    for k,v in DNMT3A_nonsense_codons_DICT_NP_072046.items(): #iterating over the list of codon numbers (and codon sequences)\n",
    "        if v in Possible_Amino_Acid_Changes.keys(): #assign the list of all the possible amino acid changes to each codon they called\n",
    "            changes = Possible_Amino_Acid_Changes_letters[v]\n",
    "    #         print(Possible_Amino_Acid_Changes.values())\n",
    "            calls_DNMT3A_nonsense_NP_072046[k]=(v, changes)\n",
    "\n",
    "    #Changing the original amino acid to a Word (e.g. Phe) rather than a codon\n",
    "    calls_DNMT3A_nonsense_NP_072046_extra = {}\n",
    "\n",
    "    for k, v in calls_DNMT3A_nonsense_NP_072046.items():\n",
    "        if v[0] in Genetic_Code:\n",
    "            Amino_acid_change = Genetic_Code.get(v[0])\n",
    "            calls_DNMT3A_nonsense_NP_072046_extra[k] = (Amino_acid_change, v[0], v[1])\n",
    "            #e.g. 12: (Phe, 'K', '*', 'T' etc..)\n",
    "\n",
    "    #Changing the original amino acid to a LETTER (rather than the codon)\n",
    "    calls_DNMT3A_nonsense_NP_072046_extra2 = {}\n",
    "\n",
    "    for k, v in calls_DNMT3A_nonsense_NP_072046_extra.items():\n",
    "        if v[0] in Amino_acid_abbreviations:\n",
    "            Amino_acid_abbrev = Amino_acid_abbreviations.get(v[0])\n",
    "            calls_DNMT3A_nonsense_NP_072046_extra2[k] = (Amino_acid_abbrev, v[1], v[2])\n",
    "            #e.g. 12: ((F), ('K', '*', 'T' etc..))\n",
    "\n",
    "    #Create a dictionary that contains details of the positions of nonsense variants at that codon\n",
    "    DNMT3A_position_in_list_dict_nonsense_NP_072046 = {}\n",
    "\n",
    "    for k, v in calls_DNMT3A_nonsense_NP_072046_extra2.items():\n",
    "        original_AA = v[0]\n",
    "        possible_AA_changes = v[2]\n",
    "        possible_AA_changes_string = ''.join(v[2])\n",
    "        positions_in_list = []\n",
    "        for i in possible_AA_changes:\n",
    "            if i == '*': #nonsense\n",
    "                position_in_list=-1\n",
    "                nonsense_change = i    \n",
    "                while True:\n",
    "                    position_in_list = possible_AA_changes_string.find(nonsense_change, position_in_list+1)\n",
    "    #               print(position_in_list)\n",
    "                    if position_in_list == -1: break\n",
    "                    positions_in_list.append(position_in_list)\n",
    "                    positions_in_list = set(positions_in_list) #convert into a 'set' to remove duplicate positions\n",
    "                    positions_in_list = list(positions_in_list) #convert back to a list so can be added to dictionary\n",
    "                DNMT3A_position_in_list_dict_nonsense_NP_072046[k]=(positions_in_list, v)\n",
    "\n",
    "        aa_positions_variants_called = []\n",
    "        for k, v in DNMT3A_position_in_list_dict_nonsense_NP_072046.items():\n",
    "            aa_positions_variants_called.append(k)\n",
    "    # print('positions at which nonsense variants called = ', aa_positions_variants_called)\n",
    "\n",
    "    #Create a dictionary that contains the position in the list of possible AA where the variant is and what it changes to\n",
    "    DNMT3A_bases_nonsense_NP_072046 = {}\n",
    "    for k, v in DNMT3A_position_in_list_dict_nonsense_NP_072046.items():\n",
    "        original_a_a = v[1][0]\n",
    "        codon_seq = v[1][1]\n",
    "        base_changes=[]\n",
    "        if codon_seq in Possible_Base_Changes.keys():\n",
    "            list_of_possible_bases = Possible_Base_Changes.get(codon_seq)\n",
    "        for i in v[0]:\n",
    "            base_change_position = i\n",
    "            base_change = list_of_possible_bases[base_change_position]\n",
    "            base_changes.append(base_change)\n",
    "        DNMT3A_bases_nonsense_NP_072046[k]=(v, base_changes, original_a_a)\n",
    "\n",
    "    #Create dictionary which contains original codon seq and details on which base is changed\n",
    "    DNMT3A_bases_nonsense2_NP_072046 = {}\n",
    "\n",
    "    for k, v in DNMT3A_bases_nonsense_NP_072046.items():\n",
    "        base_change_positions = v[0][0]\n",
    "        codon_positions = []\n",
    "        positions_in_gene = []\n",
    "        for i in base_change_positions:\n",
    "            base_change_position = i\n",
    "            if base_change_position <= 2:\n",
    "                codon_pos = 1\n",
    "            if 2 < base_change_position <= 5:\n",
    "                codon_pos = 2\n",
    "            if 5 < base_change_position <= 8:\n",
    "                codon_pos = 3\n",
    "            codon_positions.append(codon_pos)\n",
    "        DNMT3A_bases_nonsense2_NP_072046[k] = (v, codon_positions)\n",
    "\n",
    "    #Create dictionary which contains original codon seq, details on which base is changed and it's position in the coding region\n",
    "    DNMT3A_bases_nonsense3_NP_072046 = {}\n",
    "\n",
    "    for k, v in DNMT3A_bases_nonsense2_NP_072046.items():\n",
    "        codon_positions = v[1]\n",
    "        base_changes = v[0][1]\n",
    "        original_a_a = v[0][2]\n",
    "        original_bases = []\n",
    "        positions_in_gene = []\n",
    "        codon_number = k\n",
    "        for i in codon_positions:\n",
    "            codon_pos = i\n",
    "            position_in_gene = ((int(k)-1)*3)+codon_pos #codon number minus 1 (e.g. codon 307 -1 = 306, then times by 3 takes you to beginning of codon 307 (then add the position))  \n",
    "            positions_in_gene.append(position_in_gene)\n",
    "            original_base = (v[0][0][1][1][codon_pos-1])\n",
    "            original_bases.append(original_base)\n",
    "        DNMT3A_bases_nonsense3_NP_072046[k] = (v[0][0][1], positions_in_gene, original_bases, base_changes, codon_number, original_a_a)\n",
    "\n",
    "    #Create dictionary which contains original codon seq, details on which base is changed and its neighbouring bases\n",
    "    DNMT3A_bases_nonsense4_NP_072046 = {}\n",
    "\n",
    "    for k, v in DNMT3A_bases_nonsense3_NP_072046.items():\n",
    "        position_in_gene = v[1]\n",
    "        sites_trio_nonsense=[]\n",
    "\n",
    "        for i in position_in_gene:\n",
    "    #         print(i)\n",
    "            if i in intron_exon_boundary_caution_base_DNMT3A_NP072046:\n",
    "#                 print('positions at intron_exon_boundary =',i)\n",
    "                adapted_trio = intron_exon_check_DNMT3A_NP072046(i)\n",
    "                sites_trio_nonsense.append([adapted_trio])\n",
    "                DNMT3A_bases_nonsense4_NP_072046[k] = (v, sites_trio_nonsense)\n",
    "\n",
    "            else:\n",
    "                for i in position_in_gene:\n",
    "                    gene_pos = i\n",
    "                    site_trio_nonsense = [DNMT3A_exons_extra_bases_NP_072046[gene_pos:gene_pos+3] for gene_pos in range (gene_pos-1, gene_pos, 3)]\n",
    "                    sites_trio_nonsense.append(site_trio_nonsense)\n",
    "                    DNMT3A_bases_nonsense4_NP_072046[k] = (v, sites_trio_nonsense)\n",
    "\n",
    "    #Create a dictionary containing pos in coding region, original base, new base, base affected with its neighbours\n",
    "    #e.g. 938, G, A, TGG = base 938 on coding region, original base = G, changes to base A, base with neighbours either side = TGG\n",
    "    DNMT3A_sites_nonsense5_NP_072046 = {}\n",
    "\n",
    "    a=0\n",
    "    for k, v in DNMT3A_bases_nonsense4_NP_072046.items():\n",
    "        original_a_a = v[0][0][0]\n",
    "        codon_n = v[0][1] #base number in codon (out of 3 x 912)\n",
    "        codon_n_912 = v[0][4]\n",
    "        original_base = v[0][2]\n",
    "        new_base = v[0][3]\n",
    "        site_trio = v[1]\n",
    "        info=[]\n",
    "\n",
    "        change = 0\n",
    "        for i in codon_n:\n",
    "            codon_number = i\n",
    "            orig_base = original_base[change]\n",
    "            n_base = new_base[change]\n",
    "            site_tri = site_trio[change]\n",
    "            info.append([codon_number, orig_base, n_base, site_tri, codon_n_912, original_a_a])\n",
    "            change = change+1\n",
    "        DNMT3A_sites_nonsense5_NP_072046[a]=info\n",
    "        a=a+1\n",
    "    #     print(DNMT3A_sites_nonsense5_NP_072046)\n",
    "\n",
    "    # Write in a way that can iterate over mutation rates (i.e. put in to form C[C>A]G etc)   \n",
    "    DNMT3A_sites_nonsense6_NP_072046={}\n",
    "\n",
    "    separating_positions = {} # separate out all the possible changes to just 1 effective item in the list\n",
    "    a=0\n",
    "    for k, v in DNMT3A_sites_nonsense5_NP_072046.items():\n",
    "        for i in v: #because some contain more than 1 change from that codon:\n",
    "            separating_positions[a]=i\n",
    "            a=a+1\n",
    "\n",
    "    for k, v in separating_positions.items():\n",
    "        base_pos = v[0]\n",
    "        original_base = v[1]\n",
    "        new_base = v[2]\n",
    "        site_trio = v[3]\n",
    "        codon_n_912 = v[4]\n",
    "        original_a_a = v[5]\n",
    "        sequence = (str(original_base),str(new_base))\n",
    "        changes = '>'.join(sequence)\n",
    "        left_base = site_trio[0][0]\n",
    "        right_base = site_trio[0][2]\n",
    "        sequence2 = (changes, right_base)\n",
    "        changes2 = ']'.join(sequence2)\n",
    "        sequence3 = (left_base, changes2)\n",
    "        changes3 = '['.join(sequence3)\n",
    "        DNMT3A_sites_nonsense6_NP_072046[k] = (v[0],changes3, codon_n_912, original_a_a) #site_trio = the affected base and its neighbours\n",
    "\n",
    "    #Assign the mutation rates to each site...\n",
    "    DNMT3A_mutation_rates_nonsense_NP_072046 = {}\n",
    "    for k, v in DNMT3A_sites_nonsense6_NP_072046.items():\n",
    "        base_change_info = v[1]\n",
    "        if base_change_info in Mutation_rates.keys():\n",
    "            mutation_rate = Mutation_rates.get(base_change_info)\n",
    "            original_a_a = v[3]\n",
    "            aa_positioning = v[2]\n",
    "            variant_type = '*'\n",
    "            variant_id = str(original_a_a)+str(aa_positioning)+variant_type\n",
    "            DNMT3A_mutation_rates_nonsense_NP_072046[k] = (v[0],v[1], mutation_rate, variant_id)\n",
    "\n",
    "    DNMT3A_mutation_rate_nonsense_NP_072046_sum = []\n",
    "\n",
    "    mutation_rates_list = []\n",
    "    for k, v in DNMT3A_mutation_rates_nonsense_NP_072046.items():\n",
    "        codon_position_in_gene=(v[0])\n",
    "        variant = v[3]\n",
    "        site = v[1]\n",
    "        mutation_rate_value = v[2] \n",
    "        mutation_rates_list.append((v[0], variant, site, mutation_rate_value))\n",
    "        DNMT3A_mutation_rate_nonsense_NP_072046_sum.append(mutation_rate_value)\n",
    "\n",
    "    #Get the average of all the mutation rates\n",
    "    DNMT3A_mutation_rate_nonsense_NP_072046_sum = []\n",
    "    for k, v in DNMT3A_mutation_rates_nonsense_NP_072046.items():\n",
    "        mut_rate = v[2]\n",
    "        DNMT3A_mutation_rate_nonsense_NP_072046_sum.append(mut_rate)\n",
    "\n",
    "    DNMT3A_Summed_mutation_rate_nonsense_NP_072046_added = np.sum(DNMT3A_mutation_rate_nonsense_NP_072046_sum)\n",
    "\n",
    "    print('Total summed mutation rate across DNMT3A nonsense_NP_072046 for this study = ', DNMT3A_Summed_mutation_rate_nonsense_NP_072046_added)\n",
    "\n",
    "    mutation_rates_list_all = mutation_rates_list\n",
    "    return mutation_rates_list_all # e.g. lower= 0, upper = 913 for whole gene\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 299,
   "metadata": {
    "code_folding": [
     0
    ]
   },
   "outputs": [],
   "source": [
    "def mutation_rates_missense_range_of_variants_DNMT3A_NP072046(lower_range, upper_range):\n",
    "    intron_exon_boundary_caution_base_DNMT3A_NP072046 = [1279, 1429, 1474, 1554, 1852, 2083, 2082, 2173, 2322, 2478]\n",
    "\n",
    "    #The mutation rate at a base is calculated based on its trinucleotide context.  To obtain the trinucleotide context, the base either side of the mutated base needs to be ascertained.\n",
    "    #The coding sequence (+ an extra base at each end) of each gene is used when determining the trinucleotide context of the base change.  The problem with using the coding sequence is that\n",
    "    #if a base change falls at the end of an exon, its neighbouring base will actuall be an intronic base, which may be different to the 'neighbouring' base in the coding sequence.\n",
    "    #To account for this, there is an \"intron_exon_check\" function that is called at the intron/exon boundary if the neighbouring base in the intron is different to the neighbouring base in the coding sequence.\n",
    "    \n",
    "    def intron_exon_check_DNMT3A_NP072046(position_in_gene_number):\n",
    "        if position_in_gene_number == 1279:\n",
    "            gene_pos = 1279\n",
    "            first_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos-1]\n",
    "            middle_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 1429:\n",
    "            gene_pos = 1429\n",
    "            first_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos-1]\n",
    "            middle_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 1474:\n",
    "            gene_pos = 1474\n",
    "            first_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos-1]\n",
    "            middle_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 1554:\n",
    "            gene_pos = 1554\n",
    "            first_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos-1]\n",
    "            middle_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 1852:\n",
    "            gene_pos = 1852\n",
    "            first_base = 'G'\n",
    "            middle_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos]\n",
    "            third_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos+1]\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 2083:\n",
    "            gene_pos = 2083\n",
    "            first_base = 'G'\n",
    "            middle_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos]\n",
    "            third_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos+1]\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 2082:\n",
    "            gene_pos = 2082\n",
    "            first_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos-1]\n",
    "            middle_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 2173:\n",
    "            gene_pos = 2173\n",
    "            first_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos-1]\n",
    "            middle_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 2322:\n",
    "            gene_pos = 2322\n",
    "            first_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos-1]\n",
    "            middle_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 2478:\n",
    "            gene_pos = 2478\n",
    "            first_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos-1]\n",
    "            middle_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        return triplet\n",
    "\n",
    "    #DNMT3A transcript used (NP_072046):\n",
    "    #Downloaded from http://grch37.ensembl.org/Homo_sapiens/Transcript/Exons?db=core;g=ENSG00000119772;r=2:25455845-25564754;t=ENST00000321117\n",
    "    DNMT3A_exons_NP_072046 = 'ATGCCCGCCATGCCCTCCAGCGGCCCCGGGGACACCAGCAGCTCTGCTGCGGAGCGGGAGGAGGACCGAAAGGACGGAGAGGAGCAGGAGGAGCCGCGTGGCAAGGAGGAGCGCCAAGAGCCCAGCACCACGGCACGGAAGGTGGGGCGGCCTGGGAGGAAGCGCAAGCACCCCCCGGTGGAAAGCGGTGACACGCCAAAGGACCCTGCGGTGATCTCCAAGTCCCCATCCATGGCCCAGGACTCAGGCGCCTCAGAGCTATTACCCAATGGGGACTTGGAGAAGCGGAGTGAGCCCCAGCCAGAGGAGGGGAGCCCTGCTGGGGGGCAGAAGGGCGGGGCCCCAGCAGAGGGAGAGGGTGCAGCTGAGACCCTGCCTGAAGCCTCAAGAGCAGTGGAAAATGGCTGCTGCACCCCCAAGGAGGGCCGAGGAGCCCCTGCAGAAGCGGGCAAAGAACAGAAGGAGACCAACATCGAATCCATGAAAATGGAGGGCTCCCGGGGCCGGCTGCGGGGTGGCTTGGGCTGGGAGTCCAGCCTCCGTCAGCGGCCCATGCCGAGGCTCACCTTCCAGGCGGGGGACCCCTACTACATCAGCAAGCGCAAGCGGGACGAGTGGCTGGCACGCTGGAAAAGGGAGGCTGAGAAGAAAGCCAAGGTCATTGCAGGAATGAATGCTGTGGAAGAAAACCAGGGGCCCGGGGAGTCTCAGAAGGTGGAGGAGGCCAGCCCTCCTGCTGTGCAGCAGCCCACTGACCCCGCATCCCCCACTGTGGCTACCACGCCTGAGCCCGTGGGGTCCGATGCTGGGGACAAGAATGCCACCAAAGCAGGCGATGACGAGCCAGAGTACGAGGACGGCCGGGGCTTTGGCATTGGGGAGCTGGTGTGGGGGAAACTGCGGGGCTTCTCCTGGTGGCCAGGCCGCATTGTGTCTTGGTGGATGACGGGCCGGAGCCGAGCAGCTGAAGGCACCCGCTGGGTCATGTGGTTCGGAGACGGCAAATTCTCAGTGGTGTGTGTTGAGAAGCTGATGCCGCTGAGCTCGTTTTGCAGTGCGTTCCACCAGGCCACGTACAACAAGCAGCCCATGTACCGCAAAGCCATCTACGAGGTCCTGCAGGTGGCCAGCAGCCGCGCGGGGAAGCTGTTCCCGGTGTGCCACGACAGCGATGAGAGTGACACTGCCAAGGCCGTGGAGGTGCAGAACAAGCCCATGATTGAATGGGCCCTGGGGGGCTTCCAGCCTTCTGGCCCTAAGGGCCTGGAGCCACCAGAAGAAGAGAAGAATCCCTACAAAGAAGTGTACACGGACATGTGGGTGGAACCTGAGGCAGCTGCCTACGCACCACCTCCACCAGCCAAAAAGCCCCGGAAGAGCACAGCGGAGAAGCCCAAGGTCAAGGAGATTATTGATGAGCGCACAAGAGAGCGGCTGGTGTACGAGGTGCGGCAGAAGTGCCGGAACATTGAGGACATCTGCATCTCCTGTGGGAGCCTCAATGTTACCCTGGAACACCCCCTCTTCGTTGGAGGAATGTGCCAAAACTGCAAGAACTGCTTTCTGGAGTGTGCGTACCAGTACGACGACGACGGCTACCAGTCCTACTGCACCATCTGCTGTGGGGGCCGTGAGGTGCTCATGTGCGGAAACAACAACTGCTGCAGGTGCTTTTGCGTGGAGTGTGTGGACCTCTTGGTGGGGCCGGGGGCTGCCCAGGCAGCCATTAAGGAAGACCCCTGGAACTGCTACATGTGCGGGCACAAGGGTACCTACGGGCTGCTGCGGCGGCGAGAGGACTGGCCCTCCCGGCTCCAGATGTTCTTCGCTAATAACCACGACCAGGAATTTGACCCTCCAAAGGTTTACCCACCTGTCCCAGCTGAGAAGAGGAAGCCCATCCGGGTGCTGTCTCTCTTTGATGGAATCGCTACAGGGCTCCTGGTGCTGAAGGACTTGGGCATTCAGGTGGACCGCTACATTGCCTCGGAGGTGTGTGAGGACTCCATCACGGTGGGCATGGTGCGGCACCAGGGGAAGATCATGTACGTCGGGGACGTCCGCAGCGTCACACAGAAGCATATCCAGGAGTGGGGCCCATTCGATCTGGTGATTGGGGGCAGTCCCTGCAATGACCTCTCCATCGTCAACCCTGCTCGCAAGGGCCTCTACGAGGGCACTGGCCGGCTCTTCTTTGAGTTCTACCGCCTCCTGCATGATGCGCGGCCCAAGGAGGGAGATGATCGCCCCTTCTTCTGGCTCTTTGAGAATGTGGTGGCCATGGGCGTTAGTGACAAGAGGGACATCTCGCGATTTCTCGAGTCCAACCCTGTGATGATTGATGCCAAAGAAGTGTCAGCTGCACACAGGGCCCGCTACTTCTGGGGTAACCTTCCCGGTATGAACAGGCCGTTGGCATCCACTGTGAATGATAAGCTGGAGCTGCAGGAGTGTCTGGAGCATGGCAGGATAGCCAAGTTCAGCAAAGTGAGGACCATTACTACGAGGTCAAACTCCATAAAGCAGGGCAAAGACCAGCATTTTCCTGTCTTCATGAATGAGAAAGAGGACATCTTATGGTGCACTGAAATGGAAAGGGTATTTGGTTTCCCAGTCCACTATACTGACGTCTCCAACATGAGCCGCTTGGCGAGGCAGAGACTGCTGGGCCGGTCATGGAGCGTGCCAGTCATCCGCCACCTCTTCGCTCCGCTGAAGGAGTATTTTGCGTGTGTGTAA'\n",
    "\n",
    "    #Split the sequence into codons \n",
    "    DNMT3A_codons_NP_072046 = [DNMT3A_exons_NP_072046[i:i+3] for i in range(0, len(DNMT3A_exons_NP_072046), 3)]\n",
    "\n",
    "    #DNMT3A transcript used (NP_072046):\n",
    "    #Downloaded from http://grch37.ensembl.org/Homo_sapiens/Transcript/Exons?db=core;g=ENSG00000119772;r=2:25455845-25564754;t=ENST00000321117\n",
    "    DNMT3A_exons_extra_bases_NP_072046 = 'GATGCCCGCCATGCCCTCCAGCGGCCCCGGGGACACCAGCAGCTCTGCTGCGGAGCGGGAGGAGGACCGAAAGGACGGAGAGGAGCAGGAGGAGCCGCGTGGCAAGGAGGAGCGCCAAGAGCCCAGCACCACGGCACGGAAGGTGGGGCGGCCTGGGAGGAAGCGCAAGCACCCCCCGGTGGAAAGCGGTGACACGCCAAAGGACCCTGCGGTGATCTCCAAGTCCCCATCCATGGCCCAGGACTCAGGCGCCTCAGAGCTATTACCCAATGGGGACTTGGAGAAGCGGAGTGAGCCCCAGCCAGAGGAGGGGAGCCCTGCTGGGGGGCAGAAGGGCGGGGCCCCAGCAGAGGGAGAGGGTGCAGCTGAGACCCTGCCTGAAGCCTCAAGAGCAGTGGAAAATGGCTGCTGCACCCCCAAGGAGGGCCGAGGAGCCCCTGCAGAAGCGGGCAAAGAACAGAAGGAGACCAACATCGAATCCATGAAAATGGAGGGCTCCCGGGGCCGGCTGCGGGGTGGCTTGGGCTGGGAGTCCAGCCTCCGTCAGCGGCCCATGCCGAGGCTCACCTTCCAGGCGGGGGACCCCTACTACATCAGCAAGCGCAAGCGGGACGAGTGGCTGGCACGCTGGAAAAGGGAGGCTGAGAAGAAAGCCAAGGTCATTGCAGGAATGAATGCTGTGGAAGAAAACCAGGGGCCCGGGGAGTCTCAGAAGGTGGAGGAGGCCAGCCCTCCTGCTGTGCAGCAGCCCACTGACCCCGCATCCCCCACTGTGGCTACCACGCCTGAGCCCGTGGGGTCCGATGCTGGGGACAAGAATGCCACCAAAGCAGGCGATGACGAGCCAGAGTACGAGGACGGCCGGGGCTTTGGCATTGGGGAGCTGGTGTGGGGGAAACTGCGGGGCTTCTCCTGGTGGCCAGGCCGCATTGTGTCTTGGTGGATGACGGGCCGGAGCCGAGCAGCTGAAGGCACCCGCTGGGTCATGTGGTTCGGAGACGGCAAATTCTCAGTGGTGTGTGTTGAGAAGCTGATGCCGCTGAGCTCGTTTTGCAGTGCGTTCCACCAGGCCACGTACAACAAGCAGCCCATGTACCGCAAAGCCATCTACGAGGTCCTGCAGGTGGCCAGCAGCCGCGCGGGGAAGCTGTTCCCGGTGTGCCACGACAGCGATGAGAGTGACACTGCCAAGGCCGTGGAGGTGCAGAACAAGCCCATGATTGAATGGGCCCTGGGGGGCTTCCAGCCTTCTGGCCCTAAGGGCCTGGAGCCACCAGAAGAAGAGAAGAATCCCTACAAAGAAGTGTACACGGACATGTGGGTGGAACCTGAGGCAGCTGCCTACGCACCACCTCCACCAGCCAAAAAGCCCCGGAAGAGCACAGCGGAGAAGCCCAAGGTCAAGGAGATTATTGATGAGCGCACAAGAGAGCGGCTGGTGTACGAGGTGCGGCAGAAGTGCCGGAACATTGAGGACATCTGCATCTCCTGTGGGAGCCTCAATGTTACCCTGGAACACCCCCTCTTCGTTGGAGGAATGTGCCAAAACTGCAAGAACTGCTTTCTGGAGTGTGCGTACCAGTACGACGACGACGGCTACCAGTCCTACTGCACCATCTGCTGTGGGGGCCGTGAGGTGCTCATGTGCGGAAACAACAACTGCTGCAGGTGCTTTTGCGTGGAGTGTGTGGACCTCTTGGTGGGGCCGGGGGCTGCCCAGGCAGCCATTAAGGAAGACCCCTGGAACTGCTACATGTGCGGGCACAAGGGTACCTACGGGCTGCTGCGGCGGCGAGAGGACTGGCCCTCCCGGCTCCAGATGTTCTTCGCTAATAACCACGACCAGGAATTTGACCCTCCAAAGGTTTACCCACCTGTCCCAGCTGAGAAGAGGAAGCCCATCCGGGTGCTGTCTCTCTTTGATGGAATCGCTACAGGGCTCCTGGTGCTGAAGGACTTGGGCATTCAGGTGGACCGCTACATTGCCTCGGAGGTGTGTGAGGACTCCATCACGGTGGGCATGGTGCGGCACCAGGGGAAGATCATGTACGTCGGGGACGTCCGCAGCGTCACACAGAAGCATATCCAGGAGTGGGGCCCATTCGATCTGGTGATTGGGGGCAGTCCCTGCAATGACCTCTCCATCGTCAACCCTGCTCGCAAGGGCCTCTACGAGGGCACTGGCCGGCTCTTCTTTGAGTTCTACCGCCTCCTGCATGATGCGCGGCCCAAGGAGGGAGATGATCGCCCCTTCTTCTGGCTCTTTGAGAATGTGGTGGCCATGGGCGTTAGTGACAAGAGGGACATCTCGCGATTTCTCGAGTCCAACCCTGTGATGATTGATGCCAAAGAAGTGTCAGCTGCACACAGGGCCCGCTACTTCTGGGGTAACCTTCCCGGTATGAACAGGCCGTTGGCATCCACTGTGAATGATAAGCTGGAGCTGCAGGAGTGTCTGGAGCATGGCAGGATAGCCAAGTTCAGCAAAGTGAGGACCATTACTACGAGGTCAAACTCCATAAAGCAGGGCAAAGACCAGCATTTTCCTGTCTTCATGAATGAGAAAGAGGACATCTTATGGTGCACTGAAATGGAAAGGGTATTTGGTTTCCCAGTCCACTATACTGACGTCTCCAACATGAGCCGCTTGGCGAGGCAGAGACTGCTGGGCCGGTCATGGAGCGTGCCAGTCATCCGCCACCTCTTCGCTCCGCTGAAGGAGTATTTTGCGTGTGTGTAAG'\n",
    "\n",
    "    #List of all amino acid positions (across whole coding region) covered (amino acid position minus 1 (as python starts at 0))\n",
    "    DNMT3A_missense_codons_DICT_NP_072046={}\n",
    "    for i in range(lower_range, upper_range):\n",
    "        DNMT3A_missense_codons_DICT_NP_072046[i+1]= (DNMT3A_codons_NP_072046[i])\n",
    "\n",
    "    calls_DNMT3A_missense_NP_072046 = {}\n",
    "    for k,v in DNMT3A_missense_codons_DICT_NP_072046.items(): #iterating over the list of codon numbers (and codon sequences)\n",
    "        if v in Possible_Amino_Acid_Changes.keys(): #assign the list of all the possible amino acid changes to each codon they called\n",
    "            changes = Possible_Amino_Acid_Changes_letters[v]\n",
    "    #         print(Possible_Amino_Acid_Changes.values())\n",
    "            calls_DNMT3A_missense_NP_072046[k]=(v, changes)\n",
    "\n",
    "    #Changing the original amino acid to a Word (e.g. Phe) rather than a codon\n",
    "    calls_DNMT3A_missense_NP_072046_extra = {}\n",
    "\n",
    "    for k, v in calls_DNMT3A_missense_NP_072046.items():\n",
    "        if v[0] in Genetic_Code:\n",
    "            Amino_acid_change = Genetic_Code.get(v[0])\n",
    "            calls_DNMT3A_missense_NP_072046_extra[k] = (Amino_acid_change, v[0], v[1])\n",
    "            #e.g. 12: (Phe, 'K', '*', 'T' etc..)\n",
    "\n",
    "    #Changing the original amino acid to a LETTER (rather than the codon)\n",
    "    calls_DNMT3A_missense_NP_072046_extra2 = {}\n",
    "\n",
    "    for k, v in calls_DNMT3A_missense_NP_072046_extra.items():\n",
    "        if v[0] in Amino_acid_abbreviations:\n",
    "            Amino_acid_abbrev = Amino_acid_abbreviations.get(v[0])\n",
    "            calls_DNMT3A_missense_NP_072046_extra2[k] = (Amino_acid_abbrev, v[1], v[2])\n",
    "            #e.g. 12: ((F), ('K', '*', 'T' etc..))\n",
    "\n",
    "    #Create a dictionary that contains details of the positions of missense variants at that codon\n",
    "    DNMT3A_position_in_list_dict_missense_NP_072046 = {}\n",
    "\n",
    "    for k, v in calls_DNMT3A_missense_NP_072046_extra2.items():\n",
    "        original_AA = v[0]\n",
    "        possible_AA_changes = v[2]\n",
    "        possible_AA_changes_string = ''.join(v[2])\n",
    "        positions_in_list = []\n",
    "        for i in possible_AA_changes:\n",
    "            if i != original_AA: #not synonymous\n",
    "                if i != '*': #not missense\n",
    "                    position_in_list=-1\n",
    "                    missense_change = i\n",
    "                    new_aa_list = []\n",
    "                    while True:\n",
    "                        position_in_list = possible_AA_changes_string.find(missense_change, position_in_list+1)\n",
    "        #               print(position_in_list)\n",
    "                        if position_in_list == -1: break\n",
    "                        positions_in_list.append(position_in_list)\n",
    "                        positions_in_list = set(positions_in_list) #convert into a 'set' to remove duplicate positions\n",
    "                        positions_in_list = list(positions_in_list) #convert back to a list so can be added to dictionary\n",
    "                    for pos in positions_in_list:\n",
    "                        new_aa = possible_AA_changes_string[pos]\n",
    "                        new_aa_list.append(new_aa)\n",
    "                    DNMT3A_position_in_list_dict_missense_NP_072046[k]=(positions_in_list, v, new_aa_list)\n",
    "\n",
    "        aa_positions_variants_called = []\n",
    "        for k, v in DNMT3A_position_in_list_dict_missense_NP_072046.items():\n",
    "            aa_positions_variants_called.append(k)\n",
    "    # print('positions at which missense variants called = ', aa_positions_variants_called)\n",
    "\n",
    "    #Create a dictionary that contains the position in the list of possible AA where the variant is and what it changes to\n",
    "    DNMT3A_bases_missense_NP_072046 = {}\n",
    "    for k, v in DNMT3A_position_in_list_dict_missense_NP_072046.items():\n",
    "        original_a_a = v[1][0]\n",
    "        codon_seq = v[1][1]\n",
    "        base_changes=[]\n",
    "        new_a_a_list = v[2]\n",
    "        if codon_seq in Possible_Base_Changes.keys():\n",
    "            list_of_possible_bases = Possible_Base_Changes.get(codon_seq)\n",
    "        for i in v[0]:\n",
    "            base_change_position = i\n",
    "            base_change = list_of_possible_bases[base_change_position]\n",
    "            base_changes.append(base_change)\n",
    "        DNMT3A_bases_missense_NP_072046[k]=(v, base_changes, original_a_a, new_a_a_list)\n",
    "\n",
    "    #Create dictionary which contains original codon seq and details on which base is changed\n",
    "    DNMT3A_bases_missense2_NP_072046 = {}\n",
    "\n",
    "    for k, v in DNMT3A_bases_missense_NP_072046.items():\n",
    "        base_change_positions = v[0][0]\n",
    "        codon_positions = []\n",
    "        positions_in_gene = []\n",
    "        new_a_a_list = v[3]\n",
    "        for i in base_change_positions:\n",
    "            base_change_position = i\n",
    "            if base_change_position <= 2:\n",
    "                codon_pos = 1\n",
    "            if 2 < base_change_position <= 5:\n",
    "                codon_pos = 2\n",
    "            if 5 < base_change_position <= 8:\n",
    "                codon_pos = 3\n",
    "            codon_positions.append(codon_pos)\n",
    "        DNMT3A_bases_missense2_NP_072046[k] = (v, codon_positions)\n",
    "\n",
    "    #Create dictionary which contains original codon seq, details on which base is changed and it's position in the coding region\n",
    "    DNMT3A_bases_missense3_NP_072046 = {}\n",
    "\n",
    "    for k, v in DNMT3A_bases_missense2_NP_072046.items():\n",
    "        codon_positions = v[1]\n",
    "        base_changes = v[0][1]\n",
    "        original_a_a = v[0][2]\n",
    "        new_aa_list = v[0][0][2]\n",
    "        original_bases = []\n",
    "        positions_in_gene = []\n",
    "        codon_number = k\n",
    "        for i in codon_positions:\n",
    "            codon_pos = i\n",
    "            position_in_gene = ((int(k)-1)*3)+codon_pos #codon number minus 1 (e.g. codon 307 -1 = 306, then times by 3 takes you to beginning of codon 307 (then add the position))  \n",
    "            positions_in_gene.append(position_in_gene)\n",
    "            original_base = (v[0][0][1][1][codon_pos-1])\n",
    "            original_bases.append(original_base)\n",
    "        DNMT3A_bases_missense3_NP_072046[k] = (v[0][0][1], positions_in_gene, original_bases, base_changes, codon_number, original_a_a, new_aa_list)\n",
    "\n",
    "    # Create dictionary which contains original codon seq, details on which base is changed and its neighbouring bases\n",
    "    DNMT3A_bases_missense4_NP_072046 = {}\n",
    "\n",
    "    for k, v in DNMT3A_bases_missense3_NP_072046.items():\n",
    "        position_in_gene = v[1]\n",
    "        sites_trio_missense=[]\n",
    "        new_a_a_list = v[6]\n",
    "\n",
    "        for i in position_in_gene:\n",
    "    #         print(i)\n",
    "            if i in intron_exon_boundary_caution_base_DNMT3A_NP072046:\n",
    "#                 print('positions at intron_exon_boundary =',i)\n",
    "                adapted_trio = intron_exon_check_DNMT3A_NP072046(i)\n",
    "                sites_trio_missense.append([adapted_trio])\n",
    "                DNMT3A_bases_missense4_NP_072046[k] = (v, sites_trio_missense)\n",
    "\n",
    "            else:\n",
    "                for i in position_in_gene:\n",
    "                    gene_pos = i\n",
    "                    site_trio_missense = [DNMT3A_exons_extra_bases_NP_072046[gene_pos:gene_pos+3] for gene_pos in range (gene_pos-1, gene_pos, 3)]\n",
    "                    sites_trio_missense.append(site_trio_missense)\n",
    "                    DNMT3A_bases_missense4_NP_072046[k] = (v, sites_trio_missense)\n",
    "\n",
    "    #Create a dictionary containing pos in coding region, original base, new base, base affected with its neighbours\n",
    "    #e.g. 938, G, A, TGG = base 938 on coding region, original base = G, changes to base A, base with neighbours either side = TGG\n",
    "    DNMT3A_sites_missense5_NP_072046 = {}\n",
    "\n",
    "    a=0\n",
    "    for k, v in DNMT3A_bases_missense4_NP_072046.items():\n",
    "        original_a_a = v[0][0][0]\n",
    "        new_a_a = v[0][6]\n",
    "        codon_n = v[0][1] #base number in codon (out of 3 x 912)\n",
    "        codon_n_912 = v[0][4]\n",
    "        original_base = v[0][2]\n",
    "        new_base = v[0][3]\n",
    "        site_trio = v[1]\n",
    "        info=[]\n",
    "\n",
    "        change = 0\n",
    "        for i in codon_n:\n",
    "            codon_number = i\n",
    "            orig_base = original_base[change]\n",
    "            new_amino_acid = new_a_a[change]\n",
    "            n_base = new_base[change]\n",
    "            site_tri = site_trio[change]\n",
    "            info.append([codon_number, orig_base, n_base, site_tri, codon_n_912, original_a_a, new_amino_acid])\n",
    "            change = change+1\n",
    "        DNMT3A_sites_missense5_NP_072046[a]=info\n",
    "        a=a+1\n",
    "    #     print(DNMT3A_sites_missense5_NP_072046)\n",
    "\n",
    "    # Write in a way that can iterate over mutation rates (i.e. put in to form C[C>A]G etc)   \n",
    "    DNMT3A_sites_missense6_NP_072046={}\n",
    "\n",
    "    separating_positions = {} # separate out all the possible changes to just 1 effective item in the list\n",
    "    a=0\n",
    "    for k, v in DNMT3A_sites_missense5_NP_072046.items():\n",
    "        for i in v: #because some contain more than 1 change from that codon:\n",
    "            separating_positions[a]=i\n",
    "            a=a+1\n",
    "\n",
    "    for k, v in separating_positions.items():\n",
    "        base_pos = v[0]\n",
    "        original_base = v[1]\n",
    "        new_base = v[2]\n",
    "        site_trio = v[3]\n",
    "        codon_n_912 = v[4]\n",
    "        original_a_a = v[5]\n",
    "        new_a_a = v[6]\n",
    "        sequence = (str(original_base),str(new_base))\n",
    "        changes = '>'.join(sequence)\n",
    "        left_base = site_trio[0][0]\n",
    "        right_base = site_trio[0][2]\n",
    "        sequence2 = (changes, right_base)\n",
    "        changes2 = ']'.join(sequence2)\n",
    "        sequence3 = (left_base, changes2)\n",
    "        changes3 = '['.join(sequence3)\n",
    "        DNMT3A_sites_missense6_NP_072046[k] = (v[0],changes3, codon_n_912, original_a_a, new_a_a) #site_trio = the affected base and its neighbours\n",
    "\n",
    "    #Assign the mutation rates to each site...\n",
    "    DNMT3A_mutation_rates_missense_NP_072046 = {}\n",
    "    for k, v in DNMT3A_sites_missense6_NP_072046.items():\n",
    "        base_change_info = v[1]\n",
    "        if base_change_info in Mutation_rates.keys():\n",
    "            mutation_rate = Mutation_rates.get(base_change_info)\n",
    "            original_a_a = v[3]\n",
    "            new_a_a = v[4]\n",
    "            aa_positioning = v[2]\n",
    "            variant_id = str(original_a_a)+str(aa_positioning)+str(new_a_a)\n",
    "            DNMT3A_mutation_rates_missense_NP_072046[k] = (v[0],v[1], mutation_rate, variant_id)\n",
    "\n",
    "    DNMT3A_mutation_rate_missense_NP_072046_sum = []\n",
    "\n",
    "    mutation_rates_list = []\n",
    "    for k, v in DNMT3A_mutation_rates_missense_NP_072046.items():\n",
    "        codon_position_in_gene=(v[0])\n",
    "        variant = v[3]\n",
    "        site = v[1]\n",
    "        mutation_rate_value = v[2] \n",
    "        mutation_rates_list.append((v[0], variant, site, mutation_rate_value))\n",
    "        DNMT3A_mutation_rate_missense_NP_072046_sum.append(mutation_rate_value)\n",
    "\n",
    "    #Get the average of all the mutation rates\n",
    "    DNMT3A_mutation_rate_missense_NP_072046_sum = []\n",
    "    for k, v in DNMT3A_mutation_rates_missense_NP_072046.items():\n",
    "        mut_rate = v[2]\n",
    "        DNMT3A_mutation_rate_missense_NP_072046_sum.append(mut_rate)\n",
    "\n",
    "    DNMT3A_Summed_mutation_rate_missense_NP_072046_added = np.sum(DNMT3A_mutation_rate_missense_NP_072046_sum)\n",
    "\n",
    "    print('Total summed mutation rate across DNMT3A missense_NP_072046 for this study = ', DNMT3A_Summed_mutation_rate_missense_NP_072046_added)\n",
    "\n",
    "    mutation_rates_list_all = mutation_rates_list\n",
    "    return mutation_rates_list_all # e.g. lower= 0, upper = 913 for whole gene\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### TET2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 300,
   "metadata": {
    "code_folding": [
     0
    ]
   },
   "outputs": [],
   "source": [
    "def mutation_rates_nonsense_range_of_variants_TET2(lower_range, upper_range):\n",
    "    \n",
    "    #The mutation rate at a base is calculated based on its trinucleotide context.  To obtain the trinucleotide context, the base either side of the mutated base needs to be ascertained.\n",
    "    #The coding sequence (+ an extra base at each end) of each gene is used when determining the trinucleotide context of the base change.  The problem with using the coding sequence is that\n",
    "    #if a base change falls at the end of an exon, its neighbouring base will actuall be an intronic base, which may be different to the 'neighbouring' base in the coding sequence.\n",
    "    #To account for this, there is an \"intron_exon_check\" function that is called at the intron/exon boundary if the neighbouring base in the intron is different to the neighbouring base in the coding sequence.\n",
    "    \n",
    "    intron_exon_boundary_caution_base_TET2_NP01120680 = [3409, 3804, 4044, 4537]\n",
    "   \n",
    "    def intron_exon_check_TET2_NP01120680(position_in_gene_number):\n",
    "        if position_in_gene_number == 3409:\n",
    "            gene_pos = 3409\n",
    "            first_base = TET2_exons_extra_bases_NP_01120680[gene_pos-1]\n",
    "            middle_base = TET2_exons_extra_bases_NP_01120680[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 3804:\n",
    "            gene_pos = 3804\n",
    "            first_base = 'G'\n",
    "            middle_base = TET2_exons_extra_bases_NP_01120680[gene_pos]\n",
    "            third_base = TET2_exons_extra_bases_NP_01120680[gene_pos+1]\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 4044:\n",
    "            gene_pos = 4044\n",
    "            first_base = TET2_exons_extra_bases_NP_01120680[gene_pos-1]\n",
    "            middle_base = TET2_exons_extra_bases_NP_01120680[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 4537:\n",
    "            gene_pos = 4537\n",
    "            first_base = TET2_exons_extra_bases_NP_01120680[gene_pos-1]\n",
    "            middle_base = TET2_exons_extra_bases_NP_01120680[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        return triplet\n",
    "    \n",
    "    #TET2 transcript (NP_01120680):\n",
    "    #Downloaded from http://grch37.ensembl.org/Homo_sapiens/Transcript/Exons?db=core;g=ENSG00000168769;r=4:106067032-106163928;t=ENST00000380013\n",
    "    TET2_exons_NP_01120680 = 'ATGGAACAGGATAGAACCAACCATGTTGAGGGCAACAGACTAAGTCCATTCCTGATACCATCACCTCCCATTTGCCAGACAGAACCTCTGGCTACAAAGCTCCAGAATGGAAGCCCACTGCCTGAGAGAGCTCATCCAGAAGTAAATGGAGACACCAAGTGGCACTCTTTCAAAAGTTATTATGGAATACCCTGTATGAAGGGAAGCCAGAATAGTCGTGTGAGTCCTGACTTTACACAAGAAAGTAGAGGGTATTCCAAGTGTTTGCAAAATGGAGGAATAAAACGCACAGTTAGTGAACCTTCTCTCTCTGGGCTCCTTCAGATCAAGAAATTGAAACAAGACCAAAAGGCTAATGGAGAAAGACGTAACTTCGGGGTAAGCCAAGAAAGAAATCCAGGTGAAAGCAGTCAACCAAATGTCTCCGATTTGAGTGATAAGAAAGAATCTGTGAGTTCTGTAGCCCAAGAAAATGCAGTTAAAGATTTCACCAGTTTTTCAACACATAACTGCAGTGGGCCTGAAAATCCAGAGCTTCAGATTCTGAATGAGCAGGAGGGGAAAAGTGCTAATTACCATGACAAGAACATTGTATTACTTAAAAACAAGGCAGTGCTAATGCCTAATGGTGCTACAGTTTCTGCCTCTTCCGTGGAACACACACATGGTGAACTCCTGGAAAAAACACTGTCTCAATATTATCCAGATTGTGTTTCCATTGCGGTGCAGAAAACCACATCTCACATAAATGCCATTAACAGTCAGGCTACTAATGAGTTGTCCTGTGAGATCACTCACCCATCGCATACCTCAGGGCAGATCAATTCCGCACAGACCTCTAACTCTGAGCTGCCTCCAAAGCCAGCTGCAGTGGTGAGTGAGGCCTGTGATGCTGATGATGCTGATAATGCCAGTAAACTAGCTGCAATGCTAAATACCTGTTCCTTTCAGAAACCAGAACAACTACAACAACAAAAATCAGTTTTTGAGATATGCCCATCTCCTGCAGAAAATAACATCCAGGGAACCACAAAGCTAGCGTCTGGTGAAGAATTCTGTTCAGGTTCCAGCAGCAATTTGCAAGCTCCTGGTGGCAGCTCTGAACGGTATTTAAAACAAAATGAAATGAATGGTGCTTACTTCAAGCAAAGCTCAGTGTTCACTAAGGATTCCTTTTCTGCCACTACCACACCACCACCACCATCACAATTGCTTCTTTCTCCCCCTCCTCCTCTTCCACAGGTTCCTCAGCTTCCTTCAGAAGGAAAAAGCACTCTGAATGGTGGAGTTTTAGAAGAACACCACCACTACCCCAACCAAAGTAACACAACACTTTTAAGGGAAGTGAAAATAGAGGGTAAACCTGAGGCACCACCTTCCCAGAGTCCTAATCCATCTACACATGTATGCAGCCCTTCTCCGATGCTTTCTGAAAGGCCTCAGAATAATTGTGTGAACAGGAATGACATACAGACTGCAGGGACAATGACTGTTCCATTGTGTTCTGAGAAAACAAGACCAATGTCAGAACACCTCAAGCATAACCCACCAATTTTTGGTAGCAGTGGAGAGCTACAGGACAACTGCCAGCAGTTGATGAGAAACAAAGAGCAAGAGATTCTGAAGGGTCGAGACAAGGAGCAAACACGAGATCTTGTGCCCCCAACACAGCACTATCTGAAACCAGGATGGATTGAATTGAAGGCCCCTCGTTTTCACCAAGCGGAATCCCATCTAAAACGTAATGAGGCATCACTGCCATCAATTCTTCAGTATCAACCCAATCTCTCCAATCAAATGACCTCCAAACAATACACTGGAAATTCCAACATGCCTGGGGGGCTCCCAAGGCAAGCTTACACCCAGAAAACAACACAGCTGGAGCACAAGTCACAAATGTACCAAGTTGAAATGAATCAAGGGCAGTCCCAAGGTACAGTGGACCAACATCTCCAGTTCCAAAAACCCTCACACCAGGTGCACTTCTCCAAAACAGACCATTTACCAAAAGCTCATGTGCAGTCACTGTGTGGCACTAGATTTCATTTTCAACAAAGAGCAGATTCCCAAACTGAAAAACTTATGTCCCCAGTGTTGAAACAGCACTTGAATCAACAGGCTTCAGAGACTGAGCCATTTTCAAACTCACACCTTTTGCAACATAAGCCTCATAAACAGGCAGCACAAACACAACCATCCCAGAGTTCACATCTCCCTCAAAACCAGCAACAGCAGCAAAAATTACAAATAAAGAATAAAGAGGAAATACTCCAGACTTTTCCTCACCCCCAAAGCAACAATGATCAGCAAAGAGAAGGATCATTCTTTGGCCAGACTAAAGTGGAAGAATGTTTTCATGGTGAAAATCAGTATTCAAAATCAAGCGAGTTCGAGACTCATAATGTCCAAATGGGACTGGAGGAAGTACAGAATATAAATCGTAGAAATTCCCCTTATAGTCAGACCATGAAATCAAGTGCATGCAAAATACAGGTTTCTTGTTCAAACAATACACACCTAGTTTCAGAGAATAAAGAACAGACTACACATCCTGAACTTTTTGCAGGAAACAAGACCCAAAACTTGCATCACATGCAATATTTTCCAAATAATGTGATCCCAAAGCAAGATCTTCTTCACAGGTGCTTTCAAGAACAGGAGCAGAAGTCACAACAAGCTTCAGTTCTACAGGGATATAAAAATAGAAACCAAGATATGTCTGGTCAACAAGCTGCGCAACTTGCTCAGCAAAGGTACTTGATACATAACCATGCAAATGTTTTTCCTGTGCCTGACCAGGGAGGAAGTCACACTCAGACCCCTCCCCAGAAGGACACTCAAAAGCATGCTGCTCTAAGGTGGCATCTCTTACAGAAGCAAGAACAGCAGCAAACACAGCAACCCCAAACTGAGTCTTGCCATAGTCAGATGCACAGGCCAATTAAGGTGGAACCTGGATGCAAGCCACATGCCTGTATGCACACAGCACCACCAGAAAACAAAACATGGAAAAAGGTAACTAAGCAAGAGAATCCACCTGCAAGCTGTGATAATGTGCAGCAAAAGAGCATCATTGAGACCATGGAGCAGCATCTGAAGCAGTTTCACGCCAAGTCGTTATTTGACCATAAGGCTCTTACTCTCAAATCACAGAAGCAAGTAAAAGTTGAAATGTCAGGGCCAGTCACAGTTTTGACTAGACAAACCACTGCTGCAGAACTTGATAGCCACACCCCAGCTTTAGAGCAGCAAACAACTTCTTCAGAAAAGACACCAACCAAAAGAACAGCTGCTTCTGTTCTCAATAATTTTATAGAGTCACCTTCCAAATTACTAGATACTCCTATAAAAAATTTATTGGATACACCTGTCAAGACTCAATATGATTTCCCATCTTGCAGATGTGTAGAGCAAATTATTGAAAAAGATGAAGGTCCTTTTTATACCCATCTAGGAGCAGGTCCTAATGTGGCAGCTATTAGAGAAATCATGGAAGAAAGGTTTGGACAGAAGGGTAAAGCTATTAGGATTGAAAGAGTCATCTATACTGGTAAAGAAGGCAAAAGTTCTCAGGGATGTCCTATTGCTAAGTGGGTGGTTCGCAGAAGCAGCAGTGAAGAGAAGCTACTGTGTTTGGTGCGGGAGCGAGCTGGCCACACCTGTGAGGCTGCAGTGATTGTGATTCTCATCCTGGTGTGGGAAGGAATCCCGCTGTCTCTGGCTGACAAACTCTACTCGGAGCTTACCGAGACGCTGAGGAAATACGGCACGCTCACCAATCGCCGGTGTGCCTTGAATGAAGAGAGAACTTGCGCCTGTCAGGGGCTGGATCCAGAAACCTGTGGTGCCTCCTTCTCTTTTGGTTGTTCATGGAGCATGTACTACAATGGATGTAAGTTTGCCAGAAGCAAGATCCCAAGGAAGTTTAAGCTGCTTGGGGATGACCCAAAAGAGGAAGAGAAACTGGAGTCTCATTTGCAAAACCTGTCCACTCTTATGGCACCAACATATAAGAAACTTGCACCTGATGCATATAATAATCAGATTGAATATGAACACAGAGCACCAGAGTGCCGTCTGGGTCTGAAGGAAGGCCGTCCATTCTCAGGGGTCACTGCATGTTTGGACTTCTGTGCTCATGCCCACAGAGACTTGCACAACATGCAGAATGGCAGCACATTGGTATGCACTCTCACTAGAGAAGACAATCGAGAATTTGGAGGAAAACCTGAGGATGAGCAGCTTCACGTTCTGCCTTTATACAAAGTCTCTGACGTGGATGAGTTTGGGAGTGTGGAAGCTCAGGAGGAGAAAAAACGGAGTGGTGCCATTCAGGTACTGAGTTCTTTTCGGCGAAAAGTCAGGATGTTAGCAGAGCCAGTCAAGACTTGCCGACAAAGGAAACTAGAAGCCAAGAAAGCTGCAGCTGAAAAGCTTTCCTCCCTGGAGAACAGCTCAAATAAAAATGAAAAGGAAAAGTCAGCCCCATCACGTACAAAACAAACTGAAAACGCAAGCCAGGCTAAACAGTTGGCAGAACTTTTGCGACTTTCAGGACCAGTCATGCAGCAGTCCCAGCAGCCCCAGCCTCTACAGAAGCAGCCACCACAGCCCCAGCAGCAGCAGAGACCCCAGCAGCAGCAGCCACATCACCCTCAGACAGAGTCTGTCAACTCTTATTCTGCTTCTGGATCCACCAATCCATACATGAGACGGCCCAATCCAGTTAGTCCTTATCCAAACTCTTCACACACTTCAGATATCTATGGAAGCACCAGCCCTATGAACTTCTATTCCACCTCATCTCAAGCTGCAGGTTCATATTTGAATTCTTCTAATCCCATGAACCCTTACCCTGGGCTTTTGAATCAGAATACCCAATATCCATCATATCAATGCAATGGAAACCTATCAGTGGACAACTGCTCCCCATATCTGGGTTCCTATTCTCCCCAGTCTCAGCCGATGGATCTGTATAGGTATCCAAGCCAAGACCCTCTGTCTAAGCTCAGTCTACCACCCATCCATACACTTTACCAGCCAAGGTTTGGAAATAGCCAGAGTTTTACATCTAAATACTTAGGTTATGGAAACCAAAATATGCAGGGAGATGGTTTCAGCAGTTGTACCATTAGACCAAATGTACATCATGTAGGGAAATTGCCTCCTTATCCCACTCATGAGATGGATGGCCACTTCATGGGAGCCACCTCTAGATTACCACCCAATCTGAGCAATCCAAACATGGACTATAAAAATGGTGAACATCATTCACCTTCTCACATAATCCATAACTACAGTGCAGCTCCGGGCATGTTCAACAGCTCTCTTCATGCCCTGCATCTCCAAAACAAGGAGAATGACATGCTTTCCCACACAGCTAATGGGTTATCAAAGATGCTTCCAGCTCTTAACCATGATAGAACTGCTTGTGTCCAAGGAGGCTTACACAAATTAAGTGATGCTAATGGTCAGGAAAAGCAGCCATTGGCACTAGTCCAGGGTGTGGCTTCTGGTGCAGAGGACAACGATGAGGTCTGGTCAGACAGCGAGCAGAGCTTTCTGGATCCTGACATTGGGGGAGTGGCCGTGGCTCCAACTCATGGGTCAATTCTCATTGAGTGTGCAAAGCGTGAGCTGCATGCCACAACCCCTTTAAAGAATCCCAATAGGAATCACCCCACCAGGATCTCCCTCGTCTTTTACCAGCATAAGAGCATGAATGAGCCAAAACATGGCTTGGCTCTTTGGGAAGCCAAAATGGCTGAAAAAGCCCGTGAGAAAGAGGAAGAGTGTGAAAAGTATGGCCCAGACTATGTGCCTCAGAAATCCCATGGCAAAAAAGTGAAACGGGAGCCTGCTGAGCCACATGAAACTTCAGAGCCCACTTACCTGCGTTTCATCAAGTCTCTTGCCGAAAGGACCATGTCCGTGACCACAGACTCCACAGTAACTACATCTCCATATGCCTTCACTCGGGTCACAGGGCCTTACAACAGATATATATGA'\n",
    "\n",
    "    #Split the sequence into codons \n",
    "    TET2_codons_NP_01120680 = [TET2_exons_NP_01120680[i:i+3] for i in range(0, len(TET2_exons_NP_01120680), 3)]\n",
    "\n",
    "    #TET2 transcript (transcript variant 2 (TET2-001), NM_017628, NP_001120680):\n",
    "    #Downloaded from http://grch37.ensembl.org/Homo_sapiens/Transcript/Exons?db=core;g=ENSG00000168769;r=4:106067032-106163928;t=ENST00000380013\n",
    "    TET2_exons_extra_bases_NP_01120680 = 'GATGGAACAGGATAGAACCAACCATGTTGAGGGCAACAGACTAAGTCCATTCCTGATACCATCACCTCCCATTTGCCAGACAGAACCTCTGGCTACAAAGCTCCAGAATGGAAGCCCACTGCCTGAGAGAGCTCATCCAGAAGTAAATGGAGACACCAAGTGGCACTCTTTCAAAAGTTATTATGGAATACCCTGTATGAAGGGAAGCCAGAATAGTCGTGTGAGTCCTGACTTTACACAAGAAAGTAGAGGGTATTCCAAGTGTTTGCAAAATGGAGGAATAAAACGCACAGTTAGTGAACCTTCTCTCTCTGGGCTCCTTCAGATCAAGAAATTGAAACAAGACCAAAAGGCTAATGGAGAAAGACGTAACTTCGGGGTAAGCCAAGAAAGAAATCCAGGTGAAAGCAGTCAACCAAATGTCTCCGATTTGAGTGATAAGAAAGAATCTGTGAGTTCTGTAGCCCAAGAAAATGCAGTTAAAGATTTCACCAGTTTTTCAACACATAACTGCAGTGGGCCTGAAAATCCAGAGCTTCAGATTCTGAATGAGCAGGAGGGGAAAAGTGCTAATTACCATGACAAGAACATTGTATTACTTAAAAACAAGGCAGTGCTAATGCCTAATGGTGCTACAGTTTCTGCCTCTTCCGTGGAACACACACATGGTGAACTCCTGGAAAAAACACTGTCTCAATATTATCCAGATTGTGTTTCCATTGCGGTGCAGAAAACCACATCTCACATAAATGCCATTAACAGTCAGGCTACTAATGAGTTGTCCTGTGAGATCACTCACCCATCGCATACCTCAGGGCAGATCAATTCCGCACAGACCTCTAACTCTGAGCTGCCTCCAAAGCCAGCTGCAGTGGTGAGTGAGGCCTGTGATGCTGATGATGCTGATAATGCCAGTAAACTAGCTGCAATGCTAAATACCTGTTCCTTTCAGAAACCAGAACAACTACAACAACAAAAATCAGTTTTTGAGATATGCCCATCTCCTGCAGAAAATAACATCCAGGGAACCACAAAGCTAGCGTCTGGTGAAGAATTCTGTTCAGGTTCCAGCAGCAATTTGCAAGCTCCTGGTGGCAGCTCTGAACGGTATTTAAAACAAAATGAAATGAATGGTGCTTACTTCAAGCAAAGCTCAGTGTTCACTAAGGATTCCTTTTCTGCCACTACCACACCACCACCACCATCACAATTGCTTCTTTCTCCCCCTCCTCCTCTTCCACAGGTTCCTCAGCTTCCTTCAGAAGGAAAAAGCACTCTGAATGGTGGAGTTTTAGAAGAACACCACCACTACCCCAACCAAAGTAACACAACACTTTTAAGGGAAGTGAAAATAGAGGGTAAACCTGAGGCACCACCTTCCCAGAGTCCTAATCCATCTACACATGTATGCAGCCCTTCTCCGATGCTTTCTGAAAGGCCTCAGAATAATTGTGTGAACAGGAATGACATACAGACTGCAGGGACAATGACTGTTCCATTGTGTTCTGAGAAAACAAGACCAATGTCAGAACACCTCAAGCATAACCCACCAATTTTTGGTAGCAGTGGAGAGCTACAGGACAACTGCCAGCAGTTGATGAGAAACAAAGAGCAAGAGATTCTGAAGGGTCGAGACAAGGAGCAAACACGAGATCTTGTGCCCCCAACACAGCACTATCTGAAACCAGGATGGATTGAATTGAAGGCCCCTCGTTTTCACCAAGCGGAATCCCATCTAAAACGTAATGAGGCATCACTGCCATCAATTCTTCAGTATCAACCCAATCTCTCCAATCAAATGACCTCCAAACAATACACTGGAAATTCCAACATGCCTGGGGGGCTCCCAAGGCAAGCTTACACCCAGAAAACAACACAGCTGGAGCACAAGTCACAAATGTACCAAGTTGAAATGAATCAAGGGCAGTCCCAAGGTACAGTGGACCAACATCTCCAGTTCCAAAAACCCTCACACCAGGTGCACTTCTCCAAAACAGACCATTTACCAAAAGCTCATGTGCAGTCACTGTGTGGCACTAGATTTCATTTTCAACAAAGAGCAGATTCCCAAACTGAAAAACTTATGTCCCCAGTGTTGAAACAGCACTTGAATCAACAGGCTTCAGAGACTGAGCCATTTTCAAACTCACACCTTTTGCAACATAAGCCTCATAAACAGGCAGCACAAACACAACCATCCCAGAGTTCACATCTCCCTCAAAACCAGCAACAGCAGCAAAAATTACAAATAAAGAATAAAGAGGAAATACTCCAGACTTTTCCTCACCCCCAAAGCAACAATGATCAGCAAAGAGAAGGATCATTCTTTGGCCAGACTAAAGTGGAAGAATGTTTTCATGGTGAAAATCAGTATTCAAAATCAAGCGAGTTCGAGACTCATAATGTCCAAATGGGACTGGAGGAAGTACAGAATATAAATCGTAGAAATTCCCCTTATAGTCAGACCATGAAATCAAGTGCATGCAAAATACAGGTTTCTTGTTCAAACAATACACACCTAGTTTCAGAGAATAAAGAACAGACTACACATCCTGAACTTTTTGCAGGAAACAAGACCCAAAACTTGCATCACATGCAATATTTTCCAAATAATGTGATCCCAAAGCAAGATCTTCTTCACAGGTGCTTTCAAGAACAGGAGCAGAAGTCACAACAAGCTTCAGTTCTACAGGGATATAAAAATAGAAACCAAGATATGTCTGGTCAACAAGCTGCGCAACTTGCTCAGCAAAGGTACTTGATACATAACCATGCAAATGTTTTTCCTGTGCCTGACCAGGGAGGAAGTCACACTCAGACCCCTCCCCAGAAGGACACTCAAAAGCATGCTGCTCTAAGGTGGCATCTCTTACAGAAGCAAGAACAGCAGCAAACACAGCAACCCCAAACTGAGTCTTGCCATAGTCAGATGCACAGGCCAATTAAGGTGGAACCTGGATGCAAGCCACATGCCTGTATGCACACAGCACCACCAGAAAACAAAACATGGAAAAAGGTAACTAAGCAAGAGAATCCACCTGCAAGCTGTGATAATGTGCAGCAAAAGAGCATCATTGAGACCATGGAGCAGCATCTGAAGCAGTTTCACGCCAAGTCGTTATTTGACCATAAGGCTCTTACTCTCAAATCACAGAAGCAAGTAAAAGTTGAAATGTCAGGGCCAGTCACAGTTTTGACTAGACAAACCACTGCTGCAGAACTTGATAGCCACACCCCAGCTTTAGAGCAGCAAACAACTTCTTCAGAAAAGACACCAACCAAAAGAACAGCTGCTTCTGTTCTCAATAATTTTATAGAGTCACCTTCCAAATTACTAGATACTCCTATAAAAAATTTATTGGATACACCTGTCAAGACTCAATATGATTTCCCATCTTGCAGATGTGTAGAGCAAATTATTGAAAAAGATGAAGGTCCTTTTTATACCCATCTAGGAGCAGGTCCTAATGTGGCAGCTATTAGAGAAATCATGGAAGAAAGGTTTGGACAGAAGGGTAAAGCTATTAGGATTGAAAGAGTCATCTATACTGGTAAAGAAGGCAAAAGTTCTCAGGGATGTCCTATTGCTAAGTGGGTGGTTCGCAGAAGCAGCAGTGAAGAGAAGCTACTGTGTTTGGTGCGGGAGCGAGCTGGCCACACCTGTGAGGCTGCAGTGATTGTGATTCTCATCCTGGTGTGGGAAGGAATCCCGCTGTCTCTGGCTGACAAACTCTACTCGGAGCTTACCGAGACGCTGAGGAAATACGGCACGCTCACCAATCGCCGGTGTGCCTTGAATGAAGAGAGAACTTGCGCCTGTCAGGGGCTGGATCCAGAAACCTGTGGTGCCTCCTTCTCTTTTGGTTGTTCATGGAGCATGTACTACAATGGATGTAAGTTTGCCAGAAGCAAGATCCCAAGGAAGTTTAAGCTGCTTGGGGATGACCCAAAAGAGGAAGAGAAACTGGAGTCTCATTTGCAAAACCTGTCCACTCTTATGGCACCAACATATAAGAAACTTGCACCTGATGCATATAATAATCAGATTGAATATGAACACAGAGCACCAGAGTGCCGTCTGGGTCTGAAGGAAGGCCGTCCATTCTCAGGGGTCACTGCATGTTTGGACTTCTGTGCTCATGCCCACAGAGACTTGCACAACATGCAGAATGGCAGCACATTGGTATGCACTCTCACTAGAGAAGACAATCGAGAATTTGGAGGAAAACCTGAGGATGAGCAGCTTCACGTTCTGCCTTTATACAAAGTCTCTGACGTGGATGAGTTTGGGAGTGTGGAAGCTCAGGAGGAGAAAAAACGGAGTGGTGCCATTCAGGTACTGAGTTCTTTTCGGCGAAAAGTCAGGATGTTAGCAGAGCCAGTCAAGACTTGCCGACAAAGGAAACTAGAAGCCAAGAAAGCTGCAGCTGAAAAGCTTTCCTCCCTGGAGAACAGCTCAAATAAAAATGAAAAGGAAAAGTCAGCCCCATCACGTACAAAACAAACTGAAAACGCAAGCCAGGCTAAACAGTTGGCAGAACTTTTGCGACTTTCAGGACCAGTCATGCAGCAGTCCCAGCAGCCCCAGCCTCTACAGAAGCAGCCACCACAGCCCCAGCAGCAGCAGAGACCCCAGCAGCAGCAGCCACATCACCCTCAGACAGAGTCTGTCAACTCTTATTCTGCTTCTGGATCCACCAATCCATACATGAGACGGCCCAATCCAGTTAGTCCTTATCCAAACTCTTCACACACTTCAGATATCTATGGAAGCACCAGCCCTATGAACTTCTATTCCACCTCATCTCAAGCTGCAGGTTCATATTTGAATTCTTCTAATCCCATGAACCCTTACCCTGGGCTTTTGAATCAGAATACCCAATATCCATCATATCAATGCAATGGAAACCTATCAGTGGACAACTGCTCCCCATATCTGGGTTCCTATTCTCCCCAGTCTCAGCCGATGGATCTGTATAGGTATCCAAGCCAAGACCCTCTGTCTAAGCTCAGTCTACCACCCATCCATACACTTTACCAGCCAAGGTTTGGAAATAGCCAGAGTTTTACATCTAAATACTTAGGTTATGGAAACCAAAATATGCAGGGAGATGGTTTCAGCAGTTGTACCATTAGACCAAATGTACATCATGTAGGGAAATTGCCTCCTTATCCCACTCATGAGATGGATGGCCACTTCATGGGAGCCACCTCTAGATTACCACCCAATCTGAGCAATCCAAACATGGACTATAAAAATGGTGAACATCATTCACCTTCTCACATAATCCATAACTACAGTGCAGCTCCGGGCATGTTCAACAGCTCTCTTCATGCCCTGCATCTCCAAAACAAGGAGAATGACATGCTTTCCCACACAGCTAATGGGTTATCAAAGATGCTTCCAGCTCTTAACCATGATAGAACTGCTTGTGTCCAAGGAGGCTTACACAAATTAAGTGATGCTAATGGTCAGGAAAAGCAGCCATTGGCACTAGTCCAGGGTGTGGCTTCTGGTGCAGAGGACAACGATGAGGTCTGGTCAGACAGCGAGCAGAGCTTTCTGGATCCTGACATTGGGGGAGTGGCCGTGGCTCCAACTCATGGGTCAATTCTCATTGAGTGTGCAAAGCGTGAGCTGCATGCCACAACCCCTTTAAAGAATCCCAATAGGAATCACCCCACCAGGATCTCCCTCGTCTTTTACCAGCATAAGAGCATGAATGAGCCAAAACATGGCTTGGCTCTTTGGGAAGCCAAAATGGCTGAAAAAGCCCGTGAGAAAGAGGAAGAGTGTGAAAAGTATGGCCCAGACTATGTGCCTCAGAAATCCCATGGCAAAAAAGTGAAACGGGAGCCTGCTGAGCCACATGAAACTTCAGAGCCCACTTACCTGCGTTTCATCAAGTCTCTTGCCGAAAGGACCATGTCCGTGACCACAGACTCCACAGTAACTACATCTCCATATGCCTTCACTCGGGTCACAGGGCCTTACAACAGATATATATGAT'\n",
    "\n",
    "    #List of all amino acid positions (across whole coding region) covered by Coombs (amino acid position minus 1 (as python starts at 0))\n",
    "    TET2_nonsense_codons_DICT_NP_01120680={}\n",
    "    for i in range(lower_range, upper_range):\n",
    "        TET2_nonsense_codons_DICT_NP_01120680[i+1]= (TET2_codons_NP_01120680[i])\n",
    "\n",
    "    calls_TET2_nonsense_NP_01120680 = {}\n",
    "    for k,v in TET2_nonsense_codons_DICT_NP_01120680.items(): #iterating over the list of codon numbers (and codon sequences)\n",
    "        if v in Possible_Amino_Acid_Changes.keys(): #assign the list of all the possible amino acid changes to each codon they called\n",
    "            changes = Possible_Amino_Acid_Changes_letters[v]\n",
    "    #         print(Possible_Amino_Acid_Changes.values())\n",
    "            calls_TET2_nonsense_NP_01120680[k]=(v, changes)\n",
    "\n",
    "    #Changing the original amino acid to a Word (e.g. Phe) rather than a codon\n",
    "    calls_TET2_nonsense_NP_01120680_extra = {}\n",
    "\n",
    "    for k, v in calls_TET2_nonsense_NP_01120680.items():\n",
    "        if v[0] in Genetic_Code:\n",
    "            Amino_acid_change = Genetic_Code.get(v[0])\n",
    "            calls_TET2_nonsense_NP_01120680_extra[k] = (Amino_acid_change, v[0], v[1])\n",
    "            #e.g. 12: (Phe, 'K', '*', 'T' etc..)\n",
    "\n",
    "    #Changing the original amino acid to a LETTER (rather than the codon)\n",
    "    calls_TET2_nonsense_NP_01120680_extra2 = {}\n",
    "\n",
    "    for k, v in calls_TET2_nonsense_NP_01120680_extra.items():\n",
    "        if v[0] in Amino_acid_abbreviations:\n",
    "            Amino_acid_abbrev = Amino_acid_abbreviations.get(v[0])\n",
    "            calls_TET2_nonsense_NP_01120680_extra2[k] = (Amino_acid_abbrev, v[1], v[2])\n",
    "            #e.g. 12: ((F), ('K', '*', 'T' etc..))\n",
    "\n",
    "    #Create a dictionary that contains details of the positions of nonsense variants at that codon\n",
    "    TET2_position_in_list_dict_nonsense_NP_01120680 = {}\n",
    "\n",
    "    for k, v in calls_TET2_nonsense_NP_01120680_extra2.items():\n",
    "        original_AA = v[0]\n",
    "        possible_AA_changes = v[2]\n",
    "        possible_AA_changes_string = ''.join(v[2])\n",
    "        positions_in_list = []\n",
    "        for i in possible_AA_changes:\n",
    "            if i == '*': #nonsense\n",
    "                position_in_list=-1\n",
    "                nonsense_change = i    \n",
    "                while True:\n",
    "                    position_in_list = possible_AA_changes_string.find(nonsense_change, position_in_list+1)\n",
    "    #               print(position_in_list)\n",
    "                    if position_in_list == -1: break\n",
    "                    positions_in_list.append(position_in_list)\n",
    "                    positions_in_list = set(positions_in_list) #convert into a 'set' to remove duplicate positions\n",
    "                    positions_in_list = list(positions_in_list) #convert back to a list so can be added to dictionary\n",
    "                TET2_position_in_list_dict_nonsense_NP_01120680[k]=(positions_in_list, v)\n",
    "\n",
    "        aa_positions_variants_called = []\n",
    "        for k, v in TET2_position_in_list_dict_nonsense_NP_01120680.items():\n",
    "            aa_positions_variants_called.append(k)\n",
    "    # print('positions at which nonsense variants called = ', aa_positions_variants_called)\n",
    "\n",
    "    #Create a dictionary that contains the position in the list of possible AA where the variant is and what it changes to\n",
    "    TET2_bases_nonsense_NP_01120680 = {}\n",
    "    for k, v in TET2_position_in_list_dict_nonsense_NP_01120680.items():\n",
    "        original_a_a = v[1][0]\n",
    "        codon_seq = v[1][1]\n",
    "        base_changes=[]\n",
    "        if codon_seq in Possible_Base_Changes.keys():\n",
    "            list_of_possible_bases = Possible_Base_Changes.get(codon_seq)\n",
    "        for i in v[0]:\n",
    "            base_change_position = i\n",
    "            base_change = list_of_possible_bases[base_change_position]\n",
    "            base_changes.append(base_change)\n",
    "        TET2_bases_nonsense_NP_01120680[k]=(v, base_changes, original_a_a)\n",
    "\n",
    "    #Create dictionary which contains original codon seq and details on which base is changed\n",
    "    TET2_bases_nonsense2_NP_01120680 = {}\n",
    "\n",
    "    for k, v in TET2_bases_nonsense_NP_01120680.items():\n",
    "        base_change_positions = v[0][0]\n",
    "        codon_positions = []\n",
    "        positions_in_gene = []\n",
    "        for i in base_change_positions:\n",
    "            base_change_position = i\n",
    "            if base_change_position <= 2:\n",
    "                codon_pos = 1\n",
    "            if 2 < base_change_position <= 5:\n",
    "                codon_pos = 2\n",
    "            if 5 < base_change_position <= 8:\n",
    "                codon_pos = 3\n",
    "            codon_positions.append(codon_pos)\n",
    "        TET2_bases_nonsense2_NP_01120680[k] = (v, codon_positions)\n",
    "\n",
    "    #Create dictionary which contains original codon seq, details on which base is changed and it's position in the coding region\n",
    "    TET2_bases_nonsense3_NP_01120680 = {}\n",
    "\n",
    "    for k, v in TET2_bases_nonsense2_NP_01120680.items():\n",
    "        codon_positions = v[1]\n",
    "        base_changes = v[0][1]\n",
    "        original_a_a = v[0][2]\n",
    "        original_bases = []\n",
    "        positions_in_gene = []\n",
    "        codon_number = k\n",
    "        for i in codon_positions:\n",
    "            codon_pos = i\n",
    "            position_in_gene = ((int(k)-1)*3)+codon_pos #codon number minus 1 (e.g. codon 307 -1 = 306, then times by 3 takes you to beginning of codon 307 (then add the position))  \n",
    "            positions_in_gene.append(position_in_gene)\n",
    "            original_base = (v[0][0][1][1][codon_pos-1])\n",
    "            original_bases.append(original_base)\n",
    "        TET2_bases_nonsense3_NP_01120680[k] = (v[0][0][1], positions_in_gene, original_bases, base_changes, codon_number, original_a_a)\n",
    "\n",
    "    #Create dictionary which contains original codon seq, details on which base is changed and its neighbouring bases\n",
    "    TET2_bases_nonsense4_NP_01120680 = {}\n",
    "\n",
    "    for k, v in TET2_bases_nonsense3_NP_01120680.items():\n",
    "        position_in_gene = v[1]\n",
    "        sites_trio_nonsense=[]\n",
    "\n",
    "        for i in position_in_gene:\n",
    "    #         print(i)\n",
    "            if i in intron_exon_boundary_caution_base_TET2_NP01120680:\n",
    "#                 print('positions at intron_exon_boundary =',i)\n",
    "                adapted_trio = intron_exon_check_TET2_NP01120680(i)\n",
    "                sites_trio_nonsense.append([adapted_trio])\n",
    "                TET2_bases_nonsense4_NP_01120680[k] = (v, sites_trio_nonsense)\n",
    "\n",
    "            else:\n",
    "                for i in position_in_gene:\n",
    "                    gene_pos = i\n",
    "                    site_trio_nonsense = [TET2_exons_extra_bases_NP_01120680[gene_pos:gene_pos+3] for gene_pos in range (gene_pos-1, gene_pos, 3)]\n",
    "                    sites_trio_nonsense.append(site_trio_nonsense)\n",
    "                    TET2_bases_nonsense4_NP_01120680[k] = (v, sites_trio_nonsense)\n",
    "\n",
    "    #Create a dictionary containing pos in coding region, original base, new base, base affected with its neighbours\n",
    "    #e.g. 938, G, A, TGG = base 938 on coding region, original base = G, changes to base A, base with neighbours either side = TGG\n",
    "    TET2_sites_nonsense5_NP_01120680 = {}\n",
    "\n",
    "    a=0\n",
    "    for k, v in TET2_bases_nonsense4_NP_01120680.items():\n",
    "        original_a_a = v[0][0][0]\n",
    "        codon_n = v[0][1] #base number in codon (out of 3 x 912)\n",
    "        codon_n_912 = v[0][4]\n",
    "        original_base = v[0][2]\n",
    "        new_base = v[0][3]\n",
    "        site_trio = v[1]\n",
    "        info=[]\n",
    "\n",
    "        change = 0\n",
    "        for i in codon_n:\n",
    "            codon_number = i\n",
    "            orig_base = original_base[change]\n",
    "            n_base = new_base[change]\n",
    "            site_tri = site_trio[change]\n",
    "            info.append([codon_number, orig_base, n_base, site_tri, codon_n_912, original_a_a])\n",
    "            change = change+1\n",
    "        TET2_sites_nonsense5_NP_01120680[a]=info\n",
    "        a=a+1\n",
    "    #     print(TET2_sites_nonsense5_NP_01120680)\n",
    "\n",
    "    # Write in a way that can iterate over mutation rates (i.e. put in to form C[C>A]G etc)   \n",
    "    TET2_sites_nonsense6_NP_01120680={}\n",
    "\n",
    "    separating_positions = {} # separate out all the possible changes to just 1 effective item in the list\n",
    "    a=0\n",
    "    for k, v in TET2_sites_nonsense5_NP_01120680.items():\n",
    "        for i in v: #because some contain more than 1 change from that codon:\n",
    "            separating_positions[a]=i\n",
    "            a=a+1\n",
    "\n",
    "    for k, v in separating_positions.items():\n",
    "        base_pos = v[0]\n",
    "        original_base = v[1]\n",
    "        new_base = v[2]\n",
    "        site_trio = v[3]\n",
    "        codon_n_912 = v[4]\n",
    "        original_a_a = v[5]\n",
    "        sequence = (str(original_base),str(new_base))\n",
    "        changes = '>'.join(sequence)\n",
    "        left_base = site_trio[0][0]\n",
    "        right_base = site_trio[0][2]\n",
    "        sequence2 = (changes, right_base)\n",
    "        changes2 = ']'.join(sequence2)\n",
    "        sequence3 = (left_base, changes2)\n",
    "        changes3 = '['.join(sequence3)\n",
    "        TET2_sites_nonsense6_NP_01120680[k] = (v[0],changes3, codon_n_912, original_a_a) #site_trio = the affected base and its neighbours\n",
    "\n",
    "    #Assign the mutation rates to each site...\n",
    "    TET2_mutation_rates_nonsense_NP_01120680 = {}\n",
    "    for k, v in TET2_sites_nonsense6_NP_01120680.items():\n",
    "        base_change_info = v[1]\n",
    "        if base_change_info in Mutation_rates.keys():\n",
    "            mutation_rate = Mutation_rates.get(base_change_info)\n",
    "            original_a_a = v[3]\n",
    "            aa_positioning = v[2]\n",
    "            variant_type = '*'\n",
    "            variant_id = str(original_a_a)+str(aa_positioning)+variant_type\n",
    "            TET2_mutation_rates_nonsense_NP_01120680[k] = (v[0],v[1], mutation_rate, variant_id)\n",
    "\n",
    "    TET2_mutation_rate_nonsense_NP_01120680_sum = []\n",
    "\n",
    "    mutation_rates_list = []\n",
    "    for k, v in TET2_mutation_rates_nonsense_NP_01120680.items():\n",
    "        codon_position_in_gene=(v[0])\n",
    "        variant = v[3]\n",
    "        site = v[1]\n",
    "        mutation_rate_value = v[2] \n",
    "        mutation_rates_list.append((v[0], variant, site, mutation_rate_value))\n",
    "        TET2_mutation_rate_nonsense_NP_01120680_sum.append(mutation_rate_value)\n",
    "\n",
    "    #Get the average of all the mutation rates\n",
    "    TET2_mutation_rate_nonsense_NP_01120680_sum = []\n",
    "    for k, v in TET2_mutation_rates_nonsense_NP_01120680.items():\n",
    "        mut_rate = v[2]\n",
    "        TET2_mutation_rate_nonsense_NP_01120680_sum.append(mut_rate)\n",
    "\n",
    "    TET2_Summed_mutation_rate_nonsense_NP_01120680_added = np.sum(TET2_mutation_rate_nonsense_NP_01120680_sum)\n",
    "\n",
    "    print('Total summed mutation rate across TET2 nonsense_NP_01120680 for this study = ', TET2_Summed_mutation_rate_nonsense_NP_01120680_added)\n",
    "\n",
    "    mutation_rates_list_all = mutation_rates_list\n",
    "    return mutation_rates_list_all # e.g. lower= 0, upper = 913 for whole gene\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 301,
   "metadata": {
    "code_folding": [
     0
    ]
   },
   "outputs": [],
   "source": [
    "def mutation_rates_missense_range_of_variants_TET2(lower_range, upper_range):\n",
    "\n",
    "    #The mutation rate at a base is calculated based on its trinucleotide context.  To obtain the trinucleotide context, the base either side of the mutated base needs to be ascertained.\n",
    "    #The coding sequence (+ an extra base at each end) of each gene is used when determining the trinucleotide context of the base change.  The problem with using the coding sequence is that\n",
    "    #if a base change falls at the end of an exon, its neighbouring base will actuall be an intronic base, which may be different to the 'neighbouring' base in the coding sequence.\n",
    "    #To account for this, there is an \"intron_exon_check\" function that is called at the intron/exon boundary if the neighbouring base in the intron is different to the neighbouring base in the coding sequence.\n",
    "    \n",
    "    intron_exon_boundary_caution_base_TET2_NP01120680 = [3409, 3804, 4044, 4537]\n",
    "   \n",
    "    def intron_exon_check_TET2_NP01120680(position_in_gene_number):\n",
    "        if position_in_gene_number == 3409:\n",
    "            gene_pos = 3409\n",
    "            first_base = TET2_exons_extra_bases_NP_01120680[gene_pos-1]\n",
    "            middle_base = TET2_exons_extra_bases_NP_01120680[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 3804:\n",
    "            gene_pos = 3804\n",
    "            first_base = 'G'\n",
    "            middle_base = TET2_exons_extra_bases_NP_01120680[gene_pos]\n",
    "            third_base = TET2_exons_extra_bases_NP_01120680[gene_pos+1]\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 4044:\n",
    "            gene_pos = 4044\n",
    "            first_base = TET2_exons_extra_bases_NP_01120680[gene_pos-1]\n",
    "            middle_base = TET2_exons_extra_bases_NP_01120680[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 4537:\n",
    "            gene_pos = 4537\n",
    "            first_base = TET2_exons_extra_bases_NP_01120680[gene_pos-1]\n",
    "            middle_base = TET2_exons_extra_bases_NP_01120680[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        return triplet\n",
    "    \n",
    "    #TET2 transcript (NP_01120680):\n",
    "    #Downloaded from http://grch37.ensembl.org/Homo_sapiens/Transcript/Exons?db=core;g=ENSG00000168769;r=4:106067032-106163928;t=ENST00000380013\n",
    "    TET2_exons_NP_01120680 = 'ATGGAACAGGATAGAACCAACCATGTTGAGGGCAACAGACTAAGTCCATTCCTGATACCATCACCTCCCATTTGCCAGACAGAACCTCTGGCTACAAAGCTCCAGAATGGAAGCCCACTGCCTGAGAGAGCTCATCCAGAAGTAAATGGAGACACCAAGTGGCACTCTTTCAAAAGTTATTATGGAATACCCTGTATGAAGGGAAGCCAGAATAGTCGTGTGAGTCCTGACTTTACACAAGAAAGTAGAGGGTATTCCAAGTGTTTGCAAAATGGAGGAATAAAACGCACAGTTAGTGAACCTTCTCTCTCTGGGCTCCTTCAGATCAAGAAATTGAAACAAGACCAAAAGGCTAATGGAGAAAGACGTAACTTCGGGGTAAGCCAAGAAAGAAATCCAGGTGAAAGCAGTCAACCAAATGTCTCCGATTTGAGTGATAAGAAAGAATCTGTGAGTTCTGTAGCCCAAGAAAATGCAGTTAAAGATTTCACCAGTTTTTCAACACATAACTGCAGTGGGCCTGAAAATCCAGAGCTTCAGATTCTGAATGAGCAGGAGGGGAAAAGTGCTAATTACCATGACAAGAACATTGTATTACTTAAAAACAAGGCAGTGCTAATGCCTAATGGTGCTACAGTTTCTGCCTCTTCCGTGGAACACACACATGGTGAACTCCTGGAAAAAACACTGTCTCAATATTATCCAGATTGTGTTTCCATTGCGGTGCAGAAAACCACATCTCACATAAATGCCATTAACAGTCAGGCTACTAATGAGTTGTCCTGTGAGATCACTCACCCATCGCATACCTCAGGGCAGATCAATTCCGCACAGACCTCTAACTCTGAGCTGCCTCCAAAGCCAGCTGCAGTGGTGAGTGAGGCCTGTGATGCTGATGATGCTGATAATGCCAGTAAACTAGCTGCAATGCTAAATACCTGTTCCTTTCAGAAACCAGAACAACTACAACAACAAAAATCAGTTTTTGAGATATGCCCATCTCCTGCAGAAAATAACATCCAGGGAACCACAAAGCTAGCGTCTGGTGAAGAATTCTGTTCAGGTTCCAGCAGCAATTTGCAAGCTCCTGGTGGCAGCTCTGAACGGTATTTAAAACAAAATGAAATGAATGGTGCTTACTTCAAGCAAAGCTCAGTGTTCACTAAGGATTCCTTTTCTGCCACTACCACACCACCACCACCATCACAATTGCTTCTTTCTCCCCCTCCTCCTCTTCCACAGGTTCCTCAGCTTCCTTCAGAAGGAAAAAGCACTCTGAATGGTGGAGTTTTAGAAGAACACCACCACTACCCCAACCAAAGTAACACAACACTTTTAAGGGAAGTGAAAATAGAGGGTAAACCTGAGGCACCACCTTCCCAGAGTCCTAATCCATCTACACATGTATGCAGCCCTTCTCCGATGCTTTCTGAAAGGCCTCAGAATAATTGTGTGAACAGGAATGACATACAGACTGCAGGGACAATGACTGTTCCATTGTGTTCTGAGAAAACAAGACCAATGTCAGAACACCTCAAGCATAACCCACCAATTTTTGGTAGCAGTGGAGAGCTACAGGACAACTGCCAGCAGTTGATGAGAAACAAAGAGCAAGAGATTCTGAAGGGTCGAGACAAGGAGCAAACACGAGATCTTGTGCCCCCAACACAGCACTATCTGAAACCAGGATGGATTGAATTGAAGGCCCCTCGTTTTCACCAAGCGGAATCCCATCTAAAACGTAATGAGGCATCACTGCCATCAATTCTTCAGTATCAACCCAATCTCTCCAATCAAATGACCTCCAAACAATACACTGGAAATTCCAACATGCCTGGGGGGCTCCCAAGGCAAGCTTACACCCAGAAAACAACACAGCTGGAGCACAAGTCACAAATGTACCAAGTTGAAATGAATCAAGGGCAGTCCCAAGGTACAGTGGACCAACATCTCCAGTTCCAAAAACCCTCACACCAGGTGCACTTCTCCAAAACAGACCATTTACCAAAAGCTCATGTGCAGTCACTGTGTGGCACTAGATTTCATTTTCAACAAAGAGCAGATTCCCAAACTGAAAAACTTATGTCCCCAGTGTTGAAACAGCACTTGAATCAACAGGCTTCAGAGACTGAGCCATTTTCAAACTCACACCTTTTGCAACATAAGCCTCATAAACAGGCAGCACAAACACAACCATCCCAGAGTTCACATCTCCCTCAAAACCAGCAACAGCAGCAAAAATTACAAATAAAGAATAAAGAGGAAATACTCCAGACTTTTCCTCACCCCCAAAGCAACAATGATCAGCAAAGAGAAGGATCATTCTTTGGCCAGACTAAAGTGGAAGAATGTTTTCATGGTGAAAATCAGTATTCAAAATCAAGCGAGTTCGAGACTCATAATGTCCAAATGGGACTGGAGGAAGTACAGAATATAAATCGTAGAAATTCCCCTTATAGTCAGACCATGAAATCAAGTGCATGCAAAATACAGGTTTCTTGTTCAAACAATACACACCTAGTTTCAGAGAATAAAGAACAGACTACACATCCTGAACTTTTTGCAGGAAACAAGACCCAAAACTTGCATCACATGCAATATTTTCCAAATAATGTGATCCCAAAGCAAGATCTTCTTCACAGGTGCTTTCAAGAACAGGAGCAGAAGTCACAACAAGCTTCAGTTCTACAGGGATATAAAAATAGAAACCAAGATATGTCTGGTCAACAAGCTGCGCAACTTGCTCAGCAAAGGTACTTGATACATAACCATGCAAATGTTTTTCCTGTGCCTGACCAGGGAGGAAGTCACACTCAGACCCCTCCCCAGAAGGACACTCAAAAGCATGCTGCTCTAAGGTGGCATCTCTTACAGAAGCAAGAACAGCAGCAAACACAGCAACCCCAAACTGAGTCTTGCCATAGTCAGATGCACAGGCCAATTAAGGTGGAACCTGGATGCAAGCCACATGCCTGTATGCACACAGCACCACCAGAAAACAAAACATGGAAAAAGGTAACTAAGCAAGAGAATCCACCTGCAAGCTGTGATAATGTGCAGCAAAAGAGCATCATTGAGACCATGGAGCAGCATCTGAAGCAGTTTCACGCCAAGTCGTTATTTGACCATAAGGCTCTTACTCTCAAATCACAGAAGCAAGTAAAAGTTGAAATGTCAGGGCCAGTCACAGTTTTGACTAGACAAACCACTGCTGCAGAACTTGATAGCCACACCCCAGCTTTAGAGCAGCAAACAACTTCTTCAGAAAAGACACCAACCAAAAGAACAGCTGCTTCTGTTCTCAATAATTTTATAGAGTCACCTTCCAAATTACTAGATACTCCTATAAAAAATTTATTGGATACACCTGTCAAGACTCAATATGATTTCCCATCTTGCAGATGTGTAGAGCAAATTATTGAAAAAGATGAAGGTCCTTTTTATACCCATCTAGGAGCAGGTCCTAATGTGGCAGCTATTAGAGAAATCATGGAAGAAAGGTTTGGACAGAAGGGTAAAGCTATTAGGATTGAAAGAGTCATCTATACTGGTAAAGAAGGCAAAAGTTCTCAGGGATGTCCTATTGCTAAGTGGGTGGTTCGCAGAAGCAGCAGTGAAGAGAAGCTACTGTGTTTGGTGCGGGAGCGAGCTGGCCACACCTGTGAGGCTGCAGTGATTGTGATTCTCATCCTGGTGTGGGAAGGAATCCCGCTGTCTCTGGCTGACAAACTCTACTCGGAGCTTACCGAGACGCTGAGGAAATACGGCACGCTCACCAATCGCCGGTGTGCCTTGAATGAAGAGAGAACTTGCGCCTGTCAGGGGCTGGATCCAGAAACCTGTGGTGCCTCCTTCTCTTTTGGTTGTTCATGGAGCATGTACTACAATGGATGTAAGTTTGCCAGAAGCAAGATCCCAAGGAAGTTTAAGCTGCTTGGGGATGACCCAAAAGAGGAAGAGAAACTGGAGTCTCATTTGCAAAACCTGTCCACTCTTATGGCACCAACATATAAGAAACTTGCACCTGATGCATATAATAATCAGATTGAATATGAACACAGAGCACCAGAGTGCCGTCTGGGTCTGAAGGAAGGCCGTCCATTCTCAGGGGTCACTGCATGTTTGGACTTCTGTGCTCATGCCCACAGAGACTTGCACAACATGCAGAATGGCAGCACATTGGTATGCACTCTCACTAGAGAAGACAATCGAGAATTTGGAGGAAAACCTGAGGATGAGCAGCTTCACGTTCTGCCTTTATACAAAGTCTCTGACGTGGATGAGTTTGGGAGTGTGGAAGCTCAGGAGGAGAAAAAACGGAGTGGTGCCATTCAGGTACTGAGTTCTTTTCGGCGAAAAGTCAGGATGTTAGCAGAGCCAGTCAAGACTTGCCGACAAAGGAAACTAGAAGCCAAGAAAGCTGCAGCTGAAAAGCTTTCCTCCCTGGAGAACAGCTCAAATAAAAATGAAAAGGAAAAGTCAGCCCCATCACGTACAAAACAAACTGAAAACGCAAGCCAGGCTAAACAGTTGGCAGAACTTTTGCGACTTTCAGGACCAGTCATGCAGCAGTCCCAGCAGCCCCAGCCTCTACAGAAGCAGCCACCACAGCCCCAGCAGCAGCAGAGACCCCAGCAGCAGCAGCCACATCACCCTCAGACAGAGTCTGTCAACTCTTATTCTGCTTCTGGATCCACCAATCCATACATGAGACGGCCCAATCCAGTTAGTCCTTATCCAAACTCTTCACACACTTCAGATATCTATGGAAGCACCAGCCCTATGAACTTCTATTCCACCTCATCTCAAGCTGCAGGTTCATATTTGAATTCTTCTAATCCCATGAACCCTTACCCTGGGCTTTTGAATCAGAATACCCAATATCCATCATATCAATGCAATGGAAACCTATCAGTGGACAACTGCTCCCCATATCTGGGTTCCTATTCTCCCCAGTCTCAGCCGATGGATCTGTATAGGTATCCAAGCCAAGACCCTCTGTCTAAGCTCAGTCTACCACCCATCCATACACTTTACCAGCCAAGGTTTGGAAATAGCCAGAGTTTTACATCTAAATACTTAGGTTATGGAAACCAAAATATGCAGGGAGATGGTTTCAGCAGTTGTACCATTAGACCAAATGTACATCATGTAGGGAAATTGCCTCCTTATCCCACTCATGAGATGGATGGCCACTTCATGGGAGCCACCTCTAGATTACCACCCAATCTGAGCAATCCAAACATGGACTATAAAAATGGTGAACATCATTCACCTTCTCACATAATCCATAACTACAGTGCAGCTCCGGGCATGTTCAACAGCTCTCTTCATGCCCTGCATCTCCAAAACAAGGAGAATGACATGCTTTCCCACACAGCTAATGGGTTATCAAAGATGCTTCCAGCTCTTAACCATGATAGAACTGCTTGTGTCCAAGGAGGCTTACACAAATTAAGTGATGCTAATGGTCAGGAAAAGCAGCCATTGGCACTAGTCCAGGGTGTGGCTTCTGGTGCAGAGGACAACGATGAGGTCTGGTCAGACAGCGAGCAGAGCTTTCTGGATCCTGACATTGGGGGAGTGGCCGTGGCTCCAACTCATGGGTCAATTCTCATTGAGTGTGCAAAGCGTGAGCTGCATGCCACAACCCCTTTAAAGAATCCCAATAGGAATCACCCCACCAGGATCTCCCTCGTCTTTTACCAGCATAAGAGCATGAATGAGCCAAAACATGGCTTGGCTCTTTGGGAAGCCAAAATGGCTGAAAAAGCCCGTGAGAAAGAGGAAGAGTGTGAAAAGTATGGCCCAGACTATGTGCCTCAGAAATCCCATGGCAAAAAAGTGAAACGGGAGCCTGCTGAGCCACATGAAACTTCAGAGCCCACTTACCTGCGTTTCATCAAGTCTCTTGCCGAAAGGACCATGTCCGTGACCACAGACTCCACAGTAACTACATCTCCATATGCCTTCACTCGGGTCACAGGGCCTTACAACAGATATATATGA'\n",
    "\n",
    "    #Split the sequence into codons \n",
    "    TET2_codons_NP_01120680 = [TET2_exons_NP_01120680[i:i+3] for i in range(0, len(TET2_exons_NP_01120680), 3)]\n",
    "\n",
    "    #TET2 transcript (transcript variant 2 (TET2-001), NM_017628, NP_001120680):\n",
    "    #Downloaded from http://grch37.ensembl.org/Homo_sapiens/Transcript/Exons?db=core;g=ENSG00000168769;r=4:106067032-106163928;t=ENST00000380013\n",
    "    TET2_exons_extra_bases_NP_01120680 = 'GATGGAACAGGATAGAACCAACCATGTTGAGGGCAACAGACTAAGTCCATTCCTGATACCATCACCTCCCATTTGCCAGACAGAACCTCTGGCTACAAAGCTCCAGAATGGAAGCCCACTGCCTGAGAGAGCTCATCCAGAAGTAAATGGAGACACCAAGTGGCACTCTTTCAAAAGTTATTATGGAATACCCTGTATGAAGGGAAGCCAGAATAGTCGTGTGAGTCCTGACTTTACACAAGAAAGTAGAGGGTATTCCAAGTGTTTGCAAAATGGAGGAATAAAACGCACAGTTAGTGAACCTTCTCTCTCTGGGCTCCTTCAGATCAAGAAATTGAAACAAGACCAAAAGGCTAATGGAGAAAGACGTAACTTCGGGGTAAGCCAAGAAAGAAATCCAGGTGAAAGCAGTCAACCAAATGTCTCCGATTTGAGTGATAAGAAAGAATCTGTGAGTTCTGTAGCCCAAGAAAATGCAGTTAAAGATTTCACCAGTTTTTCAACACATAACTGCAGTGGGCCTGAAAATCCAGAGCTTCAGATTCTGAATGAGCAGGAGGGGAAAAGTGCTAATTACCATGACAAGAACATTGTATTACTTAAAAACAAGGCAGTGCTAATGCCTAATGGTGCTACAGTTTCTGCCTCTTCCGTGGAACACACACATGGTGAACTCCTGGAAAAAACACTGTCTCAATATTATCCAGATTGTGTTTCCATTGCGGTGCAGAAAACCACATCTCACATAAATGCCATTAACAGTCAGGCTACTAATGAGTTGTCCTGTGAGATCACTCACCCATCGCATACCTCAGGGCAGATCAATTCCGCACAGACCTCTAACTCTGAGCTGCCTCCAAAGCCAGCTGCAGTGGTGAGTGAGGCCTGTGATGCTGATGATGCTGATAATGCCAGTAAACTAGCTGCAATGCTAAATACCTGTTCCTTTCAGAAACCAGAACAACTACAACAACAAAAATCAGTTTTTGAGATATGCCCATCTCCTGCAGAAAATAACATCCAGGGAACCACAAAGCTAGCGTCTGGTGAAGAATTCTGTTCAGGTTCCAGCAGCAATTTGCAAGCTCCTGGTGGCAGCTCTGAACGGTATTTAAAACAAAATGAAATGAATGGTGCTTACTTCAAGCAAAGCTCAGTGTTCACTAAGGATTCCTTTTCTGCCACTACCACACCACCACCACCATCACAATTGCTTCTTTCTCCCCCTCCTCCTCTTCCACAGGTTCCTCAGCTTCCTTCAGAAGGAAAAAGCACTCTGAATGGTGGAGTTTTAGAAGAACACCACCACTACCCCAACCAAAGTAACACAACACTTTTAAGGGAAGTGAAAATAGAGGGTAAACCTGAGGCACCACCTTCCCAGAGTCCTAATCCATCTACACATGTATGCAGCCCTTCTCCGATGCTTTCTGAAAGGCCTCAGAATAATTGTGTGAACAGGAATGACATACAGACTGCAGGGACAATGACTGTTCCATTGTGTTCTGAGAAAACAAGACCAATGTCAGAACACCTCAAGCATAACCCACCAATTTTTGGTAGCAGTGGAGAGCTACAGGACAACTGCCAGCAGTTGATGAGAAACAAAGAGCAAGAGATTCTGAAGGGTCGAGACAAGGAGCAAACACGAGATCTTGTGCCCCCAACACAGCACTATCTGAAACCAGGATGGATTGAATTGAAGGCCCCTCGTTTTCACCAAGCGGAATCCCATCTAAAACGTAATGAGGCATCACTGCCATCAATTCTTCAGTATCAACCCAATCTCTCCAATCAAATGACCTCCAAACAATACACTGGAAATTCCAACATGCCTGGGGGGCTCCCAAGGCAAGCTTACACCCAGAAAACAACACAGCTGGAGCACAAGTCACAAATGTACCAAGTTGAAATGAATCAAGGGCAGTCCCAAGGTACAGTGGACCAACATCTCCAGTTCCAAAAACCCTCACACCAGGTGCACTTCTCCAAAACAGACCATTTACCAAAAGCTCATGTGCAGTCACTGTGTGGCACTAGATTTCATTTTCAACAAAGAGCAGATTCCCAAACTGAAAAACTTATGTCCCCAGTGTTGAAACAGCACTTGAATCAACAGGCTTCAGAGACTGAGCCATTTTCAAACTCACACCTTTTGCAACATAAGCCTCATAAACAGGCAGCACAAACACAACCATCCCAGAGTTCACATCTCCCTCAAAACCAGCAACAGCAGCAAAAATTACAAATAAAGAATAAAGAGGAAATACTCCAGACTTTTCCTCACCCCCAAAGCAACAATGATCAGCAAAGAGAAGGATCATTCTTTGGCCAGACTAAAGTGGAAGAATGTTTTCATGGTGAAAATCAGTATTCAAAATCAAGCGAGTTCGAGACTCATAATGTCCAAATGGGACTGGAGGAAGTACAGAATATAAATCGTAGAAATTCCCCTTATAGTCAGACCATGAAATCAAGTGCATGCAAAATACAGGTTTCTTGTTCAAACAATACACACCTAGTTTCAGAGAATAAAGAACAGACTACACATCCTGAACTTTTTGCAGGAAACAAGACCCAAAACTTGCATCACATGCAATATTTTCCAAATAATGTGATCCCAAAGCAAGATCTTCTTCACAGGTGCTTTCAAGAACAGGAGCAGAAGTCACAACAAGCTTCAGTTCTACAGGGATATAAAAATAGAAACCAAGATATGTCTGGTCAACAAGCTGCGCAACTTGCTCAGCAAAGGTACTTGATACATAACCATGCAAATGTTTTTCCTGTGCCTGACCAGGGAGGAAGTCACACTCAGACCCCTCCCCAGAAGGACACTCAAAAGCATGCTGCTCTAAGGTGGCATCTCTTACAGAAGCAAGAACAGCAGCAAACACAGCAACCCCAAACTGAGTCTTGCCATAGTCAGATGCACAGGCCAATTAAGGTGGAACCTGGATGCAAGCCACATGCCTGTATGCACACAGCACCACCAGAAAACAAAACATGGAAAAAGGTAACTAAGCAAGAGAATCCACCTGCAAGCTGTGATAATGTGCAGCAAAAGAGCATCATTGAGACCATGGAGCAGCATCTGAAGCAGTTTCACGCCAAGTCGTTATTTGACCATAAGGCTCTTACTCTCAAATCACAGAAGCAAGTAAAAGTTGAAATGTCAGGGCCAGTCACAGTTTTGACTAGACAAACCACTGCTGCAGAACTTGATAGCCACACCCCAGCTTTAGAGCAGCAAACAACTTCTTCAGAAAAGACACCAACCAAAAGAACAGCTGCTTCTGTTCTCAATAATTTTATAGAGTCACCTTCCAAATTACTAGATACTCCTATAAAAAATTTATTGGATACACCTGTCAAGACTCAATATGATTTCCCATCTTGCAGATGTGTAGAGCAAATTATTGAAAAAGATGAAGGTCCTTTTTATACCCATCTAGGAGCAGGTCCTAATGTGGCAGCTATTAGAGAAATCATGGAAGAAAGGTTTGGACAGAAGGGTAAAGCTATTAGGATTGAAAGAGTCATCTATACTGGTAAAGAAGGCAAAAGTTCTCAGGGATGTCCTATTGCTAAGTGGGTGGTTCGCAGAAGCAGCAGTGAAGAGAAGCTACTGTGTTTGGTGCGGGAGCGAGCTGGCCACACCTGTGAGGCTGCAGTGATTGTGATTCTCATCCTGGTGTGGGAAGGAATCCCGCTGTCTCTGGCTGACAAACTCTACTCGGAGCTTACCGAGACGCTGAGGAAATACGGCACGCTCACCAATCGCCGGTGTGCCTTGAATGAAGAGAGAACTTGCGCCTGTCAGGGGCTGGATCCAGAAACCTGTGGTGCCTCCTTCTCTTTTGGTTGTTCATGGAGCATGTACTACAATGGATGTAAGTTTGCCAGAAGCAAGATCCCAAGGAAGTTTAAGCTGCTTGGGGATGACCCAAAAGAGGAAGAGAAACTGGAGTCTCATTTGCAAAACCTGTCCACTCTTATGGCACCAACATATAAGAAACTTGCACCTGATGCATATAATAATCAGATTGAATATGAACACAGAGCACCAGAGTGCCGTCTGGGTCTGAAGGAAGGCCGTCCATTCTCAGGGGTCACTGCATGTTTGGACTTCTGTGCTCATGCCCACAGAGACTTGCACAACATGCAGAATGGCAGCACATTGGTATGCACTCTCACTAGAGAAGACAATCGAGAATTTGGAGGAAAACCTGAGGATGAGCAGCTTCACGTTCTGCCTTTATACAAAGTCTCTGACGTGGATGAGTTTGGGAGTGTGGAAGCTCAGGAGGAGAAAAAACGGAGTGGTGCCATTCAGGTACTGAGTTCTTTTCGGCGAAAAGTCAGGATGTTAGCAGAGCCAGTCAAGACTTGCCGACAAAGGAAACTAGAAGCCAAGAAAGCTGCAGCTGAAAAGCTTTCCTCCCTGGAGAACAGCTCAAATAAAAATGAAAAGGAAAAGTCAGCCCCATCACGTACAAAACAAACTGAAAACGCAAGCCAGGCTAAACAGTTGGCAGAACTTTTGCGACTTTCAGGACCAGTCATGCAGCAGTCCCAGCAGCCCCAGCCTCTACAGAAGCAGCCACCACAGCCCCAGCAGCAGCAGAGACCCCAGCAGCAGCAGCCACATCACCCTCAGACAGAGTCTGTCAACTCTTATTCTGCTTCTGGATCCACCAATCCATACATGAGACGGCCCAATCCAGTTAGTCCTTATCCAAACTCTTCACACACTTCAGATATCTATGGAAGCACCAGCCCTATGAACTTCTATTCCACCTCATCTCAAGCTGCAGGTTCATATTTGAATTCTTCTAATCCCATGAACCCTTACCCTGGGCTTTTGAATCAGAATACCCAATATCCATCATATCAATGCAATGGAAACCTATCAGTGGACAACTGCTCCCCATATCTGGGTTCCTATTCTCCCCAGTCTCAGCCGATGGATCTGTATAGGTATCCAAGCCAAGACCCTCTGTCTAAGCTCAGTCTACCACCCATCCATACACTTTACCAGCCAAGGTTTGGAAATAGCCAGAGTTTTACATCTAAATACTTAGGTTATGGAAACCAAAATATGCAGGGAGATGGTTTCAGCAGTTGTACCATTAGACCAAATGTACATCATGTAGGGAAATTGCCTCCTTATCCCACTCATGAGATGGATGGCCACTTCATGGGAGCCACCTCTAGATTACCACCCAATCTGAGCAATCCAAACATGGACTATAAAAATGGTGAACATCATTCACCTTCTCACATAATCCATAACTACAGTGCAGCTCCGGGCATGTTCAACAGCTCTCTTCATGCCCTGCATCTCCAAAACAAGGAGAATGACATGCTTTCCCACACAGCTAATGGGTTATCAAAGATGCTTCCAGCTCTTAACCATGATAGAACTGCTTGTGTCCAAGGAGGCTTACACAAATTAAGTGATGCTAATGGTCAGGAAAAGCAGCCATTGGCACTAGTCCAGGGTGTGGCTTCTGGTGCAGAGGACAACGATGAGGTCTGGTCAGACAGCGAGCAGAGCTTTCTGGATCCTGACATTGGGGGAGTGGCCGTGGCTCCAACTCATGGGTCAATTCTCATTGAGTGTGCAAAGCGTGAGCTGCATGCCACAACCCCTTTAAAGAATCCCAATAGGAATCACCCCACCAGGATCTCCCTCGTCTTTTACCAGCATAAGAGCATGAATGAGCCAAAACATGGCTTGGCTCTTTGGGAAGCCAAAATGGCTGAAAAAGCCCGTGAGAAAGAGGAAGAGTGTGAAAAGTATGGCCCAGACTATGTGCCTCAGAAATCCCATGGCAAAAAAGTGAAACGGGAGCCTGCTGAGCCACATGAAACTTCAGAGCCCACTTACCTGCGTTTCATCAAGTCTCTTGCCGAAAGGACCATGTCCGTGACCACAGACTCCACAGTAACTACATCTCCATATGCCTTCACTCGGGTCACAGGGCCTTACAACAGATATATATGAT'\n",
    "\n",
    "    #List of all amino acid positions (across whole coding region) covered (amino acid position minus 1 (as python starts at 0))\n",
    "    TET2_missense_codons_DICT_NP_01120680={}\n",
    "    for i in range(lower_range, upper_range):\n",
    "        TET2_missense_codons_DICT_NP_01120680[i+1]= (TET2_codons_NP_01120680[i])\n",
    "\n",
    "    calls_TET2_missense_NP_01120680 = {}\n",
    "    for k,v in TET2_missense_codons_DICT_NP_01120680.items(): #iterating over the list of codon numbers (and codon sequences)\n",
    "        if v in Possible_Amino_Acid_Changes.keys(): #assign the list of all the possible amino acid changes to each codon they called\n",
    "            changes = Possible_Amino_Acid_Changes_letters[v]\n",
    "    #         print(Possible_Amino_Acid_Changes.values())\n",
    "            calls_TET2_missense_NP_01120680[k]=(v, changes)\n",
    "\n",
    "    #Changing the original amino acid to a Word (e.g. Phe) rather than a codon\n",
    "    calls_TET2_missense_NP_01120680_extra = {}\n",
    "\n",
    "    for k, v in calls_TET2_missense_NP_01120680.items():\n",
    "        if v[0] in Genetic_Code:\n",
    "            Amino_acid_change = Genetic_Code.get(v[0])\n",
    "            calls_TET2_missense_NP_01120680_extra[k] = (Amino_acid_change, v[0], v[1])\n",
    "            #e.g. 12: (Phe, 'K', '*', 'T' etc..)\n",
    "\n",
    "    #Changing the original amino acid to a LETTER (rather than the codon)\n",
    "    calls_TET2_missense_NP_01120680_extra2 = {}\n",
    "\n",
    "    for k, v in calls_TET2_missense_NP_01120680_extra.items():\n",
    "        if v[0] in Amino_acid_abbreviations:\n",
    "            Amino_acid_abbrev = Amino_acid_abbreviations.get(v[0])\n",
    "            calls_TET2_missense_NP_01120680_extra2[k] = (Amino_acid_abbrev, v[1], v[2])\n",
    "            #e.g. 12: ((F), ('K', '*', 'T' etc..))\n",
    "\n",
    "    #Create a dictionary that contains details of the positions of missense variants at that codon\n",
    "    TET2_position_in_list_dict_missense_NP_01120680 = {}\n",
    "\n",
    "    for k, v in calls_TET2_missense_NP_01120680_extra2.items():\n",
    "        original_AA = v[0]\n",
    "        possible_AA_changes = v[2]\n",
    "        possible_AA_changes_string = ''.join(v[2])\n",
    "        positions_in_list = []\n",
    "        for i in possible_AA_changes:\n",
    "            if i != original_AA: #not synonymous\n",
    "                if i != '*': #not missense\n",
    "                    position_in_list=-1\n",
    "                    missense_change = i\n",
    "                    new_aa_list = []\n",
    "                    while True:\n",
    "                        position_in_list = possible_AA_changes_string.find(missense_change, position_in_list+1)\n",
    "        #               print(position_in_list)\n",
    "                        if position_in_list == -1: break\n",
    "                        positions_in_list.append(position_in_list)\n",
    "                        positions_in_list = set(positions_in_list) #convert into a 'set' to remove duplicate positions\n",
    "                        positions_in_list = list(positions_in_list) #convert back to a list so can be added to dictionary\n",
    "                    for pos in positions_in_list:\n",
    "                        new_aa = possible_AA_changes_string[pos]\n",
    "                        new_aa_list.append(new_aa)\n",
    "                    TET2_position_in_list_dict_missense_NP_01120680[k]=(positions_in_list, v, new_aa_list)\n",
    "\n",
    "        aa_positions_variants_called = []\n",
    "        for k, v in TET2_position_in_list_dict_missense_NP_01120680.items():\n",
    "            aa_positions_variants_called.append(k)\n",
    "    # print('positions at which missense variants called = ', aa_positions_variants_called)\n",
    "\n",
    "    #Create a dictionary that contains the position in the list of possible AA where the variant is and what it changes to\n",
    "    TET2_bases_missense_NP_01120680 = {}\n",
    "    for k, v in TET2_position_in_list_dict_missense_NP_01120680.items():\n",
    "        original_a_a = v[1][0]\n",
    "        codon_seq = v[1][1]\n",
    "        base_changes=[]\n",
    "        new_a_a_list = v[2]\n",
    "        if codon_seq in Possible_Base_Changes.keys():\n",
    "            list_of_possible_bases = Possible_Base_Changes.get(codon_seq)\n",
    "        for i in v[0]:\n",
    "            base_change_position = i\n",
    "            base_change = list_of_possible_bases[base_change_position]\n",
    "            base_changes.append(base_change)\n",
    "        TET2_bases_missense_NP_01120680[k]=(v, base_changes, original_a_a, new_a_a_list)\n",
    "\n",
    "    #Create dictionary which contains original codon seq and details on which base is changed\n",
    "    TET2_bases_missense2_NP_01120680 = {}\n",
    "\n",
    "    for k, v in TET2_bases_missense_NP_01120680.items():\n",
    "        base_change_positions = v[0][0]\n",
    "        codon_positions = []\n",
    "        positions_in_gene = []\n",
    "        new_a_a_list = v[3]\n",
    "        for i in base_change_positions:\n",
    "            base_change_position = i\n",
    "            if base_change_position <= 2:\n",
    "                codon_pos = 1\n",
    "            if 2 < base_change_position <= 5:\n",
    "                codon_pos = 2\n",
    "            if 5 < base_change_position <= 8:\n",
    "                codon_pos = 3\n",
    "            codon_positions.append(codon_pos)\n",
    "        TET2_bases_missense2_NP_01120680[k] = (v, codon_positions)\n",
    "\n",
    "    #Create dictionary which contains original codon seq, details on which base is changed and it's position in the coding region\n",
    "    TET2_bases_missense3_NP_01120680 = {}\n",
    "\n",
    "    for k, v in TET2_bases_missense2_NP_01120680.items():\n",
    "        codon_positions = v[1]\n",
    "        base_changes = v[0][1]\n",
    "        original_a_a = v[0][2]\n",
    "        new_aa_list = v[0][0][2]\n",
    "        original_bases = []\n",
    "        positions_in_gene = []\n",
    "        codon_number = k\n",
    "        for i in codon_positions:\n",
    "            codon_pos = i\n",
    "            position_in_gene = ((int(k)-1)*3)+codon_pos #codon number minus 1 (e.g. codon 307 -1 = 306, then times by 3 takes you to beginning of codon 307 (then add the position))  \n",
    "            positions_in_gene.append(position_in_gene)\n",
    "            original_base = (v[0][0][1][1][codon_pos-1])\n",
    "            original_bases.append(original_base)\n",
    "        TET2_bases_missense3_NP_01120680[k] = (v[0][0][1], positions_in_gene, original_bases, base_changes, codon_number, original_a_a, new_aa_list)\n",
    "\n",
    "    # Create dictionary which contains original codon seq, details on which base is changed and its neighbouring bases\n",
    "    TET2_bases_missense4_NP_01120680 = {}\n",
    "\n",
    "    for k, v in TET2_bases_missense3_NP_01120680.items():\n",
    "        position_in_gene = v[1]\n",
    "        sites_trio_missense=[]\n",
    "        new_a_a_list = v[6]\n",
    "\n",
    "        for i in position_in_gene:\n",
    "    #         print(i)\n",
    "            if i in intron_exon_boundary_caution_base_TET2_NP01120680:\n",
    "#                 print('positions at intron_exon_boundary =',i)\n",
    "                adapted_trio = intron_exon_check_TET2_NP01120680(i)\n",
    "                sites_trio_missense.append([adapted_trio])\n",
    "                TET2_bases_missense4_NP_01120680[k] = (v, sites_trio_missense)\n",
    "\n",
    "            else:\n",
    "                for i in position_in_gene:\n",
    "                    gene_pos = i\n",
    "                    site_trio_missense = [TET2_exons_extra_bases_NP_01120680[gene_pos:gene_pos+3] for gene_pos in range (gene_pos-1, gene_pos, 3)]\n",
    "                    sites_trio_missense.append(site_trio_missense)\n",
    "                    TET2_bases_missense4_NP_01120680[k] = (v, sites_trio_missense)\n",
    "\n",
    "    #Create a dictionary containing pos in coding region, original base, new base, base affected with its neighbours\n",
    "    #e.g. 938, G, A, TGG = base 938 on coding region, original base = G, changes to base A, base with neighbours either side = TGG\n",
    "    TET2_sites_missense5_NP_01120680 = {}\n",
    "\n",
    "    a=0\n",
    "    for k, v in TET2_bases_missense4_NP_01120680.items():\n",
    "        original_a_a = v[0][0][0]\n",
    "        new_a_a = v[0][6]\n",
    "        codon_n = v[0][1] #base number in codon (out of 3 x 912)\n",
    "        codon_n_912 = v[0][4]\n",
    "        original_base = v[0][2]\n",
    "        new_base = v[0][3]\n",
    "        site_trio = v[1]\n",
    "        info=[]\n",
    "\n",
    "        change = 0\n",
    "        for i in codon_n:\n",
    "            codon_number = i\n",
    "            orig_base = original_base[change]\n",
    "            new_amino_acid = new_a_a[change]\n",
    "            n_base = new_base[change]\n",
    "            site_tri = site_trio[change]\n",
    "            info.append([codon_number, orig_base, n_base, site_tri, codon_n_912, original_a_a, new_amino_acid])\n",
    "            change = change+1\n",
    "        TET2_sites_missense5_NP_01120680[a]=info\n",
    "        a=a+1\n",
    "    #     print(TET2_sites_missense5_NP_01120680)\n",
    "\n",
    "    # Write in a way that can iterate over mutation rates (i.e. put in to form C[C>A]G etc)   \n",
    "    TET2_sites_missense6_NP_01120680={}\n",
    "\n",
    "    separating_positions = {} # separate out all the possible changes to just 1 effective item in the list\n",
    "    a=0\n",
    "    for k, v in TET2_sites_missense5_NP_01120680.items():\n",
    "        for i in v: #because some contain more than 1 change from that codon:\n",
    "            separating_positions[a]=i\n",
    "            a=a+1\n",
    "\n",
    "    for k, v in separating_positions.items():\n",
    "        base_pos = v[0]\n",
    "        original_base = v[1]\n",
    "        new_base = v[2]\n",
    "        site_trio = v[3]\n",
    "        codon_n_912 = v[4]\n",
    "        original_a_a = v[5]\n",
    "        new_a_a = v[6]\n",
    "        sequence = (str(original_base),str(new_base))\n",
    "        changes = '>'.join(sequence)\n",
    "        left_base = site_trio[0][0]\n",
    "        right_base = site_trio[0][2]\n",
    "        sequence2 = (changes, right_base)\n",
    "        changes2 = ']'.join(sequence2)\n",
    "        sequence3 = (left_base, changes2)\n",
    "        changes3 = '['.join(sequence3)\n",
    "        TET2_sites_missense6_NP_01120680[k] = (v[0],changes3, codon_n_912, original_a_a, new_a_a) #site_trio = the affected base and its neighbours\n",
    "\n",
    "    #Assign the mutation rates to each site...\n",
    "    TET2_mutation_rates_missense_NP_01120680 = {}\n",
    "    for k, v in TET2_sites_missense6_NP_01120680.items():\n",
    "        base_change_info = v[1]\n",
    "        if base_change_info in Mutation_rates.keys():\n",
    "            mutation_rate = Mutation_rates.get(base_change_info)\n",
    "            original_a_a = v[3]\n",
    "            new_a_a = v[4]\n",
    "            aa_positioning = v[2]\n",
    "            variant_id = str(original_a_a)+str(aa_positioning)+str(new_a_a)\n",
    "            TET2_mutation_rates_missense_NP_01120680[k] = (v[0],v[1], mutation_rate, variant_id)\n",
    "\n",
    "    TET2_mutation_rate_missense_NP_01120680_sum = []\n",
    "\n",
    "    mutation_rates_list = []\n",
    "    for k, v in TET2_mutation_rates_missense_NP_01120680.items():\n",
    "        codon_position_in_gene=(v[0])\n",
    "        variant = v[3]\n",
    "        site = v[1]\n",
    "        mutation_rate_value = v[2] \n",
    "        mutation_rates_list.append((v[0], variant, site, mutation_rate_value))\n",
    "        TET2_mutation_rate_missense_NP_01120680_sum.append(mutation_rate_value)\n",
    "\n",
    "    #Get the average of all the mutation rates\n",
    "    TET2_mutation_rate_missense_NP_01120680_sum = []\n",
    "    for k, v in TET2_mutation_rates_missense_NP_01120680.items():\n",
    "        mut_rate = v[2]\n",
    "        TET2_mutation_rate_missense_NP_01120680_sum.append(mut_rate)\n",
    "\n",
    "    TET2_Summed_mutation_rate_missense_NP_01120680_added = np.sum(TET2_mutation_rate_missense_NP_01120680_sum)\n",
    "\n",
    "    print('Total summed mutation rate across TET2 missense_NP_01120680 for this study = ', TET2_Summed_mutation_rate_missense_NP_01120680_added)\n",
    "\n",
    "    mutation_rates_list_all = mutation_rates_list\n",
    "    return mutation_rates_list_all # e.g. lower= 0, upper = 913 for whole gene\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 302,
   "metadata": {
    "code_folding": [
     0
    ]
   },
   "outputs": [],
   "source": [
    "def mutation_rates_nonsense_range_of_variants_TET2_shorter(lower_range, upper_range):\n",
    "    \n",
    "    #TET2 transcript:\n",
    "    #Downloaded from http://grch37.ensembl.org/Homo_sapiens/Transcript/Summary?db=core;g=ENSG00000168769;r=4:106067032-106200973;t=ENST00000305737\n",
    "    TET2_exons_NP_060098 = 'ATGGAACAGGATAGAACCAACCATGTTGAGGGCAACAGACTAAGTCCATTCCTGATACCATCACCTCCCATTTGCCAGACAGAACCTCTGGCTACAAAGCTCCAGAATGGAAGCCCACTGCCTGAGAGAGCTCATCCAGAAGTAAATGGAGACACCAAGTGGCACTCTTTCAAAAGTTATTATGGAATACCCTGTATGAAGGGAAGCCAGAATAGTCGTGTGAGTCCTGACTTTACACAAGAAAGTAGAGGGTATTCCAAGTGTTTGCAAAATGGAGGAATAAAACGCACAGTTAGTGAACCTTCTCTCTCTGGGCTCCTTCAGATCAAGAAATTGAAACAAGACCAAAAGGCTAATGGAGAAAGACGTAACTTCGGGGTAAGCCAAGAAAGAAATCCAGGTGAAAGCAGTCAACCAAATGTCTCCGATTTGAGTGATAAGAAAGAATCTGTGAGTTCTGTAGCCCAAGAAAATGCAGTTAAAGATTTCACCAGTTTTTCAACACATAACTGCAGTGGGCCTGAAAATCCAGAGCTTCAGATTCTGAATGAGCAGGAGGGGAAAAGTGCTAATTACCATGACAAGAACATTGTATTACTTAAAAACAAGGCAGTGCTAATGCCTAATGGTGCTACAGTTTCTGCCTCTTCCGTGGAACACACACATGGTGAACTCCTGGAAAAAACACTGTCTCAATATTATCCAGATTGTGTTTCCATTGCGGTGCAGAAAACCACATCTCACATAAATGCCATTAACAGTCAGGCTACTAATGAGTTGTCCTGTGAGATCACTCACCCATCGCATACCTCAGGGCAGATCAATTCCGCACAGACCTCTAACTCTGAGCTGCCTCCAAAGCCAGCTGCAGTGGTGAGTGAGGCCTGTGATGCTGATGATGCTGATAATGCCAGTAAACTAGCTGCAATGCTAAATACCTGTTCCTTTCAGAAACCAGAACAACTACAACAACAAAAATCAGTTTTTGAGATATGCCCATCTCCTGCAGAAAATAACATCCAGGGAACCACAAAGCTAGCGTCTGGTGAAGAATTCTGTTCAGGTTCCAGCAGCAATTTGCAAGCTCCTGGTGGCAGCTCTGAACGGTATTTAAAACAAAATGAAATGAATGGTGCTTACTTCAAGCAAAGCTCAGTGTTCACTAAGGATTCCTTTTCTGCCACTACCACACCACCACCACCATCACAATTGCTTCTTTCTCCCCCTCCTCCTCTTCCACAGGTTCCTCAGCTTCCTTCAGAAGGAAAAAGCACTCTGAATGGTGGAGTTTTAGAAGAACACCACCACTACCCCAACCAAAGTAACACAACACTTTTAAGGGAAGTGAAAATAGAGGGTAAACCTGAGGCACCACCTTCCCAGAGTCCTAATCCATCTACACATGTATGCAGCCCTTCTCCGATGCTTTCTGAAAGGCCTCAGAATAATTGTGTGAACAGGAATGACATACAGACTGCAGGGACAATGACTGTTCCATTGTGTTCTGAGAAAACAAGACCAATGTCAGAACACCTCAAGCATAACCCACCAATTTTTGGTAGCAGTGGAGAGCTACAGGACAACTGCCAGCAGTTGATGAGAAACAAAGAGCAAGAGATTCTGAAGGGTCGAGACAAGGAGCAAACACGAGATCTTGTGCCCCCAACACAGCACTATCTGAAACCAGGATGGATTGAATTGAAGGCCCCTCGTTTTCACCAAGCGGAATCCCATCTAAAACGTAATGAGGCATCACTGCCATCAATTCTTCAGTATCAACCCAATCTCTCCAATCAAATGACCTCCAAACAATACACTGGAAATTCCAACATGCCTGGGGGGCTCCCAAGGCAAGCTTACACCCAGAAAACAACACAGCTGGAGCACAAGTCACAAATGTACCAAGTTGAAATGAATCAAGGGCAGTCCCAAGGTACAGTGGACCAACATCTCCAGTTCCAAAAACCCTCACACCAGGTGCACTTCTCCAAAACAGACCATTTACCAAAAGCTCATGTGCAGTCACTGTGTGGCACTAGATTTCATTTTCAACAAAGAGCAGATTCCCAAACTGAAAAACTTATGTCCCCAGTGTTGAAACAGCACTTGAATCAACAGGCTTCAGAGACTGAGCCATTTTCAAACTCACACCTTTTGCAACATAAGCCTCATAAACAGGCAGCACAAACACAACCATCCCAGAGTTCACATCTCCCTCAAAACCAGCAACAGCAGCAAAAATTACAAATAAAGAATAAAGAGGAAATACTCCAGACTTTTCCTCACCCCCAAAGCAACAATGATCAGCAAAGAGAAGGATCATTCTTTGGCCAGACTAAAGTGGAAGAATGTTTTCATGGTGAAAATCAGTATTCAAAATCAAGCGAGTTCGAGACTCATAATGTCCAAATGGGACTGGAGGAAGTACAGAATATAAATCGTAGAAATTCCCCTTATAGTCAGACCATGAAATCAAGTGCATGCAAAATACAGGTTTCTTGTTCAAACAATACACACCTAGTTTCAGAGAATAAAGAACAGACTACACATCCTGAACTTTTTGCAGGAAACAAGACCCAAAACTTGCATCACATGCAATATTTTCCAAATAATGTGATCCCAAAGCAAGATCTTCTTCACAGGTGCTTTCAAGAACAGGAGCAGAAGTCACAACAAGCTTCAGTTCTACAGGGATATAAAAATAGAAACCAAGATATGTCTGGTCAACAAGCTGCGCAACTTGCTCAGCAAAGGTACTTGATACATAACCATGCAAATGTTTTTCCTGTGCCTGACCAGGGAGGAAGTCACACTCAGACCCCTCCCCAGAAGGACACTCAAAAGCATGCTGCTCTAAGGTGGCATCTCTTACAGAAGCAAGAACAGCAGCAAACACAGCAACCCCAAACTGAGTCTTGCCATAGTCAGATGCACAGGCCAATTAAGGTGGAACCTGGATGCAAGCCACATGCCTGTATGCACACAGCACCACCAGAAAACAAAACATGGAAAAAGGTAACTAAGCAAGAGAATCCACCTGCAAGCTGTGATAATGTGCAGCAAAAGAGCATCATTGAGACCATGGAGCAGCATCTGAAGCAGTTTCACGCCAAGTCGTTATTTGACCATAAGGCTCTTACTCTCAAATCACAGAAGCAAGTAAAAGTTGAAATGTCAGGGCCAGTCACAGTTTTGACTAGACAAACCACTGCTGCAGAACTTGATAGCCACACCCCAGCTTTAGAGCAGCAAACAACTTCTTCAGAAAAGACACCAACCAAAAGAACAGCTGCTTCTGTTCTCAATAATTTTATAGAGTCACCTTCCAAATTACTAGATACTCCTATAAAAAATTTATTGGATACACCTGTCAAGACTCAATATGATTTCCCATCTTGCAGATGTGTAGGTAAGTGCCAGAAATGTACTGAGACACATGGCGTTTATCCAGAATTAGCAAATTTATCTTCAGATATGGGATTTTCCTTCTTTTTTTAA'\n",
    "\n",
    "    #Split the sequence into codons \n",
    "    TET2_codons_NP_060098 = [TET2_exons_NP_060098[i:i+3] for i in range(0, len(TET2_exons_NP_060098), 3)]\n",
    "\n",
    "    #TET2 transcript 1 extra base at each end:\n",
    "    #Downloaded from http://grch37.ensembl.org/Homo_sapiens/Transcript/Summary?db=core;g=ENSG00000168769;r=4:106067032-106200973;t=ENST00000305737\n",
    "    TET2_exons_extra_bases_NP_060098 = 'GATGGAACAGGATAGAACCAACCATGTTGAGGGCAACAGACTAAGTCCATTCCTGATACCATCACCTCCCATTTGCCAGACAGAACCTCTGGCTACAAAGCTCCAGAATGGAAGCCCACTGCCTGAGAGAGCTCATCCAGAAGTAAATGGAGACACCAAGTGGCACTCTTTCAAAAGTTATTATGGAATACCCTGTATGAAGGGAAGCCAGAATAGTCGTGTGAGTCCTGACTTTACACAAGAAAGTAGAGGGTATTCCAAGTGTTTGCAAAATGGAGGAATAAAACGCACAGTTAGTGAACCTTCTCTCTCTGGGCTCCTTCAGATCAAGAAATTGAAACAAGACCAAAAGGCTAATGGAGAAAGACGTAACTTCGGGGTAAGCCAAGAAAGAAATCCAGGTGAAAGCAGTCAACCAAATGTCTCCGATTTGAGTGATAAGAAAGAATCTGTGAGTTCTGTAGCCCAAGAAAATGCAGTTAAAGATTTCACCAGTTTTTCAACACATAACTGCAGTGGGCCTGAAAATCCAGAGCTTCAGATTCTGAATGAGCAGGAGGGGAAAAGTGCTAATTACCATGACAAGAACATTGTATTACTTAAAAACAAGGCAGTGCTAATGCCTAATGGTGCTACAGTTTCTGCCTCTTCCGTGGAACACACACATGGTGAACTCCTGGAAAAAACACTGTCTCAATATTATCCAGATTGTGTTTCCATTGCGGTGCAGAAAACCACATCTCACATAAATGCCATTAACAGTCAGGCTACTAATGAGTTGTCCTGTGAGATCACTCACCCATCGCATACCTCAGGGCAGATCAATTCCGCACAGACCTCTAACTCTGAGCTGCCTCCAAAGCCAGCTGCAGTGGTGAGTGAGGCCTGTGATGCTGATGATGCTGATAATGCCAGTAAACTAGCTGCAATGCTAAATACCTGTTCCTTTCAGAAACCAGAACAACTACAACAACAAAAATCAGTTTTTGAGATATGCCCATCTCCTGCAGAAAATAACATCCAGGGAACCACAAAGCTAGCGTCTGGTGAAGAATTCTGTTCAGGTTCCAGCAGCAATTTGCAAGCTCCTGGTGGCAGCTCTGAACGGTATTTAAAACAAAATGAAATGAATGGTGCTTACTTCAAGCAAAGCTCAGTGTTCACTAAGGATTCCTTTTCTGCCACTACCACACCACCACCACCATCACAATTGCTTCTTTCTCCCCCTCCTCCTCTTCCACAGGTTCCTCAGCTTCCTTCAGAAGGAAAAAGCACTCTGAATGGTGGAGTTTTAGAAGAACACCACCACTACCCCAACCAAAGTAACACAACACTTTTAAGGGAAGTGAAAATAGAGGGTAAACCTGAGGCACCACCTTCCCAGAGTCCTAATCCATCTACACATGTATGCAGCCCTTCTCCGATGCTTTCTGAAAGGCCTCAGAATAATTGTGTGAACAGGAATGACATACAGACTGCAGGGACAATGACTGTTCCATTGTGTTCTGAGAAAACAAGACCAATGTCAGAACACCTCAAGCATAACCCACCAATTTTTGGTAGCAGTGGAGAGCTACAGGACAACTGCCAGCAGTTGATGAGAAACAAAGAGCAAGAGATTCTGAAGGGTCGAGACAAGGAGCAAACACGAGATCTTGTGCCCCCAACACAGCACTATCTGAAACCAGGATGGATTGAATTGAAGGCCCCTCGTTTTCACCAAGCGGAATCCCATCTAAAACGTAATGAGGCATCACTGCCATCAATTCTTCAGTATCAACCCAATCTCTCCAATCAAATGACCTCCAAACAATACACTGGAAATTCCAACATGCCTGGGGGGCTCCCAAGGCAAGCTTACACCCAGAAAACAACACAGCTGGAGCACAAGTCACAAATGTACCAAGTTGAAATGAATCAAGGGCAGTCCCAAGGTACAGTGGACCAACATCTCCAGTTCCAAAAACCCTCACACCAGGTGCACTTCTCCAAAACAGACCATTTACCAAAAGCTCATGTGCAGTCACTGTGTGGCACTAGATTTCATTTTCAACAAAGAGCAGATTCCCAAACTGAAAAACTTATGTCCCCAGTGTTGAAACAGCACTTGAATCAACAGGCTTCAGAGACTGAGCCATTTTCAAACTCACACCTTTTGCAACATAAGCCTCATAAACAGGCAGCACAAACACAACCATCCCAGAGTTCACATCTCCCTCAAAACCAGCAACAGCAGCAAAAATTACAAATAAAGAATAAAGAGGAAATACTCCAGACTTTTCCTCACCCCCAAAGCAACAATGATCAGCAAAGAGAAGGATCATTCTTTGGCCAGACTAAAGTGGAAGAATGTTTTCATGGTGAAAATCAGTATTCAAAATCAAGCGAGTTCGAGACTCATAATGTCCAAATGGGACTGGAGGAAGTACAGAATATAAATCGTAGAAATTCCCCTTATAGTCAGACCATGAAATCAAGTGCATGCAAAATACAGGTTTCTTGTTCAAACAATACACACCTAGTTTCAGAGAATAAAGAACAGACTACACATCCTGAACTTTTTGCAGGAAACAAGACCCAAAACTTGCATCACATGCAATATTTTCCAAATAATGTGATCCCAAAGCAAGATCTTCTTCACAGGTGCTTTCAAGAACAGGAGCAGAAGTCACAACAAGCTTCAGTTCTACAGGGATATAAAAATAGAAACCAAGATATGTCTGGTCAACAAGCTGCGCAACTTGCTCAGCAAAGGTACTTGATACATAACCATGCAAATGTTTTTCCTGTGCCTGACCAGGGAGGAAGTCACACTCAGACCCCTCCCCAGAAGGACACTCAAAAGCATGCTGCTCTAAGGTGGCATCTCTTACAGAAGCAAGAACAGCAGCAAACACAGCAACCCCAAACTGAGTCTTGCCATAGTCAGATGCACAGGCCAATTAAGGTGGAACCTGGATGCAAGCCACATGCCTGTATGCACACAGCACCACCAGAAAACAAAACATGGAAAAAGGTAACTAAGCAAGAGAATCCACCTGCAAGCTGTGATAATGTGCAGCAAAAGAGCATCATTGAGACCATGGAGCAGCATCTGAAGCAGTTTCACGCCAAGTCGTTATTTGACCATAAGGCTCTTACTCTCAAATCACAGAAGCAAGTAAAAGTTGAAATGTCAGGGCCAGTCACAGTTTTGACTAGACAAACCACTGCTGCAGAACTTGATAGCCACACCCCAGCTTTAGAGCAGCAAACAACTTCTTCAGAAAAGACACCAACCAAAAGAACAGCTGCTTCTGTTCTCAATAATTTTATAGAGTCACCTTCCAAATTACTAGATACTCCTATAAAAAATTTATTGGATACACCTGTCAAGACTCAATATGATTTCCCATCTTGCAGATGTGTAGGTAAGTGCCAGAAATGTACTGAGACACATGGCGTTTATCCAGAATTAGCAAATTTATCTTCAGATATGGGATTTTCCTTCTTTTTTTAAA'\n",
    "\n",
    "    #List of all amino acid positions (across whole coding region) covered (amino acid position minus 1 (as python starts at 0))\n",
    "    TET2_nonsense_codons_DICT_NP_060098={}\n",
    "    for i in range(lower_range, upper_range):\n",
    "        TET2_nonsense_codons_DICT_NP_060098[i+1]= (TET2_codons_NP_060098[i])\n",
    "\n",
    "    calls_TET2_nonsense_NP_060098 = {}\n",
    "    for k,v in TET2_nonsense_codons_DICT_NP_060098.items(): #iterating over the list of codon numbers (and codon sequences)\n",
    "        if v in Possible_Amino_Acid_Changes.keys(): #assign the list of all the possible amino acid changes to each codon they called\n",
    "            changes = Possible_Amino_Acid_Changes_letters[v]\n",
    "    #         print(Possible_Amino_Acid_Changes.values())\n",
    "            calls_TET2_nonsense_NP_060098[k]=(v, changes)\n",
    "\n",
    "    #Changing the original amino acid to a Word (e.g. Phe) rather than a codon\n",
    "    calls_TET2_nonsense_NP_060098_extra = {}\n",
    "\n",
    "    for k, v in calls_TET2_nonsense_NP_060098.items():\n",
    "        if v[0] in Genetic_Code:\n",
    "            Amino_acid_change = Genetic_Code.get(v[0])\n",
    "            calls_TET2_nonsense_NP_060098_extra[k] = (Amino_acid_change, v[0], v[1])\n",
    "            #e.g. 12: (Phe, 'K', '*', 'T' etc..)\n",
    "\n",
    "    #Changing the original amino acid to a LETTER (rather than the codon)\n",
    "    calls_TET2_nonsense_NP_060098_extra2 = {}\n",
    "\n",
    "    for k, v in calls_TET2_nonsense_NP_060098_extra.items():\n",
    "        if v[0] in Amino_acid_abbreviations:\n",
    "            Amino_acid_abbrev = Amino_acid_abbreviations.get(v[0])\n",
    "            calls_TET2_nonsense_NP_060098_extra2[k] = (Amino_acid_abbrev, v[1], v[2])\n",
    "            #e.g. 12: ((F), ('K', '*', 'T' etc..))\n",
    "\n",
    "    #Create a dictionary that contains details of the positions of nonsense variants at that codon\n",
    "    TET2_position_in_list_dict_nonsense_NP_060098 = {}\n",
    "\n",
    "    for k, v in calls_TET2_nonsense_NP_060098_extra2.items():\n",
    "        original_AA = v[0]\n",
    "        possible_AA_changes = v[2]\n",
    "        possible_AA_changes_string = ''.join(v[2])\n",
    "        positions_in_list = []\n",
    "        for i in possible_AA_changes:\n",
    "            if i == '*': #nonsense\n",
    "                position_in_list=-1\n",
    "                nonsense_change = i    \n",
    "                while True:\n",
    "                    position_in_list = possible_AA_changes_string.find(nonsense_change, position_in_list+1)\n",
    "    #               print(position_in_list)\n",
    "                    if position_in_list == -1: break\n",
    "                    positions_in_list.append(position_in_list)\n",
    "                    positions_in_list = set(positions_in_list) #convert into a 'set' to remove duplicate positions\n",
    "                    positions_in_list = list(positions_in_list) #convert back to a list so can be added to dictionary\n",
    "                TET2_position_in_list_dict_nonsense_NP_060098[k]=(positions_in_list, v)\n",
    "\n",
    "        aa_positions_variants_called = []\n",
    "        for k, v in TET2_position_in_list_dict_nonsense_NP_060098.items():\n",
    "            aa_positions_variants_called.append(k)\n",
    "    # print('positions at which nonsense variants called = ', aa_positions_variants_called)\n",
    "\n",
    "    #Create a dictionary that contains the position in the list of possible AA where the variant is and what it changes to\n",
    "    TET2_bases_nonsense_NP_060098 = {}\n",
    "    for k, v in TET2_position_in_list_dict_nonsense_NP_060098.items():\n",
    "        original_a_a = v[1][0]\n",
    "        codon_seq = v[1][1]\n",
    "        base_changes=[]\n",
    "        if codon_seq in Possible_Base_Changes.keys():\n",
    "            list_of_possible_bases = Possible_Base_Changes.get(codon_seq)\n",
    "        for i in v[0]:\n",
    "            base_change_position = i\n",
    "            base_change = list_of_possible_bases[base_change_position]\n",
    "            base_changes.append(base_change)\n",
    "        TET2_bases_nonsense_NP_060098[k]=(v, base_changes, original_a_a)\n",
    "\n",
    "    #Create dictionary which contains original codon seq and details on which base is changed\n",
    "    TET2_bases_nonsense2_NP_060098 = {}\n",
    "\n",
    "    for k, v in TET2_bases_nonsense_NP_060098.items():\n",
    "        base_change_positions = v[0][0]\n",
    "        codon_positions = []\n",
    "        positions_in_gene = []\n",
    "        for i in base_change_positions:\n",
    "            base_change_position = i\n",
    "            if base_change_position <= 2:\n",
    "                codon_pos = 1\n",
    "            if 2 < base_change_position <= 5:\n",
    "                codon_pos = 2\n",
    "            if 5 < base_change_position <= 8:\n",
    "                codon_pos = 3\n",
    "            codon_positions.append(codon_pos)\n",
    "        TET2_bases_nonsense2_NP_060098[k] = (v, codon_positions)\n",
    "\n",
    "    #Create dictionary which contains original codon seq, details on which base is changed and it's position in the coding region\n",
    "    TET2_bases_nonsense3_NP_060098 = {}\n",
    "\n",
    "    for k, v in TET2_bases_nonsense2_NP_060098.items():\n",
    "        codon_positions = v[1]\n",
    "        base_changes = v[0][1]\n",
    "        original_a_a = v[0][2]\n",
    "        original_bases = []\n",
    "        positions_in_gene = []\n",
    "        codon_number = k\n",
    "        for i in codon_positions:\n",
    "            codon_pos = i\n",
    "            position_in_gene = ((int(k)-1)*3)+codon_pos #codon number minus 1 (e.g. codon 307 -1 = 306, then times by 3 takes you to beginning of codon 307 (then add the position))  \n",
    "            positions_in_gene.append(position_in_gene)\n",
    "            original_base = (v[0][0][1][1][codon_pos-1])\n",
    "            original_bases.append(original_base)\n",
    "        TET2_bases_nonsense3_NP_060098[k] = (v[0][0][1], positions_in_gene, original_bases, base_changes, codon_number, original_a_a)\n",
    "\n",
    "    #Create dictionary which contains original codon seq, details on which base is changed and its neighbouring bases\n",
    "    TET2_bases_nonsense4_NP_060098 = {}\n",
    "\n",
    "    for k, v in TET2_bases_nonsense3_NP_060098.items():\n",
    "        position_in_gene = v[1]\n",
    "        sites_trio_nonsense=[]\n",
    "\n",
    "        for i in position_in_gene:\n",
    "            gene_pos = i\n",
    "            site_trio_nonsense = [TET2_exons_extra_bases_NP_060098[gene_pos:gene_pos+3] for gene_pos in range (gene_pos-1, gene_pos, 3)]\n",
    "            sites_trio_nonsense.append(site_trio_nonsense)\n",
    "            TET2_bases_nonsense4_NP_060098[k] = (v, sites_trio_nonsense)\n",
    "\n",
    "    #Create a dictionary containing pos in coding region, original base, new base, base affected with its neighbours\n",
    "    #e.g. 938, G, A, TGG = base 938 on coding region, original base = G, changes to base A, base with neighbours either side = TGG\n",
    "    TET2_sites_nonsense5_NP_060098 = {}\n",
    "\n",
    "    a=0\n",
    "    for k, v in TET2_bases_nonsense4_NP_060098.items():\n",
    "        original_a_a = v[0][0][0]\n",
    "        codon_n = v[0][1] #base number in codon (out of 3 x 912)\n",
    "        codon_n_912 = v[0][4]\n",
    "        original_base = v[0][2]\n",
    "        new_base = v[0][3]\n",
    "        site_trio = v[1]\n",
    "        info=[]\n",
    "\n",
    "        change = 0\n",
    "        for i in codon_n:\n",
    "            codon_number = i\n",
    "            orig_base = original_base[change]\n",
    "            n_base = new_base[change]\n",
    "            site_tri = site_trio[change]\n",
    "            info.append([codon_number, orig_base, n_base, site_tri, codon_n_912, original_a_a])\n",
    "            change = change+1\n",
    "        TET2_sites_nonsense5_NP_060098[a]=info\n",
    "        a=a+1\n",
    "    #     print(TET2_sites_nonsense5_NP_060098)\n",
    "\n",
    "    # Write in a way that can iterate over mutation rates (i.e. put in to form C[C>A]G etc)   \n",
    "    TET2_sites_nonsense6_NP_060098={}\n",
    "\n",
    "    separating_positions = {} # separate out all the possible changes to just 1 effective item in the list\n",
    "    a=0\n",
    "    for k, v in TET2_sites_nonsense5_NP_060098.items():\n",
    "        for i in v: #because some contain more than 1 change from that codon:\n",
    "            separating_positions[a]=i\n",
    "            a=a+1\n",
    "\n",
    "    for k, v in separating_positions.items():\n",
    "        base_pos = v[0]\n",
    "        original_base = v[1]\n",
    "        new_base = v[2]\n",
    "        site_trio = v[3]\n",
    "        codon_n_912 = v[4]\n",
    "        original_a_a = v[5]\n",
    "        sequence = (str(original_base),str(new_base))\n",
    "        changes = '>'.join(sequence)\n",
    "        left_base = site_trio[0][0]\n",
    "        right_base = site_trio[0][2]\n",
    "        sequence2 = (changes, right_base)\n",
    "        changes2 = ']'.join(sequence2)\n",
    "        sequence3 = (left_base, changes2)\n",
    "        changes3 = '['.join(sequence3)\n",
    "        TET2_sites_nonsense6_NP_060098[k] = (v[0],changes3, codon_n_912, original_a_a) #site_trio = the affected base and its neighbours\n",
    "\n",
    "    #Assign the mutation rates to each site...\n",
    "    TET2_mutation_rates_nonsense_NP_060098 = {}\n",
    "    for k, v in TET2_sites_nonsense6_NP_060098.items():\n",
    "        base_change_info = v[1]\n",
    "        if base_change_info in Mutation_rates.keys():\n",
    "            mutation_rate = Mutation_rates.get(base_change_info)\n",
    "            original_a_a = v[3]\n",
    "            aa_positioning = v[2]\n",
    "            variant_type = '*'\n",
    "            variant_id = str(original_a_a)+str(aa_positioning)+variant_type\n",
    "            TET2_mutation_rates_nonsense_NP_060098[k] = (v[0],v[1], mutation_rate, variant_id)\n",
    "\n",
    "    TET2_mutation_rate_nonsense_NP_060098_sum = []\n",
    "\n",
    "    mutation_rates_list = []\n",
    "    for k, v in TET2_mutation_rates_nonsense_NP_060098.items():\n",
    "        codon_position_in_gene=(v[0])\n",
    "        variant = v[3]\n",
    "        site = v[1]\n",
    "        mutation_rate_value = v[2] \n",
    "        mutation_rates_list.append((v[0], variant, site, mutation_rate_value))\n",
    "        TET2_mutation_rate_nonsense_NP_060098_sum.append(mutation_rate_value)\n",
    "\n",
    "    #Get the average of all the mutation rates\n",
    "    TET2_mutation_rate_nonsense_NP_060098_sum = []\n",
    "    for k, v in TET2_mutation_rates_nonsense_NP_060098.items():\n",
    "        mut_rate = v[2]\n",
    "        TET2_mutation_rate_nonsense_NP_060098_sum.append(mut_rate)\n",
    "\n",
    "    TET2_Summed_mutation_rate_nonsense_NP_060098_added = np.sum(TET2_mutation_rate_nonsense_NP_060098_sum)\n",
    "\n",
    "    print('Total summed mutation rate across TET2 nonsense_NP_060098 for this study = ', TET2_Summed_mutation_rate_nonsense_NP_060098_added)\n",
    "\n",
    "    mutation_rates_list_all = mutation_rates_list\n",
    "    return mutation_rates_list_all # e.g. lower= 0, upper = 913 for whole gene\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 303,
   "metadata": {
    "code_folding": [
     0
    ]
   },
   "outputs": [],
   "source": [
    "def mutation_rates_missense_range_of_variants_TET2_shorter(lower_range, upper_range):\n",
    "    \n",
    "    #TET2 transcript:\n",
    "    #Downloaded from http://grch37.ensembl.org/Homo_sapiens/Transcript/Summary?db=core;g=ENSG00000168769;r=4:106067032-106200973;t=ENST00000305737\n",
    "    TET2_exons_NP_060098 = 'ATGGAACAGGATAGAACCAACCATGTTGAGGGCAACAGACTAAGTCCATTCCTGATACCATCACCTCCCATTTGCCAGACAGAACCTCTGGCTACAAAGCTCCAGAATGGAAGCCCACTGCCTGAGAGAGCTCATCCAGAAGTAAATGGAGACACCAAGTGGCACTCTTTCAAAAGTTATTATGGAATACCCTGTATGAAGGGAAGCCAGAATAGTCGTGTGAGTCCTGACTTTACACAAGAAAGTAGAGGGTATTCCAAGTGTTTGCAAAATGGAGGAATAAAACGCACAGTTAGTGAACCTTCTCTCTCTGGGCTCCTTCAGATCAAGAAATTGAAACAAGACCAAAAGGCTAATGGAGAAAGACGTAACTTCGGGGTAAGCCAAGAAAGAAATCCAGGTGAAAGCAGTCAACCAAATGTCTCCGATTTGAGTGATAAGAAAGAATCTGTGAGTTCTGTAGCCCAAGAAAATGCAGTTAAAGATTTCACCAGTTTTTCAACACATAACTGCAGTGGGCCTGAAAATCCAGAGCTTCAGATTCTGAATGAGCAGGAGGGGAAAAGTGCTAATTACCATGACAAGAACATTGTATTACTTAAAAACAAGGCAGTGCTAATGCCTAATGGTGCTACAGTTTCTGCCTCTTCCGTGGAACACACACATGGTGAACTCCTGGAAAAAACACTGTCTCAATATTATCCAGATTGTGTTTCCATTGCGGTGCAGAAAACCACATCTCACATAAATGCCATTAACAGTCAGGCTACTAATGAGTTGTCCTGTGAGATCACTCACCCATCGCATACCTCAGGGCAGATCAATTCCGCACAGACCTCTAACTCTGAGCTGCCTCCAAAGCCAGCTGCAGTGGTGAGTGAGGCCTGTGATGCTGATGATGCTGATAATGCCAGTAAACTAGCTGCAATGCTAAATACCTGTTCCTTTCAGAAACCAGAACAACTACAACAACAAAAATCAGTTTTTGAGATATGCCCATCTCCTGCAGAAAATAACATCCAGGGAACCACAAAGCTAGCGTCTGGTGAAGAATTCTGTTCAGGTTCCAGCAGCAATTTGCAAGCTCCTGGTGGCAGCTCTGAACGGTATTTAAAACAAAATGAAATGAATGGTGCTTACTTCAAGCAAAGCTCAGTGTTCACTAAGGATTCCTTTTCTGCCACTACCACACCACCACCACCATCACAATTGCTTCTTTCTCCCCCTCCTCCTCTTCCACAGGTTCCTCAGCTTCCTTCAGAAGGAAAAAGCACTCTGAATGGTGGAGTTTTAGAAGAACACCACCACTACCCCAACCAAAGTAACACAACACTTTTAAGGGAAGTGAAAATAGAGGGTAAACCTGAGGCACCACCTTCCCAGAGTCCTAATCCATCTACACATGTATGCAGCCCTTCTCCGATGCTTTCTGAAAGGCCTCAGAATAATTGTGTGAACAGGAATGACATACAGACTGCAGGGACAATGACTGTTCCATTGTGTTCTGAGAAAACAAGACCAATGTCAGAACACCTCAAGCATAACCCACCAATTTTTGGTAGCAGTGGAGAGCTACAGGACAACTGCCAGCAGTTGATGAGAAACAAAGAGCAAGAGATTCTGAAGGGTCGAGACAAGGAGCAAACACGAGATCTTGTGCCCCCAACACAGCACTATCTGAAACCAGGATGGATTGAATTGAAGGCCCCTCGTTTTCACCAAGCGGAATCCCATCTAAAACGTAATGAGGCATCACTGCCATCAATTCTTCAGTATCAACCCAATCTCTCCAATCAAATGACCTCCAAACAATACACTGGAAATTCCAACATGCCTGGGGGGCTCCCAAGGCAAGCTTACACCCAGAAAACAACACAGCTGGAGCACAAGTCACAAATGTACCAAGTTGAAATGAATCAAGGGCAGTCCCAAGGTACAGTGGACCAACATCTCCAGTTCCAAAAACCCTCACACCAGGTGCACTTCTCCAAAACAGACCATTTACCAAAAGCTCATGTGCAGTCACTGTGTGGCACTAGATTTCATTTTCAACAAAGAGCAGATTCCCAAACTGAAAAACTTATGTCCCCAGTGTTGAAACAGCACTTGAATCAACAGGCTTCAGAGACTGAGCCATTTTCAAACTCACACCTTTTGCAACATAAGCCTCATAAACAGGCAGCACAAACACAACCATCCCAGAGTTCACATCTCCCTCAAAACCAGCAACAGCAGCAAAAATTACAAATAAAGAATAAAGAGGAAATACTCCAGACTTTTCCTCACCCCCAAAGCAACAATGATCAGCAAAGAGAAGGATCATTCTTTGGCCAGACTAAAGTGGAAGAATGTTTTCATGGTGAAAATCAGTATTCAAAATCAAGCGAGTTCGAGACTCATAATGTCCAAATGGGACTGGAGGAAGTACAGAATATAAATCGTAGAAATTCCCCTTATAGTCAGACCATGAAATCAAGTGCATGCAAAATACAGGTTTCTTGTTCAAACAATACACACCTAGTTTCAGAGAATAAAGAACAGACTACACATCCTGAACTTTTTGCAGGAAACAAGACCCAAAACTTGCATCACATGCAATATTTTCCAAATAATGTGATCCCAAAGCAAGATCTTCTTCACAGGTGCTTTCAAGAACAGGAGCAGAAGTCACAACAAGCTTCAGTTCTACAGGGATATAAAAATAGAAACCAAGATATGTCTGGTCAACAAGCTGCGCAACTTGCTCAGCAAAGGTACTTGATACATAACCATGCAAATGTTTTTCCTGTGCCTGACCAGGGAGGAAGTCACACTCAGACCCCTCCCCAGAAGGACACTCAAAAGCATGCTGCTCTAAGGTGGCATCTCTTACAGAAGCAAGAACAGCAGCAAACACAGCAACCCCAAACTGAGTCTTGCCATAGTCAGATGCACAGGCCAATTAAGGTGGAACCTGGATGCAAGCCACATGCCTGTATGCACACAGCACCACCAGAAAACAAAACATGGAAAAAGGTAACTAAGCAAGAGAATCCACCTGCAAGCTGTGATAATGTGCAGCAAAAGAGCATCATTGAGACCATGGAGCAGCATCTGAAGCAGTTTCACGCCAAGTCGTTATTTGACCATAAGGCTCTTACTCTCAAATCACAGAAGCAAGTAAAAGTTGAAATGTCAGGGCCAGTCACAGTTTTGACTAGACAAACCACTGCTGCAGAACTTGATAGCCACACCCCAGCTTTAGAGCAGCAAACAACTTCTTCAGAAAAGACACCAACCAAAAGAACAGCTGCTTCTGTTCTCAATAATTTTATAGAGTCACCTTCCAAATTACTAGATACTCCTATAAAAAATTTATTGGATACACCTGTCAAGACTCAATATGATTTCCCATCTTGCAGATGTGTAGGTAAGTGCCAGAAATGTACTGAGACACATGGCGTTTATCCAGAATTAGCAAATTTATCTTCAGATATGGGATTTTCCTTCTTTTTTTAA'\n",
    "\n",
    "    #Split the sequence into codons \n",
    "    TET2_codons_NP_060098 = [TET2_exons_NP_060098[i:i+3] for i in range(0, len(TET2_exons_NP_060098), 3)]\n",
    "\n",
    "    #TET2 transcript 1 extra base at each end:\n",
    "    #Downloaded from http://grch37.ensembl.org/Homo_sapiens/Transcript/Summary?db=core;g=ENSG00000168769;r=4:106067032-106200973;t=ENST00000305737\n",
    "    TET2_exons_extra_bases_NP_060098 = 'GATGGAACAGGATAGAACCAACCATGTTGAGGGCAACAGACTAAGTCCATTCCTGATACCATCACCTCCCATTTGCCAGACAGAACCTCTGGCTACAAAGCTCCAGAATGGAAGCCCACTGCCTGAGAGAGCTCATCCAGAAGTAAATGGAGACACCAAGTGGCACTCTTTCAAAAGTTATTATGGAATACCCTGTATGAAGGGAAGCCAGAATAGTCGTGTGAGTCCTGACTTTACACAAGAAAGTAGAGGGTATTCCAAGTGTTTGCAAAATGGAGGAATAAAACGCACAGTTAGTGAACCTTCTCTCTCTGGGCTCCTTCAGATCAAGAAATTGAAACAAGACCAAAAGGCTAATGGAGAAAGACGTAACTTCGGGGTAAGCCAAGAAAGAAATCCAGGTGAAAGCAGTCAACCAAATGTCTCCGATTTGAGTGATAAGAAAGAATCTGTGAGTTCTGTAGCCCAAGAAAATGCAGTTAAAGATTTCACCAGTTTTTCAACACATAACTGCAGTGGGCCTGAAAATCCAGAGCTTCAGATTCTGAATGAGCAGGAGGGGAAAAGTGCTAATTACCATGACAAGAACATTGTATTACTTAAAAACAAGGCAGTGCTAATGCCTAATGGTGCTACAGTTTCTGCCTCTTCCGTGGAACACACACATGGTGAACTCCTGGAAAAAACACTGTCTCAATATTATCCAGATTGTGTTTCCATTGCGGTGCAGAAAACCACATCTCACATAAATGCCATTAACAGTCAGGCTACTAATGAGTTGTCCTGTGAGATCACTCACCCATCGCATACCTCAGGGCAGATCAATTCCGCACAGACCTCTAACTCTGAGCTGCCTCCAAAGCCAGCTGCAGTGGTGAGTGAGGCCTGTGATGCTGATGATGCTGATAATGCCAGTAAACTAGCTGCAATGCTAAATACCTGTTCCTTTCAGAAACCAGAACAACTACAACAACAAAAATCAGTTTTTGAGATATGCCCATCTCCTGCAGAAAATAACATCCAGGGAACCACAAAGCTAGCGTCTGGTGAAGAATTCTGTTCAGGTTCCAGCAGCAATTTGCAAGCTCCTGGTGGCAGCTCTGAACGGTATTTAAAACAAAATGAAATGAATGGTGCTTACTTCAAGCAAAGCTCAGTGTTCACTAAGGATTCCTTTTCTGCCACTACCACACCACCACCACCATCACAATTGCTTCTTTCTCCCCCTCCTCCTCTTCCACAGGTTCCTCAGCTTCCTTCAGAAGGAAAAAGCACTCTGAATGGTGGAGTTTTAGAAGAACACCACCACTACCCCAACCAAAGTAACACAACACTTTTAAGGGAAGTGAAAATAGAGGGTAAACCTGAGGCACCACCTTCCCAGAGTCCTAATCCATCTACACATGTATGCAGCCCTTCTCCGATGCTTTCTGAAAGGCCTCAGAATAATTGTGTGAACAGGAATGACATACAGACTGCAGGGACAATGACTGTTCCATTGTGTTCTGAGAAAACAAGACCAATGTCAGAACACCTCAAGCATAACCCACCAATTTTTGGTAGCAGTGGAGAGCTACAGGACAACTGCCAGCAGTTGATGAGAAACAAAGAGCAAGAGATTCTGAAGGGTCGAGACAAGGAGCAAACACGAGATCTTGTGCCCCCAACACAGCACTATCTGAAACCAGGATGGATTGAATTGAAGGCCCCTCGTTTTCACCAAGCGGAATCCCATCTAAAACGTAATGAGGCATCACTGCCATCAATTCTTCAGTATCAACCCAATCTCTCCAATCAAATGACCTCCAAACAATACACTGGAAATTCCAACATGCCTGGGGGGCTCCCAAGGCAAGCTTACACCCAGAAAACAACACAGCTGGAGCACAAGTCACAAATGTACCAAGTTGAAATGAATCAAGGGCAGTCCCAAGGTACAGTGGACCAACATCTCCAGTTCCAAAAACCCTCACACCAGGTGCACTTCTCCAAAACAGACCATTTACCAAAAGCTCATGTGCAGTCACTGTGTGGCACTAGATTTCATTTTCAACAAAGAGCAGATTCCCAAACTGAAAAACTTATGTCCCCAGTGTTGAAACAGCACTTGAATCAACAGGCTTCAGAGACTGAGCCATTTTCAAACTCACACCTTTTGCAACATAAGCCTCATAAACAGGCAGCACAAACACAACCATCCCAGAGTTCACATCTCCCTCAAAACCAGCAACAGCAGCAAAAATTACAAATAAAGAATAAAGAGGAAATACTCCAGACTTTTCCTCACCCCCAAAGCAACAATGATCAGCAAAGAGAAGGATCATTCTTTGGCCAGACTAAAGTGGAAGAATGTTTTCATGGTGAAAATCAGTATTCAAAATCAAGCGAGTTCGAGACTCATAATGTCCAAATGGGACTGGAGGAAGTACAGAATATAAATCGTAGAAATTCCCCTTATAGTCAGACCATGAAATCAAGTGCATGCAAAATACAGGTTTCTTGTTCAAACAATACACACCTAGTTTCAGAGAATAAAGAACAGACTACACATCCTGAACTTTTTGCAGGAAACAAGACCCAAAACTTGCATCACATGCAATATTTTCCAAATAATGTGATCCCAAAGCAAGATCTTCTTCACAGGTGCTTTCAAGAACAGGAGCAGAAGTCACAACAAGCTTCAGTTCTACAGGGATATAAAAATAGAAACCAAGATATGTCTGGTCAACAAGCTGCGCAACTTGCTCAGCAAAGGTACTTGATACATAACCATGCAAATGTTTTTCCTGTGCCTGACCAGGGAGGAAGTCACACTCAGACCCCTCCCCAGAAGGACACTCAAAAGCATGCTGCTCTAAGGTGGCATCTCTTACAGAAGCAAGAACAGCAGCAAACACAGCAACCCCAAACTGAGTCTTGCCATAGTCAGATGCACAGGCCAATTAAGGTGGAACCTGGATGCAAGCCACATGCCTGTATGCACACAGCACCACCAGAAAACAAAACATGGAAAAAGGTAACTAAGCAAGAGAATCCACCTGCAAGCTGTGATAATGTGCAGCAAAAGAGCATCATTGAGACCATGGAGCAGCATCTGAAGCAGTTTCACGCCAAGTCGTTATTTGACCATAAGGCTCTTACTCTCAAATCACAGAAGCAAGTAAAAGTTGAAATGTCAGGGCCAGTCACAGTTTTGACTAGACAAACCACTGCTGCAGAACTTGATAGCCACACCCCAGCTTTAGAGCAGCAAACAACTTCTTCAGAAAAGACACCAACCAAAAGAACAGCTGCTTCTGTTCTCAATAATTTTATAGAGTCACCTTCCAAATTACTAGATACTCCTATAAAAAATTTATTGGATACACCTGTCAAGACTCAATATGATTTCCCATCTTGCAGATGTGTAGGTAAGTGCCAGAAATGTACTGAGACACATGGCGTTTATCCAGAATTAGCAAATTTATCTTCAGATATGGGATTTTCCTTCTTTTTTTAAA'\n",
    "\n",
    "    #List of all amino acid positions (across whole coding region) covered (amino acid position minus 1 (as python starts at 0))\n",
    "    TET2_missense_codons_DICT_NP_072046={}\n",
    "    for i in range(lower_range, upper_range):\n",
    "        TET2_missense_codons_DICT_NP_072046[i+1]= (TET2_codons_NP_072046[i])\n",
    "\n",
    "    calls_TET2_missense_NP_072046 = {}\n",
    "    for k,v in TET2_missense_codons_DICT_NP_072046.items(): #iterating over the list of codon numbers (and codon sequences)\n",
    "        if v in Possible_Amino_Acid_Changes.keys(): #assign the list of all the possible amino acid changes to each codon they called\n",
    "            changes = Possible_Amino_Acid_Changes_letters[v]\n",
    "    #         print(Possible_Amino_Acid_Changes.values())\n",
    "            calls_TET2_missense_NP_072046[k]=(v, changes)\n",
    "\n",
    "    #Changing the original amino acid to a Word (e.g. Phe) rather than a codon\n",
    "    calls_TET2_missense_NP_072046_extra = {}\n",
    "\n",
    "    for k, v in calls_TET2_missense_NP_072046.items():\n",
    "        if v[0] in Genetic_Code:\n",
    "            Amino_acid_change = Genetic_Code.get(v[0])\n",
    "            calls_TET2_missense_NP_072046_extra[k] = (Amino_acid_change, v[0], v[1])\n",
    "            #e.g. 12: (Phe, 'K', '*', 'T' etc..)\n",
    "\n",
    "    #Changing the original amino acid to a LETTER (rather than the codon)\n",
    "    calls_TET2_missense_NP_072046_extra2 = {}\n",
    "\n",
    "    for k, v in calls_TET2_missense_NP_072046_extra.items():\n",
    "        if v[0] in Amino_acid_abbreviations:\n",
    "            Amino_acid_abbrev = Amino_acid_abbreviations.get(v[0])\n",
    "            calls_TET2_missense_NP_072046_extra2[k] = (Amino_acid_abbrev, v[1], v[2])\n",
    "            #e.g. 12: ((F), ('K', '*', 'T' etc..))\n",
    "\n",
    "    #Create a dictionary that contains details of the positions of missense variants at that codon\n",
    "    TET2_position_in_list_dict_missense_NP_072046 = {}\n",
    "\n",
    "    for k, v in calls_TET2_missense_NP_072046_extra2.items():\n",
    "        original_AA = v[0]\n",
    "        possible_AA_changes = v[2]\n",
    "        possible_AA_changes_string = ''.join(v[2])\n",
    "        positions_in_list = []\n",
    "        for i in possible_AA_changes:\n",
    "            if i != original_AA: #not synonymous\n",
    "                if i != '*': #not missense\n",
    "                    position_in_list=-1\n",
    "                    missense_change = i\n",
    "                    new_aa_list = []\n",
    "                    while True:\n",
    "                        position_in_list = possible_AA_changes_string.find(missense_change, position_in_list+1)\n",
    "        #               print(position_in_list)\n",
    "                        if position_in_list == -1: break\n",
    "                        positions_in_list.append(position_in_list)\n",
    "                        positions_in_list = set(positions_in_list) #convert into a 'set' to remove duplicate positions\n",
    "                        positions_in_list = list(positions_in_list) #convert back to a list so can be added to dictionary\n",
    "                    for pos in positions_in_list:\n",
    "                        new_aa = possible_AA_changes_string[pos]\n",
    "                        new_aa_list.append(new_aa)\n",
    "                    TET2_position_in_list_dict_missense_NP_072046[k]=(positions_in_list, v, new_aa_list)\n",
    "\n",
    "        aa_positions_variants_called = []\n",
    "        for k, v in TET2_position_in_list_dict_missense_NP_072046.items():\n",
    "            aa_positions_variants_called.append(k)\n",
    "    # print('positions at which missense variants called = ', aa_positions_variants_called)\n",
    "\n",
    "    #Create a dictionary that contains the position in the list of possible AA where the variant is and what it changes to\n",
    "    TET2_bases_missense_NP_072046 = {}\n",
    "    for k, v in TET2_position_in_list_dict_missense_NP_072046.items():\n",
    "        original_a_a = v[1][0]\n",
    "        codon_seq = v[1][1]\n",
    "        base_changes=[]\n",
    "        new_a_a_list = v[2]\n",
    "        if codon_seq in Possible_Base_Changes.keys():\n",
    "            list_of_possible_bases = Possible_Base_Changes.get(codon_seq)\n",
    "        for i in v[0]:\n",
    "            base_change_position = i\n",
    "            base_change = list_of_possible_bases[base_change_position]\n",
    "            base_changes.append(base_change)\n",
    "        TET2_bases_missense_NP_072046[k]=(v, base_changes, original_a_a, new_a_a_list)\n",
    "\n",
    "    #Create dictionary which contains original codon seq and details on which base is changed\n",
    "    TET2_bases_missense2_NP_072046 = {}\n",
    "\n",
    "    for k, v in TET2_bases_missense_NP_072046.items():\n",
    "        base_change_positions = v[0][0]\n",
    "        codon_positions = []\n",
    "        positions_in_gene = []\n",
    "        new_a_a_list = v[3]\n",
    "        for i in base_change_positions:\n",
    "            base_change_position = i\n",
    "            if base_change_position <= 2:\n",
    "                codon_pos = 1\n",
    "            if 2 < base_change_position <= 5:\n",
    "                codon_pos = 2\n",
    "            if 5 < base_change_position <= 8:\n",
    "                codon_pos = 3\n",
    "            codon_positions.append(codon_pos)\n",
    "        TET2_bases_missense2_NP_072046[k] = (v, codon_positions)\n",
    "\n",
    "    #Create dictionary which contains original codon seq, details on which base is changed and it's position in the coding region\n",
    "    TET2_bases_missense3_NP_072046 = {}\n",
    "\n",
    "    for k, v in TET2_bases_missense2_NP_072046.items():\n",
    "        codon_positions = v[1]\n",
    "        base_changes = v[0][1]\n",
    "        original_a_a = v[0][2]\n",
    "        new_aa_list = v[0][0][2]\n",
    "        original_bases = []\n",
    "        positions_in_gene = []\n",
    "        codon_number = k\n",
    "        for i in codon_positions:\n",
    "            codon_pos = i\n",
    "            position_in_gene = ((int(k)-1)*3)+codon_pos #codon number minus 1 (e.g. codon 307 -1 = 306, then times by 3 takes you to beginning of codon 307 (then add the position))  \n",
    "            positions_in_gene.append(position_in_gene)\n",
    "            original_base = (v[0][0][1][1][codon_pos-1])\n",
    "            original_bases.append(original_base)\n",
    "        TET2_bases_missense3_NP_072046[k] = (v[0][0][1], positions_in_gene, original_bases, base_changes, codon_number, original_a_a, new_aa_list)\n",
    "\n",
    "    # Create dictionary which contains original codon seq, details on which base is changed and its neighbouring bases\n",
    "    TET2_bases_missense4_NP_072046 = {}\n",
    "\n",
    "    for k, v in TET2_bases_missense3_NP_072046.items():\n",
    "        position_in_gene = v[1]\n",
    "        sites_trio_missense=[]\n",
    "        new_a_a_list = v[6]\n",
    "\n",
    "        for i in position_in_gene:\n",
    "            gene_pos = i\n",
    "            site_trio_missense = [TET2_exons_extra_bases_NP_072046[gene_pos:gene_pos+3] for gene_pos in range (gene_pos-1, gene_pos, 3)]\n",
    "            sites_trio_missense.append(site_trio_missense)\n",
    "            TET2_bases_missense4_NP_072046[k] = (v, sites_trio_missense)\n",
    "\n",
    "    #Create a dictionary containing pos in coding region, original base, new base, base affected with its neighbours\n",
    "    #e.g. 938, G, A, TGG = base 938 on coding region, original base = G, changes to base A, base with neighbours either side = TGG\n",
    "    TET2_sites_missense5_NP_072046 = {}\n",
    "\n",
    "    a=0\n",
    "    for k, v in TET2_bases_missense4_NP_072046.items():\n",
    "        original_a_a = v[0][0][0]\n",
    "        new_a_a = v[0][6]\n",
    "        codon_n = v[0][1] #base number in codon (out of 3 x 912)\n",
    "        codon_n_912 = v[0][4]\n",
    "        original_base = v[0][2]\n",
    "        new_base = v[0][3]\n",
    "        site_trio = v[1]\n",
    "        info=[]\n",
    "\n",
    "        change = 0\n",
    "        for i in codon_n:\n",
    "            codon_number = i\n",
    "            orig_base = original_base[change]\n",
    "            new_amino_acid = new_a_a[change]\n",
    "            n_base = new_base[change]\n",
    "            site_tri = site_trio[change]\n",
    "            info.append([codon_number, orig_base, n_base, site_tri, codon_n_912, original_a_a, new_amino_acid])\n",
    "            change = change+1\n",
    "        TET2_sites_missense5_NP_072046[a]=info\n",
    "        a=a+1\n",
    "    #     print(TET2_sites_missense5_NP_072046)\n",
    "\n",
    "    # Write in a way that can iterate over mutation rates (i.e. put in to form C[C>A]G etc)   \n",
    "    TET2_sites_missense6_NP_072046={}\n",
    "\n",
    "    separating_positions = {} # separate out all the possible changes to just 1 effective item in the list\n",
    "    a=0\n",
    "    for k, v in TET2_sites_missense5_NP_072046.items():\n",
    "        for i in v: #because some contain more than 1 change from that codon:\n",
    "            separating_positions[a]=i\n",
    "            a=a+1\n",
    "\n",
    "    for k, v in separating_positions.items():\n",
    "        base_pos = v[0]\n",
    "        original_base = v[1]\n",
    "        new_base = v[2]\n",
    "        site_trio = v[3]\n",
    "        codon_n_912 = v[4]\n",
    "        original_a_a = v[5]\n",
    "        new_a_a = v[6]\n",
    "        sequence = (str(original_base),str(new_base))\n",
    "        changes = '>'.join(sequence)\n",
    "        left_base = site_trio[0][0]\n",
    "        right_base = site_trio[0][2]\n",
    "        sequence2 = (changes, right_base)\n",
    "        changes2 = ']'.join(sequence2)\n",
    "        sequence3 = (left_base, changes2)\n",
    "        changes3 = '['.join(sequence3)\n",
    "        TET2_sites_missense6_NP_072046[k] = (v[0],changes3, codon_n_912, original_a_a, new_a_a) #site_trio = the affected base and its neighbours\n",
    "\n",
    "    #Assign the mutation rates to each site...\n",
    "    TET2_mutation_rates_missense_NP_072046 = {}\n",
    "    for k, v in TET2_sites_missense6_NP_072046.items():\n",
    "        base_change_info = v[1]\n",
    "        if base_change_info in Mutation_rates.keys():\n",
    "            mutation_rate = Mutation_rates.get(base_change_info)\n",
    "            original_a_a = v[3]\n",
    "            new_a_a = v[4]\n",
    "            aa_positioning = v[2]\n",
    "            variant_id = str(original_a_a)+str(aa_positioning)+str(new_a_a)\n",
    "            TET2_mutation_rates_missense_NP_072046[k] = (v[0],v[1], mutation_rate, variant_id)\n",
    "\n",
    "    TET2_mutation_rate_missense_NP_072046_sum = []\n",
    "\n",
    "    mutation_rates_list = []\n",
    "    for k, v in TET2_mutation_rates_missense_NP_072046.items():\n",
    "        codon_position_in_gene=(v[0])\n",
    "        variant = v[3]\n",
    "        site = v[1]\n",
    "        mutation_rate_value = v[2] \n",
    "        mutation_rates_list.append((v[0], variant, site, mutation_rate_value))\n",
    "        TET2_mutation_rate_missense_NP_072046_sum.append(mutation_rate_value)\n",
    "\n",
    "    #Get the average of all the mutation rates\n",
    "    TET2_mutation_rate_missense_NP_072046_sum = []\n",
    "    for k, v in TET2_mutation_rates_missense_NP_072046.items():\n",
    "        mut_rate = v[2]\n",
    "        TET2_mutation_rate_missense_NP_072046_sum.append(mut_rate)\n",
    "\n",
    "    TET2_Summed_mutation_rate_missense_NP_072046_added = np.sum(TET2_mutation_rate_missense_NP_072046_sum)\n",
    "\n",
    "    print('Total summed mutation rate across TET2 missense_NP_072046 for this study = ', TET2_Summed_mutation_rate_missense_NP_072046_added)\n",
    "\n",
    "    mutation_rates_list_all = mutation_rates_list\n",
    "    return mutation_rates_list_all # e.g. lower= 0, upper = 913 for whole gene\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### ASXL1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 304,
   "metadata": {
    "code_folding": [
     0
    ]
   },
   "outputs": [],
   "source": [
    "def mutation_rates_nonsense_range_of_variants_ASXL1(lower_range, upper_range):\n",
    "    \n",
    "    #The mutation rate at a base is calculated based on its trinucleotide context.  To obtain the trinucleotide context, the base either side of the mutated base needs to be ascertained.\n",
    "    #The coding sequence (+ an extra base at each end) of each gene is used when determining the trinucleotide context of the base change.  The problem with using the coding sequence is that\n",
    "    #if a base change falls at the end of an exon, its neighbouring base will actuall be an intronic base, which may be different to the 'neighbouring' base in the coding sequence.\n",
    "    #To account for this, there is an \"intron_exon_check\" function that is called at the intron/exon boundary if the neighbouring base in the intron is different to the neighbouring base in the coding sequence.\n",
    "    \n",
    "    intron_exon_boundary_caution_base = [140, 143, 252, 373, 1086, 1085]\n",
    "   \n",
    "    def intron_exon_check(position_in_gene_number):\n",
    "        if position_in_gene_number == 140:\n",
    "            gene_pos = 140\n",
    "            first_base = ASXL1_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = ASXL1_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 143:\n",
    "            gene_pos = 143\n",
    "            first_base = ASXL1_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = ASXL1_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 252:\n",
    "            gene_pos = 252\n",
    "            first_base = ASXL1_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = ASXL1_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 373:\n",
    "            gene_pos = 373\n",
    "            first_base = ASXL1_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = ASXL1_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 1086:\n",
    "            gene_pos = 1086\n",
    "            first_base = 'G'\n",
    "            middle_base = ASXL1_exons_extra_bases[gene_pos]\n",
    "            third_base = ASXL1_exons_extra_bases[gene_pos+1]\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 1085:\n",
    "            gene_pos = 1085\n",
    "            first_base = ASXL1_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = ASXL1_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        return triplet\n",
    "    \n",
    "    #ASXL1 transcript:\n",
    "    #Downloaded from http://grch37.ensembl.org\n",
    "    ASXL1_exons = 'ATGAAGGACAAACAGAAGAAGAAGAAGGAGCGCACGTGGGCCGAGGCCGCGCGCCTGGTATTAGAAAACTACTCGGATGCTCCAATGACACCAAAACAGATTCTGCAGGTCATAGAGGCAGAAGGACTAAAGGAAATGAGAAGTGGGACTTCCCCTCTCGCATGCCTCAATGCTATGCTACATTCCAATTCAAGAGGAGGAGAGGGGTTGTTTTATAAACTGCCTGGCCGAATCAGCCTTTTCACGCTCAAGAAGGATGCCCTGCAGTGGTCTCGCCATCCAGCTACAGTGGAGGGAGAGGAGCCAGAGGACACGGCTGATGTGGAGAGCTGTGGGTCTAATGAAGCCAGCACTGTGAGTGGTGAAAACGATGTATCTCTTGATGAAACATCTTCGAACGCATCCTGTTCTACAGAATCTCAGAGTCGACCTCTTTCCAATCCCAGGGACAGCTACAGAGCTTCCTCACAGGCGAACAAACAAAAGAAAAAGACTGGGGTGATGCTGCCTCGAGTTGTCCTGACTCCTCTGAAGGTAAACGGGGCCCACGTGGAATCTGCATCAGGGTTCTCGGGCTGCCACGCCGATGGCGAGAGCGGCAGCCCGTCCAGCAGCAGCAGCGGCTCTCTGGCCCTGGGCAGCGCTGCTATTCGTGGCCAGGCCGAGGTCACCCAGGACCCTGCCCCGCTCCTGAGAGGCTTCCGGAAGCCAGCCACAGGTCAAATGAAGCGCAACAGAGGGGAAGAAATAGATTTTGAGACACCTGGGTCCATTCTTGTCAACACCAACCTCCGTGCCCTGATCAACTCTCGGACCTTCCATGCCTTACCATCACACTTCCAGCAGCAGCTCCTCTTCCTCCTGCCTGAAGTAGACAGACAGGTGGGGACGGATGGCCTGTTGCGTCTCAGCAGCAGTGCACTAAATAACGAGTTTTTTACCCATGCGGCTCAGAGCTGGCGGGAGCGCCTGGCTGATGGTGAATTTACTCATGAGATGCAAGTCAGGATACGACAGGAAATGGAGAAGGAAAAGAAGGTGGAACAATGGAAAGAAAAGTTCTTTGAAGACTACTATGGACAGAAGCTGGGTTTGACCAAAGAAGAGTCATTGCAGCAGAACGTGGGCCAGGAGGAGGCTGAAATCAAAAGTGGCTTGTGTGTCCCAGGAGAATCAGTGCGTATACAGCGTGGTCCAGCCACCCGACAGCGAGATGGGCATTTTAAGAAACGCTCTCGGCCAGATCTCCGAACCAGAGCCAGAAGGAATCTGTACAAAAAACAGGAGTCAGAACAAGCAGGGGTTGCTAAGGATGCAAAATCTGTGGCCTCAGATGTTCCCCTCTACAAGGATGGGGAGGCTAAGACTGACCCAGCAGGGCTGAGCAGTCCCCATCTGCCAGGCACATCCTCTGCAGCACCCGACCTGGAGGGTCCCGAATTCCCAGTTGAGTCTGTGGCTTCTCGGATCCAGGCTGAGCCAGACAACTTGGCACGTGCCTCTGCATCTCCAGACAGAATTCCTAGCCTGCCTCAGGAAACTGTGGATCAGGAACCCAAGGATCAGAAGAGGAAATCCTTTGAGCAGGCGGCCTCTGCATCCTTTCCCGAAAAGAAGCCCCGGCTTGAAGATCGTCAGTCCTTTCGTAACACAATTGAAAGTGTTCACACCGAAAAGCCACAGCCCACTAAAGAGGAGCCCAAAGTCCCGCCCATCCGGATTCAACTTTCACGTATCAAACCACCCTGGGTGGTTAAAGGTCAGCCCACTTACCAGATATGCCCCCGGATCATCCCCACCACGGAGTCCTCCTGCCGGGGTTGGACTGGCGCCAGGACCCTCGCAGACATTAAAGCCCGTGCTCTGCAGGTCCGAGGGGCGAGAGGTCACCACTGCCATAGAGAGGCGGCCACCACTGCCATCGGAGGGGGGGGTGGCCCGGGTGGAGGTGGCGGCGGGGCCACCGATGAGGGAGGTGGCAGAGGCAGCAGCAGTGGTGATGGTGGTGAGGCCTGTGGCCACCCTGAGCCCAGGGGAGGCCCGAGCACCCCTGGAAAGTGTACGTCAGATCTACAGCGAACACAACTACTGCCGCCTTATCCTCTAAATGGGGAGCATACCCAGGCCGGAACTGCCATGTCCAGAGCTAGGAGAGAGGACCTGCCTTCTCTGAGAAAGGAGGAAAGCTGCCTACTACAGAGGGCTACAGTTGGACTCACAGATGGGCTAGGAGATGCCTCCCAACTCCCCGTTGCTCCCACTGGGGACCAGCCATGCCAGGCCTTGCCCCTACTGTCCTCCCAAACCTCAGTAGCTGAGAGATTAGTGGAGCAGCCTCAGTTGCATCCGGATGTTAGAACTGAATGTGAGTCTGGCACCACTTCCTGGGAAAGTGATGATGAGGAGCAAGGACCCACCGTTCCTGCAGACAATGGTCCCATTCTGTCTCTAGTGGGAGATGATACATTAGAGAAAGGAACTGGCCAAGCTCTTGACAGTCATCCCACTATGAAGGATCCTGTAAATGTGACCCCCAGTTCCACACCTGAATCCTCACCGACTGATTGCCTGCAGAACAGAGCATTTGATGACGAATTAGGGCTTGGTGGCTCATGCCCTCCTATGAGGGAAAGTGATACTAGACAAGAAAACTTGAAAACCAAGGCTCTCGTTTCTAACAGTTCTTTGCATTGGATACCCATCCCATCGAATGATGAGGTAGTGAAACAGCCCAAACCAGAATCCAGAGAACACATACCATCTGTTGAGCCCCAGGTTGGAGAGGAGTGGGAGAAAGCTGCTCCCACCCCTCCTGCATTGCCTGGGGATTTGACAGCTGAGGAGGGTCTAGATCCTCTTGACAGCCTTACTTCACTCTGGACTGTGCCATCTCGAGGAGGCAGTGACAGCAATGGCAGTTACTGTCAACAGGTGGACATTGAAAAGCTGAAAATCAACGGAGACTCTGAAGCACTGAGTCCTCACGGTGAGTCCACGGATACAGCCTCTGACTTTGAAGGTCACCTCACGGAGGACAGCAGTGAGGCTGACACTAGAGAAGCTGCAGTGACAAAGGGATCTTCGGTGGACAAGGATGAGAAACCCAATTGGAACCAATCTGCCCCACTGTCCAAGGTGAATGGTGACATGCGTCTGGTTACAAGGACAGATGGGATGGTTGCTCCTCAGAGCTGGGTGTCTCGAGTATGTGCGGTCCGCCAAAAGATCCCAGATTCCCTACTGCTGGCCAGTACTGAGTACCAGCCAAGAGCCGTGTGCCTGTCCATGCCTGGGTCCTCAGTGGAGGCCACTAACCCACTTGTGATGCAGTTGCTGCAGGGTAGCTTGCCCCTAGAGAAGGTTCTTCCACCAGCCCACGATGACAGCATGTCAGAATCCCCACAAGTACCACTTACAAAAGACCAGAGCCATGGCTCGCTACGCATGGGATCTTTACATGGTCTTGGAAAAAACAGTGGCATGGTTGATGGAAGCAGCCCCAGTTCTTTAAGGGCTTTGAAGGAGCCTCTTCTGCCAGATAGCTGTGAAACAGGCACTGGTCTTGCCAGGATTGAGGCCACCCAGGCTCCTGGAGCACCCCAAAAGAATTGCAAGGCAGTCCCAAGTTTTGACTCCCTCCATCCAGTGACAAATCCCATTACATCCTCTAGGAAACTGGAAGAAATGGATTCCAAAGAGCAGTTCTCTTCCTTTAGTTGTGAAGATCAGAAGGAAGTCCGTGCTATGTCACAGGACAGTAATTCAAATGCTGCTCCAGGAAAGAGCCCAGGAGATCTTACTACCTCGAGAACACCTCGTTTCTCATCTCCAAATGTGATCTCCTTTGGTCCAGAGCAGACAGGTCGGGCCCTGGGTGATCAGAGCAATGTTACAGGCCAAGGGAAGAAGCTTTTTGGCTCTGGGAATGTGGCTGCAACCCTTCAGCGCCCCAGGCCTGCGGACCCGATGCCTCTTCCTGCTGAGATCCCTCCAGTTTTTCCCAGTGGGAAGTTGGGACCAAGCACAAACTCCATGTCTGGTGGGGTACAGACTCCAAGGGAAGACTGGGCTCCAAAGCCACATGCCTTTGTTGGCAGCGTCAAGAATGAGAAGACTTTTGTGGGGGGTCCTCTTAAGGCAAATGCCGAGAACAGGAAAGCTACTGGGCATAGTCCCCTGGAACTGGTGGGTCACTTGGAAGGGATGCCCTTTGTCATGGACTTGCCCTTCTGGAAATTACCCCGAGAGCCAGGGAAGGGGCTCAGTGAGCCTCTGGAGCCTTCTTCTCTCCCCTCCCAACTCAGCATCAAGCAGGCATTTTATGGGAAGCTTTCTAAACTCCAACTGAGTTCCACCAGCTTTAATTATTCCTCTAGCTCTCCCACCTTTCCCAAAGGCCTTGCTGGAAGTGTGGTGCAGCTGAGCCACAAAGCAAACTTTGGTGCGAGCCACAGTGCATCACTTTCCTTGCAAATGTTCACTGACAGCAGCACGGTGGAAAGCATCTCGCTCCAGTGTGCGTGCAGCCTGAAAGCCATGATCATGTGCCAAGGCTGCGGTGCGTTCTGTCACGATGACTGTATTGGACCCTCAAAGCTCTGTGTATTGTGCCTTGTGGTGAGATAA'\n",
    "    \n",
    "    #Split the sequence into codons \n",
    "    ASXL1_codons = [ASXL1_exons[i:i+3] for i in range(0, len(ASXL1_exons), 3)]\n",
    "\n",
    "    #ASXL1 transcript:\n",
    "    #Downloaded from http://grch37.ensembl.org\n",
    "    ASXL1_exons_extra_bases = 'GATGAAGGACAAACAGAAGAAGAAGAAGGAGCGCACGTGGGCCGAGGCCGCGCGCCTGGTATTAGAAAACTACTCGGATGCTCCAATGACACCAAAACAGATTCTGCAGGTCATAGAGGCAGAAGGACTAAAGGAAATGAGAAGTGGGACTTCCCCTCTCGCATGCCTCAATGCTATGCTACATTCCAATTCAAGAGGAGGAGAGGGGTTGTTTTATAAACTGCCTGGCCGAATCAGCCTTTTCACGCTCAAGAAGGATGCCCTGCAGTGGTCTCGCCATCCAGCTACAGTGGAGGGAGAGGAGCCAGAGGACACGGCTGATGTGGAGAGCTGTGGGTCTAATGAAGCCAGCACTGTGAGTGGTGAAAACGATGTATCTCTTGATGAAACATCTTCGAACGCATCCTGTTCTACAGAATCTCAGAGTCGACCTCTTTCCAATCCCAGGGACAGCTACAGAGCTTCCTCACAGGCGAACAAACAAAAGAAAAAGACTGGGGTGATGCTGCCTCGAGTTGTCCTGACTCCTCTGAAGGTAAACGGGGCCCACGTGGAATCTGCATCAGGGTTCTCGGGCTGCCACGCCGATGGCGAGAGCGGCAGCCCGTCCAGCAGCAGCAGCGGCTCTCTGGCCCTGGGCAGCGCTGCTATTCGTGGCCAGGCCGAGGTCACCCAGGACCCTGCCCCGCTCCTGAGAGGCTTCCGGAAGCCAGCCACAGGTCAAATGAAGCGCAACAGAGGGGAAGAAATAGATTTTGAGACACCTGGGTCCATTCTTGTCAACACCAACCTCCGTGCCCTGATCAACTCTCGGACCTTCCATGCCTTACCATCACACTTCCAGCAGCAGCTCCTCTTCCTCCTGCCTGAAGTAGACAGACAGGTGGGGACGGATGGCCTGTTGCGTCTCAGCAGCAGTGCACTAAATAACGAGTTTTTTACCCATGCGGCTCAGAGCTGGCGGGAGCGCCTGGCTGATGGTGAATTTACTCATGAGATGCAAGTCAGGATACGACAGGAAATGGAGAAGGAAAAGAAGGTGGAACAATGGAAAGAAAAGTTCTTTGAAGACTACTATGGACAGAAGCTGGGTTTGACCAAAGAAGAGTCATTGCAGCAGAACGTGGGCCAGGAGGAGGCTGAAATCAAAAGTGGCTTGTGTGTCCCAGGAGAATCAGTGCGTATACAGCGTGGTCCAGCCACCCGACAGCGAGATGGGCATTTTAAGAAACGCTCTCGGCCAGATCTCCGAACCAGAGCCAGAAGGAATCTGTACAAAAAACAGGAGTCAGAACAAGCAGGGGTTGCTAAGGATGCAAAATCTGTGGCCTCAGATGTTCCCCTCTACAAGGATGGGGAGGCTAAGACTGACCCAGCAGGGCTGAGCAGTCCCCATCTGCCAGGCACATCCTCTGCAGCACCCGACCTGGAGGGTCCCGAATTCCCAGTTGAGTCTGTGGCTTCTCGGATCCAGGCTGAGCCAGACAACTTGGCACGTGCCTCTGCATCTCCAGACAGAATTCCTAGCCTGCCTCAGGAAACTGTGGATCAGGAACCCAAGGATCAGAAGAGGAAATCCTTTGAGCAGGCGGCCTCTGCATCCTTTCCCGAAAAGAAGCCCCGGCTTGAAGATCGTCAGTCCTTTCGTAACACAATTGAAAGTGTTCACACCGAAAAGCCACAGCCCACTAAAGAGGAGCCCAAAGTCCCGCCCATCCGGATTCAACTTTCACGTATCAAACCACCCTGGGTGGTTAAAGGTCAGCCCACTTACCAGATATGCCCCCGGATCATCCCCACCACGGAGTCCTCCTGCCGGGGTTGGACTGGCGCCAGGACCCTCGCAGACATTAAAGCCCGTGCTCTGCAGGTCCGAGGGGCGAGAGGTCACCACTGCCATAGAGAGGCGGCCACCACTGCCATCGGAGGGGGGGGTGGCCCGGGTGGAGGTGGCGGCGGGGCCACCGATGAGGGAGGTGGCAGAGGCAGCAGCAGTGGTGATGGTGGTGAGGCCTGTGGCCACCCTGAGCCCAGGGGAGGCCCGAGCACCCCTGGAAAGTGTACGTCAGATCTACAGCGAACACAACTACTGCCGCCTTATCCTCTAAATGGGGAGCATACCCAGGCCGGAACTGCCATGTCCAGAGCTAGGAGAGAGGACCTGCCTTCTCTGAGAAAGGAGGAAAGCTGCCTACTACAGAGGGCTACAGTTGGACTCACAGATGGGCTAGGAGATGCCTCCCAACTCCCCGTTGCTCCCACTGGGGACCAGCCATGCCAGGCCTTGCCCCTACTGTCCTCCCAAACCTCAGTAGCTGAGAGATTAGTGGAGCAGCCTCAGTTGCATCCGGATGTTAGAACTGAATGTGAGTCTGGCACCACTTCCTGGGAAAGTGATGATGAGGAGCAAGGACCCACCGTTCCTGCAGACAATGGTCCCATTCTGTCTCTAGTGGGAGATGATACATTAGAGAAAGGAACTGGCCAAGCTCTTGACAGTCATCCCACTATGAAGGATCCTGTAAATGTGACCCCCAGTTCCACACCTGAATCCTCACCGACTGATTGCCTGCAGAACAGAGCATTTGATGACGAATTAGGGCTTGGTGGCTCATGCCCTCCTATGAGGGAAAGTGATACTAGACAAGAAAACTTGAAAACCAAGGCTCTCGTTTCTAACAGTTCTTTGCATTGGATACCCATCCCATCGAATGATGAGGTAGTGAAACAGCCCAAACCAGAATCCAGAGAACACATACCATCTGTTGAGCCCCAGGTTGGAGAGGAGTGGGAGAAAGCTGCTCCCACCCCTCCTGCATTGCCTGGGGATTTGACAGCTGAGGAGGGTCTAGATCCTCTTGACAGCCTTACTTCACTCTGGACTGTGCCATCTCGAGGAGGCAGTGACAGCAATGGCAGTTACTGTCAACAGGTGGACATTGAAAAGCTGAAAATCAACGGAGACTCTGAAGCACTGAGTCCTCACGGTGAGTCCACGGATACAGCCTCTGACTTTGAAGGTCACCTCACGGAGGACAGCAGTGAGGCTGACACTAGAGAAGCTGCAGTGACAAAGGGATCTTCGGTGGACAAGGATGAGAAACCCAATTGGAACCAATCTGCCCCACTGTCCAAGGTGAATGGTGACATGCGTCTGGTTACAAGGACAGATGGGATGGTTGCTCCTCAGAGCTGGGTGTCTCGAGTATGTGCGGTCCGCCAAAAGATCCCAGATTCCCTACTGCTGGCCAGTACTGAGTACCAGCCAAGAGCCGTGTGCCTGTCCATGCCTGGGTCCTCAGTGGAGGCCACTAACCCACTTGTGATGCAGTTGCTGCAGGGTAGCTTGCCCCTAGAGAAGGTTCTTCCACCAGCCCACGATGACAGCATGTCAGAATCCCCACAAGTACCACTTACAAAAGACCAGAGCCATGGCTCGCTACGCATGGGATCTTTACATGGTCTTGGAAAAAACAGTGGCATGGTTGATGGAAGCAGCCCCAGTTCTTTAAGGGCTTTGAAGGAGCCTCTTCTGCCAGATAGCTGTGAAACAGGCACTGGTCTTGCCAGGATTGAGGCCACCCAGGCTCCTGGAGCACCCCAAAAGAATTGCAAGGCAGTCCCAAGTTTTGACTCCCTCCATCCAGTGACAAATCCCATTACATCCTCTAGGAAACTGGAAGAAATGGATTCCAAAGAGCAGTTCTCTTCCTTTAGTTGTGAAGATCAGAAGGAAGTCCGTGCTATGTCACAGGACAGTAATTCAAATGCTGCTCCAGGAAAGAGCCCAGGAGATCTTACTACCTCGAGAACACCTCGTTTCTCATCTCCAAATGTGATCTCCTTTGGTCCAGAGCAGACAGGTCGGGCCCTGGGTGATCAGAGCAATGTTACAGGCCAAGGGAAGAAGCTTTTTGGCTCTGGGAATGTGGCTGCAACCCTTCAGCGCCCCAGGCCTGCGGACCCGATGCCTCTTCCTGCTGAGATCCCTCCAGTTTTTCCCAGTGGGAAGTTGGGACCAAGCACAAACTCCATGTCTGGTGGGGTACAGACTCCAAGGGAAGACTGGGCTCCAAAGCCACATGCCTTTGTTGGCAGCGTCAAGAATGAGAAGACTTTTGTGGGGGGTCCTCTTAAGGCAAATGCCGAGAACAGGAAAGCTACTGGGCATAGTCCCCTGGAACTGGTGGGTCACTTGGAAGGGATGCCCTTTGTCATGGACTTGCCCTTCTGGAAATTACCCCGAGAGCCAGGGAAGGGGCTCAGTGAGCCTCTGGAGCCTTCTTCTCTCCCCTCCCAACTCAGCATCAAGCAGGCATTTTATGGGAAGCTTTCTAAACTCCAACTGAGTTCCACCAGCTTTAATTATTCCTCTAGCTCTCCCACCTTTCCCAAAGGCCTTGCTGGAAGTGTGGTGCAGCTGAGCCACAAAGCAAACTTTGGTGCGAGCCACAGTGCATCACTTTCCTTGCAAATGTTCACTGACAGCAGCACGGTGGAAAGCATCTCGCTCCAGTGTGCGTGCAGCCTGAAAGCCATGATCATGTGCCAAGGCTGCGGTGCGTTCTGTCACGATGACTGTATTGGACCCTCAAAGCTCTGTGTATTGTGCCTTGTGGTGAGATAAT'\n",
    "  \n",
    "    #List of all amino acid positions (across whole coding region) covered by (amino acid position minus 1 (as python starts at 0))\n",
    "    ASXL1_nonsense_codons_DICT={}\n",
    "    for i in range(lower_range, upper_range):\n",
    "        ASXL1_nonsense_codons_DICT[i+1]= (ASXL1_codons[i])\n",
    "\n",
    "    calls_ASXL1_nonsense = {}\n",
    "    for k,v in ASXL1_nonsense_codons_DICT.items(): #iterating over the list of codon numbers (and codon sequences)\n",
    "        if v in Possible_Amino_Acid_Changes.keys(): #assign the list of all the possible amino acid changes to each codon they called\n",
    "            changes = Possible_Amino_Acid_Changes_letters[v]\n",
    "    #         print(Possible_Amino_Acid_Changes.values())\n",
    "            calls_ASXL1_nonsense[k]=(v, changes)\n",
    "\n",
    "    #Changing the original amino acid to a Word (e.g. Phe) rather than a codon\n",
    "    calls_ASXL1_nonsense_extra = {}\n",
    "\n",
    "    for k, v in calls_ASXL1_nonsense.items():\n",
    "        if v[0] in Genetic_Code:\n",
    "            Amino_acid_change = Genetic_Code.get(v[0])\n",
    "            calls_ASXL1_nonsense_extra[k] = (Amino_acid_change, v[0], v[1])\n",
    "            #e.g. 12: (Phe, 'K', '*', 'T' etc..)\n",
    "\n",
    "    #Changing the original amino acid to a LETTER (rather than the codon)\n",
    "    calls_ASXL1_nonsense_extra2 = {}\n",
    "\n",
    "    for k, v in calls_ASXL1_nonsense_extra.items():\n",
    "        if v[0] in Amino_acid_abbreviations:\n",
    "            Amino_acid_abbrev = Amino_acid_abbreviations.get(v[0])\n",
    "            calls_ASXL1_nonsense_extra2[k] = (Amino_acid_abbrev, v[1], v[2])\n",
    "            #e.g. 12: ((F), ('K', '*', 'T' etc..))\n",
    "\n",
    "    #Create a dictionary that contains details of the positions of nonsense variants at that codon\n",
    "    ASXL1_position_in_list_dict_nonsense = {}\n",
    "\n",
    "    for k, v in calls_ASXL1_nonsense_extra2.items():\n",
    "        original_AA = v[0]\n",
    "        possible_AA_changes = v[2]\n",
    "        possible_AA_changes_string = ''.join(v[2])\n",
    "        positions_in_list = []\n",
    "        for i in possible_AA_changes:\n",
    "            if i == '*': #nonsense\n",
    "                position_in_list=-1\n",
    "                nonsense_change = i    \n",
    "                while True:\n",
    "                    position_in_list = possible_AA_changes_string.find(nonsense_change, position_in_list+1)\n",
    "    #               print(position_in_list)\n",
    "                    if position_in_list == -1: break\n",
    "                    positions_in_list.append(position_in_list)\n",
    "                    positions_in_list = set(positions_in_list) #convert into a 'set' to remove duplicate positions\n",
    "                    positions_in_list = list(positions_in_list) #convert back to a list so can be added to dictionary\n",
    "                ASXL1_position_in_list_dict_nonsense[k]=(positions_in_list, v)\n",
    "\n",
    "        aa_positions_variants_called = []\n",
    "        for k, v in ASXL1_position_in_list_dict_nonsense.items():\n",
    "            aa_positions_variants_called.append(k)\n",
    "    # print('positions at which nonsense variants called = ', aa_positions_variants_called)\n",
    "\n",
    "    #Create a dictionary that contains the position in the list of possible AA where the variant is and what it changes to\n",
    "    ASXL1_bases_nonsense = {}\n",
    "    for k, v in ASXL1_position_in_list_dict_nonsense.items():\n",
    "        original_a_a = v[1][0]\n",
    "        codon_seq = v[1][1]\n",
    "        base_changes=[]\n",
    "        if codon_seq in Possible_Base_Changes.keys():\n",
    "            list_of_possible_bases = Possible_Base_Changes.get(codon_seq)\n",
    "        for i in v[0]:\n",
    "            base_change_position = i\n",
    "            base_change = list_of_possible_bases[base_change_position]\n",
    "            base_changes.append(base_change)\n",
    "        ASXL1_bases_nonsense[k]=(v, base_changes, original_a_a)\n",
    "\n",
    "    #Create dictionary which contains original codon seq and details on which base is changed\n",
    "    ASXL1_bases_nonsense2 = {}\n",
    "\n",
    "    for k, v in ASXL1_bases_nonsense.items():\n",
    "        base_change_positions = v[0][0]\n",
    "        codon_positions = []\n",
    "        positions_in_gene = []\n",
    "        for i in base_change_positions:\n",
    "            base_change_position = i\n",
    "            if base_change_position <= 2:\n",
    "                codon_pos = 1\n",
    "            if 2 < base_change_position <= 5:\n",
    "                codon_pos = 2\n",
    "            if 5 < base_change_position <= 8:\n",
    "                codon_pos = 3\n",
    "            codon_positions.append(codon_pos)\n",
    "        ASXL1_bases_nonsense2[k] = (v, codon_positions)\n",
    "\n",
    "    #Create dictionary which contains original codon seq, details on which base is changed and it's position in the coding region\n",
    "    ASXL1_bases_nonsense3 = {}\n",
    "\n",
    "    for k, v in ASXL1_bases_nonsense2.items():\n",
    "        codon_positions = v[1]\n",
    "        base_changes = v[0][1]\n",
    "        original_a_a = v[0][2]\n",
    "        original_bases = []\n",
    "        positions_in_gene = []\n",
    "        codon_number = k\n",
    "        for i in codon_positions:\n",
    "            codon_pos = i\n",
    "            position_in_gene = ((int(k)-1)*3)+codon_pos #codon number minus 1 (e.g. codon 307 -1 = 306, then times by 3 takes you to beginning of codon 307 (then add the position))  \n",
    "            positions_in_gene.append(position_in_gene)\n",
    "            original_base = (v[0][0][1][1][codon_pos-1])\n",
    "            original_bases.append(original_base)\n",
    "        ASXL1_bases_nonsense3[k] = (v[0][0][1], positions_in_gene, original_bases, base_changes, codon_number, original_a_a)\n",
    "\n",
    "    #Create dictionary which contains original codon seq, details on which base is changed and its neighbouring bases\n",
    "    ASXL1_bases_nonsense4 = {}\n",
    "\n",
    "    for k, v in ASXL1_bases_nonsense3.items():\n",
    "        position_in_gene = v[1]\n",
    "        sites_trio_nonsense=[]\n",
    "\n",
    "        for i in position_in_gene:\n",
    "    #         print(i)\n",
    "            if i in intron_exon_boundary_caution_base:\n",
    "#                 print('positions at intron_exon_boundary =',i)\n",
    "                adapted_trio = intron_exon_check(i)\n",
    "                sites_trio_nonsense.append([adapted_trio])\n",
    "                ASXL1_bases_nonsense4[k] = (v, sites_trio_nonsense)\n",
    "\n",
    "            else:\n",
    "                for i in position_in_gene:\n",
    "                    gene_pos = i\n",
    "                    site_trio_nonsense = [ASXL1_exons_extra_bases[gene_pos:gene_pos+3] for gene_pos in range (gene_pos-1, gene_pos, 3)]\n",
    "                    sites_trio_nonsense.append(site_trio_nonsense)\n",
    "                    ASXL1_bases_nonsense4[k] = (v, sites_trio_nonsense)\n",
    "\n",
    "    #Create a dictionary containing pos in coding region, original base, new base, base affected with its neighbours\n",
    "    #e.g. 938, G, A, TGG = base 938 on coding region, original base = G, changes to base A, base with neighbours either side = TGG\n",
    "    ASXL1_sites_nonsense5 = {}\n",
    "\n",
    "    a=0\n",
    "    for k, v in ASXL1_bases_nonsense4.items():\n",
    "        original_a_a = v[0][0][0]\n",
    "        codon_n = v[0][1] #base number in codon (out of 3 x 912)\n",
    "        codon_n_912 = v[0][4]\n",
    "        original_base = v[0][2]\n",
    "        new_base = v[0][3]\n",
    "        site_trio = v[1]\n",
    "        info=[]\n",
    "\n",
    "        change = 0\n",
    "        for i in codon_n:\n",
    "            codon_number = i\n",
    "            orig_base = original_base[change]\n",
    "            n_base = new_base[change]\n",
    "            site_tri = site_trio[change]\n",
    "            info.append([codon_number, orig_base, n_base, site_tri, codon_n_912, original_a_a])\n",
    "            change = change+1\n",
    "        ASXL1_sites_nonsense5[a]=info\n",
    "        a=a+1\n",
    "    #     print(ASXL1_sites_nonsense5)\n",
    "\n",
    "    # Write in a way that can iterate over mutation rates (i.e. put in to form C[C>A]G etc)   \n",
    "    ASXL1_sites_nonsense6={}\n",
    "\n",
    "    separating_positions = {} # separate out all the possible changes to just 1 effective item in the list\n",
    "    a=0\n",
    "    for k, v in ASXL1_sites_nonsense5.items():\n",
    "        for i in v: #because some contain more than 1 change from that codon:\n",
    "            separating_positions[a]=i\n",
    "            a=a+1\n",
    "\n",
    "    for k, v in separating_positions.items():\n",
    "        base_pos = v[0]\n",
    "        original_base = v[1]\n",
    "        new_base = v[2]\n",
    "        site_trio = v[3]\n",
    "        codon_n_912 = v[4]\n",
    "        original_a_a = v[5]\n",
    "        sequence = (str(original_base),str(new_base))\n",
    "        changes = '>'.join(sequence)\n",
    "        left_base = site_trio[0][0]\n",
    "        right_base = site_trio[0][2]\n",
    "        sequence2 = (changes, right_base)\n",
    "        changes2 = ']'.join(sequence2)\n",
    "        sequence3 = (left_base, changes2)\n",
    "        changes3 = '['.join(sequence3)\n",
    "        ASXL1_sites_nonsense6[k] = (v[0],changes3, codon_n_912, original_a_a) #site_trio = the affected base and its neighbours\n",
    "\n",
    "    #Assign the mutation rates to each site...\n",
    "    ASXL1_mutation_rates_nonsense = {}\n",
    "    for k, v in ASXL1_sites_nonsense6.items():\n",
    "        base_change_info = v[1]\n",
    "        if base_change_info in Mutation_rates.keys():\n",
    "            mutation_rate = Mutation_rates.get(base_change_info)\n",
    "            original_a_a = v[3]\n",
    "            aa_positioning = v[2]\n",
    "            variant_type = '*'\n",
    "            variant_id = str(original_a_a)+str(aa_positioning)+variant_type\n",
    "            ASXL1_mutation_rates_nonsense[k] = (v[0],v[1], mutation_rate, variant_id)\n",
    "\n",
    "    ASXL1_mutation_rate_nonsense_sum = []\n",
    "\n",
    "    mutation_rates_list = []\n",
    "    for k, v in ASXL1_mutation_rates_nonsense.items():\n",
    "        codon_position_in_gene=(v[0])\n",
    "        variant = v[3]\n",
    "        site = v[1]\n",
    "        mutation_rate_value = v[2] \n",
    "        mutation_rates_list.append((v[0], variant, site, mutation_rate_value))\n",
    "        ASXL1_mutation_rate_nonsense_sum.append(mutation_rate_value)\n",
    "\n",
    "    #Get the average of all the mutation rates\n",
    "    ASXL1_mutation_rate_nonsense_sum = []\n",
    "    for k, v in ASXL1_mutation_rates_nonsense.items():\n",
    "        mut_rate = v[2]\n",
    "        ASXL1_mutation_rate_nonsense_sum.append(mut_rate)\n",
    "\n",
    "    ASXL1_Summed_mutation_rate_nonsense_added = np.sum(ASXL1_mutation_rate_nonsense_sum)\n",
    "\n",
    "    print('Total summed mutation rate across ASXL1 nonsense for this study = ', ASXL1_Summed_mutation_rate_nonsense_added)\n",
    "\n",
    "    mutation_rates_list_all = mutation_rates_list\n",
    "    return mutation_rates_list_all # e.g. lower= 0, upper = 913 for whole gene\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 305,
   "metadata": {
    "code_folding": [
     0
    ]
   },
   "outputs": [],
   "source": [
    "def mutation_rates_missense_range_of_variants_ASXL1(lower_range, upper_range):\n",
    "\n",
    "    #The mutation rate at a base is calculated based on its trinucleotide context.  To obtain the trinucleotide context, the base either side of the mutated base needs to be ascertained.\n",
    "    #The coding sequence (+ an extra base at each end) of each gene is used when determining the trinucleotide context of the base change.  The problem with using the coding sequence is that\n",
    "    #if a base change falls at the end of an exon, its neighbouring base will actuall be an intronic base, which may be different to the 'neighbouring' base in the coding sequence.\n",
    "    #To account for this, there is an \"intron_exon_check\" function that is called at the intron/exon boundary if the neighbouring base in the intron is different to the neighbouring base in the coding sequence.\n",
    "    \n",
    "    intron_exon_boundary_caution_base = [140, 143, 252, 373, 1086, 1085]\n",
    "   \n",
    "    def intron_exon_check(position_in_gene_number):\n",
    "        if position_in_gene_number == 140:\n",
    "            gene_pos = 140\n",
    "            first_base = ASXL1_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = ASXL1_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 143:\n",
    "            gene_pos = 143\n",
    "            first_base = ASXL1_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = ASXL1_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 252:\n",
    "            gene_pos = 252\n",
    "            first_base = ASXL1_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = ASXL1_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 373:\n",
    "            gene_pos = 373\n",
    "            first_base = ASXL1_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = ASXL1_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 1086:\n",
    "            gene_pos = 1086\n",
    "            first_base = 'G'\n",
    "            middle_base = ASXL1_exons_extra_bases[gene_pos]\n",
    "            third_base = ASXL1_exons_extra_bases[gene_pos+1]\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 1085:\n",
    "            gene_pos = 1085\n",
    "            first_base = ASXL1_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = ASXL1_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        return triplet\n",
    "    \n",
    "    #ASXL1 transcript:\n",
    "    #Downloaded from http://grch37.ensembl.org\n",
    "    ASXL1_exons = 'ATGAAGGACAAACAGAAGAAGAAGAAGGAGCGCACGTGGGCCGAGGCCGCGCGCCTGGTATTAGAAAACTACTCGGATGCTCCAATGACACCAAAACAGATTCTGCAGGTCATAGAGGCAGAAGGACTAAAGGAAATGAGAAGTGGGACTTCCCCTCTCGCATGCCTCAATGCTATGCTACATTCCAATTCAAGAGGAGGAGAGGGGTTGTTTTATAAACTGCCTGGCCGAATCAGCCTTTTCACGCTCAAGAAGGATGCCCTGCAGTGGTCTCGCCATCCAGCTACAGTGGAGGGAGAGGAGCCAGAGGACACGGCTGATGTGGAGAGCTGTGGGTCTAATGAAGCCAGCACTGTGAGTGGTGAAAACGATGTATCTCTTGATGAAACATCTTCGAACGCATCCTGTTCTACAGAATCTCAGAGTCGACCTCTTTCCAATCCCAGGGACAGCTACAGAGCTTCCTCACAGGCGAACAAACAAAAGAAAAAGACTGGGGTGATGCTGCCTCGAGTTGTCCTGACTCCTCTGAAGGTAAACGGGGCCCACGTGGAATCTGCATCAGGGTTCTCGGGCTGCCACGCCGATGGCGAGAGCGGCAGCCCGTCCAGCAGCAGCAGCGGCTCTCTGGCCCTGGGCAGCGCTGCTATTCGTGGCCAGGCCGAGGTCACCCAGGACCCTGCCCCGCTCCTGAGAGGCTTCCGGAAGCCAGCCACAGGTCAAATGAAGCGCAACAGAGGGGAAGAAATAGATTTTGAGACACCTGGGTCCATTCTTGTCAACACCAACCTCCGTGCCCTGATCAACTCTCGGACCTTCCATGCCTTACCATCACACTTCCAGCAGCAGCTCCTCTTCCTCCTGCCTGAAGTAGACAGACAGGTGGGGACGGATGGCCTGTTGCGTCTCAGCAGCAGTGCACTAAATAACGAGTTTTTTACCCATGCGGCTCAGAGCTGGCGGGAGCGCCTGGCTGATGGTGAATTTACTCATGAGATGCAAGTCAGGATACGACAGGAAATGGAGAAGGAAAAGAAGGTGGAACAATGGAAAGAAAAGTTCTTTGAAGACTACTATGGACAGAAGCTGGGTTTGACCAAAGAAGAGTCATTGCAGCAGAACGTGGGCCAGGAGGAGGCTGAAATCAAAAGTGGCTTGTGTGTCCCAGGAGAATCAGTGCGTATACAGCGTGGTCCAGCCACCCGACAGCGAGATGGGCATTTTAAGAAACGCTCTCGGCCAGATCTCCGAACCAGAGCCAGAAGGAATCTGTACAAAAAACAGGAGTCAGAACAAGCAGGGGTTGCTAAGGATGCAAAATCTGTGGCCTCAGATGTTCCCCTCTACAAGGATGGGGAGGCTAAGACTGACCCAGCAGGGCTGAGCAGTCCCCATCTGCCAGGCACATCCTCTGCAGCACCCGACCTGGAGGGTCCCGAATTCCCAGTTGAGTCTGTGGCTTCTCGGATCCAGGCTGAGCCAGACAACTTGGCACGTGCCTCTGCATCTCCAGACAGAATTCCTAGCCTGCCTCAGGAAACTGTGGATCAGGAACCCAAGGATCAGAAGAGGAAATCCTTTGAGCAGGCGGCCTCTGCATCCTTTCCCGAAAAGAAGCCCCGGCTTGAAGATCGTCAGTCCTTTCGTAACACAATTGAAAGTGTTCACACCGAAAAGCCACAGCCCACTAAAGAGGAGCCCAAAGTCCCGCCCATCCGGATTCAACTTTCACGTATCAAACCACCCTGGGTGGTTAAAGGTCAGCCCACTTACCAGATATGCCCCCGGATCATCCCCACCACGGAGTCCTCCTGCCGGGGTTGGACTGGCGCCAGGACCCTCGCAGACATTAAAGCCCGTGCTCTGCAGGTCCGAGGGGCGAGAGGTCACCACTGCCATAGAGAGGCGGCCACCACTGCCATCGGAGGGGGGGGTGGCCCGGGTGGAGGTGGCGGCGGGGCCACCGATGAGGGAGGTGGCAGAGGCAGCAGCAGTGGTGATGGTGGTGAGGCCTGTGGCCACCCTGAGCCCAGGGGAGGCCCGAGCACCCCTGGAAAGTGTACGTCAGATCTACAGCGAACACAACTACTGCCGCCTTATCCTCTAAATGGGGAGCATACCCAGGCCGGAACTGCCATGTCCAGAGCTAGGAGAGAGGACCTGCCTTCTCTGAGAAAGGAGGAAAGCTGCCTACTACAGAGGGCTACAGTTGGACTCACAGATGGGCTAGGAGATGCCTCCCAACTCCCCGTTGCTCCCACTGGGGACCAGCCATGCCAGGCCTTGCCCCTACTGTCCTCCCAAACCTCAGTAGCTGAGAGATTAGTGGAGCAGCCTCAGTTGCATCCGGATGTTAGAACTGAATGTGAGTCTGGCACCACTTCCTGGGAAAGTGATGATGAGGAGCAAGGACCCACCGTTCCTGCAGACAATGGTCCCATTCTGTCTCTAGTGGGAGATGATACATTAGAGAAAGGAACTGGCCAAGCTCTTGACAGTCATCCCACTATGAAGGATCCTGTAAATGTGACCCCCAGTTCCACACCTGAATCCTCACCGACTGATTGCCTGCAGAACAGAGCATTTGATGACGAATTAGGGCTTGGTGGCTCATGCCCTCCTATGAGGGAAAGTGATACTAGACAAGAAAACTTGAAAACCAAGGCTCTCGTTTCTAACAGTTCTTTGCATTGGATACCCATCCCATCGAATGATGAGGTAGTGAAACAGCCCAAACCAGAATCCAGAGAACACATACCATCTGTTGAGCCCCAGGTTGGAGAGGAGTGGGAGAAAGCTGCTCCCACCCCTCCTGCATTGCCTGGGGATTTGACAGCTGAGGAGGGTCTAGATCCTCTTGACAGCCTTACTTCACTCTGGACTGTGCCATCTCGAGGAGGCAGTGACAGCAATGGCAGTTACTGTCAACAGGTGGACATTGAAAAGCTGAAAATCAACGGAGACTCTGAAGCACTGAGTCCTCACGGTGAGTCCACGGATACAGCCTCTGACTTTGAAGGTCACCTCACGGAGGACAGCAGTGAGGCTGACACTAGAGAAGCTGCAGTGACAAAGGGATCTTCGGTGGACAAGGATGAGAAACCCAATTGGAACCAATCTGCCCCACTGTCCAAGGTGAATGGTGACATGCGTCTGGTTACAAGGACAGATGGGATGGTTGCTCCTCAGAGCTGGGTGTCTCGAGTATGTGCGGTCCGCCAAAAGATCCCAGATTCCCTACTGCTGGCCAGTACTGAGTACCAGCCAAGAGCCGTGTGCCTGTCCATGCCTGGGTCCTCAGTGGAGGCCACTAACCCACTTGTGATGCAGTTGCTGCAGGGTAGCTTGCCCCTAGAGAAGGTTCTTCCACCAGCCCACGATGACAGCATGTCAGAATCCCCACAAGTACCACTTACAAAAGACCAGAGCCATGGCTCGCTACGCATGGGATCTTTACATGGTCTTGGAAAAAACAGTGGCATGGTTGATGGAAGCAGCCCCAGTTCTTTAAGGGCTTTGAAGGAGCCTCTTCTGCCAGATAGCTGTGAAACAGGCACTGGTCTTGCCAGGATTGAGGCCACCCAGGCTCCTGGAGCACCCCAAAAGAATTGCAAGGCAGTCCCAAGTTTTGACTCCCTCCATCCAGTGACAAATCCCATTACATCCTCTAGGAAACTGGAAGAAATGGATTCCAAAGAGCAGTTCTCTTCCTTTAGTTGTGAAGATCAGAAGGAAGTCCGTGCTATGTCACAGGACAGTAATTCAAATGCTGCTCCAGGAAAGAGCCCAGGAGATCTTACTACCTCGAGAACACCTCGTTTCTCATCTCCAAATGTGATCTCCTTTGGTCCAGAGCAGACAGGTCGGGCCCTGGGTGATCAGAGCAATGTTACAGGCCAAGGGAAGAAGCTTTTTGGCTCTGGGAATGTGGCTGCAACCCTTCAGCGCCCCAGGCCTGCGGACCCGATGCCTCTTCCTGCTGAGATCCCTCCAGTTTTTCCCAGTGGGAAGTTGGGACCAAGCACAAACTCCATGTCTGGTGGGGTACAGACTCCAAGGGAAGACTGGGCTCCAAAGCCACATGCCTTTGTTGGCAGCGTCAAGAATGAGAAGACTTTTGTGGGGGGTCCTCTTAAGGCAAATGCCGAGAACAGGAAAGCTACTGGGCATAGTCCCCTGGAACTGGTGGGTCACTTGGAAGGGATGCCCTTTGTCATGGACTTGCCCTTCTGGAAATTACCCCGAGAGCCAGGGAAGGGGCTCAGTGAGCCTCTGGAGCCTTCTTCTCTCCCCTCCCAACTCAGCATCAAGCAGGCATTTTATGGGAAGCTTTCTAAACTCCAACTGAGTTCCACCAGCTTTAATTATTCCTCTAGCTCTCCCACCTTTCCCAAAGGCCTTGCTGGAAGTGTGGTGCAGCTGAGCCACAAAGCAAACTTTGGTGCGAGCCACAGTGCATCACTTTCCTTGCAAATGTTCACTGACAGCAGCACGGTGGAAAGCATCTCGCTCCAGTGTGCGTGCAGCCTGAAAGCCATGATCATGTGCCAAGGCTGCGGTGCGTTCTGTCACGATGACTGTATTGGACCCTCAAAGCTCTGTGTATTGTGCCTTGTGGTGAGATAA'\n",
    "    \n",
    "    #Split the sequence into codons \n",
    "    ASXL1_codons = [ASXL1_exons[i:i+3] for i in range(0, len(ASXL1_exons), 3)]\n",
    "\n",
    "    #ASXL1 transcript:\n",
    "    #Downloaded from http://grch37.ensembl.org\n",
    "    ASXL1_exons_extra_bases = 'GATGAAGGACAAACAGAAGAAGAAGAAGGAGCGCACGTGGGCCGAGGCCGCGCGCCTGGTATTAGAAAACTACTCGGATGCTCCAATGACACCAAAACAGATTCTGCAGGTCATAGAGGCAGAAGGACTAAAGGAAATGAGAAGTGGGACTTCCCCTCTCGCATGCCTCAATGCTATGCTACATTCCAATTCAAGAGGAGGAGAGGGGTTGTTTTATAAACTGCCTGGCCGAATCAGCCTTTTCACGCTCAAGAAGGATGCCCTGCAGTGGTCTCGCCATCCAGCTACAGTGGAGGGAGAGGAGCCAGAGGACACGGCTGATGTGGAGAGCTGTGGGTCTAATGAAGCCAGCACTGTGAGTGGTGAAAACGATGTATCTCTTGATGAAACATCTTCGAACGCATCCTGTTCTACAGAATCTCAGAGTCGACCTCTTTCCAATCCCAGGGACAGCTACAGAGCTTCCTCACAGGCGAACAAACAAAAGAAAAAGACTGGGGTGATGCTGCCTCGAGTTGTCCTGACTCCTCTGAAGGTAAACGGGGCCCACGTGGAATCTGCATCAGGGTTCTCGGGCTGCCACGCCGATGGCGAGAGCGGCAGCCCGTCCAGCAGCAGCAGCGGCTCTCTGGCCCTGGGCAGCGCTGCTATTCGTGGCCAGGCCGAGGTCACCCAGGACCCTGCCCCGCTCCTGAGAGGCTTCCGGAAGCCAGCCACAGGTCAAATGAAGCGCAACAGAGGGGAAGAAATAGATTTTGAGACACCTGGGTCCATTCTTGTCAACACCAACCTCCGTGCCCTGATCAACTCTCGGACCTTCCATGCCTTACCATCACACTTCCAGCAGCAGCTCCTCTTCCTCCTGCCTGAAGTAGACAGACAGGTGGGGACGGATGGCCTGTTGCGTCTCAGCAGCAGTGCACTAAATAACGAGTTTTTTACCCATGCGGCTCAGAGCTGGCGGGAGCGCCTGGCTGATGGTGAATTTACTCATGAGATGCAAGTCAGGATACGACAGGAAATGGAGAAGGAAAAGAAGGTGGAACAATGGAAAGAAAAGTTCTTTGAAGACTACTATGGACAGAAGCTGGGTTTGACCAAAGAAGAGTCATTGCAGCAGAACGTGGGCCAGGAGGAGGCTGAAATCAAAAGTGGCTTGTGTGTCCCAGGAGAATCAGTGCGTATACAGCGTGGTCCAGCCACCCGACAGCGAGATGGGCATTTTAAGAAACGCTCTCGGCCAGATCTCCGAACCAGAGCCAGAAGGAATCTGTACAAAAAACAGGAGTCAGAACAAGCAGGGGTTGCTAAGGATGCAAAATCTGTGGCCTCAGATGTTCCCCTCTACAAGGATGGGGAGGCTAAGACTGACCCAGCAGGGCTGAGCAGTCCCCATCTGCCAGGCACATCCTCTGCAGCACCCGACCTGGAGGGTCCCGAATTCCCAGTTGAGTCTGTGGCTTCTCGGATCCAGGCTGAGCCAGACAACTTGGCACGTGCCTCTGCATCTCCAGACAGAATTCCTAGCCTGCCTCAGGAAACTGTGGATCAGGAACCCAAGGATCAGAAGAGGAAATCCTTTGAGCAGGCGGCCTCTGCATCCTTTCCCGAAAAGAAGCCCCGGCTTGAAGATCGTCAGTCCTTTCGTAACACAATTGAAAGTGTTCACACCGAAAAGCCACAGCCCACTAAAGAGGAGCCCAAAGTCCCGCCCATCCGGATTCAACTTTCACGTATCAAACCACCCTGGGTGGTTAAAGGTCAGCCCACTTACCAGATATGCCCCCGGATCATCCCCACCACGGAGTCCTCCTGCCGGGGTTGGACTGGCGCCAGGACCCTCGCAGACATTAAAGCCCGTGCTCTGCAGGTCCGAGGGGCGAGAGGTCACCACTGCCATAGAGAGGCGGCCACCACTGCCATCGGAGGGGGGGGTGGCCCGGGTGGAGGTGGCGGCGGGGCCACCGATGAGGGAGGTGGCAGAGGCAGCAGCAGTGGTGATGGTGGTGAGGCCTGTGGCCACCCTGAGCCCAGGGGAGGCCCGAGCACCCCTGGAAAGTGTACGTCAGATCTACAGCGAACACAACTACTGCCGCCTTATCCTCTAAATGGGGAGCATACCCAGGCCGGAACTGCCATGTCCAGAGCTAGGAGAGAGGACCTGCCTTCTCTGAGAAAGGAGGAAAGCTGCCTACTACAGAGGGCTACAGTTGGACTCACAGATGGGCTAGGAGATGCCTCCCAACTCCCCGTTGCTCCCACTGGGGACCAGCCATGCCAGGCCTTGCCCCTACTGTCCTCCCAAACCTCAGTAGCTGAGAGATTAGTGGAGCAGCCTCAGTTGCATCCGGATGTTAGAACTGAATGTGAGTCTGGCACCACTTCCTGGGAAAGTGATGATGAGGAGCAAGGACCCACCGTTCCTGCAGACAATGGTCCCATTCTGTCTCTAGTGGGAGATGATACATTAGAGAAAGGAACTGGCCAAGCTCTTGACAGTCATCCCACTATGAAGGATCCTGTAAATGTGACCCCCAGTTCCACACCTGAATCCTCACCGACTGATTGCCTGCAGAACAGAGCATTTGATGACGAATTAGGGCTTGGTGGCTCATGCCCTCCTATGAGGGAAAGTGATACTAGACAAGAAAACTTGAAAACCAAGGCTCTCGTTTCTAACAGTTCTTTGCATTGGATACCCATCCCATCGAATGATGAGGTAGTGAAACAGCCCAAACCAGAATCCAGAGAACACATACCATCTGTTGAGCCCCAGGTTGGAGAGGAGTGGGAGAAAGCTGCTCCCACCCCTCCTGCATTGCCTGGGGATTTGACAGCTGAGGAGGGTCTAGATCCTCTTGACAGCCTTACTTCACTCTGGACTGTGCCATCTCGAGGAGGCAGTGACAGCAATGGCAGTTACTGTCAACAGGTGGACATTGAAAAGCTGAAAATCAACGGAGACTCTGAAGCACTGAGTCCTCACGGTGAGTCCACGGATACAGCCTCTGACTTTGAAGGTCACCTCACGGAGGACAGCAGTGAGGCTGACACTAGAGAAGCTGCAGTGACAAAGGGATCTTCGGTGGACAAGGATGAGAAACCCAATTGGAACCAATCTGCCCCACTGTCCAAGGTGAATGGTGACATGCGTCTGGTTACAAGGACAGATGGGATGGTTGCTCCTCAGAGCTGGGTGTCTCGAGTATGTGCGGTCCGCCAAAAGATCCCAGATTCCCTACTGCTGGCCAGTACTGAGTACCAGCCAAGAGCCGTGTGCCTGTCCATGCCTGGGTCCTCAGTGGAGGCCACTAACCCACTTGTGATGCAGTTGCTGCAGGGTAGCTTGCCCCTAGAGAAGGTTCTTCCACCAGCCCACGATGACAGCATGTCAGAATCCCCACAAGTACCACTTACAAAAGACCAGAGCCATGGCTCGCTACGCATGGGATCTTTACATGGTCTTGGAAAAAACAGTGGCATGGTTGATGGAAGCAGCCCCAGTTCTTTAAGGGCTTTGAAGGAGCCTCTTCTGCCAGATAGCTGTGAAACAGGCACTGGTCTTGCCAGGATTGAGGCCACCCAGGCTCCTGGAGCACCCCAAAAGAATTGCAAGGCAGTCCCAAGTTTTGACTCCCTCCATCCAGTGACAAATCCCATTACATCCTCTAGGAAACTGGAAGAAATGGATTCCAAAGAGCAGTTCTCTTCCTTTAGTTGTGAAGATCAGAAGGAAGTCCGTGCTATGTCACAGGACAGTAATTCAAATGCTGCTCCAGGAAAGAGCCCAGGAGATCTTACTACCTCGAGAACACCTCGTTTCTCATCTCCAAATGTGATCTCCTTTGGTCCAGAGCAGACAGGTCGGGCCCTGGGTGATCAGAGCAATGTTACAGGCCAAGGGAAGAAGCTTTTTGGCTCTGGGAATGTGGCTGCAACCCTTCAGCGCCCCAGGCCTGCGGACCCGATGCCTCTTCCTGCTGAGATCCCTCCAGTTTTTCCCAGTGGGAAGTTGGGACCAAGCACAAACTCCATGTCTGGTGGGGTACAGACTCCAAGGGAAGACTGGGCTCCAAAGCCACATGCCTTTGTTGGCAGCGTCAAGAATGAGAAGACTTTTGTGGGGGGTCCTCTTAAGGCAAATGCCGAGAACAGGAAAGCTACTGGGCATAGTCCCCTGGAACTGGTGGGTCACTTGGAAGGGATGCCCTTTGTCATGGACTTGCCCTTCTGGAAATTACCCCGAGAGCCAGGGAAGGGGCTCAGTGAGCCTCTGGAGCCTTCTTCTCTCCCCTCCCAACTCAGCATCAAGCAGGCATTTTATGGGAAGCTTTCTAAACTCCAACTGAGTTCCACCAGCTTTAATTATTCCTCTAGCTCTCCCACCTTTCCCAAAGGCCTTGCTGGAAGTGTGGTGCAGCTGAGCCACAAAGCAAACTTTGGTGCGAGCCACAGTGCATCACTTTCCTTGCAAATGTTCACTGACAGCAGCACGGTGGAAAGCATCTCGCTCCAGTGTGCGTGCAGCCTGAAAGCCATGATCATGTGCCAAGGCTGCGGTGCGTTCTGTCACGATGACTGTATTGGACCCTCAAAGCTCTGTGTATTGTGCCTTGTGGTGAGATAAT'\n",
    "  \n",
    "    #List of all amino acid positions (across whole coding region) covered (amino acid position minus 1 (as python starts at 0))\n",
    "    ASXL1_missense_codons_DICT={}\n",
    "    for i in range(lower_range, upper_range):\n",
    "        ASXL1_missense_codons_DICT[i+1]= (ASXL1_codons[i])\n",
    "\n",
    "    calls_ASXL1_missense = {}\n",
    "    for k,v in ASXL1_missense_codons_DICT.items(): #iterating over the list of codon numbers (and codon sequences)\n",
    "        if v in Possible_Amino_Acid_Changes.keys(): #assign the list of all the possible amino acid changes to each codon they called\n",
    "            changes = Possible_Amino_Acid_Changes_letters[v]\n",
    "    #         print(Possible_Amino_Acid_Changes.values())\n",
    "            calls_ASXL1_missense[k]=(v, changes)\n",
    "\n",
    "    #Changing the original amino acid to a Word (e.g. Phe) rather than a codon\n",
    "    calls_ASXL1_missense_extra = {}\n",
    "\n",
    "    for k, v in calls_ASXL1_missense.items():\n",
    "        if v[0] in Genetic_Code:\n",
    "            Amino_acid_change = Genetic_Code.get(v[0])\n",
    "            calls_ASXL1_missense_extra[k] = (Amino_acid_change, v[0], v[1])\n",
    "            #e.g. 12: (Phe, 'K', '*', 'T' etc..)\n",
    "\n",
    "    #Changing the original amino acid to a LETTER (rather than the codon)\n",
    "    calls_ASXL1_missense_extra2 = {}\n",
    "\n",
    "    for k, v in calls_ASXL1_missense_extra.items():\n",
    "        if v[0] in Amino_acid_abbreviations:\n",
    "            Amino_acid_abbrev = Amino_acid_abbreviations.get(v[0])\n",
    "            calls_ASXL1_missense_extra2[k] = (Amino_acid_abbrev, v[1], v[2])\n",
    "            #e.g. 12: ((F), ('K', '*', 'T' etc..))\n",
    "\n",
    "    #Create a dictionary that contains details of the positions of missense variants at that codon\n",
    "    ASXL1_position_in_list_dict_missense = {}\n",
    "\n",
    "    for k, v in calls_ASXL1_missense_extra2.items():\n",
    "        original_AA = v[0]\n",
    "        possible_AA_changes = v[2]\n",
    "        possible_AA_changes_string = ''.join(v[2])\n",
    "        positions_in_list = []\n",
    "        for i in possible_AA_changes:\n",
    "            if i != original_AA: #not synonymous\n",
    "                if i != '*': #not missense\n",
    "                    position_in_list=-1\n",
    "                    missense_change = i\n",
    "                    new_aa_list = []\n",
    "                    while True:\n",
    "                        position_in_list = possible_AA_changes_string.find(missense_change, position_in_list+1)\n",
    "        #               print(position_in_list)\n",
    "                        if position_in_list == -1: break\n",
    "                        positions_in_list.append(position_in_list)\n",
    "                        positions_in_list = set(positions_in_list) #convert into a 'set' to remove duplicate positions\n",
    "                        positions_in_list = list(positions_in_list) #convert back to a list so can be added to dictionary\n",
    "                    for pos in positions_in_list:\n",
    "                        new_aa = possible_AA_changes_string[pos]\n",
    "                        new_aa_list.append(new_aa)\n",
    "                    ASXL1_position_in_list_dict_missense[k]=(positions_in_list, v, new_aa_list)\n",
    "\n",
    "        aa_positions_variants_called = []\n",
    "        for k, v in ASXL1_position_in_list_dict_missense.items():\n",
    "            aa_positions_variants_called.append(k)\n",
    "    # print('positions at which missense variants called = ', aa_positions_variants_called)\n",
    "\n",
    "    #Create a dictionary that contains the position in the list of possible AA where the variant is and what it changes to\n",
    "    ASXL1_bases_missense = {}\n",
    "    for k, v in ASXL1_position_in_list_dict_missense.items():\n",
    "        original_a_a = v[1][0]\n",
    "        codon_seq = v[1][1]\n",
    "        base_changes=[]\n",
    "        new_a_a_list = v[2]\n",
    "        if codon_seq in Possible_Base_Changes.keys():\n",
    "            list_of_possible_bases = Possible_Base_Changes.get(codon_seq)\n",
    "        for i in v[0]:\n",
    "            base_change_position = i\n",
    "            base_change = list_of_possible_bases[base_change_position]\n",
    "            base_changes.append(base_change)\n",
    "        ASXL1_bases_missense[k]=(v, base_changes, original_a_a, new_a_a_list)\n",
    "\n",
    "    #Create dictionary which contains original codon seq and details on which base is changed\n",
    "    ASXL1_bases_missense2 = {}\n",
    "\n",
    "    for k, v in ASXL1_bases_missense.items():\n",
    "        base_change_positions = v[0][0]\n",
    "        codon_positions = []\n",
    "        positions_in_gene = []\n",
    "        new_a_a_list = v[3]\n",
    "        for i in base_change_positions:\n",
    "            base_change_position = i\n",
    "            if base_change_position <= 2:\n",
    "                codon_pos = 1\n",
    "            if 2 < base_change_position <= 5:\n",
    "                codon_pos = 2\n",
    "            if 5 < base_change_position <= 8:\n",
    "                codon_pos = 3\n",
    "            codon_positions.append(codon_pos)\n",
    "        ASXL1_bases_missense2[k] = (v, codon_positions)\n",
    "\n",
    "    #Create dictionary which contains original codon seq, details on which base is changed and it's position in the coding region\n",
    "    ASXL1_bases_missense3 = {}\n",
    "\n",
    "    for k, v in ASXL1_bases_missense2.items():\n",
    "        codon_positions = v[1]\n",
    "        base_changes = v[0][1]\n",
    "        original_a_a = v[0][2]\n",
    "        new_aa_list = v[0][0][2]\n",
    "        original_bases = []\n",
    "        positions_in_gene = []\n",
    "        codon_number = k\n",
    "        for i in codon_positions:\n",
    "            codon_pos = i\n",
    "            position_in_gene = ((int(k)-1)*3)+codon_pos #codon number minus 1 (e.g. codon 307 -1 = 306, then times by 3 takes you to beginning of codon 307 (then add the position))  \n",
    "            positions_in_gene.append(position_in_gene)\n",
    "            original_base = (v[0][0][1][1][codon_pos-1])\n",
    "            original_bases.append(original_base)\n",
    "        ASXL1_bases_missense3[k] = (v[0][0][1], positions_in_gene, original_bases, base_changes, codon_number, original_a_a, new_aa_list)\n",
    "\n",
    "    # Create dictionary which contains original codon seq, details on which base is changed and its neighbouring bases\n",
    "    ASXL1_bases_missense4 = {}\n",
    "\n",
    "    for k, v in ASXL1_bases_missense3.items():\n",
    "        position_in_gene = v[1]\n",
    "        sites_trio_missense=[]\n",
    "        new_a_a_list = v[6]\n",
    "\n",
    "        for i in position_in_gene:\n",
    "    #         print(i)\n",
    "            if i in intron_exon_boundary_caution_base:\n",
    "#                 print('positions at intron_exon_boundary =',i)\n",
    "                adapted_trio = intron_exon_check(i)\n",
    "                sites_trio_missense.append([adapted_trio])\n",
    "                ASXL1_bases_missense4[k] = (v, sites_trio_missense)\n",
    "\n",
    "            else:\n",
    "                for i in position_in_gene:\n",
    "                    gene_pos = i\n",
    "                    site_trio_missense = [ASXL1_exons_extra_bases[gene_pos:gene_pos+3] for gene_pos in range (gene_pos-1, gene_pos, 3)]\n",
    "                    sites_trio_missense.append(site_trio_missense)\n",
    "                    ASXL1_bases_missense4[k] = (v, sites_trio_missense)\n",
    "\n",
    "    #Create a dictionary containing pos in coding region, original base, new base, base affected with its neighbours\n",
    "    #e.g. 938, G, A, TGG = base 938 on coding region, original base = G, changes to base A, base with neighbours either side = TGG\n",
    "    ASXL1_sites_missense5 = {}\n",
    "\n",
    "    a=0\n",
    "    for k, v in ASXL1_bases_missense4.items():\n",
    "        original_a_a = v[0][0][0]\n",
    "        new_a_a = v[0][6]\n",
    "        codon_n = v[0][1] #base number in codon (out of 3 x 912)\n",
    "        codon_n_912 = v[0][4]\n",
    "        original_base = v[0][2]\n",
    "        new_base = v[0][3]\n",
    "        site_trio = v[1]\n",
    "        info=[]\n",
    "\n",
    "        change = 0\n",
    "        for i in codon_n:\n",
    "            codon_number = i\n",
    "            orig_base = original_base[change]\n",
    "            new_amino_acid = new_a_a[change]\n",
    "            n_base = new_base[change]\n",
    "            site_tri = site_trio[change]\n",
    "            info.append([codon_number, orig_base, n_base, site_tri, codon_n_912, original_a_a, new_amino_acid])\n",
    "            change = change+1\n",
    "        ASXL1_sites_missense5[a]=info\n",
    "        a=a+1\n",
    "    #     print(ASXL1_sites_missense5)\n",
    "\n",
    "    # Write in a way that can iterate over mutation rates (i.e. put in to form C[C>A]G etc)   \n",
    "    ASXL1_sites_missense6={}\n",
    "\n",
    "    separating_positions = {} # separate out all the possible changes to just 1 effective item in the list\n",
    "    a=0\n",
    "    for k, v in ASXL1_sites_missense5.items():\n",
    "        for i in v: #because some contain more than 1 change from that codon:\n",
    "            separating_positions[a]=i\n",
    "            a=a+1\n",
    "\n",
    "    for k, v in separating_positions.items():\n",
    "        base_pos = v[0]\n",
    "        original_base = v[1]\n",
    "        new_base = v[2]\n",
    "        site_trio = v[3]\n",
    "        codon_n_912 = v[4]\n",
    "        original_a_a = v[5]\n",
    "        new_a_a = v[6]\n",
    "        sequence = (str(original_base),str(new_base))\n",
    "        changes = '>'.join(sequence)\n",
    "        left_base = site_trio[0][0]\n",
    "        right_base = site_trio[0][2]\n",
    "        sequence2 = (changes, right_base)\n",
    "        changes2 = ']'.join(sequence2)\n",
    "        sequence3 = (left_base, changes2)\n",
    "        changes3 = '['.join(sequence3)\n",
    "        ASXL1_sites_missense6[k] = (v[0],changes3, codon_n_912, original_a_a, new_a_a) #site_trio = the affected base and its neighbours\n",
    "\n",
    "    #Assign the mutation rates to each site...\n",
    "    ASXL1_mutation_rates_missense = {}\n",
    "    for k, v in ASXL1_sites_missense6.items():\n",
    "        base_change_info = v[1]\n",
    "        if base_change_info in Mutation_rates.keys():\n",
    "            mutation_rate = Mutation_rates.get(base_change_info)\n",
    "            original_a_a = v[3]\n",
    "            new_a_a = v[4]\n",
    "            aa_positioning = v[2]\n",
    "            variant_id = str(original_a_a)+str(aa_positioning)+str(new_a_a)\n",
    "            ASXL1_mutation_rates_missense[k] = (v[0],v[1], mutation_rate, variant_id)\n",
    "\n",
    "    ASXL1_mutation_rate_missense_sum = []\n",
    "\n",
    "    mutation_rates_list = []\n",
    "    for k, v in ASXL1_mutation_rates_missense.items():\n",
    "        codon_position_in_gene=(v[0])\n",
    "        variant = v[3]\n",
    "        site = v[1]\n",
    "        mutation_rate_value = v[2] \n",
    "        mutation_rates_list.append((v[0], variant, site, mutation_rate_value))\n",
    "        ASXL1_mutation_rate_missense_sum.append(mutation_rate_value)\n",
    "\n",
    "    #Get the average of all the mutation rates\n",
    "    ASXL1_mutation_rate_missense_sum = []\n",
    "    for k, v in ASXL1_mutation_rates_missense.items():\n",
    "        mut_rate = v[2]\n",
    "        ASXL1_mutation_rate_missense_sum.append(mut_rate)\n",
    "\n",
    "    ASXL1_Summed_mutation_rate_missense_added = np.sum(ASXL1_mutation_rate_missense_sum)\n",
    "\n",
    "    print('Total summed mutation rate across ASXL1 missense for this study = ', ASXL1_Summed_mutation_rate_missense_added)\n",
    "\n",
    "    mutation_rates_list_all = mutation_rates_list\n",
    "    return mutation_rates_list_all # e.g. lower= 0, upper = 913 for whole gene\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### TP53"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 306,
   "metadata": {
    "code_folding": [
     0
    ]
   },
   "outputs": [],
   "source": [
    "def mutation_rates_nonsense_range_of_variants_TP53(lower_range, upper_range):\n",
    "    \n",
    "    #The mutation rate at a base is calculated based on its trinucleotide context.  To obtain the trinucleotide context, the base either side of the mutated base needs to be ascertained.\n",
    "    #The coding sequence (+ an extra base at each end) of each gene is used when determining the trinucleotide context of the base change.  The problem with using the coding sequence is that\n",
    "    #if a base change falls at the end of an exon, its neighbouring base will actuall be an intronic base, which may be different to the 'neighbouring' base in the coding sequence.\n",
    "    #To account for this, there is an \"intron_exon_check\" function that is called at the intron/exon boundary if the neighbouring base in the intron is different to the neighbouring base in the coding sequence.\n",
    "    \n",
    "    intron_exon_boundary_caution_base_TP53 = [74, 75, 96, 375, 672, 782, 919, 993, 1100]\n",
    "   \n",
    "    def intron_exon_check(position_in_gene_number):\n",
    "        if position_in_gene_number == 74:\n",
    "            gene_pos = 74\n",
    "            first_base = TP53_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = TP53_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 75:\n",
    "            gene_pos = 75\n",
    "            first_base = 'G'\n",
    "            middle_base = TP53_exons_extra_bases[gene_pos]\n",
    "            third_base = TP53_exons_extra_bases[gene_pos+1]\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 96:\n",
    "            gene_pos = 96\n",
    "            first_base = TP53_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = TP53_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 375:\n",
    "            gene_pos = 375\n",
    "            first_base = TP53_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = TP53_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 672:\n",
    "            gene_pos = 672\n",
    "            first_base = TP53_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = TP53_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 782:\n",
    "            gene_pos = 782\n",
    "            first_base = TP53_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = TP53_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 919:\n",
    "            gene_pos = 919\n",
    "            first_base = TP53_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = TP53_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 993:\n",
    "            gene_pos = 993\n",
    "            first_base = TP53_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = TP53_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 1100:\n",
    "            gene_pos = 1100\n",
    "            first_base = TP53_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = TP53_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        return triplet\n",
    "    \n",
    "    #TP53 transcript:\n",
    "    #Downloaded from http://grch37.ensembl.org\n",
    "    TP53_exons = 'ATGGAGGAGCCGCAGTCAGATCCTAGCGTCGAGCCCCCTCTGAGTCAGGAAACATTTTCAGACCTATGGAAACTACTTCCTGAAAACAACGTTCTGTCCCCCTTGCCGTCCCAAGCAATGGATGATTTGATGCTGTCCCCGGACGATATTGAACAATGGTTCACTGAAGACCCAGGTCCAGATGAAGCTCCCAGAATGCCAGAGGCTGCTCCCCCCGTGGCCCCTGCACCAGCAGCTCCTACACCGGCGGCCCCTGCACCAGCCCCCTCCTGGCCCCTGTCATCTTCTGTCCCTTCCCAGAAAACCTACCAGGGCAGCTACGGTTTCCGTCTGGGCTTCTTGCATTCTGGGACAGCCAAGTCTGTGACTTGCACGTACTCCCCTGCCCTCAACAAGATGTTTTGCCAACTGGCCAAGACCTGCCCTGTGCAGCTGTGGGTTGATTCCACACCCCCGCCCGGCACCCGCGTCCGCGCCATGGCCATCTACAAGCAGTCACAGCACATGACGGAGGTTGTGAGGCGCTGCCCCCACCATGAGCGCTGCTCAGATAGCGATGGTCTGGCCCCTCCTCAGCATCTTATCCGAGTGGAAGGAAATTTGCGTGTGGAGTATTTGGATGACAGAAACACTTTTCGACATAGTGTGGTGGTGCCCTATGAGCCGCCTGAGGTTGGCTCTGACTGTACCACCATCCACTACAACTACATGTGTAACAGTTCCTGCATGGGCGGCATGAACCGGAGGCCCATCCTCACCATCATCACACTGGAAGACTCCAGTGGTAATCTACTGGGACGGAACAGCTTTGAGGTGCGTGTTTGTGCCTGTCCTGGGAGAGACCGGCGCACAGAGGAAGAGAATCTCCGCAAGAAAGGGGAGCCTCACCACGAGCTGCCCCCAGGGAGCACTAAGCGAGCACTGCCCAACAACACCAGCTCCTCTCCCCAGCCAAAGAAGAAACCACTGGATGGAGAATATTTCACCCTTCAGATCCGTGGGCGTGAGCGCTTCGAGATGTTCCGAGAGCTGAATGAGGCCTTGGAACTCAAGGATGCCCAGGCTGGGAAGGAGCCAGGGGGGAGCAGGGCTCACTCCAGCCACCTGAAGTCCAAAAAGGGTCAGTCTACCTCCCGCCATAAAAAACTCATGTTCAAGACAGAAGGGCCTGACTCAGACTGA'\n",
    "    \n",
    "    #Split the sequence into codons \n",
    "    TP53_codons = [TP53_exons[i:i+3] for i in range(0, len(TP53_exons), 3)]\n",
    "\n",
    "    #TP53 transcript:\n",
    "    #Downloaded from http://grch37.ensembl.org\n",
    "    TP53_exons_extra_bases = 'CATGGAGGAGCCGCAGTCAGATCCTAGCGTCGAGCCCCCTCTGAGTCAGGAAACATTTTCAGACCTATGGAAACTACTTCCTGAAAACAACGTTCTGTCCCCCTTGCCGTCCCAAGCAATGGATGATTTGATGCTGTCCCCGGACGATATTGAACAATGGTTCACTGAAGACCCAGGTCCAGATGAAGCTCCCAGAATGCCAGAGGCTGCTCCCCCCGTGGCCCCTGCACCAGCAGCTCCTACACCGGCGGCCCCTGCACCAGCCCCCTCCTGGCCCCTGTCATCTTCTGTCCCTTCCCAGAAAACCTACCAGGGCAGCTACGGTTTCCGTCTGGGCTTCTTGCATTCTGGGACAGCCAAGTCTGTGACTTGCACGTACTCCCCTGCCCTCAACAAGATGTTTTGCCAACTGGCCAAGACCTGCCCTGTGCAGCTGTGGGTTGATTCCACACCCCCGCCCGGCACCCGCGTCCGCGCCATGGCCATCTACAAGCAGTCACAGCACATGACGGAGGTTGTGAGGCGCTGCCCCCACCATGAGCGCTGCTCAGATAGCGATGGTCTGGCCCCTCCTCAGCATCTTATCCGAGTGGAAGGAAATTTGCGTGTGGAGTATTTGGATGACAGAAACACTTTTCGACATAGTGTGGTGGTGCCCTATGAGCCGCCTGAGGTTGGCTCTGACTGTACCACCATCCACTACAACTACATGTGTAACAGTTCCTGCATGGGCGGCATGAACCGGAGGCCCATCCTCACCATCATCACACTGGAAGACTCCAGTGGTAATCTACTGGGACGGAACAGCTTTGAGGTGCGTGTTTGTGCCTGTCCTGGGAGAGACCGGCGCACAGAGGAAGAGAATCTCCGCAAGAAAGGGGAGCCTCACCACGAGCTGCCCCCAGGGAGCACTAAGCGAGCACTGCCCAACAACACCAGCTCCTCTCCCCAGCCAAAGAAGAAACCACTGGATGGAGAATATTTCACCCTTCAGATCCGTGGGCGTGAGCGCTTCGAGATGTTCCGAGAGCTGAATGAGGCCTTGGAACTCAAGGATGCCCAGGCTGGGAAGGAGCCAGGGGGGAGCAGGGCTCACTCCAGCCACCTGAAGTCCAAAAAGGGTCAGTCTACCTCCCGCCATAAAAAACTCATGTTCAAGACAGAAGGGCCTGACTCAGACTGAC'\n",
    "    \n",
    "    #List of all amino acid positions (across whole coding region) covered (amino acid position minus 1 (as python starts at 0))\n",
    "    TP53_nonsense_codons_DICT={}\n",
    "    for i in range(lower_range, upper_range):\n",
    "        TP53_nonsense_codons_DICT[i+1]= (TP53_codons[i])\n",
    "\n",
    "    calls_TP53_nonsense = {}\n",
    "    for k,v in TP53_nonsense_codons_DICT.items(): #iterating over the list of codon numbers (and codon sequences)\n",
    "        if v in Possible_Amino_Acid_Changes.keys(): #assign the list of all the possible amino acid changes to each codon they called\n",
    "            changes = Possible_Amino_Acid_Changes_letters[v]\n",
    "    #         print(Possible_Amino_Acid_Changes.values())\n",
    "            calls_TP53_nonsense[k]=(v, changes)\n",
    "\n",
    "    #Changing the original amino acid to a Word (e.g. Phe) rather than a codon\n",
    "    calls_TP53_nonsense_extra = {}\n",
    "\n",
    "    for k, v in calls_TP53_nonsense.items():\n",
    "        if v[0] in Genetic_Code:\n",
    "            Amino_acid_change = Genetic_Code.get(v[0])\n",
    "            calls_TP53_nonsense_extra[k] = (Amino_acid_change, v[0], v[1])\n",
    "            #e.g. 12: (Phe, 'K', '*', 'T' etc..)\n",
    "\n",
    "    #Changing the original amino acid to a LETTER (rather than the codon)\n",
    "    calls_TP53_nonsense_extra2 = {}\n",
    "\n",
    "    for k, v in calls_TP53_nonsense_extra.items():\n",
    "        if v[0] in Amino_acid_abbreviations:\n",
    "            Amino_acid_abbrev = Amino_acid_abbreviations.get(v[0])\n",
    "            calls_TP53_nonsense_extra2[k] = (Amino_acid_abbrev, v[1], v[2])\n",
    "            #e.g. 12: ((F), ('K', '*', 'T' etc..))\n",
    "\n",
    "    #Create a dictionary that contains details of the positions of nonsense variants at that codon\n",
    "    TP53_position_in_list_dict_nonsense = {}\n",
    "\n",
    "    for k, v in calls_TP53_nonsense_extra2.items():\n",
    "        original_AA = v[0]\n",
    "        possible_AA_changes = v[2]\n",
    "        possible_AA_changes_string = ''.join(v[2])\n",
    "        positions_in_list = []\n",
    "        for i in possible_AA_changes:\n",
    "            if i == '*': #nonsense\n",
    "                position_in_list=-1\n",
    "                nonsense_change = i    \n",
    "                while True:\n",
    "                    position_in_list = possible_AA_changes_string.find(nonsense_change, position_in_list+1)\n",
    "    #               print(position_in_list)\n",
    "                    if position_in_list == -1: break\n",
    "                    positions_in_list.append(position_in_list)\n",
    "                    positions_in_list = set(positions_in_list) #convert into a 'set' to remove duplicate positions\n",
    "                    positions_in_list = list(positions_in_list) #convert back to a list so can be added to dictionary\n",
    "                TP53_position_in_list_dict_nonsense[k]=(positions_in_list, v)\n",
    "\n",
    "        aa_positions_variants_called = []\n",
    "        for k, v in TP53_position_in_list_dict_nonsense.items():\n",
    "            aa_positions_variants_called.append(k)\n",
    "    # print('positions at which nonsense variants called = ', aa_positions_variants_called)\n",
    "\n",
    "    #Create a dictionary that contains the position in the list of possible AA where the variant is and what it changes to\n",
    "    TP53_bases_nonsense = {}\n",
    "    for k, v in TP53_position_in_list_dict_nonsense.items():\n",
    "        original_a_a = v[1][0]\n",
    "        codon_seq = v[1][1]\n",
    "        base_changes=[]\n",
    "        if codon_seq in Possible_Base_Changes.keys():\n",
    "            list_of_possible_bases = Possible_Base_Changes.get(codon_seq)\n",
    "        for i in v[0]:\n",
    "            base_change_position = i\n",
    "            base_change = list_of_possible_bases[base_change_position]\n",
    "            base_changes.append(base_change)\n",
    "        TP53_bases_nonsense[k]=(v, base_changes, original_a_a)\n",
    "\n",
    "    #Create dictionary which contains original codon seq and details on which base is changed\n",
    "    TP53_bases_nonsense2 = {}\n",
    "\n",
    "    for k, v in TP53_bases_nonsense.items():\n",
    "        base_change_positions = v[0][0]\n",
    "        codon_positions = []\n",
    "        positions_in_gene = []\n",
    "        for i in base_change_positions:\n",
    "            base_change_position = i\n",
    "            if base_change_position <= 2:\n",
    "                codon_pos = 1\n",
    "            if 2 < base_change_position <= 5:\n",
    "                codon_pos = 2\n",
    "            if 5 < base_change_position <= 8:\n",
    "                codon_pos = 3\n",
    "            codon_positions.append(codon_pos)\n",
    "        TP53_bases_nonsense2[k] = (v, codon_positions)\n",
    "\n",
    "    #Create dictionary which contains original codon seq, details on which base is changed and it's position in the coding region\n",
    "    TP53_bases_nonsense3 = {}\n",
    "\n",
    "    for k, v in TP53_bases_nonsense2.items():\n",
    "        codon_positions = v[1]\n",
    "        base_changes = v[0][1]\n",
    "        original_a_a = v[0][2]\n",
    "        original_bases = []\n",
    "        positions_in_gene = []\n",
    "        codon_number = k\n",
    "        for i in codon_positions:\n",
    "            codon_pos = i\n",
    "            position_in_gene = ((int(k)-1)*3)+codon_pos #codon number minus 1 (e.g. codon 307 -1 = 306, then times by 3 takes you to beginning of codon 307 (then add the position))  \n",
    "            positions_in_gene.append(position_in_gene)\n",
    "            original_base = (v[0][0][1][1][codon_pos-1])\n",
    "            original_bases.append(original_base)\n",
    "        TP53_bases_nonsense3[k] = (v[0][0][1], positions_in_gene, original_bases, base_changes, codon_number, original_a_a)\n",
    "\n",
    "    #Create dictionary which contains original codon seq, details on which base is changed and its neighbouring bases\n",
    "    TP53_bases_nonsense4 = {}\n",
    "\n",
    "    for k, v in TP53_bases_nonsense3.items():\n",
    "        position_in_gene = v[1]\n",
    "        sites_trio_nonsense=[]\n",
    "\n",
    "        for i in position_in_gene:\n",
    "    #         print(i)\n",
    "            if i in intron_exon_boundary_caution_base_TP53:\n",
    "#                 print('positions at intron_exon_boundary =',i)\n",
    "                adapted_trio = intron_exon_check_TP53(i)\n",
    "                sites_trio_nonsense.append([adapted_trio])\n",
    "                TP53_bases_nonsense4[k] = (v, sites_trio_nonsense)\n",
    "\n",
    "            else:\n",
    "                for i in position_in_gene:\n",
    "                    gene_pos = i\n",
    "                    site_trio_nonsense = [TP53_exons_extra_bases[gene_pos:gene_pos+3] for gene_pos in range (gene_pos-1, gene_pos, 3)]\n",
    "                    sites_trio_nonsense.append(site_trio_nonsense)\n",
    "                    TP53_bases_nonsense4[k] = (v, sites_trio_nonsense)\n",
    "\n",
    "    #Create a dictionary containing pos in coding region, original base, new base, base affected with its neighbours\n",
    "    #e.g. 938, G, A, TGG = base 938 on coding region, original base = G, changes to base A, base with neighbours either side = TGG\n",
    "    TP53_sites_nonsense5 = {}\n",
    "\n",
    "    a=0\n",
    "    for k, v in TP53_bases_nonsense4.items():\n",
    "        original_a_a = v[0][0][0]\n",
    "        codon_n = v[0][1] #base number in codon (out of 3 x 912)\n",
    "        codon_n_912 = v[0][4]\n",
    "        original_base = v[0][2]\n",
    "        new_base = v[0][3]\n",
    "        site_trio = v[1]\n",
    "        info=[]\n",
    "\n",
    "        change = 0\n",
    "        for i in codon_n:\n",
    "            codon_number = i\n",
    "            orig_base = original_base[change]\n",
    "            n_base = new_base[change]\n",
    "            site_tri = site_trio[change]\n",
    "            info.append([codon_number, orig_base, n_base, site_tri, codon_n_912, original_a_a])\n",
    "            change = change+1\n",
    "        TP53_sites_nonsense5[a]=info\n",
    "        a=a+1\n",
    "    #     print(TP53_sites_nonsense5)\n",
    "\n",
    "    # Write in a way that can iterate over mutation rates (i.e. put in to form C[C>A]G etc)   \n",
    "    TP53_sites_nonsense6={}\n",
    "\n",
    "    separating_positions = {} # separate out all the possible changes to just 1 effective item in the list\n",
    "    a=0\n",
    "    for k, v in TP53_sites_nonsense5.items():\n",
    "        for i in v: #because some contain more than 1 change from that codon:\n",
    "            separating_positions[a]=i\n",
    "            a=a+1\n",
    "\n",
    "    for k, v in separating_positions.items():\n",
    "        base_pos = v[0]\n",
    "        original_base = v[1]\n",
    "        new_base = v[2]\n",
    "        site_trio = v[3]\n",
    "        codon_n_912 = v[4]\n",
    "        original_a_a = v[5]\n",
    "        sequence = (str(original_base),str(new_base))\n",
    "        changes = '>'.join(sequence)\n",
    "        left_base = site_trio[0][0]\n",
    "        right_base = site_trio[0][2]\n",
    "        sequence2 = (changes, right_base)\n",
    "        changes2 = ']'.join(sequence2)\n",
    "        sequence3 = (left_base, changes2)\n",
    "        changes3 = '['.join(sequence3)\n",
    "        TP53_sites_nonsense6[k] = (v[0],changes3, codon_n_912, original_a_a) #site_trio = the affected base and its neighbours\n",
    "\n",
    "    #Assign the mutation rates to each site...\n",
    "    TP53_mutation_rates_nonsense = {}\n",
    "    for k, v in TP53_sites_nonsense6.items():\n",
    "        base_change_info = v[1]\n",
    "        if base_change_info in Mutation_rates.keys():\n",
    "            mutation_rate = Mutation_rates.get(base_change_info)\n",
    "            original_a_a = v[3]\n",
    "            aa_positioning = v[2]\n",
    "            variant_type = '*'\n",
    "            variant_id = str(original_a_a)+str(aa_positioning)+variant_type\n",
    "            TP53_mutation_rates_nonsense[k] = (v[0],v[1], mutation_rate, variant_id)\n",
    "\n",
    "    TP53_mutation_rate_nonsense_sum = []\n",
    "\n",
    "    mutation_rates_list = []\n",
    "    for k, v in TP53_mutation_rates_nonsense.items():\n",
    "        codon_position_in_gene=(v[0])\n",
    "        variant = v[3]\n",
    "        site = v[1]\n",
    "        mutation_rate_value = v[2] \n",
    "        mutation_rates_list.append((v[0], variant, site, mutation_rate_value))\n",
    "        TP53_mutation_rate_nonsense_sum.append(mutation_rate_value)\n",
    "\n",
    "    #Get the average of all the mutation rates\n",
    "    TP53_mutation_rate_nonsense_sum = []\n",
    "    for k, v in TP53_mutation_rates_nonsense.items():\n",
    "        mut_rate = v[2]\n",
    "        TP53_mutation_rate_nonsense_sum.append(mut_rate)\n",
    "\n",
    "    TP53_Summed_mutation_rate_nonsense_added = np.sum(TP53_mutation_rate_nonsense_sum)\n",
    "\n",
    "    print('Total summed mutation rate across TP53 nonsense for this study = ', TP53_Summed_mutation_rate_nonsense_added)\n",
    "\n",
    "    mutation_rates_list_all = mutation_rates_list\n",
    "    return mutation_rates_list_all # e.g. lower= 0, upper = 913 for whole gene\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 307,
   "metadata": {
    "code_folding": [
     0
    ]
   },
   "outputs": [],
   "source": [
    "def mutation_rates_missense_range_of_variants_TP53(lower_range, upper_range):\n",
    "\n",
    "    #The mutation rate at a base is calculated based on its trinucleotide context.  To obtain the trinucleotide context, the base either side of the mutated base needs to be ascertained.\n",
    "    #The coding sequence (+ an extra base at each end) of each gene is used when determining the trinucleotide context of the base change.  The problem with using the coding sequence is that\n",
    "    #if a base change falls at the end of an exon, its neighbouring base will actuall be an intronic base, which may be different to the 'neighbouring' base in the coding sequence.\n",
    "    #To account for this, there is an \"intron_exon_check\" function that is called at the intron/exon boundary if the neighbouring base in the intron is different to the neighbouring base in the coding sequence.\n",
    "    \n",
    "    intron_exon_boundary_caution_base_TP53 = [74, 75, 96, 375, 672, 782, 919, 993, 1100]\n",
    "   \n",
    "    def intron_exon_check(position_in_gene_number):\n",
    "        if position_in_gene_number == 74:\n",
    "            gene_pos = 74\n",
    "            first_base = TP53_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = TP53_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 75:\n",
    "            gene_pos = 75\n",
    "            first_base = 'G'\n",
    "            middle_base = TP53_exons_extra_bases[gene_pos]\n",
    "            third_base = TP53_exons_extra_bases[gene_pos+1]\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 96:\n",
    "            gene_pos = 96\n",
    "            first_base = TP53_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = TP53_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 375:\n",
    "            gene_pos = 375\n",
    "            first_base = TP53_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = TP53_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 672:\n",
    "            gene_pos = 672\n",
    "            first_base = TP53_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = TP53_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 782:\n",
    "            gene_pos = 782\n",
    "            first_base = TP53_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = TP53_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 919:\n",
    "            gene_pos = 919\n",
    "            first_base = TP53_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = TP53_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 993:\n",
    "            gene_pos = 993\n",
    "            first_base = TP53_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = TP53_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 1100:\n",
    "            gene_pos = 1100\n",
    "            first_base = TP53_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = TP53_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        return triplet\n",
    "    \n",
    "    #TP53 transcript:\n",
    "    #Downloaded from http://grch37.ensembl.org\n",
    "    TP53_exons = 'ATGGAGGAGCCGCAGTCAGATCCTAGCGTCGAGCCCCCTCTGAGTCAGGAAACATTTTCAGACCTATGGAAACTACTTCCTGAAAACAACGTTCTGTCCCCCTTGCCGTCCCAAGCAATGGATGATTTGATGCTGTCCCCGGACGATATTGAACAATGGTTCACTGAAGACCCAGGTCCAGATGAAGCTCCCAGAATGCCAGAGGCTGCTCCCCCCGTGGCCCCTGCACCAGCAGCTCCTACACCGGCGGCCCCTGCACCAGCCCCCTCCTGGCCCCTGTCATCTTCTGTCCCTTCCCAGAAAACCTACCAGGGCAGCTACGGTTTCCGTCTGGGCTTCTTGCATTCTGGGACAGCCAAGTCTGTGACTTGCACGTACTCCCCTGCCCTCAACAAGATGTTTTGCCAACTGGCCAAGACCTGCCCTGTGCAGCTGTGGGTTGATTCCACACCCCCGCCCGGCACCCGCGTCCGCGCCATGGCCATCTACAAGCAGTCACAGCACATGACGGAGGTTGTGAGGCGCTGCCCCCACCATGAGCGCTGCTCAGATAGCGATGGTCTGGCCCCTCCTCAGCATCTTATCCGAGTGGAAGGAAATTTGCGTGTGGAGTATTTGGATGACAGAAACACTTTTCGACATAGTGTGGTGGTGCCCTATGAGCCGCCTGAGGTTGGCTCTGACTGTACCACCATCCACTACAACTACATGTGTAACAGTTCCTGCATGGGCGGCATGAACCGGAGGCCCATCCTCACCATCATCACACTGGAAGACTCCAGTGGTAATCTACTGGGACGGAACAGCTTTGAGGTGCGTGTTTGTGCCTGTCCTGGGAGAGACCGGCGCACAGAGGAAGAGAATCTCCGCAAGAAAGGGGAGCCTCACCACGAGCTGCCCCCAGGGAGCACTAAGCGAGCACTGCCCAACAACACCAGCTCCTCTCCCCAGCCAAAGAAGAAACCACTGGATGGAGAATATTTCACCCTTCAGATCCGTGGGCGTGAGCGCTTCGAGATGTTCCGAGAGCTGAATGAGGCCTTGGAACTCAAGGATGCCCAGGCTGGGAAGGAGCCAGGGGGGAGCAGGGCTCACTCCAGCCACCTGAAGTCCAAAAAGGGTCAGTCTACCTCCCGCCATAAAAAACTCATGTTCAAGACAGAAGGGCCTGACTCAGACTGA'\n",
    "    \n",
    "    #Split the sequence into codons \n",
    "    TP53_codons = [TP53_exons[i:i+3] for i in range(0, len(TP53_exons), 3)]\n",
    "\n",
    "    #TP53 transcript:\n",
    "    #Downloaded from http://grch37.ensembl.org\n",
    "    TP53_exons_extra_bases = 'CATGGAGGAGCCGCAGTCAGATCCTAGCGTCGAGCCCCCTCTGAGTCAGGAAACATTTTCAGACCTATGGAAACTACTTCCTGAAAACAACGTTCTGTCCCCCTTGCCGTCCCAAGCAATGGATGATTTGATGCTGTCCCCGGACGATATTGAACAATGGTTCACTGAAGACCCAGGTCCAGATGAAGCTCCCAGAATGCCAGAGGCTGCTCCCCCCGTGGCCCCTGCACCAGCAGCTCCTACACCGGCGGCCCCTGCACCAGCCCCCTCCTGGCCCCTGTCATCTTCTGTCCCTTCCCAGAAAACCTACCAGGGCAGCTACGGTTTCCGTCTGGGCTTCTTGCATTCTGGGACAGCCAAGTCTGTGACTTGCACGTACTCCCCTGCCCTCAACAAGATGTTTTGCCAACTGGCCAAGACCTGCCCTGTGCAGCTGTGGGTTGATTCCACACCCCCGCCCGGCACCCGCGTCCGCGCCATGGCCATCTACAAGCAGTCACAGCACATGACGGAGGTTGTGAGGCGCTGCCCCCACCATGAGCGCTGCTCAGATAGCGATGGTCTGGCCCCTCCTCAGCATCTTATCCGAGTGGAAGGAAATTTGCGTGTGGAGTATTTGGATGACAGAAACACTTTTCGACATAGTGTGGTGGTGCCCTATGAGCCGCCTGAGGTTGGCTCTGACTGTACCACCATCCACTACAACTACATGTGTAACAGTTCCTGCATGGGCGGCATGAACCGGAGGCCCATCCTCACCATCATCACACTGGAAGACTCCAGTGGTAATCTACTGGGACGGAACAGCTTTGAGGTGCGTGTTTGTGCCTGTCCTGGGAGAGACCGGCGCACAGAGGAAGAGAATCTCCGCAAGAAAGGGGAGCCTCACCACGAGCTGCCCCCAGGGAGCACTAAGCGAGCACTGCCCAACAACACCAGCTCCTCTCCCCAGCCAAAGAAGAAACCACTGGATGGAGAATATTTCACCCTTCAGATCCGTGGGCGTGAGCGCTTCGAGATGTTCCGAGAGCTGAATGAGGCCTTGGAACTCAAGGATGCCCAGGCTGGGAAGGAGCCAGGGGGGAGCAGGGCTCACTCCAGCCACCTGAAGTCCAAAAAGGGTCAGTCTACCTCCCGCCATAAAAAACTCATGTTCAAGACAGAAGGGCCTGACTCAGACTGAC'\n",
    "    \n",
    "    #List of all amino acid positions (across whole coding region) covered (amino acid position minus 1 (as python starts at 0))\n",
    "    TP53_missense_codons_DICT={}\n",
    "    for i in range(lower_range, upper_range):\n",
    "        TP53_missense_codons_DICT[i+1]= (TP53_codons[i])\n",
    "\n",
    "    calls_TP53_missense = {}\n",
    "    for k,v in TP53_missense_codons_DICT.items(): #iterating over the list of codon numbers (and codon sequences)\n",
    "        if v in Possible_Amino_Acid_Changes.keys(): #assign the list of all the possible amino acid changes to each codon they called\n",
    "            changes = Possible_Amino_Acid_Changes_letters[v]\n",
    "    #         print(Possible_Amino_Acid_Changes.values())\n",
    "            calls_TP53_missense[k]=(v, changes)\n",
    "\n",
    "    #Changing the original amino acid to a Word (e.g. Phe) rather than a codon\n",
    "    calls_TP53_missense_extra = {}\n",
    "\n",
    "    for k, v in calls_TP53_missense.items():\n",
    "        if v[0] in Genetic_Code:\n",
    "            Amino_acid_change = Genetic_Code.get(v[0])\n",
    "            calls_TP53_missense_extra[k] = (Amino_acid_change, v[0], v[1])\n",
    "            #e.g. 12: (Phe, 'K', '*', 'T' etc..)\n",
    "\n",
    "    #Changing the original amino acid to a LETTER (rather than the codon)\n",
    "    calls_TP53_missense_extra2 = {}\n",
    "\n",
    "    for k, v in calls_TP53_missense_extra.items():\n",
    "        if v[0] in Amino_acid_abbreviations:\n",
    "            Amino_acid_abbrev = Amino_acid_abbreviations.get(v[0])\n",
    "            calls_TP53_missense_extra2[k] = (Amino_acid_abbrev, v[1], v[2])\n",
    "            #e.g. 12: ((F), ('K', '*', 'T' etc..))\n",
    "\n",
    "    #Create a dictionary that contains details of the positions of missense variants at that codon\n",
    "    TP53_position_in_list_dict_missense = {}\n",
    "\n",
    "    for k, v in calls_TP53_missense_extra2.items():\n",
    "        original_AA = v[0]\n",
    "        possible_AA_changes = v[2]\n",
    "        possible_AA_changes_string = ''.join(v[2])\n",
    "        positions_in_list = []\n",
    "        for i in possible_AA_changes:\n",
    "            if i != original_AA: #not synonymous\n",
    "                if i != '*': #not missense\n",
    "                    position_in_list=-1\n",
    "                    missense_change = i\n",
    "                    new_aa_list = []\n",
    "                    while True:\n",
    "                        position_in_list = possible_AA_changes_string.find(missense_change, position_in_list+1)\n",
    "        #               print(position_in_list)\n",
    "                        if position_in_list == -1: break\n",
    "                        positions_in_list.append(position_in_list)\n",
    "                        positions_in_list = set(positions_in_list) #convert into a 'set' to remove duplicate positions\n",
    "                        positions_in_list = list(positions_in_list) #convert back to a list so can be added to dictionary\n",
    "                    for pos in positions_in_list:\n",
    "                        new_aa = possible_AA_changes_string[pos]\n",
    "                        new_aa_list.append(new_aa)\n",
    "                    TP53_position_in_list_dict_missense[k]=(positions_in_list, v, new_aa_list)\n",
    "\n",
    "        aa_positions_variants_called = []\n",
    "        for k, v in TP53_position_in_list_dict_missense.items():\n",
    "            aa_positions_variants_called.append(k)\n",
    "    # print('positions at which missense variants called = ', aa_positions_variants_called)\n",
    "\n",
    "    #Create a dictionary that contains the position in the list of possible AA where the variant is and what it changes to\n",
    "    TP53_bases_missense = {}\n",
    "    for k, v in TP53_position_in_list_dict_missense.items():\n",
    "        original_a_a = v[1][0]\n",
    "        codon_seq = v[1][1]\n",
    "        base_changes=[]\n",
    "        new_a_a_list = v[2]\n",
    "        if codon_seq in Possible_Base_Changes.keys():\n",
    "            list_of_possible_bases = Possible_Base_Changes.get(codon_seq)\n",
    "        for i in v[0]:\n",
    "            base_change_position = i\n",
    "            base_change = list_of_possible_bases[base_change_position]\n",
    "            base_changes.append(base_change)\n",
    "        TP53_bases_missense[k]=(v, base_changes, original_a_a, new_a_a_list)\n",
    "\n",
    "    #Create dictionary which contains original codon seq and details on which base is changed\n",
    "    TP53_bases_missense2 = {}\n",
    "\n",
    "    for k, v in TP53_bases_missense.items():\n",
    "        base_change_positions = v[0][0]\n",
    "        codon_positions = []\n",
    "        positions_in_gene = []\n",
    "        new_a_a_list = v[3]\n",
    "        for i in base_change_positions:\n",
    "            base_change_position = i\n",
    "            if base_change_position <= 2:\n",
    "                codon_pos = 1\n",
    "            if 2 < base_change_position <= 5:\n",
    "                codon_pos = 2\n",
    "            if 5 < base_change_position <= 8:\n",
    "                codon_pos = 3\n",
    "            codon_positions.append(codon_pos)\n",
    "        TP53_bases_missense2[k] = (v, codon_positions)\n",
    "\n",
    "    #Create dictionary which contains original codon seq, details on which base is changed and it's position in the coding region\n",
    "    TP53_bases_missense3 = {}\n",
    "\n",
    "    for k, v in TP53_bases_missense2.items():\n",
    "        codon_positions = v[1]\n",
    "        base_changes = v[0][1]\n",
    "        original_a_a = v[0][2]\n",
    "        new_aa_list = v[0][0][2]\n",
    "        original_bases = []\n",
    "        positions_in_gene = []\n",
    "        codon_number = k\n",
    "        for i in codon_positions:\n",
    "            codon_pos = i\n",
    "            position_in_gene = ((int(k)-1)*3)+codon_pos #codon number minus 1 (e.g. codon 307 -1 = 306, then times by 3 takes you to beginning of codon 307 (then add the position))  \n",
    "            positions_in_gene.append(position_in_gene)\n",
    "            original_base = (v[0][0][1][1][codon_pos-1])\n",
    "            original_bases.append(original_base)\n",
    "        TP53_bases_missense3[k] = (v[0][0][1], positions_in_gene, original_bases, base_changes, codon_number, original_a_a, new_aa_list)\n",
    "\n",
    "    # Create dictionary which contains original codon seq, details on which base is changed and its neighbouring bases\n",
    "    TP53_bases_missense4 = {}\n",
    "\n",
    "    for k, v in TP53_bases_missense3.items():\n",
    "        position_in_gene = v[1]\n",
    "        sites_trio_missense=[]\n",
    "        new_a_a_list = v[6]\n",
    "\n",
    "        for i in position_in_gene:\n",
    "    #         print(i)\n",
    "            if i in intron_exon_boundary_caution_base_TP53:\n",
    "#                 print('positions at intron_exon_boundary =',i)\n",
    "                adapted_trio = intron_exon_check(i)\n",
    "                sites_trio_missense.append([adapted_trio])\n",
    "                TP53_bases_missense4[k] = (v, sites_trio_missense)\n",
    "\n",
    "            else:\n",
    "                for i in position_in_gene:\n",
    "                    gene_pos = i\n",
    "                    site_trio_missense = [TP53_exons_extra_bases[gene_pos:gene_pos+3] for gene_pos in range (gene_pos-1, gene_pos, 3)]\n",
    "                    sites_trio_missense.append(site_trio_missense)\n",
    "                    TP53_bases_missense4[k] = (v, sites_trio_missense)\n",
    "\n",
    "    #Create a dictionary containing pos in coding region, original base, new base, base affected with its neighbours\n",
    "    #e.g. 938, G, A, TGG = base 938 on coding region, original base = G, changes to base A, base with neighbours either side = TGG\n",
    "    TP53_sites_missense5 = {}\n",
    "\n",
    "    a=0\n",
    "    for k, v in TP53_bases_missense4.items():\n",
    "        original_a_a = v[0][0][0]\n",
    "        new_a_a = v[0][6]\n",
    "        codon_n = v[0][1] #base number in codon (out of 3 x 912)\n",
    "        codon_n_912 = v[0][4]\n",
    "        original_base = v[0][2]\n",
    "        new_base = v[0][3]\n",
    "        site_trio = v[1]\n",
    "        info=[]\n",
    "\n",
    "        change = 0\n",
    "        for i in codon_n:\n",
    "            codon_number = i\n",
    "            orig_base = original_base[change]\n",
    "            new_amino_acid = new_a_a[change]\n",
    "            n_base = new_base[change]\n",
    "            site_tri = site_trio[change]\n",
    "            info.append([codon_number, orig_base, n_base, site_tri, codon_n_912, original_a_a, new_amino_acid])\n",
    "            change = change+1\n",
    "        TP53_sites_missense5[a]=info\n",
    "        a=a+1\n",
    "    #     print(TP53_sites_missense5)\n",
    "\n",
    "    # Write in a way that can iterate over mutation rates (i.e. put in to form C[C>A]G etc)   \n",
    "    TP53_sites_missense6={}\n",
    "\n",
    "    separating_positions = {} # separate out all the possible changes to just 1 effective item in the list\n",
    "    a=0\n",
    "    for k, v in TP53_sites_missense5.items():\n",
    "        for i in v: #because some contain more than 1 change from that codon:\n",
    "            separating_positions[a]=i\n",
    "            a=a+1\n",
    "\n",
    "    for k, v in separating_positions.items():\n",
    "        base_pos = v[0]\n",
    "        original_base = v[1]\n",
    "        new_base = v[2]\n",
    "        site_trio = v[3]\n",
    "        codon_n_912 = v[4]\n",
    "        original_a_a = v[5]\n",
    "        new_a_a = v[6]\n",
    "        sequence = (str(original_base),str(new_base))\n",
    "        changes = '>'.join(sequence)\n",
    "        left_base = site_trio[0][0]\n",
    "        right_base = site_trio[0][2]\n",
    "        sequence2 = (changes, right_base)\n",
    "        changes2 = ']'.join(sequence2)\n",
    "        sequence3 = (left_base, changes2)\n",
    "        changes3 = '['.join(sequence3)\n",
    "        TP53_sites_missense6[k] = (v[0],changes3, codon_n_912, original_a_a, new_a_a) #site_trio = the affected base and its neighbours\n",
    "\n",
    "    #Assign the mutation rates to each site...\n",
    "    TP53_mutation_rates_missense = {}\n",
    "    for k, v in TP53_sites_missense6.items():\n",
    "        base_change_info = v[1]\n",
    "        if base_change_info in Mutation_rates.keys():\n",
    "            mutation_rate = Mutation_rates.get(base_change_info)\n",
    "            original_a_a = v[3]\n",
    "            new_a_a = v[4]\n",
    "            aa_positioning = v[2]\n",
    "            variant_id = str(original_a_a)+str(aa_positioning)+str(new_a_a)\n",
    "            TP53_mutation_rates_missense[k] = (v[0],v[1], mutation_rate, variant_id)\n",
    "\n",
    "    TP53_mutation_rate_missense_sum = []\n",
    "\n",
    "    mutation_rates_list = []\n",
    "    for k, v in TP53_mutation_rates_missense.items():\n",
    "        codon_position_in_gene=(v[0])\n",
    "        variant = v[3]\n",
    "        site = v[1]\n",
    "        mutation_rate_value = v[2] \n",
    "        mutation_rates_list.append((v[0], variant, site, mutation_rate_value))\n",
    "        TP53_mutation_rate_missense_sum.append(mutation_rate_value)\n",
    "\n",
    "    #Get the average of all the mutation rates\n",
    "    TP53_mutation_rate_missense_sum = []\n",
    "    for k, v in TP53_mutation_rates_missense.items():\n",
    "        mut_rate = v[2]\n",
    "        TP53_mutation_rate_missense_sum.append(mut_rate)\n",
    "\n",
    "    TP53_Summed_mutation_rate_missense_added = np.sum(TP53_mutation_rate_missense_sum)\n",
    "\n",
    "    print('Total summed mutation rate across TP53 missense for this study = ', TP53_Summed_mutation_rate_missense_added)\n",
    "\n",
    "    mutation_rates_list_all = mutation_rates_list\n",
    "    return mutation_rates_list_all # e.g. lower= 0, upper = 913 for whole gene\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Lists of variants targeted by each study"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- In order to calculate the number of times you expect to see a variant across all studies, you need to know which variants were call-able by which studies.\n",
    "- This can be calculated by using the mutation rate functions (above) as they will also produce a list of the variants within e.g. range of amino acid positions of interest or from a CSV file of variants."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Jaiswal 2014"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 179,
   "metadata": {
    "code_folding": [],
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total summed mutation rate across DNMT3A nonsense_NP_072046 for this study =  4.3738789e-07\n",
      "Total summed mutation rate across DNMT3A missense_NP_072046 for this study =  4.1418962309999996e-07\n"
     ]
    }
   ],
   "source": [
    "#Mutation rate across Jaiswal's DNMT3A variants\n",
    "Jaiswal_non_DNMT3A = mutation_rates_nonsense_range_of_variants_DNMT3A_NP072046(0, 913)\n",
    "Jaiswal_miss_DNMT3A = mutation_rates_list_missense_variants_DNMT3A_NP072046('Jaiswal DNMT3A missense variants called.csv')\n",
    "\n",
    "#Lists of DNMT3A variants targetable in Jaiswal 2014 (i.e. included in their panel)\n",
    "Jaiswal_nonsense_DNMT3A = []\n",
    "for i in Jaiswal_non_DNMT3A:\n",
    "    Jaiswal_nonsense_DNMT3A.append(str('p.')+i[1])\n",
    "    \n",
    "Jaiswal_missense_DNMT3A = []\n",
    "for i in Jaiswal_miss_DNMT3A:\n",
    "    Jaiswal_missense_DNMT3A.append(i[3])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 180,
   "metadata": {
    "code_folding": [],
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total summed mutation rate across TET2 nonsense_NP_01120680 for this study =  5.852107500000001e-07\n",
      "Total summed mutation rate across TET2 missense_NP_01120680 for this study =  1.1247945999999999e-08\n"
     ]
    }
   ],
   "source": [
    "#Mutation rate across Jaiswal's TET2 variants\n",
    "Jaiswal_non_TET2 = mutation_rates_nonsense_range_of_variants_TET2(0, 1137)\n",
    "Jaiswal_miss_TET2 = mutation_rates_list_missense_variants_TET2('Jaiswal TET2 missense variants called.csv')\n",
    "\n",
    "#Lists of TET2 variants targetable in Jaiswal 2014 (i.e. included in their panel)\n",
    "Jaiswal_nonsense_TET2 = []\n",
    "for i in Jaiswal_non_TET2:\n",
    "    Jaiswal_nonsense_TET2.append(str('p.')+i[1])\n",
    "    \n",
    "Jaiswal_missense_TET2 = []\n",
    "for i in Jaiswal_miss_TET2:\n",
    "    Jaiswal_missense_TET2.append(i[3])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 181,
   "metadata": {
    "code_folding": [],
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total summed mutation rate across ASXL1 nonsense for this study =  6.22373285e-07\n"
     ]
    }
   ],
   "source": [
    "#Mutation rate across Jaiswal's ASXL1 variants\n",
    "Jaiswal_non_ASXL1 = mutation_rates_nonsense_range_of_variants_ASXL1(0, 1542)\n",
    "\n",
    "#Lists of ASXL1 variants targetable in Jaiswal 2014 (i.e. included in their panel)\n",
    "Jaiswal_nonsense_ASXL1 = []\n",
    "for i in Jaiswal_non_ASXL1:\n",
    "    Jaiswal_nonsense_ASXL1.append(str('p.')+i[1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 182,
   "metadata": {
    "code_folding": [],
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total summed mutation rate across TP53 nonsense for this study =  1.60359228e-07\n",
      "Total summed mutation rate across TP53 missense for this study =  5.456417459999999e-07\n"
     ]
    }
   ],
   "source": [
    "#Mutation rate across Jaiswal's TP53 variants\n",
    "Jaiswal_non_TP53 = mutation_rates_nonsense_range_of_variants_TP53(0, 394)\n",
    "Jaiswal_miss_TP53 = mutation_rates_list_missense_variants_TP53('Jaiswal TP53 missense variants called.csv')\n",
    "\n",
    "#Lists of TP53 variants targetable in Jaiswal 2014 (i.e. included in their panel)\n",
    "Jaiswal_nonsense_TP53 = []\n",
    "for i in Jaiswal_non_TP53:\n",
    "    Jaiswal_nonsense_TP53.append(str('p.')+i[1])\n",
    "    \n",
    "Jaiswal_missense_TP53 = []\n",
    "for i in Jaiswal_miss_TP53:\n",
    "    Jaiswal_missense_TP53.append(i[3])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Genovese 2014"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 183,
   "metadata": {
    "code_folding": [
     0
    ],
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total summed mutation rate across DNMT3A nonsense_NP_072046 for this study =  3.0580037300000003e-07\n",
      "Total summed mutation rate across DNMT3A nonsense_NP_072046 for this study =  1.580314e-08\n",
      "Total summed mutation rate across DNMT3A missense_NP_072046 for this study =  5.0437521407e-06\n",
      "Total summed mutation rate across DNMT3A missense_NP_072046 for this study =  3.796521375e-07\n"
     ]
    }
   ],
   "source": [
    "#Mutation rate across Genovese's DNMT3A variants\n",
    "Genovese_nonsense_DNMT3A_1 = mutation_rates_nonsense_range_of_variants_DNMT3A_NP072046(213, 841)\n",
    "Genovese_nonsense_DNMT3A_2 = mutation_rates_nonsense_range_of_variants_DNMT3A_NP072046(866, 913)\n",
    "Genovese_missense_DNMT3A_1 = mutation_rates_missense_range_of_variants_DNMT3A_NP072046(213, 841)\n",
    "Genovese_missense_DNMT3A_2 = mutation_rates_missense_range_of_variants_DNMT3A_NP072046(866, 913)\n",
    "\n",
    "#Lists of DNMT3A variants targetable in Genovese 2014 (i.e. included in their panel)\n",
    "Genovese_nonsense_DNMT3A = []\n",
    "Genovese_missense_DNMT3A = []\n",
    "for i in Genovese_nonsense_DNMT3A_1:\n",
    "    Genovese_nonsense_DNMT3A.append(str('p.')+i[1])\n",
    "for i in Genovese_nonsense_DNMT3A_2:\n",
    "    Genovese_nonsense_DNMT3A.append(str('p.')+i[1])\n",
    "for i in Genovese_missense_DNMT3A_1:\n",
    "    Genovese_missense_DNMT3A.append(str('p.')+i[1])\n",
    "for i in Genovese_missense_DNMT3A_1:\n",
    "    Genovese_missense_DNMT3A.append(str('p.')+i[1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 184,
   "metadata": {
    "code_folding": [
     0,
     6
    ]
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total summed mutation rate across DNMT3A missense_NP_072046 for this study =  1.0806259e-08\n"
     ]
    }
   ],
   "source": [
    "#Mutation rate across Genovese's DNMT3A variants\n",
    "Genovese_missense_DNMT3A_partial_codons = mutation_rates_missense_range_of_variants_DNMT3A_NP072046(846, 848)\n",
    "base_position_list = [2540, 2541, 2542, 2543]\n",
    "Genovese_partial_missense = []\n",
    "Genovese_partial_missense_sum_values = []\n",
    "for i in Genovese_missense_DNMT3A_partial_codons:\n",
    "    if i[0] in base_position_list:\n",
    "        Genovese_partial_missense.append(i)\n",
    "        Genovese_partial_missense_sum_values.append(i[3])\n",
    "        \n",
    "Genovese_partial_missense_added = np.sum(Genovese_partial_missense_sum_values)\n",
    "\n",
    "#Lists of DNMT3A variants targetable in Genovese 2014 (i.e. included in their panel)\n",
    "for i in Genovese_partial_missense:\n",
    "    Genovese_missense_DNMT3A.append(str('p.'+i[1]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 185,
   "metadata": {
    "code_folding": [],
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total summed mutation rate across TET2 nonsense_NP_060098 for this study =  5.94141875e-07\n",
      "Total summed mutation rate across TET2 missense_NP_060098 for this study =  1.0182869000000002e-08\n"
     ]
    }
   ],
   "source": [
    "#Mutation rate across Genovese's TET2 variants\n",
    "Genovese_non_TET2 = mutation_rates_nonsense_range_of_variants_TET2_shorter(0, 1166)\n",
    "Genovese_miss_TET2 = mutation_rates_list_missense_variants_TET2_shorter('Genovese TET2 missense variants called.csv')\n",
    "\n",
    "#Lists of TET2 variants targetable in Genovese 2014 (i.e. included in their panel)\n",
    "Genovese_nonsense_TET2 = []\n",
    "for i in Genovese_non_TET2:\n",
    "    Genovese_nonsense_TET2.append(str('p.')+i[1])\n",
    "    \n",
    "Genovese_missense_TET2 = []\n",
    "for i in Genovese_miss_TET2:\n",
    "    Genovese_missense_TET2.append(i[3])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 186,
   "metadata": {
    "code_folding": [
     0
    ],
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total summed mutation rate across ASXL1 missense for this study =  4.9927510000000005e-09\n",
      "Total summed mutation rate across ASXL1 nonsense for this study =  3.9807642e-08\n",
      "Total summed mutation rate across ASXL1 nonsense for this study =  1.8562608e-08\n",
      "Total summed mutation rate across ASXL1 nonsense for this study =  2.1717856400000002e-07\n",
      "Total summed mutation rate across ASXL1 nonsense for this study =  3.21870573e-07\n",
      "Total summed mutation rate across ASXL1 nonsense for this study =  3.27627e-09\n",
      "Total summed mutation rate across ASXL1 nonsense for this study =  3.78679e-10\n"
     ]
    }
   ],
   "source": [
    "#Mutation rate across Genovese's ASXL1 variants\n",
    "Genovese_miss_ASXL1 = mutation_rates_list_missense_variants_ASXL1('Genovese ASXL1 missense variants called.csv')\n",
    "Genovese_non_ASXL1_1 = mutation_rates_nonsense_range_of_variants_ASXL1(19, 128)\n",
    "Genovese_non_ASXL1_2 = mutation_rates_nonsense_range_of_variants_ASXL1(143, 196)\n",
    "Genovese_non_ASXL1_3 = mutation_rates_nonsense_range_of_variants_ASXL1(200, 627)\n",
    "Genovese_non_ASXL1_4 = mutation_rates_nonsense_range_of_variants_ASXL1(655, 1541)\n",
    "Genovese_non_ASXL1_5 = mutation_rates_nonsense_range_of_variants_ASXL1(141, 143)\n",
    "Genovese_non_ASXL1_6 = mutation_rates_nonsense_range_of_variants_ASXL1(627, 629)\n",
    "\n",
    "#Lists of ASXL1 variants targetable in Genovese 2014 (i.e. included in their panel)\n",
    "Genovese_nonsense_ASXL1 = []\n",
    "for i in Genovese_non_ASXL1_1:\n",
    "    Genovese_nonsense_ASXL1.append(str('p.')+i[1])\n",
    "for i in Genovese_non_ASXL1_2:\n",
    "    Genovese_nonsense_ASXL1.append(str('p.')+i[1])\n",
    "for i in Genovese_non_ASXL1_3:\n",
    "    Genovese_nonsense_ASXL1.append(str('p.')+i[1])\n",
    "for i in Genovese_non_ASXL1_4:\n",
    "    Genovese_nonsense_ASXL1.append(str('p.')+i[1])\n",
    "for i in Genovese_non_ASXL1_5:\n",
    "    Genovese_nonsense_ASXL1.append(str('p.')+i[1])\n",
    "for i in Genovese_non_ASXL1_6:\n",
    "    Genovese_nonsense_ASXL1.append(str('p.')+i[1])\n",
    "    \n",
    "Genovese_missense_ASXL1 = []\n",
    "for i in Genovese_miss_ASXL1:\n",
    "    Genovese_missense_ASXL1.append(i[3])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 187,
   "metadata": {
    "code_folding": [
     0
    ],
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total summed mutation rate across TP53 nonsense for this study =  5.873716e-08\n",
      "Total summed mutation rate across TP53 missense for this study =  3.350155542e-07\n"
     ]
    }
   ],
   "source": [
    "#Mutation rate across Genovese's TP53 variants\n",
    "Genovese_non_TP53 = mutation_rates_list_nonsense_variants_TP53('Genovese TP53 nonsense variants called.csv')\n",
    "Genovese_miss_TP53 = mutation_rates_list_missense_variants_TP53('Genovese TP53 missense variants called.csv')\n",
    "\n",
    "#Lists of TP53 variants targetable in Genovese 2014 (i.e. included in their panel)\n",
    "Genovese_nonsense_TP53 = []\n",
    "for i in Genovese_non_TP53:\n",
    "    Genovese_nonsense_TP53.append(i[3])\n",
    "    \n",
    "Genovese_missense_TP53 = []\n",
    "for i in Genovese_miss_TP53:\n",
    "    Genovese_missense_TP53.append(i[3])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## McKerrel 2015"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 188,
   "metadata": {},
   "outputs": [],
   "source": [
    "#List of missense variants in DNMT3A called by McKerrel\n",
    "McKerrel_missense_DNMT3A = ['p.R882H', 'p.R882C', 'p.R882P', 'p.R882S', 'p.R882G', 'p.R882L']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Zink 2017"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 189,
   "metadata": {
    "code_folding": [],
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total summed mutation rate across DNMT3A nonsense_NP_072046 for this study =  4.3738789e-07\n",
      "Total summed mutation rate across DNMT3A missense_NP_072046 for this study =  9.851992712e-07\n"
     ]
    }
   ],
   "source": [
    "#Mutation rate across Zink's DNMT3A variants\n",
    "Zink_non_DNMT3A = mutation_rates_nonsense_range_of_variants_DNMT3A_NP072046(0, 913)\n",
    "Zink_miss_DNMT3A = mutation_rates_list_missense_variants_DNMT3A_NP072046('Zink DNMT3A missense variants in COSMIC.csv')\n",
    "\n",
    "#Lists of DNMT3A variants targetable in Zink 2017 (i.e. included in their panel)\n",
    "Zink_nonsense_DNMT3A = []\n",
    "for i in Zink_non_DNMT3A:\n",
    "    Zink_nonsense_DNMT3A.append(str('p.')+i[1])\n",
    "\n",
    "Zink_missense_DNMT3A = []\n",
    "for i in Zink_miss_DNMT3A:\n",
    "    Zink_missense_DNMT3A.append(i[3])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 190,
   "metadata": {
    "code_folding": [],
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total summed mutation rate across TET2 nonsense_NP_01120680 for this study =  9.77205187e-07\n",
      "Total summed mutation rate across TET2 missense_NP_01120680 for this study =  9.066063957e-07\n"
     ]
    }
   ],
   "source": [
    "#Mutation rate across Zink's TET2 variants\n",
    "Zink_non_TET2 = mutation_rates_nonsense_range_of_variants_TET2(0, 2003)\n",
    "Zink_miss_TET2 = mutation_rates_list_missense_variants_TET2('Zink TET2 missense variants in COSMIC.csv')\n",
    "\n",
    "#Lists of TET2 variants targetable in Zink 2017 (i.e. included in their panel)\n",
    "Zink_nonsense_TET2 = []\n",
    "for i in Zink_non_TET2:\n",
    "    Zink_nonsense_TET2.append(str('p.')+i[1])\n",
    "    \n",
    "Zink_missense_TET2 = []\n",
    "for i in Zink_miss_TET2:\n",
    "    Zink_missense_TET2.append(i[3])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 191,
   "metadata": {
    "code_folding": [],
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total summed mutation rate across ASXL1 nonsense for this study =  6.22373285e-07\n",
      "Total summed mutation rate across ASXL1 missense for this study =  3.277334481e-07\n"
     ]
    }
   ],
   "source": [
    "#Mutation rate across Zink's ASXL1 variants\n",
    "Zink_non_ASXL1 = mutation_rates_nonsense_range_of_variants_ASXL1(0, 1542)\n",
    "Zink_miss_ASXL1 = mutation_rates_list_missense_variants_ASXL1('Zink ASXL1 missense variants called.csv')\n",
    "\n",
    "#Lists of ASXL1 variants targetable in Zink 2017 (i.e. included in their panel)\n",
    "Zink_nonsense_ASXL1 = []\n",
    "for i in Zink_non_ASXL1:\n",
    "    Zink_nonsense_ASXL1.append(str('p.')+i[1])\n",
    "    \n",
    "Zink_missense_ASXL1 = []\n",
    "for i in Zink_miss_ASXL1:\n",
    "    Zink_missense_ASXL1.append(i[3])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 192,
   "metadata": {
    "code_folding": [],
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total summed mutation rate across TP53 nonsense for this study =  1.60359228e-07\n",
      "Total summed mutation rate across TP53 missense for this study =  1.2571066527e-06\n"
     ]
    }
   ],
   "source": [
    "#Mutation rate across Zink's TP53 variants\n",
    "Zink_non_TP53 = mutation_rates_nonsense_range_of_variants_TP53(0, 394)\n",
    "Zink_miss_TP53 = mutation_rates_list_missense_variants_TP53('Zink TP53 missense variants called.csv')\n",
    "\n",
    "#Lists of TP53 variants targetable in Zink 2017 (i.e. included in their panel)\n",
    "Zink_nonsense_TP53 = []\n",
    "for i in Zink_non_TP53:\n",
    "    Zink_nonsense_TP53.append(str('p.')+i[1])\n",
    "    \n",
    "Zink_missense_TP53 = []\n",
    "for i in Zink_miss_TP53:\n",
    "    Zink_missense_TP53.append(i[3])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Coombs 2017"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 193,
   "metadata": {
    "code_folding": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total summed mutation rate across DNMT3A nonsense_NP_072046 for this study =  4.3738789e-07\n",
      "Total summed mutation rate across DNMT3A missense_NP_072046 for this study =  7.571578417e-06\n"
     ]
    }
   ],
   "source": [
    "#Mutation rate across Coombs' DNMT3A variants\n",
    "Coombs_non_DNMT3A = mutation_rates_nonsense_range_of_variants_DNMT3A_NP072046(0, 913)\n",
    "Coombs_miss_DNMT3A = mutation_rates_missense_range_of_variants_DNMT3A_NP072046(0, 913)\n",
    "\n",
    "#Lists of DNMT3A variants targetable in Coombs 2017 (i.e. included in their panel)\n",
    "Coombs_nonsense_DNMT3A = []\n",
    "for i in Coombs_non_DNMT3A:\n",
    "    Coombs_nonsense_DNMT3A.append(str('p.')+i[1])\n",
    "    \n",
    "Coombs_missense_DNMT3A = []\n",
    "for i in Coombs_miss_DNMT3A:\n",
    "    Coombs_missense_DNMT3A.append(str('p.')+i[1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 194,
   "metadata": {
    "code_folding": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total summed mutation rate across TET2 nonsense_NP_01120680 for this study =  9.77205187e-07\n",
      "Total summed mutation rate across TET2 missense_NP_01120680 for this study =  1.25365212936e-05\n"
     ]
    }
   ],
   "source": [
    "#Mutation rate across Coombs' TET2 variants\n",
    "Coombs_non_TET2 = mutation_rates_nonsense_range_of_variants_TET2(0, 2003)\n",
    "Coombs_miss_TET2 = mutation_rates_missense_range_of_variants_TET2(0, 2003)\n",
    "\n",
    "#Lists of TET2 variants targetable in Coombs 2017 (i.e. included in their panel)\n",
    "Coombs_nonsense_TET2 = []\n",
    "for i in Coombs_non_TET2:\n",
    "    Coombs_nonsense_TET2.append(str('p.')+i[1])\n",
    "    \n",
    "Coombs_missense_TET2 = []\n",
    "for i in Coombs_miss_TET2:\n",
    "    Coombs_missense_TET2.append(str('p.')+i[1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 195,
   "metadata": {
    "code_folding": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total summed mutation rate across ASXL1 nonsense for this study =  6.22373285e-07\n",
      "Total summed mutation rate across ASXL1 missense for this study =  1.1103113451e-05\n"
     ]
    }
   ],
   "source": [
    "#Mutation rate across Coombs' ASXL1 variants\n",
    "Coombs_non_ASXL1 = mutation_rates_nonsense_range_of_variants_ASXL1(0, 1542)\n",
    "Coombs_miss_ASXL1 = mutation_rates_missense_range_of_variants_ASXL1(0, 1542)\n",
    "\n",
    "#Lists of ASXL1 variants targetable in Coombs 2017 (i.e. included in their panel)\n",
    "Coombs_nonsense_ASXL1 = []\n",
    "for i in Coombs_non_ASXL1:\n",
    "    Coombs_nonsense_ASXL1.append(str('p.')+i[1])\n",
    "    \n",
    "Coombs_missense_ASXL1 = []\n",
    "for i in Coombs_miss_ASXL1:\n",
    "    Coombs_missense_ASXL1.append(str('p.')+i[1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 196,
   "metadata": {
    "code_folding": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total summed mutation rate across TP53 nonsense for this study =  1.60359228e-07\n",
      "Total summed mutation rate across TP53 missense for this study =  3.0112256972e-06\n"
     ]
    }
   ],
   "source": [
    "#Mutation rate across Coombs' TP53 variants\n",
    "Coombs_non_TP53 = mutation_rates_nonsense_range_of_variants_TP53(0, 394)\n",
    "Coombs_miss_TP53 = mutation_rates_missense_range_of_variants_TP53(0, 394)\n",
    "\n",
    "#Lists of TP53 variants targetable in Coombs 2017 (i.e. included in their panel)\n",
    "Coombs_nonsense_TP53 = []\n",
    "for i in Coombs_non_TP53:\n",
    "    Coombs_nonsense_TP53.append(str('p.')+i[1])\n",
    "    \n",
    "Coombs_missense_TP53 = []\n",
    "for i in Coombs_miss_TP53:\n",
    "    Coombs_missense_TP53.append(str('p.')+i[1])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Young 2016 & 2019"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 197,
   "metadata": {
    "code_folding": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total summed mutation rate across DNMT3A nonsense_NP_072046 for this study =  4.3738789e-07\n",
      "Total summed mutation rate across DNMT3A missense_NP_072046 for this study =  7.571578417e-06\n"
     ]
    }
   ],
   "source": [
    "#Mutation rate across Young DNMT3A variants\n",
    "Young_non_DNMT3A = mutation_rates_nonsense_range_of_variants_DNMT3A_NP072046(0, 913)\n",
    "Young_miss_DNMT3A = mutation_rates_missense_range_of_variants_DNMT3A_NP072046(0, 913)\n",
    "\n",
    "#Lists of DNMT3A variants targetable in Young (i.e. included in their panel)\n",
    "Young_nonsense_DNMT3A = []\n",
    "for i in Young_non_DNMT3A:\n",
    "    Young_nonsense_DNMT3A.append(str('p.')+i[1])\n",
    "    \n",
    "Young_missense_DNMT3A = []\n",
    "for i in Young_miss_DNMT3A:\n",
    "    Young_missense_DNMT3A.append(str('p.')+i[1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 198,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total summed mutation rate across TET2 nonsense_NP_01120680 for this study =  9.77205187e-07\n",
      "Total summed mutation rate across TET2 missense_NP_01120680 for this study =  1.25365212936e-05\n",
      "Total summed mutation rate across TET2 missense_NP_01120680 for this study =  2.1975729000000004e-08\n"
     ]
    }
   ],
   "source": [
    "#Mutation rate across Young TET2 variants\n",
    "Young_non_TET2 = mutation_rates_nonsense_range_of_variants_TET2(0, 2003)\n",
    "Young_miss_TET2 = mutation_rates_missense_range_of_variants_TET2(0, 2003)\n",
    "\n",
    "#Lists of TET2 variants targetable in Young (i.e. included in their panel)\n",
    "Young_nonsense_TET2 = []\n",
    "for i in Young_non_TET2:\n",
    "    Young_nonsense_TET2.append(str('p.')+i[1])\n",
    "    \n",
    "Young_missense_TET2_all = []\n",
    "for i in Young_miss_TET2:\n",
    "    Young_missense_TET2_all.append(str('p.')+i[1])\n",
    "    \n",
    "#Excluded regions:\n",
    "excluded_TET2_Young=[]\n",
    "Young_TET2_excluded = mutation_rates_list_missense_variants_TET2('excluded_variants_Young_TET2.csv')\n",
    "for i in Young_TET2_excluded:\n",
    "    excluded_TET2_Young.append(i[3])\n",
    "\n",
    "Young_missense_TET2=[]\n",
    "for i in Young_missense_TET2_all:\n",
    "    if i not in Young_TET2_excluded:\n",
    "        Young_missense_TET2.append(i)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 199,
   "metadata": {
    "code_folding": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total summed mutation rate across ASXL1 nonsense for this study =  3.5900751800000003e-07\n",
      "Total summed mutation rate across ASXL1 missense for this study =  6.8303733942e-06\n",
      "Total summed mutation rate across ASXL1 missense for this study =  1.53910711e-08\n"
     ]
    }
   ],
   "source": [
    "#Mutation rate across Young ASXL1 variants\n",
    "Young_non_ASXL1 = mutation_rates_nonsense_range_of_variants_ASXL1(573, 1542)\n",
    "Young_miss_ASXL1 = mutation_rates_missense_range_of_variants_ASXL1(573, 1542)\n",
    "\n",
    "#Lists of ASXL1 variants targetable in Young (i.e. included in their panel)\n",
    "Young_nonsense_ASXL1 = []\n",
    "for i in Young_non_ASXL1:\n",
    "    Young_nonsense_ASXL1.append(str('p.')+i[1])\n",
    "    \n",
    "Young_missense_ASXL1_all = []\n",
    "for i in Young_miss_ASXL1:\n",
    "    Young_missense_ASXL1_all.append(str('p.')+i[1])\n",
    "    \n",
    "#Excluded regions:\n",
    "excluded_ASXL1_Young=[]\n",
    "Young_ASXL1_excluded = mutation_rates_list_missense_variants_ASXL1('excluded_variants_Young_ASXL1.csv')\n",
    "for i in Young_ASXL1_excluded:\n",
    "    excluded_ASXL1_Young.append(i[3])\n",
    "\n",
    "Young_missense_ASXL1=[]\n",
    "for i in Young_missense_ASXL1_all:\n",
    "    if i not in Young_ASXL1_excluded:\n",
    "        Young_missense_ASXL1.append(i)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 200,
   "metadata": {
    "code_folding": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total summed mutation rate across TP53 nonsense for this study =  1.60359228e-07\n",
      "Total summed mutation rate across TP53 missense for this study =  3.0112256972e-06\n",
      "Total summed mutation rate across TP53 missense for this study =  1.2712269999999998e-09\n"
     ]
    }
   ],
   "source": [
    "#Mutation rate across Young TP53 variants\n",
    "Young_non_TP53 = mutation_rates_nonsense_range_of_variants_TP53(0, 394)\n",
    "Young_miss_TP53 = mutation_rates_missense_range_of_variants_TP53(0, 394)\n",
    "\n",
    "#Lists of TP53 variants targetable in Young (i.e. included in their panel)\n",
    "Young_nonsense_TP53 = []\n",
    "for i in Young_non_TP53:\n",
    "    Young_nonsense_TP53.append(str('p.')+i[1])\n",
    "    \n",
    "Young_missense_TP53_all = []\n",
    "for i in Young_miss_TP53:\n",
    "    Young_missense_TP53_all.append(str('p.')+i[1])\n",
    "    \n",
    "#Excluded regions:\n",
    "excluded_TP53_Young=[]\n",
    "Young_TP53_excluded = mutation_rates_list_missense_variants_TP53('excluded_variants_Young_TP53.csv')\n",
    "for i in Young_TP53_excluded:\n",
    "    excluded_TP53_Young.append(i[3])\n",
    "\n",
    "Young_missense_TP53=[]\n",
    "for i in Young_missense_TP53_all:\n",
    "    if i not in Young_TP53_excluded:\n",
    "        Young_missense_TP53.append(i)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Desai 2018"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 201,
   "metadata": {
    "code_folding": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total summed mutation rate across DNMT3A nonsense_NP_072046 for this study =  4.3738789e-07\n",
      "Total summed mutation rate across DNMT3A missense_NP_072046 for this study =  7.571578417e-06\n"
     ]
    }
   ],
   "source": [
    "#Mutation rate across Desai DNMT3A variants\n",
    "Desai_non_DNMT3A = mutation_rates_nonsense_range_of_variants_DNMT3A_NP072046(0, 913)\n",
    "Desai_miss_DNMT3A = mutation_rates_missense_range_of_variants_DNMT3A_NP072046(0, 913)\n",
    "\n",
    "#Lists of DNMT3A variants targetable in Desai (i.e. included in their panel)\n",
    "Desai_nonsense_DNMT3A = []\n",
    "for i in Desai_non_DNMT3A:\n",
    "    Desai_nonsense_DNMT3A.append(str('p.')+i[1])\n",
    "    \n",
    "Desai_missense_DNMT3A = []\n",
    "for i in Desai_miss_DNMT3A:\n",
    "    Desai_missense_DNMT3A.append(str('p.')+i[1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 202,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total summed mutation rate across TET2 nonsense_NP_01120680 for this study =  9.77205187e-07\n",
      "Total summed mutation rate across TET2 missense_NP_01120680 for this study =  1.25365212936e-05\n",
      "Total summed mutation rate across TET2 missense_NP_01120680 for this study =  5.2565262e-08\n"
     ]
    }
   ],
   "source": [
    "#Mutation rate across Desai TET2 variants\n",
    "Desai_non_TET2 = mutation_rates_nonsense_range_of_variants_TET2(0, 2003)\n",
    "Desai_miss_TET2 = mutation_rates_missense_range_of_variants_TET2(0, 2003)\n",
    "\n",
    "#Lists of TET2 variants targetable in Desai (i.e. included in their panel)\n",
    "Desai_nonsense_TET2 = []\n",
    "for i in Desai_non_TET2:\n",
    "    Desai_nonsense_TET2.append(str('p.')+i[1])\n",
    "    \n",
    "Desai_missense_TET2_all = []\n",
    "for i in Desai_miss_TET2:\n",
    "    Desai_missense_TET2_all.append(str('p.')+i[1])\n",
    "    \n",
    "#Excluded regions:\n",
    "excluded_TET2_Desai=[]\n",
    "Desai_TET2_excluded = mutation_rates_list_missense_variants_TET2('excluded_variants_Desai_2018_TET2.csv')\n",
    "for i in Desai_TET2_excluded:\n",
    "    excluded_TET2_Desai.append(i[3])\n",
    "\n",
    "Desai_missense_TET2=[]\n",
    "for i in Desai_missense_TET2_all:\n",
    "    if i not in Desai_TET2_excluded:\n",
    "        Desai_missense_TET2.append(i)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 203,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total summed mutation rate across ASXL1 nonsense for this study =  6.22373285e-07\n",
      "Total summed mutation rate across ASXL1 missense for this study =  1.1103113451e-05\n",
      "Total summed mutation rate across ASXL1 missense for this study =  2.0795553099999998e-08\n",
      "Total summed mutation rate across ASXL1 nonsense for this study =  1.0517e-09\n"
     ]
    }
   ],
   "source": [
    "#Mutation rate across Desai ASXL1 variants\n",
    "Desai_non_ASXL1 = mutation_rates_nonsense_range_of_variants_ASXL1(0, 1542)\n",
    "Desai_miss_ASXL1 = mutation_rates_missense_range_of_variants_ASXL1(0, 1542)\n",
    "\n",
    "#Lists of ASXL1 variants targetable in Desai (i.e. included in their panel)\n",
    "Desai_nonsense_ASXL1_all = []\n",
    "for i in Desai_non_ASXL1:\n",
    "    Desai_nonsense_ASXL1_all.append(str('p.')+i[1])\n",
    "    \n",
    "Desai_missense_ASXL1_all = []\n",
    "for i in Desai_miss_ASXL1:\n",
    "    Desai_missense_ASXL1_all.append(str('p.')+i[1])\n",
    "    \n",
    "#Excluded regions:\n",
    "excluded_ASXL1_Desai_missense=[]\n",
    "Desai_ASXL1_excluded_missense = mutation_rates_list_missense_variants_ASXL1('excluded_variants_Desai_2018_ASXL1_missense.csv')\n",
    "for i in Desai_ASXL1_excluded_missense:\n",
    "    excluded_ASXL1_Desai_missense.append(i[3])\n",
    "\n",
    "Desai_missense_ASXL1=[]\n",
    "for i in Desai_missense_ASXL1_all:\n",
    "    if i not in Desai_ASXL1_excluded_missense:\n",
    "        Desai_missense_ASXL1.append(i)\n",
    "        \n",
    "excluded_ASXL1_Desai_nonsense=[]\n",
    "Desai_ASXL1_excluded_nonsense = mutation_rates_list_nonsense_variants_ASXL1('excluded_variants_Desai_2018_ASXL1_nonsense.csv')\n",
    "for i in Desai_ASXL1_excluded_nonsense:\n",
    "    excluded_ASXL1_Desai_nonsense.append(i[3])\n",
    "\n",
    "Desai_nonsense_ASXL1=[]\n",
    "for i in Desai_nonsense_ASXL1_all:\n",
    "    if i not in Desai_ASXL1_excluded_nonsense:\n",
    "        Desai_nonsense_ASXL1.append(i)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 204,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total summed mutation rate across TP53 nonsense for this study =  1.60359228e-07\n",
      "Total summed mutation rate across TP53 missense for this study =  3.0112256972e-06\n",
      "Total summed mutation rate across TP53 missense for this study =  4.734506e-09\n"
     ]
    }
   ],
   "source": [
    "#Mutation rate across Desai TP53 variants\n",
    "Desai_non_TP53 = mutation_rates_nonsense_range_of_variants_TP53(0, 394)\n",
    "Desai_miss_TP53 = mutation_rates_missense_range_of_variants_TP53(0, 394)\n",
    "\n",
    "#Lists of TP53 variants targetable in Desai (i.e. included in their panel)\n",
    "Desai_nonsense_TP53 = []\n",
    "for i in Desai_non_TP53:\n",
    "    Desai_nonsense_TP53.append(str('p.')+i[1])\n",
    "    \n",
    "Desai_missense_TP53_all = []\n",
    "for i in Desai_miss_TP53:\n",
    "    Desai_missense_TP53_all.append(str('p.')+i[1])\n",
    "    \n",
    "#Excluded regions:\n",
    "excluded_TP53_Desai=[]\n",
    "Desai_TP53_excluded = mutation_rates_list_missense_variants_TP53('excluded_variants_Desai_2018_TP53.csv')\n",
    "for i in Desai_TP53_excluded:\n",
    "    excluded_TP53_Desai.append(i[3])\n",
    "\n",
    "Desai_missense_TP53=[]\n",
    "for i in Desai_missense_TP53_all:\n",
    "    if i not in Desai_TP53_excluded:\n",
    "        Desai_missense_TP53.append(i)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Acuna-Hidalgo 2017"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 205,
   "metadata": {
    "code_folding": [
     0
    ]
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total summed mutation rate across DNMT3A nonsense_NP_072046 for this study =  5.679932e-09\n",
      "Total summed mutation rate across DNMT3A nonsense_NP_072046 for this study =  1.0495078000000002e-08\n",
      "Total summed mutation rate across DNMT3A nonsense_NP_072046 for this study =  1.567588e-09\n",
      "Total summed mutation rate across DNMT3A nonsense_NP_072046 for this study =  2.075691e-08\n",
      "Total summed mutation rate across DNMT3A nonsense_NP_072046 for this study =  1.1075570000000001e-09\n",
      "Total summed mutation rate across DNMT3A nonsense_NP_072046 for this study =  1.7038651e-08\n"
     ]
    }
   ],
   "source": [
    "#Mutation rate across Acuna-Hidalgo DNMT3A variants\n",
    "Acuna_complete_nonsense_DNMT3A_1 = mutation_rates_nonsense_range_of_variants_DNMT3A_NP072046(874,892)\n",
    "Acuna_complete_nonsense_DNMT3A_2 = mutation_rates_nonsense_range_of_variants_DNMT3A_NP072046(851,866)\n",
    "Acuna_complete_nonsense_DNMT3A_3 = mutation_rates_nonsense_range_of_variants_DNMT3A_NP072046(849,857)\n",
    "Acuna_complete_nonsense_DNMT3A_4 = mutation_rates_nonsense_range_of_variants_DNMT3A_NP072046(758,774)\n",
    "Acuna_complete_nonsense_DNMT3A_5 = mutation_rates_nonsense_range_of_variants_DNMT3A_NP072046(733,740)\n",
    "Acuna_complete_nonsense_DNMT3A_6 = mutation_rates_nonsense_range_of_variants_DNMT3A_NP072046(317,334)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 206,
   "metadata": {
    "code_folding": [
     0
    ]
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total summed mutation rate across DNMT3A missense_NP_072046 for this study =  1.482275953e-07\n",
      "Total summed mutation rate across DNMT3A missense_NP_072046 for this study =  9.026848400000001e-08\n",
      "Total summed mutation rate across DNMT3A missense_NP_072046 for this study =  4.58192721e-08\n",
      "Total summed mutation rate across DNMT3A missense_NP_072046 for this study =  1.1447460200000002e-07\n",
      "Total summed mutation rate across DNMT3A missense_NP_072046 for this study =  6.144157100000001e-08\n",
      "Total summed mutation rate across DNMT3A missense_NP_072046 for this study =  1.579131871e-07\n"
     ]
    }
   ],
   "source": [
    "#Mutation rate across Acuna-Hidalgo DNMT3A variants\n",
    "Acuna_complete_missense_DNMT3A_1 = mutation_rates_missense_range_of_variants_DNMT3A_NP072046(874,892)\n",
    "Acuna_complete_missense_DNMT3A_2 = mutation_rates_missense_range_of_variants_DNMT3A_NP072046(851,866)\n",
    "Acuna_complete_missense_DNMT3A_3 = mutation_rates_missense_range_of_variants_DNMT3A_NP072046(849,857)\n",
    "Acuna_complete_missense_DNMT3A_4 = mutation_rates_missense_range_of_variants_DNMT3A_NP072046(758,774)\n",
    "Acuna_complete_missense_DNMT3A_5 = mutation_rates_missense_range_of_variants_DNMT3A_NP072046(733,740)\n",
    "Acuna_complete_missense_DNMT3A_6 = mutation_rates_missense_range_of_variants_DNMT3A_NP072046(317,334)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 207,
   "metadata": {
    "code_folding": []
   },
   "outputs": [],
   "source": [
    "#Lists of DNMT3A variants targetable in Acunda-Hidalgo (i.e. included in their panel)\n",
    "Acuna_nonsense_DNMT3A = []\n",
    "for i in Acuna_complete_nonsense_DNMT3A_1:\n",
    "    Acuna_nonsense_DNMT3A.append(str('p.')+i[1])\n",
    "for i in Acuna_complete_nonsense_DNMT3A_2:\n",
    "    Acuna_nonsense_DNMT3A.append(str('p.')+i[1])\n",
    "for i in Acuna_complete_nonsense_DNMT3A_3:\n",
    "    Acuna_nonsense_DNMT3A.append(str('p.')+i[1])\n",
    "for i in Acuna_complete_nonsense_DNMT3A_4:\n",
    "    Acuna_nonsense_DNMT3A.append(str('p.')+i[1])\n",
    "for i in Acuna_complete_nonsense_DNMT3A_5:\n",
    "    Acuna_nonsense_DNMT3A.append(str('p.')+i[1])\n",
    "for i in Acuna_complete_nonsense_DNMT3A_6:\n",
    "    Acuna_nonsense_DNMT3A.append(str('p.')+i[1])\n",
    "    \n",
    "Acuna_missense_DNMT3A = []\n",
    "for i in Acuna_complete_missense_DNMT3A_1:\n",
    "    Acuna_missense_DNMT3A.append(str('p.')+i[1])\n",
    "for i in Acuna_complete_missense_DNMT3A_2:\n",
    "    Acuna_missense_DNMT3A.append(str('p.')+i[1])\n",
    "for i in Acuna_complete_missense_DNMT3A_3:\n",
    "    Acuna_missense_DNMT3A.append(str('p.')+i[1])\n",
    "for i in Acuna_complete_missense_DNMT3A_4:\n",
    "    Acuna_missense_DNMT3A.append(str('p.')+i[1])\n",
    "for i in Acuna_complete_missense_DNMT3A_5:\n",
    "    Acuna_missense_DNMT3A.append(str('p.')+i[1])\n",
    "for i in Acuna_complete_missense_DNMT3A_6:\n",
    "    Acuna_missense_DNMT3A.append(str('p.')+i[1])\n",
    "    \n",
    "for i in ['p.Y874S', 'p.Y874F', 'p.Y874C', 'p.F851Y', 'p.F851S', 'p.F851C', 'p.F851L', 'p.I858L', 'p.I858F', 'p.I858V', 'p.I858N', 'p.I858T', 'p.I858S',\\\n",
    "'p.A741T', 'p.A741P', 'p.A741S', 'p.A741E', 'p.A741V', 'p.A741G', 'p.E733A', 'p.E733V', 'p.E733G', 'p.E733D']:\n",
    "    Acuna_missense_DNMT3A.append(i)\n",
    "    \n",
    "for i in ['p.Y874*', 'p.E733*']:\n",
    "    Acuna_nonsense_DNMT3A.append(i)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 208,
   "metadata": {
    "code_folding": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total summed mutation rate across TET2 nonsense_NP_01120680 for this study =  8.039082e-09\n",
      "Total summed mutation rate across TET2 nonsense_NP_01120680 for this study =  4.007699e-09\n",
      "Total summed mutation rate across TET2 nonsense_NP_01120680 for this study =  1.368038e-08\n",
      "Total summed mutation rate across TET2 nonsense_NP_01120680 for this study =  1.2193169e-08\n",
      "Total summed mutation rate across TET2 nonsense_NP_01120680 for this study =  4.1087400000000004e-09\n",
      "Total summed mutation rate across TET2 nonsense_NP_01120680 for this study =  9.706974e-09\n"
     ]
    }
   ],
   "source": [
    "#Mutation rate across Acuna-Hidalgo TET2 variants\n",
    "Acuna_complete_nonsense_TET2_1 = mutation_rates_nonsense_range_of_variants_TET2(527, 534)\n",
    "Acuna_complete_nonsense_TET2_2 = mutation_rates_nonsense_range_of_variants_TET2(541, 546)\n",
    "Acuna_complete_nonsense_TET2_3 = mutation_rates_nonsense_range_of_variants_TET2(547, 552)\n",
    "Acuna_complete_nonsense_TET2_4 = mutation_rates_nonsense_range_of_variants_TET2(754, 773)\n",
    "Acuna_complete_nonsense_TET2_5 = mutation_rates_nonsense_range_of_variants_TET2(844, 849)\n",
    "Acuna_complete_nonsense_TET2_6 = mutation_rates_nonsense_range_of_variants_TET2(885, 890)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 209,
   "metadata": {
    "code_folding": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total summed mutation rate across TET2 missense_NP_01120680 for this study =  4.158436819999999e-08\n",
      "Total summed mutation rate across TET2 missense_NP_01120680 for this study =  3.2290779e-08\n",
      "Total summed mutation rate across TET2 missense_NP_01120680 for this study =  3.0273943e-08\n",
      "Total summed mutation rate across TET2 missense_NP_01120680 for this study =  1.0329367619999999e-07\n",
      "Total summed mutation rate across TET2 missense_NP_01120680 for this study =  2.6732733099999998e-08\n",
      "Total summed mutation rate across TET2 missense_NP_01120680 for this study =  2.21783323e-08\n"
     ]
    }
   ],
   "source": [
    "#Mutation rate across Acuna-Hidalgo TET2 variants\n",
    "Acuna_complete_missense_TET2_1 = mutation_rates_missense_range_of_variants_TET2(527, 534)\n",
    "Acuna_complete_missense_TET2_2 = mutation_rates_missense_range_of_variants_TET2(541, 546)\n",
    "Acuna_complete_missense_TET2_3 = mutation_rates_missense_range_of_variants_TET2(547, 552)\n",
    "Acuna_complete_missense_TET2_4 = mutation_rates_missense_range_of_variants_TET2(754, 773)\n",
    "Acuna_complete_missense_TET2_5 = mutation_rates_missense_range_of_variants_TET2(844, 849)\n",
    "Acuna_complete_missense_TET2_6 = mutation_rates_missense_range_of_variants_TET2(885, 890)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 210,
   "metadata": {
    "code_folding": []
   },
   "outputs": [],
   "source": [
    "#Lists of TET2 variants targetable in Acunda-Hidalgo (i.e. included in their panel)\n",
    "Acuna_nonsense_TET2 = []\n",
    "for i in Acuna_complete_nonsense_TET2_1:\n",
    "    Acuna_nonsense_TET2.append(str('p.')+i[1])\n",
    "for i in Acuna_complete_nonsense_TET2_2:\n",
    "    Acuna_nonsense_TET2.append(str('p.')+i[1])\n",
    "for i in Acuna_complete_nonsense_TET2_3:\n",
    "    Acuna_nonsense_TET2.append(str('p.')+i[1])\n",
    "for i in Acuna_complete_nonsense_TET2_4:\n",
    "    Acuna_nonsense_TET2.append(str('p.')+i[1])\n",
    "for i in Acuna_complete_nonsense_TET2_5:\n",
    "    Acuna_nonsense_TET2.append(str('p.')+i[1])\n",
    "for i in Acuna_complete_nonsense_TET2_6:\n",
    "    Acuna_nonsense_TET2.append(str('p.')+i[1])\n",
    "    \n",
    "Acuna_missense_TET2 = []\n",
    "for i in Acuna_complete_missense_TET2_1:\n",
    "    Acuna_missense_TET2.append(str('p.')+i[1])\n",
    "for i in Acuna_complete_missense_TET2_2:\n",
    "    Acuna_missense_TET2.append(str('p.')+i[1])\n",
    "for i in Acuna_complete_missense_TET2_3:\n",
    "    Acuna_missense_TET2.append(str('p.')+i[1])\n",
    "for i in Acuna_complete_missense_TET2_4:\n",
    "    Acuna_missense_TET2.append(str('p.')+i[1])\n",
    "for i in Acuna_complete_missense_TET2_5:\n",
    "    Acuna_missense_TET2.append(str('p.')+i[1])\n",
    "for i in Acuna_complete_missense_TET2_6:\n",
    "    Acuna_missense_TET2.append(str('p.')+i[1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 211,
   "metadata": {
    "code_folding": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total summed mutation rate across ASXL1 nonsense for this study =  2.738905e-08\n",
      "Total summed mutation rate across ASXL1 nonsense for this study =  5.3283495e-08\n",
      "Total summed mutation rate across ASXL1 nonsense for this study =  1.3687139e-08\n",
      "Total summed mutation rate across ASXL1 nonsense for this study =  2.314861e-08\n",
      "Total summed mutation rate across ASXL1 nonsense for this study =  6.198635e-09\n"
     ]
    }
   ],
   "source": [
    "#Mutation rate across Acuna-Hidalgo ASXL1 variants\n",
    "Acuna_complete_nonsense_ASXL1_1 = mutation_rates_nonsense_range_of_variants_ASXL1(400, 407)\n",
    "Acuna_complete_nonsense_ASXL1_2 = mutation_rates_nonsense_range_of_variants_ASXL1(413, 573)\n",
    "Acuna_complete_nonsense_ASXL1_3 = mutation_rates_nonsense_range_of_variants_ASXL1(572, 596)\n",
    "Acuna_complete_nonsense_ASXL1_4 = mutation_rates_nonsense_range_of_variants_ASXL1(688, 694)\n",
    "Acuna_complete_nonsense_ASXL1_5 = mutation_rates_nonsense_range_of_variants_ASXL1(799, 804)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 212,
   "metadata": {
    "code_folding": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total summed mutation rate across ASXL1 missense for this study =  4.93893791e-08\n",
      "Total summed mutation rate across ASXL1 missense for this study =  1.1371639964e-06\n",
      "Total summed mutation rate across ASXL1 missense for this study =  1.624548922e-07\n",
      "Total summed mutation rate across ASXL1 missense for this study =  3.0752606199999994e-08\n",
      "Total summed mutation rate across ASXL1 missense for this study =  3.3298389e-08\n"
     ]
    }
   ],
   "source": [
    "#Mutation rate across Acuna-Hidalgo ASXL1 variants\n",
    "Acuna_complete_missense_ASXL1_1 = mutation_rates_missense_range_of_variants_ASXL1(400, 407)\n",
    "Acuna_complete_missense_ASXL1_2 = mutation_rates_missense_range_of_variants_ASXL1(413, 573)\n",
    "Acuna_complete_missense_ASXL1_3 = mutation_rates_missense_range_of_variants_ASXL1(572, 596)\n",
    "Acuna_complete_missense_ASXL1_4 = mutation_rates_missense_range_of_variants_ASXL1(688, 694)\n",
    "Acuna_complete_missense_ASXL1_5 = mutation_rates_missense_range_of_variants_ASXL1(799, 804)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 213,
   "metadata": {
    "code_folding": [
     0
    ]
   },
   "outputs": [],
   "source": [
    "#Lists of ASXL1 variants targetable in Acunda-Hidalgo (i.e. included in their panel)\n",
    "Acuna_nonsense_ASXL1 = []\n",
    "for i in Acuna_complete_nonsense_ASXL1_1:\n",
    "    Acuna_nonsense_ASXL1.append(str('p.')+i[1])\n",
    "for i in Acuna_complete_nonsense_ASXL1_2:\n",
    "    Acuna_nonsense_ASXL1.append(str('p.')+i[1])\n",
    "for i in Acuna_complete_nonsense_ASXL1_3:\n",
    "    Acuna_nonsense_ASXL1.append(str('p.')+i[1])\n",
    "for i in Acuna_complete_nonsense_ASXL1_4:\n",
    "    Acuna_nonsense_ASXL1.append(str('p.')+i[1])\n",
    "for i in Acuna_complete_nonsense_ASXL1_5:\n",
    "    Acuna_nonsense_ASXL1.append(str('p.')+i[1])\n",
    "    \n",
    "Acuna_missense_ASXL1 = []\n",
    "for i in Acuna_complete_missense_ASXL1_1:\n",
    "    Acuna_missense_ASXL1.append(str('p.')+i[1])\n",
    "for i in Acuna_complete_missense_ASXL1_2:\n",
    "    Acuna_missense_ASXL1.append(str('p.')+i[1])\n",
    "for i in Acuna_complete_missense_ASXL1_3:\n",
    "    Acuna_missense_ASXL1.append(str('p.')+i[1])\n",
    "for i in Acuna_complete_missense_ASXL1_4:\n",
    "    Acuna_missense_ASXL1.append(str('p.')+i[1])\n",
    "for i in Acuna_complete_missense_ASXL1_5:\n",
    "    Acuna_missense_ASXL1.append(str('p.')+i[1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 214,
   "metadata": {
    "code_folding": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total summed mutation rate across TP53 nonsense for this study =  1.5439768000000003e-08\n",
      "Total summed mutation rate across TP53 nonsense for this study =  5.197881e-09\n",
      "Total summed mutation rate across TP53 nonsense for this study =  1.7852187e-08\n",
      "Total summed mutation rate across TP53 nonsense for this study =  1.469962e-09\n"
     ]
    }
   ],
   "source": [
    "#Mutation rate across Acuna-Hidalgo TP53 variants\n",
    "Acuna_complete_nonsense_TP53_1 = mutation_rates_nonsense_range_of_variants_TP53(185, 221)\n",
    "Acuna_complete_nonsense_TP53_2 = mutation_rates_nonsense_range_of_variants_TP53(168, 185)\n",
    "Acuna_complete_nonsense_TP53_3 = mutation_rates_nonsense_range_of_variants_TP53(117, 146)\n",
    "Acuna_complete_nonsense_TP53_4 = mutation_rates_nonsense_range_of_variants_TP53(59, 81)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 215,
   "metadata": {
    "code_folding": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total summed mutation rate across TP53 missense for this study =  2.514175651e-07\n",
      "Total summed mutation rate across TP53 missense for this study =  1.9618538400000002e-07\n",
      "Total summed mutation rate across TP53 missense for this study =  2.129085392e-07\n",
      "Total summed mutation rate across TP53 missense for this study =  2.2091162200000002e-07\n"
     ]
    }
   ],
   "source": [
    "#Mutation rate across Acuna-Hidalgo TP53 variants\n",
    "Acuna_complete_missense_TP53_1 = mutation_rates_missense_range_of_variants_TP53(185, 221)\n",
    "Acuna_complete_missense_TP53_2 = mutation_rates_missense_range_of_variants_TP53(168, 185)\n",
    "Acuna_complete_missense_TP53_3 = mutation_rates_missense_range_of_variants_TP53(117, 146)\n",
    "Acuna_complete_missense_TP53_4 = mutation_rates_missense_range_of_variants_TP53(59, 81)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 216,
   "metadata": {
    "code_folding": [
     0
    ]
   },
   "outputs": [],
   "source": [
    "#Lists of TP53 variants targetable in Acunda-Hidalgo (i.e. included in their panel)\n",
    "Acuna_nonsense_TP53 = []\n",
    "for i in Acuna_complete_nonsense_TP53_1:\n",
    "    Acuna_nonsense_TP53.append(str('p.')+i[1])\n",
    "for i in Acuna_complete_nonsense_TP53_2:\n",
    "    Acuna_nonsense_TP53.append(str('p.')+i[1])\n",
    "for i in Acuna_complete_nonsense_TP53_3:\n",
    "    Acuna_nonsense_TP53.append(str('p.')+i[1])\n",
    "for i in Acuna_complete_nonsense_TP53_4:\n",
    "    Acuna_nonsense_TP53.append(str('p.')+i[1])\n",
    "    \n",
    "Acuna_missense_TP53_all = []\n",
    "for i in Acuna_complete_missense_TP53_1:\n",
    "    Acuna_missense_TP53_all.append(str('p.')+i[1])\n",
    "for i in Acuna_complete_missense_TP53_2:\n",
    "    Acuna_missense_TP53_all.append(str('p.')+i[1])\n",
    "for i in Acuna_complete_missense_TP53_3:\n",
    "    Acuna_missense_TP53_all.append(str('p.')+i[1])\n",
    "for i in Acuna_complete_missense_TP53_4:\n",
    "    Acuna_missense_TP53_all.append(str('p.')+i[1])\n",
    "    \n",
    "#Excluded regions:\n",
    "Acuna_TP53_excluded=['p.P72R']\n",
    "\n",
    "Acuna_missense_TP53=[]\n",
    "for i in Acuna_missense_TP53_all:\n",
    "    if i not in Acuna_TP53_excluded:\n",
    "        Acuna_missense_TP53.append(i)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Lists of all possible variants in DNMT3A, TET2, ASXL1, TP53"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- Next, create lists of ALL possible variants in DNMT3A, TET2, ASXL1, TP53 so can look at each of them in turn and see how often they are observed and then use this to calculate what fitness effect would explain this number of observations.\n",
    "- The mutation rate calculator functions also produce a list of all the possible variants so use this function to generate a list of all the possible variants within each gene."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## DNMT3A"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 217,
   "metadata": {
    "code_folding": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total summed mutation rate across DNMT3A nonsense_NP_072046 for this study =  4.3738789e-07\n",
      "Total summed mutation rate across DNMT3A missense_NP_072046 for this study =  7.571578417e-06\n"
     ]
    }
   ],
   "source": [
    "#Mutation rate across DNMT3A\n",
    "all_nonsense_DNMT3A = mutation_rates_nonsense_range_of_variants_DNMT3A_NP072046(0, 913)\n",
    "all_missense_DNMT3A = mutation_rates_missense_range_of_variants_DNMT3A_NP072046(0, 913)\n",
    "\n",
    "#List of possible variants\n",
    "all_nonsense_variants_DNMT3A = []\n",
    "for i in all_nonsense_DNMT3A:\n",
    "    all_nonsense_variants_DNMT3A.append(str('p.')+i[1])\n",
    "    \n",
    "all_missense_variants_DNMT3A = []\n",
    "for i in all_missense_DNMT3A:\n",
    "    all_missense_variants_DNMT3A.append(str('p.')+i[1])\n",
    "\n",
    "all_missense_variants_DNMT3A = list(dict.fromkeys(all_missense_variants_DNMT3A))\n",
    "all_nonsense_variants_DNMT3A = list(dict.fromkeys(all_nonsense_variants_DNMT3A))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## TET2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 218,
   "metadata": {
    "code_folding": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total summed mutation rate across TET2 nonsense_NP_01120680 for this study =  9.77205187e-07\n",
      "Total summed mutation rate across TET2 missense_NP_01120680 for this study =  1.25365212936e-05\n"
     ]
    }
   ],
   "source": [
    "#Mutation rate across TET2\n",
    "all_nonsense_TET2 = mutation_rates_nonsense_range_of_variants_TET2(0, 2003)\n",
    "all_missense_TET2 = mutation_rates_missense_range_of_variants_TET2(0, 2003)\n",
    "\n",
    "#List of possible variants\n",
    "all_nonsense_variants_TET2 = []\n",
    "for i in all_nonsense_TET2:\n",
    "    all_nonsense_variants_TET2.append(str('p.')+i[1])\n",
    "    \n",
    "all_missense_variants_TET2 = []\n",
    "for i in all_missense_TET2:\n",
    "    all_missense_variants_TET2.append(str('p.')+i[1])\n",
    "\n",
    "all_missense_variants_TET2 = list(dict.fromkeys(all_missense_variants_TET2))\n",
    "all_nonsense_variants_TET2 = list(dict.fromkeys(all_nonsense_variants_TET2))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## ASXL1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 219,
   "metadata": {
    "code_folding": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total summed mutation rate across ASXL1 nonsense for this study =  6.22373285e-07\n",
      "Total summed mutation rate across ASXL1 missense for this study =  1.1103113451e-05\n"
     ]
    }
   ],
   "source": [
    "#Mutation rate across TET2\n",
    "all_nonsense_ASXL1 = mutation_rates_nonsense_range_of_variants_ASXL1(0, 1542)\n",
    "all_missense_ASXL1 = mutation_rates_missense_range_of_variants_ASXL1(0, 1542)\n",
    "\n",
    "#List of possible variants\n",
    "all_nonsense_variants_ASXL1 = []\n",
    "for i in all_nonsense_ASXL1:\n",
    "    all_nonsense_variants_ASXL1.append(str('p.')+i[1])\n",
    "    \n",
    "all_missense_variants_ASXL1 = []\n",
    "for i in all_missense_ASXL1:\n",
    "    all_missense_variants_ASXL1.append(str('p.')+i[1])\n",
    "\n",
    "all_missense_variants_ASXL1 = list(dict.fromkeys(all_missense_variants_ASXL1))\n",
    "all_nonsense_variants_ASXL1 = list(dict.fromkeys(all_nonsense_variants_ASXL1))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## TP53"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 220,
   "metadata": {
    "code_folding": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total summed mutation rate across TP53 nonsense for this study =  1.60359228e-07\n",
      "Total summed mutation rate across TP53 missense for this study =  3.0112256972e-06\n"
     ]
    }
   ],
   "source": [
    "#Mutation rate across TP53\n",
    "all_nonsense_TP53 = mutation_rates_nonsense_range_of_variants_TP53(0, 394)\n",
    "all_missense_TP53 = mutation_rates_missense_range_of_variants_TP53(0, 394)\n",
    "\n",
    "#List of possible variants\n",
    "all_nonsense_variants_TP53 = []\n",
    "for i in all_nonsense_TP53:\n",
    "    all_nonsense_variants_TP53.append(str('p.')+i[1])\n",
    "    \n",
    "all_missense_variants_TP53 = []\n",
    "for i in all_missense_TP53:\n",
    "    all_missense_variants_TP53.append(str('p.')+i[1])\n",
    "\n",
    "all_missense_variants_TP53 = list(dict.fromkeys(all_missense_variants_TP53))\n",
    "all_nonsense_variants_TP53 = list(dict.fromkeys(all_nonsense_variants_TP53))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Actual number of observations of each variant"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- Look at each variant seen across the studies and count how many times it was observed."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### DNMT3A"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 221,
   "metadata": {
    "code_folding": [
     0
    ]
   },
   "outputs": [],
   "source": [
    "#Create lists of variants by studies and counts of each variant\n",
    "import csv\n",
    "filename=\"all_studies_trimmed_all_genes.csv\"\n",
    "\n",
    "with open(filename, 'r') as csvfile:\n",
    "    read_reader = csv.reader(csvfile)  #csv.reader returns a reader object which will iterate over lines in the csvfile\n",
    "    row_count=0\n",
    "    DNMT3A_results={} #dictionary of DNMTA results (key = ID (row[13]), value = VAF (row[10]))\n",
    "\n",
    "    for row in read_reader:\n",
    "        if row_count>0:\n",
    "#             print(row)\n",
    "            if row[3] == 'DNMT3A':\n",
    "#                 print(row)\n",
    "                VAF = row[0]\n",
    "                age = row[1]\n",
    "                variant = str('p.')+row[2]\n",
    "                study = row[4]\n",
    "                DNMT3A_results[row_count]=(variant, study)\n",
    "                \n",
    "        row_count=row_count+1\n",
    "        \n",
    "all_DNMT3A_results_list = []\n",
    "for k, v in DNMT3A_results.items():\n",
    "    all_DNMT3A_results_list.append(v[0])\n",
    "        \n",
    "all_DNMT3A_df = pd.DataFrame({'variant':all_DNMT3A_results_list})\n",
    "all_grouped_df=all_DNMT3A_df.groupby(['variant']).size().reset_index(name='counts')\n",
    "all_grouped_df.to_csv('all_DNMT3A_variant_counts.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 222,
   "metadata": {
    "code_folding": [
     0
    ]
   },
   "outputs": [],
   "source": [
    "#All_DNMT3A_counts\n",
    "filename=\"all_DNMT3A_variant_counts.csv\"\n",
    "with open(filename, 'r') as csvfile:\n",
    "    read_reader = csv.reader(csvfile)  #csv.reader returns a reader object which will iterate over lines in the csvfile\n",
    "    row_count=0\n",
    "    all_studies_DNMT3A_counts={} #dictionary of DNMTA results (key = ID (row[13]), value = VAF (row[10]))\n",
    "\n",
    "    for row in read_reader:\n",
    "#         print(row)\n",
    "        if row_count>0:\n",
    "            variant = row[1]\n",
    "            counts = row[2]\n",
    "            all_studies_DNMT3A_counts[variant]=counts\n",
    "                \n",
    "        row_count=row_count+1"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### TET2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 223,
   "metadata": {
    "code_folding": [
     0
    ]
   },
   "outputs": [],
   "source": [
    "#Create lists of variants by studies and counts of each variant\n",
    "import csv\n",
    "filename=\"all_studies_trimmed_all_genes.csv\"\n",
    "\n",
    "with open(filename, 'r') as csvfile:\n",
    "    read_reader = csv.reader(csvfile)  #csv.reader returns a reader object which will iterate over lines in the csvfile\n",
    "    row_count=0\n",
    "    TET2_results={} #dictionary of DNMTA results (key = ID (row[13]), value = VAF (row[10]))\n",
    "\n",
    "    for row in read_reader:\n",
    "        if row_count>0:\n",
    "#             print(row)\n",
    "            if row[3] == 'TET2':\n",
    "#                 print(row)\n",
    "                VAF = row[0]\n",
    "                age = row[1]\n",
    "                variant = str('p.')+row[2]\n",
    "                study = row[4]\n",
    "                TET2_results[row_count]=(variant, study)\n",
    "                \n",
    "        row_count=row_count+1\n",
    "        \n",
    "all_TET2_results_list = []\n",
    "for k, v in TET2_results.items():\n",
    "    all_TET2_results_list.append(v[0])\n",
    "\n",
    "all_TET2_df = pd.DataFrame({'variant':all_TET2_results_list})\n",
    "all_grouped_df=all_TET2_df.groupby(['variant']).size().reset_index(name='counts')\n",
    "all_grouped_df.to_csv('all_TET2_variant_counts.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 224,
   "metadata": {
    "code_folding": [
     0
    ]
   },
   "outputs": [],
   "source": [
    "#All_TET2_counts\n",
    "filename=\"all_TET2_variant_counts.csv\"\n",
    "with open(filename, 'r') as csvfile:\n",
    "    read_reader = csv.reader(csvfile)  #csv.reader returns a reader object which will iterate over lines in the csvfile\n",
    "    row_count=0\n",
    "    all_studies_TET2_counts={} #dictionary of DNMTA results (key = ID (row[13]), value = VAF (row[10]))\n",
    "\n",
    "    for row in read_reader:\n",
    "#         print(row)\n",
    "        if row_count>0:\n",
    "            variant = row[1]\n",
    "            counts = row[2]\n",
    "            all_studies_TET2_counts[variant]=counts\n",
    "                \n",
    "        row_count=row_count+1"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### ASXL1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 225,
   "metadata": {
    "code_folding": [
     0
    ]
   },
   "outputs": [],
   "source": [
    "#Create lists of variants by studies and counts of each variant\n",
    "import csv\n",
    "filename=\"all_studies_trimmed_all_genes.csv\"\n",
    "\n",
    "with open(filename, 'r') as csvfile:\n",
    "    read_reader = csv.reader(csvfile)  #csv.reader returns a reader object which will iterate over lines in the csvfile\n",
    "    row_count=0\n",
    "    ASXL1_results={} #dictionary of DNMTA results (key = ID (row[13]), value = VAF (row[10]))\n",
    "\n",
    "    for row in read_reader:\n",
    "        if row_count>0:\n",
    "#             print(row)\n",
    "            if row[3] == 'ASXL1':\n",
    "#                 print(row)\n",
    "                VAF = row[0]\n",
    "                age = row[1]\n",
    "                variant = str('p.')+row[2]\n",
    "                study = row[4]\n",
    "                ASXL1_results[row_count]=(variant, study)\n",
    "                \n",
    "        row_count=row_count+1\n",
    "\n",
    "all_ASXL1_results_list = []\n",
    "\n",
    "for k, v in ASXL1_results.items():\n",
    "    all_ASXL1_results_list.append(v[0])\n",
    "        \n",
    "all_ASXL1_df = pd.DataFrame({'variant':all_ASXL1_results_list})\n",
    "all_grouped_df=all_ASXL1_df.groupby(['variant']).size().reset_index(name='counts')\n",
    "all_grouped_df.to_csv('all_ASXL1_variant_counts.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 226,
   "metadata": {
    "code_folding": [
     0
    ]
   },
   "outputs": [],
   "source": [
    "#All_ASXL1_counts\n",
    "filename=\"all_ASXL1_variant_counts.csv\"\n",
    "with open(filename, 'r') as csvfile:\n",
    "    read_reader = csv.reader(csvfile)  #csv.reader returns a reader object which will iterate over lines in the csvfile\n",
    "    row_count=0\n",
    "    all_studies_ASXL1_counts={} #dictionary of DNMTA results (key = ID (row[13]), value = VAF (row[10]))\n",
    "\n",
    "    for row in read_reader:\n",
    "#         print(row)\n",
    "        if row_count>0:\n",
    "            variant = row[1]\n",
    "            counts = row[2]\n",
    "            all_studies_ASXL1_counts[variant]=counts\n",
    "                \n",
    "        row_count=row_count+1"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### TP53"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 227,
   "metadata": {
    "code_folding": [
     0
    ]
   },
   "outputs": [],
   "source": [
    "#Create lists of variants by studies and counts of each variant\n",
    "import csv\n",
    "filename=\"all_studies_trimmed_all_genes.csv\"\n",
    "\n",
    "with open(filename, 'r') as csvfile:\n",
    "    read_reader = csv.reader(csvfile)  #csv.reader returns a reader object which will iterate over lines in the csvfile\n",
    "    row_count=0\n",
    "    TP53_results={} #dictionary of DNMTA results (key = ID (row[13]), value = VAF (row[10]))\n",
    "\n",
    "    for row in read_reader:\n",
    "        if row_count>0:\n",
    "#             print(row)\n",
    "            if row[3] == 'TP53':\n",
    "#                 print(row)\n",
    "                VAF = row[0]\n",
    "                age = row[1]\n",
    "                variant = str('p.')+row[2]\n",
    "                study = row[4]\n",
    "                TP53_results[row_count]=(variant, study)\n",
    "                \n",
    "        row_count=row_count+1\n",
    "\n",
    "all_TP53_results_list = []\n",
    "for k, v in TP53_results.items():\n",
    "    all_TP53_results_list.append(v[0])\n",
    "        \n",
    "all_TP53_df = pd.DataFrame({'variant':all_TP53_results_list})\n",
    "all_grouped_df=all_TP53_df.groupby(['variant']).size().reset_index(name='counts')\n",
    "all_grouped_df.to_csv('all_TP53_variant_counts.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 228,
   "metadata": {
    "code_folding": [
     0
    ]
   },
   "outputs": [],
   "source": [
    "#All_TP53_counts\n",
    "filename=\"all_TP53_variant_counts.csv\"\n",
    "with open(filename, 'r') as csvfile:\n",
    "    read_reader = csv.reader(csvfile)  #csv.reader returns a reader object which will iterate over lines in the csvfile\n",
    "    row_count=0\n",
    "    all_studies_TP53_counts={} #dictionary of DNMTA results (key = ID (row[13]), value = VAF (row[10]))\n",
    "\n",
    "    for row in read_reader:\n",
    "#         print(row)\n",
    "        if row_count>0:\n",
    "            variant = row[1]\n",
    "            counts = row[2]\n",
    "            all_studies_TP53_counts[variant]=counts\n",
    "                \n",
    "        row_count=row_count+1"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Functions for calculating the expected number of observations of a variant"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- The number of times you expect to see a variant depends on it's fitness effect (s), the size of the study (i.e. number of participants), the ages of the individuals in the study and the study's sequencing limit of detection.\n",
    "- The distribution of ages in the studys is presumed to be ~normally distributed, except for Acuna-Hidalgo, which has a ~uniform distribution of ages."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 229,
   "metadata": {},
   "outputs": [],
   "source": [
    "def density_of_clones(f, s, t, N, u): \n",
    "    return ((N*u)/(f*(1-2*f)))*np.exp(-f/((1-2*f)*((np.exp(s*t)-1)/(2*N*s))))\n",
    "\n",
    "def prevalence_single_s(fmin, s, t, N, u): #expected prevalence of variant of a particular fitness effect\n",
    "    expected_number_clones=integrate.quad(lambda f: density_of_clones(f, s, t, N, u), fmin, 0.4999)[0]\n",
    "    return 1.0-np.exp(-expected_number_clones)\n",
    "\n",
    "def prevalence_averaged_over_ages(fmin, s, N, u, mean, sigma, study_size): #expected prevalence across range of ages\n",
    "    expected_prev_clones=integrate.quad(lambda t: (prevalence_single_s(fmin, s, t, N, u))*\\\n",
    "                                 ((1/((2*np.pi*(sigma)**2)**0.5))*(np.exp(-(t-mean)**2/(2*(sigma**2))))), mean-2*sigma, mean+2*sigma)\n",
    "    prev = expected_prev_clones[0]\n",
    "    expect_number_clones = prev*study_size\n",
    "    return expect_number_clones\n",
    "\n",
    "def prevalence_averaged_over_uniform_ages(fmin, s, N, u, mean, study_size): #expected prevalence across uniform distribution of ages\n",
    "    expected_prevalence_clones=integrate.quad(lambda t: (prevalence_single_s(fmin, s, t, N, u))*(1/49), 20, 69)\n",
    "    prevalence = expected_prevalence_clones[0]\n",
    "    expected_number_clones = prevalence*study_size\n",
    "    return expected_number_clones"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Maximum Likelihood Estimation for s"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- This involves calculating the most likely fitness effect (s) that would explain the number of observations of the variant."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    " **Calculating the expected number of observations of each variant**\n",
    "- To calculate the expected number of observations of each variant, need to look at each variant in turn and for each variant, need to then look at each study in turn and see if that variant is targeted by that study.  If it is, need to work out how many times you expect to see that variant in that study, taking in to account the study's limit of detection, study size and ages of the individuals in the study. Then, once looked at all studies, sum up the number of observations across all the studies to get the overall expected number of observations for that variant.\n",
    "- The expected number also depends on the fitness effect of the variant - the value of s that means the observed number of variants matches as closely as possible the expected number of variants is calculated."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## DNMT3A variants"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 230,
   "metadata": {
    "code_folding": [
     0,
     96,
     107,
     119
    ]
   },
   "outputs": [],
   "source": [
    "def expected_number_variants_missense_DNMT3A(data, params):\n",
    "    s_value = params[0]\n",
    "    \n",
    "    a = 0\n",
    "    expected_number_by_variant_missense = {}\n",
    "    \n",
    "    Jaiswal_fmin = 0.056\n",
    "    Jaiswal_mean = 58\n",
    "    Jaiswal_sigma = 11.5\n",
    "    Jaiswal_total = 17182\n",
    "    \n",
    "    Genovese_fmin = 0.1194\n",
    "    Genovese_mean = 55.1\n",
    "    Genovese_total = 12380\n",
    "    Genovese_sigma = 11.95\n",
    "    \n",
    "    McKerrel_fmin = 0.008\n",
    "    McKerrel_mean = 51.88\n",
    "    McKerrel_total = 4219\n",
    "    McKerrel_sigma = 18.36\n",
    "    \n",
    "    Zink_fmin = 0.2363\n",
    "    Zink_mean = 60\n",
    "    Zink_total = 11262\n",
    "    Zink_sigma = 1.18907720e+01\n",
    "    \n",
    "    Coombs_fmin = 0.019\n",
    "    Coombs_mean = 59.82\n",
    "    Coombs_total = 5649\n",
    "    Coombs_sigma = 15.14\n",
    "    \n",
    "    Young_fmin = 0.0009\n",
    "    Young_mean = 63\n",
    "    Young_total = 89\n",
    "    Young_sigma = 6.8\n",
    "    \n",
    "    Desai_fmin = 0.0169\n",
    "    Desai_mean = 66.1\n",
    "    Desai_total = 424\n",
    "    Desai_sigma = 6 #estimated\n",
    "    \n",
    "    Acuna_fmin = 0.0015\n",
    "    Acuna_mean = 4.5\n",
    "    Acuna_total = 2014\n",
    "    \n",
    "    N = 9.39358404e+04\n",
    "\n",
    "    for i in data:\n",
    "        variant_mutation_rate = mutation_rates_variants_DNMT3A(i)\n",
    "\n",
    "        if i in Jaiswal_missense_DNMT3A:\n",
    "            Jaiswal_number = prevalence_averaged_over_ages(Jaiswal_fmin, s_value, N, variant_mutation_rate, Jaiswal_mean, Jaiswal_sigma, Jaiswal_total)\n",
    "        else:\n",
    "            Jaiswal_number = 0\n",
    "\n",
    "        if i in Genovese_missense_DNMT3A:\n",
    "            Genovese_number = prevalence_averaged_over_ages(Genovese_fmin, s_value, N, variant_mutation_rate, Genovese_mean, Genovese_sigma, Genovese_total)\n",
    "        else:\n",
    "            Genovese_number = 0 \n",
    "\n",
    "        if i in Zink_missense_DNMT3A:\n",
    "            Zink_number = prevalence_averaged_over_ages(Zink_fmin, s_value, N, variant_mutation_rate, Zink_mean, Zink_sigma, Zink_total)\n",
    "        else:\n",
    "            Zink_number = 0\n",
    "            \n",
    "        if i in McKerrel_missense_DNMT3A:\n",
    "            McKerrel_number = prevalence_averaged_over_ages(McKerrel_fmin, s_value, N, variant_mutation_rate, McKerrel_mean, McKerrel_sigma, McKerrel_total)\n",
    "        else:\n",
    "            McKerrel_number = 0\n",
    "\n",
    "        if i in Coombs_missense_DNMT3A:\n",
    "            Coombs_number = prevalence_averaged_over_ages(Coombs_fmin, s_value, N, variant_mutation_rate, Coombs_mean, Coombs_sigma, Coombs_total)\n",
    "        else:\n",
    "            Coombs_number = 0\n",
    "\n",
    "        if i in Young_missense_DNMT3A:\n",
    "            Young_number = prevalence_averaged_over_ages(Young_fmin, s_value, N, variant_mutation_rate, Young_mean, Young_sigma, Young_total)\n",
    "        else:\n",
    "            Young_number = 0\n",
    "\n",
    "        if i in Desai_missense_DNMT3A:\n",
    "            Desai_number = prevalence_averaged_over_ages(Desai_fmin, s_value, N, variant_mutation_rate, Desai_mean, Desai_sigma, Desai_total)\n",
    "        else:\n",
    "            Desai_number = 0\n",
    "\n",
    "        if i in Acuna_missense_DNMT3A:\n",
    "            Acuna_number = prevalence_averaged_over_uniform_ages(Acuna_fmin, s_value, N, variant_mutation_rate, Acuna_mean, Acuna_total)\n",
    "        else:\n",
    "            Acuna_number = 0\n",
    "\n",
    "        expected_total = Jaiswal_number + Genovese_number + Zink_number + McKerrel_number + Coombs_number + Young_number + Desai_number + Acuna_number\n",
    "\n",
    "        expected_number_by_variant_missense[i] = expected_total\n",
    "    \n",
    "    return expected_number_by_variant_missense\n",
    "\n",
    "def prob_observed_given_expected_missense_DNMT3A(params, data): #data = list of variant (e.g. p.R882H) (or all_missense_variants_DNMT3A)\n",
    "    total_square_distance = 0\n",
    "    a = 0\n",
    "    for variant in data:\n",
    "        observed_number = float(all_studies_DNMT3A_counts[variant])\n",
    "        expected_number = float(expected_number_variants_missense_DNMT3A(data, params)[variant])\n",
    "        square_distance = ((observed_number - expected_number)**2)\n",
    "        total_square_distance = total_square_distance + square_distance\n",
    "\n",
    "    return total_square_distance   \n",
    "\n",
    "def prob_observed_given_expected_missense_DNMT3A_upper_error(params, data): #data = list of variant (e.g. p.R882H) (or all_missense_variants_DNMT3A)\n",
    "    total_square_distance = 0\n",
    "    a = 0\n",
    "    for variant in data:\n",
    "        observed_number = float(all_studies_DNMT3A_counts[variant])\n",
    "        observed_number_upper_error = observed_number+np.sqrt(observed_number)\n",
    "        expected_number = float(expected_number_variants_missense_DNMT3A(data, params)[variant])\n",
    "        square_distance = ((observed_number_upper_error - expected_number)**2)\n",
    "        total_square_distance = total_square_distance + square_distance\n",
    "\n",
    "    return total_square_distance   \n",
    "\n",
    "def prob_observed_given_expected_missense_DNMT3A_lower_error(params, data): #data = list of variant (e.g. p.R882H) (or all_missense_variants_DNMT3A)\n",
    "    total_square_distance = 0\n",
    "    a = 0\n",
    "    for variant in data:\n",
    "        observed_number = float(all_studies_DNMT3A_counts[variant])\n",
    "        observed_number_lower_error = observed_number-np.sqrt(observed_number)\n",
    "        expected_number = float(expected_number_variants_missense_DNMT3A(data, params)[variant])\n",
    "        square_distance = ((observed_number_lower_error - expected_number)**2)\n",
    "        total_square_distance = total_square_distance + square_distance\n",
    "\n",
    "    return total_square_distance   "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 231,
   "metadata": {
    "code_folding": [
     0,
     91,
     102,
     114
    ]
   },
   "outputs": [],
   "source": [
    "def expected_number_variants_nonsense_DNMT3A(data, params):\n",
    "    s_value = params[0]\n",
    "    \n",
    "    a = 0\n",
    "    expected_number_by_variant_nonsense = {}\n",
    "    \n",
    "    Jaiswal_fmin = 0.056\n",
    "    Jaiswal_mean = 58\n",
    "    Jaiswal_sigma = 11.5\n",
    "    Jaiswal_total = 17182\n",
    "    \n",
    "    Genovese_fmin = 0.1194\n",
    "    Genovese_mean = 55.1\n",
    "    Genovese_total = 12380\n",
    "    Genovese_sigma = 11.95\n",
    "    \n",
    "    McKerrel_fmin = 0.008\n",
    "    McKerrel_mean = 51.88\n",
    "    McKerrel_total = 4219\n",
    "    McKerrel_sigma = 18.36\n",
    "    \n",
    "    Zink_fmin = 0.2363\n",
    "    Zink_mean = 60\n",
    "    Zink_total = 11262\n",
    "    Zink_sigma = 1.18907720e+01\n",
    "    \n",
    "    Coombs_fmin = 0.019\n",
    "    Coombs_mean = 59.82\n",
    "    Coombs_total = 5649\n",
    "    Coombs_sigma = 15.14\n",
    "    \n",
    "    Young_fmin = 0.0009\n",
    "    Young_mean = 63\n",
    "    Young_total = 89\n",
    "    Young_sigma = 6.8\n",
    "    \n",
    "    Desai_fmin = 0.0169\n",
    "    Desai_mean = 66.1\n",
    "    Desai_total = 424\n",
    "    Desai_sigma = 6 #estimated\n",
    "    \n",
    "    Acuna_fmin = 0.0015\n",
    "    Acuna_mean = 4.5\n",
    "    Acuna_total = 2014\n",
    "    \n",
    "    N = 9.39358404e+04\n",
    "\n",
    "    for i in data:\n",
    "        variant_mutation_rate = mutation_rates_variants_DNMT3A(i)\n",
    "\n",
    "        if i in Jaiswal_nonsense_DNMT3A:\n",
    "            Jaiswal_number = prevalence_averaged_over_ages(Jaiswal_fmin, s_value, N, variant_mutation_rate, Jaiswal_mean, Jaiswal_sigma, Jaiswal_total)\n",
    "        else:\n",
    "            Jaiswal_number = 0\n",
    "\n",
    "        if i in Genovese_nonsense_DNMT3A:\n",
    "            Genovese_number = prevalence_averaged_over_ages(Genovese_fmin, s_value, N, variant_mutation_rate, Genovese_mean, Genovese_sigma, Genovese_total)\n",
    "        else:\n",
    "            Genovese_number = 0 \n",
    "\n",
    "        if i in Zink_nonsense_DNMT3A:\n",
    "            Zink_number = prevalence_averaged_over_ages(Zink_fmin, s_value, N, variant_mutation_rate, Zink_mean, Zink_sigma, Zink_total)\n",
    "        else:\n",
    "            Zink_number = 0\n",
    "\n",
    "        if i in Coombs_nonsense_DNMT3A:\n",
    "            Coombs_number = prevalence_averaged_over_ages(Coombs_fmin, s_value, N, variant_mutation_rate, Coombs_mean, Coombs_sigma, Coombs_total)\n",
    "        else:\n",
    "            Coombs_number = 0\n",
    "\n",
    "        if i in Young_nonsense_DNMT3A:\n",
    "            Young_number = prevalence_averaged_over_ages(Young_fmin, s_value, N, variant_mutation_rate, Young_mean, Young_sigma, Young_total)\n",
    "        else:\n",
    "            Young_number = 0\n",
    "\n",
    "        if i in Desai_nonsense_DNMT3A:\n",
    "            Desai_number = prevalence_averaged_over_ages(Desai_fmin, s_value, N, variant_mutation_rate, Desai_mean, Desai_sigma, Desai_total)\n",
    "        else:\n",
    "            Desai_number = 0\n",
    "\n",
    "        if i in Acuna_nonsense_DNMT3A:\n",
    "            Acuna_number = prevalence_averaged_over_uniform_ages(Acuna_fmin, s_value, N, variant_mutation_rate, Acuna_mean, Acuna_total)\n",
    "        else:\n",
    "            Acuna_number = 0\n",
    "\n",
    "        expected_total = Jaiswal_number + Genovese_number + Zink_number + Coombs_number + Young_number + Desai_number + Acuna_number\n",
    "\n",
    "        expected_number_by_variant_nonsense[i] = expected_total\n",
    "    \n",
    "    return expected_number_by_variant_nonsense\n",
    "\n",
    "def prob_observed_given_expected_nonsense_DNMT3A(params, data): #data = list of variant (e.g. p.R882H) (or all_nonsense_variants_DNMT3A)\n",
    "    total_square_distance = 0\n",
    "    a = 0\n",
    "    for variant in data:\n",
    "        observed_number = float(all_studies_DNMT3A_counts[variant])\n",
    "        expected_number = float(expected_number_variants_nonsense_DNMT3A(data, params)[variant])\n",
    "        square_distance = ((observed_number - expected_number)**2)\n",
    "        total_square_distance = total_square_distance + square_distance\n",
    "\n",
    "    return total_square_distance\n",
    "\n",
    "def prob_observed_given_expected_nonsense_DNMT3A_upper_error(params, data): #data = list of variant (e.g. p.R882H) (or all_nonsense_variants_DNMT3A)\n",
    "    total_square_distance = 0\n",
    "    a = 0\n",
    "    for variant in data:\n",
    "        observed_number = float(all_studies_DNMT3A_counts[variant])\n",
    "        observed_number_upper_error = observed_number+np.sqrt(observed_number)\n",
    "        expected_number = float(expected_number_variants_nonsense_DNMT3A(data, params)[variant])\n",
    "        square_distance = ((observed_number_upper_error - expected_number)**2)\n",
    "        total_square_distance = total_square_distance + square_distance\n",
    "\n",
    "    return total_square_distance   \n",
    "\n",
    "def prob_observed_given_expected_nonsense_DNMT3A_lower_error(params, data): #data = list of variant (e.g. p.R882H) (or all_nonsense_variants_DNMT3A)\n",
    "    total_square_distance = 0\n",
    "    a = 0\n",
    "    for variant in data:\n",
    "        observed_number = float(all_studies_DNMT3A_counts[variant])\n",
    "        observed_number_lower_error = observed_number-np.sqrt(observed_number)\n",
    "        expected_number = float(expected_number_variants_nonsense_DNMT3A(data, params)[variant])\n",
    "        square_distance = ((observed_number_lower_error - expected_number)**2)\n",
    "        total_square_distance = total_square_distance + square_distance\n",
    "\n",
    "    return total_square_distance "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 233,
   "metadata": {
    "code_folding": [
     0
    ],
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/anaconda3/lib/python3.7/site-packages/scipy/integrate/quadpack.py:385: IntegrationWarning: The integral is probably divergent, or slowly convergent.\n",
      "  warnings.warn(msg, IntegrationWarning)\n"
     ]
    }
   ],
   "source": [
    "#CALCULATING S FOR MISSENSE VARIANTS\n",
    "DNMT3A_missense_s = {}\n",
    "a = 0\n",
    "for var in all_missense_variants_DNMT3A: #iterate over all the possible missense variants in DNMT3A\n",
    "    initial_guess=[0.05] #initial guess for s = 0.05 (i.e. 5%)\n",
    "    if var in all_studies_DNMT3A_counts.keys(): #look to see if the variant is seen across any of the studies\n",
    "        observed_number = float(all_studies_DNMT3A_counts[var]) #if it is, retrieve how many times it was observed\n",
    "#         print(observed_number)\n",
    "        if observed_number >1: #only calculate s for variants seen more than 1 time\n",
    "            variant_mutation_rate = mutation_rates_variants_DNMT3A(var) #trinucleotide-context dependent mutation rate for the variant\n",
    "            \n",
    "            MLE = scipy.optimize.minimize(prob_observed_given_expected_missense_DNMT3A, initial_guess, args=([var],), method='Nelder-Mead')\n",
    "            s_value = MLE.x[0]\n",
    "\n",
    "            upper_error_MLE = scipy.optimize.minimize(prob_observed_given_expected_missense_DNMT3A_upper_error, \\\n",
    "                                                      initial_guess, args=([var],), method='Nelder-Mead')\n",
    "            s_value_upper = upper_error_MLE.x[0]\n",
    "\n",
    "            lower_error_MLE = scipy.optimize.minimize(prob_observed_given_expected_missense_DNMT3A_lower_error, \\\n",
    "                                                      initial_guess, args=([var],), method='Nelder-Mead')\n",
    "            s_value_lower = lower_error_MLE.x[0]\n",
    "\n",
    "#             print('a = ', a)\n",
    "            #Create a dictionary where key = variant\n",
    "            DNMT3A_missense_s[var]=(s_value, observed_number, variant_mutation_rate, s_value_upper, s_value_lower) \n",
    "    a = a+1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 234,
   "metadata": {
    "code_folding": [
     0
    ],
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/anaconda3/lib/python3.7/site-packages/scipy/integrate/quadpack.py:385: IntegrationWarning: The integral is probably divergent, or slowly convergent.\n",
      "  warnings.warn(msg, IntegrationWarning)\n"
     ]
    }
   ],
   "source": [
    "#CALCULATING S FOR NONSENSE VARIANTS\n",
    "DNMT3A_nonsense_s = {}\n",
    "a = 0\n",
    "for var in all_nonsense_variants_DNMT3A: #iterate over all the possible nonsense variants in DNMT3A\n",
    "    initial_guess=[0.05] #initial guess for s = 0.05 (i.e. 5%)\n",
    "    if var in all_studies_DNMT3A_counts.keys(): #look to see if the variant is seen across any of the studies\n",
    "        observed_number = float(all_studies_DNMT3A_counts[var]) #if it is, retrieve how many times it was observed\n",
    "#         print(observed_number)\n",
    "        if observed_number >1: #only calculate s for variants seen more than 1 time\n",
    "            variant_mutation_rate = mutation_rates_variants_DNMT3A(var) #trinucleotide-context dependent mutation rate for the variant\n",
    "            \n",
    "            MLE = scipy.optimize.minimize(prob_observed_given_expected_nonsense_DNMT3A, initial_guess, args=([var],), method='Nelder-Mead')\n",
    "            s_value = MLE.x[0]\n",
    "\n",
    "            upper_error_MLE = scipy.optimize.minimize(prob_observed_given_expected_nonsense_DNMT3A_upper_error, \\\n",
    "                                                      initial_guess, args=([var],), method='Nelder-Mead')\n",
    "            s_value_upper = upper_error_MLE.x[0]\n",
    "\n",
    "            lower_error_MLE = scipy.optimize.minimize(prob_observed_given_expected_nonsense_DNMT3A_lower_error, \\\n",
    "                                                      initial_guess, args=([var],), method='Nelder-Mead')\n",
    "            s_value_lower = lower_error_MLE.x[0]\n",
    "\n",
    "#             print('a = ', a)\n",
    "            DNMT3A_nonsense_s[var]=(s_value, observed_number, variant_mutation_rate, s_value_upper, s_value_lower) \n",
    "    a = a+1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 235,
   "metadata": {
    "code_folding": [
     0
    ]
   },
   "outputs": [],
   "source": [
    "#Exac - retrieve the frequencies in Exac for each of the variants\n",
    "import csv\n",
    "filename=\"DNMT3A_Exac.csv\"\n",
    "\n",
    "with open(filename, 'r') as csvfile:\n",
    "    read_reader = csv.reader(csvfile)  #csv.reader returns a reader object which will iterate over lines in the csvfile\n",
    "    row_count=0\n",
    "    Exac_DNMT3A_results={} \n",
    "    \n",
    "    for row in read_reader:\n",
    "        if row_count>1:\n",
    "#             print(row)\n",
    "            variant = row[6]\n",
    "            allele_freq = float(row[14])\n",
    "            african_freq = float(row[15])/float(row[16])\n",
    "            east_asian_freq = float(row[18])/float(row[19])\n",
    "            european_freq = float(row[21])/float(row[22])\n",
    "            finnish_freq = float(row[24])/float(row[25])\n",
    "            latino_freq = float(row[27])/float(row[28])\n",
    "            south_asian_freq = float(row[33])/float(row[34])\n",
    "            Exac_DNMT3A_results[variant]=(allele_freq, african_freq, east_asian_freq, european_freq, finnish_freq, latino_freq, south_asian_freq)\n",
    "\n",
    "        row_count=row_count+1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 236,
   "metadata": {
    "code_folding": [
     0
    ]
   },
   "outputs": [],
   "source": [
    "#COSMIC - retrieve the frequencies in COSMIC for each of the variants\n",
    "import csv\n",
    "filename=\"DNMT3A_COSMIC_variant_counts.csv\"\n",
    "with open(filename, 'r') as csvfile:\n",
    "    read_reader = csv.reader(csvfile)  #csv.reader returns a reader object which will iterate over lines in the csvfile\n",
    "    row_count=0\n",
    "    DNMT3A_COSMIC_results={} #dictionary of DNMTA results (key = ID (row[13]), value = VAF (row[10]))\n",
    "\n",
    "    for row in read_reader:\n",
    "        if row_count>1:\n",
    "            variant = row[1]\n",
    "            counts = int(row[2])\n",
    "            DNMT3A_COSMIC_results[variant]=counts\n",
    "\n",
    "        row_count=row_count+1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 237,
   "metadata": {
    "code_folding": [
     0
    ],
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "#Create dictionary of results and save to .csv file\n",
    "DNMT3A_s_dictionary={}\n",
    "for k, v in DNMT3A_missense_s.items():\n",
    "#     print(k)\n",
    "    if k in Exac_DNMT3A_results.keys():\n",
    "        exac_freq = round((Exac_DNMT3A_results[k][0])*100, 4)\n",
    "        african_freq = round((Exac_DNMT3A_results[k][1])*100, 4)\n",
    "        east_asian_freq = round((Exac_DNMT3A_results[k][2])*100, 4)\n",
    "        european_freq = round((Exac_DNMT3A_results[k][3])*100, 4)\n",
    "        finnish_freq = round((Exac_DNMT3A_results[k][4])*100, 4)\n",
    "        latino_freq = round((Exac_DNMT3A_results[k][5])*100, 4)\n",
    "        south_asian_freq = round((Exac_DNMT3A_results[k][6])*100, 4)\n",
    "    if k not in Exac_DNMT3A_results.keys():\n",
    "        exac_freq = '-'\n",
    "        african_freq = '-'\n",
    "        east_asian_freq = '-'\n",
    "        european_freq = '-'\n",
    "        finnish_freq = '-'\n",
    "        latino_freq = '-'\n",
    "        south_asian_freq = '-'\n",
    "    if k in DNMT3A_COSMIC_results.keys():\n",
    "        COSMIC_number = DNMT3A_COSMIC_results[k]\n",
    "    if k not in DNMT3A_COSMIC_results.keys():\n",
    "        COSMIC_number = '-'\n",
    "    DNMT3A_s_dictionary[k]=(round(v[0]*100, 2),round(v[4]*100, 2),round(v[3]*100, 2), v[1], v[2], COSMIC_number, exac_freq, \\\n",
    "                            african_freq, east_asian_freq, european_freq, finnish_freq, latino_freq, south_asian_freq)\n",
    "\n",
    "for k, v in DNMT3A_nonsense_s.items():\n",
    "    if k in Exac_DNMT3A_results.keys():\n",
    "        exac_freq = round((Exac_DNMT3A_results[k][0])*100, 4)\n",
    "        african_freq = round((Exac_DNMT3A_results[k][1])*100, 4)\n",
    "        east_asian_freq = round((Exac_DNMT3A_results[k][2])*100, 4)\n",
    "        european_freq = round((Exac_DNMT3A_results[k][3])*100, 4)\n",
    "        finnish_freq = round((Exac_DNMT3A_results[k][4])*100, 4)\n",
    "        latino_freq = round((Exac_DNMT3A_results[k][5])*100, 4)\n",
    "        south_asian_freq = round((Exac_DNMT3A_results[k][6])*100, 4)\n",
    "    if k not in Exac_DNMT3A_results.keys():\n",
    "        exac_freq = '-'\n",
    "        african_freq = '-'\n",
    "        east_asian_freq = '-'\n",
    "        european_freq = '-'\n",
    "        finnish_freq = '-'\n",
    "        latino_freq = '-'\n",
    "        south_asian_freq = '-'\n",
    "    if k in DNMT3A_COSMIC_results.keys():\n",
    "        COSMIC_number = DNMT3A_COSMIC_results[k]\n",
    "    if k not in DNMT3A_COSMIC_results.keys():\n",
    "        COSMIC_number = '-'\n",
    "    DNMT3A_s_dictionary[k]=(round(v[0]*100, 2),round(v[4]*100, 2),round(v[3]*100, 2), v[1], v[2], COSMIC_number, exac_freq, \\\n",
    "                            african_freq, east_asian_freq, european_freq, finnish_freq, latino_freq, south_asian_freq)\n",
    "\n",
    "DNMT3A_df = pd.DataFrame.from_dict(DNMT3A_s_dictionary, orient='index', columns=['s (%)', 's lower value (%)', 's upper value (%)','observed number', 'site-specific mutation rate', 'Number of times in COSMIC',\\\n",
    "                                                                                'Exac Overall freq (%)', 'Exac African freq (%)', 'Exac East Asian freq (%)',\\\n",
    "                                                                                'Exac European freq (%)', 'Exac Finnish freq (%)', 'Exac Latino freq (%)', 'Exac South Asian freq (%)'])\n",
    "DNMT3A_df_sorted = DNMT3A_df.sort_values(by = ['s (%)'], ascending = False)\n",
    "\n",
    "DNMT3A_df_sorted.to_csv('DNMT3A_s_across_whole_gene.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 238,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>s (%)</th>\n",
       "      <th>s lower value (%)</th>\n",
       "      <th>s upper value (%)</th>\n",
       "      <th>observed number</th>\n",
       "      <th>site-specific mutation rate</th>\n",
       "      <th>Number of times in COSMIC</th>\n",
       "      <th>Exac Overall freq (%)</th>\n",
       "      <th>Exac African freq (%)</th>\n",
       "      <th>Exac East Asian freq (%)</th>\n",
       "      <th>Exac European freq (%)</th>\n",
       "      <th>Exac Finnish freq (%)</th>\n",
       "      <th>Exac Latino freq (%)</th>\n",
       "      <th>Exac South Asian freq (%)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>p.E30A</th>\n",
       "      <td>196.75</td>\n",
       "      <td>28.05</td>\n",
       "      <td>196.75</td>\n",
       "      <td>4.0</td>\n",
       "      <td>1.864760e-10</td>\n",
       "      <td>3</td>\n",
       "      <td>0.2959</td>\n",
       "      <td>0.02</td>\n",
       "      <td>0</td>\n",
       "      <td>0.3033</td>\n",
       "      <td>2.1424</td>\n",
       "      <td>0.1413</td>\n",
       "      <td>0.0247</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.F751C</th>\n",
       "      <td>30.59</td>\n",
       "      <td>18.09</td>\n",
       "      <td>196.75</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.566770e-10</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.F732C</th>\n",
       "      <td>22.84</td>\n",
       "      <td>16.11</td>\n",
       "      <td>29.32</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.859550e-10</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.F414C</th>\n",
       "      <td>21.61</td>\n",
       "      <td>15.05</td>\n",
       "      <td>27.52</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.566770e-10</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.R729G</th>\n",
       "      <td>21.22</td>\n",
       "      <td>16.27</td>\n",
       "      <td>25.69</td>\n",
       "      <td>3.0</td>\n",
       "      <td>2.470660e-10</td>\n",
       "      <td>-</td>\n",
       "      <td>0.0009</td>\n",
       "      <td>0.0097</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.W795S</th>\n",
       "      <td>20.33</td>\n",
       "      <td>14.30</td>\n",
       "      <td>25.16</td>\n",
       "      <td>2.0</td>\n",
       "      <td>3.739340e-10</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.Y908C</th>\n",
       "      <td>19.88</td>\n",
       "      <td>15.03</td>\n",
       "      <td>24.63</td>\n",
       "      <td>4.0</td>\n",
       "      <td>1.473330e-09</td>\n",
       "      <td>-</td>\n",
       "      <td>0.0033</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0045</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0086</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.Y735C</th>\n",
       "      <td>19.66</td>\n",
       "      <td>18.31</td>\n",
       "      <td>20.94</td>\n",
       "      <td>23.0</td>\n",
       "      <td>1.839360e-09</td>\n",
       "      <td>12</td>\n",
       "      <td>0.0118</td>\n",
       "      <td>0.0097</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0182</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0087</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.R882C</th>\n",
       "      <td>19.41</td>\n",
       "      <td>18.54</td>\n",
       "      <td>20.26</td>\n",
       "      <td>61.0</td>\n",
       "      <td>5.336040e-09</td>\n",
       "      <td>460</td>\n",
       "      <td>0.0355</td>\n",
       "      <td>0.0677</td>\n",
       "      <td>0.0347</td>\n",
       "      <td>0.0391</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0087</td>\n",
       "      <td>0.0364</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.P507R</th>\n",
       "      <td>19.02</td>\n",
       "      <td>13.63</td>\n",
       "      <td>23.05</td>\n",
       "      <td>2.0</td>\n",
       "      <td>4.563380e-10</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.L737R</th>\n",
       "      <td>18.75</td>\n",
       "      <td>13.80</td>\n",
       "      <td>22.43</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.864760e-10</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.M880V</th>\n",
       "      <td>18.74</td>\n",
       "      <td>16.09</td>\n",
       "      <td>21.07</td>\n",
       "      <td>6.0</td>\n",
       "      <td>6.735990e-10</td>\n",
       "      <td>7</td>\n",
       "      <td>0.0017</td>\n",
       "      <td>0.0097</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0015</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.S714C</th>\n",
       "      <td>18.45</td>\n",
       "      <td>16.30</td>\n",
       "      <td>20.33</td>\n",
       "      <td>7.0</td>\n",
       "      <td>7.420070e-10</td>\n",
       "      <td>25</td>\n",
       "      <td>0.0041</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0075</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.R882P</th>\n",
       "      <td>18.25</td>\n",
       "      <td>15.62</td>\n",
       "      <td>20.55</td>\n",
       "      <td>6.0</td>\n",
       "      <td>6.302210e-10</td>\n",
       "      <td>36</td>\n",
       "      <td>0.0058</td>\n",
       "      <td>0.029</td>\n",
       "      <td>0.0116</td>\n",
       "      <td>0.003</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0087</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.G543A</th>\n",
       "      <td>17.69</td>\n",
       "      <td>13.43</td>\n",
       "      <td>20.73</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.470660e-10</td>\n",
       "      <td>5</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.F755S</th>\n",
       "      <td>17.66</td>\n",
       "      <td>14.96</td>\n",
       "      <td>19.86</td>\n",
       "      <td>4.0</td>\n",
       "      <td>4.962520e-10</td>\n",
       "      <td>3</td>\n",
       "      <td>0.0041</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.006</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0086</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.W860R</th>\n",
       "      <td>17.64</td>\n",
       "      <td>16.34</td>\n",
       "      <td>18.83</td>\n",
       "      <td>18.0</td>\n",
       "      <td>2.463301e-09</td>\n",
       "      <td>7</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.E599D</th>\n",
       "      <td>17.30</td>\n",
       "      <td>14.38</td>\n",
       "      <td>19.62</td>\n",
       "      <td>4.0</td>\n",
       "      <td>1.244421e-09</td>\n",
       "      <td>-</td>\n",
       "      <td>0.0034</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0061</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.S770*</th>\n",
       "      <td>17.27</td>\n",
       "      <td>12.99</td>\n",
       "      <td>20.26</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.488650e-10</td>\n",
       "      <td>2</td>\n",
       "      <td>0.0017</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.003</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.N838D</th>\n",
       "      <td>17.09</td>\n",
       "      <td>13.07</td>\n",
       "      <td>19.74</td>\n",
       "      <td>2.0</td>\n",
       "      <td>4.962520e-10</td>\n",
       "      <td>5</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.R882S</th>\n",
       "      <td>17.00</td>\n",
       "      <td>11.76</td>\n",
       "      <td>20.51</td>\n",
       "      <td>2.0</td>\n",
       "      <td>4.385790e-10</td>\n",
       "      <td>40</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.W753C</th>\n",
       "      <td>16.82</td>\n",
       "      <td>12.41</td>\n",
       "      <td>19.84</td>\n",
       "      <td>2.0</td>\n",
       "      <td>6.856450e-10</td>\n",
       "      <td>-</td>\n",
       "      <td>0.0008</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0015</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.N501S</th>\n",
       "      <td>16.55</td>\n",
       "      <td>12.73</td>\n",
       "      <td>19.01</td>\n",
       "      <td>2.0</td>\n",
       "      <td>5.656330e-10</td>\n",
       "      <td>5</td>\n",
       "      <td>0.0294</td>\n",
       "      <td>0.0098</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0321</td>\n",
       "      <td>0.0457</td>\n",
       "      <td>0.0174</td>\n",
       "      <td>0.0502</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.M761V</th>\n",
       "      <td>16.45</td>\n",
       "      <td>13.82</td>\n",
       "      <td>18.57</td>\n",
       "      <td>4.0</td>\n",
       "      <td>6.735990e-10</td>\n",
       "      <td>-</td>\n",
       "      <td>0.0008</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0015</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.E733G</th>\n",
       "      <td>16.09</td>\n",
       "      <td>13.09</td>\n",
       "      <td>18.17</td>\n",
       "      <td>3.0</td>\n",
       "      <td>8.133640e-10</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.L547H</th>\n",
       "      <td>15.90</td>\n",
       "      <td>12.41</td>\n",
       "      <td>18.21</td>\n",
       "      <td>2.0</td>\n",
       "      <td>3.786790e-10</td>\n",
       "      <td>5</td>\n",
       "      <td>0.0017</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0087</td>\n",
       "      <td>0.0061</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.S770W</th>\n",
       "      <td>15.63</td>\n",
       "      <td>12.94</td>\n",
       "      <td>17.59</td>\n",
       "      <td>3.0</td>\n",
       "      <td>5.511000e-10</td>\n",
       "      <td>5</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.R882H</th>\n",
       "      <td>15.58</td>\n",
       "      <td>15.15</td>\n",
       "      <td>15.99</td>\n",
       "      <td>105.0</td>\n",
       "      <td>1.883910e-08</td>\n",
       "      <td>1069</td>\n",
       "      <td>0.0545</td>\n",
       "      <td>0.0869</td>\n",
       "      <td>0.0694</td>\n",
       "      <td>0.0571</td>\n",
       "      <td>0.0758</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0485</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.R326S</th>\n",
       "      <td>15.34</td>\n",
       "      <td>11.65</td>\n",
       "      <td>17.80</td>\n",
       "      <td>2.0</td>\n",
       "      <td>4.385790e-10</td>\n",
       "      <td>-</td>\n",
       "      <td>0.0008</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0151</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.F732S</th>\n",
       "      <td>14.96</td>\n",
       "      <td>11.84</td>\n",
       "      <td>16.95</td>\n",
       "      <td>2.0</td>\n",
       "      <td>4.962520e-10</td>\n",
       "      <td>3</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.R736H</th>\n",
       "      <td>10.88</td>\n",
       "      <td>10.32</td>\n",
       "      <td>11.33</td>\n",
       "      <td>18.0</td>\n",
       "      <td>1.883910e-08</td>\n",
       "      <td>20</td>\n",
       "      <td>0.0067</td>\n",
       "      <td>0.0097</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0091</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0087</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.C497Y</th>\n",
       "      <td>10.64</td>\n",
       "      <td>8.79</td>\n",
       "      <td>11.63</td>\n",
       "      <td>2.0</td>\n",
       "      <td>3.213910e-09</td>\n",
       "      <td>3</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.R771*</th>\n",
       "      <td>10.63</td>\n",
       "      <td>10.05</td>\n",
       "      <td>11.10</td>\n",
       "      <td>16.0</td>\n",
       "      <td>1.883910e-08</td>\n",
       "      <td>12</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.G332R</th>\n",
       "      <td>10.58</td>\n",
       "      <td>8.91</td>\n",
       "      <td>11.67</td>\n",
       "      <td>4.0</td>\n",
       "      <td>6.254924e-09</td>\n",
       "      <td>2</td>\n",
       "      <td>0.0041</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0075</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.W297*</th>\n",
       "      <td>10.55</td>\n",
       "      <td>9.21</td>\n",
       "      <td>11.37</td>\n",
       "      <td>3.0</td>\n",
       "      <td>5.105670e-09</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.R598*</th>\n",
       "      <td>10.54</td>\n",
       "      <td>9.95</td>\n",
       "      <td>11.00</td>\n",
       "      <td>11.0</td>\n",
       "      <td>1.883910e-08</td>\n",
       "      <td>9</td>\n",
       "      <td>0.0059</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0061</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0121</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.R899C</th>\n",
       "      <td>10.50</td>\n",
       "      <td>8.26</td>\n",
       "      <td>11.96</td>\n",
       "      <td>2.0</td>\n",
       "      <td>5.336040e-09</td>\n",
       "      <td>-</td>\n",
       "      <td>0.0016</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.003</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.G699D</th>\n",
       "      <td>10.47</td>\n",
       "      <td>9.15</td>\n",
       "      <td>11.28</td>\n",
       "      <td>3.0</td>\n",
       "      <td>5.336040e-09</td>\n",
       "      <td>4</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.Q886*</th>\n",
       "      <td>10.40</td>\n",
       "      <td>8.74</td>\n",
       "      <td>11.36</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.959610e-09</td>\n",
       "      <td>2</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.Q374*</th>\n",
       "      <td>10.29</td>\n",
       "      <td>8.52</td>\n",
       "      <td>11.23</td>\n",
       "      <td>2.0</td>\n",
       "      <td>3.959610e-09</td>\n",
       "      <td>1</td>\n",
       "      <td>0.0017</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0155</td>\n",
       "      <td>0.0088</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.W313*</th>\n",
       "      <td>9.96</td>\n",
       "      <td>8.25</td>\n",
       "      <td>10.85</td>\n",
       "      <td>2.0</td>\n",
       "      <td>4.870030e-09</td>\n",
       "      <td>3</td>\n",
       "      <td>0.0049</td>\n",
       "      <td>0.0096</td>\n",
       "      <td>0.0116</td>\n",
       "      <td>0.006</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.W440*</th>\n",
       "      <td>9.89</td>\n",
       "      <td>8.19</td>\n",
       "      <td>10.77</td>\n",
       "      <td>2.0</td>\n",
       "      <td>5.105670e-09</td>\n",
       "      <td>2</td>\n",
       "      <td>0.0008</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0015</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.W601*</th>\n",
       "      <td>9.86</td>\n",
       "      <td>8.63</td>\n",
       "      <td>10.60</td>\n",
       "      <td>3.0</td>\n",
       "      <td>7.793320e-09</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.G890D</th>\n",
       "      <td>9.82</td>\n",
       "      <td>8.20</td>\n",
       "      <td>10.77</td>\n",
       "      <td>3.0</td>\n",
       "      <td>5.336040e-09</td>\n",
       "      <td>1</td>\n",
       "      <td>0.0008</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0015</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.G685R</th>\n",
       "      <td>9.58</td>\n",
       "      <td>7.95</td>\n",
       "      <td>10.42</td>\n",
       "      <td>2.0</td>\n",
       "      <td>6.254924e-09</td>\n",
       "      <td>4</td>\n",
       "      <td>0.0017</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.003</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.G332E</th>\n",
       "      <td>9.57</td>\n",
       "      <td>7.09</td>\n",
       "      <td>10.95</td>\n",
       "      <td>2.0</td>\n",
       "      <td>4.687470e-09</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.L547F</th>\n",
       "      <td>9.40</td>\n",
       "      <td>7.80</td>\n",
       "      <td>10.22</td>\n",
       "      <td>2.0</td>\n",
       "      <td>7.065730e-09</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.D702N</th>\n",
       "      <td>9.40</td>\n",
       "      <td>7.80</td>\n",
       "      <td>10.23</td>\n",
       "      <td>2.0</td>\n",
       "      <td>7.065730e-09</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.W581*</th>\n",
       "      <td>9.38</td>\n",
       "      <td>7.79</td>\n",
       "      <td>10.20</td>\n",
       "      <td>2.0</td>\n",
       "      <td>7.144750e-09</td>\n",
       "      <td>2</td>\n",
       "      <td>0.0027</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0016</td>\n",
       "      <td>0.0177</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0063</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.W893*</th>\n",
       "      <td>9.38</td>\n",
       "      <td>7.79</td>\n",
       "      <td>10.20</td>\n",
       "      <td>2.0</td>\n",
       "      <td>7.144750e-09</td>\n",
       "      <td>2</td>\n",
       "      <td>0.0016</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.003</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.W306*</th>\n",
       "      <td>9.26</td>\n",
       "      <td>7.69</td>\n",
       "      <td>10.06</td>\n",
       "      <td>2.0</td>\n",
       "      <td>7.793320e-09</td>\n",
       "      <td>1</td>\n",
       "      <td>0.0008</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0015</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.A662T</th>\n",
       "      <td>9.25</td>\n",
       "      <td>7.48</td>\n",
       "      <td>10.27</td>\n",
       "      <td>2.0</td>\n",
       "      <td>9.720770e-09</td>\n",
       "      <td>-</td>\n",
       "      <td>0.0008</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0015</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.W860*</th>\n",
       "      <td>9.21</td>\n",
       "      <td>7.02</td>\n",
       "      <td>10.27</td>\n",
       "      <td>2.0</td>\n",
       "      <td>4.870030e-09</td>\n",
       "      <td>1</td>\n",
       "      <td>0.0008</td>\n",
       "      <td>0.0096</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.W327*</th>\n",
       "      <td>9.12</td>\n",
       "      <td>6.93</td>\n",
       "      <td>10.14</td>\n",
       "      <td>2.0</td>\n",
       "      <td>5.105670e-09</td>\n",
       "      <td>2</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.R326H</th>\n",
       "      <td>9.02</td>\n",
       "      <td>8.16</td>\n",
       "      <td>9.62</td>\n",
       "      <td>7.0</td>\n",
       "      <td>1.883910e-08</td>\n",
       "      <td>2</td>\n",
       "      <td>0.0025</td>\n",
       "      <td>0.0096</td>\n",
       "      <td>0</td>\n",
       "      <td>0.003</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.R771Q</th>\n",
       "      <td>8.52</td>\n",
       "      <td>6.38</td>\n",
       "      <td>9.52</td>\n",
       "      <td>2.0</td>\n",
       "      <td>7.065730e-09</td>\n",
       "      <td>4</td>\n",
       "      <td>0.0025</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.003</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.R899H</th>\n",
       "      <td>8.21</td>\n",
       "      <td>6.73</td>\n",
       "      <td>9.04</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.883910e-08</td>\n",
       "      <td>3</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.R688H</th>\n",
       "      <td>8.21</td>\n",
       "      <td>6.73</td>\n",
       "      <td>9.03</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.883910e-08</td>\n",
       "      <td>1</td>\n",
       "      <td>0.0033</td>\n",
       "      <td>0.0096</td>\n",
       "      <td>0</td>\n",
       "      <td>0.003</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0061</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.R659H</th>\n",
       "      <td>8.09</td>\n",
       "      <td>6.72</td>\n",
       "      <td>8.78</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.883910e-08</td>\n",
       "      <td>-</td>\n",
       "      <td>0.0017</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0015</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0062</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.R366H</th>\n",
       "      <td>8.09</td>\n",
       "      <td>6.72</td>\n",
       "      <td>8.78</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.883910e-08</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>102 rows  13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          s (%)  s lower value (%)  s upper value (%)  observed number  \\\n",
       "p.E30A   196.75              28.05             196.75              4.0   \n",
       "p.F751C   30.59              18.09             196.75              2.0   \n",
       "p.F732C   22.84              16.11              29.32              2.0   \n",
       "p.F414C   21.61              15.05              27.52              2.0   \n",
       "p.R729G   21.22              16.27              25.69              3.0   \n",
       "p.W795S   20.33              14.30              25.16              2.0   \n",
       "p.Y908C   19.88              15.03              24.63              4.0   \n",
       "p.Y735C   19.66              18.31              20.94             23.0   \n",
       "p.R882C   19.41              18.54              20.26             61.0   \n",
       "p.P507R   19.02              13.63              23.05              2.0   \n",
       "p.L737R   18.75              13.80              22.43              2.0   \n",
       "p.M880V   18.74              16.09              21.07              6.0   \n",
       "p.S714C   18.45              16.30              20.33              7.0   \n",
       "p.R882P   18.25              15.62              20.55              6.0   \n",
       "p.G543A   17.69              13.43              20.73              2.0   \n",
       "p.F755S   17.66              14.96              19.86              4.0   \n",
       "p.W860R   17.64              16.34              18.83             18.0   \n",
       "p.E599D   17.30              14.38              19.62              4.0   \n",
       "p.S770*   17.27              12.99              20.26              2.0   \n",
       "p.N838D   17.09              13.07              19.74              2.0   \n",
       "p.R882S   17.00              11.76              20.51              2.0   \n",
       "p.W753C   16.82              12.41              19.84              2.0   \n",
       "p.N501S   16.55              12.73              19.01              2.0   \n",
       "p.M761V   16.45              13.82              18.57              4.0   \n",
       "p.E733G   16.09              13.09              18.17              3.0   \n",
       "p.L547H   15.90              12.41              18.21              2.0   \n",
       "p.S770W   15.63              12.94              17.59              3.0   \n",
       "p.R882H   15.58              15.15              15.99            105.0   \n",
       "p.R326S   15.34              11.65              17.80              2.0   \n",
       "p.F732S   14.96              11.84              16.95              2.0   \n",
       "...         ...                ...                ...              ...   \n",
       "p.R736H   10.88              10.32              11.33             18.0   \n",
       "p.C497Y   10.64               8.79              11.63              2.0   \n",
       "p.R771*   10.63              10.05              11.10             16.0   \n",
       "p.G332R   10.58               8.91              11.67              4.0   \n",
       "p.W297*   10.55               9.21              11.37              3.0   \n",
       "p.R598*   10.54               9.95              11.00             11.0   \n",
       "p.R899C   10.50               8.26              11.96              2.0   \n",
       "p.G699D   10.47               9.15              11.28              3.0   \n",
       "p.Q886*   10.40               8.74              11.36              3.0   \n",
       "p.Q374*   10.29               8.52              11.23              2.0   \n",
       "p.W313*    9.96               8.25              10.85              2.0   \n",
       "p.W440*    9.89               8.19              10.77              2.0   \n",
       "p.W601*    9.86               8.63              10.60              3.0   \n",
       "p.G890D    9.82               8.20              10.77              3.0   \n",
       "p.G685R    9.58               7.95              10.42              2.0   \n",
       "p.G332E    9.57               7.09              10.95              2.0   \n",
       "p.L547F    9.40               7.80              10.22              2.0   \n",
       "p.D702N    9.40               7.80              10.23              2.0   \n",
       "p.W581*    9.38               7.79              10.20              2.0   \n",
       "p.W893*    9.38               7.79              10.20              2.0   \n",
       "p.W306*    9.26               7.69              10.06              2.0   \n",
       "p.A662T    9.25               7.48              10.27              2.0   \n",
       "p.W860*    9.21               7.02              10.27              2.0   \n",
       "p.W327*    9.12               6.93              10.14              2.0   \n",
       "p.R326H    9.02               8.16               9.62              7.0   \n",
       "p.R771Q    8.52               6.38               9.52              2.0   \n",
       "p.R899H    8.21               6.73               9.04              2.0   \n",
       "p.R688H    8.21               6.73               9.03              2.0   \n",
       "p.R659H    8.09               6.72               8.78              2.0   \n",
       "p.R366H    8.09               6.72               8.78              2.0   \n",
       "\n",
       "         site-specific mutation rate Number of times in COSMIC  \\\n",
       "p.E30A                  1.864760e-10                         3   \n",
       "p.F751C                 1.566770e-10                         -   \n",
       "p.F732C                 1.859550e-10                         1   \n",
       "p.F414C                 1.566770e-10                         -   \n",
       "p.R729G                 2.470660e-10                         -   \n",
       "p.W795S                 3.739340e-10                         -   \n",
       "p.Y908C                 1.473330e-09                         -   \n",
       "p.Y735C                 1.839360e-09                        12   \n",
       "p.R882C                 5.336040e-09                       460   \n",
       "p.P507R                 4.563380e-10                         -   \n",
       "p.L737R                 1.864760e-10                         1   \n",
       "p.M880V                 6.735990e-10                         7   \n",
       "p.S714C                 7.420070e-10                        25   \n",
       "p.R882P                 6.302210e-10                        36   \n",
       "p.G543A                 2.470660e-10                         5   \n",
       "p.F755S                 4.962520e-10                         3   \n",
       "p.W860R                 2.463301e-09                         7   \n",
       "p.E599D                 1.244421e-09                         -   \n",
       "p.S770*                 2.488650e-10                         2   \n",
       "p.N838D                 4.962520e-10                         5   \n",
       "p.R882S                 4.385790e-10                        40   \n",
       "p.W753C                 6.856450e-10                         -   \n",
       "p.N501S                 5.656330e-10                         5   \n",
       "p.M761V                 6.735990e-10                         -   \n",
       "p.E733G                 8.133640e-10                         1   \n",
       "p.L547H                 3.786790e-10                         5   \n",
       "p.S770W                 5.511000e-10                         5   \n",
       "p.R882H                 1.883910e-08                      1069   \n",
       "p.R326S                 4.385790e-10                         -   \n",
       "p.F732S                 4.962520e-10                         3   \n",
       "...                              ...                       ...   \n",
       "p.R736H                 1.883910e-08                        20   \n",
       "p.C497Y                 3.213910e-09                         3   \n",
       "p.R771*                 1.883910e-08                        12   \n",
       "p.G332R                 6.254924e-09                         2   \n",
       "p.W297*                 5.105670e-09                         1   \n",
       "p.R598*                 1.883910e-08                         9   \n",
       "p.R899C                 5.336040e-09                         -   \n",
       "p.G699D                 5.336040e-09                         4   \n",
       "p.Q886*                 3.959610e-09                         2   \n",
       "p.Q374*                 3.959610e-09                         1   \n",
       "p.W313*                 4.870030e-09                         3   \n",
       "p.W440*                 5.105670e-09                         2   \n",
       "p.W601*                 7.793320e-09                         1   \n",
       "p.G890D                 5.336040e-09                         1   \n",
       "p.G685R                 6.254924e-09                         4   \n",
       "p.G332E                 4.687470e-09                         -   \n",
       "p.L547F                 7.065730e-09                         1   \n",
       "p.D702N                 7.065730e-09                         -   \n",
       "p.W581*                 7.144750e-09                         2   \n",
       "p.W893*                 7.144750e-09                         2   \n",
       "p.W306*                 7.793320e-09                         1   \n",
       "p.A662T                 9.720770e-09                         -   \n",
       "p.W860*                 4.870030e-09                         1   \n",
       "p.W327*                 5.105670e-09                         2   \n",
       "p.R326H                 1.883910e-08                         2   \n",
       "p.R771Q                 7.065730e-09                         4   \n",
       "p.R899H                 1.883910e-08                         3   \n",
       "p.R688H                 1.883910e-08                         1   \n",
       "p.R659H                 1.883910e-08                         -   \n",
       "p.R366H                 1.883910e-08                         -   \n",
       "\n",
       "        Exac Overall freq (%) Exac African freq (%) Exac East Asian freq (%)  \\\n",
       "p.E30A                 0.2959                  0.02                        0   \n",
       "p.F751C                     -                     -                        -   \n",
       "p.F732C                     -                     -                        -   \n",
       "p.F414C                     -                     -                        -   \n",
       "p.R729G                0.0009                0.0097                        0   \n",
       "p.W795S                     -                     -                        -   \n",
       "p.Y908C                0.0033                     0                        0   \n",
       "p.Y735C                0.0118                0.0097                        0   \n",
       "p.R882C                0.0355                0.0677                   0.0347   \n",
       "p.P507R                     -                     -                        -   \n",
       "p.L737R                     -                     -                        -   \n",
       "p.M880V                0.0017                0.0097                        0   \n",
       "p.S714C                0.0041                     0                        0   \n",
       "p.R882P                0.0058                 0.029                   0.0116   \n",
       "p.G543A                     -                     -                        -   \n",
       "p.F755S                0.0041                     0                        0   \n",
       "p.W860R                     -                     -                        -   \n",
       "p.E599D                0.0034                     0                        0   \n",
       "p.S770*                0.0017                     0                        0   \n",
       "p.N838D                     -                     -                        -   \n",
       "p.R882S                     -                     -                        -   \n",
       "p.W753C                0.0008                     0                        0   \n",
       "p.N501S                0.0294                0.0098                        0   \n",
       "p.M761V                0.0008                     0                        0   \n",
       "p.E733G                     -                     -                        -   \n",
       "p.L547H                0.0017                     0                        0   \n",
       "p.S770W                     -                     -                        -   \n",
       "p.R882H                0.0545                0.0869                   0.0694   \n",
       "p.R326S                0.0008                     0                        0   \n",
       "p.F732S                     -                     -                        -   \n",
       "...                       ...                   ...                      ...   \n",
       "p.R736H                0.0067                0.0097                        0   \n",
       "p.C497Y                     -                     -                        -   \n",
       "p.R771*                     -                     -                        -   \n",
       "p.G332R                0.0041                     0                        0   \n",
       "p.W297*                     -                     -                        -   \n",
       "p.R598*                0.0059                  0.01                        0   \n",
       "p.R899C                0.0016                     0                        0   \n",
       "p.G699D                     -                     -                        -   \n",
       "p.Q886*                     -                     -                        -   \n",
       "p.Q374*                0.0017                     0                        0   \n",
       "p.W313*                0.0049                0.0096                   0.0116   \n",
       "p.W440*                0.0008                     0                        0   \n",
       "p.W601*                     -                     -                        -   \n",
       "p.G890D                0.0008                     0                        0   \n",
       "p.G685R                0.0017                     0                        0   \n",
       "p.G332E                     -                     -                        -   \n",
       "p.L547F                     -                     -                        -   \n",
       "p.D702N                     -                     -                        -   \n",
       "p.W581*                0.0027                     0                        0   \n",
       "p.W893*                0.0016                     0                        0   \n",
       "p.W306*                0.0008                     0                        0   \n",
       "p.A662T                0.0008                     0                        0   \n",
       "p.W860*                0.0008                0.0096                        0   \n",
       "p.W327*                     -                     -                        -   \n",
       "p.R326H                0.0025                0.0096                        0   \n",
       "p.R771Q                0.0025                     0                        0   \n",
       "p.R899H                     -                     -                        -   \n",
       "p.R688H                0.0033                0.0096                        0   \n",
       "p.R659H                0.0017                     0                        0   \n",
       "p.R366H                     -                     -                        -   \n",
       "\n",
       "        Exac European freq (%) Exac Finnish freq (%) Exac Latino freq (%)  \\\n",
       "p.E30A                  0.3033                2.1424               0.1413   \n",
       "p.F751C                      -                     -                    -   \n",
       "p.F732C                      -                     -                    -   \n",
       "p.F414C                      -                     -                    -   \n",
       "p.R729G                      0                     0                    0   \n",
       "p.W795S                      -                     -                    -   \n",
       "p.Y908C                 0.0045                     0               0.0086   \n",
       "p.Y735C                 0.0182                     0               0.0087   \n",
       "p.R882C                 0.0391                     0               0.0087   \n",
       "p.P507R                      -                     -                    -   \n",
       "p.L737R                      -                     -                    -   \n",
       "p.M880V                 0.0015                     0                    0   \n",
       "p.S714C                 0.0075                     0                    0   \n",
       "p.R882P                  0.003                     0               0.0087   \n",
       "p.G543A                      -                     -                    -   \n",
       "p.F755S                  0.006                     0               0.0086   \n",
       "p.W860R                      -                     -                    -   \n",
       "p.E599D                 0.0061                     0                    0   \n",
       "p.S770*                  0.003                     0                    0   \n",
       "p.N838D                      -                     -                    -   \n",
       "p.R882S                      -                     -                    -   \n",
       "p.W753C                 0.0015                     0                    0   \n",
       "p.N501S                 0.0321                0.0457               0.0174   \n",
       "p.M761V                 0.0015                     0                    0   \n",
       "p.E733G                      -                     -                    -   \n",
       "p.L547H                      0                     0               0.0087   \n",
       "p.S770W                      -                     -                    -   \n",
       "p.R882H                 0.0571                0.0758                    0   \n",
       "p.R326S                      0                0.0151                    0   \n",
       "p.F732S                      -                     -                    -   \n",
       "...                        ...                   ...                  ...   \n",
       "p.R736H                 0.0091                     0               0.0087   \n",
       "p.C497Y                      -                     -                    -   \n",
       "p.R771*                      -                     -                    -   \n",
       "p.G332R                 0.0075                     0                    0   \n",
       "p.W297*                      -                     -                    -   \n",
       "p.R598*                 0.0061                     0                    0   \n",
       "p.R899C                  0.003                     0                    0   \n",
       "p.G699D                      -                     -                    -   \n",
       "p.Q886*                      -                     -                    -   \n",
       "p.Q374*                      0                0.0155               0.0088   \n",
       "p.W313*                  0.006                     0                    0   \n",
       "p.W440*                 0.0015                     0                    0   \n",
       "p.W601*                      -                     -                    -   \n",
       "p.G890D                 0.0015                     0                    0   \n",
       "p.G685R                  0.003                     0                    0   \n",
       "p.G332E                      -                     -                    -   \n",
       "p.L547F                      -                     -                    -   \n",
       "p.D702N                      -                     -                    -   \n",
       "p.W581*                 0.0016                0.0177                    0   \n",
       "p.W893*                  0.003                     0                    0   \n",
       "p.W306*                 0.0015                     0                    0   \n",
       "p.A662T                 0.0015                     0                    0   \n",
       "p.W860*                      0                     0                    0   \n",
       "p.W327*                      -                     -                    -   \n",
       "p.R326H                  0.003                     0                    0   \n",
       "p.R771Q                  0.003                     0                    0   \n",
       "p.R899H                      -                     -                    -   \n",
       "p.R688H                  0.003                     0                    0   \n",
       "p.R659H                 0.0015                     0                    0   \n",
       "p.R366H                      -                     -                    -   \n",
       "\n",
       "        Exac South Asian freq (%)  \n",
       "p.E30A                     0.0247  \n",
       "p.F751C                         -  \n",
       "p.F732C                         -  \n",
       "p.F414C                         -  \n",
       "p.R729G                         0  \n",
       "p.W795S                         -  \n",
       "p.Y908C                         0  \n",
       "p.Y735C                         0  \n",
       "p.R882C                    0.0364  \n",
       "p.P507R                         -  \n",
       "p.L737R                         -  \n",
       "p.M880V                         0  \n",
       "p.S714C                         0  \n",
       "p.R882P                         0  \n",
       "p.G543A                         -  \n",
       "p.F755S                         0  \n",
       "p.W860R                         -  \n",
       "p.E599D                         0  \n",
       "p.S770*                         0  \n",
       "p.N838D                         -  \n",
       "p.R882S                         -  \n",
       "p.W753C                         0  \n",
       "p.N501S                    0.0502  \n",
       "p.M761V                         0  \n",
       "p.E733G                         -  \n",
       "p.L547H                    0.0061  \n",
       "p.S770W                         -  \n",
       "p.R882H                    0.0485  \n",
       "p.R326S                         0  \n",
       "p.F732S                         -  \n",
       "...                           ...  \n",
       "p.R736H                         0  \n",
       "p.C497Y                         -  \n",
       "p.R771*                         -  \n",
       "p.G332R                         0  \n",
       "p.W297*                         -  \n",
       "p.R598*                    0.0121  \n",
       "p.R899C                         0  \n",
       "p.G699D                         -  \n",
       "p.Q886*                         -  \n",
       "p.Q374*                         0  \n",
       "p.W313*                         0  \n",
       "p.W440*                         0  \n",
       "p.W601*                         -  \n",
       "p.G890D                         0  \n",
       "p.G685R                         0  \n",
       "p.G332E                         -  \n",
       "p.L547F                         -  \n",
       "p.D702N                         -  \n",
       "p.W581*                    0.0063  \n",
       "p.W893*                         0  \n",
       "p.W306*                         0  \n",
       "p.A662T                         0  \n",
       "p.W860*                         0  \n",
       "p.W327*                         -  \n",
       "p.R326H                         0  \n",
       "p.R771Q                         0  \n",
       "p.R899H                         -  \n",
       "p.R688H                    0.0061  \n",
       "p.R659H                    0.0062  \n",
       "p.R366H                         -  \n",
       "\n",
       "[102 rows x 13 columns]"
      ]
     },
     "execution_count": 238,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "DNMT3A_df_sorted"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 269,
   "metadata": {
    "code_folding": [
     0
    ]
   },
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAABZoAAAHvCAYAAADdMGqTAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDIuMi4zLCBodHRwOi8vbWF0cGxvdGxpYi5vcmcvIxREBQAAIABJREFUeJzs3Xd4FNX+x/H3CSEhoSRACEgNiIAVBRUuqCBNQQXBAsIVEMu9oiBXxateNQT0J1wsgAp6UQQUuxRFaUpHQHrvkEAogRBCC4GU8/tjdtfd7CYEQgn6eT3PPLt75syZM3N2RvzuyXeMtRYRERERERERERERkbMVdLE7ICIiIiIiIiIiIiKXNgWaRURERERERERERKRAFGgWERERERERERERkQJRoFlERERERERERERECkSBZhEREREREREREREpEAWaRURERERERERERKRAFGgWERERuUiMMcHGGGuM+fhi96UwMsb0MsZsMsaccp2nyq7y5saYxcaYY67yv1/svoqIiIiI/NUp0CwiIiJyjhljmroCoLktDfPYNsgY088Y0/ZC9rmwMca0BIYBa4EngIeBFGNMFDAeKAY86ypfcJ76UM81FlXPR/t/VcaYmjmuhyxjTKrrR4WvjDH3GWOCA2zXwmub7gHWu3+4mZijfL6r/JQxplwuffrAq+1bvNrK71LZGFPBGPOpMWaNMeaQMeaEMWaLMWakMabGac5Jb1c7h4wxYWd4SkVEREQKBb9/wImIiIjIOfMl8HOA8q0A1tpMV1Ap02tdEBALfAL8cN57WHi1dL0+Yq094i40xjQFSgGvWmvP9/mphzMWvwA7z/O+/oqmAZ+73pcAagJ3AR2BpcaY9tbaxFy27W+M+cpam57PfWXgXFt/B971XmGMKQY8BKTj/IABkIXzI4a3psCjwAjgtxzrUoDqwBXAVJzvywmgNtADeNAYc7O1dlMu/esBbAMuBzoA4/J5XCIiIiKFhgLNIiIiIufPcmvt53lVOINA2V9NBSDLO8jsVQ5OYE/yYIwxQHFr7bGL3ZdcbMx5fRhj+gLPAYOBycaY+tbarBzbLQVuBHq56uVHGk5w+BFyBJqB9kBp4AugM4C11vJHENzdt2I4gebfcrmu1wG35Cw0xox37fspoHeA9fWBuq59v4gTdFagWURERC45Sp0hIiIicpHkzNFsjKmJM/MS4FGvP8vPzFnf9ef984wxacaYZGPM/4wxxQPso5Ix5kNjzC5X6oDdrs9ROeqVNcYMNcZsN8akG2MOGmOWGmOezVGvuzFmiSvVwXFjzDZjzOfGmLL5POY7jDEzjDGHXftZZYx5wmt9TWOMxZlNWsTrHPxijEnEmekNMM/73Li2LWaMecUYs97V9iFjzA/GmLoB+mGMMf8wxvxunFzPR40xq40xsa71rwMjc+zLe6zCjDH9jZPuIc21rzXGmIH5OAfBrn7OM8YkucYlwZW+oUwu2zxgjJnjOm9pxpiNrvEKdq13p5V42Di5rTcAJ4E+Xm00dZ3HI642lpnAKSiuNcZ87/qunDTG7DXGzDTGtPaqc9bHnxfreAv4Gif4+kCAal8Aq4CXjDGlz6D5T4FrXYFdb48Ay4A1Z9Hl/EhwvebW10eBI8BEYAxwuzGm+nnqi4iIiMh5oxnNIiIiIudPeM6ALnDSWns0l/r7gG44wabZ/BFUzc5Rrz7OLMxPcGZdNgMex0nB0dNdyRWs+g0o4qq7HedP+5/ECWbd5DVjeDzwN+BDYDVQHLgSJ13AO672HgFGAXOA13BSA1QF2gBRwMG8ToYx5kngA1efBuDMMr0D+MgYU91a+5LrHDzs6mND1/kA2IuTMuMunMDcAGCz+9wYY0KA6UADYCxOfufSrvPymzHmFmvtClddg5PWpCOwEHgDSHUd731AHPAtzuxp732BK+2J6zx1BUa72ijqOrfN8joHLsVwZu1+jxNcPA7cjJOLurFrXNw/OGCMGQS8gDNj9m0gCSfFwn3AfwDvGcvPuY77E9e5THC1cS/wnes8Dnbt8yHgU9e5dwfYywEzcVJHfIiTAiIKuMnVxynn4Pjz42Oc8bkL+CrHOosz83cK8BLOucmPSTjf0R44gWWMk3+7Oc7s6FIF7jWe72IpIAQnHUica5VfGh3zR9qOb6y1J4wxnwODcILfr52L/oiIiIhcMNZaLVq0aNGiRYsWLedwwQnO2lyWr7zqBbvKPs6rLMC6LODGHOum4cxgDfMq+wkn2FgxR90GrjZecX0u42p32GmO6weclBVFzuKcVHb1b2yAdR/gBMmreZV9DmQGqPuYq6+35CjvixN0bpGjPBJIBH7xKuvsamM0EJSjftDp9uVadwT44Sy/H0He4+RV/g/X/jp4lTVylc0AQnPUN17vW7jqJQNROeoVdZ2DFKCCV3kosMj1XajhKuuQsw+5HENBjr+max9D8qgT7aqzOMAx9nF9/hXnx47KOa6PiTnamg+kut4PdZ2HUNdn9w8mkTjB64DjneP78PfTHN+9+F7ze4FncqnbJec+gR9xAvxBee1HixYtWrRo0aKlsC1KnSEiIiJy/vwP56F23svr56Dd+dbapTnKZuLMoKwG4ErB0BpnxuwpY0yUe8F56NgOoJVr2zSclB0NjTHV8tjvYaAk0MY1K/hMPODq3yjvvrj68yPOrOvmZ9imt7/jzPhdmaPtYJyAZBNjTKirrju497y11me2eM7PeTiMk4bh6jPtqLU221p7AsAYU8QYE+nq60xXlQZe1bu4Xl+01p7M0Y4N0Pxoa21yjrKbgEo4P17s89r+JPAWTuC7rddxgTPGJfM4jLM+/nxyz7TPa5bxv3GC5f3PoN1RODO+73V9h7sBE6y1qWfVy8AW4FzrbXFmXO8DyrjTnOTwKLDNWjvfq2w0UIU/HogpIiIicklQ6gwRERGR82eLtfaX89Du9gBl7rQV7lzJdQCDM0v2H7m0kwXOAwmNk4v5HSDeGLMOJ+g5wVo7y6v+6zgPO/sBSDbGzMFJX/C1Pf0D5650vc7Ko07507SRlzo4gewDedQpgzO79AogMUBA9kw8g5PiZK0xZhvOcf0ATM4lAOzDGNMJJ81FXZwZx968c/legTNOq/PZr80Bytz5ftcFWLfW9VrD9ToTJwfyo0BXY8wSnJQkX1trN3ptV6Djzwd3gDnnwyA9rLVLjTHfAt2MMW8Dm07XqLV2lTFmBU5qiv04x/3Pc9Bf730cANzX/Y+udBircVKQPOWuZ4ypgfPXDyOMk5/dbQNOapNHcf5SQUREROSSoECziIiIyKUnK491JsfrGJw0FIGkud9Ya983xkzAyYl7G/Ag0MsYM85a+3dXnU3GmDo4KQyaA01wcun2M8bcZq3dkY9+dcEJ8AWyNZfy/AgCVuKk0MhNildfChQMtdaOdwXa2+Cch5Y4qRVmG2NaWa8cyzkZYx7EyRG9COiNk9YiHSdQ/hO+D+w+05njaQHK8t2GK0jcxfVQv9bArTg5kF81xvSy1o5w1Tvr48+n61yvpwse/wcnX/lA12t+jMJJoVEUJ0XFr2fTwfyy1iYaY2YCjxtj+nidmx44Y9MTr9zqXtoZY8paa/PMfS4iIiJSWCjQLCIiIlK4nIvZoOAEbS1QNL+zqq21u3HSffzP9Wf+43CCjm9b14P0XOkWfnItGGPa4jxkrQ/OLNfcbHG9HjhPs7y3AOWAX/Mxo3YTTmqIqNPMas6zHVcA8DPgM1cahsE4s5TvBibksenDOAHh26216e5CY8w1ufS1BXAtsDyv/uRhm+s1UJqLq1yvPrPkrbVrgDXAf11pWH7HCeaO8KpztsefH4+5Xn/Kq5K1dqsx5n84M4VvzWfb43BShjQD+p9BupSCCMMJbJcADhljgnDSdizDOa85VQKG4KSEGXoB+iciIiJSYMrRLCIiIlKIWGuzcGa3lilgO0k4KQ8eMMbclHO9cZRzvQ83xoTl2D4TJ9CIuy+uPMI5Lfeuk4evgVNAf2NMsQD9iTTGhJymjbyMxQnOBQx2G2O803KMw5lJ+t+cuaZzfHanAymTo06wMSbCu8wV3F4ZqH4AWTgPLvT8W9y131cC1P3C9fpmzvNzBnmylwC7gUeNMdFe24cAz7v68oOrrEzOdq21KUA8UMIYE3IOjj9PxpjngI7ACuC7fGzSH2esBuWnfWvtIZx0GXHAyLPspp8c3zHv8mtxUmRsdu0b4A6cB2SOtdZ+F2AZCuzCmfUsIiIicknQjGYRERGRwmcRcIcx5gWcYFOWtfabs2jnH8B8YL4xZixO4C4YJ2fvvcAnOHmXrwJ+caXOWAcccpU9iTMbdoGrvZnGmAPAPFe/yuDkus0m9/QcAFhrE4wxTwMfAutdeWt34sxCvg5oB9TCSSNxNt7Bmfn7rjGmBTAbOApUxUnzcZQ/Hq72FdDB1fdaxpgfcR5uV8tVt66r3u84s5pfcQXlj/PHgxR3GmN+wAmuunP9PomTniPPWbg4wdN2OOfzM5wH2rUH/ALw1trfXPmHnwPcOYn34YzhA8aYG06XH9tam2mM6QV8Cywxxox0HUsn4GacWb3uGc09gKdd34WtQCZwu+u8fGGtPeX6waEgx+9Wxxjzd9f74sDlwD04+bZ/B9q7fnjJk7V2v+scxeZzv1hrR+e37hl41RjTBPgZJzBfBGcm+t9d75/yqvsozncrr5nf44FnjDE3Bnj4p4iIiEiho0CziIiISOHzT+B9nBmuJXFmwJ5xoNkV3K0HvAi0BbripGzYBUzkj9miCTi5nJviBDxDcQK+HwGDvNI7fAA84OpfGZwHEC4HelprZ+ejPyONMRtwZtE+CUQAycBGnFy7eT3I73RtnzLG3Ak8jRPYi3Ot2oMTuB/jVdcaYzriBP564AQos3DSR3zrVW+HMeYxnBzFI3BSH3zi2m4YTvC1JU6QdC9O0PBNa+2+0/T1c2NMCZzZ12/jBGd/AF4lQP5qa+3zrgfYPQX8G2c29i5gMs7s9/ycnwnGmJY436l/4/x/wAagh7X2U6+qM3EC//cAl+EEmnfgBLrfd9U5WpDj93KHa7E4M5L34XyfXsV5EOVpg8xe3sL5XhbkgZIF9SNQEWc2djROcDkR54eNt6y1G8DzlwH3AEustbvyaO97nO9ID0CBZhERESn0zLl5KPT5YYwZBNyIM7skCjiB8z9CE4H3vR+MYYyJwflHcG6+ttZ2Om+dFREREREREREREfmLKuyB5lM4sxrW48zuKA40xAk+7wEaumcBeAWaV+EEonNaa63NT443ERERERERERERETkDhT11RinvJ3G7GWPeAF4GXgJ65li90lrb7wL0TURERERERERERETwetJ1YRQoyOzizlF4xYXqi4iIiIiIiIiIiIgEVthnNOfmHtfr6gDrKhpj/gGUxXlAzUJrbaB6IiIiIiIiIiIiInIOFOoczW7GmOeBEjhPJr8RuAUnyNzCWnvAVSeG3B8GOBvoZq3debp9jRgxwgI8+eSTpqD9FhEREREREREREfkrKNSpM7w8D8QCfXCCzFOBVu4gs0saMACoD5R2LU2AWUBT4FdjTPHcdmCMecIYs9T9+fDhw3bFihU2IyPDLl261AJ26dKl9uTJk3bVqlX20KFDduvWrTYxMdHu3bvXbtq0yR45csSuWLHCZmZm+m2zZs0am5KSYrdt22Z37dpl9+3bZzdu3GiPHj1qly1bZrOysny2OXHihF27dq09ePCg3b59u925c6dNSkqyGzZssMeOHbNLly612dnZdsiU3+3gjZl2yJTfbVpaml23bp09cOCA3bFjh01ISLD79++369evt8ePH/e0v2TJEs9+jh8/btevX2/3799vExIS7I4dO+yBAwfsunXrbFpamk+fsrOz7dKlS+2xY8fshg0bbFJSkt25c6fdvn27PXjwoF27dq09ceKEzzZZWVl22bJl9ujRo3bjxo123759dteuXXbbtm02JSXFrlmzxp48edJnm8zMTLtixQp75MgRu2nTJrt3716bmJhot27dag8dOmRXrVrlt01GRoZdsWKFPXz4sN28ebPds2eP3b17t92yZYtNTU21K1eutBkZGT7HfurUKc9Ybtmyxe7evdvu2bPHbt682XqPv3ubJUuW2JMnT9rVq1d7xjLn+C9fvtxvLNPT0+2aNWt8xjLn+LvPbc7xT05O9oxlUlKSXb9+vWf8rbU+23iPf3x8vI2Pj7cHDhzwG3/v12PHjvmNf3Jysl27dq1NS0vzOV/Z2dl22bJleY5/enq6z/csKyvLLl++3B45ciTg+K9evTrX8T98+LDdtGmT3bNnj9/4nzp1ym/8V65caVNTU+3mzZvt7t277e7du+3mzZt9xt97G+/xd1/Le/bssZs2bfKMf6Br2Xv8va9l7/H3Pm/p6el5Xss5x3/JkiU2LS3Nb/zd17J7/HOOpftazjn+ga5l93fHPf5JSUl+45/zWnaPv/e17D3+a9as8Rl/9/XvHn/vaznn+HtfY3ldy4HG372Ne/zzupYL8708t2tZ93Ldy3Uv171c93Ldy3Uv171c93LdywvjvXzI9OV28No0O2TK77qX616ue/klcC/nArkkZjS7GWPKA42AgUBJ4G5r7fLTbBMMzAcaAH2stUPzqu81o/mc9Pl8G7wx0/O+b51LNROKiIiIiIiIiFwqFIsQueRckMwNl8qMZgCstUnW2glAK5wczGPzsU0m8LHr423nsXsiIiIiIiIiIiIif0mXVKDZzVqbAKwHrjbGROVjE3eKjVxTZ4iIiIiIiIiIiIjI2bkkA80uFV2vWfmo29D1uv089UVERERERERERETkL6vQBpqNMXWMMRUClAcZY94AooHfrLWHXOUNjDEhAeo3A/7l+vj5+eyziIiIiIiIiIiIyF9RYc7Yficw2BgzF9gGHATKA02AGsA+4HGv+oNwUmnMBhJdZdcBzVzvX7XW/nYB+i0iIiIiIiIiIiLyl1KYA82/AP8DGgN1gUjgOLAZ+AwYZq1N8ar/GdAeuAloDRQFkoBvgPettfMuXNdFRERERERERERE/joKbaDZWrsWeOoM6n8CfHL+eiQiIiIiIiIiIiIigRTaHM0iIiIiIiIiIiIicmlQoFlERERERERERERECkSBZhEREREREREREREpEAWaRURERERERERERKRAFGgWERERERERERERkQJRoFlERERERERERERECkSBZhEREREREREREREpEAWaRURERERERERERKRAFGgWERERERERERERkQJRoFkKlX79+mGMYfbs2Re7KyIiIiIiIiIiIpJPCjSLiIiIiIiIiIiISIEo0CyFytNPP82GDRu4+eabL3ZXREREREREREREJJ+CL3YHRLxFRUURFRV1sbshIiIiIiIiIiIiZ0AzmqXA4uPjMcbQvXt3tm3bxv3330/ZsmUpWbIkrVq1Yu3atQAcOHCAJ554gssuu4xixYpx0003MWvWLJ+2csvRPG/ePO655x4qV65MaGgoFSpUoGHDhsTFxfnUS0pK4vnnn6d27doUL16cyMhIateuTffu3dm+fbtf36dNm0abNm2IiooiNDSUyy+/nL59+5KamupXNyYmhpiYGNLS0ujbty9Vq1YlNDSUmjVrMmjQIKy1ftv88MMPNG/enMsuu4zQ0FAqVqxIkyZNGD58uF/dlJQUXnrpJa688krCwsKIiIigefPmTJ8+/bRjICIiIiIiIiIicjFpRrOcM/Hx8TRo0IArr7yS7t27Ex8fz4QJE2jatCkLFy7kzjvvpFSpUnTs2JGUlBS++uorWrduzebNm6latWqu7U6dOpW77rqLUqVK0bZtWypVqkRKSgobNmxg+PDhxMbGApCWlkbjxo3Ztm0bLVu25J577sFaS0JCApMmTeL++++nRo0annb79+9PbGwsZcqU4e677yY6OprVq1fz1ltv8fPPP7Nw4UJKlSrl05eMjAxatWrFnj17aN26NcHBwUycOJEXX3yR9PR0T18A/ve///GPf/yDChUqcM899xAVFcX+/ftZvXo1n376KT179vTUTUhIoGnTpsTHx3Prrbdy5513cvz4cSZPnsydd97JRx99xOOPP36uhkpEREREREREROScUqBZzpk5c+bw+uuv85///MdTNmDAAF577TUaNGjAgw8+yPDhwwkKcibSt2zZkq5du/Luu+/y7rvv5truyJEjyc7OZvbs2dStW9dnXXJysuf9r7/+yrZt2+jTp49fe6dOneLkyZOez7NmzSI2Npa//e1v/Pzzz0RGRnrWjR49mkceeYTY2Fi/dvbs2UPdunWZMWMGYWFhAMTGxlKrVi3effddXn75ZYoWLQrARx99REhICKtWrSI6OjrXfgN069aNhIQEvvzySzp16uQpT01NpWnTpvTu3Zu2bdtSvnz5XM+TiIiIiIiIiIjIxaJA83nSeVHGJbPPLxoWPSf7j4mJ4cUXX/Qp69atG6+99honT55k8ODBniAzQOfOnenRowcrV67MV/vuwK63QPmcA9ULCQkhJCTE83nYsGGAE8T2DjIDdO/enaFDhzJu3LiAAfBhw4b57CM6Opp27doxduxYNm3axDXXXONZFxwc7Ak859bvVatWMWfOHO6//36fIDNAZGQkcXFx3HvvvXz//fc+s6BFREREREREREQKCwWa5Zy5/vrrKVKkiE9ZxYoVAahVqxYlS5b0WVekSBHKly9PYmJinu126dKF8ePH06BBAzp27Mjtt99O48aNqVy5sk+9Jk2aUKlSJQYOHMjy5ctp06YNjRs3DtivhQsXUrRoUb799lu+/fZbv32eOnWKAwcOcPDgQcqWLespj4iIoGbNmn71q1SpAsChQ4d8+v3cc89x9dVX07FjR5o0aULjxo0pV66cX18ADh8+TL9+/fzaPnDgAAAbNmzI6zSJiIiIiIiIiIhcNAo0yzkTERHhVxYcHJzrOvf6jIy8Z2J36NCByZMn8/bbbzNq1Cg++ugjAOrXr8+bb75Jy5YtAShVqhSLFi0iNjaWH374gWnTpgHO7OGePXvyyiuveGYXHzx4kMzMTL+HCeZ07Ngxn0BzztnPOY8zKyvLU/bss88SFRXF8OHDGTZsGEOGDMEYQ5MmTRg8eDA33nijpy8AM2bMYMaMGXn2RUREREREREREpDBSoPk8OVfpKE5n8MZMz/u+df68w3nXXXdx1113cfz4cRYvXszkyZMZMWIEd999NytWrOCqq64CoHLlynzyySdYa1m/fj0zZ87kgw8+oH///mRnZzNgwADACXxnZ2eTkpJyXvvdtWtXunbtSmpqKr/99hsTJkxg1KhR3HHHHWzYsIHo6GhPEH7o0KH07t37vPZHRERERERERETkfAg6fRWRwqN48eI0a9aMd955h5dffplTp04xZcoUv3rGGK6++mp69erlmSU8ceJEz/qGDRty6NAh1q1bd0H6HRkZSZs2bRg5ciTdu3cnJSWFefPmefoCeD6LiIiIiIiIiIhcahRolkLv119/5cSJE37lSUlJAISHhwOwdu1a4uPjT1sP4F//+hcAjz/+OHv27PHb5vjx4yxatKhA/Z46dSqZmZl+5fv37/fpz4033sitt97K+PHjGTVqVMC21qxZ49lORERERERERESksPnz5lqQP43nnnuO+Ph4mjZtSkxMDCEhISxbtoyZM2dSrVo1OnXqBMAvv/zCs88+S6NGjahTpw7R0dEkJiYyadIkgoKC6Nu3r6fN5s2bM3DgQF566SWuuOIK2rRpQ/Xq1Tl27BgJCQnMmTOHW265halTp551vzt16kSxYsW45ZZbiImJwVrLvHnzWLJkCfXr16dFixaeul988QXNmjXj0UcfZdiwYTRo0IDIyEgSExNZvXo1a9euZeHChURHR5/9iRQRERERERERETlPFGiWQu/ll19mwoQJLF26lF9++YWgoCCqVq3Kyy+/TJ8+fShdujQAd9xxB3369GHu3LlMmjSJI0eOcNlll9GyZUtPANrbv//9bxo3bsywYcOYP38+kyZNIiIigkqVKvHEE0/QuXPnAvV74MCBTJs2jeXLl/Pzzz9TrFgxqlWrxqBBg3jyySc9DyYEJ7f0smXLeO+99/j+++8ZN24cWVlZVKhQgauuuopevXpx7bXXFqg/IiIiIiIiIiIi54ux1l7sPhQqI0aMsABPPvnkxe5KvvxVHgYoIiIiIiIiIoWDYhEilxxzIXaiHM0iIiIiIiIiIiIiUiAKNIuIiIiIiIiIiIhIgSjQLCIiIiIiIiIiIiIFokCziIiIiIiIiIiIiBSIAs0iIiIiIiIiIiIiUiAKNIuIiIiIiIiIiIhIgSjQLCIiIiIiIiIiIiIFEnyxOyAiIiIiIiIiIpeGxDTL/pOWbAtBxvlcOdxc7G6JSCGgQLOIiIiIiIiIiORp7eFsxidms/Go9Sl/YXUmdUoaOlQO4poI/eG8yF+Z7gAiIiIiIiIiIpKrWfuzeXNDll+Q2W3jUcubG7KYvT/7AvdMRAoTBZpFRERERERERCSgtYez+Xh7FoFDzH+wwMjtWaw9rGCzyF+VAs0iIiIiIiIiIhLQ+MTs0waZ3SwwIVGBZpG/KgWaRURERERERETET2KazTVdRm42HLUkpp3ZNiLy56BAs4iIiIiIiIiI+DnbNBhKnyHy11SoA83GmEHGmF+NMbuMMSeMMSnGmBXGmFhjTNlctmlkjPnZVTfNGLPaGNPHGFPkQvdfRERERERERORSdSLrwm4nIpe2Qh1oBv4FFAdmAEOBcUAm0A9YbYyp4l3ZGNMOmAvcBkwAPgBCgHeBry5Yr0VERERERERELnFhZzll72y3E5FLW/DF7sBplLLWpucsNMa8AbwMvAT0dJWVAkYCWUBTa+1SV/mrwEzgfmNMJ2utAs4iIiIiIiIiIqdxTUQQcOZpMJztROSvplBf+YGCzC7fuF6v8Cq7HygHfOUOMnu18Yrr45PnvJNCfHw8xhi6d+9OfHw8nTp1IioqimLFinHjjTcyefJkv21OnjzJwIEDue666wgPD6dUqVLceuutfPPNN+ek/VOnTjFs2DDq1atH6dKlCQ8PJyYmhnbt2vHLL7/41d+4cSPdu3enSpUqhIZqmm7oAAAgAElEQVSGUr58eTp37symTZv86nbv3h1jDPHx8Xz00Udce+21FCtWjPLly/PEE09w+PBhv21Wr17NQw89RExMDKGhoZQrV4569erRp08fMjIyfOpmZmYyfPhwGjZsSKlSpQgPD+eGG27g/fffJztbea5ERERERETkwqgcbqhT0pzRNleWNFQOP7NtROTPobDPaM7NPa7X1V5lzVyvUwPUnwukAY2MMaHW2pPns3N/VQkJCdx8883UqFGDhx9+mJSUFL7++mtPcPf2228HnCDwHXfcwZw5c6hTpw5PPfUUaWlpfPfdd3Ts2JGVK1fyf//3f2fdPjjB4C+//JJrrrmGrl27EhYWxp49e5g/fz5Tp06lRYsWnrpTp06lQ4cOZGRkcM8991CzZk0SExMZP348P/30E7NmzaJevXp+/XnhhReYNm0a99xzD61atWLWrFmMHDmSrVu3MnPmTE+91atX06BBA4wxtG3blurVq3PkyBG2bt3K8OHDef311ylatCiApw/Tpk2jdu3adO7cmWLFijFr1ix69erF4sWL+eyzz87ZmImIiIiIiIjkpUPlIN7ckIXNR10DtK9cqOc0ish5dEkEmo0xzwMlgAjgRuAWnCDzQK9qtV2vm3Nub63NNMbsAK4GagAbzmuH/6Jmz55Nv379iI2N9ZR17tyZO++8k8GDB3sCwW+//TZz5syhdevW/PDDDwQHO1/D2NhYbr75Zt58803uvvtuGjVqdFbtHz58mK+++or69euzePFiihTxTQ518OBBz/tDhw7x0EMPER4ezty5c7nqqqs869atW0eDBg147LHHWL58ud/xLlq0iDVr1lC1alXAmYncrFkzZs2axe+//87NN98MwJgxY0hPT2fixIm0a9fOp41Dhw4RHh7u+fzGG28wbdo0nn76aYYMGeLpe1ZWFk888QSjRo3i/vvv92tHRERERERE5Hy4JiKIx2rAx9vzDjYb4PEaRZQ2Q+Qv7JIINAPPA+W9Pk8FultrD3iVRbhe/fMW+JZHnuO+BWTuHXYhduPjhbPczk7sfU72X61aNV555RWfsjvuuIOqVavy+++/e8pGjRqFMYZ33nnHE2QGiI6O5tVXX+Wxxx7j448/9gs057d9YwzWWkJDQwkK8v8PXNmyZT3vx44dS2pqKu+//75PkBng6quv5vHHH2fIkCGsX7/eb/1rr73mCTIDBAcH88gjjzBv3jyfQLNbWFiYX19Kly7teZ+dnc37779PhQoVePfdd30C5EWKFOHtt9/m008/Zdy4cQo0i4iIiIiIyAVze3QQ5UJhQmI2G476h5uvLGloXzlIQWaRv7gi/fr1u9h9OK1+/fq91a9fv7i4uLgPgd+AB4DecXFxs/r167cXIC4urhdQFnivX79+KTnbiIuLewyoAnzcr1+/xJzrjTFPxMXFfXTXXXdVBKhVqxYbNmwgKiqKFStWULFiRZYuXUrZsmVZt24doaGh7N69m7S0NI4fP86ePXsICQlh/fr1lCtXjgHfLDmPZ+TcevC6klhrSU5O5tChQ2RmZhIfH0+JEiVYtWqV59grVKjAsmXLKFWqFNu2bSM7O5uUlBQSEhIYPXo09erVo1OnTj7blC9fnrFjx5KQkEDHjh1JTU0lLi6O8uXL88ILL7B161YiIyNZuXIlFStWZNeuXXz11VecOHGCrl27Eh8fT0pKCh9//DE33XQTbdu2ZfPmzZQuXdqzzZgxY0hISKBt27aEh4eTlJTEpk2bmDNnDl999RV79uxh+/bt1KxZkzVr1lCpUiWWLl1KVFQUAwYMYNu2bZQoUYKFCxcyffp0Jk+ezNy5c/n222/Zvn07W7dupVmzZhw/fpyoqCg+++wzNm3axOOPP07x4sV9xn/FihX89NNPVK1albvvvpulS5dSs2ZNxo4dy7fffsvSpUvZu3cvxhiOHTtGaGgoa9eupUKFCowfP54xY8ZQpkwZUlJS+P7775kzZw6TJ09m2rRpzJ8/nxUrVpCRkUGDBg247LLLWLZsmedcR0ZGsnnzZowxHDhwgNTUVDIyMkhISCA8PJzVq1d76rpfS5UqxdatW33GPysri+3bt1OiRAlWrlzpOV8VKlRg+fLlfuN/8OBBsrOz2b59O6VKlfKMy5IlS6hQoQIrV66kePHi7Nixg8zMTA4dOkRycjIAW7Zs8Rn/pUuXEh0dzapVqwgPDychIYFTp05x+PBh9u/fjzGGzZs3U6ZMGZ/rsly5cqxZs4ZixYqxc+dO0tPTOXr0KPv27aNIkSJs3LiRsmXL+mwTFRXF2rVrfa7lY8eOsWfPHooWLcqGDRsoV64cy5cv97n+169fT9GiRdmzZw/Hjx/nxIkTJCYmEhISwrp164iOjmbZsmWe81amTBk2btxIkSJF2LdvH0ePHiU9PZ1du3YRFhbGmjVrPNeW+7yVLl2aTZs2ERQURFJSEkeOHOHUqVMkJCQQFhYWcCwjIiLYunUrgM/4B7qW3d8d9/hnZ2dz8OBBn/EvWbKk3/W/fPlySpYs6RlL7/Hftm0bERERPmNZvnx5z/jHx8eTkZHB4cOHOXDggM/4r1ixgkqVKrFkyRKio6NZvXo14eHh7Ny5k5MnT3LkyBGSkpIICgryG3/3Nu7x37VrF+np6Rw7doy9e/cSHBxcoHt5zvHfsGFDwPF3X8vly5f32cY9lrmN/+rVq33GP+e1vH//fp/x976WlyxZ4vmeuce/IPdy72s55/iXL1+eFStW+Iy/97Wc817ufS3nHH/va9n7Xu6+lnMbf+9refny5T73cve1nJiYmOv4u7dZsmQJUVFRPtdyzvF3X8veY1mmTBk2bNhAcHCwZyxzjn/OsQx0LZ88eZKdO3d6xl/3ct3LdS/XvVz3ct3LdS/XvfxSuZcXO3WU2tkHuC06mAPxmyhaqjRV96+ib8MqVNiznGuqlNe9XPdy3csL6b38sssuizuPoUUPY21+suwULsaYajgpMrZYa69xlS3BSatxo7V2WYBt1uKkzrjKWptr6owRI0ZYgCefLNhzAy/GjOazVdAZzfHx8VSvXp1u3boxevRov/VNmzZlzpw5WGtJTEykSpUq3HjjjSxZ4h+MT09PJywsjJiYGHbs2HHG7budOHGCQYMG8cUXX7BlyxYAihUrxv33389bb71F+fLOBPmWLVsGfDhgTqNHj6Zbt26Ak/95zJgx7Nixg5iYGJ96s2fP5vbbbyc2NhbvH3EWLlzIG2+8wcyZMzlx4gQAtWvXJjY2loceegiABQsWcMstt5y2L97nRkRERERERORCG7wx0/O+b51L5Y/lRf7SLsgTOi/Ju4G1NsEYsx643hgTZa1NBjbhBJprAT6BZmNMMFAdyAS2X5A+nqN0FKdzqd3cIyKcDCf79u0LuH7v3r0+9c5WWFgY/fr1o1+/fuzatYu5c+cyevRoPv/8c+Lj45k3b57PflatWsV1111XoH3m5W9/+xuTJ0/m5MmTLFu2jKlTp/Lee+/RuXNnypUrR4sWLTx9ad++PePHjz9vfRERERERERERETnXLuXkORVdr1mu15mu1zsD1L0NCAd+s9aePN8dk9yVLFmSyy+/nN27d3tmGnubNWsWAPXq1Ttn+6xSpQpdunRh2rRpXHHFFcyfP9/zQMCGDRsCeALP51toaCiNGjWif//+DBvmzHqfNGkSAHXq1CEyMpJFixaRkZFxQfojIiIiIiIiIiJyLhTaQLMxpo4xpkKA8iBjzBtANE7g+JBr1XdAMtDJGHOjV/1iwOuujyPOc7clH3r06IG1lr59+5KVleUpT05OZsCAAZ46Z+vAgQMsXrzYr/z48eMcPXqU4OBgQkJCAHjkkUeIjIwkLi7O54GCbtnZ2cyePfus+wJOEPvwYf9nVCYlJQEQHh4OOA8T7NWrF3v37qV3796eFBve9u7dy/r16wvUHxERERERERERkXOtMOdauBMYbIyZC2wDDgLlgSZADWAf8Li7srX2iDHmcZyA82xjzFdACtAWqO0q//qCHoEE9PzzzzNlyhQmTZpE3bp1adOmDWlpaXz77bfs37+fF154IV+5inOze/duGjZsyJVXXkm9evWoUqUKR44cYfLkyezbt4/evXtTsmRJAMqWLct3331H+/btadiwIc2bN+fqq68mKCiInTt3snDhQg4ePEh6evpZ9+ftt99m+vTpNG3alBo1alCiRAnWrVvHlClTKF26NE888YSn7quvvsqqVav48MMP+fHHH2nWrBmVKlVi//79bNmyhQULFvDGG29w1VVXnXV/REREREREREREzrXCHGj+Bfgf0BioC0QCx3EeAvgZMMxam+K9gbV2ojGmCfAf4D6gGLAVeNZV/9J78uGfUEhICDNmzOCdd97hiy++4L333iM4OJi6desyZMgQz8PxzlZMTAxxcXHMnj2bWbNmkZycTJkyZahduzYDBw6kU6dOPvWbN2/O6tWreeutt5g2bRrz5s0jJCSEihUr0qxZM+67774C9adnz56ULl2axYsXs2DBAjIzM6lcuTI9e/bkueeeo1q1ap66RYsWZeLEiXz++eeMHj2ayZMnc+zYMcqVK0f16tUZMGAAXbp0KVB/RERERERE5NLn/cymQC6F5ziJyJ+LUezV14gRIyzAk08+ebG7ki+X2sMARURERERERKTgvOMBh079EdspHWKA8xsjUCxC5JJjLsROdDcQEREREREREbnEeAd4FfgVkcKg0D4MUEREREREREREREQuDQo0i4iIiIiIiIiIiEiBKNAsIiIiIiIiIiIiIgWiQLOIiIiIiIiIiIiIFIgCzSIiIiIiIiIiIiJSIAo0i4iIiIiIiIiIiEiBKNAsIiIiIiIiIiIiIgWiQLOIiIiIiIiIiIiIFIgCzSIiIiIiIiIiIiJSIAo0i4iIiIiIiIiIiEiBKNAsIiIiIiIiIiIiIgUSfLE7IGdu8MZMz/ttx6xfed86GlYRERERERERERG5cDSj+RJXJuSPRS593bt3xxhDfHz8xe5KoTd79myMMfTr1+9idyXfmjZtijHmYndDREREREREROScU6D5EtS3TrBn+b/rinoWd9mFFB8fjzGG7t27X9D95kdGRgYTJkzg0Ucf5ZprrqFUqVKEh4dz7bXX8tprr3H06NFct01MTKRHjx5UrFiR0NBQYmJi6NOnD4cOHbqARyBybhTm61RERERERERE/hyUY0H+tLZt20aHDh0oXrw4t99+O3fddRfHjh1j2rRpDBgwgK+//poFCxYQFRXlt12jRo3Yv38/7dq1o06dOvz+++8MHTqUqVOnsmDBAsqWLXuRjkouZWPHjiUtLe1id0NERERERERE5JxToFn+tEqWLMkHH3xAt27dKF68uKf81KlTdOjQgZ9++om4uDjee+89n+169uzJ/v37GTZsGL169fKUP/vss7z77rv85z//4cMPP7xgxyF/HlWrVr3YXRAREREREREROS+UOkPOWr9+/ahevToAY8aMwRjjWUaPHu2pl52dzYcffshNN91EiRIlKF68ODfddBMjRowgOzvbr11jDE2bNmXPnj08/PDDREdHExYWRv369fniiy/y3b9KlSrRs2dPnyAzQEhICC+//DLg5Pn1tn37dqZPn05MTAxPPfWUz7q4uDiKFy/OZ599xvHjx/Pdj2XLlnHfffcRHR1NaGgo1apVo2fPnuzduzfXbbKzs3nnnXeoU6cOxYoVo3LlyvzrX//iyJEjfnVXr17NQw89RExMDKGhoZQrV4569erRp08fMjIyfOpmZmYyfPhwGjZs6EklcsMNN/D+++/7jYV3uoXNmzfTsWNHoqOjCQoKYvbs2dSpU4eQkBCSk5MDHsPAgQMxxvDBBx/4lCcmJvL0009To0YNQkNDKVu2LG3btmXJkiUB20lKSuLRRx+lfPnyhIWFcf311zNmzJhcz10gb775JsYYhg0bFnD9nj17KFKkCDfddJNPWf/+/WncuDEVKlQgJCSEihUr0rlzZzZs2ODXxunOFwTO0Xzq1Cnef/992rRpQ7Vq1QgNDaVMmTK0aNGCKVOmBOxvTEwMMTExpKWl0bdvX6pWrUpoaCg1a9Zk0KBBWPvHQ0Lze52KiIiIiIiIiBSEZjTLWWvatCmpqakMHTqUunXrcu+993rWXX/99Z73Dz/8MF988QVVqlThsccewxjDhAkT6NmzJ/Pnz2fcuHF+bR86dIhGjRoRGRnJI488QmpqKt988w1dunRh9+7d9O3bt0B9L1q0KADBwb6XwMyZMwFo1aoVQUG+v8OULFmSxo0bM336dBYtWkTz5s1Pu5/Jkydz3333Ya3l/vvvp1q1aixbtowRI0YwadIkFixYQExMjN92//rXv5g7dy4PPvgg7dq1Y9q0aQwZMoR58+Yxf/58ihUrBjhB5gYNGmCMoW3btlSvXp0jR46wdetWhg8fzuuvv+451oyMDO655x6mTZtG7dq16dy5M8WKFWPWrFn06tWLxYsX89lnn/n1Zdu2bTRo0IBatWrRpUsXTpw4QalSpejWrRsvv/wyX375pc/Mb7exY8cSEhJCp06dPGXLly+nVatWpKSkcMcdd9ChQweSk5OZOHEit9xyCxMmTKBNmzae+gcPHqRRo0Zs376dW265hVtuuYW9e/fyz3/+k1atWp32/Lt17dqVV155hTFjxtC7d2+/9Z9//jnZ2dl069bNUzZ37lwGDhzI7bffzn333UeJEiXYsmUL3333HT/88AMLFiygbt26+T5fuUlJSeGZZ56hUaNGtGzZknLlyrF3715+/PFH2rRpw8iRI3nsscf8tsvIyKBVq1bs2bOH1q1bExwczMSJE3nxxRdJT08nNjYWyP91KiJyoQ3emJnn+gv93AkRERERESkga60Wr2X48OF2+PDhVvJnx44dFrDdunULuP6LL76wgL3hhhvs0aNHPeXHjh2z9evXt4AdN26czzaABewDDzxgs7KyPOXbt2+3pUuXtkWLFrXbtm0rUL//+c9/WsC++OKLPuXPP/+8Bexbb70VcLunnnrKAvn6jhw9etSWLVvWBgUF2blz5/qsGzhwoAVsy5Ytfcq7detmAVu2bFkbHx/vKc/KyrIdOnSwgO3fv7+n/Nlnn7WAnThxot/+U1JSfM5fbGysBezTTz9tMzMzPeWZmZm2R48efu24xxawL730kl/7iYmJNigoyNavX99v3e+//24B26FDB09ZRkaGvfzyy21oaKidPXu2T/3du3fbihUr2goVKtj09HRP+eOPP24B26dPH5/6S5YsscHBwRawsbGxfvsPpFWrVhawa9as8Vt31VVX2aJFi9rk5GRPWVJSkj1y5Ihf3ZUrV9rixYvbO++806f8dOfLWmubNGlindvuH9LT0+2uXbv86qamptqrr77ali5d2qalpfmsq1atmgVs69atfdYlJSXZiIgIGxERYU+dOuXXt9yuUxGRi+G/GzI8y0urTnkWd5mIiIjkn/d/V/+M+xORArsgcVVNFTlf6j5wsXuQf6u+PW9Njxo1CnDSKJQoUcJTXrx4cQYNGkSLFi34+OOP6dy5s892RYoUYdCgQT6ziqtXr07v3r2Ji4vjs88+88zYPFM//PADH330EZUrV+aFF17wWXf48GEAIiIiAm7rLk9NTT3tfiZNmsTBgwd56KGHuPXWW33WPffcc3z44YfMmDGDnTt3+uXufeaZZ6hWrZrnc1BQEIMHD2bixImMGjWKV1991ad+WFiY3/5Lly7teZ+dnc37779PhQoVePfddylSpIhnXZEiRXj77bf59NNPGTduHO3atfNpp3z58gHPdaVKlWjevDkzZsxg3bp1XH311Z517tQW3jOEf/rpJ7Zt28bzzz9PkyZNfNqqWLEiL7zwAn369OHXX3+lTZs2ZGRkMG7cOEqWLEm/fv186t9444106dLljFJodOvWjenTpzNmzBgGDx7sKV+6dCnr16+nffv2Pg95jI6ODthO3bp1adasGdOnTycjI8MzY9wtt/OVm9DQUCpXruxXHhERQY8ePXjuuedYsmQJt912m1+dYcOG+Yx9dHQ07dq1Y+zYsWzatIlrrrkm3/0QEbnQvGcse89u1kxmEREREZFLk/4lL+fV8uXLCQoKomnTpn7rmjRpQpEiRVixYoXfuqpVq3ryynpr2rQpcXFxAbfJj99++43OnTtTvHhxvv/+e59gbH5YV+7bnHl2A1m+fDkAzZo181sXHBzMbbfdRnx8PCtWrPALNOcMxALUqFGDKlWqEB8fT2pqKpGRkXTs2JGhQ4dy7733cv/999OiRQsaN27M5Zdf7rPt5s2bOXjwIFdccQWvv/56wP6GhYUFzD1ct25dQkNDA27TvXt3ZsyYwZgxY/jvf/8LODmHv/rqK8qVK+eTBmPhwoUAJCQk+AWOAbZs2QLAhg0baNOmDRs3biQtLY1bb701YOC/adOmZxRobt++PREREXz++ecMHDjQE2x3t9G9e3e/bX766Sc+/PBDli5dSnJyMpmZvn/mnZyczGWXXeZTltf5ys26desYPHgwc+fOZe/evaSnp/us3717t982ERER1KxZ06+8SpUqgJN+RkRERERERETkQlGgWc6rw4cPU6ZMGUJCQvzWBQcHExUVxf79+/3WlS9fPmB7FSpU8LR7phYuXEjr1q0JCgpiypQp3HzzzX513AHN3Np3P4wvtxnP3txt5AxEurnLA82Ozuv4ExISOHz4MJGRkdx8883MmzePN954g++++86TY7l27drExsby0EMPAU6uY3CCuXFxcbn2+dixYwH3mZv27dtTqlQpPv/8c958802KFCnC5MmTOXjwIH369PHJge3uw7ff5j2D3t0H9/k73Xchv8LCwnjwwQcZOXIk06dPp3Xr1mRkZPDll19Srlw5Wrdu7VN/2LBhPPPMM5QuXZqWLVtStWpVwsPDMcYwceJEVq1axcmTJwvcr0WLFtGsWTMyMzNp3rw5bdu2pVSpUgQFBbFy5UomTZoUcD+RkZEB23Of86ysrDPqh4iIiIiIiIhIQSjQfL6cx3QUl5KIiAhSUlICphjIzMwkOTk54IPSkpKSAra3b98+T7tnYt68edx1110EBQUxbdo0GjZsGLBe7dq1AWcGcCDuWbe1atU67T7dfXT3Oae9e/f61POWlJTk6Yu3QMf/t7/9jcmTJ3Py5EmWLVvG1KlTee+99+jcuTPlypWjRYsWnvrt27dn/Pjxp+27t7xmb7uDtx9//DEzZszgzjvvDJg2w7vPkyZNom3btqfdr7v+6b4LZ6Jbt26MHDmSMWPG0Lp1a09Q/JlnnvH5fmZmZhIbG0uFChVYvny5348F7tnZgeRntru3119/nRMnTjBr1iy/mf9vvvkmkyZNOqP2REREREREREQuhqDTVxHJnTv9QG6zJ2+44Qays7OZO3eu37q5c+eSlZVFvXr1/Nbt3LmT+Ph4v/LZs2d72s2vmTNn0rp1a4KDg5kxY0auQWaA22+/HYDp06eTnZ3ts+7o0aMsWLCAsLCwPNtwc/fR3WdvmZmZzJ8/HyDg8c+ZM8evbPv27ezatYuYmJiAs1lDQ0Np1KgR/fv3Z9iwYQCeIGWdOnWIjIxk0aJFZGRknLbvZ8KdcmLMmDEkJyczZcoUrrvuOq6//nqfeu5zNm/evHy1W6dOHcLDw1m5cmXAGeaBzuvpNG7cmCuuuIJJkyZx+PDhXIPiycnJpKam0qhRI78g87FjxzxpUc6FrVu3UqZMmYDpZQJ9D87G6a5TEREREREREZGCUqBZCqR06dIYY9i5c2fA9T169ADgpZdeIi0tzVOelpbGiy++CMCjjz7qt11WVhb//ve/fYK9O3bsYNiwYQQHB/P3v/89X/2bPn06d999N8WKFePXX3/lpptuyrP+5ZdfTqtWrYiPj+eDDz7wWRcbG8vx48fp2rUrxYsXP+2+7733XsqUKcOXX37JokWLfNYNGTKE7du306JFC7/8zABDhw4lISHB8zk7O5u+ffuSnZ3NI4884imfN29ewCCsexZweHg44KRT6NWrF3v37qV3796cOHHCb5u9e/eyfv360x5XTt7B2xEjRpCRkREw33G7du24/PLL+eCDD/j5558DtrVw4ULP96Ro0aJ06dKFo0eP+uV0Xrp0KePGjTvjvoITVE5PT2f48OH8/PPPXHfddX4/XERHRxMeHs6yZct80olkZGTwzDPPkJycfFb7DiQmJoaUlBRWr17tU/7JJ58wbdq0c7KP012nIiIiIiIiIiIFpdQZUiAlSpSgQYMGzJs3jy5dulCrVi2KFClC27Ztue666+jcuTOTJk3im2++4eqrr+bee+/15LjdsWMHDz74IF26dPFr97rrrmPx4sXUr1+fVq1acfjwYb7++mtSU1P573//6/ewu0A2bdpEu3btSE9Pp02bNkyaNClgGoKcQczhw4fTqFEjevfuza+//sqVV17J4sWLmTVrFrVq1eKNN97I97kZNWoUDzzwAE2aNOGBBx6gatWqLFu2jOnTp1OhQgU++uijgNs2btyY66+/no4dOxIREcG0adNYtWoV9evX54UXXvDUe/vtt5k+fTpNmzalRo0alChRgnXr1jFlyhRKly7NE0884an76quvsmrVKj788EN+/PFHmjVrRqVKldi/fz9btmxhwYIFvPHGG1x11VX5Oj5vXbt25dVXX2XAgAEEBwfTuXNnvzpFixZl/Pjx3HHHHdx11100atSI66+/nvDwcHbt2sWSJUvYvn07/8/enYe3edX5338febdjO17jOE7ahi5xk1KgHQoDQ7oMzEApkLTlx29gpmwp6cPQMjy0wPxom6QDBQoDHZiuQBmGDlBKysA1MMBT0gI/hmIncbzEafbYSuw4cRLvu87zh5bItmzpliVZy+d1Xbpk3bpv6VhH95H0Ped8T1dXVyBA/vnPf57nn7v7xIYAACAASURBVH+er33tazQ2NvLGN76Rrq4ufvjDH/K2t72Nn/70p1GV9b777uP+++9nYmJi1mhmAJfLxZ133skXvvAFrrjiCt75zncyPj7Ojh07OHPmDNdddx07duxw/NyhfPzjH+eXv/wlb3zjG3n3u99NaWkpjY2N/P73v+eWW27h2WefXfBzhDtPRUREREREREQWzFqrS9DlkUcesY888oiVyB04cMC+/e1vt+Xl5dYYYwH71FNPBe6fmpqy//qv/2qvuuoqW1BQYAsKCuxrXvMa+41vfMNOTU3NejzArl+/3h4/fty+973vtVVVVTYvL8+++tWvtk8//XTE5dqxY4cFwl5C6ejosO9///ttTU2NzcnJsatWrbJ33nmn7e3tdfz6/OlPf7Lvete7bGVlpc3JybErV660mzdvtsePH5+172233WYBe+jQIfvlL3/ZXnbZZTYvL8/W1tbau+66y/b19U3b/5e//KV9//vfb+vr621JSYktLCy0l156qf3Yxz5mjx49OuvxPR6P/e53v2uvv/56W1ZWZnNycmxtba19wxveYD/3uc/Zjo6OwL5HjhyxgL3tttvC/o/Hjh2zLpfLAvbtb3/7vPuePHnSfupTn7Jr1661BQUFtqioyF588cX25ptvtv/+7/9uJyYmpu3f1dVlP/CBD9jKykqbn59vr7zySvvUU08F6vf+++8PW76ZbrjhBgvY7Oxs293dHXKfiYkJ+5WvfMXW19fb/Px8u2zZMvu+973PHj16NFBPR44cCewfyeu1fv36kO+5n/3sZ/aaa66xS5YssaWlpfbNb36zffHFF+1TTz0163yy1toLLrjAXnDBBSGf4/7777eA3bFjx7Tt4c5TEZHF9KX2icBFREREnEv0Z6k+u0VSTkLiqsZam8i4dtJ79NFHLcAdd9yx2EXJWMYY1q9fH1UOXhEREUk9D+2bDPx99xpNuBMREXEq0Z+l+uwWSTkmEU+iHM0iIiIiIiIiIiIisiAKNIuIiIiIiIiIiIjIgijQLCIiIiIiIiIiIiILokQ6knSUN1xERERERERERCS1aESziIiIiIiIiIiIiCyIAs0iIiIiIiIiIiIisiAKNIuIiIiIiIiIiIjIgijQLCIiIiIiIiIiIiILokCziIiIiIiIiIiIiCxI9mIXQERERERERERE5vbQvsnFLoKISFgKNIuIiIiIiIiIpIiz4zbwd1muWcSSiIhMl7SBZmNMBbABuBG4AlgBjAMtwFPAU9ZaT9D+FwJH5nnIH1pr3xOv8oqIiIiIiIiIxMPda86Hb4JHN/u3a8SziCSDpA00A7cCjwJdwA6gA1gGbAS+CbzVGHOrtdbOOG4P8JMQj9cax7KKiIiIiIiIiIiIZKxkDjTvB94B/NeMkcv/CPwJuBlv0PnHM45rstZuSVQhRURERERERERERDJd0gaarbW/mWN7tzHmMeBzwLXMDjSLiIiIiIiIiIjETLj0JMHpTUQyVaqeBRO+61Bnea0x5iNABdAL/I+1tjlhJRMRERERERERkbSlBRlFQku5QLMxJhv4O9/N/w6xy5t9l+BjXgBus9Z2xLd0IiIiIiIiIiKSbsItyCgikLVly5bFLoMjW7du/RLwLuDn1tptQdtzgVHgE8CngS8CvwAuBK4Dbtq6deu3t2zZMjHrQQFjzO1bt259/MYbb6wFuPTSS2lvb6eyspLdu3dTW1tLY2MjFRUVtLW1kZeXx/HjxxkeHmZoaIgTJ06Qm5vL3r17qaqqYteuXdOOaW9vJycnhxMnTjA0NMTIyAhut5u8vDxaW1tZtmzZtGPKysp4+eWXycrKoru7m4GBAUZHR+ns7KSgoIDm5mZqamrYuXNn4JilS5eyf/9+jDH09PTQ39/P+Pg4x44do7CwkObmZmpra2loaGDFihU0NjZSWlrKwYMHsdZy+vRpzp49y+TkJEePHmXJkiXs2bMn8Pj+5yspKeHQoUN4PB7OnDlDb28vHo+Hw4cPU1xcPO2YZcuWsXv3boqLizly5AiTk5OcPXuW06dPA3Dw4EGWLl1KU1NT4Jjq6mr27NlDUVERR48eZWJigr6+Pnp6ejDGsH//fsrKyqYdU1VVRXNzM4WFhXR0dDA2NkZ/fz8nT54kKyuLffv2UVFRwa5duwL/e2VlJa2treTl5eF2uxkdHWVwcJCuri6ys7MD9e8/pqGhgcrKSvbu3Ruoy5n139bWRnV19bS6LC8vp729nezs7EBdzqz/mXXpr3+Xy8XJkyfp7+9nbGyMjo6OQP0vX758zvo/deoU586dY2JiYlb9NzY2Bq5LSkpm1f/U1BSHDx9myZIlNDU1BV6vmpoadu3aNW/9l5SUBOqloaGBmpoampqaKCoqCln/Bw4cmLP+CwsLOXbsGOPj47Pqv7y8fNp5WVVVRUtLC/n5+XR0dDA6OsrAwADd3d3T6j/4mOD695/Lg4ODnDhxgpycHNrb20Oey8H1H3wuB9f/zp07A69beXk5+/btm/NcbmlpmVb/DQ0NIevffy77639mXfrPZWBa/Yc6l/3vHX/9ezweent7p9X/zHPZX//B53Jw/R86dIjS0tJpdbls2bJA/Qefy6dOnZpW/7t37w6cY9XV1XOeyy6Xa1b9+4/x139nZ+ec57LacrXlasuTty1vaGjEU1JJjruNNRUFasvVlqstV1uutjwF23J9L09cW964cxc2v5jsU0e4qMBb//tOnMZYS3bPIV5VuzTubXlL+8t4isrIObaH112yIuPa8oPj+eR0tuIpW07ukZ1qy9WWJ31bvnz58q1RhGEdM9ba8HslCWPMncDDwD7gDdbaMxEckw38HrgG+Li19uH59n/00UctwB133LHwAouIiIhIWBoVJCIiErlQn5uJ/izN9M/uTP//JSUlJMeLKxFPEgvGmI/iDTLvBa6LJMgMYK2dBL7pu/mmOBVPREREREREREREJGNF1e1ijCkCVgCVwAjQY609HsuCzXi+jwNfBVqBG6y1PQ4f4pTvuiimBRMRERERERERERFZBMGj60NJ9Ij7iJ/NGPMK4IN4F9p7NTNGQxtjeoEdwI+B7b6RxAtmjPkU8AWgCXiztfZ0FA/zOt/14ViUSURERERERERERCRZnB0/nx65LDchmTJmCRtoNsZcDfwT8Jd4g8sTQAvQDZwBCoAK4DLgVuAW4JQx5mHgn621Y9EWzhhzL7AN2Am8Zb50GcaYa4Dd1trxGduvB/7Bd/N70ZZFREREREREREREJFkEj1hOhtzh8z6rMebfgPcCfcATwA+AP1lrR+fY/0Lgr4DbgM8Bm40xf2ut/a3TghljbsMbZJ4CfgfcacysaPxRa+13fH9/EVhrjHkBcPu2vRK43vf3vdbaPzgth4iIiIiIiIiIiIjML1x4+y3AJ4DHIxmZbK09CjwOPG6MeSWwFVgPOA40Axf5rrOAj8+xz4vAd3x//zuwAfgz4K1ADnASeAb4hrX2d1GUQURERERERERERETCCBdoXm2tHYnmga21zcAGY0xBlMdvAbY42P9bwLeieS4RERERERERERERiZ5rvjujDTLH+jFEREREREREREREJHnNG2iOlDEmxxjzamPMZbF4PBERERERERERERFJHY4CzcaYdxtjnjHGlAdtewXQBjQCe40x240xi7O0oYiIiIiIiIiIiIgknNMRzR8E1lhrzwRt+wpwMbADaAbeCXwgNsUTERERERERERERkWTnNNB8OdDgv2GMKQHeBjxjrf1L4LXAPhRoFhEREREREREREckYTgPNVUBX0O3XA9nADwCstRPAr4FXxKR0IiIiIiIiIiIiIpL0nAaaB4DSoNvrAQv8PmjbKFC8wHKJiIiIiIiIiIiISIpwumjfAeCtxpg8vAHmW4Fma+3poH0uAHpiVD4RERERERERERERSXJORzQ/AazGG3Bu9/397Rn7XAO0LbxoIiIiIiIiIiIyH/ewpWfM0j3qvXYP28UukohkKEcjmq21/2aMuQy43bfpG74LAMaY64ELgUdiVUAREREREREREZluYMKyrW2SfQPTA8v3NE+yptiwsc7FulKn4wtFRKLnNHUG1tp/BP5xjrt/D5QBQwsplIiIiIiIiIiIhNY7ZukYAW9W09n2DVgebJ9i02q4tlrBZhFJDMeB5vlYa8eB8Vg+poiIiIiIZJaH9k3Oe//da2L6M0ZEJKUMTPiDzPOzwJOHp6jMQyObRSQh5m1pjDEFC32CWDyGiIiIiIhkprPjNnARERHoHo18Xws85/bErSwiIsHCDQU4Yox5EHjMWjvm5IGNMVcC24BG4IEoyyciIiIiIhkmeMRy8OhmjWQWkUznHrYMTjk7pn3Au0BgXaGJT6FERHzCzZ34FfDPQJcx5lFjzHXzjVA2xqw2xtxhjPkfYBdwJbAjdsUVEREREREREclMrX3RjU6O9jgRESfmHRJgrf07Y8y/AJ8Hbvddpowx7UAXcBbIByqAy4BKwAAngf8DfNXpSGgREREREREREZltxOFo5oUeJyLiRNi5Z9baRuAtxphLgA8BNwCvAq6YsespYDvwY+DH1tqJGJdVRERERERERCRjFWQl9jgRESciTnJmrT0AfBrAGFMIrMA7knkE6LHWdsWlhCIiIiIiIiIiwrpSF+A8DYb3OBGR+IpqNQ1r7TBwwHcREREREREREZE4qys0LMnC0YKAS7Lg+x3TD9DiqiISD+rSEhERERERERFJETX5zvYvyoaz45az4zY+BRIR8VEXloiIiIiIiIhIiijOMawqsHSMzL+fATatzqLhzPlUGxrJLCLxpBZGRERERBaNe9jSM2bxWHAZ7+26QrPYxRIREUlqFXmGXJclx2VoH5g9Urm+2LChzsW6Ute0QLOISDwp0CwiIiIiCdfa52G728O+GT+O72meZE2xYaPvx7GIiIiEVpxjuHtNNu5hy8MHJgOdtnddkq1OWxFZFAo0i4iIiEhC7ejx8M3DU8yVKXLfgOXB9ik2rYZrqxVsFhERmU9doaE6z0y7LSKyGPTNXUREREQSprVv/iCznwWePDxFa5+m+4qIiIiIpAJHgWZjzCpjTEmYfYqNMasWViwRERERSUfb3Z6wQWY/CzznVqBZRERERCQVOB3RfAS4K8w+d/r2ExEREREJcA/bWTmZw2kfsLiHnR0jIiKSrvyL6HaPeq/1GSkiycRpjmbju4iIiIiIOBJtGozWPg91hVkxLo2IiEjqCLeI7qTHUpyjcI2ILK54LAa4DBiKw+OKiIiISAobmUrscSIiIukgkkV0AVYVWCryFGwWkcUTNtBsjPm7GZteFWIbQBawCvhboCUGZRMRERGRNFIQ5aDkaI8TERFJdZEuogvQMQK5LqXSEJHFE8mI5u9AoE2zwDt9l5n83WbDwNYFl0xERERE0sq6UhfgPH2G9zgREZHM42QRXYDu0bgVRUQkrEgCzR/wXRvg28BPgP8Msd8U0Av8j7X2XGyKJyIiIiLpoq7QsKbYOFoQsL7YUFeoacAiIpJ5ollEd3DKe5w+O0VkMYQNNFtr/83/tzHmNuAn1trvxrVUIiIiIpKWNta5eLA9sinABthQp9HMIiKSmbSIroikGkeLAVprr4tXQUREREQk/a0rdfHh1YTNN2mATauzlDZDREQylhbRFVlcD+2bnPf+u9c4CqtmBEeviDHmNcDbgcettSdD3F8D3A781FrbFJsiioiIiEg6ua7aRVUePOf20B5iSnB9sWFDnUtBZhERyWhaRFckeZwdP/+dtSxXqWnm4jT0/kngjcADc9x/EvgQcDHwdwsol4iIiIiksXWl3kCye9jy8IFJPBZcBu66JFt5JUVERHC+iO4nv/EQAPUlBrLOf5ZuGPQGyJ771N0xLV+mcg9besZs4LuLcmKnr+ARy8GjmzWSeW5OX5nXAzustSFnOlprrTHmN8CbFlwyEREREUl7dYWG6jwz7baIiIhEt4huloGCLAM9vec3FpbHoXSZp7XPw3a3Z1Z93NM8yZpiw0bNxhJxHGiuAdxh9jkBLI+uOCIiIiIiIulHeR5FJBpOFtH98t/fzcVFcOUVOfCxLwS2P/fRT8avgBliR49n3vUl9g1YHmyfYtNquLZawWbJXE7f/cNAVZh9qoCx6IojIiIiIiKS3s6O28BFRGQ+3kV0s4hkvs+qAijO0cygWGvtmz/I7GeBJw9P0doXeboTkXTjtNu8CXinMeYT1trBmXcaY0qAd/r2ExEREREREZTnUUSiF8kiuhMeqyBznGx3eyIaUQ7eYPNzbo9SaKShlr7z74LNjROz7n/s6pxEFidpOX3nP4F3xPKvjTGvDL7DGHMl8Cug0rffghhjKowxHzbGPGeMOWiMGTHG9Bljfm+M+ZAxJmTZjTF/boz5uTHmjDFm2BjTbIz5uDFG666KiIiIiIiISMpZV+ri3rXZfOmV2awogOX5sKIAvvTKbO5dm60gc5y4h62jHNkA7QMW97BmrKSzkanzl2ThX6Sye9R7vVjvQUfd59baHxpj3gr8HbDbGHMSOA6sAJYBBvg3a+33Y1C2W4FHgS5gB9Dhe46NwDeBtxpjbg1emNAY807gx8Ao8EPgDHAT8FXgDb7HFBERERERERGJuXjnY9ciuokVbRqM1j4PdYUa75hOrig9f64dGjwfxF3skczJtkil4xbOWvt+Y8wfgI8Ba/EuEAjQCvyLtfabMSrbfuAdwH9ZawNntjHmH4E/ATfjDTr/2Le9BHgSmAKutdY2+rbfC/wGuMUY8x5r7Q9iVD4RERERERERkZCC87CX5Sog7NRN//TTee//2WffEfcyRDtiNZlGukpsbPjiQ4G/h6eCgrr+IO7XP53gEiXnIpVRdaVZa58AnjDGFAJLgXPW2uFYFsxa+5s5tncbYx4DPgdciy/QDNyCN63Hd/1BZt/+o8aYzwLPA3cACjSLiIiIiIiISMwlRT72g51wvAcmpyA7i4rOTnpXrkzc88fB8d6hwN8rKooS9rwFUQ5KjvY4SQ1Lz549f6O0YlHK4HSRyicOT93wH6/LeT7e5VpQS+cLLsc0wBwhf9bt4Dkp1/uu/zvE/r/FW84/N8bkWWvH4lk4EREREREREZFEWtXSyp//+MfQvm/a9g9+8h4669fwh5tvhjWvWqTSORc8Yjl4dHMiRjL7eVMOOE+focUA089zn7o78Pdfff5Lgb9XLcJIZnC+SCVwL95BuHEVVaDZGFOFN3VFPVBkrf1w0PaLgBZr7UjMSjn9ubPx5oiG6UHly3zX+2ceY62dNMYcwZvqYzXQHo+yiYiIiIiIiIgk2hW/2cFbnngSl50derLAyvZ93Pq5z0P2Zthw/ewHkJDqCg1rio2jBQHri41yZ6exis5OKk71kOXxMOVyeWcQXJzYGQPRLFIJrI9HWWbK2rJli6MDjDEfAn4JvBO4BnjVli1btgFs3bp1NdAIHN2yZcvu2BbVa+vWrV8C3gX83Fq7LWj73UAF8NCWLVvOhDjuvcBK4OktW7a4Z95vjLl969atj9944421AJdeeint7e1UVlaye/duamtraWxspKKigra2NvLy8jh+/DjDw8MMDQ1x4sQJcnNz2bt3L1VVVezatWvaMe3t7eTk5HDixAmGhoYYGRnB7XaTl5dHa2sry5Ytm3ZMWVkZL7/8MllZWXR3dzMwMMDo6CidnZ0UFBTQ3NxMTU0NO3fuDByzdOlS9u/fjzGGnp4e+vv7GR8f59ixYxQWFtLc3ExtbS0NDQ2sWLGCxsZGSktLOXjwINZaTp8+zdmzZ5mcnOTo0aMsWbKEPXv2BB7f/3wlJSUcOnQIj8fDmTNn6O3txePxcPjwYYqLi6cds2zZMnbv3k1xcTFHjhxhcnKSs2fPcvr0aQAOHjzI0qVLaWpqChxTXV3Nnj17KCoq4ujRo0xMTNDX10dPTw/GGPbv309ZWdm0Y6qqqmhubqawsJCOjg7Gxsbo7+/n5MmTZGVlsW/fPioqKti1a1fgf6+srKS1tZW8vDzcbjejo6MMDg7S1dVFdnZ2oP79xzQ0NFBZWcnevXsDdTmz/tva2qiurp5Wl+Xl5bS3t5OdnR2oy5n1P7Mu/fXvcrk4efIk/f39jI2N0dHREaj/5cuXz1n/p06d4ty5c0xMTMyq/8bGxsB1SUnJrPqfmpri8OHDLFmyhKampsDrVVNTw65du+at/5KSkkC9NDQ0UFNTQ1NTE0VFRSHr/8CBA3PWf2FhIceOHWN8fHxW/ZeXl087L6uqqmhpaSE/P5+Ojg5GR0cZGBigu7t7Wv0HHxNc//5zeXBwkBMnTpCTk0N7e3vIczm4/oPP5eD637lzZ+B1Ky8vZ9++fXOeyy0tLdPqv6GhIWT9+89lf/3PrEv/uQxMq/9Q57L/veOvf4/HQ29v77T6n3ku++s/+FwOrv9Dhw5RWlo6rS6XLVsWqP/gc/nUqVPT6n/37t2Bc6y6unrOc9nlcs2qf/8x/vrv7Oyc81xWW662XG158rblDQ2NeEoqyXG3saaiQG252vJp5/L+IRc5XS/jWVJJXkdTWrTlTXua8ZQtJ+fITl53SZ3acrXladGW63v57La8pbMH13AfZmqS0nPHYtaWN+7chc0vJvvUES4q8DC044+88suhg8wAJujavriTM6uqOFOQvaC2vKX9ZTxFZQweaKKRGhoaGvjdWAUNTS38vi+Hl/Z38j9dw1yQNRST7+WP/1cj+fn5nOvr473r1yS0LV9Z6OLgiR7yx/rJ9oxT3tfJeE4BK3pa6StezqquXYHr0fwS3mqOUJyN2vI0a8uHX2jgr5/8Ntd+/xlKBgZYMjhIycAAPPNLaGjl5b5e8l+xKiFt+f7xQno6j+LyTFI4eo4lw70AVJ09xEheKXUnmwPvyYGiKlb0tNC/pIab67K2hmwkYsjYORqikDsb82a8o4ibgfuBvwI2W2uzgvZpBo5Za2+KcVkxxtwJPAzsA95grT0TdN9+4BLgEmvtwRDH/gF4PfB6a+0f53qORx991ALccccdMS69iIiIiISyaDksJSWk4/sjHf8nEZktXuf6rMf94H2w08HE7asvh28tLN60udGb0XR4Cgp9EaGJoAwTa0rM+fLFwGKlzvALt+gaeAP5m1ZnJWzRNUmg7c/j2fY4LmuxnO+8mcZl4L7EzBh4zj3Fj9zOU7r8x+ty4j7U3ukZ/ymgC1hvre03xrw6xD7NeAO6MWWM+SjeIPNe4IbgILNPn++6dI6HKJmxn4iIiIiIiIhI6jrYCTvb5w5+hdK4N2bT/QuzoDzX+/fJUcjxxVjTrRPtumoXVXnwnNtDe4iUBfXFhg11rrTNzRzcuRFKutX3NC+1wAOPB2YMzHmeeSxsewxqq+CaK+JapGRebNLpO+Fq4AfW2v559nEDNdEXaTZjzMeBrwKteIPMPSF2e9lXvkuBnTOOz8abO3oSOBzLsomIiIiIiEhiZHSwQySUl1oAB0Hm4OMWEGh+xZLZz3hm3Ibcni7WlXoDye5hy8MHJvFY7yDWuy7JzqiczGfHzwfay3Iz4P9+/EfeIHIkPBaeeDbugeZoF6lMBKefwrnAUJh9lgJT0RVnNmPMp4AvAE3Am621p+fY9TfAe4G/Br4/4743AYXAb621Y7Eqm4iIiIiIxI972NIzZgM/5t3DNqN+zMv8Mi7YIRLK0EhijxPqCg3VeWba7XQX3ImXUemffDMGHInhjIG5RLNIJfBivMoTzOk74ihwVZh9rsE7unjBjDH3AtvwjlB+S4h0GcGeBb4IvMcY83VrbaPvMfKBf/Lt82gsyiUiIiIiIvHT2udhu9sz6wfUPc2TrCk2bEzj6ckyv4wNdojMpaggsceJZBLfjIGojotjoBlgY52LB9vnzxvuZwALD8S1QD5OP43/E7jHGHOrtfZHM+80xnwAeCXwfxZaMGPMbXiDzFPA74A7jZnVS3TUWvsdAF/O6E14A84vGGN+AJwB3gFc5tv+w4WWS0RERERE4ifcgkv7BiwPtk+xaTVacElExDdF31GO5qDjRGQeSTxjYF2piw+vxskilc/HvVA4DzR/CXgP8H1jzC34Ft4zxvw98BfARuAA8PUYlO0i33UW8PE59nkR+I7/hrX2J8aY9XgD3TcD+cBB4BPAv1hrHY0pFxERERGRxGntmz/I7GeBJw9PUZmHRjaLSGa7eCVcVY9xMr3/6svjPtpSJC0k+YyBZFyk0lGg2Vp71hfI/S5wa9Bd/+K7/h3wN9bacHmcI3muLcCWKI77v8DbFvr8IiIiIiKSWNvdnoimgII32Pyc26NAs4jIR27F85EHcEUwts5jDE+/ZQOv7lP7KRLOgbVruYTknjGQbItUOm5VrLUd1tprgVcBdwCfBT4G/Jm1dr219nhsiygiIiIiIunOPWydLmpD+4DFPaxJiyKS4a65gl/dvgmPL93ozFbRf9tjDE/+7SZ+ccFaHmyf4oUeT0KLKZJKdvR42DJaS/sla5wFmRdpxoB/kcqafO/1Yi1SOe+IZmPMduAH1tpnfLffhDcvcoe1thloTkAZRUREREQkjQQv4ubXMxZdwLi1z0NdYdZCiyQiElPuYUvPmA2MLnQP27gGflquv46+qipe++x2Lto3PY2GAfZeWs9zb9tAW/06QCmIROYTnMpr+40b+czDD0Y0YwCXgdtviXv5klm41BnvApqCbu8AtuJdpE9ERERERGRBzo57f7gNzY49R2RkKoaFERFZoNY+D9vdnlkzNO5pnmRNsWFjHPOldlyxjucvXEtpp5s7n3iYLI+HKZeLf7n9Lo7X1s3aXymIREILTuXVVr+Ob77vw3z4e9/EZe2sNBqB2y4D923O+IU2wwWazwElQbcXZ9y1iIiIiIikjbvXnP8Z4h/dPGEt5yacP1aBBjOLSJLY0TP/gqb7BiwPtk+xaTVcWx374O7IlGVwCgZr6zhVWR3YHirI7OdP25pJsQAAIABJREFUQbRY0+xFIvaxL8x//9c/HZOnCZXK64U3Xsepiio2/Pw5Lt8fesZA1d/fQtX6V8akDKksXKC5HfjfxpgGoMu37UJfCo15WWt/u9DCiYiIiIhIZih2tEz5eRqJJyLJIHiq/XzimbJiIMqZIUpBJCmnp/f839UVMX3o1r7Qucvb6tfRVr+OFSdCzxh4c7XhAzEtSWoK93VuK/AT4D+Ctt3mu4SjVkpERERERCJSkGVYU4yjBQHrixdvsRsRkWDBU+3DiVfKCk+Ua6PGOwVRqLz8wYJnuYjMKXjEcvDo5hiNZPYLdz4cn2PGwK97LBct8cRltkIqmfdsttb+yhhTD/wlsALYArzou4iIiIiIiMTMxjoXD7aHHxEI3qmqG+oy+8eciCSHUFPtw4lHygpXlA+VyBRE/rz8AGW56iiU5DPX+fDJbzwU+PsVRw/N2v7lv79bC2wSfkQz1tpjwLcAjDFbgBestVoMUEREREREYmpdqYsPrybs9HMDbFqdldE/5EQkecw11T6S42KZsiJZUxCFyss/c7tIsvCeD/Of02eWlofcrgU2wwSajTH/DPy3tfZXvk0fAHbHvVQiIiIiIpKRrqt2UZXn/aHWHmKEYH2xYUOdK6N/xIlIcok29USsU1b8zZe/zNCkZcrOPeJyJqUgEpmurtCwptjMmqUQ6vwJJdMX2AzXffRx4BzgDzR/G2/e5uZ4FkpERERERDLXulJvINk9bHn4wCQe650Sftcl2Sn/w809bOkZs4H/aVvbBAVZc/9PGvEnkvyiTT0Rj5QV+VkwNDn3iMtgsUxBNDJlGfNA96jFZcjoQJukPiepvELJ5AU2w31rGQQKg26rlRARERERkYSoKzRU55lpt1NVa5+H7W5PyDyuS7IsNfkwGXSXcpeKpI5IptrPfdzC+FNRHBq0fPmjnwRgaQ50jMx/XKxSEA1MWLpHYdA3Ortr1Ht9T/Mka4oNGzUDRWLtYCcc74HJKcjO8t6+eGVMn8KfyuvJw9FNO4j3ApvJLFyg+SCw0RjzHNDl27bUGLMq3ANbazsWWjgREREREZFUt6PHM2/e6cEpODQEKwugwhdY10hmkdQx11T7+cQjZcWEL9btMnBxEdMCwDOfO5oURMH5lQ8NWsY9MDpPfH3fgOXB9ik2rYZrqxVslgV6qQUe/xHsbJ++/eZPwFX18JFb4ZorYvZ011W7ODrk4dcnnY9rTuQCm8km3LeXh4DvAX8I2naX7zIfG8Fji4iIiIiIpLXWvvmDzH4W7wjEXJelOEejmUVSjZOp9rFMWeHvlHpo3ySHBr3PXpZrAtvjlYKoIAv6J8PvZ/GOCq3Mi/+ig5LGtj8PDzwOnjnOsJ3tsPkBuG8zbLg+Zk/75mVZ/PpkBG/0GTL5vT5vMNha+31jzBHgRmAF8H68+Zmb4l80ERERERGR1Lbd7XGU47F7FIpz4lactDEz17Xywcpi80+1D9exFKuUFZGKZQqi4JkW29omOTkWWetm8S7wmsnBN1mAl1rmDzL7eSxsewxqq2I2sjlZZiukkrCjjq21fwT+CGCMeT/wnLV2W5zLJSIiImkseOplKJoyLiLpwD1sHf04Be8095GpaJcfSn9z5bpWPlhJBtdVu6jK8wZV20Oc+9GmrEg20bRt7QMW93B0bVtbRy9HTvYzOeUhO8tFW0cva1dVRPVYkoIe/1H4ILOfx8ITz8Y0hcZizVZIVU5/xX0A2B2PgoiIiEhmOjt+/mubFr8SkXTS2ud8cTCAAeezdDNCuFzXygcryWBdqTeQHK+UFckg2rbN6XHP7+lk2zMv8du2E9O2r7vzad60tpb73n0NN1wZ20XgJMkc7Jydkzmcxr0xXSAw0oUBEz1bIVk5CjRba/8tXgURERGRzBE8Yjl4dLNGMotIOol21flIB25lEie5rpUPVpJBLFNWJJto2zYnx33r121seuQ3WBv6rP9t2wnesuUnPPnR6/ngX66NrkDiWMLTFr3UEv1xMQo0g3e2wo6TUzFfYDMdOf41Z4xxAR8F3gvUA0XW2mzffa8GNgFfs9buj2VBRUREREREUkm0q8670iceFTNOcl0rH6xIfEXbtkV63PN7Orl9niCzn8daNv3rb7igqkQjm+Ns0dIWDY0k9rh5FOcYinPg3LilI+jh77oki6vL9Xnj5+iVMMbkAr8Gvga8AhjAOzrc7wjwQbxBaBERERERkYwV7Y/uYk3umCbR+WBFZH7Rtm2RHrftmZfwhAky+3ms5YFn/hRVeSQyO3o8PNg+NWc77E9b9EJPdClV5lVUkNjj5jEwYTkwYDkyDFP2/OWf90+xrW0y6pQy6cZp63A3cB2wFVgGfDP4TmvtOeC3wF/FpHQiIiIiIiIpyr9avRNLsqAgS0OagyUqH6yIRCaatq2+2ESUYqGto3dWTuZwXmw7TltHr6NjJDJO0xbFvN2NdlG/GC4GCN5g+8Gh0GkzIM7B9hTjNND8XuD/Wmu3WWs9EPK9dgRYteCSiYiIiIiIpLiNdS6chGNq8uNWlJSViHywIuLMq8sib9kMsKEusvDT882dUZUn2uNkftGkLYqpi1fCVfXOjrn68pjmZ/YH28OJW7A9xTgNNF8E/DHMPmeA8uiKIyIiIiIimcK/qFD3qPc6HVMdeFerzwobbDbAqgJvDkiZLt75YEXEmR09Hn7QEVkwzQCbVmdFnDajf3g8qjJFe5zMLWnSFn3k1sgXL3AZuP2WmD79ogfbU4zTQPMIsDTMPquAc9EVR0RERERE0l1rn4dtbZPc0zzJ8RHoGoXjI95FhdIxz+F11S4+U59F/RxTzeuLDZ+pz6IiT0HmUOKdD1ZEIhdpKgW/96xycW115OdiSWFuVOWK9jiZW9KkLbrmCrj3I+GDzS4D922OadqMpAm2pxCny0w0AW8xxuRaa2d1FxljSvHmZ/5DLAonIiIiIiLpZUfP/EEKf57DTatxFJxIdutKXawrdeEetjx8YBKP9f4mvuuS7EDe0l90pVeAPVb8+WCd/NiPNB+sSCo7NGh5aN9k4O/hqenb714T+5VFnYzuBGg6a7mp1vu3fxaLv/1zD9tZ5+kNr4wu5UG0x8nckipt0cYbYEU1PPEsNO6dff/Vl3tHMsc4N/NCgu11hZk5rcZpq/Mk8DTwtDHmQ8F3GGOWAk8BZcBjsSmeiIiIiIikC6eLClXmJaJUiVVXaKgOGrmsYGhkNta5eLA9slGUTvLBiqSLCQ944jyIMtrRnTtOTvG707OPvad5kjXFho11rsAMhLWrKnjT2lpHCwKuX7uCtasqHJVLwku6tEXXXOG9HOyEe74Kk1OQnQVf+oeY5mQOllTB9hTh6NPXWvt9vMHkm4FTwB0AxphGoAt4J/CItfbnMS6niIiIiIikOOU5lGg5yXXtJB+sSCp7xRLD3WuyuXtNNjku7yhhlzm/PdaiHd355BHPnAFq/yyWF3rOP/Z9774Gl4msE85lDPe++7VRlUvml7Rpiy5e6R3dfMFy73WcgsyQhMH2FOC49q21HwI+COwFqvB+lr8GOAh8yFr7sZiWUEREREREUl60I+FGpjI3z6FMF2mu63RKuSKSTOI1StM/i8UfyL7hypU88f9cHzbY7DKGJz96PTdcqbQZ8eBPW+REuqUtStpgexKLqovLWvsd4DvGmAK8qTL6rLVDsSyYiIiIiIikj2hHwg1MZvbIIJkuklzXIployp5PndEzZkPmP16oeLbF/lks/gDdh968lgurS3jgmT/xYtvxWfuvX7uCe9/9WgWZ4yzT0xZpjQDnFjSXwlo7AozEqCwiIiIiIpKmoh0JF++co5KalOtaxKu1z8N2t4ehoDb2+Mj0/Mex4g0Cxy+lUfvA9AD5DVeu5IYrV9LW0csN//RzPB4PLpeL5z/7NuVkThBv2iLCrq+QzmmLMj3Y7lTsk/aIiIhIyvGvWD6XeOT5E5HMEu1IOFeGxA/dw5aeMRsYoRuP0Ygikl56x+y8ATB//uOVBVCRt/D2JJrRnU619nmoK5z+gbF2VQXl5SXTbkviXFftoirPO+K8PUTd1xcbNgQt6Jhu/MH2Jw/P32OezsF2J/SrUURERKY5O37+C2RZroIcIhIb0Y6EK07zXyz+0YgzAzfBoxEz/UerSCoJ7rw/NGhnbY9V5/3AhKUjgvnlFugYgVyXpThn4d/rnIzujEa88kDLwiRF2qKPfeH8360HZ2//+qfj9tTXVbvYcXKK7lEYDPEeTfdguxNp/rVNREREIhH8oyf4B5JGMotIrESb5zA/TfIzhwo+jXtg97m5oyr+0YibVqMF7kRSUHlu7B5rZhsyNP9ktFm6R6E4Z+HliDSVQrRikQdaM/XiJ2nSFlWXJfwpi3MMxTkwMmU5ErRK3QPrtEZAMJ1dIiIiIiKSENHkOfxFV/zygS6WCY934a7xCF4Ii3e6bmVeZq9iL5IqEhHEXJIN/Q4DzYNT3gBZLIRLpZAF1BZAZxQreoVq59zDljHP9NuRBvY0Uy+NxHHEshN1J9xctrOV/JERRgsKqCt/FVysRSn9FGgWEREREZGEiGZRoXQJNAcHnzY3TjDuIEhk8QZ0FGjOLBqVKcGC6/u/u6b47jHnbeOAw+D0fGamUjg95u1EK8jyLuKaZSDPxbQAcTj1xWZaADk4tdBo0OSPcKmFNFNPYiX4/ZP9pxZu/Nl2Lt2/b/pO3/8uXFUPH7kVrrkiwSVMPjrLREREREQkYTJ9USHwjmZ2moa0fcBqgcAMplGZEizaPMaeOOS68KdSqM7zpvR4xRLDoUFLWa4h21gODeFoFovfjh7PvJ2SSi2U2RLdEXfFb3bwV48/gcH7fp7VCu9sh80PwH2bYcP1MX3uVBOTV94YUwn8BTAM/H/WWqVvFxERkbDcw5aeMRtYUERBFJHMkBSLCi2iySiDPa19HuoK0yRptYSlUZkyl2jzGLsS3LwW5xg+vNrlaBYLeNu6SHJAK7WQQHw74u5ekw0vtcCTTwa2zfkMHgvbHoPaqowe2ezoE8oYcwfwfuCt1tozvm1XAf8NlPt2azTGXG+tHQr9KCIiIpLpgqdCBgs3FVJE0kvSLCqUIqIdxSgi6cX7Hcl56oziReijiGYWy3a3J+KFBpVaKDMltCPu8R9FPh3AY+GJZxVoduB/AdYfZPZ5CCgDngKWATcCm4GvxKSEIiIiklY0FVJEJDrRjmIUkfRSV2hYU2xmddjPZ0kWFGQtTmeek1ks7mHr6P8CpRaSODrY6U2L4UTjXu9xGbpAoNNfb5cAzf4bvpQZ64FvWWs/bK29CWgA/iZ2RRQREZF04XQqZGtfeiwCJiISLDvKWIhG7ImI38Y619xT+EOoyY9bUSLmn8VSk++9DhUYjva7n74zSly81JLY49KA028qFUBP0O03+K6fC9r2O+CChRTKzxhzizHm68aY3xlj+o0x1hjzvTn2vdB3/1yXH8SiTCIiIhK9aKZCioikmywDTgcn1xeHDsqISGZaV+riw6uzwgabDbCqwJsvORVEmyJIqYUkLoZGEntcGnCaOuMMUBl0ez3exEB/CNpmgVj1lX0WuBIYBNzAmgiO2QP8JMT21hiVSURERKKgqZAiIufluWA4wr40A2yo02hmEZmuKg/qCqBzjpjWqkJ43wVZ/KIrdTruo00RpNRCEhdFBYk9Lg04DTS3AzcZY/4PMIU3Z3ODtbY/aJ8Lge7YFI9/wBtgPog3qL0jgmOarLVbYvT8IiIiEiMLmQpZV6hfDyKSXrJd3tE5Yx7mnelhgE2rs5Q2Q+IqeDGtUOKywJYsiH/Nixd+9PN59/vrT92UoBLFRrQLHaqNlLiIdlG/DF4M0OmZ+DCwHG/wtxOoAR7x32mMyQLeiHdU8YJZa3dYaw9Ya50NfxIREZGko6mQIiLe2R1jnvMB5g9f5KK+OPSsjfpiw2fqs7QwqiTU2XEbuEhymmvNi5HBocDF78nDUwxMpE5d+hc6dGKu1ELuYUvPmKV71HvtHk6d10GSxMUr4ap6Z8dcfXnGLgQIDkc0W2t/aozZDNzu2/S0tTY4Z/Jf4u2Y/2WMyheNWmPMR/Dmk+4F/sda2xzmGBERkZSVKqOQNBVSRDJZa5+H7W7PrBRCTx7xsKbYsGm1i593efBYcBm465JspQ2ShAn+rhD8vSJZvkPIdMFrXqy/9W2B7S8GjW72b7dA9ygU5ySwgAu0sc7Fg+3hF4+G0KmF5mpv72meZE2xYWOdSyOgJXIfuRU2PwCeCN6RLgO33xL/MiUxx58a1tongCfmuO+XQNlCC7VAb/ZdAowxLwC3WWs7FqVEIiIiMRIqqHxo8PyXnvLc89vLcp0HKPwjP/yBjljmR9ZUSAkW/F4Ofg/7tyu4IenEP8V9rp+o+wYsLw9YVhZAdb63zVWQWWTxJWNnfjRrXgxOwciUpSArNdoV70KHzNtuQujUQpG0tw+2T7FpNZotIpG55gq49yOw9bH593MZuG9zwtNmJNt36qwtW7Yk9AmjtXXr1guB9wMtW7Zs2R7i/lxgFPgE8Gngi8Av8OaMvg64aevWrd/esmXLRKjHN8bcvnXr1sdvvPHGWoBLL72U9vZ2Kisr2b17N7W1tTQ2NlJRUUFbWxt5eXkcP36c4eFhhoaGOHHiBLm5uezdu5eqqip27do17Zj29nZycnI4ceIEQ0NDjIyM4Ha7ycvLo7W1lWXLlk07pqysjJdffpmsrCy6u7sZGBhgdHSUzs5OCgoKaG5upqamhp07dwaOWbp0Kfv378cYQ09PD/39/YyPj3Ps2DEKCwtpbm6mtraWhoYGVqxYQWNjI6WlpRw8eBBrLadPn+bs2bNMTk5y9OhRlixZwp49ewKP73++kpISDh06hMfj4cyZM/T29uLxeDh8+DDFxcXTjlm2bBm7d++muLiYI0eOMDk5ydmzZzl9+jQABw8eZOnSpTQ1NQWOqa6uZs+ePRQVFXH06FEmJibo6+ujp6cHYwz79++nrKxs2jFVVVU0NzdTWFhIR0cHY2Nj9Pf3c/LkSbKysti3bx8VFRXs2rUr8L9XVlbS2tpKXl4ebreb0dFRBgcH6erqIjs7O1D//mMaGhqorKxk7969gbqcWf9tbW1UV1dPq8vy8nLa29vJzs4O1OXM+p9Zl/76d7lcnDx5kv7+fsbGxujo6AjU//Lly+es/1OnTnHu3DkmJiZm1X9jY2PguqSkZFb9T01NcfjwYZYsWUJTU1Pg9aqpqWHXrl3z1n9JSUmgXhoaGqipqaGpqYmioqKQ9X/gwIE567+wsJBjx44xPj4+q/7Ly8unnZdVVVW0tLSQn59PR0cHo6OjDAwM0N3dPa3+g48Jrn//uTw4OMiJEyfIycmhvb095LkcXP/B53Jw/e/cuTPwupWXl7Nv3745z+WWlpZp9d/Q0BCy/v3nsr/+Z9al/1wGptV/qHPZ/97x17/H46G3t3da/c88l/31H3wuB9f/oUOHKC0tnVaXy5YtC9R/8Ll86tSpafW/e/fuwDlWXV0957nscrlm1b//GH/9d3Z2znkup0tb3rSnGU/ZcnKO7KS3aDm5R3aRXVTCyoEjlOdA0XAvhWN9FDDJW7M7ecuFJRG15bs7T/P0kQl2tB6ix1XM0uPNuPOX83JTI3+YrGbo0G5qli6sLbfD3nO5fxKqzxxiOH8pK0/uoa94Oau6djFQVEVtTysTOfmU9bvJnhrjsqxBLvGcUluehm35IZaSc2wPnrLllHbsIqu0kuqeNrLyCsg+42ZVzpja8gxqyz3FVeS421hTUTCrLd8/5CKn62U8SyrJ62hKuba88ehJnt1/lizPOOV9nYznFLCipzXQ9vmvR/JLyO05TA6WvKFeaqbO6Xt5mLZ8V8tebHYOWee6WJk1kvZtuaeslpwju3jtK5bHtS0/mltDzpFdeEqqKDzektHfy4+ZpWR3HwDjYvhMD5ND/YyPjVF4rhObW0CuuzXhbfmR7GrO7t/NWG4xleeO4rJTFI6epWikl/0Hu8ibGGbKlcNflA8G2pf+omqqu1vIKSikor8jpm3513+1kwM5NeQe2UVvfiX5x1sZNPnQ66Z/eIQr8oanncuHXRXkHGvCU7acvKO75mzLc/u6eHVVLuMd7XTnVnKyaReT2fkUjPVxweWXcuXgy7z7wjyWj3YH6v+lA51870Q2y0+1M1BUzaru3YHXYKigjGW9+/G4XJQM9dB2sp8yRhk86Y7qe/nB8XxyOlvxlC0n98jOjIqxvDxkyOnaj6eonNxjTYv+vXx/Tz9mYpSss8dZV1kQn7a8cgmjl19Ebm8/WSfPzA4qXn05L7/nevLfvj7h38vbTpzBNXAag2XJ6UNkLyllWVczrnLvZ8bVFy1jz549LF++fOvcUdfYMU7SHxtjXg28Hm/KjD7ftiK8eZrfCQwDX7TWPhzzghpzLd7FAJ+21r7PwXHZwO+Ba4CPhyvbo48+agHuuOOO6AsrIiKSAKGmtkYz3TXcyA84P2JkoSM/Wvs8jqZCfqZeC2BJYiTjqLVMEa7dSvVp/NvaJh2NPlySBZcUm5T8XxMt1d8bTiTyf82k19WJZHldnnNP8SN36BlioVJn+C3Ph5r82LctD+2bDIyinPDAsnw4OQo5vq9vj12dM2t/v0jL4h62rN/y08DtF7e8I+SsD6ftbX2x4d610b0eyfJ+WAzJ9r8nvDwHO3n+Z3vIHR5mvLCQG266MlVyMidkSoPTGvgU8BfW2keCtj0I/C0wiDcv8j8bY9qttb+KURkXxFo7aYz5Jt5A85vwLmgoIiIizL2YzEwW72IylXkLS2WxkKmQIokSvABWNCloRPwWMsVdZDHEM4VWulmsDspo165wxaka716TPSvQF+61caqu0JDnmn57pmja2/YBq/e4OHfxSna9dXng5g0XL36wPZk4fTWuBl7w3zDG5AC3AX8CrgXKgd3AnUBSBJp9Tvmuixa1FCIiIkkmeDGZcCzwnNuz4ODvddUuqvK8j9Ue4gdBfbFhgxZpkQTTQlgSD619zvPSAwzENkYjEpYWT1uYRHZQRrvmRXGCPs4Wq7Mi2va2tc9DXaFWnhaJFadNTTXQGXT7aqAYeNxaOwqcMMb8J/DXMSpfrLzOd314UUshIiKSRBZz5Me6Uu8PVvew5eEDk4EfI3ddkq1RJSKSNkamojsukoXtRWJFi6dFZ7E6KOsKDWuKjeOUPPFeCHBgwtI96u2cCJaozopo29tojxNJuI99Yf77v/7pxJQjDKdnuWV6cPqNvm0vBm07BVQtsFyOGWOuMcbkhth+PfAPvpvfS2ypREREktdCRn7ESl2hoTrPUJPvvVaQWUTSSbJNcZfU5B8h2j3qvXYPx64nwmkKrVh+B5DobaxzRZxs1QA1+fEsDfSOWQ4OeVP/hOLvrHihJ37vn2jb22iPE1lUPb3nL0nGaVdbB+dHB4N3AUC3tTZ4pHAtcHahBQMwxrwLeJfvZo3v+vXGmO/4/j5trf2k7+8vAmuNMS8Abt+2VwLX+/6+11r7h1iUS0REJB1o5IeISHwl+xT3VBQ8ctS/AFnw9nRKd5OIdBaLkUJLFs7pmhcNZ+IX4G3t89AxEn4/f2fFK4qgOCf2vWnRtrd6P0vK8I9YPtgJ93wVJqcgOwvuem9SLUbo9FP4GWCrMeZZYBR4PfC1GfusAw7FoGwAr8KbAzrYat8F4BjgDzT/O7AB+DPgrUAOcNJX5m9Ya38XozKJiIikhbiN/EiRaV0iIvGWrFPc00X5rPms6SMR6Sy0eFrkknGRRCdrXsQz0LzdHfljW6B7FIpzYl+OaNrb+mJns+kyqaNLktBLLfD4j2Bn+/TtN38CrqqHj9wK11yxOGUL4vQs+Cre/MsbfbebgG3+O40xlwNXAZ+PReGstVuALRHu+y3gW7F4XhERkUwQ95EfQyPQ2Q0eD7hcsLIGigocP5+ISCrbWOfiwfbwqQn84j3FPdVlQiDHaTqLyrzoRmVq8bTwkn2RxJlrXvjT7hRmwb1r43+uRNNZMTgFI1M2og614MDu0NTs7TPbAyftrQE21EVfd+nc0ZVKkrETKC62Pw8PPD73Ig4722HzA3DfZthwfeh9EsRRy2OtHQTeYIxZ59u011ob/Ok0jHdUcWOMyiciIiJxEreRH++7MXRve88Zb2/7Sy1J0dsuIpIITqa4ryyIz5RySS2JSmehFFrzczKqfLHVFRo8584y1NePx+NhIstFW0cva1dVxPV5o+2sGJh0PrPORnBSOE0p4vS8yYSOrlSR7J1AMfVSy/xBZj+PhW2PQW3Vov7Wiuossda2zrH9KHB0AeURERGRCITqvQcc9+jHfORHCvW2i4gkSqRT3H/RpYXWMl0i01lo8bS5OR1VHq+8w5F4fk8n2555id+2nQhs6wPW3fk0b1pby33vvgbylsfluaPtdAgXL/MLDux+w4TePpOTlCKSmhKRWiipPP6jyE8aj4Unnk29QLMxJge4AagHllhrH/BtzwdK8C7Sp29JIiIiMTZf7/1MkfTox3TkR4r1tovEW/CU31A0MiqzBE9xv7f1/HvjgXXZgQChAs2SyHQWWjxtbk5Hlccr7/A0IdbA6Dg1wMix03zSwm9r1s66/7dtJ3jLlp9w88Zr+bM/q495kaLtdHDFOSY/M6WIfxDGXZdkp2dahQySqNRCSeNg5+xZouE07vUet0gLBDp+tY0xf4131PJ/AV9heg7lVwFdwP+KQdlEREQkyI4eDw+2Tzka6eTv0X+hZ+4fktdVu/hMfRb1xaG/eNcXGz5TnxV+REA0ve0iGeLsuA1cJLPVFRryXAQuCnpIsESms/Cn0HLC6eJpqSjavMOdvtlm/llmcTM0AvuOMLpnP2b/UYqm5u/U9FjLs9tf4MBBd8yLEm0A79y4dzG9h/ZNhu2UXYi6QkN1nqEm33ud7u/dTBBNaqGU9lJLYo+LAUfDKIwxVwNAIM5JAAAgAElEQVQ/AU4D/wC8Fvjf/vuttX80xhzBm6f5+zEsp4iISEaLtPc+lEh69Bc88iPa3vYP3s8GG7Ty1RLfc339084eSyQJBY9YDv4hrZHMIjKXRKezSOTiaaki2lHlp8e913HLETtjDYx8YCWwcmqcF/NLuH7kLL8pKAPgp91tsw7P/8pBlq1cxsEsw3Ofujsmn0XRrPeR54KqvAU/tWSgRKYWShpDI4k9Lgactiz34l3w72prbbcx5v4Q+zQAr1lwyURERCTASe99KJEuFuQf+RF8OyLR9pr39FI9NIHL48HjcsFFNVBUEN1jicRAxqxeLiJJKdHpLOK9eFoqisVihzHPETvPGhgWWD/az6+6W9lUeQlPFdcE7lvimeTiyVGysEyNDzK0tBCWljh66nAjjjfWufh8e2QvmgH+38sy430ksZfI1EJJI9rfRYv4e8ppoPkNwE+std3z7NMJ3Bh9kURERCRYNL33oczXox/8I+LQoJ21Peyol2h7zTtPMu3nzpkzcFW9N3Ct/M2SQBm1enkKWHCblEA3/dNP573/Z599R4JKIukgmhGiC01nocXTpovVYocxyxEbZg0Mf81nAU+ePsCx7Dy+VlrLfec6WD/WP33nwwfhqnru7msHnH/PCk7/VJbrfeZ1pS5WFUzREear4EI6K9zDdtq/r07gzJTI1EJJI9rfQym0GOASvGkz5lNIFLmfRUREJLRoe+/neqxwPfoT0TxdLHvNd7bD5gfgvs2w4frYPa7IHDJu9fIUU5672CWI3PHeocDfKyqKFrEkksoWI52FkxRa6b7QabSjykOJdEbZvBysgZEF/OvpQ1wyOUKW7/mDa88CxuH3rEjSQFXkGXJdlu5Rb77qmaLtrAjuBA5+CdQJnJkSnVooKVy80jsIx0mKwqsvX7SFAMF5oPk4MHsp0+leBRyOrjgiIiIyUyx74ed6rOAfC5sbJ0Jun1ese809FrY9BrVVGtkscZVxq5eniFQKVAWPWA4e3TxzJHNwgGZ4avb2VPqfJbZCBW5XFhDXEaJzcZpCK9QI11QXzajy+SwoR+z/z959x8dR3Xsf/8yqd1nNkiy5YBtsbFoohpBgwPQSSoCECzcdAmk8SS5JbgLENrlJnnDTCzcQLjxJSIAQQiBAmjF2MOBeZRlbbpJsq1iy1fvO88dq5ZW0krbP7Oj7fr30krS7s3uk2Zk5+zu/8ztBroFhAvP6T7xxRr7i0O9R6GdlJRlkJcHt0xNCW+9jBA0Cy0ixLi1kG5++1TM4FMiAk8uAu2+JfpvGEWxv5jXgHsMw3mea5psj7zQM42rgvcB3I9E4ERERiewofNRG9EMZbZ+I24THnp/UgWanZ43ZQSirl8f9BxaxXJIz4nESBd7ArcuAORlEPEM0Epyy0Ol419h+t4kBYa2P4SvkGrFBroER1KklyH6Wdw2Doz2e35dX9JGW4HlF3xJHIa/34UODwOKPFaWFbGHRafDgp8ctoQN4LhwP3WP5Z6dgrwTfAT4M/N0wjJ8CMwEMw7gWuAj4LHAE+EEE2ygiIjKpRXIKZ1Q74cGMtgdqw05PNo+F07/swolZY1ablKuXi2WcEpyTyJvovRFIOQsJz8hrbFaSwadOcgUU7AxEyLPTQl0DI1AB9LPGWsNgVxukuExykzxl15Ii2MXUILCMxYrSQrZw8xKYVuQZHNqwc/T955zqyWS2QYJOUL0a0zQPGYZxBfAccL/PXS/h2Yd7gZtN05yojrOIiIgEKFJTODMTQssqCVigo+3BWrt90gaaFZiKrkm5ermIxJ1IZIjKaIFcY8dbJDEYIc8oi+QaGGMZp581UfmKHjfU90CqC2ZnRuZ9qUFgGc/CHBefOokJB4GiUVrIcotO83xV1cBXfgj9A5CYAN/7oq0+KwX9KcU0zU2GYZwCXAtcAOQDLcA7wJ9N0xx/jqeIiIgELZjR+7EUp0asOWObaLQ9FNHO5pFJa1KuXi4ituUtTeDNXFbgzHojF0nsGYCjvaE9T0hikZ04Rj8r0PIVAN1uaOszyYpAXaAf7zkRUlr1h1eHfm460jDq9sW3XjOsvRoEjh7fwRjfUilWrDNwSZFr3EEgK0sLxcSccs/nLd/fbSSkd4JpmgN4sphfmuixIiIiEr5AR+/HcvdJCbFbKGWs0fbFZ8P/vhj888Uim0cmpUm5ermI2M5YpQm+sq2feVkGNztl6rdNBRLg980q7x4w/dbNHktYNWKjsQbGSGP0s4IpXwGeeuJZSeE3Z6xJcelZGeNup0Hg2MlLtroFoweBJkVpoc/7LIe3o2r07T/9WmzbMwbNuxQREYkTE43e+5OZ4Mlkjulq3L6doGOtJ37euju057NBrTFxplitXq4sxcmhorqJ/fWt9A+4SUxwUVHdxILp+VY3S2xuotIEu9pMvlM5QHka5KdE77xhp2zFWAkkwO/vfF+cCns7AlsoMCI1YoNcA8MkyEUB/fSzQilf0T4AXQPhl067fKqLXx/0XJt9M5YnokHg6LLrOWDSlhYqmmJ1C8YU0jvFMIxzgPOAKYC/w9k0TfPhcBomIiIio403eg8Muy0rkaHVwANV22nS4x7+e1gdNt9OUEYa5GRCS3vg259zqu2mg4lzRHv18lCDGJOBk4JaK7bWcNsjr9Lc3jPs9oVfeJq8zBROLs3l7e99yKLWiZ0FWprABKq7INkVmdIEE7FDtmK0BRrgv+uk0YP1gS4UGLEasROsgeENLLsNg/1pmczubAv8ucfoZ4W6hkFbBAqpxmoQOBy+1zB/4ukaJnHEJhnLEwnq3W8YRjbwAnAJ4w+SmYACzSIiIlEy1uh9UYjZThENiI3XCVq7PfCsHJfhWT1ZJIqitXp5OEGMySaeg1pP/KOCu3/xOm7T/55ubu/hnd31/O8/K/jEZQti3DqxO6tKE/gTSmAsXmdrBBPgf3zfAAUpo4OYMa8RO84aGAaw8+T5/Omam9i2ei2/27TCbzbgKOP0s0ItQ9HYHXqQ2ivag8CRdqz3RDunJNv//S8SbcFeTR4BLgX+BTwJ1ABa/E9ERCSOxTQgNkFWzhCXAQ/do7IZEnXRWL08EkEMp3NCtteKrTXjBpl93fXz15lRmM2SMzRDQzysLE0QrnifrRFMgN/EE0z29/fEvEaszxoYhz7/fYwBNwMuFz+5+z4OlZYB8OaOar4y/3y+V/kOCYxTRmOCflaoZSh6TSJS6iVag8CR4nsN881udsK1TSRcwR4FNwCbgEtM0wxvmEpEREQsZ0lAbJysHMAzjfPuWxRkDoGmcoYm0plpkQpiiL0tf25tQEFmALdp8vBz6xRoliFWliYIR7zP1gglwF/ZZlLbOfY2MasRO7gGRteASWbbifIYt7/wewD++3P3A/DMtDnUpmXwzfbDnLrbzyKCAfSzQi1fAZEp9eIdBH583/ip1RErTyJx4/pvvTTqNt8SXPd/94ZYNkf8CPbTRg7wGwWZRUREnMGygJhPVk7jfT/ANTCAOyGBwh9/STWZ/QhlerKmcgYnUplp4QQx4mHKuXhUVDexuuJwUNusqjjEl1c1UDw1b9iH4ns29AEwO9OI+3rVErhQSxMEuB5cVDhhtkaoAf5wy0FEUls/dOTmjfuYN/NK+NanP8m0w7V84bEfk+B2k5WaQNYPAutnhVK+YqRwS71cUuRiZf0Add2ebP6RIl6eROLOoaYOAFp6THKyMyxujXgF22vZA0yNRkNEREQktiwNiA1m5QCkt7aeuP3HT3u+x8liF9EW7PRkTeUMX7iZaeEEMcrSQ5yrLDG3YltNSNtVVdVSPDVvWF3q5t4TP8dzvWoJTqilCVwWjkc5YbZGqAH+ULeLqMG+0ZraAf5QO/615qktK8mvrwAgu92T/ZzUQ1D9rGDKV/gTiVIvWUkGWUme52nrJ+BBYCctOCvDvfzAB4Z+9mY3d7ebfPxj11rVJBkh2KPr58B3DcOYZprmoWg0SERERMI3skPuLzhsl4BYe96JrBy75iL4m6bny7fTGynxPj15sorrIIYErLWzd+IH+XFuVr8CHAKEXpogy6K3j1Nma4Qa4A91u2gIti3Ng9nPecmQGsR2ga5hMJ5IlXpJSzCG/d3BvKeiOYAXrwtiikRTsJep1/AsBrjGMIxlwEbguL8HmqZZHWbbREREJEhtfabfKYb+MmAtDYj5ZNL8Os4yb73T9ACm5UcnNO6E6cmTlROCGDKx7PTQIhehbifOE0ppgswET8DNCnYYnI7EOgShBvjXNbs53DW6LVb0WwL5Gz525iUsvvWaYbd97/REcoIMgnrXMHh87wCNIYyvWVXqJdr7Jd4XxBSJpmCPvgOcWLj0V+M8zgzhuUVERCQMKxvcVHWMff/IDFgFxALnb5reyNsjyQnTkyerUIMY2n/xZcnpodWSD3U7caZgSxMUB5OOGmF2m60R6joE4QT485ItLJDtI5S/YX6WEXKm7cIcF1eXmPz6YPDXNitLvUSLZpyJjC/YYPCvIeRZEyIiIhIl3gzYifhmwH7/sVfY2Tr2ZX1kJoyXAmLR5ZTpyZNVrAMAYo0F0/O5aEFpUAsCnlyay4Lp+VFslcSbQEsTGEB5mqderVXsMDgdqXUIggnwG8AXTk6wXd8n2L/hprLw2h9vpV6iRTPOJNoiMXPDakG10DTNj0WpHSIiIhKGUDJgUxMgMxHa+6Gr/UQqdFrm2OUgohkQe/KpV1idOfZzRyt72G4iMT3ZDjUDrahrbRexDgCINR66bRGXf/NPAZ979xw5zoqtNSw5Y/ysZid8yJTAeUsT/KnWTaWfAar5WQY3lbl47Uho14ZIcdJsjWAC/HedZL8gs9fUFKjrGf8xkfobrCj1MrIvk5VoXekYL804k1gKdeaG1dRLERERiXNdAya72oLbprLN5NEvXc/xPs/0vjf+8OrQfWNlMscyIBaLOsihqKhuYn99K/0DbhITXFRUN0U8QzGc6cl2rRlo1/0ZLU4JYsj4lpxRztzSXHYf9rtkzSimCQ8/t27CQLOveP2QKcFZmOM5N9d2mvx4T/9QYO2+uYlDg4RWB5qdNlsj0AC/Hc/PE5Vu8EpxwZdPidw15uYyF9+uDLyTEmqpl7H6MgCZCaZlJWQ040x8B4L3tpujbo/EQHCkZm5YKaiWGoYxACw1TfPhcR7zDWCZaZrx818QERGJY6Gu6L2jxc1VJQksSW3iL8dbMU03huGipbGZnMK8YY+NdEDMX0ftoluuYfZgRvOTT70ydP9Ft1wzaptYd7ZWbK1h+XNrR02TX/iFp7loQSkP3bYoqODReEKdZlzbafL8OJk2sa4ZGOu61nYTz0GMSHJydm5FdVPAQWavVRWHJhyg8v5PRgYdb5+eoGCFw5WlGxSlGMN+txOnzdYIJMBvN4GWbgDojdDYhO95PNUF3RM8bzilXiYKorcPQFUHvNHgjnn9YzssiCkn+CZ/HO83qKtvpnhq3sQbRkie1vYdU7A9O2PwK5DHiYiISAyEuqL32p01fOdnG0YFT1994lkKy0tYeOE5FM8sIzMhuvUJR3bU6uqbaW5uxe1243K52F3bRGGRp+NoRUbfE/+o4O5fvI7b9P+PXl1xmCuWvsjjn72UT1y2IOzXC3V68jvNpi1rBsYiC9yu4jGIEU1Oy85dsa0m5O3GOwbsOjNBxC6zNSJdHsruAX5fVpduKE71zKA63gc9froq4ZR6CSaIbkX9Y7stiDlZjZX88YMfPsOsWaW852ORS/4YKZ4Hx2MpGv+lKUB3FJ5XRERE/AhlRe+9Wyt55q+rMMcInjbWHGHls3/hhhsXc9aiUyPekffXUVuxtYblvxndcfzVz55l1qxSLltyDt++bmZE2zGRFVtrxg0ye7lNk7t+/jozCrPD7tyGMj05PQE6A/wgE6uagbHMAre7eApiRJoTpoCOpbWzN+LbTZTNF+uZCSIjWTlbY7IPwlhVumGs83WkS71YHUSfiB0WxJzsJkr+2L8/sskfEpoJe3iGYVw04qaZfm4DSACmA3cA70agbSIiIhIA74req3zqLPvjrb1cd6CWdX9d5SkYOh7T5OUXV/H5M6cA0Q0IBtJx/NUTL3N+amw7jsufWzthkNnLbZpB118dSzDTkyHwILNXtGsGxjoLXMQK2emhzZsda7tAs/msmJkg4suK2RoahAmtdMO0w7W0ba6AhB7ISINFp8GcyPTpIjmIGg/1j520IGY8siL5I1CTeQFsfwJJJXgDhs7nJvDRwS9/DDxH3pfDbpmIiIgEJC3BYF4WrPK5rav9xOJraZnDF1/bsWbDxEHmQZEMno4l0I6jGeOOY0V106hs3IkEUn81EMFMT74g3+CtpuDrp0SrZqCdPwiIRNKS00N73461nd2z+URGitVsjWgNwsRiYa9ICqYEw0vrXuPkt18kq6N99J05mfDIlzxBZ5uIh/rHTlsQM95YlfwRrMm2ALY/gZw5l+M5ZxvAQ3gCz6v8PG4AaAJWmqa5K1INFBERkfHtbTcpSjmRsQzDs5t9b29pbKax5khQzx+p4OlY7NpxjFb91fGMzIho64cjXSbtg5+Fffeld3rynjYTAg5PnRCtmoF23Z9irUjXVLWDBdPzuWhBaVADUosXTPN7foiHbD4Rq8RiEKYvQgvnRVOgJRg+fKiK97Q2YXAikDNMSzvc8zA8dA/cdGlkGxmiQPok0w7XsnDXDtK6uuhKS2PHvIV0lU2PfuN8OG1BzHhhZfJHICb7AtgjTRhoNk1zqfdnwzA+CrxomuZPotkoERERCU5WksH0NJPqLk8wuf14K6bpxjBctDQ2k1PoWUyv/mBtSM8fTvB0PHbuOEaj/mowvBkRGcD03Az2dZiUpI6enlwbbN2MQdGoGWjn/SnWcHpN1YduW8QVS18MaHDFZRg8eNt5fu+Lh2w+W6mqgbXboaMr4uUArBBvmbWxFM1BGN//6z0b+vzebieBlG54X/MRvlf5zlBwecz/gNuE5f8DpYW2yGwer0+y7LsPUdxQ5zc7uzcrE74fu+xsuyyIOdlYkfwhoQvqDGqa5qxoNURERERCMzvT4P55iazYWsNX/7SWje8OD/S9+sSzFJaXcM3l53JBdj8bQ3iNSAVPRwq14/jbtdV8J8odx0jXXw3EeBkRvoEI3w/PdqoZqA8C4msy1FRdckY5j33m0gnLxbgMg8c/e+mY2fuhzjCI1swE27rz61B9xJOROZINywGEIi/0S4gjRXMQxntd7Ro4MXMIYHlFH2kJhu0CzoGUbvg/+7YT8NCT24THnrfFMTNWX+biN1cyZ3+V3+xsE0hui312tpULYk5WVid/WCGeZ4LZ68wpIiIiIZlo8bXGmiP85smX+dD75ob0/OEET8cTagfw1eoermxwRzU4Fen6q9Fip5qBk/GDgPg3mRa2++TlC5hZlM3Dz61jVcWhUfcvXjCNB287b9wSMaHOMIjGzATbemEFbN8z9v02LAcQKLsFNO0kmoMwbX0mdd3QPuKxu9ogM8FkR4v96qCPV7ph2uFaLjje4L9cxlg27PTMELB4RoC/vsyCyh186re/GjM7e+h3C7KzrVgQczKzIvkjFBXVTeyvb6V/wE1igiukGXtOmAk27hXNMIzXGVwA0DTN2sHfA2Gaprkk7NaJiIjIhPZU1fLEE4EtvvbMv3aH9BovvF3FP7ZUj3l/qDXIQu0AJqYkRT04Fcn6q8Hy11GFnDEfb5eagfHyQUCib7ItbLfkjHKWnFHOxd94no17GzFNk5lF2Tx7/9UBnRPsNDPBltZuh4d/OfHjbFYOQMIXrUGYlQ1u9naMvcJB+wC2nHExXumGhbt2AEEEmb3Wbrc80Ayj+zI3v/ICrgDXfLAqOztWC2JOdnZP/lixtYblz60d9Zlh4Ree5qIFpTx026KA1iNxykywiYZOLx78nj7i94kEvyKNiIiIhOSfKzYEvPiaCeSkJ9MSRAbp4gXTyEpLGvo9kqsph9oBLJ5RFpPgVKTqrwZqvI7qrFmlXLbkHObOKRu1nV1qBtr9g0Agrv/WS8NqpK7OHP6hcbIu7BKMybywXVZaMukpno9Ys6ZmBzzwZKeZCbb0yz94AkmBsFE5AAlfNAZhIjXjwqra2mOVbkjr6grtCTtC3G6EcP8fvn2Z0sO1zN+zKy6zsyXyrEz+mMhEs0pXVxzm8qUvcsvNF3PuufOH3ed7TDhpJti4R7ppmq7xfhcRERFr1dU3s39/cIuvtXT24jIC+8zuDZ76jsJHcjXlUDqOReWlQ4sbRjs4Fan6q4GYqKO6f/9hfvXEy9xy88Uwb3QAxQ41A+38QSAULa0dHOrxvLfCHVSZTLSwXWjsMjPBdqpqYGNlcNs4JOCkRQKjMwgTjRkXsa6t7Vu6YfFST79szoH93BbKk2WkRbRtEPr/w9uXqV1TAcRvdrZEXqyTPwKxYmvNhJ8RAEzT5PkX3iAxM5OZs8uYkjz6ne2kmWATlc74AvCOaZrrYtQeERERGcfID53bK0JbfO22983luTerxu0YGREIngYimI4jhsGCC88edlO0g1ORqL86kWA7qitOz/X7enaoGWjHDwLB8F148cmnXmHaYEazMpkDp4XtQmOXmQm2s3Z7SJs99vDvuPv/fTXCjbHOZF4kMJKDMJGccWGHQH9ZukHK4J/7Vn4xMHrRvAkFmf0/1gDI7MHrZbj/l4U5LhZm9YS2cYSys8V+Ypn8Eajlz60NfFapabJ29UbOml8+6hhx2kywic4APwKWAusADMMYAJaapvlwlNuFYRi3AIuBM4EzgCzgadM07xxnm/cCDwDnA6lAFfC/wE9N05zkXVcREXGavGRI7O8LadsF5fn8fenCMYOnJ80qZcmSc/jEZTOH3R6JRS5G8nYc7/rF65jjddYMg/OuWkzxzOGlI2IRnPLWX62obuJDj7w29PcHWn91IsF2VB9+bt24HWgrawba8YNAKOrqm2lubsXdakbsvT5ZaGG70NlhZoLthBg4SusL7fpoJ3YIZNpBJAdhnDjj4uMfuxbwBH1rjlVRXrkr8I3POTWsDOCoDYCEmmUdhexssY9YJH8EqqK6KagZfAD79h+mrr4Z5hUNu91p56WJrlzdQIrP7wYhzF4I0QN4AsztQC0wb7wHG4ZxA/BHPG1+FmgGrgd+CFwI3BrNxoqIiMTCyA+dP6lOZXWgS/X6yE5PHjN4euUtV1A8NW/Y4yO1yMVYPnn5AhoSM/jRCxtoqBndaSsqL2XBhWePCjJDbINTC6bnM2tq9rDfwxVKR3VVxSFbBz7t9EEgWCu21vDoU2uHStLUD94eqff6ZKCF7cJjh5kJthJi4Ojfr31PhBsiVrqkyMXK+gHquj0L9Y2UmQBfOHniTH+nz7h464Mf5Nb/+nZgi+i5DLj7lqBfIyYDIINZ1tHOzpb4E+3kj0Ct2BbarNItu2qoPbdw2PXcaeelic4Q+4ErDcP4iWma3n52rBb6+yKeAHMVnszmlWM90DCMbOBxYAC42DTNDYO3Pwi8DtxiGMaHTdN8JuqtFhERiaFILL42Mng6MsgcyCIXVyx9kcc/eymfuGxBSO0B+PdFM9ieNo2WxmbefPHvmKYbw3DxvhuvGKrJ7E+8B6dC7aiu2FZj20Az2OeDQDBi9V53Oi1sFxlWzkywlSADR97A1EfXHuEjC2s0MBQBvqUSRtrbbg6VTIi2rCSDrCToGjDZ235iX8/OhLQEI6D+gNNnXFSftpC/330XV/3ysfEf6DLgoXvsG5idUw5nz8cIpj57mNnZYn++68QANLaemPHytV+viWmZs9YgFlb31djex1e29TMvy+DmwRlKTjsvTRRo/iWe8hmHDWPo4rHUMIylE2xnmqYZ1jCXaZpDgWWf1x7LLUAh8GtvkHnwOboNw3gAWAHcCyjQLCIijhLtxdcCrR3sNk3u+vnrzCjMDvlD/VBwijwyc08EvscLMjshOBVqRzXU7WItGlng0RDL9/pkoIXtAmeXIF64RgYARgo5ADAYcAp0QUADeCM1h19Xd/BbDQxF3LHeE0e1vwWtosk3k/aeDSdKozy0ICng5wh1xsW6ZjfbW8y4KGey/dJLuOqtlVB9BFraRz8gJxMe+ZJ9g8xen74V96cfjmp2tsSvQ00ddPd6UnpTk2Mfcc1OD61uzMGde2g6XM8qPAHXGRkGGQmws9Vk8a3XBPVcdk22GfcsaZrmTwzDaACuBUqBS4Bq4ED0mxaUSwe//9XPfauBTuC9hmGkmKYZYlV5ERERewpn8TVvYGDdnvqh2/Y+9crQz39J6Au4drA7gNrBvvwFJdr6YU+bSdORBvJLivxsdYJTglOhdlRD3U78u+upwOtkB/ten4wm08J2vueydXvqOd7RM/Tz9d96KagA67Fek/puz89JLugcOLHYljcgHQ+BrkNNHUM/T8vPCP8JP30r3PMwuCc+RgeAh3M9x6YGhiLD9z3nOzBy/7zEcQdKoqW206THPfz3QAedQ5lxkeLyZEzb0SO7+ofOEcPOFw8sB+D+xCOeBTU7ujxlaBadFj9Zv4tO4+9338UVjz0+frDZ7tnZEjG+19Prv/USh5o9AynnzZ0a80WbQ51VmpKeCkBXu+c6uasdTi/JJDPIS7udk20m/FMGy008A2AYhht40jTN5dFuWJBOGfy+e+Qdpmn2G4axH1gAnAQEMfdCRETE/iK1+Fp3r+fDYktrBznZGfT09LJ6/5Gg2hJq7WDfoMT0rAyaAtgm3oNTXpEofyLhqahuGqrJHCi718m2g8m4sN20vMyhDKtAjQziNQ9mjM7ONIZlj9rdyACAv9tDtug0ePDT8PAvwW2Oqtvq/X0AuKtgLq+nTRm6TwNDzrGjxc0Lte5RQeKR09AnEsyMC4DytPgY4Eky/CzON6c8fgLLfmy/9BJOW/E6eXV1pLXHcXa2OE4os0qLyktZcscNAKz6w6tDty+59RpuctBMsISlS+UDVwwAACAASURBVJcG/OBly5YBvLF06dKD0WrQGK87E/gYsH3p0qUv+Ln/fiAfeGTp0qXNfu6/AygHnl66dGmtv9cwDOPuZcuW/fLaa68tBTj55JOprKykoKCAzZs3U1payoYNG8jPz6eiooKUlBQOHTpEZ2cnHR0dHD58mOTkZHbu3ElhYSGbNm0atk1lZSVJSUkcPnyYjo4Ourq6qK2tJSUlhR07djB16tRh20yZMoV3332XhIQE6urqaGtro7u7m5qaGtLS0ti2bRvFxcVs3LhxaJvc3Fx2796NYRg0NDTQ2tpKb28vBw8eJD09nW3btlFaWsr69euZNm0aGzZsICcnh6qqKkzT5OjRoxw7doz+/n4OHDhAZmYmW7duHXp+7+tlZ2ezd+9e3G43zc3NNDU14Xa72bdvH1lZWcO2mTp1Kps3byYrK4v9+/fT39/PsWPHOHr0KABVVVXk5uayZcuWoW2KiorYunUrGRkZHDhwgL6+PlpaWmhoaMAwDHbv3s2UKVOGbVNYWMi2bdtIT0+nurqanp4eWltbqa+vJyEhgV27dpGfn8+mTZuG/vaCggJ27NhBSkoKtbW1dHd3097ezpEjR0hMTBza/95t1q9fT0FBATt37hzalyP3f0VFBUVFRcP2ZV5eHpWVlSQmJg7ty5H7f+S+9O5/l8tFfX09ra2t9PT0UF1dPbT/S0pKxtz/jY2NHD9+nL6+vlH7f8OGDUPfs7OzR+3/gYEB9u3bR2ZmJlu2bBn6fxUXF7Np06Zx9392dvbQflm/fj3FxcVs2bKFjIwMv/t/z549Y+7/9PR0Dh48SG9v76j9n5eXN+y4LCwsZPv27aSmplJdXU13dzdtbW3U1dUN2/++2/juf++x3N7ezuHDh0lKSqKystLvsey7/32PZd/9v3HjxqH/W15eHrt27RrzWN6+ffuw/b9+/Xq/+997LHv3/8h96T2WgWH739+x7H3vePe/2+2mqalp2P4feSx797/vsey7//fu3UtOTs6wfTl16tSh/e97LDc2Ng7b/5s3bx46xoqKisY8ll0u16j9793Gu/9ramrGPJadfi7Pcrdw8enTqT7axqHmrlHXucULpnH/5WV86trzhvb/lQ89j2lCX18vR4510tE7QP+Am4EBN7lJA8w7fR7v7q4O7qINpA20c+lZsyc8l3/mhvOZn93Ff3z4En67YgsFuZnQ381Dn7qM1dsOYuAmweznPdOSMY0EpjbtpiMtj/L6rbRmlTC9bjNzSwtjfi7/3ep36e/vx+12c25pQtjn8uqqnbzbNMDBxraA/8fvmZ7NAx9+77jn8h179mGmZpJUvY3z55bF/Fx+8dd+x7Nrqnjib1vYtL+J2qZ2ao628frWgzz+2ibuWDzPNufyZ97czcqKuqDf6zkJvSw6uTigc/m7DS0Y/b0kNNVwWqH/67ITz+Xd9TWcn9PHmcntNDbWk5HkYtrxPXzxPYVMq9/GmSdNi/tz+cLcHr5022JOyWhn2Seu5NW1lcwpzSM3xeDpLywO+FzemZzN5l17SDJMsrqauCqrjauK3MzuquaG2dlkHtrOraefOJbt2i9v6E7g53/ZTFN7D00t7SwsSSPZ7A2/X/7+c+n6+1u4BkwSevuGHYsGcNSVyA1TT+VPGYWjjtWDjW3Mz+lhTvlU9cvD7JfvrDtOQmsjpuEio3EPe8kl6eBW3FNKSDmwKWrn8p3k8+K/NrI3uYTpRzbRkZZH8dFdDLiSyGmvo7W9nbVHOulvOkRpVsq4/fKSjESyO+o5dKyVpP4eprTV0peYSmljBa0ZU5let5mWLM/rJGTmknV0DwWpCXQ0NdiuX17SdpCmrn4K+o5R1N/MvbNdzGzfz9Wzp5BRuy1i53Kr+uW7uxI5+e01tE4tITEniZSERHoTXRjZmXRML4LpxRy4/GxL++UHkotJ2r8R95RSUg5sUowlRv3yJ/62hYa2XtxuNwUZiVyxoCDm5/L+tqP8dcfRwILDBrz38guY13OA6u4UDm7YRFdPH3S0klhQwLmuek7OdFPT0ERG1zEM003B8f10p2RRVr9t6JzUllHETX07eE+J//0/3rm8pKRkWTD93FAZZoBTBK1kGMbFeBYDfNo0zTv93L8bmAvMNU2zys/9bwEXABeYpvnOeK/16KOPmgD33ntvBFouIiISe97F1xq6BnC5XNxxx5V8f/HoMhS+GWdVR1rY39AKJpguF2XTCpg7p5y//WNd0K//8L+dzwM+5TkCcf23XmLdnnr6TUgtLKLpSMPQfd4SGiPrlhnAf86PfVZzxDP18NQHDrT8iWEY/GPpjRNm542c3hxrvv8nf9PoYz3FcTzfem4dD/5u3C6iX8G8163eH3YwWf4HwZ4jxsrSBILK0rTaiq01LH9urd/srosWlPLQbYvCzyr+/HfZX9/CgQMNzOnvJgGTAQyqElNpdyXygeKxazH/+FMX8YXrzgzv9Se52k6TH+/px216qhXcNzeR31efyOCP1nG9o8UdVKZfoH2DHS3uMWdc+GPX49HJ59Z4+NvioY1O5P3sANaUzvCaaCFpAAyD865aTEZOFjvWbKCxZvSM0YVzS/nRnYuYOnNaNGeCxaTWhlOOgpbB7zlj3J894nEiIiKO5V18zT1Yp694qv/F9F5+4ANDgYFdh4753DPAvv1HON7S4Xe7iYRTO7hnAFKB9KyJa3qaeKbk2+0DXygCLX9iGAa33HxxXEwBt1MgeSKqky1WWdngHreO9a42k+9UDnDXSXBxkX3PdRN90F5dcZgrIrEw30+/xtMhDgzFywKqdjReyYrMBChOhayk6MUvXqh1B1zmIpi+wcIcT8DGG0Bv74OWcUpOx8vxKOJ0FdVN7K9vpaO7D8MwaOuy7vz+ycsXMLMom4efW8eqikOj7i8qL2XBhWfT0dLGymf/AmNcJ3fsOXGdfPCyBX4H9uxak3kkpwSa3wXOAU4GNvreYRhGIjAL6Af2xb5pIiIi9jRRYKC5uTWk5w2kdrBv9kddfTPrq1sYSEomIdHFmRefT06h/+D4SJVtZlCLANnZRB3Vk2aVsmTJOcydU2ZB65xNdbLFCjtaxg8ye5nA4/sGKEix5wrzK7bWTJzNReQW5tPAUGxNNBjSPgBVHTA9LTozpWs7zaAW7oPg+wZl6QZpLjgUwLqGdj8eRZxsrJkzqyoOs/gbz0dm5kwIlpxRzpIzyqmobuKa77zGsZ4BDMPF+268gpzCPOoO1LLur6vGDDJ7jbxOFqWcOIfF02cdpwSaXwfuAK4Cfj/ivouAdGC1aZo9sW6YiIiIHQUaGAjW4gXTAl4cbU9VLX//53oOHhg+fezVJ56lsLyEhReeQ/HMiYOqO1rclKUnhNReu/HtqH7okdfoH3CTmODi2fuv5tXOsSZuSbhCWdAlkPe674DK3nZz1O2aYuscvuUyvFN5fW/3l+EfrSzNWFv+3NqAryWRWJhPA0OxE+hgCEB1l+fxkX6P7mhxh7xdMH2Duu7An9vOx6MTea+fvtdUX1ZcS3V9j72YzZwJw4Lp+ZxcnM3OVk8bvYkzO9ZsmDDI7OWEBWydcmZ8HjgKfNgwjHO8NxqGkQp8a/DXR61omIiIiB0FExgIlMsweDDAerVdu3fxqydeHhVk9mqsOcLKZ//C3q2VEz/XwIQPiTve8idzS3OZNTU74OD9I7v6h772tptDX97bZGwP3bYIlxFYtkgw73WvvOQTX+Js0/Iyh77GEk6Wpp1UVDcFNUADsKriEBXVTSG/pndgKBjBDILKCcEMhgD8ZPdAxK81oV7jg9muttOkPcjXsePxOBkc6zWHvuxC1/foC3bmzIqtNTFq2WipCZDpM87Q0tjstybzeMK9TlrNtsMshmHcCNw4+Gvx4PcLDMN4avDno6Zp/geAaZqthmHchSfg/IZhGM8AzcAHgFMGb382Vm0XERGxs1ACAxNxGQaPf/bSgEbfv7/qIN98cuXEI/umybq/riIjJ2vczOY0ZyQzR5w+8AQn0DrZwbzXldE0eQRbkzxWWZrRtmJbaB/mV2yrCSvw+9BtiwJeQDWUgSEJbTCkfQC6BiIbAAz1Gh/MdvF4PE6mjNrZmZ5B4PvnJdpm4T0n/X/jQaxnzoSrJM1gz+D5s+5gbUjPsWJbDcyJz9mMdj46zgQ+OuK2kwa/AA4C/+G9wzTNFw3DWAx8A/ggnrWEqoAvAT8xzQinbYmIiMSpUAMDc4pzqKobva7u4gXTePC28wLq0O1ocfO959cFPH0M06RizcZxA82aunqCPviEZ6I62cG810XGE4sszVgIdYG9cBfmi8bAkAwXavD1vLzIXpM91/jg2xJM3yDej0cNLIuThTNzxqqZLFmJMD3DwAD6evpCeo7Wzl5SItusmLHtpxHTNJcCS4PcZg1wTTTaIyIiYncja4R2DiZ9PPnUK6zONIYy7kL9gP/RS+dz0/mzR9UODqYT9z8bj9IQ5PSxhprDtDQ2+10gcH6WEZPFMUKpvxoJvnUJnZ6xZBfj1cnW1HuJlFhkacaClQvzaWAouuwSfC1LN5iXZQSVXR1o38B7LW3oCS0nzcrjUdd/mSysmjkTroJk+M/5CXx1ezLbQ9g+Oz2ZeF1kTmcnERERB/EGIxMzMkgZ/LDXMcBQnd775yWGFRjw1g72CqYDV9tp8vbO0DqLdQdrRwWaDeCmsthnM49XdzWalLEUW+G810UmEosszViwemE+DQxFj50GQ24uc/GdysAWJQylb5AVYlTEbsej09V2mjT0mLhNcBme32ORbCDWsmrmTCQszHHxvStn8Oorbwa97ZLTy3m1MwqNigEFmkVERBzi5Qc+MKx2ne9CKVOST3TErQoM7Ghxhzx9bOR2BnDXSQkx+5AXrYzlifjWJRQR54hmlmYseRfmC2ZaczQW5tPAUOTZaTBkYY6LT50Ej++bOF26PA1eO+LmtSPuCa+dvvcvr+iP++PRqdr6TOq64Svbhi80+ZVt/czLMri5zKWgv4NZOXMmEkK5Tp40q5RXO3PidjajjkYREREHuX9e4tDXt09PGvry3gYnOjzBiERgoGsAklKSQtrWd7vMBM9UtIuL1I0Rkfh1c5mLQMNUVs3gCMRDty3CZQT2l2hhvvjhHQwJRjSDr5cUuZiT4ekD+OMCpqZ4Ml1D4ZTj0WmaekyqOjwLTfqzq83kO5UDvNEQWk1xsT+rZ85EwkO3LcII8DppGAZLlpwDQJ/7xFc8sX8oXERERCLuodsWccXSFwNawTlSgYG0BCieMfaifuOZNrOMJANmZ0JaghE3mSu+Geb+xENWgkg4fI+BeM3MiRZvluav9o1fEiDWMziCpYX5nMd7fPa7A8/wjUXwdflpnkHn2k6TH+/pHyqhcN/cRH5ffSISGcp5xSnHo5PsaHFT3TXx40w82e4FKSpn4kR2mTkTjiVnlPPBmxfzxxdWYQZwnfzEZTMBuGfDiRmd8dRfip+WioiISMRYERhYmOMipzCPwvISGoNYEHDmrFLeP8fTWYynTtZIY5UyEZksVGd8tEuKXBSmwJ9q3VT6mbY/P8vgpjiYFq6F+ZwpK8kgP9mkaYJSp7EOvpalGxSlDL+ORqJ2r1OOR6d4oTbwNE4Tz37TvnEmKxJkghHIouHnnXsqU6Zks2LFBvbtHx00H3mdrO006fE5BOKpJnn8floTERGRsMQ6MOCdhlt34TmsfPYvEEBn0TAMLh+cPhaPvIHxkdlXt09PiJvOokg44nlwKFYW5ngCV/6yNOPpPGHFwnyBfLiX4I28duUleUoX9PqJ+1kZfI1G7V6nHI/xrrbTDKpmNkBlmxlXwTgJXKAJMoDlM2fGWzR87pwy5s4p45r0ljGvkzta3LxQ6x71/o+nmuTq+YmIiExisQ4M3Fzm4t22Ms67ajHr/rpq/GCzYbDsE5eQOie0cht2EGxnUUETkclrZJZmvAZLrFqYb7wP9xKcsa5dAMkGZCZCWqK1wdemHnPcsgre2r13nURIazo45XiMVztaQitKu6PFTVn6GIW8Ja55E2Rue+RVmtt7Rt2fl5nCyaW5fOKyBTFvW7D987Gukysb3OOW7wn3vBYrCjSLiIhIzAIDQzUQmU9GThYVazbSUDN6+lhReSlfueVcvrx4xoR1ju0q3M6igiYiIuPT4FvkTXTt6jWhuQ+mJ1oXfFXtXufrGmPxv2htJ/FhyRnlNP320zGdORMrO1rGP/d6xcN5TYFmERERiamhGohZ5RTPLKOlsZk3X/w7punGMFx85I4r+fTZBbbtPAUi1M6igiYiImKVQK9dANVdnsdbca1W7V5ne2RXPw09wZXN8EpTMvOkYNXMmWh6odYd0LkX7H9eU6BZREREJhTpkg7DayDmsSvf01k8KcPgp5cWRaDF1nJSZ1FERCaHYK5dYM21S7V7J4esECNV6ktJPHLaeU2BZhEREQlKJEo6+JbDaOuD/sG+1aEuk0d29cf1AmJO6yxONqqTLSKTUbxcu1S71/m8fcDlFf1BvSfnZxnqR4mt+H7e2dtuDrt9b7vJ7EzP+9Vp57X4/RQnIiIiMRPN4FpesmdxIadwWmdxMlOdbBGZLOLl2qXavZPHzWUuvlMZWCkXA7ipTNnMYl95yWPf57TzmgLNIiIiEnMjM5a3Zxtj3hdvnNZZnGyUsSwik1G8XLtCrcGr2r3xZ2gB6QnqhhvAXSclqGyG2M54n2lWZ5747OO081p8f5ITERERsRmndRZFRMT54uHa9ciufroGQlskTkHI+DS0gHStm0o/ZTTmZxncVObS/pW45nn/Bj+rxK7vewWaRURERCLIaZ1FEYmu8Wo4QvzM8lB98/gWL9eutASDzAST9iAyqYOp3euU49FJhi8g3Y/bBJcB981NVE1miStjXSfv/cHL7G4zOfumawJ+LjvXJNdZUkREZBJTYCDyytIN5mUZWsBGRII2Xg3HeKL65vEnHq5d3iDvjhZ3TGr3OuV4dIqydIOiFGPY7yLxauR1siTNwABH1CRXoFlEREQABQYiSQvYiEignJIhqYHJ+BfMtQusu3ZFs3avU45HEbGfia6TKxvcjqhJrrOoiIjIJKbAQHRoARsREYk3gV67AKanWVvySbV7RcRpnHJeU6BZREREJAqc0lkUEZHJY6JrVwKQ4oLjfdbXLVbtXhFxGiec1xRoFhEREYkSJ3QWRURkchl57Tra47l9VgZ0B7EIX6yodq+IOE08n9cUaBYRERGJsnjuLIqIyOTkvXYVpXh+V/1iERGZiOZqioiIiIiIiIiIiEhYNCQpIiIituOt+wiwt90cdbuyqkREREREROxFn9JERETE1vKSrW6BiIiIiIhY6fpvvTT087o99aNuf/mBD8S8TTKaAs0iIiJiCXUWRUREREQkWNPyMq1ugoxBgWYRERGxnDqLIiIiIiIyFiWhxAcFmkVERMQSk6GzqFrTIiISb+Ll2hUv7RQRCZQTzmv2b6GIiIiIA6jWtIiIxJt4uXbFSztFRAIVr+c1BZpFREREoiQesg5ERER8xcu1K17aKSISKCec11xWN0BERERERERERERE4lv8h8pFRERERERERMRxnFCzVmQy0REpIiIiIiIiIiK2Fq81a0UmEwWaRURERERERETEdpSxLBJfVKNZRERERERERERERMKiQLOIiIiIiIiIiIiIhEWBZhEREREREREREREJiwLNIiIiIiIiIiIiIhIWBZpFREREREREREREJCwKNIuIiIiIiIiIiIhIWBRoFhEREREREREREZGwOC7QbBjGAcMwzDG+6qxun4iIiIiIiIiIiIjTJFrdgChpAX7k5/b2WDdERERERERERERExOmcGmg+bprmUqsbISIiIiIiIiIiIjIZOK50hoiIiIiIiIiIiIjEllMzmlMMw7gTmA50ANuA1aZpDljbLBERERERERERERHncWqguRj4zYjb9huG8XHTNFdZ0SARERERERERERERp0pYunSp1W2IqGXLluUADwNfBb4JPAMkAEuADy9btuwvS5curR+5nWEYdy9btuyX1157bSnAySefTGVlJQUFBWzevJnS0lI2bNhAfn4+FRUVpKSkcOjQITo7O+no6ODw4cMkJyezc+dOCgsL2bRp07BtKisrSUpK4vDhw3R0dNDV1UVtbS0pKSns2LGDqVOnDttmypQpvPvuuyQkJFBXV0dbWxvd3d3U1NSQlpbGtm3bKC4uZuPGjUPb5Obmsnv3bgzDoKGhgdbWVnp7ezl48CDp6els27aN0tJS1q9fz7Rp09iwYQM5OTlUVVVhmiZHjx7l2LFj9Pf3c+DAATIzM9m6devQ83tfLzs7m7179+J2u2lubqapqQm3282+ffvIysoats3UqVPZvHkzWVlZ7N+/n/7+fo4dO8bRo0cBqKqqIjc3ly1btgxtU1RUxNatW8nIyODAgQP09fXR0tJCQ0MDhmGwe/dupkyZMmybwsJCtm3bRnp6OtXV1fT09NDa2kp9fT0JCQns2rWL/Px8Nm3aNPS3FxQUsGPHDlJSUqitraW7u5v29naOHDlCYmLi0P73brN+/XoKCgrYuXPn0L4cuf8rKiooKioati/z8vKorKwkMTFxaF+O3P8j96V3/7tcLurr62ltbaWnp4fq6uqh/V9SUjLm/m9sbOT48eP09fWN2v8bNmwY+p6dnT1q/w8MDLBv3z4yMzPZsmXL0P+ruLiYTZs2jbv/s7Ozh/bL+vXrKS4uZsuWLWRkZPjd/3v27Blz/6enp3Pw4EF6e3tH7f+8vLxhx2VhYSHbt28nNTWV6upquru7aWtro66ubtj+993Gd/97j+X29nYOHz5MUlISlZWVfo9l3/3veyz77v+NGzcO/d/y8vLYtWvXmMfy9u3bh+3/9evX+93/3mPZu/9H7kvvsQwM2//+jmXve8e7/91uN01NTcP2/8hj2bv/fY9l3/2/d+9ecnJyhu3LqVOnDu1/32O5sbFx2P7fvHnz0DFWVFQ05rHscrlG7X/vNt79X1NTM+axrHO5zuU6l+tcrnO5zuU6l+tcrnO5zuU6l+tcrnO5zuWT8VxeUlKyLOygawAM0zRj8TqWMwzjv4EvAy+apnnTWI979NFHTYB77703Vk0TERERERERERERiRYjFi8ymRYD/J/B7xdZ2goRERERERERERERh5lMgeaGwe8ZlrZCRERERERERERExGEmU6D5gsHv+yxthYiIiIiIiIiIiIjDOCrQbBjGAsMw8vzcPgP42eCvv41tq0REREREREREREScLdHqBkTYrcDXDMNYCewH2oDZwLVAKvAq8N/WNU9ERERERERERETEeZwWaF4JnAKchadURgZwHHgT+A3wG9M0TeuaJyIiIiIiIiIiIuI8jgo0m6a5ClhldTtEREREREREREREJhNH1WgWERERERERERERkdhToFlEREREREREREREwqJAs4iIiIiIiIiIiIiERYFmEREREREREREREQmLAs0iIiIiIiIiIiIiEhYFmkVEREREREREREQkLAo0i4iIiIiIiIiIiEhYFGgWERERERERERERkbAo0CwiIiIiIiIiIiIiYVGgWURERERERERERETCokCziIiIiIiIiIiIiIRFgWYRERERERERERERCYsCzSIiIiIiIiIiIiISFgWaRURERERERERERCQsCjSLiIiIiIiIiIiISFgUaBYRERERERERERGRsCjQLCIiIiIiIiIiIiJhUaBZRERERERERERERMKiQLOIiIiIiIiIiIiIhEWBZhEREREREREREREJiwLNIiIiIiIiIiIiIhIWBZpFREREREREREREJCwKNIuIiIiIiIiIiIhIWBRoFhEREREREREREZGwKNAsIiIiIiIiIiIiImFRoFlEREREREREREREwqJAs4iIiIiIiIiIiIiERYFmEREREREREREREQmLAs0iIiIiIiIiIiIiEhYFmkVEREREREREREQkLAo0i4iIiIiIiIiIiEhYFGgWERERERERERERkbAo0CwiIiIiIiIiIiIiYVGgWURERERERERERETCokCziIiIiIiIiIiIiIRFgWYRERERERERERERCYsCzSIiIiIiIiIiIiISFgWaRURERERERERERCQsCjSLiIiIiIiIiIiISFgcF2g2DKPMMIz/NQzjsGEYPYZhHDAM40eGYUyxum0iIiIiIiIiIiIiTpRodQMiyTCM2cBbQBHwZ2AXcB5wH3CVYRgXmqbZZGETRURERERERERERBzHaRnNv8ATZP6CaZo3mqb5NdM0LwV+CJwC/JelrRMRERERERERERFxIMcEmg3DOAm4AjgA/HzE3d8EOoB/NwwjI8ZNExEREREREREREXE0xwSagUsHv//dNE237x2mabYBa4B04PxYN0xERERERERERETEyZwUaD5l8PvuMe7fM/j95Bi0RURERERERERERGTScNJigDmD31vGuN97e66/Ow3DuBu4+xe/+AUAjz76aEQbJyIiIiIiIiIiImIB89577zWi/SJOymieiPefafq70zTNx0zTPCeG7RERERERERERERFxBCdlNHszlnPGuD97xOP8ikV0X0RERERERERERMRJnJTR/O7g97FqMM8d/D5WDWcRERERERERERERCYFhmn4rScQdwzBmA1XAAWC2aZpun/uygCN4AuuFpml2WNJIEREREREREREREQdyTEazaZp7gb8DM4HPjrh7GZAB/FpBZhEREREREREREZHIckxGMwxlNb8FFAF/BiqBRcAleEpmvNc0zSbrWigiIiIiIiIiIiLiPI4KNAMYhlEOLAeuAvLxlMx4EVhmmmazlW0TERERERERERERcSLHBZojISHBZbrd9vi/GIaBU/aRywCb/FvFRlyAe8JHBcdJx000GIYLnzL2YkeGAXoPB89wgdPe23ovhMgA9H8bTzSuv2Df/p7T+wZ2/b87J0hLkgAAIABJREFUTbSOm0hz+vs9HHHZD1ZfYHzx1P+zzb6cHP2keDlnx9BR0zQLo/0iidF+gXjkdpuYG6xuhYdxjsnq1autbkZEXHTRRZg/s7oVYjfG58A8/ZbIPue257n+3U9G9Dmd5OVTnuD2t3utboaM4/cXJHP2H/ZY3Yy4s/HWufB1h3Wav21w+4d2Wt2KuPP7Z0+FG35sdTNszf3n+zAvPDPiz2us2WKbfrQvJ/Wp/VE/Ozai0W+NBvWFxxaP/WD1C8cXV/0/m/TrJks/KVp9nXhlrNlyMBav45jFAEVERERERERERETEGgo0i4iIiIiIiIiIiEhYVDojwnp64TPfhX+ug+ZWmFMG3/4sXH2h1S3z7+GHH2bjxo10d3eTl5fHv/3bv3HddddZ3SzH29MAp30bbjkLfvtRq1tjbxfvfYN3OptJNAwApiWm8e68qyxuVeAOvbKX3T/bTNeRDlIK0jjzuxeRf06x1c1yhD9cOmXY7wM9Xcy5+R7O+fKPLGqRs/U01FL9q6V07N6MkZjMlPOvovzj38BImGRdiaOV8LfPQt1GSC+ESx+BU26yulVRtXvP0+zf/yLHW3YzY/q1nL/o20P31dW/zYaN36Kz8wj5+adz/nn/RUbGNAtba7F9q6F6HbQdhmlnw3vu8Nzu7ocNv4bjNdDVDBd+DgrmWttWP352pJGn6pvZ3tnN7YW5PDV3BgDvtHXw4MEjbOzoIgG4OCeTn5xURklyUmRe91l46i+wvQpuvxKeWuq5fec++Mg3YW+t5/ez58NP/gNOPSkiLxsRf/zjH/nrX//Kvn37WLJkCV//+tdHPebJJ5/kySef5Ac/+AHnnHOOBa201s9WwVPvwPYjcPvZ8NS/e24/0ASzvgkZySce+9XL4cGrrWlntP3saBVPHTvI9u4Wbs8t56nyc4fu63T38x9HtvHc8Vr6TDdnpOWyevbF1jU2BPt/u5OaF/bQtruZ0utmc9Z3LwKg9qUqtn1zzdDjTLeJu3uA9//xBnIXFljV3JDt/sMv2P/qrzm+dwczLv8Q5z/4xNB91f/8A9t/tZzOxkOkF5Vxxj0PU7b4BgtbG56G135D0xsv0FX9LnkXXsfMz31v6L6jK56j7k+/pO/4UTLnnc2Mz3yH5LypFrY2RP098LfPwP5/QnczTJkDF38bZg+eiPav8PT7WquhdBFc/xTkzIh6swYGetmwcTl19W/T29tCZuZ0zjj9/1BachEDA728/c5XaG7eQUfnYS695CmmFp0X9TYFZKAftj0HjbuhtxMyC2D+dTD1VGg+ALtegeO1nlrVBXPhtJshNcfqVkdMj9vNZ/bW8s+WNpr7BpiTlsK3Z5Rw9ZRsdnZ285HdB9nb7Sndc3ZmGj85qYxT01MtbnXkTbJPh9HXPwDlU2HVYzC9GF5dA7f9J2x/BmaWWt260e68806++tWvkpyczMGDB7nvvvuYO3cup5xyitVNc7TPPgfnRv/65Bg/Kz2LT+XPsroZQWtcc4jK/17P2T+8lNzTC+lu7LS6SY5y6+vHhn7u7+rgT9eWMf3SD1rYImer/tVSkrLzOP2xtxjoaGX3wx+j8W9PU3TNJBotc/fD8zfAWffA7f+A6lXwh+vhE5sh/2SrWxc1aWlFLDj10xypW8PAQM/Q7T09x3hzzX2cd+5yppVewrbtP2HNW1/misufsbC1FkvNgVOugIZdMNA3/L78k2D2Ylj/lCVNC0RpchIPlE/lb8fb6HKfWD7nWP8AdxcXcGVuFomGwef21fLxPdX8dcHsyLxuITzwSfjb29DVM/z25/8vzCgBtxt+/gf48Ndhm43eYgUFBXzkIx9h3bp19PT0jLr/0KFDvPHGG+Tn51vQOnsozYEHroK/VUJX3+j7jz8CiQmxb1eslSal8UDRPP7WVk+XOTDsvrtrN9Fvuqk85UryEpLZ0nXcolaGLrUonbmfOYPGfx1ioOfE31f2gTmUfWDO0O81L+xm9y+2kLMgPo+JtMISFnzsPzmy9h8M9HQN3d7ZcIi3l32M93/vj5ScfyWH33qNNd+4nQ+8sIfUvCILWxy6pLwiSj74GVq2/Auzt3vo9raKtRz63fc5eelvSSmeQc2T32L/j77IKct/Z2FrQ+Tuh6xyuHMV5EyHqlfhT7fBp7ZDcia8cDNc8yuYez2sehD+9CH42DtRb5Zp9pOeXsySS39NRnoJh4+sZs1bX+Lqq/5MWmohBQXv4ZST/5033/pi1NsSFHMA0qbA+z7v+V6/09PvufSr0NcJM94L5873LOC4/XnY/Du44F6rWx0x/aZJeUoSqxbOYXpKMq8ea+W2XQfYftYplCYn8vy8mcxIScYN/PzIUT787gG2nTXP6mZHnEpnRFhGGiz9tCeo7HLBde+HWaWwsdLqlvk3a9YskpM9aQTGYMbooUOHrGyS4z2zAXLTYIlzYxIy6N2fbuLkz5zFlDOLMFwGaVMzSJuaYXWzHKn69T+SMqWIwjPfZ3VTHKu3oYYp770GV3IKSVMKyTnz/XTVTLKFaY7u8mSqnvdFcCXAzEuh7ELY8RurWxZV5WWXU1Z2GSkpucNur6n9BznZc5hefhUJCSmctvCzHG95l9bWfRa11AZKz4CS0yF5xLnelQizL4b82Z4sHpu6OT+XG/NzyU8cnoty9ZRsbi3IJTsxgfQEF58rKWBNa0fkXvdSuPFiyB+R1JSb5elTGwaYJiS4oKomYi8bEYsXL+b9738/2dnZfu//0Y9+xD333ENSUmSyv+PRzWfCjWdA/iTvAt2cM40bc6aRn5g87PZ3u9t4qfUwj5WdTWFiCgmGwdnpU8Z4FvsquWImJZfNJDk3ZdzH1fypirIb5wx99ow35RffRNniG0jJyRt2e2fjIZKycim94CoMw2DahdeQmJZB+6H4vSZOWXQlueddTmLW8Ot/y8aVTLngatLK5+JKSqbkls/SXrmenrqYrDMWWckZcNFSyJ3pCX7OvQ5yZnlmrr37AhQsgPm3QmIqvH8pNGz19AejLDExndMWfo7MjGkYhotppReTkVHGseYKEhKSmXfKRygsPBvDsNkoXWIKzLsa0vM9/8/ihZCR55nRNfVUmHYWJKVCYjLMej807be6xRGVkZDA0uklzExNwWUYXJeXw6zUZDa2d5GbmMjM1BQMw8AEEgyo6h49QO0ECjRHWX0T7K6GCCV8RMUPfvADLr/8cu68807y8/M5//zzrW6SY7V2wUOvwPdvtrol8eU/67ZTUPESF1at5I32BqubExBzwM3xHUfpOdbNisuf4x8X/Z7ty99ioLvf6qY50v7Xfsusq++I2w8t8aDomo/SvOYV3D1d9DbV0bJlNdlnXmR1s2LMz4rmpgmNO2LfFBtoaakiN/fEDKjExHQyM8ppaamysFUSC6tb21kQw6meuRdD6oXw+Ufg6x+P2cuGbeXKlSQmJnLBBRdY3RRbm/EQlD0AH/8NHG23ujWxt7armRnJ6XyzvoKCipc4bfff+WNLrdXNiorOQ200baij/Ab7lQ4KV968s8meMY/af72Me2CA2lV/xpWUQu6c06xuWsSZpunp/5y4AcAZCQjt9dC82xNgbqyAojNO3JecAbmz4WhFzJvV1X2UtrYD5OTMmfjBdtLdCu2NkFUy+r6mvZDt7JKS9b197O7qGdZnyn1nG6lvbeXz+w7x9bI4LDcTAJXOiKK+frjjQfjotTBvptWtGduXvvQl7rvvPioqKti8efNQhrNE3oOvwCffC+Xxl6Rgmf9bchqnpmSTbLh45ngN1x94iy1zL2N2SqbVTRtXz9EuzD43R/56gAufvg4j0cX6z/yD3Y9uYf4XJ199xmjqqKumcfNqFn39l1Y3xdEyTz2Po/98js0fOQvcA+Qvvonc8y63ulmxlT8PMorgnUc8Wc0HV3rKZ8y4xOqWWaK/v5OUlOEZXUlJWfT1Ry7TVexnW0cXy2vq+fO82JW0Ov4GdHTB//uLp4xGPOjs7OSxxx7j+9//vtVNsa2CTFh/P5xZBk0dntJydzwFf/uc1S2Lrdq+TnZ0t/LB7Gkcnn8db3c2ce2BNzk1JZv5qf4z5eNV7YtV5J8zlfTyLKubEnGuhARmXX0nb3/zIwz0duNKTObC//o9iWnOS+XPec9i9v3w/1Bwxe2kFs/kyPM/A8PA7VNKJC4N9MFLd8BpH4WCedDX7lmPw1dqDvS2xbRZbncfb7/zFWbNvIHsbBstUjAR9wD/v707j4+quvs4/vlNkskOhJCAyB4EWQUEF9zQ1q0uVWttq1KtS0sX1Fqf1lp91Fa7atVWbe1TW6q0tmor1LphUauCS2VxAVkEgoAsYQ3ZMzPn+ePehEmYyUImmSR836/XeV3uuefMPcP8cufOufeew6JHYeBRkNuoQ3XPJlj5Ahx9VXLa1gFqI45LVq3nssLeHB7V0bz7mPGUh8P8adsuBqd3z6eddEdzO4lEYPotEEyF+7+X7NY0LyUlhfHjx1NSUsKcOXOS3ZxuaelG+PcK+PbB2R9xwI7Oyic3JY30QAqX9R7CcVn5PLt3S7Kb1axAhncdb+j00WQUZpHeO4NhXxnLtv90smd+u4F1z82mz/jjyOnf9cbx7ipcJMLqO66g19GnMXH2uxzxh7cJlZeyafbPm6/cnaSkwYVz4KNn4Ff94K27YdRF0GNAsluWFKmpWdTWNrz9sDZURlpq9/tRLZ6PKqs5c/la7ht6KCf07NgLvtmZMONz3uSA23Z26K4PyB/+8AdOP/10+vfvhJO0dBI56TB5sDc+c98ecP9FMG+F9wTgwSTTUkgz4+a+owgGApyUU8DJ2YXMK9ua7KYl3Ma5HzHgvO53NzPAlrfns/SB73PKA//mC6+W86kH5/P2T2awa9XSZDct4XqMm0r/i65h7V3f4v1vnESwcACBjGzS8rvw3akuAv+cDilBOP1+Ly8tB6pLG5arLoVgx10ocS7CG2/eSCCQxuQjb+6w/baZi8DiR72h5sZf2HBbWQm8+ZA3EWB+J370vw0izjF99XqCAeP+Yfv/TshOSWFGv3y+vPpjttXEmLygi9Mdze3AObjyR7B1Jzx7H6R1of/lcDisMZrbySuroXgnDLrFWy+rhrCD5Zth8Y3JbVtXYgYu1uPrnUywZzoZ/bJBIzm0u+LnZjNq+neT3YxuLVy2m9odmyk8YzqBtHQCaen0OflzbPrrPQyY3gWupiZS4XiY/p9963+a6t35chDq2XM469bNrV8PhSooK9vQ9R7rlBZZX1XDp5d9xC0D+jK9sHfzFdpBJAIVVbBpGySpCS22ePHiBjdw7N69m1tvvZWLL76YSy65JMmt65zqTpk6/1leYo3P6Nl8oW5g56KtVG2roP/pQ5LdlHaxa/W7FE44nvxRRwKQP3oy+WOmsOW/L5E3YkKSW5d4hWdcSuEZlwJQ9ck6tvz9QTIHdtFJiJyDZ66E8q3whWe9GwsACsbA+3/aV66mHHat8YbV6JBmOd56+2aqqndw0gm/JRDoIne/OgdLHoPqvXDM17zO5joVO2HhgzDiNBg4JXltbEfOOa78aANba0I8O3oYaYHYHQIRoCISYVNNLYXBLvLZtpDuaG4HX/8JfLgOnr4HMjtu+LpW27VrF/Pnz6eiooJwOMzbb7/N/PnzmTRpUrKb1i199ThYcxss/b6XZhwPZ42BF76Z7JZ1XrvDNbywdwtVkTAhF+HPuz7m1bLtnJ7TNa6WD7zgMNY9upzqHZXU7Klm3Z+W0XfaoGQ3q1spee8NKko+YdApn0t2U7q11B69CRYOoGTeX3DhEKHyUna88hRZg7vfLMnN2vYehKq8mbPfvAvKNsP4y5PdqnYViYQIh6uJuDDOhb1/R0IMOPTT7CldzYYN8wiHq/lg2W/o1XNE13qsM9EiYe/RWxfxUrjWywMIh7z1BuU6V5dayDmqIhHCzhF2UBWJEHKOTdU1nLLsI77Zr4AZh/RJ/H5DUFUN4QiEw96/QyF48U1YssLLKy2D6++BvFwY1YkeYAmFQlRXVxOJRIhEIlRXVxMKhbjnnnuYNWsWDz/8MA8//DD5+fnccMMNnH/++clucocLhaGq1v98I96/Q2F4qxhWbvUuIOwog2uehGmHQc/MZLe4fYRchKpI2P/7cvXntyfmFDAoLYufbFtByEVYUL6dV8pLOD2na43dGQlFCFeHcBGHC3v/joQi9ds3zFnNIacNITWnaw/TGAmFCFdXEQmHcZGw9+9QiPxRk9n27oL6O5h3rlxCydIFXXqMZhcOEamphkgEF4kQqamuz6v8eBXOOWpKPmH9QzdT+JnLSM3pohdNnv86bP8QLnoa0qIOQCPO9+bhWPF379zv9R96Nxz06Zjz33cW3U5p6VpOPP4BUlMbdiyFwzWEw95EcpFILeFwtTd2dmfw3uPeWNdHf9W7Q7xO5W5YcD8MPd5L3dTX12zkw4oqnh49lMyUfV2uL+7ey5KyCsLOURoKc/26TeSlpDCqA+e86Chd6F7brmH9ZnjoH5AehH6n78t/6Ca45MzktSsWM2POnDncfffdRCIR+vbty8yZMznhhBOS3bRuKSvopTo56ZCRBgXdb4iyhKl1jpu3LGNF9V5SMA7PyGXOkKmMzOga/2kjvjGRml1VvHT6k6Skp9D/zKEc9vUjmq8oLbbu2UcZOO080rK7Rkx0ZUU3PMCGWXeyZe7vIJBC7pijGXD5D5LdrI73/qPw7u+9TsKBJ8CXXvRm2O7Gli3/LR8se7B+vXj904wd8w3Gjf0Wx0+9l3cW38kbb32P/N7jmTr1IB+PdtU8WPn8vvWN78DIM7wZ2OffCZX+mA9v/MZbnvq/3szsncQdG7Zw+4Z9j+vPLtnFrQP7Yhhrq2q4fcMWbt+wb/iqsmPHJ2a/D8Pt/7dvffZzcOvV3mTaM38BG7dBZjpMGQ3P/xoyOtGf3COPPMKsWbPq1+fNm8fll1/OFVdc0aBcSkoKubm5ZGVldXALk++O5+H25/atz/4v3HomjOwLNz0N2/ZCjww49XB4rAtN9thad2z9kNu3fVi/Pnv3x9xaOIrb+o1h7pCpXLVxET/dtpLBwSweGTiFw7vY+Myrf7OUVfcvqV/f9M81jPjWREbOnES4OsQnz61j8q8/lcQWJsayWT/mg4fvqF8vfv4vjL3yZsZd9b+Mu/IWXr/pS1Tt2kp6rwJGX/Y9Djm6685nsfnvD7L5iV/Xr+98bS6HfH4mhWddzrr7rqd668cEMrLpc/Ln6P+F65LY0jbYsx6WPAQp6XBf1M1MZz4EYy+BC/7uDRz/z0uh/9Fw3l87pFnl5Zv4aM3jBAJB5vxz3+TbU468jSFDzuGZZz9DecUnALzyn6sBOOfsF8nJPrRD2hdXxU4oXgiBVHg+aqiPI74A5SVQscM7T4o+Vzr7Fx3fznayvqqGh7buIN2Mfm/vmzTyoaIBBAMBZq7dyMbqWjIDxpTcLJ4fU0RGoPvd/2ud5qpHJ2Jmzr2T7FZ4bDK8+uqryW5GQpx44om4+5PdCuls7FvgGo/b1NbXfO9Jzll5ZUJfszt5euTDfOmNmmQ3Q5rw2LFBjnyiG8zc3cEWff4wuKmbndf82PjSF5YnuxVdzmN/Gw2fvS/Zzejc5l6LOy7xj3PbgqV0lvPoaN3pnDoWnWd3jPY4b20POheOryueB+u8sGld6vyvk5zXHTTnSe10rtNV2YKli5xzk9t7P92v61xEREREREREREREOpQ6mkVERERERERERESkTdTRLCIiIiIiIiIiIiJtoo5mEREREREREREREWkTdTSLiIiIiIiIiIiISJuYc11kds4OZGYlwPpkt0NERERERERERESkjQY75wraeyfqaBYRERERERERERGRNtHQGSIiIiIiIiIiIiLSJupoFhEREREREREREZE2UUeziIiIiIiIiIiIiLSJOppFREREREREREREpE3U0SwiIiIiIiIiIiIibZKa7AZ0RqmpqS4QUB+8iIh0XpFQGNN3VbtwOHQeIAerSCiCmSW7GSJJ56xjvwtCoTCBgP72RESkfYTD4e3OuYL23o86mmMIBAKUlpYmuxkiIiJx5Wf34tVRP092M7qlacU/oKSkJNnNEEmKvKx8Huy7KNnNEEm6a0qnULK9474LsrN7Mm7cjR22PxERObgsXfrD9R2xH92uIyIiIiIiIiIiIiJtoo5mEREREREREREREWkTDZ0hIiIicZ2z8nZ2hvYSMCPTghyXO5rJ2SOYtf1Fnjjspvpy31j3ACWhPfvlHZUzAgcsKV/Dr4bMqN92/qofMTBYsF/ejMKzOL3XJCZ/cC0ZFsQMcgKZnNpzItf2+ywppmvkItJ1/WzHl9lQu5J7+r5GmgUBeHj393mz8hlSLQ2APin9OSL9ZD6TczVZgVwAXq94ij/uuZmgZQCQG8hjZPAozsq5mn6pQ5PzZkREDsCyZfcRCu1lzJjrSU3Nqs9fseIhqqq2kptbRHn5xwBEImHMwCwFgLy88QwceFZS2i0iLaOOZhEREWnSLwdfzdE5I9lWu5tvFf+GVEthXfVWdoXKyEvNIeTCrK76hIxAsEHe+5XFfK3vZ3DOMavk34RdhBQLsL22lJCLsLJqY4O8DTXbmZRdVL/fx4Z/l4HpBRRXb+Vr637NoPQCLux9fBL/J0REDtz20CZW1Swiy3JZWvUSUzLPqN92Zs4VXJB7HbWumo21q3hi7138ZMfF3Jz/N9IDXkdMUdoEburzZyIuzPbwJl4o/yO3b7+QH+Q/xoC0Ecl6WyIirRYM5rFr1wcUFBwFQGXlVpwLATBgwFmkp/cCYP36uQSDuRxyyClJa6uItI5uCxIREZEWKUzrxXG5oykJ7eHQtHwWl38EwIrKjQzL6Mek7KIGeRHnGJ0xkDGZgwi5MKuqNgGwpGINk7OHMzhY2CBvQLAPBWk999vvkPS+TMgqYk3Vlg56pyIiibewcg5FaUdwXNZ5LKicE7NMmqUzNDiOmXkPUhbZzeuVT+1XJmApFKYOYnrPWxkZnMLcsgfau+kiIgmVlzeOnTvfq1/fufNd8vLGt6huKFTOmjV/5v337+L993/O2rV/pbZ2b/32HTsWs2zZfbz33k9ZvvxX7N69vH7b9u2L+PDDB/x6f6GmpjRxb0pEAHU0i4iISAttqdnFgr3LGZkxgEnZRSypWAN4ncQTs4qYkDWsQd64rCGkBVJJC6QyNmtwfSf04vI1TMgu4ojsYQ3yJmYVxdzv2qotLK1Yw8iMQzvgXYqItI+Flf/kmMyzOSbzHJZVL2BPeHvcspmBbEanT2VVzTtNvuaRGaeyumZRopsqItKusrMHEIlUU1VVgnMRdu9eTu/e41pU1zlHnz6TGTPmWkaPvgaATZvmARAKVbJp04sMHz6d8eNvZPjwy8nIKABg164PKCl5i6FDv8jYsd8hM7MfH3+8/8U8EWkbdTSLiIhIk274+PdMW34jV627j0nZRXyl4FQmZQ9ncbnfqVy+hgnZw5iYXdQgL3oYjElZw1nib1tasYaJWcOYmFXUIO/I7OEN9nvpml9w8vIb+fb63/HZvGM5N+/ojni7IiIJt6pmETvCnzAl80yGpI2hIGUgb1X+q8k6vQKFlEf2tLmMiEhnVHdX8969a0lPzyctrUeL6qWl5dCz50gCgTRSUjLo2/d4ysrWNyhTVVVCJBIiGOxR39G8Y8ci+vY9gYyMfMxS6NfvJMrKNlBbW5bw9yZyMNMYzSIiItKkuwZdxdE5IxvkTcwu4kebHqM0XMEHFcXcMeDLZKWksyNUSmm4gncr1nJx/rT68pOyi3hi52uUhivYFSpjUHohvVN7cNvG2ZSGK1hTtZmJ2Q3vaJ5d9D8MTC/oiLcoItKuFlbMYUz6VHIDeQAck3kWCyrnclrO5XHr7I5sJTuw/3BC0Xa1oIyISGfUu/d4Vq/+EzU1u+nd+4gW1wuHq9m06QX27l1LOFwFQCTije+cmprJ4MHnU1LyJuvXzyEnZzCHHnoa6em9qanZw8aNz7Bx47P1r2UWoLa2lLS0nMS+OZGDmDqaRUREpNUGBPtQkNqTf+xcSL9gHlkp6QCMyxzCP3YupCJSzdiswfXlx2UNoSxSxT92LuSIrGEA5KRk0CfNe42C1J4cGsxPynsREWlPNa6K/1Y9T4QI1209AYCQq6HClfJx7YqYdaoi5SyvfoOzc2Y0+dqLq/7NYcEjE95mEZH2Fgz2IhjsRWnpagYOPKfF9bZtW0BtbSkjRlxFWloO5eUbWb36j/Xbe/YcQc+eI4hEavnkk3+zYcMzDB8+nbS0HvTvfyq9eo1qj7cjIj51NIuIiMgBmZA9jL9sf5kzek1ukPdIyXxGZQ4iIxCsz88IBBmdMZC/bH+ZKwpP21c+y3uNoxrdMS0i0l0srppPgBR+WDCXVNLq83+z+9ssrJzboGytq2FTaDVPlN5FVqAnx2eev9/rRVyYHeHNzCufxcqa//KD/Mfa/T2IiLSHQYPOIRyuIiUliHORFtUJh2sw84bNCIUq2Lr1tfptNTWlVFZuITd3KGYpBAJpmHkjxvbpM5mtW18jI6OAjIw+hEKVlJUVq+NZJMHU0SwiIiIHZFL2cF7Ys5gJ/h3KABOyirg3PJdzs46JWf69yuIG5SdmF/H4ztf2GzZDRKS7WFg5h+Oyzic/pX+D/FOyLuEvpT9mdPBYniv7Ay+WP4rD0SelP0ekT+OsnK+SHsiqL7+mdilf33IkDkduoBcjg0dxS/7j9E/T8VNEuqb09N6trlNYeCzr1/+DDz74BWlpPSgoOIrS0o/8rY5t2xawfv0/ACMr6xAGDjwLgLy8sUQitRQXP0FNzR5SUjLp0WO4OppFEsycc8luQ6cTDAZdaWlpspshIiISV352L14d/fNkN6Nbmlb8A0pKSpLdDJGkyMvK58F+i5LdDJGku6Z0CiXbO+67IDu7J+PH39hh+xMRkYOd57GCAAAPDUlEQVTL0qU/XOScm9x8ybYJtPcORERERERERERERKR7U0eziIiIiIiIiIiIiLSJOppFREREREREREREpE3U0SwiIiIiIiIiIiIibaKOZhERERERERERERFpE3POJbsNnY6Z7QVWJrsd0i30AbYnuxHSLSiWJBEUR5IoiiVJFMWSJIpiSRJFsSSJoliSzmSkcy63vXeS2t476KJWOucmJ7sR0vWZ2TuKJUkExZIkguJIEkWxJImiWJJEUSxJoiiWJFEUS9KZmNk7HbEfDZ0hIiIiIiIiIiIiIm2ijmYRERERERERERERaRN1NMf2u2Q3QLoNxZIkimJJEkFxJImiWJJEUSxJoiiWJFEUS5IoiiXpTDokHjUZoIiIiIiIiIiIiIi0ie5oFhEREREREREREZE2UUeziIiIiIiIiIiIiLSJOpp9ZjbAzP5gZp+YWbWZFZvZvWaWl+y2SXKY2YVm9msze83MSs3MmdnsZupMNbNnzWynmVWY2Xtmdp2ZpTRR52wze8XM9phZmZm9ZWaXJf4dSTKYWb6ZXWVmT5nZR2ZW6X/Wr5vZlWYW8zisWJJYzOxnZjbfzDb4sbTTzJaY2a1mlh+njmJJmmVm0/3vOWdmV8Up0+q4MLPLzOxtv/wev/7Z7fMupKP558suTtoSp46OSRKXmZ1gZn83s83+b7LNZjbPzD4To6xiSRows8ubOCbVpXCMeoolicnMzvKPQRv9c++1ZvaEmR0bp7xiSdqFea4wszfNbK8fX0vM7JrG8WVmaWZ2rZn90cyWmllNU+f4fp3jzOznZvZfMyvxv4PXmdnvzWx4q9qqMZrBzIqAhUAhMBdYARwFnAysBI5zzu1IXgslGcxsKXAEUAZsBA4H/uycuzRO+c8CfweqgL8BO4FzgJHAk865z8eo8y3g18AOv04NcCEwALjbOXdDgt+WdDAzmwH8BtgMvAx8DPQFLgB64sXM513UwVixJPGYWQ2wGFgObAOygWOAycAnwDHOuQ1R5RVL0iwzGwi8D6QAOcDVzrnfNyrT6rgws7uA7+B9hz4JBIEvAr2Bmc65+9vrPUnHMLNioBdwb4zNZc65uxqV1zFJ4jKzm4EfAduBf+GdO/UBJgIvO+e+G1VWsST7MbMJwHlxNp8AnAI845w7O6qOYkliMrOfAd/F+5zn4B2bhgPnAqnAl51zs6PKK5ak3ZjZI8B0vN+ATwPlwKeB0TTqUzCzXsAuv+pWvLgaSIxz/KjX3wIU4PWNLgJCwLHAVH9fpzrn3mhRY51zB30CXgAc3o+e6Pxf+vm/TXYblZISFycDhwEGTPNjYXacsj38P/hqYHJUfob/h+qALzaqMwTvS2gHMCQqPw/4yK9zbLL/H5TaHEen4J1gBBrl98PrdHbA5xRLSi2Mp4w4+Xf6n/ODiiWl1iT/O+7fwBrgF/5nfFVb4wLvpNT52/MavdYO//WGtNf7Uuqw+CkGiltYVsckpabi4/P+5/kikBtje5piSaktCXjD/5zPVSwptSBe+gFhYAtQ2Gjbyf7nvFaxpNQRCe8CmgPWAn2i8tOAp/xtl0flB4EzgUP89duIcY7faB/fA/rHyL/Jr/t+S9t70A+dYWbDgNPwTpQfaLT5Vrye++lmlt3BTZMkc8697Jxb7fy/rmZciHf156/OuXeiXqMKuNlf/XqjOlcA6cD9zrniqDq7gB/7qzMOsPnSSTjnXnLOPe2cizTK3wL81l+dFrVJsSRx+XEQy+P+8rCoPMWStMQ1eBfEvoJ3zhPLgcRF3fqdfrm6OsV451vp/j7l4KFjksRk3jBiPwMqgIudc3sbl3HO1UatKpakVcxsLN4TYJuAZ6I2KZYknsF4Q82+5ZzbFr3BOfcysBcvduoolqQ9XeAv73bOba/L9L8bb/FXZ0bl1zjnnnPObW7pDpxzP3POfRJj08+ASmCsxRmqsbGDvqMZ78cVwLwYHUF7gQVAFt4Xk0g8dXH0fIxtr+KdOE81s/QW1nmuURnpnup+NIWi8hRLciDO8ZfvReUplqRJZjYK+Clwn3Pu1SaKHkhcKJYOHulmdqmZ3eSPB3hynLEodUySeKYCQ4FngV3+mKjf8+Mp1jioiiVpra/5y4edc9FjNCuWJJ7VeMMNHGVmfaI3mNmJQC7eE2F1FEvSnvr5y7UxttXlTfKHzEg0x77+iv3GuI9FHc3eeDkAq+JsX+0vR3RAW6TrihtHzrkQsA5vHKdhLayzGe/OsgFmlpXYpkpnYGapwJf91eiTC8WSNMvMbjCz28zsHjN7DW9My/fwOg3rKJYkLv8Y9CjeED43NVO8VXHhPwV2KN4YvbHupNC5VffSDy+W7sQbq/klYLWZndSonI5JEs8Uf7kVbx6Cf+F9n90LLDSz/5hZ9J2DiiVpMTPLBC4FIkDjsUkVSxKTc24n3lACfYHlZvY7M/uJmT0OzMMb5udrUVUUS9Ke6u5iHhpjW3RMHd4O+/483oWVN51zu1tSQR3N3mRcAHvibK/Lb48rA9J9HEgctbROzzjbpWv7KTAWeNY590JUvmJJWuIGvOGdrgOOx7tYcZpzriSqjGJJmvK/eBNsXe6cq2ymbGvjQudWB48/Ap/C62zOBsYBD+GNO/mcmR0RVVbHJImn0F/OADLxJjfKxTtPegE4EXgiqrxiSVrjIrxYeM5FTZjsUyxJXM65e/GGLEgFrgZuxOt02wDMajSkhmJJ2tO//OX1Zta7LtO/ceT2qHJ5idypmQ3Fm6wyhDfBd4uoo7l55i9bMk6vSDwHEkeKvW7KzK7BO1CvwJs5tlXV/aVi6SDmnOvnnDO8zp0L8K5kLzGzSa14GcXSQcrMjsK7i/lu19LZo5t5SX/Z2rhQHHVxzrnb/bkItjrnKpxzHzjnZuBNqJ2JN/lMS+mYdPCqG2rFgAudc/Odc2XOuWXA+cBG4KQ4w2jEoliSaF/1lw8dQF3F0kHMzL4LPAnMAorwLqgeiTdUwZ/N7OeteTl/qViSA/FXvOFUith3h/29wFLgM+x7WrBFQ1u0hJkV+vssAK51zi1saV11NDd/lahHo3IisRxIHLW0Tmkb2iWdjJl9E7gPWA6c7D+WFU2xJC3md+48hTepbT7wSNRmxZLsJ2rIjFXsmzykOa2Ni+bKN3cHj3R9dZPdnhiVp2OSxFM3Yeha59y70Rv8Jy7qnvw6yl8qlqRFzGw03hjgG/HGAG9MsSQxmdk0vEnQ/umcu945t9a/oLoY7wLYJuA7ZlY3bIFiSdqNP5/cuXhPtW7Bu1ntCrxj2/HADr/otpgv0Ep+J/NLeMO7XOuce7A19dXRDCv9ZbxxAg/zl/HGcBaBJuLI/1E/FO9xg7UtrHMI3hXTjc65isQ2VZLFzK4D7gc+wOtk3hKjmGJJWs05tx7v4sWYqAlLFEsSSw7e5zsKqDIzV5fwhmMB+D8/715/vVVx4Zwrx/sBluNvb0znVt1f3Q+d7Kg8HZMknrrPOd7Yj3Ud0ZmNyiuWpDnxJgGso1iSeM72ly833uB/tm/j9adN9LMVS9KunHMh59zdzrkJzrlM51wP59wZeL8BJwCVwLK27sePvVeA0cA3nXO/au1rqKN534HjNDNr8P9hZrnAcXgf2Jsd3TDpUl7yl2fE2HYikAUsdM5Vt7DOmY3KSBdnZt8D7sF7vOXkRmN6RVMsyYHq7y/rfkgpliSWauDhOGmJX+Z1f71uWI0DiQvF0sGtboiD6B/UOiZJPK/idcAcZmbBGNvH+stif6lYkmaZWQbeXX8RvO+0WBRLEk+6vyyIs70uv8ZfKpYkWaYDGcDjzrnatryQmQ0A/oM3qeCM1t7JXM85d9AnvMexHDCzUf4v/fzfJruNSslNwDQ/FmbH2d4DKMH7AT85Kj8DWOjX/WKjOkOBKrzHHIZE5ecBH/l1jk32e1dKSPzc4n+e7wC9mymrWFKKFxuHA/1i5AeAO/3PeYFiSelAE954ug64qq1xgfeosvO350XlD/Ffpyr6tZS6XgLGxPpOAwbjjRXogJui8nVMUmoqnmb7n+cdjfJPxeso3A30UiwptSKmpvuf69NNlFEsKcWLjYv8z3ILcGijbWf6x6VKIF+xpNQRCegRI28KsBPYCwxrou5txDjHb1RmELAG76alr7Slrea/4EHNzIrw/vgLgbnAh8DRwMl4j3VOdc7tiP8K0h2Z2XnAef5qP+B0vDtzXvPztjvnbmhU/km8L4u/4v3Bn4s3rs2TwEWu0R+cmc0EfoX35fI3vCuiFwID8CZpugHp0szsMrwJJMJ4M7bGGpO02Dk3K6qOYkn24w+98gu8O7/W4H3WfYGT8CYD3AJ8yjm3PKqOYklazMxuwxs+42rn3O8bbWt1XJjZ3cD1eOPHPQkEgS/gjSc+0zl3f7u9GWl3frzciPd04Dq8HzlFwFl4P6yfBc53ztVE1dExSWLyx4NcAAzHO9d+G++ixfl4P44vds49EVVesSRNMrPX8MYuPdc593QT5RRLsh//afcXgE/jfb89hXeuPQpvWA0DrnPO3RdVR7Ek7cbM3sK7uPEBXkyOwZsIsBq4wDn3QqPyN+LdqATe0BpH4PV71k0c+Hr0+b6ZrcO7IWQR8K84zZjlnCtutrHJ7pXvLAkYCPwR2Iz3x70eb8KuJu8+VOq+iX1XfeKl4hh1jsP7YbXLPwi8D3wbSGliP+fgPZ6wFygH/gtcluz3r9RhceSAVxRLSi2IpbHAA3jDr2zHe8x4j/853xbv+0qxpNTSRDN3OxxIXACX+eXK/Xr/Ac5O9ntVSki8nAQ8BqzAu9u0Fu9urheBL4N3Q0uMejomKcX7nHvjPVG6Du/32A68m4COUSwptTKWRvnfZxuaioeo8oolpVifcRpwHd4wqqX+ufc2vE640xRLSh2ZgP/B6wTejde5vA5v8uUhccq/QtN9ELMalW+uz8IB01rSVt3RLCIiIiIiIiIiIiJtoskARURERERERERERKRN1NEsIiIiIiIiIiIiIm2ijmYRERERERERERERaRN1NIuIiIiIiIiIiIhIm6ijWURERERERERERETaRB3NIiIiIiIiIiIiItIm6mgWERERERERERERkTZRR7OIiIiIiIiIiIiItIk6mkVERERERERERESkTdTRLCIiIiIiIiIiIiJt8v+ZbEJs6iSp+wAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<Figure size 1728x576 with 3 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "#Plotting the predicted fitness effect for the variants seen more than once\n",
    "plt.close('all')\n",
    "f, (ax1, ax2) = plt.subplots(2, 1, sharey=True, sharex = True, figsize=(24, 8))\n",
    "gs = matplotlib.gridspec.GridSpec(3, 1, width_ratios=[1], height_ratios=[10,0.5, 0.5])\n",
    "ax1 = plt.subplot(gs[0])\n",
    "ax2 = plt.subplot(gs[1])\n",
    "ax3 = plt.subplot(gs[2])\n",
    "gs.update(hspace=0.05)\n",
    "\n",
    "nonsense_list = []\n",
    "y_list = []\n",
    "x_list=[]\n",
    "y_list_top20 = []\n",
    "x_list_top20=[]\n",
    "y_list_nonsense = []\n",
    "x_list_nonsense = []\n",
    "x=[]\n",
    "y=[]\n",
    "OBS_EXP_DICT={}\n",
    "top20_lower_err = []\n",
    "top20_upper_err = []\n",
    "missense_lower_err = []\n",
    "missense_upper_err = []\n",
    "nonsense_lower_err = []\n",
    "nonsense_upper_err = []\n",
    "\n",
    "top20 = c1\n",
    "missense = c6\n",
    "nonsense = c10\n",
    "synonymous = c3\n",
    "\n",
    "axisfont=20\n",
    "domainfont = 12\n",
    "titlefont=12\n",
    "axislabelfont=12\n",
    "\n",
    "top_20_DNMT3A = ['p.R882C', 'p.R729W', 'p.R326C', 'p.R320*', 'p.R882H', 'p.R736H', 'p.Y735C',\\\n",
    "                'p.R736C', 'p.W860R', 'p.R771*', 'p.R598*', 'p.P904L']\n",
    "\n",
    "PWWP_col = 'mediumvioletred'\n",
    "ADD_col = 'darkviolet'\n",
    "MTase_col = 'midnightblue'\n",
    "\n",
    "for k, v in DNMT3A_missense_s.items():\n",
    "    if k in top_20_DNMT3A:\n",
    "        codon_number = int(((re.findall('\\d+', k)))[0])\n",
    "        s_value = float(v[0])*100\n",
    "        x_list_top20.append(codon_number)\n",
    "        y_list_top20.append(s_value)\n",
    "        \n",
    "        s_upper = float(v[3])*100\n",
    "        s_lower = float(v[4])*100\n",
    "        error_upper = s_upper-s_value\n",
    "        error_lower = s_value-s_lower\n",
    "        top20_lower_err.append(error_lower)\n",
    "        top20_upper_err.append(error_upper)\n",
    "\n",
    "    if k not in top_20_DNMT3A:\n",
    "        codon_number = int(((re.findall('\\d+', k)))[0])\n",
    "        s_value = float(v[0])*100\n",
    "        x_list.append(codon_number)\n",
    "        y_list.append(s_value)\n",
    "        \n",
    "        s_upper = float(v[3])*100\n",
    "        s_lower = float(v[4])*100\n",
    "        error_upper = s_upper-s_value\n",
    "        error_lower = s_value-s_lower\n",
    "        missense_lower_err.append(error_lower)\n",
    "        missense_upper_err.append(error_upper)\n",
    "\n",
    "    missense_err = [tuple(missense_lower_err), tuple(missense_upper_err)]   \n",
    "    \n",
    "for k, v in DNMT3A_nonsense_s.items():\n",
    "    if k in top_20_DNMT3A:\n",
    "        codon_number = int(((re.findall('\\d+', k)))[0])\n",
    "        s_value = float(v[0])*100\n",
    "        x_list_top20.append(codon_number)\n",
    "        y_list_top20.append(s_value)\n",
    "        \n",
    "        s_upper = float(v[3])*100\n",
    "        s_lower = float(v[4])*100\n",
    "        error_upper = s_upper-s_value\n",
    "        error_lower = s_value-s_lower\n",
    "        top20_lower_err.append(error_lower)\n",
    "        top20_upper_err.append(error_upper)\n",
    "    \n",
    "    top20_err = [tuple(top20_lower_err), tuple(top20_upper_err)]\n",
    "        \n",
    "    if k not in top_20_DNMT3A:\n",
    "        codon_number = int(((re.findall('\\d+', k)))[0])\n",
    "        s_value = float(v[0])*100\n",
    "        x_list_nonsense.append(codon_number)\n",
    "        y_list_nonsense.append(s_value)\n",
    "        \n",
    "        s_upper = float(v[3])*100\n",
    "        s_lower = float(v[4])*100\n",
    "        error_upper = s_upper-s_value\n",
    "        error_lower = s_value-s_lower\n",
    "        nonsense_lower_err.append(error_lower)\n",
    "        nonsense_upper_err.append(error_upper)\n",
    "\n",
    "    nonsense_err = [tuple(nonsense_lower_err), tuple(nonsense_upper_err)]        \n",
    "    \n",
    "msize = 12\n",
    "medgewidth = 2\n",
    "linew = 3\n",
    "csize = 4\n",
    "cthick = 3\n",
    "\n",
    "markers1, caps1, bars1 = ax1.errorbar(x_list, y_list, yerr = missense_err, fmt = 'o', ecolor = missense, elinewidth = linew, \\\n",
    "             capsize = csize, capthick = cthick, markersize = msize, markeredgewidth = medgewidth, \\\n",
    "             markeredgecolor = missense, markerfacecolor = missense, zorder = 1)    \n",
    "\n",
    "markers2, caps2, bars2 = ax1.errorbar(x_list_nonsense, y_list_nonsense, yerr = nonsense_err, fmt = 'o', ecolor = nonsense, elinewidth = linew, \\\n",
    "             capsize = csize, capthick = cthick, markersize = msize, markeredgewidth = medgewidth, \\\n",
    "             markeredgecolor = nonsense, markerfacecolor = nonsense, zorder = 2)    \n",
    "\n",
    "markers3, caps3, bars3 = ax1.errorbar(x_list_top20, y_list_top20, yerr = top20_err, fmt = 'o', ecolor = top20, elinewidth = linew, \\\n",
    "             capsize = csize, capthick = cthick, markersize = msize, markeredgewidth = medgewidth, \\\n",
    "             markeredgecolor = top20, markerfacecolor = top20, zorder = 3)    \n",
    "\n",
    "# loop through bars and caps and set the alpha value\n",
    "a = 0.7\n",
    "[bar.set_alpha(a) for bar in bars1]\n",
    "[cap.set_alpha(a) for cap in caps1]\n",
    "[bar.set_alpha(a) for bar in bars2]\n",
    "[cap.set_alpha(a) for cap in caps2]\n",
    "[bar.set_alpha(a) for bar in bars3]\n",
    "[cap.set_alpha(a) for cap in caps3]\n",
    "\n",
    "#Only show the required axis lines\n",
    "ax1.spines['top'].set_visible(False)\n",
    "ax1.spines['right'].set_visible(False)\n",
    "\n",
    "for axis in ['bottom','left']:\n",
    "    ax1.spines[axis].set_linewidth(2)\n",
    "\n",
    "for axis in ['bottom','left']:\n",
    "    ax1.spines[axis].set_color(grey3)\n",
    "    \n",
    "ax1.set_ylabel('fitness effect (s)', fontsize = axisfont, labelpad = 10)\n",
    "ax1.set_title('Fitness effects across DNMT3A', fontsize = 18)\n",
    "\n",
    "# ax1.set_yscale('log')\n",
    "\n",
    "# STACKED BAR CHART\n",
    "# define the exons + create data\n",
    "exons = [{'2':(24-1)+1, '3': (59-25)+1, '4': (150-60)+1, '5':(164-151)+1, '6':(213-165)+1, '7': (285-214)+1, '8': (338-286)+1,\\\n",
    "          '9': (374-339)+1, '10': (427-375)+1, '11': (477-428)+1, '12': (492-478)+1, '13': (518-493)+1, '14':(556-519)+1, \\\n",
    "          '15': (617-557)+1, '16': (646-618)+1, '17': (694-647)+1, '18': (725-695)+1, '19': (774-726)+1, '20': (803-775)+1, \\\n",
    "          '21': (826-804)+1, '22': (866-827)+1, '23': (912-867)+1}]\n",
    "exon_orders = [['2', '3', '4', '5', '6', '7', '8', '9', '10', '11', '12', '13', '14', '15', '16', '17', '18', '19', '20', '21', '22', '23']]\n",
    "colors2 = [c4, c0, c3, c1, c5, c6,c7, c8, c9, c10, c2,c4, c0, c3, c1, c5, c6,c7, c8, c9, c10, c2]\n",
    "names = exons[0].keys()\n",
    "values = np.array([[data[name] for name in order] for data,order in zip(exons, exon_orders)])\n",
    "lefts = np.insert(np.cumsum(values, axis=1),0,0, axis=1)[:, :-1]\n",
    "orders = np.array(exon_orders)\n",
    "\n",
    "#plot the bar chart\n",
    "for name, color in zip(names, colors2):\n",
    "    idx = np.where(orders == name)\n",
    "    value = values[idx]\n",
    "    left = lefts[idx]\n",
    "    ax2.bar(x=left, height=-10, width=value, bottom=-5, color=color, orientation = 'horizontal', label = name, edgecolor='k', linewidth=1, tick_label = name)\n",
    "    ax2.text(left + (value/2.), (-0.26), name, ha='center', va='center', fontsize = axislabelfont)\n",
    "    \n",
    "\n",
    "# STACKED BAR CHART of domains    \n",
    "domains = [{' ': (291-1)+1, 'PWWP' :(350-292)+1, '  ':(481-351)+1, 'ADD': (614-482)+1, '   ': (633-615)+1, 'MTase': (912-634)+1}]\n",
    "domain_orders = [[' ','PWWP','  ','ADD','   ','MTase']]\n",
    "# blank_col = c13\n",
    "blank_col = grey1\n",
    "colors3 = [blank_col, PWWP_col, blank_col, ADD_col, blank_col, MTase_col]\n",
    "names_domains = domains[0].keys()\n",
    "values = np.array([[data[name] for name in order] for data,order in zip(domains, domain_orders)])\n",
    "lefts = np.insert(np.cumsum(values, axis=1),0,0, axis=1)[:, :-1]\n",
    "orders = np.array(domain_orders)\n",
    "\n",
    "#plot the bar chart\n",
    "for name, color in zip(names_domains, colors3):\n",
    "    idx = np.where(orders == name)\n",
    "    value = values[idx]\n",
    "    left = lefts[idx]\n",
    "    ax3.bar(x=left, height=-10, width=value, bottom=-5, color=color, orientation = 'horizontal', alpha = 0.9, label = name, edgecolor='k', linewidth=1)\n",
    "    ax3.text(left + (value/2.), (-0.25), name, ha='center', va='center', fontsize = domainfont, color = 'k', fontweight = 530)\n",
    "\n",
    "    \n",
    "\n",
    "# CONFIGURING THE GRAPH APPEARANCE\n",
    "#Set the x and y axis limits\n",
    "ax1.set_xlim(0, 912)\n",
    "ax1.set_ylim(-0.01, 30)\n",
    "ax2.set_xlim(0, 912)\n",
    "ax2.set_ylim(-0.5, 0)\n",
    "ax3.set_xlim(0, 912)\n",
    "ax3.set_ylim(-0.5, 0)\n",
    "\n",
    "#x-axis ticks\n",
    "x1_major_ticks = []\n",
    "x1_major_tick_labels = []\n",
    "x2_major_ticks = [0, 100, 200, 300, 400, 500, 600, 700, 800, 912]\n",
    "x2_major_tick_labels = [\"0\", \"100\", \"200\", \"300\", \"400\", \"500\", \"600\", \"700\", \"800\", \"912\"]\n",
    "ax3.set_xticks(x2_major_ticks)\n",
    "ax3.set_xticklabels(x2_major_tick_labels, fontsize = axisfont)\n",
    "ax1.set_yticklabels([])\n",
    "ax1.set_yticks([])\n",
    "ax1.set_xticklabels([])\n",
    "ax1.set_xticks([])\n",
    "ax2.set_yticklabels([])\n",
    "ax2.set_yticks([])\n",
    "ax2.set_xticklabels([])\n",
    "ax2.set_xticks([])\n",
    "ax3.set_yticklabels([])\n",
    "ax3.set_yticks([])\n",
    "\n",
    "#y-axis ticks\n",
    "y1_major_ticks = [0, 5, 10, 15, 20, 25, 30, 35]\n",
    "y1_major_tick_labels = [\"0\", \"5\", \"10\", \"15\", \"20\", \"25\", \"30\", \"35\"]\n",
    "ax1.set_yticks(y1_major_ticks)\n",
    "ax1.set_yticklabels(y1_major_tick_labels, fontsize = axisfont)\n",
    "\n",
    "#set panel legends\n",
    "legend_marker_size = 2\n",
    "legend_elements = [Line2D([0], [0], marker = '_', color = missense, alpha = 1.0, markersize = legend_marker_size, \\\n",
    "                          lw=3, label='missense'),\n",
    "                   Line2D([0], [0], marker = '_', color=nonsense, alpha=1.0, markersize = legend_marker_size, \\\n",
    "                          lw=3, label='nonsense'),\n",
    "                  Line2D([0], [0], marker = '_', color=top20, alpha=1.0, markersize = legend_marker_size, \\\n",
    "                          lw=3, label='top 20 observed variant')]\n",
    "    \n",
    "ax1.legend(ncol=1, handles=legend_elements, loc='upper left', frameon=0, fontsize = 20)\n",
    "\n",
    "ax1.grid(which='major', linestyle=':', linewidth='0.7', color=grey3, zorder = 0)\n",
    "\n",
    "# plt.show()\n",
    "\n",
    "Fig1e = os.path.join('DNMT3A_variant_fitness_lower.pdf')\n",
    "plt.savefig(Fig1e, bbox_inches='tight')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 308,
   "metadata": {
    "code_folding": [
     0
    ]
   },
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAABacAAAHvCAYAAABAPWMbAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDIuMi4zLCBodHRwOi8vbWF0cGxvdGxpYi5vcmcvIxREBQAAIABJREFUeJzs3Xd4VFX+x/H3SUJCCJDQAlKjKGAFQYWlLEgVkCKiKKyAWH6KgqyKq64aQF1xrSAEXBRFBRtKEaUpIEVAeg2dUEMghNBCIOX8/rgzszOZSQjFENbP63nmmZlzzzn33HvmzgPfOfleY61FRERERERERERERKQgBV3qAYiIiIiIiIiIiIjIn4+C0yIiIiIiIiIiIiJS4BScFhEREREREREREZECp+C0iIiIiIiIiIiIiBQ4BadFREREREREREREpMApOC0iIiIiIiIiIiIiBU7BaREREZHLiDEmxBhjjTEfXeqxFEbGmH7GmM3GmDOu81TZVd7CGLPUGHPCVf63Sz1WEREREZE/OwWnRURERAoBY0wzV9A0t0eDPNoGGWMGGWM6FuSYCxtjTCtgOLAeeBR4AEgxxpQFvgeKAk+7yhf9QWOo65qLqn9E/39Wxpirc1wPWcaYVNcPEV8ZY+42xoQEaNfSq03vANvdP/ZMzlG+0FV+xhhTLpcxjfTqu7FXX/l9VDbGVDDGfGKMWWeMOWKMOWWM2WqMGWOMueos56S/q58jxpjwczylIiIiIoWC3z/gREREROSS+hL4KUD5NgBrbaYrEJXptS0IiAU+Bqb+4SMsvFq5nh+01h5zFxpjmgElgZettX/0+amLMxc/A7v/4H39Gc0EvnC9Lg5cDbQHugHLjTF3WWv35tJ2iDHmK2ttej73lYFzbf0NeM97gzGmKHA/kI7zowdAFs4PH96aAQ8Bo4DfcmxLAa4ErgFm4HxeTgE1gT7AvcaY26y1m3MZXx9gO1Ad6AKMz+dxiYiIiBQaCk6LiIiIFC4rrbVf5FXhHIJrfzYVgCzvwLRXOTjBQMmDMcYAEdbaE5d6LLnYlPP6MMYMBJ4B3gKmGWPqWWuzcrRbDtwC9HPVy480nIDyg+QITgN3AaWACUB3AGut5b+Bc/fYiuIEp3/L5breADTOWWiM+d617yeA/gG21wNqu/b9PE6gWsFpERERueworYeIiIjIZSRnzmljzNU4KzwBHvJKGZCZs74r9cACY0yaMSbZGPMfY0xEgH1UMsaMNsbscaU12Od6XzZHvTLGmGHGmB3GmHRjzGFjzHJjzNM56vU2xixzpWE4aYzZboz5whhTJp/H3MYYM9sYc9S1nzXGmEe9tl9tjLE4q1aDvc7Bz8aYvTgrygEWeJ8bV9uixpiXjDEbXX0fMcZMNcbUDjAOY4z5P2PM78bJXX3cGLPWGBPr2v4aMCbHvrznKtwYM8Q4qSjSXPtaZ4wZmo9zEOIa5wJjTJJrXna5UkuUzqXNPcaYX13nLc0Ys8k1XyGu7e6UFw8YJ1d3PHAaGODVRzPXeTzm6mOFCZwe40ZjzHeuz8ppY0yiMWaOMaatV53zPv68WMfbwNc4Adt7AlSbAKwBXjDGlDqH7j8BbnQFg709CKwA1p3HkPNjl+s5t7E+BBwDJgPjgNuNMVf+QWMRERER+cNo5bSIiIhI4VIsZxAYOG2tPZ5L/QNAL5wA1Tz+G4jNzlGvHs5qz49xVnc2Bx7BSQ/S113JFeD6DQh21d2Bk3bgcZwA2K1eK5O/B/4CjAbWAhHAtTipDN519fcgMBb4FXgFJ21BVaAdUBY4nNfJMMY8Dox0jelVnNWsbYAPjTFXWmtfcJ2DB1xjbOA6HwCJOOk82uME814FtrjPjTEmFJgF1Ac+w8lXXcp1Xn4zxjS21q5y1TU4KVe6AYuB14FU1/HeDQwGvsVZpe29L3ClZHGdp57Ap64+irjObfO8zoFLUZzVwd/hBCRPArfh5NZu5JoX948UGGPeBJ7DWZn7DpCEk/7hbuCfgPfK6Gdcx/2x61zucvXRGZjoOo9vufZ5P/CJ69y7g/LlgDk4aS1G46SnKAvc6hrj9Itw/PnxEc78tAe+yrHN4qwwng68gHNu8mMKzme0D04wGuPkE2+Bswq75AWPGs9nsSQQipOqZLBrk1+KH/PflCLfWGtPGWO+AN7ECZi/cjHGIyIiIlJgrLV66KGHHnrooYceelziB05A1+by+MqrXoir7KO8ygJsywJuybFtJs5K2XCvsh9xApQVc9St7+rjJdf70q5+h5/luKbipNMIPo9zUtk1vs8CbBuJE1iv5lX2BZAZoO7DrrE2zlE+ECdQ3TJHeRSwF/jZq6y7q49PgaAc9YPOti/XtmPA1PP8fAR5z5NX+f+59tfFq6yhq2w2EJajvvF63dJVLxkom6NeEdc5SAEqeJWHAUtcn4WrXGVdco4hl2O4kOO/2rWP9/OoE+2qszTAMQ5wvf8F5weSyjmuj8k5+loIpLpeD3OdhzDXe/ePLFE4Ae+A853j8/C3sxxfZ3yv+UTgqVzq9si5T+AHnB8FgvLajx566KGHHnrooUdheyith4iIiEjh8h+cG/t5P167CP0utNYuz1E2B2elZjUAV3qItjgrc88YY8q6Hzg3XtsJtHa1TcNJJ9LAGFMtj/0eBUoA7Vyrj8/FPa7xjfUei2s8P+Cs7m5xjn16+xvOyuLVOfoOwQliNjXGhLnqugOCz1prfVal53yfh6M4KSKuP9eBWmuzrbWnAIwxwcaYKNdY57iq1Peq3sP1/Ly19nSOfmyA7j+11ibnKLsVqITzg8cBr/angbdxguUdvY4LnDkukcdhnPfx55N7RX9eq5n/gRNgH3IO/Y7FWVne2fUZ7gVMstamntcoA1uEc613xFnZfQAo7U7BksNDwHZr7UKvsk+BKvz3pqAiIiIilwWl9RAREREpXLZaa3/+A/rdEaDMnVLDnfu5FmBwVuP+Xy79ZIFzU0bj5JZ+F0gwxmzACZROstbO9ar/Gs4N36YCycaYX3FSK3xtz37TvWtdz3PzqFP+LH3kpRZO8PtQHnVK46xivQbYGyCIey6ewkm/st4Ysx3nuKYC03IJGvswxtyHk4KjNs7KZm/euYmvwZmntfkc15YAZe78xRsCbFvver7K9TwHJ6fzQ0BPY8wynHQpX1trN3m1u6Djzwd3UDrnDTE9rLXLjTHfAr2MMe8Am8/WqbV2jTFmFU7ajIM4x/3YRRiv9z4OAe7r/gdXqo61OOlRnnDXM8ZchfNXFqOMk2/eLR4n7cpDOH8RISIiInJZUHBaRERE5M8hK49tJsfzOJwUGYGkuV9Ya0cYYybh5Pj9K3Av0M8YM95a+zdXnc3GmFo46RVaAE1xcgMPMsb81Vq7Mx/j6oETFAxkWy7l+REErMZJ75GbFK+xXFAA1Vr7vSs43w7nPLTCSfswzxjT2nrljM7JGHMvTs7rJUB/nJQb6TjB9R/xvdH5ua5QTwtQlu8+XIHlHq4bG7YFmuDkdH7ZGNPPWjvKVe+8jz+fbnI9ny3g/E+c/OtDXc/5MRYnvUcRnPQZv5zPAPPLWrvXGDMHeMQYM8Dr3PTBmZu+eOWK99LJGFPGWptnLncRERGRwkLBaREREZHL38VYdQpOoNcCRfK7ettauw8nFcl/XCkIxuMEKt+xrpsJulJB/Oh6YIzpiHOjuQE4q2lzs9X1fOgPWk2+FSgH/JKPlbubcdJWlD3L6uk8+3EFDT8HPneliHgLZzX0ncCkPJo+gBNEvt1am+4uNMbckMtYWwI3AivzGk8etrueA6XguM717LMa31q7DlgH/NuVIuZ3nADwKK8653v8+fGw6/nHvCpZa7cZY/6DsyK5ST77Ho+TzqQ5MOQcUrlciHCcYHhx4IgxJggnpcgKnPOaUyXgfZx0NcMKYHwiIiIiF0w5p0VEREQuc9baLJxVtKUvsJ8knHQM9xhjbs253TjKuV4XM8aE52ifiROcxD0WV17knFZ618nD18AZYIgxpmiA8UQZY0LP0kdePsMJ6AUMkBtjvFOGjMdZsfrvnLmzc7x3pyopnaNOiDEm0rvMFRBfHah+AFk4N2/0/Pvdtd+XAtSd4Hp+I+f5OYe838uAfcBDxphor/ahwLOusUx1lZXO2a+1NgVIAIobY0IvwvHnyRjzDNANWAVMzEeTIThz9WZ++rfWHsFJ5TEYGHOew/ST4zPmXX4jTvqOLa59A7TBuUnoZ9baiQEew4A9OKurRURERC4LWjktIiIi8r9hCdDGGPMcToAqy1r7zXn083/AQmChMeYznGBfCE4O4s7Axzh5pK8Dfnal9dgAHHGVPY6z6naRq785xphDwALXuErj5O7NJvfUIQBYa3cZY54ERgMbXXl4d+Osdr4J6ATUwElxcT7exVlh/J4xpiUwDzgOVMVJQXKc/95g7iugi2vsNYwxP+Dc4K+Gq25tV73fcVZPv+QK5J/kvzeT3G2MmYoTkHXnLn4cJ3VInqt9cQKunXDO5+c4N/W7C/AL2ltrf3PlU34GcOdYPoAzh/cYY24+W75va22mMaYf8C2wzBgzxnUs9wG34awedq+c7gM86fosbAMygdtd52WCtfaM60eKCzl+t1rGmL+5XkcA1YEOOPnDfwfucv1Ykydr7UHXOYrN536x1n6a37rn4GVjTFPgJ5xgfjDOive/uV4/4VX3IZzPVl4rzL8HnjLG3BLgBqgiIiIihY6C0yIiIiL/Gx4DRuCspC2Bs9L2nIPTroBwXeB5oCPQEyedxB5gMv9dlboLJzd1M5wgaRhOkPhD4E2v1BMjgXtc4yuNcxPGlUBfa+28fIxnjDEmHme17uNAJJAMbMLJHZzXzQzP1vcZY8wdwJM4wcDBrk37cYL947zqWmNMN5xgYR+coGYWTmqLb73q7TTGPIyTc3kUTlqGj13thuMEbFvhBFYTcQKNb1hrD5xlrF8YY4rjrPJ+ByegOxV4mQD5uK21z7pu4vcE8A+cVd97gGk4q+zzc34mGWNa4Xym/oHzf4d4oI+19hOvqnNwfizoAFyBE5zeiRMcH+Gqc/xCjt9LG9fD4qx8PoDzeXoZ52acZw1Me3kb53N5ITfVvFA/ABVxVn1H4wSk9+L8GPK2tTYePH+B0AFYZq3dk0d/3+F8RvoACk6LiIhIoWcuzo2xLz5jTBmc/+i0x1k9UAnnzzrXAZ8An3jnejPGXIOzmqUNzh3Ky+Os4FkCvJ/jrvE599UL5x/u1+H8J2MVzj8Gp138IxMRERERERERERGRwhycfgxntUkiMBfnTzjL4wSgI3FWBdzjvnmNMeYrnBUHG3H+FDUFqImz4icYeMpaOzzAft7GWdWxF2clUCjOnyuWBvpZa0fkbCMiIiIiIiIiIiIiF6YwB6eb4/y53485VkhXwMknVwXoaq39zlXeG1jjviu8V/2mwGycP/2LsdYmem1riJMPcTtwq/tmI8aYGJy7YEcAtay1CX/IQYqIiIiIiIiIiIj8SQWdvcqlYa2dY639wTsw7So/gHNTHHByHLrLP80ZmHaV/4pzc5tQoGGOzY+5nl/3ugs2rmD0SJzciQ9e0IGIiIiIiIiIiIiIiJ9CG5w+iwzXc+YF1m/uep4RoM30HHVERERERERERERE5CK57ILTxpgQnLvGQ+Cgcs761XDuCp4GzPcqj8C5yeIJ71QfXra6nmucbR+jRo2yo0aNKpz5UUREREREREREREQKocsuOA0MBW4AfrLWzsyrojEmDBiPk55jkHfqDpybKgIczaW5uzwqj/4fNcYs9zQ4etSuWrXKZmRk2OXLl1vALl++3J4+fdquWbPGHjlyxG7bts3u3bvXJiYm2s2bN9tjx47ZVatW2czMTL8269atsykpKXb79u12z5499sCBA3bTpk32+PHjdsWKFTYrK8unzalTp+z69evt4cOH7Y4dO+zu3bttUlKSjY+PtydOnLDLly+32dnZPm3S0tLshg0b7KFDh+zOnTvtrl277MGDB+3GjRvtyZMnPXWXLVvmaXPy5Em7ceNGe/DgQbtr1y67c+dOe+jQIbthwwablpbm0797fydOnLDx8fE2KSnJ7t692+7YscMePnzYrl+/3p46dcqnTVZWll2xYoU9fvy43bRpkz1w4IDds2eP3b59u01JSbHr1q2zp0+f9mmTmZlpV61aZY8dO2Y3b95sExMT7d69e+22bdvskSNH7Jo1a/zaZGRk2FWrVtmjR4/aLVu22P3799t9+/bZrVu32tTUVLt69WqbkZHhc+xnzpzxzOXWrVvtvn377P79++2WLVus9/y72yxbtsyePn3arl271jOXOed/5cqVfnOZnp5u161b5zOXOec/51y65z85Odkzl0lJSXbjxo2e+bfW5jr/CQkJNiEhwR46dMhv/r2fT5w44Tf/ycnJdv369TYtLc3nfGVnZ9sVK1bkOf/p6ek+n7OsrCy7cuVKe+zYsYDzv3bt2lzn/+jRo3bz5s12//79fvN/5swZv/lfvXq1TU1NtVu2bLH79u2z+/bts1u2bPGZf+823vPvvpb3799vN2/e7Jn/QNey9/x7X8ve8+993tLT0/O8lnPO/7Jly2xaWprf/LuvZff855xL97Wcc/4DXcvuz457/pOSkvzmP+e17J5/72vZe/7XrVvnM//u6989/97Xcs75977G8rqWA82/u417/vO6lvVdru9yfZfru1zf5fou13e5vsv1Xa7vcn2X67tc3+X6Lv8zfpdTQArtDREDMcb0B4YBm4BG1tqUPOoGA18C9wBfA/dbr4M1xlQE9gH7rLWVA7QvApwBTltri+Y1Lveq6ccff/ycj0lERERERERERESkkDEFsZPLZuW0MeYJnMD0RuD2fASmv8AJTH8D/M36R+HdK6MjCexsK6tFRERERERERERE5DxdFsFpY8wAYASwHicwfSCPuiE4K6bvAyYA3a21fjdOtNaexFk5XdwYc0WArq5xPW+5wOGLiIiIiIiIiIiISA6FPjhtjPkH8B6wGicwfTCPuqHARJwV058BD1hrs/Lofo7r+Y4A29rmqCMiIiIiIiIiIiIiF0mhDk4bY17GuQHiCqCFtTY5j7phwCSgE/Ax8KC1Nvssuxjtev6nMaaUV18xwBPAaeCT8x2/iIiIiIiIiIiIiAQWcqkHkBtjTC9gCJAFLAD6G+OXhzvBWvup6/VooB2QjJOu45UA9edZa+e531hrfzPGvAs8Daw1xkwEQoFuQGmgn7U24eIdlYiIiIiIiIiIiIhAIQ5OA1e6noOBAbnU+RX4NEf9ssArefQ7z/uNtfYZY8xa4EngUSAbWAm8Za2dds6jFhEREREREREREZGzKrTBaWvtIGDQOdRvdgH7GgeMO9/2IiIiIiIiIiIiInJuCnXOaRERERERERERERH536TgtIiIiIiIiIiIiIgUOAWnRURERERERERERKTAKTgtIiIiIiIiIiIiIgVOwWkRERERERERERERKXAKTouIiIiIiIiIiIhIgVNwWkREREREREREREQKnILTIiIiIiIiIiIiIlLgFJwWERERERERERERkQKn4LRc9gYNGoQxhnnz5l3qoYiIiIiIiIiIiEg+KTgtIiIiIiIiIiIiIgVOwWm57D355JPEx8dz2223XeqhiIiIiIiIiIiISD6FXOoBiFyosmXLUrZs2Us9DBERERERERERETkHWjktl0RCQgLGGHr37s327dvp2rUrZcqUoUSJErRu3Zr169cDcOjQIR599FGuuOIKihYtyq233srcuXN9+sot5/SCBQvo0KEDlStXJiwsjAoVKtCgQQMGDx7sUy8pKYlnn32WmjVrEhERQVRUFDVr1qR3797s2LHDb+wzZ86kXbt2lC1blrCwMKpXr87AgQNJTU31qxsTE0NMTAxpaWkMHDiQqlWrEhYWxtVXX82bb76JtdavzdSpU2nRogVXXHEFYWFhVKxYkaZNmxIXF+dXNyUlhRdeeIFrr72W8PBwIiMjadGiBbNmzTrrHIiIiIiIiIiIiFxKWjktl1RCQgL169fn2muvpXfv3iQkJDBp0iSaNWvG4sWLueOOOyhZsiTdunUjJSWFr776irZt27JlyxaqVq2aa78zZsygffv2lCxZko4dO1KpUiVSUlKIj48nLi6O2NhYANLS0mjUqBHbt2+nVatWdOjQAWstu3btYsqUKXTt2pWrrrrK0++QIUOIjY2ldOnS3HnnnURHR7N27VrefvttfvrpJxYvXkzJkiV9xpKRkUHr1q3Zv38/bdu2JSQkhMmTJ/P888+Tnp7uGQvAf/7zH/7v//6PChUq0KFDB8qWLcvBgwdZu3Ytn3zyCX379vXU3bVrF82aNSMhIYEmTZpwxx13cPLkSaZNm8Ydd9zBhx9+yCOPPHKxpkpEREREREREROSiUnBaLqlff/2V1157jX/+85+esldffZVXXnmF+vXrc++99xIXF0dQkLPIv1WrVvTs2ZP33nuP9957L9d+x4wZQ3Z2NvPmzaN27do+25KTkz2vf/nlF7Zv386AAQP8+jtz5gynT5/2vJ87dy6xsbH85S9/4aeffiIqKsqz7dNPP+XBBx8kNjbWr5/9+/dTu3ZtZs+eTXh4OACxsbHUqFGD9957jxdffJEiRYoA8OGHHxIaGsqaNWuIjo7OddwAvXr1YteuXXz55Zfcd999nvLU1FSaNWtG//796dixI+XLl8/1PImIiIiIiIiIiFwqCk4XIt2XZFzqIeTbhAZFLko/MTExPP/88z5lvXr14pVXXuH06dO89dZbnsA0QPfu3enTpw+rV6/OV//uYLC3QPmpA9ULDQ0lNDTU83748OGAE/j2DkwD9O7dm2HDhjF+/PiAQfPhw4f77CM6OppOnTrx2WefsXnzZm644QbPtpCQEE+wOrdxr1mzhl9//ZWuXbv6BKYBoqKiGDx4MJ07d+a7777zWW0tIiIiIiIiIiJSWCg4LZdUnTp1CA4O9imrWLEiADVq1KBEiRI+24KDgylfvjx79+7Ns98ePXrw/fffU79+fbp168btt99Oo0aNqFy5sk+9pk2bUqlSJYYOHcrKlStp164djRo1CjiuxYsXU6RIEb799lu+/fZbv32eOXOGQ4cOcfjwYcqUKeMpj4yM5Oqrr/arX6VKFQCOHDniM+5nnnmG66+/nm7dutG0aVMaNWpEuXLl/MYCcPToUQYNGuTX96FDhwCIj4/P6zSJiIiIiIiIiIhcMgpOyyUVGRnpVxYSEpLrNvf2jIy8V5l36dKFadOm8c477zB27Fg+/PBDAOrVq8cbb7xBq1atAChZsiRLliwhNjaWqVOnMnPmTMBZpdy3b19eeuklzyrmw4cPk5mZ6XdDxZxOnDjhE5zOuco653FmZWV5yp5++mnKli1LXFwcw4cP5/3338cYQ9OmTXnrrbe45ZZbPGMBmD17NrNnz85zLCIiIiIiIiIiIoWRgtOFyMVKlSGO9u3b0759e06ePMnSpUuZNm0ao0aN4s4772TVqlVcd911AFSuXJmPP/4Yay0bN25kzpw5jBw5kiFDhpCdnc2rr74KOMHy7OxsUlJS/tBx9+zZk549e5Kamspvv/3GpEmTGDt2LG3atCE+Pp7o6GhP4H7YsGH079//Dx2PiIiIiIiIiIjIHyHo7FVELm8RERE0b96cd999lxdffJEzZ84wffp0v3rGGK6//nr69evnWY08efJkz/YGDRpw5MgRNmzYUCDjjoqKol27dowZM4bevXuTkpLCggULPGMBPO9FREREREREREQuNwpOy/+kX375hVOnTvmVJyUlAVCsWDEA1q9fT0JCwlnrAfz9738H4JFHHmH//v1+bU6ePMmSJUsuaNwzZswgMzPTr/zgwYM+47nlllto0qQJ33//PWPHjg3Y17p16zztREREREREREREChul9ZD/Sc888wwJCQk0a9aMmJgYQkNDWbFiBXPmzKFatWrcd999APz88888/fTTNGzYkFq1ahEdHc3evXuZMmUKQUFBDBw40NNnixYtGDp0KC+88ALXXHMN7dq148orr+TEiRPs2rWLX3/9lcaNGzNjxozzHvd9991H0aJFady4MTExMVhrWbBgAcuWLaNevXq0bNnSU3fChAk0b96chx56iOHDh1O/fn2ioqLYu3cva9euZf369SxevJjo6OjzP5EiIiIiIiIiIiJ/EAWn5X/Siy++yKRJk1i+fDk///wzQUFBVK1alRdffJEBAwZQqlQpANq0acOAAQOYP38+U6ZM4dixY1xxxRW0atXKE7T29o9//INGjRoxfPhwFi5cyJQpU4iMjKRSpUo8+uijdO/e/YLGPXToUGbOnMnKlSv56aefKFq0KNWqVePNN9/k8ccf99ycEZxc2StWrOCDDz7gu+++Y/z48WRlZVGhQgWuu+46+vXrx4033nhB4xEREREREREREfmjGGvtpR7DZW/UqFEW4PHHH7/UQxERERERERERERG5UKYgdqKc0yIiIiIiIiIiIiJS4BScFhEREREREREREZECp+C0iIiIiIiIiIiIiBQ4BadFREREREREREREpMApOC0iIiIiIiIiIiIiBU7BaREREREREREREREpcApOi4iIiIiIiIiIiEiBU3BaRERERERERERERApcoQ1OG2PKGGMeNsZMMsZsM8acMsYcNcYsNMY8ZIwJylG/iDHmKWPMJ8aY1caYM8YYa4x5OI999HbVye3x2B9/pCIiIiIiIiIiIiJ/PiGXegB5uAcYBSQCc4HdQHmgC/AR0NYYc4+11rrqRwDvu14nAQeAKvnc1xRgdYDy5ec3dBERERERERERERHJS2EOTm8BOgI/Wmuz3YXGmBeB34G7cQLV37k2pQHtgNXW2kRjzCAgNp/7mmyt/fQijVtEREREREREREREzqLQpvWw1s6x1v7gHZh2lR8ARrveNvMqP2OtnW6tTSzAYYqIiIiIiIiIiIjIeSjMK6fzkuF6zrxI/dUxxgwAigL7gLnW2r0XqW8RERERERERERERyeGyC04bY0KAnq63My5St0/leJ9ljPkIGGCtTb9I+xARERERERERERERl+BBgwZd6jGck8GDB/8b6Az8ZK0T5uPvAAAgAElEQVQdkke9ZjhpP34YNGjQylzqRAIrgL8DzwMjgaXALUAroPqgQYO+C9QWwBjz6ODBgz9s3759RYAaNWoQHx9P2bJlWbVqFRUrVmT58uWUKVOGDRs2EBYWxr59+0hLS+PkyZPs37+f0NBQNm7cSLly5Vi5cqVPm/j4eIoUKcL+/fs5efIkp06dYu/evYSFhbF+/XrKly/v06ZUqVJs3ryZ4OBgDhw4wPHjx0lPT2fPnj2Eh4ezdu1aKlSowIoVKzxtoqKi2LJlC8YYDh48yLFjxzhz5gy7du2iWLFirF27looVK7Js2TIqVarE8uXLiYyMZNu2bVhrSU5O5siRI2RmZpKQkEDx4sVZs2aNp3/3/kqWLMn27dvJzs4mJSWFw4cPk52dzY4dOyhRooRPm/Lly7Nq1SpKlCjBzp07yczM5MiRIyQnJwOwbds2oqKiWL16tadNdHQ0a9asISIigoSEBDIyMjh69CgHDx7EGMOWLVsoVaqUT5ty5cqxdu1aihUrxu7duzl9+jTHjh0jKSmJ4OBgNm3aRJkyZVi5cqXn2MuWLcv69esJCwtj7969pKenc+LECRITEwkJCfHMv7vNsmXLKFu2LBs3bvTMZc7537BhA9HR0T5zWbp0aeLj4wkJCfHMZc75zzmX7vkPCgoiKSmJY8eOcfr0aXbv3u2Z/yuuuCLX+T906BCpqalkZGT4zf/y5cs9zyVLlvSb/6ysLHbs2EHx4sVZvXq153xVqFCBlStX5jn/JUuW9MzLsmXLqFChAqtXryYiIiLg/G/dujXX+S9WrBi7du3izJkzfvNfunRpn+uyXLlyrFu3jqJFi7J7927S09M5fvw4Bw4c8Jl/7zbe8+++lk+cOMH+/fspUqQI8fHxAa9l7/n3vpa953/FihWe81a6dGk2bdqU67W8bt06n/lftmxZwPl3X8vu+c85l+5rGfCZ/0DXsvuz457/7OxsDh8+7DP/Oa9l9/x7X8ve8799+3YiIyN95rJ8+fKe+fe+lg8dOuQz/6tWrfJcY9HR0bley0FBQX7z727jnv89e/bkei3ru1zf5fou13e5vsv1Xa7vcn2X67tc3+X6Ltd3ub7L9V3+Z/wuv+KKKwbnFhO9mIy1tiD2c1EYY/oDw4BNQCNrbUoedQfh3BDxEWvtR+e4nyrAGqAUUMdauyav+qNGjbIAjz/++LnsRkRERERERERERKQwMgWxk0J7Q8ScjDFP4ASmNwK35xWYvlDW2j3AT663f/2j9iMiIiIiIiIiIiLyZ3VZBKddNyscAazHCUwfKIDdHnI9RxTAvv50EhISMMbQu3dvEhISuO+++yhbtixFixbllltuYdq0aX5tTp8+zdChQ7npppsoVqwYJUuWpEmTJnzzzTcXpf8zZ84wfPhw6tatS6lSpShWrBgxMTF06tSJn3/+2a/+pk2b6N27N1WqVCEsLIzy5cvTvXt3Nm/e7Fe3d+/eGGNISEjgww8/5MYbb6Ro0aKUL1+eRx99lKNHj/q1Wbt2Lffffz8xMTGEhYVRrlw56taty4ABA8jIyPCpm5mZSVxcHA0aNKBkyZIUK1aMm2++mREjRpCdnZ3nXIiIiIiIiIiIiFwKhf6GiMaYfwBDgdVAK2ttcgHtur7reUcB7e9PadeuXdx2221cddVVPPDAA6SkpPD11197AsK333474ASO27Rpw6+//kqtWrV44oknSEtLY+LEiXTr1o3Vq1fzr3/967z7ByeA/OWXX3LDDTfQs2dPwsPD2b9/PwsXLmTGjBm0bNnSU3fGjBl06dKFjIwMOnTowNVXX83evXv5/vvv+fHHH5k7dy5169b1G89zzz3HzJkz6dChA61bt2bu3LmMGTOGbdu2MWfOHE+9tWvXUr9+fYwxdOzYkSuvvJJjx46xbds24uLieO211yhSpAiAZwwzZ86kZs2adO/enaJFizJ37lz69evH0qVL+fzzzy/anImIiIiIiIiIiFwU1tpC+wBeBiywHCh9jm0Hudo+nEedJgHKDPCCq+0hoOTZ9hUXF2fj4uKs5N/OnTut6xzbQYMG+WybMWOGBWzbtm09Zf/61788ZRkZGZ7ypKQkW61aNQvYRYsWnXf/qamp1hhj69WrZzMzM/3Gm5yc7HmdkpJio6KibJkyZeyGDRt86q1fv95GRETYm2++2ae8V69eFrBVqlSxu3bt8pRnZGTYJk2aWMAuXbrUU/70009bwE6ePNlvLCkpKTYrK8vzPjY21gL2ySef9Bl7Zmam7dOnT679iIiIiIiIiIiI5KJA4r+FduW0MaYXMATIAhYA/Y3xy8OdYK391KvN80At19s6rucHjTGNXa8XWt+bI843xmwBlgH7gEigEXADkAb0sNYeu2gHdRam8/CC2tUFs5P7X5R+qlWrxksvveRT1qZNG6pWrcrvv//uKRs7dizGGN59911CQv77sY2Ojubll1/m4Ycf5qOPPqJhw4bn1b8xBmstYWFhBAX5Z7spU6aM5/Vnn31GamoqI0aM4LrrrvOpd/311/PII4/w/vvvs3HjRr/tr7zyClWrVvW8DwkJ4cEHH2TBggX8/vvv3HbbbT71w8PD/cZSqlQpz+vs7GxGjBhBhQoVeO+99wgODvZsCw4O5p133uGTTz5h/PjxdOrUya8vERERERERERGRS6XQBqeBK13PwcCAXOr8Cnzq9f4OoGmOOg1dDzfv4PTbwG1Ac6A0kA3sBkYC71prldLjD1anTh2fgKpblSpVWLx4MQDHjx9n27ZtVKpUiVq1avnVbd68OQCrVq06r/4BSpYsSYcOHfjhhx+oU6cOd999N02aNKF+/foUK1bMp6273Zo1axg0aJBf31u2bAEgPj7eLzh9yy23BBwLwJEjRzxl3bp1Y9iwYXTu3JmuXbvSsmVLGjVqRPXq1f32dfjwYa655hpee+01v77BCXDHx8cH3CYiIiIiIiIiInKpFNrgtLV2EE5qjnNp0+wc6w88l/py8UVFRQUsDwkJ8dzIz32zwCuuuCJgXXd5amrqefXv9vXXX/Pmm28yYcIEYmNjAShatChdu3bl7bffpnz58gAcPnwYgDFjxuR5bCdOnMjXeNwrwbOysjxlt912GwsWLOD1119n4sSJnpzRNWvWJDY2lvvvv99nLFu3bmXw4MHnNBYREREREREREZFLqdAGp/+MLlaqjP81kZGRABw4cCDg9sTERJ965ys8PJxBgwYxaNAg9uzZw/z58/n000/54osvSEhIYMGCBT77WbNmDTfddNMF7TMvf/nLX5g2bRqnT59mxYoVzJgxgw8++IDu3btTrlw5WrZs6RnLXXfdxffff/+HjUVERERERERERORi80+uK1LIlChRgurVq7Nv3z62bt3qt33u3LkA1K1b96Lts0qVKvTo0YOZM2dyzTXXsHDhQs8q5QYNGgB4gtV/tLCwMBo2bMiQIUMYPtzJSz5lyhQAatWqRVRUFEuWLCEjI6NAxiMiIiIiIiIiInIxKDgtl4U+ffpgrWXgwIE+6S+Sk5N59dVXPXXO16FDh1i6dKlf+cmTJzl+/DghISGEhoYC8OCDDxIVFcXgwYN9bqrolp2dzbx58857LOAEvt3pTLwlJSUBePJgh4SE0K9fPxITE+nfvz+nTp3ya5OYmMjGjRsvaDwiIiIiIiIiIiIXm9J6yGXh2WefZfr06UyZMoXatWvTrl070tLS+Pbbbzl48CDPPfccjRs3Pu/+9+3bR4MGDbj22mupW7cuVapU4dixY0ybNo0DBw7Qv39/SpQoAUCZMmWYOHEid911Fw0aNKBFixZcf/31BAUFsXv3bhYvXszhw4dJT08/7/G88847zJo1i2bNmnHVVVdRvHhxNmzYwPTp0ylVqhSPPvqop+7LL7/MmjVrGD16ND/88APNmzenUqVKHDx4kK1bt7Jo0SJef/11v5szioiIiIiIiIiIXEoKTstlITQ0lNmzZ/Puu+8yYcIEPvjgA0JCQqhduzbvv/++5waB5ysmJobBgwczb9485s6dS3JyMqVLl6ZmzZoMHTqU++67z6d+ixYtWLt2LW+//TYzZ85kwYIFhIaGUrFiRZo3b87dd999QePp27cvpUqVYunSpSxatIjMzEwqV65M3759eeaZZ6hWrZqnbpEiRZg8eTJffPEFn376KdOmTePEiROUK1eOK6+8kldffZUePXpc0HhEREREREREREQuNmOtvdRjuOyNGjXKAjz++OOXeigiIiIiIiIiIiIiF8oUxE6Uc1pERERERERERERECpyC0yIiIiIiIiIiIiJS4BScFhEREREREREREZECp+C0iIiIiIiIiIiIiBQ4BadFREREREREREREpMApOC0iIiIiIiIiIiIiBU7BaREREREREREREREpcApOi4iIiIiIiIiIiEiBU3BaRERERERERERERAqcgtMiIiIiIiIiIiIiUuAUnBYRERERERERERGRAqfgtIiIiIiIiIiIiIgUOAWnRURERERERERERKTAKTgtcpnr3bs3xhgSEhIu9VAKvXnz5mGMYdCgQZd6KPnWrFkzjDGXehgiIiIiIiIiIhedgtNS4BISEjDG0Lt370s9FD8ZGRlMmjSJhx56iBtuuIGSJUtSrFgxbrzxRl555RWOHz+ea9u9e/fSp08fKlasSFhYGDExMQwYMIAjR44U4BGIXByF+ToVERERERERkf8NIZd6ACKFyfbt2+nSpQsRERHcfvvttG/fnhMnTjBz5kxeffVVvv76axYtWkTZsmX92jVs2JCDBw/SqVMnatWqxe+//86wYcOYMWMGixYtokyZMpfoqORy9tlnn5GWlnaphyEiIiIiIiIictEpOC3ipUSJEowcOZJevXoRERHhKT9z5gxdunThxx9/ZPDgwXzwwQc+7fr27cvBgwcZPnw4/fr185Q//fTTvPfee/zzn/9k9OjRBXYc8r+jatWql3oIIiIiIiIiIiJ/CKX1kAI1aNAgrrzySgDGjRuHMcbz+PTTTz31srOzGT16NLfeeivFixcnIiKCW2+9lVGjRpGdne3XrzGGZs2asX//fh544AGio6MJDw+nXr16TJgwId/jq1SpEn379vUJTAOEhoby4osvAk7eYm87duxg1qxZxMTE8MQTT/hsGzx4MBEREXz++eecPHky3+NYsWIFd999N9HR0YSFhVGtWjX69u1LYmJirm2ys7N59913qVWrFkWLFqVy5cr8/e9/59ixY351165dy/33309MTAxhYWGUK1eOunXrMmDAADIyMnzqZmZmEhcXR4MGDTxpTm6++WZGjBjhNxfeqSC2bNlCt27diI6OJigoiHnz5lGrVi1CQ0NJTk4OeAxDhw7FGMPIkSN9yvfu3cuTTz7JVVddRVhYGGXKlKFjx44sW7YsYD9JSUk89NBDlC9fnvDwcOrUqcO4ceNyPXeBvPHGGxhjGD58eMDt+/fvJzg4mFtvvdWnbMiQITRq1IgKFSoQGhpKxYoV6d69O/Hx8X59nO18QeCc02fOnGHEiBG0a9eOatWqERYWRunSpWnZsiXTp08PON6YmBhiYmJIS0tj4MCBVK1albCwMK6++mrefPNNrLWeuvm9TkVERERERERELoRWTkuBatasGampqQwbNozatWvTuXNnz7Y6dep4Xj/wwANMmDCBKlWq8PDDD2OMYdKkSfTt25eFCxcyfvx4v76PHDlCw4YNiYqK4sEHHyQ1NZVvvvmGHj16sG/fPgYOHHhBYy9SpAgAISG+l82cOXMAaN26NUFBvr/3lChRgkaNGjFr1iyWLFlCixYtzrqfadOmcffdd2OtpWvXrlSrVo0VK1YwatQopkyZwqJFi4iJifFr9/e//5358+dz77330qlTJ2bOnMn777/PggULWLhwIUWLFgWcwHT9+vUxxtCxY0euvPJKjh07xrZt24iLi+O1117zHGtGRgYdOnRg5syZ1KxZk+7du1O0aFHmzp1Lv379WLp0KZ9//rnfWLZv3079+vWpUaMGPXr04NSpU5QsWZJevXrx4osv8uWXX/qsMHf77LPPCA0N5b777vOUrVy5ktatW5OSkkKbNm3o0qULycnJTJ48mcaNGzNp0iTatWvnqX/48GEaNmzIjh07aNy4MY0bNyYxMZHHHnuM1q1bn/X8u/Xs2ZOXXnqJcePG0b9/f7/tX3zxBdnZ2fTq1ctTNn/+fIYOHcrtt9/O3XffTfHixdm6dSsTJ05k6tSpLFq0iNq1a+f7fOUmJSWFp556ioYNG9KqVSvKlStHYmIiP/zwA+3atWPMmDE8/PDDfu0yMjJo3bo1+/fvp23btoSEhDB58mSef/550tPTiY2NBfJ/nYqIiIiIiIiIXBBrrR4X+IiLi7NxcXFW8mfnzp0WsL169Qq4fcKECRawN998sz1+/Lin/MSJE7ZevXoWsOPHj/dpA1jA3nPPPTYrK8tTvmPHDluqVClbpEgRu3379gsa92OPPWYB+/zzz/uUP/vssxawb7/9dsB2TzzxhAXy9Rk5fvy4LVOmjA0KCrLz58/32TZ06FAL2FatWvmU9+rVywK2TJkyNiEhwVOelZVlu3TpYgE7ZMgQT/nTTz9tATt58mS//aekpPicv9jYWAvYJ5980mZmZnrKMzMzbZ8+ffz6cc8tYF944QW//vfu3WuDgoJsvXr1/Lb9/vvvFrBdunTxlGVkZNjq1avbsLAwO2/ePJ/6+/btsxUrVrQVKlSw6enpnvJHHnnEAnbAgAE+9ZctW2ZDQkIsYGNjY/32H0jr1q0tYNetW+e37brrrrNFihSxycnJnrKkpCR77Ngxv7qrV6+2ERER9o477vApP9v5stbapk2bWuer+r/S09Ptnj17/Oqmpqba66+/3pYqVcqmpaX5bKtWrZoFbNu2bX22JSUl2cjISBsZGWnPnDnjN7bcrlMRERERERER+Z9WIHFVrZwuTGrfc6lHkH9rvv3Duh47dizgpHgoXry4pzwiIoI333yTli1b8tFHH9G9e3efdsHBwbz55ps+q5evvPJK+vfvz+DBg/n88889K0PP1dSpU/nwww+pXLkyzz33nM+2o0ePAhAZGRmwrbs8NTX1rPuZMmUKhw8f5v7776dJkyY+25555hlGjx7N7Nmz2b17t18u4qeeeopq1ap53gcFBfHWW28xefJkxo4dy8svv+xTPzw83G//pUqV8rzOzs5mxIgRVKhQgffee4/g4GDPtuDgYN555x0++eQTxo8fT6dOnXz6KV++fMBzXalSJVq0aMHs2bPZsGED119/vWebO+2G90rkH3/8ke3bt/Pss8/StGlTn74qVqzIc889x4ABA/jll19o164dGRkZjB8/nhIlSjBo0CCf+rfccgs9evQ4p/QevXr1YtasWYwbN4633nrLU758+XI2btzIXXfd5XOjy+jo6ID91K5dm+bNmzNr1iwyMjI8K9PdcjtfuQkLC6Ny5cp+5ZGRkfTp04dnnnmGZcuW8de//tWvzvDhw33mPjo6mk6dOvHZZ5+xefNmbrjhhnyPQ0RERERERETkQig4LYXOypUrCQoKolmzZn7bmjZtSnBwMKtWrfLbVrVqVU+eXG/NmjVj8ODBAdvkx2+//Ub37t2JiIjgu+++8wng5od15fLNmTc4kJUrVwLQvHlzv20hISH89a9/JSEhgVWrVvkFp3MGbwGuuuoqqlSpQkJCAqmpqURFRdGtWzeGDRtG586d6dq1Ky1btqRRo0ZUr17dp+2WLVs4fPgw11xzDa+99lrA8YaHhwfMpVy7dm3CwsICtunduzezZ89m3Lhx/Pvf/wacHMpfffUV5cqV80nRsXjxYgB27drlF2wG2Lp1KwDx8fG0a9eOTZs2kZaWRpMmTQL+WNCsWbNzCk7fddddREZG8sUXXzB06FBPgN7dR+/evf3a/Pjjj4wePZrly5eTnJxMZmamz/bk5GSuuOIKn7K8zlduNmzYwFtvvcX8+fNJTEwkPT3dZ/u+ffv82kRGRnL11Vf7lVepUgVwUuOIiIiIiIiIiBQUBael0Dl69CilS5cmNDTUb1tISAhly5bl4MGDftvKly8fsL8KFSp4+j1Xixcvpm3btgQFBTF9+nRuu+02vzruIGhu/btvSJjbympv7j5yBi/d3OWBVmHndfy7du3i6NGjREVFcdttt7FgwQJef/11Jk6c6MkZXbNmTWJjY7n//vsBJ3czOAHgwYMH5zrmEydOBNxnbu666y5KlizJF198wRtvvEFwcDDTpk3j8OHDDBgwwCent3sM336b90p99xjc5+9sn4X8Cg8P595772XMmDHMmjWLtm3bkpGRwZdffkm5cuVo27atT/3hw4fz1FNPUapUKVq1akXVqlUpVqwYxhgmT57MmjVrOH369AWPa8mSJTRv3pzMzExatGhBx44dKVmyJEFBQaxevZopU6YE3E9UVFTA/tznPCsr65zGISIiIiIiIiJyIRScLkz+wFQZl5PIyEhSUlICpj/IzMwkOTk54M3ikpKSAvZ34MABT7/nYsGCBbRv356goCBmzpxJgwYNAtarWbMm4Kw0DsS9urdGjRpn3ad7jO4x55SYmOhTz1tSUpJnLN4CHf9f/vIXpk2bxunTp1mxYgUzZszggw8+oHv37pQrV46WLVt66t911118//33Zx27t7xWibsDvh999BGzZ8/mjjvuCJjSw3vMU6ZMoWPHjmfdr7v+2T4L56JXr16MGTOGcePG0bZtW08g/amnnvL5fGZmZhIbG0uFChVYuXKl3w8M7lXggeRnVb231157jVOnTjF37ly/vzB44403mDJlyjn1JyIiIiIiIiJyKQSdvYrIxeVOjZDbKs2bb76Z7Oxs5s+f77dt/vz5ZGVlUbduXb9tu3fvJiEhwa983rx5nn7za86cObRt25aQkBBmz56da2Aa4Pbbbwdg1qxZZGdn+2w7fvw4ixYtIjw8PM8+3NxjdI/ZW2ZmJgsXLgQIePy//vqrX9mOHTvYs2cPMTExAVfNhoWF0bBhQ4YMGcLw4cMBPIHNWrVqERUVxZIlS8jIyDjr2M+FOx3GuHHjSE5OZvr06dx0003UqVPHp577nC1YsCBf/daqVYtixYqxevXqgCvZA53Xs2nUqBHXXHMNU6ZM4ejRo7kG0pOTk0lNTaVhw4Z+gekTJ054UrZcDNu2baN06dIBU98E+hycj7NdpyIiIiIiIiIiF0rBaSlwpUqVwhjD7t27A27v06cPAC+88AJpaWme8rS0NJ5//nkAHnroIb92WVlZ/OMf//AJEO/cuZPhw4cTEhLC3/72t3yNb9asWdx5550ULVqUX375hVtvvTXP+tWrV6d169YkJCQwcuRIn22xsbGcPHmSnj17EhERcdZ9d+7cmdKlS/Pll1+yZMkSn23vv/8+O3bsoGXLln75pgGGDRvGrl27PO+zs7MZOHAg2dnZPPjgg57yBQsWBAzculcbFytWDHBSPfTr14/ExET69+/PqVOn/NokJiaycePGsx5XTt4B31GjRpGRkREwf3OnTp2oXr06I0eO5KeffgrY1+LFiz2fkyJFitCjRw+OHz/ul6N6+fLljB8//pzHCk4gOj09nbi4OH766Sduuukmvx87oqOjKVasGCtWrPBJdZKRkcFTTz1FcnLyee07kJiYGFJSUli7dq1P+ccff8zMmTMvyj7Odp2KiIiIiIiIiFyoQpvWwxhTBrgLaA/cCFQCzgDrgE+AT6y12V71iwB9gTrAzcB1QBHgEWvtR2fZVy/gCVebLGAV8La1dtpFPiwBihcvTv369VmwYAE9evSgRo0aBAcH07FjR2666Sa6d+/OlClT+Oabb7j++uvp3LmzJ2fvzp07uffee+nRo4dfvzfddBNLly6lXr16tG7dmqNHj/L111+TmprKv//9b78b/gWyefNmOnXqRHp6Ou3atWPKlCkBUyTkDHzGxcXRsGFD+vfvzy+//MK1117L0qVLmTt3LjVq1OD111/P97kZO3Ys99xzD02bNuWee+6hatWqrFixglmzZlGhQgU+/PDDgG0bNWpEnTp16NatG5GRkcycOZM1a9ZQr149nnvuOU+9d955h1mzZtGsWTOuuuoqihcvzoYNG5g+fTqlSpXi0Ucf9dR9+eWXWbNmDaNHj+aHH36gefPmVKpUiYMHD7J161YWLVrE66+/znXXXZev4/PWs2dPXn75ZV599VVCQkLo3r27X50iRYrw/fff06ZNG9q3b0/Dhg2pU6cOxYoVY8+ePSxbtowdO3aQmJjoCar/61//4pdffuH9999n+fLlNG7cmMTERL7++mvatWvH1KlTz2usr7zyCrGxsWRkZPitmgYICgqif//+DB06lBtvvJFOnTpx5swZ5s6dS0pKCrfffjtz5849530HMmDAAGbOnEnjxo259957iYyMZPny5SxcuJCuXbsyceLEC97H2a5TEREREREREZELZq0tlA/gMcAC+4HxwBvAWCDVVT4RMF71o1zlFjgA7Ha9fvgs+3nbVW8P8B4wEjjsKnsyP2ONi4uzcXFxVvJv69at9s4777SlS5e2xhgL2E8++cSzPSsry44cOdLWq1fPhoeH2/DwcFu3bl07YsQIm5WV5dff/7N3p9F1peWB7/+PJZVsyZZsWZNdDEUlFNChGRoyQBGqKIZAVxKK6SZpaKqqGS4OYwYWvUIIRxC6c8PUgXAdyFBFkkoqWVAhDYGQ3FAMCdBIlmV5kMujLM9zWZ7kSc/9cI6UI1mDjyW7XPL/t9ZZW3vv9937Oefx+354vNe7gbztttty9+7d+cY3vjFbWlqytrY2n/vc5+YDDzxwyXE9/PDDWfbvaNLPRAYGBvKee+7J9vb2rKmpySc96Un5nve8Jw8fPlzx7/OjH/0o77rrrmxubs6ampp84hOfmO94xzty9+7dF7W9++67E8itW+ftf3YAACAASURBVLfmJz7xiXza056WtbW1uXz58nzve9+bx44dG9P+m9/8Zt5zzz35jGc8IxsaGrKuri5vueWWfPe73539/f0XXX94eDj//M//PO+4445csmRJ1tTU5PLly/PWW2/Nj33sYzkwMDDadvv27Qnk3XffPe133LFjR86bNy+B/Pmf//kp2+7fvz8/8IEP5E/8xE/kggULsr6+Pn/8x388X/e61+Vf/MVf5Llz58a037t3b957773Z3Nyc8+fPz2c/+9l53333jeb3wx/+8LTxjffSl740gayurs59+/ZN2ObcuXP5yU9+Mp/xjGfk/Pnzs62tLd/0pjdlf3//aJ62b98+2v5Sfq/bbrttwn9zX/3qV/Onf/qnc+HChdnY2Jgvf/nL8zvf+U7ed999F42nzMwnP/nJ+eQnP3nCe3z4wx9OIB9++OExx6cbp5IkSZIkac66KjXgyGKB9poTEXcA9cA/5NgnpNuBHwFPBF6fmV8uHb8BeCnQk5l7I6IAfJgpnpyOiBcC/wZsBX4yM4+Wjt8ErCrd/+mZ2T9VrCtXrkyAFStWXOa31UxFBLfddttlrSksSZIkSZIkaYy4Gje5ZteczsxvZeZXywvTpeP7gD8q7d5edvxsZn4jM/dWcJt3lLYfGylMl67VT/EJ6lrg3gn6SZIkSZIkSZJm4JotTk/jXGl7fobXuaO0/ccJzn1jXBtJkiRJkiRJ0ix53BWnI6IaeHNpd6Ki8qVep57iSxZPTPK09ebS9pbLvYckSZIkSZIkaWLVj3UAl+H3gGcCX8/Mb87gOo2l7bFJzo8cXzyDe+gquVbXTpckSZIkSZI0sapCofBYx3DJIuI9wEeAjcDPFwqF05O17ejouJ3imtRfLRQK3ROcXwT8BnC8UCh8aoLzVcCHgPOFQuF3J4nn7R0dHZ+/8847lwPccsst9PX10dzczOrVq1m+fDldXV0sXbqU9evXU1tby+7duzl16hQnT55kz5493HDDDWzYsIGWlha6u7vH9Onr66OmpoY9e/Zw8uRJTp8+za5du6itrWXdunW0tbWN6bNkyRIeeeQRqqqq2LdvH8ePH2doaIidO3eyYMECent7aW9vZ9WqVaN9Fi9ezKZNm4gIDhw4wODgIGfPnmXHjh3U1dXR29vL8uXL6ezs5MYbb6Srq4vGxka2bNlCZnLo0CGOHj3K+fPn6e/vZ+HChaxZs2b0+iP3a2hoYOvWrQwPD3PkyBEOHz7M8PAw27ZtY9GiRWP6tLW1sXr1ahYtWsT27ds5f/48R48e5dChQwBs2bKFxYsX09PTM9qntbWVNWvWUF9fT39/P+fOnePYsWMcOHCAiGDTpk0sWbJkTJ+WlhZ6e3upq6tjYGCAM2fOMDg4yP79+6mqqmLjxo0sXbqU7u7u0e/e3NzMunXrqK2tZdeuXQwNDXHixAn27t1LdXX1aP5H+nR2dtLc3MyGDRtGczk+/+vXr6e1tXVMLpuamujr66O6uno0l+PzPz6XI/mfN28e+/fvZ3BwkDNnzjAwMDCa/2XLlk2a/4MHD/Loo49y7ty5i/Lf1dU1um1oaLgo/xcuXGDbtm0sXLiQnp6e0d+rvb2d7u7uKfPf0NAwmpfOzk7a29vp6emhvr5+wvxv3rx50vzX1dWxY8cOzp49e1H+m5qaxozLlpYW1q5dy/z58xkYGGBoaIjjx4+zb9++Mfkv71Oe/5GxfOLECfbs2UNNTQ19fX0TjuXy/JeP5fL8r1q1avR3a2pqYuPGjZOO5bVr147Jf2dn54T5HxnLI/kfn8uRsQyMyf9EY3nk385I/oeHhzl8+PCY/I8fyyP5Lx/L5fnfunUrjY2NY3LZ1tY2mv/ysXzw4MEx+V+9evXoGGttbZ10LM+bN++i/I/0Gcn/zp07Jx3LzuXO5c7lzuXO5c7lzuXO5c7lzuXO5c7lzuXO5dfjXL5s2bKOyequsykeL0+cRsQ7gT8ENgAvLb0Ycar2BeDDwNsy808mOF8PnKC4rMeiCc43AweBA5nZNtW9Vq5cmQArVqy4tC8jSZIkSZIkSdeuuBo3eVysOR0R76NYmF4HvGS6wvSlyMyTwG5gYUQsm6DJU0vbTTO9lyRJkiRJkiRprGu+OB0RHwA+DfRQLEwfmMXLf6u0feUE5141ro0kSZIkSZIkaZZc08XpiPgQxRcgrqK4lMehWb7FH5W2H4yIJWX3vQl4J3AGuG+W7ylJkiRJkiRJ173qxzqAyUTE3RRffngB+B7wnoiLljrpz8z7y/r8d+Dppd3nlLb3RsSLSn//a/n605n5/Yj4FPDrQG9EfAm4AfgloAl4d2b2z+b3kiRJkiRJkiRdw8Vp4CmlbRXwvknafAe4v2z/lcBt49q8sPQZMebliJn5GxHRC7wLeDswDHQDH8/Mr11W5JIkSZIkSZKkKV2zxenMLACFCvvcfpn3+iLwxcvpK0mSJEmSJEmq3DW95rQkSZIkSZIkaW6yOC1JkiRJkiRJuuosTkuSJEmSJEmSrjqL05IkSZIkSZKkq87itCRJkiRJkiTpqrM4LUmSJEmSJEm66ixOS5IkSZIkSZKuOovTkiRJkiRJkqSrrvpyOkVEPXAj0AycBg5k5u7ZDEySJEmSJEmSNHddcnE6In4M+G/Ay4HnMu6p64g4DDwMfBl4KDPPz2KckiRJkiRJkqQ5ZNridEQ8H/hd4GUUC9LngLXAPuAIsABYCjwNeAPweuBgRPwB8KnMPHNlQpckSZIkSZIkPV5NWZyOiC8CbwSOAV8AHgR+lJlDk7S/Cfg54G7gY8A7IuK/ZuZ3ZzFmSZIkSZIkSdLj3HQvRHwF8OvA8sz81cz87mSFaYDM7M/Mz2fmC4HnAN3AbbMXriRJkiRJkiRpLphuWY+bM/P05Vw4M3uB10TEgsvpL0mSJEmSJEmau6Z8cvpyC9OzfQ1JkiRJkiRJ0twy3bIelyQiaiLiuRHxtNm4niRJkiRJkiRpbquoOB0R/1dE/G1ENJUd+zFgPdAFbIiIhyJiuuVCJEmSJEmSJEnXsUqfnP5vwNMz80jZsU8CPw48DPQCrwbunZ3wJEmSJEmSJElzUaXF6f8AdI7sREQD8J+Bv83MlwE/BWzE4rQkSZIkSZIkaQqVFqdbgL1l+y8AqoEHATLzHPDPwI/NSnSSJEmSJEmSpDmp0uL0caCxbP82IIF/LTs2BCyaYVySJEmSJEmSpDms0hcXbgZeFRG1FIvSbwB6M/NQWZsnAwdmKT5JkiRJkiRJ0hxU6ZPTXwBuplik7iv9/Wfj2vw0sH7moUmSJEmSJEmS5qqKitOZ+UXg94A6ist7/GHpA0BE3AHcBDw8eyFKkiRJkiRJkuaaSpf1IDN/C/itSU7/K7AEODmToCRJkiRJkiRJc1vFxempZOZZ4OxsXlOSJEmSJEmSNPdMuaxHRCyY6Q1m4xqSJEmSJEmSpLllujWnt0fEeyOittILR8SzI+Lvgd+8vNAkSZIkSZIkSXPVdMXpfwI+BeyNiJUR8ZKpnoSOiJsjYkVE/ADoBp6NL0eUJEmSJEmSJI0z5ZrTmfnmiPgM8D+At5c+FyKiD9gLHAXmA0uBpwHNQAD7gQ8Cn87MM1cufEmSJEmSJEnS49G0L0TMzC7gFRHxVOAtwEuB5wD/cVzTg8BDwJeBL2fmuVmOVZIkSZIkSZI0R0xbnB6RmZuB/w4QEXXAjRSfmD4NHMjMvVckQkmSJEmSJEnSnHPJxelymXkK2Fz6SJIkSZIkSZJUkeleiPiYiYilEfHWiPi7iNgSEacj4lhE/GtEvCUiJow9Il4YEV+PiCMRcSoieiPifRFRNUHb2yMip/j83pX/ppIkSZIkSZJ0/bmsJ6evkjcAKym+ePFhYABoA14L/Anwqoh4Q2bmSIeIeDXFNa+HgL8BjgC/AHwauLV0zYl8B/j2BMf/dTa+iCRJkiRJkiRprGu5OL0J+EXgHzJzeORgRPwW8CPgdRQL1V8uHW8A/hi4ANxeepEjEfEh4FvA6yPilzPzwQnu9e3MLFzB7yJJkiRJkiRJKnPNLuuRmd/KzK+WF6ZLx/cBf1Tavb3s1OuBFuDBkcJ0qf0Q8Nul3RVXLmJJkiRJkiRJ0qW6lp+cnsq50vZ82bE7Stt/nKD9d4FTwAsjojYzz4w7/+MR8S6gAdgHfC8zfdmjJEmSJEmSJF0hj7vidERUA28u7ZYXop9W2m4a3yczz0fEduAngJuBvnFN3lj6lN/ny8DbMvPobMQtSZIkSZIkSfp3VYVC4ZIbR8STOjo6slAojH/yuLzNoo6OjvZCoXBsNgIcr6Oj4/eBu4CvZ+ZHyo6/H1gKfLxQKByZoN8bgScCDxQKhV2lY/UUX7T4a8AHKL448WGKRew7gBd1dHTcXygUcvz1ACLi7R0dHZ+/8847lwPccsst9PX10dzczOrVq1m+fDldXV0sXbqU9evXU1tby+7duzl16hQnT55kz5493HDDDWzYsIGWlha6u7vH9Onr66OmpoY9e/Zw8uRJTp8+za5du6itrWXdunW0tbWN6bNkyRIeeeQRqqqq2LdvH8ePH2doaIidO3eyYMECent7aW9vZ9WqVaN9Fi9ezKZNm4gIDhw4wODgIGfPnmXHjh3U1dXR29vL8uXL6ezs5MYbb6Srq4vGxka2bNlCZnLo0CGOHj3K+fPn6e/vZ+HChaxZs2b0+iP3a2hoYOvWrQwPD3PkyBEOHz7M8PAw27ZtY9GiRWP6tLW1sXr1ahYtWsT27ds5f/48R48e5dChQwBs2bKFxYsX09PTM9qntbWVNWvWUF9fT39/P+fOnePYsWMcOHCAiGDTpk0sWbJkTJ+WlhZ6e3upq6tjYGCAM2fOMDg4yP79+6mqqmLjxo0sXbqU7u7u0e/e3NzMunXrqK2tZdeuXQwNDXHixAn27t1LdXX1aP5H+nR2dtLc3MyGDRtGczk+/+vXr6e1tXVMLpuamujr66O6uno0l+PzPz6XI/mfN28e+/fvZ3BwkDNnzjAwMDCa/2XLlk2a/4MHD/Loo49y7ty5i/Lf1dU1um1oaLgo/xcuXGDbtm0sXLiQnp6e0d+rvb2d7u7uKfPf0NAwmpfOzk7a29vp6emhvr5+wvxv3rx50vzX1dWxY8cOzp49e1H+m5qaxozLlpYW1q5dy/z58xkYGGBoaIjjx4+zb9++Mfkv71Oe/5GxfOLECfbs2UNNTQ19fX0TjuXy/JeP5fL8r1q1avR3a2pqYuPGjZOO5bVr147Jf2dn54T5HxnLI/kfn8uRsQyMyf9EY3nk385I/oeHhzl8+PCY/I8fyyP5Lx/L5fnfunUrjY2NY3LZ1tY2mv/ysXzw4MEx+V+9evXoGGttbZ10LM+bN++i/I/0Gcn/zp07Jx3LzuXO5c7lzuXO5c7lzuXO5c7lzuXO5c7lzuXO5dfjXL5s2bKO6eq0syEyJ6y7Ttw44gJQyMyPTtHmg8BHMrNqFuIbf+33AH8AbARuzcwjZec2AU8FnpqZWybo+33gBcALMvOH09ynAegBngLclZl/P1X7lStXJsCKFS5pLUmSJEmSJOlxL67GTSp9IWJwlQK76MYR76RYmN4AvKS8MF0y8qR24ySXaBjXblKZOQj8VWn3xRWGKkmSJEmSJEmaRqXF6UvRBpyczQtGxPuAPwTWUSxM75ug2SOl7S0T9K+m+BT0eWDbJd72YGlbX1m0kiRJkiRJkqTpTPtCxIh487hDz5ngGEAV8CTgvwJrZyG2kft/APg9istsvDwzD03S9FsUX2r4SuCvx517MVAHfDczJ10ve5yfKW0vtZgtSZIkSZIkSbpE0xangfuBkYWpE3h16TPeyHIfp4BZWTA7Ij4EfARYBbxigqU8yn0J+H+AX46Iz2ZmV+ka84HfLbVZOe76twI/yMzhccffBPwScBb429n4LpIkSZIkSZKkf3cpxel7S9sA/gz4CjDRCwIvAIcpFnsfnWlgEXE3xcL0BeB7wHsiLlruuj8z74fiOtER8TaKRepvR8SDwBHgF4GnlY7/zbj+DwDzSi9L3AXMB34S+CmKS4D835nZP9PvIkmSJEmSJEkaa9ridGZ+ceTvUsH4K5n551c0qqKnlLZVwPsmafMdik92A5CZX4mI24APAq+jWGzeAvw68JnMzHH9VwIvA24FmikW4HeXrvm/MnPNbHwRSZIkSZIkSdJYcXG9VpVauXJlAqxYseKxDkWSJEmSJEmSZuqiJSyuhHmVNI6I/xQRvxMRbZOcby+df87shCdJkiRJkiRJmosqKk4Dvwm8FTgwyfn9wFsoLqMhSZIkSZIkSdKEKi1OvwB4eIK1mwEoHf8WxTWcJUmSJEmSJEmaUKXF6XZg1zRt9gDLLi8cSZIkSZIkSdL1oNLi9CmgZZo2LcCZywtHkiRJkiRJknQ9qLQ43QO8OiIWTnQyIhqAV5faSZIkSZIkSZI0oUqL01+g+GT0P0fEs8pPRMSzgX8CmkvtJEmSJEmSJEmaUHUljTPzbyLiVcCbgdURsR/YDdwItAEBfDEz/3rWI5UkSZIkSZIkzRmVPjlNZt4DvAPYQPEFic8rbdcDb8/Me2czQEmSJEmSJEnS3FPRk9MjMvMLwBciog5YDDyamadmNTJJkiRJkiRJ0px1WcXpEaWCtEVpSZIkSZIkSVJFLqs4HREtwOuAZwD1mfnWsuNPAdZm5ulZi1KSJEmSJEmSNKdUXJyOiLcAnwHmU3wBYgJvLZ1uA34AvB3401mKUZIkSZIkSZI0x1T0QsSIeDnwBWAT8BpgZfn5zFxH8cWId81WgJIkSZIkSZKkuafSJ6c/AOwFbsvMwYh47gRteoEXzDgySZIkSZIkSdKcVdGT08Dzga9l5uAUbXYB7ZcfkiRJkiRJkiRprqu0OH0DcHKaNouBC5cXjiRJkiRJkiTpelBpcbofeN40bX4aeOSyopEkSZIkSZIkXRcqLU7/PfCzEfGGiU5GxL3As4AvzzQwSZIkSZIkSdLcVekLEX8f+GXgryPi9UAjQES8C/hZ4LXAZuCzsxmkJEmSJEmSJGluqag4nZlHI+I24M+B8qenP1Pafg/4L5k53brUkiRJkiRJkqTrWKVPTpOZA8DtEfEs4AXAUuAY8MPMXDXL8UmSJEmSJEmS5qApi9MR8RDwYGb+bWn/xUB/Zg5kZi/QexVilCRJkiRJkiTNMdO9EPEu4Oll+w8D91yxaCRJkiRJkiRJ14XpitOPAg1l+3EFY5EkSZIkSZIkXSemW3O6D/iViOgE9paO3VRa3mNKmfndmQYnSZIkSZIkSZqbpitOdwBfAf6q7Njdpc90qi43KEmSJEmSJEnS3DZlcToz/ykingG8DLgRKADfKX0kSZIkSZIkSbos0z05TWbuAP4UICIKwLcz8yNXOC5JkiRJkiRJ0hw25QsRI+JTEfGKskP3UlzmQ5IkSZIkSZKkyzZlcRp4H/AzZft/Btx15cKRJEmSJEmSJF0PpitOnwDqyvbjCsYiSZIkSZIkSbpOTLfm9BbgtRHxd8De0rHFEfGk6S6cmQMzDU6SJEmSJEmSNDdN9+T0x4EfA74PbC8de2/p76k+22YaWEQsjYi3RsTfRcSWiDgdEcci4l8j4i0RMWHsEfHCiPh6RByJiFMR0RsR74uIqinu9fMR8e3S9U9ExP+JiLtn+h0kSZIkSZIkSROb8snpzPzriNgO3AncCNwD9AI9Vz403gCspPjE9sPAANAGvBb4E+BVEfGGzMyRDhHxauDLwBDwN8AR4BeATwO3lq45RkS8C/gscBj4S+As8Hrg/oj4j5n5m1fqC0qSJEmSJEnS9SrKarvTN44YBgqZ+ZErF9Love4A6oF/yMzhsuPtwI+AJwKvz8wvl443UFyGpBG4NTO7SsfnA98CXgD8SmY+WHatm4CNwEngeZnZXzq+BOik+NT4CzPzB1PFunLlygRYsWLFTL+2JEmSJEmSJD3Wrsq7B6db1mO8e4GvXIlAxsvMb2XmV8sL06Xj+4A/Ku3eXnbq9UAL8OBIYbrUfgj47dLu+OrxfwNqgT8cKUyX+hwF/kdp9x0z+yaSJEmSJEmSpPGmeyHiGJn5xSsVSIXOlbbny47dUdr+4wTtvwucAl4YEbWZeeYS+nxjXBtJkiRJkiRJ0iyp9MlpImJeRLw7In5YeoHg+bJzz42I/zcibpndMMfcvxp4c2m3vKj8tNJ20/g+mXme4osaq4GbL7HPXorLfTwhIupmGLYkSZIkSZIkqUxVoVC45MYRcQPwTYpLXSyg+DTyokKh8BGAjo6OIeABYLBQKHx7toMt3eP3gbuAr5evfd3R0fF+YCnw8UKhcGSCfm+kuE71A4VCYVfp2MeAKuCDhUJheII+7wQWAZ8rFAonxp+PiLd3dHR8/s4771wOcMstt9DX10dzczOrV69m+fLldHV1sXTpUtavX09tbS27d+/m1KlTnDx5kj179nDDDTewYcMGWlpa6O7uHtOnr6+Pmpoa9uzZw8mTJzl9+jS7du2itraWdevW0dbWNqbPkiVLeOSRR6iqqmLfvn0cP36coaEhdu7cyYIFC+jt7aW9vZ1Vq1aN9lm8eDGbNm0iIjhw4ACDg4OcPXuWHTt2UFdXR29vL8uXL6ezs5Mbb7yRrq4uGhsb2bJlC5nJoUOHOHr0KOfPn6e/v5+FCxeyZs2a0euP3K+hoYGtW7cyPDzMkSNHOHz4MMPDw2zbto1FixaN6dPW1sbq1atZtGgR27dv5/z58xw9epRDhw4BsGXLFhYvXkxPT89on9bWVtasWUN9fT39/f2cO3eOY8eOceDAASKCTZs2sWTJkjF9Wlpa6O3tpa6ujoGBAc6cOcPg4CD79++nqqqKjRs3snTpUrq7u0e/e3NzM+vWraO2tpZdu3YxNDTEiRMn2Lt3L9XV1aP5H+nT2dlJc3MzGzZsGM3l+PyvX7+e1tbWMblsamqir6+P6urq0VyOz//4XI7kf968eezfv5/BwUHOnDnDwMDAaP6XLVs2af4PHjzIo48+yrlz5y7Kf1dX1+i2oaHhovxfuHCBbdu2sXDhQnp6ekZ/r/b2drq7u6fMf0NDw2heOjs7aW9vp6enh/r6+gnzv3nz5knzX1dXx44dOzh79uxF+W9qahozLltaWli7di3z589nYGCAoaEhjh8/zr59+8bkv7xPef5HxvKJEyfYs2cPNTU19PX1TTiWy/NfPpbL879q1arR362pqYmNGzdOOpbXrl07Jv+dnZ0T5n9kLI/kf3wuR8YyMCb/E43lkX87I/kfHh7m8OHDY/I/fiyP5L98LJfnf+vWrTQ2No7JZVtb22j+y8fywYMHx+R/9erVo2OstbV10rE8b968i/I/0mck/zt37px0LDuXO5c7lzuXO5c7lzuXO5c7lzuXO5c7lzuXO5dfj3P5smXLOiav0M6eSl+I+EHgo0AB+F3gd4APZWZVWZt/Ahoy82dmN1SIiPcAf0DxJYa3ZuaRsnObgKcCT83MLRP0/T7FlyK+IDN/WDp2FqgBakpPV4/vswdYBiwrrXU9IV+IKEmSJEmSJGkOuSZfiPhG4N8y8yOlFxVOVNneDjxpxpGNExHvpFiY3gC8pLwwXXKstG2c5BIN49pV0mewglAlSZIkSZIkSdOotDj9FOCH07Q5AjRdXjgTi4j3AX8IrKNYmJ7oKeZHStuL1rsurVP9FIovUNx2iX2WAfXArsw8dfnRS5IkSZIkSZLGq7Q4fRpYPE2bJwGPXl44F4uIDwCfBnooFqYPTNL0W6XtKyc492KgDvh+Zp65xD6vGtdGkiRJkiRJkjRLKi1O9wCvKL0Y8SIR0Qj8HPCjmQZWut6HgN8DVgEvzcxDUzT/EnAI+OWIeH7ZNeZTXB8bYOW4PvcBZ4B3RcRNZX2WAL9V2v2jGXwFSZIkSZIkSdIEqits/8fAA8ADEfGW8hMRsZhisXcJs1DQjYi7gY8AF4DvAe+JuGgd7v7MvB8gMwcj4m0Ui9TfjogHKS4x8ovA00rH/6a8c2Zuj4j3A58BuiLib4CzwOuBJwCfzMwfzPS7SJIkSZIkSZLGqqg4nZl/HREvA+6lWPQ9ChARXcBPALXA5zLz67MQ21NK2yrgfZO0+Q5wf1l8X4mI24APAq8D5gNbgF8HPpOZF73AMTM/GxH9wG8Cb6b4NPkG4Lcz84uz8D0kSZIkSZIkSePEBPXa6TtF3AO8F3gWMPI483rgU5l536xF9zixcuXKBFixYsVjHYokSZIkSZIkzdRFS1hcCZUu6wFAaSmN+yNiAcVlPI5l5snZDEySJEmSJEmSNHddVnF6RGaeBk7PUiySJEmSJEmSpOvEvMc6AEmSJEmSJEnS9cfitCRJkiRJkiTpqrM4LUmSJEmSJEm66ixOS5IkSZIkSZKuOovTkiRJkiRJkqSrzuK0JEmSJEmSJOmqq56Ni0REM/CzwCng/8vMC7NxXUmSJEmSJEnS3FTRk9MRsSIi/k9ENJUdex7QB3wJ+Drw/Yion90wJUmSJEmSJElzSaXLevwSkJl5pOzYx4ElwH0Ui9M/CbxjdsKTJEmSJEmSJM1FlRannwr0juyUlvO4DfjTzHxrZv4C0An8l9kLUZIkSZIkSZI011RanF4KHCjbv7W0/buyY98DnjyToCRJkiRJkiRJc1ulxekjQHPZ/m3AMPD9smMJzJ9hXJIkSZIkSZKkOazS4nQf8AsRsTQiFlNcg7ozMwfL2twE7Jul+CRJkiRJkiRJc1B1he3/APgKsAs4D9QBHxg5GRFVwIsY+yS1riG7TiXrjg1z+gIsqIJnNs7jCXXxWIclSZIkSZIk6TpTUXE6M/93RLwDeHvp0AOZ+ZdlTV5GcUmPb85SfJol644N89CuYTYez3Fnhnn6ouC1T5jHMxsrfZBekiRJkiRJki5PpU9Ok5lfAL4wyblvAktmGpRm18MHhvmTbRcYX5YeZpOdPgAAIABJREFUsfF48j/7LvC2m+H2VgvUkiRJkiRJkq48K5Fz3LpjUxemRyTwx9susO7Y8NUIS5IkSZIkSdJ1rqLidEQ8NyJ+NSIay47VR8QXI+LRiNgTEe+d/TB1uR7aNTxtYXpEAn+3y+K0JEmSJEmSpCuv0ienPwB8MDOPlR37n8B/LV1rKfCpiHjFLMWnGdh1KidYY3pqfceTXacq6yNJkiRJkiRJlap0zennA98e2YmIGuBu4EfA7UATsBp4D/BPsxKhLtvlLtGx7tgwT6irmuVoJEmSJEnS9eLjG89Pef79T6/4NWiS5qBKn5xuBXaW7T8fWAR8PjOHMnMP8PfAs2YpPs3A6QtXt58kSZIkSdJ4R8/m6EeSylX631Q5rs+LSse+U3bsINAyw7g0CxZc5sPPl9tPkiRJkiQJxj4ZXf4UtU9MSypX6ZPTA8DPlO2/GtiVmdvKji0Hjs40MM3cMxsrTe/M+kmSJEmSJEnSpaq0Cvm3wAsj4ksR8ZfAC4AvjWvzTGDrbASnmXlCXfD0RVFRn2csCp5QV1kfSZIkSZIkSapUpcXpTwM/AF4L/BdgDfCRkZMR8R+A5zF2mQ89hl77hHlcaqk5gNc8waemJUmSJEmSJF15FVUiM/NEZt5K8YWHzwKen5nHypqcAl4DrJy9EDUTz2ycx1tvrpq2QB3A226uckkPSZIkSZIkSVfFZa1Cn5nrJjneD/TPIB5dAS9pnUdLLfzdrmH6jl/8ZtxnLApe84R5FqYlSZIkSZIkXTWXVZyOiBrgpcAzgIWZ+dHS8flAA3AoM4dnLUrN2DMbi8XnXaeSP9h8nuGEeQHvfWq1a0xLkiRJkiRJuuoqLk5HxCuBPwXaKa4GkcBHS6efA/wb8Cbgr2cpRs2iJ9QFrbUxZl+SJEmSJEmSrraK1nGIiOcDX6FYkP414K/Kz2fmD4HtFNedliRJkiRJkiRpQpUuMvwhii89fH5mfgbYPEGbTuDZMw1MkiRJkiRJkjR3VVqcvhX4Smbum6LNTmDZ5Yf07yLi9RHx2Yj4XkQMRkRGxF9O0X5hRHw0IvoiYigiHo2If4mI/zxJ+3tK15zs847Z+B6SJEmSJEmSpLEqXXN6IXBomjZ1VF70nsxvU3wK+wSwC3j6ZA0jYjHwPeCZwHrg80A98IvAP0TEe0tPe0/k74GeCY53XX7okiRJkiRJkqTJVFqc3g38xDRtngNsu7xwLvJrFIvSW4DbgIenaFugWJh+CPilzDwPEBEtwI+AT0TENzJzoqVIvpKZ989SzJIkSZIkSZKkaVT6hPM3gJ+LiBdNdDIiXgW8EPjaTAMDyMyHM3NzZuYlNH9tafs7I4Xp0jUOAp8EagCX6ZAkSZIkSZKka0ClT07/T+CXgX+KiM8CNwFExJ3Ai4F3AnuBT81ijJeqvbSd6KntkWMvnaTvcyLifcB8ik+HP5yZu2Y5PkmSJEmSJElSSUXF6czcHRGvAP4WeH/Zqf8NBLAVeG1mTrcu9ZVwiOKLGJ8CbBh37ubSdrI1q987bv9CRPwJ8L7MHJq9ECVJkiRJkiRJAFWFQqGiDoVCYW9HR8fngG5gB7AR+DbwaeDdmblvlmMEoKOj4ybgHmBtoVB4aILzTwOeB7R2dHQ8VCgUEiAilgL3AUuA6o6Ojt8vFArnS30agVUU17b+78DngP8DPB94OfBjhULhy5PFFBFv7+jo+Pydd965HOCWW26hr6+P5uZmVq9ezfLly+nq6mLp0qWsX7+e2tpadu/ezalTpzh58iR79uzhhhtuYMOGDbS0tNDd3T2mT19fHzU1NezZs4eTJ09y+vRpdu3aRW1tLevWraOtrW1MnyVLlvDII49QVVXFvn37OH78OENDQ+zcuZMFCxbQ29tLe3s7XV2rGF6yjJrt3TzrxiVs2rSJiODAgQMMDg5y9uxZduzYQV1dHb29vSxfvpzOzk5uvPFGurq6aGxsZMuWLWQmhw4d4ujRo5w/f57+/n4WLlzImjVrRmNqb29n1apVNDQ0sHXrVoaHhzly5AiHDx9meHiYbdu2sWjRojF92traWL16NYsWLWL79u2cP3+eo0ePcuhQ8f88tmzZwuLFi+np6Rnt09raypo1a6ivr6e/v59z585x7NgxDhw4QESwadMmlixZMqZPS0sLvb291NXVMTAwwJkzZxgcHGT//v1UVVWxceNGli5dSnd39+h3b25uZt26ddTW1rJr1y6GhoY4ceIEe/fupbq6ejT/I306Oztpbm5mw4YNo7kcn//169fT2to6JpdNTU309fVRXV09msvx+R/5bcfnf968eezfv5/BwUHOnDnDwMDAaP6XLVs2ps/ixYtH83/w4EEeffRRzp07d1H+u7q6RrcNDQ0X5f/ChQts27aNhQsX0tPTM/p7tbe3093dPWX+GxoaRvPS2dlJe3s7PT091NfXT5j/zZs3T5r/uro6duzYwdmzZy/Kf1NT05hx2dLSwtq1a5k/fz4DAwMMDQ1x/Phx9u3bNyb/5X3K8z8ylk+cOMGePXuoqamhr69vwrFcnv/ysVye/1WrVo3+bk1NTWzcuHHSsbx27dox+e/s7Jww/yNjeST/43M5MpaBMfmfaCyP/NsZyf/w8DCHDx8ek//xY3kk/+VjuTz/W7dupbGxcUwu29raRvNfPpYPHjw4Jv+rV68eHWOtra2TjuV58+ZdlP+RPiP537lz56Rj+Vqeyycby87lzuXO5c7lzuXO5c7lzuXO5c7l19pc/sjJoGbvJobrl1A7sMa53LncufxxMJcvW7asY6pa7WyJS1vO+bEXEbdTfCHiA5n5pgnOtwM/BJ4MrAP+BagDXg0cp/hUdR1Qm5lnp7nXE4E1FAvaz8nMNVO1X7lyZQKsWLGisi/1GPn4xtEluXn/0ytd2UWSJEmSJOnSWYeQHpfiatyk0hciXrNKT2z/JPAZoB74VYqF6a8BLwMWAMemK0yXrrUT+Hpp98VXJGBJkiRJkiRJuo5d1n9XRcTzgZ+i+GRx1QRNMjM/OpPALkdmHqS4fvSYNaQj4iUUq/2dFVzuYGlbPzvRSZIkSZIkSZJGVFScjogG4CFgpNg7mQSuenF6Cm8rbR+ooM9Pl7bbZjkWSZIkSZIkSbruVfrk9MeBO4DvUXzJ4E7g/JQ9rpKImAfUZeaJccffCvwK0MO44nRE/Gxmfm/csaD4csQXAIeAf7yScUuSJEmSJEnS9ajS4vSrgW7gJZk5fAXiGSMi7gLuKu22l7YviIj7S38fyszfLP1dB+yPiH8GtpSO/SzF5Ue2Aq/JzHPjbvHdiNhEcbmP3UAjcCvwTOAU8MbMHJzdbyVJkiRJkiRJqrQ43Qj8xdUoTJc8B7h73LGbSx+AHcBIcfoM8CDwIuDlpWNbgQ8Dnxr/RHXJJygWr+8AmoBhYAD4XKmPS3pIkiRJkiRJ0hVQaXF6M9B2JQKZSGYWgMIltj0HvKXC67+/8qgkSZIkSZIkSTM1r8L2nwN+ISJuvBLBSJIkSZIkSZKuD5U+Of0Niktg/FtEdACrgEcnapiZAzOMTZIkSZIkSZI0R1VanO4HEgjgT6Zol5dxbUmSJEmSJEnSdaLSAvKfUyw8S5IkSZIkSZJ02SoqTmfmPVcoDkmSJEmSJEnSdaTSFyJKkiRJkiRJkjRjFRWnI+JCRHxomjYfjIjzMwtLkiRJkiRJkjSXVfrkdJQ+l9JOkiRJkiRJkqQJXYllPZYAQ1fgupIkSZIkSZKkOWLaFyJGxIvHHbppgmMAVcCTgDcCj8xCbJIkSZIkSZKkOWra4jTwbSBLfydwd+kzkQCGgd+YcWSSJEmSJEmSpDnrUorTH6FYlA7gdygWq78zQbsLwGHg4czcOFsBSpIkSZIkSZLmnmmL05lZGPk7Iu4GvpKZn7mSQUmSJEmSJEmS5rZLeXJ6VGY+5UoFIkmSJEmSJEm6fsx7rAOQJEmSJEmSJF1/pnxyOiK+RekliJm5q7R/KTIzXzrj6CRJkiRJkiRJc9J0y3rcXtrWjdufTl5OMJIkSZIkSZKk68OUxenMnDfVviRJkiRJkiRJl2PKYnNEvCcifupqBSNJkiRJkiRJuj5M9yT0/wJeObITERci4kNXNiRJkiRJkiRJ0lw3XXF6CKgt24/SR5IkSZIkSZKkyzZdcXo78HMR0VZ2zJcdSpIkSZIkSZJmZLri9OeB/wTsiYgLpWOF0vIeU33OX9mwJUmSJEmSJEmPZ9VTnczMz0TEAeBOYDnwEmAA6L/yoUmSJEmSJEmS5qopi9MAmfkg8CBARAwD92XmR650YJIkSZIkSZKkuWu6ZT3G6wC+fQXikCRJkiRJkiRdR6Z9crpcZnZcqUAkSZIkSZIkSdePSp+cliRJkiRJkiRpxixOS5IkSZIkSZKuOovTkiRJkiRJkqSrzuK0JEmSJEmSJOmqszgtSZIkSZIkSbrqLE5LkiRJkiRJkq66a7o4HRGvj4jPRsT3ImIwIjIi/nKK9gsj4qMR0RcRQxHxaET8S0T852nuc3dE/CgiTkTEsYj4dkT8/Ox/I0mSJEmSJEkSXOPFaeC3gXcBzwF2T9UwIhYDPyj1uQB8HvgS8B+Bf4iI90zS7xPA/cAy4I+Bvyz1+WpEvGtWvoUkSZIkSZIkaYzqxzqAafwasAvYAtwGPDxF2wLwTOAh4Jcy8zxARLQAPwI+ERHfyMzNIx0i4oXAbwBbgZ/MzKOl4x8HVpX6fC0z+2f5e0mSJEmSJEnSde2afnI6Mx/OzM2ZmZfQ/LWl7e+MFKZL1zgIfBKoAd4xrs/I/sdGCtOlPv3A54Ba4N7LDF+SJEmSJEmSNIlrujhdofbSdtsE50aOvXTc8TtK23+coM83xrWRJEmSJEmSJM2SuVScPlTaPmWCczeXtk8fORAR9cCNwInM3DtBn5HlP26ZtQglSZIkSZIkSQBUFQqFxzqGS9LR0XETcA+wtlAoPDTB+acBzwNaOzo6HioUCgkQEUuB+4AlQHVHR8fvFwqF8x0dHS0U15s+WCgUPjXB9c4BHwJqCoXC704UU0S8vaOj4/N33nnncoBbbrmFvr4+mpubWb16NcuXL6erq4ulS5eyfv16amtr2b17N6dOneLkyZPs2bOHG264gQ0bNtDS0kJ3d/eYPn19fdTU1LBnzx5OnjzJ6dOn2bVrF7W1taxbt462trYxfZYsWcIjjzxCVVUV+/bt4/jx4wwNDbFz504WLFhAb28v7e3tdHWtYnjJMmq2d/OsG5ewadMmIoIDBw4wODjI2bNn2bFjB3V1dfT29rJ8+XI6Ozu58cYb6erqorGxkS1btpCZHDp0iKNHj3L+/Hn6+/tZuHAha9asGY2pvb2dVatW0dDQwNatWxkeHubIkSMcPnyY4eFhtm3bxqJFi8b0aWtrY/Xq1SxatIjt27dz/vx5jh49yqFDxf9/2LJlC4sXL6anp2e0T2trK2vWrKG+vp7+/n7OnTvHsWPHOHDgABHBpk2bWLJkyZg+LS0t9Pb2UldXx8DAAGfOnGFwcJD9+/dTVVXFxo0bWbp0Kd3d3aPfvbm5mXXr1lFbW8uuXbsYGhrixIkT7N27l+rq6tH8j/Tp7OykubmZDRs2jOZyfP7Xr19Pa2vrmFw2NTXR19dHdXX1aC7H53/ktx2f/3nz5rF//34GBwc5c+YMAwMDo/lftmzZmD6LFy8ezf/Bgwd59NFHOXfu3EX57+rqGt02NDRclP8LFy6wbds2Fi5cSE9Pz+jv1d7eTnd395T5b2hoGM1LZ2cn7e3t9PT0UF9fP2H+N2/ePGn+6+rq2LFjB2fPnr0o/01NTWPGZUtLC2vXrmX+/PkMDAwwNDTE8ePH2bdv35j8l/cpz//IWD5x4gR79uyhpqaGvr6+Ccdyef7Lx3J5/letWjX6uzU1NbFx48ZJx/LatWvH5L+zs3PC/I+M5ZH8j8/lyFgGxuR/orE88m9nJP/Dw8McPnx4TP7Hj+WR/JeP5fL8b926lcbGxjG5bGtrG81/+Vg+ePDgmPyvXr16dIy1trZOOpbnzZt3Uf5H+ozkf+fOnZOO5Wt5Lp9sLDuXO5c7lzuXO5c7lzuXO5c7lzuXX2tz+SMng5q9mxiuX0LtwBrncudy5/LHwVy+bNmyjsnqtLMpLm0558deRNxO8YWID2TmmyY43w78EHgysA74F6AOeDVwHFhW2q/NzLMRsRzYDezOzCdMcL0a4CxwJjPnTxXbypUrE2DFihWX/f2upo9vHF2Sm/c//Vp/J6YkSZIkSXo8sw4hPS7F1bjJnFnWIzP3AT8JfAaoB36VYmH6a8DLgAXAscw8W+pyrLRtnOSSjePaSZIkSZIkSZJmyZwpTgNk5sHMfG9m3pyZN2RmW2a+heI61AF0lrU9SfHJ6YURsWyCyz21tN10xQOXJEmSJEmSpOvMnCpOT+Ftpe0D445/q7R95QR9XjWujSRJkiRJkiRplsyZ4nREzIuIhRMcfyvwK0APFxen/6i0/WBELCnrcxPwTuAMxZcpSpIkSZIkSZJm0TW9Cn1E3AXcVdptL21fEBH3l/4+lJm/Wfq7DtgfEf8MbCkd+1ngp4CtwGsy81z59TPz+xHxKeDXgd6I+BJwA/BLQBPw7szsn/UvJkmSJEmSJEnXuWu6OA08B7h73LGbSx+AHcBIcfoM8CDwIuDlpWNbgQ8Dn8rMExPdIDN/IyJ6gXcBbweGgW7g45n5tVn6HpIkSZIkSZKkMtd0cTozC0DhEtueA95ymff5IvDFy+krSZIkSZIkSarcnFlzWpIkSZIkSZL0+GFxWpIkSZIkSZJ01VmcliRJkiRJkiRddRanJUmSJEmSJElXncVpSZIkSZIkSdJVZ3FakiRJkiRJknTVWZyWJEmSJEmSJF11FqclSZIkSZIkSVedxWlJkiRJkiRJ0lVncVqSJOn/b+++w+woy8aPf5/t2U3fdBKSkNBC770qTUABUV+k2lFB1JfXgiCJvaGiKOJPXvIqigoiqHQQBClCGgkppFfS2ybZlj3n+f0xZzebZTfZhN1zdjffz3XNNXtmnpm5z8595sy5z5xnJEmSJElZZ3FakiRJkiRJkpR1FqclSZIkSZIkSVlncVqSJEmSJEmSlHUWpyVJkiRJkiRJWWdxWpIkSZIkSZKUdRanJUmSJEmSJElZZ3FakiRJkiRJkpR1FqclSZIkSZIkSVlncVqSJEmSJEmSlHUWpyVJkiRJkiRJWWdxWpIkSZIkSZKUdRanJUmSJEmSJElZZ3FakiRJkiRJkpR1FqclSZIkSZIkSVlncVqSJEmSJEmSlHUWpyVJkiRJkiRJWWdxWpIkSZIkSZKUdRanJUmSJEmSJElZZ3FakiRJkiRJkpR1FqclSZIkSZIkSVlncVqSJEmSJEmSlHUWpyVJkiRJkiRJWdehi9MhhEtDCD8PIbwQQqgIIcQQwr07aF8cQvhsCOHVEMKaEMLmEMLMEMLPQgjDm2l/TWadLQ3Xtu8zlCRJkiRJkqQ9U0GuA9iJm4HDgM3AUuCAlhqGEAqAZ4CTgFnAfUANcAxwPXBVCOHEGOOMZhZ/GJjSzPQJ7yh6SZIkSZIkSVKzOnpx+gskRem5wGnAsztoezFJYfoZ4OwYY7p+RghhHPB14Ebgo80s+1CMcXwbxSxJkiRJkiRJ2okO3a1HjPHZGOOcGGNsRfN9MuNHGhemMx7OjPu3XXSSJEmSJEmSpN3V0a+c3hXTM+PzQgi3NylQX5AZP93CsoeHED4PlADLgGdjjEvbKU5JkiRJkiRJ2uN1peL0I8CDwCXAtBDC00AtcBRwMvBz4I4Wlr2hyeNUCOE3wOdjjNXtFK8kSZIkSZIk7bG6THE6xhhDCJeS9C19CzCm0exngD/EGFNNFltAcrPEJ0n6tu5FUsj+LvApoCfw4XYOXZIkSZIkSZL2OPljx47NdQytMm7cuBHANcC0sWPHPth0fgihBPgDcDXweeATwDeBF4DLgC+MGzdu6tixY9+sX2bs2LGLxo4d++rYsWPXjh07duvYsWM3jx07dsa4ceP+mtnWMePGjXto7NixK5uLKYTwyXHjxt11/vnnDwHYb7/9mDlzJv369WPy5MkMGTKECRMmUF5ezvTp0ykuLmbZsmVUVlayZcsW3nrrLYqKipgxYwb9+/dn0qRJ2y0zc+ZMCgsLeeutt9iyZQtVVVUsXbqU4uJi3njjDQYOHLjdMn369OHNN98kPz+fFStWsGnTJqqrq1myZAndunVj6tSpDBo0iAkTJpLuM5jCBZM4dK8+zJ49mxACq1atoqKigtraWhYtWkRpaSlTp05lyJAhvPbaa+y1115MmDCBXr16MXfuXGKMrFmzhvXr11NXV8fChQvp3r07r7/+ekNMgwYNYuLEifTs2ZN58+aRTqdZt24da9euJZ1OM3/+fHr06LHdMgMHDmTy5Mn06NGDBQsWUFdXx/r161mzZg0Ac+fOpXfv3kyZMqVhmQEDBvD6669TVlbGwoUL2bp1Kxs3bmTVqlWEEJg9ezZ9+vTZbpn+/fszdepUSktLWbx4MTU1NVRUVLBy5Ury8/OZNWsW5eXlTJo0qeG59+vXjzfeeIPi4mKWLl1KdXU1mzdvZvny5RQUFDTs//plXnvtNfr168eMGTMa9mXT/T99+nQGDBiw3b7s27cvM2fOpKCgoGFfNt3/9f/bpvs/Ly+PlStXUlFRQU1NDYsXL27Y/4MHD95umd69ezfs/9WrV7Nhwwa2bt36tv0/YcKEhnHPnj3ftv9TqRTz58+ne/fuTJkypeH/NWjQICZNmrTD/d+zZ8+G/fLaa68xaNAgpkyZQllZWbP7f86cOS3u/9LSUhYtWkRtbe3b9n/fvn23e13279+fadOmUVJSwuLFi6murmbTpk2sWLFiu/3feJnG+7/+tbx582beeustCgsLmTlzZrOv5cb7v/FrufH+nzhxYsP/rW/fvsyaNavF1/K0adO22/+vvfZas/u//rVcv/+b7sv61zKw3f5v7rVcnzv1+z+dTrN27drt9n/T13L9/m/8Wm68/+fNm0evXr2225cDBw5s2P+NX8urV6/ebv9Pnjy54TU2YMCAFl/LeXl5b9v/9cvU7/8lS5a0+FruyMfyll7LHss9lnss91jusdxjucdyj+Ueyz2Wd7Rj+ZtbAoXLZ5Mu60Px4tc9lnss91jeCY7lgwcPHrezem1bCK2712DuhRBOB54Ffh9jvKKZ+WOBW4EbYow/azLvMGAKsCjGOKKV27sXuBz4XIzx5ztqe+edd0aAT3/6061Zdc79cFZdw9//c0CXuXhekiRJkiR1QNYhpE4pZGMjednYSJbU3/Tw2aYzYoyvA+uA4SGE8laub3VmXNYGsUmSJEmSJEmSGulKxenizLh/0xkhhGKS/qMhuUliaxyXGc9/h3FJkiRJkiRJkproSsXpFzLjmzLF6MbGktz88bUY46b6iSGEU5quJCS+CpwArAEeb59wJUmSJEmSJGnP1aE7+gkhXARclHk4KDM+IYQwPvP3mhjjjZm/vw1cCLwLmBVCeByoAk4Cjs38fUOTTTwfQpgNvAYsA3pl2h8MVAKXxxgr2vp5SZIkSZIkSdKerkMXp4HDgaubTNsnMwAsAm4EiDEuCyEcCXwZOB/4CMmV4cuB8cD3Y4yzmqzrRySF6zOBvkAaWAz8AvhxjNEuPSRJkiRJkiSpHXTo4nSMcSxJlxytbb+apFh9487aZtr/z24FJkmSJEmSJEl6R7pSn9OSJEmSJEmSpE7C4rQkSZIkSZIkKessTkuSJEmSJEmSss7itCRJkiRJkiQp6yxOS5IkSZIkSZKyzuK0JEmSJEmSJCnrLE5LkiRJkiRJkrLO4rQkSZIkSZIkKessTkuSJEmSJEmSss7itCRJkiRJkiQp6yxOS5IkSZIkSZKyzuK0JEmSJEmSJCnrLE5LkiRJkiRJkrLO4rQkSZIkSZIkKessTkuSJEmSJEmSss7itCRJkiRJkiQp6yxOS5IkSZIkSZKyzuK0JEmSJEmSJCnrLE5LkiRJkiRJkrLO4rQkSZIkSZIkKessTkuSJEmSJEmSss7itCRJkiRJkiQp6yxOS5IkSZIkSZKyzuK0JEmSJEmSJCnrLE5LkiRJkiRJkrLO4rQkSZIkSZIkKessTkuSJEmSJEmSss7itCRJkiRJkiQp6yxOS5IkSZIkSZKyzuK0JEmSJEmSJCnrLE5LkiRJkiRJkrLO4rQkSZIkSZIkKessTkuSJEmSJEmSsq5DF6dDCJeGEH4eQnghhFARQoghhHt30L44hPDZEMKrIYQ1IYTNIYSZIYSfhRCG72C5qzPLbA4hbAwhPBdCuKB9npUkSZIkSZIkqUMXp4GbgeuAw4FlO2oYQigAngHuAHoA9wG/AlYB1wOvhxDGNLPcj4DxwGDg/wH3AocAfw8hXNdWT0SSJEmSJEmStE1BrgPYiS8AS4G5wGnAsztoezFwEkmB+uwYY7p+RghhHPB14Ebgo42mnwj8NzAPOCbGuD4z/YfAROBHIYR/xBgXtuFzkiRJkiRJkqQ9Xoe+cjrG+GyMcU6MMbai+T6Z8SONC9MZD2fG/ZtMvzYz/nZ9YTqz3YXAL4Bi4CO7FrUkSZIkSZIkaWc6dHF6F03PjM8LITR9XvX9Rz/dZPqZmfHjzazvsSZtJEmSJEmSJEltpKN367ErHgEeBC4BpoUQngZqgaOAk4Gfk/RHDUAIoQzYC9gcY1zezPrmZMb7tWfQkiRJkiRJkrQn6jLF6RhjDCFcStK39C1A45sfPgP8IcaYajStV2a8sYVV1k/v3doY7rzzztY2zanujf6+c0e9eEuSJEmSJL1D1iGkTil++tOfDu29kS7TrUcIoQT4E8lNDz8LDCYpQL8HGA48H0J4326susX+rkMInwwhTNiNdUqSJEmSJEnSHq3LXDkNfAX4AHBDjPGuRtMfy1xRPQW4nW3YVUA1AAAgAElEQVQ3R6y/MroXzdvZldXEGH8N/Hq3I5YkSZIkSZKkPVSXuXKabTc9fNsPRGKMrwPrgOEhhPLMtC3AMqB7CGFwM+vbNzOe3Q6xSpIkSZIkSdIerSsVp4sz4/5NZ4QQioGemYe1jWb9MzM+t5n1ndekjSRJkiRJkiSpjXSl4vQLmfFNmWJ0Y2NJujB5Lca4qdH0X2XGXwsh9KmfGEIYQdJvdQ1wT3sEK0mSJEmSJEl7shBji/f7y7kQwkXARZmHg4BzgPlsK0SviTHemGm7F/AKMBRYCDwOVAEnAcdm/n5XjPHlJtu4DfgisBR4ACgCPgSUA9fHGO9op6cnSZIkSZIkSXusjl6cHgvcuoMmi2KMIxq17w98GTgfGElyZfhykq45vh9jnNXCdq4GrgPGAGlgEvDDGOM/3vmzkCRJkiRJkiQ11aGL051Jfn5eTKc7xv8yhEBX2a95ATrIv1UdSB7Jt0htqSu9btpaCHnE2Nb/cbW5EMAc3nUhD7pafpsLuykA/t92pD3ef6Hjnu919XODjvp/72ra63XT1rp6vr8TnfJc2HOBHetM538dZl/uGedJneWYnUVrYoxvu7dfWyto7w3sKdLpSJyQ6ygS4ejI888/n+sw2sSpp56KHauoqXAdxEMvbdt1Tn2AC9/8WJuus6v4+/53c9nLtTtvqJy674Qijrp/Tq7D6HQmfmBfuKmLnWh/J3DZh2bkOopO574/jYH33Z7rMDq09MM3EE86vM3XG16c0mHOoxvrSufUzfE8Ozva47y1PXgu3LLOeC7seeGOdarzvw5yXrennCe117lOZxVenLIoG9vpSjdElCRJkiRJkiR1EhanJUmSJEmSJElZZ7ceOVZTC5/5Hjz9KqyrgNFD4TufhfNOynVkLfvmN7/JxIkTqa6upm/fvnz4wx/mggsuyHVYXd6cVXDId+DSI+Deq3MdTcd2+rzneKVyHQUhALBXQTfePODcHEfVessemcfsOyZTtXwLxf26cfj3TqX86EG5DqtLuP/MPts9TtVUMfqSazn6v3+ao4i6tppVS1n8m7FsmT2ZUFBEn+PPZdhHvkbI34NOP9bMhCc+CysmQml/OPOHsP/FuY6q3c2e83sWLHiIDRtnM3zv8zn+uO80zFux8mUmTPwWlZXLKS8/lOOP/TZlZXvlMNocm/88LH4VNr0Fex0FR16eTE/XwYTfwoYlULUOTroO+u2b21ibccfy1YxfuY5pldVc1r834/cdDsArm7Zwy6LlTNxSRT5weq/u/GyfoQwuKmyb7f4Jxv8Dps2Fy86B8WOT6TPmw1W3wrylyeOjDoSf3Qhj9mmTzbaJv/zlLzz++OPMnz+fd73rXdx0001va3PPPfdwzz338OMf/5ijjz46B1Hm1h3/gvGvwLTlcNlRMP7KZPrCtTDyVigr2tb2y2fBLeflJs72dseauYxfv4hp1Ru5rPcwxg87pmFeZbqOG5dP5c8blrI1pjmsW2+eH3V67oLdDQvuncGSB+ewafY6hlwwiiO+dyoAS/82l6m3vtjQLqYj6eoUp/zlffQ+uF+uwt1ts+//JQse/S0b5r3B8LM+xPG33N0wb/HT9zPtN9+gcvUySgcM5bBrv8nQ096Xw2jfmVWP/Y61zz1I1eI36XvSBYy47gcN89Y882dW/PUutm5YQ/cDjmL4Z75LUd+BOYx2N9XVwBOfgQVPQ/U66DMaTv8OjMociBY8k5z7VSyGIcfBheOh1/B2DyuVqmXCxG+wYuXL1NZupHv3vTns0M8zZPCppFK1vPzKl1i37g22VL7FmWeMZ+CAY9s9plZJ1cHUP8Pq2VBbCd37wYEXwMAxsG4hzHoENixN+t7uty8ccgmU9Mp11G2iJp3mM/OW8vTGTazbmmJ0t2K+M3ww5/XpyYzKaq6avYh51UmXQkd178bP9hnKmNKSHEfdPvagT4cdU10Khg2Ef/0a9h4Ej74IH/wqTPsjjBiS6+iad8UVV/DlL3+ZoqIiFi1axA033MC+++7L/vvvn+vQurTP/hmOaf/3tC7jjiFH8PHykbkOY5etfnEZM3/0Gkf95Ex6H9qf6tWVuQ6pS/nAP9c3/F1XtYW/nj+Uvc98fw4j6toW/2YshT37cuivXyK1pYLZ37yG1U/8ngHv2UO+YUvXwQPvgyOuhcuegsX/gvsvhI9OhvL9ch1du+rWbQAHjfkUy1e8SCpV0zC9pmY9/37xBo495hvsNeQMpk77GS++9N+cfdYfcxhtjpX0gv3PhlWzILV1+3nl+8Co0+C18TkJrTWGFBVy87CBPLFhE1XpbbcQWl+X4pOD+nFO7x4UhMB185fykTmLefygUW2z3f5w88fgiZehqmb76Q98H4YPhnQafnE//NdNMLUDpVi/fv246qqrePXVV6mpqXnb/GXLlvHcc89RXl6eg+g6hiG94OZz4YmZULX17fM3/BAK8rMfV7YNKezGzQMO4IlNK6mKqe3mfXLpJOpimpn7n0Pf/CKmVG3IUZS7r2RAKft+5jBWv7CMVM225zf0vaMZ+t7RDY+XPDib2b+cQq+DOudrolv/wRx0zVdZ/p+nSNVUNUyvXLWMl8ddwyk/+AuDjz+Ht156jBe/dhnvfXAOJX0H5DDi3VfYdwCD3/8ZNk55gVhb3TB90/T/sOwPt7Hf2HspHjScJfd8iwU//QL7f+MPOYx2N6XroMcwuOJf0GtvmPso/PWD8PFpUNQdHrwE3vMb2PdC+Nct8NcPwTWvtHtYMdZRWjqId535W8pKB/PW8ud58aUvct65D9OtpD/9+h3J/vtdyb9f+kK7x7JLYgq69YGTr0/GK2ck5z1nfhm2VsLwE+GYA5ObWE57ACb/AU74dK6jbhN1MTKsuJB/HTyavYuLeHR9BR+ctZBpR+zPkKICHjhgBMOLi0gDv1i+hv96cyFTjzgg12G3C7v1yLGybjD2U0khOi8PLjgFRg6BiTNzHVnLRo4cSVFRcrlCyFyZumzZslyG1OX9cQL07gbv6tq1DAFv/nwS+33mCPocPoCQF+g2sIxuA8tyHVaXtPiff6G4zwD6H35yrkPpsmpXLaHPie8hr6iYwj796XX4KVQt2YNuzrNmVnI17LFfgLx8GHEmDD0J3vhdriNrd8OGnsXQoe+muLj3dtOXLH2KXj1Hs/ewc8nPL+aQgz/Lho1vUlExP0eRdgBDDoPBh0JRk2N9XgGMOh3KRyVXC3VQl5T35qLy3pQXbH/Ny3l9evKBfr3pWZBPaX4e1w3ux4sVW9puu2fCRadDeZOLp3r3SM6rQ4AYIT8P5i5ps822idNOO41TTjmFnj17Njv/pz/9Kddeey2FhW1zlXlndMnhcNFhUL6HnwJd0msvLuq1F+UFRdtNf7N6E3+reItfDz2K/gXF5IfAUaV9WlhLxzX47BEMfvcIinoX77Ddkr/OZehFoxs+e3Y2w06/mKGnvY/iXn23m165ehmFPXoz5IRzCSGw10nvoaBbGZuXdd73xD7HnUPvY8+ioMf27/8bJz5LnxPOo9uwfckrLGLwpZ9l88zXqFmRlXutta2iMjh1LPQekRRM970Aeo1MfiX35oPQ7yA48ANQUAKnjIVVryfnhO2soKCUQw6+ju5lexFCHnsNOZ2ysqGsXzed/PwiDtj/Kvr3P4oQOtg3ewXFcMB5UFqe/D8HHQxlfZNfjg0cA3sdAYUlUFAEI0+BtQtyHXGbKcvPZ+zegxlRUkxeCFzQtxcjS4qYuLmK3gUFjCgpJoRABPIDzK1++xfaXYXF6Q5m5VqYvRja6KKSdvPjH/+Ys846iyuuuILy8nKOP/74XIfUZVVUwdcfgdsuyXUknctXV0yj3/S/cdLcZ3lu86pch9MqMZVmwxtrqFlfzTNn/ZmnTr2Pad94iVR1Xa5D65IWPHYvI8+7vNN+0OkMBrznata9+Ajpmipq165g45Tn6Xn4qbkOK4uauQt8jLD6jeyH0kFs3DiX3r23/dKqoKCU7mXD2Lhxbg6jUjY8X7GZg7L4U9Tep0PJSXD9D+Gmj2Rts+/Ys88+S0FBASeccEKuQ+nQhn8dht4MH/kdrNmc62iy7z9V6xheVMqtK6fTb/rfOGT2k/xl49Jch9UuKpdtYu2EFQx7X8fr1uid6nvAUfQcfgBLX/g76VSKpf96mLzCYnqPPiTXobW5GGNyDrRtAkDXuGhh80pYNzspSq+eDgMO2zavqAx6j4I107MeVlX1GjZtWkivXqN33rgjqa6Azauhx+C3z1s7D3p23e4uV9ZuZXZVzXbnS71fmUrJS69z/fxl3DS0E3aD00p269GBbK2Dy2+Bq8+HA0bkOpod++IXv8gNN9zA9OnTmTx5csOV1Gp7tzwCHzsRhnW+iyFy5vuDD2FMcU+KQh5/3LCECxe+xJR9382o4u65Dm2HatZUEbemWf74Qk76/QWEgjxe+8xTzL5zCgd+Yc/rb7I9bVmxmNWTn+e4m+7KdShdWvcxx7Lm6T8z+aojIJ2i/LSL6X3sWbkOK3vKD4CyAfDKD5Orpxc9m3TtMfyMXEeWM3V1lRQXb3/lWGFhD7bWtd0Vtep4pm6p4htLVvLwAdnrbmvDc7ClCv7vH0kXH51BZWUlv/71r7nttttyHUqH1a87vPY/cPhQWLsl6fbu8vHwxHW5jiy7lm6t5I3qCt7fcy/eOvACXq5cy/kL/82Y4p4cWNL8Ffmd1dKH5lJ+9EBKh/XIdShtLi8/n5HnXcHLt15FqraavIIiTvr2fRR063o/Geh15GnM/8nn6Xf2ZZQMGsHyB+6AEEg36uakU0pthb9dDodcDf0OgK2bk3uMNFbSC2o3ZTWsdHorL7/yJUaOeB89e3agmy7sTDoFE38Hw46FHk0KsRuXwZtPwHEfz01s7WxrOnL57EVcPaAvBzQqTm84/lC2pFL836r1DC/uur+o8srpDiKdhitvgaICuOPLuY6mdfLz8zn00ENZvXo1Dz30UK7D6ZKmLIWnZ8EX9tw6xm45rrScHvmFFOflc3XfEZxUWs6jm1bkOqydyitJvi8ceeUYSgaUUty3hH0+cjCr/tXBfo/cBSx47F76HXoS3Yd0vn7JO4uYTjPnWx+l93Fnc8S9r3PY/75K3ZYKlt37g50v3FXkF8KlD8HcR+Bng+A/t8GBH4SeQ3MdWc4UFJSydev2lzlurdtMYUHX+yCuxNyqGs6bMZ/bR+7FKb2y+yVxWTe49v3JDRJXrcvqpnfL//7v/3LOOecwZEgHvfFMB9C9GI4envQ3PbAn3PFBeHJW8kvDPUm3kE9hCNw88ECK8vI4rXt/zigbwJObV+Y6tDa39OG5DL2o6101DbDi1WeY8ouvcuYvnuZDz2/hXb98hle/ey3rZ0/JdWhtruchJzLkg59j/o+uY9pnTqNowFDySsooLO/EV8HGNPztSsgvgnPuSKYVdoeaiu3b1VRAUfa+XIkxzcuvfIW8vEKOPurmrG33HYtpmPS7pCu8Qy/dft7m1fDKXcnNEMs7eDcDuyEdI1fOWURRXuCOfd7+OaEsP59rB5Vz1ZzFrKpt5kYMXYBXTncAMcLHvgkr18Gjt0NhJ9srqVTKPqfbyXNzYOE62PuW5PHmGkhFmLEcJn0lt7F1JiFAbO7n9R1MUa9iSgaVgb1MtLuFj93LgVd+KddhdGmpzRvYunY5A869krzCYvIKi+l3xvtZ9sefMPTKTvItbFsYcChc+a9tj//vxOTqmj1Ur16jWbDg4YbHdXWVbN68pPP95FStsqi6lndPn8stQwdy5YC+O1+gHaTTUFkNy1ZBjkJotUmTJm130ceGDRu49dZb+fCHP8zll1+e4+g6pvpTpo5/lte2Di3ptfNGXcC6iSupXlXJkHNG5DqUdrF+zusMOPxkyg88CoDyMUdTftAxrHjtn/TZ7/AcR9f2Bpx7BQPOvQKA6rcWsOIvv6TbsE56U6UY4ZGPwZaV8KFHkwsSAPofBNP+b1u72i2wfl7S5UdWwor859Wbqa5Zy2mn/Iq8vE5ypW2MMPk+qNkEx38qKVDXq1wHL/0S9jsbhh2TuxjbSYyRj81dwsraOh4dsw+Fec0XA9JAZTrNstqtDCjqJPt1F3jldAfw6e/CzAXw959At+x1xbdb1q9fzzPPPENlZSWpVIpXX32VZ555hiOPPDLXoXVJnzwJ5o2FKV9NhmtPhvMPgic+m+vIOq4NqVqe2LSC6nSKupjm9+sX8/zmNZzTvXN8Kz/skn1Z8LsZ1KytonZjDQv+bzoDT98712F1Kaunvkzl6rfY+8z35zqULq2gZ1+KBgxl9ZN/IKbqqNtSwdrn/krp8K55h+kWrZoKddXJ3cZf+RFsXg6HXpPrqNpdOl1HKlVDOqaIMZX8na5j6F7vZmPFHJYseZJUqoY3pt9J7177da6fnLa1dCr5WXBMJ0NqazINIFWXPN6uXccqw9XFSHU6TSpGUhGq02nqYmRZTS1nTp/LZwf159rB/dp+u3VQXQOpNKRSyd91dfDUKzB5VjKtYjN88SfQpwcc2IF+KFNXV0dNTQ3pdJp0Ok1NTQ11dXX85Cc/Yfz48dx9993cfffdlJeXc+ONN3LxxRfnOuSsq0tB9dbM/k0nf9el4D8L4c2VyZcOazfD5x6A0/eFXt1yHXH7qItpqtOpzOsrNpzfntq9P3sXlvLdVbOoi2le3LKG57as5pzunas/0nRdmlRNHTEdiank73RdumH+kofmMPjsERR079xdSKbr6kjVVJNOpYjpVPJ3XR3lBx7NqtdfbLhSet2bk1k95cVO3ed0TNWRrq2BdJqYTpOurWmYVrV4NjFGale/xaK7bmbAe66moHsn/aLl8U/Dmpnwwb9DYaMD0H4XJ/cWmfWX5Pzv399ILlTol53z3wkTx1FRMZ9TT/4FBQXbF5dSqVpSqeSGeun0VlKpmqQv8I5g6p+TvruP+2RyJXq9qg3w4h0w8uRk6II+PW8pMyur+fuYkXTL31aifWrDJiZvriQVIxV1Kb64YBl98vM5MIv378imTnaNbtezaDnc9SAUF8Ggc7ZNv+smuPy83MXVkhACDz30ELfddhvpdJqBAwdy/fXXc8opp+Q6tC6ptCgZ6nUvhpJC6N/1ulxrM1tj5OYV05lVs4l8AgeU9OChESeyf0nn+Kft95kjqF1fzT/PeYD84nyGnDeSfT992M4XVKstePR3DDv9IgrLOkdOdGajbvwFS8Z/mxUP/xry8ulx0HEMveZruQ4ru6b9Dl7/TVJUHHYKXPZUclfyLm76jF/xxvRfNjxeuOjvHHzQZzjk4Os4+cSfMmHSt3n5P1+mvO+hnHjiHt6/7uwn4c3Htz1eOgH2Pze5c/0z34aqTH8UL9+ZjM/6enJH+w7iW0tWMG7Jtq4E7l29nluHDSQQmF9dy7glKxi3ZFvXWptPOLRttns3jPt/2x7f+xjc+onkpuLX/xCWroJuxXDMGHj851DSgV52v/3tbxk/fnzD4yeffJJrrrmGj370o9u1y8/Pp0ePHpSWlmY5wtz71uMw7rFtj+99DW49D/YfCDf9HVZtgp4lcNYBcF8nuuHlrvrWypmMWzWz4fG9GxZz64ADGTvoIB4ecSIfXzqR7616k+FFpfx22DEc0Mn6m55z5xRm3zG54fGyv81jv+uOYP/rjyRVU8dbjy3g6J+/K4cRto3p47/DG3d/q+Hxwsf/wMEfu5lDPv51DvnYLfz7psuoXr+S4t79GXP1lxl8XOe9P8fyv/yS5ff/vOHxuhceZvAHrmfA+dew4PYvUrNyMXklZfQ74/0M+dDncxjpO7BxEUy+C/KL4fZGF0CddxccfDlc8pekI/y/XQFDjoOL/piVsLZsWcbceX8mL6+Ih/627Qbkxxw1lhEjLuSRR9/Dlsq3AHjuX58A4MILnqJ72V5Zia9Fletg4UuQVwCPN+qG5LAPwZbVULk2OU9qfK50wQ+zH2c7WFRdy10r11IcAoNe3XbTzLtGDaUoL4/r5y9lac1WuuUFjulRyuMHjaIkr2teYxw6zDclnVwIIcYJuY4iEY6G559/PtdhtIlTTz2VeEeuo1BHE66D2LQfqne6zqkPcOGbH2vTdXYVf9//bi57uTbXYWgn7juhiKPu7wJ3PM+yiR/YF27qYudC3wlc9qEZuY6i07nvT2PgfbfnOoyO7eEbiCe1/U/Nw4tT6Cjn0Y11pXPq5nienR3tcd7aHjwXbllnPBf2vHDHOtX5Xwc5r9tjzpPa6VynswovTpkYYzy6vbfTNUvukiRJkiRJkqQOzeK0JEmSJEmSJCnrLE5LkiRJkiRJkrLO4rQkSZIkSZIkKessTkuSJEmSJEmSsi7E2EnuUNrBhRBWA4tyHYckSZIkSZIkvUPDY4z923sjFqclSZIkSZIkSVlntx6SJEmSJEmSpKyzOC1JkiRJkiRJyjqL05IkSZIkSZKkrLM4LUmSJEmSJEnKOovTkiRJkiRJkqSsK8h1AF1FQUFBzMuz1i9J6rjSdSmC71XtIhLxPEB7qnRdmhBCrsOQci6G7L4X1NWlyMvztSdJah+pVGpNjLF/e2/H4nQbycvLo6KiItdhSJLUovKy3jx/4A9yHUaXdPrCr7F69epchyHlRJ/Scn45cGKuw5By7nMVx7B6TfbeC8rKenHIIV/J2vYkSXuWKVO+sSgb2/ESH0mSJEmSJElS1lmcliRJkiRJkiRlnd16SJKkNnPhm+NYV7eJvBDoFoo4qccYji7bj/FrnuL+fW9qaPeZBb9gdd3Gt007tvt+RGDylnn8bMS1DfMunv1NhhX1f9u0aweczzm9j+ToN26gJBQRAnTP68ZZvY7ghkHvIz/4Pbykzuv7a69iydY3+cnAFygMRQDcveGrvFL1CAWhEIB++UM4rPgM3tP9E5Tm9QDg35V/5Z6NN1MUSgDokdeH/YuO5fzun2BQwcjcPBlJ2k3Tp99OXd0mDjroixQUlDZMnzXrLqqrV9Kjxyi2bFkMQDqdIgQIIR+APn0OZdiw83MSt6TWsTgtSZLa1I+Hf4Ljuu/Pqq0buG7hnRSEfBbUrGR93Wb6FHSnLqaYU/0WJXlF202bVrWQTw18DzFGxq9+mlRMkx/yWLO1grqY5s3qpdtNW1K7hiPLRjVs977RX2JYcX8W1qzkUwt+zt7F/bm078k5/E9I0u5bU7eM2bUTKQ09mFL9T47pdm7DvPO6f5RLenyerbGGpVtnc/+mH/HdtR/m5vI/UZyXFG5GFR7OTf1+TzqmWJNaxhNb7mHcmkv5Wvl9DC3cL1dPS5J2S1FRH9avf4P+/Y8FoKpqJTHWATB06PkUF/cGYNGihykq6sHgwWfmLFZJu8bLiSRJUrsYUNibk3qMYXXdRvYqLGfSlrkAzKpayj4lgziybNR209IxMqZkGAd125u6mGJ29TIAJlfO4+iy0QwvGrDdtKFF/ehf2Ott2x1RPJDDS0cxr3pFlp6pJLW9l6oeYlThYZxUehEvVj3UbJvCUMzIokO4vs8v2ZzewL+r/vq2NnkhnwEFe3Nlr1vZv+gYHt78i/YOXZLaXJ8+h7Bu3dSGx+vWvU6fPoe2atm6ui3Mm/d7pk37EdOm/YD58//I1q2bGuavXTuJ6dNvZ+rU7zFjxs/YsGFGw7w1ayYyc+YvMsv9gdrairZ7UpIAi9OSJKmdrKhdz4ubZrB/yVCOLBvF5Mp5QFJYPqJ0FIeX7rPdtENKR1CYV0BhXgEHlw5vKFxP2jKPw8tGcVjZPttNO6J0VLPbnV+9gimV89i/ZK8sPEtJah8vVf2N47tdwPHdLmR6zYtsTK1psW23vDLGFJ/I7NoJO1znUSVnMad2YluHKkntrqxsKOl0DdXVq4kxzYYNM+jb95BWLRtjpF+/oznooBsYM+ZzACxb9iQAdXVVLFv2FKNHX8mhh36F0aOvoaSkPwDr17/B6tX/YeTI/+Lgg/+bbt0GsXjx278ElPTOWJyWJElt6sbFv+H0GV/h4wtu58iyUXyk/1kcWTaaSVsyhegt8zi8bB+OKBu13bTGXXQcWTqayZl5UyrncUTpPhxROmq7aUeVjd5uu1fM+yFnzPgKX1j0a97X5wTe2+e4bDxdSWpzs2snsjb1Fsd0O48RhQfRP38Y/6n6xw6X6Z03gC3pje+4jSR1VPVXT2/aNJ/i4nIKC3u2arnCwu706rU/eXmF5OeXMHDgyWzevGi7NtXVq0mn6ygq6tlQnF67diIDB55CSUk5IeQzaNBpbN68hK1bN7f5c5P2ZPY5LUmS2tSP9v44x3Xff7tpR5SN4pvL7qMiVckblQv51tCrKM0vZm1dBRWpSl6vnM+Hy09vaH9k2SjuX/cCFalK1tdtZu/iAfQt6MnYpfdSkapkXvVyjijb/srpe0f9D8OK+2fjKUpSu3qp8iEOKj6RHnl9ADi+2/m8WPUwZ3e/psVlNqRXUpb39q6OGlvfijaS1FH17Xsoc+b8H7W1G+jb97BWL5dK1bBs2RNs2jSfVKoagHQ66a+6oKAbw4dfzOrVr7Bo0UN07z6cvfY6m+LivtTWbmTp0kdYuvTRhnWFkMfWrRUUFnZv2ycn7cEsTkuSpHY3tKgf/Qt68eC6lxhU1IfS/GIADuk2ggfXvURluoaDS4c3tD+kdASb09U8uO4lDivdB4Du+SX0K0zW0b+gF3sVlefkuUhSe6qN1bxW/Thp0nx+5SkA1MVaKmMFi7fOanaZ6vQWZtS8zAXdr93huidVP82+RUe1ecySlA1FRb0pKupNRcUchg27sNXLrVr1Ilu3VrDffh+nsLA7W7YsZc6cexrm9+q1H7167Uc6vZW33nqaJUseYfToKyks7MmQIWfRu/eB7fF0JGVYnJYkSVlxeNk+/GHNs5zb++jtpv129TMc2G1vSvKKGqaX5BUxpmQYf1jzLB8dcPa29qXJOo5tcmW2JHUVk6qfIY98vtH/YfBy4yQAABDdSURBVAoobJh+54Yv8FLVw9u13RprWVY3h/srfkRpXi9O7nbx29aXjinWppbz5JbxvFn7Gl8rv6/dn4MktZe9976QVKqa/PwiYky3aplUqpYQki496uoqWbnyhYZ5tbUVVFWtoEePkYSQT15eISEkPeD263c0K1e+QElJf0pK+lFXV8XmzQstVkttzOK0JEnKiiPLRvPExkkcnrkSGuDw0lH8NPUw7y09vtn2U6sWbtf+iLJR/HndC2/r0kOSuoqXqh7ipNKLKc8fst30M0sv5w8V32FM0Qk8tvl/eWrL74hE+uUP4bDi0zm/+ycpzittaD9v6xQ+veIoIpEeeb3Zv+hYbin/M0MKPX5K6ryKi/vu8jIDBpzAokUP8sYbP6SwsCf9+x9LRcXczNzIqlUvsmjRg0CgtHQww4adD0CfPgeTTm9l4cL7qa3dSH5+N3r2HG1xWmpjIcaY6xi6hKKiolhRUZHrMCRJalF5WW+eH/ODXIfRJZ2+8GusXr0612FIOdGntJxfDpqY6zCknPtcxTGsXpO994Kysl4ceuhXsrY9SdKeZcqUb0yMMR6985bvTF57b0CSJEmSJEmSpKYsTkuSJEmSJEmSss7itCRJkiRJkiQp6yxOS5IkSZIkSZKyzuK0JEmSJEmSJCnrQowx1zF0CSGETcCbuY5DXUI/YE2ug1CnZx6prZhLaivmktqKuaS2Yi6prZhLagvmkTqa/WOMPdp7IwXtvYE9yJsxxqNzHYQ6vxDCBHNJ75R5pLZiLqmtmEtqK+aS2oq5pLZiLqktmEfqaEIIE7KxHbv1kCRJkiRJkiRlncVpSZIkSZIkSVLWWZxuO7/OdQDqMswltQXzSG3FXFJbMZfUVswltRVzSW3FXFJbMI/U0WQlJ70hoiRJkiRJkiQp67xyWpIkSZIkSZKUdRanJUmSJEmSJElZZ3H6HQghDA0h/G8I4a0QQk0IYWEI4achhD65jk3ZF0K4NITw8xDCCyGEihBCDCHcu5NlTgwhPBpCWBdCqAwhTA0hfD6EkL+DZS4IITwXQtgYQtgcQvhPCOHqtn9GypUQQnkI4eMhhL+GEOaGEKoy+/vfIYSPhRCaPXabT2pOCOH7IYRnQghLMrm0LoQwOYRwawihvIVlzCXtVAjhysx7XQwhfLyFNrucFyGEq0MIr2bab8wsf0H7PAtlW+Z8ObYwrGhhGY9JalEI4ZQQwl9CCMszn8mWhxCeDCG8p5m25pK2E0K4ZgfHpPoh1cxy5pKaFUI4P3MMWpo5954fQrg/hHBCC+3NJbWLkPhoCOGVEMKmTH5NDiF8rml+hRAKQwg3hBDuCSFMCSHU7ugcP7PMSSGEH4QQXgshrM68By8IIfwmhDB6l2K1z+ndE0IYBbwEDAAeBmYBxwJnAG8CJ8UY1+YuQmVbCGEKcBiwGVgKHAD8PsZ4RQvt3wf8BagG/gSsAy4E9gceiDF+oJllrgN+DqzNLFMLXAoMBW6LMd7Yxk9LORBCuBa4E1gOPAssBgYClwC9SPLmA7HRAdx8UktCCLXAJGAGsAooA44HjgbeAo6PMS5p1N5c0k6FEIYB04B8oDvwiRjjb5q02eW8CCH8CPhvkvfRB4Ai4L+AvsD1McY72us5KTtCCAuB3sBPm5m9Ocb4oybtPSapRSGEm4FvAmuAf5CcO/UDjgCejTF+qVFbc0lvE0I4HLiohdmnAGcCj8QYL2i0jLmkZoUQvg98iWQ/P0RybBoNvBcoAK6KMd7bqL25pHYTQvgtcCXJZ8C/A1uAdwNjaFJTCCH0BtZnFl1JklfDaOYcv9H6VwD9SWqjE4E64ATgxMy2zooxvtyqYGOMDrsxAE8AkeSDUuPpP85M/1WuY3TIek6cAewLBOD0TB7c20LbnpkDRA1wdKPpJZkXdgT+q8kyI0jetNYCIxpN7wPMzSxzQq7/Dw5tkktnkpyU5DWZPoikUB2B95tPDq3Mp5IWpn87s59/aS457MqQeZ97GpgH/DCzjz/+TvOC5EQ2Zub3abKutZn1jWiv5+WQtfxZCCxsZVuPSQ47yo8PZPbnU0CPZuYXmksO72QAXs7s5/eaSw6tyJdBQApYAQxoMu+MzH6eby45ZGMg+dItAvOBfo2mFwJ/zcy7ptH0IuA8YHDm8ViaOcdvso0vA0OamX5TZtlprY3Xbj12QwhhH+BskpPrXzSZfSvJNwRXhhDKshyacijG+GyMcU7MvBp34lKSb5j+GGOc0Ggd1cDNmYefbrLMR4Fi4I4Y48JGy6wHvpN5eO1uhq8OJMb4zxjj32OM6SbTVwC/yjw8vdEs80ktyuRBc/6cGe/baJq5pNb4HMmXaB8hOedpzu7kRf3jb2fa1S+zkOR8qzizTe05PCapWSHp4uz7QCXw4RjjpqZtYoxbGz00l7RLQggHk/zSbBnwSKNZ5pJaMpyk69z/xBhXNZ4RY3wW2ESSO/XMJbWnSzLj22KMa+onZt4bb8k8vL7R9NoY42MxxuWt3UCM8fsxxreamfV9oAo4OLTQjWRTFqd3z5mZ8ZPNFI82AS8CpSRvZlJz6nPo8WbmPU9yon1iCKG4lcs81qSNuq76D1p1jaaZT9odF2bGUxtNM5e0QyGEA4HvAbfHGJ/fQdPdyQtzac9RHEK4IoRwU6Z/wzNa6FvTY5JaciIwEngUWJ/p4/XLmXxqrl9Xc0m76lOZ8d0xxsZ9TptLaskckq4Qjg0h9Gs8I4RwKtCD5Jdn9cwltadBmfH8ZubVTzsy051HW4tsq1e8rc/+5lic3j37Z8azW5g/JzPeLwuxqHNqMYdijHXAApI+qfZp5TLLSa5eGxpCKG3bUNVRhBAKgKsyDxufkJhP2qkQwo0hhLEhhJ+EEF4g6aNzKkmhsZ65pBZljkG/I+le6KadNN+lvMj82mwvkj6Hm7tiw3OrrmUQSS59m6Tv6X8Cc0IIpzVp5zFJLTkmM15Jcl+Ff5C8n/0UeCmE8K8QQuMrFM0ltVoIoRtwBZAGmva1ai6pWTHGdSTdHAwEZoQQfh1C+G4I4c/AkyRdEH2q0SLmktpT/dXSI5uZ1zinDmiHbX+A5MuYV2KMG1qzgMXp3dMrM97Ywvz66e3xDYS6ht3JodYu06uF+er8vgccDDwaY3yi0XTzSa1xI0nXU58HTib5guPsGOPqRm3MJe3I10luMnZNjLFqJ213NS88t9pz3AO8i6RAXQYcAtxF0o/mYyGEwxq19ZiklgzIjK8FupHc4KkHyXnSE8CpwP2N2ptL2hUfJMmFx2Kjm0ZnmEtqUYzxpyTdKRQAnwC+QlKoWwKMb9Ldh7mk9vSPzPiLIYS+9RMzF5uMa9SuT1tuNIQwkuSGnXUkNzlvFYvT7SNkxq3pe1hqzu7kkHnXhYUQPkdycJ9FcsfdXVo8Mzaf9mAxxkExxkBSELqE5BvzySGEI3dhNebSHiqEcCzJ1dK3xdbedXsnq8yMdzUvzKNOLsY4LnNvhZUxxsoY4xsxxmtJbirejeQGPK3lMWnPVd8NTAAujTE+E2PcHGOcDlwMLAVOa6GLj+aYS2rsk5nxXbuxrLm0BwshfAl4ABgPjCL5EvYokm4Ufh9C+MGurC4zNpe0O/5I0tXLKLZdyf9TYArwHrb9KrFV3W60RghhQGab/YEbYowvtXZZi9O7Z2ffRvVs0k5qandyqLXLVLyDuNQBhRA+C9wOzADOyPxkrDHzSa2WKQj9leTGvuXAbxvNNpf0No2685jNthuo7Myu5sXO2u/sSiF1fvU3/D210TSPSWpJ/U1T58cYX288I/PLjvpfmB2bGZtLapUQwhiSPs2XkvRp3pS5pGaFEE4nuRHc32KMX4wxzs98CTuJ5EuzZcB/hxDqu1Qwl9RuMvfHey/Jr2dXkFzg9lGSY9vJwNpM01XNrmAXZQrT/yTpeuaGGOMvd2V5i9O7583MuKV+D/fNjFvqk1pqMYcyRYCRJD+DmN/KZQaTfCu7NMZY2bahKpdCCJ8H7gDeIClMr2immfmkXRZjXETyhcdBjW7aYi6pOd1J9u+BQHUIIdYPJF3FAPy/zLSfZh7vUl7EGLeQfGjrnpnflOdWXV/9h6OyRtM8Jqkl9fu5pb4s64vX3Zq0N5e0My3dCLGeuaSWXJAZP9t0RmbfvkpSgzsiM9lcUruKMdbFGG+LMR4eY+wWY+wZYzyX5DPg4UAVMP2dbieTe88BY4DPxhh/tqvrsDi9e+oPNmeHELb7H4YQegAnkezkV7IdmDqNf2bG5zYz71SgFHgpxljTymXOa9JGXUAI4cvAT0h+enNGkz7KGjOftLuGZMb1H77MJTWnBri7hWFyps2/M4/ru/zYnbwwl/Zs9d0vNP4Q7jFJLXmepGizbwihqJn5B2fGCzNjc0k7FUIoIbm6ME3yntYcc0ktKc6M+7cwv356bWZsLilXrgRKgD/HGLe+kxWFEIYC/yK5seK1u3rFdIMYo8NuDCQ/FYvA9U2m/zgz/Ve5jtEhp/lxeiYP7m1hfk9gNckH/qMbTS8BXsos+19NlhkJVJP8/GJEo+l9gLmZZU7I9XN3aLMcuiWzTycAfXfS1nxyaCk3DgAGNTM9D/h2Zj+/aC457O5A0j9wBD7+TvOC5GfUMTO/T6PpIzLrqW68LofONwAHNfeeBgwn6fswAjc1mu4xyWFH+XRvZn9+q8n0s0iKixuA3uaSwy7k1JWZ/fr3HbQxlxxayo0PZvblCmCvJvPOyxyXqoByc8khGwPQs5lpxwDrgE3APjtYdizNnOM3abM3MI/kQqePvJNYQ2aF2kUhhFEkB4wBwMPATOA44AySn5yeGGNc2/Ia1NWEEC4CLso8HAScQ3L1zwuZaWtijDc2af8AyZvLH0kOEO8l6aPnAeCDsckLNIRwPfAzkjejP5F863opMJTkJlU3ok4vhHA1yU00UiR3um2uj9WFMcbxjZYxn/Q2mW5hfkhyhdk8kn09EDiN5IaIK4B3xRhnNFrGXFKrhRDGknTt8YkY42+azNvlvAgh3AZ8kaQ/vAeAIuBDJP2jXx9jvKPdnozaXSZfvkLyK8QFJB+MRgHnk3wYfxS4OMZY22gZj0lqVqZ/yxeB0STn26+SfNFxMckH6g/HGO9v1N5c0g6FEF4g6Yv1vTHGv++gnbmkt8n8qv4J4N0k729/JTnXPpCky48AfD7GeHujZcwltZsQwn9IvhB5gyQnDyK5GWINcEmM8Ykm7b9CcnETJN1+HEZS96y/eeK/G5/vhxAWkFxEMhH4RwthjI8xLtxpsLmu5HfmARgG3AMsJzkgLCK5adkOr3J06JoD275ZamlY2MwyJ5F8EFufOWhMA74A5O9gOxeS/GxiE7AFeA24OtfP3yGruRSB58wnh1bk0sHAL0i6hllD8hPojZn9PLal9ytzyaG1Azu5qmJ38gK4OtNuS2a5fwEX5Pq5OrRJvpwG3AfMIrmqdSvJVWNPAVdBcuFMM8t5THJoaT/3Jfnl6gKSz2NrSS4cOt5cctjFXDow8362ZEf50Ki9ueTQ3D4uBD5P0sVrRebcexVJ4e5sc8khmwPwPySF4w0kBekFJDegHtFC++fYcQ1ifJP2O6tZROD01sTqldOSJEmSJEmSpKzzhoiSJEmSJEmSpKyzOC1JkiRJkiRJyjqL05IkSZIkSZKkrLM4LUmSJEmSJEnKOovTkiRJkiRJkqSsszgtSZIkSZIkSco6i9OSJEmSJEmSpKyzOC1JkiRJkiRJyjqL05IkSZIkSZKkrLM4LUmSJEmSJEnKuv8PWhcdnN0uu2sAAAAASUVORK5CYII=\n",
      "text/plain": [
       "<Figure size 1728x576 with 3 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "#Plotting the predicted fitness effect for the variants seen more than once\n",
    "plt.close('all')\n",
    "f, (ax1, ax2) = plt.subplots(2, 1, sharey=True, sharex = True, figsize=(24, 8))\n",
    "gs = matplotlib.gridspec.GridSpec(3, 1, width_ratios=[1], height_ratios=[10,0.5, 0.5])\n",
    "ax1 = plt.subplot(gs[0])\n",
    "ax2 = plt.subplot(gs[1])\n",
    "ax3 = plt.subplot(gs[2])\n",
    "gs.update(hspace=0.05)\n",
    "\n",
    "nonsense_list = []\n",
    "y_list = []\n",
    "x_list=[]\n",
    "y_list_top20 = []\n",
    "x_list_top20=[]\n",
    "y_list_nonsense = []\n",
    "x_list_nonsense = []\n",
    "x=[]\n",
    "y=[]\n",
    "OBS_EXP_DICT={}\n",
    "top20_lower_err = []\n",
    "top20_upper_err = []\n",
    "missense_lower_err = []\n",
    "missense_upper_err = []\n",
    "nonsense_lower_err = []\n",
    "nonsense_upper_err = []\n",
    "\n",
    "top20 = c1\n",
    "missense = c6\n",
    "nonsense = c10\n",
    "synonymous = c3\n",
    "\n",
    "axisfont=20\n",
    "domainfont = 12\n",
    "titlefont=12\n",
    "axislabelfont=12\n",
    "\n",
    "top_20_DNMT3A = ['p.R882C', 'p.R729W', 'p.R326C', 'p.R320*', 'p.R882H', 'p.R736H', 'p.Y735C',\\\n",
    "                'p.R736C', 'p.W860R', 'p.R771*', 'p.R598*', 'p.P904L']\n",
    "\n",
    "for k, v in DNMT3A_missense_s.items():\n",
    "    if k in top_20_DNMT3A:\n",
    "        codon_number = int(((re.findall('\\d+', k)))[0])\n",
    "        s_value = float(v[0])*100\n",
    "        x_list_top20.append(codon_number)\n",
    "        y_list_top20.append(s_value)\n",
    "        \n",
    "        s_upper = float(v[3])*100\n",
    "        s_lower = float(v[4])*100\n",
    "        error_upper = s_upper-s_value\n",
    "        error_lower = s_value-s_lower\n",
    "        top20_lower_err.append(error_lower)\n",
    "        top20_upper_err.append(error_upper)\n",
    "\n",
    "    if k not in top_20_DNMT3A:\n",
    "        codon_number = int(((re.findall('\\d+', k)))[0])\n",
    "        s_value = float(v[0])*100\n",
    "        x_list.append(codon_number)\n",
    "        y_list.append(s_value)\n",
    "        \n",
    "        s_upper = float(v[3])*100\n",
    "        s_lower = float(v[4])*100\n",
    "        error_upper = s_upper-s_value\n",
    "        error_lower = s_value-s_lower\n",
    "        missense_lower_err.append(error_lower)\n",
    "        missense_upper_err.append(error_upper)\n",
    "\n",
    "    missense_err = [tuple(missense_lower_err), tuple(missense_upper_err)]   \n",
    "    \n",
    "for k, v in DNMT3A_nonsense_s.items():\n",
    "    if k in top_20_DNMT3A:\n",
    "        codon_number = int(((re.findall('\\d+', k)))[0])\n",
    "        s_value = float(v[0])*100\n",
    "        x_list_top20.append(codon_number)\n",
    "        y_list_top20.append(s_value)\n",
    "        \n",
    "        s_upper = float(v[3])*100\n",
    "        s_lower = float(v[4])*100\n",
    "        error_upper = s_upper-s_value\n",
    "        error_lower = s_value-s_lower\n",
    "        top20_lower_err.append(error_lower)\n",
    "        top20_upper_err.append(error_upper)\n",
    "    \n",
    "    top20_err = [tuple(top20_lower_err), tuple(top20_upper_err)]\n",
    "        \n",
    "    if k not in top_20_DNMT3A:\n",
    "        codon_number = int(((re.findall('\\d+', k)))[0])\n",
    "        s_value = float(v[0])*100\n",
    "        x_list_nonsense.append(codon_number)\n",
    "        y_list_nonsense.append(s_value)\n",
    "        \n",
    "        s_upper = float(v[3])*100\n",
    "        s_lower = float(v[4])*100\n",
    "        error_upper = s_upper-s_value\n",
    "        error_lower = s_value-s_lower\n",
    "        nonsense_lower_err.append(error_lower)\n",
    "        nonsense_upper_err.append(error_upper)\n",
    "\n",
    "    nonsense_err = [tuple(nonsense_lower_err), tuple(nonsense_upper_err)]        \n",
    "    \n",
    "msize = 12\n",
    "medgewidth = 2\n",
    "linew = 3\n",
    "csize = 4\n",
    "cthick = 3\n",
    "\n",
    "markers1, caps1, bars1 = ax1.errorbar(x_list, y_list, yerr = missense_err, fmt = 'o', ecolor = missense, elinewidth = linew, \\\n",
    "             capsize = csize, capthick = cthick, markersize = msize, markeredgewidth = medgewidth, \\\n",
    "             markeredgecolor = missense, markerfacecolor = missense, zorder = 1)    \n",
    "\n",
    "markers2, caps2, bars2 = ax1.errorbar(x_list_nonsense, y_list_nonsense, yerr = nonsense_err, fmt = 'o', ecolor = nonsense, elinewidth = linew, \\\n",
    "             capsize = csize, capthick = cthick, markersize = msize, markeredgewidth = medgewidth, \\\n",
    "             markeredgecolor = nonsense, markerfacecolor = nonsense, zorder = 2)    \n",
    "\n",
    "markers3, caps3, bars3 = ax1.errorbar(x_list_top20, y_list_top20, yerr = top20_err, fmt = 'o', ecolor = top20, elinewidth = linew, \\\n",
    "             capsize = csize, capthick = cthick, markersize = msize, markeredgewidth = medgewidth, \\\n",
    "             markeredgecolor = top20, markerfacecolor = top20, zorder = 3)    \n",
    "\n",
    "# loop through bars and caps and set the alpha value\n",
    "a = 0.7\n",
    "[bar.set_alpha(a) for bar in bars1]\n",
    "[cap.set_alpha(a) for cap in caps1]\n",
    "[bar.set_alpha(a) for bar in bars2]\n",
    "[cap.set_alpha(a) for cap in caps2]\n",
    "[bar.set_alpha(a) for bar in bars3]\n",
    "[cap.set_alpha(a) for cap in caps3]\n",
    "\n",
    "#Only show the required axis lines\n",
    "ax1.spines['top'].set_visible(False)\n",
    "ax1.spines['right'].set_visible(False)\n",
    "\n",
    "for axis in ['bottom','left']:\n",
    "    ax1.spines[axis].set_linewidth(2)\n",
    "\n",
    "for axis in ['bottom','left']:\n",
    "    ax1.spines[axis].set_color(grey3)\n",
    "    \n",
    "ax1.set_ylabel('fitness effect (s)', fontsize = axisfont, labelpad = 10)\n",
    "ax1.set_title('Fitness effects across DNMT3A', fontsize = 18)\n",
    "\n",
    "# ax1.set_yscale('log')\n",
    "\n",
    "# STACKED BAR CHART\n",
    "# define the exons + create data\n",
    "exons = [{'2':(24-1)+1, '3': (59-25)+1, '4': (150-60)+1, '5':(164-151)+1, '6':(213-165)+1, '7': (285-214)+1, '8': (338-286)+1,\\\n",
    "          '9': (374-339)+1, '10': (427-375)+1, '11': (477-428)+1, '12': (492-478)+1, '13': (518-493)+1, '14':(556-519)+1, \\\n",
    "          '15': (617-557)+1, '16': (646-618)+1, '17': (694-647)+1, '18': (725-695)+1, '19': (774-726)+1, '20': (803-775)+1, \\\n",
    "          '21': (826-804)+1, '22': (866-827)+1, '23': (912-867)+1}]\n",
    "exon_orders = [['2', '3', '4', '5', '6', '7', '8', '9', '10', '11', '12', '13', '14', '15', '16', '17', '18', '19', '20', '21', '22', '23']]\n",
    "colors2 = [c4, c0, c3, c1, c5, c6,c7, c8, c9, c10, c2,c4, c0, c3, c1, c5, c6,c7, c8, c9, c10, c2]\n",
    "names = exons[0].keys()\n",
    "values = np.array([[data[name] for name in order] for data,order in zip(exons, exon_orders)])\n",
    "lefts = np.insert(np.cumsum(values, axis=1),0,0, axis=1)[:, :-1]\n",
    "orders = np.array(exon_orders)\n",
    "\n",
    "#plot the bar chart\n",
    "for name, color in zip(names, colors2):\n",
    "    idx = np.where(orders == name)\n",
    "    value = values[idx]\n",
    "    left = lefts[idx]\n",
    "    ax2.bar(x=left, height=-10, width=value, bottom=-5, color=color, orientation = 'horizontal', label = name, edgecolor='k', linewidth=1, tick_label = name)\n",
    "    ax2.text(left + (value/2.), (-0.26), name, ha='center', va='center', fontsize = axislabelfont)\n",
    "    \n",
    "\n",
    "# STACKED BAR CHART of domains    \n",
    "domains = [{' ': (291-1)+1, 'PWWP' :(350-292)+1, '  ':(481-351)+1, 'ADD': (614-482)+1, '   ': (633-615)+1, 'MTase': (912-634)+1}]\n",
    "domain_orders = [[' ','PWWP','  ','ADD','   ','MTase']]\n",
    "# blank_col = c13\n",
    "blank_col = grey1\n",
    "colors3 = [blank_col, PWWP_col, blank_col, ADD_col, blank_col, MTase_col]\n",
    "names_domains = domains[0].keys()\n",
    "values = np.array([[data[name] for name in order] for data,order in zip(domains, domain_orders)])\n",
    "lefts = np.insert(np.cumsum(values, axis=1),0,0, axis=1)[:, :-1]\n",
    "orders = np.array(domain_orders)\n",
    "\n",
    "#plot the bar chart\n",
    "for name, color in zip(names_domains, colors3):\n",
    "    idx = np.where(orders == name)\n",
    "    value = values[idx]\n",
    "    left = lefts[idx]\n",
    "    ax3.bar(x=left, height=-10, width=value, bottom=-5, color=color, orientation = 'horizontal', alpha = 0.9, label = name, edgecolor='k', linewidth=1)\n",
    "    ax3.text(left + (value/2.), (-0.25), name, ha='center', va='center', fontsize = domainfont, color = 'k', fontweight = 530)\n",
    "\n",
    "    \n",
    "\n",
    "# CONFIGURING THE GRAPH APPEARANCE\n",
    "#Set the x and y axis limits\n",
    "ax1.set_xlim(0, 912)\n",
    "ax1.set_ylim(180, 220)\n",
    "ax2.set_xlim(0, 912)\n",
    "ax2.set_ylim(-0.5, 0)\n",
    "ax3.set_xlim(0, 912)\n",
    "ax3.set_ylim(-0.5, 0)\n",
    "\n",
    "#x-axis ticks\n",
    "x1_major_ticks = []\n",
    "x1_major_tick_labels = []\n",
    "x2_major_ticks = [0, 100, 200, 300, 400, 500, 600, 700, 800, 912]\n",
    "x2_major_tick_labels = [\"0\", \"100\", \"200\", \"300\", \"400\", \"500\", \"600\", \"700\", \"800\", \"912\"]\n",
    "ax3.set_xticks(x2_major_ticks)\n",
    "ax3.set_xticklabels(x2_major_tick_labels, fontsize = axisfont)\n",
    "ax1.set_yticklabels([])\n",
    "ax1.set_yticks([])\n",
    "ax1.set_xticklabels([])\n",
    "ax1.set_xticks([])\n",
    "ax2.set_yticklabels([])\n",
    "ax2.set_yticks([])\n",
    "ax2.set_xticklabels([])\n",
    "ax2.set_xticks([])\n",
    "ax3.set_yticklabels([])\n",
    "ax3.set_yticks([])\n",
    "\n",
    "#y-axis ticks\n",
    "y1_major_ticks = [180, 185, 190, 195, 200, 205, 210, 215, 220]\n",
    "y1_major_tick_labels = [\"180\", \"185\", \"190\", \"195\", \"200\", \"205\", \"210\", \"215\", \"220\"]\n",
    "ax1.set_yticks(y1_major_ticks)\n",
    "ax1.set_yticklabels(y1_major_tick_labels, fontsize = axisfont)\n",
    "\n",
    "#set panel legends\n",
    "legend_marker_size = 2\n",
    "legend_elements = [Line2D([0], [0], marker = '_', color = missense, alpha = 1.0, markersize = legend_marker_size, \\\n",
    "                          lw=3, label='missense'),\n",
    "                   Line2D([0], [0], marker = '_', color=nonsense, alpha=1.0, markersize = legend_marker_size, \\\n",
    "                          lw=3, label='nonsense'),\n",
    "                  Line2D([0], [0], marker = '_', color=top20, alpha=1.0, markersize = legend_marker_size, \\\n",
    "                          lw=3, label='top 20 observed variant')]\n",
    "    \n",
    "ax1.legend(ncol=1, handles=legend_elements, loc='upper left', frameon=0, fontsize = 20)\n",
    "\n",
    "ax1.grid(which='major', linestyle=':', linewidth='0.7', color=grey3, zorder = 0)\n",
    "\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## TET2 variants"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 241,
   "metadata": {
    "code_folding": [
     0,
     91,
     102,
     114
    ]
   },
   "outputs": [],
   "source": [
    "def expected_number_variants_missense_TET2(data, params):\n",
    "    s_value = params[0]\n",
    "    \n",
    "    a = 0\n",
    "    expected_number_by_variant_missense = {}\n",
    "    \n",
    "    Jaiswal_fmin = 0.056\n",
    "    Jaiswal_mean = 58\n",
    "    Jaiswal_sigma = 11.5\n",
    "    Jaiswal_total = 17182\n",
    "    \n",
    "    Genovese_fmin = 0.1194\n",
    "    Genovese_mean = 55.1\n",
    "    Genovese_total = 12380\n",
    "    Genovese_sigma = 11.95\n",
    "    \n",
    "    McKerrel_fmin = 0.008\n",
    "    McKerrel_mean = 51.88\n",
    "    McKerrel_total = 4219\n",
    "    McKerrel_sigma = 18.36\n",
    "    \n",
    "    Zink_fmin = 0.2363\n",
    "    Zink_mean = 60\n",
    "    Zink_total = 11262\n",
    "    Zink_sigma = 1.18907720e+01\n",
    "    \n",
    "    Coombs_fmin = 0.019\n",
    "    Coombs_mean = 59.82\n",
    "    Coombs_total = 5649\n",
    "    Coombs_sigma = 15.14\n",
    "    \n",
    "    Young_fmin = 0.0009\n",
    "    Young_mean = 63\n",
    "    Young_total = 89\n",
    "    Young_sigma = 6.8\n",
    "    \n",
    "    Desai_fmin = 0.0169\n",
    "    Desai_mean = 66.1\n",
    "    Desai_total = 424\n",
    "    Desai_sigma = 6 #estimated\n",
    "    \n",
    "    Acuna_fmin = 0.0015\n",
    "    Acuna_mean = 4.5\n",
    "    Acuna_total = 2014\n",
    "    \n",
    "    N = 9.39358404e+04\n",
    "\n",
    "    for i in data:\n",
    "        variant_mutation_rate = mutation_rates_variants_TET2(i)\n",
    "\n",
    "        if i in Jaiswal_missense_TET2:\n",
    "            Jaiswal_number = prevalence_averaged_over_ages(Jaiswal_fmin, s_value, N, variant_mutation_rate, Jaiswal_mean, Jaiswal_sigma, Jaiswal_total)\n",
    "        else:\n",
    "            Jaiswal_number = 0\n",
    "\n",
    "        if i in Genovese_missense_TET2:\n",
    "            Genovese_number = prevalence_averaged_over_ages(Genovese_fmin, s_value, N, variant_mutation_rate, Genovese_mean, Genovese_sigma, Genovese_total)\n",
    "        else:\n",
    "            Genovese_number = 0 \n",
    "\n",
    "        if i in Zink_missense_TET2:\n",
    "            Zink_number = prevalence_averaged_over_ages(Zink_fmin, s_value, N, variant_mutation_rate, Zink_mean, Zink_sigma, Zink_total)\n",
    "        else:\n",
    "            Zink_number = 0\n",
    "\n",
    "        if i in Coombs_missense_TET2:\n",
    "            Coombs_number = prevalence_averaged_over_ages(Coombs_fmin, s_value, N, variant_mutation_rate, Coombs_mean, Coombs_sigma, Coombs_total)\n",
    "        else:\n",
    "            Coombs_number = 0\n",
    "\n",
    "        if i in Young_missense_TET2:\n",
    "            Young_number = prevalence_averaged_over_ages(Young_fmin, s_value, N, variant_mutation_rate, Young_mean, Young_sigma, Young_total)\n",
    "        else:\n",
    "            Young_number = 0\n",
    "\n",
    "        if i in Desai_missense_TET2:\n",
    "            Desai_number = prevalence_averaged_over_ages(Desai_fmin, s_value, N, variant_mutation_rate, Desai_mean, Desai_sigma, Desai_total)\n",
    "        else:\n",
    "            Desai_number = 0\n",
    "\n",
    "        if i in Acuna_missense_TET2:\n",
    "            Acuna_number = prevalence_averaged_over_uniform_ages(Acuna_fmin, s_value, N, variant_mutation_rate, Acuna_mean, Acuna_total)\n",
    "        else:\n",
    "            Acuna_number = 0\n",
    "\n",
    "        expected_total = Jaiswal_number + Genovese_number + Zink_number + Coombs_number + Young_number + Desai_number + Acuna_number\n",
    "\n",
    "        expected_number_by_variant_missense[i] = expected_total\n",
    "    \n",
    "    return expected_number_by_variant_missense\n",
    "\n",
    "def prob_observed_given_expected_missense_TET2(params, data): #data = list of variant (e.g. p.R882H) (or all_missense_variants_TET2)\n",
    "    total_square_distance = 0\n",
    "    a = 0\n",
    "    for variant in data:\n",
    "        observed_number = float(all_studies_TET2_counts[variant])\n",
    "        expected_number = float(expected_number_variants_missense_TET2(data, params)[variant])\n",
    "        square_distance = ((observed_number - expected_number)**2)\n",
    "        total_square_distance = total_square_distance + square_distance\n",
    "\n",
    "    return total_square_distance\n",
    "\n",
    "def prob_observed_given_expected_missense_TET2_upper_error(params, data): #data = list of variant (e.g. p.R882H) (or all_missense_variants_TET2)\n",
    "    total_square_distance = 0\n",
    "    a = 0\n",
    "    for variant in data:\n",
    "        observed_number = float(all_studies_TET2_counts[variant])\n",
    "        observed_number_upper_error = observed_number+np.sqrt(observed_number)\n",
    "        expected_number = float(expected_number_variants_missense_TET2(data, params)[variant])\n",
    "        square_distance = ((observed_number_upper_error - expected_number)**2)\n",
    "        total_square_distance = total_square_distance + square_distance\n",
    "\n",
    "    return total_square_distance   \n",
    "\n",
    "def prob_observed_given_expected_missense_TET2_lower_error(params, data): #data = list of variant (e.g. p.R882H) (or all_missense_variants_TET2)\n",
    "    total_square_distance = 0\n",
    "    a = 0\n",
    "    for variant in data:\n",
    "        observed_number = float(all_studies_TET2_counts[variant])\n",
    "        observed_number_lower_error = observed_number-np.sqrt(observed_number)\n",
    "        expected_number = float(expected_number_variants_missense_TET2(data, params)[variant])\n",
    "        square_distance = ((observed_number_lower_error - expected_number)**2)\n",
    "        total_square_distance = total_square_distance + square_distance\n",
    "\n",
    "    return total_square_distance   "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 242,
   "metadata": {
    "code_folding": [
     0,
     91,
     102,
     114
    ]
   },
   "outputs": [],
   "source": [
    "def expected_number_variants_nonsense_TET2(data, params):\n",
    "    s_value = params[0]\n",
    "    \n",
    "    a = 0\n",
    "    expected_number_by_variant_nonsense = {}\n",
    "    \n",
    "    Jaiswal_fmin = 0.056\n",
    "    Jaiswal_mean = 58\n",
    "    Jaiswal_sigma = 11.5\n",
    "    Jaiswal_total = 17182\n",
    "    \n",
    "    Genovese_fmin = 0.1194\n",
    "    Genovese_mean = 55.1\n",
    "    Genovese_total = 12380\n",
    "    Genovese_sigma = 11.95\n",
    "    \n",
    "    McKerrel_fmin = 0.008\n",
    "    McKerrel_mean = 51.88\n",
    "    McKerrel_total = 4219\n",
    "    McKerrel_sigma = 18.36\n",
    "    \n",
    "    Zink_fmin = 0.2363\n",
    "    Zink_mean = 60\n",
    "    Zink_total = 11262\n",
    "    Zink_sigma = 1.18907720e+01\n",
    "    \n",
    "    Coombs_fmin = 0.019\n",
    "    Coombs_mean = 59.82\n",
    "    Coombs_total = 5649\n",
    "    Coombs_sigma = 15.14\n",
    "    \n",
    "    Young_fmin = 0.0009\n",
    "    Young_mean = 63\n",
    "    Young_total = 89\n",
    "    Young_sigma = 6.8\n",
    "    \n",
    "    Desai_fmin = 0.0169\n",
    "    Desai_mean = 66.1\n",
    "    Desai_total = 424\n",
    "    Desai_sigma = 6 #estimated\n",
    "    \n",
    "    Acuna_fmin = 0.0015\n",
    "    Acuna_mean = 4.5\n",
    "    Acuna_total = 2014\n",
    "    \n",
    "    N = 9.39358404e+04\n",
    "\n",
    "    for i in data:\n",
    "        variant_mutation_rate = mutation_rates_variants_TET2(i)\n",
    "\n",
    "        if i in Jaiswal_nonsense_TET2:\n",
    "            Jaiswal_number = prevalence_averaged_over_ages(Jaiswal_fmin, s_value, N, variant_mutation_rate, Jaiswal_mean, Jaiswal_sigma, Jaiswal_total)\n",
    "        else:\n",
    "            Jaiswal_number = 0\n",
    "\n",
    "        if i in Genovese_nonsense_TET2:\n",
    "            Genovese_number = prevalence_averaged_over_ages(Genovese_fmin, s_value, N, variant_mutation_rate, Genovese_mean, Genovese_sigma, Genovese_total)\n",
    "        else:\n",
    "            Genovese_number = 0 \n",
    "\n",
    "        if i in Zink_nonsense_TET2:\n",
    "            Zink_number = prevalence_averaged_over_ages(Zink_fmin, s_value, N, variant_mutation_rate, Zink_mean, Zink_sigma, Zink_total)\n",
    "        else:\n",
    "            Zink_number = 0\n",
    "\n",
    "        if i in Coombs_nonsense_TET2:\n",
    "            Coombs_number = prevalence_averaged_over_ages(Coombs_fmin, s_value, N, variant_mutation_rate, Coombs_mean, Coombs_sigma, Coombs_total)\n",
    "        else:\n",
    "            Coombs_number = 0\n",
    "\n",
    "        if i in Young_nonsense_TET2:\n",
    "            Young_number = prevalence_averaged_over_ages(Young_fmin, s_value, N, variant_mutation_rate, Young_mean, Young_sigma, Young_total)\n",
    "        else:\n",
    "            Young_number = 0\n",
    "\n",
    "        if i in Desai_nonsense_TET2:\n",
    "            Desai_number = prevalence_averaged_over_ages(Desai_fmin, s_value, N, variant_mutation_rate, Desai_mean, Desai_sigma, Desai_total)\n",
    "        else:\n",
    "            Desai_number = 0\n",
    "\n",
    "        if i in Acuna_nonsense_TET2:\n",
    "            Acuna_number = prevalence_averaged_over_uniform_ages(Acuna_fmin, s_value, N, variant_mutation_rate, Acuna_mean, Acuna_total)\n",
    "        else:\n",
    "            Acuna_number = 0\n",
    "\n",
    "        expected_total = Jaiswal_number + Genovese_number + Zink_number + Coombs_number + Young_number + Desai_number + Acuna_number\n",
    "\n",
    "        expected_number_by_variant_nonsense[i] = expected_total\n",
    "    \n",
    "    return expected_number_by_variant_nonsense\n",
    "\n",
    "def prob_observed_given_expected_nonsense_TET2(params, data): #data = list of variant (e.g. p.R882H) (or all_nonsense_variants_TET2)\n",
    "    total_square_distance = 0\n",
    "    a = 0\n",
    "    for variant in data:\n",
    "        observed_number = float(all_studies_TET2_counts[variant])\n",
    "        expected_number = float(expected_number_variants_nonsense_TET2(data, params)[variant])\n",
    "        square_distance = ((observed_number - expected_number)**2)\n",
    "        total_square_distance = total_square_distance + square_distance\n",
    "\n",
    "    return total_square_distance\n",
    "\n",
    "def prob_observed_given_expected_nonsense_TET2_upper_error(params, data): #data = list of variant (e.g. p.R882H) (or all_nonsense_variants_TET2)\n",
    "    total_square_distance = 0\n",
    "    a = 0\n",
    "    for variant in data:\n",
    "        observed_number = float(all_studies_TET2_counts[variant])\n",
    "        observed_number_upper_error = observed_number+np.sqrt(observed_number)\n",
    "        expected_number = float(expected_number_variants_nonsense_TET2(data, params)[variant])\n",
    "        square_distance = ((observed_number_upper_error - expected_number)**2)\n",
    "        total_square_distance = total_square_distance + square_distance\n",
    "\n",
    "    return total_square_distance   \n",
    "\n",
    "def prob_observed_given_expected_nonsense_TET2_lower_error(params, data): #data = list of variant (e.g. p.R882H) (or all_nonsense_variants_TET2)\n",
    "    total_square_distance = 0\n",
    "    a = 0\n",
    "    for variant in data:\n",
    "        observed_number = float(all_studies_TET2_counts[variant])\n",
    "        observed_number_lower_error = observed_number-np.sqrt(observed_number)\n",
    "        expected_number = float(expected_number_variants_nonsense_TET2(data, params)[variant])\n",
    "        square_distance = ((observed_number_lower_error - expected_number)**2)\n",
    "        total_square_distance = total_square_distance + square_distance\n",
    "\n",
    "    return total_square_distance   "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 243,
   "metadata": {
    "code_folding": [
     0
    ],
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "#CALCULATING S FOR MISSENSE VARIANTS\n",
    "TET2_missense_s = {}\n",
    "a = 0\n",
    "for var in all_missense_variants_TET2:\n",
    "    initial_guess=[0.05]\n",
    "    if var in all_studies_TET2_counts.keys():\n",
    "        observed_number = float(all_studies_TET2_counts[var])\n",
    "#         print(observed_number)\n",
    "        if observed_number >1:\n",
    "            variant_mutation_rate = mutation_rates_variants_TET2(var)\n",
    "            \n",
    "            MLE = scipy.optimize.minimize(prob_observed_given_expected_missense_TET2, initial_guess, args=([var],), method='Nelder-Mead')\n",
    "            s_value = MLE.x[0]\n",
    "\n",
    "            upper_error_MLE = scipy.optimize.minimize(prob_observed_given_expected_missense_TET2_upper_error, \\\n",
    "                                                      initial_guess, args=([var],), method='Nelder-Mead')\n",
    "            s_value_upper = upper_error_MLE.x[0]\n",
    "\n",
    "            lower_error_MLE = scipy.optimize.minimize(prob_observed_given_expected_missense_TET2_lower_error, \\\n",
    "                                                      initial_guess, args=([var],), method='Nelder-Mead')\n",
    "            s_value_lower = lower_error_MLE.x[0]\n",
    "\n",
    "#             print('a = ', a)\n",
    "            TET2_missense_s[var]=(s_value, observed_number, variant_mutation_rate, s_value_upper, s_value_lower) \n",
    "    a = a+1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 244,
   "metadata": {
    "code_folding": [
     0
    ],
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/anaconda3/lib/python3.7/site-packages/scipy/integrate/quadpack.py:385: IntegrationWarning: The integral is probably divergent, or slowly convergent.\n",
      "  warnings.warn(msg, IntegrationWarning)\n"
     ]
    }
   ],
   "source": [
    "#CALCULATING S FOR NONSENSE VARIANTS\n",
    "TET2_nonsense_s = {}\n",
    "a = 0\n",
    "for var in all_nonsense_variants_TET2:\n",
    "    initial_guess=[0.05]\n",
    "    if var in all_studies_TET2_counts.keys():\n",
    "        observed_number = float(all_studies_TET2_counts[var])\n",
    "#         print(observed_number)\n",
    "        if observed_number >1:\n",
    "            variant_mutation_rate = mutation_rates_variants_TET2(var)\n",
    "            \n",
    "            MLE = scipy.optimize.minimize(prob_observed_given_expected_nonsense_TET2, initial_guess, args=([var],), method='Nelder-Mead')\n",
    "            s_value = MLE.x[0]\n",
    "\n",
    "            upper_error_MLE = scipy.optimize.minimize(prob_observed_given_expected_nonsense_TET2_upper_error, \\\n",
    "                                                      initial_guess, args=([var],), method='Nelder-Mead')\n",
    "            s_value_upper = upper_error_MLE.x[0]\n",
    "\n",
    "            lower_error_MLE = scipy.optimize.minimize(prob_observed_given_expected_nonsense_TET2_lower_error, \\\n",
    "                                                      initial_guess, args=([var],), method='Nelder-Mead')\n",
    "            s_value_lower = lower_error_MLE.x[0]\n",
    "\n",
    "#             print('a = ', a)\n",
    "            TET2_nonsense_s[var]=(s_value, observed_number, variant_mutation_rate, s_value_upper, s_value_lower) \n",
    "    a = a+1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 245,
   "metadata": {
    "code_folding": [
     0
    ]
   },
   "outputs": [],
   "source": [
    "#Exac - retrieve the frequencies in Exac for each of the variants\n",
    "import csv\n",
    "filename=\"TET2_Exac.csv\"\n",
    "\n",
    "with open(filename, 'r') as csvfile:\n",
    "    read_reader = csv.reader(csvfile)  #csv.reader returns a reader object which will iterate over lines in the csvfile\n",
    "    row_count=0\n",
    "    Exac_TET2_results={} \n",
    "    \n",
    "    for row in read_reader:\n",
    "#         print(row)\n",
    "        if row_count>1:\n",
    "            variant = row[6]\n",
    "            allele_freq = float(row[14])\n",
    "            african_freq = float(row[15])/float(row[16])\n",
    "            east_asian_freq = float(row[18])/float(row[19])\n",
    "            european_freq = float(row[21])/float(row[22])\n",
    "            finnish_freq = float(row[24])/float(row[25])\n",
    "            latino_freq = float(row[27])/float(row[28])\n",
    "            south_asian_freq = float(row[33])/float(row[34])\n",
    "            Exac_TET2_results[variant]=(allele_freq, african_freq, east_asian_freq, european_freq, finnish_freq, latino_freq, south_asian_freq)\n",
    "\n",
    "        row_count=row_count+1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 246,
   "metadata": {
    "code_folding": [
     0
    ]
   },
   "outputs": [],
   "source": [
    "#COSMIC - retrieve the frequencies in COSMIC for each of the variants\n",
    "import csv\n",
    "filename=\"TET2_COSMIC_variant_counts.csv\"\n",
    "\n",
    "with open(filename, 'r') as csvfile:\n",
    "    read_reader = csv.reader(csvfile)  #csv.reader returns a reader object which will iterate over lines in the csvfile\n",
    "    row_count=0\n",
    "    TET2_COSMIC_results={} \n",
    "    \n",
    "    for row in read_reader:\n",
    "#         print(row)\n",
    "        if row_count>1:\n",
    "            variant = row[1]\n",
    "            counts = int(row[2])\n",
    "            TET2_COSMIC_results[variant]=counts\n",
    "\n",
    "        row_count=row_count+1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 247,
   "metadata": {
    "code_folding": [
     0
    ],
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "#Create dictionary of results and save to .csv file\n",
    "TET2_s_dictionary={}\n",
    "for k, v in TET2_missense_s.items():\n",
    "#     print(k)\n",
    "    if k in Exac_TET2_results.keys():\n",
    "        exac_freq = round((Exac_TET2_results[k][0])*100, 4)\n",
    "        african_freq = round((Exac_TET2_results[k][1])*100, 4)\n",
    "        east_asian_freq = round((Exac_TET2_results[k][2])*100, 4)\n",
    "        european_freq = round((Exac_TET2_results[k][3])*100, 4)\n",
    "        finnish_freq = round((Exac_TET2_results[k][4])*100, 4)\n",
    "        latino_freq = round((Exac_TET2_results[k][5])*100, 4)\n",
    "        south_asian_freq = round((Exac_TET2_results[k][6])*100, 4)\n",
    "    if k not in Exac_TET2_results.keys():\n",
    "        exac_freq = '-'\n",
    "        african_freq = '-'\n",
    "        east_asian_freq = '-'\n",
    "        european_freq = '-'\n",
    "        finnish_freq = '-'\n",
    "        latino_freq = '-'\n",
    "        south_asian_freq = '-'\n",
    "    if k in TET2_COSMIC_results.keys():\n",
    "        COSMIC_number = TET2_COSMIC_results[k]\n",
    "    if k not in TET2_COSMIC_results.keys():\n",
    "        COSMIC_number = '-'\n",
    "    TET2_s_dictionary[k]=(round(v[0]*100, 2),round(v[4]*100, 2),round(v[3]*100, 2), v[1], v[2], COSMIC_number, exac_freq, \\\n",
    "                            african_freq, east_asian_freq, european_freq, finnish_freq, latino_freq, south_asian_freq)\n",
    "\n",
    "for k, v in TET2_nonsense_s.items():\n",
    "    if k in Exac_TET2_results.keys():\n",
    "        exac_freq = round((Exac_TET2_results[k][0])*100, 4)\n",
    "        african_freq = round((Exac_TET2_results[k][1])*100, 4)\n",
    "        east_asian_freq = round((Exac_TET2_results[k][2])*100, 4)\n",
    "        european_freq = round((Exac_TET2_results[k][3])*100, 4)\n",
    "        finnish_freq = round((Exac_TET2_results[k][4])*100, 4)\n",
    "        latino_freq = round((Exac_TET2_results[k][5])*100, 4)\n",
    "        south_asian_freq = round((Exac_TET2_results[k][6])*100, 4)\n",
    "    if k not in Exac_TET2_results.keys():\n",
    "        exac_freq = '-'\n",
    "        african_freq = '-'\n",
    "        east_asian_freq = '-'\n",
    "        european_freq = '-'\n",
    "        finnish_freq = '-'\n",
    "        latino_freq = '-'\n",
    "        south_asian_freq = '-'\n",
    "    if k in TET2_COSMIC_results.keys():\n",
    "        COSMIC_number = TET2_COSMIC_results[k]\n",
    "    if k not in TET2_COSMIC_results.keys():\n",
    "        COSMIC_number = '-'\n",
    "    TET2_s_dictionary[k]=(round(v[0]*100, 2),round(v[4]*100, 2),round(v[3]*100, 2), v[1], v[2], COSMIC_number, exac_freq, \\\n",
    "                            african_freq, east_asian_freq, european_freq, finnish_freq, latino_freq, south_asian_freq)\n",
    "\n",
    "TET2_df = pd.DataFrame.from_dict(TET2_s_dictionary, orient='index', columns=['s (%)', 's lower value (%)', 's upper value (%)','observed number', 'site-specific mutation rate', 'Number of times in COSMIC',\\\n",
    "                                                                                'Exac Overall freq (%)', 'Exac African freq (%)', 'Exac East Asian freq (%)',\\\n",
    "                                                                                'Exac European freq (%)', 'Exac Finnish freq (%)', 'Exac Latino freq (%)', 'Exac South Asian freq (%)'])\n",
    "TET2_df_sorted = TET2_df.sort_values(by = ['s (%)'], ascending = False)\n",
    "\n",
    "TET2_df_sorted.to_csv('TET2_s_across_whole_gene.csv')\n",
    "# TET2_s_dictionary"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 248,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>s (%)</th>\n",
       "      <th>s lower value (%)</th>\n",
       "      <th>s upper value (%)</th>\n",
       "      <th>observed number</th>\n",
       "      <th>site-specific mutation rate</th>\n",
       "      <th>Number of times in COSMIC</th>\n",
       "      <th>Exac Overall freq (%)</th>\n",
       "      <th>Exac African freq (%)</th>\n",
       "      <th>Exac East Asian freq (%)</th>\n",
       "      <th>Exac European freq (%)</th>\n",
       "      <th>Exac Finnish freq (%)</th>\n",
       "      <th>Exac Latino freq (%)</th>\n",
       "      <th>Exac South Asian freq (%)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>p.L532*</th>\n",
       "      <td>17.32</td>\n",
       "      <td>13.02</td>\n",
       "      <td>20.34</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.460430e-10</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.Q916*</th>\n",
       "      <td>15.39</td>\n",
       "      <td>13.61</td>\n",
       "      <td>16.80</td>\n",
       "      <td>5.0</td>\n",
       "      <td>1.089990e-09</td>\n",
       "      <td>43</td>\n",
       "      <td>0.0008</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0061</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.S588*</th>\n",
       "      <td>12.79</td>\n",
       "      <td>10.38</td>\n",
       "      <td>14.17</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.095600e-09</td>\n",
       "      <td>4</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.R1359H</th>\n",
       "      <td>11.19</td>\n",
       "      <td>8.70</td>\n",
       "      <td>12.66</td>\n",
       "      <td>2.0</td>\n",
       "      <td>3.959610e-09</td>\n",
       "      <td>6</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.C1135Y</th>\n",
       "      <td>10.70</td>\n",
       "      <td>8.79</td>\n",
       "      <td>11.75</td>\n",
       "      <td>2.0</td>\n",
       "      <td>3.213910e-09</td>\n",
       "      <td>3</td>\n",
       "      <td>0.0008</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0087</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.Q3*</th>\n",
       "      <td>10.64</td>\n",
       "      <td>8.79</td>\n",
       "      <td>11.63</td>\n",
       "      <td>2.0</td>\n",
       "      <td>3.213910e-09</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.Q530*</th>\n",
       "      <td>10.58</td>\n",
       "      <td>8.24</td>\n",
       "      <td>11.74</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.412750e-09</td>\n",
       "      <td>4</td>\n",
       "      <td>0.0008</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0015</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.R550*</th>\n",
       "      <td>10.29</td>\n",
       "      <td>9.37</td>\n",
       "      <td>10.96</td>\n",
       "      <td>7.0</td>\n",
       "      <td>9.720770e-09</td>\n",
       "      <td>61</td>\n",
       "      <td>0.0058</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0116</td>\n",
       "      <td>0.0075</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0087</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.R1516*</th>\n",
       "      <td>8.21</td>\n",
       "      <td>6.73</td>\n",
       "      <td>9.04</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.883910e-08</td>\n",
       "      <td>26</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          s (%)  s lower value (%)  s upper value (%)  observed number  \\\n",
       "p.L532*   17.32              13.02              20.34              2.0   \n",
       "p.Q916*   15.39              13.61              16.80              5.0   \n",
       "p.S588*   12.79              10.38              14.17              2.0   \n",
       "p.R1359H  11.19               8.70              12.66              2.0   \n",
       "p.C1135Y  10.70               8.79              11.75              2.0   \n",
       "p.Q3*     10.64               8.79              11.63              2.0   \n",
       "p.Q530*   10.58               8.24              11.74              2.0   \n",
       "p.R550*   10.29               9.37              10.96              7.0   \n",
       "p.R1516*   8.21               6.73               9.04              2.0   \n",
       "\n",
       "          site-specific mutation rate Number of times in COSMIC  \\\n",
       "p.L532*                  2.460430e-10                         -   \n",
       "p.Q916*                  1.089990e-09                        43   \n",
       "p.S588*                  1.095600e-09                         4   \n",
       "p.R1359H                 3.959610e-09                         6   \n",
       "p.C1135Y                 3.213910e-09                         3   \n",
       "p.Q3*                    3.213910e-09                         -   \n",
       "p.Q530*                  2.412750e-09                         4   \n",
       "p.R550*                  9.720770e-09                        61   \n",
       "p.R1516*                 1.883910e-08                        26   \n",
       "\n",
       "         Exac Overall freq (%) Exac African freq (%) Exac East Asian freq (%)  \\\n",
       "p.L532*                      -                     -                        -   \n",
       "p.Q916*                 0.0008                     0                        0   \n",
       "p.S588*                      -                     -                        -   \n",
       "p.R1359H                     -                     -                        -   \n",
       "p.C1135Y                0.0008                     0                        0   \n",
       "p.Q3*                        -                     -                        -   \n",
       "p.Q530*                 0.0008                     0                        0   \n",
       "p.R550*                 0.0058                     0                   0.0116   \n",
       "p.R1516*                     -                     -                        -   \n",
       "\n",
       "         Exac European freq (%) Exac Finnish freq (%) Exac Latino freq (%)  \\\n",
       "p.L532*                       -                     -                    -   \n",
       "p.Q916*                       0                     0                    0   \n",
       "p.S588*                       -                     -                    -   \n",
       "p.R1359H                      -                     -                    -   \n",
       "p.C1135Y                      0                     0               0.0087   \n",
       "p.Q3*                         -                     -                    -   \n",
       "p.Q530*                  0.0015                     0                    0   \n",
       "p.R550*                  0.0075                     0               0.0087   \n",
       "p.R1516*                      -                     -                    -   \n",
       "\n",
       "         Exac South Asian freq (%)  \n",
       "p.L532*                          -  \n",
       "p.Q916*                     0.0061  \n",
       "p.S588*                          -  \n",
       "p.R1359H                         -  \n",
       "p.C1135Y                         0  \n",
       "p.Q3*                            -  \n",
       "p.Q530*                          0  \n",
       "p.R550*                          0  \n",
       "p.R1516*                         -  "
      ]
     },
     "execution_count": 248,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "TET2_df_sorted"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 271,
   "metadata": {
    "code_folding": [
     0
    ]
   },
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAABZ8AAAHvCAYAAAA7GaHXAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDIuMi4zLCBodHRwOi8vbWF0cGxvdGxpYi5vcmcvIxREBQAAIABJREFUeJzs3Xd4VFX+x/H3CSEVktACUoMFUFQQVBBwCVVBqaIguBKx7MIKsnZdNWBZYa0gAq6KoqKrohRRCChBiqD00HuoIZCE0EIg5fz+mPKbyUxCCyTq5/U8eSZz7rnnfu89c+eRb47fa6y1iIiIiIiIiIiIiIgUp4CSDkBERERERERERERE/niUfBYRERERERERERGRYqfks4iIiIiIiIiIiIgUOyWfRURERERERERERKTYKfksIiIiIiIiIiIiIsVOyWcRERERERERERERKXZKPouIiIj8jhljAo0x1hjzQUnHUhoZYwYbYzYZY045r1NNZ3s7Y8yvxphjzvZ7SjpWEREREZE/GiWfRUREREohY0ysMyla2E/zIvYNMMYMM8Z0vZgxlzbGmA7AaGAt8BDwVyDDGFMZ+BYIAR51ti+6QDE0cc5F7Qsx/p+RMWbPae4Nz59WHn+gKfJ+Msa8fBbjfuCMpbsx5mNjzEZjzHFjzD5jzBxjTMeSvk4iIiIipUFgSQcgIiIiIkX6AvjBT/tWAGttrjEmFMj12BYAxAMfAtMveISlVwfn633W2iOuRmNMLBABPG+tvdDXpwmOufgR2HWBj/VnMRgI93jfEHgamAxMK9B3k8fvy4G3CxlzK/A1sLFA+2jgFPC4n/4AHwDpzuNuBioC9wMJxpinrbUjT3cyIiIiIn9kSj6LiIiIlG4rrLWfFdXBWpt9sYL5nakG5Hkmnj3aATIucjy/O8YYA4Rba4+VdCwu1topnu+NMe1xJJ9X+7tXjDGuf/PsOc29lAasLrDvCCC7iP3ustbOLbDPu85xhhtjxltrDxd5QiIiIiJ/YCq7ISIiIvI7VrDmszHmciDHufl+jzIBuQX7O0sSLDDGZBlj0owx/zXGhPs5Rg1jzHhjzG5n7eS9zveVC/SrZIwZZYzZbozJNsakG2OWGWMeLdAvzhiz1BiT6SxVsM0Y85kxptIZnvMtztIGh53HWW2Mechj++XGGIujnEYZj2vwozFmD44V4QALPK+Nc98QY8xzxpj1zrEPGWOmG2Ma+YnDGGP+Zoz5zThqRx81xiQZY+Kd218G3i9wLM+5CjXGvGgcNamznMda40x4nu4aBDrjXGCMSXXOy05jzLvGmIqF7HOnMeZn53XLcpaKGOVKzhpj2jvj+6tx1MreAJwEhnqMEeu8jkecYyw3xsT5OdY1xphvnJ+Vk8aYFGPMXGNMJ48+53z+pUXBxLOz7TjwPRAMXHHRgxIREREpRbTyWURERKR0CyuY5AVOWmuPFtJ/P9AfmAjM4/8TrfkF+jUFeji3fwa0BR7EUb5jkKuTMaYu8AtQxtl3O46E2kCgjTHmBo+Vxd8CNwHjgSQcpRGuBGKBN53j3QdMAH4GXgBOALWBzkBlHCUMCmWMGQi864zpJSALuAV4zxhT11r7jPMa/NUZY3Pn9QBIwVFu4zYcpRFewlEqId85dhAwG2gGfIKj5EIF53X5xRjTylq70tnX4CiJ0htYDLwCZDrP9w5gOI4yDtUKHAv+v2TDeOBe4GPnGGWd17ZtUdfAKQR4DPgGmAocB27EUdu6pXNeXH+EwBgzEngSWAe8AaQClzlj/RfgubL5Med5f+i8ljudY3THUdoiBXjNecy7gY+c196VdK8CzAXynOe4C8fc3uCMcWYxnP+5CPJzLwHkW2uLexV8TefrgWIeV0REROR3xVhrSzoGERERESnAOOoSJxay+UtrbR9nv0AcK50/tNY+UFibx7iubflAM2vtMo9tCTgSxVHW2hPOtu9xJKqbWGv3efRthiMBHG+tfdm52jYdeMdaO6SI85oOtAKqWGvzzvByuPatCWxznv+9Bba9C/wNuMxa60qWfgb0sdYGFuj7AI4VyTdbaxd6tD8BjAQ6Wmt/9GiPwvHQwo3W2vbOtr7AJBxJ/gHW2nyP/gGu94Udy7ntCDDPWnvWD4Y0xgQAwa558mj/G46k7h3W2m+dbS1wPFDxR+B2a+1Jj/7GOv9BYBzlK+bgmMcG1to0j35lgR1AGHCVtXa/sz0Yxx8SbgCusNZuN8b0xJEUd8dQyDmc8/kXGMcV9/PW2pf9bHd95gtz2FobVcjYe3CU3bj8LOJpAvwGzLfWXqhEuoiIiMjvgspuiIiIiJRu/8Xx4DzPH58E2zlY6Jl4dpoLBAF1AJwJ5U44VtaeMsZUdv3gSALvADo6983CkeBrboypU8RxDwPlgc7O1cNn405nfBM8Y3HG8x2O1dntznJMT/fgWBm8qsDYgcBPQGtnshWgH2CBxz0TzwAF3xfhMHCNMabh2QZqrc33+ANBGWNMlDNWVxmIZh7d+zlfn/ZMPDvH8bcS5WPPxLPTDUAN4ANX4tm5/0ngdRz/rnAlkV01jjsbY8oXcRrnfP7n6Bd876UOQLfiOoAxpiqOxPtxHKvQRURERP7UVHZDREREpHTb4rkKtxht99PmKnnhqr3cADA4VhT/rZBx8sDx0EPjqO38JpBsjFmHIxE6xVrruYL7ZRwrn6cDacaYn3GUYfjyDB5qd6XztbAV4QBVTzNGURrgSG4fLKJPRRxlJ67A8QC7gknas/EIjpXTa40x23Cc13RgRiFJYS/GmD44SmQ0wlGywlMFj9+vwDFPSWcY12Y/bXWdr+v8bFvrfL3U+ToX+BxHuZF7jTFLcZQz+dJau9Fjv/M6/3Nw8ALdSwA4k/8/4vgMdrbWbj3NLiIiIiJ/eEo+i4iIiPw5FVXywhR4nYijLrQ/Wa5frLVjjDFTcNRU/gtwFzDYGDPJWnuPs88mY0wDoD2OVcqtgQ+AYcaYv1hrd5xBXP0ovJbu+ST8AoBVwBNF9HHVBjY4Vj6fM2vtt87ke2cc16ED8AAwzxjT0bNmc0HGmLtw1JxeAgwB9gDZOJLn3+P9fzie7QrzLD9tZzyGM3Hcz/ngwE7AzTjqTT9vjBlsrR3n7HfO51/aGMfDMn8ELge6WmvnlWxEIiIiIqWDks8iIiIifzzFtWp0q3Ossme6YtRauxdHqZD/OmvtTsKRiHzD9bA+Z6mG750/GGO6AtOAoThWwxZmi/P1Qq1g3QJUAX46g5W3m3CUlah8mtXPRY5jrU0HPgU+dZYheQ3HaubbgSlF7PpXHEniNtbabFejMebqQmJtD1wDrCgqniJsc776K5FxlfPVazW9tXYNsAb4j7OEy2/ACGCcR59zPf9Sw5l4/gnHyvmu1to5JRySiIiISKmhms8iIiIifzDOB/ll4ygRcT7jpOIol3CnMeaGgtuNQxXn72HGmNAC++fiSD7iisVZmqCgFZ59ivAlcAp40RgT4ieeKGNM0GnGKMonOOoa+02AO+v5ukzCsRr4PwVrVxd47yolUrFAn0BjTKRnmzPhvcpffz/ycDw00v3f887jPuen7+fO11cLXp+zqLu9FNgL3G+MifbYPwh43BnLdGdbxYLjWmszgGSgnDEmqBjOv1RwJtV/xJF47mGtnV3CIYmIiIiUKlr5LCIiIvLHtAS4xRjzJLAbyLPWfnUO4/wNWAgsNMZ8AqzE8d+QdYHuwIc46jhfBfzoLLuxDjjkbBuIY9XsIud4c40xB4EFzrgqAvfhSF4WVtoDAGvtTmPMw8B4YL0x5jNgF47VytfieHBcPRwlKM7FmzhWCL9ljGkPzAOOArVxlAg5iqM0BMD/gJ7O2OsZY77D8QC9es6+jZz9fsOx+vk5Z6L+OP//sMZdxpjpOBKuB3DUTB6Io7TH96eJdbLzfOcaYz4FgoEegE9S3lr7izHmDRwripcZY74G9uOYwzuNMdedrt62tTbXGDMY+BpYaox533kufYAbgRetta6VzwOAh52fha1ALtDGeV0+t9aecv4R4nzO/1zUNMbcU8i2XzziPyPOBPuPQGMcf4yo5Gf8RacpJSMiIiLyh6bks4iIiMgf09+BMThWwpbHsVL2rJPPzoRvE+BpoCtwL45yD7uBqTiSoAA7cdSGjsWRBA3GkQR+DxjpURriXeBOZ3wVcTzkcAUw6Ezq5Fpr3zfGbMCx2nYgEAmkARuBf1H0wwJPN/YpY8ytwMPAPcBw56Z9OJL5Ez36WmNMb+AfOJKt8Tiu8XYcCVpXvx3GmAdw1Dweh+PBgB869xuNIyHbAQjH8SDDKcCr1tr9p4n1M2NMORyrtN/AkbCdDjyPn3rY1trHjTErncd9Cseq7d3ADByr5M/k+kwxxnTA8Zl6Cse/JTYAA6y1H3l0nYvjjwFdgEtwJJ934Eh+j3H2OXo+53+OmuIo8eHPffh/CGdRygDXOX/v5/wp6K84zl1ERETkT8lcmAdJFw9jzEjgehwrSCoDJ3D8w2YqMMZZI67gPi1w/AdxcxwrP7YCE4B3nP8LqoiIiIiIiIiIiIhcYKU9+XwKx0qY9ThWcITjSCpfj2MFSnNr7W6P/t2Ab3Cs3vgSxwqQLkB9YLK19s6LegIiIiIiIiIiIiIif1KlPfkc4vn0bo/2V4BngXHW2kHOtggcq5wjgZbW2mWuMXD8r383AXdba/93seIXERERERERERER+bMKOH2XkuMv8ezkqld4hUdbLxwPm/mfK/HsMYbrqd8Diz1IEREREREREREREfFRqpPPRejifE3yaGvrfJ3lp/98HA/GaWGMCb6QgYmIiIiIiIiIiIiI4wnVpZ4x5nGgHI6SGtcDrXAknkd4dKvvfN1ccH9rba4xZgfQELgUx1O5/Ro3bpwFGDhwoCmW4EVERERERERERET+hH4vK58fB+KBoTgSz7OAjtbagx59Ip2vhwsZw9Ue5W+jMeYhY4y7XMfhw4ftypUrbU5Ojl22bJkF7LJly+zJkyft6tWr7aFDh+zWrVvtnj17bEpKit20aZM9cuSIXblypc3NzfXZZ82aNTYjI8Nu27bN7t692+7fv99u3LjRHj161C5fvtzm5eV57XPixAm7du1am56ebrdv32537dplU1NT7YYNG+yxY8fssmXLbH5+vtc+WVlZdt26dfbgwYN2x44ddufOnfbAgQN2/fr19vjx4+6+S5cude9z/Phxu379envgwAG7c+dOu2PHDnvw4EG7bt06m5WV5TW+63jHjh2zGzZssKmpqXbXrl12+/btNj093a5du9aeOHHCa5+8vDy7fPlye/ToUbtx40a7f/9+u3v3brtt2zabkZFh16xZY0+ePOm1T25url25cqU9cuSI3bRpk01JSbF79uyxW7dutYcOHbKrV6/22ScnJ8euXLnSHj582G7evNnu27fP7t27127ZssVmZmbaVatW2ZycHK9zP3XqlHsut2zZYvfu3Wv37dtnN2/ebD3n37XP0qVL7cmTJ21SUpJ7LgvO/4oVK3zmMjs7265Zs8ZrLgvOf8G5dM1/Wlqaey5TU1Pt+vXr3fNvrS10/pOTk21ycrI9ePCgz/x7vh47dsxn/tPS0uzatWttVlaW1/XKz8+3y5cvL3L+s7OzvT5neXl5dsWKFfbIkSN+5z8pKanQ+T98+LDdtGmT3bdvn8/8nzp1ymf+V61aZTMzM+3mzZvt3r177d69e+3mzZu95t9zH8/5d93L+/bts5s2bXLPv7972XP+Pe9lz/n3vG7Z2dlF3ssF53/p0qU2KyvLZ/5d97Jr/gvOpeteLjj//u5l12fHNf+pqak+81/wXnbNv+e97Dn/a9as8Zp/1/3vmn/Pe7ng/HveY0Xdy/7m37WPa/6Lupf1Xa7vcn2X67tc3+X6Ltd3ub7L9V2u73J9l+u7XN/l+i7/M36XU4JK9QMHCzLGVAVa4FjxXB643Vq7wrltM44a0FdYa7f62fcXHA8dvMlau6SwY3isfC7+ExARERERERERERG5uEqswsPvZeUzANbaVGvtFKAjUAn4xGOza2VzpM+ODhEF+omIiIiIiIiIiIjIBfK7Sj67WGt3AuuBhsaYys7mTc7XegX7G2MCgbpALrD9ogQpIiIiIiIiIiIi8if2u0w+O1V3vuY5X+c6X2/10/cvQBjwi7X25IUOTEREREREREREROTPrtQmn40xDYwx1fy0BxhjXgGicSSTDzk3TQbSgD7GmOs9+ocALzvfjrvAYYuIiIiIiIiIiIgIEFjSARThVuA1Y8x8YBuQDlQFWgOXAvuBB12drbVHjDEP4khCzzPG/A/IALoC9Z3tX17UMxARERERERERERH5kyrNyecfgf8CLYFGQBRwHNgMfAqMttZmeO5grZ1qjGkN/Au4AwgBtgKPOvvbixe+iIiIiIiIiIiIyJ9XqU0+W2vXAv84h/0WAZ2LPyIREREREREREREROVOltuaziIiIiIiIiIiIiPx+KfksIiIiIiIiIiIiIsVOyWcRERERERERERERKXZKPouIiIiIiIiIiIhIsVPyWURERERERERERESKnZLPIiIiIiIiIiIiIlLslHwWERERERERERERkWKn5LOIiIiIiIiIiIiIFDsln0VERERERERERESk2Cn5LKXKsGHDMMYwb968kg5FREREREREREREzoOSzyIiIiIiIiIiIiJS7JR8llLl4YcfZsOGDdx4440lHYqIiIiIiIiIiIich8CSDkDEU+XKlalcuXJJhyEiIiIiIiIiIiLnSSuf5bwlJydjjCEuLo5t27bRq1cvKlWqRPny5enYsSNr164F4ODBgzz00ENccsklhISEcMMNN5CYmOg1VmE1nxcsWECXLl2oWbMmwcHBVKtWjebNmzN8+HCvfqmpqTz++OPUr1+f8PBwoqKiqF+/PnFxcWzfvt0n9oSEBDp37kzlypUJDg7msssu44knniAzM9Onb0xMDDExMWRlZfHEE09Qu3ZtgoODufzyyxk5ciTWWp99pk+fTrt27bjkkksIDg6mevXqtG7dmrFjx/r0zcjI4JlnnuHKK68kNDSUyMhI2rVrx+zZs087ByIiIiIiIiIiIqWNVj5LsUlOTqZZs2ZceeWVxMXFkZyczJQpU4iNjWXx4sXceuutRERE0Lt3bzIyMvjf//5Hp06d2Lx5M7Vr1y503FmzZnHbbbcRERFB165dqVGjBhkZGWzYsIGxY8cSHx8PQFZWFi1btmTbtm106NCBLl26YK1l586dTJs2jV69enHppZe6x33xxReJj4+nYsWK3H777URHR5OUlMTrr7/ODz/8wOLFi4mIiPCKJScnh44dO7Jv3z46depEYGAgU6dO5emnnyY7O9sdC8B///tf/va3v1GtWjW6dOlC5cqVOXDgAElJSXz00UcMGjTI3Xfnzp3ExsaSnJzMzTffzK233srx48eZMWMGt956K++99x4PPvhgcU2ViIiIiIiIiIjIBafksxSbn3/+mZdffpl//etf7raXXnqJF154gWbNmnHXXXcxduxYAgIcC+47dOjAvffey1tvvcVbb71V6Ljvv/8++fn5zJs3j0aNGnltS0tLc//+008/sW3bNoYOHeoz3qlTpzh58qT7fWJiIvHx8dx000388MMPREVFubd9/PHH3HfffcTHx/uMs2/fPho1asScOXMIDQ0FID4+nnr16vHWW2/x7LPPUrZsWQDee+89goKCWL16NdHR0YXGDdC/f3927tzJF198QZ8+fdztmZmZxMbGMmTIELp27UrVqlULvU4iIiIiIiIiIiKliZLPF0jfJTklHcIZ+7x52WIZJyYmhqefftqrrX///rzwwgucPHmS1157zZ14Bujbty8DBgxg1apVZzS+K9nryV99aH/9goKCCAoKcr8fPXo04EhseyaeAeLi4hg1ahSTJk3ymxQfPXq01zGio6Pp1q0bn3zyCZs2beLqq692bwsMDHQnowuLe/Xq1fz888/06tXLK/EMEBUVxfDhw+nevTvffPON12ppERERERERERGR0kzJZyk2jRs3pkyZMl5t1atXB6BevXqUL1/ea1uZMmWoWrUqe/bsKXLcfv368e2339KsWTN69+5NmzZtaNmyJTVr1vTq17p1a2rUqMGIESNYsWIFnTt3pmXLln7jWrx4MWXLluXrr7/m66+/9jnmqVOnOHjwIOnp6VSqVMndHhkZyeWXX+7Tv1atWgAcOnTIK+7HHnuMhg0b0rt3b1q3bk3Lli2pUqWKTywAhw8fZtiwYT5jHzx4EIANGzYUdZlERERERERERERKFSWfpdhERkb6tAUGBha6zbU9J6foVeI9e/ZkxowZvPHGG0yYMIH33nsPgKZNm/Lqq6/SoUMHACIiIliyZAnx8fFMnz6dhIQEwLHKeNCgQTz33HPuVcjp6enk5ub6PLCwoGPHjnklnwuuki54nnl5ee62Rx99lMqVKzN27FhGjx7N22+/jTGG1q1b89prr3H99de7YwGYM2cOc+bMKTIWERERERERERGR3wslny+Q4iplIQ633XYbt912G8ePH+fXX39lxowZjBs3jttvv52VK1dy1VVXAVCzZk0+/PBDrLWsX7+euXPn8u677/Liiy+Sn5/PSy+9BDiS4fn5+WRkZFzQuO+9917uvfdeMjMz+eWXX5gyZQoTJkzglltuYcOGDURHR7sT86NGjWLIkCEXNB4REREREREREZGLJeD0XURKj/DwcNq2bcubb77Js88+y6lTp5g5c6ZPP2MMDRs2ZPDgwe7VxFOnTnVvb968OYcOHWLdunUXJe6oqCg6d+7M+++/T1xcHBkZGSxYsMAdC+B+LyIiIiIiIiIi8keg5LOUej/99BMnTpzwaU9NTQUgLCwMgLVr15KcnHzafgD//Oc/AXjwwQfZt2+fzz7Hjx9nyZIl5xX3rFmzyM3N9Wk/cOCAVzzXX389N998M99++y0TJkzwO9aaNWvc+4mIiIiIiIiIiPweqOyGlHqPPfYYycnJxMbGEhMTQ1BQEMuXL2fu3LnUqVOHPn36APDjjz/y6KOP0qJFCxo0aEB0dDR79uxh2rRpBAQE8MQTT7jHbNeuHSNGjOCZZ57hiiuuoHPnztStW5djx46xc+dOfv75Z1q1asWsWbPOOe4+ffoQEhJCq1atiImJwVrLggULWLp0KU2bNqV9+/buvp9//jlt27bl/vvvZ/To0TRr1oyoqCj27NlDUlISa9euZfHixURHR5/7hRQREREREREREbmIlHyWUu/ZZ59lypQpLFu2jB9//JGAgABq167Ns88+y9ChQ6lQoQIAt9xyC0OHDmX+/PlMmzaNI0eOcMkll9ChQwd3UtrTU089RcuWLRk9ejQLFy5k2rRpREZGUqNGDR566CH69u17XnGPGDGChIQEVqxYwQ8//EBISAh16tRh5MiRDBw40P3wQ3DUql6+fDnvvPMO33zzDZMmTSIvL49q1apx1VVXMXjwYK655przikdERERERERERORiMtbako6hVBk3bpwFGDhwYEmHIiIiIiIiIiIiInK+TEkdWDWfRURERERERERERKTYKfksIiIiIiIiIiIiIsVOyWcRERERERERERERKXZKPouIiIiIiIiIiIhIsVPyWURERERERERERESKnZLPIiIiIiIiIiIiIlLslHwWERERERERERERkWKn5LOIiIiIiIiIiIiIFDsln0VERERERERERESk2Cn5LCIiIiIiIiIiIiLFTslnERERERERERERESl2Sj6LiIiIiIiIiIiISLFT8llEREREREREREREip2SzyIiIiIiIiIiIiJS7JR8FhEREREREREREZFip+SziIiIiIiIiIiIiBQ7JZ9FREREREREREREpNgp+SznLTk5GWMMcXFxJCcn06dPHypXrkxISAjXX389M2bM8Nnn5MmTjBgxgmuvvZawsDAiIiK4+eab+eqrr4pl/FOnTjF69GiaNGlChQoVCAsLIyYmhm7duvHjjz/69N+4cSNxcXHUqlWL4OBgqlatSt++fdm0aZNP37i4OIwxJCcn895773HNNdcQEhJC1apVeeihhzh8+LDPPklJSdx9993ExMQQHBxMlSpVaNKkCUOHDiUnJ8erb25uLmPHjqV58+ZEREQQFhbGddddx5gxY8jPzy9yLkREREREREREREoLY60t6RhKlXHjxlmAgQMHlnQovxvJycnUrVuX2NhY1q1bx6WXXspNN91ERkYGX375JTk5Ofz444+0adMGcCSGO3bsyM8//0yDBg24/fbbycrKYvLkyRw4cIBnnnmGf//73+c8PkDfvn354osvuPrqq2nbti2hoaHs27ePhQsX0rNnT15//XV331mzZtGzZ09ycnLo0qULl19+OXv27OHbb78lODiYxMREmjRp4u4fFxfHxIkTufPOO0lISKBLly5UrVqVxMREVq5cSZs2bZg7d667f1JSEs2aNcMYQ9euXalbty5Hjhxh69atJCYmkpGRQbly5QDcMSQkJFC/fn1iY2MJCQkhMTGRpKQk7rnnHj799NMLNpciIiIiIiIiIvKHY0rsyNZa/Xj8jB071o4dO9bKmduxY4cFLGCHDRvmtW3WrFkWsJ06dXK3/fvf/3a35eTkuNtTU1NtnTp1LGAXLVp0zuNnZmZaY4xt2rSpzc3N9Yk3LS3N/XtGRoaNioqylSpVsuvWrfPqt3btWhseHm6vu+46r/b+/ftbwNaqVcvu3LnT3Z6Tk2NvvvlmC9hff/3V3f7oo49awE6dOtUnloyMDJuXl+d+Hx8fbwH78MMPe8Wem5trBwwYUOg4IiIiIiIiIiIihSixXGtgiWS8/wRM99ElHcIZs1OHFMs4derU4bnnnvNqu+WWW6hduza//fabu23ChAkYY3jzzTcJDPz/j2B0dDTPP/88DzzwAB988AEtWrQ4p/GNMVhrCQ4OJiDAt7JMpUqV3L9/8sknZGZmMmbMGK666iqvfg0bNuTBBx/k7bffZv369T7bX3jhBWrXru1+HxgYyH333ceCBQv47bffuPHGG736h4aG+sRSoUIF9+/5+fmMGTOGatWq8dZbb1GmTBn3tjJlyvDGG2/w0UcfMWmwSU9OAAAgAElEQVTSJLp16+YzloiIiIiIiIiISGmi5LMUm8aNG3slTF1q1arF4sWLATh69Chbt26lRo0aNGjQwKdv27ZtAVi5cuU5jQ8QERFBly5d+O6772jcuDF33HEHN998M82aNSMsLMxrX9d+q1evZtiwYT5jb968GYANGzb4JJ+vv/56v7EAHDp0yN3Wu3dvRo0aRffu3enVqxft27enZcuWXHbZZT7HSk9P54orruDll1/2GRscCewNGzb43SYiIiIiIiIiIlKaKPksxSYqKspve2BgoPtBea6H8V1yySV++7raMzMzz2l8ly+//JKRI0fy+eefEx8fD0BISAi9evXi9ddfp2rVqgCkp6cD8P777xd5bseOHTujeFwrufPy8txtN954IwsWLOCVV15h8uTJ7prN9evXJz4+nrvvvtsrli1btjB8+PCzikVERERERERERKS0UfL5AimuUhZ/NJGRkQDs37/f7/aUlBSvfucqNDSUYcOGMWzYMHbv3s38+fP5+OOP+eyzz0hOTmbBggVex1m9ejXXXnvteR2zKDfddBMzZszg5MmTLF++nFmzZvHOO+/Qt29fqlSpQvv27d2x9OjRg2+//faCxSIiIiIiIiIiInIx+BbEFbmAypcvz2WXXcbevXvZsmWLz/bExEQAmjRpUmzHrFWrFv369SMhIYErrriChQsXulcZN2/eHMCdjL7QgoODadGiBS+++CKjRzvqgk+bNg2ABg0aEBUVxZIlS8jJybko8YiIiIiIiIiIiFwoSj7LRTdgwACstTzxxBNe5SnS0tJ46aWX3H3O1cGDB/n111992o8fP87Ro0cJDAwkKCgIgPvuu4+oqCiGDx/u9dBCl/z8fObNm3fOsYAjse0qN+IpNTUVwF2HOjAwkMGDB5OSksKQIUM4ceKEzz4pKSmsX7/+vOIRERERERERERG5GFR2Qy66xx9/nJkzZzJt2jQaNWpE586dycrK4uuvv+bAgQM8+eSTtGrV6pzH37t3L82bN+fKK6+kSZMm1KpViyNHjjBjxgz279/PkCFDKF++PACVKlVi8uTJ9OjRg+bNm9OuXTsaNmxIQEAAu3btYvHixaSnp5OdnX3O8bzxxhvMnj2b2NhYLr30UsqVK8e6deuYOXMmFSpU4KGHHnL3ff7551m9ejXjx4/nu+++o23bttSoUYMDBw6wZcsWFi1axCuvvOLz8EMREREREREREZHSptQmn40xlYAewG3ANUAN4BSwBvgI+Mham+/RPwbYUcSQX1pr+1yoeOXMBQUFMWfOHN58800+//xz3nnnHQIDA2nUqBFvv/22+wF85yomJobhw4czb948EhMTSUtLo2LFitSvX58RI0bQp4/3x6Bdu3YkJSXx+uuvk5CQwIIFCwgKCqJ69eq0bduWO+6447ziGTRoEBUqVODXX39l0aJF5ObmUrNmTQYNGsRjjz1GnTp13H3Lli3L1KlT+eyzz/j444+ZMWMGx44do0qVKtStW5eXXnqJfv36nVc8IiIiIiIiIiIiF4Ox1pZ0DH4ZY/4OjANSgERgF1AV6AlEAt8Ad1rnCXgkn1cDU/0MudZaO/l0xx03bpwFGDhw4Hmfg4iIiIiIiIiIiEgJMyV14FK78hnYDHQFvi+wwvlZ4DfgDhyJ6G8K7LfKWjvsYgUpIiIiIiIiIiIiIr5K7QMHrbVzrbXfeSaene37gfHOt7EXPTAREREREREREREROa3SvPK5KDnO11w/26obY/4GVALSgcXW2qSLFpmIiIiIiIiIiIiI/P6Sz8aYQOBe59tZfrp0cP547jMP6G+t3XVhoxMRERERERERERERgDLDhg0r6RjOyvDhw/8DdAd+sNa+6NEeBGQDjwJPAyOBmUAM0AboMnz48AnDhg3L8RkUMMY8NHz48Pduu+226gD16tVjw4YNVK5cmZUrV1K9enWWLVtGpUqVWLduHcHBwezdu5esrCyOHz/Ovn37CAoKYv369VSpUoUVK1Z47bNhwwbKli3Lvn37OH78OCdOnGDPnj0EBwezdu1aqlat6rVPhQoV2LRpE2XKlGH//v0cPXqU7Oxsdu/eTWhoKElJSVSrVo3ly5e794mKimLz5s0YYzhw4ABHjhzh1KlT7Ny5k7CwMJKSkqhevTpLly6lRo0aLFu2jMjISLZu3Yq1lrS0NA4dOkRubi7JycmUK1eO1atXu8d3HS8iIoJt27aRn59PRkYG6enp5Ofns337dsqXL++1T9WqVVm5ciXly5dnx44d5ObmcujQIdLS0gDYunUrUVFRrFq1yr1PdHQ0q1evJjw8nOTkZHJycjh8+DAHDhzAGMPmzZupUKGC1z5VqlQhKSmJsLAwdu3axcmTJzly5AipqamUKVOGjRs3UqlSJVasWOE+98qVK7N27VqCg4PZs2cP2dnZHDt2jJSUFAIDA93z79pn6dKlVK5cmfXr17vnsuD8r1u3jujoaK+5rFixIhs2bCAwMNA9lwXnv+BcuuY/ICCA1NRUjhw5wsmTJ9m1a5d7/i+55JJC5//gwYNkZmaSk5PjM//Lli1zv0ZERPjMf15eHtu3b6dcuXKsWrXKfb2qVavGihUripz/iIgI97wsXbqUatWqsWrVKsLDw/3O/5YtWwqd/7CwMHbu3MmpU6d85r9ixYpe92WVKlVYs2YNISEh7Nq1i+zsbI4ePcr+/fu95t9zH8/5d93Lx44dY9++fZQtW5YNGzb4vZc959/zXvac/+XLl7uvW8WKFdm4cWOh9/KaNWu85n/p0qV+5991L7vmv+Bcuu5lwGv+/d3Lrs+Oa/7z8/NJT0/3mv+C97Jr/j3vZc/537ZtG5GRkV5zWbVqVff8e97LBw8e9Jr/lStXuu+x6OjoQu/lgIAAn/l37eOa/927dxd6L+u7XN/l+i7Xd7m+y/Vdru9yfZfru1zf5fou13e5vsv1Xf5n/C6/5JJLhp9DGrZYGGttSR37rBljhgCjgI1AS2ttxhnsEwgsBJoBQ621o4rqP27cOAswcODA8w9YREREREREREREpGSZkjpwqX3gYEHGmH/gSDyvB9qcSeIZwFqbC3zgfPuXCxSeiIiIiIiIiIiIiHj4XSSfjTFDgTHAWhyJ5/1nOcRB52t4sQYmIiIiIiIiIiIiIn6V+uSzMeYp4C1gFY7E84FzGKa583V7sQUmIiIiIiIiIiIiIoUq1clnY8zzwAhgOdDOWptWRN9mxpggP+1tgX863352QQIVERERERERERERES+BJR1AYYwx/YEXgTxgATDEGJ/a2MnW2o+dv48EGhpj5gF7nG3XAm2dvz9vrf3lQsYsIiIiIiIiIiIiIg6lNvkM1HW+lgGGFtLnZ+Bj5++fAj2AG4BOQFkgFfgKGGOtXXDBIhURERERERERERERL6U2+WytHQYMO4v+HwIfXqh4REREREREREREROTMleqazyJ/NnFxcRhjSE5OLulQSr158+ZhjGHYsGElHcoZi42NxU/5IBERERERERGRPyQln+W8JCcnY4whLi6upEPxkZOTw5QpU7j//vu5+uqriYiIICwsjGuuuYYXXniBo0ePFrrvnj17GDBgANWrVyc4OJiYmBiGDh3KoUOHLuIZiBSP0nyfioiIiIiIiMgfV6ktuyFyvrZt20bPnj0JDw+nTZs23HbbbRw7doyEhAReeuklvvzySxYtWkTlypV99mvRogUHDhygW7duNGjQgN9++41Ro0Yxa9YsFi1aRKVKlUrorOT37JNPPiErK6ukwxARERERERERuSiUfJY/rPLly/Puu+/Sv39/wsPD3e2nTp2iZ8+efP/99wwfPpx33nnHa79BgwZx4MABRo8ezeDBg93tjz76KG+99Rb/+te/GD9+/EU7D/njqF27dkmHICIiIiIiIiJy0ajshpyzYcOGUbduXQAmTpyIMcb98/HHH7v75efnM378eG644QbKlStHeHg4N9xwA+PGjSM/P99nXGMMsbGx7Nu3j7/+9a9ER0cTGhpK06ZN+fzzz884vho1ajBo0CCvxDNAUFAQzz77LOCoG+xp+/btzJ49m5iYGP7xj394bRs+fDjh4eF8+umnHD9+/IzjWL58OXfccQfR0dEEBwdTp04dBg0aREpKSqH75Ofn8+abb9KgQQNCQkKoWbMm//znPzly5IhP36SkJO6++25iYmIIDg6mSpUqNGnShKFDh5KTk+PVNzc3l7Fjx9K8eXN3GZLrrruOMWPG+MyFZ6mGzZs307t3b6KjowkICGDevHk0aNCAoKAg0tLS/J7DiBEjMMbw7rvverXv2bOHhx9+mEsvvZTg4GAqVapE165dWbp0qd9xUlNTuf/++6latSqhoaE0btyYiRMnFnrt/Hn11VcxxjB69Gi/2/ft20eZMmW44YYbvNpefPFFWrZsSbVq1QgKCqJ69er07duXDRs2+IxxuusF/ms+nzp1ijFjxtC5c2fq1KlDcHAwFStWpH379sycOdNvvDExMcTExJCVlcUTTzxB7dq1CQ4O5vLLL2fkyJFYa919z/Q+FREREREREREpblr5LOcsNjaWzMxMRo0aRaNGjejevbt7W+PGjd2///Wvf+Xzzz+nVq1aPPDAAxhjmDJlCoMGDWLhwoVMmjTJZ+xDhw7RokULoqKiuO+++8jMzOSrr76iX79+7N27lyeeeOK8Yi9btiwAgYHet8DcuXMB6NixIwEB3n+bKV++PC1btmT27NksWbKEdu3anfY4M2bM4I477sBaS69evahTpw7Lly9n3LhxTJs2jUWLFhETE+Oz3z//+U/mz5/PXXfdRbdu3UhISODtt99mwYIFLFy4kJCQEMCReG7WrBnGGLp27UrdunU5cuQIW7duZezYsbz88svuc83JyaFLly4kJCRQv359+vbtS0hICImJiQwePJhff/2VTz/91CeWbdu20axZM+rVq0e/fv04ceIEERER9O/fn2effZYvvvjCa4W4yyeffEJQUBB9+vRxt61YsYKOHTuSkZHBLbfcQs+ePUlLS2Pq1Km0atWKKVOm0LlzZ3f/9PR0WrRowfbt22nVqhWtWrUiJSWFv//973Ts2PG019/l3nvv5bnnnmPixIkMGTLEZ/tnn31Gfn4+/fv3d7fNnz+fESNG0KZNG+644w7KlSvHli1bmDx5MtOnT2fRokU0atTojK9XYTIyMnjkkUdo0aIFHTp0oEqVKqSkpPDdd9/RuXNn3n//fR544AGf/XJycujYsSP79u2jU6dOBAYGMnXqVJ5++mmys7OJj48Hzvw+FREREREREREpdtZa/Xj8jB071o4dO9bKmdmxY4cFbP/+/f1u//zzzy1gr7vuOnv06FF3+7Fjx2zTpk0tYCdNmuS1D2ABe+edd9q8vDx3+/bt222FChVs2bJl7bZt284r7r///e8WsE8//bRX++OPP24B+/rrr/vd7x//+IcFzugzcvToUVupUiUbEBBg58+f77VtxIgRFrAdOnTwau/fv78FbKVKlWxycrK7PS8vz/bs2dMC9sUXX3S3P/rooxawU6dO9Tl+RkaG1/WLj4+3gH344Ydtbm6uuz03N9cOGDDAZxzX3AL2mWee8Rl/z549NiAgwDZt2tRn22+//WYB27NnT3dbTk6Oveyyy2xwcLCdN2+eV/+9e/fa6tWr22rVqtns7Gx3+4MPPmgBO3ToUK/+S5cutYGBgRaw8fHxPsf3p2PHjhawa9as8dl21VVX2bJly9q0tDR3W2pqqj1y5IhP31WrVtnw8HB76623erWf7npZa23r1q2t42v3/2VnZ9vdu3f79M3MzLQNGza0FSpUsFlZWV7b6tSpYwHbqVMnr22pqak2MjLSRkZG2lOnTvnEVth9KiIiIiIiIiJ/aCWWa9XK5wul0Z0lHcGZW/31BRt6woQJgKMEQ7ly5dzt4eHhjBw5kvbt2/PBBx/Qt29fr/3KlCnDyJEjvVYf161blyFDhjB8+HA+/fRT98rOszV9+nTee+89atasyZNPPum17fDhwwBERkb63dfVnpmZedrjTJs2jfT0dO6++25uvvlmr22PPfYY48ePZ86cOezatcunFvAjjzxCnTp13O8DAgJ47bXXmDp1KhMmTOD555/36h8aGupz/AoVKrh/z8/PZ8yYMVSrVo233nqLMmXKuLeVKVOGN954g48++ohJkybRrVs3r3GqVq3q91rXqFGDdu3aMWfOHNatW0fDhg3d21xlMTxXEn///fds27aNxx9/nNatW3uNVb16dZ588kmGDh3KTz/9ROfOncnJyWHSpEmUL1+eYcOGefW//vrr6dev31mV3+jfvz+zZ89m4sSJvPbaa+72ZcuWsX79enr06OH1IMno6Gi/4zRq1Ii2bdsye/ZscnJy3CvLXQq7XoUJDg6mZs2aPu2RkZEMGDCAxx57jKVLl/KXv/zFp8/o0aO95j46Oppu3brxySefsGnTJq6++uozjkNEREREREREpLgp+SwX1IoVKwgICCA2NtZnW+vWrSlTpgwrV6702Va7dm13nVpPsbGxDB8+3O8+Z+KXX36hb9++hIeH880333glaM+EddbSLVi3158VK1YA0LZtW59tgYGB/OUvfyE5OZmVK1f6JJ8LJmcBLr30UmrVqkVycjKZmZlERUXRu3dvRo0aRffu3enVqxft27enZcuWXHbZZV77bt68mfT0dK644gpefvllv/GGhob6rWXcqFEjgoOD/e4TFxfHnDlzmDhxIv/5z38ARw3j//3vf1SpUsWrhMbixYsB2Llzp08yGWDLli0AbNiwgc6dO7Nx40aysrK4+eab/f4xIDY29qySzz169CAyMpLPPvuMESNGuBPwrjHi4uJ89vn+++8ZP348y5YtIy0tjdzcXK/taWlpXHLJJV5tRV2vwqxbt47XXnuN+fPnk5KSQnZ2ttf2vXv3+uwTGRnJ5Zdf7tNeq1YtwFG6RkRERERERESkJCn5LBfU4cOHqVixIkFBQT7bAgMDqVy5MgcOHPDZVrVqVb/jVatWzT3u2Vq8eDGdOnUiICCAmTNncuONN/r0cSU5Cxvf9cC/wlZGe3KNUTA56eJq97eKuqjz37lzJ4cPHyYqKoobb7yRBQsW8MorrzB58mR3zeb69esTHx/P3XffDThqJ4MjwTt8+PBCYz527JjfYxamR48eRERE8Nlnn/Hqq69SpkwZZsyYQXp6OkOHDvWqqe2K4euvi15p74rBdf1O91k4U6Ghodx11128//77zJ49m06dOpGTk8MXX3xBlSpV6NSpk1f/0aNH88gjj1ChQgU6dOhA7dq1CQsLwxjD1KlTWb16NSdPnjzvuJYsWULbtm3Jzc2lXbt2dO3alYiICAICAli1ahXTpk3ze5yoqCi/47mueV5e3lnFISIiIiIiIiJS3JR8vlAuYCmL35PIyEgyMjL8lifIzc0lLS3N78PYUlNT/Y63f/9+97hnY8GCBdx2220EBASQkJBA8+bN/farX78+4Fgp7I9rdW69evVOe0xXjK6YC0pJSfHq5yk1NdUdiyd/53/TTTcxY8YMTp48yfLly5k1axbvvPMOffv2pUqVKrRv397dv0ePHnz77benjd1TUau8XQndDz74gDlz5nDrrbf6LbnhGfO0adPo2rXraY/r6n+6z8LZ6N+/P++//z4TJ06kU6dO7kT5I4884vX5zM3NJT4+nmrVqrFixQqfPyC4VnH7cyar4j29/PLLnDhxgsTERJ//Q+DVV19l2rRpZzWeiIiIiIiIiEhpEXD6LiKFc5UuKGyV5XXXXUd+fj7z58/32TZ//nzy8vJo0qSJz7Zdu3aRnJzs0z5v3jz3uGdq7ty5dOrUicDAQObMmVNo4hmgTZs2AMyePZv8/HyvbUePHmXRokWEhoYWOYaLK0ZXzJ5yc3NZuHAhgN/z//nnn33atm/fzu7du4mJifG76jU4OJgWLVrw4osvMnr0aAB34rJBgwZERUWxZMkScnJyThv72XCVq5g4cSJpaWnMnDmTa6+9lsaNG3v1c12zBQsWnNG4DRo0ICwsjFWrVvldie7vup5Oy5YtueKKK5g2bRqHDx8uNFGelpZGZmYmLVq08Ek8Hzt2zF1SpThs3bqVihUr+i1N4+9zcC5Od5+KiIiIiIiIiFwISj7LealQoQLGGHbt2uV3+4ABAwB45plnyMrKcrdnZWXx9NNPA3D//ff77JeXl8dTTz3llQDesWMHo0ePJjAwkHvuueeM4ps9eza33347ISEh/PTTT9xwww1F9r/sssvo2LEjycnJvPvuu17b4uPjOX78OPfeey/h4eGnPXb37t2pWLEiX3zxBUuWLPHa9vbbb7N9+3bat2/vU+8ZYNSoUezcudP9Pj8/nyeeeIL8/Hzuu+8+d/uCBQv8JmZdq4XDwsIARymGwYMHk5KSwpAhQzhx4oTPPikpKaxfv/6051WQZ0J33Lhx5OTk+K2f3K1bNy677DLeffddfvjhB79jLV682P05KVu2LP369ePo0aM+NaKXLVvGpEmTzjpWcCSas7OzGTt2LD/88APXXnutzx8zoqOjCQsLY/ny5V6lSHJycnjkkUdIS0s7p2P7ExMTQ0ZGBklJSV7tH374IQkJCcVyjNPdpyIiIiIiIiIiF4LKbsh5KVeuHM2aNWPBggX069ePevXqUaZMGbp27cq1115L3759mTZtGl999RUNGzake/fu7pq5O3bs4K677qJfv34+41577bX8+uuvNG3alI4dO3L48GG+/PJLMjMz+c9//uPzQD1/Nm3aRLdu3cjOzqZz585MmzbNbwmDgonNsWPH0qJFC4YMGcJPP/3ElVdeya+//kpiYiL16tXjlVdeOeNrM2HCBO68805at27NnXfeSe3atVm+fDmzZ8+mWrVqvPfee373bdmyJY0bN6Z3795ERkaSkJDA6tWradq0KU8++aS73xtvvMHs2bOJjY3l0ksvpVy5cqxbt46ZM2dSoUIFHnroIXff559/ntWrVzN+/Hi+++472rZtS40aNThw4ABbtmxh0aJFvPLKK1x11VVndH6e7r33Xp5//nleeuklAgMD6du3r0+fsmXL8u2333LLLbdw22230aJFCxo3bkxYWBi7d+9m6dKlbN++nZSUFHfS/N///jc//fQTb7/9NsuWLaNVq1akpKTw5Zdf0rlzZ6ZPn35Osb7wwgvEx8eTk5Pjs+oZICAggCFDhjBixAiuueYaunXrxqlTp0hMTCQjI4M2bdqQmJh41sf2Z+jQoSQkJNCqVSvuuusuIiMjWbZsGQsXLqRXr15Mnjz5vI9xuvtUREREREREROSCsNbqx+Nn7NixduzYsVbO3JYtW+ztt99uK1asaI0xFrAfffSRe3teXp599913bdOmTW1oaKgNDQ21TZo0sWPGjLF5eXk+4wG2devWdu/evbZfv362SpUqNjg42F533XV20qRJZxxXYmKiBU7748+uXbtsXFycrVatmi1btqytXbu2HTJkiE1PTz/r6/Pbb7/Z7t2728qVK9uyZcvaWrVq2b///e927969Pn379+9vAbtt2zb7+uuv2/r169vg4GBbvXp1+8gjj9jDhw979U9ISLBxcXH2yiuvtBERETYsLMzWq1fPDh482CYnJ/uMn5+fbz/55BPbtm1bW6FCBVu2bFlbvXp127JlS/vKK6/YXbt2ufvu2LHDArZ///6nPcedO3fagIAAC9jbb7+9yL6pqan2qaeesg0bNrShoaE2PDzcXn755faOO+6wn376qc3JyfHqn5KSYu+77z5buXJlGxISYhs1amQ/+ugj9/zGx8efNr6C2rVrZwEbGBho9+/f77dPTk6OfeONN+yVV15pQ0JCbNWqVe0999xjk5OT3fO0Y8cOd/8zuV6tW7f2+5n77rvvbLNmzWy5cuVsZGSk7dChg/3555/tRx995HM/WWttnTp1bJ06dfweIz4+3gI2MTHRq/1096mIiIiIiIiI/GGVWK7VWGsvZq671Bs3bpwFGDhwYEmH8qdljKF169bnVNNXREREREREREREvJiSOrBqPouIiIiIiIiIiIhIsVPyWURERERERERERESKnZLPIiIiIiIiIiIiIlLsAks6AJGCVIdcRERERERERETk908rn0VERERERERERESk2Cn5LCIiIiIiIiIiIiLFTslnERERERERERERESl2Sj6LiIiIiIiIiIiISLFT8llEREREREREREREip2SzyIiIiIiIiIi8n/s3Xt4nNV96PvvkmTLlm3Z1t0ydoHsEFLRmLQcCGkLBCe0KQdyaepk95akKezQ7JK2T9jt3gWeEdDT9nCaNEkTGsitadkh3m1CYLe0SQ2BtGmwfJUty/huSRaWbfki32Rb0jp/aEZ7JOvikV5JlvT9PM88M+/7rjXzwz+/y+I3S2tJUuIsPkuSJEmSJEmSEmfxWZIkSZIkSZKUOIvPkiRJkiRJkqTEWXyWJEmSJEmSJCXO4rMkSZIkSZIkKXEWnyVJkiRJkiRJibP4LEmSJEmSJElKnMVnSZIkSZIkSVLiLD5LkiRJkiRJkhJn8VmSJEmSJEmSlDiLz5IkSZIkSZKkxFl8liRJkiRJkiQlzuKzJEmSJEmSJClxFp8lSZIkSZIkSYmz+CxJkiRJkiRJSpzFZ0mSJEmSJElS4iw+S5IkSZIkSZISZ/FZkiRJkiRJkpQ4i8+SJEmSJEmSpMRZfJYkSZIkSZIkJc7isyRJkiRJkiQpcQWj6RRCmAcsBcqAs8ChGOOBJAOTJEmSJEmSJE1dl1x8DiG8Afgt4F3AWxkwazqE0A68BPwD8O0YY1eCcUqSJEmSJEmSppARi88hhBuAx4B30ltwvgBsAQ4CR4G5QCnwJuBXgA8Ah0MInwU+HWM8Nz6hS5IkSZIkSZIuV8MWn0MIfwP8GnACeBJ4BlgbY+wcov2VwC8AHwb+BJpPTLAAACAASURBVPh4COE3YoyvJBizJEmSJEmSJOkyN9KGg3cAfwBUxxh/J8b4ylCFZ4AY474Y45dijG8Hrgc2ALcmF64kSZIkSZIkaSoYadmNq2OMZ0fzxjHGeuB9IYS5o+kfQigF3gfcCfwUvRscnqd3yY+vAV+LMfYM0u/twIPA24A5wC7gq8DnY4zdo4lFkiRJkiRJkpSbYYvPoy08J/QevwI8AbxO70aGTUAl8H7gy8C7Qwi/EmOMmQ4hhPfQu+FhJ/Atetekvgv4DPCz6feUJF2G7nrsuWGvP//g3RMUiSRJkiRJSsKIGw5eihDCLOA64EyM8bUk3hPYAdwN/GP2DOcQwv8A1gK/TG8h+h/S54uBp4Bu4LYY47r0+YeAF4EPhBA+FGN8JqH4JEnj5ED76b7XS0vnTWIkkiRJkiRptHIqPocQVgEfAD4eYzyaPvcG4AXgDenj7wKrYoxdYwksxvjiEOcPhhD+mt4NDW8jXXxOx1UOfCNTeE637wwhPAisAe6jd9NESdJlJntmc/YsaGc8S5IkSZI0NY204eBAvwVcmyk8p/0F8J/oXRqjHngP8NFkwhvShfRzdoH79vTzPw/S/hXgDPD2EELheAYmSZIkSZIkScq9+PyTQF3mIL3UxS8Bq2OM7wRuBLYzjsXnEEIB8Jvpw+xC85vSzzsG9knPwt5L70zvq8crNkmSJEmSJElSr1yLz+X0bgCYcTO9Bd1nAGKMF4Dvk16CY5z8Gb3rS/9TjPFfss4vTD+fGKJf5vyi8QpMkiRJkiRJktQrP5VKXXLj2traTwK7UqnUC+njjwE/C9yfSqXOpM/dBtySSqX+JOlgQwj3A4/QO7v6/06lUmezYvtdoBT4fCqVOjqwb21t7W8Dy4Avp1KplkHe+97a2tov3XnnndUA11xzDY2NjZSVlbFx40aqq6tZt24dpaWlNDQ0UFhYyIEDBzhz5gynT5+mtbWV2bNns23bNsrLy9mwYUO/Po2NjcyaNYvW1lZOnz7N2bNnaWlpobCwkK1bt1JZWdmvz+LFi3nttdfIz8/n4MGDnDx5ks7OTpqbm5k7dy719fVUVVWxfv36vj6LFi1ix44dhBA4dOgQHR0dnD9/nv3791NUVER9fT3V1dXU1dWxdOlS1q1bx8KFC9m1axcxRo4cOcKxY8fo6upi3759zJ8/n82bN/e9f+bziouL2b17Nz09PRw9epT29nZ6enrYs2cPCxYs6NensrKSjRs3smDBAvbu3UtXVxfHjh3jyJEjAOzatYtFixaxadOmvj4VFRVs3ryZefPmsW/fPi5cuMCJEyc4dOgQIQR27NjB4sWL+/UpLy+nvr6eoqIimpqaOHfuHB0dHbS1tZGfn8/27dspLS1lw4YNff/tZWVlbN26lcLCQlpaWujs7OTUqVO8/vrrFBQU9OU/06euro6ysjK2bdvWl8uB+W9oaKCioqJfLktKSmhsbKSgoKAvlwPzPzCXmfzn5eXR1tZGR0cH586do6mpqS//S5YsGTL/hw8f5vjx41y4cOGi/K9bt67vubi4+KL8d3d3s2fPHubPn8+mTZv6/ryqqqrYsGHDsPkvLi7uy0tdXR1VVVVs2rSJefPmDZr/nTt3Dpn/oqIi9u/fz/nz5y/Kf0lJSb/7sry8nC1btjBnzhyampro7Ozk5MmTHDx4sF/+s/tk5z9zL586dYrW1lZmzZpFY2PjoPdydv6z7+Xs/K9fv77vz62kpITt27cPeS9v2bKlX/7r6uoGzX/mXs7kf2AuM/cy0C//g93Lmb87mfz39PTQ3t7eL/8D7+VM/rPv5ez87969m4ULF/bLZWVlZV/+s+/lw4cP98v/xo0b++6xiooK6uvr+V7jUTrPnqWnJ7LyTYtoa2sjLy/vovxn+mTy39zcPOS97FjuWO5Y7ljuWD6xY/lg97JjuWO5Y7ljuWO5Y7ljuWO5Y/nEj+VLliypzakIm6AQY7z0xiH8O1AJ1AARaABOxRjfmtXmm8DNMcYrEw00hE8AfwVsA1bGGA8OuF4H3ADcEGNcP0j/rem4fzLG2DjU5zzxxBMR4L777ksweklSLtxwUJIkSZKkxITJ+uBcl914kt41k3cCjenXXx3Q5iZ6i9KJCSH8Hr2F563AOwYWntNeSz9fM0j/AuAqejco3JNkbJIkSZIkSZKki+VUfI4x/g29ay4X0bvG8l+lHwCEEG4HrgReSirAEMIfAp8BNtFbeD40RNMX08+/OMi1W+iN+UcxxnNJxSZJkiRJkiRJGlyuM5+JMf6PGGNZ+vHJ2H/djn8DFgN/mURwIYSH6C12r6d3qY0jwzT/e+AI8KEQwg1Z7zEHeCx9+EQScUmSJEmSJEmShleQ5JvFGM8D55N4rxDCh+ndXLAb+CFwfwgXLU+yL8b49fRnd4QQ7qG3CP2DEMIzwFHgbuBN6fPfSiI2SZIkSZIkSdLwhi0+hxDmxhjPjuUDxvAeV6Wf84HfG6LNy8DXMwcxxmdDCLcCfwz8MjAH2AX8AfC5mMvuipIkSZIkSZKkURtp2Y29IYRPhhAKc33jEMKKEMJ3gU+NJrAYYyrGGEZ43DZIv3+PMf5SjHFxjHFujPGnYoyfiTF2jyYOSZIkSZIkSVLuRio+fw/4NPB6COGJEMI7Qghzh2ocQrg6hHBfCOE/gA3AChLcfFCSJEmSJEmSNDUMu+xGjPE3QwifA/4f4N70ozuE0Ai8Dhyjd2mLUnrXVS4DAtBG79IXn4kxnhu/8CVJkiRJkiRJl6MRNxyMMa4D7gghvBH4GLASuB74qQFNDwPfBv4B+IcY44WEY5UkSZIkSZIkTREjFp8zYow7gT8CCCEUAUvpnfF8FjgUY3x9XCKUJEmSJEmSJE05l1x8zhZjPAPsTD8kSZIkSZIkSepnpA0HJUmSJEmSJEnKmcVnSZIkSZIkSVLiLD5LkiRJkiRJkhJn8VmSJEmSJEmSlDiLz5IkSZIkSZKkxFl8liRJkiRJkiQlLqficwhheQiheIQ2C0IIy8cWliRJkiRJkiRpKst15vNe4JMjtLk/3U6SJEmSJEmSNEPlWnwO6YckSZIkSZIkSUMajzWfK4HT4/C+kiRJkiRJkqQpomCkBiGE3xxw6vpBzgHkA8uB3wC2JBCbJEmSJEmSJGmKGrH4DHwdiOnXEXhP+jFQZjmOM0DtmCOTJEmSJEmSJE1Zl1J8/mj6OQBfBZ4FvjtIu26gHfiPGOPxZMKTJEmSJEmSJE1FIxafY4x/k3kdQvgw8GyM8RvjGpUkSZIkSZIkaUq7lJnPfWKM7xivQCRJkiRJkiRJ00deLo1DCD8dQng4hFA5xPWq9PXrkwlPkiRJkiRJkjQV5VR8Bj4F/DZwaIjrbcDHgD8YS1CSpJmroamdvW0d7Gw9zt62Dhqa2ic7JEmSJEmSNAo5LbsB3Ay8FGOMg12MMcYQwovALWOOTJI0o6zZ3Mwjq1/llYbWfuevu/9pbqmp5uFVN7FyxbJJik6SJEmSJOUq15nPVUDLCG1agSWjC0eSNBN95fsN3JF69qLCc8YrDa3ckXqWr/5rwwRHJkmSJEmSRivX4vMZoHyENuXAudGFI0maadZsbubeL75Iz+C/VNOnJ0bu+cKLrNncPEGRSZIkSZKksci1+LwJeE8IYf5gF0MIxcB70u0kSRrRI6tfHbHwnNETI4+uXjvOEUmSJEmSpCTkWnx+kt6Zzd8PIbwl+0IIYQXwPaAs3U6SpGE1NLUPudTGUF5uOOAmhJIkSZIkTQE5FZ9jjN8CvgHcBGwMIbSGEOpCCK3ABuBG4Bsxxm8mH6okabpZUz+6JTRG20+SJEmSJE2cXGc+E2P8CPBxYBu9GxD+TPq5Abg3xvjRJAOUJE1fHWfOT2g/SZIkSZI0cQpG0ynG+CTwZAihCFgEHI8xnkk0MknStFdcNHtC+0mSJEmSpIkzquJzRrrgbNFZkjQqK9+ybEL7SZIkSZKkiZPzshsAIYTyEMLHQwifDSF8ecD5G0MIc5MLUZI0XdUsL+WWmuqc+txas5Sa5aXjFJEkSZIkSUpKzsXnEMLHgH3AF4DfBbLXeK4E/gP41SSCkyRNfw+vuom8EC6pbV4IPLTqxnGOSJIkSZIkJSGn4nMI4V3Ak8AO4H3AE9nXY4xb6d148L1JBShJmt5WrljGk79z+4gF6LwQeOoTt7NyhUtuSJIkSZI0FeS65vMfAq8Dt8YYO0IIbx2kTT1w85gjkyTNGB97Vw1XVhTz6Oq1vNxw4KLrt9Ys5aFVN1p4lqaxux57btjrzz949wRFIkmSJCkpuRafbwCeiTF2DNOmBagafUiSpJlo5YplrFyxjIamdj74+At0dfdQkJ/Htx54t2s8SzPMgfbTfa+Xls6bxEgkSZIkjUWuxefZwOkR2iwCukcXjiRppqtZXspVlcX9jiVNf9kzm7NnQTvjWZIkSZq6ct1wcB/wMyO0uQl4bVTRSJIkSZIkSZKmhVyLz98Ffj6E8CuDXQwhfBR4C/APYw1MkiRJkiRJkjR15brsxv8LfAj4ZgjhA8BCgBDCfwV+Hng/sBP4fJJBSpIkSZIkSZKmlpyKzzHGYyGEW4FvANmznz+Xfv4h8KsxxpHWhZYkSZIkSZIkTWO5znwmxtgE3BZCeAtwM1AKnAB+HGNcn3B8kiRJkiRJkqQpaNjicwjh28AzMcbV6eNbgH0xxqYYYz1QPwExSpIkSZIkSZKmmJE2HHwvcG3W8UvAR8YtGkmSJEmSJEnStDBS8fk4UJx1HMYxFkmSJEmSJEnSNDHSms+NwH8OIdQBr6fPXZlefmNYMcZXxhqcJEmSJEmSJGlqGqn4XAs8C/zPrHMfTj9Gkj/aoCRJkiRJkiRJU9uwxecY4/dCCG8G3gksBVLAy+mHJEmSlKiGpnb2tnXQ1d1DQX4eDU3t1CwvneywJEmSJI3CSDOfiTHuB74CEEJIAT+IMT4yznFJkiRpBlmzuZlHVr/KKw2t/c5fd//T3FJTzcOrbmLlimWTFJ0kSZKk0Rh2w8EQwqdDCHdknfoovctwSJIkSYn4yvcbuCP17EWF54xXGlq5I/UsX/3XhgmOTJIkSdJYDFt8Bn4PeFvW8VeB945fOJIkSZpJ1mxu5t4vvkhPjMO264mRe77wIms2N09QZJIkSZLGaqTi8ymgKOs4jGMskiRJmmEeWf3qiIXnjJ4YeXT12nGOSJIkSVJSRlrzeRfw/hDCd4DX0+cWhRCWj/TGMcamsQYnSZKk6auhqX3IpTaG8nLDATchlCRJkqaIkWY+Pw68AfgRsDd97pPp18M99iQRXAjhAyGEz4cQfhhC6AghxBDC3w3R9sr09aEezyQRkyRJkpKxpn50S2iMtp8kSZKkiTXszOcY4zdDCHuBO4GlwEeAemDT+IcGwIPACnqX/2gBrr2EPpsZfFPErQnGJUmSpDHqOHN+QvtJkiRJmlgjLbtBjPHHwI8BQggfAb4TY3xknOPK+H16i867gFuBly6hz6YYY2o8g5IkSdLYFRfNntB+kiRJkibWiMXnAT4KbByPQAYTY+wrNofgXoeSJEnTycq3LJvQfpIkSZImVk7F5xjj34xXIAmqDiH8F6AUaAf+I8ZYP8kxSZIkaYCa5aXcUlOd06aDt9YsdbNBSZIkaYoYacPBi4QQ8kIIvxtC+HEI4UQIoSvr2ltDCF8MIVyTbJg5eRfw18CfpJ83hxBeCiEsn8SYJEmSNIiHV91E3iX+hlteCDy06sZxjkiSJElSUvJTqdQlNw4hzAb+Bfg4MBc4AyxIpVKPANTW1nYCTwMdqVTqB0kGWltbeyW9Gx5uSaVS3x7k+mygE/gD4I+APwdeAK4E3gHcVVtb+9VUKnVhsPcPIdxbW1v7pTvvvLMa4JprrqGxsZGysjI2btxIdXU169ato7S0lIaGBgoLCzlw4ABnzpzh9OnTtLa2Mnv2bLZt20Z5eTkbNmzo16exsZFZs2bR2trK6dOnOXv2LC0tLRQWFrJ161YqKyv79Vm8eDGvvfYa+fn5HDx4kJMnT9LZ2UlzczNz586lvr6eqqoq1q9f39dn0aJF7NixgxAChw4doqOjg/Pnz7N//36Kioqor6+nurqauro6li5dyrp161i4cCG7du0ixsiRI0c4duwYXV1d7Nu3j/nz57N58+a+9898XnFxMbt376anp4ejR4/S3t5OT08Pe/bsYcGCBf36VFZWsnHjRhYsWMDevXvp6uri2LFjHDlyBIBdu3axaNEiNm3a1NenoqKCzZs3M2/ePPbt28eFCxc4ceIEhw4dIoTAjh07WLx4cb8+5eXl1NfXU1RURFNTE+fOnaOjo4O2tjby8/PZvn07paWlbNiwoe+/vaysjK1bt1JYWEhLSwudnZ2cOnWK119/nYKCgr78Z/rU1dVRVlbGtm3b+nI5MP8NDQ1UVFT0y2VJSQmNjY0UFBT05XJg/gfmMpP/vLw82tra6Ojo4Ny5czQ1NfXlf8mSJUPm//Dhwxw/fpwLFy5clP9169b1PRcXF1+U/+7ubvbs2cP8+fPZtGlT359XVVUVGzZsGDb/xcXFfXmpq6ujqqqKTZs2MW/evEHzv3PnziHzX1RUxP79+zl//vxF+S8pKel3X5aXl7NlyxbmzJlDU1MTnZ2dnDx5koMHD/bLf3af7Pxn7uVTp07R2trKrFmzaGxsHPRezs5/9r2cnf/169f3/bmVlJSwffv2Ie/lLVu29Mt/XV3doPnP3MuZ/A/MZeZeBvrlf7B7OfN3J5P/np4e2tvb++V/4L2cyX/2vZyd/927d7Nw4cJ+uaysrOzLf/a9fPjw4X7537hxY989VlFRQX19Pd9rPErn2bP09ERWvmkRbW1t5OXlXZT/TJ9M/pubm4e8lx3LHcsdyy/PsfznfvrNdJ08wr/tOk4c5mfBvAB//fHbuLKww7F8iozlg93LjuWO5Y7l03Ms9+dyx3LHcsdyx/LLeyxfsmRJ7TA/ao+rEONwP+YPaBzCHwOPAingMeBh4KEYY35Wm+8BxTHGtyUaaAi30bvh4NMxxl/PoV8B8G/ATcDvxRg/O1z7J554IgLcd999ow9WkjQmdz32XN/r5x+8exIjkTRR1mxu5tHVa3m54cBF126tWcpDq25k5QrXepYkSZJGYdI208t1w8FfA/49xvgIQAhhsMr1XuCusQaWlBhjVwjhy/QWn28Bhi0+S5IkaeKtXLGMlSuW0dDUzgcff4Gu7h4K8vP41gPvdo1nSZIkaYrKtfh8FfCPI7Q5CpSMLpxxczj9PG9So5AkSdKwapaXclVlcb9jSZIkSVNTrhsOngUWjdBmOXB8dOGMm8wSIHsmNQpJkiRJkiRJmiFyLT5vAu5Ibzx4kRDCQuAXgLVjDSxXIYSbBosrhHA78Pvpw7+b2KgkSZIkSZIkaWbKddmNp4CngadDCB/LvhBCWAR8DVgM/HUSwYUQ3gu8N31YlX6+OYTw9fTrIzHGT6Vf/zlQE0L4AdCSPvcW4Pb064dijD9KIi5JkiRJkiRJ0vByKj7HGL8ZQngn8FHgbuAYQAhhHVADFAJfiDH+U0LxXQ98eMC5q9MPgP1Apvj8t8D7gP8LeDcwC2gDVgN/FWP8YUIxSZIkSZIkSZJGkOvMZ2KMHwsh/BD4JL0ziwPw00AD8OkY49eSCi7GmAJSl9j2K8BXkvpsSZIkSZIkSdLo5Vx8Bogxfh34eghhLr3LbJyIMZ5OMjBJkiRJkiRJ0tQ1quJzRozxLHA2oVgkSbrs3PXYc8Nef/7BuycoEkmSJEmSppYxFZ8lSZpJDrT/n1/yWVo6bxIjkSRJkiTp8mfxWZKkYWTPbM6eBe2MZ0mSJEmShpc32QFIkiRJkiRJkqYfi8+SJEmSJEmSpMRZfJYkSZIkSZIkJc7isyRJkiRJkiQpcYlsOBhCKAN+HjgD/GuMsTuJ95UkSZIkSZIkTU05FZ9DCPcBHwHeHWM8mj73M8A/AyXpZutCCLfHGE8nGagkSZKmr7see67v9dqdbRedf/7Buyc8JkmSJEljk+vM5w8CMVN4TnscWAx8DagE7gQ+DvxFIhFKkiRpRllaMn+yQ5AkSZKUgFyLz28E/jFzkF5u41bgyzHG/5I+9yrwq1h8liRJ0iVyZrMkSZI0/eS64WApcCjr+GfTz9/JOvdD4CfGEpQkSZIkSZIkaWrLtfh8FCjLOr4V6AF+lHUuAnPGGJckSZIkSZIkaQrLtfjcCNwVQigNISyidw3ouhhjR1abK4GDCcUnSdJloaGpnb1tHexsPc7etg4amtonOyRJkiRJki5rua75/FngWaAF6AKKgD/MXAwh5AM/R/+Z0JIkTVlrNjfzyOpXeaWhtd/56+5/mltqqnl41U2sXLFskqKTJEmSJOnyldPM5xjjc8DHgQbgNeBTMca/y2ryTnqX3PiXxCKUJGmSfOX7DdyRevaiwnPGKw2t3JF6lq/+a8MERyZJkiRJ0uUv15nPxBifBJ4c4tq/AIvHGpQkaea567Hn+l6v3dl20fnnH7x7QuNZs7mZe7/4Ij0xDtuuJ0bu+cKL/ER5sTOgJUmSJEnKkuuaz5IkjbulJfP7HpPlkdWvjlh4zuiJkUdXrx3niCRJkiRJmlpymvkcQngrcDPwdIzxRPrcPOCLwHuAM8Cfxxg/m3SgkqTpbaJnNg+noal9yKU2hvJywwEamtqpWV46TlFJkiRJkjS15Drz+Q+BP84UntP+FPiN9HuVAp8OIdyRUHySJE24NfXNE9pPkiRJkqTpKNc1n28AfpA5CCHMAj4MrAVuA0qAjcD9wPcSiVCSpAnWceb8hPaTJEmXr8e3dw17/YFrc95KSZKkGSPXmc8VQPa0rhuABcCXYoydMcZW4LvAWxKKT5KkCVdcNHtC+0mSpKnh2PnY95AkSSPL9SvaOKDPz6XPvZx17jBQPsa4JEmaNCvfsmxC+0mSpMtX9szm7FnQzniWJGlkuc58bgLelnX8HqAlxrgn61w1cGysgUmSNFlqlpdyS011Tn1urVnqZoOSJEmSJGXJtfi8Gnh7COHvQwh/B9wM/P2ANtcBu5MITpKkyfLwqpvIC+GS2uaFwEOrbhzniCRJkiRJmlpyLT5/BvgP4P3ArwKbgUcyF0MIPwn8DP2X4ZAkacpZuWIZT/7O7SMWoPNC4KlP3M7KFS65IUmSJElStpwWqYoxngJ+NoRwXfrUthhjT1aTM8D7gHUJxSdJ0qT52LtquLKimEdXr+XlhgMXXb+1ZikPrbrRwrMkSZIkSYMY1Q4JMcatQ5zfB+wbQzySJF1WVq5YxsoVy2hoaueDj79AV3cPBfl5fOuBd7vGsyRJkiRJwxhV8TmEMAtYCbwZmB9jfDR9fg5QDBwZMCNakqQprWZ5KVdVFvc7liRJkiRJQ8t1zWdCCL9I7+zmfwT+AkhlXb4eeB34YAKxSZIkSZIkSZKmqJyKzyGEG4BngQj8PvA/s6/HGH8M7KV33WdJkiRJkiRJ0gyV68znh+jdVPCGGOPngJ2DtKkDVow1MEmSJEmSJEnS1JVr8flngWdjjAeHadMMLBl9SJIkSZIkSZKkqS7XDQfnA0dGaFPEKNaSliTNXHc99tyw159/8O4JikSSNJU9vr1r2OsPXDuq/dYlSZI0Srn+9HUAqBmhzfXAntGFc/kYrBBi8UOSxt+B9tN9r5eWzpvESCRJU9mx87Hv9eLZYRIj0XTRciZy6FykJ0Je6D2+osi/W5IkDSfX4vMLwMdDCD8XY/y3gRdDCO8G3g78WRLBTSaLH5I0cbK/3Mv+8s8v/SRJucie2Zw9C9oZzxqLrSd6+HZLD9tPxn7n/1t9F9cuCLz/ijyuW+gv/0qSNJhcfwr7U+BDwPdCCJ8HrgQIIdwJ3AJ8Angd+HSCMU6K7IKzxQ9JkiRJmnleOtTDl/d0E4e4vv1k5E8bu7nnaritwgK0JEkD5VR8jjEeCCHcAawGHsi69BwQgN3A+2OMI60LLUmSJEnSZWvrieELzxkReGpPN2WFOANakqQBcv79sxjjhhDCm4A7gZuBUuAE8GPguzHG4Xf5kCRJkiTpMvftlp4RC88ZEfhOS4/FZ0mSBhjV4mcxxm56ZztfvCufJEmSJElTWMuZeNEazyNpPBndhFCSpAH8WlaSJEmSpCxbT/RMaD9JkqarUc18DiHcANwILAbyB2kSY4yPjiUwSZIkSRqNljORQ+ciPRHyAs5GVc7Odk9sP0mSpqucis8hhGLg28A76N1gcCgRmNLF571tHXR191CQn0dDUzs1y0snOyRJmvYamtodfyVJo7b1RA/fbum5aLmE/1bfxbULAu+/Is81eXVJ5g42xWoc+0mSNF3lOvP5ceB24IfA14BmYFpuMNjQfLTv9XX3P80tNdU8vOomVq5YNolRSdL0tGZzM4+sfpVXGlr7nR9u/L3rseG3HXj+wbsTj1OSdPl66VAPX97TPeQGcdtPRv60sZt7robbKixAa3i9X1LkvoSGX25IktRfrsXn9wAbgHfEGGfUYlavNLRyR+pZnvrE7fzWO2smOxxJmja+8v0G7v3ii/TEwcsFlzL+Hmg/3fd6aem8cYlTknT52npi+MJzRgSe2tNNWaFFQg3viqLAtQtCTpsOvnlBcHkXSZIGyLX4vBD425lWeM7oiZF7vvAiP1Fe7AxoSUrAms3NwxaeMwYbf7NnNmfPgnbGsyTNPN9u6Rmx8JwRge+09Fh81ojef0Uef9o48pca0Lsm5fuu8O+UJEkD5fqv406gcjwCmSp6YuTR1WsnOwxJmhYeWf3qiIXnjKHG38w60Ttbj7O3rYOGpvakw5QkXcZazsScZqcCNJ6MtJzJrY9mnusW5vHbV+cPu9kR9Bae77k63y80JEkaRK4zn78A/FkIYWmM8cB4BDQVvNxwwE2wJGmMGpraL1rjeSTZ4+9o1omWJE0/W0+M7pcyt57o4Yoi/zMXzwAAIABJREFUd4fT8N5RkUd5Ye9s+cZBvuR484LA+9zIUpKkIeVafH6B3g0H/z2EUAusB44P1jDG2DTG2C5ra+qbLT5L0hisqW8edb8fv3ZwzOtES5Kmh7PdE9tPM891C3uLyy1nIp/d2UVPhLwAn3xjgWs8S5I0glyLz/voXSYtAF8epl0cxXtPKR1nzk92CJI0pY12HN245zDfeGn7qNaJliRNP3NHOXl5tP00c11RFKgoDP2OJUnS8HItEH8DLnkvj2mtuGj2ZIcgSVPaaMfRH25rzXmdaIvPkjR99S53kPvSGy6TIEmSNP5yKj7HGD8yTnFcJITwAeBW4HpgBbAAeDrG+OvD9Hk78CDwNmAOsAv4KvD5GGOiv1i38i0WMiRpLEY7ju4+eCKn9mNdp/+ux57re712Z9tF559/8O5Rva8kKRlXFAWuXRBy2nTwzQuCs1YlSZImwOX8df+DwH+lt/g84uaGIYT3AK8AtwDfoXdzxNnAZ4Bnkgzs1pqlrvcsSWNUs7yUW2qqc+rzhqqFo/qs0a4vPdDSkvl9D0nS5eP9V+RxqaXkALzvisv5f4MkSZKmj5xmPocQuoFUjPHRYdr8MVAbYxzrms+/D7TQO3v5VuClYT6zGHgK6AZuizGuS59/CHgR+EAI4UMxxjEXofNC4KFVN471bSRJwMOrbuKO1LOXtIxGXgj8/E9W5zzzGca2Tr8zmyXp8nfdwjx++2r48p7uYdcIDMA9V+e75IYkSdIEyfWnrpB+XEq7MYkxvhRj3BnjJS3s+QGgHHgmU3hOv0cnvTOoAe4ba0x5IfDUJ2537VBJSsjKFct48nduJy8M/89GZvx969Xlo/oc1+mXpOnvHRV5/Pc35/PmBYP/m/LmBYH//uZ8bquw8CxJkjRRxjo7eTCLgc5xeN/h3J5+/udBrr0CnAHeHkIojDGeG80H3FqzlIdW3WjhWZIS9rF31XBlRTGPrl7Lyw0Xr7KUPf42NLWP6jNcp1+SZobrFuZx3cI8Ws5EPruzi54IeQE++cYC13iWJEmaBCMWn0MItww4deUg5wDygeXArwGvJRBbLt6Uft4x8EKMsSuEsBeoAa4GGi/lDWuWldDV3UNBfh7feuDdrvEsSeNo5YplfcXlDz7+wpDjb2ad6FcaWi/5vV2nX5JmniuKAhWFod+xJEmSJl5+KpUatkFtbe1e4MPAR9Knrk+/Hvj4TeC9QBHwQCqV2pJUkLW1tVemP2NLKpX69iDXHwBKgcdTqdTRQa7/GrAMeDqVSrUM9hkhhHtra2u/dOedd1YDvN6zmNmhi4qSBfxcdTfV1dWsW7eO0tJSGhoaKCws5MCBA5w5c4bTp0/T2trK7Nmz2bZtG+Xl5WzYsKFfn8bGRmbNmkVrayunT5/m7NmztLS0UFhYyNatW6msrOzXZ/Hixbz22mvk5+dz8OBBTp48SWdnJ83NzcydO5f6+nqqqqpYv359X59FixaxY8cOQggcOnSIjo4Ozp8/z/79+ykqKqK+vp7q6mrq6upYunQp69atY+HChezatYsYI0eOHOHYsWN0dXWxb98+5s+fz+bNm/veP/N5xcXF7N69m56eHo4ePUp7ezs9PT3s2bOHBQsW9OtTWVnJxo0bWbBgAXv37qWrq4tjx45x5MgRAHbt2sWiRYvYtGlTX5+Kigo2b97MvHnz2LdvHxcuXODEiRMcOnSIEAI7duxg8eLF/fqUl5dTX19PUVERTU1NnDt3jo6ODtra2sjPz2f79u2UlpayYcOGvv/2srIytm7dSmFhIS0tLXR2dnLq1Clef/11CgoKaGxspKysrK9PXV0dZWVlbNu2rS+XA/Pf0NBARUVFv1yWlJTQ2NhIQUFBXy4H5n9gLjP5z8vLo62tjY6ODs6dO0dTU1Nf/pcsWTJk/g8fPszx48e5cOHCRflft25d33NxcfFF+e/u7mbPnj3Mnz+fTZs29f15VVVVsWHDhmHzX1xc3JeXuro6qqqq2LRpE/PmzRs0/zt37hwy/0VFRezfv5/z589flP+SkhI2btzYL/9btmxhzpw5NDU10dnZycmTJzl48GC//Gf3yc5/5l4+deoUra2tzJo1i8bGxkHv5ez8Z9/L2flfv359359bSUkJ27dvH/Je3rJlS7/819XVDZr/zL2cyf/AXGbuZaBf/ge7lzN/dzL57+npob29vV/+B97Lmfxn38vZ+d+9ezcLFy7sl8vKysq+/Gffy4cPH+6X/40bN/bdYzVv/An+1ysNlC+ax5z8Hn7t5iv67uW8vDx27NjB9W+6km/+cMewa3pm5IXAH/7Ccla88Yp++Xcsdyx3LHcsdywfv7G8oqJiyHs5M5Zn5z/TJ5P/5ubmIe/lXMby+q0N9CwoY9b+TbztjUsdyx3LExnLXzt0gtB1jvz2Zn6qvMix3LHcsXycx3J/Lncs9+fyZMbyJUuW1F7C/0KPizDSksohhBQQ6V3H+WHgB8DLgzTtBtqBl2KM2xMNMoTb6N1w8OkY468Pcn0H8EbgjTHGXYNc/xFwM3BzjPHHw33WE088EQH+qX1p3zk3m5KkiXPXY8/1vR5q/P3K9xu494svDrtRYWad6N96Z03iMUqSLn+Pb+/qe/3AteOx2qBmIv9eSZKmqEn7NbAR/7WMMaYyr0MIHwaejTF+bjyDGoUT6eeFQ1wvHtBOkjSF5bJOtCRJkiRJmhw5fVUbY7xqvAIZo9eAG4BrgPXZF0IIBcBVQBewZ+JDkySNh0tdJ1qSJEmSJE2OvMkOICEvpp9/cZBrt9C7DvWPYoznJi4kSdJEqFleylWVxbyxehFXVRZbeJYkSZIk6TIx7MznEMKL9K73/OEYY0v6+FLEGOPKMUd36f4e+HPgQyGEz8cY1wGEEOYAj6XbPDGB8UiSJEmSJEnSjDbSshu3pZ+LBhyPZPhdDC9BCOG9wHvTh1Xp55tDCF9Pvz4SY/wUQIyxI4RwD71F6B+EEJ4BjgJ3A29Kn//WWGOSJEmSJEmSJF2aYYvPMca84Y7H2fXAhwecuzr9ANgPfCpzIcb4bAjhVuCPgV8G5gC7gD8APhdjHHNBXJIkSZIkSZJ0aUZaduN+4McxxrUTFE+fGGMKSOXY59+BXxqPeCRJl5e7Hnuu7/XanW0XnX/+wbsnPCZJkiRJkvR/jDST+S/J2sQvhNAdQnhofEOSJCk3S0vm9z0kSZIkSdLlYaQ1nzuBwqzjkH5IkjSpnNksSZIkSdLlbaSZz3uBXwghVGadc+1kSZIkSZIkSdKwRpr5/CV6l95oDaFvwnMqhJAaoV+MMY703pIkSZIkSWOSvRfIYPyNOUmaPMMWiGOMnwshHALuBKqBdwBNwL7xD02SJEmSJOnSHWg/3fd6aem8SYxEkgQjz3wmxvgM8AxACKEH+FqM8ZHxDkySJEmSJGkk2TObs2dBO+NZkibfSGs+D1QL/GAc4pAkSZIkSZIkTSM5rcscY6wdr0AuN2t3tvW9znxz6remkiRJ0uXr8e1dfa93n4oXnX/gWrelkSRJmkj+9DWEpSXzJzsESZIkSaNUMnuyI9B04ZcaU0dDUzt72zro6u6hID+PhqZ2apaXTnZYkjSj+a/kEDZ85j9PdgiSJEmScmARUOPNLzUuT2s2N/PI6ld5paG13/nr7n+aW2qqeXjVTaxcsWySopOkmc2fziRJkiRJGoJfalzevvL9Bu794ov0xDjo9VcaWrkj9SxPfeJ2fuudNRMcnSQp1w0HJUmSJEmSJt2azc3DFp4zemLkni+8yJrNzRMUmSQpw+KzJEmSJEmach5Z/eqIheeMnhh5dPXacY5IkjSQxWdJkiRJkjSlNDS1X7TG80hebjhAQ1P7OEUkSRqMi1dJkibdXY891/d67c62i84//+DdEx6TJEmSLl9r6ke3hMaa+mZqlpcmHI0kaSgWnyVJl5WlJfMnOwRJkiRd5jrOnJ/QfpKk0bH4LEmadM5sliRJUi6Ki2ZPaD9J0ui45rMkSZIkSZpSVr5l2YT2kySNjsVnSZIkSZI0pdQsL+WWmuqc+txas9T1niVpgll8liRJkiRJU87Dq24iL4RLapsXAg+tunGcI5IkDWTxWZIkSZIkTTkrVyzjyd+5fcQCdF4IPPWJ21m5wiU3JGmiueGgJEmSJEmakj72rhqurCjm0dVrebnhwEXXb61ZykOrbrTwLEmTxOKzJEmSJEmaslauWMbKFctoaGrng4+/QFd3DwX5eXzrgXe7xrMkTTKLz5IkSZIkacqrWV7KVZXF/Y4lSZPLNZ8lSZIkSZIkSYmz+CxJkiRJkiRJSpzFZ0mSJEmSJElS4iw+S5IkSZIkSZISZ/FZkiRJkiRJkpQ4i8+SJEmSJEmSpMRZfJYkSZIkSZIkJc7isyRJkiRJkiQpcRafJUmSJEmSJEmJs/gsSZIkSZIkSUpcwWQHIEmSJEmSNFp3PfZc3+u1O9suOv/8g3dPeEySpF4WnyVJkiRJ0rSwtGT+ZIcgScpi8VmSJEmSJE1ZzmyWpMuXaz5LkiRJkiRJkhJn8VmSJEmSJEmSlDiLz5IkSZIkSZKkxFl8liRJkiRJkiQlzuKzJEmSJEmSJClxFp8lSZIkSZIkSYmz+CxJkiRJkiRJSpzFZ0mSJEmSJElS4iw+S5IkSZIkSZISZ/FZkiRJkiRJkpQ4i8+SJEmSJEmSpMRZfJYkSZIkSZIkJc7isyRJkiRJkiQpcRafJUmSJEmSJEmJs/gsSZIkSZIkSUrctCs+hxD2hRDiEI+Dkx2fJEmSJEmSJM0EBZMdwDg5AfzlIOdPTXQgkiRJkiRJkjQTTdfi8/EYY2qyg5AkSZIkSZKkmWraLbshSZIkSZIkSZp803Xmc2EI4deB5cBpoB54JcbYPblhSZIkSZIkSdLMMF2Lz1XA3w44tzeE8NEY48uTEZAkSZIkSZIkzSTTcdmNrwEr6S1AzwN+CvgScCXwQghhxeSFJkmSJEmSJEkzQ34qlZrsGBKVSqVeTqVSe1Op1OlUKnUhlUodSqVS/7u2tnYBcAtQlUqlvjWwXwjh3tra2i/deeed1QDXXHMNjY2NlJWVsXHjRqqrq1m3bh2lpaU0NDRQWFjIgQMHOHPmDKdPn6a1tZXZs2ezbds2ysvL2bBhQ78+jY2NzJo1i9bWVk6fPs3Zs2dpaWmhsLCQrVu3UllZ2a/P4sWLee2118jPz+fgwYOcPHmSzs5OmpubmTt3LvX19VRVVbF+/fq+PosWLWLHjh2EEDh06BAdHR2cP3+e/fv3U1RURH19PdXV1dTV1bF06VLWrVvHwoUL2bVrFzFGjhw5wrFjx+jq6mLfvn3Mnz+fzZs3971/5vOKi4vZvXs3PT09HD16lPb2dnp6etizZw8LFizo16eyspKNGzeyYMEC9u7dS1dXF8eOHePIkSMA7Nq1i0WLFrFp06a+PhUVFWzevJl58+axb98+Lly4wIkTJzh06BAhBHbs2MHixYv79SkvL6e+vp6ioiKampo4d+4cHR0dtLW1kZ+fz/bt2yktLWXDhg19/+1lZWVs3bqVwsJCWlpa6Ozs5NSpU7z++usUFBT05T/Tp66ujrKyMrZt29aXy4H5b2hooKKiol8uS0pKaGxspKCgoC+XA/M/MJeZ/Ofl5dHW1kZHRwfnzp2jqampL/9LliwZMv+HDx/m+PHjXLhw4aL8r1u3ru+5uLj4ovx3d3ezZ88e5s+fz6ZNm/r+vKqqqtiwYcOw+S8uLu7LS11dHVVVVWzatIl58+YNmv+dO3cOmf+ioiL279/P+fPnL8p/SUlJv/uyvLycLVu2MGfOHJqamujs7OTkyZMcPHiwX/6z+2TnP3Mvnzp1itbWVmbNmkVjY+Og93J2/rPv5ez8r1+/vu/PraSkhO3btw95L2/ZsqVf/uvq6gbNf+ZezuR/YC4z9zLQL/+D3cuZvzuZ/Pf09NDe3t4v/wPv5Uz+s+/l7Pzv3r2bhQsX9stlZWVlX/6z7+XDhw/3y//GjRv77rGKiooh7+W8vLyL8p/pk8l/c3PzkPeyY7ljuWO5Y7ljuWO5Y7ljuWO5Y7ljuWO5Y7lj+Uwcy5csWVKbTOU1dyHGOFmfPaFCCP8J2AkcjTGWDtXuiSeeiAD33XffRIUmSZIkSZIkSeMlTNYHT8dlN4ZyKP08b1KjkCRJkiRJkqQZYCYVn29OP++Z1CgkSZIkSZIkaQaYVsXnEEJNCKFkkPM/AfxV+vDvJjYqSZIkSZIkSZp5CiY7gIT9CvBHIYSXgL3ASeANwJ3AHOCfgP9v8sKTJEmSJEmSpJlhuhWfXwLeBLyV3mU25gHHgX8D/hb42zhTdliUJEmSJEmSpEk0rYrPMcaXgZcnOw5JkiRJkiRJmumm1ZrPkiRJkiRJkqTLg8VnSZIkSZIkSVLiLD5LkiRJkiRJkhJn8VmSJEmSJEmSlDiLz5IkSZIkSZKkxFl8liRJkiRJkiQlzuKzJEmSJEmSJClxFp8lSZIkSZIkSYmz+CxJkiRJkiRJSpzFZ0mSJEmSJElS4iw+S5IkSZIkSZISZ/FZkiRJkiRJkpQ4i8+SJEmSJEmSpMRZfJYkSZIkSZIkJc7isyRJkiRJkiQpcRafJUmSJEmSJEmJs/gsSZIkSZIkSUqcxWdJkiRJkiRJUuIsPkuSJEmSJEmSEmfxWZIkSZIkSZKUOIvPkiRJkiRJkqTEWXyWJEmSJEmSJCXO4rMkSZIkSZIkKXEWnyVJkiRJkiRJibP4LEmSJEmSJElKnMVnSZIkSZIkSVLiLD5LkiRJkiRJkhJn8VmSJEmSJEmSlDiLz5IkSZIkSZKkxFl8liRJkiRJ/397dx4lR1kufvz7ZA87hIQtQCCEhEXvj0WWIBBwZ5cbUDwicAUXQOSqR0TlAh43FAQVRb1E8V4UEFSQKxoUDGERJYiyhCUhiQbISliyJzPz/v6oGtI0PZklNVMz3d/POXVquta35326qvrpt96SJKlwJp8lSZIkSZIkSYUz+SxJkiRJkiRJKpzJZ0mSJEmSJElS4Uw+S5IkSZIkSZIKZ/JZkiRJkiRJklQ4k8+SJEmSJEmSpMKZfJYkSZIkSZIkFc7ksyRJkiRJkiSpcCafJUmSJEmSJEmFM/ksSZIkSZIkSSqcyWdJkiRJkiRJUuFMPkuSJEmSJEmSCmfyWZIkSZIkSZJUOJPPkiRJkiRJkqTCmXyWJEmSJEmSJBXO5LMkSZIkSZIkqXAmnyVJkiRJkiRJhTP5LEmSJEmSJEkqnMlnSZIkSZIkSVLhTD5LkiRJkiRJkgpn8lmSJEmSJEmSVDiTz5IkSZIkSZKkwpl8liRJkiRJkiQVru6SzxExMiJ+HBEvRMTqiJgTEVdFxJZll02SJEmSJEmSGsWAsgtQpIgYDTwAjABuA54CDgA+Cbw7Ig5JKb1YYhElSZIkSZIkqSHUW8vn75Mlns9LKZ2QUvpcSulI4EpgLPCVUksnSZIkSZIkSQ2ibpLPEbEr8E5gDvC9qtkXA8uBUyNi4x4umiRJkiRJkiQ1nLpJPgNH5uM7U0otlTNSSkuB+4GNgIN6umCSJEmSJEmS1Gjqqc/nsfn4mTbmzyBrGb07cFd7G7vmmmsKKpYkSZIkSZIklSZ9/OMfjzJ2XE8tnzfPx6+0Mb91+ha1ZkbERyJiWuGlkiRJkiRJkqQGVE8tn9vTmt1PtWamlH4E/KjniiNJkiRJkiRJ9aueWj63tmzevI35m1UtJ0mSJEmSJEnqJvWUfH46H+/exvwx+bitPqElSZIkSZIkSQWJlGr2QtHnRMRoYCYwBxidUmqpmLcpMI8s2T48pbS8lEJKkiRJkiRJUoOom5bPKaVngTuBUcA5VbMvBTYG/sfEsyRJkiRJkiR1v7pp+QyvtX5+ABgB3AY8CRwIHEHW3cb4lNKL5ZVQkiRJkiRJkhpDXSWfASJiR+BLwLuBYWTdbdwKXJpSWlJm2SRJkiRJkiSpUdRd8lmSJEmSJEmSVL666fNZkiRJkiRJktR7mHyWJEmSJEmSJBVuQNkF6I369++XWlrsjkSSJEmqBxFBkd0N9gso8utCP6CluM31qKL/t10vRz9SKvC/GAFlvK/oB0W+j0ZTVr1JUi82cOAg1q5dszilNLyM/Zt8rqGlJZGmlV0KSZIkSUWI/RNTp04tbHuHHXYY6erCNkecC+nNE4vbYA+KR2/h2Kc/XHYxuH3sJE7585rCtnfDwYPY7+YZhW2vox4+aQx83uRpl301OOV908suhST1KjfctCfAP8vav91uSJIkSZIkSZIKZ/JZkiRJkiRJklQ4k8+SJEmSJEmSpMLZ57MkqeF88CK466+wfBVsOww++yE484SySyVJKtPcuXM544wzOPzww7nooos2aFsTroIH58CAvKnPDlvA0/+14WXsihtfnsulC6bzrzUr2HbgEK7bcX8O3biU5w29zh37/PR1r5tXNTPqA3vwposO7pb9LZs3h2nfPI/Fjz9I/4GD2fGIE9n3/CvoN6C4r8Qrn5vJ3GsvZfmsxxm42VbscOoFbHngOwvbfo+adjU8eh0segz2PAWOvW7dvNl3weRz4NV/wfYHZvM237mkgkrlembGz5g9+1ZefuUZdt7paA468KsANDev4c8PfpYlSx5n+YoXOPKI69hmxAEll1Yqh8lnSVLDufB0mHQRDB4ET82BCR+FfcbCfnuUXTJJUlmuvPJKxo0bV9j2rj4Zzhxf2Oa65A9LF3DBvMe4aacDOWCjrZjXtKrcAlU46pHTXvu7acVa7jzk52z/7lHdtr9p3zyPIVsO5723/4s1y17mT+e9hxm/+gFjTz63kO2n5iae/cbHGf6OUxhz0XUsnf5Xnr3sowzd8TaGbL9LIfvoUZtsD4d8EWZNhqaV66avWAy/OhGOuhbGHAv3XAS/fh+c/mB5ZZVKNHToCPba86PMm38/zc2rXzdv6633Zezup3LfA/9ZUumk3sFuNyRJDWev0VniGSDy4dnnyiyRJKlMd911F5tuuin77rtv2UUp1MULnuC/ttmDgzYeRr8Idhg4lB0GDi27WG8wb/IcBm81lK3237bb9rH8hdns9LaJ9B88hKHDtmW7g97FK7OmF7b9Vc/PYu2ShYw45gyif382e9PBbDx2X16cemth++hR406EsSfA0GGvn/70r2DrvWCPk2DAEDj0Elj4D1j8VCnFlMq248h3MHLk2xk8eIvXTe/ffxDjxn6I4cP3I6J/SaWTegeTz5KkhnT212GjQ2DcRNhuazjqkLJLJEkqw/Lly5k0aRJnn312odu98Dew9QVwyLdgyjOFbrpDmlNi2sqXWNS0mt2e+h0jn/wt5z7/CCtbmnu+MO2Y++sZjDxhNyKi2/ax+8mf4J9/+AVNq1awYuHzzHvw92x3UIFdYqRUc9qqf80obh+9waInYMS/rXs9aGPYYjQsfqK8MkmSejWTz5KkhvT9z8HSqXDvtXDiEetaQkuSGsu1117L0UcfzTbbbFPYNi87AWZdAs9/GT4yHo79ITy7qLDNd8iCplWsTYlbXnmee0dP4O9j3s4jK1/mywuf7NmCtGPFC8t48aH57HjCmG7dz4h9D+WV2dO55e3DuO34Xdhq3H6MPPz4wrY/ZIddGbD5Viz4zX+Tmtby6j/uZdn0h2hZs7L9lfuStctgyOavnzZkc1iztJzySJJ6PZPPkqSG1b8/vPX/wXML4Zpbyi6NJKmnzZgxg4cffpiTTz650O0eOAo2HQKDB8JpB8Ehu8IdPdwwdGh+m/cnho1mu4FD2XrAYD41fAx3vDq/ZwvSjuduncFW+23DRjtu2m37SC0tTDn/GHaccAIn3f0yJ/5+HmuWvsTfv3dhYfuIAQMZ/dlreOXhKfzjrPEsuP3HbDn+PQzcqvu6EinFwE1g9auvn7b6VRjUffUnSerbfOCgJKnhNTXb57MkNaJHHnmE+fPnc9JJJwGwcuVKmpubmTNnDpMmTSpsPxFQo1OGbrXlgEGMHDi0W7uyKMJzt81kt7P+rf0FN8CaV5ewYsFcxkw8m/6DBtN/0GB2Pfo0Hv3Rxexz7tcL289GO49j7Jd+/trrp75wMsMmvLew7fcKw/eCx3667vWa5fDSs1k/0JIk1WDLZ0lSQ1m4BG6cDMtWQHMzTP4z3DAZjty/7JJJknracccdxw033MCkSZOYNGkSxx13HAcffDBXXHFFl7f58gqYPB1Wrc1+3PzZQzB1JrxrjwIL3kFnbDmK7y6eycKmVbzUtIarFs3gmM16T0vcJX9bwKoFK9j+3aO6dT+Dt9iajbffhZm/+iEtTU2sWfoys+/4X7bY7c2F7mfFP5+iZc1qWlavZP5vrmXtS4sYNuHEQvfRY1qaoGkVpOZsaFqVTdv9vbDocXjql9m0+74EI94MW48ru8RSKVpammhuXk1Laial5uzvliYAmpvX0Ny8Ol9uLc3Nq0m1+oeX6pwtnyVJDSUCrvklfOxr0JJg523hqk/D8RPKLpkkqacNGTKEIUOGvPZ66NChDBo0iC222KLL21zbDF/8P3hqAfTvB+O2gVvPgrHFdSndYRdtsweLm1ez+1OTGdKvHydvviNfGFFCFrwNc2+dwbbv2JkBm3T/gxcO/dpN/O2qzzD9+suJfv3ZZt/D2feTlxe6jyX33Mriu28mNTWxyR77M+ai6+g3cHCh++gx930Z7rt03evHr4e3XgyHXQIn/hImnwu/+SBsfyCccGNpxZTK9sT0H/D4E99/7fWcf97O3nudzZv2Ppff3nEUy1e8AMCUe84C4Nhj/sAmG+9QSlmlsoS/urxRRKQ0rexSSJIkSSpC7A9Tp04tbHuHHXYY6erCNkecC+nNE4vbYA+KR2/h2Kc/XHYxuH3sJE7585rCtnfDwYPY7+YZhW2vox4+aQx83u/oXfbV4JT3TS+7FJLUq9xw054AD6eUSrnf1243JEmSJEmSJEmFM/ksSZIkSZJLWd4XAAAQuUlEQVQkSSqcyWdJkiRJkiRJUuFMPkuSJEmSJEmSCmfyWZIkSZIkSZJUuEjJJ+lWi4ilwNNll0O9ytbA4rILoV7HuFAtxoWqGROqxbhQLcaFqhkTqsW4UC3GhdqyNbBxSml4GTsfUMZO+4CnU0r7l10I9R4RMc2YUDXjQrUYF6pmTKgW40K1GBeqZkyoFuNCtRgXakseG6PK2r/dbkiSJEmSJEmSCmfyWZIkSZIkSZJUOJPPtf2o7AKo1zEmVItxoVqMC1UzJlSLcaFajAtVMyZUi3GhWowLtaXU2PCBg5IkSZIkSZKkwtnyWZIkSZIkSZJUOJPPkiRJkiRJkqTCmXzORcTIiPhxRLwQEasjYk5EXBURW5ZdNm2YiBgWEWdGxK8jYmZErIyIVyLivoj4cETU/BxExPiIuCMilkTEioh4NCLOj4j+69nXMRExJd/+soj4S0Sc1n3vTkWKiFMjIuXDmW0s0+k6jojTIuKv+fKv5Osf0z3vQkWIiEMj4pcRMS8/J8yLiDsj4qgay3qsaAARcXQeA8/l55FZEXFzRBzcxvLGRR2IiIkR8d2IuDciXs3PD9e3s06P1L3nlnJ0JiYiYkxEXBARd0fE3IhYExELIuK2iDiinf10qn4jon8eZ4/mx6gleRyO39D3rPZ15VhRtf6kimvQ3dpYptN1HBFDI+LSiHg6IlZFxMKI+EVE7NGV96nO6eI5JPLP/5S8jldGxOy83nZvYx2PF31EZ2MiIgZHxDl5/S7O6/jJiPhOROy8nvWMiT4memneKiIOiYhvRMRDEbEosu/GsyPi2rbOVzWllBp+AEYDC4AE3Ap8Hbg7f/0UMKzsMjpsUP1+LK/LF4CfAV8Dfgy8nE+/hbz/84p1jgeagGXAJOCbeSwk4OY29nNuPn8x8D3gSmBuPu3ysv8PDu3GyY55TCzN6+zMIuoYuDyfPzdf/nvAi/m0c8t+3w416+yLef0sAn4CfJXsAQ0PAd+oWtZjRQMMwGUVdXZtfp1wC7AGaAE+aFzU5wD8Pa+DpcCT+d/Xr2f5Hql7zy19IyaAG/P5TwA/JLsG/VUeIwk4r4j6BQK4mXXfXb6Zx9+yfF/Hl/1/q/ehs8eKqnWPrVg3AbsVUcfAYOC+fJ2H8nPZz4G1wHLgwLL/b/U+dOEcMgS4vaKer86PGz8FZgHH1FjH40UfGjp5DhlQ8Rl+EvhuXt/35NNeBvY0JupjoJfmrYD5QDNwL3BVHl/358svAw7u0Psr+x/cGwZgcv6P+0TV9G/l039QdhkdNqh+jyS7qOtXNX1b4F95Hf97xfTNgIXAamD/iulDgAfy5d9fta1RwKr8oD6qYvqWwMx8nQ59KB1KiZEA/gg8mx+w35B87kodA+Pz6TOBLau29WK+vVHd9b4cuhQLJ+V19gdg0xrzB1b87bGiAYb8XNGcX3iNqJp3RF5ns4yL+hzyOh6TnycmsP4viT1S955b+lRMnA7sU2P64WQ/Xq0GttvQ+gVOyde5HxhSMf0t+T4WUuOc5lBOXFStNzw/v9wITKHt5HOn6xi4MF/nZiq+B5ElK1p/FOnXlffr0D1xQZYISmQNH95QN1Rch+avPV70saGT55DW7yV/rI4H4NJ83o+NifoY6KV5K+ACYPsa5f18vvxjHXl/Dd/tRkTsCrwTmEN2sK90MdmvwqdGxMY9XDQVJKV0d0rp9pRSS9X0+cAP8pcTKmZNJLsQvDGlNK1i+VVkLSIBPl61m/8ga11wdUppTsU6L5FdPED2S5Z6p/PIDvZnkH3ma+lKHbe+/kq+XOs6c8iON4PzfaoXyG9lugxYAXwgpbS0epmU0tqKlx4rGsPOZN2U/SWltLByRkrpT2QtV4ZXTDYu6khK6U8ppRkpv8puR0/VveeWEnUmJlJK16WUHqkx/R6yROMgskRBpa7Ub2tcfTGPt9Z1HgJuIovLie2VV13XyWNFpR/l43PaWa5TdRwRwbpY+mzl96CU0m1kLdj2JPshRN2kM3EREaPJ6uwh4AvV313z7a2tmuTxoo/p5LFi13z82xrxcFs+Hl413Zjoo3pr3iqldFlK6YUaRb4MWAnsHRHD2nt/DZ98Jks4AdxZo5KXkv36sxFwUE8XTD2i9QTeVDGtNSZ+X2P5qWSJqfERMbiD6/yuahn1Inmfd18Hvp1SmrqeRbtSx8ZF3zIe2AW4A3gpsj5+L4iIT0btfn09VjSGGWQtFA+IiK0rZ0TEYcCmZC1SWhkXjaun6t54qQ+1rkGhk/Wbx9N4svi6tyPrqHeIiNOBE4CPpZReXM9yXanj0cBOwDMppdkdXEflOoUsP/NTYLOI+GBEXBgRH1lPv6oeL+rbE/n4PTX6+23tv/mPVdONifrUG/NWqaI8ze0tPKCDG61nY/PxM23Mn0HWMnp34K4eKZF6REQMAD6Uv6z88LUZEymlpoiYDexF9kvkkx1YZ15ELAdGRsRGKaUVRZRfGy6Pgf8lu43l8+0s3qk6zu+W2AFYllKaV2N7M/JxzQeHqBRvyccLgL8Bb6qcGRFTgYkppUX5JI8VDSCltCQiLiDrimt6RNxKdqvaaOA4si5aPlqxinHRuLq97j231If8IVFvI/tiOLVielfqdzegP1n3P9WJ7LbWUcnyGPg22e32t7azeFfquCPfcavXUblar0M3J+sKsLIlYYqIa8j6iW8GjxcN4rdkzwk4EXgsIv5I1iBiP+CtZH1AX926sDFRn3px3uokskY4D6aUXm7vfdjyOTu4A7zSxvzW6Vv0QFnUs74O7A3ckVKaXDG9KzHR0XU2b2O+yvFfwD7A6Smlle0s29k69tjS94zIxx8DhgJvJzuh7k32bIDDyPpNbOWxokGklK4iu/AfAJwFfI7sgmsucF1VdxzGRePqibr33NLH5S2QfkZ22+sllbdF070xZEz0EnkLxp+SPajpvA6sYlw0htbr0C8B08gaQWxK9kPVs8DZwEUVyxsXdS7vmmMicAlZ0vA84DNk/UZPBX7e+mNEzpioT70ubxURu5D9+NEEfHp9y7Yy+dy+yMed7b9LvVhEnEf2IXkKOLWzq+fjzsSEcdTLRMQBZK2dr0gp/bmITebjztaxMdF79M/HQdbC+a6U0rKU0hPAe4HngMPb6IKjFo8VdSIiPkv2hOnryFo8b0zW6mQW8LOI+EZnNpePjYvG05N1b6z0QhHRn+yOq0PI+tG8vIub8vjRt/0nWV/LZ1X9+NBVnlfqQ+t16DzgvSmlx/Pr0LvJEpAtwKciYlAnt2tc9FERMYTsXPEZsn7htyNLCh5F9kySqRFxfBc2bUz0Eb0xbxURI8i66BgOfDKl9EBHNmzyuf3M/mZVy6mPi4hzyG5zmw4ckVJaUrVIV2Kio+u82omiqptUdLfxDK9vQbA+na3j9pZv71dH9bzWL4CzUkr/qJyRt4xv/aX5gHzssaIBRMQEsgdq/Cal9KmU0qyU0oqU0t/IfpR4Hvh0/gBjMC4aWU/UveeWPipPPF9PdtfEL4AP1njgVFfq1+8yfUhEjAG+AvwkpXRHB1frzmOLcdF7tF6H/r76jsz8unQ2WUvoPfLJHi/qX+uddl9IKf0wpTQ/pfRqSul3ZD9IDCTLa7QyJupIb8xb5Ynnu8la4n8ypfT9NrbzBiaf4el83FYfNmPycVv9ZakPiYjzyfpFepzsAzy/xmJtxkSetNyF7PaCWR1cZzuyVnLP2Vdnr7EJWV3tAayKiNQ6ABfny/x3Pu2q/HWn6jiltJwsKbVJPr+ax5bep7WO2+qzqvVLwdCq5T1W1LfWB7r8qXpGXk9/Jbue2iefbFw0rm6ve88tfVNe/zcA7wd+DnygVt+aXazfmWQP+tk1309H1lF59iLrcuWMyuvP/Br08HyZGfm0E/LXXaljv+P2PZ26DvV40RDWdw36D2AJsHNEDMunGRN1ojfmrfL5U4A9gXNSSt/p6PsBk8+w7oP8zuoniEbEpmS3xa0EHuzpgqlY+QOjrgT+TvYBXtjGonfn43fXmHcYsBHwQEppdQfXeU/VMirfamBSG8Mj+TL35a9bu+ToSh0bF33LVLIT9Jg2bmncOx/PycceKxpD6xOih7cxv3X6mnxsXDSunqp746UPyc8nt5C1Xvsf4NSqPjqrdap+83h6gCy+Du3IOirVHNq+Bm1NLtycv54DXa7jZ8keqL173jdnR9ZRue7Kx3tXz8j7im9NAs6pmOXxor61eQ2ax0RrK9U1FbOMiT6uN+atImIkcA8wDvhYZ1o8vyal1PAD2a3UCfhE1fRv5dN/UHYZHTa4ji/K63IasFU7y24GLCJLUO5fMX0I2YE5Ae+vWmcXYBXwIjCqYvqWZL8mJuDgsv8PDh2KlUvy+jpzQ+sYGJ9PnwlsWTF9VL6dVZXbcih/ILslOgFfrpr+DrK+9l4GtsineaxogAE4Oa+X+cAOVfPek8fFSmCYcVHfAzAhr4vr25jfI3XvuaX3DB2IicHAb/NlrgX6dWCbna5f4JR8nfuBIRXT35LH40Jgs7L/X40ytBcX61lvSr7ebjXmdbqOgQvzdW6ujD3g+Hz6Ex2JSYeeiQtgENmPBi3AO6rmfTlfd0rVdI8XfXjoQEx8P5//R2Bw1byv5fP+akzUz0AvzFsBO+XHpmbgjK6+t8g31tAiYjRZ5YwAbgOeBA4ke4roM8D4lNKL5ZVQGyIiTiN7SFQz2RM5a/VXNCeldF3FOieQtVJZBdxIdkvLcWR929wCnJyqPjwR8QngO2Qf5JvIfoGcCIwke6jdZ4p8X+oeEXEJWdcbZ6WUrq2a1+k6jogrgE+RPazuFrILy/cBw8h+8Lq6296MOi3vx+p+YDfgXrIuFXYm69s3kd0qfXPF8h4r6lx+V9Rk4O3AUuDXZInoPchuhwzg/JTStyvWMS7qRF6Xrbe+bwu8i+z2xXvzaYsr66an6t5zS3k6ExMR8RPgdGAx65II1aaklKZU7aNT9RsRQdaP9ESyhxLdni/7PrIvof+eUrqtq+9Z7evssaKNbUwh63pjTEppZtW8Ttdx3jLybrJk1DSylrU7kbXCXwMcmVL6SxferjqoC+eQtwJ3kn3mfw38kywBeBhZkumtKaXXdX/g8aJv6eQ5ZAeyO/BHkrV4/z1Zg4dDyJ5BsxJ4W0qp9U7d1n0YE31Qb81bRcRssh8vHgb+r43iX5dSmrPeN1h2Zr+3DMCOwE/Ini67huxA/23a+bXBofcPrGvJur5hSo31DgHuIOtfayXwGNnTqfuvZ1/Hkt2OsBRYDjwEnFb2/8ChS/FyZhvzO13HwGn5csvz9e4Bjin7vTq0WV9bkd35Mjs/H7xI9sPkQW0s77GizgeyB7qcT/YF4FWy7lkWkl2AvdO4qN+hA9cQc8qqe88tvT8mWNeSdX3DJUXULzAgj7PH8rh7KY/D8WX/zxph6MqxosY2WuPlDS2fu1rHZP0DXwrMIGsdt4isJfSeZf/PGmHo4jlkT7KE0EKy69C5wA+BkevZj8eLPjJ0NibIuty4nKyB5CrW5ap+AowzJupn6EBslJK36kCZEjChvfdny2dJkiRJkiRJUuF84KAkSZIkSZIkqXAmnyVJkiRJkiRJhTP5LEmSJEmSJEkqnMlnSZIkSZIkSVLhTD5LkiRJkiRJkgpn8lmSJEmSJEmSVDiTz5IkSZIkSZKkwpl8liRJkiRJkiQVzuSzJEmSJEmSJKlwJp8lSZIkSZIkSYX7/yIBAVfnF7KmAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 1728x576 with 2 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "#Plotting the predicted fitness effect for the variants seen more than once\n",
    "plt.close('all')\n",
    "f, (ax1, ax2) = plt.subplots(2, 1, sharey=True, sharex = True, figsize=(24, 8))\n",
    "gs = matplotlib.gridspec.GridSpec(2, 1, width_ratios=[1], height_ratios=[10,0.5])\n",
    "ax1 = plt.subplot(gs[0])\n",
    "ax2 = plt.subplot(gs[1])\n",
    "gs.update(hspace=0.05)\n",
    "\n",
    "nonsense_list = []\n",
    "y_list = []\n",
    "x_list=[]\n",
    "y_list_top20 = []\n",
    "x_list_top20=[]\n",
    "y_list_nonsense = []\n",
    "x_list_nonsense = []\n",
    "x=[]\n",
    "y=[]\n",
    "OBS_EXP_DICT={}\n",
    "top20_lower_err = []\n",
    "top20_upper_err = []\n",
    "missense_lower_err = []\n",
    "missense_upper_err = []\n",
    "nonsense_lower_err = []\n",
    "nonsense_upper_err = []\n",
    "\n",
    "top20 = c1\n",
    "missense = c6\n",
    "nonsense = c10\n",
    "synonymous = c3\n",
    "\n",
    "axisfont=20\n",
    "domainfont = 12\n",
    "titlefont=12\n",
    "axislabelfont=12\n",
    "\n",
    "top_20_TET2 = []\n",
    "\n",
    "#ERROR NEEDS TO BE IN THE FORMAT: [tuple(lower_err),tuple(upper_err)]\n",
    "\n",
    "for k, v in TET2_missense_s.items():\n",
    "    if k in top_20_TET2:\n",
    "        codon_number = int(((re.findall('\\d+', k)))[0])\n",
    "        s_value = float(v[0])*100\n",
    "        x_list_top20.append(codon_number)\n",
    "        y_list_top20.append(s_value)\n",
    "        \n",
    "        s_upper = float(v[3])*100\n",
    "        s_lower = float(v[4])*100\n",
    "        error_upper = s_upper-s_value\n",
    "        error_lower = s_value-s_lower\n",
    "        top20_lower_err.append(error_lower)\n",
    "        top20_upper_err.append(error_upper)\n",
    "\n",
    "    if k not in top_20_TET2:\n",
    "        codon_number = int(((re.findall('\\d+', k)))[0])\n",
    "        s_value = float(v[0])*100\n",
    "        x_list.append(codon_number)\n",
    "        y_list.append(s_value)\n",
    "        \n",
    "        s_upper = float(v[3])*100\n",
    "        s_lower = float(v[4])*100\n",
    "        error_upper = s_upper-s_value\n",
    "        error_lower = s_value-s_lower\n",
    "        missense_lower_err.append(error_lower)\n",
    "        missense_upper_err.append(error_upper)\n",
    "\n",
    "    missense_err = [tuple(missense_lower_err), tuple(missense_upper_err)]   \n",
    "    \n",
    "for k, v in TET2_nonsense_s.items():\n",
    "    if k in top_20_TET2:\n",
    "        codon_number = int(((re.findall('\\d+', k)))[0])\n",
    "        s_value = float(v[0])*100\n",
    "        x_list_top20.append(codon_number)\n",
    "        y_list_top20.append(s_value)\n",
    "        \n",
    "        s_upper = float(v[3])*100\n",
    "        s_lower = float(v[4])*100\n",
    "        error_upper = s_upper-s_value\n",
    "        error_lower = s_value-s_lower\n",
    "        top20_lower_err.append(error_lower)\n",
    "        top20_upper_err.append(error_upper)\n",
    "    \n",
    "    top20_err = [tuple(top20_lower_err), tuple(top20_upper_err)]\n",
    "        \n",
    "    if k not in top_20_TET2:\n",
    "        codon_number = int(((re.findall('\\d+', k)))[0])\n",
    "        s_value = float(v[0])*100\n",
    "        x_list_nonsense.append(codon_number)\n",
    "        y_list_nonsense.append(s_value)\n",
    "        \n",
    "        s_upper = float(v[3])*100\n",
    "        s_lower = float(v[4])*100\n",
    "        error_upper = s_upper-s_value\n",
    "        error_lower = s_value-s_lower\n",
    "        nonsense_lower_err.append(error_lower)\n",
    "        nonsense_upper_err.append(error_upper)\n",
    "\n",
    "    nonsense_err = [tuple(nonsense_lower_err), tuple(nonsense_upper_err)]        \n",
    "    \n",
    "msize = 12\n",
    "medgewidth = 2\n",
    "linew = 3\n",
    "csize = 4\n",
    "cthick = 3\n",
    "\n",
    "markers1, caps1, bars1 = ax1.errorbar(x_list, y_list, yerr = missense_err, fmt = 'o', ecolor = missense, elinewidth = linew, \\\n",
    "             capsize = csize, capthick = cthick, markersize = msize, markeredgewidth = medgewidth, \\\n",
    "             markeredgecolor = missense, markerfacecolor = missense, zorder = 1)    \n",
    "\n",
    "markers2, caps2, bars2 = ax1.errorbar(x_list_nonsense, y_list_nonsense, yerr = nonsense_err, fmt = 'o', ecolor = nonsense, elinewidth = linew, \\\n",
    "             capsize = csize, capthick = cthick, markersize = msize, markeredgewidth = medgewidth, \\\n",
    "             markeredgecolor = nonsense, markerfacecolor = nonsense, zorder = 2)    \n",
    "\n",
    "markers3, caps3, bars3 = ax1.errorbar(x_list_top20, y_list_top20, yerr = top20_err, fmt = 'o', ecolor = top20, elinewidth = linew, \\\n",
    "             capsize = csize, capthick = cthick, markersize = msize, markeredgewidth = medgewidth, \\\n",
    "             markeredgecolor = top20, markerfacecolor = top20, zorder = 3)    \n",
    "\n",
    "# loop through bars and caps and set the alpha value\n",
    "a = 0.7\n",
    "[bar.set_alpha(a) for bar in bars1]\n",
    "[cap.set_alpha(a) for cap in caps1]\n",
    "[bar.set_alpha(a) for bar in bars2]\n",
    "[cap.set_alpha(a) for cap in caps2]\n",
    "[bar.set_alpha(a) for bar in bars3]\n",
    "[cap.set_alpha(a) for cap in caps3]\n",
    "\n",
    "#Only show the required axis lines\n",
    "ax1.spines['top'].set_visible(False)\n",
    "ax1.spines['right'].set_visible(False)\n",
    "\n",
    "for axis in ['bottom','left']:\n",
    "    ax1.spines[axis].set_linewidth(2)\n",
    "\n",
    "for axis in ['bottom','left']:\n",
    "    ax1.spines[axis].set_color(grey3)\n",
    "    \n",
    "ax1.set_ylabel('fitness effect (s)', fontsize = axisfont, labelpad = 10)\n",
    "ax1.set_title('Fitness effects across TET2', fontsize = 18)\n",
    "\n",
    "# ax1.set_yscale('log')\n",
    "\n",
    "# STACKED BAR CHART\n",
    "# define the exons + create data\n",
    "exons = [{'3':(1137-1)+1, '4': (1167-1138)+1, '5': (1198-1168)+1, '6':(1268-1199)+1, '7':(1318-1269)+1, '8': (1348-1319)+1, '9': (1394-1349)+1,'10': (1513-1395)+1, '11': (2002-1514)+1}]\n",
    "exon_orders = [['3', '4', '5', '6', '7', '8', '9', '10', '11']]\n",
    "colors2 = [c4, c0, c3, c1, c5, c6,c7, c8, c9]\n",
    "names = exons[0].keys()\n",
    "values = np.array([[data[name] for name in order] for data,order in zip(exons, exon_orders)])\n",
    "lefts = np.insert(np.cumsum(values, axis=1),0,0, axis=1)[:, :-1]\n",
    "orders = np.array(exon_orders)\n",
    "\n",
    "#plot the bar chart\n",
    "for name, color in zip(names, colors2):\n",
    "    idx = np.where(orders == name)\n",
    "    value = values[idx]\n",
    "    left = lefts[idx]\n",
    "    ax2.bar(x=left, height=-10, width=value, bottom=-5, color=color, orientation = 'horizontal', label = name, edgecolor='k', linewidth=1, tick_label = name)\n",
    "    ax2.text(left + (value/2.), (-0.26), name, ha='center', va='center', fontsize = 12)\n",
    "\n",
    "# CONFIGURING THE GRAPH APPEARANCE\n",
    "#Set the x and y axis limits\n",
    "ax1.set_xlim(0, 2004)\n",
    "ax1.set_ylim(-0.01, 30)\n",
    "ax2.set_xlim(0, 2004)\n",
    "ax2.set_ylim(-0.5, 0)\n",
    "\n",
    "#x-axis ticks\n",
    "x1_major_ticks = []\n",
    "x1_major_tick_labels = []\n",
    "x2_major_ticks = [0, 200, 400, 600, 800, 1000, 1200, 1400, 1600, 1800, 2002]\n",
    "x2_major_tick_labels = [\"0\", \"200\", \"400\", \"600\", \"800\", \"1000\", \"1200\", \"1400\", \"1600\", \"1800\", \"2002\"]\n",
    "ax1.set_xticks(x1_major_ticks)\n",
    "ax1.set_xticklabels(x1_major_tick_labels, fontsize = axisfont)\n",
    "ax2.set_xticks(x2_major_ticks)\n",
    "ax2.set_xticklabels(x2_major_tick_labels, fontsize = axisfont)\n",
    "\n",
    "#y-axis ticks\n",
    "y1_major_ticks = [0, 5, 10, 15, 20, 25, 30]\n",
    "y1_major_tick_labels = [\"0\", \"5\", \"10\", \"15\", \"20\", \"25\", \"30\"]\n",
    "ax1.set_yticks(y1_major_ticks)\n",
    "ax1.set_yticklabels(y1_major_tick_labels, fontsize = axisfont)\n",
    "\n",
    "#set panel legends\n",
    "legend_marker_size = 2\n",
    "legend_elements = [Line2D([0], [0], marker = '_', color = missense, alpha = 1.0, markersize = legend_marker_size, \\\n",
    "                          lw=3, label='missense'),\n",
    "                   Line2D([0], [0], marker = '_', color=nonsense, alpha=1.0, markersize = legend_marker_size, \\\n",
    "                          lw=3, label='nonsense'),\n",
    "                  Line2D([0], [0], marker = '_', color=top20, alpha=1.0, markersize = legend_marker_size, \\\n",
    "                          lw=3, label='top 20 observed variant')]\n",
    "    \n",
    "ax1.legend(ncol=1, handles=legend_elements, loc='upper left', frameon=0, fontsize = 20)\n",
    "\n",
    "ax1.grid(which='major', linestyle=':', linewidth='0.7', color=grey3, zorder = 0)\n",
    "\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## ASXL1 variants"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 250,
   "metadata": {
    "code_folding": [
     0,
     86,
     97,
     109
    ]
   },
   "outputs": [],
   "source": [
    "def expected_number_variants_missense_ASXL1(data, params):\n",
    "    s_value = params[0]\n",
    "    \n",
    "    a = 0\n",
    "    expected_number_by_variant_missense = {}\n",
    "    \n",
    "    Jaiswal_fmin = 0.056\n",
    "    Jaiswal_mean = 58\n",
    "    Jaiswal_sigma = 11.5\n",
    "    Jaiswal_total = 17182\n",
    "    \n",
    "    Genovese_fmin = 0.1194\n",
    "    Genovese_mean = 55.1\n",
    "    Genovese_total = 12380\n",
    "    Genovese_sigma = 11.95\n",
    "    \n",
    "    McKerrel_fmin = 0.008\n",
    "    McKerrel_mean = 51.88\n",
    "    McKerrel_total = 4219\n",
    "    McKerrel_sigma = 18.36\n",
    "    \n",
    "    Zink_fmin = 0.2363\n",
    "    Zink_mean = 60\n",
    "    Zink_total = 11262\n",
    "    Zink_sigma = 1.18907720e+01\n",
    "    \n",
    "    Coombs_fmin = 0.019\n",
    "    Coombs_mean = 59.82\n",
    "    Coombs_total = 5649\n",
    "    Coombs_sigma = 15.14\n",
    "    \n",
    "    Young_fmin = 0.0009\n",
    "    Young_mean = 63\n",
    "    Young_total = 89\n",
    "    Young_sigma = 6.8\n",
    "    \n",
    "    Desai_fmin = 0.0169\n",
    "    Desai_mean = 66.1\n",
    "    Desai_total = 424\n",
    "    Desai_sigma = 6 #estimated\n",
    "    \n",
    "    Acuna_fmin = 0.0015\n",
    "    Acuna_mean = 4.5\n",
    "    Acuna_total = 2014\n",
    "    \n",
    "    N = 9.39358404e+04\n",
    "\n",
    "    for i in data:\n",
    "        variant_mutation_rate = mutation_rates_variants_ASXL1(i)\n",
    "\n",
    "        if i in Genovese_missense_ASXL1:\n",
    "            Genovese_number = prevalence_averaged_over_ages(Genovese_fmin, s_value, N, variant_mutation_rate, Genovese_mean, Genovese_sigma, Genovese_total)\n",
    "        else:\n",
    "            Genovese_number = 0 \n",
    "\n",
    "        if i in Zink_missense_ASXL1:\n",
    "            Zink_number = prevalence_averaged_over_ages(Zink_fmin, s_value, N, variant_mutation_rate, Zink_mean, Zink_sigma, Zink_total)\n",
    "        else:\n",
    "            Zink_number = 0\n",
    "\n",
    "        if i in Coombs_missense_ASXL1:\n",
    "            Coombs_number = prevalence_averaged_over_ages(Coombs_fmin, s_value, N, variant_mutation_rate, Coombs_mean, Coombs_sigma, Coombs_total)\n",
    "        else:\n",
    "            Coombs_number = 0\n",
    "\n",
    "        if i in Young_missense_ASXL1:\n",
    "            Young_number = prevalence_averaged_over_ages(Young_fmin, s_value, N, variant_mutation_rate, Young_mean, Young_sigma, Young_total)\n",
    "        else:\n",
    "            Young_number = 0\n",
    "\n",
    "        if i in Desai_missense_ASXL1:\n",
    "            Desai_number = prevalence_averaged_over_ages(Desai_fmin, s_value, N, variant_mutation_rate, Desai_mean, Desai_sigma, Desai_total)\n",
    "        else:\n",
    "            Desai_number = 0\n",
    "\n",
    "        if i in Acuna_missense_ASXL1:\n",
    "            Acuna_number = prevalence_averaged_over_uniform_ages(Acuna_fmin, s_value, N, variant_mutation_rate, Acuna_mean, Acuna_total)\n",
    "        else:\n",
    "            Acuna_number = 0\n",
    "\n",
    "        expected_total = Genovese_number + Zink_number + Coombs_number + Young_number + Desai_number + Acuna_number\n",
    "\n",
    "        expected_number_by_variant_missense[i] = expected_total\n",
    "    \n",
    "    return expected_number_by_variant_missense\n",
    "\n",
    "def prob_observed_given_expected_missense_ASXL1(params, data): #data = list of variant (e.g. p.R882H) (or all_missense_variants_ASXL1)\n",
    "    total_square_distance = 0\n",
    "    a = 0\n",
    "    for variant in data:\n",
    "        observed_number = float(all_studies_ASXL1_counts[variant])\n",
    "        expected_number = float(expected_number_variants_missense_ASXL1(data, params)[variant])\n",
    "        square_distance = ((observed_number - expected_number)**2)\n",
    "        total_square_distance = total_square_distance + square_distance\n",
    "\n",
    "    return total_square_distance\n",
    "\n",
    "def prob_observed_given_expected_missense_ASXL1_upper_error(params, data): #data = list of variant (e.g. p.R882H) (or all_missense_variants_ASXL1)\n",
    "    total_square_distance = 0\n",
    "    a = 0\n",
    "    for variant in data:\n",
    "        observed_number = float(all_studies_ASXL1_counts[variant])\n",
    "        observed_number_upper_error = observed_number+np.sqrt(observed_number)\n",
    "        expected_number = float(expected_number_variants_missense_ASXL1(data, params)[variant])\n",
    "        square_distance = ((observed_number_upper_error - expected_number)**2)\n",
    "        total_square_distance = total_square_distance + square_distance\n",
    "\n",
    "    return total_square_distance   \n",
    "\n",
    "def prob_observed_given_expected_missense_ASXL1_lower_error(params, data): #data = list of variant (e.g. p.R882H) (or all_missense_variants_ASXL1)\n",
    "    total_square_distance = 0\n",
    "    a = 0\n",
    "    for variant in data:\n",
    "        observed_number = float(all_studies_ASXL1_counts[variant])\n",
    "        observed_number_lower_error = observed_number-np.sqrt(observed_number)\n",
    "        expected_number = float(expected_number_variants_missense_ASXL1(data, params)[variant])\n",
    "        square_distance = ((observed_number_lower_error - expected_number)**2)\n",
    "        total_square_distance = total_square_distance + square_distance\n",
    "\n",
    "    return total_square_distance   "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 251,
   "metadata": {
    "code_folding": [
     0,
     91,
     102,
     114
    ]
   },
   "outputs": [],
   "source": [
    "def expected_number_variants_nonsense_ASXL1(data, params):\n",
    "    s_value = params[0]\n",
    "    \n",
    "    a = 0\n",
    "    expected_number_by_variant_nonsense = {}\n",
    "    \n",
    "    Jaiswal_fmin = 0.056\n",
    "    Jaiswal_mean = 58\n",
    "    Jaiswal_sigma = 11.5\n",
    "    Jaiswal_total = 17182\n",
    "    \n",
    "    Genovese_fmin = 0.1194\n",
    "    Genovese_mean = 55.1\n",
    "    Genovese_total = 12380\n",
    "    Genovese_sigma = 11.95\n",
    "    \n",
    "    McKerrel_fmin = 0.008\n",
    "    McKerrel_mean = 51.88\n",
    "    McKerrel_total = 4219\n",
    "    McKerrel_sigma = 18.36\n",
    "    \n",
    "    Zink_fmin = 0.2363\n",
    "    Zink_mean = 60\n",
    "    Zink_total = 11262\n",
    "    Zink_sigma = 1.18907720e+01\n",
    "    \n",
    "    Coombs_fmin = 0.019\n",
    "    Coombs_mean = 59.82\n",
    "    Coombs_total = 5649\n",
    "    Coombs_sigma = 15.14\n",
    "    \n",
    "    Young_fmin = 0.0009\n",
    "    Young_mean = 63\n",
    "    Young_total = 89\n",
    "    Young_sigma = 6.8\n",
    "    \n",
    "    Desai_fmin = 0.0169\n",
    "    Desai_mean = 66.1\n",
    "    Desai_total = 424\n",
    "    Desai_sigma = 6 #estimated\n",
    "    \n",
    "    Acuna_fmin = 0.0015\n",
    "    Acuna_mean = 4.5\n",
    "    Acuna_total = 2014\n",
    "    \n",
    "    N = 9.39358404e+04\n",
    "\n",
    "    for i in data:\n",
    "        variant_mutation_rate = mutation_rates_variants_ASXL1(i)\n",
    "\n",
    "        if i in Jaiswal_nonsense_ASXL1:\n",
    "            Jaiswal_number = prevalence_averaged_over_ages(Jaiswal_fmin, s_value, N, variant_mutation_rate, Jaiswal_mean, Jaiswal_sigma, Jaiswal_total)\n",
    "        else:\n",
    "            Jaiswal_number = 0\n",
    "\n",
    "        if i in Genovese_nonsense_ASXL1:\n",
    "            Genovese_number = prevalence_averaged_over_ages(Genovese_fmin, s_value, N, variant_mutation_rate, Genovese_mean, Genovese_sigma, Genovese_total)\n",
    "        else:\n",
    "            Genovese_number = 0 \n",
    "\n",
    "        if i in Zink_nonsense_ASXL1:\n",
    "            Zink_number = prevalence_averaged_over_ages(Zink_fmin, s_value, N, variant_mutation_rate, Zink_mean, Zink_sigma, Zink_total)\n",
    "        else:\n",
    "            Zink_number = 0\n",
    "\n",
    "        if i in Coombs_nonsense_ASXL1:\n",
    "            Coombs_number = prevalence_averaged_over_ages(Coombs_fmin, s_value, N, variant_mutation_rate, Coombs_mean, Coombs_sigma, Coombs_total)\n",
    "        else:\n",
    "            Coombs_number = 0\n",
    "\n",
    "        if i in Young_nonsense_ASXL1:\n",
    "            Young_number = prevalence_averaged_over_ages(Young_fmin, s_value, N, variant_mutation_rate, Young_mean, Young_sigma, Young_total)\n",
    "        else:\n",
    "            Young_number = 0\n",
    "\n",
    "        if i in Desai_nonsense_ASXL1:\n",
    "            Desai_number = prevalence_averaged_over_ages(Desai_fmin, s_value, N, variant_mutation_rate, Desai_mean, Desai_sigma, Desai_total)\n",
    "        else:\n",
    "            Desai_number = 0\n",
    "\n",
    "        if i in Acuna_nonsense_ASXL1:\n",
    "            Acuna_number = prevalence_averaged_over_uniform_ages(Acuna_fmin, s_value, N, variant_mutation_rate, Acuna_mean, Acuna_total)\n",
    "        else:\n",
    "            Acuna_number = 0\n",
    "\n",
    "        expected_total = Jaiswal_number + Genovese_number + Zink_number + Coombs_number + Young_number + Desai_number + Acuna_number\n",
    "\n",
    "        expected_number_by_variant_nonsense[i] = expected_total\n",
    "    \n",
    "    return expected_number_by_variant_nonsense\n",
    "\n",
    "def prob_observed_given_expected_nonsense_ASXL1(params, data): #data = list of variant (e.g. p.R882H) (or all_nonsense_variants_ASXL1)\n",
    "    total_square_distance = 0\n",
    "    a = 0\n",
    "    for variant in data:\n",
    "        observed_number = float(all_studies_ASXL1_counts[variant])\n",
    "        expected_number = float(expected_number_variants_nonsense_ASXL1(data, params)[variant])\n",
    "        square_distance = ((observed_number - expected_number)**2)\n",
    "        total_square_distance = total_square_distance + square_distance\n",
    "\n",
    "    return total_square_distance   \n",
    "\n",
    "def prob_observed_given_expected_nonsense_ASXL1_upper_error(params, data): #data = list of variant (e.g. p.R882H) (or all_nonsense_variants_ASXL1)\n",
    "    total_square_distance = 0\n",
    "    a = 0\n",
    "    for variant in data:\n",
    "        observed_number = float(all_studies_ASXL1_counts[variant])\n",
    "        observed_number_upper_error = observed_number+np.sqrt(observed_number)\n",
    "        expected_number = float(expected_number_variants_nonsense_ASXL1(data, params)[variant])\n",
    "        square_distance = ((observed_number_upper_error - expected_number)**2)\n",
    "        total_square_distance = total_square_distance + square_distance\n",
    "\n",
    "    return total_square_distance   \n",
    "\n",
    "def prob_observed_given_expected_nonsense_ASXL1_lower_error(params, data): #data = list of variant (e.g. p.R882H) (or all_nonsense_variants_ASXL1)\n",
    "    total_square_distance = 0\n",
    "    a = 0\n",
    "    for variant in data:\n",
    "        observed_number = float(all_studies_ASXL1_counts[variant])\n",
    "        observed_number_lower_error = observed_number-np.sqrt(observed_number)\n",
    "        expected_number = float(expected_number_variants_nonsense_ASXL1(data, params)[variant])\n",
    "        square_distance = ((observed_number_lower_error - expected_number)**2)\n",
    "        total_square_distance = total_square_distance + square_distance\n",
    "\n",
    "    return total_square_distance   "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 252,
   "metadata": {
    "code_folding": [
     0
    ],
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "#CALCULATING S FOR MISSENSE VARIANTS\n",
    "ASXL1_missense_s = {}\n",
    "a = 0\n",
    "for var in all_missense_variants_ASXL1:\n",
    "    initial_guess=[0.05]\n",
    "    if var in all_studies_ASXL1_counts.keys():\n",
    "        observed_number = float(all_studies_ASXL1_counts[var])\n",
    "#         print(observed_number)\n",
    "        if observed_number >1:\n",
    "            variant_mutation_rate = mutation_rates_variants_ASXL1(var)\n",
    "            \n",
    "            MLE = scipy.optimize.minimize(prob_observed_given_expected_missense_ASXL1, initial_guess, args=([var],), method='Nelder-Mead')\n",
    "            s_value = MLE.x[0]\n",
    "\n",
    "            upper_error_MLE = scipy.optimize.minimize(prob_observed_given_expected_missense_ASXL1_upper_error, \\\n",
    "                                                      initial_guess, args=([var],), method='Nelder-Mead')\n",
    "            s_value_upper = upper_error_MLE.x[0]\n",
    "\n",
    "            lower_error_MLE = scipy.optimize.minimize(prob_observed_given_expected_missense_ASXL1_lower_error, \\\n",
    "                                                      initial_guess, args=([var],), method='Nelder-Mead')\n",
    "            s_value_lower = lower_error_MLE.x[0]\n",
    "\n",
    "#             print('a = ', a)\n",
    "            ASXL1_missense_s[var]=(s_value, observed_number, variant_mutation_rate, s_value_upper, s_value_lower) \n",
    "    a = a+1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 253,
   "metadata": {
    "code_folding": [
     0
    ],
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "#CALCULATING S FOR NONSENSE VARIANTS\n",
    "ASXL1_nonsense_s = {}\n",
    "a = 0\n",
    "for var in all_nonsense_variants_ASXL1:\n",
    "    initial_guess=[0.05]\n",
    "    if var in all_studies_ASXL1_counts.keys():\n",
    "        observed_number = float(all_studies_ASXL1_counts[var])\n",
    "#         print(observed_number)\n",
    "        if observed_number >1:\n",
    "            variant_mutation_rate = mutation_rates_variants_ASXL1(var)\n",
    "            \n",
    "            MLE = scipy.optimize.minimize(prob_observed_given_expected_nonsense_ASXL1, initial_guess, args=([var],), method='Nelder-Mead')\n",
    "            s_value = MLE.x[0]\n",
    "\n",
    "            upper_error_MLE = scipy.optimize.minimize(prob_observed_given_expected_nonsense_ASXL1_upper_error, \\\n",
    "                                                      initial_guess, args=([var],), method='Nelder-Mead')\n",
    "            s_value_upper = upper_error_MLE.x[0]\n",
    "\n",
    "            lower_error_MLE = scipy.optimize.minimize(prob_observed_given_expected_nonsense_ASXL1_lower_error, \\\n",
    "                                                      initial_guess, args=([var],), method='Nelder-Mead')\n",
    "            s_value_lower = lower_error_MLE.x[0]\n",
    "\n",
    "#             print('a = ', a)\n",
    "            ASXL1_nonsense_s[var]=(s_value, observed_number, variant_mutation_rate, s_value_upper, s_value_lower) \n",
    "    a = a+1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 254,
   "metadata": {
    "code_folding": [
     0
    ]
   },
   "outputs": [],
   "source": [
    "#Exac - retrieve the frequencies in Exac for each of the variants\n",
    "import csv\n",
    "filename=\"ASXL1_Exac.csv\"\n",
    "\n",
    "with open(filename, 'r') as csvfile:\n",
    "    read_reader = csv.reader(csvfile)  #csv.reader returns a reader object which will iterate over lines in the csvfile\n",
    "    row_count=0\n",
    "    Exac_ASXL1_results={} \n",
    "    \n",
    "    for row in read_reader:\n",
    "#         print(row)\n",
    "        if row_count>1:\n",
    "            variant = row[6]\n",
    "            allele_freq = float(row[14])\n",
    "            african_freq = float(row[15])/float(row[16])\n",
    "            east_asian_freq = float(row[18])/float(row[19])\n",
    "            european_freq = float(row[21])/float(row[22])\n",
    "            finnish_freq = float(row[24])/float(row[25])\n",
    "            latino_freq = float(row[27])/float(row[28])\n",
    "            south_asian_freq = float(row[33])/float(row[34])\n",
    "            Exac_ASXL1_results[variant]=(allele_freq, african_freq, east_asian_freq, european_freq, finnish_freq, latino_freq, south_asian_freq)\n",
    "\n",
    "        row_count=row_count+1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 255,
   "metadata": {
    "code_folding": [
     0
    ]
   },
   "outputs": [],
   "source": [
    "#COSMIC - retrieve the frequencies in COSMIC for each of the variants\n",
    "import csv\n",
    "filename=\"ASXL1_COSMIC_variant_counts.csv\"\n",
    "\n",
    "with open(filename, 'r') as csvfile:\n",
    "    read_reader = csv.reader(csvfile)  #csv.reader returns a reader object which will iterate over lines in the csvfile\n",
    "    row_count=0\n",
    "    ASXL1_COSMIC_results={} \n",
    "    \n",
    "    for row in read_reader:\n",
    "#         print(row)\n",
    "        if row_count>1:\n",
    "            variant = row[1]\n",
    "            counts = int(row[2])\n",
    "            ASXL1_COSMIC_results[variant]=counts\n",
    "\n",
    "        row_count=row_count+1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 256,
   "metadata": {
    "code_folding": [
     0
    ],
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "#Create dictionary of results and save to .csv file\n",
    "ASXL1_s_dictionary={}\n",
    "for k, v in ASXL1_missense_s.items():\n",
    "#     print(k)\n",
    "    if k in Exac_ASXL1_results.keys():\n",
    "        exac_freq = round((Exac_ASXL1_results[k][0])*100, 4)\n",
    "        african_freq = round((Exac_ASXL1_results[k][1])*100, 4)\n",
    "        east_asian_freq = round((Exac_ASXL1_results[k][2])*100, 4)\n",
    "        european_freq = round((Exac_ASXL1_results[k][3])*100, 4)\n",
    "        finnish_freq = round((Exac_ASXL1_results[k][4])*100, 4)\n",
    "        latino_freq = round((Exac_ASXL1_results[k][5])*100, 4)\n",
    "        south_asian_freq = round((Exac_ASXL1_results[k][6])*100, 4)\n",
    "    if k not in Exac_ASXL1_results.keys():\n",
    "        exac_freq = '-'\n",
    "        african_freq = '-'\n",
    "        east_asian_freq = '-'\n",
    "        european_freq = '-'\n",
    "        finnish_freq = '-'\n",
    "        latino_freq = '-'\n",
    "        south_asian_freq = '-'\n",
    "    if k in ASXL1_COSMIC_results.keys():\n",
    "        COSMIC_number = ASXL1_COSMIC_results[k]\n",
    "    if k not in ASXL1_COSMIC_results.keys():\n",
    "        COSMIC_number = '-'\n",
    "    ASXL1_s_dictionary[k]=(round(v[0]*100, 2),round(v[4]*100, 2),round(v[3]*100, 2), v[1], v[2], COSMIC_number, exac_freq, \\\n",
    "                            african_freq, east_asian_freq, european_freq, finnish_freq, latino_freq, south_asian_freq)\n",
    "\n",
    "for k, v in ASXL1_nonsense_s.items():\n",
    "    if k in Exac_ASXL1_results.keys():\n",
    "        exac_freq = round((Exac_ASXL1_results[k][0])*100, 4)\n",
    "        african_freq = round((Exac_ASXL1_results[k][1])*100, 4)\n",
    "        east_asian_freq = round((Exac_ASXL1_results[k][2])*100, 4)\n",
    "        european_freq = round((Exac_ASXL1_results[k][3])*100, 4)\n",
    "        finnish_freq = round((Exac_ASXL1_results[k][4])*100, 4)\n",
    "        latino_freq = round((Exac_ASXL1_results[k][5])*100, 4)\n",
    "        south_asian_freq = round((Exac_ASXL1_results[k][6])*100, 4)\n",
    "    if k not in Exac_ASXL1_results.keys():\n",
    "        exac_freq = '-'\n",
    "        african_freq = '-'\n",
    "        east_asian_freq = '-'\n",
    "        european_freq = '-'\n",
    "        finnish_freq = '-'\n",
    "        latino_freq = '-'\n",
    "        south_asian_freq = '-'\n",
    "    if k in ASXL1_COSMIC_results.keys():\n",
    "        COSMIC_number = ASXL1_COSMIC_results[k]\n",
    "    if k not in ASXL1_COSMIC_results.keys():\n",
    "        COSMIC_number = '-'\n",
    "    ASXL1_s_dictionary[k]=(round(v[0]*100, 2),round(v[4]*100, 2),round(v[3]*100, 2), v[1], v[2], COSMIC_number, exac_freq, \\\n",
    "                            african_freq, east_asian_freq, european_freq, finnish_freq, latino_freq, south_asian_freq)\n",
    "\n",
    "ASXL1_df = pd.DataFrame.from_dict(ASXL1_s_dictionary, orient='index', columns=['s (%)', 's lower value (%)', 's upper value (%)','observed number', 'site-specific mutation rate', 'Number of times in COSMIC',\\\n",
    "                                                                                'Exac Overall freq (%)', 'Exac African freq (%)', 'Exac East Asian freq (%)',\\\n",
    "                                                                                'Exac European freq (%)', 'Exac Finnish freq (%)', 'Exac Latino freq (%)', 'Exac South Asian freq (%)'])\n",
    "ASXL1_df_sorted = ASXL1_df.sort_values(by = ['s (%)'], ascending = False)\n",
    "\n",
    "ASXL1_df_sorted.to_csv('ASXL1_s_across_whole_gene.csv')\n",
    "# ASXL1_s_dictionary"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 257,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>s (%)</th>\n",
       "      <th>s lower value (%)</th>\n",
       "      <th>s upper value (%)</th>\n",
       "      <th>observed number</th>\n",
       "      <th>site-specific mutation rate</th>\n",
       "      <th>Number of times in COSMIC</th>\n",
       "      <th>Exac Overall freq (%)</th>\n",
       "      <th>Exac African freq (%)</th>\n",
       "      <th>Exac East Asian freq (%)</th>\n",
       "      <th>Exac European freq (%)</th>\n",
       "      <th>Exac Finnish freq (%)</th>\n",
       "      <th>Exac Latino freq (%)</th>\n",
       "      <th>Exac South Asian freq (%)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>p.I1220F</th>\n",
       "      <td>26.41</td>\n",
       "      <td>18.97</td>\n",
       "      <td>196.75</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.101150e-10</td>\n",
       "      <td>1</td>\n",
       "      <td>0.0099</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0181</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.Y591*</th>\n",
       "      <td>13.38</td>\n",
       "      <td>11.27</td>\n",
       "      <td>14.77</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1.107557e-09</td>\n",
       "      <td>23</td>\n",
       "      <td>0.0025</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0046</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.C856*</th>\n",
       "      <td>12.91</td>\n",
       "      <td>10.47</td>\n",
       "      <td>14.32</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.042000e-09</td>\n",
       "      <td>1</td>\n",
       "      <td>0.0008</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0015</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.S392*</th>\n",
       "      <td>12.84</td>\n",
       "      <td>10.46</td>\n",
       "      <td>14.21</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.095600e-09</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.Q588*</th>\n",
       "      <td>12.36</td>\n",
       "      <td>9.72</td>\n",
       "      <td>13.80</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.089990e-09</td>\n",
       "      <td>11</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.Q708*</th>\n",
       "      <td>11.16</td>\n",
       "      <td>9.18</td>\n",
       "      <td>12.22</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.412750e-09</td>\n",
       "      <td>8</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.Q768*</th>\n",
       "      <td>11.16</td>\n",
       "      <td>9.18</td>\n",
       "      <td>12.22</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.412750e-09</td>\n",
       "      <td>3</td>\n",
       "      <td>0.0017</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.003</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.Q803*</th>\n",
       "      <td>10.40</td>\n",
       "      <td>8.74</td>\n",
       "      <td>11.36</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.959610e-09</td>\n",
       "      <td>2</td>\n",
       "      <td>0.0025</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0015</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0174</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.Q432*</th>\n",
       "      <td>10.08</td>\n",
       "      <td>7.88</td>\n",
       "      <td>11.16</td>\n",
       "      <td>2.0</td>\n",
       "      <td>3.213910e-09</td>\n",
       "      <td>-</td>\n",
       "      <td>0.0025</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0045</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.R417*</th>\n",
       "      <td>9.88</td>\n",
       "      <td>8.33</td>\n",
       "      <td>10.78</td>\n",
       "      <td>3.0</td>\n",
       "      <td>5.336040e-09</td>\n",
       "      <td>8</td>\n",
       "      <td>0.0074</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0116</td>\n",
       "      <td>0.009</td>\n",
       "      <td>0.0302</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.W1037*</th>\n",
       "      <td>9.38</td>\n",
       "      <td>7.79</td>\n",
       "      <td>10.20</td>\n",
       "      <td>2.0</td>\n",
       "      <td>7.144750e-09</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.R404*</th>\n",
       "      <td>9.12</td>\n",
       "      <td>8.28</td>\n",
       "      <td>9.71</td>\n",
       "      <td>7.0</td>\n",
       "      <td>1.883910e-08</td>\n",
       "      <td>6</td>\n",
       "      <td>0.0058</td>\n",
       "      <td>0.0096</td>\n",
       "      <td>0.0116</td>\n",
       "      <td>0.006</td>\n",
       "      <td>0.0151</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.W583*</th>\n",
       "      <td>9.12</td>\n",
       "      <td>6.93</td>\n",
       "      <td>10.14</td>\n",
       "      <td>2.0</td>\n",
       "      <td>5.105670e-09</td>\n",
       "      <td>3</td>\n",
       "      <td>0.0017</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.003</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.R693*</th>\n",
       "      <td>9.02</td>\n",
       "      <td>8.16</td>\n",
       "      <td>9.62</td>\n",
       "      <td>7.0</td>\n",
       "      <td>1.883910e-08</td>\n",
       "      <td>75</td>\n",
       "      <td>0.005</td>\n",
       "      <td>0.0098</td>\n",
       "      <td>0.0233</td>\n",
       "      <td>0.0046</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          s (%)  s lower value (%)  s upper value (%)  observed number  \\\n",
       "p.I1220F  26.41              18.97             196.75              3.0   \n",
       "p.Y591*   13.38              11.27              14.77              3.0   \n",
       "p.C856*   12.91              10.47              14.32              2.0   \n",
       "p.S392*   12.84              10.46              14.21              2.0   \n",
       "p.Q588*   12.36               9.72              13.80              2.0   \n",
       "p.Q708*   11.16               9.18              12.22              2.0   \n",
       "p.Q768*   11.16               9.18              12.22              2.0   \n",
       "p.Q803*   10.40               8.74              11.36              3.0   \n",
       "p.Q432*   10.08               7.88              11.16              2.0   \n",
       "p.R417*    9.88               8.33              10.78              3.0   \n",
       "p.W1037*   9.38               7.79              10.20              2.0   \n",
       "p.R404*    9.12               8.28               9.71              7.0   \n",
       "p.W583*    9.12               6.93              10.14              2.0   \n",
       "p.R693*    9.02               8.16               9.62              7.0   \n",
       "\n",
       "          site-specific mutation rate Number of times in COSMIC  \\\n",
       "p.I1220F                 3.101150e-10                         1   \n",
       "p.Y591*                  1.107557e-09                        23   \n",
       "p.C856*                  1.042000e-09                         1   \n",
       "p.S392*                  1.095600e-09                         -   \n",
       "p.Q588*                  1.089990e-09                        11   \n",
       "p.Q708*                  2.412750e-09                         8   \n",
       "p.Q768*                  2.412750e-09                         3   \n",
       "p.Q803*                  3.959610e-09                         2   \n",
       "p.Q432*                  3.213910e-09                         -   \n",
       "p.R417*                  5.336040e-09                         8   \n",
       "p.W1037*                 7.144750e-09                         1   \n",
       "p.R404*                  1.883910e-08                         6   \n",
       "p.W583*                  5.105670e-09                         3   \n",
       "p.R693*                  1.883910e-08                        75   \n",
       "\n",
       "         Exac Overall freq (%) Exac African freq (%) Exac East Asian freq (%)  \\\n",
       "p.I1220F                0.0099                     0                        0   \n",
       "p.Y591*                 0.0025                     0                        0   \n",
       "p.C856*                 0.0008                     0                        0   \n",
       "p.S392*                      -                     -                        -   \n",
       "p.Q588*                      -                     -                        -   \n",
       "p.Q708*                      -                     -                        -   \n",
       "p.Q768*                 0.0017                     0                        0   \n",
       "p.Q803*                 0.0025                     0                        0   \n",
       "p.Q432*                 0.0025                     0                        0   \n",
       "p.R417*                 0.0074                     0                   0.0116   \n",
       "p.W1037*                     -                     -                        -   \n",
       "p.R404*                 0.0058                0.0096                   0.0116   \n",
       "p.W583*                 0.0017                     0                        0   \n",
       "p.R693*                  0.005                0.0098                   0.0233   \n",
       "\n",
       "         Exac European freq (%) Exac Finnish freq (%) Exac Latino freq (%)  \\\n",
       "p.I1220F                 0.0181                     0                    0   \n",
       "p.Y591*                  0.0046                     0                    0   \n",
       "p.C856*                  0.0015                     0                    0   \n",
       "p.S392*                       -                     -                    -   \n",
       "p.Q588*                       -                     -                    -   \n",
       "p.Q708*                       -                     -                    -   \n",
       "p.Q768*                   0.003                     0                    0   \n",
       "p.Q803*                  0.0015                     0               0.0174   \n",
       "p.Q432*                  0.0045                     0                    0   \n",
       "p.R417*                   0.009                0.0302                    0   \n",
       "p.W1037*                      -                     -                    -   \n",
       "p.R404*                   0.006                0.0151                    0   \n",
       "p.W583*                   0.003                     0                    0   \n",
       "p.R693*                  0.0046                     0                    0   \n",
       "\n",
       "         Exac South Asian freq (%)  \n",
       "p.I1220F                         0  \n",
       "p.Y591*                          0  \n",
       "p.C856*                          0  \n",
       "p.S392*                          -  \n",
       "p.Q588*                          -  \n",
       "p.Q708*                          -  \n",
       "p.Q768*                          0  \n",
       "p.Q803*                          0  \n",
       "p.Q432*                          0  \n",
       "p.R417*                          0  \n",
       "p.W1037*                         -  \n",
       "p.R404*                          0  \n",
       "p.W583*                          0  \n",
       "p.R693*                          0  "
      ]
     },
     "execution_count": 257,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ASXL1_df_sorted"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 272,
   "metadata": {
    "code_folding": [
     0
    ]
   },
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAABZ8AAAHvCAYAAAA7GaHXAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDIuMi4zLCBodHRwOi8vbWF0cGxvdGxpYi5vcmcvIxREBQAAIABJREFUeJzs3XmcjeX/x/HXNcYMY5mxDVkHhVIRhVB2RSFSxC+k8v2mSIu+1bcaWr6RJBL6KqXSKkspWyFLyL6O3VjHMsbYxjDL9fvjPud8z5lzZgwGo97Px2MeZ+a6r/u6P/d9nfukz1zzuY21FhERERERERERERGRnBR0pQMQERERERERERERkb8eJZ9FREREREREREREJMcp+SwiIiIiIiIiIiIiOU7JZxERERERERERERHJcUo+i4iIiIiIiIiIiEiOU/JZRERERERERERERHKcks8iIiIiVwFjTLAxxhpjPr7SseRGxpg+xpjNxpizrutU1tXezBiz1Bhz0tX+f1c6VhERERGRvwsln0VERESuIGNMY1dSNLOvelnsG2SMGWCMaXs5Y85tjDEtgBHAeqAX8DCQYIwpDkwC8gHPutoXXaIYarnmovylGF88v4A54LovXjxHv+7GmEXGmIPGmGRjzF5jzBxjzEBjTIhX3++MMenGmEYBxilgjNlmjDlsjIn0GtsaY6ZkI97OxphPjTFrjTGp3r8UEREREfm7CL7SAYiIiIgIAF8DvwRo3wZgrU01xuQHUr22BQHRwCfAj5c8wtyrhev1EWvtcXejMaYxUBh41Vp7qa9PLZy5+BXYfYmP9Xd1L1AS2A70BAZl0u9boAOwABgCJALlgbrAv4FhwFlX395AI+BTY8zN1tqTXuO8A1QGHrDWHrqAeJ/CeV+sAXYA113AGCIiIiJXNSWfRURERHKHldbaL7PqYK1NvlzBXGVKAWneiWevdoCEyxzPVccYY4ACGZKvuc2jwBbgJeAHY8wd1toF3h2MMXVxEs8TrbUPZBzAGFMK8JyjtTbeGPME8ANOsrm3q18z4AngG2vtxAuMtyuwz/WLozEo+SwiIiJ/Qyq7ISIiInIVyFjz2RhzLZDi2vyoV5mO1Iz9jTENjTELjDFJxph4Y8x/jTEFAhyjjDFmjDFmj6t28j7Xz8Uz9CtmjBlujNnhKmlwxBiz3BjzbIZ+PYwxy4wxicaYU8aY7caYL40xxbJ5zncZY2YbY465jrPGGNPLa/u1xhiLU04jj9c1+NUYsxdnRTjAAu9r49o3nzHmFWPMRtfYR40xPxpjagSIwxhj/mGM+dNVO/qEq5RCtGv7m8DYDMfynqv8xpjXjVOTOsl1rHXGmMxW7nofO9gV5wJXCYmzxphdxpgPjTFFM9nnAWPM767rlmSM2eSar2DX9uau+B42Tq3sGOAM0M9rjMau63jcNcYKY0yPAMe6yRjzg+u9csYYE2ec8hatvPpc8Pl7jXEN0AoYD0wDjuAkozNyJ3h/CzSOtfaAtTY1Q9sknL88+KdxaoQXBsYBB3FWL18Qa+2ujMcSERER+bvRymcRERGR3CEsY5IXOGOtPZFJ/wNAd5xk3Dz+l2hNz9CvNtDetf1LoCnwOE75jt7uTsaYisAfQB5XX3eZgCeAJsaY27xWFk8CbgfGAGuBAsD1QGPgPdd4j+Ak8H4HXgNO45Q+aA0Ux0keZso4q1E/dMX0BpAE3AV8ZIypaK19yXUNHnbFWM91PQDicMpt3IOToHwDZ8VsumvsEGAWThmGz3HqRRdxXZc/jDENrbWrXH0NTmKyE7AYeAunjMP1wP3AQOB7nFXW3scCV8kU13XqBnzmGiOv69o2zeoauOQDnsNZmTsFOAXUwalt3cA1L+5fQmCMGQy8AGwAhuIkUCu7Yv03Xqt+XeMWwZnvA8Au1xj3ARNd13GI65gP4ZSmqGitdSfdSwBzgDTXOe7GmdvbXDFOz4Hzd+sBGOALa+1ZY8xXQE9jTJ8M98h21+uDxphvrLWJ2Rz/KZz37zic91x5oJ21Nsv3qYiIiIhkzVhrr3QMIiIiIn9bxqlLPDeTzd9aazu7+gXjrHT+xFr7WGZtXuO6t6UDda21y722zcRJtEVYa0+72n7GSVTXstbu9+pbFycZF22tfdO12vYI8IG1tm8W5/Uj0BAoYa1Ny+blcO9bFieJ+K21tluGbR8C/wAqW2vdydIvgc7W2uAMfR/DWZF8h7V2oVd7f2Aw0NJa+6tXewTOQws3WWubu9q6ABNwkvw9rbXpXv2D3D9ndizXtuPAPGvteT8Y0hgTBIS658mr/R84Sd37XSt3McbUx3mg4q/AvdbaM179jXX9w98Y0xyYjTOP1ay18V798gI7gTDgBmvtAVd7KM4vEm4DrrPW7jDGdMBJintiyOQcLvj8vcbYAuz2mpfawHKgl7V2bIa+v+Cskj6F895d6vqaY61NyuIYbfhf7fTPrbXdA/Rx31dTrbX3nUf8Y3Det+WstXuzu5+IiIjI1U5lN0RERERyh//iPDjP++vNHBh3oXfi2WUOEAJUAHAllFvhrKw9a4wp7v7CSQLvBFq69k3CSb7VM8ZUyOK4x4BCQGvX6uHz8YArvnHesbji+QlndXaz8xzT2//hrAxenWHsYJxyDY1cyVZw6vZa4HnvxDNAxp+zcAy4yRhT/XwDtdame/2CII8xJsIV6xxXl7pe3bu6Xl/0Tjy7xgm04uQz78Szy21AGeBjd+LZtf8Z4F2c/39wJ5GPuV5bG2MKZXEaF3z+AMaYO3FWSo/3imcFsA7nwYMZ3Qc8DWzEWV39Cs77Js4Y0y9Af7cj/O8vB6ZdSKwiIiIi4kvJZxEREZHcYau19tcMX+tzYNwdAdrcpQTctZer4ZQ0+AdwOMBXZaAkeB56+CxQE4g1xqw3xowwxjTJcIw3gb04K0kPGWMmGmMeNcYUzEbM17te5waIxV3KoWQ2xslMNeDGAGMfxikPEQy46ylfB+wNkKQ9H0/jlKNYb4zZZowZa4xpk92kvDGmszFmGU7pkqOuON2lPYp4db0OpwTG2mzGtSVAW0XX64YA29zvx0qu1znAVzjlRo4YYxYZY6KNMdUy7HdR5+8a/yywxjh1vq81Ts3zWTi/BLnBu7O19qy1doS1tg5O+ZU7cVa65wGGGWMCPYgwDKcsSAKwz9UvIpvxiYiIiEgmVPNZRERE5K8tq5IXJsPreJy60IF4yhVYa0caYybj1FS+E3gQ6GOMmWCt/T9Xn82uJGRznFXKjYCPgQHGmDuttTuzEVdX4FAmfbZl0p4dQcBqoH8WfRK8YrmoOnXW2knGmN9x6l03wlnV/hgwzxjT0rtmc0bGmAdxak4vAfriJPSTcVaG/4zvYpLzXWEeqARFtsdwrabu6npwYCvgDpx606+6ajGPdvW7mPMvDHTEOd81mXTrCTyfSYxJwAKcB0H+DvyCk8z+PkPXQTjJ+07ACVe/YcAjWV4EEREREcmSks8iIiIiV6+cenjHNtdYeb1rIGd5YGv34ZQK+a+rDu4EnETkUPfD+lylGn52fWGMaQtMBfrhrIbNzFbX6+HsxnOetgIlgN8yKUfhbTNOWYni51j9nOU4rgfXfQF84VrxOwTngX/3ApOz2PVhnCRxE9eqcwCMMTdmEmtz4CZgZVbxZMH9wL5AJTLcK4x9VtNba9fhlMB4x1XC5U+cZO5orz4Xev6dcepP/yvjcV2eAR42xryUVRLbZYnrtYx3o6vu+lPA99ba71xtnwKPGGO+t9b+co5xRURERCQTKrshIiIicpVyPcgvmf+ViLjQcQ7ilDB4wBhzW8btxlHC9X2YMSZ/hv1TcZKPuGNx1SXOaKV3nyx8i1Nm4XVjTL4A8UQYY0LOMUZWPsdJQAZMgBtjvEt6TMBZDfxOxjIRGX4+6XotmqFPsDEm3LvNlfBeHah/AGk4dYg9/253HfeVAH2/cr2+nfH6nEeJi2U4ZSceNcZEeu0fgrO6OB3XQ/mMMUUzjmutTQBigYLGmJAcOP9HgXhgqLV2YsYvYBwQiZPExhhT1RhTKZOx3A8I3Oh1XgWBT3FKmfT26vsMzirz/6r8hoiIiMiF08pnERERkavbEuAuY8wLwB4gzb168zz9A1gILDTGfA6swvm3YkWcpN0nOHWcbwB+dZXd2IBTg/gG4AmcVbOLXOPNMcYcxil5sAcnyfgITvIys9IeAFhrdxljngLGABuNMV8Cu3FWK98MtAOq4CQHL8R7OCuEhxljmgPzcEotlMcpEXICpzQEwDdAB1fsVYwxP+E8QK+Kq28NV78/cVY/v+JK1J/ifw9r3G2M+REn4XoIp2byEzilPX4+R6wTXec7xxjzBRAKtAf8kvLW2j+MMUNxVhQvN8Z8DxzAmcMHjDG3WGtPZtwvwxipxpg+OGUplhljxrrOpTNQB3jdWutegdwTeMr1XtgGpAJNXNflK2vtWdcvIS7o/F2ru+vgPPwws/IxU4GPcJLUk4FbgAnGmLnAfJz3SEGgHk75jmPAG177DwWigPu9V7Zba48ZY3qRefmNKsaYQL8AACdRftoY0winFAlALddrX2PMcSDdWvufzM5dRERE5K9CyWcRERGRq9s/gZE4K2EL4ayUPe/ksyvhWwt4EWiL8+C9JJzE8RScJCjALpza0I1xkqChOAm+j4DBXqUhPgQecMVXFOchhyuB3tbaedmIZ6wxJgZnte0TQDjOCthNwL9xVqpeEFdS9G6cUgv/Bwx0bdqPk8wf79XXGmM6AU/iJFujca7xDrzqBltrdxpjHsOpeTwayIuTsH8SGIGTkG0BFADicBKlb1trD5wj1i9dq3OfxkmUJuCsPH6VAPWwrbXPG2NWuY77L5xV23uAaTir5LNzfSYbY1rgvKf+hfP/DDFAT2vtp15d5+D8MqANcA1O8nknTvJ7pKvPiYs4/0ddr5OyiDXeVcv5bmNMaZyHVP4L55cLj+KsigbnPToOeMedPDfGtAR6AV9ba/2OYa2dbowZB/QMUH7jenyT2N7G4DwcsgXOe9Wbu854GqDks4iIiPzlmXOXubtyjDGDgVtxVpYUx/lH3C6c/wEa6aodl3Gf+jj/UK6HsyJkG84/ND/IYsWEiIiIiIiIiIiIiOSg3J58PouzQmYjzsqOAjhJ5VtxVqbUs9bu8erfDvgBZ1XHtzgrQ9oAVYGJ1toHLusJiIiIiIiIiIiIiPxN5fbkcz7vp3p7tb8FvAyMttb2drUVxlnlHA40sNYud4+B8yeBtwMPWWu/uVzxi4iIiIiIiIiIiPxdBZ27y5UTKPHs4q5jeJ1XW0ech9B84048e43hfhjIEzkepIiIiIiIiIiIiIj4ydXJ5yy0cb2u9Wpr6nqdEaD/fJwH5tQ3xoReysBERERERERERERExHlyda5njHkeKIhTUuNWoCFO4nmQV7eqrtctGfe31qYaY3YC1YFKOE/rDmj06NEW4IknnjA5EryIiIiIiIiIiIjI39DVsvL5eSAa6IeTeJ4BtLTWHvbqE+56PZbJGO72iEAbjTG9jDGech3Hjh2zq1atsikpKXb58uUWsMuXL7dnzpyxa9assUePHrXbtm2ze/futXFxcXbz5s32+PHjdtWqVTY1NdVvn3Xr1tmEhAS7fft2u2fPHnvgwAG7adMme+LECbtixQqblpbms8/p06ft+vXr7ZEjR+yOHTvs7t277cGDB21MTIw9efKkXb58uU1PT/fZJykpyW7YsMEePnzY7ty50+7atcseOnTIbty40Z46dcrTd9myZZ59Tp06ZTdu3GgPHTpkd+3aZXfu3GkPHz5sN2zYYJOSknzGdx/v5MmTNiYmxh48eNDu3r3b7tixwx45csSuX7/enj592meftLQ0u2LFCnvixAm7adMme+DAAbtnzx67fft2m5CQYNetW2fPnDnjs09qaqpdtWqVPX78uN28ebONi4uze/futdu2bbNHjx61a9as8dsnJSXFrlq1yh47dsxu2bLF7t+/3+7bt89u3brVJiYm2tWrV9uUlBSfcz979qxnLrdu3Wr37dtn9+/fb7ds2WK959+9z7Jly+yZM2fs2rVrPXOZcf5XrlzpN5fJycl23bp1PnOZcf4zzqV7/uPj4z1zefDgQbtx40bP/FtrM53/2NhYGxsbaw8fPuw3/96vJ0+e9Jv/+Ph4u379epuUlORzvdLT0+2KFSuynP/k5GSf91laWppduXKlPX78eMD5X7t2babzf+zYMbt582a7f/9+v/k/e/as3/yvXr3aJiYm2i1btth9+/bZffv22S1btvjMv/c+3vPvvpf3799vN2/e7Jn/QPey9/x738ve8+993ZKTk7O8lzPO/7Jly2xSUpLf/LvvZff8Z5xL972ccf4D3cvu9457/g8ePOg3/xnvZff8e9/L3vO/bt06n/l33//u+fe+lzPOv/c9ltW9HGj+3fu45z+re1mf5fos12e5Psv1Wa7Pcn2W67Ncn+X6LM+tn+XD5qy3Q//cb99btEOf5fos12e5Pstz/LOcKyhXP3AwI2NMSaA+zornQsC91tqVrm1bcGpAX2et3RZg3z9wHjp4u7V2SWbH8Fr5nPMnICIiIiIiIiKSwZBNqZ7v+1e7Kv5IXUSuLleswsPVsvIZAGvtQWvtZKAlUAz43Guze2VzuN+OjsIZ+omIiIiIiIiIiIjIJXJVJZ/drLW7gI1AdWNMcVfzZtdrlYz9jTHBQEUgFdhxWYIUERERERERERER+Ru7KpPPLqVdr2mu1zmu17sD9L0TCAP+sNaeudSBiYiIiIiIiIiIiPzd5drkszGmmjGmVID2IGPMW0AkTjL5qGvTRCAe6GyMudWrfz7gTdePoy9x2CIiIiIiIiIiIiIC5OYq9ncDQ4wx84HtwBGgJNAIqAQcAB53d7bWHjfGPI6ThJ5njPkGSADaAlVd7d9e1jMQERERERERERER+ZvKzcnnX4H/Ag2AGkAEcArYAnwBjLDWJnjvYK2dYoxpBPwbuB/IB2wDnnX1t5cvfBEREREREREREZG/r1ybfLbWrgeevID9FgGtcz4iEREREREREREREcmuXFvzWURERERERERERESuXko+i4iIiIiIiIiIiEiOU/JZRERERERERERERHKcks8iIiIiIiIiIiIikuOUfBYRERERERERERGRHKfks4iIiIiIiIiIiIjkOCWfRURERERERERERCTHKfksIiIiIiIiIiIiIjlOyWcRERERERERERERyXFKPkuuMmDAAIwxzJs370qHIiIiIiIiIiIiIhdByWcRERERERERERERyXFKPkuu8tRTTxETE0OdOnWudCgiIiIiIiIiIiJyEYKvdAAi3ooXL07x4sWvdBgiIiIiIiIiIiJykbTyWS5abGwsxhh69OjB9u3b6dixI8WKFaNQoUK0bNmS9evXA3D48GF69erFNddcQ758+bjtttuYO3euz1iZ1XxesGABbdq0oWzZsoSGhlKqVCnq1avHwIEDffodPHiQ559/nqpVq1KgQAEiIiKoWrUqPXr0YMeOHX6xz5w5k9atW1O8eHFCQ0OpXLky/fv3JzEx0a9vVFQUUVFRJCUl0b9/f8qXL09oaCjXXnstgwcPxlrrt8+PP/5Is2bNuOaaawgNDaV06dI0atSIUaNG+fVNSEjgpZde4vrrryd//vyEh4fTrFkzZs2adc45EBERERERERERyW208llyTGxsLHXr1uX666+nR48exMbGMnnyZBo3bszixYu5++67KVy4MJ06dSIhIYFvvvmGVq1asWXLFsqXL5/puDNmzOCee+6hcOHCtG3bljJlypCQkEBMTAyjRo0iOjoagKSkJBo0aMD27dtp0aIFbdq0wVrLrl27mDp1Kh07dqRSpUqecV9//XWio6MpWrQo9957L5GRkaxdu5Z3332XX375hcWLF1O4cGGfWFJSUmjZsiX79++nVatWBAcHM2XKFF588UWSk5M9sQD897//5R//+AelSpWiTZs2FC9enEOHDrF27Vo+/fRTevfu7em7a9cuGjduTGxsLHfccQd33303p06dYtq0adx999189NFHPP744zk1VSIiIiIiIiIiIpecks+SY37//XfefPNN/v3vf3va3njjDV577TXq1q3Lgw8+yKhRowgKchbct2jRgm7dujFs2DCGDRuW6bhjx44lPT2defPmUaNGDZ9t8fHxnu9/++03tm/fTr9+/fzGO3v2LGfOnPH8PHfuXKKjo7n99tv55ZdfiIiI8Gz77LPPeOSRR4iOjvYbZ//+/dSoUYPZs2eTP39+AKKjo6lSpQrDhg3j5ZdfJm/evAB89NFHhISEsGbNGiIjIzONG6B79+7s2rWLr7/+ms6dO3vaExMTady4MX379qVt27aULFky0+skIiIiIiIiIiKSmyj5fIl0WZJypUPItq/q5c2RcaKionjxxRd92rp3785rr73GmTNnGDJkiCfxDNClSxd69uzJ6tWrszW+O9nrLVB96ED9QkJCCAkJ8fw8YsQIwElseyeeAXr06MHw4cOZMGFCwKT4iBEjfI4RGRlJu3bt+Pzzz9m8eTM33nijZ1twcLAnGZ1Z3GvWrOH333+nY8eOPolngIiICAYOHMh9993HDz/84LNaWkREREREREREJDdT8llyTM2aNcmTJ49PW+nSpQGoUqUKhQoV8tmWJ08eSpYsyd69e7Mct2vXrkyaNIm6devSqVMnmjRpQoMGDShbtqxPv0aNGlGmTBkGDRrEypUrad26NQ0aNAgY1+LFi8mbNy/ff/8933//vd8xz549y+HDhzly5AjFihXztIeHh3Pttdf69S9XrhwAR48e9Yn7ueeeo3r16nTq1IlGjRrRoEEDSpQo4RcLwLFjxxgwYIDf2IcPHwYgJiYmq8skIiIiIiIiIiKSqyj5LDkmPDzcry04ODjTbe7tKSlZrxLv0KED06ZNY+jQoYwbN46PPvoIgNq1a/P222/TokULAAoXLsySJUuIjo7mxx9/ZObMmYCzyrh379688sornlXIR44cITU11e+BhRmdPHnSJ/mccZV0xvNMS0vztD377LMUL16cUaNGMWLECN5//32MMTRq1IghQ4Zw6623emIBmD17NrNnz84yFhERERERERERkauFks+XSE6VshDHPffcwz333MOpU6dYunQp06ZNY/To0dx7772sWrWKG264AYCyZcvyySefYK1l48aNzJkzhw8//JDXX3+d9PR03njjDcBJhqenp5OQkHBJ4+7WrRvdunUjMTGRP/74g8mTJzNu3DjuuusuYmJiiIyM9CTmhw8fTt++fS9pPCIiIiIiIiIiIpdL0Lm7iOQeBQoUoGnTprz33nu8/PLLnD17lunTp/v1M8ZQvXp1+vTp41lNPGXKFM/2evXqcfToUTZs2HBZ4o6IiKB169aMHTuWHj16kJCQwIIFCzyxAJ6fRURERERERERE/gqUfJZc77fffuP06dN+7QcPHgQgLCwMgPXr1xMbG3vOfgDPPPMMAI8//jj79+/32+fUqVMsWbLkouKeMWMGqampfu2HDh3yiefWW2/ljjvuYNKkSYwbNy7gWOvWrfPsJyIiIiIiIiIicjVQ2Q3J9Z577jliY2Np3LgxUVFRhISEsGLFCubMmUOFChXo3LkzAL/++ivPPvss9evXp1q1akRGRrJ3716mTp1KUFAQ/fv394zZrFkzBg0axEsvvcR1111H69atqVixIidPnmTXrl38/vvvNGzYkBkzZlxw3J07dyZfvnw0bNiQqKgorLUsWLCAZcuWUbt2bZo3b+7p+9VXX9G0aVMeffRRRowYQd26dYmIiGDv3r2sXbuW9evXs3jxYiIjIy/8QoqIiIiIiIiIiFxGSj5Lrvfyyy8zefJkli9fzq+//kpQUBDly5fn5Zdfpl+/fhQpUgSAu+66i379+jF//nymTp3K8ePHueaaa2jRooUnKe3tX//6Fw0aNGDEiBEsXLiQqVOnEh4eTpkyZejVqxddunS5qLgHDRrEzJkzWblyJb/88gv58uWjQoUKDB48mCeeeMLz8ENwalWvWLGCDz74gB9++IEJEyaQlpZGqVKluOGGG+jTpw833XTTRcUjIiIiIiIiIiJyORlr7ZWOIVcZPXq0BXjiiSeudCgiIiIiIiIi8jcwZNP/Sjb2r6Z1giKS48yVOrBqPouIiIiIiIiIiIhIjlPyWURERERERERERERynJLPIiIiIiIiIiIiIpLjlHwWERERERERERERkRyn5LOIiIiIiIiIiIiI5Dgln0VEREREREREREQkxyn5LCIiIiIiIiIiIiI5LvhKByAiIiIiIiIi8ne1N8ly6Iwl3UKQcX4uG2audFgiIjlCyWcRERERERERkcts/bF0Ju1NZ9MJ69P+wtpUqhUydCgbxI3h+oN1Ebm66VNMREREREREROQymnsonbdj0vwSz26bTljejklj3qH0yxyZiEjOUvJZREREREREROQyWX8snY93pBE47fw/Fhi7I431x5SAFpGrl5LPIiIiIiIiIiKXyaS96edMPLtZYPJeJZ9F5Oql5LOIiIiIiIiIyGWwN8lmWmojMzEnLHuTzm8fEZHcQslnEREREREREZHL4EJLaKj0hohcrZR8FhERERERERG5DE6nXd79RESuNCWfRUREREREREQug/x5Lu9+IiJXmpLPIiIiIiIiIiKXwY3hF5aGudD9RESuNH16yUWLjY3FGEOPHj2IjY2lc+fOFC9enHz58nHrrbcybdo0v33OnDnDoEGDuPnmmwkLC6Nw4cLccccdfPfddzky/tmzZxkxYgS1atWiSJEihIWFERUVRbt27fj111/9+m/atIkePXpQrlw5QkNDKVmyJF26dGHz5s1+fXv06IExhtjYWD766CNuuukm8uXLR8mSJenVqxfHjh3z22ft2rU89NBDREVFERoaSokSJahVqxb9+vUjJSXFp29qaiqjRo2iXr16FC5cmLCwMG655RZGjhxJerrqfImIiIiIiFytyoYZqhUy57XP9YUMZcPObx8Rkdwi+EoHIH8du3btok6dOlSqVImHH36YhIQEvv32W0/Ct0mTJoCTGL7rrrv4/fffqVatGk9+vVX8AAAgAElEQVQ++SRJSUlMnDiRTp06sXr1av7zn/9c8PjgJIi//vprbrzxRrp160b+/PnZv38/CxcuZMaMGTRv3tzTd8aMGXTo0IGUlBTatGnDtddey969e5k0aRI///wzc+fOpVatWn7xvPDCC8ycOZM2bdrQsmVL5s6dy9ixY9m2bRtz5szx9Fu7di1169bFGEPbtm2pWLEix48fZ9u2bYwaNYo333yTvHnzAnhimDlzJlWrVqVLly7ky5ePuXPn0qdPH5YuXcoXX3yRY3MmIiIiIiIil1eHskG8HZOGzUZfA7Qvq3WDInL1UvJZcsy8efMYMGAA0dHRnrYuXbpw9913M2TIEE9yeOjQofz++++0atWKH3/8keBg520YHR1NnTp1ePvtt7n33nupX7/+BY1/7NgxvvnmG2rXrs3SpUvJk8e3ONaRI0c83x89epSHHnqIsLAw5s+fzw033ODZtmHDBurWrctjjz3GypUr/c53yZIlrFu3jvLlywPOiuWmTZsyd+5c/vzzT+rUqQPA+PHjSU5OZsqUKbRr185njKNHjxIWFub5+a233mLmzJk89dRTvP/++57Y09LS6NWrF+PGjaNjx45+44iIiIiIiMjV4cbwIB6rBB/vyDoBbYDHK+VRyQ0Ruaop+XyJmPtGXOkQss1O6Zsj41SoUIFXXnnFp+2uu+6ifPny/Pnnn562cePGYYzhvffe8ySeASIjI3n11Vd57LHH+Pjjj/2Sz9kd3xiDtZbQ0FCCgvz/I12sWDHP959//jmJiYmMHDnSJ/EMUL16dR5//HHef/99Nm7c6Lf9tdde8ySeAYKDg3nkkUdYsGCBT/LZLX/+/H6xFClSxPN9eno6I0eOpFSpUgwbNswnaZ4nTx6GDh3Kp59+yoQJE5R8FhERERERuYo1iQyiRChM3ptOzAn/FPT1hQztywYp8SwiVz0lnyXH1KxZ02+VMUC5cuVYvHgxACdOnGDbtm2UKVOGatWq+fVt2rQpAKtWrbqg8QEKFy5MmzZt+Omnn6hZsyb3338/d9xxB3Xr1vVZZQx49luzZg0DBgzwG3vLli0AxMTE+CWfb7311oCxgLOi2a1Tp04MHz6c++67j44dO9K8eXMaNGhA5cqV/Y515MgRrrvuOt58802/scFJYMfExATcJiIiIiIiIlePG8Od5PLeJMvwramkWwgy8PR1warxLCJ/GUo+S46JiIgI2B4cHOx5UJ77YXzXXHNNwL7u9sTExAsa3+3bb79l8ODBfPXVV54yHfny5aNjx468++67lCxZEvhfCY6xY8dmeW4nT57MVjzuldxpaWmetjp16rBgwQLeeustJk6c6KnZXLVqVaKjo3nooYd8Ytm6dSsDBw48r1hERERERETk6lQ2zBAZanx+FhH5q1Dy+RLJqVIWfzXh4eEAHDhwIOD2uLg4n34XKn/+/AwYMIABAwawZ88e5s+fz2effcaXX35JbGwsCxYs8DnOmjVruPnmmy/qmFm5/fbbmTZtGmfOnGHFihXMmDGDDz74gC5dulCiRAmaN2/uiaV9+/ZMmjTpksUiIiIiIiIiIiJyOah4kFxWhQoVonLlyuzbt4+tW7f6bZ87dy4AtWrVyrFjlitXjq5duzJz5kyuu+46Fi5c6FllXK9ePQBPMvpSCw0NpX79+rz++uuMGOHUBZ86dSoA1apVIyIigiVLlpCSknJZ4hEREREREREREblUlHyWy65nz55Ya+nfv79PeYr4+HjeeOMNT58LdfjwYZYuXerXfurUKU6cOEFwcDAhISEAPPLII0RERDBw4ECfhxa6paenM2/evAuOBZzEtrvciLeDBw8CeOpQBwcH06dPH+Li4ujbty+nT5/22ycuLo6NGzdeVDwiIiIiIiIiIiKXg8puyGX3/PPPM336dKZOnUqNGjVo3bo1SUlJfP/99xw6dIgXXniBhg0bXvD4+/bto169elx//fXUqlWLcuXKcfz4caZNm8aBAwfo27cvhQoVAqBYsWJMnDiR9u3bU69ePZo1a0b16tUJCgpi9+7dLF68mCNHjpCcnHzB8QwdOpRZs2bRuHFjKlWqRMGCBdmwYQPTp0+nSJEi9OrVy9P31VdfZc2aNYwZM4affvqJpk2bUqZMGQ4dOsTWrVtZtGgRb731lt/DD0VERERERERERHKbXJt8NsYUA9oD9wA3AWWAs8A64FPgU2ttulf/KGBnFkN+a63tfKnilewLCQlh9uzZvPfee3z11Vd88MEHBAcHU6NGDd5//33PA/guVFRUFAMHDmTevHnMnTuX+Ph4ihYtStWqVRk0aBCdO/u+DZo1a8batWt59913mTlzJgsWLCAkJITSpUvTtGlT7r///ouKp3fv3hQpUoSlS5eyaNEiUlNTKVu2LL179+a5556jQoUKnr558+ZlypQpfPnll3z22WdMmzaNkydPUqJECSpWrMgbb7xB165dLyoeERERERERERGRy8FYa690DAEZY/4JjAbigLnAbqAk0AEIB34AHrCuE/BKPq8BpgQYcr21duK5jjt69GgL8MQTT1z0OYiIiIiIiIiInMuQTame7/tXy7XrBEXk6mWu1IFz8yfaFqAt8HOGFc4vA38C9+Mkon/IsN9qa+2AyxWkiIiIiIiIiIiIiPjLtQ8ctNbOsdb+5J14drUfAMa4fmx82QMTERERERERERERkXPKzSufs5Liek0NsK20MeYfQDHgCLDYWrv2skUmIiIiIiIiIiIiIldf8tkYEwx0c/04I0CXFq4v733mAd2ttbsvbXQiIiIiIiIiIiIiApBnwIABVzqG8zJw4MB3gPuAX6y1r3u1hwDJwLPAi8BgYDoQBTQB2gwcOHDcgAEDUvwGBYwxvQYOHPjRPffcUxqgSpUqxMTEULx4cVatWkXp0qVZvnw5xYoVY8OGDYSGhrJv3z6SkpI4deoU+/fvJyQkhI0bN1KiRAlWrlzps09MTAx58+Zl//79nDp1itOnT7N3715CQ0NZv349JUuW9NmnSJEibN68mTx58nDgwAFOnDhBcnIye/bsIX/+/Kxdu5ZSpUqxYsUKzz4RERFs2bIFYwyHDh3i+PHjnD17ll27dhEWFsbatWspXbo0y5Yto0yZMixfvpzw8HC2bduGtZb4+HiOHj1KamoqsbGxFCxYkDVr1njGdx+vcOHCbN++nfT0dBISEjhy5Ajp6ens2LGDQoUK+exTsmRJVq1aRaFChdi5cyepqakcPXqU+Ph4ALZt20ZERASrV6/27BMZGcmaNWsoUKAAsbGxpKSkcOzYMQ4dOoQxhi1btlCkSBGffUqUKMHatWsJCwtj9+7dnDlzhuPHj3Pw4EHy5MnDpk2bKFasGCtXrvSce/HixVm/fj2hoaHs3buX5ORkTp48SVxcHMHBwZ75d++zbNkyihcvzsaNGz1zmXH+N2zYQGRkpM9cFi1alJiYGIKDgz1zmXH+M86le/6DgoI4ePAgx48f58yZM+zevdsz/9dcc02m83/48GESExNJSUnxm//ly5d7XgsXLuw3/2lpaezYsYOCBQuyevVqz/UqVaoUK1euzHL+Cxcu7JmXZcuWUapUKVavXk2BAgUCzv/WrVsznf+wsDB27drF2bNn/ea/aNGiPvdliRIlWLduHfny5WP37t0kJydz4sQJDhw44DP/3vt4z7/7Xj558iT79+8nb968xMTEBLyXveff+172nv8VK1Z4rlvRokXZtGlTpvfyunXrfOZ/2bJlAefffS+75z/jXLrvZcBn/gPdy+73jnv+09PTOXLkiM/8Z7yX3fPvfS97z//27dsJDw/3mcuSJUt65t/7Xj58+LDP/K9atcpzj0VGRmZ6LwcFBfnNv3sf9/zv2bMn03tZn+X6LNdnuT7L9Vmuz3J9luuzXJ/l+izPrZ/lqzdtA5tO0MkjlEpL1Ge5Psv1Wa7P8hz9LL/mmmsGXkAaNkcYa+2VOvZ5M8b0BYYDm4AG1tqEbOwTDCwE6gL9rLXDs+o/evRoC/DEE09cfMAiIiIiIiIiIucwZNP/qor2r3bV/ZG6iOR+5kodONc+cDAjY8yTOInnjUCT7CSeAay1qcDHrh/vvEThiYiIiIiIiIiIiIiXqyL5bIzpB4wE1uMkng+c5xCHXa8FcjQwEREREREREREREQko1yefjTH/AoYBq3ESz4cuYJh6rtcdORaYiIiIiIiIiIiIiGQqVyefjTGvAoOAFUAza218Fn3rGmNCArQ3BZ5x/fjlJQlURERERERERERERHzk2ir2xpjuwOtAGrAA6GuMX23sWGvtZ67vBwPVjTHzgL2utpuBpq7vX7XW/nEpYxYRERERERERERERR65NPgMVXa95gH6Z9Pkd+Mz1/RdAe+A2oBWQFzgIfAeMtNYuuGSRioiIiIiIiIiIiIiPXJt8ttYOAAacR/9PgE8uVTwiIiIiIiIiIiIikn25uuazyN9Njx49MMYQGxt7pUPJ9ebNm4cxhgEDBlzpULKtcePGBCgfJCIiIiIiIiLyl6Tks1yU2NhYjDH06NHjSofiJyUlhcmTJ/Poo49y4403UrhwYcLCwrjpppt47bXXOHHiRKb77t27l549e1K6dGlCQ0OJioqiX79+HD169DKegUjOyM33qYiIiIiIiIj8deXashsiF2v79u106NCBAgUK0KRJE+655x5OnjzJzJkzeeONN/j2229ZtGgRxYsX99uvfv36HDp0iHbt2lGtWjX+/PNPhg8fzowZM1i0aBHFihW7QmclV7PPP/+cpKSkKx2GiIiIiIiIiMhloeSz/GUVKlSIDz/8kO7du1OgQAFP+9mzZ+nQoQM///wzAwcO5IMPPvDZr3fv3hw6dIgRI0bQp08fT/uzzz7LsGHD+Pe//82YMWMu23nIX0f58uWvdAgiIiIiIiIiIpeNym7IBRswYAAVK1YEYPz48RhjPF+fffaZp196ejpjxozhtttuo2DBghQoUIDbbruN0aNHk56e7jeuMYbGjRuzf/9+Hn74YSIjI8mfPz+1a9fmq6++ynZ8ZcqUoXfv3j6JZ4CQkBBefvllwKkb7G3Hjh3MmjWLqKgonnzySZ9tAwcOpECBAnzxxRecOnUq23GsWLGC+++/n8jISEJDQ6lQoQK9e/cmLi4u033S09N57733qFatGvny5aNs2bI888wzHD9+3K/v2rVreeihh4iKiiI0NJQSJUpQq1Yt+vXrR0pKik/f1NRURo0aRb169TxlSG655RZGjhzpNxfepRq2bNlCp06diIyMJCgoiHnz5lGtWjVCQkKIj48PeA6DBg3CGMOHH37o0753716eeuopKlWqRGhoKMWKFaNt27YsW7Ys4DgHDx7k0UcfpWTJkuTPn5+aNWsyfvz4TK9dIG+//TbGGEaMGBFw+/79+8mTJw+33XabT9vrr79OgwYNKFWqFCEhIZQuXZouXboQExPjN8a5rhcErvl89uxZRo4cSevWralQoQKhoaEULVqU5s2bM3369IDxRkVFERUVRVJSEv3796d8+fKEhoZy7bXXMnjwYKy1nr7ZvU9FRERERERERHKaVj7LBWvcuDGJiYkMHz6cGjVqcN9993m21axZ0/P9ww8/zFdffUW5cuV47LHHMMYwefJkevfuzcKFC5kwYYLf2EePHqV+/fpERETwyCOPkJiYyHfffUfXrl3Zt28f/fv3v6jY8+bNC0BwsO8tMGfOHABatmxJUJDv72YKFSpEgwYNmDVrFkuWLKFZs2bnPM60adO4//77sdbSsWNHKlSowIoVKxg9ejRTp05l0aJFREVF+e33zDPPMH/+fB588EHatWvHzJkzef/991mwYAELFy4kX758gJN4rlu3LsYY2rZtS8WKFTl+/Djbtm1j1KhRvPnmm55zTUlJoU2bNsycOZOqVavSpUsX8uXLx9y5c+nTpw9Lly7liy++8Itl+/bt1K1blypVqtC1a1dOnz5N4cKF6d69Oy+//DJff/21zwpxt88//5yQkBA6d+7saVu5ciUtW7YkISGBu+66iw4dOhAfH8+UKVNo2LAhkydPpnXr1p7+R44coX79+uzYsYOGDRvSsGFD4uLi+Oc//0nLli3Pef3dunXrxiuvvML48ePp27ev3/Yvv/yS9PR0unfv7mmbP38+gwYNokmTJtx///0ULFiQrVu3MnHiRH788UcWLVpEjRo1sn29MpOQkMDTTz9N/fr1adGiBSVKlCAuLo6ffvqJ1q1bM3bsWB577DG//VJSUmjZsiX79++nVatWBAcHM2XKFF588UWSk5OJjo4Gsn+fioiIiIiIiIjkOGutvry+Ro0aZUeNGmUle3bu3GkB271794Dbv/rqKwvYW265xZ44ccLTfvLkSVu7dm0L2AkTJvjsA1jAPvDAAzYtLc3TvmPHDlukSBGbN29eu3379ouK+5///KcF7IsvvujT/vzzz1vAvvvuuwH3e/LJJy2QrffIiRMnbLFixWxQUJCdP3++z7ZBgwZZwLZo0cKnvXv37hawxYoVs7GxsZ72tLQ026FDBwvY119/3dP+7LPPWsBOmTLF7/gJCQk+1y86OtoC9qmnnrKpqame9tTUVNuzZ0+/cdxzC9iXXnrJb/y9e/faoKAgW7t2bb9tf/75pwVshw4dPG0pKSm2cuXKNjQ01M6bN8+n/759+2zp0qVtqVKlbHJysqf98ccft4Dt16+fT/9ly5bZ4OBgC9jo6Gi/4wfSsmVLC9h169b5bbvhhhts3rx5bXx8vKft4MGD9vjx4359V69ebQsUKGDvvvtun/ZzXS9rrW3UqJF1Pnb/Jzk52e7Zs8evb2Jioq1evbotUqSITUpK8tlWoUIFC9hWrVr5bDt48KANDw+34eHh9uzZs36xZXafioiIiIjIlfVOTIrnS0TkErhiuVatfL5UajxwpSPIvjXfX7Khx40bBzglGAoWLOhpL1CgAIMHD6Z58+Z8/PHHdOnSxWe/PHnyMHjwYJ/VxxUrVqRv374MHDiQL774wrOy83z9+OOPfPTRR5QtW5YXXnjBZ9uxY8cACA8PD7ivuz0xMfGcx5k6dSpHjhzhoYce4o477vDZ9txzzzFmzBhmz57N7t27/WoBP/3001SoUMHzc1BQEEOGDGHKlCmMGzeOV1991ad//vz5/Y5fpEgRz/fp6emMHDmSUqVKMWzYMPLkyePZlidPHoYOHcqnn37KhAkTaNeunc84JUuWDHity5QpQ7NmzZg9ezYbNmygevXqnm3ushjeK4l//vlntm/fzvPPP0+jRo18xipdujQvvPAC/fr147fffqN169akpKQwYcIEChUqxIABA3z633rrrXTt2vW8ym90796dWbNmMX78eIYMGeJpX758ORs3bqR9+/Y+D5KMjIwMOE6NGjVo2rQps2bNIiUlxbOy3C2z65WZ0NBQypYt69ceHh5Oz549ee6551i2bBl33nmnX58RI0b4zH1kZCTt2rXj888/Z/Pmzdx4443ZjkNEREREREREJKcp+SyX1MqVKwkKCqJx48Z+2xo1akSePHlYtWqV37by5ct76tR6a9y4MQMHDgy4T3b88ccfdOnShQIFCvDDDz/4JGizw7pq6Was2xvIypUrAWjatKnftuDgYO68805iY2NZtWqVX/I5Y3IWoFKlSpQrV47Y2FgSExOJiIigU6dODB8+nPvuu4+OHTvSvHlzGjRoQOXKlX323bJlC0eOHOG6667jzTffDBhv/vz5A9YyrlGjBqGhoQH36dGjB7Nnz2b8+PG88847gFPD+JtvvqFEiRI+JTQWL14MwK5du/ySyQBbt24FICYmhtatW7Np0yaSkpK44447Av4yoHHjxueVfG7fvj3h4eF8+eWXDBo0yJOAd4/Ro0cPv31+/vlnxowZw/Lly4mPjyc1NdVne3x8PNdcc41PW1bXKzMbNmxgyJAhzJ8/n7i4OJKTk32279u3z2+f8PBwrr32Wr/2cuXKAU7pGhERERERERGRK0nJZ7mkjh07RtGiRQkJCfHbFhwcTPHixTl06JDftpIlSwYcr1SpUp5xz9fixYtp1aoVQUFBTJ8+nTp16vj1cSc5Mxvf/cC/zFZGe3OPkTE56eZuD7SKOqvz37VrF8eOHSMiIoI6deqwYMEC3nrrLSZOnOip2Vy1alWio6N56KGHAKd2MjgJ3oEDB2Ya88mTJwMeMzPt27encOHCfPnll7z99tvkyZOHadOmceTIEfr16+dTU9sdw/ffZ73S3h2D+/qd672QXfnz5+fBBx9k7NixzJo1i1atWpGSksLXX39NiRIlaNWqlU//ESNG8PTTT1OkSBFatGhB+fLlCQsLwxjDlClTWLNmDWfOnLnouJYsWULTpk1JTU2lWbNmtG3blsKFCxMUFMTq1auZOnVqwONEREQEHM99zdPS0s4rDhERERERERGRnKbk86VyCUtZXE3Cw8NJSEgIWJ4gNTWV+Pj4gA9jO3jwYMDxDhw44Bn3fCxYsIB77rmHoKAgZs6cSb169QL2q1q1KuCsFA7EvTq3SpUq5zymO0Z3zBnFxcX59PN28OBBTyzeAp3/7bffzrRp0zhz5gwrVqxgxowZfPDBB3Tp0oUSJUrQvHlzT//27dszadKkc8buLatV3u6E7scff8zs2bO5++67A5bc8I556tSptG3b9pzHdfc/13vhfHTv3p2xY8cyfvx4WrVq5UmUP/300z7vz9TUVKKjoylVqhQrV670+wWCexV3INlZFe/tzTff5PTp08ydO9fvLwTefvttpk6del7jiYiIiIiIiIjkFkHn7iKSOXfpgsxWWd5yyy2kp6czf/58v23z588nLS2NWrVq+W3bvXs3sbGxfu3z5s3zjJtdc+bMoVWrVgQHBzN79uxME88ATZo0AWDWrFmkp6f7bDtx4gSLFi0if/78WY7h5o7RHbO31NRUFi5cCBDw/H///Xe/th07drBnzx6ioqICrnoNDQ2lfv36vP7664wYMQLAk7isVq0aERERLFmyhJSUlHPGfj7c5SrGjx9PfHw806dP5+abb6ZmzZo+/dzXbMGCBdkat1q1aoSFhbF69eqAK9EDXddzadCgAddddx1Tp07l2LFjmSbK4+PjSUxMpH79+n6J55MnT3pKquSEbdu2UbRo0YClaQK9Dy7Eue5TEREREREREZFLQclnuShFihTBGMPu3bsDbu/ZsycAL730EklJSZ72pKQkXnzxRQAeffRRv/3S0tL417/+5ZMA3rlzJyNGjCA4OJj/+7//y1Z8s2bN4t577yVfvnz89ttv3HbbbVn2r1y5Mi1btiQ2NpYPP/zQZ1t0dDSnTp2iW7duFChQ4JzHvu+++yhatChff/01S5Ys8dn2/vvvs2PHDpo3b+5X7xlg+PDh7Nq1y/Nzeno6/fv3Jz09nUceecTTvmDBgoCJWfdq4bCwMMApxdCnTx/i4uLo27cvp0+f9tsnLi6OjRs3nvO8MvJO6I4ePZqUlJSA9ZPbtWtH5cqV+fDDD/nll18CjrV48WLP+yRv3rx07dqVEydO+NWIXr58ORMmTDjvWMFJNCcnJzNq1Ch++eUXbr75Zr9fZkRGRhIWFsaKFSt8SpGkpKTw9NNPEx8ff0HHDiQqKoqEhATWrl3r0/7JJ58wc+bMHDnGue5TEREREREREZFLQWU35KIULFiQunXrsmDBArp27UqVKlXIkycPbdu25eabb6ZLly5MnTqV7777jurVq3Pfffd5aubu3LmTBx98kK5du/qNe/PNN7N06VJq165Ny5YtOXbsGN9++y2JiYm88847fg/UC2Tz5s20a9eO5ORkWrduzdSpUwOWMMiY2Bw1ahT169enb9++/Pbbb1x//fUsXbqUuXPnUqVKFd56661sX5tx48bxwAMP0KhRIx544AHKly/PihUrmDVrFqVKleKjjz4KuG+DBg2oWbMmnTp1Ijw8nJkzZ7JmzRpq167NCy+84Ok3dOhQZs2aRePGjalUqRIFCxZkw4YNTJ8+nSJFitCrVy9P31dffZU1a9YwZswYfvrpJ5o2bUqZMmU4dOgQW7duZdGiRbz11lvccMMN2To/b926dePVV1/ljTfeIDg4mC5duvj1yZs3L5MmTeKuu+7innvuoX79+tSsWZOwsDD27NnDsmXL2LFjB3FxcZ6k+X/+8x9+++033n//fZYvX07Dhg2Ji4vj22+/pXXr1vz4448XFOtrr71GdHQ0KSkpfqueAYKCgujbty+DBg3ipptuol27dpw9e5a5c+eSkJBAkyZNmDt37nkfO5B+/foxc+ZMGjZsyIMPPkh4eDjLly9n4cKFdOzYkYkTJ170Mc51n4qIiIiIiIiIXBLWWn15fY0aNcqOGjXKSvZt3brV3nvvvbZo0aLWGGMB++mnn3q2p6Wl2Q8//NDWrl3b5s+f3+bPn9/WqlXLjhw50qalpfmNB9hGjRrZffv22a5du9oSJUrY0NBQe8stt9gJEyZkO665c+da4Jxfgezevdv26NHDlipVyubNm9eWL1/e/j979x4d51Ufev+7JdmyJVu2dbf02kloSUgFcSg5MYGeGOwkhaYJoYDhHCgBUvIS8ja0XbDoOiVhlMBpe9JCGwopSZNwaQ7Eq9AU2lJCneBQKLF8lS3L8d2yLVu25Yt8k2/a7x+aUUeyLh5pdP9+1po1M8+z98xv9JtnW/5pP/t54IEHYmtra8Y/n5UrV8a77rorlpaWxilTpsR58+bFT3ziE3Hfvn2XtL377rsjELdv3x7/4i/+Il5zzTUxPz8/VlVVxU996lPx+PHj3dr/+Mc/jh/5yEfitddeG4uKimJBQUG8+uqr4+///u/HXbt2XfL6HR0d8Vvf+lZcvHhxnDNnTpwyZUqsqqqKb33rW+MXv/jF2NTU1NV2586dEYh33333gJ9x9+7dMScnJwLxt3/7t/tt29LSEj/72c/Gmu+3DxEAACAASURBVJqaOH369FhYWBh/9Vd/Nb7nPe+J3/72t+P58+e7td+/f3/86Ec/GktLS+O0adPiggUL4jPPPNOV389//vMDxtfTkiVLIhDz8vLigQMHem1z/vz5+Jd/+Zfx2muvjdOmTYsVFRXxQx/6UNy1a1dXnnbu3NnV/nJ+XosWLer1O/fDH/4wLly4MM6YMSPOmjUr3nrrrXHFihXxmWeeueR4ijHGK664Il5xxRW9vsfnP//5CMSXXnqp2/aBjlNJkiRJo+f/NJ7vuknSMBi1WmuIMY5krXvMe/zxxyPAfffdN9qhTFohBBYtWjSoNX0lSZIkSRpvHt18oevxZ17nSeqSsi6M1hu75rMkSZIkSZIkKessPkuSJEmSJEmSss7isyRJkiRJkiQp61xISGOO65BLkiRJkiRJ458znyVJkiRJkiRJWWfxWZIkSZIkSZKUdRafJUmSJEmSJElZZ/FZkiRJkiRJkpR1Fp8lSZIkSZIkSVln8VmSJEmSJEmSlHUWnyVJkiRJkiRJWWfxWZIkSZIkSZKUdRafJUmSJEmSJElZZ/FZkiRJkiRJkpR1Fp8lSZIkSZIkSVln8VmSJEmSJEmSlHUWnyVJkiRJkiRJWWfxWZIkSZIkSZKUdRafJUmSJEmSJElZZ/FZkiRJkiRJkpR1Fp8lSZIkSZIkSVln8VmSJEmSJEmSlHUWnyVJkiRJkiRJWWfxWZIkSZIkSZKUdRafJUmSJEmSJElZZ/FZkiRJkiRJkpR1Fp8lSZIkSZIkSVln8VmSJEmSJEmSlHUWnyVJkiRJkiRJWWfxWZIkSZIkSZKUdRafJUmSJEmSJElZlzeYTiGEQqAaKAXOAAdjjPuyGZgkSZIkSZIkafy67OJzCOFXgI8BtwJvpMes6RBCK/AS8D3g+zHGC1mMU5IkSZIkSZI0jgxYfA4h3AB8AbiFzoLzeWADcAA4AkwHSoBrgPcB7wUOhRD+GvhSjPHs8IQuSZIkSZIkSRqr+i0+hxC+CXwQOA48AXwXWBljbO+j/ZXAbwJ3A18EPhFC+N0Y48tZjFmSJEmSJEmSNMYNdMHB24A/AqpijJ+MMb7cV+EZIMa4K8b49RjjW4DrgTXAouyFK0mSJEmSJEkaDwZaduM1McYzg3nhGGM98O4QwvTB9A8hlADvBm4H3kDnBQ7P0bnkxzPAMzHGjl76vQX4HPBmYBqwDXga+EqM8eJgYpEkSZIkSZIkZabf4vNgC89Zeo33AY8D++m8kGETUAH8DvB3wDtDCO+LMcZUhxDCu+i84GE78Byda1LfAXwZeGvyNSVJkiRJkiRJw2zACw5ejhDCFOD1wOkY46vZeE1gC3An8C/pM5xDCP8LWAm8h85C9PeS24uAJ4GLwNtijKuS2x8EXgTeG0L4QIzxu1mKT5IkSZIkSZLUh4HWfO4mhLA0hLAshFCctu1XgAZgFbAphPD9EMKQi9oxxhdjjD/subRGjPEA8LfJp29L2/VeoAz4bqrwnGzfTucyHAD3DTUuSZIkSZIkSdLAMio+Ax8DXhdjPJK27S+BX6VzaYx64F3AR7MTXp/OJ+8vpG1bnLz/t17avwycBt4SQsgfzsAkSZIkSZIkSZkXn38NqEs9SS518VvAshjjLcCNwGaGsficnFX94eTT9ELzNcn7LT37xBgvADvpXGbkNcMVmyRJkiRJkiSpU6bF5zI6LwCYchOdBd3vAsQYzwM/AX4lK9H17s/oXF/6X2OMP07bPit5f7yPfqnts4crMEmSJEmSJElSp9xEInHZjWtraz8FbEskEj9KPr8HeCvwQCKROJ3c9jbg5kQi8cVsBxtCeAB4mM7Z1b+dSCTOpMX2+0AJ8JVEInGkZ9/a2trfA+YBf5dIJPb28tr31tbWfv3222+vArj66qtpbGyktLSUtWvXUlVVxapVqygpKaGhoYH8/Hz27dvH6dOnOXXqFM3NzUydOpVNmzZRVlbGmjVruvVpbGxkypQpNDc3c+rUKc6cOcPevXvJz89n48aNVFRUdOszZ84cXn31VXJzczlw4AAnTpygvb2dPXv2MH36dOrr66msrGT16tVdfWbPns2WLVsIIXDw4EHa2to4d+4cu3fvpqCggPr6eqqqqqirq6O6uppVq1Yxa9Ystm3bRoyRw4cPc/ToUS5cuMCuXbuYMWMG69ev73r91PsVFRWxfft2Ojo6OHLkCK2trXR0dLBjxw5mzpzZrU9FRQVr165l5syZ7Ny5kwsXLnD06FEOHz4MwLZt25g9ezbr1q3r6lNeXs769espLCxk165dnD9/nuPHj3Pw4EFCCGzZsoU5c+Z061NWVkZ9fT0FBQU0NTVx9uxZ2traaGlpITc3l82bN1NSUsKaNWu6PntpaSkbN24kPz+fvXv30t7ezsmTJ9m/fz95eXld+U/1qauro7S0lE2bNnXlsmf+GxoaKC8v75bL4uJiGhsbycvL68plz/z3zGUq/zk5ObS0tNDW1sbZs2dpamrqyv/cuXP7zP+hQ4c4duwY58+fvyT/q1at6rovKiq6JP8XL15kx44dzJgxg3Xr1nX9vCorK1mzZk2/+S8qKurKS11dHZWVlaxbt47CwsJe879169Y+819QUMDu3bs5d+7cJfkvLi7udlyWlZWxYcMGpk2bRlNTE+3t7Zw4cYIDBw50y396n/T8p47lkydP0tzczJQpU2hsbOz1WE7Pf/qxnJ7/1atXd/3ciouL2bx5c5/H8oYNG7rlv66urtf8p47lVP575jJ1LAPd8t/bsZz67qTy39HRQWtra7f89zyWU/lPP5bT8799+3ZmzZrVLZcVFRVd+U8/lg8dOtQt/2vXru06xsrLy/s8lnNyci7Jf6pPKv979uzp81h2LHcsdyx3LHcsdyx3LHcsdyx3LB+rY/m6zdsgdpBzspXKi8ccyx3LHcsdy7M6ls+dO7c2oyJsFoUY4+U3DuHnQAVQA0Q6LzR4Msb4xrQ23wFuijFemdVAQ7gf+BtgE7AkeeHB9P11wA3ADTHG1b3035iM+9dijI19vc/jjz8eAe67z2sTSpIkSZKk4ffo5v+6pNVnXpc3ipFImqDCaL1xpstuPEHnmslbgcbk46d7tFlIZ1E6a0IIf0Bn4Xkj8PaeheekV5P3V/fSPw+4is4LFO7IZmySJEmSJEmSpEtlVHyOMX6TzjWXC+hcY/lvkjcAQgiLgSuBl7IVYAjhs8CXgXV0Fp4P9tH0xeT9O3rZdzOdMf8ixng2W7FJkiRJkiRJknqX6cxnYoz/K8ZYmrx9KnZft+M/gDnAX2UjuBDCg3QWu1fTudTG4X6a/wNwGPhACOGGtNeYBnwh+fTxbMQlSZIkSZIkSepfVhcSijGeA85l47VCCHfTeXHBi8DPgAdCuGR5kl0xxm8k37sthPBxOovQPw0hfBc4AtwJXJPc/lw2YpMkSZIkSZIk9a/f4nMIYXqM8cxQ3mAIr3FV8j4X+IM+2qwAvpF6EmN8PoSwCPgT4D3ANGAb8EfAYzGTqytKkiRJkiRJkgZtoGU3doYQPhVCyM/0hUMIC0II/wR8ejCBxRgTMcYwwO1tvfT7eYzxt2KMc2KM02OMb4gxfjnGeHEwcUiSJEmSJEmSMjdQ8fkF4EvA/hDC4yGEt4cQpvfVOITwmhDCfSGE/wTWAAvI4sUHJUmSJEmSJEnjQ7/LbsQYPxxCeAz438C9ydvFEEIjsB84SufSFiV0rqtcCgSghc6lL74cYzw7fOFLkiRJkiRJksaiAS84GGNcBdwWQngtcA+wBLgeeEOPpoeA7wPfA74XYzyf5VglSZIkSZIkSePEgMXnlBjjVuCPAUIIBUA1nTOezwAHY4z7hyVCSZIkSZIkSdK4c9nF53QxxtPA1uRNkiRJkiRJGXh084Wux9tPxku2f+Z1gyrZSNKY4kgmSZIkSZI0ioqnjnYEkjQ8LD5LkiRJkiSNMGc2S5oMckY7AEmSJEmSJEnSxGPxWZIkSZIkSZKUdRafJUmSJEmSJElZZ/FZkiRJkiRJkpR1GRWfQwjzQwhFA7SZGUKYP7SwJEmSJEmSJEnjWaYzn3cCnxqgzQPJdpIkSZIkSZKkSSrT4nNI3iRJkiRJkiRJ6tNwrPlcAZwahteVJEmSJEmSJI0TeQM1CCF8uMem63vZBpALzAd+F9iQhdgkSZIkSZIkSePUgMVn4BtATD6OwLuSt55Sy3GcBmqHHJkkSZIkSZIkady6nOLzR5P3AXgaeB74p17aXQRagf+MMR7LTniSJEmSJEmSpPFowOJzjPGbqcchhLuB52OM3xrWqCRJkiRJkiRJ49rlzHzuEmN8+3AFIkmSJEmSJEmaOHIyaRxC+PUQwkMhhIo+9lcm91+fnfAkSZIkSZIkSeNRRsVn4NPA7wEH+9jfAtwD/NFQgpIkSZIkSZIkjW+ZFp9vAl6KMcbedia3vwi8daiBSZIkSZIkSZLGr0yLz5XA3gHaNANzBxeOJEmSJEmSJGkiyLT4fBooG6BNGXB2cOFIkiRJkiRJkiaCTIvP64B3hRBm9LYzhFAEvCvZTpIkSZIkSZI0SWVafH6CzpnNPwkhXJe+I4SwAHgBKE22kyRJkiRJkiRNUnmZNI4xPhdCeCfwYWBtCKEF2AdUAxVAAL4ZY/xO1iOVJEmSJEmSJI0bmc58Jsb4EeATwCY6L0D4puR9A3BvjPGj2QxQkiRJkiRJkjT+ZDTzOSXG+ATwRAihAJgNHIsxns5qZJIkSZIkSZKkcWtQxeeUZMHZorMkSZIkSZIkqZtBFZ9DCGXAe4BrgcIY4++lbb8K2BBjPJO1KCVJkiRJkiRJ40rGxecQwj3AY8A0Oi8wGIHfS+6uAP4TuBd4KksxSpIkSZIkSZLGmYwuOBhCuBV4AtgCvBt4PH1/jHEjnRcevCtbAUqSJEmSJEmSxp9MZz5/FtgPLIoxtoUQ3thLm3rgpiFHJkmSJEmSJEkatzKa+QzcAPxzjLGtnzZ7gcrBhyRJkiRJkiRJGu8yLT5PBU4N0GY2cHFw4UiSJEmSJEmSJoJMi8+7gDcN0GYh8OqgopEkSZIkSZIkTQiZFp//CfjvIYT39bYzhPBR4Drge0MNTJIkSZIkSZI0fmV6wcH/A3wA+E4I4b3ALIAQwv8H/Hfgd4CtwFeyGaQkSZIkSZIkaXzJqPgcYzwaQlgEfAtIn/38WPL+Z8D/jDEOtC60JEmSJEmSJGkCy3TmMzHGJuBtIYTrgJuAEuA48MsY4+osxydJkiRJkiRJGof6LT6HEL4PfDfGuCz5/GZgV4yxKcZYD9SPQIySJEmSJEmSpHFmoAsO3gW8Lu35S8BHhi0aSZIkSZIkSdKEMFDx+RhQlPY8DGMskiRJkiRJkqQJYqA1nxuB/xFCqAP2J7ddmVx+o18xxpeHGpwkSZIkSZIkaXwaqPhcCzwP/N+0bXcnbwPJHWxQkiRJkiRJkqTxrd/ic4zxhRDCtcAtQDWQAFYkb5IkSZIkSZIk9Wqgmc/EGHcDTwGEEBLAT2OMDw9zXJIkSZIkSZKkcazfCw6GEL4UQrgtbdNH6VyGQ5IkSZIkSZKkPvVbfAb+AHhz2vOngbuGLxxJkiRJkiRJ0kQwUPH5JFCQ9jwMYyySJEmSJEmSpAlioDWftwG/E0L4R2B/ctvsEML8gV44xtg01OAkSZIkSZIkSePTQMXnR4G/B36Rtu1TyVt/4mW89oBCCO8FFgHXAwuAmcCzMcYP9dL2SmBnPy/3XIzxA0ONSZIkSZIkSZI0sH4LxDHG74QQdgK3A9XAR4B6YN3whwbA5+gsOp8E9gKvu4w+6+n9oogbsxiXJEmSJEmSJKkfA85OjjH+EvglQAjhI8A/xhgfHua4Uv6QzqLzNjpnQL90GX3WxRgTwxmUJEmSJEmSJKl/mS6N8VFg7XAE0psYY1exOQSvdShJkiRJkiRJ40VGxecY4zeHK5Asqgoh/L9ACdAK/GeMsX6UY5IkSZIkSZKkSSXjiwKGEHKA+4EPAtcChTHGvOS+NwIfB/4qxrglm4Fm4NbkrUsI4afA3THGplGJSJIkSZIkSZImmdxEInHZjUMIU4EfA58ApgOngZmJROJhgNra2nbgWaAtkUj8NJuB1tbWXknnBQ83JBKJ7/eyfyrQDvwR8MfAnwM/Aq4E3g7cUVtb+3QikTjf2+uHEO6tra39+u23314FcPXVV9PY2EhpaSlr166lqqqKVatWUVJSQkNDA/n5+ezbt4/Tp09z6tQpmpubmTp1Kps2baKsrIw1a9Z069PY2MiUKVNobm7m1KlTnDlzhr1795Kfn8/GjRupqKjo1mfOnDm8+uqr5ObmcuDAAU6cOEF7ezt79uxh+vTp1NfXU1lZyerVq7v6zJ49my1bthBC4ODBg7S1tXHu3Dl2795NQUEB9fX1VFVVUVdXR3V1NatWrWLWrFls27aNGCOHDx/m6NGjXLhwgV27djFjxgzWr1/f9fqp9ysqKmL79u10dHRw5MgRWltb6ejoYMeOHcycObNbn4qKCtauXcvMmTPZuXMnFy5c4OjRoxw+fBiAbdu2MXv2bNatW9fVp7y8nPXr11NYWMiuXbs4f/48x48f5+DBg4QQ2LJlC3PmzOnWp6ysjPr6egoKCmhqauLs2bO0tbXR0tJCbm4umzdvpqSkhDVr1nR99tLSUjZu3Eh+fj579+6lvb2dkydPsn//fvLy8rryn+pTV1dHaWkpmzZt6splz/w3NDRQXl7eLZfFxcU0NjaSl5fXlcue+e+Zy1T+c3JyaGlpoa2tjbNnz9LU1NSV/7lz5/aZ/0OHDnHs2DHOnz9/Sf5XrVrVdV9UVHRJ/i9evMiOHTuYMWMG69at6/p5VVZWsmbNmn7zX1RU1JWXuro6KisrWbduHYWFhb3mf+vWrX3mv6CggN27d3Pu3LlL8l9cXNztuCwrK2PDhg1MmzaNpqYm2tvbOXHiBAcOHOiW//Q+6flPHcsnT56kubmZKVOm0NjY2OuxnJ7/9GM5Pf+rV6/u+rkVFxezefPmPo/lDRs2dMt/XV1dr/lPHcup/PfMZepYBrrlv7djOfXdSeW/o6OD1tbWbvnveSyn8p9+LKfnf/v27cyaNatbLisqKrryn34sHzp0qFv+165d23WMlZeX93ks5+TkXJL/VJ9U/vfs2dPnsexY7ljuWO5Y7ljuWO5Y7ljuWO5Y7ljuWO5Y7lg+GcfyuXPn1l5uDTbbQozx8huH8CfAI0AC+ALwEPBgjDE3rc0LQFGM8c1ZDTSEt9F5wcFnY4wfyqBfHvAfwELgD2KMf91f+8cffzwC3HfffYMPVpIkSZIkSZLGhlG7mF6my258EPh5jPFhgBBCb5XrncAdQw0sW2KMF0IIf0dn8flmoN/isyRJmjzu+MIP+t3/w8/dOUKRSJIkSdLEk2nx+SrgXwZocwQoHlw4w+ZQ8r5wVKOQJElj1r7WU12Pq0v8lUGSJEmShirT4vMZYPYAbeYDxwYXzrBJLQGyY1SjkCRJY0r6zOb0WdDOeJYkSZKkocvJsP064LbkhQcvEUKYBfwmsHKogWUqhLCwt7hCCIuBP0w+/fuRjUqSJEmSJEmSJqdMZz4/CTwLPBtCuCd9RwhhNvAMMAf422wEF0K4C7gr+bQyeX9TCOEbyceHY4yfTj7+c6AmhPBTYG9y23XA4uTjB2OMv8hGXJIkSZIkSZKk/mVUfI4xfieEcAvwUeBO4ChACGEVUAPkA1+NMf5rluK7Hri7x7bXJG8Au4FU8fnbwLuB/wa8E5gCtADLgL+JMf4sSzFJkiRJkiRJkgaQ6cxnYoz3hBB+BnyKzpnFAfh1oAH4UozxmWwFF2NMAInLbPsU8FS23luSJEnSyEtff703rskuSZI0fmRcfAaIMX4D+EYIYTqdy2wcjzGe6r+XJEmSJF2+fa3/9V+M6pLCUYxEkiRJgzGo4nNKjPEMcCZLsUiSxjlnq0mShir934r0f1f8N0SSJGn8GVLxWZKkvjhbTZIkSZKkyc3isyQpa5ytJkmSJEmSUnJGOwBJkiRJkiRJ0sRj8VmSJEmSJEmSlHUuuyFJkiaV3i6MeeLMOVZvP0SMkSvLi2hoaqVmfskoRCdJkiRJE4fFZ0mSNGlt3nuUA8dOc7L9fNe2hj1HeP0Dz3JzTRUPLV3IkgXzRjFCSZIkSRq/srLsRgihNITw7hDCb4YQcrPxmpIkScPhh5+7kx9+7k7uWvgrbDtwvFvhOd3LDc3clniep/+9YYQjlCRJkqSJIaPicwjhvhDCKyGE4rRtbwIagX8A/hX4RQihMLthSpIkZc/y9Xu492svDtiuI0Y+/tUXWb5+zwhEJUmSJEkTS6Yzn98PxBjjkbRtjwJzgGfoLD7/N+AT2QlPkiQp+x5e9godMV5W244YeWTZymGOSJIkSZImnkyLz68F6lNPQgilwCLgqRjj78UY7wDqgP+ZvRAlSZKyp6GplZcbmjPqs6JhHw1NrcMUkSRJkiRNTJkWn0uAg2nP35q8/8e0bT8DrhhKUJIkScNlef3gltAYbD9Jg9PQ1MrOlja2Nh9jZ0ubfwCSJEkah/IybH8EKE17vgjoAH6Rti0C04YYlyRpHEsVDC5c7CAvN4eGplZq5peMdlgSAG2nz41oP0mZWb5+Dw8ve+WSMxRe/8Cz3FxTxUNLF7JkwbxRik6SJEmZyLT43AjcEUL4E+AinWtA18UY29LaXAkcyE54kqTxxIKBxoOigqkj2k/S5XvqJw3c+7UX+1yT/eWGZm5LPM+T9y/mY7fUjHB0kiRJylSmy278NTAX2AvsASqBr6V2hhBygd8A1mcrQEnS+PDUTxq4LfF8n2vppgoGT/97wwhHJnW35LrB/QFksP0kXZ7l6/f0W3hO6YiRj3/1RZavdykcSZKksS6j4nOM8QfAJ4AG4FXg0zHGv09rcgudS278OGsRSpLGPAsGGk9q5pdwc01VRn0W1VS7dIw0zB5e9sqA/46kdMTII8tWDnNEkiRJGqpMZz4TY3wixnhD8vblHvt+HGOcE2N8InshSpLGOgsGGm8eWrqQnBAuq21OCDy49MZhjkia3BqaWvs8c6YvKxr2eRFCSZKkMS7j4rMkSeksGGg8WrJgHk98cvGA7XJC4Mn7F7tWuTTMltcP7oyYwfaTJEnSyMio+BxCeGMI4ZMhhFlp2wpDCN8MIRwLITSHED6V/TAlSWOVBQONV2++ppIrymYyfWpur/sX1VTzQuIuL2omjYC20+dGtJ8kSZJGRl6G7T8L/PcY49fStv0p8LvASaAE+FIIoTHG+EKWYpQkjWEWDDTeLF+/h4eXvdLrjP3pU3Mpn1XAvzx4p2s8SyOoqGDqiPaTJEnSyMh02Y0bgJ+mnoQQpgB3AyuBcuAq4DDwQJbikySNcRYMNJ489ZMGbks83+dSMWfOXWT3oRO8suXACEcmTW5Lrhvc0jaD7SdJkqSRkWnxuRxIP0/6BmAm8PUYY3uMsRn4J+C6LMUnSRrjLBhovFi+fg/3fu3Fy7o45se/+iLL17s0jDRSauaXcHNNVUZ9FtVUe4aCJEnSGJdp8TnSfamO30huW5G27RBQNsS4JEnjhAUDjRcPL3vlsgrPAB0x8siylcMckaR0Dy1dSE4Il9U2JwQeXHrjMEckSZKkocq0+NwEvDnt+buAvTHGHWnbqoCjQw1MkjR+WDDQWNfQ1NrnUht9WdGwj4am1mGKSFJPSxbM44lPLh7w35OcEHjy/sUsWeAZNJIkSWNdpsXnZcBbQgj/EEL4e+Am4B96tHk9sD0bwUmSxgcLBhrrltcPbgmNwfaTNDj33FrDC4m7WFRT3ev+RTXVvJC4i4/dUjPCkUmSJGkw8gZu0s2XgXcAv5N8vg54OLUzhPBrwJuA/52V6CRJ48Y9t9ZwZXkRjyxbyYqGfZfsX1RTzYNLb7TwrFHRdvrciPaTNHhLFsxjyYJ5NDS18v5Hf8SFix3k5ebw3Gfe6ZJNkiRJ40xGxecY40ngrSGE1yc3bYoxdqQ1OQ28G1iVpfgkSeOIBQONVUUFU0e0n4bmji/8oN/9P/zcnSMUSXZMtM8zUmrml3BVRVG355IkSRpfMp35DECMcWMf23cBu4YQjyRpArBgoLFmyXWDm3E/2H7Knn2tp7oeV5cUjmIk2THRPo8kSZLUn0EVn0MIU4AlwLXAjBjjI8nt04Ai4HCPGdGSJEmjpmZ+CTfXVGV00cFFNdX+4WSUpM8ETp81PF5nCE+0zyNJkiRdrkwvOEgI4R10zm7+F+AvgUTa7uuB/cD7sxCbJElS1jy0dOGAF8VMyQmBB5feOMwRSZIkSdLEllHxOYRwA/A8EIE/BP5v+v4Y4y+BnXSu+yxJkjRmLFkwjyc+ufiyCtBP3r/Yi2NKkiRJ0hBlOvP5QTovKnhDjPExYGsvbeqABUMNTJIkKdvuubWGFxJ3saimutf9xTPyefPVFXzslpoRjkySJEmSJp5M13x+K/B8jPFAP232ALcPPiRJkqThs2TBPJYsmEdDUyvvf/RHXLjYQV5uDqVF05g5fepohydJkiRJE0amxecZwOEB2hQwiLWkJUmSRlLN/BKuqiga7TAkSZIkacLKtEi8DxjoPNTrgR2DC0eSJEmSJEmSNBFkWnz+EfCbIYTf6G1nCOGdwFuAfx5qYJIkSZrcGppa2dnSxtbmY+xsaaOhqXW0QxqSifZ5JEmSpIFkuuzGnwIfAF4IIXwFuBIghHA7cDNwP7Af+FIWY5QkSdIksnz9Hh5e9govNzR32/76B57l5poqHlq6kCUL5o1SdJmbaJ9HkiRJulwZzXyOMe4DbgOagc8A7wMC8IPk8/3AO2KMA60LLUmSJF3iqZ80cFvi+UsKtSkvNzRzW+J5nv73hhGObHAm2ueRJEmSMpHpzGdijGtCCNcAtwM3ASXAceCXwD/FGC9kN0RJkiRNBsvX7+Her71IR4z9Mh6PhQAAIABJREFUtuuIkY9/9UWuKCsa0zOGR/vz3PGFH/S7/4efuzNr7yVJkiT1JuPiM0CM8SKds537/41WkiRpDFu5taXX7amincW5kfXwslcGLNSmdMTII8tWjuni81j6PPtaT3U9ri4pHJb3kCRJknoaVPFZkiRpoqkunjHaIUxqDU2tfS5N0ZcVDftoaGqlZn7JMEU1eGPh86T/8SR9FrR/VJEkSdJIGVTxOYRwA3AjMAfI7aVJjDE+MpTAJEkaCk831+W48bUVXY/9Toyu5fV7Bt1vLBafJ9rnkSRJkgYjo+JzCKEI+D7wdjovNNiXCFh8liSNCZ5uLo19bafPjWi/4TbRPo8kSZI0GJnOfH4UWAz8DHgG2AN4gUFJ0pjj6ebS+FJUMHVE+w23ifZ5JEmSpMHItPj8LmAN8PYYY8cwxCNJkqRJaMl1g7vQ3mD7DbeJ9nkkSZKkwcjJsP0s4CULz5IkScqmmvkl3FxTlVGfRTXVY3Z95LH0eRqaWtnZ0sbW5mPsbGmjoak16+8hSZIk9SbTmc9bgYoBW0mSlEVePFCaHB5aupDbEs/TEeOAbXNC4MGlN45AVIM32p9n+fo9PLzsFV5uaO62/fUPPMvNNVU8tHQhSxY401qSJEnDJ9OZz18F7gghVA9HMJIkDWRf66mum6SJZcmCeTzxycXkhP6ua91ZqH3y/sVjvnA6mp/nqZ80cFvi+UsKzykvNzRzW+J5nv73hqy9pyRJktRTpjOff0TnBQd/HkKoBVYDx3prGGNsGmJskiQBXjxQmkzuubWGK8uLeGTZSlY07Ltk/6Kaah5ceuOYLzynjMbnWb5+D/d+7cUBZ1x3xMjHv/oiV5QVjZufpyRJksaXTIvPu4AIBODv+mkXB/HakiRlXWqt0wsXO8jLzaGhqXXMrhErqdOSBfNYsmAeDU2tvP/RH3Udv8995p3j8vgd6c/z8LJXLmupD+gsQD+ybKXFZ0mSJA2LTAvE36KzsCxJ0iXSZyWv3NpyyfaRnKnsWqfS+Fczv4SrKoq6PR/PRuLzNDS19rnURl9WNOwbU3+YG0v/lkiSJGloMio+xxg/MkxxXCKE8F5gEXA9sACYCTwbY/xQP33eAnwOeDMwDdgGPA18JcZ4cdiDliR1qS6eMWrv/dRPGvo95Ty11umT9y/mY7fUjHB0kjR8ltfvGXS/sVJ8Tjea/5ZIkiRp6Mby0hifo7PofBLYC7yuv8YhhHcB3wPageeAI8AdwJeBtwLvG85gJUnDPxvtcpbQcK1TSZNZ2+lzI9pvODizWZIkaeLIqPgcQrgIJGKMj/TT5k+A2hjjUAvbf0hn0XkbnTOgX+rnPYuAJ4GLwNtijKuS2x8EXgTeG0L4QIzxu0OMSZI0CjJZQsO1TiVNZkUFU0e0nzQRpC/10hv/ICJJ0uBlWiAOydvltBuSGGNXsTmEAV/uvUAZ8K1U4Tn5Gu0hhM8By4H7AIvPkjTOZLKExsKrK8f9WqfKvv6KCiu3tnDjaytGMBppeC25bnB/TBtsP2mi2dd6qutxdUnhKEYiSdLEMBzLbsyhc+mLkbQ4ef9vvex7GTgNvCWEkB9jPDtyYUmShiLTJTTue8cbBvc+Y3StU2WfRQVNdDXzS7i5piqjP8Qtqql2DNSklj6zOf0Pls54liRp6AYsPocQbu6x6cpetgHkAvOBDwKvZiG2TFyTvN/Sc0eM8UIIYSdQA7wGaBzJwCRJg5fpEhr/snrnoN5nLK11quzrr6gw0KnW0nj00NKF3JZ4/rLGz5wQeHDpjSMQlSRJkiaj3EQi0W+D2trancDdwEeSm65PPu55+zBwF1AAfCaRSGzIVpC1tbVXJt9jQyKR+H4v+z8DlACPJhKJI73s/yAwD3g2kUjs7e09Qgj31tbWfv3222+vArj66qtpbGyktLSUtWvXUlVVxapVqygpKaGhoYH8/Hz27dvH6dOnOXXqFM3NzUydOpVNmzZRVlbGmjVruvVpbGxkypQpNDc3c+rUKc6cOcPevXvJz89n48aNVFRUdOszZ84cXn31VXJzczlw4AAnTpygvb2dPXv2MH36dOrr66msrGT16tVdfWbPns2WLVsIIXDw4EHa2to4d+4cu3fvpqCggPr6eqqqqqirq6O6uppVq1Yxa9Ystm3bRoyRw4cPc/ToUS5cuMCuXbuYMWMG69ev73r91PsVFRWxfft2Ojo6OHLkCK2trXR0dLBjxw5mzpzZrU9FRQVr165l5syZ7Ny5kwsXLnD06FEOHz4MwLZt25g9ezbr1q3r6lNeXs769espLCxk165dnD9/nuPHj3Pw4EFCCGzZsoU5c+Z061NWVkZ9fT0FBQU0NTVx9uxZ2traaGlpITc3l82bN1NSUsKaNWu6PntpaSkbN24kPz+fvXv30t7ezsmTJ9m/fz95eXld+U/1qauro7S0lE2bNnXlsmf+GxoaKC8v75bL4uJiGhsbycvL68plz/z3zGUq/zk5ObS0tNDW1sbZs2dpamrqyv/cuXP7zP+hQ4c4duwY58+fvyT/q1at6rovKiq6JP8XL15kx44dzJgxg3Xr1nX9vCorK1mzZk2/+S8qKurKS11dHZWVlaxbt47CwsJe879169Y+819QUMDu3bs5d+7cJfkvLi7udlyWlZWxYcMGpk2bRlNTE+3t7Zw4cYIDBw50y396n/T8p47lkydP0tzczJQpU2hsbOz1WE7Pf/qxnJ7/1atXd/3ciouL2bx5c5/H8oYNG7rlv66urtf8p47lVP575jJ1LAPd8t/bsZz67qTy39HRQWtra7f89zyWU/lPP5bT8799+3ZmzZrVLZcVFRVd+U8/lg8dOtQt/2vXru06xsrLy7uO5eUrG/jTH2zK6N+LY6cGV0R+2zXFXF2W71g+Ccbyx3+wkqlTp3LixEne9ab/h2dXbKbj4kUuXLzIwuo8x/IxOJY/9/NthJDDqVMn+dDbrh33Y/mzKzpfp+3ECT5224JhGcvnlxZSmHOel19tpb/yc06AP/vAG3nLVYXDNpb3PJZzcnIuyX+qTyr/e/bs6fNY9vdyx/Lh/r38pW0nOH78ONOmTePqghOO5f5e7u/ljuWO5eNwLPf38kvzP3fu3NpB/Wc5C0IcYEZECCEBRDrXcX4I+CmwopemF4FW4KUY4+asBhnC2+i84OCzMcYP9bJ/C/Ba4LUxxm297P8FcBNwU4zxl/291+OPPx4B7rvvvixELkkarMf+eR2f+ruXR+S9Nj72QU85nyT6m/ns6dVj00TL0Uh+nuXr9/DIspWsaNh3yb5FNdU8uPRGL7gq9TDRxhxJkpKGfH2+wRpw2Y0YYyL1OIRwN/B8jPGx4QxqEI4n72f1sb+oRztJ0hg32KUwriyfya6DJy67vWudSpqoliyYx5IF82hoauX9j/6ICxc7yMvN4bnPvNNxT+pFQ1MrO1vauo4VL0gsSdLQZXTBwRjjVcMVyBC9CtwAXA2sTt8RQsgDrgIuADtGPjRJUm8GWmv31uvnD+p1f+tNV/K3/7bRtU7Vp5VbW4DO72Dqceo5ONNNE0/N/BKuqijq9lzSf1m+fg8PL3vlkgt1vv6BZ7m5poqHli70LAFJkgYpo+LzGPYinRc6fAfwnR77bqZzHeqXY4xnRzowSdLA9rWe6npcXVIIwJLrBvefvE++8zp+/TXl3Pu1F/stQOeEwJP3L/Y/k5NcdfGM0Q5hwhvoD00W+yWNpqd+0tDv7wwvNzRzW+J5nrx/MR+7pWaEo5Mkafzrt/gcQniRzvWe744x7k0+vxwxxrhkyNFdvn8A/hz4QAjhKzHGVQAhhGnAF5JtHh/BeCRJA0gVnHqeDv5nH35r16y8m2uqLpmF1J/UEho180u4srzItU7VTep06jmF+Zd81zQyevtDkySNluXr9wz4x2qAjhj5+Fdf5IqyIn93kCQpQwPNfH5b8r6gx/OBDHyu8wBCCHcBdyWfVibvbwohfCP5+HCM8dMAMca2EMLH6SxC/zSE8F3gCHAncE1y+3NDjUmSlD2Xc4rrQ0sXclvi+UEtoeFap0rxdOrRdTl/aJKk0fDwslcu63cM6CxAP7Jspf9eSJKUoX6LzzHGnP6eD7Prgbt7bHtN8gawG/h0akeM8fkQwiLgT4D3ANOAbcAfAY/FeJm/VUiShl0mp7g+8cnFQ1pCw7VOJzdPpx59Fv8ljUUNTa0ZnV0FsKJhnxchlCQpQ/0Wk0MID4QQRuVKTDHGRIwx9HO7spc+P48x/laMcU6McXqM8Q0xxi/HGC+OwkeQJPUi01Ncrywv4oXEXSyqqe613aKaal5I3GXhUJfI9Lu2fP2eEYps8njqJw3clni+zwJPqvj/9L83jHBkkia75fWDG/MH20+SpMlqoJnMf0XnRfwACCFcDCE8OLwhSZImssGe4vrTL76HjY99kJp5xVxTNZuaecVsfOyD/PSL73HWpHo1mO+assfiv6SxrO30uRHtJ0nSZDXQms/tQH7a85C8SZKUsaGe4uoSGrpcnk49+oayluodX/hB176VW1u6Hqe2p9aRHi8m2ueRJoKigqkj2k+SpMlqoJnPO4HfDCFUpG1z7WRJ0qB4iqtGit+10TWU4n9P1cUzum4TwUT7PNJ4teS6wZ01Ndh+kiRNVgPNfP46nUtvNIfQNeE5EUJIDNAvxhgHem1J0iTjKa4aKX7XRtdQiv8180sm3EzgifZ5pImgZn4JN9dUZfSHskU11Z4dI0lShvotEMcYHwshHARuB6qAtwNNwK7hD02SNNF4iqtGit+10WXxX9J48NDShdyWeP6ylgjKCYEHl944AlFJkjSxDDg7Ocb4XeC7ACGEDuCZGOPDwx2YJGni8RRXjRS/a6PL4r+k8WDJgnk88cnFA14cNScEnrx/sRc4liRpEAZa87mnWuCnwxCHJGkSSJ3imglPcdVg+F0bXRb/JY0X99xawwuJu1hUU93r/kU11byQuIuP3VIzwpFJkjQxZLQuc4yxdrgCkSRNDp7iqpHid230uJbq2HDHF37Q9Xjl1pZLtrsWtdRpyYJ5LFkwj4amVt7/6I+4cLGDvNwcnvvMOx2XJEkaokxnPkuSNCSpU1xz/utCtr3yFFcNld+10fXQ0oUD/uxTLP4Pv+riGV03Sb2rmV/CVRVFvLZqNldVFFl4liQpCzKa+SxJUjbcc2sNV5YX8ciylaxo2HfJ/kU11Ty49EaLgRoyv2ujx7VUR58zmyVJkjTaLD5LkkaFp7hqpPhdGz0W/yVJkqTJzeKzJGlUpU5xTX8uDQe/a6PD4r8kSZI0eVl8liRJ0rCz+C9JkiRNPhafJUmj4o4v/KDr8cqtLZdsT1+rNJO2kiRJkiRpbLD4LEkaddXFM4alrSRJkiRJGj0WnyVJoyKT2crObJbGL89ckCRJkiYvi8+SJEkaEZ65IEmSJE0uFp8lSZI0bJzZLEmSJE1eOaMdgCRJkiRJkiRp4nHmsyRpQnKdWUmSJEmSRpfFZ0nShOc6s5IkqS/+wVqSpOFj8VmSNCH5H0VJkpQp/2AtSVJ2WXyWJEmSJE1a/sFakqTh4wUHJUmSJEmSJElZZ/FZkiRJkiRJkpR1Fp8lSZIkSZIkSVln8VmSJEmSJEmSlHUWnyVJkiRJkiRJWWfxWZIkSZIkSZKUdRafJUmSJEmSJElZZ/FZkiRJkiRJkpR1Fp8lSZIkSZIkSVmXN9oBSJIkDac7vvCDrscrt7Zcsv2Hn7tzxGOSJEmSpMnA4rMkSZo0qotnjHYIkiRJkjRpWHyWJEkTmjObJUmSJGl0uOazJEmSJEmSJCnrLD5LkiRJkiRJkrLO4rMkSZIkSZIkKessPkuSJEmSJEmSss7isyRJkiRJkiQp6yw+S5IkSZIkSZKyzuKzJEmSJEmSJCnrLD5LkiRJkiRJkrLO4rMkSZIkSZIkKessPkuSJEmSJEmSss7isyRJkiRJkiQp6yw+S5IkSZIkSZKyzuKzJEmSJEmSJCnrLD5LkiRJkiRJkrLO4rMkSZIkSZIkKessPkuSJEmSJEmSss7isyRJkiRJkiQp6yZc8TmEsCuEEPu4HRjt+CRJkiRJkiRpMsgb7QCGyXHgr3rZfnKkA5EkSZIkSZKkyWiiFp+PxRgTox2EJEmSJEmSJE1WE27ZDUmSJEmSJEnS6JuoM5/zQwgfAuYDp4B64OUY48XRDUuSJEmSJEmSJoeJWnyuBL7dY9vOEMJHY4wrRiMgSZIkSZIkSZpMJuKyG88AS+gsQBcCbwC+DlwJ/CiEsGD0QpMkSZIkSZKkySE3kUiMdgxZlUgkViQSiZ2JROJUIpE4n0gkDiYSiX+ura2dCdwMVCYSied69gsh3FtbW/v122+/vQrg6quvprGxkdLSUtauXUtVVRWrVq2ipKSEhoYG8vPz2bdvH6dPn+bUqVM0NzczdepUNm3aRFlZGWvWrOnWp7GxkSlTptDc3MypU6c4c+YMe/fuJT8/n40bN1JRUdGtz5w5c3j11VfJzc3lwIEDnDhxgvb2dvbs2cP06dOpr6+nsrKS1atXd/WZPXs2W7ZsIYTAwYMHaWtr49y5c+zevZuCggLq6+upqqqirq6O6upqVq1axaxZs9i2bRsxRg4fPszRo0e5cOECu3btYsaMGaxfv77r9VPvV1RUxPbt2+no6ODIkSO0trbS0dHBjh07mDlzZrc+FRUVrF27lpkzZ7Jz504uXLjA0aNHOXz4MADbtm1j9uzZrFu3rqtPeXk569evp7CwkF27dnH+/HmOHz/OwYMHCSGwZcsW5syZ061PWVkZ9fX1FBQU0NTUxNmzZ2lra6OlpYXc3Fw2b95MSUkJa9as6frspaWlbNy4kfz8fPbu3Ut7ezsnT55k//795OXldeU/1aeuro7S0lI2bdrUlcue+W9oaKC8vLxbLouLi2lsbCQvL68rlz3z3zOXqfzn5OTQ0tJCW1sbZ8+epampqSv/c+fO7TP/hw4d4tixY5w/f/6S/K9atarrvqio6JL8X7x4kR07djBjxgzWrVvX9fOqrKxkzZo1/ea/qKioKy91dXVUVlaybt06CgsLe83/1q1b+8x/QUEBu3fv5ty5c5fkv7i4uNtxWVZWxoYNG5g2bRpNTU20t7dz4sQJDhw40C3/6X3S8586lk+ePElzczNTpkyhsbGx12M5Pf/px3J6/levXt31cysuLmbz5s19HssbNmzolv+6urpe8586llP575nL1LEMdMt/b8dy6ruTyn9HRwetra3d8t/zWE7lP/1YTs//9u3bmTVrVrdcVlRUdOU//Vg+dOhQt/yvXbu26xgrLy/v81jOycm5JP+pPqn879mzp89j2bHcsdyx3LHcsdyx3LHcsdyx3LHcsdyx3LHcsXwyjuVz586tzU7lNXMhxjha7z2iQgi/CmwFjsQYS/pq9/jjj0eA++67b6RCkyRJkiRJkqThEkbrjSfisht9OZi8LxzVKCRJkiRJkiRpEphMxeebkvc7RjUKSZIkSZIkSZoEJlTxOYRQE0Io7mX7FcDfJJ/+/chGJUmSJEmSJEmTT95oB5Bl7wP+OITwErATOAH8CnA7MA34V+AvRi88SZIkSZIkSZocJlrx+SXgGuCNdC6zUQgcA/4D+Dbw7ThZrrAoSZIkSZIkSaNoQhWfY4wrgBWjHYckSZIkSZIkTXYTas1nSZIkSZIkSdLYYPFZkiRJkiRJkpR1Fp8lSZIkSZIkSVln8VmSJEmSJEmSlHUWnyVJkiRJkiRJ/3979x0vR13vf/z1Se89IZSEEiJoAEGQEhACKldBBHujxIZdFPWngkq496J4BYRLEVQEESuoIF5REAiRcpF6QxEIJSSUAAkkpJH6/f3xnQObw56T3XO2HM55PR+Peew5U3Zm3ju7O/PZme/UnMVnSZIkSZIkSVLNWXyWJEmSJEmSJNWcxWdJkiRJkiRJUs1ZfJYkSZIkSZIk1ZzFZ0mSJEmSJElSzVl8liRJkiRJkiTVnMVnSZIkSZIkSVLNWXyWJEmSJEmSJNWcxWdJkiRJkiRJUs1ZfJYkSZIkSZIk1ZzFZ0mSJEmSJElSzVl8liRJkiRJkiTVnMVnSZIkSZIkSVLNWXyWJEmSJEmSJNWcxWdJkiRJkiRJUs1ZfJYkSZIkSZIk1ZzFZ0mSJEmSJElSzVl8liRJkiRJkiTVnMVnSZIkSZIkSVLNWXyWJEmSJEmSJNWcxWdJkiRJkiRJUs1ZfJYkSZIkSZIk1ZzFZ0mSJEmSJElSzVl8liRJkiRJkiTVnMVnSZIkSZIkSVLNWXyWJEmSJEmSJNWcxWdJkiRJkiRJUs1ZfJYkSZIkSZIk1ZzFZ0mSJEmSJElSzVl8liRJkiRJkiTVnMVnSZIkSZIkSVLNWXyWJEmSJEmSJNWcxWdJkiRJkiRJUs1ZfJYkSZIkSZIk1ZzFZ0mSJEmSJElSzVl8liRJkiRJkiTVnMVnSZIkSZIkSVLNWXyWJEmSJEmSJNWcxWdJkiRJkiRJUs1ZfJYkSZIkSZIk1ZzFZ0mSJEmSJElSzVl8liRJkiRJkiTVnMVnSZIkSZIkSVLNWXyWJEmSJEmSJNWcxWdJkiRJkiRJUs1ZfJYkSZIkSZIk1ZzFZ0mSJEmSJElSzVl8liRJkiRJkiTVXLcrPkfEFhHxs4h4MiJWRcTciDg9IkY2e9kkSZIkSZIkqafo0+wFqKWImATcBIwDLgfuB3YHjgHeFhF7p5QWNXERJUmSJEmSJKlH6G5nPp9DLjx/MaV0WErpGymlA4AfAtsBJzV16SRJkiRJkiSph+g2xeeI2AY4EJgLnN1q8AnAcuCIiBjc4EWTJEmSJEmSpB6n2xSfgQOKx6tSSutLB6SUlgI3AoOAPRu9YJIkSZIkSZLU03SnNp+3Kx4fbGP4HPKZ0a8BrtnYk/3oRz+q0WJJkiRJkiRJUtOkz3zmM9GMGXenM5+HF49L2hje0n9EuYERcXRE3FbzpZIkSZIkSZKkHqg7nfm8MS3V/VRuYErpx8CPG7c4kiRJkiRJktR9daczn1vObB7exvBhrcaTJEmSJEmSJNVJdyo+P1A8vqaN4ZOLx7bahJYkSZIkSZIk1UikVLYVilediJgEPATMBSallNaXDBsKPEUuto9NKS1vykJKkiRJkiRJUg/Rbc58Tik9DFwFbAV8rtXgE4HBwEUWniVJkiRJkiSp/rrNmc/w0tnPNwHjgMuBfwF7APuTm9uYmlJa1LwllCRJkiRJkqSeoVsVnwEiYgLw78DbgNHk5jYuA05MKT3XzGWTJEmSJEmSpJ6i2xWfJUmSJEmSJEnN123afJYkSZIkSZIkdR0WnyVJkiRJkiRJNden2QvQFfXu3SutX9/x5kgigno3Z9IroBOL2Pn5A+ubN/sNNCLvDefXi5QauPYR8GpqHid6QSPzqaWmZB1A41/frvQelmqp2d+PkiRJktSV9OvXh9Wr1y5MKY1txvwtPpexfn0i3dbx6WO3xKxZs2q3QGXsu+++pLPqOot2xech7fTe5i1AiZh9KYc88PGGze+K7c7nQzevbtj8fr1XP3a9ZE7D5tdZt79vMhz3Kq38fDf40Afua+gsf/3b18GhZzR0ngDrLz+GtPfODZ+vVG9x412d+g6XJEmSpO4kdlsL8Fiz5m+zG5IkSZIkSZKkmrP4LEmSJEmSJEmqOZvdqNCsO+CUX8Dt98OTz8IFJ8D0Qzr+fBdffDGzZs1i3rx59O3blylTpnD00UezzTbb1G6hS5x9PZx3I8x9Lv8/ZTx8621w8A51mV3FvvvMvzh+wb18bvQkztp8l7rP74Ez7+DBs+7coF//MQM58MYP122eKxc+xV3nHM9TN/+VNSuWMmSzrXnj185i3Bv2rds823L3Z6ex+tknXtF/2C77Mfm4nzZ8eSqyfh38YwbcczEsewqGbApTPgL7zoBezf0IW7NmObPv/m8ef+LvrFr1HCNHvJY37PJNRo/esXYzWfgQPHwdLJ4PLy6BXT4ME/d4eXhK8MBfYe5NsGYljNwSdnovDNu0dsugisxasoxTnniG25ev5MnVa7hg2wlM32T0S8P/sGgx5y1YxB3LVrBw7Tqu22ES04YPbeISS5KkzmjvGHHNWvjWOXDlTfDw4zBsMOy/G5z8BZg4vrnLLUlSI1l8rtCyFbDDtnDkwXDkCZ1/vjvvvJPDDjuM7bffnpQSP/vZzzj22GO56KKLGDZsWOdn0MoWI+H7h8LkcbB+Pfz8Fjjsx3D712GnzWs+u4r87/JF/GTRo+w0YHhD5zt46+FM/cVBL/0fvaNu81q9dDFXf2oaY3eayn6nXE7/kWNY9sSj9B/VlDbe2f57v88bQGHN4mf419ffxaipB7UzVZPd/H24/Ww45Ocwdkd4Zjb8+Sjo0x/2+XZTF+2ft36bxYsfZM89vseggZsw97EruO76j3PQ265g0KBNajOTdatg6HiY8Ea44+JXDn/oGnjoOnjDh2HIOHjgb3DTOfDm46HvgNosgyqybN16dhg8gCPHjeLIOa9sTmv5uvVMHTqYw8eO5Mg585qwhJIkqZbaO0Zc8SLccT8c/zHY+TWwZBl85XR42xdg9q+hj0fikqQewq+8Ch20T+4App/Y+ec79dRTN/j/+OOP56CDDuLuu+9m77337vwMWjl0pw3/P+md8KMb4OZHm1N8XrJuDR+Z/0/On7Ab//50Y2/w1qtPMGDsoIbM618Xn8LA0ePZ64QLXuo3ZLOtGzLvcvoOH73B/wuvvYTeA4cwcq+3N2mJKvDETTD5kNwBjNgKJr8TnrylqYu1du2LzH/8avbZ+3Q2Gbc7ADvu8HmeeHImDz38G3ba8ZjazGiTKbkDuOOXGw5LCR6+Hia/BTYrbh74ho/Ald+CJ26HrWr/WaK2HTRqGAeNyj8eTi9TXD5i3CgAFq5Z29DlkiRJ9dHeMeLwIXD1ORv2O+84mPJ++Ndc2HHbhiyiJEnC5PoxAAAbRElEQVRNZ5vPXcSKFStYv349Q4fW/xLsdevhN7fBslUwtUl10KMfv533Dt+cA4aMa/i8l89fytVv+jV/P+C33P7la1k+/4W6zevxWX9i9JTdufFbH+YPB23OlUfuxoOXnENKqW7zrFRKiYXXXsKoNx1Kr/4Dm704bdtiH3jsOlh4f/7/2ftg7rUwqblna6e0jpTW0btX/w369+49gGefvaMxC7FiEax6AcZtV7IA/WDMJHju0cYsgyRJkirywvL8ONJWtyRJPYhnPncRZ555JpMnT2bKlCl1m8fdT8Bep8KLa2FIf/jjJ2HHJpz1/JNFj/DQ6mX8YsLuDZ/3iJ3Gssv39mXINiNY9dxK5vzoLm784J+Z9ud3029k7ZsoWPbko8z5w7ls94EvMu2Ir7F4zmxuP+1LALzmfZ+t+fyqsXT2Dax+5nHGvPl9TV2Ojdrr67B6Kfz4ddCrN6xfC1OPh12bm1/fvoMZM3pn7r3vXIYPn8yAAWN4bN7/sGjRXQwZMrExC7FqaX7s3+oIpv9QWLmkMcsgSZKkjVq9Br7yQzjkTbBFjVpnkyTp1cDicxdw1llnMXv2bM4++2x69+5dt/lstwnc9U1YvAJ+fxcc9QuYeQzssFndZvkKD7y4lOMW3MM/Jk2jX6/Gn3i/yX4TNvh/5OvHcc1bfsf8y+Yw6aM1vElci/XrGbX9ruz82ZMAGLXdLiydP4c5v/9R04vPz/79dwyatBODtn5dU5djo+77Ldx9ERz6Kxg7BZ6+C64+BkZsDTt/vKmLtueeJ3PLP7/F5VfsT0RvRo58HRMnHsTzzze2KRmiVbvlKUH9mjKXJElSFdauhcO/DYuXwZ9Oa/bSSJLUWBafm+zMM8/kmmuu4YwzzmCzzepbBe7XB7Yt7nO325Zw6zz44XVw/kfqOtsN3LxiEQvXrWaHB69+qd86ErOWL+TcRY+wfIfD6N+rfgX41voM7svQbUeyfG59mt4YMGZThm392g36Ddtqe5b/7qy6zK9Sa5YsYsmt1zDhEzW4e2a9Xfs12OOrMOWD+f9xO8KSx+Dm7zW9+Dx0yETecsBFrF27gjVrljNw4FhuvOlYhgzeojEL0HLG84svwMCRL/dfveyVZ0NLkiSp4dauhQ8dD3c/BDPPg9Ejmr1EkiQ1lsXnJjrjjDO49tprOeOMM9hyyy0bPv/1CVY1+L5Xhw3fjN0GvXWDfh+dfxuT+w/huHHb0y8aezb0ulVrWfboYsbssWldnn/sjnuxdN6DG/RbOm8Og8c3qFmGNiya+Xuibz9G7X1wU5ejImtX5OY2SvXqDWl9c5anjD59BtGnzyBWr17CUwtuZOfXf6UxMx40GvoPg2cfgJHFZ8i6NbDoYZhyaGOWQZIkSWWtWQsf/Cbc8zDM/DGMH9PsJZIkqfEsPldo2Qp4aH7+e/16mLcA7noARg2HieOrf77TTjuNq666ipNOOomhQ4eyaNEiAAYOHMigQYNquOTZNy6Hg6fAhJGw9EX41W0wcw78z6drPqt2jejdjxG9+23Qb3Cv3ozq3Y8dBgyv+/zv/f4tbLL/RAZtOoRVz63kwXPuYt2KtWzxrvrcbnq7Dx7D1Ufvy70Xfo+Jb34fzz94Fw9ecjY7ffo/6jK/SqSUWHjNJYza+2B6DxzStOWo2LaHwM0n52Y2xkyBp++EW06DHY9s9pLx1FM3kFjPsKHbsHTZPO76vx8wbOhWbLP1u2o3k7WrYPmzxT8JVj4PSx6HvoNg0CiYtB88eBUM2QSGjM1/9+4Pm+9au2VQRZatW8dDK1cBsJ7EvFVruGvZCkb17cPE/v14bs1a5q1azeJ16wB4aOVqRvRewfh+fRnfr28zF12SJHVAe8eIm42B930dbr0PrvhhbhFtwcI87vAhMLD2t5uRJKlLsvhcodvug/1LCrUnnJe7o94BF86o/vkuu+wyAL785S9v0H/69Ol87GMf68SSlrfgBTj857BgKQwfADttDld+Bv6tizf3W2svLljOHcfOZPXiF+k/cgAjdh7HPr87hEGb16eJgtGv2403ff9SZp/7be654LsM3mQCOx49g8nvaXDVv8Sye29h1VNz2foLpzRtGapy4Jkw69vw18/CimdgyKawyydhn+80e8lYs2Yp/zf7dFasXEC/fsOZsMWB7LTjMfTqVcNC4uJ5cGNJMy33X5m7CbvDGz4C2745n+08+1JYsyKfAT31M9DXI5pGu23ZCva/5+GX/j9h/gJOmL+Ao8aN5MLJW/Kn55bw0ZYjVOCTD+e/T5iwCTMm1ufqC0mSVD/tHSPOOBouvz733/XwDae74ASYfkjjllOSpGay+FyhabtBuq12zzdr1qzaPVkFLjyiobOrysxJ0xo2r11/eEDD5tVi870PYvO9D2r4fNsydIc92fWSOc1ejMr1HwpvPT13XczEiW9n4sS313cmYybDoWe0PTwCtn977tRU04YPJe29c5vDp28ymumbjG7gEkmSpHra2DFiLY8fJUl6tWpsA7uSJEmSJEmSpB7B4rMkSZIkSZIkqeYsPkuSJEmSJEmSas7isyRJkiRJkiSp5iw+S5IkSZIkSZJqLlJKzV6GLicilgIPNHs5uqkxwMJmL0Q3Zbb1Ya71Y7b1Y7b1Y7b1Y7b1Y7b1Ya71Y7b1Y7b1Y7b1Y7b1Y7Y9wxhgcEppbDNm3qcZM30VeCCltFuzF6I7iojbzLY+zLY+zLV+zLZ+zLZ+zLZ+zLZ+zLY+zLV+zLZ+zLZ+zLZ+zLZ+zLZnKF7nrZo1f5vdkCRJkiRJkiTVnMVnSZIkSZIkSVLNWXwu78fNXoBuzGzrx2zrw1zrx2zrx2zrx2zrx2zrx2zrw1zrx2zrx2zrx2zrx2zrx2x7hqa+zt5wUJIkSZIkSZJUc575LEmSJEmSJEmqOYvPkiRJkiRJkqSas/hciIgtIuJnEfFkRKyKiLkRcXpEjGz2snUVEfHeiDgzIv4RES9ERIqIizcyzdSI+EtEPBcRKyJidkR8KSJ6tzPNOyJiZkQsiYhlEXFLRBxV+zXqGiJidER8IiL+GBEPRcTKYt1viIiPR0TZ96nZViYivh8R10TE/CLb5yLizog4ISJGtzGN2XZARBxRfC6kiPhEG+NUnVNEHBUR/yzGX1JM/476rEXzFd8/qY1uQRvTuM1WISLeFBG/j4iniu/8pyLiqog4qMy4ZrsRETG9nW22pVtXZjqzrVBEHFxso48X32WPRMQlEbFXG+Ob7UZE9rGI+N+IWFrkdGdEfLGtnMx1Q9GFjw3iVb7vUE22EdE3Io6JiAsi4q6IWB3t7Iu1mraqnCKid/F6zY6X96v/EhFTO7O+jVRltpMj4usRcW3kY4nVEfF0RFweEftvZD49Ktsqc50QEecU7+UFkffFniym/WhE9G1nPj0qV+jYZ22r6c+Pl/fHtm1jnKpzioiBEXFiRDwQES9GxDMR8buIeG1H1lNVv4+2ivb3vX9Twfz6R8Q9xfiPtzHOWyPi1Mg1leeKcW+oeuVSSj2+AyYBTwMJuAw4Gbi2+P9+YHSzl7ErdMBdRSZLgX8Vf1/czviHAmuBZcD5wA+KPBNwSRvTfL4YvhA4G/ghML/od0qzM6hTrp8u1u9J4JfA94CfAYuL/pdStM9uth3KdzXwv0WmJwNnArcW6/0EMMFsa5LzhGKbXVqs9ydqkRNwSjF8fjH+2cCiot/nm73edcpybpHljDLdV8uM7zZbXb7fKtbzWeAC4LvkG3DcCvyX2XYo053b2F5nANcU6/5ns+1wvt8vWe+fkr/LLiV/v60HDjfbDuV6UbF+Txe5ngHci/te1WTYJY8N6Ab7DtVkC4wohidgATCPNvbFOpMTEMAlvHyM/IPidVxWvK6HNju3OmT7m2L4vcB55OO0PxTrm4Avmm2Hcp0GLAGuAs4l74udV7LtXgf0Mdfqsy0z7SEl0yZg21rkBPQHbiimuZW8r/IrYA2wHNij2bm9Grsq30dbFcPvovw++HsrmN+pJdvG422Mc1kxfCVwd/H3DVWvW7PD7Qod8LciwC+06n9a0f/cZi9jV+iA/YHJxYfTtI28EYYBzwCrgN1K+g8Abiqm/WCrabYCXiy+QLYq6T8SeKiYZq9m51CHXA8ovhR6teo/npe/gN9jth3Od0Ab/U8q1vscs+10xgH8HXiYvLPyigOejuQETC36PwSMbPVci4rn26pe69XEPOcCcysc1222umzfV6zf1cDQMsP7mm3NM7+5WO93mm2H8hsPrCMXlMa1GrZ/sd6PmG3VuR7Wkh0wpqR/X+CPxbDp5rrRHLvcsQHdZN+hymz7AW8HNi3+n8FGis8dyQn4UDHNjZTsXwNvLF7XZyjz3drVuiqznQ7sUqb/fuQfAFe15N7Ts+3ANturTP++5MJzAt5vrtVn22q6seT9h98AM2m7+Fx1TsA3i2kuKX0tyT8yJvIPNq94je1q91rzcvH5wg7Oaxr5JIaWkyHbKj7vBUwBepfM0+JzBwLfpgjv0dZvDmAo+dee5cDgZi9rV+oqeCN8rBj+8zLDDiiGXd+q/78X/U+s5vm6cwccV6z3mWZb82xfX6z31Wbb6SyPKb649qWNA56O5MTLZ6V9tMw0bT7fq72juuKz22zlufYiF5qWA2PNtiGZ71Cs8+NAb7PtUIZ7FOt2eRvDXwCWmm3VubZ8v3yuzLCW7fZ2c60q02l0gWMDuuG+w8ayLTP+DDZefK46J2BW0X//ap6vK3fVZttq2qtodZKQ2dYk12OKaY83185lS/4xdQEwmvaLz1XlRC6MPlb037qa57Or3WtNJ4rP5B+E51LUQmin+NzGPKsuPtvmc97ZAbgqpbS+dEBKaSn5159BwJ6NXrBXuZZc/1pm2CxgBTA1IvpXOM2VrcbpKdYUj2tL+pltbRxSPM4u6We2VSra9DoZOCOlNKudUTuSU0/Otn9EHB4Rx0Vux3H/KN8eptts5aYCWwN/AZ6P3Ibu14t8y7Wba7ad96ni8fyUUmmbz2ZbuTnks+t2j4gxpQMiYl/yiRJ/L+lttpUZXzw+UmZYS783RMSI4m9z7bxGZWjulakqp+J1mUp+nf5RyTQ9QLnjNDDbDiv2dVvuvzG71WBzrUJETCdf5fPplNKidsbrSE6TgInAgymlRyucRvWzWUR8qjhu/FRE7FTBNP9Nvoro43Vetpf0adSMurDtiscH2xg+BzgQeA253UJVps1cU0prI+JR8qn725DbstnYNE9FxHJgi4gYlFJaUYdl7lIiog9wZPFv6Zes2XZARHwVGAIMB3YD9iHv1JxcMprZVqHYRn9Bbh7muI2MXlVOETEY2BxYllJ6qszzzSkeX9Oxpe/yxpOzLfVoRHw0pXR9ST+32cq9sXh8GrgD2LF0YETMIreN9mzRy2w7ISIGAoeTr4r4aavBZluhlNJzEfF1clNw90XEZeTLiycB7yQ3IfOpkknMtjILi8etywzbpuTv7cn3jTDXzqt7hu47VKaDOW1LvuT6kZRS62JrW9N0WxGxJfBmcsFuVkl/s61C8aPq58ln0o4F3krO41fAn0vGM9cqFNvnGeQzZi/byOgdyamSGlrraVQ/by26l0TETOColNK81iNHxLuAo8hXx7xieL145nMuREFu8L6clv4j2hiu8jqSa6XTDG9jeHdzMvnSz7+klP5W0t9sO+arwAnAl8iF578CB5YUmsBsq/UdYBdyu5grNzJutTn15M/mC8gHNOOBweQi6Xnky5yujIjXl4zrNlu5ccXjp4GBwFvIZ43uQL73w77kdutamG3nvJ+czZUppfmthpltFVJKpwPvJp808kngG+T2y+eTL7V8pmR0s61MS1Hj2IgY1dKz+FH1xJLxRhaP5tp5jciwJ+87VKOer0W3z7Y4U/SX5BuuzUgpPV8y2GyrM4Z8fPYd4DPkH1ZPIR9bpJLxzLVCEdEL+Dm5+dgvVjCJ2b56rQD+A9iVvL8yktwe/XXkJjuuKX64eUlEbEI+rrwypXR+IxfW4vPGRfGY2h1L1epIrj3mtYiILwJfId9t9ohqJy8ezbZESml8SinIBb13k8+suTMi3lDF05htISJ2J5/tfGpK6eZaPGXxWG1O3SpXgJTSiSmla1NKT6eUVqSU7kkpfZp85uNAcluOlXKbfVlLsyVBPsP5mpTSspTSvcC7yO0S79dGExzlmG37ji4ez+vAtGZbIiL+H3ApcCH5wHww+UDjEeCXEfFf1Txd8djTs/0N+bLgSeQzyn8cEaeT7xh/EC+ftbWujelbM9fOa2SGZl4Zt+dWimYhfgHsDfyWXCjtCLMFUkr3F8dnfYAtgS+T9x9mlf4wWM1TVjFud831y+QC5Cdb/TDSUX6/dVEppWdSSt9JKd2RUlpcdLPILTfcQj6r/ROtJvsJ+caen2zw4lp8ZuNnHQxrNZ4q05FcK53mhU4sV5cXEZ8jXyZzH7mR/udajWK2nVAU9P5I/lAeTb6JQguzrUBJcxsPAt+ucLJqc9rY+Bv7xb07Ord43Lekn9ts5Vp2wB9JKf1f6YDizP2WK0x2Lx7NtoMi4nXk9gMfJ7ex3ZrZVigipgHfB/6UUjo2pfRI8aPUHeQfTZ4AvhIRLU1FmG0Fivu8vJN8VdQC8g/9HyNvs/uQmzYBaDmr3Fw7rxEZuu9QmY7k1OOPmYvC88XkK09+Bxze6uxcMNsOSSmtSynNSymdQW5Kak/yTQRbmGsFImIycBJwQUqp3P5XOfX8bO422b6aFM2ntDR599JxY0QcSb7v1TEppScavVwWn+GB4rGt9mgmF49ttWej8trMtShcbU2+OcMjFU6zKflMn8e7c7t4EfEl4CzgHnLheUGZ0cy2BlJKj5EL/FNKbuJktpUZQl7f1wIvRkRq6ciXzgH8pOh3evF/VTmllJaTiypDiuGt9cTP5pYiSOnlU26zlWtZ78VtDG8pTg9sNb7ZVq+tGw22MNvKvaN4vK71gGJd/0nen9+l6G22FUoprU0pnZpS2jmlNDClNCyl9DbyvsHOwErg3mJ0c+28umfovkNlOpjTQ+QrAbYpXq9Kpuk2inX+NfBBcnvEHy7XPq7Z1kTLzeqmtfQw14pNITcH89HS47PiGG2/Ypw5Rb/Div87kpM1tK6vpWnR0uPGliu+f15m+wDYvKRfzZtMsfj88s78gUX7OC+JiKHkS2pWkm82ospdWzy+rcywfYFBwE0ppVUVTvP2VuN0O8UNhX5IvuRz/1ZtOJYy29rZrHhsKY6YbWVWAee30d1ZjHND8X9LkxwdyaknZtueluYgSg/M3WYrN4tc2JgcEf3KDN+heJxbPJptB0TEAPJZpOvJnwHlmG3l+hePY9sY3tJ/dfFotp13BDAA+F1KaU3Rz1w7r1EZmntlqsqpeF1uIr9Ob6pkmu6i2Ge4lHzG80XAEW38sNrCbDtn8+KxdXHfXDduLm0fo7Wc1HZJ8f9c6HBOD5NvOP+aiCh3497umO2rzZ7FY+lx4820vX1AbkO65f/S7+LaSCn1+I58qW0CvtCq/2lF/3ObvYxdrSP/EpnId1AtN3wY+deWVcBuJf0HkD/cEvDBVtNsDbxIvtRxq5L+I8m/yCVgr2ave53y/HaxfrcBozYyrtlWnuv2wPgy/XuRL0lKwI1mW9PMZxTr/InO5kS+bD8Vw0eW9N+qeJ4XS5+rO3TkMxZe8RlAbgdvTpHHcSX93Wary/fiYv3+s1X/t5KLpYuBEWbbqYyPKNbzinbGMdvK83x/sW4LgM1bDXt7sd2uBEabbdXZDivT743Ac8BSYBu32arynEYXODagG+47bCzbMuPPoMy+WGdzAj5UTHMjMKCk/xuL1/WZcu+rrtxVsN32B/6nGOenQK8KnrPHZ1tBrnsAg8r0HwJcXUx7krlWn207080sptu2zLCqcwK+WUxzSen7Aji06H9vJe8Xu46/1sX7qF+Z/gcU74cETK1wXol8JdHGxtuqGPeGatcniifo0SJiEnmnZxxwOfAv8gu5P/lSgakppUVtP0PPUFya0XJ5xnjg38i/pPyj6LcwpfTVVuNfSt7wf0PemX8nsF3R//2p1QYYEV8A/pv8BfJb8pk87wW2IN/Y7Kt0MxFxFPkmQuuAMynfNtLclNKFJdOYbQWKZkx+QD7j8WHyum9CvuxoG/LB/JtTSveVTGO2nRARM8hNb3wypfTTVsOqzikiTgWOJbfDeSnQD/gAub3uL6SUzqrbyjRBkd83yFflPEougEwCDiYfoP8FeFdKaXXJNG6zFYqIceQd623J313/JBf230XekfpwSumSkvHNtkoR8Q9ym7nvTCld0c54ZluB4qq8vwFvIX8e/JH83fVacpMcAXwp5XYyW6Yx2wpExC3kwv095GynkG82uAp4d0rpb63GN9dWuuqxQXfYd+hAtt8gn3QBudmY15OPb1tunnlDmf2yqnKKiCC3dfxe8k3RryjG/QB5H+U9KaXLO7XiDVBNthFxATAdWAicQ/kbqM1MKc1sNY8el22VuV5GLqxdTz6DdgUwgfyj6gjytvtvKaVlrebR43KF6j8P2niOmeRj4MkppYdaDas6p4joTz6zeSr5BLprgInkKwRWAweklG7pwOr2aFW+j2aS911mkt8TADuRi88A304p/WeF803AEymlLcoM24eXb1w4BHgP+QeJliZySClN3+hMGl2976od+cPuAuAp8pvlMfJN39o9C7Undbz8K3pb3dwy0+xNLpY8T97Bv5t8B9be7cznEPIX0VJgOXArcFSz17+JuSbyTo3ZVp/tDsDZ5KZMFpIv31pSrPeMtt7fZtupzFu257Jn23QkJ+CoYrzlxXTXA+9o9rrWKb/9yG0K3k8+C3cN+Uyxq4EjIf9oXGY6t9nKMx5FvrLpUfL3/SLyD897mm2ns31t8f6f314+Zlt1rn2BL5GbgHuh+C57BvgzcKDZdjjXrwG3F5+1q4rPhHNp56xYc33Fus2gix4b8Crfd6g2W14+q7Gt7sJa5AT0KV6vu4vX7/ni9azo7Lqu0FWTbQW5JmCG2Vad68HAL8kn+i0h7+8+A/wdOBro0858elSu1WbbznO0bMuvOPO5ozmR75NyIvlHrlXkY5ZLgNc1O7NXa1fl++jj5H3BucCy4jWYR/6h9k1VzjfRxpnP5B/g2v0crGQenvksSZIkSZIkSao5bzgoSZIkSZIkSao5i8+SJEmSJEmSpJqz+CxJkiRJkiRJqjmLz5IkSZIkSZKkmrP4LEmSJEmSJEmqOYvPkiRJkiRJkqSas/gsSZIkSZIkSao5i8+SJEmSJEmSpJqz+CxJkiRJkiRJqjmLz5IkSZIkSZKkmvv/vddOH1IhuKcAAAAASUVORK5CYII=\n",
      "text/plain": [
       "<Figure size 1728x576 with 2 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "#Plotting the predicted fitness effect for the variants seen more than once\n",
    "plt.close('all')\n",
    "f, (ax1, ax2) = plt.subplots(2, 1, sharey=True, sharex = True, figsize=(24, 8))\n",
    "gs = matplotlib.gridspec.GridSpec(2, 1, width_ratios=[1], height_ratios=[10,0.5])\n",
    "ax1 = plt.subplot(gs[0])\n",
    "ax2 = plt.subplot(gs[1])\n",
    "gs.update(hspace=0.05)\n",
    "\n",
    "nonsense_list = []\n",
    "y_list = []\n",
    "x_list=[]\n",
    "y_list_top20 = []\n",
    "x_list_top20=[]\n",
    "y_list_nonsense = []\n",
    "x_list_nonsense = []\n",
    "x=[]\n",
    "y=[]\n",
    "OBS_EXP_DICT={}\n",
    "top20_lower_err = []\n",
    "top20_upper_err = []\n",
    "missense_lower_err = []\n",
    "missense_upper_err = []\n",
    "nonsense_lower_err = []\n",
    "nonsense_upper_err = []\n",
    "\n",
    "top20 = c1\n",
    "missense = c6\n",
    "nonsense = c10\n",
    "synonymous = c3\n",
    "\n",
    "axisfont=20\n",
    "domainfont = 12\n",
    "titlefont=12\n",
    "axislabelfont=12\n",
    "\n",
    "top_20_ASXL1 = []\n",
    "\n",
    "#ERROR NEEDS TO BE IN THE FORMAT: [tuple(lower_err),tuple(upper_err)]\n",
    "\n",
    "for k, v in ASXL1_missense_s.items():\n",
    "    if k in top_20_ASXL1:\n",
    "        codon_number = int(((re.findall('\\d+', k)))[0])\n",
    "        s_value = float(v[0])*100\n",
    "        x_list_top20.append(codon_number)\n",
    "        y_list_top20.append(s_value)\n",
    "        \n",
    "        s_upper = float(v[3])*100\n",
    "        s_lower = float(v[4])*100\n",
    "        error_upper = s_upper-s_value\n",
    "        error_lower = s_value-s_lower\n",
    "        top20_lower_err.append(error_lower)\n",
    "        top20_upper_err.append(error_upper)\n",
    "\n",
    "    if k not in top_20_ASXL1:\n",
    "        codon_number = int(((re.findall('\\d+', k)))[0])\n",
    "        s_value = float(v[0])*100\n",
    "        x_list.append(codon_number)\n",
    "        y_list.append(s_value)\n",
    "        \n",
    "        s_upper = float(v[3])*100\n",
    "        s_lower = float(v[4])*100\n",
    "        error_upper = s_upper-s_value\n",
    "        error_lower = s_value-s_lower\n",
    "        missense_lower_err.append(error_lower)\n",
    "        missense_upper_err.append(error_upper)\n",
    "\n",
    "    missense_err = [tuple(missense_lower_err), tuple(missense_upper_err)]   \n",
    "    \n",
    "for k, v in ASXL1_nonsense_s.items():\n",
    "    if k in top_20_ASXL1:\n",
    "        codon_number = int(((re.findall('\\d+', k)))[0])\n",
    "        s_value = float(v[0])*100\n",
    "        x_list_top20.append(codon_number)\n",
    "        y_list_top20.append(s_value)\n",
    "        \n",
    "        s_upper = float(v[3])*100\n",
    "        s_lower = float(v[4])*100\n",
    "        error_upper = s_upper-s_value\n",
    "        error_lower = s_value-s_lower\n",
    "        top20_lower_err.append(error_lower)\n",
    "        top20_upper_err.append(error_upper)\n",
    "    \n",
    "    top20_err = [tuple(top20_lower_err), tuple(top20_upper_err)]\n",
    "        \n",
    "    if k not in top_20_ASXL1:\n",
    "        codon_number = int(((re.findall('\\d+', k)))[0])\n",
    "        s_value = float(v[0])*100\n",
    "        x_list_nonsense.append(codon_number)\n",
    "        y_list_nonsense.append(s_value)\n",
    "        \n",
    "        s_upper = float(v[3])*100\n",
    "        s_lower = float(v[4])*100\n",
    "        error_upper = s_upper-s_value\n",
    "        error_lower = s_value-s_lower\n",
    "        nonsense_lower_err.append(error_lower)\n",
    "        nonsense_upper_err.append(error_upper)\n",
    "\n",
    "    nonsense_err = [tuple(nonsense_lower_err), tuple(nonsense_upper_err)]        \n",
    "    \n",
    "msize = 12\n",
    "medgewidth = 2\n",
    "linew = 3\n",
    "csize = 4\n",
    "cthick = 3\n",
    "\n",
    "markers1, caps1, bars1 = ax1.errorbar(x_list, y_list, yerr = missense_err, fmt = 'o', ecolor = missense, elinewidth = linew, \\\n",
    "             capsize = csize, capthick = cthick, markersize = msize, markeredgewidth = medgewidth, \\\n",
    "             markeredgecolor = missense, markerfacecolor = missense, zorder = 1)    \n",
    "\n",
    "markers2, caps2, bars2 = ax1.errorbar(x_list_nonsense, y_list_nonsense, yerr = nonsense_err, fmt = 'o', ecolor = nonsense, elinewidth = linew, \\\n",
    "             capsize = csize, capthick = cthick, markersize = msize, markeredgewidth = medgewidth, \\\n",
    "             markeredgecolor = nonsense, markerfacecolor = nonsense, zorder = 2)    \n",
    "\n",
    "markers3, caps3, bars3 = ax1.errorbar(x_list_top20, y_list_top20, yerr = top20_err, fmt = 'o', ecolor = top20, elinewidth = linew, \\\n",
    "             capsize = csize, capthick = cthick, markersize = msize, markeredgewidth = medgewidth, \\\n",
    "             markeredgecolor = top20, markerfacecolor = top20, zorder = 3)    \n",
    "\n",
    "# loop through bars and caps and set the alpha value\n",
    "a = 0.7\n",
    "[bar.set_alpha(a) for bar in bars1]\n",
    "[cap.set_alpha(a) for cap in caps1]\n",
    "[bar.set_alpha(a) for bar in bars2]\n",
    "[cap.set_alpha(a) for cap in caps2]\n",
    "[bar.set_alpha(a) for bar in bars3]\n",
    "[cap.set_alpha(a) for cap in caps3]\n",
    "\n",
    "#Only show the required axis lines\n",
    "ax1.spines['top'].set_visible(False)\n",
    "ax1.spines['right'].set_visible(False)\n",
    "\n",
    "for axis in ['bottom','left']:\n",
    "    ax1.spines[axis].set_linewidth(2)\n",
    "\n",
    "for axis in ['bottom','left']:\n",
    "    ax1.spines[axis].set_color(grey3)\n",
    "    \n",
    "ax1.set_ylabel('fitness effect (s)', fontsize = axisfont, labelpad = 10)\n",
    "ax1.set_title('Fitness effects across ASXL1', fontsize = 18)\n",
    "\n",
    "# ax1.set_yscale('log')\n",
    "\n",
    "# # STACKED BAR CHART\n",
    "# define the exons + create data\n",
    "exons = [{'1':(19-1)+1, '2': (47-20)+1, '3': (84-48)+1, '4':(125-85)+1, '5':(157-126)+1, '6': (189-158)+1,\\\n",
    "          '7': (240-190)+1,'8': (294-241)+1, '9': (327-295)+1, '10': (362-328)+1, '11': (573-363)+1, \\\n",
    "          '12': (1541-574)+1}]\n",
    "exon_orders = [['1', '2', '3', '4', '5', '6', '7', '8', '9', '10', '11', '12']]\n",
    "colors2 = [c4, c0, c3, c1, c5, c6,c7, c8, c9, c10, c2,c4]\n",
    "names = exons[0].keys()\n",
    "values = np.array([[data[name] for name in order] for data,order in zip(exons, exon_orders)])\n",
    "lefts = np.insert(np.cumsum(values, axis=1),0,0, axis=1)[:, :-1]\n",
    "orders = np.array(exon_orders)\n",
    "\n",
    "# #plot the bar chart\n",
    "for name, color in zip(names, colors2):\n",
    "    idx = np.where(orders == name)\n",
    "    value = values[idx]\n",
    "    left = lefts[idx]\n",
    "    ax2.bar(x=left, height=-10, width=value, bottom=-5, color=color, orientation = 'horizontal', label = name, edgecolor='k', linewidth=1, tick_label = name)\n",
    "    ax2.text(left + (value/2.), (-0.26), name, ha='center', va='center', fontsize = 14)\n",
    "    \n",
    "# # CONFIGURING THE GRAPH APPEARANCE\n",
    "# #Set the x and y axis limits\n",
    "ax1.set_xlim(0, 1542)\n",
    "ax1.set_ylim(-0.01, 30)\n",
    "ax2.set_xlim(0, 1542)\n",
    "ax2.set_ylim(-0.5, 0)\n",
    "\n",
    "#x-axis ticks\n",
    "x1_major_ticks = []\n",
    "x1_major_tick_labels = []\n",
    "x2_major_ticks = [0, 100, 200, 300, 400, 500, 600, 700, 800, 900, 1000, 1100, 1200, 1300, 1400, 1541]\n",
    "x2_major_tick_labels = [\"0\", \"100\", \"200\", \"300\", \"400\", \"500\", \"600\", \"700\", \"800\", \"900\", \"1000\", \"1100\", \"1200\", \"1300\", \"1400\", \"1541\"]\n",
    "ax1.set_xticks(x1_major_ticks)\n",
    "ax1.set_xticklabels(x1_major_tick_labels, fontsize = axisfont)\n",
    "ax2.set_xticks(x2_major_ticks)\n",
    "ax2.set_xticklabels(x2_major_tick_labels, fontsize = axisfont)\n",
    "\n",
    "#y-axis ticks\n",
    "y1_major_ticks = [0, 5, 10, 15, 20, 25, 30]\n",
    "y1_major_tick_labels = [\"0\", \"5\", \"10\", \"15\", \"20\", \"25\", \"30\"]\n",
    "ax1.set_yticks(y1_major_ticks)\n",
    "ax1.set_yticklabels(y1_major_tick_labels, fontsize = axisfont)\n",
    "\n",
    "#set panel legends\n",
    "legend_marker_size = 2\n",
    "legend_elements = [Line2D([0], [0], marker = '_', color = missense, alpha = 1.0, markersize = legend_marker_size, \\\n",
    "                          lw=3, label='missense'),\n",
    "                   Line2D([0], [0], marker = '_', color=nonsense, alpha=1.0, markersize = legend_marker_size, \\\n",
    "                          lw=3, label='nonsense'),\n",
    "                  Line2D([0], [0], marker = '_', color=top20, alpha=1.0, markersize = legend_marker_size, \\\n",
    "                          lw=3, label='top 20 observed variant')]\n",
    "    \n",
    "ax1.legend(ncol=1, handles=legend_elements, loc='upper left', frameon=0, fontsize = 20)\n",
    "\n",
    "ax1.grid(which='major', linestyle=':', linewidth='0.7', color=grey3, zorder = 0)\n",
    "\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 273,
   "metadata": {
    "code_folding": [
     0
    ]
   },
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAABawAAAHvCAYAAAC43ZjsAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDIuMi4zLCBodHRwOi8vbWF0cGxvdGxpYi5vcmcvIxREBQAAIABJREFUeJzs3Xd4VFX+x/H3SUJCT2ihN1HACisKLKAgTUEpIorCCgjqriiIBduqCag/UVQEaS4uigrqihRFISA1sID0Lj3UUEIIEEJIO78/7szsTGYSQg/6eT1PnknOPffc771n7jz6ncP3GmstIiIiIiIiIiIiIiJXWtCVDkBEREREREREREREBJSwFhEREREREREREZF8QglrEREREREREREREckXlLAWERERERERERERkXxBCWsRERERERERERERyReUsBYRERERERERERGRfEEJaxEREZE/IGNMiDHGGmM+u9Kx5EfGmL7GmC3GmDTXdarkam9hjFlmjEl2tf/tSscqIiIiIvJnooS1iIiIyFXEGNPMlUjN6adhLvsGGWOijTHtL2fM+Y0xphUwHNgAPAk8CiQaY0oDk4GCwPOu9sWXKIZbXXNR5VKML54vbQ667otXztKvhzFmsTHmkDEm1Rizzxgz1xgz0BgT6tX3P8aYLGNM0wDjFDHGbDfGHDHGRHqNbY0xU/MQ78PGmM+NMeuMMRneX6SIiIiI/JmEXOkAREREROS8fAP8EqB9O4C1NsMYUwjI8NoWBEQB/wZ+vOQR5l+tXK+PWWtPuBuNMc2A4sAb1tpLfX1uxZmLX4E9l/hYf1b3AWWBHUAvYHAO/b4DOgGxwBAgCagCNAD+CQwF0lx9+wBNgc+NMbdYa5O9xnkfqAE8aK09fB7xPoPzvlgL7ASuO48xRERERK56SliLiIiIXJ1WWWu/zq2DtTb1cgVzlSkHZHonq73aARIvczxXHWOMAYpkS9jmN72BrcCrwA/GmDustbHeHYwxDXCS1ZOstQ9mH8AYUw7wnKO1NsEY8xTwA06Cuo+rXwvgKeBba+2k84y3G7Df9WXTGJSwFhERkT8plQQRERER+QPKXsPaGHMtkO7a3NurhEhG9v7GmCbGmFhjTIoxJsEY8y9jTJEAx6hojBljjNnrqgW93/V36Wz9ShljhhljdrrKLRw1xqwwxjyfrV9PY8xyY0ySMeaUMWaHMeZrY0ypPJ7z3caY2caY467jrDXGPOm1/VpjjMUp9RHsdQ1+Ncbsw1l5DhDrfW1c+xY0xrxujNnkGvuYMeZHY0ydAHEYY8zfjTG/uWphn3SVeYhybX8bGJvtWN5zVcgYM8g4NbZTXMdab4zJaYWw97FDXHHGuspbpBljdhtjRhpjSuawz4PGmAWu65ZijPndNV8hru0tXfE9apza35uBM0B/rzGaua7jCdcYK40xPQMc62ZjzA+u98oZY0y8cUpvtPHqc97n7zVGeaANMB6YDhzFSWBn504Kzwk0jrX2oLU2I1vbZJx/4fAP49Q8Lw6MAw7hrJI+L9ba3dmPJSIiIvJnpBXWIiIiIlenwtkTw8AZa+3JHPofBHrgJPDm87/kbFa2fvWA+13bvwaaA0/glBbp4+5kjKkO/BcIdvV1lzB4CrjLGHO71wrmycBfgTHAOqAIcD3QDPjINd5jOEm/BcCbwGmcsgxtgdI4CcccGWfV60hXTG8BKcDdwKfGmOrW2ldd1+BRV4wNXdcDIB6nFMi9OEnNt3BW5ma5xg4FZuGUiPgSp/51Cdd1+a8xpom1drWrr8FJZnYBlgDv4JSYuB54ABgIfI+zmtv7WOAq5+K6Tt2BL1xjFHBd2+a5XQOXgsALOCuApwKngPo4tbobu+bF/cUFxpj3gJeAjcCHOEnXGq5Y/4nX6mLXuCVw5vsgsNs1Rkdgkus6DnEd8xGcshnVrbXuRH0ZYC6Q6TrHPThze7srxhkX4fzdegIG+Mpam2aMmQj0Msb0zXaP7HC9PmSM+dZam5TH8Z/Bef+Ow3nPVQE6WGtzfZ+KiIiIyNkZa+2VjkFERERE8sg4dZbn5bD5O2vtw65+ITgrqv9trX08pzavcd3bsoAG1toVXtticJJzEdba0662n3GS27daaw949W2Ak8CLsta+7VrVexT4xFrbL5fz+hFoApSx1mbm8XK4962Ek3j8zlrbPdu2kcDfgRrWWneC9WvgYWttSLa+j+OsfL7DWrvIq30A8B7Q2lr7q1d7BM6DG3+31rZ0tXUFJuB8MdDLWpvl1T/I/XdOx3JtOwHMt9ae88MxjTFBQJh7nrza/46TCH7AtUIYY0wjnIdK/grcZ60949XfWNf/KBhjWgKzceaxtrU2watfAWAXUBi4wVp70NUehvPlw+3AddbancaYTjiJdE8MOZzDeZ+/1xhbgT1e81IPWAE8aa0dm63vLzirsU/hvHeXuX7mWmtTcjlGO/5XC/5La22PAH3c99U0a23Hc4h/DM77trK1dl9e9xMRERH5I1BJEBEREZGr079wHh7o/fP2RRh3kXey2mUuEApUBXAlodvgrOBNM8aUdv/gJI53Aa1d+6bgJOwaGmOq5nLc40AxoK1rlfK5eNAV3zjvWFzx/ISzCrzFOY7p7W84K5DXZBs7BKeURFNXghacOsQWeNE7WQ2Q/e9cHAduNsbceK6BWmuzvL5UCDbGRLhinevq0sCrezfX6yveyWrXOIFWtXzhnax2uR2oCHzmTla79j8DfIDz/xvuxPNx12tbY0yxXE7jvM8fwBhzJ86K7PFe8awE1uM8fDG7jsCzwCacVdyv47xv4o0x/QP0dzvK//6FwvTziVVERERE/ClhLSIiInJ12mat/TXbz4aLMO7OAG3uMgfuWtK1ccot/B04EuCnBlAWPA9+fB6oC8QZYzYYY4YbY+7Kdoy3gX04K1YPG2MmGWN6G2OK5iHm612v8wLE4i4zUTYP4+SkNnBTgLGP4JSuCAHc9aGvA/YFSOyei2dxSmVsMMZsN8aMNca0y2si3xjzsDFmOU5ZlWOuON1lR0p4db0OpzzHujzGtTVAW3XX68YA29zvx2tcr3OBiTilUI4aYxYbY6KMMbWz7XdB5+8aPw1Ya5y65dcap4b7LJwvTm7w7mytTbPWDrfW1scpDXMnzor6YGCoMSbQwxgL45QsSQT2u/pF5DE+EREREcmFaliLiIiIiLfcynGYbK/jcepcB+IppWCtHWGMmYJTI/pO4CGgrzFmgrX2b64+W1yJy5Y4q6GbAp8B0caYO621u/IQVzfgcA59tufQnhdBwBpgQC59Er1iuaCae9baycaYBTj1u5virJ5/HJhvjGntXYM6O2PMQzg1tJcC/XC+BEjFWYH+M74LVs51JXug8hh5HsO1arub6+GJbYA7cOpnv+GqLT3a1e9Czr840BnnfNfm0K0X8GIOMaYAsTgPw1wA/IKTAP8+W9fBOAn/LsBJV7+hwGO5XgQREREROSslrEVERET+PC7Ww0u2u8Yq4F3TOdcDW7sfp4zJv1x1fSfgJC8/dD+w0FVG4mfXD8aY9sA0oD/OqtucbHO9HslrPOdoG1AGmJNDqQxvW3BKXpQ+yyrrXMdxPbzvK+Ar18riITgPPbwPmJLLro/iJJbvcq1uB8AYc1MOsbYEbgZW5RZPLtwPLQxUvsO9ktln1b61dj1OeY73XeVlfsNJAI/26nO+5/8wTj3tl7Mf1+U54FFjzKu5Jb5dlrpeK3o3uurIPwN8b639j6vtc+AxY8z31tpfzjKuiIiIiORCJUFERERE/iRcDzNM5X/lK853nEM45RUeNMbcnn27cZRx/V7YGFMo2/4ZOAlL3LG46ixnt8q7Ty6+wykBMcgYUzBAPBHGmNCzjJGbL3GSlgGT5sYY73IjE3BWHb+fvYRFtr+TXa8ls/UJMcaEe7e5kuRrAvUPIBOnrrLnv/Ndx309QN+Jrtd3s1+fcyi/sRynJEZvY0yk1/6hOKuYs3A9mNAYUzL7uNbaRCAOKGqMCb0I598bSAA+tNZOyv4DjAMicRLfGGNqGWOuyWEs90MSN3mdV1Hgc5wyK328+j6Hs5r9XyoNIiIiInJhtMJaRERE5M9lKXC3MeYlYC+Q6V4leo7+DiwCFhljvgRW4/y3ZXWcRN+/cepS3wD86ioJshGnpvINwFM4q3MXu8aba4w5glOOYS9OYvIxnIRnTmVHALDW7jbGPAOMATYZY74G9uCsir4F6ADUxEkono+PcFYiDzXGtATm45SBqIJTvuQkTtkKgG+BTq7YaxpjfsJ5iGBNV986rn6/4ayyft2V3D/F/x5YuccY8yNOkvYwTg3op3DKjvx8llgnuc53rjHmKyAMuB/wS+Rba/9rjPkQZ+XyCmPM98BBnDl80BjzF2ttcvb9so2RYYzpi1MyY7kxZqzrXB4G6gODrLXulc69gGdc74XtQAZwl+u6TLTWprm+uDiv83etIq+P8wDInErbTAM+xUlsTwH+AkwwxswDFuK8R4oCDXFKixwH3vLa/0OgGvCA9wp6a+1xY8yT5FwapKYxJtCXBuAk108bY5rilEkBuNX12s8YcwLIstb+X07nLiIiIvJHooS1iIiIyJ/LP4AROCtui+GsyD3nhLUrSXwr8ArQHufhgyk4yeapOIlTgN04ta6b4SROw3CSgp8C73mVrRgJPOiKryTOgx5XAX2stfPzEM9YY8xmnFW9TwHhOCttfwf+ibMi9ry4Eqn34JSB+Bsw0LXpAM4XAOO9+lpjTBfgaZwEbRTONd6JVx1ka+0uY8zjODWcRwMFcJL8TwPDcZK4rYAiQDxOcvVda+3Bs8T6tWsV8LM4ydVEnBXObxCgvre19kVjzGrXcV/GWR2+F5iOsxo/L9dnijGmFc576mWc/8fYDPSy1n7u1XUuzhcI7YDyOAnrXTgJ8xGuPicv4Px7u14n5xJrgqs29T3GmAo4D+p8GecLid44q6/BeY+OA953J9yNMa2BJ4FvrLV+x7DWzjDGjAN6BSgNcj2+iW9vY3AekNkK573qzV03PRNQwlpERET+FMzZy/BdGcaYUjj/U3MvTl29ijj/1HM9zj/D+9xam+XV/zqc1Sx34zwApSzOCp6lwMfW2nm5HKsHzn+k34DzH4OrgQ+stdMv/pmJiIiIiIiIiIiISCD5OWH9D5zVJvE4Kx/24CShO+GsmPkBeND94BtjzLc4T+nehPPPUxOBWjgrfoKBZ621wwMc5wOcVR37cFYCheL8E8aSQF9r7Yjs+4iIiIiIiIiIiIjIxZefE9bNcf4J4M/ZVlKXw6n5VxnobK39wdXeE1jrfsq8V/+mwGycGoHVrLXxXtsa4dRN3AHcbq095mqvBqx0Hb+2tTbukpykiIiIiIiIiIiIiHgEnb3LlWGtnWut/ck7We1qP4hT5w2cWoju9i+yJ6td7QtwHowTCjTKtvkfrtd33Mlq1z5xOHUUw/B/YIqIiIiIiIiIiIiIXAL5NmF9Fumu14wL7N/c9TozwD4zsvURERERERERERERkUvoqktYG2NCcJ5CD4ETzdn7V8V5yngKsNCrvQjOgxyTvcuEeNnmeq15tmOMHj3ajh49On/WVhERERERERERERG5Slx1CWtgMHAT8Iu1Nia3jsaYMGACTmmPaO+yHzgPbgQ4nsPu7vaIXMZ/0hizwrPD8eN29erVNj093a5YscICdsWKFfbMmTN27dq19tixY3b79u123759Nj4+3m7ZssWeOHHCrl692mZkZPjts379epuYmGh37Nhh9+7daw8ePGh///13e/LkSbty5UqbmZnps8/p06fthg0b7NGjR+3OnTvtnj177KFDh+zmzZttcnKyXbFihc3KyvLZJyUlxW7cuNEeOXLE7tq1y+7evdsePnzYbtq0yZ46dcrTd/ny5Z59Tp06ZTdt2mQPHz5sd+/ebXft2mWPHDliN27caFNSUnzGdx8vOTnZbt682R46dMju2bPH7ty50x49etRu2LDBnj592mefzMxMu3LlSnvy5En7+++/24MHD9q9e/faHTt22MTERLt+/Xp75swZn30yMjLs6tWr7YkTJ+yWLVtsfHy83bdvn92+fbs9duyYXbt2rd8+6enpdvXq1fb48eN269at9sCBA3b//v1227ZtNikpya5Zs8amp6f7nHtaWppnLrdt22b3799vDxw4YLdu3Wq959+9z/Lly+2ZM2fsunXrPHOZff5XrVrlN5epqal2/fr1PnOZff6zz6V7/hMSEjxzeejQIbtp0ybP/Ftrc5z/uLg4GxcXZ48cOeI3/96vycnJfvOfkJBgN2zYYFNSUnyuV1ZWll25cmWu85+amurzPsvMzLSrVq2yJ06cCDj/69aty3H+jx8/brds2WIPHDjgN/9paWl+879mzRqblJRkt27davfv32/3799vt27d6jP/3vt4z7/7Xj5w4IDdsmWLZ/4D3cve8+99L3vPv/d1S01NzfVezj7/y5cvtykpKX7z776X3fOffS7d93L2+Q90L7vfO+75P3TokN/8Z7+X3fPvfS97z//69et95t99/7vn3/tezj7/3vdYbvdyoPl37+Oe/9zuZX2W67Ncn+X6LNdnuT7L9Vmuz3J9luuzXJ/l+izXZ7k+y/+Mn+VcQfn2oYuBGGP6AcOA34HG1trEXPoGA98ADwLfAY9Yr5M1xlQA9gP7rbWVAuxfAEgDzlhrC+YWl3t19VNPPXXO5yQiIiIiIiIiIiKSz5grdeCrZoW1MeZpnGT1JuCuPCSrv8ZJVv8H+Jv1z8y7V1CHE9jZVmCLiIiIiIiIiIiIyEV0VSSsjTH9gRHABpxk9cFc+obgrKx+GJgIdLXW+j2c0Vp7CmeFdVFjTPkAQ13net16geGLiIiIiIiIiIiISB7k+4S1MeZlYCiwBidZfTiXvqHAJJyV1V8Cj1prM3MZfq7r9Z4A29pk6yMiIiIiIiIiIiIil1C+TlgbY97AecjiSqCFtTYhl75hwBSgA/Bv4DFrbdZZDjHG9fpPY0wJr7GqAU8DZ4DPzzd+EREREREREREREcm7kCsdQE6MMT2AQUAmEAv0M8av1nectfYL1+9jgLZAAk6pjzcD9J9vrZ3v/sNa+19jzEfA88A6Y8wkIBToApQE+lpr4y7eWYmIiIiIiIiIiIhITvJtwhqo7noNBvrn0GcB8EW2/qWBN3MZd773H9baF4wx64BngCeBLGAVMMRaO/2coxYRERERERERERGR85JvE9bW2mgg+hz6N7uAY40Hxp/v/iIiIiIiIiIiIiJy4fJ1DWsRERERERERERER+fNQwlpERERERERERERE8gUlrEVEREREREREREQkX1DCWkRERERERERERETyBSWsRURERERERERERCRfUMJaRERERERERERERPIFJaxFREREREREREREJF9QwlpERERERERERERE8gUlrEVEREREREREREQkX1DCWq560dHRGGOYP3/+lQ5FRERERERERERELoAS1iIiIiIiIiIiIiKSLyhhLVe9Z555hs2bN1O/fv0rHYqIiIiIiIiIiIhcgJArHYDIhSpdujSlS5e+0mGIiIiIiIiIiIjIBdIKa7ki4uLiMMbQs2dPduzYQefOnSlVqhTFihWjdevWbNiwAYAjR47w5JNPUr58eQoWLMjtt9/OvHnzfMbKqYZ1bGws7dq1o1KlSoSFhVGuXDkaNmzIwIEDffodOnSIF198kVq1alGkSBEiIiKoVasWPXv2ZOfOnX6xx8TE0LZtW0qXLk1YWBg1atRgwIABJCUl+fWtVq0a1apVIyUlhQEDBlClShXCwsK49tpree+997DW+u3z448/0qJFC8qXL09YWBgVKlSgadOmjBo1yq9vYmIir776Ktdffz2FChUiPDycFi1aMGvWrLPOgYiIiIiIiIiISH6jFdZyRcXFxdGgQQOuv/56evbsSVxcHFOmTKFZs2YsWbKEe+65h+LFi9OlSxcSExP59ttvadOmDVu3bqVKlSo5jjtz5kzuvfdeihcvTvv27alYsSKJiYls3ryZUaNGERUVBUBKSgqNGzdmx44dtGrVinbt2mGtZffu3UybNo3OnTtzzTXXeMYdNGgQUVFRlCxZkvvuu4/IyEjWrVvHBx98wC+//MKSJUsoXry4Tyzp6em0bt2aAwcO0KZNG0JCQpg6dSqvvPIKqampnlgA/vWvf/H3v/+dcuXK0a5dO0qXLs3hw4dZt24dn3/+OX369PH03b17N82aNSMuLo477riDe+65h1OnTjF9+nTuuecePv30U5544omLNVUiIiIiIiIiIiKXnBLWckUtWLCAt99+m3/+85+etrfeeos333yTBg0a8NBDDzFq1CiCgpx/DNCqVSu6d+/O0KFDGTp0aI7jjh07lqysLObPn0+dOnV8tiUkJHh+nzNnDjt27KB///5+46WlpXHmzBnP3/PmzSMqKoq//vWv/PLLL0RERHi2ffHFFzz22GNERUX5jXPgwAHq1KnD7NmzKVSoEABRUVHUrFmToUOH8tprr1GgQAEAPv30U0JDQ1m7di2RkZE5xg3Qo0cPdu/ezTfffMPDDz/saU9KSqJZs2b069eP9u3bU7Zs2Ryvk4iIiIiIiIiISH6ihHU+0nVp+pUOIc8mNixwUcapVq0ar7zyik9bjx49ePPNNzlz5gxDhgzxJKsBunbtSq9evVizZk2exncniL0FqncdqF9oaCihoaGev4cPHw44yXDvZDVAz549GTZsGBMmTAiYSB8+fLjPMSIjI+nQoQNffvklW7Zs4aabbvJsCwkJ8SSwc4p77dq1LFiwgM6dO/skqwEiIiIYOHAgHTt25IcffvBZlS0iIiIiIiIiIpKfKWEtV1TdunUJDg72aatQoQIANWvWpFixYj7bgoODKVu2LPv27ct13G7dujF58mQaNGhAly5duOuuu2jcuDGVKlXy6de0aVMqVqzI4MGDWbVqFW3btqVx48YB41qyZAkFChTg+++/5/vvv/c7ZlpaGkeOHOHo0aOUKlXK0x4eHs61117r179y5coAHDt2zCfuF154gRtvvJEuXbrQtGlTGjduTJkyZfxiATh+/DjR0dF+Yx85cgSAzZs353aZRERERERERERE8hUlrOWKCg8P92sLCQnJcZt7e3p67qvRO3XqxPTp0/nwww8ZN24cn376KQD16tXj3XffpVWrVgAUL16cpUuXEhUVxY8//khMTAzgrGbu06cPr7/+ume189GjR8nIyPB7aGN2ycnJPgnr7Kuxs59nZmamp+3555+ndOnSjBo1iuHDh/Pxxx9jjKFp06YMGTKE2267zRMLwOzZs5k9e3ausYiIiIiIiIiIiFwtlLDORy5WmQ1x3Hvvvdx7772cOnWKZcuWMX36dEaPHs19993H6tWrueGGGwCoVKkS//73v7HWsmnTJubOncvIkSMZNGgQWVlZvPXWW4CTQM/KyiIxMfGSxt29e3e6d+9OUlIS//3vf5kyZQrjxo3j7rvvZvPmzURGRnqS+cOGDaNfv36XNB4REREREREREZHLJejsXUSubkWKFKF58+Z89NFHvPbaa6SlpTFjxgy/fsYYbrzxRvr27etZtTx16lTP9oYNG3Ls2DE2btx4WeKOiIigbdu2jB07lp49e5KYmEhsbKwnFsDzt4iIiIiIiIiIyB+BEtbyhzRnzhxOnz7t137o0CEAChcuDMCGDRuIi4s7az+A5557DoAnnniCAwcO+O1z6tQpli5dekFxz5w5k4yMDL/2w4cP+8Rz2223cccddzB58mTGjRsXcKz169d79hMREREREREREbkaqCSI/CG98MILxMXF0axZM6pVq0ZoaCgrV65k7ty5VK1alYcffhiAX3/9leeff55GjRpRu3ZtIiMj2bdvH9OmTSMoKIgBAwZ4xmzRogWDBw/m1Vdf5brrrqNt27ZUr16d5ORkdu/ezYIFC2jSpAkzZ84877gffvhhChYsSJMmTahWrRrWWmJjY1m+fDn16tWjZcuWnr4TJ06kefPm9O7dm+HDh9OgQQMiIiLYt28f69atY8OGDSxZsoTIyMjzv5AiIiIiIiIiIiKXkRLW8of02muvMWXKFFasWMGvv/5KUFAQVapU4bXXXqN///6UKFECgLvvvpv+/fuzcOFCpk2bxokTJyhfvjytWrXyJLK9vfzyyzRu3Jjhw4ezaNEipk2bRnh4OBUrVuTJJ5+ka9euFxT34MGDiYmJYdWqVfzyyy8ULFiQqlWr8t577/HUU095HgAJTu3tlStX8sknn/DDDz8wYcIEMjMzKVeuHDfccAN9+/bl5ptvvqB4RERERERERERELidjrb3SMVz1Ro8ebQGeeuqpKx2KiIiIiIiIiIiIyIUyV+rAqmEtIiIiIiIiIiIiIvmCEtYiIiIiIiIiIiIiki8oYS0iIiIiIiIiIiIi+YIS1iIiIiIiIiIiIiKSLyhhLSIiIiIiIiIiIiL5ghLWIiIiIiIiIiIiIpIvKGEtIiIiIiIiIiIiIvmCEtYiIiIiIiIiIiIiki8oYS0iIiIiIiIiIiIi+YIS1iIiIiIiIiIiIiKSL+TbhLUxppQx5nFjzBRjzHZjzGljzHFjzCJjTG9jTFC2/gWMMc8aYz43xqwxxqQZY6wx5vFcjtHT1Senn39c+jMVEREREREREREREYCQKx1ALh4ERgPxwDxgD1AW6AR8BrQxxjxorbWu/kWAj12/HwIOApXzeKxpwJoA7SvOL3QREREREREREREROVf5OWG9FWgP/GytzXI3GmNeA34DHsBJXv/g2pQCtAXWWGvjjTHRQFQejzXVWvvFRYpbRERERERERERERM5Dvi0JYq2da639yTtZ7Wo/CIxx/dnMqz3NWjvDWht/GcMUERERERERERERkYskP6+wzk266zXjIo1X1xjTHygI7AfmWWv3XaSxRURERERERERERCQPrrqEtTEmBOju+nPmRRr22Wx/ZxpjPgP6W2tTL9IxRERERERERERERCQXwdHR0Vc6hnMycODA94GOwC/W2kG59GuGUzLkp+jo6FU59AkHVgLPAa8AI4FlwG1AK6BGdHT0D4H2BTDGPDlw4MBP77333goANWsBq029AAAgAElEQVTWZPPmzZQuXZrVq1dToUIFVqxYQalSpdi4cSNhYWHs37+flJQUTp06xYEDBwgNDWXTpk2UKVOGVatW+eyzefNmChQowIEDBzh16hSnT59m3759hIWFsWHDBsqWLeuzT4kSJdiyZQvBwcEcPHiQkydPkpqayt69eylUqBDr1q2jXLlyrFy50rNPREQEW7duxRjD4cOHOXHiBGlpaezevZvChQuzbt06KlSowPLly6lYsSIrVqwgPDyc7du3Y60lISGBY8eOkZGRQVxcHEWLFmXt2rWe8d3HK168ODt27CArK4vExESOHj1KVlYWO3fupFixYj77lC1bltWrV1OsWDF27dpFRkYGx44dIyEhAYDt27cTERHBmjVrPPtERkaydu1aihQpQlxcHOnp6Rw/fpzDhw9jjGHr1q2UKFHCZ58yZcqwbt06ChcuzJ49ezhz5gwnTpzg0KFDBAcH8/vvv1OqVClWrVrlOffSpUuzYcMGwsLC2LdvH6mpqSQnJxMfH09ISIhn/t37LF++nNKlS7Np0ybPXGaf/40bNxIZGekzlyVLlmTz5s2EhIR45jL7/GefS/f8BwUFcejQIU6cOMGZM2fYs2ePZ/7Lly+f4/wfOXKEpKQk0tPT/eZ/xYoVntfixYv7zX9mZiY7d+6kaNGirFmzxnO9ypUrx6pVq3Kd/+LFi3vmZfny5ZQrV441a9ZQpEiRgPO/bdu2HOe/cOHC7N69m7S0NL/5L1mypM99WaZMGdavX0/BggXZs2cPqampnDx5koMHD/rMv/c+3vPvvpeTk5M5cOAABQoUYPPmzQHvZe/5976Xved/5cqVnutWsmRJfv/99xzv5fXr1/vM//LlywPOv/teds9/9rl038uAz/wHupfd7x33/GdlZXH06FGf+c9+L7vn3/te9p7/HTt2EB4e7jOXZcuW9cy/97185MgRn/lfvXq15x6LjIzM8V4OCgrym3/3Pu7537t3b473sj7L9Vmuz3J9luuzXJ/l+izXZ7k+y/VZrs9yfZbrs1yf5X/Gz/Ly5csPzCkneqkZa+2VOvY5M8b0A4YBvwONrbWJufSNxnno4hPW2s/O8TiVgbVACaCutXZtbv1Hjx5tAZ566qlzOcyfWlxcHNWrV6dHjx5ER0fzyiuv8Ouvv5KcnMxNN91EdHQ09913n88+Z86cYejQoUycOJHt27cTEhJCnTp16Nu3Lw899NAFj5+WlsaYMWP44osv2LVrF2fOnCEyMtJzjJYtW/r0//333xk8eDBz5szh8OHDRERE0KJFC6KioqhVq5ZP3549ezJ+/Hh27dpFTEwMI0aMYNu2bYSHh9OhQweGDBlCeHi4zz7r1q3j3XffZcmSJcTHx1O8eHEqV67MnXfeyZAhQyhQoICnb0ZGBv/617/48ssv2bRpExkZGdSqVYvevXvTp08fgoLybbl6ERERERERERHJf8yVOvBVUxLEGPM0TrJ6E9Ait2T1hbLW7jXG/AJ0A+7ESV7LJbB7927q16/PNddcw6OPPkpiYiLfffcdHTp04Ndff+Wuu+4CnGTy3XffzYIFC6hduzZPP/00KSkpTJo0iS5durBmzRr+7//+77zHByep/M0333DTTTfRvXt3ChUqxIEDB1i0aBEzZ870SVjPnDmTTp06kZ6eTrt27bj22mvZt28fkydP5ueff2bevHnceuutfvG89NJLxMTE0K5dO1q3bs28efMYO3Ys27dvZ+7cuZ5+69ato0GDBhhjaN++PdWrV+fEiRNs376dUaNG8fbbb3sS1u4YYmJiqFWrFl27dqVgwYLMmzePvn37smzZMr766quLNmciIiIiIiIiIiKXylWRsHY9EHEosAEnWX34Mhz2iOu1yGU41p/W/PnziY6OJioqytPWtWtX7rnnHoYMGeJJKH/44YcsWLCANm3a8OOPPxIS4rx1o6KiqF+/Pu+++y733XcfjRo1Oq/xjx8/zrfffku9evVYtmwZwcHBPuMcPXrU8/uxY8d45JFHKFy4MAsXLuSGG27wbNu4cSMNGjTg8ccfZ9Uq/0o0S5cuZf369VSpUgVwVkY3b96cefPm8dtvv1G/fn0Axo8fT2pqKlOnTqVDhw4+Yxw7dozChQt7/n7nnXeIiYnhmWee4eOPP/bEnpmZyZNPPsm4cePo3Lmz3zgiIiIiIiIiIiL5Tb5PWBtjXgYGA2uAVtbahMt06Aau152X6XiYjsMv16EumJ3a76KMU7VqVV5//XWftrvvvpsqVarw22+/edrGjRuHMYaPPvrIk6wGiIyM5I033uDxxx/ns88+80tY53V8YwzWWsLCwgKWzyhVqpTn9y+//JKkpCRGjBjhk6wGuPHGG3niiSf4+OOP2bRpk9/2N99805OsBggJCeGxxx4jNjbWJ2HtVqhQIb9YSpQo4fk9KyuLESNGUK5cOYYOHeqTaA8ODubDDz/k888/Z8KECUpYi4iIiIiIiIhIvpevE9bGmDeAQTgPRmx9scuAGGPusNbGZmszOA9g/CuQAMy8mMcUX3Xr1vVbzQxQuXJllixZAsDJkyfZvn07FStWpHbt2n59mzdvDsDq1avPa3yA4sWL065dO3766Sfq1q3LAw88wB133EGDBg18VjMDnv3Wrl1LoIeWbt26FYDNmzf7Jaxvu+22gLGAs3LarUuXLgwbNoyOHTvSuXNnWrZsSePGjalRo4bfsY4ePcp1113H22+/7Tc2OEnvzZs3B9wmIiIiIiIiIiKSn+TbhLUxpgdOsjoTiAX6OblkH3HW2i+89nkFcGc067peHzPGNHH9vijbAxgXGmO2AsuB/UA40Bi4CUgBullrT1y0kxI/ERERAdtDQkLIysoCnHIdAOXLlw/Y192elJR0XuO7fffdd7z33ntMnDjRU0KkYMGCdO7cmQ8++ICyZcsC/ysPMnbs2FzPLTk5OU/xuFeMZ2Zmetrq169PbGws77zzDpMmTfLUoK5VqxZRUVE88sgjPrFs27aNgQNzfnhroFhERERERERERETym3ybsAaqu16Dgf459FkAfOH19z1A02x9Grl+3LwT1h8A9YHmQEkgC9gDjAQ+stZetnIgcPHKbPzRhIeHA3Dw4MGA2+Pj4336na9ChQoRHR1NdHQ0e/fuZeHChXzxxRd8/fXXxMXFERsb63OctWvXcsstt1zQMXPz17/+lenTp3PmzBlWrlzJzJkz+eSTT+jatStlypShZcuWnljuv/9+Jk+efMliERERERERERERuRz8i/XmE9baaGutOctPs2z7NDtL/57Z+g+w1ja11law1ha01ha21ta21j5zuZPVkrNixYpRo0YN9u/fz7Zt2/y2z5s3D4Bbb731oh2zcuXKdOvWjZiYGK677joWLVrkWc3csGFDAE8C+1ILCwujUaNGDBo0iOHDnTrn06ZNA6B27dpERESwdOlS0tPTL0s8IiIiIiIiIiIil0q+TViLeOvVqxfWWgYMGOBTOiMhIYG33nrL0+d8HTlyhGXLlvm1nzp1ipMnTxISEkJoaCgAjz32GBEREQwcONDnwY1uWVlZzJ8//7xjAScZ7i6F4u3QoUMAnrraISEh9O3bl/j4ePr168fp06f99omPj2fTpk0XFI+IiIiIiIiIiMjlkJ9Lgoh4vPjii8yYMYNp06ZRp04d2rZtS0pKCt9//z2HDx/mpZdeokmTJmcfKAf79++nYcOGXH/99dx6661UrlyZEydOMH36dA4ePEi/fv0oVqwYAKVKlWLSpEncf//9NGzYkBYtWnDjjTcSFBTEnj17WLJkCUePHiU1NfW84/nwww+ZNWsWzZo145prrqFo0aJs3LiRGTNmUKJECZ588klP3zfeeIO1a9cyZswYfvrpJ5o3b07FihU5fPgw27ZtY/Hixbzzzjt+D4AUERERERERERHJb5SwlqtCaGgos2fP5qOPPmLixIl88sknhISEUKdOHT7++GPPQwjPV7Vq1Rg4cCDz589n3rx5JCQkULJkSWrVqsXgwYN5+OGHffq3aNGCdevW8cEHHxATE0NsbCyhoaFUqFCB5s2b88ADD1xQPH369KFEiRIsW7aMxYsXk5GRQaVKlejTpw8vvPACVatW9fQtUKAAU6dO5euvv+aLL75g+vTpJCcnU6ZMGapXr85bb71Ft27dLigeERERERERERGRy8FYa690DFe90aNHW4CnnnrqSociIiIiIiIiIiIicqHMlTqwaliLiIiIiIiIiIiISL6ghLWIiIiIiIiIiIiI5AtKWIuIiIiIiIiIiIhIvqCEtYiIiIiIiIiIiIjkC0pYi4iIiIiIiIiIiEi+oIS1iIiIiIiIiIiIiOQLSliLiIiIiIiIiIiISL6ghLWIiIiIiIiIiIiI5AtKWIuIiIiIiIiIiIhIvqCEtYiIiIiIiIiIiIjkC0pYi4iIiIiIiIiIiEi+oIS1iIiIiIiIiIiIiOQLSliLiIiIiIiIiIiISL6ghLXIVa5nz54YY4iLi7vSoeR78+fPxxhDdHT0lQ4lz5o1a4Yx5kqHISIiIiIiIiJyWShhLZddXFwcxhh69ux5pUPxk56ezpQpU+jduzc33XQTxYsXp3Dhwtx88828+eabnDx5Msd99+3bR69evahQoQJhYWFUq1aN/v37c+zYsct4BiIXR36+T0VERERERETkjyvkSgcgkp/s2LGDTp06UaRIEe666y7uvfdekpOTiYmJ4a233uK7775j8eLFlC5d2m+/Ro0acfjwYTp06EDt2rX57bffGDZsGDNnzmTx4sWUKlXqCp2VXM2+/PJLUlJSrnQYIiIiIiIiIiKXhRLWIl6KFSvGyJEj6dGjB0WKFPG0p6Wl0alTJ37++WcGDhzIJ5984rNfnz59OHz4MMOHD6dv376e9ueff56hQ4fyz3/+kzFjxly285A/jipVqlzpEERERERERERELhuVBJHLKjo6murVqwMwfvx4jDGeny+++MLTLysrizFjxnD77bdTtGhRihQpwu23387o0aPJysryG9cYQ7NmzThw4ACPPvookZGRFCpUiHr16jFx4sQ8x1exYkX69Onjk6wGCA0N5bXXXgOcOsjedu7cyaxZs6hWrRpPP/20z7aBAwdSpEgRvvrqK06dOpXnOFauXMkDDzxAZGQkYWFhVK1alT59+hAfH5/jPllZWXz00UfUrl2bggULUqlSJZ577jlOnDjh13fdunU88sgjVKtWjbCwMMqUKcOtt95K//79SU9P9+mbkZHBqFGjaNiwoadEyl/+8hdGjBjhNxfeZSS2bt1Kly5diIyMJCgoiPnz51O7dm1CQ0NJSEgIeA6DBw/GGMPIkSN92vft28czzzzDNddcQ1hYGKVKlaJ9+/YsX7484DiHDh2id+/elC1blkKFClG3bl3Gjx+f47UL5N1338UYw/DhwwNuP3DgAMHBwdx+++0+bYMGDaJx48aUK1eO0NBQKlSoQNeuXdm8ebPfGGe7XhC4hnVaWhojRoygbdu2VK1albCwMEqWLEnLli2ZMWNGwHirVatGtWrVSElJYcCAAVSpUoWwsDCuvfZa3nvvPay1nr55vU9FRERERERERC42rbCWy6pZs2YkJSUxbNgw6tSpQ8eOHT3b6tat6/n90UcfZeLEiVSuXJnHH38cYwxTpkyhT58+LFq0iAkTJviNfezYMRo1akRERASPPfYYSUlJ/Oc//6Fbt27s37+fAQMGXFDsBQoUACAkxPe2mTt3LgCtW7cmKMj3O6BixYrRuHFjZs2axdKlS2nRosVZjzN9+nQeeOABrLV07tyZqlWrsnLlSkaPHs20adNYvHgx1apV89vvueeeY+HChTz00EN06NCBmJgYPv74Y2JjY1m0aBEFCxYEnGR1gwYNMMbQvn17qlevzokTJ9i+fTujRo3i7bff9pxreno67dq1IyYmhlq1atG1a1cKFizIvHnz6Nu3L8uWLeOrr77yi2XHjh00aNCAmjVr0q1bN06fPk3x4sXp0aMHr732Gt98843PSnS3L7/8ktDQUB5++GFP26pVq2jdujWJiYncfffddOrUiYSEBKZOnUqTJk2YMmUKbdu29fQ/evQojRo1YufOnTRp0oQmTZoQHx/PP/7xD1q3bn3W6+/WvXt3Xn/9dcaPH0+/fv38tn/99ddkZWXRo0cPT9vChQsZPHgwd911Fw888ABFixZl27ZtTJo0iR9//JHFixdTp06dPF+vnCQmJvLss8/SqFEjWrVqRZkyZYiPj+enn36ibdu2jB07lscff9xvv/T0dFq3bs2BAwdo06YNISEhTJ06lVdeeYXU1FSioqKAvN+nIiIiIiIiIiIXnbVWPxf4M2rUKDtq1CgrebNr1y4L2B49egTcPnHiRAvYv/zlL/bkyZOe9uTkZFuvXj0L2AkTJvjsA1jAPvjggzYzM9PTvnPnTluiRAlboEABu2PHjguK+x//+IcF7CuvvOLT/uKLL1rAfvDBBwH3e/rppy2Qp/fIyZMnbalSpWxQUJBduHChz7bBgwdbwLZq1cqnvUePHhawpUqVsnFxcZ72zMxM26lTJwvYQYMGedqff/55C9ipU6f6HT8xMdHn+kVFRVnAPvPMMzYjI8PTnpGRYXv16uU3jntuAfvqq6/6jb9v3z4bFBRk69Wr57ftt99+s4Dt1KmTpy09Pd3WqFHDhoWF2fnz5/v0379/v61QoYItV66cTU1N9bQ/8cQTFrD9+/f36b98+XIbEhJiARsVFeV3/EBat25tAbt+/Xq/bTfccIMtUKCATUhI8LQdOnTInjhxwq/vmjVrbJEiRew999zj036262WttU2bNrXOR/X/pKam2r179/r1TUpKsjfeeKMtUaKETUlJ8dlWtWpVC9g2bdr4bDt06JANDw+34eHhNi0tzS+2nO5TEREREREREflDu2K5Vq2wzk/qPHilI8i7td9fsqHHjRsHOOUhihYt6mkvUqQI7733Hi1btuSzzz6ja9euPvsFBwfz3nvv+axyrl69Ov369WPgwIF89dVXnhWk5+rHH3/k008/pVKlSrz00ks+244fPw5AeHh4wH3d7UlJSWc9zrRp0zh69CiPPPIId9xxh8+2F154gTFjxjB79mz27NnjV9v42WefpWrVqp6/g4KCGDJkCFOnTmXcuHG88cYbPv0LFSrkd/wSJUp4fs/KymLEiBGUK1eOoUOHEhwc7NkWHBzMhx9+yOeff86ECRPo0KGDzzhly5YNeK0rVqxIixYtmD17Nhs3buTGG2/0bHOX7PBesfzzzz+zY8cOXnzxRZo2beozVoUKFXjppZfo378/c+bMoW3btqSnpzNhwgSKFStGdHS0T//bbruNbt26nVNpkB49ejBr1izGjx/PkCFDPO0rVqxg06ZN3H///T4P04yMjAw4Tp06dWjevDmzZs0iPT3ds4LdLafrlZOwsDAqVark1x4eHk6vXr144YUXWL58OXfeeadfn+HDh/vMfWRkJB06dODLL79ky5Yt3HTTTXmOQ0RERERERETkYlPCWvKdVatWERQURLNmzfy2NW3alODgYFavXu23rUqVKp66u96aNWvGwIEDA+6TF//973/p2rUrRYoU4YcffvBJ6uaFddUGzl6HOJBVq1YB0Lx5c79tISEh3HnnncTFxbF69Wq/hHX2hC7ANddcQ+XKlYmLiyMpKYmIiAi6dOnCsGHD6NixI507d6Zly5Y0btyYGjVq+Oy7detWjh49ynXXXcfbb78dMN5ChQoFrM1cp04dwsLCAu7Ts2dPZs+ezfjx43n//fcBpybzt99+S5kyZXzKeyxZsgSA3bt3+yWgAbZt2wbA5s2badu2Lb///jspKSnccccdAb9AaNas2TklrO+//37Cw8P5+uuvGTx4sCdp7x6jZ8+efvv8/PPPjBkzhhUrVpCQkEBGRobP9oSEBMqXL+/Tltv1ysnGjRsZMmQICxcuJD4+ntTUVJ/t+/fv99snPDyca6+91q+9cuXKgFNWR0RERERERETkSlLCWvKd48ePU7JkSUJDQ/22hYSEULp0aQ4fPuy3rWzZsgHHK1eunGfcc7VkyRLatGlDUFAQM2bMoH79+n593InRnMZ3P/QwpxXY3txjZE9ournbA63Wzu38d+/ezfHjx4mIiKB+/frExsbyzjvvMGnSJE8N6lq1ahEVFcUjjzwCOLWgwUkKDxw4MMeYk5OTAx4zJ/fffz/Fixfn66+/5t133yU4OJjp06dz9OhR+vfv71Mj3B3D99/nvqLfHYP7+p3tvZBXhQoV4qGHHmLs2LHMmjWLNm3akJ6ezjfffEOZMmVo06aNT//hw4fz7LPPUqJECVq1akWVKlUoXLgwxhimTp3K2rVrOXPmzAXHtXTpUpo3b05GRgYtWrSgffv2FC9enKCgINasWcO0adMCHiciIiLgeO5rnpmZeU5xiIiIiIiIiIhcbEpY5yeXsMzG1SQ8PJzExMSApRMyMjJISEgI+EC6Q4cOBRzv4MGDnnHPRWxsLPfeey9BQUHExMTQsGHDgP1q1aoFOCuSA3GvAq5Zs+ZZj+mO0R1zdvHx8T79vB06dMgTi7dA5//Xv/6V6dOnc+bMGVauXMnMmTP55JNP6Nq1K2XKlKFly5ae/vfffz+TJ08+a+zecltN7k4Cf/bZZ8yePZt77rknYDkQ75inTZtG+/btz3pcd/+zvRfORY8ePRg7dizjx4+nTZs2nuT6s88+6/P+zMjIICoqinLlyrFq1Sq/Lx3cq8UDycvqe29vv/02p0+fZt68eX7/EuHdd99l2rRp5zSeiIiIiIiIiEh+EXT2LiIXl7usQk6rOf/yl7+QlZXFwoUL/bYtXLiQzMxMbr31Vr9te/bsIS4uzq99/vz5nnHzau7cubRp04aQkBBmz56dY7Ia4K677gJg1qxZZGVl+Ww7efIkixcvplChQrmO4eaO0R2zt4yMDBYtWgQQ8PwXLFjg17Zz50727t1LtWrVAq6uDQsLo1GjRgwaNIjhw4cDeJKdtWvXJiIigqVLl5Kenn7W2M+Fu5TG+PHjSUhIYMaMGdxyyy3UrVvXp5/7msXGxuZp3Nq1a1O4cGHWrFkTcMV7oOt6No0bN+a6665j2rRpHD9+PMfkekJCAklJSTRq1MgvWZ2cnOwp93IxbN++nZIlSwYsmxPofXA+znafioiIiIiIiIhcCkpYy2VXokQJjDHs2bMn4PZevXoB8Oqrr5KSkuJpT0lJ4ZVXXgGgd+/efvtlZmby8ssv+ySNd+3axfDhwwkJCeFvf/tbnuKbNWsW9913HwULFmTOnDncfvvtufavUaMGrVu3Ji4ujpEjR/psi4qK4tSpU3Tv3p0iRYqc9dgdO3akZMmSfPPNNyxdutRn28cff8zOnTtp2bKlX/1qgGHDhrF7927P31lZWQwYMICsrCwee+wxT3tsbGzAZK57VXLhwoUBp0xE3759iY+Pp1+/fpw+fdpvn/j4eDZt2nTW88rOOwk8evRo0tPTA9aD7tChAzVq1GDkyJH88ssvAcdasmSJ531SoEABunXrxsmTJ/1qXq9YsYIJEyacc6zgJKdTU1MZNWoUv/zyC7fccovfFyCRkZEULlyYlStX+pRJSU9P59lnnyUhIeG8jh1ItWrVSExMZN26dT7t//73v4mJibkoxzjbfSoiIiIiIiIicimoJIhcdkWLFqVBgwbExsbSrVs3atasSXBwMO3bt+eWW26ha9euTJs2jf/85z/ceOONdOzY0VMDeNeuXTz00EN069bNb9xbbrmFZcuWUa9ePVq3bs3x48f57rvvSEpK4v333/d7qGAgW7ZsoUOHDqSmptK2bVumTZsWsLxC9mToqFGjaNSoEf369WPOnDlcf/31LFu2jHnz5lGzZk3eeeedPF+bcePG8eCDD9K0aVMefPBBqlSpwsqVK5k1axblypXj008/Dbhv48aNqVu3Ll26dCE8PJyYmBjWrl1LvXr1eOmllzz9PvzwQ2bNmkWzZs245pprKFq0KBs3bmTGjBmUKFGCJ5980tP3jTfeYO3atYwZM4affvqJ5s2bU7FiRQ4fPsy2bdtYvHgx77zzDjfccEOezs9b9+7deeONN3jrrbcICQmha9eufn0KFCjA5MmTufvuu7n33ntp1KgRdevWpXDhwuzdu5fly5ezc+dO4uPjPYn2//u//2POnDl8/PHHrFixgiZNmhAfH893331H27Zt+fHHH88r1jfffJOoqCjS09P9VlcDBAUF0a9fPwYPHszNN99Mhw4dSEtLY968eSQmJnLXXXcxb968cz52IP379ycmJoYmTZrw0EMPER4ezooVK1i0aBGdO3dm0qRJF3yMs92nIiIiIiIiIiKXhLVWPxf4M2rUKDtq1Cgrebdt2zZ733332ZIlS1pjjAXs559/7tmemZlpR44caevVq2cLFSpkCxUqZG+99VY7YsQIm5mZ+f/s3Wl0Xmd56P3/ZUuRI9uyLWuyUyDQkkDpYTjQ0hBKIAyFhpaU4YU2lCRleOMCCaWw6ClQpLT0sBhCy3AMlDaBNm3oAgqlhULfkjAcoJEsy/Igx6Ms2/I8yZM86Xo/6JH6SH4k+bElLKL/by2trb33fe99bV2+7w9Xdu593vWAvOGGG3Lnzp15yy23ZH19fVZVVeUznvGMvP/++y84rgcffDCBCX9K6enpydtuuy2bmpqysrIyH/vYx+add96ZBw4cKPvv8/DDD+fNN9+cdXV1WVlZmY95zGPyjjvuyJ07d57X9tZbb00gN2/enB/96Efz2muvzaqqqly6dGneddddeeTIkRHtv/3tb+dtt92WT37yk7Ompiarq6vzmmuuybe//e3Z3d193vUHBgbyi1/8Yt544425aNGirKyszKVLl+b111+fH/zgB7Onp2e47datWxPIW2+9dcJn3LZtW86aNSuBfPnLXz5u2z179q171kkAACAASURBVOR73vOefMpTnpJXXnllzp07N3/hF34hX/WqV+Xf/d3f5ZkzZ0a037VrV95+++1ZV1eXc+bMyac97Wl57733Duf3Ax/4wITxjfbCF74wgayoqMjdu3eXbHPmzJn82Mc+lk9+8pNzzpw52djYmK9//euzu7t7OE9bt24dbn8hf68bbrih5L+5b3zjG/nsZz87582blwsWLMgXv/jF+b3vfS/vvffe88ZTZubjHve4fNzjHlfyHh/4wAcSyAcffHDE8YnGqSRJkiRJetS6bLXWyMyfZn38UWn58uUJsGzZsssdyowVEdxwww0XtUaxJEmSJEmSpBHict3YNawlSZIkSZIkSdOCBWtJkiRJkiRJ0rRgwVqSJEmSJEmSNC1M24J1RCyOiDdFxD9HxKaIOBkRRyLihxHxxoiYNap9ZUTcFRH3RkRHRJyOiIyIN13AvW6NiIcj4ljhHg9FxMun7uk02TLT9aslSZIkSZKkn3EVlzuAcbwGWA7sAh4EeoBG4JXA54GXRcRr8r+/GjkX+MvC73uA3cBjJrpJRHwU+CNgB/DXwBXA64BvRMTbM/NTk/ZEkiRJkiRJkqQxTds3rIENwG8BP5eZt2Tm/8rM3weeBGwHXsVg8XrICeA3gKWZ2QT87UQ3iIjnMFis3gw8NTP/MDPfCjwTOAh8NCKunrxHkiRJkiRJkiSNZdoWrDPzu5n5jcwcGHV8N/CZwu7zi46fzsxvZeauMm5zR2H7wcw8VHStbuDTQBVw+0WEL0mSJEmSJEkq07QtWE/gTGF79hKvc2Nh++8lzn1rVBtJkiRJkiRJ0hT6mStYR0QF8IbCbqlC84VeZy5wFXBsjLeyNxa211zsPSRJkiRJkiRJF252c3Pz5Y6hLC0tLR8Gbga+mZl3j9Pu+QwuGfKN5ubm9hLn6xlcv3pfc3PzPSXOnwHeD1Q2Nzf/eal7RMRbWlpaPnvTTTctBbjmmmvo6uqirq6OlStXsnTpUtra2li8eDFr166lqqqKnTt3cuLECY4fP05vby9XXHEF69ato76+nvb29hF9urq6qKyspLe3l+PHj3Py5El27NhBVVUVa9asobGxcUSfRYsW8cgjjzB79mx2797N0aNH6e/vZ/v27Vx55ZV0dnbS1NTEihUrhvssXLiQDRs2EBHs3buXvr4+Tp8+zbZt26iurqazs5OlS5fS2trKVVddRVtbGwsWLGDTpk1kJvv37+fQoUOcPXuW7u5u5s2bx6pVq4avP3S/mpoaNm/ezMDAAAcPHuTAgQMMDAywZcsW5s+fP6JPY2MjK1euZP78+WzdupWzZ89y6NAh9u/fD8CmTZtYuHAhHR0dw30aGhpYtWoVc+fOpbu7mzNnznDkyBH27t1LRLBhwwYWLVo0ok99fT2dnZ1UV1fT09PDqVOn6OvrY8+ePcyePZv169ezePFi2tvbh5+9rq6ONWvWUFVVxY4dO+jv7+fYsWPs2rWLioqK4fwP9WltbaWuro5169YN53J0/teuXUtDQ8OIXNbW1tLV1UVFRcVwLkfnf3Quh/I/a9Ys9uzZQ19fH6dOnaKnp2c4/0uWLBkz//v27ePw4cOcOXPmvPy3tbUNb2tqas7L/7lz59iyZQvz5s2jo6Nj+O/V1NREe3v7uPmvqakZzktraytNTU10dHQwd+7ckvnfuHHjmPmvrq5m27ZtnD59+rz819bWjhiX9fX1rF69mjlz5tDT00N/fz9Hjx5l9+7dI/Jf3Kc4/0Nj+dixY/T29lJZWUlXV1fJsVyc/+KxXJz/FStWDP/damtrWb9+/ZhjefXq1SPy39raWjL/Q2N5KP+jczk0loER+S81lof+7Qzlf2BggAMHDozI/+ixPJT/4rFcnP/NmzezYMGCEblsbGwczn/xWN63b9+I/K9cuXJ4jDU0NIw5lmfNmnVe/of6DOV/+/btY45l53Lncudy53Lncudy53Lncudy53Lncudy53Ln8pk4ly9ZsqRlrLrrVIvMvFz3LltE3An8FbAeuD4zD47Tthn4APDmzPx8ifNLgZ3Azsz8uRLnK4HTwKnMnDNeXMuXL0+AZcuWXfjDSJIkSZIkSdL0FJfrxj8zS4JExFsZLFavA14wXrH6Ah0pbBeMcX7BqHaSJEmSJEmSpCn0M1Gwjoh3AJ8C1jBYrN59qdfMzOMMvmE9LyKWlGjyxMJ2w6XeS5IkSZIkSZI0sWlfsI6I9wAfBzoYLFbvncTLf7ewfWmJcy8b1UaSJEmSJEmSNIWmdcE6It4PfAhYAbwwM/dP8i0+U9i+NyIWFd33auCtwCng3km+pyRJkiRJkiSphIrLHcBYIuJW4G7gHPAD4M6I89b67s7M+4r6/DHwpMLu0wvb2yPiuYXff1j8AcbM/FFE3AO8E+iMiC8DVwCvBWqBt2dm92Q+lyRJkiRJkiSptGlbsAYeX9jOBt4xRpvvAfcV7b8UuGFUm+cUfoZ8vvhkZv5RRHQCbwPeAgwA7cBHMvNfLypySZIkSZIkSVLZpm3BOjObgeYy+zz/Iu/1BeALF9NXkiRJkiRJkjQ5pvUa1pIkSZIkSZKkmcOCtSRJkiRJkiRpWrBgLUmSJEmSJEmaFixYS5IkSZIkSZKmBQvWkiRJkiRJkqRpwYK1JEmSJEmSJGlasGAtSZIkSZIkSZoWLFhLkiRJkiRJkqYFC9aSJEmSJEmSpGnBgrUkSZIkSZIkaVqwYC1JkiRJkiRJmhYsWEuSJEmSJEmSpgUL1pIkSZIkSZKkaaHiYjpFxFzgKqAOOAnszcydkxmYJEmSJEmSJGlmueCCdUT8PPD7wIuBZzDq7eyIOAA8CHwF+Gpmnp3EOCVJkiRJkiRJj3ITFqwj4lnAnwMvYrBIfQZYDewGDgJXAouBa4HXAK8G9kXEXwH3ZOapqQldkiRJkiRJkvRoMm7BOiK+ANwCHAE+BzwAPJyZ/WO0vxr4deBW4IPAHRHxe5n5/UmMWZIkSZIkSZL0KDTRRxdfArwTWJqZf5CZ3x+rWA2Qmd2Z+dnMfA7wdKAduGHywpUkSZIkSZIkPVpNtCTIEzLz5MVcODM7gd+OiCsvpr8kSZIkSZIkaWYZ9w3riy1WT/Y1JEmSJEmSJEmPfhMtCXJBIqIyIp4REddOxvUkSZIkSZIkSTNPWQXriPh/IuKfIqK26NjPA2uBNmBdRHw1IiZaakSSJEmSJEmSpBHKfcP694EnZebBomMfA34BeBDoBF4B3D454UmSJEmSJEmSZopyC9a/CLQO7UREDfAbwD9l5ouAXwHWY8FakiRJkiRJklSmcgvW9cCuov3rgArgAYDMPAP8B/DzkxKdJEmSJEmSJGnGKLdgfRRYULR/A5DAD4uO9QPzLzEuSZIkSZIkSdIMU+7HETcCL4uIKgYL1a8BOjNzf1GbxwF7Jyk+SZIkSZIkSdIMUe4b1p8DnsBg4bqr8PvfjmrzbGDtpYcmSZIkSZIkSZpJyipYZ+YXgA8B1QwuDfKpwg8AEXEjcDXw4OSFKEmSJEmSJEmaCcpdEoTM/BPgT8Y4/UNgEXD8UoKSJEmSJEmSJM08ZResx5OZp4HTk3lNSZIkSZIkSdLMMO6SIBFx5aXeYDKuIUmSJEmSJEl69JtoDeutEXFXRFSVe+GIeFpEfB1418WFJkmSJEmSJEmaSSYqWH8HuAfYFRHLI+IF470xHRFPiIhlEfFjoB14Gn6AUZIkSZIkSZJ0AcZdwzoz3xARnwD+AnhL4edcRHQBu4BDwBxgMXAtUAcEsAd4L/DxzDw1deFLkiRJkiRJkh4tJnrDmsxsy8yXAE8CPgKsAn4ReAnwWuAVwHMLzb8K/C7w2Mz80KUUqyNicUS8KSL+OSI2RcTJiDgSET+MiDdGRMnYI+I5EfHNiDgYESciojMi3hERs0u0fX5E5Dg/H7rY+CVJkiRJkiRJ5Rn3DetimbkR+GOAiKgGrmLwzeqTwN7M3DXJsb0GWM7gm9wPAj1AI/BK4PPAyyLiNZmZQx0i4hXAV4B+4EvAQeA3gY8D1xeuWcr3gIdKHP/hZDyIJEmSJEmSJGliF1ywLpaZJ4CNhZ+psgH4LeDfMnNg6GBE/AnwMPAqBovXXykcrwH+GjgHPD8z2wrH3w98F3h1RLwuMx8oca+HMrN5Cp9FkiRJkiRJkjSBCZcEuVwy87uZ+Y3iYnXh+G7gM4Xd5xedejVQDzwwVKwutO8H3lfYXTZ1EUuSJEmSJEmSLsVFvWE9DZwpbM8WHbuxsP33Eu2/D5wAnhMRVSXW1v6FiHgbUAPsBn5QWAJFkiRJkiRJkvRT8jNXsI6ICuANhd3i4vS1he2G0X0y82xEbAWeAjwB6BrV5JbCT/F9vgK8OTMPTUbckiRJkiRJkqTxzW5ubr7cMZSlpaXlw8DNwDcz8+6i4+9m8COQH2lubj5Yot8twGOA+5ubm3cUjs1l8GOOfwi8h8GPMz7IYGH7RuC5LS0t9zU3N+fo6wFExFtaWlo+e9NNNy0FuOaaa+jq6qKuro6VK1eydOlS2traWLx4MWvXrqWqqoqdO3dy4sQJjh8/Tm9vL1dccQXr1q2jvr6e9vb2EX26urqorKykt7eX48ePc/LkSXbs2EFVVRVr1qyhsbFxRJ9FixbxyCOPMHv2bHbv3s3Ro0fp7+9n+/btXHnllXR2dtLU1MSKFSuG+yxcuJANGzYQEezdu5e+vj5Onz7Ntm3bqK6uprOzk6VLl9La2spVV11FW1sbCxYsYNOmTWQm+/fv59ChQ5w9e5bu7m7mzZvHqlWrhq8/dL+amho2b97MwMAABw8e5MCBAwwMDLBlyxbmz58/ok9jYyMrV65k/vz5bN26lbNnz3Lo0CH2798PwKZNm1i4cCEdHR3DfRoaGli1ahVz586lu7ubM2fOcOTIEfbu3UtEsGHDBhYtWjSiT319PZ2dnVRXV9PT08OpU6fo6+tjz549zJ49m/Xr17N48WLa29uHn72uro41a9ZQVVXFjh076O/v59ixY+zatYuKiorh/A/1aW1tpa6ujnXr1g3ncnT+165dS0NDw4hc1tbW0tXVRUVFxXAuR+d/dC6H8j9r1iz27NlDX18fp06doqenZzj/S5YsGTP/+/bt4/Dhw5w5c+a8/Le1tQ1va2pqzsv/uXPn2LJlC/PmzaOjo2P479XU1ER7e/u4+a+pqRnOS2trK01NTXR0dDB37tyS+d+4ceOY+a+urmbbtm2cPn36vPzX1taOGJf19fWsXr2aOXPm0NPTQ39/P0ePHmX37t0j8l/cpzj/Q2P52LFj9Pb2UllZSVdXV8mxXJz/4rFcnP8VK1YM/91qa2tZv379mGN59erVI/Lf2tpaMv9DY3ko/6NzOTSWgRH5LzWWh/7tDOV/YGCAAwcOjMj/6LE8lP/isVyc/82bN7NgwYIRuWxsbBzOf/FY3rdv34j8r1y5cniMNTQ0jDmWZ82adV7+h/oM5X/79u1jjmXncudy53Lncudy53Lncudy53Lncudy53LncufymTiXL1mypGWiOu1UicyStdhpKSLuBP4KWA9cn5kHi85tAJ4IPDEzN5Xo+yPgOuC6zPzJBPepATqAxwM3Z+bXx2u/fPnyBFi2zCWyJUmSJEmSJP3Mi8t142n70cXRIuKtDBar1wEvKC5WFxwpbBeMcYmaUe3GlJl9wD8Udp9XZqiSJEmSJEmSpItQVsE6Ih5bePt4vDbzI+KxlxbWedd8B/ApYA2DxerdJZo9UtheU6J/BYNvS58FtlzgbfcVtnPLi1aSJEmSJEmSdDHKfcN6K3DXBG3uLLSbFBExtLZ0B4PF6r1jNP1uYfvSEueeB1QDP8rMUxd4618tbC+0wC1JkiRJkiRJugTlFqyDn+L6JRHxfuBDwArghZm5f5zmXwb2A6+LiGcVXWMO8OeF3eWjrn99RJz3N4iI1wOvBU4D/3RJDyFJkiRJkiRJuiAVU3DNRuD4pV4kIm4F7gbOAT8A7ow4r1benZn3weC60xHxZgYL1w9FxAPAQeC3gGsLx780qv/9wKzCBxl3AHOAXwZ+hcHlQ/7fzOy+1GeRJEmSJEmSJE1swoJ1RLxh1KGnlzgGMBt4LPB7wOpJiO3xRdd9xxhtvgfcN7STmV+LiBuA9wKvYrAAvQl4J/CJzMxR/ZcDLwKuB+oYfHt8Z+Gaf5mZqybhOSRJkiRJkiRJFyDOr+GOahAxAIzfqNC0sD0BvDIzv3OJsf3MWL58eQIsW7bscociSZIkSZIkSZfqp7Ys9GgXsiTI7YVtAH8LfA34eol254ADwI8z8/DkhCdJkiRJkiRJmikmLFhn5heGfi+sK/21zPzilEYlSZIkSZIkSZpxyvroYma+YKoCkSRJkiRJkiTNbLPKaRwR/zMi/jQiGsc431Q4//TJCU+SJEmSJEmSNFOUVbAG3gW8Cdg7xvk9wBuBd15KUJIkSZIkSZKkmafcgvV1wIOZmaVOFo5/F7j+UgOTJEmSJEmSJM0s5Rasm4AdE7TpBZZcXDiSJEmSJEmSpJmq3IL1CaB+gjb1wKmLC0eSJEmSJEmSNFOVW7DuAF4REfNKnYyIGuAVhXaSJEmSJEmSJF2wcgvWn2PwDer/iIinFp+IiKcB3wHqCu0kSZIkSZIkSbpgFeU0zswvRcTLgDcAKyNiD7ATuApoBAL4Qmb+46RHKkmSJEmSJEl6VCv3DWsy8zbgDmAdgx9hfGZhuxZ4S2bePpkBSpIkSZIkSZJmhrLesB6SmZ8DPhcR1cBC4HBmnpjUyCRJkiRJkiRJM8pFFayHFIrUFqolSZIkSZIkSZfsogrWEVEPvAp4MjA3M99UdPzxwOrMPDlpUUqSJEmSJEmSHvXKLlhHxBuBTwBzGPzIYgJvKpxuBH4MvAX4m0mKUZIkSZIkSZI0A5T10cWIeDHwOWAD8NvA8uLzmbmGwY8v3jxZAUqSJEmSJEmSZoZy37B+D7ALuCEz+yLiGSXadALXXXJkkiRJkiRJkqQZpaw3rIFnAf+amX3jtNkBNF18SJIkSZIkSZKkmajcgvUVwPEJ2iwEzl1cOJIkSZIkSZKkmarcgnU38MwJ2jwbeOSiopEkSZIkSZIkzVjlFqy/DvxaRLym1MmIuB14KvCVSw1MkiRJkiRJkjSzlPvRxQ8DrwP+MSJeDSwAiIi3Ab8GvBLYCHxyMoOUJEmSJEmSJD36lVWwzsxDEXED8EWg+C3rTxS2PwB+NzMnWudakiRJkiRJkqQRyn3DmszsAZ4fEU8FrgMWA0eAn2TmikmOT5IkSZIkSZI0Q4xbsI6IrwIPZOY/FfafB3RnZk9mdgKdP4UYJUmSJEmSJEkzwEQfXbwZeFLR/oPAbVMWjSRJkiRJkiRpxpqoYH0YqCnajymMRZIkSZIkSZI0g020hnUX8DsR0QrsKhy7urA0yLgy8/uXGpwkSZIkSZIkaeaYqGDdAnwN+IeiY7cWfiYy+2KDkiRJkiRJkiTNPOMWrDPzOxHxZOBFwFVAM/C9wo8kSZIkSZIkSZNmojesycxtwN8AREQz8FBm3j3FcUmSJEmSJEmSZphxP7oYEfdExEuKDt3O4BIhkiRJkiRJkiRNqnEL1sA7gF8t2v9b4OapC0eSJEmSJEmSNFNNVLA+BlQX7ccUxiJJkiRJkiRJmsEmWsN6E/DKiPhnYFfh2MKIeOxEF87MnksNTpIkSZIkSZI0c0z0hvVHgJ8HfgRsLRy7q/D7eD9bLjWwiFgcEW+KiH+OiE0RcTIijkTEDyPijRFRMvaIeE5EfDMiDkbEiYjojIh3RMTsce718oh4qHD9YxHxXxFx66U+gyRJkiRJkiTpwo37hnVm/mNEbAVuAq4CbgM6gY6pD43XAMsZfLP7QaAHaAReCXweeFlEvCYzc6hDRLwC+ArQD3wJOAj8JvBx4PrCNUeIiLcBnwQOAH8PnAZeDdwXEf8jM981VQ8oSZIkSZIkSfpvUVTvnbhxxADQnJl3T11Iw/e6EZgL/FtmDhQdbwIeBh4DvDozv1I4XsPgEiYLgOszs61wfA7wXeA64Hcy84Gia10NrAeOA8/MzO7C8UVAK4Nvlz8nM388XqzLly9PgGXLll3qY0uSJEmSJEnS5XbZvmU40ZIgo90OfG0qAhktM7+bmd8oLlYXju8GPlPYfX7RqVcD9cADQ8XqQvt+4H2F3dEV5d8HqoBPDRWrC30OAX9R2L3j0p5EkiRJkiRJknQhJvro4giZ+YWpCqRMZwrbs0XHbixs/71E++8DJ4DnRERVZp66gD7fGtVGkiRJkiRJkjSFyn3DmoiYFRFvj4ifFD5SeLbo3DMi4v9ExDWTG+aI+1cAbyjsFheary1sN4zuk5lnGfwYZAXwhAvss4vBpUJ+LiKqLzFsSZIkSZIkSdIEZjc3N19w44i4Avg2g8tkXMngW8vzm5ub7wZoaWnpB+4H+pqbmx+a7GAL9/gwcDPwzeK1tFtaWt4NLAY+0tzcfLBEv1sYXPf6/ubm5h2FYx8EZgPvbW5uHijR563AfODTzc3Nx0afj4i3tLS0fPamm25aCnDNNdfQ1dVFXV0dK1euZOnSpbS1tbF48WLWrl1LVVUVO3fu5MSJExw/fpze3l6uuOIK1q1bR319Pe3t7SP6dHV1UVlZSW9vL8ePH+fkyZPs2LGDqqoq1qxZQ2Nj44g+ixYt4pFHHmH27Nns3r2bo0eP0t/fz/bt27nyyivp7OykqamJFStWDPdZuHAhGzZsICLYu3cvfX19nD59mm3btlFdXU1nZydLly6ltbWVq666ira2NhYsWMCmTZvITPbv38+hQ4c4e/Ys3d3dzJs3j1WrVg1ff+h+NTU1bN68mYGBAQ4ePMiBAwcYGBhgy5YtzJ8/f0SfxsZGVq5cyfz589m6dStnz57l0KFD7N+/H4BNmzaxcOFCOjo6hvs0NDSwatUq5s6dS3d3N2fOnOHIkSPs3buXiGDDhg0sWrRoRJ/6+no6Ozuprq6mp6eHU6dO0dfXx549e5g9ezbr169n8eLFtLe3Dz97XV0da9asoaqqih07dtDf38+xY8fYtWsXFRUVw/kf6tPa2kpdXR3r1q0bzuXo/K9du5aGhoYRuaytraWrq4uKiorhXI7O/+hcDuV/1qxZ7Nmzh76+Pk6dOkVPT89w/pcsWTJm/vft28fhw4c5c+bMeflva2sb3tbU1JyX/3PnzrFlyxbmzZtHR0fH8N+rqamJ9vb2cfNfU1MznJfW1laampro6Ohg7ty5JfO/cePGMfNfXV3Ntm3bOH369Hn5r62tHTEu6+vrWb16NXPmzKGnp4f+/n6OHj3K7t27R+S/uE9x/ofG8rFjx+jt7aWyspKurq6SY7k4/8VjuTj/K1asGP671dbWsn79+jHH8urVq0fkv7W1tWT+h8byUP5H53JoLAMj8l9qLA/92xnK/8DAAAcOHBiR/9FjeSj/xWO5OP+bN29mwYIFI3LZ2Ng4nP/isbxv374R+V+5cuXwGGtoaBhzLM+aNeu8/A/1Gcr/9u3bxxzLzuXO5c7lzuXO5c7lzuXO5c7lzuXO5c7lzuXO5TNxLl+yZEnL2BXaqVXuRxffC/wZ0Az8OfCnwPszc3ZRm+8ANZn5q5MbKkTEncBfMfihxOsz82DRuQ3AE4EnZuamEn1/xOCHF6/LzJ8Ujp0GKoHKwlvYo/v0AkuAJYW1s0vyo4uSJEmSJEmSHkV+Zj66eAvwfzPz7sLHEEtVu7cCj73kyEaJiLcyWKxeB7yguFhdcKSwXTDGJWpGtSunT18ZoUqSJEmSJEmSLkK5BevHAz+ZoM1BoPbiwiktIt4BfApYw2CxutTbzo8Utuetn11Y9/rxDH6kccsF9lkCzAV2ZOaJi49ekiRJkiRJknQhyi1YnwQWTtDmscDhiwvnfBHxHuDjQAeDxeq9YzT9bmH70hLnngdUAz/KzFMX2Odlo9pIkiRJkiRJkqZQuQXrDuAlhY8vniciFgC/Djx8qYEVrvd+4EPACuCFmbl/nOZfBvYDr4uIZxVdYw6D620DLB/V517gFPC2iLi6qM8i4E8Ku5+5hEeQJEmSJEmSJF2gijLb/zVwP3B/RLyx+ERELGSwALyISSjyRsStwN3AOeAHwJ0R56313Z2Z9wFkZl9EvJnBwvVDEfEAg8uT/BZwbeH4l4o7Z+bWiHg38AmgLSK+BJwGXg38HPCxzPzxpT6LJEmSJEmSJGliZRWsM/MfI+JFwO0MFoIPAUREG/AUoAr4dGZ+cxJie3xhOxt4xxhtvgfcVxTf1yLiBuC9wKuAOcAm4J3AJzLzvI9EZuYnI6IbeBfwBgbfOl8HvC8zvzAJzyFJkiRJkiRJugBRooY7caeI24C7gKcCQ689rwXuycx7Jy26nxHLly9PgGXLll3uUCRJkiRJkiTpUp231MVPS7lLggBQWIbjvoi4ksElQI5k5vHJDEySJEmSJEmSNLNcVMF6SGaeBE5OUiySJEmSJEmSpBls1uUOQJIkSZIkSZIksGAtSZIkSZIkSZomLFhLkiRJkiRJkqYFC9aSJEmSJEmSpGnBgrUkSZIkSZIkaVqwYC1JkiRJkiRJmhYqJuMiEVEH/BpwAvj/MvPcZFxXkiRJkiRJkjRzlPWGdUQsi4j/iojaomPPBLqALwPfBH4UEXMnN0xJkiRJkiRJ0qNduUuCvBbIzDxYdOwjwCLgXgYL1r8M3DE54UmSJEmSJEmSZopyC9ZPBDqHdgpLgdwA/E1mvikzfxNoBX538kKUJEmSJEmSJM0E5RasFwN7i/avL2z/uejYD4DHXUpQkiRJkiRJkqSZp9yC9UGgrmj/BmAA+FHRsQTmXGJckiRJkiRJkqQZptyCdRfwmxGxOCIWMrimdWtm9hW1uRrYPUnxSZIkSZIkSZJmiHIL1n8FLAF2ANuBJuD/DJ2MiNnAc4FVkxWgJEmSJEmSJGlmqCincWb+S0TcAbylAx9dXwAAIABJREFUcOj+zPz7oiYvYnA5kG9PUnySJEmSJEmSpBmirII1QGZ+DvjcGOe+DSy61KAkSZIkSZIkSTNPuUuCSJIkSZIkSZI0JcoqWEfEMyLiDyJiQdGxuRHxhYg4HBG9EXHX5IcpSZIkSZIkSXq0K/cN6/cA783MI0XH/jfwe4VrLQbuiYiXTFJ8kiRJkiRJkqQZotyC9bOAh4Z2IqISuBV4GGgAHg/sB+6cpPgkSZIkSZIkSTNEuQXrBmB70f6zgPnAZzOzPzN7ga8DT52k+CRJkiRJkiRJM0RFme1zVJ/nFo59r+jYPqD+EuOSJEmSJElSkY+sPzvu+Xc/qdwyjyRNP+XOZD3ArxbtvwLYkZlbio4tBQ5damCSJEmSJEkq7dDpHP590RVxGSORpMlVbsH6n4CWiPgy0A9cB/zlqDa/BGyehNgkSZIkSZJUUPwGdfHb1r5ZLenRpNwZ7ePAS4FXFvY7gLuHTkbELwLPBP5iUqKTJEmSJEmSJM0YZRWsM/MYcH1E/FLh0LrMHChqcgL4baBtkuKTJEmSJEmSJM0QF/X/jGTmmjGOdwPdlxCPJEmSJEmSJGmGuqiCdURUAi8EngzMy8w/KxyfA9QA+0e9eS1JkiRJkiRJ0rhmldshIl7K4FvU/wZ8DGguOv10YBfw2kmITZIkSZIkSZI0g5RVsI6IZwFfAxL4Q+Afis9n5k+ArQyuYy1JkiRJkiRJ0gUr9w3r9zP4YcVnZeYngI0l2rQCT7vUwCRJkiRJkiRJM0u5Bevrga9l5u5x2mwHllx8SJIkSZIkSZKkmajcgvU8YP8Ebaov4rqSJEmSJEmSpBmu3MLyTuApE7R5OrDl4sIZKSJeHRGfjIgfRERfRGRE/P047edFxJ9FRFdE9EfE4Yj4z4j4jTHa31a45lg/d0zGc0iSJEmSJEmSJlZRZvtvAXdExHMz84ejT0bEy4DnAB+ajOCA9zG4HvYxYAfwpLEaRsRC4AfALwFrgc8Cc4HfAv4tIu4qrLtdyteBjhLH2y4+dEmSJEmSJElSOcotWP9v4HXAdyLik8DVABFxE/A84K3ALuCeSYrvDxksVG8CbgAeHKdtM4PF6q8Cr83Ms4XY6oGHgY9GxLcys9SHIr+WmfdNUsySJEmSJEmSpItQ1pIgmbkTeAnQC7wbeA0QwL8U9ncBL83Mida5vtD7PZiZGzMzL6D5KwvbPx0qVheusQ/4GFAJuMSHJEmSJEmSJE1T5b5hTWa2R8S1wE3AdcBi4AjwE+DrxcXin7KmwrbU+tlDx144Rt+nR8Q7gDkMrtP9YGbumOT4JEmSJEmSJEnjKLtgDZCZ5xh8q/pfJjecS7IfWAI8Hlg36twTCtux1sC+a9T+uYj4PPCOzOyfvBAlSZIkSZIkSWOZ3dzcfLljuCAtLS1XA7cBq5ubm79a4vy1wDOBhpaWlq82NzcnQEQsBu4FFgEVLS0tH25ubj5b6LMAWMHgWtl/DHwa+C/gWcCLgZ9vbm7+ylgxRcRbWlpaPnvTTTctBbjmmmvo6uqirq6OlStXsnTpUtra2li8eDFr166lqqqKnTt3cuLECY4fP05vby9XXHEF69ato76+nvb29hF9urq6qKyspLe3l+PHj3Py5El27NhBVVUVa9asobGxcUSfRYsW8cgjjzB79mx2797N0aNH6e/vZ/v27Vx55ZV0dnbS1NTEihUrhvssXLiQDRs2EBHs3buXvr4+Tp8+zbZt26iurqazs5OlS5fS2trKVVddRVtbGwsWLGDTpk1kJvv37+fQoUOcPXuW7u5u5s2bx6pVq4avP3S/mpoaNm/ezMDAAAcPHuTAgQMMDAywZcsW5s+fP6JPY2MjK1euZP78+WzdupWzZ89y6NAh9u8fXGlm06ZNLFy4kI6OjuE+DQ0NrFq1irlz59Ld3c2ZM2c4cuQIe/fuJSLYsGEDixYtGtGnvr6ezs5Oqqur6enp4dSpU/T19bFnzx5mz57N+vXrWbx4Me3t7cPPXldXx5o1a6iqqmLHjh309/dz7Ngxdu3aRUVFxXD+h/q0trZSV1fHunXrhnM5Ov9r166loaFhRC5ra2vp6uqioqJiOJej8z86l0P5nzVrFnv27KGvr49Tp07R09MznP8lS5aMmf99+/Zx+PBhzpw5c17+29rahrc1NTXn5f/cuXNs2bKFefPm0dHRMfz3ampqor29fdz819TUDOeltbWVpqYmOjo6mDt3bsn8b9y4ccz8V1dXs23bNk6fPn1e/mtra0eMy/r6elavXs2cOXPo6emhv7+fo0ePsnv37hH5L+5TnP+hsXzs2DF6e3uprKykq6ur5Fguzn/xWC7O/4oVK4b/brW1taxfv37Msbx69eoR+W9tbS2Z/6GxPJT/0bkcGsvAiPyXGstD/3aG8j8wMMCBAwdG5H/0WB7Kf/FYLs7/5s2bWbBgwYhcNjY2Due/eCzv27dvRP5Xrlw5PMYaGhrGHMuzZs06L/9DfYbyv3379jHHsnO5c7lzuXO5c7lzuXO5c7lzuXP5dJ3LO9Zvghxg1rEDNJ077FzuXO5c7lw+qXP5kiVLWsar1U6luLDloUd1ingW8CsMFoFnl2iSmflnlxjb6Hs+n8GPLt6fma8vcb6JwWVJHgesAf4TqAZeARxl8O3raqAqM09PcK/HAKsYfL6nZ+aq8dovX748AZYtW1beQ0mSJEmSJF2Ej6z/7xVZ3/2ki/of6CVpPHG5blzWjBYRNcBXgRcwftAJTGrBeiKZuTsifhl4H/CbwB8Ah4B/LcSyBTgyUbG6cK3tEfFN4BbgeQwWryVJkiRJkiRJU6jc/wT3EeBG4AcMLrOxHbhcH1k8T2buY3A96hFrUkfEUIG9tYzL7Sts505OdJIkSZIkSZKk8ZRbsH4F0A68IDMHpiCeqfLmwvb+Mvo8u7DdMsmxSJIkSZIkSZJKmFVm+wXAg9OxWB0RsyJiXonjbwJ+B+hgVME6In6tRPuIiP8FXAfsB/59aiKWJEmSJEmSJBUr9w3rjUDjVARSSkTcDNxc2G0qbK+LiPsKv+/PzHcVfq8G9kTEfwCbCsd+jcGPQ24Gfjszz4y6xfcjYgODS4XsZLAgfz3wS8AJ4JbM7Jvcp5IkSZIkSZIklVJuwfrTwIci4qrM3DkVAY3ydODWUceeUPgB2AYMFaxPAQ8AzwVeXDi2GfgAcE9mHitx/Y8yWNC+EagFBoAeBp/znsx0ORBJkiRJkiRJ+ikpt2D9LQaLu/83IlqAFcDhUg0zs+cSYyMzm4HmC2x7Bnhjmdd/d/lRSZIkSZIkSZKmQrkF624ggQA+P067vIhrS5IkSZIkSZJmsHKLyl9ksBgtSZIkSZIkSdKkKqtgnZm3TVEckiRJkiRJkqQZbtblDkCSJEmSJEmSJCizYB0R5yLi/RO0eW9EnL20sCRJkiRJkiRJM025b1hH4edC2kmSJEmSJEmSdMGmYkmQRUD/FFxXkiRJkiRJkvQoNuFHFyPieaMOXV3iGMBs4LHALcAjkxCbJEmSJEmSJGkGmbBgDTwEZOH3BG4t/JQSwADwR5ccmSRJkiRJkiRpRrmQgvXdDBaqA/hTBgvY3yvR7hxwAHgwM9dPVoCSJEmSJEmSpJlhwoJ1ZjYP/R4RtwJfy8xPTGVQkiRJkiRJkqSZ50LesB6WmY+fqkAkSZIkSZIkSTPbrMsdgCRJkiRJkiRJMMEb1hHxXQofWszMHYX9C5GZ+cJLjk6SJEmSJEmSNGNMtCTI8wvb6lH7E8mLCUaSJEmSJEmSNHONW7DOzFnj7UuSJEmSJEmSNFnGLUBHxJ0R8Ss/rWAkSZIkSZIkSTPXRG9M/yXw0qGdiDgXEe+f2pAkSZIkSZIkSTPRRAXrfqCqaD8KP5IkSZIkSZIkTaqJCtZbgV+PiMaiY35QUZIkSZIkSZI06SYqWH8W+J9Ab0ScKxxrLiwNMt7P2akNW5IkSZIkSZL0aFMx3snM/ERE7AVuApYCLwB6gO6pD02SJEmSJEmSNJOMW7AGyMwHgAcAImIAuDcz757qwCRJkiRJkiRJM8tES4KM1gI8NAVxSJIkSZIkSZJmuAnfsC6WmS1TFYgkSZIkSZIkaWYr9w1rSZIkSZIkSZKmhAVrSZIkSZIkSdK0YMFakiRJkiRJkjQtWLCWJEmSJEmSJE0LFqwlSZIkSZIkSdOCBWtJkiRJkiRJ0rRgwVqSJEmSJEmSNC1YsJYkSZIkSZIkTQsWrCVJkiRJkiRJ04IFa0mSJEmSJEnStGDBWpIkSZIkSZI0LUzrgnVEvDoiPhkRP4iIvojIiPj7cdrPi4g/i4iuiOiPiMMR8Z8R8RsT3OfWiHg4Io5FxJGIeCgiXj75TyRJkiRJkiRJGsu0LlgD7wPeBjwd2Dlew4hYCPy40Occ8Fngy8D/AP4tIu4co99HgfuAJcBfA39f6PONiHjbpDyFJEmSJEmSJGlCFZc7gAn8IbAD2ATcADw4Tttm4JeArwKvzcyzABFRDzwMfDQivpWZG4c6RMRzgD8CNgO/nJmHCsc/Aqwo9PnXzOye5OeSJEmSJEmSJI0yrd+wzswHM3NjZuYFNH9lYfunQ8XqwjX2AR8DKoE7RvUZ2v/gULG60Kcb+DRQBdx+keFLkiRJkiRJksowrQvWZWoqbLeUODd07IWjjt9Y2P57iT7fGtVGkiRJkiRJkjSFHk0F6/2F7eNLnHtCYfukoQMRMRe4CjiWmbtK9BlaOuSaSYtQkiRJkiRJkjSm2c3NzZc7hgvS0tJyNXAbsLq5ufmrJc5fCzwTaGhpaflqc3NzAkTEYuBeYBFQ0dLS8uHm5uazLS0t9QyuX72vubn5nhLXOwO8H6hsbm7+81IxRcRbWlpaPnvTTTctBbjmmmvo6uqirq6OlStXsnTpUtra2li8eDFr166lqqqKnTt3cuLECY4fP05vby9XXHEF69ato76+nvb29hF9urq6qKyspLe3l+PHj3Py5El27NhBVVUVa9asobGxcUSfRYsW8cgjjzB79mx2797N0aNH6e/vZ/v27Vx55ZV0dnbS1NTEihUrhvssXLiQDRs2EBHs3buXvr4+Tp8+zbZt26iurqazs5OlS5fS2trKVVddRVtbGwsWLGDTpk1kJvv37+fQoUOcPXuW7u5u5s2bx6pVq4avP3S/mpoaNm/ezMDAAAcPHuTAgQMMDAywZcsW5s+fP6JPY2MjK1euZP78+WzdupWzZ89y6NAh9u8f/G8SmzZtYuHChXR0dAz3aWhoYNWqVcydO5fu7m7OnDnDkSNH2Lt3LxHBhg0bWLRo0Yg+9fX1dHZ2Ul1dTU9PD6dOnaKvr489e/Ywe/Zs1q9fz+LFi2lvbx9+9rq6OtasWUNVVRU7duygv7+fY8eOsWvXLioqKobzP9SntbWVuro61q1bN5zL0flfu3YtDQ0NI3JZW1tLV1cXFRUVw7kcnf/RuRzK/6xZs9izZw99fX2cOnWKnp6e4fwvWbJkzPzv27ePw4cPc+bMmfPy39bWNrytqak5L//nzp1jy5YtzJs3j46OjuG/V1NTE+3t7ePmv6amZjgvra2tNDU10dHRwdy5c0vmf+PGjWPmv7q6mm3btnH69Onz8l9bWztiXNbX17N69WrmzJlDT08P/f39HD16lN27d4/If3Gf4vwPjeVjx47R29tLZWUlXV1dJcdycf6Lx3Jx/lesWDH8d6utrWX9+vVjjuXVq1ePyH9ra2vJ/A+N5aH8j87l0FgGRuS/1Fge+rczlP+BgQEOHDgwIv+jx/JQ/ovHcnH+N2/ezIIFC0bksrGxcTj/xWN53759I/K/cuXK4THW0NAw5lieNWvWefkf6jOU/+3bt485lp3Lncudy53Lncudy53Lncudy53Lp+tc3rF+E+QAs44doOncYedy53LncufySZ3LlyxZ0jJWnXaqxYUtD335RcTzGfzo4v2Z+foS55uAnwCPA9YA/wlUA68AjgJLCvtVmXk6IpYCO4GdmflzJa5XCZwGTmXmnPFiW758eQIsW7bsop9PkiRJkiTpQn1k/fDnu3j3kyouYySSHqXict34UbMkSGbuBn4Z+AQwF/gDBovV/wq8CLgSOJKZpwtdjhS2C8a45IJR7SRJkiRJkiRJU+hRU7AGyMx9mXlXZj4hM6/IzMbMfCOD61oH0FrU9jiDb1jPi4glJS73xMJ2w5QHLkmSJEmSJEl6dBWsx/Hmwvb+Uce/W9i+tESfl41qI0mSJEmSJEmaQo+agnVEzIqIeSWOvwn4HaCD8wvWnyls3xsRi4r6XA28FTjF4AcbJUmSJEmSJElTbFqvyh8RNwM3F3abCtvrIuK+wu/7M/Ndhd+rgT0R8R/ApsKxXwN+BdgM/HZmnim+fmb+KCLuAd4JdEbEl4ErgNcCtcDbM7N70h9MkiRJkiRJknSeaV2wBp4O3Drq2BMKPwDbgKGC9SngAeC5wIsLxzYDHwDuycxjpW6QmX8UEZ3A24C3/P/t3XmcHFW58PHfk22y7xv7GkEDCIILAdlcroAIesFdwA31KuKCrwoq4X0vV7mCGgEBNxBxBRWuCypbDCAXsxASICEJELJACAlkmezJnPePqgmdTs+ke2Z6ujP5fT+f+tR0VZ06p57q7ql6uuoU0ARMA76dUvpTB22HJEmSJEmSJGkH6jphnVIaD4wvc9lNwEfbWM/PgJ+1pawkSZIkSZIkqWN0mT6sJUmSJEmSJEk7NxPWkiRJkiRJkqS6YMJakiRJkiRJklQXTFhLkiRJkiRJkuqCCWtJkiRJkiRJUl0wYS1JkiRJkiRJqgsmrCVJkiRJkiRJdcGEtSRJkiRJkiSpLpiwliRJkiRJkiTVBRPWkiRJkiRJkqS6YMJakiRJkiRJklQXTFhLkiRJkiRJkuqCCWtJkiRJkiRJUl0wYS1JkiRJkiRJqgsmrCVJkiRJkiRJdcGEtSRJkiRJkiSpLpiwliRJkiRJkiTVBRPWkiRJkiRJkqS6YMJakiRJkiRJklQXTFhLkiRJkiRJkuqCCWtJkiRJkiRJUl0wYS1JkiRJkiRJqgsmrCVJkiRJkiRJdcGEtSRJkiRJkiSpLpiwliRJkiRJkiTVBRPWkiRJkiRJkqS6YMJakiRJkiRJklQXTFhLkiRJkiRJkuqCCWtJkiRJkiRJUl0wYS1JkiRJkiRJqgsmrCVJkiRJkiRJdcGEtSRJkiRJkiSpLpiwliRJkiRJkiTVBRPWkiRJkiRJkqS6YMJakiRJkiRJklQX6jphHRFnRsRVEXFfRKyKiBQRN7eyfENEfDoi/hURyyKiMSJmRcT3I2KfEsufm6+zpeGT1d1CSZIkSZIkSVKzHrVuwA58DXg10AgsAg5uacGI6AHcDRwDzAZ+BWwAXgucD5wdEeNSSo+XKH47ML3E9Cntar0kSZIkSZIkqWz1nrD+PFmieh5wPHBvK8u+kyxZfTfw1pRSU/OMiLgU+AZwIfCREmVvSynd2EFtliRJkiRJkiS1QV13CZJSujelNDellMpYfP98/OfCZHXu9nw8ouNaJ0mSJEmSJEnqSPV+hXUlHsvHJ0fEhKKk9dvz8V0tlD08Ij4H9AYWA/emlBZVqZ2SJEmSJEmSpBK6UsL6z8DvgXcBMyPiLmAjcCRwLHAVcHULZS8oer0lIn4MfC6ltL5K7ZUkSZIkSZIkFeg+fvz4WrehLJdeeum+wLnAzPHjx/++eP748eO59NJLbwESWdL6aGAcsB8wEZiQUlpYtM5BwFSyvrK/AlwDPAQcBbwFOGD8+PG/a6lNEXHepZdeev2pp566O8ArXvEKZs2axfDhw3n44YfZfffdmTJlCsOGDeOxxx6joaGBxYsXs3btWtasWcOzzz5Lr169ePzxxxkxYgTTpk3bpsysWbPo2bMnzz77LGvWrGHdunUsWrSIhoYGHn30UUaNGrVNmSFDhvDEE0/QvXt3lixZwurVq1m/fj0LFy6kT58+zJgxg9GjRzN16tStZQYPHsycOXOICJYuXcqqVavYuHEjzzzzDH379mXGjBnsvvvuTJ48mT322IMpU6YwaNAg5s2bR0qJZcuW8dJLL7F582bmz59P//79eeSRR7auv7m+gQMH8uSTT9LU1MSLL77I8uXLaWpq4qmnnmLAgAHblBk1ahQPP/wwAwYM4Omnn2bz5s289NJLLFu2DIB58+YxePBgpk+fvrXMyJEjeeSRR+jXrx/z589n06ZNrFy5kqVLlxIRzJkzhyFDhmxTZsSIEcyYMYO+ffuyYMECNmzYwKpVq3j++efp3r07s2fPZtiwYUybNm3rtg8fPpxHH32UhoYGFi1axPr162lsbOS5556jR48eW/d/c5nJkyczfPhwHn/88a37snj/P/bYY4wcOXKbfTl06FBmzZpFjx49tu7L4v1fvC+b93+3bt14/vnnWbVqFRs2bGDBggVb9/9uu+3W4v5/4YUXWLFiBZs2bdpu/0+ZMmXreODAgdvt/y1btvDUU0/Rv39/pk+fvjVeo0ePZtq0aa3u/4EDB27dL5MnT2b06NFMnz6dfv36ldz/c+fObXH/9+3bl2eeeYaNGzdut/+HDh26zedyxIgRzJw5k969e7NgwQLWr1/P6tWrWbJkyTb7v7BM4f5v/iw3Njby7LPP0rNnT2bNmlXys1y4/ws/y4X7f+rUqVvjNnToUGbPnt3iZ3nmzJnb7P/JkyeX3P/Nn+Xm/V+8L5s/y8A2+7/UZ7n5vdO8/5uamli+fPk2+7/4s9y8/ws/y4X7/8knn2TQoEHb7MtRo0Zt3f+Fn+UXXnhhm/3/8MMPb/2MjRw5ssXPcrdu3bbb/81lmvf/woULW/ws+13ud7nf5X6X+13ud7nf5X6X+13ud3m9fpdPnz0PUhPdGpczessKv8v9Lve73O/yDv0u32233S5tKSdabVFe99C1FxEnkD108RcppQ+WmN8buAk4mezhircDa8kexPh9YB/grJTS7cVlS6xrL+ARYAhweErpkdaWv/baaxPApz71qQq2SJIkSZIkqW2+PXvz1r+/dHBXuoFeUp2IWlVc1w9drNBXgLOAi1NK16eUlqSUVqWU7gDOBHoCE8pZUX4l9l/yl8dVpbWSJEmSJEmSpG10pYR184MV7y2ekV8h/SKwT0QMK3N9L+Tjfh3QNkmSJEmSJEnSDnSlhHVDPh5RPCMiGoCB+cuNZa7v9fn4qXa2S5IkSZIkSZJUhq6UsL4vH1+UJ6gLjQd6AJNTSqubJ0bEG4tXEpmvkj20cRnw1+o0V5IkSZIkSZJUqK575Y+IM4Az8pej8/HREXFj/veylNKF+d+XAacBbwJmR8RfgXVkD118Xf73BUVVTIqIOcBkYDEwKF/+ELIHNn4gpbSqo7dLkiRJkiRJkrS9uk5YA4cD5xRN2z8fAJ4BLgRIKS2OiNcAXwZOBT5MdgX5c8CNwOUppdlF67qCLJl9EjAUaAIWANcA30kp2R2IJEmSJEmSJHWSuk5Yp5TGk3XnUe7yL5AlsC/c0bL58l9qU8MkSZIkSZIkSR2uK/VhLUmSJEmSJEnaiZmwliRJkiRJkiTVBRPWkiRJkiRJkqS6YMJakiRJkiRJklQXTFhLkiRJkiRJkuqCCWtJkiRJkiRJUl0wYS1JkiRJkiRJqgsmrCVJkiRJkiRJdcGEtSRJkiRJkiSpLpiwliRJkiRJkiTVBRPWkiRJkiRJkqS6YMJakiRJkiRJklQXTFhLkiRJkiRJkuqCCWtJkiRJkiRJUl0wYS1JkiRJkiRJqgsmrCVJkiRJkiRJdcGEtSRJkiRJkiSpLpiwliRJkiRJkiTVBRPWkiRJkiRJkqS6YMJakiRJkiRJklQXTFhLkiRJkiRJkuqCCWtJkiRJkiRJUl0wYS1JkiRJkiRJqgsmrCVJkiRJkiRJdcGEtSRJkiRJkiSpLpiwliRJkiRJkiTVBRPWkiRJkiRJkqS6YMJakiRJkiRJklQXTFhLkiRJkiRJkuqCCWtJkiRJkiRJUl0wYS1JkiRJkiRJqgsmrCVJkiRJkiRJdcGEtSRJkiRJkiSpLpiwliRJkiRJkiTVBRPWkiRJkiRJkqS6YMJakiRJkiRJklQX6jphHRFnRsRVEXFfRKyKiBQRN7eyfENEfDoi/hURyyKiMSJmRcT3I2KfVsqdk5dpjIiVETExIt5ena2SJEmSJEmSJJVS1wlr4GvAZ4DDgcWtLRgRPYC7gauBAcCvgOuApcD5wCMR8aoS5a4AbgR2A34E3AwcCvwxIj7TURsiSZIkSZIkSWpdj1o3YAc+DywC5gHHA/e2suw7gWPIktZvTSk1Nc+IiEuBbwAXAh8pmD4O+CLwJPDalNJL+fRvA1OBKyLiTyml+R24TZIkSZIkSZKkEur6CuuU0r0ppbkppVTG4vvn4z8XJqtzt+fjEUXTP5mPL2tOVuf1zgeuARqAD1fWakmSJEmSJElSW9R1wrpCj+XjkyOieLua+6O+q2j6Sfn4ryXWd0fRMpIkSZIkSZKkKqr3LkEq8Wfg98C7gJkRcRewETgSOBa4iqx/awAioh+wB9CYUnquxPrm5uNXlNuAa6+9tm0tlyRJkiRJqkD/gr+vba0DVUlqm/SpT30qalFxl7nCOu825ExgPHAQ8FmyPqtPBCYBv0wpbSkoMigfr2xhlc3TB7dUZ0ScFxFT2tFsSZIkSZIkSVKuy1xhHRG9gZuAk4FPk/VbvZbsQYzfByZFxFkppdtbXktJLfafnVL6IfDDtrVYkiRJkiRJklSoy1xhDXwFOAu4OKV0fUppSUppVUrpDrIrr3sCEwqWb76CehCl7egKbEmSJEmSJElSB+pKCevmBytu13NTSukR4EVgn4gYlk9bAywG+kfEbiXWNyYfz6lCWyVJkiRJkiRJRbpSwrohH48onhERDcDA/OXGgln35OO3lVjfyUXLSJIkSZIkSZKqqCslrO/LxxflCep2kmGYAAAgAElEQVRC48n6656cUlpdMP26fHxxRAxpnhgR+5L1g70BuKEajZUkSZIkSZIkbStSavGZgjUXEWcAZ+QvRwP/BjzFy8npZSmlC/Nl9wD+F9gTmA/8FVhH9tDF1+V/vyml9GBRHVcCXwAWAbcCvYD3AMOA81NKV1dp8yRJkiRJkiRJBeo9YT0euKSVRZ5JKe1bsPwI4MvAqcB+ZFeQP0fWrcflKaXZLdRzDvAZ4FVAEzAN+HZK6U/t3wpJkiRJkiRJUjnqOmEtSZIkSZIkSdp1dKU+rCVJkiRJkiRJOzET1pIkSZIkSZKkutCj1g3oKrp375aamtrevUpEUO3uWboFtKOJ7a+frIPwetAZ8d62vm6k1IlbHwE7U3c/0Q06Mz4dqSaxDqDz9289fYaljlTr/4+SJEmSVE969erBxo2bl6WURtSifhPWHaSpKZGmtL18HJWYNGlSxzWohOOOO450dVWraFV8BtJhZ9auAQVixq2c9sRHO62+Px70E9734MZOq+9XR/fiyFvmdlp97TX1rDFw0U6aLfqv4H3vebxTq/zVb14Fp0/o1DoBmm6/gHTM4Z1er1Rt8cD0dv0PlyRJkqSuJI7aDPBMreq3SxBJkiRJkiRJUl0wYS1JkiRJkiRJqgt2CVJFk6bBFT+HqbPh2Rfghkvg3NPavr6bb76ZSZMmsWDBAnr27MnYsWM577zz2H///Tuu0QWu+Qdc/wDMfzF7PXY0fO1tcOohVamubP+1dBYXL3mMTw87gKv3OKLq9T1x1TTmXP3wNtMahvfhrQ+8v2p1rlv2HNN/cDHPPfhXNq1dTf/d9+O1X7qaka85rmp1tmTmf5zAxhcWbzd94BHHM+aiH3d6e8rStAXuGw+P3gyNz0H/3WDsB+C48dCttl97mzatYcbM77No8V1s2PAiQwa/ktcc8VWGDTu04ypZNg+evBdWLIT1K+GI98Per395fkrwxF9h/j9h0zoYsg8cdiYM3K3j2qAdmrSykSsWL2XqmnU8u3ETNxy4F+eOGrZ1/u+Xr+D6JcuZ1riWZZu3cO8hB3DCoAE1bLEkSWqv1s4RN22Gr/0A7vgnPLkIBvaDE4+Cb50Pe4+ubbslSepMJqyrqHEtHHIgnH0qnH1J+9f38MMPc8YZZ3DwwQeTUuKnP/0pX/jCF7jpppsYOHBg+ysosucQuPx0GDMSmprgZw/BGT+EqV+Gw/bo8OrK8r9rlvOj5U9zWO9BnVpvv/0GMe7np2x9Hd2janVtXL2COz9xAiMOG8fxV9xOw5DhNC5+moahNennnoO/+bvsDZDbtGIps778ToaOO6WVUjX24OUw9Ro47Wcw4lBYOgP+dA70aIBjv17Tpv1r8tdZsWIOb3j9N+nbZxTzn/kj9/7jo5zytj/St++ojqlkywYYMBr2ei1Mu3n7+fPuhnn3wmveD/1HwhN/g3/+AN50MfTs3TFt0A41bmnikH69OXvkUM6eu33XYGu2NDFuQD8+OGIIZ89dUIMWSpKkjtbaOeLa9TBtNlz8ETj8FbCyEb74PXjb+TDjV9DDs3dJ0i7Cf3lVdMqx2QBw7qXtX9+VV165zeuLL76YU045hZkzZ3LMMce0v4Iipx+27evL3gHX3g8PPl2bhPXKLZv4wMJ/8ZO9juL/Pt+5D7nr1iPoPaJvp9Q16+Yr6DNsNEdfcsPWaf13369T6i6l56Bh27xeds8tdO/TnyFHn1yjFpVh8T9hzGnZADB4XxjzDnj2oZo2a/Pm9SxcdCfHHvM9Ro18HQCHHvIZFj87kXlP/prDDr2gYyoaNTYbAKb9Ytt5KcGT/4Axb4bd8wcovuYDcMfXYPFU2Lfjv0tU2ilDB3LK0OzHxnNLJKQ/NHIoAMs2be7UdkmSpOpp7RxxUH+48wfbTrv+Ihj7bpg1Hw49sFOaKElSzdmH9U5s7dq1NDU1MWBA9W8R39IEv54CjRtgXI1yp+ctmsqZg/bgpP4jO73uNQtXc+cbf8VdJ/2GqZ+/hzULV1WtrkWT/odhY1/HA197P78/ZQ/uOPso5tzyA1JKVauzXCkllt1zC0PfeDrdGvrUujkt2/NYeOZeWDY7e/3C4zD/HjigtleFp7SFlLbQvVvDNtO7d+/NCy9M65xGrF0OG1bByIMKGtALhh8ALz7dOW2QJElSWVatycZD7BVMkrQL8QrrndhVV13FmDFjGDt2bNXqmLkYjr4S1m+G/g3wh4/DoTW4uvpHy59i3sZGfr7X6zq97sGHjeCIbx5H//0Hs+HFdcy9djoPvPdPnPCnd9FrSMd3n9D47NPM/f11HPSez3LCh77EirkzmPqdzwHwirP+o8Prq8TqGfezcekihr/prJq2Y4eO/jJsXA0/fBV06w5Nm2HcxXBkbePXs2c/hg87nMcev45Bg8bQu/dwnlnwZ5Yvn07//nt3TiM2rM7GDUVnPQ0DYN3KzmmDJEmSdmjjJvjid+G0N8KeHdRznCRJOwMT1jupq6++mhkzZnDNNdfQvXv3qtVz0CiY/lVYsRZ+Nx3O+TlMvAAO2b1qVW7nifWruWjJo9x3wAn06tb5NwWMOn6vbV4PefVI7n7zb1l421wO+HAHPiivWVMTQw8+ksP/4zIAhh50BKsXzmXu766tecL6hbt+S98DDqPvfq+qaTt26PHfwMyb4PRfwoix8Px0uPMCGLwfHP7RmjbtDW/4Fg/962vc/scTiejOkCGvYu+9T+Gllzq3mxuiqB/2lKB6XbNLkiSpAps3wwe/Disa4X++U+vWSJLUuUxY74Suuuoq7r77biZMmMDuu1c3c9yrBxyYP+vvqH1g8gL47r3wkw9UtdptPLh2Ocu2bOSQOXdunbaFxKQ1y7hu+VOsOeQMGrpVL2lfrEe/ngw4cAhr5lenW5Dew3dj4H6v3GbawH0PZs1vr65KfeXatHI5KyffzV4f64AniFbbPV+C118IY9+bvR55KKx8Bh78Zs0T1gP6782bT7qJzZvXsmnTGvr0GcED//wC/fvt2TkNaL6yev0q6DPk5ekbG7e/6lqSJEmdbvNmeN/FMHMeTLwehg2udYskSepcJqx3MhMmTOCee+5hwoQJ7LPPPp1ef1OCDZ38/K8zBu3OUX3fss20Dy+cwpiG/lw08mB6Rededb1lw2Yan17B8NfvVpX1jzj0aFYvmLPNtNUL5tJvdCd1GdGC5RN/R/TsxdBjTq1pO8qyeW3WFUihbt0hNdWmPSX06NGXHj36snHjSp5b8gCHv/qLnVNx32HQMBBeeAKG5N8hWzbB8idh7Omd0wZJkiSVtGkzvPer8OiTMPGHMHp4rVskSVLnM2FdRY1rYd7C7O+mJliwBKY/AUMHwd6jK1/fd77zHf7+979z2WWXMWDAAJYvXw5Anz596Nu3bwe2PPOV2+HUsbDXEFi9Hn45BSbOhT9/ssOratXg7r0Y3L3XNtP6devO0O69OKT3oKrX/9jlDzHqxL3pu1t/Nry4jjk/mM6WtZvZ853VeUz3Qe+9gDvPO47Hbvwme7/pLF6aM505t1zDYZ/8f1WprxwpJZbdfQtDjzmV7n3616wdZTvwNHjwW1kXIMPHwvMPw0PfgUPPrnXLeO65+0k0MXDA/qxuXMD0R77NwAH7sv9+7+y4SjZvgDUv5C8SrHsJVi6Cnn2h71A44HiY83foPwr6j8j+7t4AexzZcW3QDjVu2cK8dRsAaCKxYMMmpjeuZWjPHuzd0IsXN21mwYaNrNiyBYB56zYyuPtaRvfqyehePWvZdEmS1EatnSPuPhzO+jJMfhz++N2st7Yly7JlB/WHPh3/+BxJkuqSCesqmvI4nFiQ3L3k+mw45+1w4/jK13fbbbcB8PnPf36b6eeeey4f+chH2tHS0pasgg/+DJashkG94bA94I5Pwb/VeffFHW39kjVM+8JENq5YT8OQ3gw+fCTH/vY0+u5Rne4Thr3qKN54+a3MuO7rPHrDf9Fv1F4cet54xvx7J/9SUKDxsYfY8Nx89jv/ipq1oSJvvQomfR3++h+wdin03w2O+Dgc+41at4xNm1bzyIzvsXbdEnr1GsRee76Vww69gG7dOjABuWIBPFDQhczsO7Jhr9fBaz4AB74pu6p6xq2waW12pfW4T0FPz4I605TGtZz46JNbX1+ycAmXLFzCOSOHcOOYffifF1fy4eYzWuDjT2Z/X7LXKMbvXZ07PCRJUnW1do44/jy4/R/Z9CM/uG25Gy6Bc0/rvHZKklRLJqyr6ISjIE3puPVNmjSp41ZWhhs/1KnVVWTiASd0Wl1HfvekTqur2R7HnMIex5zS6fW2ZMAhb+DIW+bWuhnlaxgAb/leNtSZvfc+mb33Prm6lQwfA6dPaHl+BBx8cjaoZk4YNIB0zOEtzj931DDOHTWsE1skSZKqbUfniB15/ihJ0s6qczv/lSRJkiRJkiSpBSasJUmSJEmSJEl1wYS1JEmSJEmSJKkumLCWJEmSJEmSJNUFE9aSJEmSJEmSpLoQKaVat6FLiIjVwBO1bkcXNRxYVutGdEHGtXqMbfUY2+oxttVjbKvH2FaHca0eY1s9xrZ6jG31GNvqMbbVY2x3DcOBfimlEbWovEctKu2inkgpHVXrRnRFETHF2HY841o9xrZ6jG31GNvqMbbVY2yrw7hWj7GtHmNbPca2eoxt9Rjb6jG2u4Z8P+9bq/rtEkSSJEmSJEmSVBdMWEuSJEmSJEmS6oIJ647zw1o3oAszttVhXKvH2FaPsa0eY1s9xrZ6jG11GNfqMbbVY2yrx9hWj7GtHmNbPcZ211DT/exDFyVJkiRJkiRJdcErrCVJkiRJkiRJdcGEtSRJkiRJkiSpLpiwboeI2DMifhoRz0bEhoiYHxHfi4ghtW5bPYiIMyPiqoi4LyJWRUSKiJt3UGZcRPwlIl6MiLURMSMiPhcR3Vsp8/aImBgRKyOiMSIeiohzOn6L6kdEDIuIj0XEHyJiXkSsy7f//oj4aESU/Gwb3/JExOURcXdELMxj+2JEPBwRl0TEsBbKGNs2iIgP5d8NKSI+1sIyFccpIs6JiH/ly6/My7+9OltRe/n/n9TCsKSFMr5nKxARb4yI30XEc/n//Oci4u8RcUqJZY3tDkTEua28Z5uHLSXKGdsyRcSp+Xt0Uf6/7KmIuCUijm5heWO7A5H5SET8b0SszuP0cER8tqU4GddtRR2fH8ROfuxQSWwjomdEXBARN0TE9IjYGK0cixWVrShOEdE9318z4uXj6r9ExLj2bG9nqTCuYyLiyxFxT2TnERsj4vmIuD0iTtxBPbtUXKHi2O4VET/IP8tLIjsWezYv++GI6NlKPca2jO/aovI/iZePxw5sYZmK4xQRfSLi0oh4IiLWR8TSiPhtRLyyLdupij9H+0brx96/LqO+hoh4NF9+UQvLvCUirowsp/Jivuz9FW9cSsmhDQNwAPA8kIDbgG8B9+SvZwPDat3GWg/A9Dweq4FZ+d83t7L86cBmoBH4CfDtPJYJuKWFMp/J5y8DrgG+CyzMp11R6xhUMbafzLfxWeAXwDeBnwIr8um3kvdRb3zbFN+NwP/mMf0WcBUwOd/uxcBexrZD4rxX/p5dnW/3xzoiTsAV+fyF+fLXAMvzaZ+p9XZXKZbz81iOLzFcWGJ537OVxfdr+Xa+ANwA/BfZQ0gmA/9tbNsU08NbeL+OB+7Ot/1PxrbN8b28YLt/TPa/7Fay/29NwAeNbZvielO+fc/ncZ0APIbHXpXEsC7PD+gCxw6VxBYYnM9PwBJgAS0ci7UnTkAAt/DyOfK38/3YmO/X02sdtw6O66/z+Y8B15Odo/0+39YEfNa4tjm2JwArgb8D15Edi11f8N69F+hhbCuPbYmypxWUTcCBHREnoAG4Py8zmexY5ZfAJmAN8Ppax21nHCr8HO2bz59O6WPwM8uo78qC98aiFpa5LZ+/DpiZ/31/xdtW6+DurAPwtzzo5xdN/04+/bpat7HWA3AiMCb/MjthBx+cgcBSYANwVMH03sA/87LvLSqzL7A+/4ezb8H0IcC8vMzRtY5DlWJ7Uv6PpFvR9NG8/E/7341vm+Pbu4Xpl+Xb/QNj2+4YB3AX8CTZAc52J0ltiRMwLp8+DxhStK7l+fr2rdZ21TCe84H5ZS7re7ay2J6Vb9+dwIAS83sa2w6P+YP5dr/D2LYpfqOBLWRJqJFF807Mt/spY1txXM9ojh0wvGB6T+AP+bxzjesO41h35wd0kWOHCmPbCzgZ2C1/PZ4dJKzbEifgfXmZByg4vgZem+/XpZT431pPQ4VxPRc4osT048l+MNzQHPNdPa5tfM92KzG9J1myOgHvNraVx7ao3Aiy44dfAxNpOWFdcZyAr+Zlbincl2Q/TCayH3q228cOHbeveTlhfWMb6zqB7MKH5gsoW0pYHw2MBboX1GnCupPeEPvnAX+6+AMFDCD7VWkN0K/Wba2XoYwPzkfy+T8rMe+kfN4/iqb/33z6pZWsr6sPwEX5tl9lfDs8tq/Ot/tOY9vuWF6Q/7M7jhZOktoSJ16++u3DJcq0uL6dfaCyhLXv2fLj2o0sObUGGGFsOyXmh+TbvAjobmzbFMPX59t2ewvzVwGrjW3FcW3+//LpEvOa37dTjWtFMT2BOjg/oAseO+wotiWWH8+OE9YVxwmYlE8/sZL11etQaVyLyv6doouKjGuHxfaCvOzFxrZ9sSX7AXYJMIzWE9YVxYksmfpMPn2/Stbn0HH7mnYkrMl+RJ5PnguhlYR1C3VWnLC2D+u2OSkf/z2l1FQ4I6W0muxXpr7AGzq7YTux5pj+tcS8ScBaYFxENJRZ5o6iZXYlm/Lx5oJpxrdjnJaPZxRMM7YVyvso+xYwIaU0qZVF2xKnXTm2DRHxwYi4KLJ+KU+M0v17+p4t3zhgP+AvwEuR9Qn85Ty+pfoBNrbt94l8/JOUUmEf1sa2fHPJruR7XUQML5wREceRXVxxV8FkY1ue0fn4qRLzmqe9JiIG538b1/brrBga9/JUFKd8v4wj20/3lVOmiyt1jgbGtV3yY93m54nMKJptbCsQEeeS3U30yZTS8laWa0ucDgD2BuaklJ4us4yqZ/eI+ER+3viJiDisjDLfJ7tb6aNVbttWPTqroi7moHw8p4X5c4G3Aq8g64dRO9ZiTFNKmyPiabJbCvYn65dnR2Wei4g1wJ4R0TeltLYKba47EdEDODt/WfiP2fi2QURcCPQHBgFHAceSHQh9q2AxY1uB/D36c7Kuay7aweIVxSki+gF7AI0ppedKrG9uPn5F21pf90aTxbbQ0xHx4ZTSPwqm+Z4t32vz8fPANODQwpkRMYmsr7cX8knGth0iog/wQbK7L35cNNvYliml9GJEfJmsm7rHI+I2slufDwDeQda9zScKihjb8izLx/uVmLd/wd8Hkz0Hw7i2X9Vj6LFDedoYpwPJbgd/KqVUnKRtqUyXFBH7AG8iS/BNKphuXCuU/xD7GbIrdkcAbyGLyS+BPxUsZ2wrkL9HJ5BdmXvbDhZvS5zKyaEVl1H1vCUftoqIicA5KaUFxQtHxDuBc8juwtlufrV4hXXbDMrHK1uY3zx9cAvztb22xLTcMoNamN8VfYvsttS/pJT+VjDd+LbNhcAlwOfIktV/Bd5akJwCY1upbwBHkPXzuW4Hy1Yap135u/kGshOh0UA/ssTq9WS3YN0REa8uWNb3bPlG5uNPAn2AN5NdnXoI2bMsjiPrh6+ZsW2fd5PF5o6U0sKieca2Aiml7wHvIrs45ePAV8j6Y19Idhvo0oLFjW15mhMhX4iIoc0T8x9iLy1Ybkg+Nq7t1xkx3JWPHSpRzX3RpWObX436C7IHzo1PKb1UMNu4Vm442fnZN4BPkf0YewXZuUUqWM7YlikiugE/I+va9rNlFDG2O6+1wP8DjiQ7XhlC1sf+vWTdidyd/9izVUSMIjuvvCOl9JPObKwJ6+qIfJxaXUqVaEtMd6n9EBGfBb5I9pTeD1VaPB8b3wIppdEppSBLAr6L7AqehyPiNRWsxtjmIuJ1ZFdVX5lSerAjVpmPK41Tl4orQErp0pTSPSml51NKa1NKj6aUPkl2hWUfsr4py+V79mXNXaoE2ZXUd6eUGlNKjwHvJOtn+fgWugcpxdi27rx8fH0byhrbAhHxf4BbgRvJTub7kZ2cPAX8IiL+u5LV5eNdPba/Jrtl+QCyK9d/GBHfA6aT3Y7efHXYlhbKFzOu7deZMTTm5fH9XCDvruLnwDHAb8gSq21hXHMppdn5+VkPYB/g82THD5MKf0ysZJUVLNtVY/t5sqTlx4t+UGkr/7/VqZTS0pTSN1JK01JKK/JhElkPEQ+RXT3/saJiPyJ7uOnHO7m5JqzbaEdXNwwsWk471paYlltmVTvatVOIiE+T3cLzONmDCl4sWsT4tkOeBPwD2Rf5MLIHSTQztmUo6ApkDvD1MotVGqcdLb+jX/a7ouvy8XEF03zPlq/5oP2plNIjhTPyOwSa72R5XT42tm0UEa8i6w9xEVmf4cWMbZki4gTgcuB/UkpfSCk9lf+QNY3sh5bFwBcjorkbC2Nbhvy5Ne8gu/tqCdnFAR8he88eS9btCkDz1evGtf06I4YeO5SnLXHapc+Z82T1zWR3t/wW+GDRFcBgXNsspbQlpbQgpTSBrJurN5A9SLGZsS1DRIwBLgNuSCmVOv4qpZrfzV0mtjuTvGuX5u74tp43RsTZZM/xuiCltLiz22XCum2eyMct9a8zJh+31D+PttdiTPNE135kD6h4qswyu5FdTbSoq/fxFxGfA64GHiVLVi8psZjx7QAppWfIfhQYW/AgK2Nbnv5k2/tKYH1EpOaB7LY+gB/l076Xv64oTimlNWSJmP75/GK74ndzc+Kk8NYu37Pla97uFS3Mb05o9yla3thWrqWHLTYztuV7ez6+t3hGvq3/IjsHOCKfbGzLlFLanFK6MqV0eEqpT0ppYErpbWTHBocD64DH8sWNa/tVPYYeO5SnjXGaR3bHwf75/iqnTJeQb++vgPeS9a38/lJ9/RrXDtP8wL4TmicY27KNJeuu5sOF52f5Odrx+TJz82ln5K/bEidzaPWvudvTwvPG5jvLf1bi/QGwR8G0Du/OxYR12zSfALw17+9nq4gYQHbLzzqyB66oPPfk47eVmHcc0Bf4Z0ppQ5llTi5apkvKH6r0XbLbUU8s6pOykPHtOLvn4+aEirEtzwbgJy0MD+fL3J+/bu4upC1x2hVj25rmrioKT+Z9z5ZvElkyZExE9Cox/5B8PD8fG9s2iIjeZFerNpF9B5RibMvXkI9HtDC/efrGfGxs2+9DQG/gtymlTfk049p+nRVD416eiuKU75d/ku2nN5ZTpivIjxduJbuy+ibgQy38ENvMuLbfHvm4+EcBY7tj82n5HK35Qrhb8tfzoc1xehJYALwiIko9vLgrxnZn84Z8XHje+CAtvz8g6xO7+XXh/+KOkVJyaMNAdhtwAs4vmv6dfPp1tW5jPQ1kv3YmsqfOlpo/kOwXnQ3AUQXTe5N9GSbgvUVl9gPWk92CuW/B9CFkv/ol4Ohab3sVY/r1fBunAEN3sKzxLT+uBwOjS0zvRna7VAIeMLYdGvPx+TZ/rL1xIutSIOXzhxRM3zdfz/rCdXWFgezKiO2+A8j69Zubx+Oigum+ZyuL78359v1n0fS3kCVYVwCDjW27YvyhfDv/2Moyxrb8eL4737YlwB5F807O37frgGHGtuLYDiwx7bXAi8BqYH/fsxXF8wTq4PyALnjssKPYllh+PCWOxdobJ+B9eZkHgN4F01+b79elpT5X9TqU8Z5tAP6cL/NjoFsZ69zl41pmbF8P9C0xvT9wZ172MmNbeWxbKTcxL3dgiXkVxwn4al7mlsLPBnB6Pv2xcj4zDm3f1/nnqFeJ6Sfln4cEjCuzrkR2x9KOlts3X/b+Srcn8hWoQhFxANmB0kjgdmAW2c4/kew2hnEppeUtr6Hry28Zab5tZDTwb2S/1tyXT1uWUrqwaPlbyT4ovyY7+H8HcFA+/d2p6A0bEecD3yf7h/MbsquFzgT2JHuw24V0QRFxDtmDlLYAV1G6r6f5KaUbC8oY3zLkXax8m+zKyifJtn0U2S1R+5MlAN6UUnq8oIyxbYeIGE/WLcjHU0o/LppXcZwi4krgC2T9it4K9ALeQ9b/+PkppaurtjE1kMfvK2R3/zxNljQ5ADiV7KT+L8A7U0obC8r4ni1TRIwkOxg/kOz/17/Ifgx4J9nB1/tTSrcULG9sKxQR95H1AfyOlNIfW1nO2JYhv/vvb8Cbyb4P/kD2v+uVZN2FBPC5lPX72VzG2JYhIh4iS/Y/ShbbsWQPXNwAvCul9Lei5Y1rkXo9P+gKxw5tiO1XyC7UgKxLm1eTnd82P0D0/hLHZRXFKSKCrP/mM8keDP/HfNn3kB2j/HtK6fZ2bXiVVRLXiLgBOBdYBvyA0g+Qm5hSmlhUxy4XV6g4treRJeP+QXal7lpgL7IfYgeTvXf/LaXUWFSHsS3j+6CFdUwkOwcek1KaVzSv4jhFRAPZFdTjyC66uxvYm+xuhI3ASSmlh9qwubu0Cj9HE8mOXSaSfSYADiNLWAN8PaX0n2XWm4DFKaU9S8w7lpcf3tgf+HeyHzGau+8hpXTuDivp7Ix/VxrIviBvAJ4j+4A9Q/bgu1avdt1VBl7+pb6lYX6JMseQJVdeIjshmEn21NrurdRzGtk/rtXAGmAycE6tt7/GsU1kB0PGt/LYHgJcQ9bNyjKyW8tW5ts9vqXPt7FtV8yb388lr+ppS5yAc/Ll1uTl/gG8vdbbWqX4HU/WT+Jssqt9N5FdkXYncDZkP06XKOd7tvwYDyW7g+ppsv/3y8l+rH6DsW13bF+Zf/4XthYfY1txXHsCnyPrnm5V/r9sKfAn4K3Gts1x/RIwNf+u3ZB/J1xHK1ffGtfttm08dXp+wE5+7CWIzMcAAAENSURBVFBpbHn56smWhhs7Ik5Aj3x/zcz330v5/izrKr5aD5XEtYyYJmC8cW1TbE8FfkF2ceBKsuPdpcBdwHlAj1bqMbZlfNeWWEfz+3m7K6zbGiey575cSvbD2Aayc5ZbgFfVOmY761Dh5+ijZMeC84HGfB8sIPtx940V1pto4Qprsh/uWv0uLKcOr7CWJEmSJEmSJNUFH7ooSZIkSZIkSaoLJqwlSZIkSZIkSXXBhLUkSZIkSZIkqS6YsJYkSZIkSZIk1QUT1pIkSZIkSZKkumDCWpIkSZIkSZJUF0xYS5IkSZIkSZLqgglrSZIkSZIkSVJdMGEtSZIkSZIkSaoLJqwlSZIkSZIkSXXh/wNLMq1n8JBW9gAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<Figure size 1728x576 with 2 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "#Plotting the predicted fitness effect for the variants seen more than once\n",
    "plt.close('all')\n",
    "f, (ax1, ax2) = plt.subplots(2, 1, sharey=True, sharex = True, figsize=(24, 8))\n",
    "gs = matplotlib.gridspec.GridSpec(2, 1, width_ratios=[1], height_ratios=[10,0.5])\n",
    "ax1 = plt.subplot(gs[0])\n",
    "ax2 = plt.subplot(gs[1])\n",
    "gs.update(hspace=0.05)\n",
    "\n",
    "nonsense_list = []\n",
    "y_list = []\n",
    "x_list=[]\n",
    "y_list_top20 = []\n",
    "x_list_top20=[]\n",
    "y_list_nonsense = []\n",
    "x_list_nonsense = []\n",
    "x=[]\n",
    "y=[]\n",
    "OBS_EXP_DICT={}\n",
    "top20_lower_err = []\n",
    "top20_upper_err = []\n",
    "missense_lower_err = []\n",
    "missense_upper_err = []\n",
    "nonsense_lower_err = []\n",
    "nonsense_upper_err = []\n",
    "\n",
    "top20 = c1\n",
    "missense = c6\n",
    "nonsense = c10\n",
    "synonymous = c3\n",
    "\n",
    "axisfont=20\n",
    "domainfont = 12\n",
    "titlefont=12\n",
    "axislabelfont=12\n",
    "\n",
    "#ERROR NEEDS TO BE IN THE FORMAT: [tuple(lower_err),tuple(upper_err)]\n",
    "\n",
    "for k, v in ASXL1_missense_s.items():\n",
    "    if k in top_20_ASXL1:\n",
    "        codon_number = int(((re.findall('\\d+', k)))[0])\n",
    "        s_value = float(v[0])*100\n",
    "        x_list_top20.append(codon_number)\n",
    "        y_list_top20.append(s_value)\n",
    "        \n",
    "        s_upper = float(v[3])*100\n",
    "        s_lower = float(v[4])*100\n",
    "        error_upper = s_upper-s_value\n",
    "        error_lower = s_value-s_lower\n",
    "        top20_lower_err.append(error_lower)\n",
    "        top20_upper_err.append(error_upper)\n",
    "\n",
    "    if k not in top_20_ASXL1:\n",
    "        codon_number = int(((re.findall('\\d+', k)))[0])\n",
    "        s_value = float(v[0])*100\n",
    "        x_list.append(codon_number)\n",
    "        y_list.append(s_value)\n",
    "        \n",
    "        s_upper = float(v[3])*100\n",
    "        s_lower = float(v[4])*100\n",
    "        error_upper = s_upper-s_value\n",
    "        error_lower = s_value-s_lower\n",
    "        missense_lower_err.append(error_lower)\n",
    "        missense_upper_err.append(error_upper)\n",
    "\n",
    "    missense_err = [tuple(missense_lower_err), tuple(missense_upper_err)]   \n",
    "    \n",
    "for k, v in ASXL1_nonsense_s.items():\n",
    "    if k in top_20_ASXL1:\n",
    "        codon_number = int(((re.findall('\\d+', k)))[0])\n",
    "        s_value = float(v[0])*100\n",
    "        x_list_top20.append(codon_number)\n",
    "        y_list_top20.append(s_value)\n",
    "        \n",
    "        s_upper = float(v[3])*100\n",
    "        s_lower = float(v[4])*100\n",
    "        error_upper = s_upper-s_value\n",
    "        error_lower = s_value-s_lower\n",
    "        top20_lower_err.append(error_lower)\n",
    "        top20_upper_err.append(error_upper)\n",
    "    \n",
    "    top20_err = [tuple(top20_lower_err), tuple(top20_upper_err)]\n",
    "        \n",
    "    if k not in top_20_ASXL1:\n",
    "        codon_number = int(((re.findall('\\d+', k)))[0])\n",
    "        s_value = float(v[0])*100\n",
    "        x_list_nonsense.append(codon_number)\n",
    "        y_list_nonsense.append(s_value)\n",
    "        \n",
    "        s_upper = float(v[3])*100\n",
    "        s_lower = float(v[4])*100\n",
    "        error_upper = s_upper-s_value\n",
    "        error_lower = s_value-s_lower\n",
    "        nonsense_lower_err.append(error_lower)\n",
    "        nonsense_upper_err.append(error_upper)\n",
    "\n",
    "    nonsense_err = [tuple(nonsense_lower_err), tuple(nonsense_upper_err)]        \n",
    "    \n",
    "msize = 12\n",
    "medgewidth = 2\n",
    "linew = 3\n",
    "csize = 4\n",
    "cthick = 3\n",
    "\n",
    "markers1, caps1, bars1 = ax1.errorbar(x_list, y_list, yerr = missense_err, fmt = 'o', ecolor = missense, elinewidth = linew, \\\n",
    "             capsize = csize, capthick = cthick, markersize = msize, markeredgewidth = medgewidth, \\\n",
    "             markeredgecolor = missense, markerfacecolor = missense, zorder = 1)    \n",
    "\n",
    "markers2, caps2, bars2 = ax1.errorbar(x_list_nonsense, y_list_nonsense, yerr = nonsense_err, fmt = 'o', ecolor = nonsense, elinewidth = linew, \\\n",
    "             capsize = csize, capthick = cthick, markersize = msize, markeredgewidth = medgewidth, \\\n",
    "             markeredgecolor = nonsense, markerfacecolor = nonsense, zorder = 2)    \n",
    "\n",
    "markers3, caps3, bars3 = ax1.errorbar(x_list_top20, y_list_top20, yerr = top20_err, fmt = 'o', ecolor = top20, elinewidth = linew, \\\n",
    "             capsize = csize, capthick = cthick, markersize = msize, markeredgewidth = medgewidth, \\\n",
    "             markeredgecolor = top20, markerfacecolor = top20, zorder = 3)    \n",
    "\n",
    "# loop through bars and caps and set the alpha value\n",
    "a = 0.7\n",
    "[bar.set_alpha(a) for bar in bars1]\n",
    "[cap.set_alpha(a) for cap in caps1]\n",
    "[bar.set_alpha(a) for bar in bars2]\n",
    "[cap.set_alpha(a) for cap in caps2]\n",
    "[bar.set_alpha(a) for bar in bars3]\n",
    "[cap.set_alpha(a) for cap in caps3]\n",
    "\n",
    "#Only show the required axis lines\n",
    "ax1.spines['top'].set_visible(False)\n",
    "ax1.spines['right'].set_visible(False)\n",
    "\n",
    "for axis in ['bottom','left']:\n",
    "    ax1.spines[axis].set_linewidth(2)\n",
    "\n",
    "for axis in ['bottom','left']:\n",
    "    ax1.spines[axis].set_color(grey3)\n",
    "    \n",
    "ax1.set_ylabel('fitness effect (s)', fontsize = axisfont, labelpad = 10)\n",
    "ax1.set_title('Fitness effects across ASXL1', fontsize = 18)\n",
    "\n",
    "# ax1.set_yscale('log')\n",
    "\n",
    "\n",
    "# # STACKED BAR CHART\n",
    "# define the exons + create data\n",
    "exons = [{'1':(19-1)+1, '2': (47-20)+1, '3': (84-48)+1, '4':(125-85)+1, '5':(157-126)+1, '6': (189-158)+1,\\\n",
    "          '7': (240-190)+1,'8': (294-241)+1, '9': (327-295)+1, '10': (362-328)+1, '11': (573-363)+1, \\\n",
    "          '12': (1541-574)+1}]\n",
    "exon_orders = [['1', '2', '3', '4', '5', '6', '7', '8', '9', '10', '11', '12']]\n",
    "colors2 = [c4, c0, c3, c1, c5, c6,c7, c8, c9, c10, c2,c4]\n",
    "names = exons[0].keys()\n",
    "values = np.array([[data[name] for name in order] for data,order in zip(exons, exon_orders)])\n",
    "lefts = np.insert(np.cumsum(values, axis=1),0,0, axis=1)[:, :-1]\n",
    "orders = np.array(exon_orders)\n",
    "\n",
    "# #plot the bar chart\n",
    "for name, color in zip(names, colors2):\n",
    "    idx = np.where(orders == name)\n",
    "    value = values[idx]\n",
    "    left = lefts[idx]\n",
    "    ax2.bar(x=left, height=-10, width=value, bottom=-5, color=color, orientation = 'horizontal', label = name, edgecolor='k', linewidth=1, tick_label = name)\n",
    "    ax2.text(left + (value/2.), (-0.26), name, ha='center', va='center', fontsize = 14)\n",
    "    \n",
    "# # CONFIGURING THE GRAPH APPEARANCE\n",
    "# #Set the x and y axis limits\n",
    "ax1.set_xlim(0, 1542)\n",
    "ax1.set_ylim(180, 220)\n",
    "ax2.set_xlim(0, 1542)\n",
    "ax2.set_ylim(-0.5, 0)\n",
    "\n",
    "#x-axis ticks\n",
    "x1_major_ticks = []\n",
    "x1_major_tick_labels = []\n",
    "x2_major_ticks = [0, 100, 200, 300, 400, 500, 600, 700, 800, 900, 1000, 1100, 1200, 1300, 1400, 1541]\n",
    "x2_major_tick_labels = [\"0\", \"100\", \"200\", \"300\", \"400\", \"500\", \"600\", \"700\", \"800\", \"900\", \"1000\", \"1100\", \"1200\", \"1300\", \"1400\", \"1541\"]\n",
    "ax1.set_xticks(x1_major_ticks)\n",
    "ax1.set_xticklabels(x1_major_tick_labels, fontsize = axisfont)\n",
    "ax2.set_xticks(x2_major_ticks)\n",
    "ax2.set_xticklabels(x2_major_tick_labels, fontsize = axisfont)\n",
    "\n",
    "#y-axis ticks\n",
    "y1_major_ticks = [180, 185, 190, 195, 200, 205, 210, 215, 220]\n",
    "y1_major_tick_labels = [\"180\", \"185\", \"190\", \"195\", \"200\", \"205\", \"210\", \"215\", \"220\"]\n",
    "ax1.set_yticks(y1_major_ticks)\n",
    "ax1.set_yticklabels(y1_major_tick_labels, fontsize = axisfont)\n",
    "\n",
    "#set panel legends\n",
    "legend_marker_size = 2\n",
    "legend_elements = [Line2D([0], [0], marker = '_', color = missense, alpha = 1.0, markersize = legend_marker_size, \\\n",
    "                          lw=3, label='missense'),\n",
    "                   Line2D([0], [0], marker = '_', color=nonsense, alpha=1.0, markersize = legend_marker_size, \\\n",
    "                          lw=3, label='nonsense'),\n",
    "                  Line2D([0], [0], marker = '_', color=top20, alpha=1.0, markersize = legend_marker_size, \\\n",
    "                          lw=3, label='top 20 observed variant')]\n",
    "    \n",
    "ax1.legend(ncol=1, handles=legend_elements, loc='upper left', frameon=0, fontsize = 20)\n",
    "\n",
    "ax1.grid(which='major', linestyle=':', linewidth='0.7', color=grey3, zorder = 0)\n",
    "\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## TP53 variants"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 260,
   "metadata": {
    "code_folding": [
     0,
     86,
     97,
     109
    ]
   },
   "outputs": [],
   "source": [
    "def expected_number_variants_missense_TP53(data, params):\n",
    "    s_value = params[0]\n",
    "    \n",
    "    a = 0\n",
    "    expected_number_by_variant_missense = {}\n",
    "    \n",
    "    Jaiswal_fmin = 0.056\n",
    "    Jaiswal_mean = 58\n",
    "    Jaiswal_sigma = 11.5\n",
    "    Jaiswal_total = 17182\n",
    "    \n",
    "    Genovese_fmin = 0.1194\n",
    "    Genovese_mean = 55.1\n",
    "    Genovese_total = 12380\n",
    "    Genovese_sigma = 11.95\n",
    "    \n",
    "    McKerrel_fmin = 0.008\n",
    "    McKerrel_mean = 51.88\n",
    "    McKerrel_total = 4219\n",
    "    McKerrel_sigma = 18.36\n",
    "    \n",
    "    Zink_fmin = 0.2363\n",
    "    Zink_mean = 60\n",
    "    Zink_total = 11262\n",
    "    Zink_sigma = 1.18907720e+01\n",
    "    \n",
    "    Coombs_fmin = 0.019\n",
    "    Coombs_mean = 59.82\n",
    "    Coombs_total = 5649\n",
    "    Coombs_sigma = 15.14\n",
    "    \n",
    "    Young_fmin = 0.0009\n",
    "    Young_mean = 63\n",
    "    Young_total = 89\n",
    "    Young_sigma = 6.8\n",
    "    \n",
    "    Desai_fmin = 0.0169\n",
    "    Desai_mean = 66.1\n",
    "    Desai_total = 424\n",
    "    Desai_sigma = 6 #estimated\n",
    "    \n",
    "    Acuna_fmin = 0.0015\n",
    "    Acuna_mean = 4.5\n",
    "    Acuna_total = 2014\n",
    "    \n",
    "    N = 9.39358404e+04\n",
    "\n",
    "    for i in data:\n",
    "        variant_mutation_rate = mutation_rates_variants_TP53(i)\n",
    "\n",
    "        if i in Genovese_missense_TP53:\n",
    "            Genovese_number = prevalence_averaged_over_ages(Genovese_fmin, s_value, N, variant_mutation_rate, Genovese_mean, Genovese_sigma, Genovese_total)\n",
    "        else:\n",
    "            Genovese_number = 0 \n",
    "\n",
    "        if i in Zink_missense_TP53:\n",
    "            Zink_number = prevalence_averaged_over_ages(Zink_fmin, s_value, N, variant_mutation_rate, Zink_mean, Zink_sigma, Zink_total)\n",
    "        else:\n",
    "            Zink_number = 0\n",
    "\n",
    "        if i in Coombs_missense_TP53:\n",
    "            Coombs_number = prevalence_averaged_over_ages(Coombs_fmin, s_value, N, variant_mutation_rate, Coombs_mean, Coombs_sigma, Coombs_total)\n",
    "        else:\n",
    "            Coombs_number = 0\n",
    "\n",
    "        if i in Young_missense_TP53:\n",
    "            Young_number = prevalence_averaged_over_ages(Young_fmin, s_value, N, variant_mutation_rate, Young_mean, Young_sigma, Young_total)\n",
    "        else:\n",
    "            Young_number = 0\n",
    "\n",
    "        if i in Desai_missense_TP53:\n",
    "            Desai_number = prevalence_averaged_over_ages(Desai_fmin, s_value, N, variant_mutation_rate, Desai_mean, Desai_sigma, Desai_total)\n",
    "        else:\n",
    "            Desai_number = 0\n",
    "\n",
    "        if i in Acuna_missense_TP53:\n",
    "            Acuna_number = prevalence_averaged_over_uniform_ages(Acuna_fmin, s_value, N, variant_mutation_rate, Acuna_mean, Acuna_total)\n",
    "        else:\n",
    "            Acuna_number = 0\n",
    "\n",
    "        expected_total = Genovese_number + Zink_number + Coombs_number + Young_number + Desai_number + Acuna_number\n",
    "\n",
    "        expected_number_by_variant_missense[i] = expected_total\n",
    "    \n",
    "    return expected_number_by_variant_missense\n",
    "\n",
    "def prob_observed_given_expected_missense_TP53(params, data): #data = list of variant (e.g. p.R882H) (or all_missense_variants_TP53)\n",
    "    total_square_distance = 0\n",
    "    a = 0\n",
    "    for variant in data:\n",
    "        observed_number = float(all_studies_TP53_counts[variant])\n",
    "        expected_number = float(expected_number_variants_missense_TP53(data, params)[variant])\n",
    "        square_distance = ((observed_number - expected_number)**2)\n",
    "        total_square_distance = total_square_distance + square_distance\n",
    "\n",
    "    return total_square_distance\n",
    "\n",
    "def prob_observed_given_expected_missense_TP53_upper_error(params, data): #data = list of variant (e.g. p.R882H) (or all_missense_variants_TP53)\n",
    "    total_square_distance = 0\n",
    "    a = 0\n",
    "    for variant in data:\n",
    "        observed_number = float(all_studies_TP53_counts[variant])\n",
    "        observed_number_upper_error = observed_number+np.sqrt(observed_number)\n",
    "        expected_number = float(expected_number_variants_missense_TP53(data, params)[variant])\n",
    "        square_distance = ((observed_number_upper_error - expected_number)**2)\n",
    "        total_square_distance = total_square_distance + square_distance\n",
    "\n",
    "    return total_square_distance   \n",
    "\n",
    "def prob_observed_given_expected_missense_TP53_lower_error(params, data): #data = list of variant (e.g. p.R882H) (or all_missense_variants_TP53)\n",
    "    total_square_distance = 0\n",
    "    a = 0\n",
    "    for variant in data:\n",
    "        observed_number = float(all_studies_TP53_counts[variant])\n",
    "        observed_number_lower_error = observed_number-np.sqrt(observed_number)\n",
    "        expected_number = float(expected_number_variants_missense_TP53(data, params)[variant])\n",
    "        square_distance = ((observed_number_lower_error - expected_number)**2)\n",
    "        total_square_distance = total_square_distance + square_distance\n",
    "\n",
    "    return total_square_distance   "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 261,
   "metadata": {
    "code_folding": [
     0,
     91,
     102,
     114
    ]
   },
   "outputs": [],
   "source": [
    "def expected_number_variants_nonsense_TP53(data, params):\n",
    "    s_value = params[0]\n",
    "    \n",
    "    a = 0\n",
    "    expected_number_by_variant_nonsense = {}\n",
    "    \n",
    "    Jaiswal_fmin = 0.056\n",
    "    Jaiswal_mean = 58\n",
    "    Jaiswal_sigma = 11.5\n",
    "    Jaiswal_total = 17182\n",
    "    \n",
    "    Genovese_fmin = 0.1194\n",
    "    Genovese_mean = 55.1\n",
    "    Genovese_total = 12380\n",
    "    Genovese_sigma = 11.95\n",
    "    \n",
    "    McKerrel_fmin = 0.008\n",
    "    McKerrel_mean = 51.88\n",
    "    McKerrel_total = 4219\n",
    "    McKerrel_sigma = 18.36\n",
    "    \n",
    "    Zink_fmin = 0.2363\n",
    "    Zink_mean = 60\n",
    "    Zink_total = 11262\n",
    "    Zink_sigma = 1.18907720e+01\n",
    "    \n",
    "    Coombs_fmin = 0.019\n",
    "    Coombs_mean = 59.82\n",
    "    Coombs_total = 5649\n",
    "    Coombs_sigma = 15.14\n",
    "    \n",
    "    Young_fmin = 0.0009\n",
    "    Young_mean = 63\n",
    "    Young_total = 89\n",
    "    Young_sigma = 6.8\n",
    "    \n",
    "    Desai_fmin = 0.0169\n",
    "    Desai_mean = 66.1\n",
    "    Desai_total = 424\n",
    "    Desai_sigma = 6 #estimated\n",
    "    \n",
    "    Acuna_fmin = 0.0015\n",
    "    Acuna_mean = 4.5\n",
    "    Acuna_total = 2014\n",
    "\n",
    "    N = 9.39358404e+04\n",
    "    \n",
    "    for i in data:\n",
    "        variant_mutation_rate = mutation_rates_variants_TP53(i)\n",
    "\n",
    "        if i in Jaiswal_nonsense_TP53:\n",
    "            Jaiswal_number = prevalence_averaged_over_ages(Jaiswal_fmin, s_value, N, variant_mutation_rate, Jaiswal_mean, Jaiswal_sigma, Jaiswal_total)\n",
    "        else:\n",
    "            Jaiswal_number = 0\n",
    "\n",
    "        if i in Genovese_nonsense_TP53:\n",
    "            Genovese_number = prevalence_averaged_over_ages(Genovese_fmin, s_value, N, variant_mutation_rate, Genovese_mean, Genovese_sigma, Genovese_total)\n",
    "        else:\n",
    "            Genovese_number = 0 \n",
    "\n",
    "        if i in Zink_nonsense_TP53:\n",
    "            Zink_number = prevalence_averaged_over_ages(Zink_fmin, s_value, N, variant_mutation_rate, Zink_mean, Zink_sigma, Zink_total)\n",
    "        else:\n",
    "            Zink_number = 0\n",
    "\n",
    "        if i in Coombs_nonsense_TP53:\n",
    "            Coombs_number = prevalence_averaged_over_ages(Coombs_fmin, s_value, N, variant_mutation_rate, Coombs_mean, Coombs_sigma, Coombs_total)\n",
    "        else:\n",
    "            Coombs_number = 0\n",
    "\n",
    "        if i in Young_nonsense_TP53:\n",
    "            Young_number = prevalence_averaged_over_ages(Young_fmin, s_value, N, variant_mutation_rate, Young_mean, Young_sigma, Young_total)\n",
    "        else:\n",
    "            Young_number = 0\n",
    "\n",
    "        if i in Desai_nonsense_TP53:\n",
    "            Desai_number = prevalence_averaged_over_ages(Desai_fmin, s_value, N, variant_mutation_rate, Desai_mean, Desai_sigma, Desai_total)\n",
    "        else:\n",
    "            Desai_number = 0\n",
    "\n",
    "        if i in Acuna_nonsense_TP53:\n",
    "            Acuna_number = prevalence_averaged_over_uniform_ages(Acuna_fmin, s_value, N, variant_mutation_rate, Acuna_mean, Acuna_total)\n",
    "        else:\n",
    "            Acuna_number = 0\n",
    "\n",
    "        expected_total = Jaiswal_number + Genovese_number + Zink_number + Coombs_number + Young_number + Desai_number + Acuna_number\n",
    "\n",
    "        expected_number_by_variant_nonsense[i] = expected_total\n",
    "    \n",
    "    return expected_number_by_variant_nonsense\n",
    "\n",
    "def prob_observed_given_expected_nonsense_TP53(params, data): #data = list of variant (e.g. p.R882H) (or all_nonsense_variants_TP53)\n",
    "    total_square_distance = 0\n",
    "    a = 0\n",
    "    for variant in data:\n",
    "        observed_number = float(all_studies_TP53_counts[variant])\n",
    "        expected_number = float(expected_number_variants_nonsense_TP53(data, params)[variant])\n",
    "        square_distance = ((observed_number - expected_number)**2)\n",
    "        total_square_distance = total_square_distance + square_distance\n",
    "\n",
    "    return total_square_distance\n",
    "\n",
    "def prob_observed_given_expected_nonsense_TP53_upper_error(params, data): #data = list of variant (e.g. p.R882H) (or all_nonsense_variants_TP53)\n",
    "    total_square_distance = 0\n",
    "    a = 0\n",
    "    for variant in data:\n",
    "        observed_number = float(all_studies_TP53_counts[variant])\n",
    "        observed_number_upper_error = observed_number+np.sqrt(observed_number)\n",
    "        expected_number = float(expected_number_variants_nonsense_TP53(data, params)[variant])\n",
    "        square_distance = ((observed_number_upper_error - expected_number)**2)\n",
    "        total_square_distance = total_square_distance + square_distance\n",
    "\n",
    "    return total_square_distance   \n",
    "\n",
    "def prob_observed_given_expected_nonsense_TP53_lower_error(params, data): #data = list of variant (e.g. p.R882H) (or all_nonsense_variants_TP53)\n",
    "    total_square_distance = 0\n",
    "    a = 0\n",
    "    for variant in data:\n",
    "        observed_number = float(all_studies_TP53_counts[variant])\n",
    "        observed_number_lower_error = observed_number-np.sqrt(observed_number)\n",
    "        expected_number = float(expected_number_variants_nonsense_TP53(data, params)[variant])\n",
    "        square_distance = ((observed_number_lower_error - expected_number)**2)\n",
    "        total_square_distance = total_square_distance + square_distance\n",
    "\n",
    "    return total_square_distance   "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 262,
   "metadata": {
    "code_folding": [
     0
    ],
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "#CALCULATING S FOR MISSENSE VARIANTS\n",
    "TP53_missense_s = {}\n",
    "a = 0\n",
    "for var in all_missense_variants_TP53:\n",
    "    initial_guess=[0.05]\n",
    "    if var in all_studies_TP53_counts.keys():\n",
    "        observed_number = float(all_studies_TP53_counts[var])\n",
    "#         print(observed_number)\n",
    "        if observed_number >1:\n",
    "            variant_mutation_rate = mutation_rates_variants_TP53(var)\n",
    "            \n",
    "            MLE = scipy.optimize.minimize(prob_observed_given_expected_missense_TP53, initial_guess, args=([var],), method='Nelder-Mead')\n",
    "            s_value = MLE.x[0]\n",
    "\n",
    "            upper_error_MLE = scipy.optimize.minimize(prob_observed_given_expected_missense_TP53_upper_error, \\\n",
    "                                                      initial_guess, args=([var],), method='Nelder-Mead')\n",
    "            s_value_upper = upper_error_MLE.x[0]\n",
    "\n",
    "            lower_error_MLE = scipy.optimize.minimize(prob_observed_given_expected_missense_TP53_lower_error, \\\n",
    "                                                      initial_guess, args=([var],), method='Nelder-Mead')\n",
    "            s_value_lower = lower_error_MLE.x[0]\n",
    "\n",
    "#             print('a = ', a)\n",
    "            TP53_missense_s[var]=(s_value, observed_number, variant_mutation_rate, s_value_upper, s_value_lower) \n",
    "    a = a+1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 263,
   "metadata": {
    "code_folding": [
     0
    ],
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "#CALCULATING S FOR NONSENSE VARIANTS\n",
    "TP53_nonsense_s = {}\n",
    "a = 0\n",
    "for var in all_nonsense_variants_TP53:\n",
    "    initial_guess=[0.05]\n",
    "    if var in all_studies_TP53_counts.keys():\n",
    "        observed_number = float(all_studies_TP53_counts[var])\n",
    "#         print(observed_number)\n",
    "        if observed_number >1:\n",
    "            variant_mutation_rate = mutation_rates_variants_TP53(var)\n",
    "            \n",
    "            MLE = scipy.optimize.minimize(prob_observed_given_expected_nonsense_TP53, initial_guess, args=([var],), method='Nelder-Mead')\n",
    "            s_value = MLE.x[0]\n",
    "\n",
    "            upper_error_MLE = scipy.optimize.minimize(prob_observed_given_expected_nonsense_TP53_upper_error, \\\n",
    "                                                      initial_guess, args=([var],), method='Nelder-Mead')\n",
    "            s_value_upper = upper_error_MLE.x[0]\n",
    "\n",
    "            lower_error_MLE = scipy.optimize.minimize(prob_observed_given_expected_nonsense_TP53_lower_error, \\\n",
    "                                                      initial_guess, args=([var],), method='Nelder-Mead')\n",
    "            s_value_lower = lower_error_MLE.x[0]\n",
    "\n",
    "#             print('a = ', a)\n",
    "            TP53_nonsense_s[var]=(s_value, observed_number, variant_mutation_rate, s_value_upper, s_value_lower) \n",
    "    a = a+1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 264,
   "metadata": {
    "code_folding": [
     0
    ]
   },
   "outputs": [],
   "source": [
    "#Exac - retrieve the frequencies in Exac for each of the variants\n",
    "import csv\n",
    "filename=\"TP53_Exac.csv\"\n",
    "\n",
    "with open(filename, 'r') as csvfile:\n",
    "    read_reader = csv.reader(csvfile)  #csv.reader returns a reader object which will iterate over lines in the csvfile\n",
    "    row_count=0\n",
    "    Exac_TP53_results={} \n",
    "    \n",
    "    for row in read_reader:\n",
    "#         print(row)\n",
    "        if row_count>1:\n",
    "            variant = row[6]\n",
    "            allele_freq = float(row[14])\n",
    "            african_freq = float(row[15])/float(row[16])\n",
    "            east_asian_freq = float(row[18])/float(row[19])\n",
    "            european_freq = float(row[21])/float(row[22])\n",
    "#             print(row[24])\n",
    "            if row[25]=='0':\n",
    "                finnish_freq = '-'\n",
    "            if row[25]!='0':\n",
    "#                 print(row)\n",
    "                finnish_freq = float(row[24])/float(row[25])\n",
    "            latino_freq = float(row[27])/float(row[28])\n",
    "            south_asian_freq = float(row[33])/float(row[34])\n",
    "            Exac_TP53_results[variant]=(allele_freq, african_freq, east_asian_freq, european_freq, finnish_freq, latino_freq, south_asian_freq)\n",
    "\n",
    "        row_count=row_count+1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 265,
   "metadata": {
    "code_folding": [
     0
    ]
   },
   "outputs": [],
   "source": [
    "#COSMIC - retrieve the frequencies in COSMIC for each of the variants\n",
    "import csv\n",
    "filename=\"TP53_COSMIC_variant_counts.csv\"\n",
    "with open(filename, 'r') as csvfile:\n",
    "    read_reader = csv.reader(csvfile)  #csv.reader returns a reader object which will iterate over lines in the csvfile\n",
    "    row_count=0\n",
    "    TP53_COSMIC_results={} \n",
    "    \n",
    "    for row in read_reader:\n",
    "#         print(row)\n",
    "        if row_count>1:\n",
    "            variant = row[1]\n",
    "            counts = int(row[2])\n",
    "            TP53_COSMIC_results[variant]=counts\n",
    "\n",
    "        row_count=row_count+1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 266,
   "metadata": {
    "code_folding": [
     0
    ],
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "#Create dictionary of results and save to .csv file\n",
    "TP53_s_dictionary={}\n",
    "for k, v in TP53_missense_s.items():\n",
    "#     print(k)\n",
    "    if k in Exac_TP53_results.keys():\n",
    "        exac_freq = round((Exac_TP53_results[k][0])*100, 4)\n",
    "        african_freq = round((Exac_TP53_results[k][1])*100, 4)\n",
    "        east_asian_freq = round((Exac_TP53_results[k][2])*100, 4)\n",
    "        european_freq = round((Exac_TP53_results[k][3])*100, 4)\n",
    "        finnish_freq = round((Exac_TP53_results[k][4])*100, 4)\n",
    "        latino_freq = round((Exac_TP53_results[k][5])*100, 4)\n",
    "        south_asian_freq = round((Exac_TP53_results[k][6])*100, 4)\n",
    "    if k not in Exac_TP53_results.keys():\n",
    "        exac_freq = '-'\n",
    "        african_freq = '-'\n",
    "        east_asian_freq = '-'\n",
    "        european_freq = '-'\n",
    "        finnish_freq = '-'\n",
    "        latino_freq = '-'\n",
    "        south_asian_freq = '-'\n",
    "    if k in TP53_COSMIC_results.keys():\n",
    "        COSMIC_number = TP53_COSMIC_results[k]\n",
    "    if k not in TP53_COSMIC_results.keys():\n",
    "        COSMIC_number = '-'\n",
    "    TP53_s_dictionary[k]=(round(v[0]*100, 2),round(v[4]*100, 2),round(v[3]*100, 2), v[1], v[2], COSMIC_number, exac_freq, \\\n",
    "                            african_freq, east_asian_freq, european_freq, finnish_freq, latino_freq, south_asian_freq)\n",
    "\n",
    "for k, v in TP53_nonsense_s.items():\n",
    "    if k in Exac_TP53_results.keys():\n",
    "        exac_freq = round((Exac_TP53_results[k][0])*100, 4)\n",
    "        african_freq = round((Exac_TP53_results[k][1])*100, 4)\n",
    "        east_asian_freq = round((Exac_TP53_results[k][2])*100, 4)\n",
    "        european_freq = round((Exac_TP53_results[k][3])*100, 4)\n",
    "        finnish_freq = round((Exac_TP53_results[k][4])*100, 4)\n",
    "        latino_freq = round((Exac_TP53_results[k][5])*100, 4)\n",
    "        south_asian_freq = round((Exac_TP53_results[k][6])*100, 4)\n",
    "    if k not in Exac_TP53_results.keys():\n",
    "        exac_freq = '-'\n",
    "        african_freq = '-'\n",
    "        east_asian_freq = '-'\n",
    "        european_freq = '-'\n",
    "        finnish_freq = '-'\n",
    "        latino_freq = '-'\n",
    "        south_asian_freq = '-'\n",
    "    if k in TP53_COSMIC_results.keys():\n",
    "        COSMIC_number = TP53_COSMIC_results[k]\n",
    "    if k not in TP53_COSMIC_results.keys():\n",
    "        COSMIC_number = '-'\n",
    "    TP53_s_dictionary[k]=(round(v[0]*100, 2),round(v[4]*100, 2),round(v[3]*100, 2), v[1], v[2], COSMIC_number, exac_freq, \\\n",
    "                            african_freq, east_asian_freq, european_freq, finnish_freq, latino_freq, south_asian_freq)\n",
    "\n",
    "TP53_df = pd.DataFrame.from_dict(TP53_s_dictionary, orient='index', columns=['s (%)', 's lower value (%)', 's upper value (%)','observed number', 'site-specific mutation rate', 'Number of times in COSMIC',\\\n",
    "                                                                                'Exac Overall freq (%)', 'Exac African freq (%)', 'Exac East Asian freq (%)',\\\n",
    "                                                                                'Exac European freq (%)', 'Exac Finnish freq (%)', 'Exac Latino freq (%)', 'Exac South Asian freq (%)'])\n",
    "TP53_df_sorted = TP53_df.sort_values(by = ['s (%)'], ascending = False)\n",
    "\n",
    "TP53_df_sorted.to_csv('TP53_s_across_whole_gene.csv')\n",
    "# TP53_s_dictionary"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 267,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>s (%)</th>\n",
       "      <th>s lower value (%)</th>\n",
       "      <th>s upper value (%)</th>\n",
       "      <th>observed number</th>\n",
       "      <th>site-specific mutation rate</th>\n",
       "      <th>Number of times in COSMIC</th>\n",
       "      <th>Exac Overall freq (%)</th>\n",
       "      <th>Exac African freq (%)</th>\n",
       "      <th>Exac East Asian freq (%)</th>\n",
       "      <th>Exac European freq (%)</th>\n",
       "      <th>Exac Finnish freq (%)</th>\n",
       "      <th>Exac Latino freq (%)</th>\n",
       "      <th>Exac South Asian freq (%)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>p.N235S</th>\n",
       "      <td>24.27</td>\n",
       "      <td>19.53</td>\n",
       "      <td>29.95</td>\n",
       "      <td>5.0</td>\n",
       "      <td>6.154280e-10</td>\n",
       "      <td>6</td>\n",
       "      <td>0.0239</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0345</td>\n",
       "      <td>0.0913</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.S241C</th>\n",
       "      <td>15.55</td>\n",
       "      <td>12.05</td>\n",
       "      <td>17.67</td>\n",
       "      <td>2.0</td>\n",
       "      <td>7.420070e-10</td>\n",
       "      <td>9</td>\n",
       "      <td>0.0008</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0152</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.Y220C</th>\n",
       "      <td>14.91</td>\n",
       "      <td>12.67</td>\n",
       "      <td>16.50</td>\n",
       "      <td>4.0</td>\n",
       "      <td>1.473330e-09</td>\n",
       "      <td>145</td>\n",
       "      <td>0.0025</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0045</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.G245C</th>\n",
       "      <td>14.45</td>\n",
       "      <td>11.27</td>\n",
       "      <td>16.26</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.042000e-09</td>\n",
       "      <td>10</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.K132E</th>\n",
       "      <td>14.03</td>\n",
       "      <td>10.48</td>\n",
       "      <td>15.97</td>\n",
       "      <td>2.0</td>\n",
       "      <td>9.527030e-10</td>\n",
       "      <td>9</td>\n",
       "      <td>0.0008</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0015</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.R273H</th>\n",
       "      <td>13.25</td>\n",
       "      <td>11.77</td>\n",
       "      <td>14.30</td>\n",
       "      <td>5.0</td>\n",
       "      <td>3.959610e-09</td>\n",
       "      <td>106</td>\n",
       "      <td>0.0026</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0047</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.R110L</th>\n",
       "      <td>12.75</td>\n",
       "      <td>9.98</td>\n",
       "      <td>14.22</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.914550e-09</td>\n",
       "      <td>7</td>\n",
       "      <td>0.0017</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0015</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0061</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.R248Q</th>\n",
       "      <td>11.82</td>\n",
       "      <td>10.29</td>\n",
       "      <td>12.82</td>\n",
       "      <td>4.0</td>\n",
       "      <td>5.597280e-09</td>\n",
       "      <td>189</td>\n",
       "      <td>0.0058</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0231</td>\n",
       "      <td>0.0075</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.V216M</th>\n",
       "      <td>10.51</td>\n",
       "      <td>7.77</td>\n",
       "      <td>11.96</td>\n",
       "      <td>2.0</td>\n",
       "      <td>3.213910e-09</td>\n",
       "      <td>60</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.R110C</th>\n",
       "      <td>10.50</td>\n",
       "      <td>8.26</td>\n",
       "      <td>11.96</td>\n",
       "      <td>2.0</td>\n",
       "      <td>5.336040e-09</td>\n",
       "      <td>-</td>\n",
       "      <td>0.0008</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0151</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.R248W</th>\n",
       "      <td>10.39</td>\n",
       "      <td>8.26</td>\n",
       "      <td>11.56</td>\n",
       "      <td>2.0</td>\n",
       "      <td>5.336040e-09</td>\n",
       "      <td>129</td>\n",
       "      <td>0.0008</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0015</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.W146*</th>\n",
       "      <td>9.12</td>\n",
       "      <td>6.93</td>\n",
       "      <td>10.14</td>\n",
       "      <td>2.0</td>\n",
       "      <td>5.105670e-09</td>\n",
       "      <td>9</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.R283H</th>\n",
       "      <td>8.21</td>\n",
       "      <td>6.73</td>\n",
       "      <td>9.04</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.883910e-08</td>\n",
       "      <td>2</td>\n",
       "      <td>0.0066</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.003</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0262</td>\n",
       "      <td>0.0183</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         s (%)  s lower value (%)  s upper value (%)  observed number  \\\n",
       "p.N235S  24.27              19.53              29.95              5.0   \n",
       "p.S241C  15.55              12.05              17.67              2.0   \n",
       "p.Y220C  14.91              12.67              16.50              4.0   \n",
       "p.G245C  14.45              11.27              16.26              2.0   \n",
       "p.K132E  14.03              10.48              15.97              2.0   \n",
       "p.R273H  13.25              11.77              14.30              5.0   \n",
       "p.R110L  12.75               9.98              14.22              2.0   \n",
       "p.R248Q  11.82              10.29              12.82              4.0   \n",
       "p.V216M  10.51               7.77              11.96              2.0   \n",
       "p.R110C  10.50               8.26              11.96              2.0   \n",
       "p.R248W  10.39               8.26              11.56              2.0   \n",
       "p.W146*   9.12               6.93              10.14              2.0   \n",
       "p.R283H   8.21               6.73               9.04              2.0   \n",
       "\n",
       "         site-specific mutation rate Number of times in COSMIC  \\\n",
       "p.N235S                 6.154280e-10                         6   \n",
       "p.S241C                 7.420070e-10                         9   \n",
       "p.Y220C                 1.473330e-09                       145   \n",
       "p.G245C                 1.042000e-09                        10   \n",
       "p.K132E                 9.527030e-10                         9   \n",
       "p.R273H                 3.959610e-09                       106   \n",
       "p.R110L                 1.914550e-09                         7   \n",
       "p.R248Q                 5.597280e-09                       189   \n",
       "p.V216M                 3.213910e-09                        60   \n",
       "p.R110C                 5.336040e-09                         -   \n",
       "p.R248W                 5.336040e-09                       129   \n",
       "p.W146*                 5.105670e-09                         9   \n",
       "p.R283H                 1.883910e-08                         2   \n",
       "\n",
       "        Exac Overall freq (%) Exac African freq (%) Exac East Asian freq (%)  \\\n",
       "p.N235S                0.0239                     0                        0   \n",
       "p.S241C                0.0008                     0                        0   \n",
       "p.Y220C                0.0025                     0                        0   \n",
       "p.G245C                     -                     -                        -   \n",
       "p.K132E                0.0008                     0                        0   \n",
       "p.R273H                0.0026                     0                        0   \n",
       "p.R110L                0.0017                     0                        0   \n",
       "p.R248Q                0.0058                     0                   0.0231   \n",
       "p.V216M                     -                     -                        -   \n",
       "p.R110C                0.0008                     0                        0   \n",
       "p.R248W                0.0008                     0                        0   \n",
       "p.W146*                     -                     -                        -   \n",
       "p.R283H                0.0066                     0                        0   \n",
       "\n",
       "        Exac European freq (%) Exac Finnish freq (%) Exac Latino freq (%)  \\\n",
       "p.N235S                 0.0345                0.0913                    0   \n",
       "p.S241C                      0                0.0152                    0   \n",
       "p.Y220C                 0.0045                     0                    0   \n",
       "p.G245C                      -                     -                    -   \n",
       "p.K132E                 0.0015                     0                    0   \n",
       "p.R273H                 0.0047                     0                    0   \n",
       "p.R110L                 0.0015                     0                    0   \n",
       "p.R248Q                 0.0075                     0                    0   \n",
       "p.V216M                      -                     -                    -   \n",
       "p.R110C                      0                0.0151                    0   \n",
       "p.R248W                 0.0015                     0                    0   \n",
       "p.W146*                      -                     -                    -   \n",
       "p.R283H                  0.003                     0               0.0262   \n",
       "\n",
       "        Exac South Asian freq (%)  \n",
       "p.N235S                         0  \n",
       "p.S241C                         0  \n",
       "p.Y220C                         0  \n",
       "p.G245C                         -  \n",
       "p.K132E                         0  \n",
       "p.R273H                         0  \n",
       "p.R110L                    0.0061  \n",
       "p.R248Q                         0  \n",
       "p.V216M                         -  \n",
       "p.R110C                         0  \n",
       "p.R248W                         0  \n",
       "p.W146*                         -  \n",
       "p.R283H                    0.0183  "
      ]
     },
     "execution_count": 267,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "TP53_df_sorted"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 274,
   "metadata": {
    "code_folding": [
     0
    ]
   },
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAABZoAAAHvCAYAAADdMGqTAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDIuMi4zLCBodHRwOi8vbWF0cGxvdGxpYi5vcmcvIxREBQAAIABJREFUeJzs3Xd4VFX+x/H3CSkkQAolIDUgAooKgggCLqEqIB0VYVdiXWUFWRXXHrDCig0p+rMCouuKAorSlI6g9N4h1BBIQmghkHJ+f9yZ2ZnMBJBQ9fN6njyTnHvuud97z9x5yDeH7zXWWkREREREREREREREzlbQxQ5ARERERERERERERC5vSjSLiIiIiIiIiIiISKEo0SwiIiIiIiIiIiIihaJEs4iIiIiIiIiIiIgUihLNIiIiIiIiIiIiIlIoSjSLiIiIiIiIiIiISKEo0SwiIiJyGTPGBBtjrDHmo4sdy6XIGNPXGLPRGHPSdZ0qutpbGmN+NcYcdbX/9WLHKiIiIiJyOVOiWUREROQSZIyJdyVAC/pqdIp9g4wxA40xHS9kzJcaY0xrYBiwBngI+BuQbowpDXwLFAUed7UvOE8x1HPNReXzMf6fkTFm92nuDe+vpl5/jPH+yjLGbDLGDDXGxPyO8aPz9b3XGDPdtU+WMeaAMeYXY8w9xhj9riUiIiJ/KsEXOwAREREROaUvgR8DtG8BsNbmGGPCgRyvbUFAIvAx8N15j/DS1dr1eq+19rC70RgTD0QCL1hrz/f1qYczFz8BO8/zsf4s+gLFvH6uDTwNjAcm5eu70ev7pcA7ru9LArcDTwCtjDENrLXZXn3XAoMDHPtYvp/rA2nAcOAAUMI17migCfD3MzslERERkcufEs0iIiIil7Zl1trPT9XBWpt1oYK5zJQDcr2TzF7tAOkXOJ7LjjHGAMWstUcvdixu1toJ3j8bY1rhJJpXBrpXjDHu33l259s+zBjzI9AWJznsPe6+0913rlgeDdD8jjFmKvCAMeY5a23q6cYRERER+SPQf+cSERERuYzlr9FsjKkOuFdm3u/1X/5z8vd3lRWYZ4zJNMakGmP+zxhTLMAxKhhj3jfG7HLVOt7j+rl0vn6ljDHvGmO2ucoIpBljlhhjHs/XL8EYs9gYk2GMOWaM2WqM+dwYU+oMz/lWY8wMY8wh13FWGmMe8tpe3RhjcUpiFPG6Bj8ZY3bjrPQGmOd9bVz7FjXGPG+MWeca+6Ax5jtjTJ0AcRhjzN+NMb8Zp9bzEWPMKmNMomv7K8CH+Y7lPVfhxpiXjFNDOtN1rNXGmEArafMfO9gV5zxjTIprXnYYY0YYY0oWsM8dxpg5ruuWaYzZ4JqvYNf2Vq74/mac2tbrgRNAf68x4l3X8bBrjKXGmIQAx7rOGPON671ywhiTbIyZaYxp69XnrM//HJrueq2ef4PrGpc4y3F34vyuFXW2gYmIiIhcbrSiWUREROTSFpE/oQucsNYeKaD/PqA3zn/dn83/kqp5+frVB7q4tn8OtAAexCnB0cfdyRhTFfgFKOLquw24CngEaG6ckgPuFcPfAjcD7wOrcMobXA3EA2+5xrsX+ASYA7wIHAcqA+2A0jhlCApkjHkEGOGK6WUgE7gV+MAYU9Va+4zrGvzNFWMj1/UASMYpmdEeuN+1/yb3tTHGhOIkHhsCY3DqO8e4rssvxpim1trlrr4Gp6zJXcBC4FUgw3W+3YBBwNc4q6e9jwWusieu63QP8JlrjBDXtW1xqmvgUhSn7MM3wESckg434dSibmLylYIwxgwBnsIpCfEmkAJc6Yr1OcB7xfITrvP+2HUtd7jG6IxTniIZeMN1zLuBT13X3p1gLwPMBHJd57gTZ24buGKccg7O/1y5yvWaf9VxE5z3VogxJgPnGj9rrU0ONIhxajcH41y3tjjntR7Yfj6CFhEREbkUGWvtxY5BRERERPIxTh3hWQVs/spa28PVLxhnBfPH1toHCmrzGte9LQ9oaK1d4rVtGk5SONpae9zV9gNOUrqetXavV9+GOMneRGvtK65VtGnAe9bafqc4r++ApkAZa23uGV4O974Vga2u878n37YROPVwr7TWuhOjnwM9rLXB+fo+gLPS+BZr7Xyv9gHAEKCNtfYnr/ZonAcKbrDWtnK19QTG4ST077PW5nn1D3L/XNCxXNsOA7Ottb/7oY3GedBcmHuevNr/jpPA7Wat/dbV1hjnYYc/Abdba0949TfW9QuBcUpQzMCZx1reJR+MMSE4SdMI4Bpr7T5XexjOHw0aAFdZa7cZY7riJMA9MRRwDmd9/vnGccf9grX2lQDb3e/5KTgJYHASwh1x6jAfdcWe6uo/BZgPbABCgeY4fyzYA9zkPvd8x1gBuFe957niedham1SYcxMRERG5nKh0hoiIiMil7f9wHmrn/eWXTDsL872TzC4zcRJrVQBcyeO2OKs5TxpjSru/cBK+24E2rn0zcZJ5jYwxVU5x3EM4D0xr51oV/Hvc4YrvE+9YXPF8j7PquuXvHNPbX3FW/K7IN3Yw8DPQzJVYBegFWOBJ7yQzQP6fT+EQcJ0xpvbvDdRam+f1x4AixphoV6wzXV0aenXv5Xp92jvJ7Bon0KqTzwLUFW4AVAA+8k60usYbivN7hTthfMj12u40pSfO+vzPUlucB/YdwFldPhRYDbT2Pl9rbVtr7avW2m+stV9aax8C7gUq4TzYMZC/49wLvXFWfYcC0eftTEREREQuQSqdISIiInJp2+y9uvYc2hagzV22wl0ruRZgcJJofy9gnFxwHkhonFrMbwFJxpi1OEnPCdZa75XZr+CsaP4OSDXGzMFZafrVGTxw7mrXa0ErvQHKnmaMU6mFkyA8cIo+JXFKR1yF83C5wjzo7TGcFdFrjDFbcc7rO2ByAQlgH8aYHjhlLurglJ3wFuP1/VU487TqDOPaFKCtqut1bYBta1yv1VyvM4EvcFYB32OMWYxTkuQra+0Gr/0Kdf5n4Rf+lyg+ASRZa3edyY7W2jHGqbndvoDtv3r9OMYY8wYw3xhznbVW5TNERETkT0GJZhEREZE/p1OVrTD5Xkfj1HEOJNP9jbV2uDFmAk4y7i/AnUBfY8w4a+1fXX02GmNqAa1wVh83Az4CBhpj/nKapJw7nl7A/gL6bCmg/UwEASuAAafok+4VS6GSodbab12J9nY416E18AAw2xjTxrvGcn7GmDtxakQvAvoBu4EsnET5D/j+z8Xfu3I8M0DbGY/hShL3cj3Ury1wC0596BeMMX2ttaNc/c76/M/SgUL+0WYHThmZMzEaeBKnVMegQhxTRERE5LKhRLOIiIjIH8+5Wg26xTVWyJkm6Ky1e3DKffyfqzbuOJyk45vuB+m5yi384PrCGNMRmAT0x1nlWpDNrtfCJgxPNX4Z4OczWFG7Eac0ROnTrGo+5TjW2jRgLDDWVUrkDZxVyrcDE06x699wEsLNrbVZ7kZjzLUFxNoKuA5Ydqp4TmGr6zVQmYtrXK8+q+SttatxSlP821WG5TecmsijvPqc7flfUK7YrsR5iOKZCHe9ljw/EYmIiIhcelSjWUREROQPxvWQvSwKmeSy1qbglDy4wxjTIP924yjj+j7CGBOeb/8cnEQj7lhcdYTzW+bd5xS+Ak4CLxljigaIJ9oYE3qaMU5lDE4d4oDJbmOMd1mOcTirfP+dv9Z0vp/d5UBK5usTbIyJ8m5zJbdXBOofQC7OQ+c8/553Hff5AH2/cL2+nv/6/I462YtxHoZ3vzEm1mv/UJyVu3k4ZS8wxpTMP661Nh1IAoobY0LPwfmfF8aYUgVs6gdcgVML3N03xJVAD6Sv63XROQxPRERE5JKmFc0iIiIif0yLgFuNMU8Bu4Bca+1/z2KcvwPzcerNjgGW4/wbsirQGfgYp+7yNcBPrtIZa4GDrrZHcFbDLnCNN9MYcwCY54qrJM6D1vIouDwHANbaHcaYR4H3gXXGmM+BnTirkK8HOgE1cMpInI23cFb+vm2MaQXMBo4AlXHKfBzBKe8A8B+gqyv2GsaY73EeblfD1beOq99vOKuan3cl5Y/xvwcp7jTGfIeTXN2PU+P4EZzyHD+cJtbxrvOdaYwZC4QBXQC/BLy19hdjzJs4K4WXGGO+BvbhzOEdxpgbTlcf21qbY4zpC3wNLDbGfOg6lx7ATcBL1lr3iub7gEdd74UtQA7Q3HVdvrDWnnT9waEw53++3GuMuQeYipMYDwXica71ZuAlr75ROOcwAadO9X6gHM481Mf5I83Z3HMiIiIilyUlmkVERET+mB4GhuOscC2BswL2dye9XMndesDTQEecmrOZOEniiTgJT3Dq147GScp1wUl87gY+AIZ4lXcYAdzhiq8kzgMIlwF9rLWzzyCeD40x63FW0T6Ck+xLBTYAz3HqB/mdbuyTxpjbgEeBv/K/2rp7cRL3o736WmPMXcA/cBKriTjXeBtOMtbdb7sx5gGcGsWjcB7a97Frv2E4ydfWQDGchwxOAF631u47TayfG2OK46y+fhMnOfsd8AIB6ldba580xix3HfdfOKuxdwGTcVa/n8n1mWCMaY3znvoXzu8S64H7rLWfenWdiZP474CzCjgHJ7H+BM57Epyk/Vmf/3n0G0696B44f8AAJ/bXcd7Hh7z6HsX5o8ctwK1ANM55rQH6AB+6/neBiIiIyJ+COT8PdD43jDFDgBtxVoaUBo7j/BIzERjuqunm7huH84/Agnxlre1x3oIVERERERERERER+ZO61BPNJ3FWuKzDWZlRDGiEk3zeCzSy1u5y9Y3DSTSvxElE57fGWjs+QLuIiIiIiIiIiIiIFMKlXjoj0vsp2m7GmFeBZ4FncP5bmrcV1tqBFyA2EREREREREREREcHrKdWXokBJZhd3fcGrLlQsIiIiIiIiIiIiIhLYpb6iuSAdXK+rAmwrb4z5O1AK5+EyC621gfqJiIiIiIiIiIiIyDlwSddodjPGPAkUx3mq+I1AU5wkcytr7QFXnzgKfhjgbKC3tXbn6Y41atQoC/DII4+YwsYtIiIiIiIiIiIi8mdwSZfO8PIkkAj0x0kyTwXauJPMLpnAy0B9IMb11QyYBcQDPxtjihV0AGPMQ8aYJe6fDx06ZJcvX26zs7PtkiVLLGCXLFliT5w4YVeuXGkPHjxot2zZYnfv3m2Tk5Ptxo0b7eHDh+3y5cttTk6O3z6rV6+26enpduvWrXbXrl123759dsOGDfbIkSN26dKlNjc312ef48eP2zVr1ti0tDS7bds2u3PnTpuSkmLXr19vjx49apcsWWLz8vJ89snMzLRr1661Bw4csNu3b7c7duyw+/fvt+vWrbPHjh3z9F28eLFnn2PHjtl169bZ/fv32x07dtjt27fbAwcO2LVr19rMzEyf8d3HO3r0qF2/fr1NSUmxO3futNu2bbNpaWl2zZo19vjx4z775Obm2qVLl9ojR47YDRs22H379tldu3bZrVu32vT0dLt69Wp74sQJn31ycnLs8uXL7eHDh+3GjRttcnKy3b17t92yZYs9ePCgXblypd8+2dnZdvny5fbQoUN206ZNdu/evXbPnj128+bNNiMjw65YscJmZ2f7nPvJkyc9c7l582a7Z88eu3fvXrtp0ybrPf/ufRYvXmxPnDhhV61a5ZnL/PO/bNkyv7nMysqyq1ev9pnL/POffy7d85+amuqZy5SUFLtu3TrP/FtrC5z/pKQkm5SUZA8cOOA3/96vR48e9Zv/1NRUu2bNGpuZmelzvfLy8uzSpUtPOf9ZWVk+77Pc3Fy7bNkye/jw4YDzv2rVqgLn/9ChQ3bjxo127969fvN/8uRJv/lfsWKFzcjIsJs2bbJ79uyxe/bssZs2bfKZf+99vOfffS/v3bvXbty40TP/ge5l7/n3vpe959/7umVlZZ3yXs4//4sXL7aZmZl+8+++l93zn38u3fdy/vkPdC+73zvu+U9JSfGb//z3snv+ve9l7/lfvXq1z/y773/3/Hvfy/nn3/seO9W9HGj+3fu45/9U97I+y/VZrs9yfZbrs1yf5fos12e5Psv1Wa7Pcn2W67Ncn+V/xs9yLpDLYkWzmzGmLNAYGAyUAG631i47zT7BwHygIdDfWvvuqfp7rWg+JzGLiIiIiIiIiIiIXEQXpHLD5bKiGQBrbYq1dgLQBqcG85gz2CcH+Mj141/OY3giIiIiIiIiIiIif0qXVaLZzVq7A1gH1DbGlD6DXdwlNgosnSEiIiIiIiIiIiIiZ+eyTDS7lHe95p5B30au123nKRYRERERERERERGRP61LNtFsjKlljCkXoD3IGPMqEAv8Yq096GpvaIwJDdC/BfBP14+fn8+YRURERERERERERP6Mgi92AKdwG/CGMWYusBVIA8oCzYBqwD7gQa/+Q3BKacwGdrvargdauL5/wVr7ywWIW0RERERERERERORP5VJONP8E/B/QBKgDRAPHgE3AWGCYtTbdq/9YoAvQAGgLhAApwH+B4dbaeRcudBEREREREREREZE/j0s20WytXQP843f0/xj4+PxFJCIiIiIiIiIiIiKBXLI1mkVERERERERERETk8qBEs4iIiIiIiIiIiIgUihLNIiIiIiIiIiIiIlIoSjSLiIiIiIiIiIiISKEo0SwiIiIiIiIiIiIihaJEs4iIiIiIiIiIiIgUihLNIiIiIiIiIiIiIlIoSjSLiIiIiIiIiIiISKEo0SwiIiIiIiIiIiIihaJEs1xSBg4ciDGG2bNnX+xQRERERERERERE5Awp0SwiIiIiIiIiIiIihaJEs1xSHn30UdavX89NN910sUMRERERERERERGRMxR8sQMQ8Va6dGlKly59scMQERERERERERGR30ErmqXQkpKSMMaQkJDA1q1b6d69O6VKlaJEiRK0adOGNWvWAHDgwAEeeughrrjiCooWLUqDBg2YNWuWz1gF1WieN28eHTp0oGLFioSFhVGuXDkaNWrEoEGDfPqlpKTw5JNPUrNmTYoVK0Z0dDQ1a9YkISGBbdu2+cU+bdo02rVrR+nSpQkLC+PKK69kwIABZGRk+PWNi4sjLi6OzMxMBgwYQOXKlQkLC6N69eoMGTIEa63fPt999x0tW7bkiiuuICwsjPLly9OsWTNGjhzp1zc9PZ1nnnmGq6++mvDwcKKiomjZsiXTp08/7RyIiIiIiIiIiIhcTFrRLOdMUlISDRs25OqrryYhIYGkpCQmTJhAfHw8Cxcu5LbbbiMyMpK77rqL9PR0/vOf/9C2bVs2bdpE5cqVCxx36tSptG/fnsjISDp27EiFChVIT09n/fr1jBw5ksTERAAyMzNp0qQJW7dupXXr1nTo0AFrLTt27GDSpEl0796datWqecZ96aWXSExMpGTJktx+++3ExsayatUqhg4dyo8//sjChQuJjIz0iSU7O5s2bdqwd+9e2rZtS3BwMBMnTuTpp58mKyvLEwvA//3f//H3v/+dcuXK0aFDB0qXLs3+/ftZtWoVn376KX369PH03bFjB/Hx8SQlJXHLLbdw2223cezYMSZPnsxtt93GBx98wIMPPniupkpEREREREREROScUqJZzpk5c+bwyiuv8Nxzz3naXn75ZV588UUaNmzInXfeyciRIwkKchbSt27dmnvuuYe3336bt99+u8BxP/zwQ/Ly8pg9ezZ16tTx2Zaamur5/ueff2br1q3079/fb7yTJ09y4sQJz8+zZs0iMTGRm2++mR9//JHo6GjPts8++4x7772XxMREv3H27t1LnTp1mDFjBuHh4QAkJiZSo0YN3n77bZ599llCQkIA+OCDDwgNDWXlypXExsYWGDdA79692bFjB19++SU9evTwtGdkZBAfH0+/fv3o2LEjZcuWLfA6iYiIiIiIiIiIXCxKNJ8nPRdlX+wQztgXjULOyThxcXE8/fTTPm29e/fmxRdf5MSJE7zxxhueJDNAz549ue+++1ixYsUZje9O7HoLVM85UL/Q0FBCQ0M9Pw8bNgxwktjeSWaAhIQE3n33XcaNGxcwAT5s2DCfY8TGxtKpUyfGjBnDxo0bufbaaz3bgoODPYnnguJeuXIlc+bMoXv37j5JZoDo6GgGDRpE586d+eabb3xWQYuIiIiIiIiIiFwqlGiWc6Zu3boUKVLEp618+fIA1KhRgxIlSvhsK1KkCGXLlmX37t2nHLdXr158++23NGzYkLvuuovmzZvTpEkTKlas6NOvWbNmVKhQgcGDB7Ns2TLatWtHkyZNAsa1cOFCQkJC+Prrr/n666/9jnny5EkOHDhAWloapUqV8rRHRUVRvXp1v/6VKlUC4ODBgz5xP/HEE9SuXZu77rqLZs2a0aRJE8qUKeMXC8ChQ4cYOHCg39gHDhwAYP369ae6TCIiIiIiIiIiIheNEs1yzkRFRfm1BQcHF7jNvT07+9Srv7t27crkyZN58803+eSTT/jggw8AqF+/Pq+//jqtW7cGIDIykkWLFpGYmMh3333HtGnTAGf1cJ8+fXj++ec9q4vT0tLIycnxe5hgfkePHvVJNOdf/Zz/PHNzcz1tjz/+OKVLl2bkyJEMGzaMd955B2MMzZo144033uDGG2/0xAIwY8YMZsyYccpYRERERERERERELkVKNJ8n56ochTjat29P+/btOXbsGL/++iuTJ09m1KhR3H777SxfvpxrrrkGgIoVK/Lxxx9jrWXdunXMnDmTESNG8NJLL5GXl8fLL78MOInvvLw80tPTz2vc99xzD/fccw8ZGRn88ssvTJgwgU8++YRbb72V9evXExsb60nCv/vuu/Tr1++8xiMiIiIiIiIiInI+BJ2+i8ilo1ixYrRo0YK33nqLZ599lpMnTzJlyhS/fsYYateuTd++fT2rhCdOnOjZ3qhRIw4ePMjatWsvSNzR0dG0a9eODz/8kISEBNLT05k3b54nFsDzs4iIiIiIiIiIyOVGiWa55P38888cP37crz0lJQWAiIgIANasWUNSUtJp+wH885//BODBBx9k7969fvscO3aMRYsWFSruqVOnkpOT49e+f/9+n3huvPFGbrnlFr799ls++eSTgGOtXr3as5+IiIiIiIiIiMilRqUz5JL3xBNPkJSURHx8PHFxcYSGhrJ06VJmzpxJlSpV6NGjBwA//fQTjz/+OI0bN6ZWrVrExsaye/duJk2aRFBQEAMGDPCM2bJlSwYPHswzzzzDVVddRbt27ahatSpHjx5lx44dzJkzh6ZNmzJ16tSzjrtHjx4ULVqUpk2bEhcXh7WWefPmsXjxYurXr0+rVq08fb/44gtatGjB/fffz7Bhw2jYsCHR0dHs3r2bVatWsWbNGhYuXEhsbOzZX0gREREREREREZHzRIlmueQ9++yzTJgwgSVLlvDTTz8RFBRE5cqVefbZZ+nfvz8xMTEA3HrrrfTv35+5c+cyadIkDh8+zBVXXEHr1q09CWhv//rXv2jSpAnDhg1j/vz5TJo0iaioKCpUqMBDDz1Ez549CxX34MGDmTZtGsuWLePHH3+kaNGiVKlShSFDhvDII494HkwITm3ppUuX8t577/HNN98wbtw4cnNzKVeuHNdccw19+/bluuuuK1Q8IiIiIiIiIiIi54ux1l7sGC4po0aNsgCPPPLIxQ5FREREREREREREpLDMhTiIajSLiIiIiIiIiIiISKEo0SwiIiIiIiIiIiIihaJEs4iIiIiIiIiIiIgUihLNIiIiIiIiIiIiIlIoSjSLiIiIiIiIiIiISKEo0SwiIiIiIiIiIiIihaJEs4iIiIiIiIiIiIgUihLNIiIiIiIiIiIiIlIoSjSLiIiIiIiIiIiISKEo0SwiIiIiIiIiIiIihaJEs4iIiIiIiIiIiIgUihLNIiIiIiIiIiIiIlIoSjSLiIiIiIiIiIiISKEo0SwiIiIiIiIiIiIihaJEs4iIiIiIiIiIiIgUihLNIiIiIiIiIiIiIlIoSjSLiIiIiIiIiIiISKFc0olmY8wQY8zPxphdxpjjxph0Y8xyY0yiMaZUAfs0Nsb86OqbaYxZZYzpb4wpcqHj/7NISkrCGENCQgJJSUn06NGD0qVLU7RoUW688UYmT57st8+JEycYPHgw119/PREREURGRnLLLbfw3//+95yMf/LkSYYNG0a9evWIiYkhIiKCuLg4OnXqxE8//eTXf8OGDSQkJFCpUiXCwsIoW7YsPXv2ZOPGjX59ExISMMaQlJTEBx98wHXXXUfRokUpW7YsDz30EIcOHfLbZ9WqVdx9993ExcURFhZGmTJlqFevHv379yc7O9unb05ODiNHjqRRo0ZERkYSERHBDTfcwPDhw8nLyzvlXIiIiIiIiIiIiFwMxlp7sWMokDHmJLAMWAfsB4oBjYAbgb1AI2vtLq/+nYBvgCzgKyAd6ADUBMZba+843TFHjRplAR555JFzei5/ZElJSVStWpX4+HjWrl1LtWrVuPnmm0lPT+err74iOzubn376iebNmwNOErhNmzbMmTOHWrVqcfvtt5OZmcn48ePZv38/zzzzDK+99tpZjw/Qs2dPvvzyS6699lpatGhBeHg4e/fuZf78+XTt2pWhQ4d6+k6dOpWuXbuSnZ1Nhw4dqF69Ort37+bbb78lLCyMWbNmUa9ePU//hIQERo8ezR133MG0adPo0KEDZcuWZdasWSxfvpzmzZszc+ZMT/9Vq1bRsGFDjDF07NiRqlWrcvjwYbZs2cKsWbNIT0+nePHiAJ4Ypk2bRs2aNYmPj6do0aLMmjWLVatW8de//pWxY8eet7kUEREREREREZE/HHNBjmKtvWS/gKIFtL8KWGCkV1skTjL6BHCj9xjAL67+PU53zJEjR9qRI0daOXPbt2+3rutrBw4c6LNt6tSpFrBt27b1tL322muetuzsbE97SkqKrVKligXsggULznr8jIwMa4yx9evXtzk5OX7xpqamer5PT0+30dHRtlSpUnbt2rU+/dasWWOLFStmb7jhBp/23r17W8BWqlTJ7tixw9OenZ1tb7nlFgvYX3/91dP++OOPW8BOnDjRL5b09HSbm5vr+TkxMdEC9tFHH/WJPScnx953330FjiMiIiIiIiIiIlKAC5LLDb4mOCmFAAAgAElEQVQg2eyzZK3NKmDTf4Fngau82roDZYAx1tol3mMYY54HfgYeAf5znsL1YToPuxCHOSfsxH7nZJwqVarw/PPP+7TdeuutVK5cmd9++83T9sknn2CM4a233iI4+H9vwdjYWF544QUeeOABPvroIxo3bnxW4xtjsNYSFhZGUJB/dZhSpf5XdWXMmDFkZGQwfPhwrrnmGp9+tWvX5sEHH+Sdd95h3bp1fttffPFFKleu7Pk5ODiYe++9l3nz5vHbb79x0003+fQPDw/3iyUmJsbzfV5eHsOHD6dcuXK8/fbbFCnyv2ovRYoU4c033+TTTz9l3LhxdOrUyW8sERERERERERGRi+WSTjSfQgfX6yqvthau16kB+s8FMoHGxpgwa+2J8xncn1XdunV9kqNulSpVYuHChQAcOXKELVu2UKFCBWrVquXXt0ULZxqXL19+VuMDREZG0qFDB77//nvq1q1Lt27duOWWW2jYsCERERE++7r3W7lyJQMHDvQbe9OmTQCsX7/eL9F84403BowF4ODBg562u+66i3fffZfOnTvTvXt3WrVqRZMmTbjyyiv9jpWWlsZVV13FK6+84jc2OMnq9evXB9wmIiIiIiIiIiJysVwWiWZjzJNAcSAKpz5zU5wk82CvbjVdr5vy72+tzTHGbAdqA9UAZerOg+jo6IDtwcHBnofYuR+Ud8UVVwTs627PyMg4q/HdvvrqK4YMGcIXX3xBYmIiAEWLFqV79+4MHTqUsmXLApCWlgbAhx9+eMpzO3r06BnF416hnZub62m76aabmDdvHq+++irjx4/31FiuWbMmiYmJ3H333T6xbN68mUGDBv2uWERERERERERERC6myyLRDDwJlPX6eSqQYK094NUW5Xo9VMAY7vbA2cpz7FyVo/ijiYpypmnfvn0BtycnJ/v0O1vh4eEMHDiQgQMHsmvXLubOnctnn33G559/TlJSEvPmzfM5zsqVK7n++usLdcxTufnmm5k8eTInTpxg6dKlTJ06lffee4+ePXtSpkwZWrVq5YmlS5cufPvtt+ctFhERERG5+N7YkHPK7QNqXS6/qomIiIg4igQqF3CpGThw4NCBAwcOGjRo0Ps4D/a7A+g3aNCgWQMHDkwGGDRoUF+gFPDewIED0/OPMWjQoAeASsBHAwcO3J1/uzHmoUGDBn3Qvn378gA1atRg/fr1lC5dmuXLl1O+fHmWLFlCqVKlWLt2LWFhYezZs4fMzEyOHTvG3r17CQ0NZd26dZQpU4Zly5b57LN+/XpCQkLYu3cvx44d4/jx4+zevZuwsDDWrFlD2bJlffaJiYlh48aNFClShH379nHkyBGysrLYtWsX4eHhrFq1inLlyrF06VLPPtHR0WzatAljDPv37+fw4cOcPHmSHTt2EBERwapVqyhfvjyLFy+mQoUKLFmyhKioKLZs2YK1ltTUVA4ePEhOTg5JSUkUL16clStXesZ3Hy8yMpKtW7eSl5dHeno6O3bs4LPPPqNKlSp07tzZZ5+yZcsyYsQIkpOT6dGjB0WKFGH8+PHs3r2bdu3akZ6eTnR0NCtWrKB8+fK8/fbbzJ07l4YNG9K+fXuSkpJIT0/no48+olq1arRp04ZNmzYRExPj2ee9994jOTmZTp06ERERwc6dOzlx4gSHDx8mJSWFmJgYQkNDeeyxxxg9ejRr1qyhadOmVKxYkcWLF7No0SJiY2OpUaMGR48eJTk5meDgYM/8L1u2jAoVKrB48WJKly7N2LFj2bhxI126dCE4ONhn/ufMmcPkyZOpUaMGHTp0YMmSJZQsWZL169cTHBzMgQMHiI6OplGjRpQoUYKffvqJEydO0K1bN7Zv384333zD7t27efjhh9m8eTNBQUGkpKRw+PBhTpw4wc6dOz3zf8UVVxQ4/wcOHCAjI4Ps7Gy/+V+yZInnNTIy0m/+c3Nz2bZtG8WLF2fFihWe90q5cuVYtmyZ3/ynpaWRl5fHtm3biIyM9MzL4sWLKVeuHCtWrKBYsWJs376dnJwcDh48SGpqKuCs3vae/yVLlhAbG8vKlSuJiIhgx44dnDx5kkOHDrF//36MMWzatImSJUv63JdlypRh9erVFC1alJ07d5KVlcWRI0fYt28fRYoUYcOGDZQqVcpnn9KlS7NmzRqfe/no0aPs3buXkJAQ1q9fH/BeXrduXcB7OTQ0lLVr1xIbG8vSpUs9161kyZJs2LChwHt59erVPvfy4sWLPfe/9/y772X3/OefS/e9DPjMf6B72f3ecc9/Xl4eaWlpPvNfokQJv/t/2bJllChRwjOX3vO/detWoqKifOaybNmynvlPSkoiOzubQ4cOceDAAZ/5X758uecei42NZdWqVQHv5aCgIL/5d+/jnv9du3aRlZUV8F7WZ/mpP8u97+X881+2bFmWL1/uM//e9/KWLVsKvJfzz7/3vez9We6+lwuaf+972f25nP9e3r17d4Hzn/+z3Ptezj//7nvZey69P8vdc5l//vPPZaB7WZ/l+izXZ7k+y933clJoOUK2LyUvpjy5W5ZyJCSSoJSthBgIOpJGudwMfZbrs1yf5fos12f5Jf5Zrn+X67P8cvksv+KKKwr+r/PnkLHWXojjnFPGmCo4JTI2W2uvdbUtximrcaO1dmmAfdbglM64xlpbYOmMUaNGWYBHHnnkfIT+h5SUlETVqlXp3bs3n332md/2+Ph45syZg/u99tprr/Hcc8/RqVMnvvnmG0/d5dTUVBo0aOBZcdy0adOzGv/AgQNs27aNhg0b+vQ7cuQINWrUIDU1lfT0dEqUKEFaWhrVq1cnJCSEyZMn+z3ALy8vj7lz5xIfH+9pS0hIYPTo0Wzfvp24uDif/rNnz6Z58+YkJiZ6aj7PmzeP66+/3m+V9tChQxkwYABPPfUUQ4YMAZwHDL788ss8/PDDvPXWW34PEExOTubgwYN+9aJFRERE5PLlvbpZK5lFRETkPDAX4iCX5b9irLU7jDHrgLrGmNLW2lRgI06iuQbgk2g2xgQDVYEcYNuFjld8Pfnkk0yZMoVJkyZRp04d2rVrR2ZmJl9//TX79+/nqaee8iSZz8aePXto1KgRV199NfXq1aNSpUocPnyYyZMns2/fPvr160eJEiUAKFWqFOPHj6dLly40atSIli1bUrt2bYKCgti5cycLFy4kLS2NrKyss47nzTffZPr06cTHx1OtWjWKFy/O2rVrmTJlCjExMTz00EOevi+88AIrV67k/fff5/vvv6dFixZUqFCB/fv3s3nzZhYsWMCrr76qRLOIiIiIiIiIiFxSLstEs0t516v7qWszgV7AbcCX+fr+BYgA5lprT1yY8KQgoaGhzJgxg7feeosvvviC9957j+DgYOrUqcM777zjeTje2YqLi2PQoEHMnj2bWbNmkZqaSsmSJalZsyaDBw+mR48ePv1btmzJqlWrGDp0KNOmTWPevHmEhoZSvnx5WrRoQbdu3QoVT58+fYiJieHXX39lwYIF5OTkULFiRfr06cMTTzxBlSpVPH1DQkKYOHEin3/+OZ999hmTJ0/m6NGjlClThqpVq/Lyyy/Tq1evQsUjIiIiIiIiIiJyrl2ypTOMMbWADGvtvnztQcDLwLPAL9baJq72SGArEAk0sdYucbUXxUlC3wzcba39z6mOq9IZIiIiIiJyIal0hoiIiJxnf/rSGbcBbxhj5uIkkNOAskAzoBqwD3jQ3dlae9gY8yAwHphtjPkPkA50BGq62r+6oGcgIiIiIiIiIiIi8idwKSeafwL+D2gC1AGigWM4DwEcCwyz1qZ772CtnWiMaQY8B3QDigJbgMdd/S/N5dsiIiIiIiIiIiIil7FLNtFsrV0D/OMs9lsAtDv3EYmIiIiIiIiIiIhIIEEXOwARERERERERERERubwp0SwiIiIiIiIiIiIihaJEs4iIiIiIiIiIiIgUihLNIiIiIiIiIiIiIlIoSjSLiIiIiIiIiIiISKEo0SwiIiIiIiIiIiIihaJEs4iIiIiIiIiIiIgUihLNIiIiIiIiIiIiIlIoSjSLiIiIiIiIiIiISKEo0SwiIiIiIiIiIiIihaJEs8glJCEhAWMMSUlJFzuUS97s2bMxxjBw4MCLHcoZi4+PxxhzscMQERERERERETnnlGiWQklKSsIYQ0JCwsUOxU92djYTJkzg/vvv59prryUyMpKIiAiuu+46XnzxRY4cOVLgvrt37+a+++6jfPnyhIWFERcXR//+/Tl48OAFPAORc+NSvk9FRERERERE5I8h+GIHIHK+bN26la5du1KsWDGaN29O+/btOXr0KNOmTePll1/mq6++YsGCBZQuXdpvv8aNG7N//346depErVq1+O2333j33XeZOnUqCxYsoFSpUhfprORyNmbMGDIzMy92GCIiIiIiIiIi55wSzfKHVaJECUaMGEHv3r0pVqyYp/3kyZN07dqVH374gUGDBvHee+/57NenTx/279/PsGHD6Nu3r6f98ccf5+233+a5557j/fffv2DnIX8clStXvtghiIiIiIiIiIicFyqdIWdt4MCBVK1aFYDRo0djjPF8ffbZZ55+eXl5vP/++zRo0IDixYtTrFgxGjRowKhRo8jLy/Mb1xhDfHw8e/fu5W9/+xuxsbGEh4dTv359vvjiizOOr0KFCvTp08cnyQwQGhrKs88+Czh1fr1t27aN6dOnExcXxz/+8Q+fbYMGDaJYsWKMHTuWY8eOnXEcS5cupVu3bsTGxhIWFkaVKlXo06cPycnJBe6Tl5fHW2+9Ra1atShatCgVK1bkn//8J4cPH/bru2rVKu6++27i4uIICwujTJky1KtXj/79+5Odne3TNycnh5EjR9KoUSNPKZEbbriB4cOH+82Fd7mFTZs2cddddxEbG0tQUBCzZ8+mVq1ahIaGkpqaGvAcBg8ejDGGESNG+LTv3r2bRx99lGrVqhEWFkapUqXo2LEjixcvDjhOSkoK999/P2XLliU8PJy6desyevToAq9dIK+//jrGGIYNGxZw+969eylSpAgNGjTwaXvppZdo0qQJ5cqVIzQ0lPLly9OzZ0/Wr1/vN8bprhcErtF88uRJhg8fTrt27ahSpQphYWGULFmSVq1aMWXKlIDxxsXFERcXR2ZmJgMGDKBy5cqEhYVRvXp1hgwZgrXW0/dM71MRERERERERkcLQimY5a/Hx8WRkZPDuu+9Sp04dOnfu7NlWt25dz/d/+9vf+OKLL6hUqRIPPPAAxhgmTJhAnz59mD9/PuPGjfMb++DBgzRu3Jjo6GjuvfdeMjIy+O9//0uvXr3Ys2cPAwYMKFTsISEhAAQH+94CM2fOBKBNmzYEBfn+HaZEiRI0adKE6dOns2jRIlq2bHna40yePJlu3bphraV79+5UqVKFpUuXMmrUKCZNmsSCBQuIi4vz2++f//wnc+fO5c4776RTp05MmzaNd955h3nz5jF//nyKFi0KOEnmhg0bYoyhY8eOVK1alcOHD7NlyxZGjhzJK6+84jnX7OxsOnTowLRp06hZsyY9e/akaNGizJo1i759+/Lrr78yduxYv1i2bt1Kw4YNqVGjBr169eL48eNERkbSu3dvnn32Wb788kufld9uY8aMITQ0lB49enjali1bRps2bUhPT+fWW2+la9eupKamMnHiRJo2bcqECRNo166dp39aWhqNGzdm27ZtNG3alKZNm5KcnMzDDz9MmzZtTnv93e655x6ef/55Ro8eTb9+/fy2f/755+Tl5dG7d29P29y5cxk8eDDNmzenW7duFC9enM2bNzN+/Hi+++47FixYQJ06dc74ehUkPT2dxx57jMaNG9O6dWvKlClDcnIy33//Pe3atePDDz/kgQce8NsvOzubNm3asHfvXtq2bUtwcDATJ07k6aefJisri8TERODM71MRERERERERkUKx1urL62vkyJF25MiRVs7M9u3bLWB79+4dcPsXX3xhAXvDDTfYI0eOeNqPHj1q69evbwE7btw4n30AC9g77rjD5ubmetq3bdtmY2JibEhIiN26dWuh4n744YctYJ9++mmf9ieffNICdujQoQH3+8c//mGBM3qPHDlyxJYqVcoGBQXZuXPn+mwbPHiwBWzr1q192nv37m0BW6pUKZuUlORpz83NtV27drWAfemllzztjz/+uAXsxIkT/Y6fnp7uc/0SExMtYB999FGbk5Pjac/JybH33Xef3zjuuQXsM8884zf+7t27bVBQkK1fv77ftt9++80CtmvXrp627Oxse+WVV9qwsDA7e/Zsn/579uyx5cuXt+XKlbNZWVme9gcffNACtn///j79Fy9ebIODgy1gExMT/Y4fSJs2bSxgV69e7bftmmuusSEhITY1NdXTlpKSYg8fPuzXd8WKFbZYsWL2tttu82k/3fWy1tpmzZpZ52P3f7KysuyuXbv8+mZkZNjatWvbmJgYm5mZ6bOtSpUqFrBt27b12ZaSkmKjoqJsVFSUPXnypF9sBd2nIiIicnH9e32250tERETkPLggeVWtaD5f6txxsSM4cyu/Pm9Df/LJJ4BTRqF48eKe9mLFijFkyBBatWrFRx99RM+ePX32K1KkCEOGDPFZVVy1alX69evHoEGDGDt2rGfF5u/13Xff8cEHH1CxYkWeeuopn22HDh0CICoqKuC+7vaMjIzTHmfSpEmkpaVx9913c8stt/hse+KJJ3j//feZMWMGO3fu9Kvd+9hjj1GlShXPz0FBQbzxxhtMnDiRTz75hBdeeMGnf3h4uN/xY2JiPN/n5eUxfPhwypUrx9tvv02RIkU824oUKcKbb77Jp59+yrhx4+jUqZPPOGXLlg14rStUqEDLli2ZMWMGa9eupXbt2p5t7tIW3iuEf/jhB7Zu3cqTTz5Js2bNfMYqX748Tz31FP379+fnn3+mXbt2ZGdnM27cOEqUKMHAgQN9+t9444306tXrd5XQ6N27N9OnT2f06NG88cYbnvYlS5awbt06unTp4vOQx9jY2IDj1KlThxYtWjB9+nSys7M9K8bdCrpeBQkLC6NixYp+7VFRUdx333088cQTLF68mL/85S9+fYYNG+Yz97GxsXTq1IkxY8awceNGrr322jOOQ0RERERERESkMJRolvNq2bJlBAUFER8f77etWbNmFClShOXLl/ttq1y5sqeurLf4+HgGDRoUcJ8z8csvv9CzZ0+KFSvGN99845OMPRPWVfs2f53dQJYtWwZAixYt/LYFBwfzl7/8haSkJJYvX+6XaM6fiAWoVq0alSpVIikpiYyMDKKjo7nrrrt499136dy5M927d6dVq1Y0adKEK6+80mffTZs2kZaWxlVXXcUrr7wSMN7w8PCAtYfr1KlDWFhYwH0SEhKYMWMGo0eP5t///jfg1Bz+z3/+Q5kyZXzKYCxcuBCAHTt2+CWOATZv3gzA+vXradeuHRs2bCAzM5NbbrklYOI/Pj7+dyWau3TpQlRUFJ9//jmDBw/2JNvdYyQkJPjt88MPP/D++++zZMkSUlNTycnJ8dmemprKFVdc4dN2qutVkLVr1/LGG28wd+5ckpOTycrK8tm+Z88ev32ioqKoXr26X3ulSpUAp/yMiIiIiIiIiMiFokSznFeHDh2iZMmShIaG+m0LDg6mdOnS7N+/329b2bJlA45Xrlw5z7i/18KFC2nbti1BQUFMmTKFm266ya+PO6FZ0Pjuh/EVtOLZm3uM/IlIN3d7oNXRpzr/HTt2cOjQIaKjo7npppuYN28er776KuPHj/fUWK5ZsyaJiYncfffdgFPrGJxk7qBBgwqM+ejRowGPWZAuXboQGRnJ559/zuuvv06RIkWYPHkyaWlp9O/f36cGtjuGr78+9Qp6dwzu63e698KZCg8P58477+TDDz9k+vTptG3bluzsbL788kvKlClD27ZtffoPGzaMxx57jJiYGFq3bk3lypWJiIjAGMPEiRNZuXIlJ06cKHRcixYtokWLFuTk5NCyZUs6duxIZGQkQUFBrFixgkmTJgU8TnR0dMDx3Nc8Nzf3d8UhIiIiIiIiIlIYSjSfL+exHMXlJCoqivT09IAlBnJyckhNTQ34oLSUlJSA4+3bt88z7u8xb9482rdvT1BQENOmTaNRo0YB+9WsWRNwVgAH4l51W6NGjdMe0x2jO+b8kpOTffp5S0lJ8cTiLdD533zzzUyePJkTJ06wdOlSpk6dynvvvUfPnj0pU6YMrVq18vTv0qUL33777Wlj93aq1dvu5O1HH33EjBkzuO222wKWzfCOedKkSXTs2PG0x3X3P9174ffo3bs3H374IaNHj6Zt27aepPhjjz3m8/7MyckhMTGRcuXKsWzZMr8/FrhXZwdyJqvdvb3yyiscP36cWbNm+a38f/3115k0adLvGk9ERERERERE5GIIOn0XkYK5yw8UtHryhhtuIC8vj7lz5/ptmzt3Lrm5udSrV89v286dO0lKSvJrnz17tmfcMzVz5kzatm1LcHAwM2bMKDDJDNC8eXMApk+fTl5ens+2I0eOsGDBAsLDw085hps7RnfM3nJycpg/fz5AwPOfM2eOX9u2bdvYtWsXcXFxAVezhoWF0bhxY1566SWGDRsG4ElS1qpVi+joaBYtWkR2dvZpY/893CUnRo8eTWpqKlOmTOH666+nbt26Pv3c12zevHlnNG6tWrWIiIhgxYoVAVeYB7qup9OkSROuuuoqJk2axKFDhwpMiqemppKRkUHjxo39ksxHjx71lEU5F7Zs2ULJkiUDlpcJ9D44G6e7T0VERERERERECkuJZimUmJgYjDHs3Lkz4Pb77rsPgGeeeYbMzExPe2ZmJk8//TQA999/v99+ubm5/Otf//JJ9m7fvp1hw4YRHBzMX//61zOKb/r06dx+++0ULVqUn3/+mQYNGpyy/5VXXkmbNm1ISkpixIgRPtsSExM5duwY99xzD8WKFTvtsTt37kzJkiX58ssvWbRokc+2d955h23bttGqVSu/+swA7777Ljt27PD8nJeXx4ABA8jLy+Pee+/1tM+bNy9gEta9CjgiIgJwyin07duX5ORk+vXrx/Hjx/32SU5OZt26dac9r/y8k7ejRo0iOzs7YL3jTp06ceWVVzJixAh+/PHHgGMtXLjQ8z4JCQmhV69eHDlyxK+m85IlSxg3btzvjhWcpHJWVhYjR47kxx9/5Prrr/f7w0VsbCwREREsXbrUp5xIdnY2jz32GKmpqWd17EDi4uJIT09n1apVPu0ff/wx06ZNOyfHON19KiIiIiIiIiJSWCqdIYVSvHhxGjZsyLx58+jVqxc1atSgSJEidOzYkeuvv56ePXsyadIk/vvf/1K7dm06d+7sqXG7fft27rzzTnr16uU37vXXX8+vv/5K/fr1adOmDYcOHeKrr74iIyODf//7334Puwtk48aNdOrUiaysLNq1a8ekSZMCliHIn8QcOXIkjRs3pl+/fvz8889cffXV/Prrr8yaNYsaNWrw6quvnvG1+eSTT7jjjjto1qwZd9xxB5UrV2bp0qVMnz6dcuXK8cEHHwTct0mTJtStW5e77rqLqKgopk2bxsqVK6lfvz5PPfWUp9+bb77J9OnTiY+Pp1q1ahQvXpy1a9cyZcoUYmJieOihhzx9X3jhBVauXMn777/P999/T4sWLahQoQL79+9n8+bNLFiwgFdffZVrrrnmjM7P2z333MMLL7zAyy+/THBwMD179vTrExISwrfffsutt95K+/btady4MXXr1iUiIoJdu3axePFitm3bRnJysidB/tprr/Hzzz/zzjvvsGTJEpo2bUpycjJfffUV7dq147vvvjurWF988UUSExPJzs72W80MEBQURL9+/Rg8eDDXXXcdnTp14uTJk8yaNYv09HSaN2/OrFmzfvexA+nfvz/Tpk2jadOm3HnnnURFRbFkyRLmz59P9+7dGT9+fKGPcbr7VERERERERESk0Ky1+vL6GjlypB05cqSVM7d582Z7++2325IlS1pjjAXsp59+6tmem5trR4wYYevXr2/Dw8NteHi4rVevnh0+fLjNzc31Gw+wzZo1s3v27LG9evWyZcqUsWFhYfaGG26w48aNO+O4Zs2aZYHTfgWyc+dOm5CQYMuVK2dDQkJs5cqVbb9+/WxaWtrvvj6//fab7dy5sy1durQNCQmxlSpVsg8//LDds2ePX9/evXtbwG7dutUOHTrU1qxZ04aFhdny5cvbxx57zB46dMin/7Rp02xCQoK9+uqrbWRkpI2IiLA1atSwffv2tUlJSX7j5+Xl2TFjxtgWLVrYmJgYGxISYsuXL2+bNGliX331Vbtz505P3+3bt1vA9u7d+7TnuGPHDhsUFGQBe/vtt5+yb0pKiv3Xv/5la9eubcPDw22xYsVs9erVbbdu3ezYsWNtdna2T//k5GR777332tKlS9uiRYva/2fv3sPjvOpD33+XJFu2bEm2rIvtiBBMCU6TEtqyGy7dJCFAS3OAxIVuzoZdbgmblLOBchra05Yihe6mNA+FQHtSSLj0wibllCTAOaWbnpJwKaVItmVZ9ii+yLasyLZ8iyRbli1p1v5jRsrorpFGl5G+n+eZ5533smZ+1vK75p3fu2atG2+8MX75y18eqd+Pf/zj08Y31m233RaBWFRUFE+ePDnhMQMDA/FTn/pUvO666+KaNWtiTU1NfMc73hGPHj06Uk9HjhwZOX4mf6+bb755wv9z3/72t+NNN90U169fH8vLy+PrXve6+P3vfz9++ctfHnc+xRjj85///Pj85z9/wvf4+Mc/HoH45JNPjto+3XkqSZIWz58lBkYekiRJ82BB8qohxriQee0l76GHHooA99xzz2KHsmKFELj55ptnNQavJEmSlG8eaB0ceX7vdn90KkmSci4sxJs4RrMkSZIkSZIkaU5MNEuSJEmSJEmS5sREsyRJkrRfXu4AACAASURBVCRJkiRpThwATEuO44ZLkiRJkiRJ+cUezZIkSZIkSZKkOTHRLEmSJEmSJEmaExPNkiRJkiRJkqQ5MdEsSZIkSZIkSZoTE82SJEmSJEmSpDkx0SxJkiRJkiRJmhMTzZIkSZIkSZKkOTHRLEmSJEmSJEmaExPNkiRJkiRJkqQ5MdEsSZIkSZIkSZoTE82SJEmSJEmSpDkx0SxJkiRJkiRJmhMTzZIkSZIkSZKkOVmyieYQwqYQwl0hhMdDCIdCCJdCCN0hhB+FEN4bQigYc/w1IYQ4xePRxfq3SJIkSZIkSdJyVrTYAUzhrcBDwAngSaAdqAF2AI8AbwghvDXGGMeU2wM8McHrtcxjrJIkSZIkSZK0Yi3lRPMB4E3A/xdjTA5vDCH8PvBT4NdJJZ2/MaZcU4yxbqGClCRJkiRJkqSVbskOnRFj/F6M8duZSeb09pPAX6VXb1nwwCRJkiRJkiRJoyzlHs1TGUgvByfYtzWE8F+BTcBZ4N9ijM0LFpkkSZIkSZIkrTB5l2gOIRQBv5le/acJDnld+pFZ5ingnTHG9vmNTpIkSZIkSZJWnsK6urrFjiEr9fX1fwbcAfxjjPG+jO2rgX7gI8DvAZ8EvgNcA9wKvLG+vv5LdXV1A+NeFAghvK++vv7zt99++1aAa6+9lkQiQWVlJbt372br1q00NjayadMm9u3bR3FxMc888wx9fX1cvHiRzs5OVq9ezf79+6mqqmLXrl2jyiQSCVatWkVnZycXL17k0qVLdHR0UFxcTEtLCzU1NaPKbNy4kaeffprCwkJOnjxJb28v/f39HD9+nLVr19Lc3MzmzZvZuXPnSJkNGzZw4MABQgh0dXXR09PDlStXOHbsGCUlJTQ3N7N161YaGhq46qqraGxspLy8nEOHDhFj5MyZM5w/f57BwUGOHj3K+vXr2bNnz8jrD79fWVkZhw8fJplMcu7cOc6ePUsymaStrY3S0tJRZWpqati9ezelpaUcOXKEwcFBzp8/z5kzZwA4dOgQGzZsoKmpaaRMdXU1e/bsYd26dRw9epSBgQG6u7vp6uoihMCBAwfYuHHjqDJVVVU0NzdTUlJCe3s7ly9fpqenh1OnTlFYWEhrayubNm1i165dI//2yspKWlpaKC4upqOjg/7+fi5cuMCJEycoKioaqf/hMg0NDVRWVrJ///6Ruhxb//v27aO6unpUXVZUVJBIJCgqKhqpy7H1P7Yuh+u/oKCAU6dO0dPTw+XLl2lvbx+p/y1btkxa/6dPn+bZZ59lYGBgXP03NjaOLMvKysbV/9DQEG1tbaxfv56mpqaRv9fmzZvZtWvXlPVfVlY2Ui8NDQ1s3ryZpqYm1q1bN2H9Hzx4cNL6Lykp4dixY1y5cmVc/VdUVIw6L6uqqti7dy9r1qyhvb2d/v5+ent7OXny5Kj6zyyTWf/D5/KFCxfo7Oxk1apVJBKJCc/lzPrPPJcz63/nzp0jf7eKigpaW1snPZf37t07qv4bGhomrP/hc3m4/sfW5fC5DIyq/4nO5eH/O8P1n0wmOXv27Kj6H3suD9d/5rmcWf+HDx+mvLx8VF3W1NSM1H/muXz69OlR9b979+6Rc6y6unrSc7mgoGBc/Q+XGa7/48ePT3ou25bbltuW25bbltuWL9W2vKGhkVhSRlFXGy9ch225bbltuW25bXketuVel9uWL+W2fMuWLfWzSMNmLcQYF+J9ciKE8EHgQaAVeFWM8dwMyhQBPwJuAj4cY3xwquMfeuihCHDPPffMPWBJkiRJmsYDrc+NCHjv9rz70akkSVr6wkK8yZKdDHCsEMIHSCWZ9wO3ziTJDBBjHAQeSa++ep7CkyRJkqSsdfRFui5HTvanlh19+dMRSJIkKVNe3C4PIXwY+DTQAtwWY+zK8iVOp5frchqYJEmSJM1CS3eSxzqStPaOTix/tHmQ7aWBHbUF3FCeN/2CJEmSln6P5hDC75JKMjeR6smcbZIZ4OXpZVvOApMkSZKkWXiyK8n9iaFxSeZhrb2R+xNDPNWVXODIJEmSZm9JJ5pDCB8D/hTYSaon85kpjr0phLB6gu2vAX47vfp38xKoJEmSJM1AS3eSR9qGmG6AjAg83DZES7fJZkmSlB+W7NAZIYR3AvcBQ8APgQ+GMG7c6qMxxq+kn38SuD6E8BTQkd72EuA16ecfizH+eD5jliRJkqSpPNaRnDbJPCwCj3ckHUJDkiTlhSWbaAZekF4WAh+e5JjvA19JP/9b4E7gPwBvAFYBp4CvA38RY/zhvEUqSZIkSdPo6IuTDpcxmURvaoLA2pIFmSxekiRp1maVaA4hrAOuAiqBS0BXjPGZXAYWY6wD6rI4/ovAF3MZgyRJkiTlymyHwWjpTlJbUpjjaCRJknJrxonmEMILgfcArwN+njHjO4cQzgJPAt8AHosxDuYwTkmSJEnKa5eGFracJEnSQpo20RxCeBnwx8BrSSWXB4C9wEngHLAW2AS8GHgr8BbgdAjhQeDPY4yX5yd0SZIkScofa2fZKXm25SRJkhbSlInmEMJfA28HuoEvAI8CP40x9k9y/DXArwDvBP478P4Qwn+JMf4ghzFLkiRJUt5JTeqX/fAZTgYoSZLywXRXLK8HPgJsjTH+VozxB5MlmQFijEdjjJ+PMb4SeCmwC7g5d+FKkiRJUn6qLQlsL81uUr/rSoMTAUqSpLww3dAZ22KMl2bzwjHGZuDOEMLa2ZSXJEmSpOVmR20B9yeGiDM4NgB31tqbWZIk5Ycpr1pmm2TO9WtIkiRJ0nJwQ3kBd20rZLo+ygG4e1uhw2ZIkqS8Me1kgDMRQlgF3AD0xRifzsVrSpIkSdJydGt1AVXF8HhHkkTv+L7N15UG7qwtMMksSZLySlaJ5hDCbwBvAd4fYzyX3vZC4DvAC9Pr3wR+I8Y4mONYJUmSJGlZuKE8lUju6Is8eHCQZISCAB96UZFjMkuSpLyUbY/m95CaGPBcxrZPAT8DfA/YBLwZeDfwcE4ilCRJkqRlqrYkUF0cRq1LkiTlo2x/i/WzQMPwSgihDPg14OsxxtcCvwS0kko0S5IkSZIkSZJWgGwTzVXAiYz1V5DqFf0oQIxxAPhn0sNoSJIkSZIkSZKWv2wTzb1Aecb6zUAEfpSxrR8onWNckiRJkiRJkqQ8ke0YzQeBN4QQikklmN8KNMcYz2Qc83ygK0fxSZIkSZIkSZKWuGx7NH8B2EYq4ZxIP//SmGNuAvbNPTRJkiRJkiRJUj7IKtEcY/xr4E+BElJDaPxF+gFACOE1wDXAk7kLUZIkSZIkSZK0lGU7dAYxxt8Hfn+S3T8CNgIX5xKUJEmSJEmSJCl/ZJ1onkqM8QpwJZevKUmSJEmSJEla2qYcOiOEsHaub5CL15AkSZIkSZIkLV3TjdF8JITwoRBCcbYvHEK4MYTwTeB3ZheaJEmSJEmSJCkfTJdo/i7w58CJEMJDIYRbp+qhHELYFkK4J4Twb8Au4EacGFCSJEmSJEmSlrUpx2iOMf5mCOGzwJ8A70s/hkIICeAEcB5YA2wCXgxUAgE4BfwB8OkY4+X5C1+SJEmSJEmStNimnQwwxtgIvD6E8CLgvcBtwEuBnxtz6GngMeAbwDdijAM5jlWSJEmSJEmStARNm2geFmM8CPweQAihBLiKVE/mS0BXjPHEvEQoSZIkSZIkSVrSZpxozhRj7AMOph+SJEmSJEmSpBVsuskAJUmSJEmSJEmakolmSZIkSZIkSdKcmGiWJEmSJEmSJM2JiWZJkiRJkiRJ0pyYaJYkSZIkSZIkzYmJZkmSJEmSJEnSnGSVaA4hXB1CKJvmmNIQwtVzC0uSJEmSJEmSlC+y7dF8BPjQNMd8MH2cJEmSJEmSJGkFyDbRHNIPSZIkSZIkSZKA+RmjuQa4OA+vK0mSJEmSJElagoqmOyCE8JtjNr10gm0AhcDVwH8B9uYgNkmSJEmSJElSHpg20Qx8BYjp5xF4c/ox1vCQGn1A/ZwjkyRJkiRJkiTlhZkkmt+dXgbgS8ATwDcnOG4IOAv8W4zx2dyEJ0mSJEmSJEla6qZNNMcY/3r4eQjhncATMca/mdeoJEmSJEmSJEl5YyY9mkfEGG+dr0AkSZIkSZIkSfmpIJuDQwi/EEL4oxBCzST7N6f3vzQ34UmSJEmSJEmSlrqsEs3A7wB3AV2T7D8FvBf4yFyCkiRJkiRJkiTlj2wTza8Anowxxol2prd/D3jVXAOTJEmSJEmSJOWHbBPNm4GOaY7pBLbMLhxJkiRJkiRJUr7JNtHcB1RNc0wVcHl24TwnhLAphHBXCOHxEMKhEMKlEEJ3COFHIYT3hhAmjD2E8MoQwj+GEM6FEPpCCM0hhA+HEArnGpMkSZIkSZIkabxsE81NwJtDCOsn2hlCKAPenD5urt4KPAzcBPw78BngG8ANwCPA10MIYcz7vxn4AfBq4HHgL4HVwKeBR3MQkyRJkiRJkiRpjGwTzV8g1WP5n0MIL8ncEUK4EfguUJk+bq4OAG8CamOMb48x/l8xxvcA24HjwK8DOzLev4xUYnoIuCXG+N4Y473AS4F/A94SQnhbDuKSJEmSJEmSJGXIKtEcY/x74G9I9TLeHULoDCE0hBA6gV3ALwF/E2P82lwDizF+L8b47Rhjcsz2k8BfpVdvydj1FlJJ8EdjjI0Zx/cDf5hevWeucUmSJEmSJEmSRsu2RzMxxncB7wf2k5oc8BfTy33A+2KM785lgJMYSC8HM7a9Jr38pwmO/wGp8aVfGUIons/AJEmSJEmSJGmlKZpNoRjjF4AvhBBKgA3AszHGvpxGNokQQhHwm+nVzKTyi9PLA2PLxBgHQwhHgOuBbUBiXoOUJEmSJEmSpBWksK6ubtaF6+rqBurq6nrr6uoGpj86N+rr6/8MuAP4xxjjfRnb7wU2AQ/U1dWdm6Dc24HnAV+tq6vrGLs/hPC++vr6z99+++1bAa699loSiQSVlZXs3r2brVu30tjYyKZNm9i3bx/FxcU888wz9PX1cfHiRTo7O1m9ejX79++nqqqKXbt2jSqTSCRYtWoVnZ2dXLx4kUuXLtHR0UFxcTEtLS3U1NSMKrNx40aefvppCgsLOXnyJL29vfT393P8+HHWrl1Lc3MzmzdvZufOnSNlNmzYwIEDBwgh0NXVRU9PD1euXOHYsWOUlJTQ3NzM1q1baWho4KqrrqKxsZHy8nIOHTpEjJEzZ85w/vx5BgcHOXr0KOvXr2fPnj0jrz/8fmVlZRw+fJhkMsm5c+c4e/YsyWSStrY2SktLR5Wpqalh9+7dlJaWcuTIEQYHBzl//jxnzpwB4NChQ2zYsIGmpqaRMtXV1ezZs4d169Zx9OhRBgYG6O7upqurixACBw4cYOPGjaPKVFVV0dzcTElJCe3t7Vy+fJmenh5OnTpFYWEhra2tbNq0iV27do382ysrK2lpaaG4uJiOjg76+/u5cOECJ06coKioaKT+h8s0NDRQWVnJ/v37R+pybP3v27eP6urqUXVZUVFBIpGgqKhopC7H1v/Yuhyu/4KCAk6dOkVPTw+XL1+mvb19pP63bNkyaf2fPn2aZ599loGBgXH139jYOLIsKysbV/9DQ0O0tbWxfv16mpqaRv5emzdvZteuXVPWf1lZ2Ui9NDQ0sHnzZpqamli3bt2E9X/w4MFJ67+kpIRjx45x5cqVcfVfUVEx6rysqqpi7969rFmzhvb2dvr7++nt7eXkyZOj6j+zTGb9D5/LFy5coLOzk1WrVpFIJCY8lzPrP/Nczqz/nTt3jvzdKioqaG1tnfRc3rt376j6b2homLD+h8/l4fofW5fD5zIwqv4nOpeH/+8M138ymeTs2bOj6n/suTxc/5nncmb9Hz58mPLy8lF1WVNTM1L/mefy6dOnR9X/7t27R86x6urqSc/lgoKCcfU/XGa4/o8fPz7puWxbbltuW25bbltuW75U2/KGhkZiSRlFXW28cB225bbltuW25bblediWe11uW76U2/ItW7bUZ5eBnZ0QY8y+UAhVpCbjuw5YF2O8K2P7C4C9McZLuQw0/fofBB4EWoFXxRjPZew7ALwIeFGM8dAEZX8MvAJ4RYzxJ5O9x0MPPRQB7rnH4ZwlSZIkzb8HWp8bEfDe7bP60akkSdJUwkK8SdZXMSGE9wKfBdaQCjICd6V31wD/BrwP+GKOYhx+3w+QSjLvB27LTDKndaeX5ZO8RNmY4yRJkiRJkiRJOZDVZIAhhNcBXyA1DvKdwEOZ+2OMLaQmBbwjVwGm3/fDwF8ALcCtMcaTExz2dHp57QTli0j1tB4E2nIZmyRJkiRJkiStdFklmoHfBU4AN8cYvwV0TXBMM/Czcw1sWAjhd4FPA02kkswTvSfA99LLX51g36uBEuDHMcbLuYpNkiRJkiRJkpT90BkvAx6NMfZMcUwHsHn2IT0nhPAx4D5gJ/D6CYbLyPQPwCeBt4UQPhdjbEy/xhrgj9PHPDRZYUmSJElaKJnjMh++EMdtd6xmSZKUb7K9elkNXJzmmA3A0OzCeU4I4Z2kksxDwA+BD4YwbtzqozHGrwDEGHtCCHeTSjg/FUJ4FDgHvAl4cXr73881LkmSJEnKpYrVix2BJEnS3GWbaD4K/OI0x9zEc+Mlz8UL0stC4MOTHPN94CvDKzHGJ0IINwN/APw6qQkLDwEfAT4bY4wTvYgkSZIkLSR7LEuSpOUm2zGavwn8xxDCWyfaGUJ4N/AS4BtzDSzGWBdjDNM8bpmg3L/GGH8txrgxxrg2xvhzMcZPxxjn3MtakiRJkiRJkjRetrfR/wx4G/C1EMJbgHKAEML/AfxHYAdwEPhcLoOUJEmSJEmSJC1dWSWaY4zn00NT/A2Q2av5s+nlD4H/HGOcbhxnSZIkSZIkSdIykfXAYDHGduCWEMJLgFcAm4Bu4Ccxxp05jk+SJEmSJEmStMRNmWgOITwGPBpj/Hp6/dXA0Rhje4yxGWhegBglSZIkSZIkSUvYdJMB3gFsz1h/EnjXvEUjSZIkSZIkSco70yWanwXKMtbDPMYiSZIkSZIkScpD043RnAD+9xBCA3Aive2a9BAaU4ox/mCuwUmSJEmSJEmSlr7pEs31wBPA/8jY9s70YzqFsw1KkiRJkiRJkpQ/pkw0xxi/G0K4DngtcBVQB3w//ZAkSZIkSZIkadoezcQYjwFfBAgh1AFPxRjvm+e4JEmSJEmSJEl5YsrJAEMIfx5CeH3GpneTGkpDkiRJkiRJkiRgmkQz8GHg5RnrXwLumL9wJEmSJEmSJEn5ZrpE8wWgJGM9zGMskiRJkiRJkqQ8NN0YzYeAHSGEx4ET6W0bQghXT/fCMcb2uQYnSZIkSZIkSVr6pks0PwD8HfDjjG0fSj+mEmfw2pIkSZIkSZKkZWDKZHCM8WshhCPA7cBVwLuAZqBp/kOTJEmSJEmSJOWDaXsdxxh/AvwEIITwLuDxGON98xyXJEmSJEmSJClPZDu8xbuB3fMRiCRJkiRpdh5oHZxy/73bHdlQkiTNr6yuNmKMfz1fgUiSJEmS5u78lTjyfOPqsIiRSJKklSTr29ohhALgA8DbgeuAdTHGovS+nwfuBj4TYzyQy0AlSZIkSRPL7LGc2bvZnsySJGmhFGRzcAhhNfDPwGeAFwK9QOYt8iPAe0gloSVJkiRJkiRJK0BWiWbgXuBWoB6oAR7J3BljfBb4AfArOYlOkiRJkiRJkrTkZZtofjvwrzHG+2KMSSBOcMwR4Oo5RyZJkiRJkiRJygvZJppfAPxkmmPOARWzC0eSJEmSJEmSlG+yTTRfAjZMc8zVwLOzC0eSJEmSJEmSlG+yTTQ3Aa9PTwo4TgihnNT4zD+da2CSJEmSJEmSpPyQbaL5YeB5wFdDCGWZO0IIG4CvABuBv8pJdJIkSZIkSZKkJa8om4NjjF8LIbwWeDfwJuA8QAihEbgeKAb+Msb4j7kOVJIkSZIkSZK0NGXbo5kY43uB9wD7gSogAL8AHALeG2P8bzmNUJIkSZIkSZK0pGXVo3lYjPErwFdCCGtJDZXRHWO8mMvAJEmSJEmSJEn5YVaJ5mExxkvApRzFIkmSJEmTeqB1cMr9926f09cbSZIkzYFXYpIkSZLyzvkrceT5xtVhESORJEkSmGiWJEmSlCcyeyxn9m62J7MkSdLi84pMkiRJklYohyORJEm54lWDJEmSJMnhSCRJ0pyYaJYkSZKkFcrhSCRJUq549SBJkiRJy0RHX6TrciQZoSCk1mtL7J0sSZLmX04SzSGESuA/An3A/x9jHMrF60qSpPnn+JySlP9aupM81pGktTeO2v7R5kG2lwZ21BZwQ3nBIkUnSZJWgqyuNEII94QQ/j2EUJGx7ReBBPAPwD8CPw4hrMttmJIkaSGcvxJHHpKk/PBkV5L7E0PjkszDWnsj9yeGeKorucCRSZKklSTbLkr/CYgxxnMZ2x4ANgJfBmqA24H3A5/KSYSSJGleOT6nJOWvlu4kj7QNMd3twQg83DZEZTH2bJYkSfMi2yuMFwHNwyvpITNuBr4YY7wrxvhGoAH4z7kLUZIkSZI0kcc6ktMmmYdF4PEOezVLkqT5kW2ieRPQlbH+qvTy8YxtPwSeP5egJEmSJElT6+iLkw6XMZlEb6Sjz+GRJElS7mWbaD4HVGas3wwkgR9nbIvAmjnGJUmSJEmaQkv37Honz7acJEnSVLJNNCeAN4YQNoUQNpAas7khxtiTccw1wMlcBBdCeEsI4XMhhB+GEHpCCDGE8HeTHHtNev9kj0dzEZMkSZKkxdXRF+m6HDnZn1qu1B66l4YWtpwkSdJUsp3l50HgCaADGARKgN8d3hlCKAR+mdE9nOfiD4EbgQvp99w+gzJ70jGO1ZKjmCRJkiQtgpbuJI91JMcNF/HR5kG2lwZ21BasqInu1hYubDlJkqSpZJVojjF+K4TwfuB96U1fjTFm9jB+LalhM/5njuL7bVIJ5kOkhul4cgZlmmKMdTl6f0mSJElLwJNdSR5pG5p04rvW3sj9iSHu3ga3VK+MZHMqqZ79MBgrKRkvSZIWTrY9mokxfgH4wiT7/iewca5BZbzeSGI5hJCrl5UkSZKUR1q6p04yD4vAw21DVBavjGRqbUlge2nIakLA60oDtSV+t5IkSbm3HK++toYQ/msI4ffTy5csdkCSJEmSZu+xjuS0SeZhEXi8Y+VMdrejtoCZpo0DcGftcvwKKEmSloKsrjJCCD8fQvitEEJ5xrZ1IYS/DiE8G0LoDCF8KPdhZuV1wF8B/z293BNCeDKEcPXihiVJkiQpWx19MaseuwCJ3pUzQeAN5QXcta1w2mRzAO7eVjhpT28nWJQkSXNVWFdXN+OD6+vrPwO8I8b4JxnbPgXcDQwApcAb6uvrf1JXV3c4l4HW19dfA7wL2FtXV/fYBPtXA/3AR4DfAz4JfAe4BrgVeGN9ff2X6urqBiZ6/RDC++rr6z9/++23bwW49tprSSQSVFZWsnv3brZu3UpjYyObNm1i3759FBcX88wzz9DX18fFixfp7Oxk9erV7N+/n6qqKnbt2jWqTCKRYNWqVXR2dnLx4kUuXbpER0cHxcXFtLS0UFNTM6rMxo0befrppyksLOTkyZP09vbS39/P8ePHWbt2Lc3NzWzevJmdO3eOlNmwYQMHDhwghEBXVxc9PT1cuXKFY8eOUVJSQnNzM1u3bqWhoYGrrrqKxsZGysvLOXToEDFGzpw5w/nz5xkcHOTo0aOsX7+ePXv2jLz+8PuVlZVx+PBhkskk586d4+zZsySTSdra2igtLR1Vpqamht27d1NaWsqRI0cYHBzk/PnznDlzBoBDhw6xYcMGmpqaRspUV1ezZ88e1q1bx9GjRxkYGKC7u5uuri5CCBw4cICNGzeOKlNVVUVzczMlJSW0t7dz+fJlenp6OHXqFIWFhbS2trJp0yZ27do18m+vrKykpaWF4uJiOjo66O/v58KFC5w4cYKioqKR+h8u09DQQGVlJfv37x+py7H1v2/fPqqrq0fVZUVFBYlEgqKiopG6HFv/Y+tyuP4LCgo4deoUPT09XL58mfb29pH637Jly6T1f/r0aZ599lkGBgbG1X9jY+PIsqysbFz9Dw0N0dbWxvr162lqahr5e23evJldu3ZNWf9lZWUj9dLQ0MDmzZtpampi3bp1E9b/wYMHJ63/kpISjh07xpUrV8bVf0VFxajzsqqqir1797JmzRra29vp7++nt7eXkydPjqr/zDKZ9T98Ll+4cIHOzk5WrVpFIpGY8FzOrP/Mczmz/nfu3Dnyd6uoqKC1tXXSc3nv3r2j6r+hoWHC+h8+l4frf2xdDp/LwKj6n+hcHv6/M1z/yWSSs2fPjqr/sefycP1nnsuZ9X/48GHKy8tH1WVNTc1I/Weey6dPnx5V/7t37x45x6qrqyc9lwsKCsbV/3CZ4fo/fvz4pOeybfnM2vKWA4dJriljVXszL3/RVbbltuW25bbli96Wt/QWcP7o01ws2cTzTjbRXbqFq0/s4kLJJracSTBYuIoNvSdYPdjH6oE+yi+cYKhgFas6E2yvrV4RbfnVayPXxHMM9Jyjd2CIymeP0l9cSu2pZrpLt/AL53bzzpduoejYnnFt+e7203z12CBP7T1EV0E5G55ppmPNFp5uauTHg5X0tTVTU25bbltuW25b7nW5ORbb8nxuy7ds2VKfRRp21kKMM79THUI4BPx7jPHt6fVVwBkgAdwCVAC7gYYY4/+W00BDuIXUZIBfjTG+I4tyRcCPgJuAD8cYH5zq+IceeigC3HPPPbMPVpKkPPVA6+DI83u3Zz2VgyTl3OMdQ/w/sxgK4621BdxZWzgPES1tHX2RBw8OkoxQEOBDLyqadEzm6SZYhOd6Qq+UCRYlSVqmFmSChmyvFqqB4xnrLyPVi/nzMcb+GGMn8E1gyYyLHGMcBB5Jr756MWORJEmSn0xBDAAAIABJREFUlJ21s8wVz7ZcvqstCVQXBzavSS0nSzJnO8FiS/fKGfdakiTNTraJ5ghkdm/65fS272dsOw1UzTGuXDudXq5b1CgkSZIkZWWyMYXnq9xK4QSLkiQp17K9+moHXp6x/magI8bYlrFtK3B+roHl2HDMbVMeJUnSCuZEUJKWotqSwPbS7H7teV3p5D155QSLkiRpfmQ7+OLXgfoQwj+QmnjvFcBnxhxzA5DTiQBnIoRwE7A7xnhlzPbXAL+dXv27hY5LkqSlrqU7yWMdyXFJh482D7K9NLCjtsCegZIW1Y7aAu5PTD/MA6QGILyz1jZrKrMdBqOlO0ltyQodk0SSJE0r20Tzp4FfBXak15uA+4Z3hhB+FvhF4E9yEVwI4Q7gjvTq5vTyFSGEr6Sfn4kx/k76+SeB60MITwEd6W0vAV6Tfv6xGOOPcxGXJEnLxXQTQbX2Ru5PDHH3NpwIStKiuaG8gLu2MeOJ67w5NrVLQwtbTpIkrQxZJZpjjBeAV4UQbkhv2h9jzLwd3gfcCTTmKL6XAu8cs21b+gFwDBhONP9t+r3/A/AGYBVwilQv7L+IMf4wRzFJkrQsZDsRVGWxY55KWjy3VhdQVZwaKzgxwbAP15UG7vQXGDPiBItLwwOtg1Puv3d7tv3CJElaXLP65Ioxtkyy/ShwdA7xjH29OqBuhsd+Efhirt5bkpTf/PI2vdlMBGUCR9JiuqE8lUju6Is8eHCQZISCAB96UZFjMmch1ZZnP3yGnwHz5/yV5z6RN672/7IkKT/N6lt2CGEVcBtwHbA+xviJ9PY1QBmpIS2clliStCT45W28uUwEZTJH0mKrLQlUF4dR65q54QkWs/kccILF3Mu86Z15g9yb4ZKkfJX1J1gI4VdJ9RzeTGoYtAh8Ir37pcC/Au8AvpajGCVJyppf3qbmRFCStLI5waIkScq1rK4WQggvA54glVz+beB/ZO6PMf4EOEJqrGRJkrREORGUJK1sqQkWC5muj7ITLEqSpJnKtlvXx0hN+PeyGOPJEMLHJzimAfiFOUcmSZLmjRNBSZKcYFGSJOVStonmVwFPxBhPTnHMceD22YckSZLmmxNBSZLACRaXgo6+SNflOPK3dz4ESVK+yjbRvB44M80xJWQ5JIckSVpYTgQlScrkBIsLr6U7yWMdyXGfxR9tHmR7aWCHvcklSXkm20+tZ4DrpznmpUDb7MKRJEkLZUdtwbRjcw5zIihJknLnya4k9yeGJr3h29obuT8xxFNds5u8V5KkxZDtN8bvAL8SQvjliXaGEN4AvBL4f+camCRJml9OBCVJ0sJr6U7ySNsQ0/2mKAIPtw3R0m2yWZKUH7IdOuN+4G3Ad0MInwOuAQgh3A68GvgAcAL48xzGKEmS5okTQUmStLAe60hOm2QeFkl9Rvs5LEnKB1klmmOMz4QQXg98Hbg3Y9e3SHV4OgzsiDFON46zJElaIpwISpKkhdHRF7OaHwEg0RudIFCSlBey7dFMjHFXCOHFwO3AK4BNQDfwE+CbMcbB3IYoSdLsOZP7zDkRlCRJ82u2w2C0dCepLSnMcTSSJOVW1olmgBjjEKlezN/KbTiSJOWGM7lLkqSl5tLQwpaTJGkh+Q1bkrTsOJO7JElaitbOslPybMtJkrSQZtWjOYTwMuCXgI3ARB95Mcb4ibkEJknSbGQ7k3tlMfZsliRJCyJ1zZH9jW6vVSRJ+SCrRHMIoQx4DLiV1OR/k4mAiWZJ0oJzJndJkrRU1ZYEtpeGrCYEvK40OG+CJCkvZNuj+QHgNcAPgS8DxwEn/5MkLQnO5C5J0vx4oHXqr333bp/Vj2VXpB21BdyfmP7XV5Dq3XVnrTfEJUn5IdurgTcDu4BbY4wObClJWlKcyV2SpPl3/spzKdKNq71Rm60bygu4axvTDvUVgLu3FfrLK0lS3sg20VwO/K1JZknSUuRM7pK0vGX2qj18IY7bbq/a+ZP5t82sB//ms3NrdQFVxakhvBIT/BrrutLAnbUFJpklSXkl26uCg0DNfAQiSdJcOZO7JK0cFasXOwJpbm4oTyWSO/oiDx4cJBmhIMCHXlTkkF6SpLyUbaL5L4E/DSFcFWN8Zj4CkiRptpzJXZKWN3vPajmqLQlUF4dR65Ik5aNsr9S+Q2oywH8NIdQDO4FnJzowxtg+x9gkScqKM7lLkiRJkrQ4sk00HwUiqXkJHpniuDiL15Ykac6cyV2SpPnT0RfpuhxHhnno6IvesJUkSUD2yeC/gRl9d5ckaVE4k7skSbnX0p3ksY7kuF8NfbR5kO2lgR1OXCdJ0oqXVaI5xviueYpDkqSccSZ3SZJy58mu5JQ3cFt7I/cnhrh7G9xS7WerJEkrlcNbSJKWJWdylyStJA+0Do48P3whjts+24kUW7qnTjIPi8DDbUNUFjvJriRJK1VWVxshhCGgLsb4iSmO+QOgPsZoEluStGgyv3D3Djy3/WvtQ8Dsv3BLkuYus42eiG303FSszt1rPdaRnPHYiZHUr4lMNEuStDJlewUX0o+ZHCdJ0pIwkFzsCCRJkzl/5bk05sbVfo2YrflIznf0xXFjMk8n0RudIFCSpBVqProKbAT65+F1JUmascwv3O9vHJhwuyRpcWS2xZm9m22jl5aW7tndqW3pTlJbUpjjaJaf+RruRJKkxTLtJ1cI4dVjNl0zwTaAQuBq4O3A0zmITZIkSZK0SC4NLWy5lSyXw51IkrRYZnKL9CkYGZYrAu9MPyYSgCTwf845MkmSJEnSolk7y07Jsy230thjWZK03Mzkk+0+UgnmAPwRqcTz9yc4bgg4CzwZY2zNVYCSJEmSpIWXmtQv++EznAxQkqSVadpEc4yxbvh5COGdwBMxxs/OZ1CSJEmSpMVVWxLYXhqymhDwutLgRICSJK1QWd1qjjG+wCSzJEmSJK0MO2oLmGnaOAB31tqbWZKklcpBoSRJkiRphXqgdXDk+eELcdz2e7cXcdc2eKRtiKn6NQfg7m2FDpshSdIKNmWiOYTwPdITAMYYO9LrMxFjjLfNOTpJkiRJ0oKoWD3x9lurC6gqhsc7kiQmGEbjutLAnbUFJpklSVrhpuvRfEt6WTJmfTozH8RLkiRJkrQo7t0+sx+53lCeSiR39EU+1vJcL+hP3FDkmMySJAmYJtEcYyyYal2SpKWuoy9yOTl63S/EkrQ0dPRFui5HkhEKgm10PqgtCRQXjF6XJEmC6YfO+CDwkxjjTxcoHkmScqKlO8ljHUlax/zE96PNg2wvDezwJ76StGhso/OXN3AlSdJkprt6+wzwq8MrIYShEMLH5jckSZLm5smuJPcnhsYlMIa19kbuTwzxVFdywv2SpPljG52fWrqT3LdvkI82D3I5ycjjo82D3LdvkJZu60uSpJVuukRzP1CcsR7SD0mSlqSW7iSPtA1NO1lABB5uG/KLsSQtINvo/OTNAUmSNBPTJZqPAL8SQqjJ2OZEf5KkJeuxjuSMP6gi8HiHX4olaaHYRucfbw5IkqSZmi7R/HngF4DOEMJQeltdegiNqR6DU7ymJEnzoqMvTtrbajKJ3khHn/dQJWm+2UbnJ28OSJKkmZpyMsAY42dDCF3A7cBW4FagHTg634GFEN4C3Ay8FLgRKAW+GmN8xxRlXgn8IfByYA1wCPgS8LkY49Bk5SRJy8Nse1G1dCepLSnMcTT544HW5+4PH74Qx22/d/uUlwuSNCO20flnLjcHnCBQkqSVZ9pvjjHGR4FHAUIISeDLMcb75jswUgnjG4ELQAewfaqDQwhvBr5BalzpvwfOAW8EPg28CnjrfAYrSVp8l2Z5S3G25ZajitWLHYGk5co2Ov94c0CSJGUj2y5K9cBT8xDHRH6bVIL5EKmezU9OdmAIoQx4GBgCbokxNqa3fwz4HvCWEMLb0klzSdIytXaW32lnW265sMeypIVgG51/vDkgSZKyMd0YzaPEGOtjjD+Yr2DGvNeTMcaDMcaZ/FbrLUAV8Ohwkjn9Gv2kekYD3DMPYUqSlpAbyrP6WJtzOUnSzNlG5x9vDkiSpGwsly5Mr0kv/2mCfT8A+oBXhhCKY4yXFy4sSdJCqi0JbC8NWY0neV1pcBzJPPbGP/7WlPu//YdvWqBIJE3HNjr/pJL82Q+f4c0BSZJWpuVyBfDi9PLA2B0xxkHgCKmk+raFDEqStPB21BYw05REAO6sXS4fhXrm7MWRh6SlyTY6vwzfHMiGNwckSVq5lkuP5vL0snuS/cPbNyxALJKkRXRDeQF3bYOH26YeIDIAd28rtNdVnsvssZzZu9mezNLSNNxGP9I2xFT9mm2jl44dtQXcn5i6voZ5c0CSpJWtsK6ubrFjmFZ9ff01wLuAvXV1dY9NsP+/AZuAz9XV1Z2bYP9dwPOAR+rq6jomeo8Qwvvq6+s/f/vtt28FuPbaa0kkElRWVrJ79262bt1KY2MjmzZtYt++fRQXF/PMM8/Q19fHxYsX6ezsZPXq1ezfv5+qqip27do1qkwikWDVqlV0dnZy8eJFLl26REdHB8XFxbS0tFBTUzOqzMaNG3n66acpLCzk5MmT9Pb20t/fz/Hjx1m7di3Nzc1s3ryZnTt3jpTZsGEDBw4cIIRAV1cXPT09XLlyhWPHjlFSUkJzczNbt26loaGBq666isbGRsrLyzl06BAxRs6cOcP58+cZHBzk6NGjrF+/nj179oy8/vD7lZWVcfjwYZLJJOfOnePs2bMkk0na2tooLS0dVaampobdu3dTWlrKkSNHGBwc5Pz585w5cwaAQ4cOsWHDBpqamkbKVFdXs2fPHtatW8fRo0cZGBigu7ubrq4uQggcOHCAjRs3jipTVVVFc3MzJSUltLe3c/nyZXp6ejh16hSFhYW0trayadMmdu3aNfJvr6yspKWlheLiYjo6Oujv7+fChQucOHGCoqKikfofLtPQ0EBlZSX79+8fqcux9b9v3z6qq6tH1WVFRQWJRIKioqKRuhxb/2Prcrj+CwoKOHXqFD09PVy+fJn29vaR+t+yZcuk9X/69GmeffZZBgYGxtV/Y2PjyLKsrGxc/Q8NDdHW1sb69etpamoa+Xtt3ryZXbt2TVn/ZWVlI/XS0NDA5s2baWpqYt26dRPW/8GDByet/5KSEo4dO8aVK1fG1X9FRcWo87Kqqoq9e/eyZs0a2tvb6e/vp7e3l5MnT46q/8wymfU/fC5fuHCBzs5OVq1aRSKRmPBczqz/zHM5s/537tw58nerqKigtbV10nN57969o+q/oaFhwvofPpeH639sXQ6fy8Co+p/oXB7+vzNc/8lkkrNnz46q/7Hn8nD9Z57LmfV/+PBhysvLR9VlTU3NSP1nnsunT58eVf+7d+8eOceqq6snPZcLCgrG1f9wmeH6P378OP39/VTGi3R2nmAwFFJx5gAXSzbxvJNNdJdu4eoTu4jrN/G2glZ+sXqNbfkyasv/7qkEyWSSwcEhXl67yrbctty2fIm25VeeOcgrr6ng0qE9nFibapd7SyrZenofg0VruH6gkx01V/iZoovjrsu8Ll/4tvzE0838zKZ1nDh+jMKhAdZe7mZ93xligOpzh+lbs4HnndpDT+kW7rjcxC9vq/G63LbctnwFtOUzvS6f6Fy2LV/+1+W25UuvLd+yZUv9NOnXnAgzm2tvcYUQbgGeBL4aY3zHBPsbgJcBL4sx7pxgfwtwPfCzMcbEVO/10EMPRYB77nHuQElaDjr6Ig8eHCQZoSDAh15U5E96lyl7NEv5xzY6f7R0J3m8I0ligjG2rysN3FlbYA90SZKWrgW5wFouQ2c8TSrRfC0wKtEcQigCXgAMAm0LH5okaTHVlgSqi8OodUnS0mAbnT9uKE8lkr05IEmSJrNcbjl/L7381Qn2vRooAX4cY7y8cCFJkqSFsq/9LEdO9XCw81mOnOphX/vZxQ5Jkpal4ZsDm9ekliaZJUnSsOXSo/kfgE8CbwshfC7G2AgQQlgD/HH6mIcWKzhJkjQ//mXPce77+r/zg32do7bf8MGv8urrt/JHv3ETt934vEWKTpIkSZJWjiWbaA4h3AHckV7dnF6+IoTwlfTzMzHG3wGIMfaEEO4mlXB+KoTwKHAOeBPw4vT2v1+o2CVJ0vz74j/v433/9/dITjLfxA/2dfL6uid4+AOv4T2vvX6Bo5MkSZKklWXJJpqBlwLvHLNtW/oBcAz4neEdMcYnQgg3A38A/DqwBjgEfAT4bMyHWQ8lSdKM/Mue41MmmYclY+Tuv/wez68qs2ezJEmSJM2jJTtGc4yxLsYYpnhcM0GZf40x/lqMcWOMcW2M8edijJ+OMQ4twj9BkiTNk/u+/u/TJpmHJWPkE1//6TxHJEmSJEkr25JNNEuSJE1kX/vZcWMyT+f7+55xgkBJkiRJmkcmmiVJUl75l+bjC1pOkiRJkjS9pTxGsyRJ0jg9fVcWtJwkScvZA62DU+6/d7tpA0nSzPiJIUmS8kpZyeoFLSdJ0kpx/spz8x9sXB0WMRJJUj4y0SxJkvLKbS953oKWkyRpOcvssZzZu9mezJKkbDlGsyRJyivXX72JV1+/NasyN19/FddfvWmeIpIkKf919EW6LkdO9qeWHX1x+kKSJGXwFqUkSco7f/QbN/H6uidIxum/BBeEwMd+45cWICpJkvJPS3eSxzqStPaO/kz9aPMg20sDO2oLuKHcPmqSpOn5aSFJkvLObTc+jy/81msoCFOPH1kQAg9/4DXcdqPDZkiSNNaTXUnuTwyNSzIPa+2N3J8Y4qmu5AJHJknKRyaaJUlSXnrv667nu3V3cPP1V024/+brr+K7dXfwntdev8CRSZK09LV0J3mkbYjpfhsUgYfbhmjpNtksSZqaQ2dIkqS8dduNz+O2G5/Hvvaz/KcHvsPgUJKiwgL+/t43OCazJElTeKwjOW2SeVgEHu9IOoSGJGlKJpolSVLeu/7qTbygpmzUuiRJmlhHX5x0uIzJJHpTEwTWlkw9bJUkaeUy0SxJWpYeaB0ceX74Qhy3/d7tfgRKkjRTfq4uL7MdBqOlO0ltSWGOo5EkLRdeDUiSlr2K1YsdgSRJy4efq/nv0tDClpMkrQwmmiVJy5I9qyRJyh0/V5eXtbPslDzbcpKklcGR/CVJkiRJWkFmO6mfkwFKkqbip4QkSZIkSStIbUlge2l2k/pdVxqcCFCSNCV//yRJkiRpQTmxnLT4dtQWcH9iiDj9oQTgzlr7qUmSpuYnhSRJkqRFU7H6uYekhXNDeQF3bStkuj7KAbh7W6HDZkiSpmVXAUmSJEkLyh7L0tJwa3UBVcXweEeSRO/4vs3XlQburC0wySxJmhGv8CRJkiRJWqFuKE8lkjv6Ig8eHCQZoSDAh15U5JjMkqSsmGiWJEmSJGmFqy0JVBeHUeuSJGXD379IkiRJkiRJkubERLMkSZIkSZIkaU5MNEuSJEmSJEmS5sREsyRJkiRJkiRpTkw0S5IkSZIkSZLmpGixA5AkSZqtN/7xt0ae//TgqXHbv/2Hb1rwmCRJkiRpJTLRLEmSloWrKtYvdgiSJEmStGKZaJYkSXnLHsuSJM3NA62DI88PX4jjtt+73bSBJGlm/MSQJEmSJElUrF7sCCRJ+cxEsyRJkiRJK5Q9liVJuVKw2AFIkiRJkiRJkvKbiWZJkiRJkiRJ0pyYaJYkSZIkSZIkzYmJZkmSJEmSJEnSnJholiRJkiRJkiTNiYlmSZIkSZIkSdKcmGiWJEmSJEmSJM2JiWZJ+l/t3Xt4XXWZ6PHv26ZN2/Se3igFCwhFwSrKcJUKKAoC4w1mZIRBB0dBcJhRn4Ojg7YiHucoiloH5XEcR+CIyhzLOCDgoFAHUHql9H6h9/v93rRJfuePtRPSkLRJd5KV7Hw/z7Oelb1u+03Wu9/uvnvt35IkSZIkSVJRbDRLkiRJkiRJkopio1mSJEmSJEmSVBQbzZIkSZIkSZKkotholiRJkiRJkiQVpeQazRGxIiJSM9OGvOOTJEmSJEmSpFJTlncA7WQncG8Ty/d0dCCSJEmSJEmSVOpKtdG8I6U0Me8gJEmSJEmSJKk7KLmhMyRJkiRJkiRJHatUr2guj4jrgROBvcAcYGpKqSbfsCRJkiRJkiSp9JRqo3kU8ECjZcsj4mMppWfzCEiSJEmSJEmSSlXPiRMn5h1Dm5o0adIg4C7gDuDLwMNAT+CdwIcnTZr0XxMnTtzYeL+I+MSkSZN+eOWVV44GOO2001iwYAHDhg1j1qxZjB49munTp1NZWcm8efMoLy9n7dq17Nu3j71797Ju3Tp69+7N/PnzGT58ODNnzjxsnwULFtCrVy/WrVvH3r172b9/P2vWrKG8vJy5c+cycuTIw/YZMmQIixYtomfPnmzYsIHdu3dz4MABVq9eTd++fZkzZw6jRo1ixowZ9fsMHjyYxYsXExFs2rSJXbt2cfDgQVauXEm/fv2YM2cOo0ePZtq0aRx//PFMnz6dQYMGsXTpUlJKbNmyhe3bt1NdXc2KFSvo378/L730Uv3x655v4MCBLFu2jNraWrZt28bWrVupra3llVdeYcCAAYftM3LkSGbNmsWAAQNYvnw51dXVbN++nS1btgCwdOlSBg8ezOzZs+v3GTFiBC+99BIVFRWsWLGCQ4cOsXPnTjZt2kREsHjxYoYMGXLYPsOHD2fOnDn069ePVatWUVVVxa5du9i4cSM9e/Zk4cKFVFZWMnPmzPrffdiwYcydO5fy8nLWrFnDgQMH2LNnD+vXr6esrKz+/NftM23aNIYNG8b8+fPrz2Xj8z9v3jxGjBhx2LkcOnQoCxYsoKysrP5cNj7/jc9l3fnv0aMHGzduZNeuXVRVVbFq1ar683/cccc1e/43b97Mjh07OHTo0GvO//Tp0+vnAwcOfM35r6mp4ZVXXqF///7Mnj27/u81atQoZs6cecTzP3DgwPrzMm3aNEaNGsXs2bOpqKho8vwvWbKk2fPfr18/Vq5cycGDB19z/ocOHXrY63L48OG8/PLL9OnTh1WrVnHgwAF2797Nhg0bDjv/DfdpeP7rXst79uxh3bp19OrViwULFjT5Wm54/hu+lhue/xkzZtT/3YYOHcrChQubfS2//PLLh53/adOmNXn+617Ldee/8bmsey0Dh53/pl7LdblTd/5ra2vZunXrYee/8Wu57vw3fC03PP/Lli1j0KBBh53LkSNH1p//hq/lzZs3H3b+Z82aVf8aGzFiRLOv5R49erzm/NftU3f+V69e3exr2VpuLbeWW8ut5dZya7m13FpuLbeWW8ut5dby7ljLjzvuuElFN11bIFJKHfE8uYuIbwKfBaaklD7Q3Hb33XdfArjllls6KjRJkiRJkiRJai/REU/SnW4G+IPCfEKuUUiSJEmSJElSielOjeZNhXlFrlFIkiRJkiRJUonpTo3m8wvzV3KNQpIkSZIkSZJKTEk1miPijIgY2sTy1wGTCw8f7NioJEmSJEmSJKm0leUdQBu7Fvh8RPweWA7sBk4BrgT6AI8D38wvPEmSJEmSJEkqPaXWaP49MA44i2yojApgB/A/wAPAAymllF94kiRJkiRJklR6SqrRnFJ6Fng27zgkSZIkSZIkqTspqTGaJUmSJEmSJEkdz0azJEmSJEmSJKkoNpolSZIkSZIkSUWx0SxJkiRJkiRJKoqNZkmSJEmSJElSUWw0S5IkSZIkSZKKYqNZkiRJkiRJklQUG82SJEmSJEmSpKLYaJYkSZIkSZIkFcVGsyRJkiRJkiSpKDaaJUmSJEmSJElFsdEsSZIkSZIkSSqKjWZJkiRJkiRJUlFsNEuSJEmSJEmSimKjWZIkSZIkSZJUFBvNkiRJkiRJkqSi2GiWJEmSJEmSJBXFRrMkSZIkSZIkqSg2miVJkiRJkiRJRbHRLEmSJEmSJEkqio1mSZIkSZIkSVJRbDRLkiRJkiRJkopio1mSJEmSJEmSVBQbzZIkSZIkSZKkotholiRJkiRJkiQVxUazJEmSJEmSJKkoNpolSZIkSZIkSUWx0SxJkiRJkiRJKoqNZkmSJEmSJElSUWw0S5IkSZIkSZKKYqNZkiRJkiRJklQUG82SJEmSJEmSpKLYaJYkSZIkSZIkFcVGsyRJkiRJkiSpKDaaJUmSJEmSJElFsdEsSZIkSZIkSSqKjWZJkiRJkiRJUlFsNEuSJEmSJEmSimKjWZIkSZIkSZJUFBvNkiRJkiRJkqSi2GiWJEmSJEmSJBXFRrMkSZIkSZIkqSg2miVJkiRJkiRJRSm5RnNEjImIH0fEuoioiogVEXFvRAzJOzZJkiRJkiRJKkVleQfQliLiFOB5YATwKLAQOAe4Hbg8Ii5MKW3NMURJkiRJkiRJKjmldkXzv5A1mf8upfT+lNLnU0qXAt8GxgF35xqdJEmSJEmSJJWgkmk0R8TJwLuBFcD3G63+MrAXuCEiKjo4NEmSJEmSJEkqaSXTaAYuLcyfSinVNlyRUtoNPAf0A87r6MAkSZIkSZIkqZSV0hjN4wrzxc2sX0J2xfNpwNNHO9h9993XRmFJkiRJkiRJUm7SLbfcEu39JKV0RfOgwnxnM+vrlg9uamVEfCIiprd5VJIkSZIkSZJU4krpiuajqevap6ZWppTuB+7vuHAkSZIkSZIkqTSU0hXNdVcsD2pm/cBG20mSJEmSJEmS2kApNZoXFeanNbP+1MK8uTGcJUmSJEmSJEnHIFJqciSJLiciTgGWAiuAU1JKtQ3WDQDWkzXWh6eU9uYSpCRJkiRJkiSVoJK5ojmltAx4ChgL3Npo9SSgAvipTWZJkiRJkiRJalslc0Uz1F/V/DwwAngUWACcC1xCNmTGBSmlrflFKEmSJEmSJEmlp6QazQARcQLwFeByoJJsyIwpwKSU0rY8Y5MkSZIkSZKkUlRyjWZJkiRJkiRJUscqmTGaJUmSJEmSJEn5sNEsSZIkSZIkSSpKWd6do/xjAAAV30lEQVQBdEY9e/ZItbUOKdKWIoL2HKalR4CnTFJ30QOozTsIqQS09/sTdU4RPUjJKtrtRICvd3Wk6AHWmuL52u0mAvA8t7MtKaXh7f0kNpqbUFubSNPzjqK0xNmJqVOnttvxJ0yYQJrcboeXpE4lboM0/pq8w5C6vJjzCFcvuinvMNTBfj3uX7nuhYN5h6EO9rPze/O2Xy7JOwx1IzOuPRW+YOOsaF8LrvvL+XlHoXb2s5+/Ed73nbzDKG2P3r6yI57GoTMkSZIkSZIkSUWx0SxJkiRJkiRJKoqNZkmSJEmSJElSURyjuYupOgif+jr894uwbRe8fgx87Va44sK8I2sbd911FzNmzODAgQMMHTqUv/qrv+Kqq67KOyypSUs2wZu+BtecBQ/emHc0kjqLi5c9wx/3baMsAoDjy/qy6PTLc45KetXax5axePIs9q/fS/mwvrzl6xOoPHtU3mGpnfzy0iGHPa6p2s/rP3gzZ3/23pwiUldRtWkNq340kb2LZxFlvRly3uWc8LEvEj1tI3R5WxbAk7fChhnQbzhc+g0Y94G8o+p2Fi95iOXLp7Bj52Jed+KVnHfu1+rXbdj4AtNnfJV9+9ZTWTme8865m4qK43OMtpt6ZSqsehF2r4Pj3wZv/Ui2vLYapv8UdqyG/dvgwttg2Kn5xtpJ+C9EF1NdAyeMhGfvhxNHwePPwV/8I7z8MIwdnXd0xbv++uu544476N27NytXruT222/n1FNPZdy4cXmHJr3Grb+AP3td3lFI6owmjz6Lj1eelHcY0mtsfm4tC745jbd9+1IGjx/Ogc378g5J7eza322v/7l6/15+deUYTrz0QzlGpK5i1Y8m0mvgUMbf/zw1e3ex+K6PsvnJhxjxXq+w6NJqq+GR98FZN8N1v4VVz8Ivr4a/mQWVp+UdXbfSt+8IznjjJ1m/4Tlqaqrql1dVbed/nrudc/7sKxw/+hLmvPxdnnv+s7z7sodzjLab6jMIxr0bNi2EmkOHr6s8GU55B0z7SS6hdVYOndHFVPSFiZ/Mmso9esBVF8FJo2HGgrwjaxsnnXQSvXv3BiAKV4KtXbs2z5CkJj08HQb3hXf6XkyS1IUs+t5MTvvUWQx5ywiiR9B3ZAV9R1bkHZY6yKrf/QflQ0Yw/C1vzzsUdQEHN61myAXvpUfvcnoNGc6gt1zE/tVL8g5LxdqyMLs685x/gB49YeylMOZCmPtA3pF1OyeMuYwxY95Fefngw5avXvNbBg18PSeecDk9e5bzpjNvZcfOReza9UpOkXZjo98Mx42H3o3eK/Uog1MuhspToNC7UsZGcxe3cSssXgVnnJJ3JG3nW9/6FpdddhnXX389lZWVnHfeeXmHJB1m13740mNwzwfzjkRSZ/WPG15m2Lz/5MKlv+eZPZvyDkcCINXUsmPuFqq2H+Dpy37Bbyf8jJe/8jw1B6rzDk0dZPlvHuSkKz5Sf0GHdCQj3nsj2557jNqq/RzcuoGds6cy8C0T8g5LRUtNLEqweW7Hh6Im7dy5lMGDX/1Wd1lZP/pXnMDOnUtzjEpqGRvNXdihavjInXDjlXD62LyjaTuf+cxneOKJJ5g8eTITJkyov8JZ6izufAxuugBOGHL0bSV1P/983Jt45fQrWPuGK/nE0JO4esXzLKvak3dYElVb9pMO1bL+iRVc+NBVTJjyAXbO38ri+2bnHZo6wN4Nq9g8ayonvfeGvENRF9H/jedwYPUSZv31Wbx880VUnHwmg8+5LO+wVKzK06FiBPzxG9lQAK88lQ2fccihlDqL6up99Oo14LBlvXoN4FD13pwiklrORnMXVVsLN9wJvctg8h15R9P2evbsyfjx49m8eTNTpkzJOxyp3uw18N8L4R8uyTsSSZ3Vuf0qGdCzF+U9enLj0LFc2K+Sx3dvyDssiR59stuznHTDG+kzoh/lQ/tw8sfOZNOzq3OOTB1h+W8eZNj4C+k/2vHjdXSptpYlX/0bBp/7bs568CXe/OMXqd67i7UP/p+8Q1OxevaCa6bA0sfgu6PgT/fAG/4CBo7JOzIVlJX149Chwy9SOFS9h15lDnWlzs+bAXZBKcFNd8HGbfD4d6BXCZ/Fmpoax2hWp/LMElixDU68M3u8pwpqEsxfDzM/n29skjqnCEhNfU1V6mC9B5XTZ1QFOGpCt7TiNw/yhhv+V95hqIuo2bODQ1vXM+LyG+jRq5wevcoZdsmHWPvwtxlzQwle6dTdjBgPNzz76uN/vwDe5E0eO4tBg17P8uWP1j+urt7Hnj2rGTTo9TlGJbWMVzR3Qbf8b1iwHH79bejbJ+9o2s727dt5+umn2bdvHzU1Nbz44os8/fTTvPWtb807NKneJy6EZRNh9j9m081vhyvPgCdvzTsySZ3BjpqDPLl7Awdqa6hOtTy0fRVT92zhPf1H5R2aBMAJHzyV5Q/Mp2rrfg7urGL5v89j5MUn5h2W2tnmOS+wb/M6Trz0Q3mHoi6ibOBQeo8Yw+an/i+ppprqvbvY+syv6Pe60/MOTW1h0xyoPpANl/HHb8Ke9TD+o3lH1e3U1lZTU1NFbaohpZrs59pqxhz/LnbuWsLq1U9RU1PF3Hn3MXjQaQwceHLeIXc/tTXZEDOpNptqDmXLAGqqs8eHbefFJSV8LWxpWrkefvj/oLw3jHrPq8t/+AX4yBX5xdUWIoIpU6Zwzz33UFtby8iRI/n0pz/NRRddlHdoUr1+vbOpTv9y6NMLhg9ofh9J3cehlPinDfNYWLWbngSn9xnAlLEXMK6PRUKdw2mfOouD2w/wu/c8Qs/ynoy+4iROveXNeYeldrb88Qc44eL306vCWqSWO+Vz32f1T+5mw6P3Q4+eDDjjXMZ89It5h6W28PID8NKPssbYCRfBdb+FsvK8o+p25s3/AXPn/Uv94xUrf82ZZ3yKN515G2+/4F6mz7ybF/50B5VDx3PBBffkGGk3tvgpWPTEq4/XTIdxl8PpV8DTd8P+bdnyF+7L5pd9CfpVdnycnUgku+2vEREpTc87itISZ8PUqVPb7fgTJkwgTW63w0tSpxK3QRp/Td5hSF1ezHmEqxfdlHcY6mC/HvevXPfCwbzDUAf72fm9edsvl+QdhrqRGdeeCl+w31K0rwXX/eX8vKNQO/vZz98I7/tO3mGUtkdvn5FSOru9n8ahMyRJkiRJkiRJRbHRLEmSJEmSJEkqio1mSZIkSZIkSVJRbDRLkiRJkiRJkopio1mSJEmSJEmSVJRIybugNhYRu4FFecchHcUwYEveQUgtYK6qKzBP1RWYp+oqzFV1BeapugLzVG1lXEppQHs/SVl7P0EXtSildHbeQUhHEhHTzVN1BeaqugLzVF2BeaquwlxVV2CeqiswT9VWImJ6RzyPQ2dIkiRJkiRJkopio1mSJEmSJEmSVBQbzU27P+8ApBYwT9VVmKvqCsxTdQXmqboKc1VdgXmqrsA8VVvpkFzyZoCSJEmSJEmSpKJ4RbMkSZIkSZIkqSg2miVJkiRJkiRJRbHRXBARYyLixxGxLiKqImJFRNwbEUPyjk3dSyH3UjPThmb2uSAiHo+IbRGxLyLmRMTfR0TPjo5fpSUiromI70XEHyJiVyEPHzzKPq3Ox4i4KiKeiYidEbEnIv4UETe2/W+kUtSaPI2IsUeosSkiHj7C89wYES8WcnRnIWevar/fTKUkIioj4uMR8auIWBoR+wt59D8RcVNENPm+3JqqjtTaPLWmKi8R8c8R8XRErC7k6baImBURX46Iymb2sZ6qw7UmV62pak5ra15E9I+IuyJiQUQciIgdhf3f28zxJ0TEAxExNyK2FvZZHhH/GRHvbFWsjtEMEXEK8DwwAngUWAicA1wCLAIuTCltzS9CdScRsQIYDNzbxOo9KaVvNtr+fcB/AAeAnwPbgKuBccAjKaVr2zVglbSImA28GdgDrAFOBx5KKV3fzPatzseIuA34HrC1sM9B4BpgDHBPSulzbfxrqcS0Jk8jYiywHHgJmNLE4eamlB5pYr9vAp8tHP8RoDfwYWAo8OmU0uS2+F1UuiLiZuA+YD3we2AVMBL4IDCIrHZemxq8ObemqqO1Nk+tqcpLRBwEZgLzgU1ABXAecDawDjgvpbS6wfbWU+WiNblqTVVzWplHg4E/AGcC84CnC9v/OTAcuD2l9N1Gx58I/C3wJ7I82gucWNinP/DVlNKdLQo2pdTtJ+BJIJG9ABsu/1Zh+Q/yjtGp+0zACmBFC7cdWCgyVcDZDZb3IfvwJAEfzvt3cuq6E9kHbqcCAVxcyKkHm9m21fkIjCV7w78VGNtg+RBgaWGf8/P+Ozh17qmVeTq2sP4nrTj+BYV9lgJDGh1rayGHx+b9d3Dq3BNwKVlTo0ej5aPImnkJ+FCD5dZUpw6fjiFPralOuUxAn2aW313Ir39psMx66pTb1MpctaY6tUUe3VtY9h9AWYPlw8k+yDgInNrC4x8PbARqgONaEmu3HzojIk4G3k3W3Pt+o9VfJuvi3xARFR0cmtQS15AVi4dTStPrFqaUDgD/VHh4Sx6BqTSklH6fUlqSCv/KHMWx5OPfAOXA5JTSigb7bAe+Vnh48zGGr26ilXl6LOpy8O5CbtY97wqy9w7lwMfa6blVIlJKv0sp/TqlVNto+QbgB4WHFzdYZU1VhzuGPD0W1lQVrVALm/KLwvzUBsusp8pNK3P1WFhTu4FW5tEHC/MvpZSqGxxjM3AP0ItG9au546eU1pJ9INcDOLklsXb7RjPZp/YATzXxhmo38BzQj+ySdKmjlEfE9RHxhYi4PSIuaWbssLr8faKJdVOBfcAFEVHebpFKrzqWfDzSPr9ptI3UlkZHxCcLdfaTETH+CNuap2pvhwrz6gbLrKnqbJrK0zrWVHUWVxfmcxoss56qM2oqV+tYU9VSTeXRqML8lSa2r1vWonGXI2IEcC7ZN0IWtWSfspZsVOLGFeaLm1m/hOyK59PIxjWROsIo4IFGy5ZHxMdSSs82WNZs/qaUqiNiOXAG2SdPC9olUulVx5KPR9pnfUTsBcZERL+U0r52iFnd12WFqV5EPAPcmFJa1WBZBdlXxvaklNY3cZwlhflp7RSnSlxElAF/XXjY8D+J1lR1GkfI0zrWVOUiIj5HNn7oILKxSt9O1nD5eoPNrKfKXQtztY41VU1qYR5tAY4DTiIb07mhuquST2/m+GcDV5H1i8eQjdE8kGyo4S0tidFGc3ZyAHY2s75u+eAOiEUC+DeygdvnAbvJCsFtwCeA30TE+Smllwrbmr/qTI4lH1uyT0VhO9/Eqy3sA+4iu8FK3Sf644GJZGM9Px0Rb0kp7S2ss86qvX2d7GYtj6eUnmyw3JqqzqS5PLWmKm+fI7thZZ0ngI8WviJex3qqzqAluWpN1dG0JI/+i+zGfhMj4rqUUg1ARFQCnylsUx4RfVNK+xsd/2yyYYTr7AY+llJqfCFksxw64+iiMG+vcR+lw6SUJhXGx9uYUtqXUpqbUrqZ7OaUfcn+kWkp81edybHkozmsNpVS2pRS+lJKaWZKaUdhmkr27aU/Aa8HPn4sh27TQNUtRMTfkd0lfiFwQ2t3L8ytqWpXR8pTa6ryllIalVIKsm+EfpDsIp1ZEfHWVhzGeqp215JctabqaFpY874ErASuBWZHxL0RcT/Z1c21vPrhWE0Tx/9B4fh9gTeSXQj504j4QeNtm2Oj+dVPeAY1s35go+2kvNS9sCc0WGb+qjM5lnxs6T67iohLOqrCjTJ+VHjYmjp7tCtJpCZFxK3Ad8je9F+SUtrWaBNrqnLXgjxtkjVVHa1wkc6vyBpylcBPG6y2nqrTOEquNrePNVWHOVIeFW7g+2fAd8m+efEp4H1kVzq/i6yJvDOldPAIxz+QUlqQUrod+CHwyYi4piWx2Wh+dTDr5sasqbtzY3NjOEsdZVNhXtFgWbP5WxhL7ySyG7Y0NQi81NaOJR+PtM9xZPm+xrHv1EHqvnJWX2cLX01cC/Qv5GRjvk9Qq0XE3wOTgblkzbsNTWxmTVWuWpinR2JNVYdLKa0k+2DkjIgYVlhsPVWn00yuHok1Va/RXB6llDanlG5PKZ2cUuqdUhqZUrqJrN4FMK0VT1N3U8mLW7KxjWb4fWH+7og47O8REQOAC4H9wB87OjCpkfML84ZvgH5XmF/exPYTgH7A8ymlqvYMTCo4lnw80j5XNNpGam/nFeaNP5wzT9VmIuIO4NvAbLLm3aZmNrWmKjetyNMjsaYqL6ML87qvhVtP1Vk1ztUjsaaqOa3Jo78tzB9qxfGPL8yrW7Jxt280p5SWAU8BY4FbG62eRPZp0U8bDLYutZuIOCMihjax/HVkV5QAPNhg1SNkdxT9cOHuoHXb9wG+Wnh4XzuFKzV2LPn4b0AVcFtEjG2wzxDgC4WHLR4PSjqaiDg3Ino3sfxS4B8KDx9stLouB79YyM26fcaSvXeoIstl6Ygi4k6ym6rNAN55lLt3W1OVi9bkqTVVeYiI0yNiVBPLe0TE3cAIssbx9sIq66ly0dpctaaqKceQRz0ion8T238cuI7sQ+SHGq17R+OLbwvLTwG+WHj4WIviTckxwQt/uOfJTs6jwALgXLK7ei4GLkgpbc0vQnUXETER+DzZlfbLye7weQpwJdAHeBz4QMOxdCLi/WRvng4ADwPbgD8HxhWW/0Xyha5jVMiv9xcejgLeQ/Yp+h8Ky7aklD7XaPtW5WNEfJps/KitwM+Bg8A1wBjgnobHl5rSmjyNiGeAM4BngDWF9eOBSws/35lSqvtPZ8PnuIfsLs1ryHK5N/CXZGOifTqlNLnxPlJDEXEj8BOyq02+R9PjJa5IKf2kwT7WVHWo1uapNVV5KAzr8g1gKrCMrN6NBN5BdmOsDWQfksxvsI/1VB2utblqTVVTjiGP+gMbgd8CSwuHuQg4p7D/u1JKKxo9xw5gB9lNJ1cDZWS9qMsLP38vpfR3LYrX/lMmIk4AvkL2R6wE1gNTgEktvemFVKyIeAdwM3AWWbOkguzFPht4AHigqaZxRFxI9inT+WQN6aXAj4HvppRa8vUJqUmFDz++fIRNVqaUxjbap9X5GBFXA58D3kr2bZv5wOSU0r8X+SuoG2hNnkbETcAHgDOBYUAvsjdiL5Dl3B+aO0ihAXMb2R2Ya4GZwDdSSv9V/G+hUteCPAV4NqV0caP9rKnqMK3NU2uq8hARZwK3kA1zOQYYDOwlu0jsMbL6+Jr/w1tP1dFam6vWVDXlGPKoF9mV7m8vbA9Zg/kR4FsppT1NPMftZDcWPBMYDvQky70/AT9KKT3Z4nhtNEuSJEmSJEmSitHtx2iWJEmSJEmSJBXHRrMkSZIkSZIkqSg2miVJkiRJkiRJRbHRLEmSJEmSJEkqio1mSZIkSZIkSVJRbDRLkiRJkiRJkopio1mSJEmSJEmSVBQbzZIkSZIkSZKkotholiRJkiRJkiQVxUazJEmSJEmSJKko/x/0pF+naZ6Y+wAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<Figure size 1728x576 with 2 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "#Plotting the predicted fitness effect for the variants seen more than once\n",
    "plt.close('all')\n",
    "f, (ax1, ax2) = plt.subplots(2, 1, sharey=True, sharex = True, figsize=(24, 8))\n",
    "gs = matplotlib.gridspec.GridSpec(2, 1, width_ratios=[1], height_ratios=[10,0.5])\n",
    "ax1 = plt.subplot(gs[0])\n",
    "ax2 = plt.subplot(gs[1])\n",
    "gs.update(hspace=0.05)\n",
    "\n",
    "nonsense_list = []\n",
    "y_list = []\n",
    "x_list=[]\n",
    "y_list_top20 = []\n",
    "x_list_top20=[]\n",
    "y_list_nonsense = []\n",
    "x_list_nonsense = []\n",
    "x=[]\n",
    "y=[]\n",
    "OBS_EXP_DICT={}\n",
    "top20_lower_err = []\n",
    "top20_upper_err = []\n",
    "missense_lower_err = []\n",
    "missense_upper_err = []\n",
    "nonsense_lower_err = []\n",
    "nonsense_upper_err = []\n",
    "\n",
    "top20 = c1\n",
    "missense = c6\n",
    "nonsense = c10\n",
    "synonymous = c3\n",
    "\n",
    "axisfont=20\n",
    "domainfont = 12\n",
    "titlefont=12\n",
    "axislabelfont=12\n",
    "\n",
    "top_20_TP53 = []\n",
    "\n",
    "for k, v in TP53_missense_s.items():\n",
    "    if k in top_20_TP53:\n",
    "        codon_number = int(((re.findall('\\d+', k)))[0])\n",
    "        s_value = float(v[0])*100\n",
    "        x_list_top20.append(codon_number)\n",
    "        y_list_top20.append(s_value)\n",
    "        \n",
    "        s_upper = float(v[3])*100\n",
    "        s_lower = float(v[4])*100\n",
    "        error_upper = s_upper-s_value\n",
    "        error_lower = s_value-s_lower\n",
    "        top20_lower_err.append(error_lower)\n",
    "        top20_upper_err.append(error_upper)\n",
    "\n",
    "    if k not in top_20_TP53:\n",
    "        codon_number = int(((re.findall('\\d+', k)))[0])\n",
    "        s_value = float(v[0])*100\n",
    "        x_list.append(codon_number)\n",
    "        y_list.append(s_value)\n",
    "        \n",
    "        s_upper = float(v[3])*100\n",
    "        s_lower = float(v[4])*100\n",
    "        error_upper = s_upper-s_value\n",
    "        error_lower = s_value-s_lower\n",
    "        missense_lower_err.append(error_lower)\n",
    "        missense_upper_err.append(error_upper)\n",
    "\n",
    "    missense_err = [tuple(missense_lower_err), tuple(missense_upper_err)]   \n",
    "    \n",
    "for k, v in TP53_nonsense_s.items():\n",
    "    if k in top_20_TP53:\n",
    "        codon_number = int(((re.findall('\\d+', k)))[0])\n",
    "        s_value = float(v[0])*100\n",
    "        x_list_top20.append(codon_number)\n",
    "        y_list_top20.append(s_value)\n",
    "        \n",
    "        s_upper = float(v[3])*100\n",
    "        s_lower = float(v[4])*100\n",
    "        error_upper = s_upper-s_value\n",
    "        error_lower = s_value-s_lower\n",
    "        top20_lower_err.append(error_lower)\n",
    "        top20_upper_err.append(error_upper)\n",
    "    \n",
    "    top20_err = [tuple(top20_lower_err), tuple(top20_upper_err)]\n",
    "        \n",
    "    if k not in top_20_TP53:\n",
    "        codon_number = int(((re.findall('\\d+', k)))[0])\n",
    "        s_value = float(v[0])*100\n",
    "        x_list_nonsense.append(codon_number)\n",
    "        y_list_nonsense.append(s_value)\n",
    "        \n",
    "        s_upper = float(v[3])*100\n",
    "        s_lower = float(v[4])*100\n",
    "        error_upper = s_upper-s_value\n",
    "        error_lower = s_value-s_lower\n",
    "        nonsense_lower_err.append(error_lower)\n",
    "        nonsense_upper_err.append(error_upper)\n",
    "\n",
    "    nonsense_err = [tuple(nonsense_lower_err), tuple(nonsense_upper_err)]        \n",
    "    \n",
    "msize = 12\n",
    "medgewidth = 2\n",
    "linew = 3\n",
    "csize = 4\n",
    "cthick = 3\n",
    "\n",
    "markers1, caps1, bars1 = ax1.errorbar(x_list, y_list, yerr = missense_err, fmt = 'o', ecolor = missense, elinewidth = linew, \\\n",
    "             capsize = csize, capthick = cthick, markersize = msize, markeredgewidth = medgewidth, \\\n",
    "             markeredgecolor = missense, markerfacecolor = missense, zorder = 1)    \n",
    "\n",
    "markers2, caps2, bars2 = ax1.errorbar(x_list_nonsense, y_list_nonsense, yerr = nonsense_err, fmt = 'o', ecolor = nonsense, elinewidth = linew, \\\n",
    "             capsize = csize, capthick = cthick, markersize = msize, markeredgewidth = medgewidth, \\\n",
    "             markeredgecolor = nonsense, markerfacecolor = nonsense, zorder = 2)    \n",
    "\n",
    "markers3, caps3, bars3 = ax1.errorbar(x_list_top20, y_list_top20, yerr = top20_err, fmt = 'o', ecolor = top20, elinewidth = linew, \\\n",
    "             capsize = csize, capthick = cthick, markersize = msize, markeredgewidth = medgewidth, \\\n",
    "             markeredgecolor = top20, markerfacecolor = top20, zorder = 3)    \n",
    "\n",
    "# loop through bars and caps and set the alpha value\n",
    "a = 0.7\n",
    "[bar.set_alpha(a) for bar in bars1]\n",
    "[cap.set_alpha(a) for cap in caps1]\n",
    "[bar.set_alpha(a) for bar in bars2]\n",
    "[cap.set_alpha(a) for cap in caps2]\n",
    "[bar.set_alpha(a) for bar in bars3]\n",
    "[cap.set_alpha(a) for cap in caps3]\n",
    "\n",
    "#Only show the required axis lines\n",
    "ax1.spines['top'].set_visible(False)\n",
    "ax1.spines['right'].set_visible(False)\n",
    "\n",
    "for axis in ['bottom','left']:\n",
    "    ax1.spines[axis].set_linewidth(2)\n",
    "\n",
    "for axis in ['bottom','left']:\n",
    "    ax1.spines[axis].set_color(grey3)\n",
    "    \n",
    "ax1.set_ylabel('fitness effect (s)', fontsize = axisfont, labelpad = 10)\n",
    "ax1.set_title('Fitness effects across TP53', fontsize = 18)\n",
    "\n",
    "# ax1.set_yscale('log')\n",
    "\n",
    "# STACKED BAR CHART\n",
    "# define the exons + create data\n",
    "exons = [{'2':(25-1)+1, '3': (32-26)+1, '4': (125-33)+1, '5':(187-126)+1, '6':(224-188)+1, '7': (261-225)+1, '8': (307-262)+1,'9': (331-308)+1, '10': (367-332)+1, '11': (393-368)+1}]\n",
    "exon_orders = [['2', '3', '4', '5', '6', '7', '8', '9', '10', '11']]\n",
    "colors2 = [c4, c0, c3, c1, c5, c6,c7, c8, c9, c10]\n",
    "names = exons[0].keys()\n",
    "values = np.array([[data[name] for name in order] for data,order in zip(exons, exon_orders)])\n",
    "lefts = np.insert(np.cumsum(values, axis=1),0,0, axis=1)[:, :-1]\n",
    "orders = np.array(exon_orders)\n",
    "\n",
    "#plot the bar chart\n",
    "for name, color in zip(names, colors2):\n",
    "    idx = np.where(orders == name)\n",
    "    value = values[idx]\n",
    "    left = lefts[idx]\n",
    "    ax2.bar(x=left, height=-10, width=value, bottom=-5, color=color, orientation = 'horizontal', label = name, edgecolor='k', linewidth=1, tick_label = name)\n",
    "    ax2.text(left + (value/2.), (-0.26), name, ha='center', va='center', fontsize = 12)\n",
    "    \n",
    "# CONFIGURING THE GRAPH APPEARANCE\n",
    "#Set the x and y axis limits\n",
    "ax1.set_xlim(0, 393)\n",
    "ax1.set_ylim(0, 35)\n",
    "# ax1.set_ylim(10**1, 10**5)\n",
    "ax2.set_xlim(0, 393)\n",
    "ax2.set_ylim(-0.5, 0)\n",
    "\n",
    "#x-axis ticks\n",
    "x1_major_ticks = []\n",
    "x1_major_tick_labels = []\n",
    "x2_major_ticks = [0, 50, 100, 150, 200, 250, 300, 350, 393]\n",
    "x2_major_tick_labels = [\"0\", \"50\", \"100\", \"150\", \"200\", \"250\", \"300\", \"350\", \"393\"]\n",
    "ax1.set_xticks(x1_major_ticks)\n",
    "ax1.set_xticklabels(x1_major_tick_labels, fontsize = axisfont)\n",
    "ax2.set_xticks(x2_major_ticks)\n",
    "ax2.set_xticklabels(x2_major_tick_labels, fontsize = axisfont)\n",
    "\n",
    "#y-axis ticks\n",
    "y1_major_ticks = [0, 5, 10, 15, 20, 25, 30, 35]\n",
    "y1_major_tick_labels = [\"0\", \"5\", \"10\", \"15\", \"20\", \"25\", \"30\", \"35\"]\n",
    "ax1.set_yticks(y1_major_ticks)\n",
    "ax1.set_yticklabels(y1_major_tick_labels, fontsize = axisfont)\n",
    "\n",
    "#set panel legends\n",
    "legend_marker_size = 2\n",
    "legend_elements = [Line2D([0], [0], marker = '_', color = missense, alpha = 1.0, markersize = legend_marker_size, \\\n",
    "                          lw=3, label='missense'),\n",
    "                   Line2D([0], [0], marker = '_', color=nonsense, alpha=1.0, markersize = legend_marker_size, \\\n",
    "                          lw=3, label='nonsense'),\n",
    "                  Line2D([0], [0], marker = '_', color=top20, alpha=1.0, markersize = legend_marker_size, \\\n",
    "                          lw=3, label='top 20 observed variant')]\n",
    "    \n",
    "ax1.legend(ncol=1, handles=legend_elements, loc='upper left', frameon=0, fontsize = 20)\n",
    "\n",
    "ax1.grid(which='major', linestyle=':', linewidth='0.7', color=grey3, zorder = 0)\n",
    "\n",
    "plt.show()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.0"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": true,
   "toc_position": {},
   "toc_section_display": true,
   "toc_window_display": true
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
